{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "829b6482-0102-4a0f-a3b7-c98386924d32",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "b98d02ba-452d-4192-b1a6-93640491404e",
   "metadata": {},
   "source": [
    "## Example Prompts\n",
    "1. Load the 2015 cell cycle dataset\n",
    "2. Convert the dataset to gene coordinates\n",
    "3. Build a DMD or linear model of the 2015 data\n",
    "4. Perform biomarker or sensor selection on this data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "217b2753-0d3c-4265-a087-4b7e0483de09",
   "metadata": {},
   "outputs": [],
   "source": [
    "from BRAD import brad"
   ]
  },
  {
   "cell_type": "raw",
   "id": "ee43e3c5-c012-4c0e-8a1a-4e0e109d1d39",
   "metadata": {},
   "source": [
    "/force WRITE make a summary of our conversation. Provide math equations in latex formatting to describe the Dynamic Mode Decomposition"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "051356db-4523-4b58-adae-95c2f63621e8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Would you like to use a database with BRAD [Y/N]?\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      " N\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32m2024-07-03 09:41:17 INFO semantic_router.utils.logger local\u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome to RAG! The chat log from this conversation will be saved to /home/jpic/BRAD/2024-07-03_08-51-24/log.json. How can I help?\n",
      "==================================================\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  /force CODE Perform time varying biomarker selection on this data from times 0 to 3. set verbose to true.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-07-03 09:41:58,126 - INFO - CODE\n",
      "2024-07-03 09:41:58,127 - INFO - CODER\n",
      "2024-07-03 09:41:58,138 - INFO - input_variables=['user_query'] template=\"You must select which code to run to help a user.\\n\\n**Available Scripts**\\nScript Name: geneformerCoordinatesTimeSeries, \\t Description: This script organizes datasets into a format to work with time series gene expression using individual geneformer embeddings as the state space. From a .h5ad file, a .pkl file is constructed here that will contain the state type being geneformer, the state names being geneformer coordinates, and the time series data organized as states (geneformer coordinates) by timepoint (hours) by replicates (experiment replicates).\\nScript Name: geneCoordinatesTimeSeries, \\t Description: This script organizes datasets into a format to work with time series gene expression using individual gene expresession as the state space. From a .h5ad file, a .pkl file is constructed here that will contain the state type being gene expression, the state names being gene names, and the time series data organized as states (genes) by timepoint (hours) by replicates (experiment replicates).\\nScript Name: linearModels, \\t Description: This file builds linear and linear time variant (LTI and LTV) models of gene expressions.\\nScript Name: geneformerEmbeddings, \\t Description: Get gene former embeddings of your data\\nScript Name: biomarkerSelectionTimeVarying, \\t Description: This script performs time varying sensor or biomarker selection on dynamical systems models of cellular processes. It selects new sensors at each time point.\\nScript Name: buildLinearModel, \\t Description: This script builds linear models of time series data using Dynamic Mode Decomposition (DMD).\\nScript Name: test, \\t Description: This script was made for Joshua to test and learn python\\nScript Name: datasetSelector, \\t Description: This script is a file chooser to load a gene expression dataset from the lab's turbo partition to give access to the data to BRAD.\\nScript Name: biomarkerSelection, \\t Description: This script performs sensor selection on models of dynamical systems. This is applied to models of cellular processes for biomarker selection.\\nScript Name: buildLTVModel, \\t Description: This script builds linear time varying models of time series data using Data Guided Control model.\\n\\n\\n**User Query**\\n{user_query}\\n\\n**Task**\\nBased on the user's query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\\n\\n**Response Template**\\nSCRIPT: <selected script>\\nREASON: <reasoning why the script fits the user prompt>\\n\"\n",
      "2024-07-03 09:41:58,139 - INFO - FIRST LLM CALL\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "BRAD >> 14: \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-07-03 09:41:59,894 - INFO - SCRIPT: biomarkerSelectionTimeVarying\n",
      "REASON: The user query specifically asks for time varying biomarker selection, which is the exact functionality described in the biomarkerSelectionTimeVarying script. Additionally, the script is designed to select new sensors at each time point, which matches the user's request to perform selection from times 0 to 3.\n",
      "2024-07-03 09:41:59,895 - INFO - Warning! the type doesn't exist\n",
      "2024-07-03 09:41:59,896 - INFO - scriptPath=/home/jpic/BRAD-Tools/\n",
      "2024-07-03 09:41:59,897 - INFO - pyPath=/home/jpic/BRAD-Tools/\n",
      "2024-07-03 09:41:59,898 - INFO - ALL SCRIPTS FOUND. BUILDING TEMPLATE\n",
      "2024-07-03 09:41:59,899 - INFO - input_variables=['history', 'input'] template='Current conversation:\\n{history}\\n\\n**PYTHON SCRIPT**\\nYou must run this python script:\\n/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py\\n\\n**PYTHON SCRIPT DOCUMENTATION**:\\nThis is the doc string of this python script:\\n\"\"\"\\nThis script performs time varying sensor or biomarker selection on dynamical systems models of cellular processes. It selects new sensors at each time point.\\n\\nArguments (four arguments):\\n    1. output directory: chatstatus[\\'output-directory\\']\\n    2. output file: <name of output file>\\n    3. input file: the file created previously with the LTI or LTV model\\n    4. start time: (int, default 0) the first time to select sensors at (inclusive)\\n    5. end time: (int, default 10) the last time to select sensors at (not inclusive)\\n    6. verbose: (optional, True or False) if sensor selection should be run in verbose mode so that more output will be displayed to the user.\\n\\nUsage:\\nCommand Line:\\n```\\npython <path/to/script/>biomarkerSelectionTimeVarying.py <output path> <output file> <input file> <start time> <end time> <verbose>\\n```\\n\\nBRAD Line:\\n```\\nsubprocess.call([sys.executable, \\'<path/to/script/>/biomarkerSelectionTimeVarying.py\\', chatstatus[\\'output-directory\\'], <output file>, <input file>, <start time>, <end time>, <verbose>], capture_output=True, text=True)\\n```\\n\\n**OUTPUT FILE NAME INSTRUCTIONS**\\n1. Output path should be chatstatus[\\'output-directory\\']\\n2. Output file name should be `S3-<descriptive name>.csv`\\n\"\"\"\\n\\n\\n**CALL PYTHON SCRIPTS FROM PYTHON**:\\nUse the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\\n\\nTo call a Python script `example_script.py` which has no arguments:\\n\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\']], capture_output=True, text=True)\\n)\\n```\\n\\nTo call a Python script `example_script.py` with one argument:\\n\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\'], \\'arg1\\'], capture_output=True, text=True)\\n)\\n```\\n\\nTo call a Python script `example_script.py` with two arguments:\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\'], \\'arg1\\', \\'arg2\\'], capture_output=True, text=True)\\n)\\n```\\n\\nNote that chatstatus[\\'output-directory\\'] is ALWAYS passed as the first argument.\\n\\nThe following files were previously created by BRAD and could be used as input to a function if necessary:\\nREPORT.tex\\nlog.json\\n.ipynb_checkpoints\\nREPORT.log\\nS1-2015.pkl\\nENRICHR-GO_Biological_Process_2021.csv\\nS0-2015.h5ad\\nS2-2015.pkl\\nS3-biomarker-selection.csv\\nimages\\n\\nQuery:{input}\\n\\n**INSTRUCTIONS**\\n1. Given the user query and the documentation, identify each of the arguments found in the user\\'s query that should be passed to the Python script.\\n2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\\n3. The last line of your response should say \"Execute: <Python code to execute>\"\\n4. Format the response/output as:\\n    Arguments: \\n    Python Code Explanation: <2 sentences maximum>\\n    Execute: <your code here>\\n\\n**IMPORTANT**\\nThe code to execute from your response must be formatted as:\\n    Execute: subprocess.call([sys.executable, \\'<path to python script>\\', \\'<argument 1>\\', \\'<argument 2>\\', ..., \\'<argument n>\\'], capture_output=True, text=True))\\nThis output should be exactly one line and no longer. Stop the response after this line.\\n'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation:\n",
      "\n",
      "\n",
      "**PYTHON SCRIPT**\n",
      "You must run this python script:\n",
      "/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py\n",
      "\n",
      "**PYTHON SCRIPT DOCUMENTATION**:\n",
      "This is the doc string of this python script:\n",
      "\"\"\"\n",
      "This script performs time varying sensor or biomarker selection on dynamical systems models of cellular processes. It selects new sensors at each time point.\n",
      "\n",
      "Arguments (four arguments):\n",
      "    1. output directory: chatstatus['output-directory']\n",
      "    2. output file: <name of output file>\n",
      "    3. input file: the file created previously with the LTI or LTV model\n",
      "    4. start time: (int, default 0) the first time to select sensors at (inclusive)\n",
      "    5. end time: (int, default 10) the last time to select sensors at (not inclusive)\n",
      "    6. verbose: (optional, True or False) if sensor selection should be run in verbose mode so that more output will be displayed to the user.\n",
      "\n",
      "Usage:\n",
      "Command Line:\n",
      "```\n",
      "python <path/to/script/>biomarkerSelectionTimeVarying.py <output path> <output file> <input file> <start time> <end time> <verbose>\n",
      "```\n",
      "\n",
      "BRAD Line:\n",
      "```\n",
      "subprocess.call([sys.executable, '<path/to/script/>/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], <output file>, <input file>, <start time>, <end time>, <verbose>], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "**OUTPUT FILE NAME INSTRUCTIONS**\n",
      "1. Output path should be chatstatus['output-directory']\n",
      "2. Output file name should be `S3-<descriptive name>.csv`\n",
      "\"\"\"\n",
      "\n",
      "\n",
      "**CALL PYTHON SCRIPTS FROM PYTHON**:\n",
      "Use the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\n",
      "\n",
      "To call a Python script `example_script.py` which has no arguments:\n",
      "\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory']], capture_output=True, text=True)\n",
      ")\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with one argument:\n",
      "\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory'], 'arg1'], capture_output=True, text=True)\n",
      ")\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with two arguments:\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory'], 'arg1', 'arg2'], capture_output=True, text=True)\n",
      ")\n",
      "```\n",
      "\n",
      "Note that chatstatus['output-directory'] is ALWAYS passed as the first argument.\n",
      "\n",
      "The following files were previously created by BRAD and could be used as input to a function if necessary:\n",
      "REPORT.tex\n",
      "log.json\n",
      ".ipynb_checkpoints\n",
      "REPORT.log\n",
      "S1-2015.pkl\n",
      "ENRICHR-GO_Biological_Process_2021.csv\n",
      "S0-2015.h5ad\n",
      "S2-2015.pkl\n",
      "S3-biomarker-selection.csv\n",
      "images\n",
      "\n",
      "Query:Perform time varying biomarker selection on this data from times 0 to 3. set verbose to true.\n",
      "\n",
      "**INSTRUCTIONS**\n",
      "1. Given the user query and the documentation, identify each of the arguments found in the user's query that should be passed to the Python script.\n",
      "2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\n",
      "3. The last line of your response should say \"Execute: <Python code to execute>\"\n",
      "4. Format the response/output as:\n",
      "    Arguments: \n",
      "    Python Code Explanation: <2 sentences maximum>\n",
      "    Execute: <your code here>\n",
      "\n",
      "**IMPORTANT**\n",
      "The code to execute from your response must be formatted as:\n",
      "    Execute: subprocess.call([sys.executable, '<path to python script>', '<argument 1>', '<argument 2>', ..., '<argument n>'], capture_output=True, text=True))\n",
      "This output should be exactly one line and no longer. Stop the response after this line.\n",
      "\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-07-03 09:42:03,772 - INFO - LLM OUTPUT PARSER\n",
      "2024-07-03 09:42:03,773 - INFO - Arguments: \n",
      "output directory: chatstatus['output-directory']\n",
      "output file: S3-biomarker-selection-times-varying.csv\n",
      "input file: S2-2015.pkl\n",
      "start time: 0\n",
      "end time: 3\n",
      "verbose: True\n",
      "\n",
      "Python Code Explanation: \n",
      "Using the subprocess module, we are executing the biomarkerSelectionTimeVarying.py script with the required arguments. \n",
      "The script will perform time varying biomarker selection on the input data from times 0 to 3 in verbose mode.\n",
      "\n",
      "Execute: subprocess.call([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-biomarker-selection-times-varying.csv', 'S2-2015.pkl', '0', '3', 'True'], capture_output=True, text=True)\n",
      "2024-07-03 09:42:03,773 - INFO - [\"Arguments: \\noutput directory: chatstatus['output-directory']\\noutput file: S3-biomarker-selection-times-varying.csv\\ninput file: S2-2015.pkl\\nstart time: 0\\nend time: 3\\nverbose: True\\n\\nPython Code Explanation: \\nUsing the subprocess module, we are executing the biomarkerSelectionTimeVarying.py script with the required arguments. \\nThe script will perform time varying biomarker selection on the input data from times 0 to 3 in verbose mode.\\n\\n\", \" subprocess.call([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-biomarker-selection-times-varying.csv', 'S2-2015.pkl', '0', '3', 'True'], capture_output=True, text=True)\"]\n",
      "2024-07-03 09:42:03,774 - INFO - subprocess.call([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-biomarker-selection-times-varying.csv', 'S2-2015.pkl', '0', '3', 'True'], capture_output=True, text=True)\n",
      "2024-07-03 09:42:03,774 - INFO - subprocess.run([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-biomarker-selection-times-varying.csv', 'S2-2015.pkl', '0', '3', 'True'], capture_output=True, text=True)\n",
      "2024-07-03 09:42:03,775 - INFO - Replacing: S2-2015.pkl with /home/jpic/BRAD/2024-07-03_08-51-24/S2-2015.pkl\n",
      "2024-07-03 09:42:03,775 - INFO - New String: subprocess.run([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-biomarker-selection-times-varying.csv', '/home/jpic/BRAD/2024-07-03_08-51-24/S2-2015.pkl', '0', '3', 'True'], capture_output=True, text=True)\n",
      "2024-07-03 09:42:03,776 - INFO - Executing Code!!\n",
      "2024-07-03 09:42:03,776 - INFO - EVAL\n",
      "2024-07-03 09:42:03,777 - INFO - subprocess.run([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-biomarker-selection-times-varying.csv', '/home/jpic/BRAD/2024-07-03_08-51-24/S2-2015.pkl', '0', '3', 'True'], capture_output=True, text=True)\n",
      "2024-07-03 09:42:03,777 - INFO - END EVAL CHECK\n",
      "2024-07-03 09:42:03,777 - INFO - Finalized PYTHON Call:\n",
      "2024-07-03 09:42:03,778 - INFO -    \n",
      "2024-07-03 09:42:03,778 - INFO - response = subprocess.run([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-biomarker-selection-times-varying.csv', '/home/jpic/BRAD/2024-07-03_08-51-24/S2-2015.pkl', '0', '3', 'True'], capture_output=True, text=True)\n",
      "2024-07-03 09:42:03,779 - INFO -    \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-07-03 09:42:29,359 - INFO - Debug: PYTHON code executed successfully.\n",
      "2024-07-03 09:42:29,361 - INFO - Code Execution Output:\n",
      "2024-07-03 09:42:29,361 - INFO - CompletedProcess(args=['/sw/pkgs/arc/python3.11-anaconda/2024.02-1/bin/python', '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', '/home/jpic/BRAD/2024-07-03_08-51-24/', 'S3-biomarker-selection-times-varying.csv', '/home/jpic/BRAD/2024-07-03_08-51-24/S2-2015.pkl', '0', '3', 'True'], returncode=1, stdout='Time varying biomarker selection script has been called!\\n*******************************\\n Time Varying Sensor Selection \\n*******************************\\nOutput Path: /home/jpic/BRAD/2024-07-03_08-51-24/\\nOutput File: S3-biomarker-selection-times-varying.csv\\nInput File : /home/jpic/BRAD/2024-07-03_08-51-24/S2-2015.pkl\\nstartTime  : 0\\nendTime    : 3\\nverbose    : True\\n..............................\\n         Model Loaded         \\nt: 0/3\\nt: 1/3\\nt: 2/3\\n..............................\\n      Sensors Selected        \\n', stderr='Traceback (most recent call last):\\n  File \"/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py\", line 107, in <module>\\n    main()\\n  File \"/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py\", line 99, in main\\n    pd.DataFrame(sensors[time][\\'sensors\\']).to_csv(output_file_path)    \\n    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \"/home/jpic/.local/lib/python3.11/site-packages/pandas/core/generic.py\", line 3772, in to_csv\\n    return DataFrameRenderer(formatter).to_csv(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \"/home/jpic/.local/lib/python3.11/site-packages/pandas/io/formats/format.py\", line 1186, in to_csv\\n    csv_formatter.save()\\n  File \"/home/jpic/.local/lib/python3.11/site-packages/pandas/io/formats/csvs.py\", line 240, in save\\n    with get_handle(\\n         ^^^^^^^^^^^\\n  File \"/home/jpic/.local/lib/python3.11/site-packages/pandas/io/common.py\", line 737, in get_handle\\n    check_parent_directory(str(handle))\\n  File \"/home/jpic/.local/lib/python3.11/site-packages/pandas/io/common.py\", line 600, in check_parent_directory\\n    raise OSError(rf\"Cannot save file into a non-existent directory: \\'{parent}\\'\")\\nOSError: Cannot save file into a non-existent directory: \\'/home/jpic/BRAD/2024-07-03_08-51-24/0/_\\'\\n')\n",
      "2024-07-03 09:42:29,362 - INFO - Debug: PYTHON code output saved to output.\n",
      "2024-07-03 09:42:29,363 - INFO - \n",
      "\n",
      "\n",
      "route\n",
      "\n",
      "\n",
      "\n",
      "2024-07-03 09:42:29,363 - INFO - CODE\n",
      "2024-07-03 09:42:29,364 - INFO - \n",
      "\n",
      "PLANNER:\n",
      "\n",
      "{'module': 'CODE', 'steps': [{'llm': \"model='meta/llama3-70b-instruct'\", 'prompt': 'input_variables=[\\'user_query\\'] template=\"You must select which code to run to help a user.\\\\n\\\\n**Available Scripts**\\\\nScript Name: geneformerCoordinatesTimeSeries, \\\\t Description: This script organizes datasets into a format to work with time series gene expression using individual geneformer embeddings as the state space. From a .h5ad file, a .pkl file is constructed here that will contain the state type being geneformer, the state names being geneformer coordinates, and the time series data organized as states (geneformer coordinates) by timepoint (hours) by replicates (experiment replicates).\\\\nScript Name: geneCoordinatesTimeSeries, \\\\t Description: This script organizes datasets into a format to work with time series gene expression using individual gene expresession as the state space. From a .h5ad file, a .pkl file is constructed here that will contain the state type being gene expression, the state names being gene names, and the time series data organized as states (genes) by timepoint (hours) by replicates (experiment replicates).\\\\nScript Name: linearModels, \\\\t Description: This file builds linear and linear time variant (LTI and LTV) models of gene expressions.\\\\nScript Name: geneformerEmbeddings, \\\\t Description: Get gene former embeddings of your data\\\\nScript Name: biomarkerSelectionTimeVarying, \\\\t Description: This script performs time varying sensor or biomarker selection on dynamical systems models of cellular processes. It selects new sensors at each time point.\\\\nScript Name: buildLinearModel, \\\\t Description: This script builds linear models of time series data using Dynamic Mode Decomposition (DMD).\\\\nScript Name: test, \\\\t Description: This script was made for Joshua to test and learn python\\\\nScript Name: datasetSelector, \\\\t Description: This script is a file chooser to load a gene expression dataset from the lab\\'s turbo partition to give access to the data to BRAD.\\\\nScript Name: biomarkerSelection, \\\\t Description: This script performs sensor selection on models of dynamical systems. This is applied to models of cellular processes for biomarker selection.\\\\nScript Name: buildLTVModel, \\\\t Description: This script builds linear time varying models of time series data using Data Guided Control model.\\\\n\\\\n\\\\n**User Query**\\\\n{user_query}\\\\n\\\\n**Task**\\\\nBased on the user\\'s query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\\\\n\\\\n**Response Template**\\\\nSCRIPT: <selected script>\\\\nREASON: <reasoning why the script fits the user prompt>\\\\n\"', 'input': 'Perform time varying biomarker selection on this data from times 0 to 3. set verbose to true.', 'output': 'content=\"SCRIPT: biomarkerSelectionTimeVarying\\\\nREASON: The user query specifically asks for time varying biomarker selection, which is the exact functionality described in the biomarkerSelectionTimeVarying script. Additionally, the script is designed to select new sensors at each time point, which matches the user\\'s request to perform selection from times 0 to 3.\" response_metadata={\\'role\\': \\'assistant\\', \\'content\\': \"SCRIPT: biomarkerSelectionTimeVarying\\\\nREASON: The user query specifically asks for time varying biomarker selection, which is the exact functionality described in the biomarkerSelectionTimeVarying script. Additionally, the script is designed to select new sensors at each time point, which matches the user\\'s request to perform selection from times 0 to 3.\", \\'token_usage\\': {\\'prompt_tokens\\': 550, \\'total_tokens\\': 625, \\'completion_tokens\\': 75}, \\'model_name\\': \\'meta/llama3-70b-instruct\\'} id=\\'run-d08d0bd0-556b-4118-b1f2-f559f5829802-0\\' role=\\'assistant\\'', 'parsedOutput': {'scriptName': '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', 'scriptType': 'python', 'scriptPath': '/home/jpic/BRAD-Tools/'}, 'purpose': 'Select which code to run'}, {'llm': \"model='meta/llama3-70b-instruct'\", 'prompt': 'input_variables=[\\'history\\', \\'input\\'] template=\\'Current conversation:\\\\n{history}\\\\n\\\\n**PYTHON SCRIPT**\\\\nYou must run this python script:\\\\n/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py\\\\n\\\\n**PYTHON SCRIPT DOCUMENTATION**:\\\\nThis is the doc string of this python script:\\\\n\"\"\"\\\\nThis script performs time varying sensor or biomarker selection on dynamical systems models of cellular processes. It selects new sensors at each time point.\\\\n\\\\nArguments (four arguments):\\\\n    1. output directory: chatstatus[\\\\\\'output-directory\\\\\\']\\\\n    2. output file: <name of output file>\\\\n    3. input file: the file created previously with the LTI or LTV model\\\\n    4. start time: (int, default 0) the first time to select sensors at (inclusive)\\\\n    5. end time: (int, default 10) the last time to select sensors at (not inclusive)\\\\n    6. verbose: (optional, True or False) if sensor selection should be run in verbose mode so that more output will be displayed to the user.\\\\n\\\\nUsage:\\\\nCommand Line:\\\\n```\\\\npython <path/to/script/>biomarkerSelectionTimeVarying.py <output path> <output file> <input file> <start time> <end time> <verbose>\\\\n```\\\\n\\\\nBRAD Line:\\\\n```\\\\nsubprocess.call([sys.executable, \\\\\\'<path/to/script/>/biomarkerSelectionTimeVarying.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\'], <output file>, <input file>, <start time>, <end time>, <verbose>], capture_output=True, text=True)\\\\n```\\\\n\\\\n**OUTPUT FILE NAME INSTRUCTIONS**\\\\n1. Output path should be chatstatus[\\\\\\'output-directory\\\\\\']\\\\n2. Output file name should be `S3-<descriptive name>.csv`\\\\n\"\"\"\\\\n\\\\n\\\\n**CALL PYTHON SCRIPTS FROM PYTHON**:\\\\nUse the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\\\\n\\\\nTo call a Python script `example_script.py` which has no arguments:\\\\n\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\']], capture_output=True, text=True)\\\\n)\\\\n```\\\\n\\\\nTo call a Python script `example_script.py` with one argument:\\\\n\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\'], \\\\\\'arg1\\\\\\'], capture_output=True, text=True)\\\\n)\\\\n```\\\\n\\\\nTo call a Python script `example_script.py` with two arguments:\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\'], \\\\\\'arg1\\\\\\', \\\\\\'arg2\\\\\\'], capture_output=True, text=True)\\\\n)\\\\n```\\\\n\\\\nNote that chatstatus[\\\\\\'output-directory\\\\\\'] is ALWAYS passed as the first argument.\\\\n\\\\nThe following files were previously created by BRAD and could be used as input to a function if necessary:\\\\nREPORT.tex\\\\nlog.json\\\\n.ipynb_checkpoints\\\\nREPORT.log\\\\nS1-2015.pkl\\\\nENRICHR-GO_Biological_Process_2021.csv\\\\nS0-2015.h5ad\\\\nS2-2015.pkl\\\\nS3-biomarker-selection.csv\\\\nimages\\\\n\\\\nQuery:{input}\\\\n\\\\n**INSTRUCTIONS**\\\\n1. Given the user query and the documentation, identify each of the arguments found in the user\\\\\\'s query that should be passed to the Python script.\\\\n2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\\\\n3. The last line of your response should say \"Execute: <Python code to execute>\"\\\\n4. Format the response/output as:\\\\n    Arguments: \\\\n    Python Code Explanation: <2 sentences maximum>\\\\n    Execute: <your code here>\\\\n\\\\n**IMPORTANT**\\\\nThe code to execute from your response must be formatted as:\\\\n    Execute: subprocess.call([sys.executable, \\\\\\'<path to python script>\\\\\\', \\\\\\'<argument 1>\\\\\\', \\\\\\'<argument 2>\\\\\\', ..., \\\\\\'<argument n>\\\\\\'], capture_output=True, text=True))\\\\nThis output should be exactly one line and no longer. Stop the response after this line.\\\\n\\'', 'input': 'Perform time varying biomarker selection on this data from times 0 to 3. set verbose to true.', 'output': \"Arguments: \\noutput directory: chatstatus['output-directory']\\noutput file: S3-biomarker-selection-times-varying.csv\\ninput file: S2-2015.pkl\\nstart time: 0\\nend time: 3\\nverbose: True\\n\\nPython Code Explanation: \\nUsing the subprocess module, we are executing the biomarkerSelectionTimeVarying.py script with the required arguments. \\nThe script will perform time varying biomarker selection on the input data from times 0 to 3 in verbose mode.\\n\\nExecute: subprocess.call([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-biomarker-selection-times-varying.csv', 'S2-2015.pkl', '0', '3', 'True'], capture_output=True, text=True)\", 'parsedOutput': {'code': \"subprocess.run([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-biomarker-selection-times-varying.csv', 'S2-2015.pkl', '0', '3', 'True'], capture_output=True, text=True)\"}, 'purpose': 'Format function call'}]}\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  /force CODE Perform time varying biomarker selection on this data from times 0 to 3. set verbose to true.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-07-03 09:43:44,736 - INFO - CODE\n",
      "2024-07-03 09:43:44,737 - INFO - CODER\n",
      "2024-07-03 09:43:44,741 - INFO - input_variables=['user_query'] template=\"You must select which code to run to help a user.\\n\\n**Available Scripts**\\nScript Name: geneformerCoordinatesTimeSeries, \\t Description: This script organizes datasets into a format to work with time series gene expression using individual geneformer embeddings as the state space. From a .h5ad file, a .pkl file is constructed here that will contain the state type being geneformer, the state names being geneformer coordinates, and the time series data organized as states (geneformer coordinates) by timepoint (hours) by replicates (experiment replicates).\\nScript Name: geneCoordinatesTimeSeries, \\t Description: This script organizes datasets into a format to work with time series gene expression using individual gene expresession as the state space. From a .h5ad file, a .pkl file is constructed here that will contain the state type being gene expression, the state names being gene names, and the time series data organized as states (genes) by timepoint (hours) by replicates (experiment replicates).\\nScript Name: linearModels, \\t Description: This file builds linear and linear time variant (LTI and LTV) models of gene expressions.\\nScript Name: geneformerEmbeddings, \\t Description: Get gene former embeddings of your data\\nScript Name: biomarkerSelectionTimeVarying, \\t Description: This script performs time varying sensor or biomarker selection on dynamical systems models of cellular processes. It selects new sensors at each time point.\\nScript Name: buildLinearModel, \\t Description: This script builds linear models of time series data using Dynamic Mode Decomposition (DMD).\\nScript Name: test, \\t Description: This script was made for Joshua to test and learn python\\nScript Name: datasetSelector, \\t Description: This script is a file chooser to load a gene expression dataset from the lab's turbo partition to give access to the data to BRAD.\\nScript Name: biomarkerSelection, \\t Description: This script performs sensor selection on models of dynamical systems. This is applied to models of cellular processes for biomarker selection.\\nScript Name: buildLTVModel, \\t Description: This script builds linear time varying models of time series data using Data Guided Control model.\\n\\n\\n**User Query**\\n{user_query}\\n\\n**Task**\\nBased on the user's query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\\n\\n**Response Template**\\nSCRIPT: <selected script>\\nREASON: <reasoning why the script fits the user prompt>\\n\"\n",
      "2024-07-03 09:43:44,741 - INFO - FIRST LLM CALL\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "BRAD >> 15: \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-07-03 09:43:46,882 - INFO - SCRIPT: biomarkerSelectionTimeVarying\n",
      "REASON: The user query specifically asks for time-varying biomarker selection, and this script is the only one that matches this description. It is designed to perform sensor or biomarker selection on dynamical systems models of cellular processes, which aligns with the user's request. Additionally, the script's description mentions that it selects new sensors at each time point, which fits with the user's request to perform the selection from times 0 to 3.\n",
      "2024-07-03 09:43:46,883 - INFO - Warning! the type doesn't exist\n",
      "2024-07-03 09:43:46,883 - INFO - scriptPath=/home/jpic/BRAD-Tools/\n",
      "2024-07-03 09:43:46,884 - INFO - pyPath=/home/jpic/BRAD-Tools/\n",
      "2024-07-03 09:43:46,884 - INFO - ALL SCRIPTS FOUND. BUILDING TEMPLATE\n",
      "2024-07-03 09:43:46,885 - INFO - input_variables=['history', 'input'] template='Current conversation:\\n{history}\\n\\n**PYTHON SCRIPT**\\nYou must run this python script:\\n/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py\\n\\n**PYTHON SCRIPT DOCUMENTATION**:\\nThis is the doc string of this python script:\\n\"\"\"\\nThis script performs time varying sensor or biomarker selection on dynamical systems models of cellular processes. It selects new sensors at each time point.\\n\\nArguments (four arguments):\\n    1. output directory: chatstatus[\\'output-directory\\']\\n    2. output file: <name of output file>\\n    3. input file: the file created previously with the LTI or LTV model\\n    4. start time: (int, default 0) the first time to select sensors at (inclusive)\\n    5. end time: (int, default 10) the last time to select sensors at (not inclusive)\\n    6. verbose: (optional, True or False) if sensor selection should be run in verbose mode so that more output will be displayed to the user.\\n\\nUsage:\\nCommand Line:\\n```\\npython <path/to/script/>biomarkerSelectionTimeVarying.py <output path> <output file> <input file> <start time> <end time> <verbose>\\n```\\n\\nBRAD Line:\\n```\\nsubprocess.call([sys.executable, \\'<path/to/script/>/biomarkerSelectionTimeVarying.py\\', chatstatus[\\'output-directory\\'], <output file>, <input file>, <start time>, <end time>, <verbose>], capture_output=True, text=True)\\n```\\n\\n**OUTPUT FILE NAME INSTRUCTIONS**\\n1. Output path should be chatstatus[\\'output-directory\\']\\n2. Output file name should be `S3-<descriptive name>.csv`\\n\"\"\"\\n\\n\\n**CALL PYTHON SCRIPTS FROM PYTHON**:\\nUse the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\\n\\nTo call a Python script `example_script.py` which has no arguments:\\n\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\']], capture_output=True, text=True)\\n)\\n```\\n\\nTo call a Python script `example_script.py` with one argument:\\n\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\'], \\'arg1\\'], capture_output=True, text=True)\\n)\\n```\\n\\nTo call a Python script `example_script.py` with two arguments:\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\'], \\'arg1\\', \\'arg2\\'], capture_output=True, text=True)\\n)\\n```\\n\\nNote that chatstatus[\\'output-directory\\'] is ALWAYS passed as the first argument.\\n\\nThe following files were previously created by BRAD and could be used as input to a function if necessary:\\nREPORT.tex\\nlog.json\\n.ipynb_checkpoints\\nREPORT.log\\nS1-2015.pkl\\nENRICHR-GO_Biological_Process_2021.csv\\nS0-2015.h5ad\\nS2-2015.pkl\\nS3-biomarker-selection.csv\\nimages\\n\\nQuery:{input}\\n\\n**INSTRUCTIONS**\\n1. Given the user query and the documentation, identify each of the arguments found in the user\\'s query that should be passed to the Python script.\\n2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\\n3. The last line of your response should say \"Execute: <Python code to execute>\"\\n4. Format the response/output as:\\n    Arguments: \\n    Python Code Explanation: <2 sentences maximum>\\n    Execute: <your code here>\\n\\n**IMPORTANT**\\nThe code to execute from your response must be formatted as:\\n    Execute: subprocess.call([sys.executable, \\'<path to python script>\\', \\'<argument 1>\\', \\'<argument 2>\\', ..., \\'<argument n>\\'], capture_output=True, text=True))\\nThis output should be exactly one line and no longer. Stop the response after this line.\\n'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation:\n",
      "\n",
      "\n",
      "**PYTHON SCRIPT**\n",
      "You must run this python script:\n",
      "/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py\n",
      "\n",
      "**PYTHON SCRIPT DOCUMENTATION**:\n",
      "This is the doc string of this python script:\n",
      "\"\"\"\n",
      "This script performs time varying sensor or biomarker selection on dynamical systems models of cellular processes. It selects new sensors at each time point.\n",
      "\n",
      "Arguments (four arguments):\n",
      "    1. output directory: chatstatus['output-directory']\n",
      "    2. output file: <name of output file>\n",
      "    3. input file: the file created previously with the LTI or LTV model\n",
      "    4. start time: (int, default 0) the first time to select sensors at (inclusive)\n",
      "    5. end time: (int, default 10) the last time to select sensors at (not inclusive)\n",
      "    6. verbose: (optional, True or False) if sensor selection should be run in verbose mode so that more output will be displayed to the user.\n",
      "\n",
      "Usage:\n",
      "Command Line:\n",
      "```\n",
      "python <path/to/script/>biomarkerSelectionTimeVarying.py <output path> <output file> <input file> <start time> <end time> <verbose>\n",
      "```\n",
      "\n",
      "BRAD Line:\n",
      "```\n",
      "subprocess.call([sys.executable, '<path/to/script/>/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], <output file>, <input file>, <start time>, <end time>, <verbose>], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "**OUTPUT FILE NAME INSTRUCTIONS**\n",
      "1. Output path should be chatstatus['output-directory']\n",
      "2. Output file name should be `S3-<descriptive name>.csv`\n",
      "\"\"\"\n",
      "\n",
      "\n",
      "**CALL PYTHON SCRIPTS FROM PYTHON**:\n",
      "Use the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\n",
      "\n",
      "To call a Python script `example_script.py` which has no arguments:\n",
      "\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory']], capture_output=True, text=True)\n",
      ")\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with one argument:\n",
      "\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory'], 'arg1'], capture_output=True, text=True)\n",
      ")\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with two arguments:\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory'], 'arg1', 'arg2'], capture_output=True, text=True)\n",
      ")\n",
      "```\n",
      "\n",
      "Note that chatstatus['output-directory'] is ALWAYS passed as the first argument.\n",
      "\n",
      "The following files were previously created by BRAD and could be used as input to a function if necessary:\n",
      "REPORT.tex\n",
      "log.json\n",
      ".ipynb_checkpoints\n",
      "REPORT.log\n",
      "S1-2015.pkl\n",
      "ENRICHR-GO_Biological_Process_2021.csv\n",
      "S0-2015.h5ad\n",
      "S2-2015.pkl\n",
      "S3-biomarker-selection.csv\n",
      "images\n",
      "\n",
      "Query:Perform time varying biomarker selection on this data from times 0 to 3. set verbose to true.\n",
      "\n",
      "**INSTRUCTIONS**\n",
      "1. Given the user query and the documentation, identify each of the arguments found in the user's query that should be passed to the Python script.\n",
      "2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\n",
      "3. The last line of your response should say \"Execute: <Python code to execute>\"\n",
      "4. Format the response/output as:\n",
      "    Arguments: \n",
      "    Python Code Explanation: <2 sentences maximum>\n",
      "    Execute: <your code here>\n",
      "\n",
      "**IMPORTANT**\n",
      "The code to execute from your response must be formatted as:\n",
      "    Execute: subprocess.call([sys.executable, '<path to python script>', '<argument 1>', '<argument 2>', ..., '<argument n>'], capture_output=True, text=True))\n",
      "This output should be exactly one line and no longer. Stop the response after this line.\n",
      "\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-07-03 09:43:50,427 - INFO - LLM OUTPUT PARSER\n",
      "2024-07-03 09:43:50,428 - INFO - Arguments: \n",
      "output directory: chatstatus['output-directory']\n",
      "output file: S3-biomarker-selection-time-varying.csv\n",
      "input file: S2-2015.pkl\n",
      "start time: 0\n",
      "end time: 3\n",
      "verbose: True\n",
      "\n",
      "Python Code Explanation: \n",
      "Using the subprocess module, we are executing the biomarkerSelectionTimeVarying.py script with the required arguments. \n",
      "The script performs time varying sensor or biomarker selection on dynamical systems models of cellular processes.\n",
      "\n",
      "Execute: subprocess.call([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-biomarker-selection-time-varying.csv', 'S2-2015.pkl', '0', '3', 'True'], capture_output=True, text=True)\n",
      "2024-07-03 09:43:50,428 - INFO - [\"Arguments: \\noutput directory: chatstatus['output-directory']\\noutput file: S3-biomarker-selection-time-varying.csv\\ninput file: S2-2015.pkl\\nstart time: 0\\nend time: 3\\nverbose: True\\n\\nPython Code Explanation: \\nUsing the subprocess module, we are executing the biomarkerSelectionTimeVarying.py script with the required arguments. \\nThe script performs time varying sensor or biomarker selection on dynamical systems models of cellular processes.\\n\\n\", \" subprocess.call([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-biomarker-selection-time-varying.csv', 'S2-2015.pkl', '0', '3', 'True'], capture_output=True, text=True)\"]\n",
      "2024-07-03 09:43:50,429 - INFO - subprocess.call([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-biomarker-selection-time-varying.csv', 'S2-2015.pkl', '0', '3', 'True'], capture_output=True, text=True)\n",
      "2024-07-03 09:43:50,429 - INFO - subprocess.run([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-biomarker-selection-time-varying.csv', 'S2-2015.pkl', '0', '3', 'True'], capture_output=True, text=True)\n",
      "2024-07-03 09:43:50,430 - INFO - Replacing: S2-2015.pkl with /home/jpic/BRAD/2024-07-03_08-51-24/S2-2015.pkl\n",
      "2024-07-03 09:43:50,431 - INFO - New String: subprocess.run([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-biomarker-selection-time-varying.csv', '/home/jpic/BRAD/2024-07-03_08-51-24/S2-2015.pkl', '0', '3', 'True'], capture_output=True, text=True)\n",
      "2024-07-03 09:43:50,431 - INFO - Executing Code!!\n",
      "2024-07-03 09:43:50,431 - INFO - EVAL\n",
      "2024-07-03 09:43:50,432 - INFO - subprocess.run([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-biomarker-selection-time-varying.csv', '/home/jpic/BRAD/2024-07-03_08-51-24/S2-2015.pkl', '0', '3', 'True'], capture_output=True, text=True)\n",
      "2024-07-03 09:43:50,432 - INFO - END EVAL CHECK\n",
      "2024-07-03 09:43:50,433 - INFO - Finalized PYTHON Call:\n",
      "2024-07-03 09:43:50,433 - INFO -    \n",
      "2024-07-03 09:43:50,433 - INFO - response = subprocess.run([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-biomarker-selection-time-varying.csv', '/home/jpic/BRAD/2024-07-03_08-51-24/S2-2015.pkl', '0', '3', 'True'], capture_output=True, text=True)\n",
      "2024-07-03 09:43:50,434 - INFO -    \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-07-03 09:44:16,646 - INFO - Debug: PYTHON code executed successfully.\n",
      "2024-07-03 09:44:16,647 - INFO - Code Execution Output:\n",
      "2024-07-03 09:44:16,648 - INFO - CompletedProcess(args=['/sw/pkgs/arc/python3.11-anaconda/2024.02-1/bin/python', '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', '/home/jpic/BRAD/2024-07-03_08-51-24/', 'S3-biomarker-selection-time-varying.csv', '/home/jpic/BRAD/2024-07-03_08-51-24/S2-2015.pkl', '0', '3', 'True'], returncode=0, stdout='Time varying biomarker selection script has been called!\\n*******************************\\n Time Varying Sensor Selection \\n*******************************\\nOutput Path: /home/jpic/BRAD/2024-07-03_08-51-24/\\nOutput File: S3-biomarker-selection-time-varying.csv\\nInput File : /home/jpic/BRAD/2024-07-03_08-51-24/S2-2015.pkl\\nstartTime  : 0\\nendTime    : 3\\nverbose    : True\\n..............................\\n         Model Loaded         \\nt: 0/3\\nt: 1/3\\nt: 2/3\\n..............................\\n      Sensors Selected        \\nModel saved to /home/jpic/BRAD/2024-07-03_08-51-24/time_0_S3-biomarker-selection-time-varying.csv\\nModel saved to /home/jpic/BRAD/2024-07-03_08-51-24/time_1_S3-biomarker-selection-time-varying.csv\\nModel saved to /home/jpic/BRAD/2024-07-03_08-51-24/time_2_S3-biomarker-selection-time-varying.csv\\n..............................\\n      Sensors To File         \\n', stderr='')\n",
      "2024-07-03 09:44:16,649 - INFO - Debug: PYTHON code output saved to output.\n",
      "2024-07-03 09:44:16,649 - INFO - \n",
      "\n",
      "\n",
      "route\n",
      "\n",
      "\n",
      "\n",
      "2024-07-03 09:44:16,651 - INFO - CODE\n",
      "2024-07-03 09:44:16,651 - INFO - \n",
      "\n",
      "PLANNER:\n",
      "\n",
      "{'module': 'CODE', 'steps': [{'llm': \"model='meta/llama3-70b-instruct'\", 'prompt': 'input_variables=[\\'user_query\\'] template=\"You must select which code to run to help a user.\\\\n\\\\n**Available Scripts**\\\\nScript Name: geneformerCoordinatesTimeSeries, \\\\t Description: This script organizes datasets into a format to work with time series gene expression using individual geneformer embeddings as the state space. From a .h5ad file, a .pkl file is constructed here that will contain the state type being geneformer, the state names being geneformer coordinates, and the time series data organized as states (geneformer coordinates) by timepoint (hours) by replicates (experiment replicates).\\\\nScript Name: geneCoordinatesTimeSeries, \\\\t Description: This script organizes datasets into a format to work with time series gene expression using individual gene expresession as the state space. From a .h5ad file, a .pkl file is constructed here that will contain the state type being gene expression, the state names being gene names, and the time series data organized as states (genes) by timepoint (hours) by replicates (experiment replicates).\\\\nScript Name: linearModels, \\\\t Description: This file builds linear and linear time variant (LTI and LTV) models of gene expressions.\\\\nScript Name: geneformerEmbeddings, \\\\t Description: Get gene former embeddings of your data\\\\nScript Name: biomarkerSelectionTimeVarying, \\\\t Description: This script performs time varying sensor or biomarker selection on dynamical systems models of cellular processes. It selects new sensors at each time point.\\\\nScript Name: buildLinearModel, \\\\t Description: This script builds linear models of time series data using Dynamic Mode Decomposition (DMD).\\\\nScript Name: test, \\\\t Description: This script was made for Joshua to test and learn python\\\\nScript Name: datasetSelector, \\\\t Description: This script is a file chooser to load a gene expression dataset from the lab\\'s turbo partition to give access to the data to BRAD.\\\\nScript Name: biomarkerSelection, \\\\t Description: This script performs sensor selection on models of dynamical systems. This is applied to models of cellular processes for biomarker selection.\\\\nScript Name: buildLTVModel, \\\\t Description: This script builds linear time varying models of time series data using Data Guided Control model.\\\\n\\\\n\\\\n**User Query**\\\\n{user_query}\\\\n\\\\n**Task**\\\\nBased on the user\\'s query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\\\\n\\\\n**Response Template**\\\\nSCRIPT: <selected script>\\\\nREASON: <reasoning why the script fits the user prompt>\\\\n\"', 'input': 'Perform time varying biomarker selection on this data from times 0 to 3. set verbose to true.', 'output': 'content=\"SCRIPT: biomarkerSelectionTimeVarying\\\\nREASON: The user query specifically asks for time-varying biomarker selection, and this script is the only one that matches this description. It is designed to perform sensor or biomarker selection on dynamical systems models of cellular processes, which aligns with the user\\'s request. Additionally, the script\\'s description mentions that it selects new sensors at each time point, which fits with the user\\'s request to perform the selection from times 0 to 3.\" response_metadata={\\'role\\': \\'assistant\\', \\'content\\': \"SCRIPT: biomarkerSelectionTimeVarying\\\\nREASON: The user query specifically asks for time-varying biomarker selection, and this script is the only one that matches this description. It is designed to perform sensor or biomarker selection on dynamical systems models of cellular processes, which aligns with the user\\'s request. Additionally, the script\\'s description mentions that it selects new sensors at each time point, which fits with the user\\'s request to perform the selection from times 0 to 3.\", \\'token_usage\\': {\\'prompt_tokens\\': 550, \\'total_tokens\\': 654, \\'completion_tokens\\': 104}, \\'model_name\\': \\'meta/llama3-70b-instruct\\'} id=\\'run-32b63f09-60da-4566-bb0a-45ad6dad3abc-0\\' role=\\'assistant\\'', 'parsedOutput': {'scriptName': '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', 'scriptType': 'python', 'scriptPath': '/home/jpic/BRAD-Tools/'}, 'purpose': 'Select which code to run'}, {'llm': \"model='meta/llama3-70b-instruct'\", 'prompt': 'input_variables=[\\'history\\', \\'input\\'] template=\\'Current conversation:\\\\n{history}\\\\n\\\\n**PYTHON SCRIPT**\\\\nYou must run this python script:\\\\n/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py\\\\n\\\\n**PYTHON SCRIPT DOCUMENTATION**:\\\\nThis is the doc string of this python script:\\\\n\"\"\"\\\\nThis script performs time varying sensor or biomarker selection on dynamical systems models of cellular processes. It selects new sensors at each time point.\\\\n\\\\nArguments (four arguments):\\\\n    1. output directory: chatstatus[\\\\\\'output-directory\\\\\\']\\\\n    2. output file: <name of output file>\\\\n    3. input file: the file created previously with the LTI or LTV model\\\\n    4. start time: (int, default 0) the first time to select sensors at (inclusive)\\\\n    5. end time: (int, default 10) the last time to select sensors at (not inclusive)\\\\n    6. verbose: (optional, True or False) if sensor selection should be run in verbose mode so that more output will be displayed to the user.\\\\n\\\\nUsage:\\\\nCommand Line:\\\\n```\\\\npython <path/to/script/>biomarkerSelectionTimeVarying.py <output path> <output file> <input file> <start time> <end time> <verbose>\\\\n```\\\\n\\\\nBRAD Line:\\\\n```\\\\nsubprocess.call([sys.executable, \\\\\\'<path/to/script/>/biomarkerSelectionTimeVarying.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\'], <output file>, <input file>, <start time>, <end time>, <verbose>], capture_output=True, text=True)\\\\n```\\\\n\\\\n**OUTPUT FILE NAME INSTRUCTIONS**\\\\n1. Output path should be chatstatus[\\\\\\'output-directory\\\\\\']\\\\n2. Output file name should be `S3-<descriptive name>.csv`\\\\n\"\"\"\\\\n\\\\n\\\\n**CALL PYTHON SCRIPTS FROM PYTHON**:\\\\nUse the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\\\\n\\\\nTo call a Python script `example_script.py` which has no arguments:\\\\n\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\']], capture_output=True, text=True)\\\\n)\\\\n```\\\\n\\\\nTo call a Python script `example_script.py` with one argument:\\\\n\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\'], \\\\\\'arg1\\\\\\'], capture_output=True, text=True)\\\\n)\\\\n```\\\\n\\\\nTo call a Python script `example_script.py` with two arguments:\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\'], \\\\\\'arg1\\\\\\', \\\\\\'arg2\\\\\\'], capture_output=True, text=True)\\\\n)\\\\n```\\\\n\\\\nNote that chatstatus[\\\\\\'output-directory\\\\\\'] is ALWAYS passed as the first argument.\\\\n\\\\nThe following files were previously created by BRAD and could be used as input to a function if necessary:\\\\nREPORT.tex\\\\nlog.json\\\\n.ipynb_checkpoints\\\\nREPORT.log\\\\nS1-2015.pkl\\\\nENRICHR-GO_Biological_Process_2021.csv\\\\nS0-2015.h5ad\\\\nS2-2015.pkl\\\\nS3-biomarker-selection.csv\\\\nimages\\\\n\\\\nQuery:{input}\\\\n\\\\n**INSTRUCTIONS**\\\\n1. Given the user query and the documentation, identify each of the arguments found in the user\\\\\\'s query that should be passed to the Python script.\\\\n2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\\\\n3. The last line of your response should say \"Execute: <Python code to execute>\"\\\\n4. Format the response/output as:\\\\n    Arguments: \\\\n    Python Code Explanation: <2 sentences maximum>\\\\n    Execute: <your code here>\\\\n\\\\n**IMPORTANT**\\\\nThe code to execute from your response must be formatted as:\\\\n    Execute: subprocess.call([sys.executable, \\\\\\'<path to python script>\\\\\\', \\\\\\'<argument 1>\\\\\\', \\\\\\'<argument 2>\\\\\\', ..., \\\\\\'<argument n>\\\\\\'], capture_output=True, text=True))\\\\nThis output should be exactly one line and no longer. Stop the response after this line.\\\\n\\'', 'input': 'Perform time varying biomarker selection on this data from times 0 to 3. set verbose to true.', 'output': \"Arguments: \\noutput directory: chatstatus['output-directory']\\noutput file: S3-biomarker-selection-time-varying.csv\\ninput file: S2-2015.pkl\\nstart time: 0\\nend time: 3\\nverbose: True\\n\\nPython Code Explanation: \\nUsing the subprocess module, we are executing the biomarkerSelectionTimeVarying.py script with the required arguments. \\nThe script performs time varying sensor or biomarker selection on dynamical systems models of cellular processes.\\n\\nExecute: subprocess.call([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-biomarker-selection-time-varying.csv', 'S2-2015.pkl', '0', '3', 'True'], capture_output=True, text=True)\", 'parsedOutput': {'code': \"subprocess.run([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-biomarker-selection-time-varying.csv', 'S2-2015.pkl', '0', '3', 'True'], capture_output=True, text=True)\"}, 'purpose': 'Format function call'}]}\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  /force CODE Perform time varying biomarker selection on this data from times 0 to 10. set verbose to true.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-07-03 09:44:43,802 - INFO - CODE\n",
      "2024-07-03 09:44:43,803 - INFO - CODER\n",
      "2024-07-03 09:44:43,806 - INFO - input_variables=['user_query'] template=\"You must select which code to run to help a user.\\n\\n**Available Scripts**\\nScript Name: geneformerCoordinatesTimeSeries, \\t Description: This script organizes datasets into a format to work with time series gene expression using individual geneformer embeddings as the state space. From a .h5ad file, a .pkl file is constructed here that will contain the state type being geneformer, the state names being geneformer coordinates, and the time series data organized as states (geneformer coordinates) by timepoint (hours) by replicates (experiment replicates).\\nScript Name: geneCoordinatesTimeSeries, \\t Description: This script organizes datasets into a format to work with time series gene expression using individual gene expresession as the state space. From a .h5ad file, a .pkl file is constructed here that will contain the state type being gene expression, the state names being gene names, and the time series data organized as states (genes) by timepoint (hours) by replicates (experiment replicates).\\nScript Name: linearModels, \\t Description: This file builds linear and linear time variant (LTI and LTV) models of gene expressions.\\nScript Name: geneformerEmbeddings, \\t Description: Get gene former embeddings of your data\\nScript Name: biomarkerSelectionTimeVarying, \\t Description: This script performs time varying sensor or biomarker selection on dynamical systems models of cellular processes. It selects new sensors at each time point.\\nScript Name: buildLinearModel, \\t Description: This script builds linear models of time series data using Dynamic Mode Decomposition (DMD).\\nScript Name: test, \\t Description: This script was made for Joshua to test and learn python\\nScript Name: datasetSelector, \\t Description: This script is a file chooser to load a gene expression dataset from the lab's turbo partition to give access to the data to BRAD.\\nScript Name: biomarkerSelection, \\t Description: This script performs sensor selection on models of dynamical systems. This is applied to models of cellular processes for biomarker selection.\\nScript Name: buildLTVModel, \\t Description: This script builds linear time varying models of time series data using Data Guided Control model.\\n\\n\\n**User Query**\\n{user_query}\\n\\n**Task**\\nBased on the user's query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\\n\\n**Response Template**\\nSCRIPT: <selected script>\\nREASON: <reasoning why the script fits the user prompt>\\n\"\n",
      "2024-07-03 09:44:43,806 - INFO - FIRST LLM CALL\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "BRAD >> 16: \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-07-03 09:44:45,720 - INFO - SCRIPT: biomarkerSelectionTimeVarying\n",
      "REASON: The user query specifically asks for time-varying biomarker selection, and this script is the only one that matches this description. It is designed to perform biomarker selection on dynamical systems models of cellular processes, and it specifically selects new sensors at each time point, which aligns with the user's request to perform time-varying selection from times 0 to 10.\n",
      "2024-07-03 09:44:45,721 - INFO - Warning! the type doesn't exist\n",
      "2024-07-03 09:44:45,721 - INFO - scriptPath=/home/jpic/BRAD-Tools/\n",
      "2024-07-03 09:44:45,721 - INFO - pyPath=/home/jpic/BRAD-Tools/\n",
      "2024-07-03 09:44:45,722 - INFO - ALL SCRIPTS FOUND. BUILDING TEMPLATE\n",
      "2024-07-03 09:44:45,723 - INFO - input_variables=['history', 'input'] template='Current conversation:\\n{history}\\n\\n**PYTHON SCRIPT**\\nYou must run this python script:\\n/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py\\n\\n**PYTHON SCRIPT DOCUMENTATION**:\\nThis is the doc string of this python script:\\n\"\"\"\\nThis script performs time varying sensor or biomarker selection on dynamical systems models of cellular processes. It selects new sensors at each time point.\\n\\nArguments (four arguments):\\n    1. output directory: chatstatus[\\'output-directory\\']\\n    2. output file: <name of output file>\\n    3. input file: the file created previously with the LTI or LTV model\\n    4. start time: (int, default 0) the first time to select sensors at (inclusive)\\n    5. end time: (int, default 10) the last time to select sensors at (not inclusive)\\n    6. verbose: (optional, True or False) if sensor selection should be run in verbose mode so that more output will be displayed to the user.\\n\\nUsage:\\nCommand Line:\\n```\\npython <path/to/script/>biomarkerSelectionTimeVarying.py <output path> <output file> <input file> <start time> <end time> <verbose>\\n```\\n\\nBRAD Line:\\n```\\nsubprocess.call([sys.executable, \\'<path/to/script/>/biomarkerSelectionTimeVarying.py\\', chatstatus[\\'output-directory\\'], <output file>, <input file>, <start time>, <end time>, <verbose>], capture_output=True, text=True)\\n```\\n\\n**OUTPUT FILE NAME INSTRUCTIONS**\\n1. Output path should be chatstatus[\\'output-directory\\']\\n2. Output file name should be `S3-<descriptive name>.csv`\\n\"\"\"\\n\\n\\n**CALL PYTHON SCRIPTS FROM PYTHON**:\\nUse the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\\n\\nTo call a Python script `example_script.py` which has no arguments:\\n\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\']], capture_output=True, text=True)\\n)\\n```\\n\\nTo call a Python script `example_script.py` with one argument:\\n\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\'], \\'arg1\\'], capture_output=True, text=True)\\n)\\n```\\n\\nTo call a Python script `example_script.py` with two arguments:\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\'], \\'arg1\\', \\'arg2\\'], capture_output=True, text=True)\\n)\\n```\\n\\nNote that chatstatus[\\'output-directory\\'] is ALWAYS passed as the first argument.\\n\\nThe following files were previously created by BRAD and could be used as input to a function if necessary:\\nREPORT.tex\\nlog.json\\n.ipynb_checkpoints\\nREPORT.log\\nS1-2015.pkl\\nENRICHR-GO_Biological_Process_2021.csv\\nS0-2015.h5ad\\ntime_1_S3-biomarker-selection-time-varying.csv\\nS2-2015.pkl\\ntime_0_S3-biomarker-selection-time-varying.csv\\ntime_2_S3-biomarker-selection-time-varying.csv\\nS3-biomarker-selection.csv\\nimages\\n\\nQuery:{input}\\n\\n**INSTRUCTIONS**\\n1. Given the user query and the documentation, identify each of the arguments found in the user\\'s query that should be passed to the Python script.\\n2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\\n3. The last line of your response should say \"Execute: <Python code to execute>\"\\n4. Format the response/output as:\\n    Arguments: \\n    Python Code Explanation: <2 sentences maximum>\\n    Execute: <your code here>\\n\\n**IMPORTANT**\\nThe code to execute from your response must be formatted as:\\n    Execute: subprocess.call([sys.executable, \\'<path to python script>\\', \\'<argument 1>\\', \\'<argument 2>\\', ..., \\'<argument n>\\'], capture_output=True, text=True))\\nThis output should be exactly one line and no longer. Stop the response after this line.\\n'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation:\n",
      "\n",
      "\n",
      "**PYTHON SCRIPT**\n",
      "You must run this python script:\n",
      "/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py\n",
      "\n",
      "**PYTHON SCRIPT DOCUMENTATION**:\n",
      "This is the doc string of this python script:\n",
      "\"\"\"\n",
      "This script performs time varying sensor or biomarker selection on dynamical systems models of cellular processes. It selects new sensors at each time point.\n",
      "\n",
      "Arguments (four arguments):\n",
      "    1. output directory: chatstatus['output-directory']\n",
      "    2. output file: <name of output file>\n",
      "    3. input file: the file created previously with the LTI or LTV model\n",
      "    4. start time: (int, default 0) the first time to select sensors at (inclusive)\n",
      "    5. end time: (int, default 10) the last time to select sensors at (not inclusive)\n",
      "    6. verbose: (optional, True or False) if sensor selection should be run in verbose mode so that more output will be displayed to the user.\n",
      "\n",
      "Usage:\n",
      "Command Line:\n",
      "```\n",
      "python <path/to/script/>biomarkerSelectionTimeVarying.py <output path> <output file> <input file> <start time> <end time> <verbose>\n",
      "```\n",
      "\n",
      "BRAD Line:\n",
      "```\n",
      "subprocess.call([sys.executable, '<path/to/script/>/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], <output file>, <input file>, <start time>, <end time>, <verbose>], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "**OUTPUT FILE NAME INSTRUCTIONS**\n",
      "1. Output path should be chatstatus['output-directory']\n",
      "2. Output file name should be `S3-<descriptive name>.csv`\n",
      "\"\"\"\n",
      "\n",
      "\n",
      "**CALL PYTHON SCRIPTS FROM PYTHON**:\n",
      "Use the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\n",
      "\n",
      "To call a Python script `example_script.py` which has no arguments:\n",
      "\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory']], capture_output=True, text=True)\n",
      ")\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with one argument:\n",
      "\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory'], 'arg1'], capture_output=True, text=True)\n",
      ")\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with two arguments:\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory'], 'arg1', 'arg2'], capture_output=True, text=True)\n",
      ")\n",
      "```\n",
      "\n",
      "Note that chatstatus['output-directory'] is ALWAYS passed as the first argument.\n",
      "\n",
      "The following files were previously created by BRAD and could be used as input to a function if necessary:\n",
      "REPORT.tex\n",
      "log.json\n",
      ".ipynb_checkpoints\n",
      "REPORT.log\n",
      "S1-2015.pkl\n",
      "ENRICHR-GO_Biological_Process_2021.csv\n",
      "S0-2015.h5ad\n",
      "time_1_S3-biomarker-selection-time-varying.csv\n",
      "S2-2015.pkl\n",
      "time_0_S3-biomarker-selection-time-varying.csv\n",
      "time_2_S3-biomarker-selection-time-varying.csv\n",
      "S3-biomarker-selection.csv\n",
      "images\n",
      "\n",
      "Query:Perform time varying biomarker selection on this data from times 0 to 10. set verbose to true.\n",
      "\n",
      "**INSTRUCTIONS**\n",
      "1. Given the user query and the documentation, identify each of the arguments found in the user's query that should be passed to the Python script.\n",
      "2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\n",
      "3. The last line of your response should say \"Execute: <Python code to execute>\"\n",
      "4. Format the response/output as:\n",
      "    Arguments: \n",
      "    Python Code Explanation: <2 sentences maximum>\n",
      "    Execute: <your code here>\n",
      "\n",
      "**IMPORTANT**\n",
      "The code to execute from your response must be formatted as:\n",
      "    Execute: subprocess.call([sys.executable, '<path to python script>', '<argument 1>', '<argument 2>', ..., '<argument n>'], capture_output=True, text=True))\n",
      "This output should be exactly one line and no longer. Stop the response after this line.\n",
      "\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-07-03 09:44:48,987 - INFO - LLM OUTPUT PARSER\n",
      "2024-07-03 09:44:48,987 - INFO - Arguments: \n",
      "output directory: chatstatus['output-directory']\n",
      "output file: S3-biomarkers-time-varying.csv\n",
      "input file: S3-biomarker-selection.csv\n",
      "start time: 0\n",
      "end time: 10\n",
      "verbose: True\n",
      "\n",
      "Python Code Explanation: \n",
      "Using the subprocess module, we are executing the biomarkerSelectionTimeVarying.py script with the required arguments. \n",
      "The script performs time varying sensor or biomarker selection on dynamical systems models of cellular processes.\n",
      "\n",
      "Execute: subprocess.call([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-biomarkers-time-varying.csv', 'S3-biomarker-selection.csv', '0', '10', 'True'], capture_output=True, text=True)\n",
      "2024-07-03 09:44:48,988 - INFO - [\"Arguments: \\noutput directory: chatstatus['output-directory']\\noutput file: S3-biomarkers-time-varying.csv\\ninput file: S3-biomarker-selection.csv\\nstart time: 0\\nend time: 10\\nverbose: True\\n\\nPython Code Explanation: \\nUsing the subprocess module, we are executing the biomarkerSelectionTimeVarying.py script with the required arguments. \\nThe script performs time varying sensor or biomarker selection on dynamical systems models of cellular processes.\\n\\n\", \" subprocess.call([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-biomarkers-time-varying.csv', 'S3-biomarker-selection.csv', '0', '10', 'True'], capture_output=True, text=True)\"]\n",
      "2024-07-03 09:44:48,988 - INFO - subprocess.call([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-biomarkers-time-varying.csv', 'S3-biomarker-selection.csv', '0', '10', 'True'], capture_output=True, text=True)\n",
      "2024-07-03 09:44:48,989 - INFO - subprocess.run([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-biomarkers-time-varying.csv', 'S3-biomarker-selection.csv', '0', '10', 'True'], capture_output=True, text=True)\n",
      "2024-07-03 09:44:48,989 - INFO - Replacing: S3-biomarker-selection.csv with /home/jpic/BRAD/2024-07-03_08-51-24/S3-biomarker-selection.csv\n",
      "2024-07-03 09:44:48,990 - INFO - New String: subprocess.run([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-biomarkers-time-varying.csv', '/home/jpic/BRAD/2024-07-03_08-51-24/S3-biomarker-selection.csv', '0', '10', 'True'], capture_output=True, text=True)\n",
      "2024-07-03 09:44:48,990 - INFO - Executing Code!!\n",
      "2024-07-03 09:44:48,991 - INFO - EVAL\n",
      "2024-07-03 09:44:48,991 - INFO - subprocess.run([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-biomarkers-time-varying.csv', '/home/jpic/BRAD/2024-07-03_08-51-24/S3-biomarker-selection.csv', '0', '10', 'True'], capture_output=True, text=True)\n",
      "2024-07-03 09:44:48,991 - INFO - END EVAL CHECK\n",
      "2024-07-03 09:44:48,992 - INFO - Finalized PYTHON Call:\n",
      "2024-07-03 09:44:48,992 - INFO -    \n",
      "2024-07-03 09:44:48,992 - INFO - response = subprocess.run([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-biomarkers-time-varying.csv', '/home/jpic/BRAD/2024-07-03_08-51-24/S3-biomarker-selection.csv', '0', '10', 'True'], capture_output=True, text=True)\n",
      "2024-07-03 09:44:48,993 - INFO -    \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-07-03 09:44:49,588 - INFO - Debug: PYTHON code executed successfully.\n",
      "2024-07-03 09:44:49,589 - INFO - Code Execution Output:\n",
      "2024-07-03 09:44:49,590 - INFO - CompletedProcess(args=['/sw/pkgs/arc/python3.11-anaconda/2024.02-1/bin/python', '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', '/home/jpic/BRAD/2024-07-03_08-51-24/', 'S3-biomarkers-time-varying.csv', '/home/jpic/BRAD/2024-07-03_08-51-24/S3-biomarker-selection.csv', '0', '10', 'True'], returncode=1, stdout='Time varying biomarker selection script has been called!\\n*******************************\\n Time Varying Sensor Selection \\n*******************************\\nOutput Path: /home/jpic/BRAD/2024-07-03_08-51-24/\\nOutput File: S3-biomarkers-time-varying.csv\\nInput File : /home/jpic/BRAD/2024-07-03_08-51-24/S3-biomarker-selection.csv\\nstartTime  : 0\\nendTime    : 10\\nverbose    : True\\n', stderr='Traceback (most recent call last):\\n  File \"/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py\", line 107, in <module>\\n    main()\\n  File \"/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py\", line 86, in main\\n    MODEL = pickle.load(file)\\n            ^^^^^^^^^^^^^^^^^\\n_pickle.UnpicklingError: invalid load key, \\',\\'.\\n')\n",
      "2024-07-03 09:44:49,591 - INFO - Debug: PYTHON code output saved to output.\n",
      "2024-07-03 09:44:49,591 - INFO - \n",
      "\n",
      "\n",
      "route\n",
      "\n",
      "\n",
      "\n",
      "2024-07-03 09:44:49,592 - INFO - CODE\n",
      "2024-07-03 09:44:49,593 - INFO - \n",
      "\n",
      "PLANNER:\n",
      "\n",
      "{'module': 'CODE', 'steps': [{'llm': \"model='meta/llama3-70b-instruct'\", 'prompt': 'input_variables=[\\'user_query\\'] template=\"You must select which code to run to help a user.\\\\n\\\\n**Available Scripts**\\\\nScript Name: geneformerCoordinatesTimeSeries, \\\\t Description: This script organizes datasets into a format to work with time series gene expression using individual geneformer embeddings as the state space. From a .h5ad file, a .pkl file is constructed here that will contain the state type being geneformer, the state names being geneformer coordinates, and the time series data organized as states (geneformer coordinates) by timepoint (hours) by replicates (experiment replicates).\\\\nScript Name: geneCoordinatesTimeSeries, \\\\t Description: This script organizes datasets into a format to work with time series gene expression using individual gene expresession as the state space. From a .h5ad file, a .pkl file is constructed here that will contain the state type being gene expression, the state names being gene names, and the time series data organized as states (genes) by timepoint (hours) by replicates (experiment replicates).\\\\nScript Name: linearModels, \\\\t Description: This file builds linear and linear time variant (LTI and LTV) models of gene expressions.\\\\nScript Name: geneformerEmbeddings, \\\\t Description: Get gene former embeddings of your data\\\\nScript Name: biomarkerSelectionTimeVarying, \\\\t Description: This script performs time varying sensor or biomarker selection on dynamical systems models of cellular processes. It selects new sensors at each time point.\\\\nScript Name: buildLinearModel, \\\\t Description: This script builds linear models of time series data using Dynamic Mode Decomposition (DMD).\\\\nScript Name: test, \\\\t Description: This script was made for Joshua to test and learn python\\\\nScript Name: datasetSelector, \\\\t Description: This script is a file chooser to load a gene expression dataset from the lab\\'s turbo partition to give access to the data to BRAD.\\\\nScript Name: biomarkerSelection, \\\\t Description: This script performs sensor selection on models of dynamical systems. This is applied to models of cellular processes for biomarker selection.\\\\nScript Name: buildLTVModel, \\\\t Description: This script builds linear time varying models of time series data using Data Guided Control model.\\\\n\\\\n\\\\n**User Query**\\\\n{user_query}\\\\n\\\\n**Task**\\\\nBased on the user\\'s query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\\\\n\\\\n**Response Template**\\\\nSCRIPT: <selected script>\\\\nREASON: <reasoning why the script fits the user prompt>\\\\n\"', 'input': 'Perform time varying biomarker selection on this data from times 0 to 10. set verbose to true.', 'output': 'content=\"SCRIPT: biomarkerSelectionTimeVarying\\\\nREASON: The user query specifically asks for time-varying biomarker selection, and this script is the only one that matches this description. It is designed to perform biomarker selection on dynamical systems models of cellular processes, and it specifically selects new sensors at each time point, which aligns with the user\\'s request to perform time-varying selection from times 0 to 10.\" response_metadata={\\'role\\': \\'assistant\\', \\'content\\': \"SCRIPT: biomarkerSelectionTimeVarying\\\\nREASON: The user query specifically asks for time-varying biomarker selection, and this script is the only one that matches this description. It is designed to perform biomarker selection on dynamical systems models of cellular processes, and it specifically selects new sensors at each time point, which aligns with the user\\'s request to perform time-varying selection from times 0 to 10.\", \\'token_usage\\': {\\'prompt_tokens\\': 550, \\'total_tokens\\': 641, \\'completion_tokens\\': 91}, \\'model_name\\': \\'meta/llama3-70b-instruct\\'} id=\\'run-d3c2dc33-079b-4ec9-ae28-e95081ec01c4-0\\' role=\\'assistant\\'', 'parsedOutput': {'scriptName': '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', 'scriptType': 'python', 'scriptPath': '/home/jpic/BRAD-Tools/'}, 'purpose': 'Select which code to run'}, {'llm': \"model='meta/llama3-70b-instruct'\", 'prompt': 'input_variables=[\\'history\\', \\'input\\'] template=\\'Current conversation:\\\\n{history}\\\\n\\\\n**PYTHON SCRIPT**\\\\nYou must run this python script:\\\\n/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py\\\\n\\\\n**PYTHON SCRIPT DOCUMENTATION**:\\\\nThis is the doc string of this python script:\\\\n\"\"\"\\\\nThis script performs time varying sensor or biomarker selection on dynamical systems models of cellular processes. It selects new sensors at each time point.\\\\n\\\\nArguments (four arguments):\\\\n    1. output directory: chatstatus[\\\\\\'output-directory\\\\\\']\\\\n    2. output file: <name of output file>\\\\n    3. input file: the file created previously with the LTI or LTV model\\\\n    4. start time: (int, default 0) the first time to select sensors at (inclusive)\\\\n    5. end time: (int, default 10) the last time to select sensors at (not inclusive)\\\\n    6. verbose: (optional, True or False) if sensor selection should be run in verbose mode so that more output will be displayed to the user.\\\\n\\\\nUsage:\\\\nCommand Line:\\\\n```\\\\npython <path/to/script/>biomarkerSelectionTimeVarying.py <output path> <output file> <input file> <start time> <end time> <verbose>\\\\n```\\\\n\\\\nBRAD Line:\\\\n```\\\\nsubprocess.call([sys.executable, \\\\\\'<path/to/script/>/biomarkerSelectionTimeVarying.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\'], <output file>, <input file>, <start time>, <end time>, <verbose>], capture_output=True, text=True)\\\\n```\\\\n\\\\n**OUTPUT FILE NAME INSTRUCTIONS**\\\\n1. Output path should be chatstatus[\\\\\\'output-directory\\\\\\']\\\\n2. Output file name should be `S3-<descriptive name>.csv`\\\\n\"\"\"\\\\n\\\\n\\\\n**CALL PYTHON SCRIPTS FROM PYTHON**:\\\\nUse the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\\\\n\\\\nTo call a Python script `example_script.py` which has no arguments:\\\\n\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\']], capture_output=True, text=True)\\\\n)\\\\n```\\\\n\\\\nTo call a Python script `example_script.py` with one argument:\\\\n\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\'], \\\\\\'arg1\\\\\\'], capture_output=True, text=True)\\\\n)\\\\n```\\\\n\\\\nTo call a Python script `example_script.py` with two arguments:\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\'], \\\\\\'arg1\\\\\\', \\\\\\'arg2\\\\\\'], capture_output=True, text=True)\\\\n)\\\\n```\\\\n\\\\nNote that chatstatus[\\\\\\'output-directory\\\\\\'] is ALWAYS passed as the first argument.\\\\n\\\\nThe following files were previously created by BRAD and could be used as input to a function if necessary:\\\\nREPORT.tex\\\\nlog.json\\\\n.ipynb_checkpoints\\\\nREPORT.log\\\\nS1-2015.pkl\\\\nENRICHR-GO_Biological_Process_2021.csv\\\\nS0-2015.h5ad\\\\ntime_1_S3-biomarker-selection-time-varying.csv\\\\nS2-2015.pkl\\\\ntime_0_S3-biomarker-selection-time-varying.csv\\\\ntime_2_S3-biomarker-selection-time-varying.csv\\\\nS3-biomarker-selection.csv\\\\nimages\\\\n\\\\nQuery:{input}\\\\n\\\\n**INSTRUCTIONS**\\\\n1. Given the user query and the documentation, identify each of the arguments found in the user\\\\\\'s query that should be passed to the Python script.\\\\n2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\\\\n3. The last line of your response should say \"Execute: <Python code to execute>\"\\\\n4. Format the response/output as:\\\\n    Arguments: \\\\n    Python Code Explanation: <2 sentences maximum>\\\\n    Execute: <your code here>\\\\n\\\\n**IMPORTANT**\\\\nThe code to execute from your response must be formatted as:\\\\n    Execute: subprocess.call([sys.executable, \\\\\\'<path to python script>\\\\\\', \\\\\\'<argument 1>\\\\\\', \\\\\\'<argument 2>\\\\\\', ..., \\\\\\'<argument n>\\\\\\'], capture_output=True, text=True))\\\\nThis output should be exactly one line and no longer. Stop the response after this line.\\\\n\\'', 'input': 'Perform time varying biomarker selection on this data from times 0 to 10. set verbose to true.', 'output': \"Arguments: \\noutput directory: chatstatus['output-directory']\\noutput file: S3-biomarkers-time-varying.csv\\ninput file: S3-biomarker-selection.csv\\nstart time: 0\\nend time: 10\\nverbose: True\\n\\nPython Code Explanation: \\nUsing the subprocess module, we are executing the biomarkerSelectionTimeVarying.py script with the required arguments. \\nThe script performs time varying sensor or biomarker selection on dynamical systems models of cellular processes.\\n\\nExecute: subprocess.call([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-biomarkers-time-varying.csv', 'S3-biomarker-selection.csv', '0', '10', 'True'], capture_output=True, text=True)\", 'parsedOutput': {'code': \"subprocess.run([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-biomarkers-time-varying.csv', 'S3-biomarker-selection.csv', '0', '10', 'True'], capture_output=True, text=True)\"}, 'purpose': 'Format function call'}]}\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  /force CODE Perform time varying biomarker selection on this data from times 5-7. set verbose to true.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-07-03 10:17:31,156 - INFO - CODE\n",
      "2024-07-03 10:17:31,157 - INFO - CODER\n",
      "2024-07-03 10:17:31,161 - INFO - input_variables=['user_query'] template=\"You must select which code to run to help a user.\\n\\n**Available Scripts**\\nScript Name: geneformerCoordinatesTimeSeries, \\t Description: This script organizes datasets into a format to work with time series gene expression using individual geneformer embeddings as the state space. From a .h5ad file, a .pkl file is constructed here that will contain the state type being geneformer, the state names being geneformer coordinates, and the time series data organized as states (geneformer coordinates) by timepoint (hours) by replicates (experiment replicates).\\nScript Name: geneCoordinatesTimeSeries, \\t Description: This script organizes datasets into a format to work with time series gene expression using individual gene expresession as the state space. From a .h5ad file, a .pkl file is constructed here that will contain the state type being gene expression, the state names being gene names, and the time series data organized as states (genes) by timepoint (hours) by replicates (experiment replicates).\\nScript Name: linearModels, \\t Description: This file builds linear and linear time variant (LTI and LTV) models of gene expressions.\\nScript Name: geneformerEmbeddings, \\t Description: Get gene former embeddings of your data\\nScript Name: biomarkerSelectionTimeVarying, \\t Description: This script performs time varying sensor or biomarker selection on dynamical systems models of cellular processes. It selects new sensors at each time point.\\nScript Name: buildLinearModel, \\t Description: This script builds linear models of time series data using Dynamic Mode Decomposition (DMD).\\nScript Name: test, \\t Description: This script was made for Joshua to test and learn python\\nScript Name: datasetSelector, \\t Description: This script is a file chooser to load a gene expression dataset from the lab's turbo partition to give access to the data to BRAD.\\nScript Name: biomarkerSelection, \\t Description: This script performs sensor selection on models of dynamical systems. This is applied to models of cellular processes for biomarker selection.\\nScript Name: buildLTVModel, \\t Description: This script builds linear time varying models of time series data using Data Guided Control model.\\n\\n\\n**User Query**\\n{user_query}\\n\\n**Task**\\nBased on the user's query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\\n\\n**Response Template**\\nSCRIPT: <selected script>\\nREASON: <reasoning why the script fits the user prompt>\\n\"\n",
      "2024-07-03 10:17:31,161 - INFO - FIRST LLM CALL\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "BRAD >> 17: \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-07-03 10:17:32,889 - INFO - SCRIPT: biomarkerSelectionTimeVarying\n",
      "REASON: The user query specifically asks for time varying biomarker selection, which is the exact functionality described in the biomarkerSelectionTimeVarying script. Additionally, the script is designed to select new sensors at each time point, which fits with the user's request to perform the selection on a specific time range (times 5-7).\n",
      "2024-07-03 10:17:32,889 - INFO - Warning! the type doesn't exist\n",
      "2024-07-03 10:17:32,890 - INFO - scriptPath=/home/jpic/BRAD-Tools/\n",
      "2024-07-03 10:17:32,890 - INFO - pyPath=/home/jpic/BRAD-Tools/\n",
      "2024-07-03 10:17:32,891 - INFO - ALL SCRIPTS FOUND. BUILDING TEMPLATE\n",
      "2024-07-03 10:17:32,892 - INFO - input_variables=['history', 'input'] template='Current conversation:\\n{history}\\n\\n**PYTHON SCRIPT**\\nYou must run this python script:\\n/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py\\n\\n**PYTHON SCRIPT DOCUMENTATION**:\\nThis is the doc string of this python script:\\n\"\"\"\\nThis script performs time varying sensor or biomarker selection on dynamical systems models of cellular processes. It selects new sensors at each time point.\\n\\nArguments (four arguments):\\n    1. output directory: chatstatus[\\'output-directory\\']\\n    2. output file: <name of output file>\\n    3. input file: the file created previously with the LTI or LTV model\\n    4. start time: (int, default 0) the first time to select sensors at (inclusive)\\n    5. end time: (int, default 10) the last time to select sensors at (not inclusive)\\n    6. verbose: (optional, True or False) if sensor selection should be run in verbose mode so that more output will be displayed to the user.\\n\\nUsage:\\nCommand Line:\\n```\\npython <path/to/script/>biomarkerSelectionTimeVarying.py <output path> <output file> <input file> <start time> <end time> <verbose>\\n```\\n\\nBRAD Line:\\n```\\nsubprocess.call([sys.executable, \\'<path/to/script/>/biomarkerSelectionTimeVarying.py\\', chatstatus[\\'output-directory\\'], <output file>, <input file>, <start time>, <end time>, <verbose>], capture_output=True, text=True)\\n```\\n\\n**OUTPUT FILE NAME INSTRUCTIONS**\\n1. Output path should be chatstatus[\\'output-directory\\']\\n2. Output file name should be `S3-<descriptive name>.csv`\\n\"\"\"\\n\\n\\n**CALL PYTHON SCRIPTS FROM PYTHON**:\\nUse the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\\n\\nTo call a Python script `example_script.py` which has no arguments:\\n\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\']], capture_output=True, text=True)\\n)\\n```\\n\\nTo call a Python script `example_script.py` with one argument:\\n\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\'], \\'arg1\\'], capture_output=True, text=True)\\n)\\n```\\n\\nTo call a Python script `example_script.py` with two arguments:\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\'], \\'arg1\\', \\'arg2\\'], capture_output=True, text=True)\\n)\\n```\\n\\nNote that chatstatus[\\'output-directory\\'] is ALWAYS passed as the first argument.\\n\\nThe following files were previously created by BRAD and could be used as input to a function if necessary:\\nREPORT.tex\\nlog.json\\n.ipynb_checkpoints\\nREPORT.log\\nS1-2015.pkl\\nENRICHR-GO_Biological_Process_2021.csv\\nS0-2015.h5ad\\ntime_1_S3-biomarker-selection-time-varying.csv\\nS2-2015.pkl\\ntime_0_S3-biomarker-selection-time-varying.csv\\ntime_2_S3-biomarker-selection-time-varying.csv\\nimages\\n\\nQuery:{input}\\n\\n**INSTRUCTIONS**\\n1. Given the user query and the documentation, identify each of the arguments found in the user\\'s query that should be passed to the Python script.\\n2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\\n3. The last line of your response should say \"Execute: <Python code to execute>\"\\n4. Format the response/output as:\\n    Arguments: \\n    Python Code Explanation: <2 sentences maximum>\\n    Execute: <your code here>\\n\\n**IMPORTANT**\\nThe code to execute from your response must be formatted as:\\n    Execute: subprocess.call([sys.executable, \\'<path to python script>\\', \\'<argument 1>\\', \\'<argument 2>\\', ..., \\'<argument n>\\'], capture_output=True, text=True))\\nThis output should be exactly one line and no longer. Stop the response after this line.\\n'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation:\n",
      "\n",
      "\n",
      "**PYTHON SCRIPT**\n",
      "You must run this python script:\n",
      "/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py\n",
      "\n",
      "**PYTHON SCRIPT DOCUMENTATION**:\n",
      "This is the doc string of this python script:\n",
      "\"\"\"\n",
      "This script performs time varying sensor or biomarker selection on dynamical systems models of cellular processes. It selects new sensors at each time point.\n",
      "\n",
      "Arguments (four arguments):\n",
      "    1. output directory: chatstatus['output-directory']\n",
      "    2. output file: <name of output file>\n",
      "    3. input file: the file created previously with the LTI or LTV model\n",
      "    4. start time: (int, default 0) the first time to select sensors at (inclusive)\n",
      "    5. end time: (int, default 10) the last time to select sensors at (not inclusive)\n",
      "    6. verbose: (optional, True or False) if sensor selection should be run in verbose mode so that more output will be displayed to the user.\n",
      "\n",
      "Usage:\n",
      "Command Line:\n",
      "```\n",
      "python <path/to/script/>biomarkerSelectionTimeVarying.py <output path> <output file> <input file> <start time> <end time> <verbose>\n",
      "```\n",
      "\n",
      "BRAD Line:\n",
      "```\n",
      "subprocess.call([sys.executable, '<path/to/script/>/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], <output file>, <input file>, <start time>, <end time>, <verbose>], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "**OUTPUT FILE NAME INSTRUCTIONS**\n",
      "1. Output path should be chatstatus['output-directory']\n",
      "2. Output file name should be `S3-<descriptive name>.csv`\n",
      "\"\"\"\n",
      "\n",
      "\n",
      "**CALL PYTHON SCRIPTS FROM PYTHON**:\n",
      "Use the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\n",
      "\n",
      "To call a Python script `example_script.py` which has no arguments:\n",
      "\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory']], capture_output=True, text=True)\n",
      ")\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with one argument:\n",
      "\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory'], 'arg1'], capture_output=True, text=True)\n",
      ")\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with two arguments:\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory'], 'arg1', 'arg2'], capture_output=True, text=True)\n",
      ")\n",
      "```\n",
      "\n",
      "Note that chatstatus['output-directory'] is ALWAYS passed as the first argument.\n",
      "\n",
      "The following files were previously created by BRAD and could be used as input to a function if necessary:\n",
      "REPORT.tex\n",
      "log.json\n",
      ".ipynb_checkpoints\n",
      "REPORT.log\n",
      "S1-2015.pkl\n",
      "ENRICHR-GO_Biological_Process_2021.csv\n",
      "S0-2015.h5ad\n",
      "time_1_S3-biomarker-selection-time-varying.csv\n",
      "S2-2015.pkl\n",
      "time_0_S3-biomarker-selection-time-varying.csv\n",
      "time_2_S3-biomarker-selection-time-varying.csv\n",
      "images\n",
      "\n",
      "Query:Perform time varying biomarker selection on this data from times 5-7. set verbose to true.\n",
      "\n",
      "**INSTRUCTIONS**\n",
      "1. Given the user query and the documentation, identify each of the arguments found in the user's query that should be passed to the Python script.\n",
      "2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\n",
      "3. The last line of your response should say \"Execute: <Python code to execute>\"\n",
      "4. Format the response/output as:\n",
      "    Arguments: \n",
      "    Python Code Explanation: <2 sentences maximum>\n",
      "    Execute: <your code here>\n",
      "\n",
      "**IMPORTANT**\n",
      "The code to execute from your response must be formatted as:\n",
      "    Execute: subprocess.call([sys.executable, '<path to python script>', '<argument 1>', '<argument 2>', ..., '<argument n>'], capture_output=True, text=True))\n",
      "This output should be exactly one line and no longer. Stop the response after this line.\n",
      "\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-07-03 10:17:36,212 - INFO - LLM OUTPUT PARSER\n",
      "2024-07-03 10:17:36,213 - INFO - Arguments: \n",
      "output directory: chatstatus['output-directory']\n",
      "output file: S3-biomarker-selection-time-varying.csv\n",
      "input file: time_2_S3-biomarker-selection-time-varying.csv\n",
      "start time: 5\n",
      "end time: 7\n",
      "verbose: True\n",
      "\n",
      "Python Code Explanation: This code will execute the biomarkerSelectionTimeVarying.py script with the given arguments, performing time varying biomarker selection on the data from times 5-7 with verbose mode enabled.\n",
      "\n",
      "Execute: subprocess.call([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-biomarker-selection-time-varying.csv', 'time_2_S3-biomarker-selection-time-varying.csv', '5', '7', 'True'], capture_output=True, text=True)\n",
      "2024-07-03 10:17:36,213 - INFO - [\"Arguments: \\noutput directory: chatstatus['output-directory']\\noutput file: S3-biomarker-selection-time-varying.csv\\ninput file: time_2_S3-biomarker-selection-time-varying.csv\\nstart time: 5\\nend time: 7\\nverbose: True\\n\\nPython Code Explanation: This code will execute the biomarkerSelectionTimeVarying.py script with the given arguments, performing time varying biomarker selection on the data from times 5-7 with verbose mode enabled.\\n\\n\", \" subprocess.call([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-biomarker-selection-time-varying.csv', 'time_2_S3-biomarker-selection-time-varying.csv', '5', '7', 'True'], capture_output=True, text=True)\"]\n",
      "2024-07-03 10:17:36,214 - INFO - subprocess.call([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-biomarker-selection-time-varying.csv', 'time_2_S3-biomarker-selection-time-varying.csv', '5', '7', 'True'], capture_output=True, text=True)\n",
      "2024-07-03 10:17:36,214 - INFO - subprocess.run([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-biomarker-selection-time-varying.csv', 'time_2_S3-biomarker-selection-time-varying.csv', '5', '7', 'True'], capture_output=True, text=True)\n",
      "2024-07-03 10:17:36,215 - INFO - Replacing: time_2_S3-biomarker-selection-time-varying.csv with /home/jpic/BRAD/2024-07-03_08-51-24/time_2_S3-biomarker-selection-time-varying.csv\n",
      "2024-07-03 10:17:36,216 - INFO - New String: subprocess.run([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-biomarker-selection-time-varying.csv', '/home/jpic/BRAD/2024-07-03_08-51-24/time_2_S3-biomarker-selection-time-varying.csv', '5', '7', 'True'], capture_output=True, text=True)\n",
      "2024-07-03 10:17:36,216 - INFO - Executing Code!!\n",
      "2024-07-03 10:17:36,217 - INFO - EVAL\n",
      "2024-07-03 10:17:36,217 - INFO - subprocess.run([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-biomarker-selection-time-varying.csv', '/home/jpic/BRAD/2024-07-03_08-51-24/time_2_S3-biomarker-selection-time-varying.csv', '5', '7', 'True'], capture_output=True, text=True)\n",
      "2024-07-03 10:17:36,217 - INFO - END EVAL CHECK\n",
      "2024-07-03 10:17:36,218 - INFO - Finalized PYTHON Call:\n",
      "2024-07-03 10:17:36,218 - INFO -    \n",
      "2024-07-03 10:17:36,218 - INFO - response = subprocess.run([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-biomarker-selection-time-varying.csv', '/home/jpic/BRAD/2024-07-03_08-51-24/time_2_S3-biomarker-selection-time-varying.csv', '5', '7', 'True'], capture_output=True, text=True)\n",
      "2024-07-03 10:17:36,219 - INFO -    \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-07-03 10:17:37,014 - INFO - Debug: PYTHON code executed successfully.\n",
      "2024-07-03 10:17:37,015 - INFO - Code Execution Output:\n",
      "2024-07-03 10:17:37,016 - INFO - CompletedProcess(args=['/sw/pkgs/arc/python3.11-anaconda/2024.02-1/bin/python', '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', '/home/jpic/BRAD/2024-07-03_08-51-24/', 'S3-biomarker-selection-time-varying.csv', '/home/jpic/BRAD/2024-07-03_08-51-24/time_2_S3-biomarker-selection-time-varying.csv', '5', '7', 'True'], returncode=1, stdout='Time varying biomarker selection script has been called!\\n*******************************\\n Time Varying Sensor Selection \\n*******************************\\nOutput Path: /home/jpic/BRAD/2024-07-03_08-51-24/\\nOutput File: S3-biomarker-selection-time-varying.csv\\nInput File : /home/jpic/BRAD/2024-07-03_08-51-24/time_2_S3-biomarker-selection-time-varying.csv\\nstartTime  : 5\\nendTime    : 7\\nverbose    : True\\n', stderr='Traceback (most recent call last):\\n  File \"/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py\", line 107, in <module>\\n    main()\\n  File \"/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py\", line 86, in main\\n    MODEL = pickle.load(file)\\n            ^^^^^^^^^^^^^^^^^\\n_pickle.UnpicklingError: invalid load key, \\',\\'.\\n')\n",
      "2024-07-03 10:17:37,017 - INFO - Debug: PYTHON code output saved to output.\n",
      "2024-07-03 10:17:37,017 - INFO - \n",
      "\n",
      "\n",
      "route\n",
      "\n",
      "\n",
      "\n",
      "2024-07-03 10:17:37,018 - INFO - CODE\n",
      "2024-07-03 10:17:37,019 - INFO - \n",
      "\n",
      "PLANNER:\n",
      "\n",
      "{'module': 'CODE', 'steps': [{'llm': \"model='meta/llama3-70b-instruct'\", 'prompt': 'input_variables=[\\'user_query\\'] template=\"You must select which code to run to help a user.\\\\n\\\\n**Available Scripts**\\\\nScript Name: geneformerCoordinatesTimeSeries, \\\\t Description: This script organizes datasets into a format to work with time series gene expression using individual geneformer embeddings as the state space. From a .h5ad file, a .pkl file is constructed here that will contain the state type being geneformer, the state names being geneformer coordinates, and the time series data organized as states (geneformer coordinates) by timepoint (hours) by replicates (experiment replicates).\\\\nScript Name: geneCoordinatesTimeSeries, \\\\t Description: This script organizes datasets into a format to work with time series gene expression using individual gene expresession as the state space. From a .h5ad file, a .pkl file is constructed here that will contain the state type being gene expression, the state names being gene names, and the time series data organized as states (genes) by timepoint (hours) by replicates (experiment replicates).\\\\nScript Name: linearModels, \\\\t Description: This file builds linear and linear time variant (LTI and LTV) models of gene expressions.\\\\nScript Name: geneformerEmbeddings, \\\\t Description: Get gene former embeddings of your data\\\\nScript Name: biomarkerSelectionTimeVarying, \\\\t Description: This script performs time varying sensor or biomarker selection on dynamical systems models of cellular processes. It selects new sensors at each time point.\\\\nScript Name: buildLinearModel, \\\\t Description: This script builds linear models of time series data using Dynamic Mode Decomposition (DMD).\\\\nScript Name: test, \\\\t Description: This script was made for Joshua to test and learn python\\\\nScript Name: datasetSelector, \\\\t Description: This script is a file chooser to load a gene expression dataset from the lab\\'s turbo partition to give access to the data to BRAD.\\\\nScript Name: biomarkerSelection, \\\\t Description: This script performs sensor selection on models of dynamical systems. This is applied to models of cellular processes for biomarker selection.\\\\nScript Name: buildLTVModel, \\\\t Description: This script builds linear time varying models of time series data using Data Guided Control model.\\\\n\\\\n\\\\n**User Query**\\\\n{user_query}\\\\n\\\\n**Task**\\\\nBased on the user\\'s query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\\\\n\\\\n**Response Template**\\\\nSCRIPT: <selected script>\\\\nREASON: <reasoning why the script fits the user prompt>\\\\n\"', 'input': 'Perform time varying biomarker selection on this data from times 5-7. set verbose to true.', 'output': 'content=\"SCRIPT: biomarkerSelectionTimeVarying\\\\nREASON: The user query specifically asks for time varying biomarker selection, which is the exact functionality described in the biomarkerSelectionTimeVarying script. Additionally, the script is designed to select new sensors at each time point, which fits with the user\\'s request to perform the selection on a specific time range (times 5-7).\" response_metadata={\\'role\\': \\'assistant\\', \\'content\\': \"SCRIPT: biomarkerSelectionTimeVarying\\\\nREASON: The user query specifically asks for time varying biomarker selection, which is the exact functionality described in the biomarkerSelectionTimeVarying script. Additionally, the script is designed to select new sensors at each time point, which fits with the user\\'s request to perform the selection on a specific time range (times 5-7).\", \\'token_usage\\': {\\'prompt_tokens\\': 549, \\'total_tokens\\': 630, \\'completion_tokens\\': 81}, \\'model_name\\': \\'meta/llama3-70b-instruct\\'} id=\\'run-8db082a1-81ce-4bf7-88ea-23ea5288f58f-0\\' role=\\'assistant\\'', 'parsedOutput': {'scriptName': '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', 'scriptType': 'python', 'scriptPath': '/home/jpic/BRAD-Tools/'}, 'purpose': 'Select which code to run'}, {'llm': \"model='meta/llama3-70b-instruct'\", 'prompt': 'input_variables=[\\'history\\', \\'input\\'] template=\\'Current conversation:\\\\n{history}\\\\n\\\\n**PYTHON SCRIPT**\\\\nYou must run this python script:\\\\n/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py\\\\n\\\\n**PYTHON SCRIPT DOCUMENTATION**:\\\\nThis is the doc string of this python script:\\\\n\"\"\"\\\\nThis script performs time varying sensor or biomarker selection on dynamical systems models of cellular processes. It selects new sensors at each time point.\\\\n\\\\nArguments (four arguments):\\\\n    1. output directory: chatstatus[\\\\\\'output-directory\\\\\\']\\\\n    2. output file: <name of output file>\\\\n    3. input file: the file created previously with the LTI or LTV model\\\\n    4. start time: (int, default 0) the first time to select sensors at (inclusive)\\\\n    5. end time: (int, default 10) the last time to select sensors at (not inclusive)\\\\n    6. verbose: (optional, True or False) if sensor selection should be run in verbose mode so that more output will be displayed to the user.\\\\n\\\\nUsage:\\\\nCommand Line:\\\\n```\\\\npython <path/to/script/>biomarkerSelectionTimeVarying.py <output path> <output file> <input file> <start time> <end time> <verbose>\\\\n```\\\\n\\\\nBRAD Line:\\\\n```\\\\nsubprocess.call([sys.executable, \\\\\\'<path/to/script/>/biomarkerSelectionTimeVarying.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\'], <output file>, <input file>, <start time>, <end time>, <verbose>], capture_output=True, text=True)\\\\n```\\\\n\\\\n**OUTPUT FILE NAME INSTRUCTIONS**\\\\n1. Output path should be chatstatus[\\\\\\'output-directory\\\\\\']\\\\n2. Output file name should be `S3-<descriptive name>.csv`\\\\n\"\"\"\\\\n\\\\n\\\\n**CALL PYTHON SCRIPTS FROM PYTHON**:\\\\nUse the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\\\\n\\\\nTo call a Python script `example_script.py` which has no arguments:\\\\n\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\']], capture_output=True, text=True)\\\\n)\\\\n```\\\\n\\\\nTo call a Python script `example_script.py` with one argument:\\\\n\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\'], \\\\\\'arg1\\\\\\'], capture_output=True, text=True)\\\\n)\\\\n```\\\\n\\\\nTo call a Python script `example_script.py` with two arguments:\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\'], \\\\\\'arg1\\\\\\', \\\\\\'arg2\\\\\\'], capture_output=True, text=True)\\\\n)\\\\n```\\\\n\\\\nNote that chatstatus[\\\\\\'output-directory\\\\\\'] is ALWAYS passed as the first argument.\\\\n\\\\nThe following files were previously created by BRAD and could be used as input to a function if necessary:\\\\nREPORT.tex\\\\nlog.json\\\\n.ipynb_checkpoints\\\\nREPORT.log\\\\nS1-2015.pkl\\\\nENRICHR-GO_Biological_Process_2021.csv\\\\nS0-2015.h5ad\\\\ntime_1_S3-biomarker-selection-time-varying.csv\\\\nS2-2015.pkl\\\\ntime_0_S3-biomarker-selection-time-varying.csv\\\\ntime_2_S3-biomarker-selection-time-varying.csv\\\\nimages\\\\n\\\\nQuery:{input}\\\\n\\\\n**INSTRUCTIONS**\\\\n1. Given the user query and the documentation, identify each of the arguments found in the user\\\\\\'s query that should be passed to the Python script.\\\\n2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\\\\n3. The last line of your response should say \"Execute: <Python code to execute>\"\\\\n4. Format the response/output as:\\\\n    Arguments: \\\\n    Python Code Explanation: <2 sentences maximum>\\\\n    Execute: <your code here>\\\\n\\\\n**IMPORTANT**\\\\nThe code to execute from your response must be formatted as:\\\\n    Execute: subprocess.call([sys.executable, \\\\\\'<path to python script>\\\\\\', \\\\\\'<argument 1>\\\\\\', \\\\\\'<argument 2>\\\\\\', ..., \\\\\\'<argument n>\\\\\\'], capture_output=True, text=True))\\\\nThis output should be exactly one line and no longer. Stop the response after this line.\\\\n\\'', 'input': 'Perform time varying biomarker selection on this data from times 5-7. set verbose to true.', 'output': \"Arguments: \\noutput directory: chatstatus['output-directory']\\noutput file: S3-biomarker-selection-time-varying.csv\\ninput file: time_2_S3-biomarker-selection-time-varying.csv\\nstart time: 5\\nend time: 7\\nverbose: True\\n\\nPython Code Explanation: This code will execute the biomarkerSelectionTimeVarying.py script with the given arguments, performing time varying biomarker selection on the data from times 5-7 with verbose mode enabled.\\n\\nExecute: subprocess.call([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-biomarker-selection-time-varying.csv', 'time_2_S3-biomarker-selection-time-varying.csv', '5', '7', 'True'], capture_output=True, text=True)\", 'parsedOutput': {'code': \"subprocess.run([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-biomarker-selection-time-varying.csv', 'time_2_S3-biomarker-selection-time-varying.csv', '5', '7', 'True'], capture_output=True, text=True)\"}, 'purpose': 'Format function call'}]}\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  /force CODE Perform time varying biomarker selection on this data from times 5-7. set verbose to true.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-07-03 10:18:10,083 - INFO - CODE\n",
      "2024-07-03 10:18:10,084 - INFO - CODER\n",
      "2024-07-03 10:18:10,087 - INFO - input_variables=['user_query'] template=\"You must select which code to run to help a user.\\n\\n**Available Scripts**\\nScript Name: geneformerCoordinatesTimeSeries, \\t Description: This script organizes datasets into a format to work with time series gene expression using individual geneformer embeddings as the state space. From a .h5ad file, a .pkl file is constructed here that will contain the state type being geneformer, the state names being geneformer coordinates, and the time series data organized as states (geneformer coordinates) by timepoint (hours) by replicates (experiment replicates).\\nScript Name: geneCoordinatesTimeSeries, \\t Description: This script organizes datasets into a format to work with time series gene expression using individual gene expresession as the state space. From a .h5ad file, a .pkl file is constructed here that will contain the state type being gene expression, the state names being gene names, and the time series data organized as states (genes) by timepoint (hours) by replicates (experiment replicates).\\nScript Name: linearModels, \\t Description: This file builds linear and linear time variant (LTI and LTV) models of gene expressions.\\nScript Name: geneformerEmbeddings, \\t Description: Get gene former embeddings of your data\\nScript Name: biomarkerSelectionTimeVarying, \\t Description: This script performs time varying sensor or biomarker selection on dynamical systems models of cellular processes. It selects new sensors at each time point.\\nScript Name: buildLinearModel, \\t Description: This script builds linear models of time series data using Dynamic Mode Decomposition (DMD).\\nScript Name: test, \\t Description: This script was made for Joshua to test and learn python\\nScript Name: datasetSelector, \\t Description: This script is a file chooser to load a gene expression dataset from the lab's turbo partition to give access to the data to BRAD.\\nScript Name: biomarkerSelection, \\t Description: This script performs sensor selection on models of dynamical systems. This is applied to models of cellular processes for biomarker selection.\\nScript Name: buildLTVModel, \\t Description: This script builds linear time varying models of time series data using Data Guided Control model.\\n\\n\\n**User Query**\\n{user_query}\\n\\n**Task**\\nBased on the user's query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\\n\\n**Response Template**\\nSCRIPT: <selected script>\\nREASON: <reasoning why the script fits the user prompt>\\n\"\n",
      "2024-07-03 10:18:10,087 - INFO - FIRST LLM CALL\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "BRAD >> 18: \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-07-03 10:18:11,923 - INFO - SCRIPT: biomarkerSelectionTimeVarying\n",
      "REASON: The user query specifically asks for time varying biomarker selection, which is the exact functionality described in the biomarkerSelectionTimeVarying script. Additionally, the script's description mentions selecting new sensors at each time point, which matches the user's request to perform selection from times 5-7. The verbose option can also be set to true as per the user's request.\n",
      "2024-07-03 10:18:11,923 - INFO - Warning! the type doesn't exist\n",
      "2024-07-03 10:18:11,925 - INFO - scriptPath=/home/jpic/BRAD-Tools/\n",
      "2024-07-03 10:18:11,926 - INFO - pyPath=/home/jpic/BRAD-Tools/\n",
      "2024-07-03 10:18:11,926 - INFO - ALL SCRIPTS FOUND. BUILDING TEMPLATE\n",
      "2024-07-03 10:18:11,927 - INFO - input_variables=['history', 'input'] template='Current conversation:\\n{history}\\n\\n**PYTHON SCRIPT**\\nYou must run this python script:\\n/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py\\n\\n**PYTHON SCRIPT DOCUMENTATION**:\\nThis is the doc string of this python script:\\n\"\"\"\\nThis script performs time varying sensor or biomarker selection on dynamical systems models of cellular processes. It selects new sensors at each time point.\\n\\nArguments (four arguments):\\n    1. output directory: chatstatus[\\'output-directory\\']\\n    2. output file: <name of output file>\\n    3. input file: the file created previously with the LTI or LTV model\\n    4. start time: (int, default 0) the first time to select sensors at (inclusive)\\n    5. end time: (int, default 10) the last time to select sensors at (not inclusive)\\n    6. verbose: (optional, True or False) if sensor selection should be run in verbose mode so that more output will be displayed to the user.\\n\\nUsage:\\nCommand Line:\\n```\\npython <path/to/script/>biomarkerSelectionTimeVarying.py <output path> <output file> <input file> <start time> <end time> <verbose>\\n```\\n\\nBRAD Line:\\n```\\nsubprocess.call([sys.executable, \\'<path/to/script/>/biomarkerSelectionTimeVarying.py\\', chatstatus[\\'output-directory\\'], <output file>, <input file>, <start time>, <end time>, <verbose>], capture_output=True, text=True)\\n```\\n\\n**OUTPUT FILE NAME INSTRUCTIONS**\\n1. Output path should be chatstatus[\\'output-directory\\']\\n2. Output file name should be `S3-<descriptive name>.csv`\\n\"\"\"\\n\\n\\n**CALL PYTHON SCRIPTS FROM PYTHON**:\\nUse the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\\n\\nTo call a Python script `example_script.py` which has no arguments:\\n\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\']], capture_output=True, text=True)\\n)\\n```\\n\\nTo call a Python script `example_script.py` with one argument:\\n\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\'], \\'arg1\\'], capture_output=True, text=True)\\n)\\n```\\n\\nTo call a Python script `example_script.py` with two arguments:\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\'], \\'arg1\\', \\'arg2\\'], capture_output=True, text=True)\\n)\\n```\\n\\nNote that chatstatus[\\'output-directory\\'] is ALWAYS passed as the first argument.\\n\\nThe following files were previously created by BRAD and could be used as input to a function if necessary:\\nREPORT.tex\\nlog.json\\n.ipynb_checkpoints\\nREPORT.log\\nS1-2015.pkl\\nENRICHR-GO_Biological_Process_2021.csv\\nS0-2015.h5ad\\nS2-2015.pkl\\nimages\\n\\nQuery:{input}\\n\\n**INSTRUCTIONS**\\n1. Given the user query and the documentation, identify each of the arguments found in the user\\'s query that should be passed to the Python script.\\n2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\\n3. The last line of your response should say \"Execute: <Python code to execute>\"\\n4. Format the response/output as:\\n    Arguments: \\n    Python Code Explanation: <2 sentences maximum>\\n    Execute: <your code here>\\n\\n**IMPORTANT**\\nThe code to execute from your response must be formatted as:\\n    Execute: subprocess.call([sys.executable, \\'<path to python script>\\', \\'<argument 1>\\', \\'<argument 2>\\', ..., \\'<argument n>\\'], capture_output=True, text=True))\\nThis output should be exactly one line and no longer. Stop the response after this line.\\n'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation:\n",
      "\n",
      "\n",
      "**PYTHON SCRIPT**\n",
      "You must run this python script:\n",
      "/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py\n",
      "\n",
      "**PYTHON SCRIPT DOCUMENTATION**:\n",
      "This is the doc string of this python script:\n",
      "\"\"\"\n",
      "This script performs time varying sensor or biomarker selection on dynamical systems models of cellular processes. It selects new sensors at each time point.\n",
      "\n",
      "Arguments (four arguments):\n",
      "    1. output directory: chatstatus['output-directory']\n",
      "    2. output file: <name of output file>\n",
      "    3. input file: the file created previously with the LTI or LTV model\n",
      "    4. start time: (int, default 0) the first time to select sensors at (inclusive)\n",
      "    5. end time: (int, default 10) the last time to select sensors at (not inclusive)\n",
      "    6. verbose: (optional, True or False) if sensor selection should be run in verbose mode so that more output will be displayed to the user.\n",
      "\n",
      "Usage:\n",
      "Command Line:\n",
      "```\n",
      "python <path/to/script/>biomarkerSelectionTimeVarying.py <output path> <output file> <input file> <start time> <end time> <verbose>\n",
      "```\n",
      "\n",
      "BRAD Line:\n",
      "```\n",
      "subprocess.call([sys.executable, '<path/to/script/>/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], <output file>, <input file>, <start time>, <end time>, <verbose>], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "**OUTPUT FILE NAME INSTRUCTIONS**\n",
      "1. Output path should be chatstatus['output-directory']\n",
      "2. Output file name should be `S3-<descriptive name>.csv`\n",
      "\"\"\"\n",
      "\n",
      "\n",
      "**CALL PYTHON SCRIPTS FROM PYTHON**:\n",
      "Use the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\n",
      "\n",
      "To call a Python script `example_script.py` which has no arguments:\n",
      "\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory']], capture_output=True, text=True)\n",
      ")\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with one argument:\n",
      "\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory'], 'arg1'], capture_output=True, text=True)\n",
      ")\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with two arguments:\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory'], 'arg1', 'arg2'], capture_output=True, text=True)\n",
      ")\n",
      "```\n",
      "\n",
      "Note that chatstatus['output-directory'] is ALWAYS passed as the first argument.\n",
      "\n",
      "The following files were previously created by BRAD and could be used as input to a function if necessary:\n",
      "REPORT.tex\n",
      "log.json\n",
      ".ipynb_checkpoints\n",
      "REPORT.log\n",
      "S1-2015.pkl\n",
      "ENRICHR-GO_Biological_Process_2021.csv\n",
      "S0-2015.h5ad\n",
      "S2-2015.pkl\n",
      "images\n",
      "\n",
      "Query:Perform time varying biomarker selection on this data from times 5-7. set verbose to true.\n",
      "\n",
      "**INSTRUCTIONS**\n",
      "1. Given the user query and the documentation, identify each of the arguments found in the user's query that should be passed to the Python script.\n",
      "2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\n",
      "3. The last line of your response should say \"Execute: <Python code to execute>\"\n",
      "4. Format the response/output as:\n",
      "    Arguments: \n",
      "    Python Code Explanation: <2 sentences maximum>\n",
      "    Execute: <your code here>\n",
      "\n",
      "**IMPORTANT**\n",
      "The code to execute from your response must be formatted as:\n",
      "    Execute: subprocess.call([sys.executable, '<path to python script>', '<argument 1>', '<argument 2>', ..., '<argument n>'], capture_output=True, text=True))\n",
      "This output should be exactly one line and no longer. Stop the response after this line.\n",
      "\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-07-03 10:18:14,812 - INFO - LLM OUTPUT PARSER\n",
      "2024-07-03 10:18:14,812 - INFO - Arguments: \n",
      "output directory: chatstatus['output-directory']\n",
      "output file: S3-DescriptiveName.csv\n",
      "input file: S2-2015.pkl (assuming it is the input file created previously with the LTI or LTV model)\n",
      "start time: 5\n",
      "end time: 7\n",
      "verbose: True\n",
      "\n",
      "Python Code Explanation: The script performs time varying sensor or biomarker selection on dynamical systems models of cellular processes. It selects new sensors at each time point.\n",
      "Execute: subprocess.call([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-DescriptiveName.csv', 'S2-2015.pkl', '5', '7', 'True'], capture_output=True, text=True)\n",
      "2024-07-03 10:18:14,813 - INFO - [\"Arguments: \\noutput directory: chatstatus['output-directory']\\noutput file: S3-DescriptiveName.csv\\ninput file: S2-2015.pkl (assuming it is the input file created previously with the LTI or LTV model)\\nstart time: 5\\nend time: 7\\nverbose: True\\n\\nPython Code Explanation: The script performs time varying sensor or biomarker selection on dynamical systems models of cellular processes. It selects new sensors at each time point.\\n\", \" subprocess.call([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-DescriptiveName.csv', 'S2-2015.pkl', '5', '7', 'True'], capture_output=True, text=True)\"]\n",
      "2024-07-03 10:18:14,813 - INFO - subprocess.call([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-DescriptiveName.csv', 'S2-2015.pkl', '5', '7', 'True'], capture_output=True, text=True)\n",
      "2024-07-03 10:18:14,814 - INFO - subprocess.run([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-DescriptiveName.csv', 'S2-2015.pkl', '5', '7', 'True'], capture_output=True, text=True)\n",
      "2024-07-03 10:18:14,814 - INFO - Replacing: S2-2015.pkl with /home/jpic/BRAD/2024-07-03_08-51-24/S2-2015.pkl\n",
      "2024-07-03 10:18:14,815 - INFO - New String: subprocess.run([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-DescriptiveName.csv', '/home/jpic/BRAD/2024-07-03_08-51-24/S2-2015.pkl', '5', '7', 'True'], capture_output=True, text=True)\n",
      "2024-07-03 10:18:14,815 - INFO - Executing Code!!\n",
      "2024-07-03 10:18:14,816 - INFO - EVAL\n",
      "2024-07-03 10:18:14,816 - INFO - subprocess.run([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-DescriptiveName.csv', '/home/jpic/BRAD/2024-07-03_08-51-24/S2-2015.pkl', '5', '7', 'True'], capture_output=True, text=True)\n",
      "2024-07-03 10:18:14,816 - INFO - END EVAL CHECK\n",
      "2024-07-03 10:18:14,817 - INFO - Finalized PYTHON Call:\n",
      "2024-07-03 10:18:14,817 - INFO -    \n",
      "2024-07-03 10:18:14,818 - INFO - response = subprocess.run([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-DescriptiveName.csv', '/home/jpic/BRAD/2024-07-03_08-51-24/S2-2015.pkl', '5', '7', 'True'], capture_output=True, text=True)\n",
      "2024-07-03 10:18:14,818 - INFO -    \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-07-03 10:18:33,396 - INFO - Debug: PYTHON code executed successfully.\n",
      "2024-07-03 10:18:33,397 - INFO - Code Execution Output:\n",
      "2024-07-03 10:18:33,398 - INFO - CompletedProcess(args=['/sw/pkgs/arc/python3.11-anaconda/2024.02-1/bin/python', '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', '/home/jpic/BRAD/2024-07-03_08-51-24/', 'S3-DescriptiveName.csv', '/home/jpic/BRAD/2024-07-03_08-51-24/S2-2015.pkl', '5', '7', 'True'], returncode=0, stdout='Time varying biomarker selection script has been called!\\n*******************************\\n Time Varying Sensor Selection \\n*******************************\\nOutput Path: /home/jpic/BRAD/2024-07-03_08-51-24/\\nOutput File: S3-DescriptiveName.csv\\nInput File : /home/jpic/BRAD/2024-07-03_08-51-24/S2-2015.pkl\\nstartTime  : 5\\nendTime    : 7\\nverbose    : True\\n..............................\\n         Model Loaded         \\nt: 5/2\\nt: 6/2\\n..............................\\n      Sensors Selected        \\nModel saved to /home/jpic/BRAD/2024-07-03_08-51-24/time_5_S3-DescriptiveName.csv\\nModel saved to /home/jpic/BRAD/2024-07-03_08-51-24/time_6_S3-DescriptiveName.csv\\n..............................\\n      Sensors To File         \\n', stderr='')\n",
      "2024-07-03 10:18:33,398 - INFO - Debug: PYTHON code output saved to output.\n",
      "2024-07-03 10:18:33,400 - INFO - \n",
      "\n",
      "\n",
      "route\n",
      "\n",
      "\n",
      "\n",
      "2024-07-03 10:18:33,400 - INFO - CODE\n",
      "2024-07-03 10:18:33,400 - INFO - \n",
      "\n",
      "PLANNER:\n",
      "\n",
      "{'module': 'CODE', 'steps': [{'llm': \"model='meta/llama3-70b-instruct'\", 'prompt': 'input_variables=[\\'user_query\\'] template=\"You must select which code to run to help a user.\\\\n\\\\n**Available Scripts**\\\\nScript Name: geneformerCoordinatesTimeSeries, \\\\t Description: This script organizes datasets into a format to work with time series gene expression using individual geneformer embeddings as the state space. From a .h5ad file, a .pkl file is constructed here that will contain the state type being geneformer, the state names being geneformer coordinates, and the time series data organized as states (geneformer coordinates) by timepoint (hours) by replicates (experiment replicates).\\\\nScript Name: geneCoordinatesTimeSeries, \\\\t Description: This script organizes datasets into a format to work with time series gene expression using individual gene expresession as the state space. From a .h5ad file, a .pkl file is constructed here that will contain the state type being gene expression, the state names being gene names, and the time series data organized as states (genes) by timepoint (hours) by replicates (experiment replicates).\\\\nScript Name: linearModels, \\\\t Description: This file builds linear and linear time variant (LTI and LTV) models of gene expressions.\\\\nScript Name: geneformerEmbeddings, \\\\t Description: Get gene former embeddings of your data\\\\nScript Name: biomarkerSelectionTimeVarying, \\\\t Description: This script performs time varying sensor or biomarker selection on dynamical systems models of cellular processes. It selects new sensors at each time point.\\\\nScript Name: buildLinearModel, \\\\t Description: This script builds linear models of time series data using Dynamic Mode Decomposition (DMD).\\\\nScript Name: test, \\\\t Description: This script was made for Joshua to test and learn python\\\\nScript Name: datasetSelector, \\\\t Description: This script is a file chooser to load a gene expression dataset from the lab\\'s turbo partition to give access to the data to BRAD.\\\\nScript Name: biomarkerSelection, \\\\t Description: This script performs sensor selection on models of dynamical systems. This is applied to models of cellular processes for biomarker selection.\\\\nScript Name: buildLTVModel, \\\\t Description: This script builds linear time varying models of time series data using Data Guided Control model.\\\\n\\\\n\\\\n**User Query**\\\\n{user_query}\\\\n\\\\n**Task**\\\\nBased on the user\\'s query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\\\\n\\\\n**Response Template**\\\\nSCRIPT: <selected script>\\\\nREASON: <reasoning why the script fits the user prompt>\\\\n\"', 'input': 'Perform time varying biomarker selection on this data from times 5-7. set verbose to true.', 'output': 'content=\"SCRIPT: biomarkerSelectionTimeVarying\\\\nREASON: The user query specifically asks for time varying biomarker selection, which is the exact functionality described in the biomarkerSelectionTimeVarying script. Additionally, the script\\'s description mentions selecting new sensors at each time point, which matches the user\\'s request to perform selection from times 5-7. The verbose option can also be set to true as per the user\\'s request.\" response_metadata={\\'role\\': \\'assistant\\', \\'content\\': \"SCRIPT: biomarkerSelectionTimeVarying\\\\nREASON: The user query specifically asks for time varying biomarker selection, which is the exact functionality described in the biomarkerSelectionTimeVarying script. Additionally, the script\\'s description mentions selecting new sensors at each time point, which matches the user\\'s request to perform selection from times 5-7. The verbose option can also be set to true as per the user\\'s request.\", \\'token_usage\\': {\\'prompt_tokens\\': 549, \\'total_tokens\\': 639, \\'completion_tokens\\': 90}, \\'model_name\\': \\'meta/llama3-70b-instruct\\'} id=\\'run-1958c187-3597-4a16-9adb-1acf9f065d1f-0\\' role=\\'assistant\\'', 'parsedOutput': {'scriptName': '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', 'scriptType': 'python', 'scriptPath': '/home/jpic/BRAD-Tools/'}, 'purpose': 'Select which code to run'}, {'llm': \"model='meta/llama3-70b-instruct'\", 'prompt': 'input_variables=[\\'history\\', \\'input\\'] template=\\'Current conversation:\\\\n{history}\\\\n\\\\n**PYTHON SCRIPT**\\\\nYou must run this python script:\\\\n/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py\\\\n\\\\n**PYTHON SCRIPT DOCUMENTATION**:\\\\nThis is the doc string of this python script:\\\\n\"\"\"\\\\nThis script performs time varying sensor or biomarker selection on dynamical systems models of cellular processes. It selects new sensors at each time point.\\\\n\\\\nArguments (four arguments):\\\\n    1. output directory: chatstatus[\\\\\\'output-directory\\\\\\']\\\\n    2. output file: <name of output file>\\\\n    3. input file: the file created previously with the LTI or LTV model\\\\n    4. start time: (int, default 0) the first time to select sensors at (inclusive)\\\\n    5. end time: (int, default 10) the last time to select sensors at (not inclusive)\\\\n    6. verbose: (optional, True or False) if sensor selection should be run in verbose mode so that more output will be displayed to the user.\\\\n\\\\nUsage:\\\\nCommand Line:\\\\n```\\\\npython <path/to/script/>biomarkerSelectionTimeVarying.py <output path> <output file> <input file> <start time> <end time> <verbose>\\\\n```\\\\n\\\\nBRAD Line:\\\\n```\\\\nsubprocess.call([sys.executable, \\\\\\'<path/to/script/>/biomarkerSelectionTimeVarying.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\'], <output file>, <input file>, <start time>, <end time>, <verbose>], capture_output=True, text=True)\\\\n```\\\\n\\\\n**OUTPUT FILE NAME INSTRUCTIONS**\\\\n1. Output path should be chatstatus[\\\\\\'output-directory\\\\\\']\\\\n2. Output file name should be `S3-<descriptive name>.csv`\\\\n\"\"\"\\\\n\\\\n\\\\n**CALL PYTHON SCRIPTS FROM PYTHON**:\\\\nUse the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\\\\n\\\\nTo call a Python script `example_script.py` which has no arguments:\\\\n\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\']], capture_output=True, text=True)\\\\n)\\\\n```\\\\n\\\\nTo call a Python script `example_script.py` with one argument:\\\\n\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\'], \\\\\\'arg1\\\\\\'], capture_output=True, text=True)\\\\n)\\\\n```\\\\n\\\\nTo call a Python script `example_script.py` with two arguments:\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\'], \\\\\\'arg1\\\\\\', \\\\\\'arg2\\\\\\'], capture_output=True, text=True)\\\\n)\\\\n```\\\\n\\\\nNote that chatstatus[\\\\\\'output-directory\\\\\\'] is ALWAYS passed as the first argument.\\\\n\\\\nThe following files were previously created by BRAD and could be used as input to a function if necessary:\\\\nREPORT.tex\\\\nlog.json\\\\n.ipynb_checkpoints\\\\nREPORT.log\\\\nS1-2015.pkl\\\\nENRICHR-GO_Biological_Process_2021.csv\\\\nS0-2015.h5ad\\\\nS2-2015.pkl\\\\nimages\\\\n\\\\nQuery:{input}\\\\n\\\\n**INSTRUCTIONS**\\\\n1. Given the user query and the documentation, identify each of the arguments found in the user\\\\\\'s query that should be passed to the Python script.\\\\n2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\\\\n3. The last line of your response should say \"Execute: <Python code to execute>\"\\\\n4. Format the response/output as:\\\\n    Arguments: \\\\n    Python Code Explanation: <2 sentences maximum>\\\\n    Execute: <your code here>\\\\n\\\\n**IMPORTANT**\\\\nThe code to execute from your response must be formatted as:\\\\n    Execute: subprocess.call([sys.executable, \\\\\\'<path to python script>\\\\\\', \\\\\\'<argument 1>\\\\\\', \\\\\\'<argument 2>\\\\\\', ..., \\\\\\'<argument n>\\\\\\'], capture_output=True, text=True))\\\\nThis output should be exactly one line and no longer. Stop the response after this line.\\\\n\\'', 'input': 'Perform time varying biomarker selection on this data from times 5-7. set verbose to true.', 'output': \"Arguments: \\noutput directory: chatstatus['output-directory']\\noutput file: S3-DescriptiveName.csv\\ninput file: S2-2015.pkl (assuming it is the input file created previously with the LTI or LTV model)\\nstart time: 5\\nend time: 7\\nverbose: True\\n\\nPython Code Explanation: The script performs time varying sensor or biomarker selection on dynamical systems models of cellular processes. It selects new sensors at each time point.\\nExecute: subprocess.call([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-DescriptiveName.csv', 'S2-2015.pkl', '5', '7', 'True'], capture_output=True, text=True)\", 'parsedOutput': {'code': \"subprocess.run([sys.executable, '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py', chatstatus['output-directory'], 'S3-DescriptiveName.csv', 'S2-2015.pkl', '5', '7', 'True'], capture_output=True, text=True)\"}, 'purpose': 'Format function call'}]}\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n"
     ]
    }
   ],
   "source": [
    "brad.chat()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5d4a18cd-1eb1-417d-a1ed-f6e6163cc49f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Time varying biomarker selection script has been called!\n",
      "*******************************\n",
      " Time Varying Sensor Selection \n",
      "*******************************\n",
      "Output Path: /home/jpic/BRAD/2024-07-03_08-51-24/\n",
      "Output File: S3-biomarker-selection-time-varying.csv\n",
      "Input File : /home/jpic/BRAD/2024-07-03_08-51-24/S2-2015.pkl\n",
      "startTime  : 0\n",
      "endTime    : 3\n",
      "verbose    : True\n",
      "..............................\n",
      "         Model Loaded         \n",
      "t: 0/3\n",
      "t: 1/3\n",
      "t: 2/3\n",
      "..............................\n",
      "      Sensors Selected        \n",
      "', stderr='Traceback (most recent call last):\n",
      "  File \"/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py\", line 107, in <module>\n",
      "    main()\n",
      "  File \"/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py\", line 99, in main\n",
      "    pd.DataFrame(sensors['time']['sensors']).to_csv(output_file_path)    \n",
      "                 ~~~~~~~^^^^^^^^\n",
      "KeyError: 'time'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "X = \"\"\"Time varying biomarker selection script has been called!\\n*******************************\\n Time Varying Sensor Selection \\n*******************************\\nOutput Path: /home/jpic/BRAD/2024-07-03_08-51-24/\\nOutput File: S3-biomarker-selection-time-varying.csv\\nInput File : /home/jpic/BRAD/2024-07-03_08-51-24/S2-2015.pkl\\nstartTime  : 0\\nendTime    : 3\\nverbose    : True\\n..............................\\n         Model Loaded         \\nt: 0/3\\nt: 1/3\\nt: 2/3\\n..............................\\n      Sensors Selected        \\n', stderr='Traceback (most recent call last):\\n  File \"/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py\", line 107, in <module>\\n    main()\\n  File \"/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py\", line 99, in main\\n    pd.DataFrame(sensors[\\'time\\'][\\'sensors\\']).to_csv(output_file_path)    \\n                 ~~~~~~~^^^^^^^^\\nKeyError: \\'time\\'\\n\"\"\"\n",
    "print(X)"
   ]
  },
  {
   "cell_type": "raw",
   "id": "6c25cb0c-694a-46be-9e8b-13542e2adda3",
   "metadata": {},
   "source": [
    "/force CODE Load the cell cycle dataset\n",
    "/force CODE Convert the dataset to gene coordinates\n",
    "/force CODE Build a DMD or linear model of the 2015 data\n",
    "/force CODE Perform biomarker or sensor selection on this data\n",
    "/force CODE Perform time varying biomarker selection on this data from times 0 to 3. set verbose to true.\n",
    "/force DATABASE load the biomarker file we created and perform gene enrichrment on the state column\n",
    "/force WRITE make a summary of our conversation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "65e7e8d4-94b4-44a4-aa8d-5fe7225a6335",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0, 1, 2, 3, 4])"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "np.arange(0,5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f2f28feb-6182-4a80-b9fe-b555b0612808",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "69b9392f-89e6-439a-8e47-028ed7a92098",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "from a print statement\n"
     ]
    }
   ],
   "source": [
    "exec(\"response = subprocess.run([sys.executable, '/home/jpic/BRAD-Tools/test.py'], capture_output=True, text=True)\")\n",
    "print(response.stdout.strip())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8a3eb4f3-c749-4395-affa-766570fadc60",
   "metadata": {},
   "outputs": [],
   "source": [
    "string = \"\"\"response = subprocess.run([sys.executable, '/home/jpic/BRAD-Tools/datasetSelector.py', '', 'S0-2015.h5ad', '2015'], capture_output=True, text=True)\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "71888e74-8d73-4d97-b81d-be391da67d79",
   "metadata": {},
   "outputs": [],
   "source": [
    "exec(string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a21a554f-0001-47e0-8a2d-3c375f76293b",
   "metadata": {},
   "outputs": [],
   "source": [
    "response = subprocess.run([sys.executable, '/home/jpic/BRAD-Tools/datasetSelector.py', '', 'S0-2015.h5ad', '2015'], capture_output=True, text=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ee1f08c7-12bc-4bd6-8aa6-994e898a2d9c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "******************************\n",
      "       Dataset Selector       \n",
      "******************************\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(response.stdout)"
   ]
  },
  {
   "cell_type": "raw",
   "id": "81dade34-4ddc-48ba-a1bf-78ba8448e580",
   "metadata": {},
   "source": [
    "Load the cell cycle dataset"
   ]
  },
  {
   "cell_type": "raw",
   "id": "14f1f6a7-280e-494b-9d3b-2f1dfb79fff6",
   "metadata": {},
   "source": [
    "/force CODE Load the cell cycle dataset\n",
    "/force CODE Convert the dataset to gene coordinates\n",
    "/force CODE Build a DMD or linear model of the 2015 data\n",
    "/force CODE Perform biomarker or sensor selection on this data\n",
    "/force DATABASE load the biomarker file we created and perform gene enrichrment on the state column\n",
    "/force WRITE make a summary of all this\n"
   ]
  },
  {
   "cell_type": "raw",
   "id": "ab7e9731-1fb0-4858-8f85-9627ce479ac5",
   "metadata": {},
   "source": [
    "/force DATABASE load the biomarker file we created and perform gene enrichrment on the biomarker column\n",
    "/force DATABASE load the biomarkers from the previous step and perform gene ernichment on the biomarker column\n",
    "/force DATABASE perform gene enrichment on the state variables from the selected biomarkers"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
