{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5d807cd5-a3bf-438d-8fcd-180ffe6d1554",
   "metadata": {},
   "source": [
    "# Data Engineering for the 2015 and 2018 data\n",
    "\n",
    "This notebook formats the 2015 and 2018 data to be used as input/output for the pipeline being build in the `DMD-Demo.ipynb`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4684c9d2-b6ce-4699-b368-7d651455d67d",
   "metadata": {},
   "source": [
    "## Dataset chooser"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c2ca658-2097-4156-baa6-0c15acfad177",
   "metadata": {},
   "outputs": [],
   "source": [
    "def main():\n",
    "    outputPath = sys.argv[1] # chatstatus['output-directory']\n",
    "    outputFile = sys.argv[2] # output file name\n",
    "    dataset    = sys.argv[3]\n",
    "\n",
    "    outputFile = os.path.join(outputPath, outputFile)\n",
    "    \n",
    "    print('******************************')\n",
    "    print('Dataset Selector')\n",
    "    print('******************************')\n",
    "\n",
    "    out_path = \"/nfs/turbo/umms-indikar/shared/projects/geneformer/data/rajapakse_lab_data.h5ad\"\n",
    "    ad = an.read(out_path)\n",
    "    \n",
    "    if dataset == '2015':\n",
    "        ds = 'chen_2015'\n",
    "    elif dataset == '2018':\n",
    "        ds = 'liu_2018'\n",
    "    else:\n",
    "        print('Dataset is invalid!')\n",
    "\n",
    "    adDs = ad[ad.obs['dataset'] == 'liu_2018']\n",
    "    adDs.write(outputFile)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ef592a0d-6d59-4212-b378-4d310c686a21",
   "metadata": {},
   "outputs": [],
   "source": [
    "import anndata as an"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2fd47fcb-93f9-4256-93bd-d354d7e830d8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/__init__.py:55: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 66 × 19393\n",
       "    obs: 'dataset', 'sample_id', 'timepoint', 'hour', 'n_counts', 'control'\n",
       "    var: 'gene_id', 'token_id', 'Chromosome', 'Source', 'Feature', 'Start', 'End', 'Score', 'Strand', 'Frame', 'gene_version', 'gene_source', 'gene_biotype', 'transcript_id', 'transcript_version', 'transcript_name', 'transcript_source', 'transcript_biotype', 'tag', 'ccds_id', 'exon_number', 'exon_id', 'exon_version', 'protein_id', 'protein_version', 'transcript_support_level', 'ensembl_id'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out_path = \"/nfs/turbo/umms-indikar/shared/projects/geneformer/data/rajapakse_lab_data.h5ad\"\n",
    "ad = an.read(out_path)\n",
    "#adata\n",
    "ad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "03fb612f-21f1-4b87-9d06-75ca18833109",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "48"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(ad.obs['dataset'] == 'liu_2018')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "7d4e55e9-46f1-487c-ae84-e8d8d2c8713f",
   "metadata": {},
   "outputs": [],
   "source": [
    "adDs = ad[ad.obs['dataset'] == 'liu_2018']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81faa7a5-3a60-4892-b46d-50a9e47be351",
   "metadata": {},
   "source": [
    "## Standard TPM formatting\n",
    "\n",
    "Take the `.h5ad` files and make pickle files where `X` is state variables by time by replicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a02d002b-b2a5-458e-864d-756fecb14aba",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "162d430e-a198-4bfd-9184-cdb68a8bfaaf",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/__init__.py:55: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "out_path = '/home/jpic/BRAD/2024-06-30_16-39-42/S0-2018.h5ad'\n",
    "ad = an.read(out_path)\n",
    "\n",
    "# stateType = 'Gene Name'\n",
    "# states = list(ad.var.index)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "3970a79e-4f89-4b47-a661-be99fd86354e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 48 × 19393\n",
       "    obs: 'dataset', 'sample_id', 'timepoint', 'hour', 'n_counts', 'control', 'order', 'replicate', 'batch'\n",
       "    var: 'gene_id', 'token_id', 'Chromosome', 'Source', 'Feature', 'Start', 'End', 'Score', 'Strand', 'Frame', 'gene_version', 'gene_source', 'gene_biotype', 'transcript_id', 'transcript_version', 'transcript_name', 'transcript_source', 'transcript_biotype', 'tag', 'ccds_id', 'exon_number', 'exon_id', 'exon_version', 'protein_id', 'protein_version', 'transcript_support_level', 'ensembl_id'"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "969f4cd1-6343-461c-b79f-bb8ec87fc033",
   "metadata": {},
   "outputs": [],
   "source": [
    "n = ad.shape[1]\n",
    "T = ad.obs['order'].nunique()\n",
    "R = ad.obs['replicate'].nunique()\n",
    "X = np.zeros((n, T, R))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "2f267b4f-893b-457e-8225-977a7d4ee9c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "timepoints = ad.obs['order'].unique()\n",
    "replicates = ad.obs['replicate'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "8fbf867a-be5a-4573-8335-2e5656b98a03",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(ad.obs.shape[0]):\n",
    "    order = ad.obs['order'][i]\n",
    "    replicate = ad.obs['replicate'][i]\n",
    "\n",
    "    time = np.where(timepoints == order)[0][0]\n",
    "    rep = np.where(replicates == replicate)[0][0]\n",
    "\n",
    "    X[:, time, rep] = ad.X[i, :]\n",
    "\n",
    "states = list(ad.var.index)\n",
    "metadata = ad.var\n",
    "\n",
    "# Please write code to save X, states, and metadata to a pkl file at output_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "ff2ebfe9-8337-4175-9b07-c533cd0c4eaa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>token_id</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_biotype</th>\n",
       "      <th>tag</th>\n",
       "      <th>ccds_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_version</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>protein_version</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>ensembl_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>5150.0</td>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>58345177.0</td>\n",
       "      <td>58353492.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>ENSG00000148584</td>\n",
       "      <td>9064.0</td>\n",
       "      <td>10</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>50799408.0</td>\n",
       "      <td>50885675.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000148584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>13826.0</td>\n",
       "      <td>12</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>9067663.0</td>\n",
       "      <td>9116229.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>ENSG00000166535</td>\n",
       "      <td>11812.0</td>\n",
       "      <td>12</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>8822620.0</td>\n",
       "      <td>8887001.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000166535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A3GALT2</th>\n",
       "      <td>ENSG00000184389</td>\n",
       "      <td>15327.0</td>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>33306765.0</td>\n",
       "      <td>33321098.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000184389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11A</th>\n",
       "      <td>ENSG00000203995</td>\n",
       "      <td>17515.0</td>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>52842510.0</td>\n",
       "      <td>52894998.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000203995</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>ENSG00000162378</td>\n",
       "      <td>10655.0</td>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>52726452.0</td>\n",
       "      <td>52827336.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000162378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>ENSG00000159840</td>\n",
       "      <td>10336.0</td>\n",
       "      <td>7</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>143381294.0</td>\n",
       "      <td>143391111.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000159840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>ENSG00000074755</td>\n",
       "      <td>1277.0</td>\n",
       "      <td>17</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>4004444.0</td>\n",
       "      <td>4143030.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000074755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>ENSG00000036549</td>\n",
       "      <td>533.0</td>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>77562415.0</td>\n",
       "      <td>77683419.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000036549</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19393 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   gene_id  token_id Chromosome          Source Feature  \\\n",
       "gene_name                                                                 \n",
       "A1BG       ENSG00000121410    5150.0         19  ensembl_havana    gene   \n",
       "A1CF       ENSG00000148584    9064.0         10  ensembl_havana    gene   \n",
       "A2M        ENSG00000175899   13826.0         12  ensembl_havana    gene   \n",
       "A2ML1      ENSG00000166535   11812.0         12  ensembl_havana    gene   \n",
       "A3GALT2    ENSG00000184389   15327.0          1  ensembl_havana    gene   \n",
       "...                    ...       ...        ...             ...     ...   \n",
       "ZYG11A     ENSG00000203995   17515.0          1  ensembl_havana    gene   \n",
       "ZYG11B     ENSG00000162378   10655.0          1  ensembl_havana    gene   \n",
       "ZYX        ENSG00000159840   10336.0          7  ensembl_havana    gene   \n",
       "ZZEF1      ENSG00000074755    1277.0         17  ensembl_havana    gene   \n",
       "ZZZ3       ENSG00000036549     533.0          1  ensembl_havana    gene   \n",
       "\n",
       "                 Start          End Score Strand Frame  ...  \\\n",
       "gene_name                                               ...   \n",
       "A1BG        58345177.0   58353492.0     .      -     .  ...   \n",
       "A1CF        50799408.0   50885675.0     .      -     .  ...   \n",
       "A2M          9067663.0    9116229.0     .      -     .  ...   \n",
       "A2ML1        8822620.0    8887001.0     .      +     .  ...   \n",
       "A3GALT2     33306765.0   33321098.0     .      -     .  ...   \n",
       "...                ...          ...   ...    ...   ...  ...   \n",
       "ZYG11A      52842510.0   52894998.0     .      +     .  ...   \n",
       "ZYG11B      52726452.0   52827336.0     .      +     .  ...   \n",
       "ZYX        143381294.0  143391111.0     .      +     .  ...   \n",
       "ZZEF1        4004444.0    4143030.0     .      -     .  ...   \n",
       "ZZZ3        77562415.0   77683419.0     .      -     .  ...   \n",
       "\n",
       "           transcript_biotype tag ccds_id  exon_number  exon_id  exon_version  \\\n",
       "gene_name                                                                       \n",
       "A1BG                      NaN NaN     NaN          NaN      NaN           NaN   \n",
       "A1CF                      NaN NaN     NaN          NaN      NaN           NaN   \n",
       "A2M                       NaN NaN     NaN          NaN      NaN           NaN   \n",
       "A2ML1                     NaN NaN     NaN          NaN      NaN           NaN   \n",
       "A3GALT2                   NaN NaN     NaN          NaN      NaN           NaN   \n",
       "...                       ...  ..     ...          ...      ...           ...   \n",
       "ZYG11A                    NaN NaN     NaN          NaN      NaN           NaN   \n",
       "ZYG11B                    NaN NaN     NaN          NaN      NaN           NaN   \n",
       "ZYX                       NaN NaN     NaN          NaN      NaN           NaN   \n",
       "ZZEF1                     NaN NaN     NaN          NaN      NaN           NaN   \n",
       "ZZZ3                      NaN NaN     NaN          NaN      NaN           NaN   \n",
       "\n",
       "           protein_id  protein_version  transcript_support_level  \\\n",
       "gene_name                                                          \n",
       "A1BG              NaN              NaN                       NaN   \n",
       "A1CF              NaN              NaN                       NaN   \n",
       "A2M               NaN              NaN                       NaN   \n",
       "A2ML1             NaN              NaN                       NaN   \n",
       "A3GALT2           NaN              NaN                       NaN   \n",
       "...               ...              ...                       ...   \n",
       "ZYG11A            NaN              NaN                       NaN   \n",
       "ZYG11B            NaN              NaN                       NaN   \n",
       "ZYX               NaN              NaN                       NaN   \n",
       "ZZEF1             NaN              NaN                       NaN   \n",
       "ZZZ3              NaN              NaN                       NaN   \n",
       "\n",
       "                ensembl_id  \n",
       "gene_name                   \n",
       "A1BG       ENSG00000121410  \n",
       "A1CF       ENSG00000148584  \n",
       "A2M        ENSG00000175899  \n",
       "A2ML1      ENSG00000166535  \n",
       "A3GALT2    ENSG00000184389  \n",
       "...                    ...  \n",
       "ZYG11A     ENSG00000203995  \n",
       "ZYG11B     ENSG00000162378  \n",
       "ZYX        ENSG00000159840  \n",
       "ZZEF1      ENSG00000074755  \n",
       "ZZZ3       ENSG00000036549  \n",
       "\n",
       "[19393 rows x 27 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ad.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "7d6eb063-ed9a-482b-a8ff-03a852cac3ec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "data_id\n",
       "S1a    0.0\n",
       "S1b    0.0\n",
       "S2a    0.0\n",
       "S2b    0.0\n",
       "S3a    1.0\n",
       "S3b    1.0\n",
       "S4a    2.0\n",
       "S4b    2.0\n",
       "S5a    3.0\n",
       "S5b    3.0\n",
       "S6a    4.0\n",
       "S6b    4.0\n",
       "S7a    5.0\n",
       "S7b    5.0\n",
       "S8a    6.0\n",
       "S8b    6.0\n",
       "S9a    7.0\n",
       "S9b    7.0\n",
       "Name: timepoint, dtype: float64"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ad.obs['timepoint']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67272b83-0a0b-4383-a101-00b8052eed9b",
   "metadata": {},
   "source": [
    "# Scratch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "07340080-ef32-4979-b15d-8c73b26289e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append('/home/jpic/BRAD-Tools-Imports/')\n",
    "from classes import model # LinearTimeInvariant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d9bc1c72-01c1-44c0-b7ed-27ade4217c7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "# Specify the path to the pickle file\n",
    "pickle_file_path = '/home/jpic/BRAD/2024-07-02_08-26-09/S2-2015.pkl'\n",
    "\n",
    "# Load the pickle file\n",
    "with open(pickle_file_path, 'rb') as file:\n",
    "    data = pickle.load(file)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "378b2358-45b2-476a-93b5-722687bb2abc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "classes.model.LinearTimeInvariant"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "8f48c68b-51a1-4243-8135-103a20069b2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append('/home/jpic/BRAD-Tools')\n",
    "from classes import model # LinearTimeInvariant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "c93e3614-a128-4c3a-a5d9-f1630b7ccfb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "# Specify the path to the pickle file\n",
    "pickle_file_path = '/home/jpic/BRAD-Tools/test.pkl'\n",
    "\n",
    "# Load the pickle file\n",
    "with open(pickle_file_path, 'rb') as file:\n",
    "    data = pickle.load(file)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "4aec1dd5-d1fb-4102-a8c8-71fe7ec04c2d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "You must select which code to run to help a user.\n",
      "\n",
      "**Available Scripts**\n",
      "Script Name: geneformerCoordinatesTimeSeries, \t Description: This script organizes datasets into a format to work with time series gene expression using individual geneformer embeddings as the state space. From a .h5ad file, a .pkl file is constructed here that will contain the state type being geneformer, the state names being geneformer coordinates, and the time series data organized as states (geneformer coordinates) by timepoint (hours) by replicates (experiment replicates).\n",
      "Script Name: geneCoordinatesTimeSeries, \t Description: This script organizes datasets into a format to work with time series gene expression using individual gene expresession as the state space. From a .h5ad file, a .pkl file is constructed here that will contain the state type being gene expression, the state names being gene names, and the time series data organized as states (genes) by timepoint (hours) by replicates (experiment replicates).\n",
      "Script Name: linearModels, \t Description: This file builds linear and linear time variant (LTI and LTV) models of gene expressions.\n",
      "Script Name: geneformerEmbeddings, \t Description: Get gene former embeddings of your data\n",
      "Script Name: buildLinearModel, \t Description: This script builds linear models of time series data using Dynamic Mode Decomposition (DMD).\n",
      "Script Name: test, \t Description: This script was made for Joshua to test and learn python\n",
      "Script Name: datasetSelector, \t Description: This script is a file chooser to load a gene expression dataset from the lab's turbo partition to give access to the data to BRAD.\n",
      "Script Name: biomarkerSelection, \t Description: This script performs sensor selection on models of dynamical systems. This is applied to models of cellular processes for biomarker selection.\n",
      "Script Name: buildLTVModel, \t Description: This script builds linear time varying models of time series data using Data Guided Control model.\n",
      "\n",
      "\n",
      "**User Query**\n",
      "{user_query}\n",
      "\n",
      "**Task**\n",
      "Based on the user's query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\n",
      "\n",
      "**Response Template**\n",
      "SCRIPT: <selected script>\n",
      "REASON: <reasoning why the script fits the user prompt>\n",
      "\n"
     ]
    }
   ],
   "source": [
    "template=\"You must select which code to run to help a user.\\n\\n**Available Scripts**\\nScript Name: geneformerCoordinatesTimeSeries, \\t Description: This script organizes datasets into a format to work with time series gene expression using individual geneformer embeddings as the state space. From a .h5ad file, a .pkl file is constructed here that will contain the state type being geneformer, the state names being geneformer coordinates, and the time series data organized as states (geneformer coordinates) by timepoint (hours) by replicates (experiment replicates).\\nScript Name: geneCoordinatesTimeSeries, \\t Description: This script organizes datasets into a format to work with time series gene expression using individual gene expresession as the state space. From a .h5ad file, a .pkl file is constructed here that will contain the state type being gene expression, the state names being gene names, and the time series data organized as states (genes) by timepoint (hours) by replicates (experiment replicates).\\nScript Name: linearModels, \\t Description: This file builds linear and linear time variant (LTI and LTV) models of gene expressions.\\nScript Name: geneformerEmbeddings, \\t Description: Get gene former embeddings of your data\\nScript Name: buildLinearModel, \\t Description: This script builds linear models of time series data using Dynamic Mode Decomposition (DMD).\\nScript Name: test, \\t Description: This script was made for Joshua to test and learn python\\nScript Name: datasetSelector, \\t Description: This script is a file chooser to load a gene expression dataset from the lab's turbo partition to give access to the data to BRAD.\\nScript Name: biomarkerSelection, \\t Description: This script performs sensor selection on models of dynamical systems. This is applied to models of cellular processes for biomarker selection.\\nScript Name: buildLTVModel, \\t Description: This script builds linear time varying models of time series data using Data Guided Control model.\\n\\n\\n**User Query**\\n{user_query}\\n\\n**Task**\\nBased on the user's query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\\n\\n**Response Template**\\nSCRIPT: <selected script>\\nREASON: <reasoning why the script fits the user prompt>\\n\"\n",
    "print(template)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "dbc34bbc-ba8d-4549-bb23-b45f0c61b5c4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tuesday, July 02, 2024\n"
     ]
    }
   ],
   "source": [
    "from datetime import datetime\n",
    "\n",
    "# Get the current date\n",
    "today = datetime.now()\n",
    "\n",
    "# Format the date in a human-readable format\n",
    "formatted_date = today.strftime(\"%A, %B %d, %Y\")\n",
    "print(formatted_date)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "287e6353-13ed-4299-bf3b-94e36f86cc0c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "% Use this template for summarizing a process building models of gene expression\n",
      "\\documentclass{article}\n",
      "\\usepackage[margin=1in]{geometry}\n",
      "\\usepackage{graphicx}\n",
      "\\usepackage{url}\n",
      "\\begin{document}\n",
      "\n",
      "\\title{TITLE}\n",
      "\\author{BRAD-TITLE}\n",
      "\\maketitle\n",
      "\n",
      "\\section*{Summary of Process}\n",
      "\n",
      "\\section*{Summary of Outputs}\n",
      "\\subsection{Data}\n",
      "% {BRAD PUT SUMMARY OF DATASET}\n",
      "\n",
      "\\subsection{Coordinates}\n",
      "% {BRAD DISCUSS THE DATA REPRESENTATION}\n",
      "\n",
      "\\subsection{Model}\n",
      "% {BRAD DISCUSS THE MODEL}\n",
      "\n",
      "\\subsection{Analysis}\n",
      "% {BRAD DISCUSS THE ANALYSIS}\n",
      "\n",
      "\n",
      "\n",
      "\\end{document}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "def getLatexTemplate(path2latexTemplate):\n",
    "    \"\"\"\n",
    "    Reads the LaTeX template file and returns it as a string.\n",
    "    \n",
    "    Args:\n",
    "    path2latexTemplate (str): The path to the LaTeX template file.\n",
    "    \n",
    "    Returns:\n",
    "    str: The contents of the LaTeX template file as a string.\n",
    "    \"\"\"\n",
    "    try:\n",
    "        with open(path2latexTemplate, 'r') as file:\n",
    "            latex_template = file.read()\n",
    "        return latex_template\n",
    "    except FileNotFoundError:\n",
    "        return f\"Error: The file at {path2latexTemplate} was not found.\"\n",
    "    except IOError:\n",
    "        return f\"Error: An error occurred while reading the file at {path2latexTemplate}.\"\n",
    "\n",
    "# Example usage\n",
    "path = '/home/jpic/latex-test/templates/dmdBiomarkerEnrichment.tex'\n",
    "template_content = getLatexTemplate(path)\n",
    "print(template_content)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "43210755-cd91-4cef-a8d3-99ebc308ad4c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "% Use this template for summarizing a process building models of gene expression\n",
      "\\documentclass{article}\n",
      "\\usepackage[margin=1in]{geometry}\n",
      "\\usepackage{graphicx}\n",
      "\\usepackage{url}\n",
      "\\begin{document}\n",
      "\n",
      "\\title{TITLE}\n",
      "\\author{MY TITLE}\n",
      "\\maketitle\n",
      "\n",
      "\\section*{Summary of Process}\n",
      "\n",
      "\\section*{Summary of Outputs}\n",
      "\\subsection{Data}\n",
      "% {BRAD PUT SUMMARY OF DATASET}\n",
      "\n",
      "\\subsection{Coordinates}\n",
      "% {BRAD DISCUSS THE DATA REPRESENTATION}\n",
      "\n",
      "\\subsection{Model}\n",
      "% {BRAD DISCUSS THE MODEL}\n",
      "\n",
      "\\subsection{Analysis}\n",
      "% {BRAD DISCUSS THE ANALYSIS}\n",
      "\n",
      "\n",
      "\n",
      "\\end{document}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(template_content.replace('BRAD-TITLE', 'MY TITLE'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "06c9da63-ba3b-433e-9070-ee8a313fbcf7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'% Use this template for summarizing a process building models of gene expression\\n\\\\documentclass{article}\\n\\\\usepackage[margin=1in]{geometry}\\n\\\\usepackage{graphicx}\\n\\\\usepackage{url}\\n\\\\begin{document}\\n\\n\\\\title{TITLE}\\n\\\\author{BRAD-TITLE}\\n\\\\maketitle\\n\\n\\\\section*{Summary of Process}\\n\\n\\\\section*{Summary of Outputs}\\n\\\\subsection{Data}\\n% {BRAD PUT SUMMARY OF DATASET}\\n\\n\\\\subsection{Coordinates}\\n% {BRAD DISCUSS THE DATA REPRESENTATION}\\n\\n\\\\subsection{Model}\\n% {BRAD DISCUSS THE MODEL}\\n\\n\\\\subsection{Analysis}\\n% {BRAD DISCUSS THE ANALYSIS}\\n\\n\\n\\n\\\\end{document}\\n'"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "template_content"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ddc33a06-9718-4c89-a452-944eb13367c7",
   "metadata": {},
   "source": [
    "# Writer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "abc30944-8bdd-45af-9f99-1d3e58f44b1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "25473816-974e-4c3a-bc6a-8be38194525d",
   "metadata": {},
   "outputs": [],
   "source": [
    "chatlog = json.load(open(os.path.join('/home/jpic/BRAD/2024-07-02_08-57-36', 'log.json')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "4795a5b0-7e79-4852-95d3-c873345ff264",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "**Chat History**\n",
      "\n",
      "Human: Load the 2015 cell cycle dataset\n",
      "\n",
      "BRAD: \n",
      "\n",
      "\n",
      "Human: Convert the dataset to gene coordinates\n",
      "\n",
      "BRAD: \n",
      "\n",
      "\n",
      "Human: Build a DMD or linear model of the 2015 data\n",
      "\n",
      "BRAD: \n",
      "\n",
      "\n",
      "Human: Build a DMD or linear model of this data\n",
      "\n",
      "BRAD: \n",
      "\n",
      "\n",
      "Human: Perform biomarker or sensor selection on this data\n",
      "\n",
      "BRAD: \n",
      "\n",
      "\n",
      "Human: load the biomarker file we created and perform gene enrichrment on the state column\n",
      "\n",
      "BRAD: The following table was generated by quering the gene list against GO_Biological_Process_2021:\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "chathistory = \"**Chat History**\\n\\n\"\n",
    "for i in chatlog.keys():\n",
    "    chathistory += \"Human: \" + chatlog[i]['prompt'] + '\\n\\n'\n",
    "    chathistory += \"BRAD: \" + chatlog[i]['output'] + '\\n\\n\\n'\n",
    "print(chathistory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3ace2f5a-6320-4bae-ba0b-3ebb74087a40",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'0': {'prompt': 'Load the 2015 cell cycle dataset',\n",
       "  'output': '',\n",
       "  'process': {'module': 'CODE', 'steps': []},\n",
       "  'status': {'databases': \"{'RAG': None}\",\n",
       "   'current table': {'key': None, 'tab': None},\n",
       "   'current documents': None,\n",
       "   'queue pointer': 0,\n",
       "   'queue': []}},\n",
       " '1': {'prompt': 'Convert the dataset to gene coordinates',\n",
       "  'output': '',\n",
       "  'process': {'module': 'CODE', 'steps': []},\n",
       "  'status': {'databases': \"{'RAG': None}\",\n",
       "   'current table': {'key': None, 'tab': None},\n",
       "   'current documents': None,\n",
       "   'queue pointer': 0,\n",
       "   'queue': []}},\n",
       " '2': {'prompt': 'Build a DMD or linear model of the 2015 data',\n",
       "  'output': '',\n",
       "  'process': {'module': 'CODE', 'steps': []},\n",
       "  'status': {'databases': \"{'RAG': None}\",\n",
       "   'current table': {'key': None, 'tab': None},\n",
       "   'current documents': None,\n",
       "   'queue pointer': 0,\n",
       "   'queue': []}},\n",
       " '3': {'prompt': 'Build a DMD or linear model of this data',\n",
       "  'output': '',\n",
       "  'process': {'module': 'CODE', 'steps': []},\n",
       "  'status': {'databases': \"{'RAG': None}\",\n",
       "   'current table': {'key': None, 'tab': None},\n",
       "   'current documents': None,\n",
       "   'queue pointer': 0,\n",
       "   'queue': []}},\n",
       " '4': {'prompt': 'Perform biomarker or sensor selection on this data',\n",
       "  'output': '',\n",
       "  'process': {'module': 'CODE', 'steps': []},\n",
       "  'status': {'databases': \"{'RAG': None}\",\n",
       "   'current table': {'key': None, 'tab': None},\n",
       "   'current documents': None,\n",
       "   'queue pointer': 0,\n",
       "   'queue': []}},\n",
       " '5': {'prompt': 'load the biomarker file we created and perform gene enrichrment on the state column',\n",
       "  'output': 'The following table was generated by quering the gene list against GO_Biological_Process_2021:',\n",
       "  'process': {'module': 'DATABASE',\n",
       "   'steps': [{'llm': \"model='meta/llama3-70b-instruct'\",\n",
       "     'prompt': \"input_variables=['history', 'input'] template='Current conversation:\\\\n{history}\\\\n    \\\\nGENEONTOLOGY: The Gene Ontology (GO) is an initiative to unify the representation of gene and gene product attributes across all species via the aims: 1) maintain and develop its controlled vocabulary of gene and gene product attributes; 2) annotate genes and gene products, and assimilate and disseminate annotation data; and 3) provide tools for easy access to all aspects of the data provided by the project, and to enable functional interpretation of experimental data using the GO.\\\\n    \\\\nENRICHR: is a tool used to lookup sets of genes and their functional association. ENRICHR has access to many gene-set libraries including Allen_Brain_Atlas_up, ENCODE_Histone_Modifications_2015, Enrichr_Libraries_Most_Popular_Genes, FANTOM6_lncRNA_KD_DEGs, GO_Biological_Process_2023, GTEx, Human_Gene_Atlas, KEGG, REACTOME, Transcription_Factor_PPIs, WikiPathways and many others databases.\\\\n    \\\\nAvailable Tables:\\\\n\\\\n\\\\nQuery:{input}\\\\n\\\\n**INSTRUCTIONS**\\\\n1. From the query, decide if GENEONTOLOGY or ENRICHR should be used\\\\n2. Identify genes in the users input that should be searched. propose genes with similar names, correct the users spelling, or make small modifications to the list, but do not propose genes that are not found in the humans input.\\\\n3. Indicate if genes need to be loaded from a file: True/False\\\\n4. Use the following output format with no additional information or explination.\\\\n\\\\ndatabase: <ENRICHR or GENEONTOLOGY>\\\\ngenes: <None or Gene 1, Gene 2, Gene 3, ...>\\\\nload: <True/False>\\\\n'\",\n",
       "     'input': 'load the biomarker file we created and perform gene enrichrment on the state column',\n",
       "     'output': 'database: ENRICHR\\ngenes: <genes from the \"state\" column of the biomarker file>\\nload: True',\n",
       "     'parsedOutput': {'database': 'ENRICHR',\n",
       "      'genes': ['<genes from the state column of the biomarker file>'],\n",
       "      'load': 'True'},\n",
       "     'purpose': 'Select database'},\n",
       "    {'llm': \"model='meta/llama3-70b-instruct'\",\n",
       "     'prompt': \"input_variables=['user_query'] template='**INSTRUCTIONS**\\\\nYou are a digital assistant responsible for determining which file we should load based on a users prompt. You many choose either the files listed under AVAILABLE FILES or found in the USER QUERY. Based upon the user query, specify which if any fields should be loaded from the table. Rather than specifying a table name, the user might specify ot load data from a particular step. In this case, files are saved with the prefix `S<number>-<File Name>` to denote which step to load them from.\\\\n\\\\n**EXAMPLES**\\\\n\\\\nHere the files are selected by name:\\\\n```\\\\nAvailable Files:\\\\nS1-<AAA>.tsv\\\\nS2-<BBB>.csv\\\\n\\\\nUser Query:\\\\nUse the <XXX> output from <BBB> for the this phase of anlaysis.\\\\n\\\\nOutput:\\\\nFile: S2-<BBB>.csv\\\\nFields: <XXX>\\\\n```\\\\n\\\\nHere the files are selected by Step:\\\\n```\\\\nAvailable Files:\\\\nS1-<AAA>.tsv\\\\nS2-<BBB>.csv\\\\n\\\\nUser Query:\\\\nUse the <YYY> identified in Step 1 as input to the codes.\\\\n\\\\nOutput:\\\\nFile: S1-<AAA>.tsv\\\\nFields: <YYY>\\\\n```\\\\n\\\\n**USER QUERY**\\\\n{user_query}\\\\n\\\\nYou must select one of the following available files based on that users query:\\\\nlog.json\\\\nS3-biomarker_selection.csv\\\\nS1-2015.pkl\\\\nS0-2015.h5ad\\\\nS2-DMDmodel.pkl\\\\n\\\\n**OUTPUT**\\\\nFile: <which file you selected>\\\\nFields: <which fields in the file are important>\\\\n'\",\n",
       "     'input': 'load the biomarker file we created and perform gene enrichrment on the state column',\n",
       "     'output': 'Based on the user query, I select the file `S3-biomarker_selection.csv` as it mentions \"the biomarker file\" which is likely to be the `S3-biomarker_selection.csv` file.\\n\\nOutput:\\nFile: S3-biomarker_selection.csv\\nFields: state',\n",
       "     'parsedOutput': {'File': 'S3-biomarker_selection.csv', 'Fields': 'state'},\n",
       "     'purpose': 'Select File'},\n",
       "    {'file': '/home/jpic/BRAD/2024-07-02_08-57-36/S3-biomarker_selection.csv',\n",
       "     'delimiter': ','},\n",
       "    {'func': 'utils.save',\n",
       "     'new file': '/home/jpic/BRAD/2024-07-02_08-57-36/ENRICHR-GO_Biological_Process_2021.csv'},\n",
       "    {'func': 'utils.savefig',\n",
       "     'new file': '/home/jpic/BRAD/2024-07-02_08-57-36/images/ENRICHR-GO_Biological_Process_2021.png'},\n",
       "    {'module': 'ENRICHR',\n",
       "     'genes': \"['MT-CO1', 'ACTB', 'MT-ND4', 'MT-CO3', 'MT-CO2', 'MT-ND5', 'ACTG1', 'B2M', 'LCP1', 'GAPDH', 'CD74', 'MT-CYB', 'TUBB', 'ENO1', 'EEF2', 'HLA-DRA', 'MT-ATP6', 'PKM', 'RPS6', 'TPT1', 'MT-ND4L', 'HSPA8', 'LDHA', 'HNRNPA2B1', 'MT-ND1', 'HSP90AB1', 'MT-ND6', 'HLA-B', 'MS4A1', 'TMSB4X', 'SRRM2', 'RPL3', 'MT-ND2', 'RPS4X', 'HSP90AA1', 'PABPC1', 'EEF1A1', 'LAPTM5', 'RACK1', 'NCL', 'PGK1', 'DDX17', 'RPL8', 'MKI67', 'RPLP1', 'PFN1', 'RPL4', 'RPL11', 'TFRC', 'RPL5', 'MYH9', 'FNBP1', 'PTMA', 'MT-ND3', 'TUBA1B', 'MT-ATP8', 'HLA-DPA1', 'RPS3', 'HNRNPU', 'ARHGDIB', 'DDX5', 'IKZF2', 'HNRNPH1', 'CORO1A', 'RPLP0', 'ACTR2', 'RPS8', 'MSN', 'COTL1', 'IRF4', 'ATP5F1B', 'CFL1', 'CANX', 'HLA-E', 'RPL19', 'RPL13A', 'ADAR', 'EIF4G2', 'SRGN', 'RPS18', 'RPS23', 'NAP1L1', 'HLA-C', 'OGT', 'SEPTIN6', 'FLNA', 'PLEK', 'FTL', 'HLA-A', 'MACF1', 'SLC25A6', 'TOP2A', 'ENTPD1', 'RHOA', 'RPL37', 'RPS24', 'ST6GAL1', 'TPM4', 'HSPD1', 'TPI1']\",\n",
       "     'plot': True,\n",
       "     'save': True,\n",
       "     'database': 'GO_Biological_Process_2021',\n",
       "     'default db': False}],\n",
       "   'database-function': '<function queryEnrichr at 0x14ab30073600>'},\n",
       "  'status': {'databases': \"{'RAG': None}\",\n",
       "   'current table': {'key': None, 'tab': None},\n",
       "   'current documents': None,\n",
       "   'queue pointer': 0,\n",
       "   'queue': []}}}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chatlog"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef08b0a1-ca99-48c4-9fc0-cda45d2b2c56",
   "metadata": {},
   "source": [
    "# Debug"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6a7a2cdf-4516-4b9c-b8be-6dfacdd153b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "import os\n",
    "import sys\n",
    "import numpy as np\n",
    "sys.path.append('/home/jpic/BRAD-Tools-Imports/')\n",
    "from classes import model # LinearTimeInvariant\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d0d29c6b-20a2-4342-9f1f-312f207de9c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "inputFile = '/home/jpic/BRAD/2024-07-01_13-44-24/S2-2015.pkl'\n",
    "with open(inputFile, 'rb') as file:\n",
    "    MODEL = pickle.load(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f8005209-145b-4570-975f-aa5d1d07bc9e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "classes.model.LinearTimeInvariant"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(MODEL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "42057316-6c4a-4a69-b4e8-4811ae59c3b1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DEBUG:root:begin gram matrix comp.\n",
      "DEBUG:root:begin eigen decomp.\n"
     ]
    }
   ],
   "source": [
    "ss = obsvEnergyMaximization(MODEL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6f69915c-c714-4338-a83b-de83ccacb533",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>state</th>\n",
       "      <th>ev1</th>\n",
       "      <th>weight</th>\n",
       "      <th>rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MT-CO1</td>\n",
       "      <td>4.287626e-01</td>\n",
       "      <td>4.287626e-01</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACTB</td>\n",
       "      <td>3.408404e-01</td>\n",
       "      <td>3.408404e-01</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MT-ND4</td>\n",
       "      <td>3.046256e-01</td>\n",
       "      <td>3.046256e-01</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MT-CO3</td>\n",
       "      <td>2.860840e-01</td>\n",
       "      <td>2.860840e-01</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MT-CO2</td>\n",
       "      <td>2.057278e-01</td>\n",
       "      <td>2.057278e-01</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19388</th>\n",
       "      <td>AGBL4</td>\n",
       "      <td>2.352274e-22</td>\n",
       "      <td>2.352274e-22</td>\n",
       "      <td>19389.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19389</th>\n",
       "      <td>OR5A1</td>\n",
       "      <td>2.314715e-22</td>\n",
       "      <td>2.314715e-22</td>\n",
       "      <td>19390.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19390</th>\n",
       "      <td>TPPP3</td>\n",
       "      <td>1.952676e-22</td>\n",
       "      <td>1.952676e-22</td>\n",
       "      <td>19391.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19391</th>\n",
       "      <td>HNRNPUL2-BSCL2</td>\n",
       "      <td>1.541600e-22</td>\n",
       "      <td>1.541600e-22</td>\n",
       "      <td>19392.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19392</th>\n",
       "      <td>PRR35</td>\n",
       "      <td>1.246884e-22</td>\n",
       "      <td>1.246884e-22</td>\n",
       "      <td>19393.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19393 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                state           ev1        weight     rank\n",
       "0              MT-CO1  4.287626e-01  4.287626e-01      1.0\n",
       "1                ACTB  3.408404e-01  3.408404e-01      2.0\n",
       "2              MT-ND4  3.046256e-01  3.046256e-01      3.0\n",
       "3              MT-CO3  2.860840e-01  2.860840e-01      4.0\n",
       "4              MT-CO2  2.057278e-01  2.057278e-01      5.0\n",
       "...               ...           ...           ...      ...\n",
       "19388           AGBL4  2.352274e-22  2.352274e-22  19389.0\n",
       "19389           OR5A1  2.314715e-22  2.314715e-22  19390.0\n",
       "19390           TPPP3  1.952676e-22  1.952676e-22  19391.0\n",
       "19391  HNRNPUL2-BSCL2  1.541600e-22  1.541600e-22  19392.0\n",
       "19392           PRR35  1.246884e-22  1.246884e-22  19393.0\n",
       "\n",
       "[19393 rows x 4 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ss['sensors']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f4efc2af-0d51-4fc2-9976-d38d088dc9aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "logger = logging.getLogger()\n",
    "logger.setLevel(logging.DEBUG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3fe76eac-0f33-45c7-b9c5-5144a63f3e02",
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "logger.setLevel(logging.DEBUG)\n",
    "from scipy.sparse.linalg import eigs\n",
    "\n",
    "def obsvEnergyMaximization(MODEL, gramT=10):\n",
    "    \"\"\"\n",
    "    Perform observer energy maximization on a LinearTimeInvariant or LinearTimeVariant model.\n",
    "\n",
    "    Params:\n",
    "    --------------\n",
    "    model (Model):\n",
    "        An instance of LinearTimeInvariant or LinearTimeVariant.\n",
    "\n",
    "    gramT (int, optional):\n",
    "        Number of timepoints over which to compute the Gram matrix. Default is 10.\n",
    "\n",
    "    Returns:\n",
    "    --------------\n",
    "    result (dict):\n",
    "        The result of observer energy maximization specific to the input model. The result includes:\n",
    "        - 'sensors': DataFrame containing information about the selected sensors.\n",
    "        - 'dmd': Dictionary with Dynamic Mode Decomposition (DMD) results.\n",
    "        - 'G': Gram matrix.\n",
    "        - 'evals': Eigenvalues of the Gram matrix.\n",
    "        - 'evecs': Eigenvectors of the Gram matrix.\n",
    "\n",
    "    References:\n",
    "    --------------\n",
    "    Hasnain, A., Balakrishnan, S., Joshy, D. M., Smith, J., Haase, S. B., & Yeung, E. (2023).\n",
    "    Learning perturbation-inducible cell states from observability analysis of transcriptome dynamics.\n",
    "    Nature Communications, 14(1), 3148. [Nature Publishing Group UK London]\n",
    "    \"\"\"\n",
    "    # Auth: Joshua Pickard\n",
    "    #       jpic@umich.edu\n",
    "    # Date: January 21, 2024\n",
    "    #\n",
    "    # Note: This developed in the github.com/Jpickard1/bioObsv and github.com/CooperStansbury/DMD_gene\n",
    "    # repositories based upon the AqibHasnain/transcriptome-dynamics-dmd-observability\n",
    "    if isinstance(MODEL, model.LinearTimeInvariant):\n",
    "        # Perform observer energy maximization for LinearTimeInvariant model\n",
    "        # You can access model-specific attributes and methods here\n",
    "        # A = model.dmd_res['Atilde']\n",
    "        # u = model.dmd_res['u_r']\n",
    "        # x0_embedded = model.dmd_res['data_embedded'][:,0,:]\n",
    "        logging.debug('begin gram matrix comp.')\n",
    "        G = MODEL.gram_matrix(T=gramT, reduced=True)\n",
    "    elif isinstance(MODEL, model.LinearTimeVariant):\n",
    "        # Perform observer energy maximization for LinearTimeVariant model\n",
    "        # You can access model-specific attributes and methods here\n",
    "        # need to implement for time invariant models\n",
    "        G = MODEL.gram_matrix(T=gramT)\n",
    "    else:\n",
    "        raise ValueError(\"Unsupported model type. Supported types: LinearTimeInvariant, LinearTimeVariant\")\n",
    "\n",
    "    logging.debug('begin eigen decomp.')\n",
    "    D, V = eigs(G, k=1)\n",
    "    V = np.abs(V) # this line was missing. it is used in the original Hasnain code.\n",
    "\n",
    "    # this line we will change based on the annotated data object\n",
    "    obs = pd.DataFrame({'state'  : MODEL.states,\n",
    "                        'ev1'    : V[:,0],\n",
    "                        'weight' : np.real(V[:,0])})\n",
    "\n",
    "    obs['rank'] = obs['weight'].rank(ascending=False)\n",
    "    obs = obs.sort_values(by='rank', ascending=True)\n",
    "    obs = obs.reset_index(drop=True)\n",
    "\n",
    "    return {'sensors': obs,\n",
    "            'G'      : G,\n",
    "            'evals'  : D,\n",
    "            'evecs'  : V}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "20bffe2d-3dc7-40ce-8cac-6f82a43cab74",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on LinearTimeInvariant in module classes.model object:\n",
      "\n",
      "class LinearTimeInvariant(Model)\n",
      " |  LinearTimeInvariant(data=None, A=None, DMD=True, dmdRank=None, states=None, S=None, XpXf=None)\n",
      " |  \n",
      " |  Linear Time-Invariant (LTI) model class.\n",
      " |  \n",
      " |  This class extends the base `Model` class and represents a specific type of model\n",
      " |  with additional parameters specific to LTI systems.\n",
      " |  \n",
      " |  Params:\n",
      " |  --------------\n",
      " |  data (np.ndarray):\n",
      " |      3 dimensional array (measurements/genes by time by replicates) from which the model will be built\n",
      " |  \n",
      " |  Attributes:\n",
      " |  --------------\n",
      " |  dmd_res:\n",
      " |      Dictionary containing Dynamic Mode Decomposition (DMD) results for the LTI model.\n",
      " |  \n",
      " |  Method resolution order:\n",
      " |      LinearTimeInvariant\n",
      " |      Model\n",
      " |      builtins.object\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __init__(self, data=None, A=None, DMD=True, dmdRank=None, states=None, S=None, XpXf=None)\n",
      " |      Initialize self.  See help(type(self)) for accurate signature.\n",
      " |  \n",
      " |  evaluateOutput(self, x, t=None)\n",
      " |      Evaluate state method specific to LinearTimeInvariant model.\n",
      " |      \n",
      " |      Params:\n",
      " |      --------------\n",
      " |      x (np.array):\n",
      " |          a model state(s)\n",
      " |      \n",
      " |      Returns:\n",
      " |      --------------\n",
      " |      result: The evaluation result specific to LTI model.\n",
      " |  \n",
      " |  evaluateState(self, x)\n",
      " |      Evaluate state method specific to LinearTimeInvariant model.\n",
      " |      \n",
      " |      Params:\n",
      " |      --------------\n",
      " |      x (np.array):\n",
      " |          a model state\n",
      " |      \n",
      " |      Returns:\n",
      " |      --------------\n",
      " |      result: The evaluation result specific to LTI model.\n",
      " |  \n",
      " |  gram_matrix(self, T=50, reduced=True, randomIC=False)\n",
      " |      Compute the Gram matrix for the Linear Time-Invariant (LTI) model.\n",
      " |      \n",
      " |      Params:\n",
      " |      --------------\n",
      " |      A (np.array):\n",
      " |          Matrix representation of the Koopman operator from DMD\n",
      " |      x0 (np.array):\n",
      " |          Initial conditions from measurements from self.data\n",
      " |      nT (int, optional):\n",
      " |          Number of timepoints over which to compute the Gram matrix (default is 50).\n",
      " |      reduced (bool, optional):\n",
      " |          If True, will compute reduced G from reduced data and KO and will also return full G after inverse projection (default is True).\n",
      " |      projection_matrix (np.array, optional):\n",
      " |          The matrix used to project data and KO to low-dimensional space (first r eigenvectors of Data.T @ Data) (default is an empty array).\n",
      " |      \n",
      " |      Returns:\n",
      " |      --------------\n",
      " |      G (np.array):\n",
      " |          The Gram matrix for the LTI model.\n",
      " |      Gfull (np.array, optional):\n",
      " |          The full Gram matrix after inverse projection (returned only if reduced is True).\n",
      " |          \n",
      " |      References:\n",
      " |      --------------\n",
      " |      Hasnain, A., Balakrishnan, S., Joshy, D. M., Smith, J., Haase, S. B., & Yeung, E. (2023).\n",
      " |      Learning perturbation-inducible cell states from observability analysis of transcriptome dynamics.\n",
      " |      Nature Communications, 14(1), 3148. [Nature Publishing Group UK London]\n",
      " |  \n",
      " |  gram_matrix_TV(self, T=0, x0=None)\n",
      " |  \n",
      " |  observabilityGramian(self, T=10, reduced=False)\n",
      " |      Constructs the observability Gramian\n",
      " |      \n",
      " |      Params:\n",
      " |      --------------\n",
      " |      T (int):\n",
      " |          number of time points to use in the finite horizon summation\n",
      " |      reduced (bool, optional):\n",
      " |          if model reduction should be used to compute this\n",
      " |          \n",
      " |      Notes:\n",
      " |      --------------\n",
      " |      currently this is only implemented where reduced is True and needs to be implemented where reduced is False\n",
      " |  \n",
      " |  obsv(self, t=None, reduced=False)\n",
      " |      Compute the observability matrix.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      t : int, optional\n",
      " |          Number of terms to compute in the observability matrix. Defaults to the number of states (n).\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      ob : np.ndarray\n",
      " |          Observability matrix [c; ca; ca^2; ...].\n",
      " |  \n",
      " |  setMeasurments(self, states)\n",
      " |      Set model output and measurement matrix C\n",
      " |      \n",
      " |      Params:\n",
      " |      --------------\n",
      " |      states:\n",
      " |          a list of model states to measure\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from Model:\n",
      " |  \n",
      " |  evaluate(self, x)\n",
      " |      Evaluate: evaluates the model update and output at a state x.\n",
      " |      \n",
      " |      Params:\n",
      " |      --------------\n",
      " |      x (np.array):\n",
      " |          a model state\n",
      " |  \n",
      " |  setMeasurements(self, states)\n",
      " |      Set the measurements for the model output.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors inherited from Model:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(MODEL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "fa8d2f87-8aac-4bcc-a9d9-70a0e60d6b0a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "This is a test report with some filenames like \\texttt{data.csv} and \\texttt{script.py}.\n",
      "Some math: a = b + c. And an underscore\\_not\\_in\\_math. Also, some more math: E = mc^2 x\\_2\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "def ensureLatexFormatting(report):\n",
    "    \"\"\"\n",
    "    This function ensures appropriate LaTeX formatting.\n",
    "    \n",
    "    For instance:\n",
    "        - any \"_\" not in a math environment should be immediately proceeded by \"\\\" to make them all \"\\_\". \n",
    "          If they are already immediately proceeded by \"\\\", then no change is required.\n",
    "        - all math should be in an appropriate math environment including $ for intext characters or \n",
    "          \\begin{equation} <math here> \\end{equation} for full lines of math.\n",
    "        - any file names that have .csv, .tsv, .pkl, .h5ad should be in the \\texttt{<file name>} environment.\n",
    "    \"\"\"\n",
    "    \n",
    "    # Escape underscores not already escaped and not in math environments\n",
    "    report = re.sub(r'(?<!\\\\)_', r'\\\\_', report)\n",
    "    \n",
    "    # Ensure all inline math is enclosed in $\n",
    "    report = re.sub(r'\\$(.*?)\\$', r'$\\1$', report)  # Ensure $...$ is correct\n",
    "    report = re.sub(r'\\\\begin\\{equation\\}(.*?)\\\\end\\{equation\\}', r'\\\\begin{equation}\\1\\\\end{equation}', report)  # Ensure \\begin{equation}...\\end{equation} is correct\n",
    "    \n",
    "    # Ensure file names with specific extensions are in \\texttt{}\n",
    "    report = re.sub(r'(\\S+\\.(csv|tsv|pkl|h5ad|py|txt))', r'\\\\texttt{\\1}', report)\n",
    "    \n",
    "    return report\n",
    "\n",
    "# Example usage\n",
    "report = \"\"\"\n",
    "This is a test report with some filenames like data.csv and script.py.\n",
    "Some math: a = b + c. And an underscore\\_not_in_math. Also, some more math: E = mc^2 x_2\n",
    "\"\"\"\n",
    "\n",
    "formatted_report = ensureLatexFormatting(report)\n",
    "print(formatted_report)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "853a928e-13c9-4ce1-84ea-07c9113f8740",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "This is a test report with some filenames like \\texttt{data.csv} and script.py.\n",
      "Some math: a = b + c. And an underscore\\_not\\_in\\_math. Also, some more math: $E = mc^2$ andegin{equation} E = mc^2 \\end{equation}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "def ensureLatexFormatting(report):\n",
    "    \"\"\"\n",
    "    This function ensures appropriate LaTeX formatting.\n",
    "    \n",
    "    For instance:\n",
    "        - any \"_\" not in a math environment should be immediately proceeded by \"\\\" to make them all \"\\_\". \n",
    "          If they are already immediately proceeded by \"\\\", then no change is required.\n",
    "        - all math should be in an appropriate math environment including $ for intext characters or \n",
    "          \\begin{equation} <math here> \\end{equation} for full lines of math.\n",
    "        - any file names that have .csv, .tsv, .pkl, .h5ad should be in the \\texttt{<file name>} environment.\n",
    "    \"\"\"\n",
    "\n",
    "    # Function to escape underscores not already escaped\n",
    "    def escape_underscores(text):\n",
    "        return re.sub(r'(?<!\\\\)_', r'\\\\_', text)\n",
    "\n",
    "    # Process the report to handle different parts\n",
    "    parts = re.split(r'(\\$.*?\\$|\\\\begin\\{equation\\}.*?\\\\end\\{equation\\})', report, flags=re.DOTALL)\n",
    "    \n",
    "    processed_parts = []\n",
    "    for part in parts:\n",
    "        if part.startswith('$') and part.endswith('$'):\n",
    "            # Inline math, do not change underscores\n",
    "            processed_parts.append(part)\n",
    "        elif part.startswith('\\\\begin{equation}') and part.endswith('\\\\end{equation}'):\n",
    "            # Equation environment, do not change underscores\n",
    "            processed_parts.append(part)\n",
    "        else:\n",
    "            # Regular text, escape underscores\n",
    "            processed_parts.append(escape_underscores(part))\n",
    "    \n",
    "    report = ''.join(processed_parts)\n",
    "\n",
    "    # Ensure file names with specific extensions are in \\texttt{}\n",
    "    report = re.sub(r'(\\S+\\.(csv|tsv|pkl|h5ad))', r'\\\\texttt{\\1}', report)\n",
    "    \n",
    "    return report\n",
    "\n",
    "# Example usage\n",
    "report = \"\"\"\n",
    "This is a test report with some filenames like data.csv and script.py.\n",
    "Some math: a = b + c. And an underscore_not_in_math. Also, some more math: $E = mc^2$ and \\begin{equation} E = mc^2 \\end{equation}\n",
    "\"\"\"\n",
    "\n",
    "formatted_report = ensureLatexFormatting(report)\n",
    "print(formatted_report)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "f0d54a11-4f86-48fc-97f9-7514abb9a06e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "This is a test report with some filenames like \\texttt{data.csv} and script.py.\n",
      "Some math: $a = b + c$. And an underscore\\_not\\_in\\_math. Also, some more math: $E = mc^2 andegin{equation} E = mc^2 \\end{equation$}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "def ensureLatexFormatting(report):\n",
    "    \"\"\"\n",
    "    This function ensures appropriate LaTeX formatting.\n",
    "    \n",
    "    For instance:\n",
    "        - any \"_\" not in a math environment should be immediately proceeded by \"\\\" to make them all \"\\_\". \n",
    "          If they are already immediately proceeded by \"\\\", then no change is required.\n",
    "        - all math should be in an appropriate math environment including $ for intext characters or \n",
    "          \\begin{equation} <math here> \\end{equation} for full lines of math.\n",
    "        - any file names that have .csv, .tsv, .pkl, .h5ad should be in the \\texttt{<file name>} environment.\n",
    "    \"\"\"\n",
    "\n",
    "    # Function to escape underscores not already escaped\n",
    "    def escape_underscores(text):\n",
    "        return re.sub(r'(?<!\\\\)_', r'\\\\_', text)\n",
    "\n",
    "    # Process the report to handle different parts\n",
    "    parts = re.split(r'(\\$.*?\\$|\\\\begin\\{equation\\}.*?\\\\end\\{equation\\})', report, flags=re.DOTALL)\n",
    "    \n",
    "    processed_parts = []\n",
    "    for part in parts:\n",
    "        if part.startswith('$') and part.endswith('$'):\n",
    "            # Inline math, do not change underscores\n",
    "            processed_parts.append(part)\n",
    "        elif part.startswith('\\\\begin{equation}') and part.endswith('\\\\end{equation}'):\n",
    "            # Equation environment, do not change underscores\n",
    "            processed_parts.append(part)\n",
    "        else:\n",
    "            # Regular text, escape underscores and wrap math expressions in $\n",
    "            part = escape_underscores(part)\n",
    "            part = re.sub(r'(\\b[a-zA-Z]\\w*\\s*=\\s*[^.]+\\b)', r'$\\1$', part)  # Heuristic to detect simple equations\n",
    "            processed_parts.append(part)\n",
    "    \n",
    "    report = ''.join(processed_parts)\n",
    "\n",
    "    # Ensure file names with specific extensions are in \\texttt{}\n",
    "    report = re.sub(r'(\\S+\\.(csv|tsv|pkl|h5ad))', r'\\\\texttt{\\1}', report)\n",
    "    \n",
    "    return report\n",
    "\n",
    "# Example usage\n",
    "report = \"\"\"\n",
    "This is a test report with some filenames like data.csv and script.py.\n",
    "Some math: a = b + c. And an underscore_not_in_math. Also, some more math: E = mc^2 and \\begin{equation} E = mc^2 \\end{equation}\n",
    "\"\"\"\n",
    "\n",
    "formatted_report = ensureLatexFormatting(report)\n",
    "print(formatted_report)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "127946ee-b15c-412e-a4f2-75dd095a83d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "This is a test report with some filenames like \\texttt{data.csv}, \\texttt{script.py}, and \\texttt{example.cpp}.\n",
      "Some math: $a = b + c$. And an underscore\\_not\\_in\\_math. Also, some more math: $E = mc^2$ and more complex math $a = b / (c + d$). Also\n",
      "we could have some stuff like $x_34$\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "def ensureLatexFormatting(report):\n",
    "    \"\"\"\n",
    "    This function ensures appropriate LaTeX formatting.\n",
    "    \n",
    "    For instance:\n",
    "        - any \"_\" not in a math environment should be immediately proceeded by \"\\\" to make them all \"\\_\". \n",
    "          If they are already immediately proceeded by \"\\\", then no change is required.\n",
    "        - all math should be in an appropriate math environment including $ for intext characters or \n",
    "          \\begin{equation} <math here> \\end{equation} for full lines of math.\n",
    "        - any file names that have .csv, .tsv, .pkl, .h5ad, .txt, .py, .cpp should be in the \\texttt{<file name>} environment.\n",
    "    \"\"\"\n",
    "\n",
    "    # Function to escape underscores not already escaped\n",
    "    def escape_underscores(text):\n",
    "        return re.sub(r'(?<!\\\\)_', r'\\\\_', text)\n",
    "\n",
    "    # Process the report to handle different parts\n",
    "    parts = re.split(r'(\\$.*?\\$)', report, flags=re.DOTALL)\n",
    "    \n",
    "    processed_parts = []\n",
    "    for part in parts:\n",
    "        if part.startswith('$') and part.endswith('$'):\n",
    "            # Inline math, do not change underscores\n",
    "            processed_parts.append(part)\n",
    "        else:\n",
    "            # Regular text, escape underscores and wrap math expressions in $\n",
    "            part = escape_underscores(part)\n",
    "            part = re.sub(r'(\\b[a-zA-Z]\\w*\\s*=\\s*[^.]+\\b)', r'$\\1$', part)  # Heuristic to detect simple equations\n",
    "            processed_parts.append(part)\n",
    "    \n",
    "    report = ''.join(processed_parts)\n",
    "\n",
    "    # Ensure file names with specific extensions are in \\texttt{}\n",
    "    report = re.sub(r'(\\S+\\.(csv|tsv|pkl|h5ad|txt|py|cpp))', r'\\\\texttt{\\1}', report)\n",
    "    \n",
    "    return report\n",
    "\n",
    "# Example usage\n",
    "report = \"\"\"\n",
    "This is a test report with some filenames like data.csv, script.py, and example.cpp.\n",
    "Some math: a = b + c. And an underscore_not_in_math. Also, some more math: $E = mc^2$ and more complex math a = b / (c + d). Also\n",
    "we could have some stuff like $x_34$\n",
    "\"\"\"\n",
    "\n",
    "formatted_report = ensureLatexFormatting(report)\n",
    "print(formatted_report)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d493edb9-113c-4bfc-8cc0-05d2f0acb1f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import sys\n",
    "response = subprocess.run([sys.executable,\n",
    "                           '/home/jpic/BRAD-Tools/biomarkerSelectionTimeVarying.py',\n",
    "                           '.',\n",
    "                           'S3-biomarker-selection-time-varying.csv',\n",
    "                           '/home/jpic/BRAD/2024-07-03_08-51-24/S2-2015.pkl',\n",
    "                           '0', '3', 'True'],\n",
    "                          capture_output=True,\n",
    "                          text=True\n",
    "                         )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6f99fa26-39c2-4f59-8aeb-ed0d15faf616",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pickle\n",
    "import pandas as pd\n",
    "\n",
    "with open('/home/jpic/BRAD/2024-07-03_16-28-48/GF_embedding.pkl', 'rb') as file:\n",
    "    data = pickle.load(file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "cba605dc-2212-4f4b-9a79-b4eaffd6627d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "sys.path.append('/home/jpic/BRAD-Tools-Imports/')\n",
    "from classes import model # LinearTimeInvariant\n",
    "\n",
    "with open('/home/jpic/BRAD/2024-07-03_16-28-48/S2-GeneFormer_Embeddings_LinearModel.pkl', 'rb') as file:\n",
    "    data = pickle.load(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "6ca6856c-7578-4730-85e3-f0c870433438",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(512, 512)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.f['A'].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "847a40e7-f24d-4b9c-9f35-86420e682cae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(512, 9, 2)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data['X'].shape"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
