{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7e66f1dc-e94f-4e50-a16c-41d2ca120666",
   "metadata": {},
   "source": [
    "# Do Experiment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47a7f3ca-8aa9-4101-af16-0d7a795196b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "from BRAD import brad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3eddf6ad-5eeb-4f6f-809c-199cbb0c957b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome to RAG! The chat log from this conversation will be saved to /home/jpic/BRAD/2024-06-07-23:15:59.212259.json. How can I help?\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter your NVIDIA API key:  ········\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Would you like to use a database with BRAD [Y/N]?\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      " Y\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fri Jun  7 23:16:02 2024 INFO Use pytorch device_name: cpu\n",
      "Fri Jun  7 23:16:02 2024 INFO Load pretrained SentenceTransformer: BAAI/bge-base-en-v1.5\n",
      "/home/jpic/.local/lib/python3.11/site-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.\n",
      "  warnings.warn(\n",
      "Fri Jun  7 23:16:03 2024 INFO Anonymized telemetry enabled. See                     https://docs.trychroma.com/telemetry for more information.\n",
      "Fri Jun  7 23:16:04 2024 INFO Collection DB_cosine_cSize_700_cOver_200 is not created.\n",
      "Fri Jun  7 23:16:04 2024 INFO local\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  what transcription factors are involved in cell reprogramming of skeletal muscle?\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fri Jun  7 23:16:12 2024 INFO RAG\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "RAG >> 1: "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/langchain_core/_api/deprecation.py:119: LangChainDeprecationWarning: The method `Chain.__call__` was deprecated in langchain 0.1.0 and will be removed in 0.3.0. Use invoke instead.\n",
      "  warn_deprecated(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "132 (12)\n",
      "\n",
      "Myf5. As noted above, these transcription factors are necessary and sufficient for skeletal muscle formation, and the instruction to express either of them effectively speciﬁes the skeletal muscle lineage. In addition, auto-regulation and cross- regulation exists among the myogenic bHLH proteins (Thayer et al., 1989), such that a transient embryonically instructed induction results in stable expression of these factors. The convergence of multiple embryonic signals on this simple binary decision has been referred to as a ‘nodal point’ in muscle cell speciﬁcation (Weintraub et al., 1991a). The identiﬁcation of other bHLH proteins that regulate complex programs of differentiation,\n",
      "\n",
      "have lacked the current\n",
      "\n",
      "Instructive chromatin and muscle lineage speciﬁcation How are Myod and Myf5 capable of accessing the appropriate muscle-speciﬁc promoters and specifying the skeletal muscle lineage? As noted above, these factors can recruit histone- modifying and chromatin-remodeling complexes to muscle promoters and reset the cellular chromatin structure and transcriptional program. Rather than acting independently of the pre-existing chromatin, however, recent studies indicate that chromatin-associated complexes instruct these factors about where to bind. As such, chromatin context establishes an instructive environment for the activity of these master regulatory transcription\n",
      "\n",
      "of other complex biological events.\n",
      "\n",
      "of DNase\n",
      "\n",
      "hypersensitivity\n",
      "\n",
      "and\n",
      "\n",
      "and Groudine, 1976) –\n",
      "\n",
      "suggested\n",
      "\n",
      "is\n",
      "\n",
      "likely\n",
      "\n",
      "This review describes generally how Myod regulates the program of skeletal muscle differentiation: the simple story of a transcription factor activating its target genes [reviews covering other aspects of skeletal muscle development and Myod function, such as cell cycle regulation, have been published previously (e.g. Buckingham et al., 2003; Kitzmann and Fernandez, 2001; McKinsey et al., 2001; Pownall et al., 2002; Puri and Sartorelli, 2000)]. It also describes how a single factor integrates information from co-regulators and from chromatin-associated proteins to achieve\n",
      "\n",
      "Review\n",
      "\n",
      "Development\n",
      "\n",
      "The circuitry of a master switch: Myod and the regulation of skeletal muscle gene transcription Stephen J. Tapscott\n",
      "\n",
      "Division of Human Biology, Fred Hutchinson Cancer Research Center, 1100 Fairview Avenue North, Seattle, WA 98109, USA e-mail: stapscot@fhcrc.org\n",
      "\n",
      "Development 132, 2685-2695 Published by The Company of Biologists 2005 doi:10.1242/dev.01874\n",
      "\n",
      "Summary\n",
      "\n",
      "The expression of Myod is sufficient to convert a ﬁbroblast to a skeletal muscle cell, and, as such, is a model system in developmental biology for studying how a single initiating event can orchestrate a highly complex and predictable response. Recent ﬁndings indicate that Myod functions in an instructive\n",
      "\n",
      "might be regulated by the expression of a very small number of genes, or possibly a single gene, a so-called ‘master switch’ (Holtzer et al., 1975a; Holtzer et al., 1975b; Weintraub et al., 1973). It was in this context that a screen for genes that regulate skeletal myogenesis led to the identiﬁcation of Myod1 (Lassar et al., 1986), a basic-helix-loop-helix (bHLH) transcription factor that can induce skeletal muscle differentiation in cells from many different lineages.\n",
      "\n",
      "The cloning of Myod, a master switch for skeletal muscle In 1979, Taylor and Jones demonstrated that treating the mouse ﬁbroblast cell line 10T1/2 with the demethylating agent 5- azacytidine generated clones with a skeletal\n",
      "\n",
      "1689-1702.\n",
      "\n",
      "Penn, B. H., Bergstrom, D. A., Dilworth, F. J., Bengal, E. and Tapscott, S. J. (2004). A MyoD-generated feed-forward circuit temporally patterns gene expression during skeletal muscle differentiation. Genes Dev. 18, 2348- 2353.\n",
      "\n",
      "Perry, R. L., Parker, M. H. and Rudnicki, M. A. (2001). Activated MEK1 binds the nuclear MyoD transcriptional complex to repress transactivation. Mol. Cell 8, 291-301.\n",
      "\n",
      "Pichaud, F. and Desplan, C. (2002). Pax genes and eye organogenesis. Curr.\n",
      "\n",
      "Opin. Genet. Dev. 12, 430-434.\n",
      "\n",
      "Pownall, M. E., Gustafsson, M. K. and Emerson, C. P., Jr (2002). Myogenic regulatory factors and the speciﬁcation of muscle progenitors in vertebrate embryos. Annu. Rev. Cell Dev.\n",
      "\n",
      "Dev. 12, 430-434.\n",
      "\n",
      "Pownall, M. E., Gustafsson, M. K. and Emerson, C. P., Jr (2002). Myogenic regulatory factors and the speciﬁcation of muscle progenitors in vertebrate embryos. Annu. Rev. Cell Dev. Biol. 18, 747-783.\n",
      "\n",
      "Puri, P. L. and Sartorelli, V. (2000). Regulation of muscle regulatory factors post-transcriptional\n",
      "\n",
      "proteins,\n",
      "\n",
      "by DNA-binding, modiﬁcations. J. Cell Physiol. 185, 155-173.\n",
      "\n",
      "interacting\n",
      "\n",
      "and\n",
      "\n",
      "Puri, P. L., Avantaggiati, M. L., Balsano, C., Sang, N., Graessmann, A., Giordano, A. and Levrero, M. (1997a). p300 is required for MyoD- dependent cell cycle arrest and muscle-speciﬁc gene transcription. EMBO J. 16, 369-383.\n",
      "\n",
      "Puri, P. L., Sartorelli, V., Yang, X. J., Hamamori, Y.,\n",
      "\n",
      "for skeletal muscle speciﬁcation, are sufficient to establish the myogenic program, and they directly regulate gene expression throughout the differentiation program. Clearly, however, these master regulatory factors do not act in isolation. There is now is evidence developmentally established, as evidenced by the emerging role of the Pbx/Meis complex in marking loci for Myod activation. In addition, the binding and activity of Myod is regulated by other factors to achieve a temporal patterning of gene expression through a feed-forward mechanism. It is likely that other master regulatory factors, such as Neurod in neurogenesis and E-proteins in lymphocyte differentiation, regulate and are\n",
      "\n",
      "of these genes (Liu et al., 2000; Mal and Harter, 2003). In contrast to the differentiating muscle cell, where Myod is associated with HATs, in the myoblast (the replicating muscle precursor cell), Myod is associated with histone deacetylases (HDACs) and might actively suppress gene expression (Fulco et al., 2003; Mal and Harter, 2003; Mal et al., 2001).\n",
      "\n",
      "These ﬁndings strongly suggest that Myod acts to negatively regulate the transcription of some genes in the myoblast and that muscle differentiation is initiated when Myod switches from its association with repressive factors to activating factors. The differentiation of Myod-expressing cell lines, such as the C2C12 myoblast cell line\n",
      "Human: what transcription factors are involved in cell reprogramming of skeletal muscle?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "and the other functions slightly later in the myotome and also during the activation of adult muscle satellite cells in muscle regeneration (Asakura et al., 1995; Goldhamer et al., 1995); however, the factors regulating these functions remain largely unknown.\n",
      "\n",
      "It appears, therefore, that complex signaling events from surrounding embryonic tissues are integrated by complex regulatory elements at the Myf5 and Myod loci to accomplish a simple binary decision: whether or not to express Myod or\n",
      "\n",
      "2687\n",
      "\n",
      "Development\n",
      "\n",
      "2688\n",
      "\n",
      "Development 132 (12)\n",
      "\n",
      "Myf5. As noted above, these transcription factors are necessary and sufficient for skeletal muscle formation, and the instruction to express either of them\n",
      "\n",
      "have lacked the current\n",
      "\n",
      "Instructive chromatin and muscle lineage speciﬁcation How are Myod and Myf5 capable of accessing the appropriate muscle-speciﬁc promoters and specifying the skeletal muscle lineage? As noted above, these factors can recruit histone- modifying and chromatin-remodeling complexes to muscle promoters and reset the cellular chromatin structure and transcriptional program. Rather than acting independently of the pre-existing chromatin, however, recent studies indicate that chromatin-associated complexes instruct these factors about where to bind. As such, chromatin context establishes an instructive environment for the activity of these master regulatory transcription\n",
      "\n",
      "of other complex biological events.\n",
      "\n",
      "of DNase\n",
      "\n",
      "hypersensitivity\n",
      "\n",
      "and\n",
      "\n",
      "and Groudine, 1976) –\n",
      "\n",
      "suggested\n",
      "\n",
      "is\n",
      "\n",
      "likely\n",
      "\n",
      "This review describes generally how Myod regulates the program of skeletal muscle differentiation: the simple story of a transcription factor activating its target genes [reviews covering other aspects of skeletal muscle development and Myod function, such as cell cycle regulation, have been published previously (e.g. Buckingham et al., 2003; Kitzmann and Fernandez, 2001; McKinsey et al., 2001; Pownall et al., 2002; Puri and Sartorelli, 2000)]. It also describes how a single factor integrates information from co-regulators and from chromatin-associated proteins to achieve\n",
      "\n",
      "Review\n",
      "\n",
      "Development\n",
      "\n",
      "The circuitry of a master switch: Myod and the regulation of skeletal muscle gene transcription Stephen J. Tapscott\n",
      "\n",
      "Division of Human Biology, Fred Hutchinson Cancer Research Center, 1100 Fairview Avenue North, Seattle, WA 98109, USA e-mail: stapscot@fhcrc.org\n",
      "\n",
      "Development 132, 2685-2695 Published by The Company of Biologists 2005 doi:10.1242/dev.01874\n",
      "\n",
      "Summary\n",
      "\n",
      "The expression of Myod is sufficient to convert a ﬁbroblast to a skeletal muscle cell, and, as such, is a model system in developmental biology for studying how a single initiating event can orchestrate a highly complex and predictable response. Recent ﬁndings indicate that Myod functions in an instructive\n",
      "\n",
      "might be regulated by the expression of a very small number of genes, or possibly a single gene, a so-called ‘master switch’ (Holtzer et al., 1975a; Holtzer et al., 1975b; Weintraub et al., 1973). It was in this context that a screen for genes that regulate skeletal myogenesis led to the identiﬁcation of Myod1 (Lassar et al., 1986), a basic-helix-loop-helix (bHLH) transcription factor that can induce skeletal muscle differentiation in cells from many different lineages.\n",
      "\n",
      "The cloning of Myod, a master switch for skeletal muscle In 1979, Taylor and Jones demonstrated that treating the mouse ﬁbroblast cell line 10T1/2 with the demethylating agent 5- azacytidine generated clones with a skeletal\n",
      "\n",
      "1689-1702.\n",
      "\n",
      "Penn, B. H., Bergstrom, D. A., Dilworth, F. J., Bengal, E. and Tapscott, S. J. (2004). A MyoD-generated feed-forward circuit temporally patterns gene expression during skeletal muscle differentiation. Genes Dev. 18, 2348- 2353.\n",
      "\n",
      "Perry, R. L., Parker, M. H. and Rudnicki, M. A. (2001). Activated MEK1 binds the nuclear MyoD transcriptional complex to repress transactivation. Mol. Cell 8, 291-301.\n",
      "\n",
      "Pichaud, F. and Desplan, C. (2002). Pax genes and eye organogenesis. Curr.\n",
      "\n",
      "Opin. Genet. Dev. 12, 430-434.\n",
      "\n",
      "Pownall, M. E., Gustafsson, M. K. and Emerson, C. P., Jr (2002). Myogenic regulatory factors and the speciﬁcation of muscle progenitors in vertebrate embryos. Annu. Rev. Cell Dev.\n",
      "\n",
      "Dev. 12, 430-434.\n",
      "\n",
      "Pownall, M. E., Gustafsson, M. K. and Emerson, C. P., Jr (2002). Myogenic regulatory factors and the speciﬁcation of muscle progenitors in vertebrate embryos. Annu. Rev. Cell Dev. Biol. 18, 747-783.\n",
      "\n",
      "Puri, P. L. and Sartorelli, V. (2000). Regulation of muscle regulatory factors post-transcriptional\n",
      "\n",
      "proteins,\n",
      "\n",
      "by DNA-binding, modiﬁcations. J. Cell Physiol. 185, 155-173.\n",
      "\n",
      "interacting\n",
      "\n",
      "and\n",
      "\n",
      "Puri, P. L., Avantaggiati, M. L., Balsano, C., Sang, N., Graessmann, A., Giordano, A. and Levrero, M. (1997a). p300 is required for MyoD- dependent cell cycle arrest and muscle-speciﬁc gene transcription. EMBO J. 16, 369-383.\n",
      "\n",
      "Puri, P. L., Sartorelli, V., Yang, X. J., Hamamori, Y.,\n",
      "\n",
      "for skeletal muscle speciﬁcation, are sufficient to establish the myogenic program, and they directly regulate gene expression throughout the differentiation program. Clearly, however, these master regulatory factors do not act in isolation. There is now is evidence developmentally established, as evidenced by the emerging role of the Pbx/Meis complex in marking loci for Myod activation. In addition, the binding and activity of Myod is regulated by other factors to achieve a temporal patterning of gene expression through a feed-forward mechanism. It is likely that other master regulatory factors, such as Neurod in neurogenesis and E-proteins in lymphocyte differentiation, regulate and are\n",
      "\n",
      "of these genes (Liu et al., 2000; Mal and Harter, 2003). In contrast to the differentiating muscle cell, where Myod is associated with HATs, in the myoblast (the replicating muscle precursor cell), Myod is associated with histone deacetylases (HDACs) and might actively suppress gene expression (Fulco et al., 2003; Mal and Harter, 2003; Mal et al., 2001).\n",
      "\n",
      "These ﬁndings strongly suggest that Myod acts to negatively regulate the transcription of some genes in the myoblast and that muscle differentiation is initiated when Myod switches from its association with repressive factors to activating factors. The differentiation of Myod-expressing cell lines, such as the C2C12 myoblast cell line\n",
      "Human: what transcription factors are involved in cell reprogramming of skeletal muscle?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "and the other functions slightly later in the myotome and also during the activation of adult muscle satellite cells in muscle regeneration (Asakura et al., 1995; Goldhamer et al., 1995); however, the factors regulating these functions remain largely unknown.\n",
      "\n",
      "It appears, therefore, that complex signaling events from surrounding embryonic tissues are integrated by complex regulatory elements at the Myf5 and Myod loci to accomplish a simple binary decision: whether or not to express Myod or\n",
      "\n",
      "2687\n",
      "\n",
      "Development\n",
      "\n",
      "2688\n",
      "\n",
      "Development 132 (12)\n",
      "\n",
      "Myf5. As noted above, these transcription factors are necessary and sufficient for skeletal muscle formation, and the instruction to express either of them\n",
      "\n",
      "132 (12)\n",
      "\n",
      "Myf5. As noted above, these transcription factors are necessary and sufficient for skeletal muscle formation, and the instruction to express either of them effectively speciﬁes the skeletal muscle lineage. In addition, auto-regulation and cross- regulation exists among the myogenic bHLH proteins (Thayer et al., 1989), such that a transient embryonically instructed induction results in stable expression of these factors. The convergence of multiple embryonic signals on this simple binary decision has been referred to as a ‘nodal point’ in muscle cell speciﬁcation (Weintraub et al., 1991a). The identiﬁcation of other bHLH proteins that regulate complex programs of differentiation,\n",
      "\n",
      "of other complex biological events.\n",
      "\n",
      "of DNase\n",
      "\n",
      "hypersensitivity\n",
      "\n",
      "and\n",
      "\n",
      "and Groudine, 1976) –\n",
      "\n",
      "suggested\n",
      "\n",
      "is\n",
      "\n",
      "likely\n",
      "\n",
      "This review describes generally how Myod regulates the program of skeletal muscle differentiation: the simple story of a transcription factor activating its target genes [reviews covering other aspects of skeletal muscle development and Myod function, such as cell cycle regulation, have been published previously (e.g. Buckingham et al., 2003; Kitzmann and Fernandez, 2001; McKinsey et al., 2001; Pownall et al., 2002; Puri and Sartorelli, 2000)]. It also describes how a single factor integrates information from co-regulators and from chromatin-associated proteins to achieve\n",
      "\n",
      "Review\n",
      "\n",
      "Development\n",
      "\n",
      "The circuitry of a master switch: Myod and the regulation of skeletal muscle gene transcription Stephen J. Tapscott\n",
      "\n",
      "Division of Human Biology, Fred Hutchinson Cancer Research Center, 1100 Fairview Avenue North, Seattle, WA 98109, USA e-mail: stapscot@fhcrc.org\n",
      "\n",
      "Development 132, 2685-2695 Published by The Company of Biologists 2005 doi:10.1242/dev.01874\n",
      "\n",
      "Summary\n",
      "\n",
      "The expression of Myod is sufficient to convert a ﬁbroblast to a skeletal muscle cell, and, as such, is a model system in developmental biology for studying how a single initiating event can orchestrate a highly complex and predictable response. Recent ﬁndings indicate that Myod functions in an instructive\n",
      "\n",
      "might be regulated by the expression of a very small number of genes, or possibly a single gene, a so-called ‘master switch’ (Holtzer et al., 1975a; Holtzer et al., 1975b; Weintraub et al., 1973). It was in this context that a screen for genes that regulate skeletal myogenesis led to the identiﬁcation of Myod1 (Lassar et al., 1986), a basic-helix-loop-helix (bHLH) transcription factor that can induce skeletal muscle differentiation in cells from many different lineages.\n",
      "\n",
      "The cloning of Myod, a master switch for skeletal muscle In 1979, Taylor and Jones demonstrated that treating the mouse ﬁbroblast cell line 10T1/2 with the demethylating agent 5- azacytidine generated clones with a skeletal\n",
      "\n",
      "1689-1702.\n",
      "\n",
      "Penn, B. H., Bergstrom, D. A., Dilworth, F. J., Bengal, E. and Tapscott, S. J. (2004). A MyoD-generated feed-forward circuit temporally patterns gene expression during skeletal muscle differentiation. Genes Dev. 18, 2348- 2353.\n",
      "\n",
      "Perry, R. L., Parker, M. H. and Rudnicki, M. A. (2001). Activated MEK1 binds the nuclear MyoD transcriptional complex to repress transactivation. Mol. Cell 8, 291-301.\n",
      "\n",
      "Pichaud, F. and Desplan, C. (2002). Pax genes and eye organogenesis. Curr.\n",
      "\n",
      "Opin. Genet. Dev. 12, 430-434.\n",
      "\n",
      "Pownall, M. E., Gustafsson, M. K. and Emerson, C. P., Jr (2002). Myogenic regulatory factors and the speciﬁcation of muscle progenitors in vertebrate embryos. Annu. Rev. Cell Dev.\n",
      "\n",
      "Dev. 12, 430-434.\n",
      "\n",
      "Pownall, M. E., Gustafsson, M. K. and Emerson, C. P., Jr (2002). Myogenic regulatory factors and the speciﬁcation of muscle progenitors in vertebrate embryos. Annu. Rev. Cell Dev. Biol. 18, 747-783.\n",
      "\n",
      "Puri, P. L. and Sartorelli, V. (2000). Regulation of muscle regulatory factors post-transcriptional\n",
      "\n",
      "proteins,\n",
      "\n",
      "by DNA-binding, modiﬁcations. J. Cell Physiol. 185, 155-173.\n",
      "\n",
      "interacting\n",
      "\n",
      "and\n",
      "\n",
      "Puri, P. L., Avantaggiati, M. L., Balsano, C., Sang, N., Graessmann, A., Giordano, A. and Levrero, M. (1997a). p300 is required for MyoD- dependent cell cycle arrest and muscle-speciﬁc gene transcription. EMBO J. 16, 369-383.\n",
      "\n",
      "Puri, P. L., Sartorelli, V., Yang, X. J., Hamamori, Y.,\n",
      "\n",
      "for skeletal muscle speciﬁcation, are sufficient to establish the myogenic program, and they directly regulate gene expression throughout the differentiation program. Clearly, however, these master regulatory factors do not act in isolation. There is now is evidence developmentally established, as evidenced by the emerging role of the Pbx/Meis complex in marking loci for Myod activation. In addition, the binding and activity of Myod is regulated by other factors to achieve a temporal patterning of gene expression through a feed-forward mechanism. It is likely that other master regulatory factors, such as Neurod in neurogenesis and E-proteins in lymphocyte differentiation, regulate and are\n",
      "\n",
      "of these genes (Liu et al., 2000; Mal and Harter, 2003). In contrast to the differentiating muscle cell, where Myod is associated with HATs, in the myoblast (the replicating muscle precursor cell), Myod is associated with histone deacetylases (HDACs) and might actively suppress gene expression (Fulco et al., 2003; Mal and Harter, 2003; Mal et al., 2001).\n",
      "\n",
      "These ﬁndings strongly suggest that Myod acts to negatively regulate the transcription of some genes in the myoblast and that muscle differentiation is initiated when Myod switches from its association with repressive factors to activating factors. The differentiation of Myod-expressing cell lines, such as the C2C12 myoblast cell line\n",
      "Human: what transcription factors are involved in cell reprogramming of skeletal muscle?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "and the other functions slightly later in the myotome and also during the activation of adult muscle satellite cells in muscle regeneration (Asakura et al., 1995; Goldhamer et al., 1995); however, the factors regulating these functions remain largely unknown.\n",
      "\n",
      "It appears, therefore, that complex signaling events from surrounding embryonic tissues are integrated by complex regulatory elements at the Myf5 and Myod loci to accomplish a simple binary decision: whether or not to express Myod or\n",
      "\n",
      "2687\n",
      "\n",
      "Development\n",
      "\n",
      "2688\n",
      "\n",
      "Development 132 (12)\n",
      "\n",
      "Myf5. As noted above, these transcription factors are necessary and sufficient for skeletal muscle formation, and the instruction to express either of them\n",
      "\n",
      "132 (12)\n",
      "\n",
      "Myf5. As noted above, these transcription factors are necessary and sufficient for skeletal muscle formation, and the instruction to express either of them effectively speciﬁes the skeletal muscle lineage. In addition, auto-regulation and cross- regulation exists among the myogenic bHLH proteins (Thayer et al., 1989), such that a transient embryonically instructed induction results in stable expression of these factors. The convergence of multiple embryonic signals on this simple binary decision has been referred to as a ‘nodal point’ in muscle cell speciﬁcation (Weintraub et al., 1991a). The identiﬁcation of other bHLH proteins that regulate complex programs of differentiation,\n",
      "\n",
      "have lacked the current\n",
      "\n",
      "Instructive chromatin and muscle lineage speciﬁcation How are Myod and Myf5 capable of accessing the appropriate muscle-speciﬁc promoters and specifying the skeletal muscle lineage? As noted above, these factors can recruit histone- modifying and chromatin-remodeling complexes to muscle promoters and reset the cellular chromatin structure and transcriptional program. Rather than acting independently of the pre-existing chromatin, however, recent studies indicate that chromatin-associated complexes instruct these factors about where to bind. As such, chromatin context establishes an instructive environment for the activity of these master regulatory transcription\n",
      "\n",
      "Review\n",
      "\n",
      "Development\n",
      "\n",
      "The circuitry of a master switch: Myod and the regulation of skeletal muscle gene transcription Stephen J. Tapscott\n",
      "\n",
      "Division of Human Biology, Fred Hutchinson Cancer Research Center, 1100 Fairview Avenue North, Seattle, WA 98109, USA e-mail: stapscot@fhcrc.org\n",
      "\n",
      "Development 132, 2685-2695 Published by The Company of Biologists 2005 doi:10.1242/dev.01874\n",
      "\n",
      "Summary\n",
      "\n",
      "The expression of Myod is sufficient to convert a ﬁbroblast to a skeletal muscle cell, and, as such, is a model system in developmental biology for studying how a single initiating event can orchestrate a highly complex and predictable response. Recent ﬁndings indicate that Myod functions in an instructive\n",
      "\n",
      "might be regulated by the expression of a very small number of genes, or possibly a single gene, a so-called ‘master switch’ (Holtzer et al., 1975a; Holtzer et al., 1975b; Weintraub et al., 1973). It was in this context that a screen for genes that regulate skeletal myogenesis led to the identiﬁcation of Myod1 (Lassar et al., 1986), a basic-helix-loop-helix (bHLH) transcription factor that can induce skeletal muscle differentiation in cells from many different lineages.\n",
      "\n",
      "The cloning of Myod, a master switch for skeletal muscle In 1979, Taylor and Jones demonstrated that treating the mouse ﬁbroblast cell line 10T1/2 with the demethylating agent 5- azacytidine generated clones with a skeletal\n",
      "\n",
      "1689-1702.\n",
      "\n",
      "Penn, B. H., Bergstrom, D. A., Dilworth, F. J., Bengal, E. and Tapscott, S. J. (2004). A MyoD-generated feed-forward circuit temporally patterns gene expression during skeletal muscle differentiation. Genes Dev. 18, 2348- 2353.\n",
      "\n",
      "Perry, R. L., Parker, M. H. and Rudnicki, M. A. (2001). Activated MEK1 binds the nuclear MyoD transcriptional complex to repress transactivation. Mol. Cell 8, 291-301.\n",
      "\n",
      "Pichaud, F. and Desplan, C. (2002). Pax genes and eye organogenesis. Curr.\n",
      "\n",
      "Opin. Genet. Dev. 12, 430-434.\n",
      "\n",
      "Pownall, M. E., Gustafsson, M. K. and Emerson, C. P., Jr (2002). Myogenic regulatory factors and the speciﬁcation of muscle progenitors in vertebrate embryos. Annu. Rev. Cell Dev.\n",
      "\n",
      "Dev. 12, 430-434.\n",
      "\n",
      "Pownall, M. E., Gustafsson, M. K. and Emerson, C. P., Jr (2002). Myogenic regulatory factors and the speciﬁcation of muscle progenitors in vertebrate embryos. Annu. Rev. Cell Dev. Biol. 18, 747-783.\n",
      "\n",
      "Puri, P. L. and Sartorelli, V. (2000). Regulation of muscle regulatory factors post-transcriptional\n",
      "\n",
      "proteins,\n",
      "\n",
      "by DNA-binding, modiﬁcations. J. Cell Physiol. 185, 155-173.\n",
      "\n",
      "interacting\n",
      "\n",
      "and\n",
      "\n",
      "Puri, P. L., Avantaggiati, M. L., Balsano, C., Sang, N., Graessmann, A., Giordano, A. and Levrero, M. (1997a). p300 is required for MyoD- dependent cell cycle arrest and muscle-speciﬁc gene transcription. EMBO J. 16, 369-383.\n",
      "\n",
      "Puri, P. L., Sartorelli, V., Yang, X. J., Hamamori, Y.,\n",
      "\n",
      "for skeletal muscle speciﬁcation, are sufficient to establish the myogenic program, and they directly regulate gene expression throughout the differentiation program. Clearly, however, these master regulatory factors do not act in isolation. There is now is evidence developmentally established, as evidenced by the emerging role of the Pbx/Meis complex in marking loci for Myod activation. In addition, the binding and activity of Myod is regulated by other factors to achieve a temporal patterning of gene expression through a feed-forward mechanism. It is likely that other master regulatory factors, such as Neurod in neurogenesis and E-proteins in lymphocyte differentiation, regulate and are\n",
      "\n",
      "of these genes (Liu et al., 2000; Mal and Harter, 2003). In contrast to the differentiating muscle cell, where Myod is associated with HATs, in the myoblast (the replicating muscle precursor cell), Myod is associated with histone deacetylases (HDACs) and might actively suppress gene expression (Fulco et al., 2003; Mal and Harter, 2003; Mal et al., 2001).\n",
      "\n",
      "These ﬁndings strongly suggest that Myod acts to negatively regulate the transcription of some genes in the myoblast and that muscle differentiation is initiated when Myod switches from its association with repressive factors to activating factors. The differentiation of Myod-expressing cell lines, such as the C2C12 myoblast cell line\n",
      "Human: what transcription factors are involved in cell reprogramming of skeletal muscle?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "and the other functions slightly later in the myotome and also during the activation of adult muscle satellite cells in muscle regeneration (Asakura et al., 1995; Goldhamer et al., 1995); however, the factors regulating these functions remain largely unknown.\n",
      "\n",
      "It appears, therefore, that complex signaling events from surrounding embryonic tissues are integrated by complex regulatory elements at the Myf5 and Myod loci to accomplish a simple binary decision: whether or not to express Myod or\n",
      "\n",
      "2687\n",
      "\n",
      "Development\n",
      "\n",
      "2688\n",
      "\n",
      "Development 132 (12)\n",
      "\n",
      "Myf5. As noted above, these transcription factors are necessary and sufficient for skeletal muscle formation, and the instruction to express either of them\n",
      "\n",
      "132 (12)\n",
      "\n",
      "Myf5. As noted above, these transcription factors are necessary and sufficient for skeletal muscle formation, and the instruction to express either of them effectively speciﬁes the skeletal muscle lineage. In addition, auto-regulation and cross- regulation exists among the myogenic bHLH proteins (Thayer et al., 1989), such that a transient embryonically instructed induction results in stable expression of these factors. The convergence of multiple embryonic signals on this simple binary decision has been referred to as a ‘nodal point’ in muscle cell speciﬁcation (Weintraub et al., 1991a). The identiﬁcation of other bHLH proteins that regulate complex programs of differentiation,\n",
      "\n",
      "have lacked the current\n",
      "\n",
      "Instructive chromatin and muscle lineage speciﬁcation How are Myod and Myf5 capable of accessing the appropriate muscle-speciﬁc promoters and specifying the skeletal muscle lineage? As noted above, these factors can recruit histone- modifying and chromatin-remodeling complexes to muscle promoters and reset the cellular chromatin structure and transcriptional program. Rather than acting independently of the pre-existing chromatin, however, recent studies indicate that chromatin-associated complexes instruct these factors about where to bind. As such, chromatin context establishes an instructive environment for the activity of these master regulatory transcription\n",
      "\n",
      "of other complex biological events.\n",
      "\n",
      "of DNase\n",
      "\n",
      "hypersensitivity\n",
      "\n",
      "and\n",
      "\n",
      "and Groudine, 1976) –\n",
      "\n",
      "suggested\n",
      "\n",
      "is\n",
      "\n",
      "likely\n",
      "\n",
      "This review describes generally how Myod regulates the program of skeletal muscle differentiation: the simple story of a transcription factor activating its target genes [reviews covering other aspects of skeletal muscle development and Myod function, such as cell cycle regulation, have been published previously (e.g. Buckingham et al., 2003; Kitzmann and Fernandez, 2001; McKinsey et al., 2001; Pownall et al., 2002; Puri and Sartorelli, 2000)]. It also describes how a single factor integrates information from co-regulators and from chromatin-associated proteins to achieve\n",
      "\n",
      "might be regulated by the expression of a very small number of genes, or possibly a single gene, a so-called ‘master switch’ (Holtzer et al., 1975a; Holtzer et al., 1975b; Weintraub et al., 1973). It was in this context that a screen for genes that regulate skeletal myogenesis led to the identiﬁcation of Myod1 (Lassar et al., 1986), a basic-helix-loop-helix (bHLH) transcription factor that can induce skeletal muscle differentiation in cells from many different lineages.\n",
      "\n",
      "The cloning of Myod, a master switch for skeletal muscle In 1979, Taylor and Jones demonstrated that treating the mouse ﬁbroblast cell line 10T1/2 with the demethylating agent 5- azacytidine generated clones with a skeletal\n",
      "\n",
      "1689-1702.\n",
      "\n",
      "Penn, B. H., Bergstrom, D. A., Dilworth, F. J., Bengal, E. and Tapscott, S. J. (2004). A MyoD-generated feed-forward circuit temporally patterns gene expression during skeletal muscle differentiation. Genes Dev. 18, 2348- 2353.\n",
      "\n",
      "Perry, R. L., Parker, M. H. and Rudnicki, M. A. (2001). Activated MEK1 binds the nuclear MyoD transcriptional complex to repress transactivation. Mol. Cell 8, 291-301.\n",
      "\n",
      "Pichaud, F. and Desplan, C. (2002). Pax genes and eye organogenesis. Curr.\n",
      "\n",
      "Opin. Genet. Dev. 12, 430-434.\n",
      "\n",
      "Pownall, M. E., Gustafsson, M. K. and Emerson, C. P., Jr (2002). Myogenic regulatory factors and the speciﬁcation of muscle progenitors in vertebrate embryos. Annu. Rev. Cell Dev.\n",
      "\n",
      "Dev. 12, 430-434.\n",
      "\n",
      "Pownall, M. E., Gustafsson, M. K. and Emerson, C. P., Jr (2002). Myogenic regulatory factors and the speciﬁcation of muscle progenitors in vertebrate embryos. Annu. Rev. Cell Dev. Biol. 18, 747-783.\n",
      "\n",
      "Puri, P. L. and Sartorelli, V. (2000). Regulation of muscle regulatory factors post-transcriptional\n",
      "\n",
      "proteins,\n",
      "\n",
      "by DNA-binding, modiﬁcations. J. Cell Physiol. 185, 155-173.\n",
      "\n",
      "interacting\n",
      "\n",
      "and\n",
      "\n",
      "Puri, P. L., Avantaggiati, M. L., Balsano, C., Sang, N., Graessmann, A., Giordano, A. and Levrero, M. (1997a). p300 is required for MyoD- dependent cell cycle arrest and muscle-speciﬁc gene transcription. EMBO J. 16, 369-383.\n",
      "\n",
      "Puri, P. L., Sartorelli, V., Yang, X. J., Hamamori, Y.,\n",
      "\n",
      "for skeletal muscle speciﬁcation, are sufficient to establish the myogenic program, and they directly regulate gene expression throughout the differentiation program. Clearly, however, these master regulatory factors do not act in isolation. There is now is evidence developmentally established, as evidenced by the emerging role of the Pbx/Meis complex in marking loci for Myod activation. In addition, the binding and activity of Myod is regulated by other factors to achieve a temporal patterning of gene expression through a feed-forward mechanism. It is likely that other master regulatory factors, such as Neurod in neurogenesis and E-proteins in lymphocyte differentiation, regulate and are\n",
      "\n",
      "of these genes (Liu et al., 2000; Mal and Harter, 2003). In contrast to the differentiating muscle cell, where Myod is associated with HATs, in the myoblast (the replicating muscle precursor cell), Myod is associated with histone deacetylases (HDACs) and might actively suppress gene expression (Fulco et al., 2003; Mal and Harter, 2003; Mal et al., 2001).\n",
      "\n",
      "These ﬁndings strongly suggest that Myod acts to negatively regulate the transcription of some genes in the myoblast and that muscle differentiation is initiated when Myod switches from its association with repressive factors to activating factors. The differentiation of Myod-expressing cell lines, such as the C2C12 myoblast cell line\n",
      "Human: what transcription factors are involved in cell reprogramming of skeletal muscle?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "and the other functions slightly later in the myotome and also during the activation of adult muscle satellite cells in muscle regeneration (Asakura et al., 1995; Goldhamer et al., 1995); however, the factors regulating these functions remain largely unknown.\n",
      "\n",
      "It appears, therefore, that complex signaling events from surrounding embryonic tissues are integrated by complex regulatory elements at the Myf5 and Myod loci to accomplish a simple binary decision: whether or not to express Myod or\n",
      "\n",
      "2687\n",
      "\n",
      "Development\n",
      "\n",
      "2688\n",
      "\n",
      "Development 132 (12)\n",
      "\n",
      "Myf5. As noted above, these transcription factors are necessary and sufficient for skeletal muscle formation, and the instruction to express either of them\n",
      "\n",
      "132 (12)\n",
      "\n",
      "Myf5. As noted above, these transcription factors are necessary and sufficient for skeletal muscle formation, and the instruction to express either of them effectively speciﬁes the skeletal muscle lineage. In addition, auto-regulation and cross- regulation exists among the myogenic bHLH proteins (Thayer et al., 1989), such that a transient embryonically instructed induction results in stable expression of these factors. The convergence of multiple embryonic signals on this simple binary decision has been referred to as a ‘nodal point’ in muscle cell speciﬁcation (Weintraub et al., 1991a). The identiﬁcation of other bHLH proteins that regulate complex programs of differentiation,\n",
      "\n",
      "have lacked the current\n",
      "\n",
      "Instructive chromatin and muscle lineage speciﬁcation How are Myod and Myf5 capable of accessing the appropriate muscle-speciﬁc promoters and specifying the skeletal muscle lineage? As noted above, these factors can recruit histone- modifying and chromatin-remodeling complexes to muscle promoters and reset the cellular chromatin structure and transcriptional program. Rather than acting independently of the pre-existing chromatin, however, recent studies indicate that chromatin-associated complexes instruct these factors about where to bind. As such, chromatin context establishes an instructive environment for the activity of these master regulatory transcription\n",
      "\n",
      "of other complex biological events.\n",
      "\n",
      "of DNase\n",
      "\n",
      "hypersensitivity\n",
      "\n",
      "and\n",
      "\n",
      "and Groudine, 1976) –\n",
      "\n",
      "suggested\n",
      "\n",
      "is\n",
      "\n",
      "likely\n",
      "\n",
      "This review describes generally how Myod regulates the program of skeletal muscle differentiation: the simple story of a transcription factor activating its target genes [reviews covering other aspects of skeletal muscle development and Myod function, such as cell cycle regulation, have been published previously (e.g. Buckingham et al., 2003; Kitzmann and Fernandez, 2001; McKinsey et al., 2001; Pownall et al., 2002; Puri and Sartorelli, 2000)]. It also describes how a single factor integrates information from co-regulators and from chromatin-associated proteins to achieve\n",
      "\n",
      "Review\n",
      "\n",
      "Development\n",
      "\n",
      "The circuitry of a master switch: Myod and the regulation of skeletal muscle gene transcription Stephen J. Tapscott\n",
      "\n",
      "Division of Human Biology, Fred Hutchinson Cancer Research Center, 1100 Fairview Avenue North, Seattle, WA 98109, USA e-mail: stapscot@fhcrc.org\n",
      "\n",
      "Development 132, 2685-2695 Published by The Company of Biologists 2005 doi:10.1242/dev.01874\n",
      "\n",
      "Summary\n",
      "\n",
      "The expression of Myod is sufficient to convert a ﬁbroblast to a skeletal muscle cell, and, as such, is a model system in developmental biology for studying how a single initiating event can orchestrate a highly complex and predictable response. Recent ﬁndings indicate that Myod functions in an instructive\n",
      "\n",
      "1689-1702.\n",
      "\n",
      "Penn, B. H., Bergstrom, D. A., Dilworth, F. J., Bengal, E. and Tapscott, S. J. (2004). A MyoD-generated feed-forward circuit temporally patterns gene expression during skeletal muscle differentiation. Genes Dev. 18, 2348- 2353.\n",
      "\n",
      "Perry, R. L., Parker, M. H. and Rudnicki, M. A. (2001). Activated MEK1 binds the nuclear MyoD transcriptional complex to repress transactivation. Mol. Cell 8, 291-301.\n",
      "\n",
      "Pichaud, F. and Desplan, C. (2002). Pax genes and eye organogenesis. Curr.\n",
      "\n",
      "Opin. Genet. Dev. 12, 430-434.\n",
      "\n",
      "Pownall, M. E., Gustafsson, M. K. and Emerson, C. P., Jr (2002). Myogenic regulatory factors and the speciﬁcation of muscle progenitors in vertebrate embryos. Annu. Rev. Cell Dev.\n",
      "\n",
      "Dev. 12, 430-434.\n",
      "\n",
      "Pownall, M. E., Gustafsson, M. K. and Emerson, C. P., Jr (2002). Myogenic regulatory factors and the speciﬁcation of muscle progenitors in vertebrate embryos. Annu. Rev. Cell Dev. Biol. 18, 747-783.\n",
      "\n",
      "Puri, P. L. and Sartorelli, V. (2000). Regulation of muscle regulatory factors post-transcriptional\n",
      "\n",
      "proteins,\n",
      "\n",
      "by DNA-binding, modiﬁcations. J. Cell Physiol. 185, 155-173.\n",
      "\n",
      "interacting\n",
      "\n",
      "and\n",
      "\n",
      "Puri, P. L., Avantaggiati, M. L., Balsano, C., Sang, N., Graessmann, A., Giordano, A. and Levrero, M. (1997a). p300 is required for MyoD- dependent cell cycle arrest and muscle-speciﬁc gene transcription. EMBO J. 16, 369-383.\n",
      "\n",
      "Puri, P. L., Sartorelli, V., Yang, X. J., Hamamori, Y.,\n",
      "\n",
      "for skeletal muscle speciﬁcation, are sufficient to establish the myogenic program, and they directly regulate gene expression throughout the differentiation program. Clearly, however, these master regulatory factors do not act in isolation. There is now is evidence developmentally established, as evidenced by the emerging role of the Pbx/Meis complex in marking loci for Myod activation. In addition, the binding and activity of Myod is regulated by other factors to achieve a temporal patterning of gene expression through a feed-forward mechanism. It is likely that other master regulatory factors, such as Neurod in neurogenesis and E-proteins in lymphocyte differentiation, regulate and are\n",
      "\n",
      "of these genes (Liu et al., 2000; Mal and Harter, 2003). In contrast to the differentiating muscle cell, where Myod is associated with HATs, in the myoblast (the replicating muscle precursor cell), Myod is associated with histone deacetylases (HDACs) and might actively suppress gene expression (Fulco et al., 2003; Mal and Harter, 2003; Mal et al., 2001).\n",
      "\n",
      "These ﬁndings strongly suggest that Myod acts to negatively regulate the transcription of some genes in the myoblast and that muscle differentiation is initiated when Myod switches from its association with repressive factors to activating factors. The differentiation of Myod-expressing cell lines, such as the C2C12 myoblast cell line\n",
      "Human: what transcription factors are involved in cell reprogramming of skeletal muscle?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "and the other functions slightly later in the myotome and also during the activation of adult muscle satellite cells in muscle regeneration (Asakura et al., 1995; Goldhamer et al., 1995); however, the factors regulating these functions remain largely unknown.\n",
      "\n",
      "It appears, therefore, that complex signaling events from surrounding embryonic tissues are integrated by complex regulatory elements at the Myf5 and Myod loci to accomplish a simple binary decision: whether or not to express Myod or\n",
      "\n",
      "2687\n",
      "\n",
      "Development\n",
      "\n",
      "2688\n",
      "\n",
      "Development 132 (12)\n",
      "\n",
      "Myf5. As noted above, these transcription factors are necessary and sufficient for skeletal muscle formation, and the instruction to express either of them\n",
      "\n",
      "132 (12)\n",
      "\n",
      "Myf5. As noted above, these transcription factors are necessary and sufficient for skeletal muscle formation, and the instruction to express either of them effectively speciﬁes the skeletal muscle lineage. In addition, auto-regulation and cross- regulation exists among the myogenic bHLH proteins (Thayer et al., 1989), such that a transient embryonically instructed induction results in stable expression of these factors. The convergence of multiple embryonic signals on this simple binary decision has been referred to as a ‘nodal point’ in muscle cell speciﬁcation (Weintraub et al., 1991a). The identiﬁcation of other bHLH proteins that regulate complex programs of differentiation,\n",
      "\n",
      "have lacked the current\n",
      "\n",
      "Instructive chromatin and muscle lineage speciﬁcation How are Myod and Myf5 capable of accessing the appropriate muscle-speciﬁc promoters and specifying the skeletal muscle lineage? As noted above, these factors can recruit histone- modifying and chromatin-remodeling complexes to muscle promoters and reset the cellular chromatin structure and transcriptional program. Rather than acting independently of the pre-existing chromatin, however, recent studies indicate that chromatin-associated complexes instruct these factors about where to bind. As such, chromatin context establishes an instructive environment for the activity of these master regulatory transcription\n",
      "\n",
      "of other complex biological events.\n",
      "\n",
      "of DNase\n",
      "\n",
      "hypersensitivity\n",
      "\n",
      "and\n",
      "\n",
      "and Groudine, 1976) –\n",
      "\n",
      "suggested\n",
      "\n",
      "is\n",
      "\n",
      "likely\n",
      "\n",
      "This review describes generally how Myod regulates the program of skeletal muscle differentiation: the simple story of a transcription factor activating its target genes [reviews covering other aspects of skeletal muscle development and Myod function, such as cell cycle regulation, have been published previously (e.g. Buckingham et al., 2003; Kitzmann and Fernandez, 2001; McKinsey et al., 2001; Pownall et al., 2002; Puri and Sartorelli, 2000)]. It also describes how a single factor integrates information from co-regulators and from chromatin-associated proteins to achieve\n",
      "\n",
      "Review\n",
      "\n",
      "Development\n",
      "\n",
      "The circuitry of a master switch: Myod and the regulation of skeletal muscle gene transcription Stephen J. Tapscott\n",
      "\n",
      "Division of Human Biology, Fred Hutchinson Cancer Research Center, 1100 Fairview Avenue North, Seattle, WA 98109, USA e-mail: stapscot@fhcrc.org\n",
      "\n",
      "Development 132, 2685-2695 Published by The Company of Biologists 2005 doi:10.1242/dev.01874\n",
      "\n",
      "Summary\n",
      "\n",
      "The expression of Myod is sufficient to convert a ﬁbroblast to a skeletal muscle cell, and, as such, is a model system in developmental biology for studying how a single initiating event can orchestrate a highly complex and predictable response. Recent ﬁndings indicate that Myod functions in an instructive\n",
      "\n",
      "might be regulated by the expression of a very small number of genes, or possibly a single gene, a so-called ‘master switch’ (Holtzer et al., 1975a; Holtzer et al., 1975b; Weintraub et al., 1973). It was in this context that a screen for genes that regulate skeletal myogenesis led to the identiﬁcation of Myod1 (Lassar et al., 1986), a basic-helix-loop-helix (bHLH) transcription factor that can induce skeletal muscle differentiation in cells from many different lineages.\n",
      "\n",
      "The cloning of Myod, a master switch for skeletal muscle In 1979, Taylor and Jones demonstrated that treating the mouse ﬁbroblast cell line 10T1/2 with the demethylating agent 5- azacytidine generated clones with a skeletal\n",
      "\n",
      "Dev. 12, 430-434.\n",
      "\n",
      "Pownall, M. E., Gustafsson, M. K. and Emerson, C. P., Jr (2002). Myogenic regulatory factors and the speciﬁcation of muscle progenitors in vertebrate embryos. Annu. Rev. Cell Dev. Biol. 18, 747-783.\n",
      "\n",
      "Puri, P. L. and Sartorelli, V. (2000). Regulation of muscle regulatory factors post-transcriptional\n",
      "\n",
      "proteins,\n",
      "\n",
      "by DNA-binding, modiﬁcations. J. Cell Physiol. 185, 155-173.\n",
      "\n",
      "interacting\n",
      "\n",
      "and\n",
      "\n",
      "Puri, P. L., Avantaggiati, M. L., Balsano, C., Sang, N., Graessmann, A., Giordano, A. and Levrero, M. (1997a). p300 is required for MyoD- dependent cell cycle arrest and muscle-speciﬁc gene transcription. EMBO J. 16, 369-383.\n",
      "\n",
      "Puri, P. L., Sartorelli, V., Yang, X. J., Hamamori, Y.,\n",
      "\n",
      "for skeletal muscle speciﬁcation, are sufficient to establish the myogenic program, and they directly regulate gene expression throughout the differentiation program. Clearly, however, these master regulatory factors do not act in isolation. There is now is evidence developmentally established, as evidenced by the emerging role of the Pbx/Meis complex in marking loci for Myod activation. In addition, the binding and activity of Myod is regulated by other factors to achieve a temporal patterning of gene expression through a feed-forward mechanism. It is likely that other master regulatory factors, such as Neurod in neurogenesis and E-proteins in lymphocyte differentiation, regulate and are\n",
      "\n",
      "of these genes (Liu et al., 2000; Mal and Harter, 2003). In contrast to the differentiating muscle cell, where Myod is associated with HATs, in the myoblast (the replicating muscle precursor cell), Myod is associated with histone deacetylases (HDACs) and might actively suppress gene expression (Fulco et al., 2003; Mal and Harter, 2003; Mal et al., 2001).\n",
      "\n",
      "These ﬁndings strongly suggest that Myod acts to negatively regulate the transcription of some genes in the myoblast and that muscle differentiation is initiated when Myod switches from its association with repressive factors to activating factors. The differentiation of Myod-expressing cell lines, such as the C2C12 myoblast cell line\n",
      "Human: what transcription factors are involved in cell reprogramming of skeletal muscle?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "and the other functions slightly later in the myotome and also during the activation of adult muscle satellite cells in muscle regeneration (Asakura et al., 1995; Goldhamer et al., 1995); however, the factors regulating these functions remain largely unknown.\n",
      "\n",
      "It appears, therefore, that complex signaling events from surrounding embryonic tissues are integrated by complex regulatory elements at the Myf5 and Myod loci to accomplish a simple binary decision: whether or not to express Myod or\n",
      "\n",
      "2687\n",
      "\n",
      "Development\n",
      "\n",
      "2688\n",
      "\n",
      "Development 132 (12)\n",
      "\n",
      "Myf5. As noted above, these transcription factors are necessary and sufficient for skeletal muscle formation, and the instruction to express either of them\n",
      "\n",
      "132 (12)\n",
      "\n",
      "Myf5. As noted above, these transcription factors are necessary and sufficient for skeletal muscle formation, and the instruction to express either of them effectively speciﬁes the skeletal muscle lineage. In addition, auto-regulation and cross- regulation exists among the myogenic bHLH proteins (Thayer et al., 1989), such that a transient embryonically instructed induction results in stable expression of these factors. The convergence of multiple embryonic signals on this simple binary decision has been referred to as a ‘nodal point’ in muscle cell speciﬁcation (Weintraub et al., 1991a). The identiﬁcation of other bHLH proteins that regulate complex programs of differentiation,\n",
      "\n",
      "have lacked the current\n",
      "\n",
      "Instructive chromatin and muscle lineage speciﬁcation How are Myod and Myf5 capable of accessing the appropriate muscle-speciﬁc promoters and specifying the skeletal muscle lineage? As noted above, these factors can recruit histone- modifying and chromatin-remodeling complexes to muscle promoters and reset the cellular chromatin structure and transcriptional program. Rather than acting independently of the pre-existing chromatin, however, recent studies indicate that chromatin-associated complexes instruct these factors about where to bind. As such, chromatin context establishes an instructive environment for the activity of these master regulatory transcription\n",
      "\n",
      "of other complex biological events.\n",
      "\n",
      "of DNase\n",
      "\n",
      "hypersensitivity\n",
      "\n",
      "and\n",
      "\n",
      "and Groudine, 1976) –\n",
      "\n",
      "suggested\n",
      "\n",
      "is\n",
      "\n",
      "likely\n",
      "\n",
      "This review describes generally how Myod regulates the program of skeletal muscle differentiation: the simple story of a transcription factor activating its target genes [reviews covering other aspects of skeletal muscle development and Myod function, such as cell cycle regulation, have been published previously (e.g. Buckingham et al., 2003; Kitzmann and Fernandez, 2001; McKinsey et al., 2001; Pownall et al., 2002; Puri and Sartorelli, 2000)]. It also describes how a single factor integrates information from co-regulators and from chromatin-associated proteins to achieve\n",
      "\n",
      "Review\n",
      "\n",
      "Development\n",
      "\n",
      "The circuitry of a master switch: Myod and the regulation of skeletal muscle gene transcription Stephen J. Tapscott\n",
      "\n",
      "Division of Human Biology, Fred Hutchinson Cancer Research Center, 1100 Fairview Avenue North, Seattle, WA 98109, USA e-mail: stapscot@fhcrc.org\n",
      "\n",
      "Development 132, 2685-2695 Published by The Company of Biologists 2005 doi:10.1242/dev.01874\n",
      "\n",
      "Summary\n",
      "\n",
      "The expression of Myod is sufficient to convert a ﬁbroblast to a skeletal muscle cell, and, as such, is a model system in developmental biology for studying how a single initiating event can orchestrate a highly complex and predictable response. Recent ﬁndings indicate that Myod functions in an instructive\n",
      "\n",
      "might be regulated by the expression of a very small number of genes, or possibly a single gene, a so-called ‘master switch’ (Holtzer et al., 1975a; Holtzer et al., 1975b; Weintraub et al., 1973). It was in this context that a screen for genes that regulate skeletal myogenesis led to the identiﬁcation of Myod1 (Lassar et al., 1986), a basic-helix-loop-helix (bHLH) transcription factor that can induce skeletal muscle differentiation in cells from many different lineages.\n",
      "\n",
      "The cloning of Myod, a master switch for skeletal muscle In 1979, Taylor and Jones demonstrated that treating the mouse ﬁbroblast cell line 10T1/2 with the demethylating agent 5- azacytidine generated clones with a skeletal\n",
      "\n",
      "1689-1702.\n",
      "\n",
      "Penn, B. H., Bergstrom, D. A., Dilworth, F. J., Bengal, E. and Tapscott, S. J. (2004). A MyoD-generated feed-forward circuit temporally patterns gene expression during skeletal muscle differentiation. Genes Dev. 18, 2348- 2353.\n",
      "\n",
      "Perry, R. L., Parker, M. H. and Rudnicki, M. A. (2001). Activated MEK1 binds the nuclear MyoD transcriptional complex to repress transactivation. Mol. Cell 8, 291-301.\n",
      "\n",
      "Pichaud, F. and Desplan, C. (2002). Pax genes and eye organogenesis. Curr.\n",
      "\n",
      "Opin. Genet. Dev. 12, 430-434.\n",
      "\n",
      "Pownall, M. E., Gustafsson, M. K. and Emerson, C. P., Jr (2002). Myogenic regulatory factors and the speciﬁcation of muscle progenitors in vertebrate embryos. Annu. Rev. Cell Dev.\n",
      "\n",
      "for skeletal muscle speciﬁcation, are sufficient to establish the myogenic program, and they directly regulate gene expression throughout the differentiation program. Clearly, however, these master regulatory factors do not act in isolation. There is now is evidence developmentally established, as evidenced by the emerging role of the Pbx/Meis complex in marking loci for Myod activation. In addition, the binding and activity of Myod is regulated by other factors to achieve a temporal patterning of gene expression through a feed-forward mechanism. It is likely that other master regulatory factors, such as Neurod in neurogenesis and E-proteins in lymphocyte differentiation, regulate and are\n",
      "\n",
      "of these genes (Liu et al., 2000; Mal and Harter, 2003). In contrast to the differentiating muscle cell, where Myod is associated with HATs, in the myoblast (the replicating muscle precursor cell), Myod is associated with histone deacetylases (HDACs) and might actively suppress gene expression (Fulco et al., 2003; Mal and Harter, 2003; Mal et al., 2001).\n",
      "\n",
      "These ﬁndings strongly suggest that Myod acts to negatively regulate the transcription of some genes in the myoblast and that muscle differentiation is initiated when Myod switches from its association with repressive factors to activating factors. The differentiation of Myod-expressing cell lines, such as the C2C12 myoblast cell line\n",
      "Human: what transcription factors are involved in cell reprogramming of skeletal muscle?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "and the other functions slightly later in the myotome and also during the activation of adult muscle satellite cells in muscle regeneration (Asakura et al., 1995; Goldhamer et al., 1995); however, the factors regulating these functions remain largely unknown.\n",
      "\n",
      "It appears, therefore, that complex signaling events from surrounding embryonic tissues are integrated by complex regulatory elements at the Myf5 and Myod loci to accomplish a simple binary decision: whether or not to express Myod or\n",
      "\n",
      "2687\n",
      "\n",
      "Development\n",
      "\n",
      "2688\n",
      "\n",
      "Development 132 (12)\n",
      "\n",
      "Myf5. As noted above, these transcription factors are necessary and sufficient for skeletal muscle formation, and the instruction to express either of them\n",
      "\n",
      "132 (12)\n",
      "\n",
      "Myf5. As noted above, these transcription factors are necessary and sufficient for skeletal muscle formation, and the instruction to express either of them effectively speciﬁes the skeletal muscle lineage. In addition, auto-regulation and cross- regulation exists among the myogenic bHLH proteins (Thayer et al., 1989), such that a transient embryonically instructed induction results in stable expression of these factors. The convergence of multiple embryonic signals on this simple binary decision has been referred to as a ‘nodal point’ in muscle cell speciﬁcation (Weintraub et al., 1991a). The identiﬁcation of other bHLH proteins that regulate complex programs of differentiation,\n",
      "\n",
      "have lacked the current\n",
      "\n",
      "Instructive chromatin and muscle lineage speciﬁcation How are Myod and Myf5 capable of accessing the appropriate muscle-speciﬁc promoters and specifying the skeletal muscle lineage? As noted above, these factors can recruit histone- modifying and chromatin-remodeling complexes to muscle promoters and reset the cellular chromatin structure and transcriptional program. Rather than acting independently of the pre-existing chromatin, however, recent studies indicate that chromatin-associated complexes instruct these factors about where to bind. As such, chromatin context establishes an instructive environment for the activity of these master regulatory transcription\n",
      "\n",
      "of other complex biological events.\n",
      "\n",
      "of DNase\n",
      "\n",
      "hypersensitivity\n",
      "\n",
      "and\n",
      "\n",
      "and Groudine, 1976) –\n",
      "\n",
      "suggested\n",
      "\n",
      "is\n",
      "\n",
      "likely\n",
      "\n",
      "This review describes generally how Myod regulates the program of skeletal muscle differentiation: the simple story of a transcription factor activating its target genes [reviews covering other aspects of skeletal muscle development and Myod function, such as cell cycle regulation, have been published previously (e.g. Buckingham et al., 2003; Kitzmann and Fernandez, 2001; McKinsey et al., 2001; Pownall et al., 2002; Puri and Sartorelli, 2000)]. It also describes how a single factor integrates information from co-regulators and from chromatin-associated proteins to achieve\n",
      "\n",
      "Review\n",
      "\n",
      "Development\n",
      "\n",
      "The circuitry of a master switch: Myod and the regulation of skeletal muscle gene transcription Stephen J. Tapscott\n",
      "\n",
      "Division of Human Biology, Fred Hutchinson Cancer Research Center, 1100 Fairview Avenue North, Seattle, WA 98109, USA e-mail: stapscot@fhcrc.org\n",
      "\n",
      "Development 132, 2685-2695 Published by The Company of Biologists 2005 doi:10.1242/dev.01874\n",
      "\n",
      "Summary\n",
      "\n",
      "The expression of Myod is sufficient to convert a ﬁbroblast to a skeletal muscle cell, and, as such, is a model system in developmental biology for studying how a single initiating event can orchestrate a highly complex and predictable response. Recent ﬁndings indicate that Myod functions in an instructive\n",
      "\n",
      "might be regulated by the expression of a very small number of genes, or possibly a single gene, a so-called ‘master switch’ (Holtzer et al., 1975a; Holtzer et al., 1975b; Weintraub et al., 1973). It was in this context that a screen for genes that regulate skeletal myogenesis led to the identiﬁcation of Myod1 (Lassar et al., 1986), a basic-helix-loop-helix (bHLH) transcription factor that can induce skeletal muscle differentiation in cells from many different lineages.\n",
      "\n",
      "The cloning of Myod, a master switch for skeletal muscle In 1979, Taylor and Jones demonstrated that treating the mouse ﬁbroblast cell line 10T1/2 with the demethylating agent 5- azacytidine generated clones with a skeletal\n",
      "\n",
      "1689-1702.\n",
      "\n",
      "Penn, B. H., Bergstrom, D. A., Dilworth, F. J., Bengal, E. and Tapscott, S. J. (2004). A MyoD-generated feed-forward circuit temporally patterns gene expression during skeletal muscle differentiation. Genes Dev. 18, 2348- 2353.\n",
      "\n",
      "Perry, R. L., Parker, M. H. and Rudnicki, M. A. (2001). Activated MEK1 binds the nuclear MyoD transcriptional complex to repress transactivation. Mol. Cell 8, 291-301.\n",
      "\n",
      "Pichaud, F. and Desplan, C. (2002). Pax genes and eye organogenesis. Curr.\n",
      "\n",
      "Opin. Genet. Dev. 12, 430-434.\n",
      "\n",
      "Pownall, M. E., Gustafsson, M. K. and Emerson, C. P., Jr (2002). Myogenic regulatory factors and the speciﬁcation of muscle progenitors in vertebrate embryos. Annu. Rev. Cell Dev.\n",
      "\n",
      "Dev. 12, 430-434.\n",
      "\n",
      "Pownall, M. E., Gustafsson, M. K. and Emerson, C. P., Jr (2002). Myogenic regulatory factors and the speciﬁcation of muscle progenitors in vertebrate embryos. Annu. Rev. Cell Dev. Biol. 18, 747-783.\n",
      "\n",
      "Puri, P. L. and Sartorelli, V. (2000). Regulation of muscle regulatory factors post-transcriptional\n",
      "\n",
      "proteins,\n",
      "\n",
      "by DNA-binding, modiﬁcations. J. Cell Physiol. 185, 155-173.\n",
      "\n",
      "interacting\n",
      "\n",
      "and\n",
      "\n",
      "Puri, P. L., Avantaggiati, M. L., Balsano, C., Sang, N., Graessmann, A., Giordano, A. and Levrero, M. (1997a). p300 is required for MyoD- dependent cell cycle arrest and muscle-speciﬁc gene transcription. EMBO J. 16, 369-383.\n",
      "\n",
      "Puri, P. L., Sartorelli, V., Yang, X. J., Hamamori, Y.,\n",
      "\n",
      "of these genes (Liu et al., 2000; Mal and Harter, 2003). In contrast to the differentiating muscle cell, where Myod is associated with HATs, in the myoblast (the replicating muscle precursor cell), Myod is associated with histone deacetylases (HDACs) and might actively suppress gene expression (Fulco et al., 2003; Mal and Harter, 2003; Mal et al., 2001).\n",
      "\n",
      "These ﬁndings strongly suggest that Myod acts to negatively regulate the transcription of some genes in the myoblast and that muscle differentiation is initiated when Myod switches from its association with repressive factors to activating factors. The differentiation of Myod-expressing cell lines, such as the C2C12 myoblast cell line\n",
      "Human: what transcription factors are involved in cell reprogramming of skeletal muscle?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "and the other functions slightly later in the myotome and also during the activation of adult muscle satellite cells in muscle regeneration (Asakura et al., 1995; Goldhamer et al., 1995); however, the factors regulating these functions remain largely unknown.\n",
      "\n",
      "It appears, therefore, that complex signaling events from surrounding embryonic tissues are integrated by complex regulatory elements at the Myf5 and Myod loci to accomplish a simple binary decision: whether or not to express Myod or\n",
      "\n",
      "2687\n",
      "\n",
      "Development\n",
      "\n",
      "2688\n",
      "\n",
      "Development 132 (12)\n",
      "\n",
      "Myf5. As noted above, these transcription factors are necessary and sufficient for skeletal muscle formation, and the instruction to express either of them\n",
      "\n",
      "132 (12)\n",
      "\n",
      "Myf5. As noted above, these transcription factors are necessary and sufficient for skeletal muscle formation, and the instruction to express either of them effectively speciﬁes the skeletal muscle lineage. In addition, auto-regulation and cross- regulation exists among the myogenic bHLH proteins (Thayer et al., 1989), such that a transient embryonically instructed induction results in stable expression of these factors. The convergence of multiple embryonic signals on this simple binary decision has been referred to as a ‘nodal point’ in muscle cell speciﬁcation (Weintraub et al., 1991a). The identiﬁcation of other bHLH proteins that regulate complex programs of differentiation,\n",
      "\n",
      "have lacked the current\n",
      "\n",
      "Instructive chromatin and muscle lineage speciﬁcation How are Myod and Myf5 capable of accessing the appropriate muscle-speciﬁc promoters and specifying the skeletal muscle lineage? As noted above, these factors can recruit histone- modifying and chromatin-remodeling complexes to muscle promoters and reset the cellular chromatin structure and transcriptional program. Rather than acting independently of the pre-existing chromatin, however, recent studies indicate that chromatin-associated complexes instruct these factors about where to bind. As such, chromatin context establishes an instructive environment for the activity of these master regulatory transcription\n",
      "\n",
      "of other complex biological events.\n",
      "\n",
      "of DNase\n",
      "\n",
      "hypersensitivity\n",
      "\n",
      "and\n",
      "\n",
      "and Groudine, 1976) –\n",
      "\n",
      "suggested\n",
      "\n",
      "is\n",
      "\n",
      "likely\n",
      "\n",
      "This review describes generally how Myod regulates the program of skeletal muscle differentiation: the simple story of a transcription factor activating its target genes [reviews covering other aspects of skeletal muscle development and Myod function, such as cell cycle regulation, have been published previously (e.g. Buckingham et al., 2003; Kitzmann and Fernandez, 2001; McKinsey et al., 2001; Pownall et al., 2002; Puri and Sartorelli, 2000)]. It also describes how a single factor integrates information from co-regulators and from chromatin-associated proteins to achieve\n",
      "\n",
      "Review\n",
      "\n",
      "Development\n",
      "\n",
      "The circuitry of a master switch: Myod and the regulation of skeletal muscle gene transcription Stephen J. Tapscott\n",
      "\n",
      "Division of Human Biology, Fred Hutchinson Cancer Research Center, 1100 Fairview Avenue North, Seattle, WA 98109, USA e-mail: stapscot@fhcrc.org\n",
      "\n",
      "Development 132, 2685-2695 Published by The Company of Biologists 2005 doi:10.1242/dev.01874\n",
      "\n",
      "Summary\n",
      "\n",
      "The expression of Myod is sufficient to convert a ﬁbroblast to a skeletal muscle cell, and, as such, is a model system in developmental biology for studying how a single initiating event can orchestrate a highly complex and predictable response. Recent ﬁndings indicate that Myod functions in an instructive\n",
      "\n",
      "might be regulated by the expression of a very small number of genes, or possibly a single gene, a so-called ‘master switch’ (Holtzer et al., 1975a; Holtzer et al., 1975b; Weintraub et al., 1973). It was in this context that a screen for genes that regulate skeletal myogenesis led to the identiﬁcation of Myod1 (Lassar et al., 1986), a basic-helix-loop-helix (bHLH) transcription factor that can induce skeletal muscle differentiation in cells from many different lineages.\n",
      "\n",
      "The cloning of Myod, a master switch for skeletal muscle In 1979, Taylor and Jones demonstrated that treating the mouse ﬁbroblast cell line 10T1/2 with the demethylating agent 5- azacytidine generated clones with a skeletal\n",
      "\n",
      "1689-1702.\n",
      "\n",
      "Penn, B. H., Bergstrom, D. A., Dilworth, F. J., Bengal, E. and Tapscott, S. J. (2004). A MyoD-generated feed-forward circuit temporally patterns gene expression during skeletal muscle differentiation. Genes Dev. 18, 2348- 2353.\n",
      "\n",
      "Perry, R. L., Parker, M. H. and Rudnicki, M. A. (2001). Activated MEK1 binds the nuclear MyoD transcriptional complex to repress transactivation. Mol. Cell 8, 291-301.\n",
      "\n",
      "Pichaud, F. and Desplan, C. (2002). Pax genes and eye organogenesis. Curr.\n",
      "\n",
      "Opin. Genet. Dev. 12, 430-434.\n",
      "\n",
      "Pownall, M. E., Gustafsson, M. K. and Emerson, C. P., Jr (2002). Myogenic regulatory factors and the speciﬁcation of muscle progenitors in vertebrate embryos. Annu. Rev. Cell Dev.\n",
      "\n",
      "Dev. 12, 430-434.\n",
      "\n",
      "Pownall, M. E., Gustafsson, M. K. and Emerson, C. P., Jr (2002). Myogenic regulatory factors and the speciﬁcation of muscle progenitors in vertebrate embryos. Annu. Rev. Cell Dev. Biol. 18, 747-783.\n",
      "\n",
      "Puri, P. L. and Sartorelli, V. (2000). Regulation of muscle regulatory factors post-transcriptional\n",
      "\n",
      "proteins,\n",
      "\n",
      "by DNA-binding, modiﬁcations. J. Cell Physiol. 185, 155-173.\n",
      "\n",
      "interacting\n",
      "\n",
      "and\n",
      "\n",
      "Puri, P. L., Avantaggiati, M. L., Balsano, C., Sang, N., Graessmann, A., Giordano, A. and Levrero, M. (1997a). p300 is required for MyoD- dependent cell cycle arrest and muscle-speciﬁc gene transcription. EMBO J. 16, 369-383.\n",
      "\n",
      "Puri, P. L., Sartorelli, V., Yang, X. J., Hamamori, Y.,\n",
      "\n",
      "for skeletal muscle speciﬁcation, are sufficient to establish the myogenic program, and they directly regulate gene expression throughout the differentiation program. Clearly, however, these master regulatory factors do not act in isolation. There is now is evidence developmentally established, as evidenced by the emerging role of the Pbx/Meis complex in marking loci for Myod activation. In addition, the binding and activity of Myod is regulated by other factors to achieve a temporal patterning of gene expression through a feed-forward mechanism. It is likely that other master regulatory factors, such as Neurod in neurogenesis and E-proteins in lymphocyte differentiation, regulate and are\n",
      "Human: what transcription factors are involved in cell reprogramming of skeletal muscle?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "==================================================\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  what transcription factors are involved in cell reprogramming of stem cells?\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fri Jun  7 23:17:05 2024 INFO RAG\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "RAG >> 2: \n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "Transcription factors orchestrate dynamic interplay between genome topology and gene regulation during cell reprogramming\n",
      "\n",
      "Ralph Stadhouders  François Le Dily1,2,3, Javier Quilez  Yasmina Cuartero1,2,3, Jochen Hecht2,6, Guillaume J. Filion1,2, Miguel Beato1,2, Marc A. Marti-Renom  and Thomas Graf\n",
      "\n",
      "1,2,8*, Enrique Vidal\n",
      "\n",
      "1,2, François Serra\n",
      "\n",
      "1,2,3, Bruno Di Stefano\n",
      "\n",
      "1,2,4,  5, Clara Berenguer1,2,\n",
      "\n",
      "1,2, Antonio Gomez1,2, Samuel Collombet\n",
      "\n",
      "1,2*\n",
      "\n",
      "1,2,3,7*\n",
      "\n",
      "Chromosomal architecture is known to influence gene expression, yet its role in controlling cell fate remains poorly understood. Reprogramming of somatic cells into pluripotent stem cells (PSCs) by the transcription factors (TFs) OCT4, SOX2,\n",
      "\n",
      "equally to this manuscript.\n",
      "\n",
      "doi: 10.1111/nyas.12989\n",
      "\n",
      "24\n",
      "\n",
      "Ann. N.Y. Acad. Sci. 1370 (2016) 24–35 C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "Figure 1. Established TF reprogramming strategies. TF reprogramming has been shown to work across a variety of cell types. This ﬁgure shows the starting and end cell populations, as well as the TF cocktails used to derive each cell type of choice. Adapted from Ref. 25.\n",
      "\n",
      "The novel technology of induced pluripotent stem cells (iPSCs) from Yamanaka and colleagues instigated the entire ﬁeld of regenerative medicine. Using only four transcription factors (TFs), Oct3/4, Sox2, Klf4, and c-Myc,\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "\n",
      "et al. Rapid chromatin switch in the direct repro- gramming of ﬁbroblasts to neurons. Cell Rep. 20, 3236–3247 (2017).\n",
      "\n",
      "42. Dall’Agnese, A. et al. Transcription factor-directed re-wiring of\n",
      "\n",
      "chromatin architecture for somatic cell nuclear reprogramming toward trans-differentiation. Mol. Cell 76, 453–472 (2019). 43. Stone, N. R. et al. Context-speciﬁc transcription factor functions\n",
      "\n",
      "regulate epigenomic and transcriptional dynamics during cardiac reprogramming. Cell Stem Cell 25, 87–102 (2019).\n",
      "\n",
      "44. Nam, Y. J. et al. Reprogramming of human ﬁbroblasts toward a cardiac fate. Proc. Natl Acad. Sci. USA 110, 5588–5593 (2013). 45. Zhou, Y. et al. Single-cell transcriptomic analyses of cell fate\n",
      "\n",
      "2, Peter D. Adams Pier Lorenzo Puri Alessandra Sacco1, Peter Andersen 2 & Alexandre R. Colas\n",
      "\n",
      "1,6, Michael S. Yu1,\n",
      "\n",
      "4, Li Qian 5,\n",
      "\n",
      "1\n",
      "\n",
      "Check for updates\n",
      "\n",
      "Deﬁning the mechanisms safeguarding cell fate identity in differentiated cells is crucial to improve 1) - our understanding of how differentiation is main- tained in healthy tissues or altered in a disease state, and 2) - our ability to use cell fate reprogramming for regenerative purposes. Here, using a genome- wide transcription factor screen followed by validation steps in a variety of reprogramming assays (cardiac, neural and iPSC in ﬁbroblasts and endothelial cells), we identiﬁed a set of four transcription factors (ATF7IP, JUNB, SP7,\n",
      "\n",
      "Ann. N.Y. Acad. Sci. ISSN 0077-8923\n",
      "\n",
      "A N N A L S O F T H E N E W Y O R K A C A D E M Y O F SC I E N C E S Issue: Hematopoietic Stem Cells IX\n",
      "\n",
      "Converting cell fates: generating hematopoietic stem cells de novo via transcription factor reprogramming\n",
      "\n",
      "Michael G. Daniel,1,2,3 Ihor R. Lemischka,1,2,4,∗ and Kateri Moore1,2,∗ 1Department of Developmental and Regenerative Biology, Icahn School of Medicine, New York, New York. 2Black Family Stem Cell Institute, Icahn School of Medicine, New York, New York. 3The Graduate School of Biomedical Science, Icahn School of Medicine, New York, New York. 4Department of Pharmacology and Systems Therapeutics, Icahn School of Medicine, New York, New York\n",
      "\n",
      "Address\n",
      "\n",
      "Cahan, P. et al. CellNet: network biology applied to stem cell engineering. Cell 158, 903–915 (2014).\n",
      "\n",
      "10. Rackham, O. J. et al. A predictive computational framework for direct reprogramming between human cell types. Nat. Genet. 48, 331–335 (2016).\n",
      "\n",
      "11. D’Alessio, A. C. et al. A systematic approach to identify candidate trans-\n",
      "\n",
      "cription factors that control cell identity. Stem Cell Rep. 5, 763–775 (2015). 12. Lambert, S. A. et al. The human transcription factors. Cell 175, 598–599 (2018). 13. Nakatake, Y. et al. Generation and profiling of 2,135 human ESC lines for the systematic analyses of cell states perturbed by inducing single transcription factors. Cell Rep. 31, 107655 (2020).\n",
      "\n",
      "14.\n",
      "\n",
      "Cahan, P. et al. CellNet: network biology applied to stem cell engineering. Cell 158, 903–915 (2014).\n",
      "\n",
      "10. Rackham, O. J. et al. A predictive computational framework for direct reprogramming between human cell types. Nat. Genet. 48, 331–335 (2016).\n",
      "\n",
      "11. D’Alessio, A. C. et al. A systematic approach to identify candidate trans-\n",
      "\n",
      "cription factors that control cell identity. Stem Cell Rep. 5, 763–775 (2015). 12. Lambert, S. A. et al. The human transcription factors. Cell 175, 598–599 (2018). 13. Nakatake, Y. et al. Generation and profiling of 2,135 human ESC lines for the systematic analyses of cell states perturbed by inducing single transcription factors. Cell Rep. 31, 107655 (2020).\n",
      "\n",
      "14.\n",
      "\n",
      "If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the first iPSCs.\n",
      "\n",
      "Direct cell conversion\n",
      "\n",
      "+\n",
      "\n",
      "MYOD\n",
      "\n",
      "Fibroblast\n",
      "\n",
      "+\n",
      "\n",
      "GATA1\n",
      "\n",
      "Myeloblast\n",
      "\n",
      "+\n",
      "\n",
      "CEBP\n",
      "\n",
      "B lymphocyte\n",
      "\n",
      "Diﬀerentiated\n",
      "\n",
      "Transdiﬀerentiation\n",
      "\n",
      "Myoblast\n",
      "\n",
      "Megakaryocyte, eosinophil, erythrocyte\n",
      "\n",
      "Macrophage\n",
      "\n",
      "Identifying candidate reprogramming factors. According to previous experiments, genes that contribute to the identities of fertilized eggs or ES cells were thought to be, in large part, responsible for the reconstitution of specific transcriptional networks during reprogramming. Thus, in 2000, we\n",
      "Human: what transcription factors are involved in cell reprogramming of stem cells?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "equally to this manuscript.\n",
      "\n",
      "doi: 10.1111/nyas.12989\n",
      "\n",
      "24\n",
      "\n",
      "Ann. N.Y. Acad. Sci. 1370 (2016) 24–35 C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "Figure 1. Established TF reprogramming strategies. TF reprogramming has been shown to work across a variety of cell types. This ﬁgure shows the starting and end cell populations, as well as the TF cocktails used to derive each cell type of choice. Adapted from Ref. 25.\n",
      "\n",
      "The novel technology of induced pluripotent stem cells (iPSCs) from Yamanaka and colleagues instigated the entire ﬁeld of regenerative medicine. Using only four transcription factors (TFs), Oct3/4, Sox2, Klf4, and c-Myc,\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "\n",
      "et al. Rapid chromatin switch in the direct repro- gramming of ﬁbroblasts to neurons. Cell Rep. 20, 3236–3247 (2017).\n",
      "\n",
      "42. Dall’Agnese, A. et al. Transcription factor-directed re-wiring of\n",
      "\n",
      "chromatin architecture for somatic cell nuclear reprogramming toward trans-differentiation. Mol. Cell 76, 453–472 (2019). 43. Stone, N. R. et al. Context-speciﬁc transcription factor functions\n",
      "\n",
      "regulate epigenomic and transcriptional dynamics during cardiac reprogramming. Cell Stem Cell 25, 87–102 (2019).\n",
      "\n",
      "44. Nam, Y. J. et al. Reprogramming of human ﬁbroblasts toward a cardiac fate. Proc. Natl Acad. Sci. USA 110, 5588–5593 (2013). 45. Zhou, Y. et al. Single-cell transcriptomic analyses of cell fate\n",
      "\n",
      "2, Peter D. Adams Pier Lorenzo Puri Alessandra Sacco1, Peter Andersen 2 & Alexandre R. Colas\n",
      "\n",
      "1,6, Michael S. Yu1,\n",
      "\n",
      "4, Li Qian 5,\n",
      "\n",
      "1\n",
      "\n",
      "Check for updates\n",
      "\n",
      "Deﬁning the mechanisms safeguarding cell fate identity in differentiated cells is crucial to improve 1) - our understanding of how differentiation is main- tained in healthy tissues or altered in a disease state, and 2) - our ability to use cell fate reprogramming for regenerative purposes. Here, using a genome- wide transcription factor screen followed by validation steps in a variety of reprogramming assays (cardiac, neural and iPSC in ﬁbroblasts and endothelial cells), we identiﬁed a set of four transcription factors (ATF7IP, JUNB, SP7,\n",
      "\n",
      "Ann. N.Y. Acad. Sci. ISSN 0077-8923\n",
      "\n",
      "A N N A L S O F T H E N E W Y O R K A C A D E M Y O F SC I E N C E S Issue: Hematopoietic Stem Cells IX\n",
      "\n",
      "Converting cell fates: generating hematopoietic stem cells de novo via transcription factor reprogramming\n",
      "\n",
      "Michael G. Daniel,1,2,3 Ihor R. Lemischka,1,2,4,∗ and Kateri Moore1,2,∗ 1Department of Developmental and Regenerative Biology, Icahn School of Medicine, New York, New York. 2Black Family Stem Cell Institute, Icahn School of Medicine, New York, New York. 3The Graduate School of Biomedical Science, Icahn School of Medicine, New York, New York. 4Department of Pharmacology and Systems Therapeutics, Icahn School of Medicine, New York, New York\n",
      "\n",
      "Address\n",
      "\n",
      "Cahan, P. et al. CellNet: network biology applied to stem cell engineering. Cell 158, 903–915 (2014).\n",
      "\n",
      "10. Rackham, O. J. et al. A predictive computational framework for direct reprogramming between human cell types. Nat. Genet. 48, 331–335 (2016).\n",
      "\n",
      "11. D’Alessio, A. C. et al. A systematic approach to identify candidate trans-\n",
      "\n",
      "cription factors that control cell identity. Stem Cell Rep. 5, 763–775 (2015). 12. Lambert, S. A. et al. The human transcription factors. Cell 175, 598–599 (2018). 13. Nakatake, Y. et al. Generation and profiling of 2,135 human ESC lines for the systematic analyses of cell states perturbed by inducing single transcription factors. Cell Rep. 31, 107655 (2020).\n",
      "\n",
      "14.\n",
      "\n",
      "Cahan, P. et al. CellNet: network biology applied to stem cell engineering. Cell 158, 903–915 (2014).\n",
      "\n",
      "10. Rackham, O. J. et al. A predictive computational framework for direct reprogramming between human cell types. Nat. Genet. 48, 331–335 (2016).\n",
      "\n",
      "11. D’Alessio, A. C. et al. A systematic approach to identify candidate trans-\n",
      "\n",
      "cription factors that control cell identity. Stem Cell Rep. 5, 763–775 (2015). 12. Lambert, S. A. et al. The human transcription factors. Cell 175, 598–599 (2018). 13. Nakatake, Y. et al. Generation and profiling of 2,135 human ESC lines for the systematic analyses of cell states perturbed by inducing single transcription factors. Cell Rep. 31, 107655 (2020).\n",
      "\n",
      "14.\n",
      "\n",
      "If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the first iPSCs.\n",
      "\n",
      "Direct cell conversion\n",
      "\n",
      "+\n",
      "\n",
      "MYOD\n",
      "\n",
      "Fibroblast\n",
      "\n",
      "+\n",
      "\n",
      "GATA1\n",
      "\n",
      "Myeloblast\n",
      "\n",
      "+\n",
      "\n",
      "CEBP\n",
      "\n",
      "B lymphocyte\n",
      "\n",
      "Diﬀerentiated\n",
      "\n",
      "Transdiﬀerentiation\n",
      "\n",
      "Myoblast\n",
      "\n",
      "Megakaryocyte, eosinophil, erythrocyte\n",
      "\n",
      "Macrophage\n",
      "\n",
      "Identifying candidate reprogramming factors. According to previous experiments, genes that contribute to the identities of fertilized eggs or ES cells were thought to be, in large part, responsible for the reconstitution of specific transcriptional networks during reprogramming. Thus, in 2000, we\n",
      "Human: what transcription factors are involved in cell reprogramming of stem cells?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "Transcription factors orchestrate dynamic interplay between genome topology and gene regulation during cell reprogramming\n",
      "\n",
      "Ralph Stadhouders  François Le Dily1,2,3, Javier Quilez  Yasmina Cuartero1,2,3, Jochen Hecht2,6, Guillaume J. Filion1,2, Miguel Beato1,2, Marc A. Marti-Renom  and Thomas Graf\n",
      "\n",
      "1,2,8*, Enrique Vidal\n",
      "\n",
      "1,2, François Serra\n",
      "\n",
      "1,2,3, Bruno Di Stefano\n",
      "\n",
      "1,2,4,  5, Clara Berenguer1,2,\n",
      "\n",
      "1,2, Antonio Gomez1,2, Samuel Collombet\n",
      "\n",
      "1,2*\n",
      "\n",
      "1,2,3,7*\n",
      "\n",
      "Chromosomal architecture is known to influence gene expression, yet its role in controlling cell fate remains poorly understood. Reprogramming of somatic cells into pluripotent stem cells (PSCs) by the transcription factors (TFs) OCT4, SOX2,\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "\n",
      "et al. Rapid chromatin switch in the direct repro- gramming of ﬁbroblasts to neurons. Cell Rep. 20, 3236–3247 (2017).\n",
      "\n",
      "42. Dall’Agnese, A. et al. Transcription factor-directed re-wiring of\n",
      "\n",
      "chromatin architecture for somatic cell nuclear reprogramming toward trans-differentiation. Mol. Cell 76, 453–472 (2019). 43. Stone, N. R. et al. Context-speciﬁc transcription factor functions\n",
      "\n",
      "regulate epigenomic and transcriptional dynamics during cardiac reprogramming. Cell Stem Cell 25, 87–102 (2019).\n",
      "\n",
      "44. Nam, Y. J. et al. Reprogramming of human ﬁbroblasts toward a cardiac fate. Proc. Natl Acad. Sci. USA 110, 5588–5593 (2013). 45. Zhou, Y. et al. Single-cell transcriptomic analyses of cell fate\n",
      "\n",
      "2, Peter D. Adams Pier Lorenzo Puri Alessandra Sacco1, Peter Andersen 2 & Alexandre R. Colas\n",
      "\n",
      "1,6, Michael S. Yu1,\n",
      "\n",
      "4, Li Qian 5,\n",
      "\n",
      "1\n",
      "\n",
      "Check for updates\n",
      "\n",
      "Deﬁning the mechanisms safeguarding cell fate identity in differentiated cells is crucial to improve 1) - our understanding of how differentiation is main- tained in healthy tissues or altered in a disease state, and 2) - our ability to use cell fate reprogramming for regenerative purposes. Here, using a genome- wide transcription factor screen followed by validation steps in a variety of reprogramming assays (cardiac, neural and iPSC in ﬁbroblasts and endothelial cells), we identiﬁed a set of four transcription factors (ATF7IP, JUNB, SP7,\n",
      "\n",
      "Ann. N.Y. Acad. Sci. ISSN 0077-8923\n",
      "\n",
      "A N N A L S O F T H E N E W Y O R K A C A D E M Y O F SC I E N C E S Issue: Hematopoietic Stem Cells IX\n",
      "\n",
      "Converting cell fates: generating hematopoietic stem cells de novo via transcription factor reprogramming\n",
      "\n",
      "Michael G. Daniel,1,2,3 Ihor R. Lemischka,1,2,4,∗ and Kateri Moore1,2,∗ 1Department of Developmental and Regenerative Biology, Icahn School of Medicine, New York, New York. 2Black Family Stem Cell Institute, Icahn School of Medicine, New York, New York. 3The Graduate School of Biomedical Science, Icahn School of Medicine, New York, New York. 4Department of Pharmacology and Systems Therapeutics, Icahn School of Medicine, New York, New York\n",
      "\n",
      "Address\n",
      "\n",
      "Cahan, P. et al. CellNet: network biology applied to stem cell engineering. Cell 158, 903–915 (2014).\n",
      "\n",
      "10. Rackham, O. J. et al. A predictive computational framework for direct reprogramming between human cell types. Nat. Genet. 48, 331–335 (2016).\n",
      "\n",
      "11. D’Alessio, A. C. et al. A systematic approach to identify candidate trans-\n",
      "\n",
      "cription factors that control cell identity. Stem Cell Rep. 5, 763–775 (2015). 12. Lambert, S. A. et al. The human transcription factors. Cell 175, 598–599 (2018). 13. Nakatake, Y. et al. Generation and profiling of 2,135 human ESC lines for the systematic analyses of cell states perturbed by inducing single transcription factors. Cell Rep. 31, 107655 (2020).\n",
      "\n",
      "14.\n",
      "\n",
      "Cahan, P. et al. CellNet: network biology applied to stem cell engineering. Cell 158, 903–915 (2014).\n",
      "\n",
      "10. Rackham, O. J. et al. A predictive computational framework for direct reprogramming between human cell types. Nat. Genet. 48, 331–335 (2016).\n",
      "\n",
      "11. D’Alessio, A. C. et al. A systematic approach to identify candidate trans-\n",
      "\n",
      "cription factors that control cell identity. Stem Cell Rep. 5, 763–775 (2015). 12. Lambert, S. A. et al. The human transcription factors. Cell 175, 598–599 (2018). 13. Nakatake, Y. et al. Generation and profiling of 2,135 human ESC lines for the systematic analyses of cell states perturbed by inducing single transcription factors. Cell Rep. 31, 107655 (2020).\n",
      "\n",
      "14.\n",
      "\n",
      "If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the first iPSCs.\n",
      "\n",
      "Direct cell conversion\n",
      "\n",
      "+\n",
      "\n",
      "MYOD\n",
      "\n",
      "Fibroblast\n",
      "\n",
      "+\n",
      "\n",
      "GATA1\n",
      "\n",
      "Myeloblast\n",
      "\n",
      "+\n",
      "\n",
      "CEBP\n",
      "\n",
      "B lymphocyte\n",
      "\n",
      "Diﬀerentiated\n",
      "\n",
      "Transdiﬀerentiation\n",
      "\n",
      "Myoblast\n",
      "\n",
      "Megakaryocyte, eosinophil, erythrocyte\n",
      "\n",
      "Macrophage\n",
      "\n",
      "Identifying candidate reprogramming factors. According to previous experiments, genes that contribute to the identities of fertilized eggs or ES cells were thought to be, in large part, responsible for the reconstitution of specific transcriptional networks during reprogramming. Thus, in 2000, we\n",
      "Human: what transcription factors are involved in cell reprogramming of stem cells?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "Transcription factors orchestrate dynamic interplay between genome topology and gene regulation during cell reprogramming\n",
      "\n",
      "Ralph Stadhouders  François Le Dily1,2,3, Javier Quilez  Yasmina Cuartero1,2,3, Jochen Hecht2,6, Guillaume J. Filion1,2, Miguel Beato1,2, Marc A. Marti-Renom  and Thomas Graf\n",
      "\n",
      "1,2,8*, Enrique Vidal\n",
      "\n",
      "1,2, François Serra\n",
      "\n",
      "1,2,3, Bruno Di Stefano\n",
      "\n",
      "1,2,4,  5, Clara Berenguer1,2,\n",
      "\n",
      "1,2, Antonio Gomez1,2, Samuel Collombet\n",
      "\n",
      "1,2*\n",
      "\n",
      "1,2,3,7*\n",
      "\n",
      "Chromosomal architecture is known to influence gene expression, yet its role in controlling cell fate remains poorly understood. Reprogramming of somatic cells into pluripotent stem cells (PSCs) by the transcription factors (TFs) OCT4, SOX2,\n",
      "\n",
      "equally to this manuscript.\n",
      "\n",
      "doi: 10.1111/nyas.12989\n",
      "\n",
      "24\n",
      "\n",
      "Ann. N.Y. Acad. Sci. 1370 (2016) 24–35 C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "Figure 1. Established TF reprogramming strategies. TF reprogramming has been shown to work across a variety of cell types. This ﬁgure shows the starting and end cell populations, as well as the TF cocktails used to derive each cell type of choice. Adapted from Ref. 25.\n",
      "\n",
      "The novel technology of induced pluripotent stem cells (iPSCs) from Yamanaka and colleagues instigated the entire ﬁeld of regenerative medicine. Using only four transcription factors (TFs), Oct3/4, Sox2, Klf4, and c-Myc,\n",
      "\n",
      "et al. Rapid chromatin switch in the direct repro- gramming of ﬁbroblasts to neurons. Cell Rep. 20, 3236–3247 (2017).\n",
      "\n",
      "42. Dall’Agnese, A. et al. Transcription factor-directed re-wiring of\n",
      "\n",
      "chromatin architecture for somatic cell nuclear reprogramming toward trans-differentiation. Mol. Cell 76, 453–472 (2019). 43. Stone, N. R. et al. Context-speciﬁc transcription factor functions\n",
      "\n",
      "regulate epigenomic and transcriptional dynamics during cardiac reprogramming. Cell Stem Cell 25, 87–102 (2019).\n",
      "\n",
      "44. Nam, Y. J. et al. Reprogramming of human ﬁbroblasts toward a cardiac fate. Proc. Natl Acad. Sci. USA 110, 5588–5593 (2013). 45. Zhou, Y. et al. Single-cell transcriptomic analyses of cell fate\n",
      "\n",
      "2, Peter D. Adams Pier Lorenzo Puri Alessandra Sacco1, Peter Andersen 2 & Alexandre R. Colas\n",
      "\n",
      "1,6, Michael S. Yu1,\n",
      "\n",
      "4, Li Qian 5,\n",
      "\n",
      "1\n",
      "\n",
      "Check for updates\n",
      "\n",
      "Deﬁning the mechanisms safeguarding cell fate identity in differentiated cells is crucial to improve 1) - our understanding of how differentiation is main- tained in healthy tissues or altered in a disease state, and 2) - our ability to use cell fate reprogramming for regenerative purposes. Here, using a genome- wide transcription factor screen followed by validation steps in a variety of reprogramming assays (cardiac, neural and iPSC in ﬁbroblasts and endothelial cells), we identiﬁed a set of four transcription factors (ATF7IP, JUNB, SP7,\n",
      "\n",
      "Ann. N.Y. Acad. Sci. ISSN 0077-8923\n",
      "\n",
      "A N N A L S O F T H E N E W Y O R K A C A D E M Y O F SC I E N C E S Issue: Hematopoietic Stem Cells IX\n",
      "\n",
      "Converting cell fates: generating hematopoietic stem cells de novo via transcription factor reprogramming\n",
      "\n",
      "Michael G. Daniel,1,2,3 Ihor R. Lemischka,1,2,4,∗ and Kateri Moore1,2,∗ 1Department of Developmental and Regenerative Biology, Icahn School of Medicine, New York, New York. 2Black Family Stem Cell Institute, Icahn School of Medicine, New York, New York. 3The Graduate School of Biomedical Science, Icahn School of Medicine, New York, New York. 4Department of Pharmacology and Systems Therapeutics, Icahn School of Medicine, New York, New York\n",
      "\n",
      "Address\n",
      "\n",
      "Cahan, P. et al. CellNet: network biology applied to stem cell engineering. Cell 158, 903–915 (2014).\n",
      "\n",
      "10. Rackham, O. J. et al. A predictive computational framework for direct reprogramming between human cell types. Nat. Genet. 48, 331–335 (2016).\n",
      "\n",
      "11. D’Alessio, A. C. et al. A systematic approach to identify candidate trans-\n",
      "\n",
      "cription factors that control cell identity. Stem Cell Rep. 5, 763–775 (2015). 12. Lambert, S. A. et al. The human transcription factors. Cell 175, 598–599 (2018). 13. Nakatake, Y. et al. Generation and profiling of 2,135 human ESC lines for the systematic analyses of cell states perturbed by inducing single transcription factors. Cell Rep. 31, 107655 (2020).\n",
      "\n",
      "14.\n",
      "\n",
      "Cahan, P. et al. CellNet: network biology applied to stem cell engineering. Cell 158, 903–915 (2014).\n",
      "\n",
      "10. Rackham, O. J. et al. A predictive computational framework for direct reprogramming between human cell types. Nat. Genet. 48, 331–335 (2016).\n",
      "\n",
      "11. D’Alessio, A. C. et al. A systematic approach to identify candidate trans-\n",
      "\n",
      "cription factors that control cell identity. Stem Cell Rep. 5, 763–775 (2015). 12. Lambert, S. A. et al. The human transcription factors. Cell 175, 598–599 (2018). 13. Nakatake, Y. et al. Generation and profiling of 2,135 human ESC lines for the systematic analyses of cell states perturbed by inducing single transcription factors. Cell Rep. 31, 107655 (2020).\n",
      "\n",
      "14.\n",
      "\n",
      "If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the first iPSCs.\n",
      "\n",
      "Direct cell conversion\n",
      "\n",
      "+\n",
      "\n",
      "MYOD\n",
      "\n",
      "Fibroblast\n",
      "\n",
      "+\n",
      "\n",
      "GATA1\n",
      "\n",
      "Myeloblast\n",
      "\n",
      "+\n",
      "\n",
      "CEBP\n",
      "\n",
      "B lymphocyte\n",
      "\n",
      "Diﬀerentiated\n",
      "\n",
      "Transdiﬀerentiation\n",
      "\n",
      "Myoblast\n",
      "\n",
      "Megakaryocyte, eosinophil, erythrocyte\n",
      "\n",
      "Macrophage\n",
      "\n",
      "Identifying candidate reprogramming factors. According to previous experiments, genes that contribute to the identities of fertilized eggs or ES cells were thought to be, in large part, responsible for the reconstitution of specific transcriptional networks during reprogramming. Thus, in 2000, we\n",
      "Human: what transcription factors are involved in cell reprogramming of stem cells?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "Transcription factors orchestrate dynamic interplay between genome topology and gene regulation during cell reprogramming\n",
      "\n",
      "Ralph Stadhouders  François Le Dily1,2,3, Javier Quilez  Yasmina Cuartero1,2,3, Jochen Hecht2,6, Guillaume J. Filion1,2, Miguel Beato1,2, Marc A. Marti-Renom  and Thomas Graf\n",
      "\n",
      "1,2,8*, Enrique Vidal\n",
      "\n",
      "1,2, François Serra\n",
      "\n",
      "1,2,3, Bruno Di Stefano\n",
      "\n",
      "1,2,4,  5, Clara Berenguer1,2,\n",
      "\n",
      "1,2, Antonio Gomez1,2, Samuel Collombet\n",
      "\n",
      "1,2*\n",
      "\n",
      "1,2,3,7*\n",
      "\n",
      "Chromosomal architecture is known to influence gene expression, yet its role in controlling cell fate remains poorly understood. Reprogramming of somatic cells into pluripotent stem cells (PSCs) by the transcription factors (TFs) OCT4, SOX2,\n",
      "\n",
      "equally to this manuscript.\n",
      "\n",
      "doi: 10.1111/nyas.12989\n",
      "\n",
      "24\n",
      "\n",
      "Ann. N.Y. Acad. Sci. 1370 (2016) 24–35 C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "Figure 1. Established TF reprogramming strategies. TF reprogramming has been shown to work across a variety of cell types. This ﬁgure shows the starting and end cell populations, as well as the TF cocktails used to derive each cell type of choice. Adapted from Ref. 25.\n",
      "\n",
      "The novel technology of induced pluripotent stem cells (iPSCs) from Yamanaka and colleagues instigated the entire ﬁeld of regenerative medicine. Using only four transcription factors (TFs), Oct3/4, Sox2, Klf4, and c-Myc,\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "\n",
      "2, Peter D. Adams Pier Lorenzo Puri Alessandra Sacco1, Peter Andersen 2 & Alexandre R. Colas\n",
      "\n",
      "1,6, Michael S. Yu1,\n",
      "\n",
      "4, Li Qian 5,\n",
      "\n",
      "1\n",
      "\n",
      "Check for updates\n",
      "\n",
      "Deﬁning the mechanisms safeguarding cell fate identity in differentiated cells is crucial to improve 1) - our understanding of how differentiation is main- tained in healthy tissues or altered in a disease state, and 2) - our ability to use cell fate reprogramming for regenerative purposes. Here, using a genome- wide transcription factor screen followed by validation steps in a variety of reprogramming assays (cardiac, neural and iPSC in ﬁbroblasts and endothelial cells), we identiﬁed a set of four transcription factors (ATF7IP, JUNB, SP7,\n",
      "\n",
      "Ann. N.Y. Acad. Sci. ISSN 0077-8923\n",
      "\n",
      "A N N A L S O F T H E N E W Y O R K A C A D E M Y O F SC I E N C E S Issue: Hematopoietic Stem Cells IX\n",
      "\n",
      "Converting cell fates: generating hematopoietic stem cells de novo via transcription factor reprogramming\n",
      "\n",
      "Michael G. Daniel,1,2,3 Ihor R. Lemischka,1,2,4,∗ and Kateri Moore1,2,∗ 1Department of Developmental and Regenerative Biology, Icahn School of Medicine, New York, New York. 2Black Family Stem Cell Institute, Icahn School of Medicine, New York, New York. 3The Graduate School of Biomedical Science, Icahn School of Medicine, New York, New York. 4Department of Pharmacology and Systems Therapeutics, Icahn School of Medicine, New York, New York\n",
      "\n",
      "Address\n",
      "\n",
      "Cahan, P. et al. CellNet: network biology applied to stem cell engineering. Cell 158, 903–915 (2014).\n",
      "\n",
      "10. Rackham, O. J. et al. A predictive computational framework for direct reprogramming between human cell types. Nat. Genet. 48, 331–335 (2016).\n",
      "\n",
      "11. D’Alessio, A. C. et al. A systematic approach to identify candidate trans-\n",
      "\n",
      "cription factors that control cell identity. Stem Cell Rep. 5, 763–775 (2015). 12. Lambert, S. A. et al. The human transcription factors. Cell 175, 598–599 (2018). 13. Nakatake, Y. et al. Generation and profiling of 2,135 human ESC lines for the systematic analyses of cell states perturbed by inducing single transcription factors. Cell Rep. 31, 107655 (2020).\n",
      "\n",
      "14.\n",
      "\n",
      "Cahan, P. et al. CellNet: network biology applied to stem cell engineering. Cell 158, 903–915 (2014).\n",
      "\n",
      "10. Rackham, O. J. et al. A predictive computational framework for direct reprogramming between human cell types. Nat. Genet. 48, 331–335 (2016).\n",
      "\n",
      "11. D’Alessio, A. C. et al. A systematic approach to identify candidate trans-\n",
      "\n",
      "cription factors that control cell identity. Stem Cell Rep. 5, 763–775 (2015). 12. Lambert, S. A. et al. The human transcription factors. Cell 175, 598–599 (2018). 13. Nakatake, Y. et al. Generation and profiling of 2,135 human ESC lines for the systematic analyses of cell states perturbed by inducing single transcription factors. Cell Rep. 31, 107655 (2020).\n",
      "\n",
      "14.\n",
      "\n",
      "If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the first iPSCs.\n",
      "\n",
      "Direct cell conversion\n",
      "\n",
      "+\n",
      "\n",
      "MYOD\n",
      "\n",
      "Fibroblast\n",
      "\n",
      "+\n",
      "\n",
      "GATA1\n",
      "\n",
      "Myeloblast\n",
      "\n",
      "+\n",
      "\n",
      "CEBP\n",
      "\n",
      "B lymphocyte\n",
      "\n",
      "Diﬀerentiated\n",
      "\n",
      "Transdiﬀerentiation\n",
      "\n",
      "Myoblast\n",
      "\n",
      "Megakaryocyte, eosinophil, erythrocyte\n",
      "\n",
      "Macrophage\n",
      "\n",
      "Identifying candidate reprogramming factors. According to previous experiments, genes that contribute to the identities of fertilized eggs or ES cells were thought to be, in large part, responsible for the reconstitution of specific transcriptional networks during reprogramming. Thus, in 2000, we\n",
      "Human: what transcription factors are involved in cell reprogramming of stem cells?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "Transcription factors orchestrate dynamic interplay between genome topology and gene regulation during cell reprogramming\n",
      "\n",
      "Ralph Stadhouders  François Le Dily1,2,3, Javier Quilez  Yasmina Cuartero1,2,3, Jochen Hecht2,6, Guillaume J. Filion1,2, Miguel Beato1,2, Marc A. Marti-Renom  and Thomas Graf\n",
      "\n",
      "1,2,8*, Enrique Vidal\n",
      "\n",
      "1,2, François Serra\n",
      "\n",
      "1,2,3, Bruno Di Stefano\n",
      "\n",
      "1,2,4,  5, Clara Berenguer1,2,\n",
      "\n",
      "1,2, Antonio Gomez1,2, Samuel Collombet\n",
      "\n",
      "1,2*\n",
      "\n",
      "1,2,3,7*\n",
      "\n",
      "Chromosomal architecture is known to influence gene expression, yet its role in controlling cell fate remains poorly understood. Reprogramming of somatic cells into pluripotent stem cells (PSCs) by the transcription factors (TFs) OCT4, SOX2,\n",
      "\n",
      "equally to this manuscript.\n",
      "\n",
      "doi: 10.1111/nyas.12989\n",
      "\n",
      "24\n",
      "\n",
      "Ann. N.Y. Acad. Sci. 1370 (2016) 24–35 C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "Figure 1. Established TF reprogramming strategies. TF reprogramming has been shown to work across a variety of cell types. This ﬁgure shows the starting and end cell populations, as well as the TF cocktails used to derive each cell type of choice. Adapted from Ref. 25.\n",
      "\n",
      "The novel technology of induced pluripotent stem cells (iPSCs) from Yamanaka and colleagues instigated the entire ﬁeld of regenerative medicine. Using only four transcription factors (TFs), Oct3/4, Sox2, Klf4, and c-Myc,\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "\n",
      "et al. Rapid chromatin switch in the direct repro- gramming of ﬁbroblasts to neurons. Cell Rep. 20, 3236–3247 (2017).\n",
      "\n",
      "42. Dall’Agnese, A. et al. Transcription factor-directed re-wiring of\n",
      "\n",
      "chromatin architecture for somatic cell nuclear reprogramming toward trans-differentiation. Mol. Cell 76, 453–472 (2019). 43. Stone, N. R. et al. Context-speciﬁc transcription factor functions\n",
      "\n",
      "regulate epigenomic and transcriptional dynamics during cardiac reprogramming. Cell Stem Cell 25, 87–102 (2019).\n",
      "\n",
      "44. Nam, Y. J. et al. Reprogramming of human ﬁbroblasts toward a cardiac fate. Proc. Natl Acad. Sci. USA 110, 5588–5593 (2013). 45. Zhou, Y. et al. Single-cell transcriptomic analyses of cell fate\n",
      "\n",
      "Ann. N.Y. Acad. Sci. ISSN 0077-8923\n",
      "\n",
      "A N N A L S O F T H E N E W Y O R K A C A D E M Y O F SC I E N C E S Issue: Hematopoietic Stem Cells IX\n",
      "\n",
      "Converting cell fates: generating hematopoietic stem cells de novo via transcription factor reprogramming\n",
      "\n",
      "Michael G. Daniel,1,2,3 Ihor R. Lemischka,1,2,4,∗ and Kateri Moore1,2,∗ 1Department of Developmental and Regenerative Biology, Icahn School of Medicine, New York, New York. 2Black Family Stem Cell Institute, Icahn School of Medicine, New York, New York. 3The Graduate School of Biomedical Science, Icahn School of Medicine, New York, New York. 4Department of Pharmacology and Systems Therapeutics, Icahn School of Medicine, New York, New York\n",
      "\n",
      "Address\n",
      "\n",
      "Cahan, P. et al. CellNet: network biology applied to stem cell engineering. Cell 158, 903–915 (2014).\n",
      "\n",
      "10. Rackham, O. J. et al. A predictive computational framework for direct reprogramming between human cell types. Nat. Genet. 48, 331–335 (2016).\n",
      "\n",
      "11. D’Alessio, A. C. et al. A systematic approach to identify candidate trans-\n",
      "\n",
      "cription factors that control cell identity. Stem Cell Rep. 5, 763–775 (2015). 12. Lambert, S. A. et al. The human transcription factors. Cell 175, 598–599 (2018). 13. Nakatake, Y. et al. Generation and profiling of 2,135 human ESC lines for the systematic analyses of cell states perturbed by inducing single transcription factors. Cell Rep. 31, 107655 (2020).\n",
      "\n",
      "14.\n",
      "\n",
      "Cahan, P. et al. CellNet: network biology applied to stem cell engineering. Cell 158, 903–915 (2014).\n",
      "\n",
      "10. Rackham, O. J. et al. A predictive computational framework for direct reprogramming between human cell types. Nat. Genet. 48, 331–335 (2016).\n",
      "\n",
      "11. D’Alessio, A. C. et al. A systematic approach to identify candidate trans-\n",
      "\n",
      "cription factors that control cell identity. Stem Cell Rep. 5, 763–775 (2015). 12. Lambert, S. A. et al. The human transcription factors. Cell 175, 598–599 (2018). 13. Nakatake, Y. et al. Generation and profiling of 2,135 human ESC lines for the systematic analyses of cell states perturbed by inducing single transcription factors. Cell Rep. 31, 107655 (2020).\n",
      "\n",
      "14.\n",
      "\n",
      "If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the first iPSCs.\n",
      "\n",
      "Direct cell conversion\n",
      "\n",
      "+\n",
      "\n",
      "MYOD\n",
      "\n",
      "Fibroblast\n",
      "\n",
      "+\n",
      "\n",
      "GATA1\n",
      "\n",
      "Myeloblast\n",
      "\n",
      "+\n",
      "\n",
      "CEBP\n",
      "\n",
      "B lymphocyte\n",
      "\n",
      "Diﬀerentiated\n",
      "\n",
      "Transdiﬀerentiation\n",
      "\n",
      "Myoblast\n",
      "\n",
      "Megakaryocyte, eosinophil, erythrocyte\n",
      "\n",
      "Macrophage\n",
      "\n",
      "Identifying candidate reprogramming factors. According to previous experiments, genes that contribute to the identities of fertilized eggs or ES cells were thought to be, in large part, responsible for the reconstitution of specific transcriptional networks during reprogramming. Thus, in 2000, we\n",
      "Human: what transcription factors are involved in cell reprogramming of stem cells?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "Transcription factors orchestrate dynamic interplay between genome topology and gene regulation during cell reprogramming\n",
      "\n",
      "Ralph Stadhouders  François Le Dily1,2,3, Javier Quilez  Yasmina Cuartero1,2,3, Jochen Hecht2,6, Guillaume J. Filion1,2, Miguel Beato1,2, Marc A. Marti-Renom  and Thomas Graf\n",
      "\n",
      "1,2,8*, Enrique Vidal\n",
      "\n",
      "1,2, François Serra\n",
      "\n",
      "1,2,3, Bruno Di Stefano\n",
      "\n",
      "1,2,4,  5, Clara Berenguer1,2,\n",
      "\n",
      "1,2, Antonio Gomez1,2, Samuel Collombet\n",
      "\n",
      "1,2*\n",
      "\n",
      "1,2,3,7*\n",
      "\n",
      "Chromosomal architecture is known to influence gene expression, yet its role in controlling cell fate remains poorly understood. Reprogramming of somatic cells into pluripotent stem cells (PSCs) by the transcription factors (TFs) OCT4, SOX2,\n",
      "\n",
      "equally to this manuscript.\n",
      "\n",
      "doi: 10.1111/nyas.12989\n",
      "\n",
      "24\n",
      "\n",
      "Ann. N.Y. Acad. Sci. 1370 (2016) 24–35 C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "Figure 1. Established TF reprogramming strategies. TF reprogramming has been shown to work across a variety of cell types. This ﬁgure shows the starting and end cell populations, as well as the TF cocktails used to derive each cell type of choice. Adapted from Ref. 25.\n",
      "\n",
      "The novel technology of induced pluripotent stem cells (iPSCs) from Yamanaka and colleagues instigated the entire ﬁeld of regenerative medicine. Using only four transcription factors (TFs), Oct3/4, Sox2, Klf4, and c-Myc,\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "\n",
      "et al. Rapid chromatin switch in the direct repro- gramming of ﬁbroblasts to neurons. Cell Rep. 20, 3236–3247 (2017).\n",
      "\n",
      "42. Dall’Agnese, A. et al. Transcription factor-directed re-wiring of\n",
      "\n",
      "chromatin architecture for somatic cell nuclear reprogramming toward trans-differentiation. Mol. Cell 76, 453–472 (2019). 43. Stone, N. R. et al. Context-speciﬁc transcription factor functions\n",
      "\n",
      "regulate epigenomic and transcriptional dynamics during cardiac reprogramming. Cell Stem Cell 25, 87–102 (2019).\n",
      "\n",
      "44. Nam, Y. J. et al. Reprogramming of human ﬁbroblasts toward a cardiac fate. Proc. Natl Acad. Sci. USA 110, 5588–5593 (2013). 45. Zhou, Y. et al. Single-cell transcriptomic analyses of cell fate\n",
      "\n",
      "2, Peter D. Adams Pier Lorenzo Puri Alessandra Sacco1, Peter Andersen 2 & Alexandre R. Colas\n",
      "\n",
      "1,6, Michael S. Yu1,\n",
      "\n",
      "4, Li Qian 5,\n",
      "\n",
      "1\n",
      "\n",
      "Check for updates\n",
      "\n",
      "Deﬁning the mechanisms safeguarding cell fate identity in differentiated cells is crucial to improve 1) - our understanding of how differentiation is main- tained in healthy tissues or altered in a disease state, and 2) - our ability to use cell fate reprogramming for regenerative purposes. Here, using a genome- wide transcription factor screen followed by validation steps in a variety of reprogramming assays (cardiac, neural and iPSC in ﬁbroblasts and endothelial cells), we identiﬁed a set of four transcription factors (ATF7IP, JUNB, SP7,\n",
      "\n",
      "Cahan, P. et al. CellNet: network biology applied to stem cell engineering. Cell 158, 903–915 (2014).\n",
      "\n",
      "10. Rackham, O. J. et al. A predictive computational framework for direct reprogramming between human cell types. Nat. Genet. 48, 331–335 (2016).\n",
      "\n",
      "11. D’Alessio, A. C. et al. A systematic approach to identify candidate trans-\n",
      "\n",
      "cription factors that control cell identity. Stem Cell Rep. 5, 763–775 (2015). 12. Lambert, S. A. et al. The human transcription factors. Cell 175, 598–599 (2018). 13. Nakatake, Y. et al. Generation and profiling of 2,135 human ESC lines for the systematic analyses of cell states perturbed by inducing single transcription factors. Cell Rep. 31, 107655 (2020).\n",
      "\n",
      "14.\n",
      "\n",
      "Cahan, P. et al. CellNet: network biology applied to stem cell engineering. Cell 158, 903–915 (2014).\n",
      "\n",
      "10. Rackham, O. J. et al. A predictive computational framework for direct reprogramming between human cell types. Nat. Genet. 48, 331–335 (2016).\n",
      "\n",
      "11. D’Alessio, A. C. et al. A systematic approach to identify candidate trans-\n",
      "\n",
      "cription factors that control cell identity. Stem Cell Rep. 5, 763–775 (2015). 12. Lambert, S. A. et al. The human transcription factors. Cell 175, 598–599 (2018). 13. Nakatake, Y. et al. Generation and profiling of 2,135 human ESC lines for the systematic analyses of cell states perturbed by inducing single transcription factors. Cell Rep. 31, 107655 (2020).\n",
      "\n",
      "14.\n",
      "\n",
      "If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the first iPSCs.\n",
      "\n",
      "Direct cell conversion\n",
      "\n",
      "+\n",
      "\n",
      "MYOD\n",
      "\n",
      "Fibroblast\n",
      "\n",
      "+\n",
      "\n",
      "GATA1\n",
      "\n",
      "Myeloblast\n",
      "\n",
      "+\n",
      "\n",
      "CEBP\n",
      "\n",
      "B lymphocyte\n",
      "\n",
      "Diﬀerentiated\n",
      "\n",
      "Transdiﬀerentiation\n",
      "\n",
      "Myoblast\n",
      "\n",
      "Megakaryocyte, eosinophil, erythrocyte\n",
      "\n",
      "Macrophage\n",
      "\n",
      "Identifying candidate reprogramming factors. According to previous experiments, genes that contribute to the identities of fertilized eggs or ES cells were thought to be, in large part, responsible for the reconstitution of specific transcriptional networks during reprogramming. Thus, in 2000, we\n",
      "Human: what transcription factors are involved in cell reprogramming of stem cells?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "Transcription factors orchestrate dynamic interplay between genome topology and gene regulation during cell reprogramming\n",
      "\n",
      "Ralph Stadhouders  François Le Dily1,2,3, Javier Quilez  Yasmina Cuartero1,2,3, Jochen Hecht2,6, Guillaume J. Filion1,2, Miguel Beato1,2, Marc A. Marti-Renom  and Thomas Graf\n",
      "\n",
      "1,2,8*, Enrique Vidal\n",
      "\n",
      "1,2, François Serra\n",
      "\n",
      "1,2,3, Bruno Di Stefano\n",
      "\n",
      "1,2,4,  5, Clara Berenguer1,2,\n",
      "\n",
      "1,2, Antonio Gomez1,2, Samuel Collombet\n",
      "\n",
      "1,2*\n",
      "\n",
      "1,2,3,7*\n",
      "\n",
      "Chromosomal architecture is known to influence gene expression, yet its role in controlling cell fate remains poorly understood. Reprogramming of somatic cells into pluripotent stem cells (PSCs) by the transcription factors (TFs) OCT4, SOX2,\n",
      "\n",
      "equally to this manuscript.\n",
      "\n",
      "doi: 10.1111/nyas.12989\n",
      "\n",
      "24\n",
      "\n",
      "Ann. N.Y. Acad. Sci. 1370 (2016) 24–35 C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "Figure 1. Established TF reprogramming strategies. TF reprogramming has been shown to work across a variety of cell types. This ﬁgure shows the starting and end cell populations, as well as the TF cocktails used to derive each cell type of choice. Adapted from Ref. 25.\n",
      "\n",
      "The novel technology of induced pluripotent stem cells (iPSCs) from Yamanaka and colleagues instigated the entire ﬁeld of regenerative medicine. Using only four transcription factors (TFs), Oct3/4, Sox2, Klf4, and c-Myc,\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "\n",
      "et al. Rapid chromatin switch in the direct repro- gramming of ﬁbroblasts to neurons. Cell Rep. 20, 3236–3247 (2017).\n",
      "\n",
      "42. Dall’Agnese, A. et al. Transcription factor-directed re-wiring of\n",
      "\n",
      "chromatin architecture for somatic cell nuclear reprogramming toward trans-differentiation. Mol. Cell 76, 453–472 (2019). 43. Stone, N. R. et al. Context-speciﬁc transcription factor functions\n",
      "\n",
      "regulate epigenomic and transcriptional dynamics during cardiac reprogramming. Cell Stem Cell 25, 87–102 (2019).\n",
      "\n",
      "44. Nam, Y. J. et al. Reprogramming of human ﬁbroblasts toward a cardiac fate. Proc. Natl Acad. Sci. USA 110, 5588–5593 (2013). 45. Zhou, Y. et al. Single-cell transcriptomic analyses of cell fate\n",
      "\n",
      "2, Peter D. Adams Pier Lorenzo Puri Alessandra Sacco1, Peter Andersen 2 & Alexandre R. Colas\n",
      "\n",
      "1,6, Michael S. Yu1,\n",
      "\n",
      "4, Li Qian 5,\n",
      "\n",
      "1\n",
      "\n",
      "Check for updates\n",
      "\n",
      "Deﬁning the mechanisms safeguarding cell fate identity in differentiated cells is crucial to improve 1) - our understanding of how differentiation is main- tained in healthy tissues or altered in a disease state, and 2) - our ability to use cell fate reprogramming for regenerative purposes. Here, using a genome- wide transcription factor screen followed by validation steps in a variety of reprogramming assays (cardiac, neural and iPSC in ﬁbroblasts and endothelial cells), we identiﬁed a set of four transcription factors (ATF7IP, JUNB, SP7,\n",
      "\n",
      "Ann. N.Y. Acad. Sci. ISSN 0077-8923\n",
      "\n",
      "A N N A L S O F T H E N E W Y O R K A C A D E M Y O F SC I E N C E S Issue: Hematopoietic Stem Cells IX\n",
      "\n",
      "Converting cell fates: generating hematopoietic stem cells de novo via transcription factor reprogramming\n",
      "\n",
      "Michael G. Daniel,1,2,3 Ihor R. Lemischka,1,2,4,∗ and Kateri Moore1,2,∗ 1Department of Developmental and Regenerative Biology, Icahn School of Medicine, New York, New York. 2Black Family Stem Cell Institute, Icahn School of Medicine, New York, New York. 3The Graduate School of Biomedical Science, Icahn School of Medicine, New York, New York. 4Department of Pharmacology and Systems Therapeutics, Icahn School of Medicine, New York, New York\n",
      "\n",
      "Address\n",
      "\n",
      "Cahan, P. et al. CellNet: network biology applied to stem cell engineering. Cell 158, 903–915 (2014).\n",
      "\n",
      "10. Rackham, O. J. et al. A predictive computational framework for direct reprogramming between human cell types. Nat. Genet. 48, 331–335 (2016).\n",
      "\n",
      "11. D’Alessio, A. C. et al. A systematic approach to identify candidate trans-\n",
      "\n",
      "cription factors that control cell identity. Stem Cell Rep. 5, 763–775 (2015). 12. Lambert, S. A. et al. The human transcription factors. Cell 175, 598–599 (2018). 13. Nakatake, Y. et al. Generation and profiling of 2,135 human ESC lines for the systematic analyses of cell states perturbed by inducing single transcription factors. Cell Rep. 31, 107655 (2020).\n",
      "\n",
      "14.\n",
      "\n",
      "If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the first iPSCs.\n",
      "\n",
      "Direct cell conversion\n",
      "\n",
      "+\n",
      "\n",
      "MYOD\n",
      "\n",
      "Fibroblast\n",
      "\n",
      "+\n",
      "\n",
      "GATA1\n",
      "\n",
      "Myeloblast\n",
      "\n",
      "+\n",
      "\n",
      "CEBP\n",
      "\n",
      "B lymphocyte\n",
      "\n",
      "Diﬀerentiated\n",
      "\n",
      "Transdiﬀerentiation\n",
      "\n",
      "Myoblast\n",
      "\n",
      "Megakaryocyte, eosinophil, erythrocyte\n",
      "\n",
      "Macrophage\n",
      "\n",
      "Identifying candidate reprogramming factors. According to previous experiments, genes that contribute to the identities of fertilized eggs or ES cells were thought to be, in large part, responsible for the reconstitution of specific transcriptional networks during reprogramming. Thus, in 2000, we\n",
      "Human: what transcription factors are involved in cell reprogramming of stem cells?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "Transcription factors orchestrate dynamic interplay between genome topology and gene regulation during cell reprogramming\n",
      "\n",
      "Ralph Stadhouders  François Le Dily1,2,3, Javier Quilez  Yasmina Cuartero1,2,3, Jochen Hecht2,6, Guillaume J. Filion1,2, Miguel Beato1,2, Marc A. Marti-Renom  and Thomas Graf\n",
      "\n",
      "1,2,8*, Enrique Vidal\n",
      "\n",
      "1,2, François Serra\n",
      "\n",
      "1,2,3, Bruno Di Stefano\n",
      "\n",
      "1,2,4,  5, Clara Berenguer1,2,\n",
      "\n",
      "1,2, Antonio Gomez1,2, Samuel Collombet\n",
      "\n",
      "1,2*\n",
      "\n",
      "1,2,3,7*\n",
      "\n",
      "Chromosomal architecture is known to influence gene expression, yet its role in controlling cell fate remains poorly understood. Reprogramming of somatic cells into pluripotent stem cells (PSCs) by the transcription factors (TFs) OCT4, SOX2,\n",
      "\n",
      "equally to this manuscript.\n",
      "\n",
      "doi: 10.1111/nyas.12989\n",
      "\n",
      "24\n",
      "\n",
      "Ann. N.Y. Acad. Sci. 1370 (2016) 24–35 C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "Figure 1. Established TF reprogramming strategies. TF reprogramming has been shown to work across a variety of cell types. This ﬁgure shows the starting and end cell populations, as well as the TF cocktails used to derive each cell type of choice. Adapted from Ref. 25.\n",
      "\n",
      "The novel technology of induced pluripotent stem cells (iPSCs) from Yamanaka and colleagues instigated the entire ﬁeld of regenerative medicine. Using only four transcription factors (TFs), Oct3/4, Sox2, Klf4, and c-Myc,\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "\n",
      "et al. Rapid chromatin switch in the direct repro- gramming of ﬁbroblasts to neurons. Cell Rep. 20, 3236–3247 (2017).\n",
      "\n",
      "42. Dall’Agnese, A. et al. Transcription factor-directed re-wiring of\n",
      "\n",
      "chromatin architecture for somatic cell nuclear reprogramming toward trans-differentiation. Mol. Cell 76, 453–472 (2019). 43. Stone, N. R. et al. Context-speciﬁc transcription factor functions\n",
      "\n",
      "regulate epigenomic and transcriptional dynamics during cardiac reprogramming. Cell Stem Cell 25, 87–102 (2019).\n",
      "\n",
      "44. Nam, Y. J. et al. Reprogramming of human ﬁbroblasts toward a cardiac fate. Proc. Natl Acad. Sci. USA 110, 5588–5593 (2013). 45. Zhou, Y. et al. Single-cell transcriptomic analyses of cell fate\n",
      "\n",
      "2, Peter D. Adams Pier Lorenzo Puri Alessandra Sacco1, Peter Andersen 2 & Alexandre R. Colas\n",
      "\n",
      "1,6, Michael S. Yu1,\n",
      "\n",
      "4, Li Qian 5,\n",
      "\n",
      "1\n",
      "\n",
      "Check for updates\n",
      "\n",
      "Deﬁning the mechanisms safeguarding cell fate identity in differentiated cells is crucial to improve 1) - our understanding of how differentiation is main- tained in healthy tissues or altered in a disease state, and 2) - our ability to use cell fate reprogramming for regenerative purposes. Here, using a genome- wide transcription factor screen followed by validation steps in a variety of reprogramming assays (cardiac, neural and iPSC in ﬁbroblasts and endothelial cells), we identiﬁed a set of four transcription factors (ATF7IP, JUNB, SP7,\n",
      "\n",
      "Ann. N.Y. Acad. Sci. ISSN 0077-8923\n",
      "\n",
      "A N N A L S O F T H E N E W Y O R K A C A D E M Y O F SC I E N C E S Issue: Hematopoietic Stem Cells IX\n",
      "\n",
      "Converting cell fates: generating hematopoietic stem cells de novo via transcription factor reprogramming\n",
      "\n",
      "Michael G. Daniel,1,2,3 Ihor R. Lemischka,1,2,4,∗ and Kateri Moore1,2,∗ 1Department of Developmental and Regenerative Biology, Icahn School of Medicine, New York, New York. 2Black Family Stem Cell Institute, Icahn School of Medicine, New York, New York. 3The Graduate School of Biomedical Science, Icahn School of Medicine, New York, New York. 4Department of Pharmacology and Systems Therapeutics, Icahn School of Medicine, New York, New York\n",
      "\n",
      "Address\n",
      "\n",
      "Cahan, P. et al. CellNet: network biology applied to stem cell engineering. Cell 158, 903–915 (2014).\n",
      "\n",
      "10. Rackham, O. J. et al. A predictive computational framework for direct reprogramming between human cell types. Nat. Genet. 48, 331–335 (2016).\n",
      "\n",
      "11. D’Alessio, A. C. et al. A systematic approach to identify candidate trans-\n",
      "\n",
      "cription factors that control cell identity. Stem Cell Rep. 5, 763–775 (2015). 12. Lambert, S. A. et al. The human transcription factors. Cell 175, 598–599 (2018). 13. Nakatake, Y. et al. Generation and profiling of 2,135 human ESC lines for the systematic analyses of cell states perturbed by inducing single transcription factors. Cell Rep. 31, 107655 (2020).\n",
      "\n",
      "14.\n",
      "\n",
      "If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the first iPSCs.\n",
      "\n",
      "Direct cell conversion\n",
      "\n",
      "+\n",
      "\n",
      "MYOD\n",
      "\n",
      "Fibroblast\n",
      "\n",
      "+\n",
      "\n",
      "GATA1\n",
      "\n",
      "Myeloblast\n",
      "\n",
      "+\n",
      "\n",
      "CEBP\n",
      "\n",
      "B lymphocyte\n",
      "\n",
      "Diﬀerentiated\n",
      "\n",
      "Transdiﬀerentiation\n",
      "\n",
      "Myoblast\n",
      "\n",
      "Megakaryocyte, eosinophil, erythrocyte\n",
      "\n",
      "Macrophage\n",
      "\n",
      "Identifying candidate reprogramming factors. According to previous experiments, genes that contribute to the identities of fertilized eggs or ES cells were thought to be, in large part, responsible for the reconstitution of specific transcriptional networks during reprogramming. Thus, in 2000, we\n",
      "Human: what transcription factors are involved in cell reprogramming of stem cells?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "Transcription factors orchestrate dynamic interplay between genome topology and gene regulation during cell reprogramming\n",
      "\n",
      "Ralph Stadhouders  François Le Dily1,2,3, Javier Quilez  Yasmina Cuartero1,2,3, Jochen Hecht2,6, Guillaume J. Filion1,2, Miguel Beato1,2, Marc A. Marti-Renom  and Thomas Graf\n",
      "\n",
      "1,2,8*, Enrique Vidal\n",
      "\n",
      "1,2, François Serra\n",
      "\n",
      "1,2,3, Bruno Di Stefano\n",
      "\n",
      "1,2,4,  5, Clara Berenguer1,2,\n",
      "\n",
      "1,2, Antonio Gomez1,2, Samuel Collombet\n",
      "\n",
      "1,2*\n",
      "\n",
      "1,2,3,7*\n",
      "\n",
      "Chromosomal architecture is known to influence gene expression, yet its role in controlling cell fate remains poorly understood. Reprogramming of somatic cells into pluripotent stem cells (PSCs) by the transcription factors (TFs) OCT4, SOX2,\n",
      "\n",
      "equally to this manuscript.\n",
      "\n",
      "doi: 10.1111/nyas.12989\n",
      "\n",
      "24\n",
      "\n",
      "Ann. N.Y. Acad. Sci. 1370 (2016) 24–35 C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "Figure 1. Established TF reprogramming strategies. TF reprogramming has been shown to work across a variety of cell types. This ﬁgure shows the starting and end cell populations, as well as the TF cocktails used to derive each cell type of choice. Adapted from Ref. 25.\n",
      "\n",
      "The novel technology of induced pluripotent stem cells (iPSCs) from Yamanaka and colleagues instigated the entire ﬁeld of regenerative medicine. Using only four transcription factors (TFs), Oct3/4, Sox2, Klf4, and c-Myc,\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "\n",
      "et al. Rapid chromatin switch in the direct repro- gramming of ﬁbroblasts to neurons. Cell Rep. 20, 3236–3247 (2017).\n",
      "\n",
      "42. Dall’Agnese, A. et al. Transcription factor-directed re-wiring of\n",
      "\n",
      "chromatin architecture for somatic cell nuclear reprogramming toward trans-differentiation. Mol. Cell 76, 453–472 (2019). 43. Stone, N. R. et al. Context-speciﬁc transcription factor functions\n",
      "\n",
      "regulate epigenomic and transcriptional dynamics during cardiac reprogramming. Cell Stem Cell 25, 87–102 (2019).\n",
      "\n",
      "44. Nam, Y. J. et al. Reprogramming of human ﬁbroblasts toward a cardiac fate. Proc. Natl Acad. Sci. USA 110, 5588–5593 (2013). 45. Zhou, Y. et al. Single-cell transcriptomic analyses of cell fate\n",
      "\n",
      "2, Peter D. Adams Pier Lorenzo Puri Alessandra Sacco1, Peter Andersen 2 & Alexandre R. Colas\n",
      "\n",
      "1,6, Michael S. Yu1,\n",
      "\n",
      "4, Li Qian 5,\n",
      "\n",
      "1\n",
      "\n",
      "Check for updates\n",
      "\n",
      "Deﬁning the mechanisms safeguarding cell fate identity in differentiated cells is crucial to improve 1) - our understanding of how differentiation is main- tained in healthy tissues or altered in a disease state, and 2) - our ability to use cell fate reprogramming for regenerative purposes. Here, using a genome- wide transcription factor screen followed by validation steps in a variety of reprogramming assays (cardiac, neural and iPSC in ﬁbroblasts and endothelial cells), we identiﬁed a set of four transcription factors (ATF7IP, JUNB, SP7,\n",
      "\n",
      "Ann. N.Y. Acad. Sci. ISSN 0077-8923\n",
      "\n",
      "A N N A L S O F T H E N E W Y O R K A C A D E M Y O F SC I E N C E S Issue: Hematopoietic Stem Cells IX\n",
      "\n",
      "Converting cell fates: generating hematopoietic stem cells de novo via transcription factor reprogramming\n",
      "\n",
      "Michael G. Daniel,1,2,3 Ihor R. Lemischka,1,2,4,∗ and Kateri Moore1,2,∗ 1Department of Developmental and Regenerative Biology, Icahn School of Medicine, New York, New York. 2Black Family Stem Cell Institute, Icahn School of Medicine, New York, New York. 3The Graduate School of Biomedical Science, Icahn School of Medicine, New York, New York. 4Department of Pharmacology and Systems Therapeutics, Icahn School of Medicine, New York, New York\n",
      "\n",
      "Address\n",
      "\n",
      "Cahan, P. et al. CellNet: network biology applied to stem cell engineering. Cell 158, 903–915 (2014).\n",
      "\n",
      "10. Rackham, O. J. et al. A predictive computational framework for direct reprogramming between human cell types. Nat. Genet. 48, 331–335 (2016).\n",
      "\n",
      "11. D’Alessio, A. C. et al. A systematic approach to identify candidate trans-\n",
      "\n",
      "cription factors that control cell identity. Stem Cell Rep. 5, 763–775 (2015). 12. Lambert, S. A. et al. The human transcription factors. Cell 175, 598–599 (2018). 13. Nakatake, Y. et al. Generation and profiling of 2,135 human ESC lines for the systematic analyses of cell states perturbed by inducing single transcription factors. Cell Rep. 31, 107655 (2020).\n",
      "\n",
      "14.\n",
      "\n",
      "Cahan, P. et al. CellNet: network biology applied to stem cell engineering. Cell 158, 903–915 (2014).\n",
      "\n",
      "10. Rackham, O. J. et al. A predictive computational framework for direct reprogramming between human cell types. Nat. Genet. 48, 331–335 (2016).\n",
      "\n",
      "11. D’Alessio, A. C. et al. A systematic approach to identify candidate trans-\n",
      "\n",
      "cription factors that control cell identity. Stem Cell Rep. 5, 763–775 (2015). 12. Lambert, S. A. et al. The human transcription factors. Cell 175, 598–599 (2018). 13. Nakatake, Y. et al. Generation and profiling of 2,135 human ESC lines for the systematic analyses of cell states perturbed by inducing single transcription factors. Cell Rep. 31, 107655 (2020).\n",
      "\n",
      "14.\n",
      "Human: what transcription factors are involved in cell reprogramming of stem cells?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "==================================================\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  What are the differences between direct lineage reprogramming and induced pluripotency?\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fri Jun  7 23:18:03 2024 INFO RAG\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "RAG >> 3: \n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "cell types, makes it chal- lenging to generate multiple lineages in the same culture.\n",
      "\n",
      "An alternative approach is direct activation of TFs, which can both shorten the time for cell conversion and isolate it from other developmental events. TF induction has been shown\n",
      "\n",
      "to transdifferentiate cells between lineages1, reprogram somatic cells to pluripotency2 and differentiate stem cells3. Previous studies on applying TFs for cell fate engineering have used either experi- mental testing of a limited set of developmentally relevant TFs4–8 or computational prediction of TFs inferred from genome-scale datasets with modest accuracy9–11. However, the field has lacked a global experimental study of\n",
      "\n",
      "acquisition of pluripotency during somatic cell transdifferentiation with iPSC repro- gramming factors. Nat. Biotechnol. 33: 769–774.\n",
      "\n",
      "73. Bar-Nur, O. et al. 2015. Lineage conversion induced by pluripotency factors involves transient passage through an iPSC stage. Nat. Biotechnol. 33: 761–768.\n",
      "\n",
      "74. Szabo, E. et al. 2010. Direct conversion of human ﬁbroblasts\n",
      "\n",
      "to multilineage blood progenitors. Nature 468: 521–526.\n",
      "\n",
      "75. Mitchell, R. et al. 2014. Molecular evidence for OCT4- induced plasticity in adult human ﬁbroblasts required for direct cell fate conversion to lineage speciﬁc progenitors. Stem Cells 32: 2178–2187.\n",
      "\n",
      "76. Pulecio, J. et al. 2014. Conversion of human ﬁbroblasts into\n",
      "\n",
      "Cell Stem Cell 1, 55–70 (2007). 161. Park, I. H. et al. Reprogramming of human somatic\n",
      "\n",
      "cells to pluripotency with defined factors. Nature 451, 141–146 (2008).\n",
      "\n",
      "162. Hanna, J. et al. Direct cell reprogramming is a stochastic process amenable to acceleration. Nature 462, 595–601 (2009).\n",
      "\n",
      "163. Yang, J. et al. Stat3 activation is limiting for reprogramming to ground state pluripotency. Cell Stem Cell 7, 319–328 (2010).\n",
      "\n",
      "164. Silva, J. et al. Promotion of reprogramming to ground state pluripotency by signal inhibition. PLoS Biol. 6, e253 (2008).\n",
      "\n",
      "Acknowledgements The authors would like to thank all Yamanaka laboratory members, past and present, and are also grateful to Y. Miyake, R. Kato, E.\n",
      "\n",
      "seminal discovery of induced PSCs (iPSCs) (FIG. 2), which are pluripotent cells that are derived from somatic cells, and that functionally bear a striking resemblance to ES cells. The discovery of iPSCs has also paved the way for direct cell fate conversion (also known as transdifferentiation) experiments, in which the expression of sets\n",
      "\n",
      "of tissue‑specific transcription factors can induce a different fate, across germ layers and by bypassing the pluripotent state, opening new doors in the field of cellular reprogramming. iPSCs are promising for many applications, including drug discovery, disease modelling and regenerative medicine23 (FIG. 2).\n",
      "\n",
      "Somatic cell nuclear transfer\n",
      "\n",
      "and adult ﬁbroblast cultures by deﬁned factors. Cell 126, 663–676 (2006). 4.\n",
      "\n",
      "5. Xu, J., Du, Y. & Deng, H. Direct lineage reprogramming: strategies,\n",
      "\n",
      "mechanisms, and applications. Cell Stem Cell 16, 119–134 (2015). Takahashi, K. et al. Induction of pluripotent stem cells from adult human ﬁbroblasts by deﬁned factors. Cell 131, 861–872 (2007). Vierbuchen, T. et al. Direct conversion of ﬁbroblasts to functional neurons by deﬁned factors. Nature 463, 1035–1041 (2010). Ieda, M. et al. Direct reprogramming of ﬁbroblasts into functional cardiomyocytes by deﬁned factors. Cell 142, 375–386 (2010). 9. Haridhasapavalan, K. K., Raina, K., Dey, C., Adhikari, P. & Thummer,\n",
      "\n",
      "6.\n",
      "\n",
      "7.\n",
      "\n",
      "8.\n",
      "\n",
      "R. P. An insight\n",
      "\n",
      "generation of induced pluripotent stem cells (iPSCs; left). It is now known that fully differentiated cells can also undergo reprogram- ming. In fact, many cells within the population initiate the process; however, owing to the fact that reprogramming is complex and comprises many events, the majority of cells never complete the process, and only a small number become iPSCs (right). b | Reprogramming can potentially occur either in a deterministic (top) or a stochastic (bottom) fashion. In the deter- ministic process, all cells follow exactly the same path towards pluripotency, and thus cells are reprogrammed with a fixed latency. In the stochastic repro- gramming process, transitions\n",
      "\n",
      "iPSCs, as well as modification of reporter lines to facilitate differentiation towards specific cell types. The finding that it is possible to modulate cell identity has also introduced the possibility of direct reprogramming, in which cell fate conversion is achieved by the overexpression of multiple transcription factors without reverting to a pluripotent state.\n",
      "\n",
      "Nevertheless, some important issues regarding the generation of iPSCs remain unsolved. In pre‑implantation embryos, PSCs are referred to as ‘naive’, and they become ‘primed’ during post‑implantation development. Mouse ES cells and iPSCs have homogeneous characteristics, resembling cells of the early epiblasts of mouse\n",
      "\n",
      "(2007).\n",
      "\n",
      "42. Yu, J. et al. Human induced pluripotent stem cells free of vector and transgene sequences. Science 324, 797–801 (2009).\n",
      "\n",
      "43. Tanabe, K., Nakamura, M., Narita, M., Takahashi, K. & Yamanaka, S. Maturation, not initiation, is the major roadblock during reprogramming toward pluripotency from human fibroblasts. Proc. Natl Acad. Sci. USA 110, 12172–12179 (2013). 44. Nakagawa, M. et al. Generation of induced pluripotent stem cells without Myc from mouse and human fibroblasts. Nat. Biotechnol. 26, 101–106 (2008).\n",
      "\n",
      "45. Jiang, J. et al. A core Klf circuitry regulates self‑\n",
      "\n",
      "renewal of embryonic stem cells. Nat. Cell Biol. 10, 353–360 (2008).\n",
      "\n",
      "46. Nakagawa, M., Takizawa, N., Narita, M.,\n",
      "\n",
      "L. Qian, Y. Hayashi, P. Delgado-Olguin, S. Ding, B. G. Bruneau, D. Srivastava, Direct reprogramming of human fibroblasts toward a . Stem Cell Reports cardiomyocyte-like state. ​\n",
      "\n",
      "1\n",
      "\n",
      ", 235–247 (2013).\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "12. L. Qian, E. C. Berry, J.-D. Fu, M. Ieda, D. Srivastava, Reprogramming of mouse fibroblasts Nat. Protoc. ​ 8\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "13. Y.-J. Nam, K. Song, X. Luo, E. Daniel, K. Lambeth, K. West, J. A. Hill, J. M. DiMaio, L. A.\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in\n",
      "Human: What are the differences between direct lineage reprogramming and induced pluripotency?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "cell types, makes it chal- lenging to generate multiple lineages in the same culture.\n",
      "\n",
      "An alternative approach is direct activation of TFs, which can both shorten the time for cell conversion and isolate it from other developmental events. TF induction has been shown\n",
      "\n",
      "to transdifferentiate cells between lineages1, reprogram somatic cells to pluripotency2 and differentiate stem cells3. Previous studies on applying TFs for cell fate engineering have used either experi- mental testing of a limited set of developmentally relevant TFs4–8 or computational prediction of TFs inferred from genome-scale datasets with modest accuracy9–11. However, the field has lacked a global experimental study of\n",
      "\n",
      "acquisition of pluripotency during somatic cell transdifferentiation with iPSC repro- gramming factors. Nat. Biotechnol. 33: 769–774.\n",
      "\n",
      "73. Bar-Nur, O. et al. 2015. Lineage conversion induced by pluripotency factors involves transient passage through an iPSC stage. Nat. Biotechnol. 33: 761–768.\n",
      "\n",
      "74. Szabo, E. et al. 2010. Direct conversion of human ﬁbroblasts\n",
      "\n",
      "to multilineage blood progenitors. Nature 468: 521–526.\n",
      "\n",
      "75. Mitchell, R. et al. 2014. Molecular evidence for OCT4- induced plasticity in adult human ﬁbroblasts required for direct cell fate conversion to lineage speciﬁc progenitors. Stem Cells 32: 2178–2187.\n",
      "\n",
      "76. Pulecio, J. et al. 2014. Conversion of human ﬁbroblasts into\n",
      "\n",
      "Cell Stem Cell 1, 55–70 (2007). 161. Park, I. H. et al. Reprogramming of human somatic\n",
      "\n",
      "cells to pluripotency with defined factors. Nature 451, 141–146 (2008).\n",
      "\n",
      "162. Hanna, J. et al. Direct cell reprogramming is a stochastic process amenable to acceleration. Nature 462, 595–601 (2009).\n",
      "\n",
      "163. Yang, J. et al. Stat3 activation is limiting for reprogramming to ground state pluripotency. Cell Stem Cell 7, 319–328 (2010).\n",
      "\n",
      "164. Silva, J. et al. Promotion of reprogramming to ground state pluripotency by signal inhibition. PLoS Biol. 6, e253 (2008).\n",
      "\n",
      "Acknowledgements The authors would like to thank all Yamanaka laboratory members, past and present, and are also grateful to Y. Miyake, R. Kato, E.\n",
      "\n",
      "seminal discovery of induced PSCs (iPSCs) (FIG. 2), which are pluripotent cells that are derived from somatic cells, and that functionally bear a striking resemblance to ES cells. The discovery of iPSCs has also paved the way for direct cell fate conversion (also known as transdifferentiation) experiments, in which the expression of sets\n",
      "\n",
      "of tissue‑specific transcription factors can induce a different fate, across germ layers and by bypassing the pluripotent state, opening new doors in the field of cellular reprogramming. iPSCs are promising for many applications, including drug discovery, disease modelling and regenerative medicine23 (FIG. 2).\n",
      "\n",
      "Somatic cell nuclear transfer\n",
      "\n",
      "and adult ﬁbroblast cultures by deﬁned factors. Cell 126, 663–676 (2006). 4.\n",
      "\n",
      "5. Xu, J., Du, Y. & Deng, H. Direct lineage reprogramming: strategies,\n",
      "\n",
      "mechanisms, and applications. Cell Stem Cell 16, 119–134 (2015). Takahashi, K. et al. Induction of pluripotent stem cells from adult human ﬁbroblasts by deﬁned factors. Cell 131, 861–872 (2007). Vierbuchen, T. et al. Direct conversion of ﬁbroblasts to functional neurons by deﬁned factors. Nature 463, 1035–1041 (2010). Ieda, M. et al. Direct reprogramming of ﬁbroblasts into functional cardiomyocytes by deﬁned factors. Cell 142, 375–386 (2010). 9. Haridhasapavalan, K. K., Raina, K., Dey, C., Adhikari, P. & Thummer,\n",
      "\n",
      "6.\n",
      "\n",
      "7.\n",
      "\n",
      "8.\n",
      "\n",
      "R. P. An insight\n",
      "\n",
      "generation of induced pluripotent stem cells (iPSCs; left). It is now known that fully differentiated cells can also undergo reprogram- ming. In fact, many cells within the population initiate the process; however, owing to the fact that reprogramming is complex and comprises many events, the majority of cells never complete the process, and only a small number become iPSCs (right). b | Reprogramming can potentially occur either in a deterministic (top) or a stochastic (bottom) fashion. In the deter- ministic process, all cells follow exactly the same path towards pluripotency, and thus cells are reprogrammed with a fixed latency. In the stochastic repro- gramming process, transitions\n",
      "\n",
      "iPSCs, as well as modification of reporter lines to facilitate differentiation towards specific cell types. The finding that it is possible to modulate cell identity has also introduced the possibility of direct reprogramming, in which cell fate conversion is achieved by the overexpression of multiple transcription factors without reverting to a pluripotent state.\n",
      "\n",
      "Nevertheless, some important issues regarding the generation of iPSCs remain unsolved. In pre‑implantation embryos, PSCs are referred to as ‘naive’, and they become ‘primed’ during post‑implantation development. Mouse ES cells and iPSCs have homogeneous characteristics, resembling cells of the early epiblasts of mouse\n",
      "\n",
      "(2007).\n",
      "\n",
      "42. Yu, J. et al. Human induced pluripotent stem cells free of vector and transgene sequences. Science 324, 797–801 (2009).\n",
      "\n",
      "43. Tanabe, K., Nakamura, M., Narita, M., Takahashi, K. & Yamanaka, S. Maturation, not initiation, is the major roadblock during reprogramming toward pluripotency from human fibroblasts. Proc. Natl Acad. Sci. USA 110, 12172–12179 (2013). 44. Nakagawa, M. et al. Generation of induced pluripotent stem cells without Myc from mouse and human fibroblasts. Nat. Biotechnol. 26, 101–106 (2008).\n",
      "\n",
      "45. Jiang, J. et al. A core Klf circuitry regulates self‑\n",
      "\n",
      "renewal of embryonic stem cells. Nat. Cell Biol. 10, 353–360 (2008).\n",
      "\n",
      "46. Nakagawa, M., Takizawa, N., Narita, M.,\n",
      "\n",
      "L. Qian, Y. Hayashi, P. Delgado-Olguin, S. Ding, B. G. Bruneau, D. Srivastava, Direct reprogramming of human fibroblasts toward a . Stem Cell Reports cardiomyocyte-like state. ​\n",
      "\n",
      "1\n",
      "\n",
      ", 235–247 (2013).\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "12. L. Qian, E. C. Berry, J.-D. Fu, M. Ieda, D. Srivastava, Reprogramming of mouse fibroblasts Nat. Protoc. ​ 8\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "13. Y.-J. Nam, K. Song, X. Luo, E. Daniel, K. Lambeth, K. West, J. A. Hill, J. M. DiMaio, L. A.\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in\n",
      "Human: What are the differences between direct lineage reprogramming and induced pluripotency?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "cell types, makes it chal- lenging to generate multiple lineages in the same culture.\n",
      "\n",
      "An alternative approach is direct activation of TFs, which can both shorten the time for cell conversion and isolate it from other developmental events. TF induction has been shown\n",
      "\n",
      "to transdifferentiate cells between lineages1, reprogram somatic cells to pluripotency2 and differentiate stem cells3. Previous studies on applying TFs for cell fate engineering have used either experi- mental testing of a limited set of developmentally relevant TFs4–8 or computational prediction of TFs inferred from genome-scale datasets with modest accuracy9–11. However, the field has lacked a global experimental study of\n",
      "\n",
      "cell types, makes it chal- lenging to generate multiple lineages in the same culture.\n",
      "\n",
      "An alternative approach is direct activation of TFs, which can both shorten the time for cell conversion and isolate it from other developmental events. TF induction has been shown\n",
      "\n",
      "to transdifferentiate cells between lineages1, reprogram somatic cells to pluripotency2 and differentiate stem cells3. Previous studies on applying TFs for cell fate engineering have used either experi- mental testing of a limited set of developmentally relevant TFs4–8 or computational prediction of TFs inferred from genome-scale datasets with modest accuracy9–11. However, the field has lacked a global experimental study of\n",
      "\n",
      "Cell Stem Cell 1, 55–70 (2007). 161. Park, I. H. et al. Reprogramming of human somatic\n",
      "\n",
      "cells to pluripotency with defined factors. Nature 451, 141–146 (2008).\n",
      "\n",
      "162. Hanna, J. et al. Direct cell reprogramming is a stochastic process amenable to acceleration. Nature 462, 595–601 (2009).\n",
      "\n",
      "163. Yang, J. et al. Stat3 activation is limiting for reprogramming to ground state pluripotency. Cell Stem Cell 7, 319–328 (2010).\n",
      "\n",
      "164. Silva, J. et al. Promotion of reprogramming to ground state pluripotency by signal inhibition. PLoS Biol. 6, e253 (2008).\n",
      "\n",
      "Acknowledgements The authors would like to thank all Yamanaka laboratory members, past and present, and are also grateful to Y. Miyake, R. Kato, E.\n",
      "\n",
      "seminal discovery of induced PSCs (iPSCs) (FIG. 2), which are pluripotent cells that are derived from somatic cells, and that functionally bear a striking resemblance to ES cells. The discovery of iPSCs has also paved the way for direct cell fate conversion (also known as transdifferentiation) experiments, in which the expression of sets\n",
      "\n",
      "of tissue‑specific transcription factors can induce a different fate, across germ layers and by bypassing the pluripotent state, opening new doors in the field of cellular reprogramming. iPSCs are promising for many applications, including drug discovery, disease modelling and regenerative medicine23 (FIG. 2).\n",
      "\n",
      "Somatic cell nuclear transfer\n",
      "\n",
      "and adult ﬁbroblast cultures by deﬁned factors. Cell 126, 663–676 (2006). 4.\n",
      "\n",
      "5. Xu, J., Du, Y. & Deng, H. Direct lineage reprogramming: strategies,\n",
      "\n",
      "mechanisms, and applications. Cell Stem Cell 16, 119–134 (2015). Takahashi, K. et al. Induction of pluripotent stem cells from adult human ﬁbroblasts by deﬁned factors. Cell 131, 861–872 (2007). Vierbuchen, T. et al. Direct conversion of ﬁbroblasts to functional neurons by deﬁned factors. Nature 463, 1035–1041 (2010). Ieda, M. et al. Direct reprogramming of ﬁbroblasts into functional cardiomyocytes by deﬁned factors. Cell 142, 375–386 (2010). 9. Haridhasapavalan, K. K., Raina, K., Dey, C., Adhikari, P. & Thummer,\n",
      "\n",
      "6.\n",
      "\n",
      "7.\n",
      "\n",
      "8.\n",
      "\n",
      "R. P. An insight\n",
      "\n",
      "generation of induced pluripotent stem cells (iPSCs; left). It is now known that fully differentiated cells can also undergo reprogram- ming. In fact, many cells within the population initiate the process; however, owing to the fact that reprogramming is complex and comprises many events, the majority of cells never complete the process, and only a small number become iPSCs (right). b | Reprogramming can potentially occur either in a deterministic (top) or a stochastic (bottom) fashion. In the deter- ministic process, all cells follow exactly the same path towards pluripotency, and thus cells are reprogrammed with a fixed latency. In the stochastic repro- gramming process, transitions\n",
      "\n",
      "iPSCs, as well as modification of reporter lines to facilitate differentiation towards specific cell types. The finding that it is possible to modulate cell identity has also introduced the possibility of direct reprogramming, in which cell fate conversion is achieved by the overexpression of multiple transcription factors without reverting to a pluripotent state.\n",
      "\n",
      "Nevertheless, some important issues regarding the generation of iPSCs remain unsolved. In pre‑implantation embryos, PSCs are referred to as ‘naive’, and they become ‘primed’ during post‑implantation development. Mouse ES cells and iPSCs have homogeneous characteristics, resembling cells of the early epiblasts of mouse\n",
      "\n",
      "(2007).\n",
      "\n",
      "42. Yu, J. et al. Human induced pluripotent stem cells free of vector and transgene sequences. Science 324, 797–801 (2009).\n",
      "\n",
      "43. Tanabe, K., Nakamura, M., Narita, M., Takahashi, K. & Yamanaka, S. Maturation, not initiation, is the major roadblock during reprogramming toward pluripotency from human fibroblasts. Proc. Natl Acad. Sci. USA 110, 12172–12179 (2013). 44. Nakagawa, M. et al. Generation of induced pluripotent stem cells without Myc from mouse and human fibroblasts. Nat. Biotechnol. 26, 101–106 (2008).\n",
      "\n",
      "45. Jiang, J. et al. A core Klf circuitry regulates self‑\n",
      "\n",
      "renewal of embryonic stem cells. Nat. Cell Biol. 10, 353–360 (2008).\n",
      "\n",
      "46. Nakagawa, M., Takizawa, N., Narita, M.,\n",
      "\n",
      "L. Qian, Y. Hayashi, P. Delgado-Olguin, S. Ding, B. G. Bruneau, D. Srivastava, Direct reprogramming of human fibroblasts toward a . Stem Cell Reports cardiomyocyte-like state. ​\n",
      "\n",
      "1\n",
      "\n",
      ", 235–247 (2013).\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "12. L. Qian, E. C. Berry, J.-D. Fu, M. Ieda, D. Srivastava, Reprogramming of mouse fibroblasts Nat. Protoc. ​ 8\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "13. Y.-J. Nam, K. Song, X. Luo, E. Daniel, K. Lambeth, K. West, J. A. Hill, J. M. DiMaio, L. A.\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in\n",
      "Human: What are the differences between direct lineage reprogramming and induced pluripotency?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "cell types, makes it chal- lenging to generate multiple lineages in the same culture.\n",
      "\n",
      "An alternative approach is direct activation of TFs, which can both shorten the time for cell conversion and isolate it from other developmental events. TF induction has been shown\n",
      "\n",
      "to transdifferentiate cells between lineages1, reprogram somatic cells to pluripotency2 and differentiate stem cells3. Previous studies on applying TFs for cell fate engineering have used either experi- mental testing of a limited set of developmentally relevant TFs4–8 or computational prediction of TFs inferred from genome-scale datasets with modest accuracy9–11. However, the field has lacked a global experimental study of\n",
      "\n",
      "cell types, makes it chal- lenging to generate multiple lineages in the same culture.\n",
      "\n",
      "An alternative approach is direct activation of TFs, which can both shorten the time for cell conversion and isolate it from other developmental events. TF induction has been shown\n",
      "\n",
      "to transdifferentiate cells between lineages1, reprogram somatic cells to pluripotency2 and differentiate stem cells3. Previous studies on applying TFs for cell fate engineering have used either experi- mental testing of a limited set of developmentally relevant TFs4–8 or computational prediction of TFs inferred from genome-scale datasets with modest accuracy9–11. However, the field has lacked a global experimental study of\n",
      "\n",
      "acquisition of pluripotency during somatic cell transdifferentiation with iPSC repro- gramming factors. Nat. Biotechnol. 33: 769–774.\n",
      "\n",
      "73. Bar-Nur, O. et al. 2015. Lineage conversion induced by pluripotency factors involves transient passage through an iPSC stage. Nat. Biotechnol. 33: 761–768.\n",
      "\n",
      "74. Szabo, E. et al. 2010. Direct conversion of human ﬁbroblasts\n",
      "\n",
      "to multilineage blood progenitors. Nature 468: 521–526.\n",
      "\n",
      "75. Mitchell, R. et al. 2014. Molecular evidence for OCT4- induced plasticity in adult human ﬁbroblasts required for direct cell fate conversion to lineage speciﬁc progenitors. Stem Cells 32: 2178–2187.\n",
      "\n",
      "76. Pulecio, J. et al. 2014. Conversion of human ﬁbroblasts into\n",
      "\n",
      "seminal discovery of induced PSCs (iPSCs) (FIG. 2), which are pluripotent cells that are derived from somatic cells, and that functionally bear a striking resemblance to ES cells. The discovery of iPSCs has also paved the way for direct cell fate conversion (also known as transdifferentiation) experiments, in which the expression of sets\n",
      "\n",
      "of tissue‑specific transcription factors can induce a different fate, across germ layers and by bypassing the pluripotent state, opening new doors in the field of cellular reprogramming. iPSCs are promising for many applications, including drug discovery, disease modelling and regenerative medicine23 (FIG. 2).\n",
      "\n",
      "Somatic cell nuclear transfer\n",
      "\n",
      "and adult ﬁbroblast cultures by deﬁned factors. Cell 126, 663–676 (2006). 4.\n",
      "\n",
      "5. Xu, J., Du, Y. & Deng, H. Direct lineage reprogramming: strategies,\n",
      "\n",
      "mechanisms, and applications. Cell Stem Cell 16, 119–134 (2015). Takahashi, K. et al. Induction of pluripotent stem cells from adult human ﬁbroblasts by deﬁned factors. Cell 131, 861–872 (2007). Vierbuchen, T. et al. Direct conversion of ﬁbroblasts to functional neurons by deﬁned factors. Nature 463, 1035–1041 (2010). Ieda, M. et al. Direct reprogramming of ﬁbroblasts into functional cardiomyocytes by deﬁned factors. Cell 142, 375–386 (2010). 9. Haridhasapavalan, K. K., Raina, K., Dey, C., Adhikari, P. & Thummer,\n",
      "\n",
      "6.\n",
      "\n",
      "7.\n",
      "\n",
      "8.\n",
      "\n",
      "R. P. An insight\n",
      "\n",
      "generation of induced pluripotent stem cells (iPSCs; left). It is now known that fully differentiated cells can also undergo reprogram- ming. In fact, many cells within the population initiate the process; however, owing to the fact that reprogramming is complex and comprises many events, the majority of cells never complete the process, and only a small number become iPSCs (right). b | Reprogramming can potentially occur either in a deterministic (top) or a stochastic (bottom) fashion. In the deter- ministic process, all cells follow exactly the same path towards pluripotency, and thus cells are reprogrammed with a fixed latency. In the stochastic repro- gramming process, transitions\n",
      "\n",
      "iPSCs, as well as modification of reporter lines to facilitate differentiation towards specific cell types. The finding that it is possible to modulate cell identity has also introduced the possibility of direct reprogramming, in which cell fate conversion is achieved by the overexpression of multiple transcription factors without reverting to a pluripotent state.\n",
      "\n",
      "Nevertheless, some important issues regarding the generation of iPSCs remain unsolved. In pre‑implantation embryos, PSCs are referred to as ‘naive’, and they become ‘primed’ during post‑implantation development. Mouse ES cells and iPSCs have homogeneous characteristics, resembling cells of the early epiblasts of mouse\n",
      "\n",
      "(2007).\n",
      "\n",
      "42. Yu, J. et al. Human induced pluripotent stem cells free of vector and transgene sequences. Science 324, 797–801 (2009).\n",
      "\n",
      "43. Tanabe, K., Nakamura, M., Narita, M., Takahashi, K. & Yamanaka, S. Maturation, not initiation, is the major roadblock during reprogramming toward pluripotency from human fibroblasts. Proc. Natl Acad. Sci. USA 110, 12172–12179 (2013). 44. Nakagawa, M. et al. Generation of induced pluripotent stem cells without Myc from mouse and human fibroblasts. Nat. Biotechnol. 26, 101–106 (2008).\n",
      "\n",
      "45. Jiang, J. et al. A core Klf circuitry regulates self‑\n",
      "\n",
      "renewal of embryonic stem cells. Nat. Cell Biol. 10, 353–360 (2008).\n",
      "\n",
      "46. Nakagawa, M., Takizawa, N., Narita, M.,\n",
      "\n",
      "L. Qian, Y. Hayashi, P. Delgado-Olguin, S. Ding, B. G. Bruneau, D. Srivastava, Direct reprogramming of human fibroblasts toward a . Stem Cell Reports cardiomyocyte-like state. ​\n",
      "\n",
      "1\n",
      "\n",
      ", 235–247 (2013).\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "12. L. Qian, E. C. Berry, J.-D. Fu, M. Ieda, D. Srivastava, Reprogramming of mouse fibroblasts Nat. Protoc. ​ 8\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "13. Y.-J. Nam, K. Song, X. Luo, E. Daniel, K. Lambeth, K. West, J. A. Hill, J. M. DiMaio, L. A.\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in\n",
      "Human: What are the differences between direct lineage reprogramming and induced pluripotency?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "cell types, makes it chal- lenging to generate multiple lineages in the same culture.\n",
      "\n",
      "An alternative approach is direct activation of TFs, which can both shorten the time for cell conversion and isolate it from other developmental events. TF induction has been shown\n",
      "\n",
      "to transdifferentiate cells between lineages1, reprogram somatic cells to pluripotency2 and differentiate stem cells3. Previous studies on applying TFs for cell fate engineering have used either experi- mental testing of a limited set of developmentally relevant TFs4–8 or computational prediction of TFs inferred from genome-scale datasets with modest accuracy9–11. However, the field has lacked a global experimental study of\n",
      "\n",
      "cell types, makes it chal- lenging to generate multiple lineages in the same culture.\n",
      "\n",
      "An alternative approach is direct activation of TFs, which can both shorten the time for cell conversion and isolate it from other developmental events. TF induction has been shown\n",
      "\n",
      "to transdifferentiate cells between lineages1, reprogram somatic cells to pluripotency2 and differentiate stem cells3. Previous studies on applying TFs for cell fate engineering have used either experi- mental testing of a limited set of developmentally relevant TFs4–8 or computational prediction of TFs inferred from genome-scale datasets with modest accuracy9–11. However, the field has lacked a global experimental study of\n",
      "\n",
      "acquisition of pluripotency during somatic cell transdifferentiation with iPSC repro- gramming factors. Nat. Biotechnol. 33: 769–774.\n",
      "\n",
      "73. Bar-Nur, O. et al. 2015. Lineage conversion induced by pluripotency factors involves transient passage through an iPSC stage. Nat. Biotechnol. 33: 761–768.\n",
      "\n",
      "74. Szabo, E. et al. 2010. Direct conversion of human ﬁbroblasts\n",
      "\n",
      "to multilineage blood progenitors. Nature 468: 521–526.\n",
      "\n",
      "75. Mitchell, R. et al. 2014. Molecular evidence for OCT4- induced plasticity in adult human ﬁbroblasts required for direct cell fate conversion to lineage speciﬁc progenitors. Stem Cells 32: 2178–2187.\n",
      "\n",
      "76. Pulecio, J. et al. 2014. Conversion of human ﬁbroblasts into\n",
      "\n",
      "Cell Stem Cell 1, 55–70 (2007). 161. Park, I. H. et al. Reprogramming of human somatic\n",
      "\n",
      "cells to pluripotency with defined factors. Nature 451, 141–146 (2008).\n",
      "\n",
      "162. Hanna, J. et al. Direct cell reprogramming is a stochastic process amenable to acceleration. Nature 462, 595–601 (2009).\n",
      "\n",
      "163. Yang, J. et al. Stat3 activation is limiting for reprogramming to ground state pluripotency. Cell Stem Cell 7, 319–328 (2010).\n",
      "\n",
      "164. Silva, J. et al. Promotion of reprogramming to ground state pluripotency by signal inhibition. PLoS Biol. 6, e253 (2008).\n",
      "\n",
      "Acknowledgements The authors would like to thank all Yamanaka laboratory members, past and present, and are also grateful to Y. Miyake, R. Kato, E.\n",
      "\n",
      "and adult ﬁbroblast cultures by deﬁned factors. Cell 126, 663–676 (2006). 4.\n",
      "\n",
      "5. Xu, J., Du, Y. & Deng, H. Direct lineage reprogramming: strategies,\n",
      "\n",
      "mechanisms, and applications. Cell Stem Cell 16, 119–134 (2015). Takahashi, K. et al. Induction of pluripotent stem cells from adult human ﬁbroblasts by deﬁned factors. Cell 131, 861–872 (2007). Vierbuchen, T. et al. Direct conversion of ﬁbroblasts to functional neurons by deﬁned factors. Nature 463, 1035–1041 (2010). Ieda, M. et al. Direct reprogramming of ﬁbroblasts into functional cardiomyocytes by deﬁned factors. Cell 142, 375–386 (2010). 9. Haridhasapavalan, K. K., Raina, K., Dey, C., Adhikari, P. & Thummer,\n",
      "\n",
      "6.\n",
      "\n",
      "7.\n",
      "\n",
      "8.\n",
      "\n",
      "R. P. An insight\n",
      "\n",
      "generation of induced pluripotent stem cells (iPSCs; left). It is now known that fully differentiated cells can also undergo reprogram- ming. In fact, many cells within the population initiate the process; however, owing to the fact that reprogramming is complex and comprises many events, the majority of cells never complete the process, and only a small number become iPSCs (right). b | Reprogramming can potentially occur either in a deterministic (top) or a stochastic (bottom) fashion. In the deter- ministic process, all cells follow exactly the same path towards pluripotency, and thus cells are reprogrammed with a fixed latency. In the stochastic repro- gramming process, transitions\n",
      "\n",
      "iPSCs, as well as modification of reporter lines to facilitate differentiation towards specific cell types. The finding that it is possible to modulate cell identity has also introduced the possibility of direct reprogramming, in which cell fate conversion is achieved by the overexpression of multiple transcription factors without reverting to a pluripotent state.\n",
      "\n",
      "Nevertheless, some important issues regarding the generation of iPSCs remain unsolved. In pre‑implantation embryos, PSCs are referred to as ‘naive’, and they become ‘primed’ during post‑implantation development. Mouse ES cells and iPSCs have homogeneous characteristics, resembling cells of the early epiblasts of mouse\n",
      "\n",
      "(2007).\n",
      "\n",
      "42. Yu, J. et al. Human induced pluripotent stem cells free of vector and transgene sequences. Science 324, 797–801 (2009).\n",
      "\n",
      "43. Tanabe, K., Nakamura, M., Narita, M., Takahashi, K. & Yamanaka, S. Maturation, not initiation, is the major roadblock during reprogramming toward pluripotency from human fibroblasts. Proc. Natl Acad. Sci. USA 110, 12172–12179 (2013). 44. Nakagawa, M. et al. Generation of induced pluripotent stem cells without Myc from mouse and human fibroblasts. Nat. Biotechnol. 26, 101–106 (2008).\n",
      "\n",
      "45. Jiang, J. et al. A core Klf circuitry regulates self‑\n",
      "\n",
      "renewal of embryonic stem cells. Nat. Cell Biol. 10, 353–360 (2008).\n",
      "\n",
      "46. Nakagawa, M., Takizawa, N., Narita, M.,\n",
      "\n",
      "L. Qian, Y. Hayashi, P. Delgado-Olguin, S. Ding, B. G. Bruneau, D. Srivastava, Direct reprogramming of human fibroblasts toward a . Stem Cell Reports cardiomyocyte-like state. ​\n",
      "\n",
      "1\n",
      "\n",
      ", 235–247 (2013).\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "12. L. Qian, E. C. Berry, J.-D. Fu, M. Ieda, D. Srivastava, Reprogramming of mouse fibroblasts Nat. Protoc. ​ 8\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "13. Y.-J. Nam, K. Song, X. Luo, E. Daniel, K. Lambeth, K. West, J. A. Hill, J. M. DiMaio, L. A.\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in\n",
      "Human: What are the differences between direct lineage reprogramming and induced pluripotency?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "cell types, makes it chal- lenging to generate multiple lineages in the same culture.\n",
      "\n",
      "An alternative approach is direct activation of TFs, which can both shorten the time for cell conversion and isolate it from other developmental events. TF induction has been shown\n",
      "\n",
      "to transdifferentiate cells between lineages1, reprogram somatic cells to pluripotency2 and differentiate stem cells3. Previous studies on applying TFs for cell fate engineering have used either experi- mental testing of a limited set of developmentally relevant TFs4–8 or computational prediction of TFs inferred from genome-scale datasets with modest accuracy9–11. However, the field has lacked a global experimental study of\n",
      "\n",
      "cell types, makes it chal- lenging to generate multiple lineages in the same culture.\n",
      "\n",
      "An alternative approach is direct activation of TFs, which can both shorten the time for cell conversion and isolate it from other developmental events. TF induction has been shown\n",
      "\n",
      "to transdifferentiate cells between lineages1, reprogram somatic cells to pluripotency2 and differentiate stem cells3. Previous studies on applying TFs for cell fate engineering have used either experi- mental testing of a limited set of developmentally relevant TFs4–8 or computational prediction of TFs inferred from genome-scale datasets with modest accuracy9–11. However, the field has lacked a global experimental study of\n",
      "\n",
      "acquisition of pluripotency during somatic cell transdifferentiation with iPSC repro- gramming factors. Nat. Biotechnol. 33: 769–774.\n",
      "\n",
      "73. Bar-Nur, O. et al. 2015. Lineage conversion induced by pluripotency factors involves transient passage through an iPSC stage. Nat. Biotechnol. 33: 761–768.\n",
      "\n",
      "74. Szabo, E. et al. 2010. Direct conversion of human ﬁbroblasts\n",
      "\n",
      "to multilineage blood progenitors. Nature 468: 521–526.\n",
      "\n",
      "75. Mitchell, R. et al. 2014. Molecular evidence for OCT4- induced plasticity in adult human ﬁbroblasts required for direct cell fate conversion to lineage speciﬁc progenitors. Stem Cells 32: 2178–2187.\n",
      "\n",
      "76. Pulecio, J. et al. 2014. Conversion of human ﬁbroblasts into\n",
      "\n",
      "Cell Stem Cell 1, 55–70 (2007). 161. Park, I. H. et al. Reprogramming of human somatic\n",
      "\n",
      "cells to pluripotency with defined factors. Nature 451, 141–146 (2008).\n",
      "\n",
      "162. Hanna, J. et al. Direct cell reprogramming is a stochastic process amenable to acceleration. Nature 462, 595–601 (2009).\n",
      "\n",
      "163. Yang, J. et al. Stat3 activation is limiting for reprogramming to ground state pluripotency. Cell Stem Cell 7, 319–328 (2010).\n",
      "\n",
      "164. Silva, J. et al. Promotion of reprogramming to ground state pluripotency by signal inhibition. PLoS Biol. 6, e253 (2008).\n",
      "\n",
      "Acknowledgements The authors would like to thank all Yamanaka laboratory members, past and present, and are also grateful to Y. Miyake, R. Kato, E.\n",
      "\n",
      "seminal discovery of induced PSCs (iPSCs) (FIG. 2), which are pluripotent cells that are derived from somatic cells, and that functionally bear a striking resemblance to ES cells. The discovery of iPSCs has also paved the way for direct cell fate conversion (also known as transdifferentiation) experiments, in which the expression of sets\n",
      "\n",
      "of tissue‑specific transcription factors can induce a different fate, across germ layers and by bypassing the pluripotent state, opening new doors in the field of cellular reprogramming. iPSCs are promising for many applications, including drug discovery, disease modelling and regenerative medicine23 (FIG. 2).\n",
      "\n",
      "Somatic cell nuclear transfer\n",
      "\n",
      "generation of induced pluripotent stem cells (iPSCs; left). It is now known that fully differentiated cells can also undergo reprogram- ming. In fact, many cells within the population initiate the process; however, owing to the fact that reprogramming is complex and comprises many events, the majority of cells never complete the process, and only a small number become iPSCs (right). b | Reprogramming can potentially occur either in a deterministic (top) or a stochastic (bottom) fashion. In the deter- ministic process, all cells follow exactly the same path towards pluripotency, and thus cells are reprogrammed with a fixed latency. In the stochastic repro- gramming process, transitions\n",
      "\n",
      "iPSCs, as well as modification of reporter lines to facilitate differentiation towards specific cell types. The finding that it is possible to modulate cell identity has also introduced the possibility of direct reprogramming, in which cell fate conversion is achieved by the overexpression of multiple transcription factors without reverting to a pluripotent state.\n",
      "\n",
      "Nevertheless, some important issues regarding the generation of iPSCs remain unsolved. In pre‑implantation embryos, PSCs are referred to as ‘naive’, and they become ‘primed’ during post‑implantation development. Mouse ES cells and iPSCs have homogeneous characteristics, resembling cells of the early epiblasts of mouse\n",
      "\n",
      "(2007).\n",
      "\n",
      "42. Yu, J. et al. Human induced pluripotent stem cells free of vector and transgene sequences. Science 324, 797–801 (2009).\n",
      "\n",
      "43. Tanabe, K., Nakamura, M., Narita, M., Takahashi, K. & Yamanaka, S. Maturation, not initiation, is the major roadblock during reprogramming toward pluripotency from human fibroblasts. Proc. Natl Acad. Sci. USA 110, 12172–12179 (2013). 44. Nakagawa, M. et al. Generation of induced pluripotent stem cells without Myc from mouse and human fibroblasts. Nat. Biotechnol. 26, 101–106 (2008).\n",
      "\n",
      "45. Jiang, J. et al. A core Klf circuitry regulates self‑\n",
      "\n",
      "renewal of embryonic stem cells. Nat. Cell Biol. 10, 353–360 (2008).\n",
      "\n",
      "46. Nakagawa, M., Takizawa, N., Narita, M.,\n",
      "\n",
      "L. Qian, Y. Hayashi, P. Delgado-Olguin, S. Ding, B. G. Bruneau, D. Srivastava, Direct reprogramming of human fibroblasts toward a . Stem Cell Reports cardiomyocyte-like state. ​\n",
      "\n",
      "1\n",
      "\n",
      ", 235–247 (2013).\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "12. L. Qian, E. C. Berry, J.-D. Fu, M. Ieda, D. Srivastava, Reprogramming of mouse fibroblasts Nat. Protoc. ​ 8\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "13. Y.-J. Nam, K. Song, X. Luo, E. Daniel, K. Lambeth, K. West, J. A. Hill, J. M. DiMaio, L. A.\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in\n",
      "Human: What are the differences between direct lineage reprogramming and induced pluripotency?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "cell types, makes it chal- lenging to generate multiple lineages in the same culture.\n",
      "\n",
      "An alternative approach is direct activation of TFs, which can both shorten the time for cell conversion and isolate it from other developmental events. TF induction has been shown\n",
      "\n",
      "to transdifferentiate cells between lineages1, reprogram somatic cells to pluripotency2 and differentiate stem cells3. Previous studies on applying TFs for cell fate engineering have used either experi- mental testing of a limited set of developmentally relevant TFs4–8 or computational prediction of TFs inferred from genome-scale datasets with modest accuracy9–11. However, the field has lacked a global experimental study of\n",
      "\n",
      "cell types, makes it chal- lenging to generate multiple lineages in the same culture.\n",
      "\n",
      "An alternative approach is direct activation of TFs, which can both shorten the time for cell conversion and isolate it from other developmental events. TF induction has been shown\n",
      "\n",
      "to transdifferentiate cells between lineages1, reprogram somatic cells to pluripotency2 and differentiate stem cells3. Previous studies on applying TFs for cell fate engineering have used either experi- mental testing of a limited set of developmentally relevant TFs4–8 or computational prediction of TFs inferred from genome-scale datasets with modest accuracy9–11. However, the field has lacked a global experimental study of\n",
      "\n",
      "acquisition of pluripotency during somatic cell transdifferentiation with iPSC repro- gramming factors. Nat. Biotechnol. 33: 769–774.\n",
      "\n",
      "73. Bar-Nur, O. et al. 2015. Lineage conversion induced by pluripotency factors involves transient passage through an iPSC stage. Nat. Biotechnol. 33: 761–768.\n",
      "\n",
      "74. Szabo, E. et al. 2010. Direct conversion of human ﬁbroblasts\n",
      "\n",
      "to multilineage blood progenitors. Nature 468: 521–526.\n",
      "\n",
      "75. Mitchell, R. et al. 2014. Molecular evidence for OCT4- induced plasticity in adult human ﬁbroblasts required for direct cell fate conversion to lineage speciﬁc progenitors. Stem Cells 32: 2178–2187.\n",
      "\n",
      "76. Pulecio, J. et al. 2014. Conversion of human ﬁbroblasts into\n",
      "\n",
      "Cell Stem Cell 1, 55–70 (2007). 161. Park, I. H. et al. Reprogramming of human somatic\n",
      "\n",
      "cells to pluripotency with defined factors. Nature 451, 141–146 (2008).\n",
      "\n",
      "162. Hanna, J. et al. Direct cell reprogramming is a stochastic process amenable to acceleration. Nature 462, 595–601 (2009).\n",
      "\n",
      "163. Yang, J. et al. Stat3 activation is limiting for reprogramming to ground state pluripotency. Cell Stem Cell 7, 319–328 (2010).\n",
      "\n",
      "164. Silva, J. et al. Promotion of reprogramming to ground state pluripotency by signal inhibition. PLoS Biol. 6, e253 (2008).\n",
      "\n",
      "Acknowledgements The authors would like to thank all Yamanaka laboratory members, past and present, and are also grateful to Y. Miyake, R. Kato, E.\n",
      "\n",
      "seminal discovery of induced PSCs (iPSCs) (FIG. 2), which are pluripotent cells that are derived from somatic cells, and that functionally bear a striking resemblance to ES cells. The discovery of iPSCs has also paved the way for direct cell fate conversion (also known as transdifferentiation) experiments, in which the expression of sets\n",
      "\n",
      "of tissue‑specific transcription factors can induce a different fate, across germ layers and by bypassing the pluripotent state, opening new doors in the field of cellular reprogramming. iPSCs are promising for many applications, including drug discovery, disease modelling and regenerative medicine23 (FIG. 2).\n",
      "\n",
      "Somatic cell nuclear transfer\n",
      "\n",
      "and adult ﬁbroblast cultures by deﬁned factors. Cell 126, 663–676 (2006). 4.\n",
      "\n",
      "5. Xu, J., Du, Y. & Deng, H. Direct lineage reprogramming: strategies,\n",
      "\n",
      "mechanisms, and applications. Cell Stem Cell 16, 119–134 (2015). Takahashi, K. et al. Induction of pluripotent stem cells from adult human ﬁbroblasts by deﬁned factors. Cell 131, 861–872 (2007). Vierbuchen, T. et al. Direct conversion of ﬁbroblasts to functional neurons by deﬁned factors. Nature 463, 1035–1041 (2010). Ieda, M. et al. Direct reprogramming of ﬁbroblasts into functional cardiomyocytes by deﬁned factors. Cell 142, 375–386 (2010). 9. Haridhasapavalan, K. K., Raina, K., Dey, C., Adhikari, P. & Thummer,\n",
      "\n",
      "6.\n",
      "\n",
      "7.\n",
      "\n",
      "8.\n",
      "\n",
      "R. P. An insight\n",
      "\n",
      "iPSCs, as well as modification of reporter lines to facilitate differentiation towards specific cell types. The finding that it is possible to modulate cell identity has also introduced the possibility of direct reprogramming, in which cell fate conversion is achieved by the overexpression of multiple transcription factors without reverting to a pluripotent state.\n",
      "\n",
      "Nevertheless, some important issues regarding the generation of iPSCs remain unsolved. In pre‑implantation embryos, PSCs are referred to as ‘naive’, and they become ‘primed’ during post‑implantation development. Mouse ES cells and iPSCs have homogeneous characteristics, resembling cells of the early epiblasts of mouse\n",
      "\n",
      "(2007).\n",
      "\n",
      "42. Yu, J. et al. Human induced pluripotent stem cells free of vector and transgene sequences. Science 324, 797–801 (2009).\n",
      "\n",
      "43. Tanabe, K., Nakamura, M., Narita, M., Takahashi, K. & Yamanaka, S. Maturation, not initiation, is the major roadblock during reprogramming toward pluripotency from human fibroblasts. Proc. Natl Acad. Sci. USA 110, 12172–12179 (2013). 44. Nakagawa, M. et al. Generation of induced pluripotent stem cells without Myc from mouse and human fibroblasts. Nat. Biotechnol. 26, 101–106 (2008).\n",
      "\n",
      "45. Jiang, J. et al. A core Klf circuitry regulates self‑\n",
      "\n",
      "renewal of embryonic stem cells. Nat. Cell Biol. 10, 353–360 (2008).\n",
      "\n",
      "46. Nakagawa, M., Takizawa, N., Narita, M.,\n",
      "\n",
      "L. Qian, Y. Hayashi, P. Delgado-Olguin, S. Ding, B. G. Bruneau, D. Srivastava, Direct reprogramming of human fibroblasts toward a . Stem Cell Reports cardiomyocyte-like state. ​\n",
      "\n",
      "1\n",
      "\n",
      ", 235–247 (2013).\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "12. L. Qian, E. C. Berry, J.-D. Fu, M. Ieda, D. Srivastava, Reprogramming of mouse fibroblasts Nat. Protoc. ​ 8\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "13. Y.-J. Nam, K. Song, X. Luo, E. Daniel, K. Lambeth, K. West, J. A. Hill, J. M. DiMaio, L. A.\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in\n",
      "Human: What are the differences between direct lineage reprogramming and induced pluripotency?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "cell types, makes it chal- lenging to generate multiple lineages in the same culture.\n",
      "\n",
      "An alternative approach is direct activation of TFs, which can both shorten the time for cell conversion and isolate it from other developmental events. TF induction has been shown\n",
      "\n",
      "to transdifferentiate cells between lineages1, reprogram somatic cells to pluripotency2 and differentiate stem cells3. Previous studies on applying TFs for cell fate engineering have used either experi- mental testing of a limited set of developmentally relevant TFs4–8 or computational prediction of TFs inferred from genome-scale datasets with modest accuracy9–11. However, the field has lacked a global experimental study of\n",
      "\n",
      "cell types, makes it chal- lenging to generate multiple lineages in the same culture.\n",
      "\n",
      "An alternative approach is direct activation of TFs, which can both shorten the time for cell conversion and isolate it from other developmental events. TF induction has been shown\n",
      "\n",
      "to transdifferentiate cells between lineages1, reprogram somatic cells to pluripotency2 and differentiate stem cells3. Previous studies on applying TFs for cell fate engineering have used either experi- mental testing of a limited set of developmentally relevant TFs4–8 or computational prediction of TFs inferred from genome-scale datasets with modest accuracy9–11. However, the field has lacked a global experimental study of\n",
      "\n",
      "acquisition of pluripotency during somatic cell transdifferentiation with iPSC repro- gramming factors. Nat. Biotechnol. 33: 769–774.\n",
      "\n",
      "73. Bar-Nur, O. et al. 2015. Lineage conversion induced by pluripotency factors involves transient passage through an iPSC stage. Nat. Biotechnol. 33: 761–768.\n",
      "\n",
      "74. Szabo, E. et al. 2010. Direct conversion of human ﬁbroblasts\n",
      "\n",
      "to multilineage blood progenitors. Nature 468: 521–526.\n",
      "\n",
      "75. Mitchell, R. et al. 2014. Molecular evidence for OCT4- induced plasticity in adult human ﬁbroblasts required for direct cell fate conversion to lineage speciﬁc progenitors. Stem Cells 32: 2178–2187.\n",
      "\n",
      "76. Pulecio, J. et al. 2014. Conversion of human ﬁbroblasts into\n",
      "\n",
      "Cell Stem Cell 1, 55–70 (2007). 161. Park, I. H. et al. Reprogramming of human somatic\n",
      "\n",
      "cells to pluripotency with defined factors. Nature 451, 141–146 (2008).\n",
      "\n",
      "162. Hanna, J. et al. Direct cell reprogramming is a stochastic process amenable to acceleration. Nature 462, 595–601 (2009).\n",
      "\n",
      "163. Yang, J. et al. Stat3 activation is limiting for reprogramming to ground state pluripotency. Cell Stem Cell 7, 319–328 (2010).\n",
      "\n",
      "164. Silva, J. et al. Promotion of reprogramming to ground state pluripotency by signal inhibition. PLoS Biol. 6, e253 (2008).\n",
      "\n",
      "Acknowledgements The authors would like to thank all Yamanaka laboratory members, past and present, and are also grateful to Y. Miyake, R. Kato, E.\n",
      "\n",
      "seminal discovery of induced PSCs (iPSCs) (FIG. 2), which are pluripotent cells that are derived from somatic cells, and that functionally bear a striking resemblance to ES cells. The discovery of iPSCs has also paved the way for direct cell fate conversion (also known as transdifferentiation) experiments, in which the expression of sets\n",
      "\n",
      "of tissue‑specific transcription factors can induce a different fate, across germ layers and by bypassing the pluripotent state, opening new doors in the field of cellular reprogramming. iPSCs are promising for many applications, including drug discovery, disease modelling and regenerative medicine23 (FIG. 2).\n",
      "\n",
      "Somatic cell nuclear transfer\n",
      "\n",
      "and adult ﬁbroblast cultures by deﬁned factors. Cell 126, 663–676 (2006). 4.\n",
      "\n",
      "5. Xu, J., Du, Y. & Deng, H. Direct lineage reprogramming: strategies,\n",
      "\n",
      "mechanisms, and applications. Cell Stem Cell 16, 119–134 (2015). Takahashi, K. et al. Induction of pluripotent stem cells from adult human ﬁbroblasts by deﬁned factors. Cell 131, 861–872 (2007). Vierbuchen, T. et al. Direct conversion of ﬁbroblasts to functional neurons by deﬁned factors. Nature 463, 1035–1041 (2010). Ieda, M. et al. Direct reprogramming of ﬁbroblasts into functional cardiomyocytes by deﬁned factors. Cell 142, 375–386 (2010). 9. Haridhasapavalan, K. K., Raina, K., Dey, C., Adhikari, P. & Thummer,\n",
      "\n",
      "6.\n",
      "\n",
      "7.\n",
      "\n",
      "8.\n",
      "\n",
      "R. P. An insight\n",
      "\n",
      "generation of induced pluripotent stem cells (iPSCs; left). It is now known that fully differentiated cells can also undergo reprogram- ming. In fact, many cells within the population initiate the process; however, owing to the fact that reprogramming is complex and comprises many events, the majority of cells never complete the process, and only a small number become iPSCs (right). b | Reprogramming can potentially occur either in a deterministic (top) or a stochastic (bottom) fashion. In the deter- ministic process, all cells follow exactly the same path towards pluripotency, and thus cells are reprogrammed with a fixed latency. In the stochastic repro- gramming process, transitions\n",
      "\n",
      "(2007).\n",
      "\n",
      "42. Yu, J. et al. Human induced pluripotent stem cells free of vector and transgene sequences. Science 324, 797–801 (2009).\n",
      "\n",
      "43. Tanabe, K., Nakamura, M., Narita, M., Takahashi, K. & Yamanaka, S. Maturation, not initiation, is the major roadblock during reprogramming toward pluripotency from human fibroblasts. Proc. Natl Acad. Sci. USA 110, 12172–12179 (2013). 44. Nakagawa, M. et al. Generation of induced pluripotent stem cells without Myc from mouse and human fibroblasts. Nat. Biotechnol. 26, 101–106 (2008).\n",
      "\n",
      "45. Jiang, J. et al. A core Klf circuitry regulates self‑\n",
      "\n",
      "renewal of embryonic stem cells. Nat. Cell Biol. 10, 353–360 (2008).\n",
      "\n",
      "46. Nakagawa, M., Takizawa, N., Narita, M.,\n",
      "\n",
      "L. Qian, Y. Hayashi, P. Delgado-Olguin, S. Ding, B. G. Bruneau, D. Srivastava, Direct reprogramming of human fibroblasts toward a . Stem Cell Reports cardiomyocyte-like state. ​\n",
      "\n",
      "1\n",
      "\n",
      ", 235–247 (2013).\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "12. L. Qian, E. C. Berry, J.-D. Fu, M. Ieda, D. Srivastava, Reprogramming of mouse fibroblasts Nat. Protoc. ​ 8\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "13. Y.-J. Nam, K. Song, X. Luo, E. Daniel, K. Lambeth, K. West, J. A. Hill, J. M. DiMaio, L. A.\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in\n",
      "Human: What are the differences between direct lineage reprogramming and induced pluripotency?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "cell types, makes it chal- lenging to generate multiple lineages in the same culture.\n",
      "\n",
      "An alternative approach is direct activation of TFs, which can both shorten the time for cell conversion and isolate it from other developmental events. TF induction has been shown\n",
      "\n",
      "to transdifferentiate cells between lineages1, reprogram somatic cells to pluripotency2 and differentiate stem cells3. Previous studies on applying TFs for cell fate engineering have used either experi- mental testing of a limited set of developmentally relevant TFs4–8 or computational prediction of TFs inferred from genome-scale datasets with modest accuracy9–11. However, the field has lacked a global experimental study of\n",
      "\n",
      "cell types, makes it chal- lenging to generate multiple lineages in the same culture.\n",
      "\n",
      "An alternative approach is direct activation of TFs, which can both shorten the time for cell conversion and isolate it from other developmental events. TF induction has been shown\n",
      "\n",
      "to transdifferentiate cells between lineages1, reprogram somatic cells to pluripotency2 and differentiate stem cells3. Previous studies on applying TFs for cell fate engineering have used either experi- mental testing of a limited set of developmentally relevant TFs4–8 or computational prediction of TFs inferred from genome-scale datasets with modest accuracy9–11. However, the field has lacked a global experimental study of\n",
      "\n",
      "acquisition of pluripotency during somatic cell transdifferentiation with iPSC repro- gramming factors. Nat. Biotechnol. 33: 769–774.\n",
      "\n",
      "73. Bar-Nur, O. et al. 2015. Lineage conversion induced by pluripotency factors involves transient passage through an iPSC stage. Nat. Biotechnol. 33: 761–768.\n",
      "\n",
      "74. Szabo, E. et al. 2010. Direct conversion of human ﬁbroblasts\n",
      "\n",
      "to multilineage blood progenitors. Nature 468: 521–526.\n",
      "\n",
      "75. Mitchell, R. et al. 2014. Molecular evidence for OCT4- induced plasticity in adult human ﬁbroblasts required for direct cell fate conversion to lineage speciﬁc progenitors. Stem Cells 32: 2178–2187.\n",
      "\n",
      "76. Pulecio, J. et al. 2014. Conversion of human ﬁbroblasts into\n",
      "\n",
      "Cell Stem Cell 1, 55–70 (2007). 161. Park, I. H. et al. Reprogramming of human somatic\n",
      "\n",
      "cells to pluripotency with defined factors. Nature 451, 141–146 (2008).\n",
      "\n",
      "162. Hanna, J. et al. Direct cell reprogramming is a stochastic process amenable to acceleration. Nature 462, 595–601 (2009).\n",
      "\n",
      "163. Yang, J. et al. Stat3 activation is limiting for reprogramming to ground state pluripotency. Cell Stem Cell 7, 319–328 (2010).\n",
      "\n",
      "164. Silva, J. et al. Promotion of reprogramming to ground state pluripotency by signal inhibition. PLoS Biol. 6, e253 (2008).\n",
      "\n",
      "Acknowledgements The authors would like to thank all Yamanaka laboratory members, past and present, and are also grateful to Y. Miyake, R. Kato, E.\n",
      "\n",
      "seminal discovery of induced PSCs (iPSCs) (FIG. 2), which are pluripotent cells that are derived from somatic cells, and that functionally bear a striking resemblance to ES cells. The discovery of iPSCs has also paved the way for direct cell fate conversion (also known as transdifferentiation) experiments, in which the expression of sets\n",
      "\n",
      "of tissue‑specific transcription factors can induce a different fate, across germ layers and by bypassing the pluripotent state, opening new doors in the field of cellular reprogramming. iPSCs are promising for many applications, including drug discovery, disease modelling and regenerative medicine23 (FIG. 2).\n",
      "\n",
      "Somatic cell nuclear transfer\n",
      "\n",
      "and adult ﬁbroblast cultures by deﬁned factors. Cell 126, 663–676 (2006). 4.\n",
      "\n",
      "5. Xu, J., Du, Y. & Deng, H. Direct lineage reprogramming: strategies,\n",
      "\n",
      "mechanisms, and applications. Cell Stem Cell 16, 119–134 (2015). Takahashi, K. et al. Induction of pluripotent stem cells from adult human ﬁbroblasts by deﬁned factors. Cell 131, 861–872 (2007). Vierbuchen, T. et al. Direct conversion of ﬁbroblasts to functional neurons by deﬁned factors. Nature 463, 1035–1041 (2010). Ieda, M. et al. Direct reprogramming of ﬁbroblasts into functional cardiomyocytes by deﬁned factors. Cell 142, 375–386 (2010). 9. Haridhasapavalan, K. K., Raina, K., Dey, C., Adhikari, P. & Thummer,\n",
      "\n",
      "6.\n",
      "\n",
      "7.\n",
      "\n",
      "8.\n",
      "\n",
      "R. P. An insight\n",
      "\n",
      "generation of induced pluripotent stem cells (iPSCs; left). It is now known that fully differentiated cells can also undergo reprogram- ming. In fact, many cells within the population initiate the process; however, owing to the fact that reprogramming is complex and comprises many events, the majority of cells never complete the process, and only a small number become iPSCs (right). b | Reprogramming can potentially occur either in a deterministic (top) or a stochastic (bottom) fashion. In the deter- ministic process, all cells follow exactly the same path towards pluripotency, and thus cells are reprogrammed with a fixed latency. In the stochastic repro- gramming process, transitions\n",
      "\n",
      "iPSCs, as well as modification of reporter lines to facilitate differentiation towards specific cell types. The finding that it is possible to modulate cell identity has also introduced the possibility of direct reprogramming, in which cell fate conversion is achieved by the overexpression of multiple transcription factors without reverting to a pluripotent state.\n",
      "\n",
      "Nevertheless, some important issues regarding the generation of iPSCs remain unsolved. In pre‑implantation embryos, PSCs are referred to as ‘naive’, and they become ‘primed’ during post‑implantation development. Mouse ES cells and iPSCs have homogeneous characteristics, resembling cells of the early epiblasts of mouse\n",
      "\n",
      "L. Qian, Y. Hayashi, P. Delgado-Olguin, S. Ding, B. G. Bruneau, D. Srivastava, Direct reprogramming of human fibroblasts toward a . Stem Cell Reports cardiomyocyte-like state. ​\n",
      "\n",
      "1\n",
      "\n",
      ", 235–247 (2013).\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "12. L. Qian, E. C. Berry, J.-D. Fu, M. Ieda, D. Srivastava, Reprogramming of mouse fibroblasts Nat. Protoc. ​ 8\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "13. Y.-J. Nam, K. Song, X. Luo, E. Daniel, K. Lambeth, K. West, J. A. Hill, J. M. DiMaio, L. A.\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in\n",
      "Human: What are the differences between direct lineage reprogramming and induced pluripotency?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "cell types, makes it chal- lenging to generate multiple lineages in the same culture.\n",
      "\n",
      "An alternative approach is direct activation of TFs, which can both shorten the time for cell conversion and isolate it from other developmental events. TF induction has been shown\n",
      "\n",
      "to transdifferentiate cells between lineages1, reprogram somatic cells to pluripotency2 and differentiate stem cells3. Previous studies on applying TFs for cell fate engineering have used either experi- mental testing of a limited set of developmentally relevant TFs4–8 or computational prediction of TFs inferred from genome-scale datasets with modest accuracy9–11. However, the field has lacked a global experimental study of\n",
      "\n",
      "cell types, makes it chal- lenging to generate multiple lineages in the same culture.\n",
      "\n",
      "An alternative approach is direct activation of TFs, which can both shorten the time for cell conversion and isolate it from other developmental events. TF induction has been shown\n",
      "\n",
      "to transdifferentiate cells between lineages1, reprogram somatic cells to pluripotency2 and differentiate stem cells3. Previous studies on applying TFs for cell fate engineering have used either experi- mental testing of a limited set of developmentally relevant TFs4–8 or computational prediction of TFs inferred from genome-scale datasets with modest accuracy9–11. However, the field has lacked a global experimental study of\n",
      "\n",
      "acquisition of pluripotency during somatic cell transdifferentiation with iPSC repro- gramming factors. Nat. Biotechnol. 33: 769–774.\n",
      "\n",
      "73. Bar-Nur, O. et al. 2015. Lineage conversion induced by pluripotency factors involves transient passage through an iPSC stage. Nat. Biotechnol. 33: 761–768.\n",
      "\n",
      "74. Szabo, E. et al. 2010. Direct conversion of human ﬁbroblasts\n",
      "\n",
      "to multilineage blood progenitors. Nature 468: 521–526.\n",
      "\n",
      "75. Mitchell, R. et al. 2014. Molecular evidence for OCT4- induced plasticity in adult human ﬁbroblasts required for direct cell fate conversion to lineage speciﬁc progenitors. Stem Cells 32: 2178–2187.\n",
      "\n",
      "76. Pulecio, J. et al. 2014. Conversion of human ﬁbroblasts into\n",
      "\n",
      "Cell Stem Cell 1, 55–70 (2007). 161. Park, I. H. et al. Reprogramming of human somatic\n",
      "\n",
      "cells to pluripotency with defined factors. Nature 451, 141–146 (2008).\n",
      "\n",
      "162. Hanna, J. et al. Direct cell reprogramming is a stochastic process amenable to acceleration. Nature 462, 595–601 (2009).\n",
      "\n",
      "163. Yang, J. et al. Stat3 activation is limiting for reprogramming to ground state pluripotency. Cell Stem Cell 7, 319–328 (2010).\n",
      "\n",
      "164. Silva, J. et al. Promotion of reprogramming to ground state pluripotency by signal inhibition. PLoS Biol. 6, e253 (2008).\n",
      "\n",
      "Acknowledgements The authors would like to thank all Yamanaka laboratory members, past and present, and are also grateful to Y. Miyake, R. Kato, E.\n",
      "\n",
      "seminal discovery of induced PSCs (iPSCs) (FIG. 2), which are pluripotent cells that are derived from somatic cells, and that functionally bear a striking resemblance to ES cells. The discovery of iPSCs has also paved the way for direct cell fate conversion (also known as transdifferentiation) experiments, in which the expression of sets\n",
      "\n",
      "of tissue‑specific transcription factors can induce a different fate, across germ layers and by bypassing the pluripotent state, opening new doors in the field of cellular reprogramming. iPSCs are promising for many applications, including drug discovery, disease modelling and regenerative medicine23 (FIG. 2).\n",
      "\n",
      "Somatic cell nuclear transfer\n",
      "\n",
      "and adult ﬁbroblast cultures by deﬁned factors. Cell 126, 663–676 (2006). 4.\n",
      "\n",
      "5. Xu, J., Du, Y. & Deng, H. Direct lineage reprogramming: strategies,\n",
      "\n",
      "mechanisms, and applications. Cell Stem Cell 16, 119–134 (2015). Takahashi, K. et al. Induction of pluripotent stem cells from adult human ﬁbroblasts by deﬁned factors. Cell 131, 861–872 (2007). Vierbuchen, T. et al. Direct conversion of ﬁbroblasts to functional neurons by deﬁned factors. Nature 463, 1035–1041 (2010). Ieda, M. et al. Direct reprogramming of ﬁbroblasts into functional cardiomyocytes by deﬁned factors. Cell 142, 375–386 (2010). 9. Haridhasapavalan, K. K., Raina, K., Dey, C., Adhikari, P. & Thummer,\n",
      "\n",
      "6.\n",
      "\n",
      "7.\n",
      "\n",
      "8.\n",
      "\n",
      "R. P. An insight\n",
      "\n",
      "generation of induced pluripotent stem cells (iPSCs; left). It is now known that fully differentiated cells can also undergo reprogram- ming. In fact, many cells within the population initiate the process; however, owing to the fact that reprogramming is complex and comprises many events, the majority of cells never complete the process, and only a small number become iPSCs (right). b | Reprogramming can potentially occur either in a deterministic (top) or a stochastic (bottom) fashion. In the deter- ministic process, all cells follow exactly the same path towards pluripotency, and thus cells are reprogrammed with a fixed latency. In the stochastic repro- gramming process, transitions\n",
      "\n",
      "iPSCs, as well as modification of reporter lines to facilitate differentiation towards specific cell types. The finding that it is possible to modulate cell identity has also introduced the possibility of direct reprogramming, in which cell fate conversion is achieved by the overexpression of multiple transcription factors without reverting to a pluripotent state.\n",
      "\n",
      "Nevertheless, some important issues regarding the generation of iPSCs remain unsolved. In pre‑implantation embryos, PSCs are referred to as ‘naive’, and they become ‘primed’ during post‑implantation development. Mouse ES cells and iPSCs have homogeneous characteristics, resembling cells of the early epiblasts of mouse\n",
      "\n",
      "(2007).\n",
      "\n",
      "42. Yu, J. et al. Human induced pluripotent stem cells free of vector and transgene sequences. Science 324, 797–801 (2009).\n",
      "\n",
      "43. Tanabe, K., Nakamura, M., Narita, M., Takahashi, K. & Yamanaka, S. Maturation, not initiation, is the major roadblock during reprogramming toward pluripotency from human fibroblasts. Proc. Natl Acad. Sci. USA 110, 12172–12179 (2013). 44. Nakagawa, M. et al. Generation of induced pluripotent stem cells without Myc from mouse and human fibroblasts. Nat. Biotechnol. 26, 101–106 (2008).\n",
      "\n",
      "45. Jiang, J. et al. A core Klf circuitry regulates self‑\n",
      "\n",
      "renewal of embryonic stem cells. Nat. Cell Biol. 10, 353–360 (2008).\n",
      "\n",
      "46. Nakagawa, M., Takizawa, N., Narita, M.,\n",
      "Human: What are the differences between direct lineage reprogramming and induced pluripotency?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "==================================================\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  What are the potential applications of cell reprogramming in regenerative medicine?\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fri Jun  7 23:19:07 2024 INFO SCRAPE\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "RAG >> 4: \n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation:\n",
      "\n",
      "    \n",
      "    Query:What are the potential applications of cell reprogramming in regenerative medicine?\n",
      "    \n",
      "    From the query, decide if ARXIV, PUBMED, or BIORXIV should be searched, and propose no more than 10 search terms for this query and database. Separate each term with a comma, and provide no extra information/explination for either the database or search terms. Format your output as follows with no additions:\n",
      "    \n",
      "    Database: <ARXIV, PUBMED, or BIORXIV>\n",
      "    Search Terms: <improved search terms>\n",
      "    \u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "searching on PUBMED...\n",
      "Search Terms: ['cell reprogramming', 'regenerative medicine', 'applications', 'therapy', 'stem cells', 'differentiation', 'tissue engineering', 'transdifferentiation', 'reprogramming techniques', 'clinical trials.']\n",
      "Directory '/home/jpic/RAG-DEV/experimental/specialized_docs' created successfully\n",
      "36383700 could not be gathered.\n",
      "pdf collection complete!\n",
      "Directory '/home/jpic/RAG-DEV/experimental/specialized_docs' created successfully\n",
      "Error occurred while searching on PUBMED: 'NoneType' object has no attribute 'attrs'\n",
      "==================================================\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  What does the literature say about the potential applications of cell reprogramming in regenerative medicine?\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fri Jun  7 23:21:01 2024 INFO SCRAPE\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "RAG >> 5: \n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation:\n",
      "Human: What are the potential applications of cell reprogramming in regenerative medicine?\n",
      "BRAD:  Database: PUBMED\n",
      "Search Terms: cell reprogramming, regenerative medicine, applications, therapy, stem cells, differentiation, tissue engineering, transdifferentiation, reprogramming techniques, clinical trials.\n",
      "    \n",
      "    Query:What does the literature say about the potential applications of cell reprogramming in regenerative medicine?\n",
      "    \n",
      "    From the query, decide if ARXIV, PUBMED, or BIORXIV should be searched, and propose no more than 10 search terms for this query and database. Separate each term with a comma, and provide no extra information/explination for either the database or search terms. Format your output as follows with no additions:\n",
      "    \n",
      "    Database: <ARXIV, PUBMED, or BIORXIV>\n",
      "    Search Terms: <improved search terms>\n",
      "    \u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "searching on PUBMED...\n",
      "Search Terms: ['cell reprogramming', 'regenerative medicine', 'applications', 'therapy', 'stem cells', 'differentiation', 'tissue engineering', 'transdifferentiation', 'reprogramming techniques', 'clinical trials.']\n",
      "Directory '/home/jpic/RAG-DEV/experimental/specialized_docs' created successfully\n",
      "36383700 could not be gathered.\n",
      "pdf collection complete!\n",
      "Directory '/home/jpic/RAG-DEV/experimental/specialized_docs' created successfully\n",
      "Error occurred while searching on PUBMED: 'NoneType' object has no attribute 'attrs'\n",
      "==================================================\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  /force RAG What are the potential applications of cell reprogramming in regenerative medicine?\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fri Jun  7 23:21:39 2024 INFO RAG\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "RAG >> 6: \n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "application. Functionally, these cells only possess short-term engraftment potential, neces- sitating further studies into this reprogramming method.\n",
      "\n",
      "We have graphically summarized the above- discussed efforts in Figure 4 and present speciﬁc reprogramming strategies. TFs, starting cells, cul- ture conditions, media, and cytokines, as well as the outcomes in terms of lineage production and engraftment, are compiled in Table 1.\n",
      "\n",
      "32\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Perspectives on the various reprogramming strategies\n",
      "\n",
      "Obtaining HSCs de novo remains the holy grail of regenerative HSC biology because of the multiple issues associated with HSC transplant and HLA matching. Generating and studying these cells in\n",
      "\n",
      "capable of multilineage engraftment potential, however, limits their use.\n",
      "\n",
      "TF reprogramming somatic cells\n",
      "\n",
      "Building upon the framework established by other studies of cell fate conversion (Fig. 1), our labora- tory was the ﬁrst to attempt somatic cell program- ming into hematopoietic stem and progenitor cells (HSPCs) using nonpluripotent TFs.26 Several labo- ratories have followed and are now trying to directly reprogram somatic cells into HSCs via TF overex- pression. Most of these studies bypass a pluripo- tent state by not using pluripotency factors, which can generate plastic intermediates.72,73 Variations between each strategy, such as starting cell popula- tions, TF cocktails, or\n",
      "\n",
      "C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "31\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "Table 1. Reprogramming strategies and materials\n",
      "\n",
      "comprising Erg, Gata2, Lmo2, Runx1c, and Scl was demonstrated by the group of George Lacaud. As in our studies, these reprogrammed cells appeared to travel through an HE intermediate and had an endothelial and hematopoietic gene expression pro- ﬁle. The generated progenitors formed erythroid, megakaryocytic, and myeloid cells, and could form B and T cells upon co-culture with OP9 and OP9- DL1 stroma, respectively. The use of p53 null ﬁbrob- lasts made reprogramming more efﬁcient and more rapid, shortening the transition from 8 to 5 days.87 It was\n",
      "\n",
      "equally to this manuscript.\n",
      "\n",
      "doi: 10.1111/nyas.12989\n",
      "\n",
      "24\n",
      "\n",
      "Ann. N.Y. Acad. Sci. 1370 (2016) 24–35 C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "Figure 1. Established TF reprogramming strategies. TF reprogramming has been shown to work across a variety of cell types. This ﬁgure shows the starting and end cell populations, as well as the TF cocktails used to derive each cell type of choice. Adapted from Ref. 25.\n",
      "\n",
      "The novel technology of induced pluripotent stem cells (iPSCs) from Yamanaka and colleagues instigated the entire ﬁeld of regenerative medicine. Using only four transcription factors (TFs), Oct3/4, Sox2, Klf4, and c-Myc,\n",
      "\n",
      "Cell Stem Cell 1, 55–70 (2007). 161. Park, I. H. et al. Reprogramming of human somatic\n",
      "\n",
      "cells to pluripotency with defined factors. Nature 451, 141–146 (2008).\n",
      "\n",
      "162. Hanna, J. et al. Direct cell reprogramming is a stochastic process amenable to acceleration. Nature 462, 595–601 (2009).\n",
      "\n",
      "163. Yang, J. et al. Stat3 activation is limiting for reprogramming to ground state pluripotency. Cell Stem Cell 7, 319–328 (2010).\n",
      "\n",
      "164. Silva, J. et al. Promotion of reprogramming to ground state pluripotency by signal inhibition. PLoS Biol. 6, e253 (2008).\n",
      "\n",
      "Acknowledgements The authors would like to thank all Yamanaka laboratory members, past and present, and are also grateful to Y. Miyake, R. Kato, E.\n",
      "\n",
      "that Drive Neuronal Fate and Reprogramming. Cell Stem . Cell ​ .\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "41. A. Ebrahimi, K. Sevinç, G. Gürhan Sevinç, A. P. Cribbs, M. Philpott, F. Uyulur, T. Morova, J. E. Dunford, S. Göklemez, Ş. Arı, U. Oppermann, T. T. Önder, Bromodomain inhibition of the coactivators CBP/EP300 facilitate cellular reprogramming. 519–528 (2019).\n",
      "\n",
      "Nat. Chem. Biol. ​\n",
      "\n",
      "15\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      ",\n",
      "\n",
      "​\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made\n",
      "\n",
      "available under a\n",
      "\n",
      "CC-BY\n",
      "\n",
      "cells. This work augments study of the hemangioblast as well as hematopoiesis in vitro.\n",
      "\n",
      "Rapid HSC reprogramming Rapid reprogramming of fetal and adult mouse into HSPCs using a TF cocktail ﬁbroblasts\n",
      "\n",
      "Ann. N.Y. Acad. Sci. 1370 (2016) 24–35 C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "ESC/iPSC\n",
      "\n",
      "Slukvin - GATA2 + ETV2 or - GATA2 + TAL1\n",
      "\n",
      "H\n",
      "\n",
      "Thomson - Gata2 - Lmo2 - Mycn - Pitx2 - Sox17 - Tal1\n",
      "\n",
      "M\n",
      "\n",
      "Daley - HoxB4\n",
      "\n",
      "M\n",
      "\n",
      "CD34+CD45+ Myeloid Progenitor\n",
      "\n",
      "Hemogenic Endothelium\n",
      "\n",
      "Endothelial cell\n",
      "\n",
      "Daley\n",
      "\n",
      "Moore - Gata2 - Gfi1b - cFos - Etv6\n",
      "\n",
      "M\n",
      "\n",
      "Lacaud - Erg - Gata2 - Lmo2 - Runx1c - Scl\n",
      "\n",
      "M\n",
      "\n",
      "Hemangioblast\n",
      "\n",
      "HSC\n",
      "\n",
      "Rafii - FOSB - GFI1 - RUNX1 - SPI1\n",
      "\n",
      "hemogenic induc- tion (Lis et al., 2017; Pereira et al., 2013; Sandler et al., 2014) and has recently been implicated during EHT from mouse ESCs (Obier et al., 2016). Our studies corroborate that direct cellular reprogramming informs the process of HSC speciﬁcation and of- fers tractable means to systematically interrogate pathways of human HSC formation. It will be interesting to address the role of other enriched pathways such as cytoskeletal regulation by Rho GTPase and the histamine H1 receptor pathway. Genes that are co-targeted during hematopoietic reprogramming by GATA2 and GFI1B are additional candidates for further study (e.g., ITGA2, PLCL1, ADRA1B, RHOJ, SEMA3A, and PLCB1; see\n",
      "\n",
      "and tran- scription to oppose the lineage reprogramming process. Finally, we will test whether the targeted inhibition of such fate stabilizers might enhance our ability to use reprogramming TFs to improve adult organ function post-injury.\n",
      "\n",
      "To uncover fate stabilizers, we employed a genome-wide TF siRNA screen in a cardiac reprogramming (CR) assay using mouse embryonic ﬁbroblasts, followed by the combinatorial evaluation of top-performing hits, and validated our ﬁndings in human primary cells (human ﬁbroblasts and adult endothelial cells) in reprogram- ming assays probing three different lineages (cardiac, neural, iPSC). This approach led us to identify a conserved set of 4 TFs (ATF7IP,\n",
      "Human: /force RAG What are the potential applications of cell reprogramming in regenerative medicine?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "through a plas- tic intermediate. Moreover, although the in vivo hematopoietic niche can provide all signals neces- sary for reprogramming, we cannot dissect the niche in this study, and thus cannot determine the ele- ments that support reprogramming. Furthermore, the inability to generate any hematopoietic cells other than macrophages and monocyte-like pro- genitors also limits future studies with this process unless they focus on these lineages.\n",
      "\n",
      "Hemogenic induction As previously stated, our laboratory became the ﬁrst to induce a hemogenic program in somatic cells via nonpluripotency TF overexpression in mouse embryonic ﬁbroblasts (MEFs). We ﬁrst performed a large screen by transducing\n",
      "\n",
      "capable of multilineage engraftment potential, however, limits their use.\n",
      "\n",
      "TF reprogramming somatic cells\n",
      "\n",
      "Building upon the framework established by other studies of cell fate conversion (Fig. 1), our labora- tory was the ﬁrst to attempt somatic cell program- ming into hematopoietic stem and progenitor cells (HSPCs) using nonpluripotent TFs.26 Several labo- ratories have followed and are now trying to directly reprogram somatic cells into HSCs via TF overex- pression. Most of these studies bypass a pluripo- tent state by not using pluripotency factors, which can generate plastic intermediates.72,73 Variations between each strategy, such as starting cell popula- tions, TF cocktails, or\n",
      "\n",
      "C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "31\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "Table 1. Reprogramming strategies and materials\n",
      "\n",
      "comprising Erg, Gata2, Lmo2, Runx1c, and Scl was demonstrated by the group of George Lacaud. As in our studies, these reprogrammed cells appeared to travel through an HE intermediate and had an endothelial and hematopoietic gene expression pro- ﬁle. The generated progenitors formed erythroid, megakaryocytic, and myeloid cells, and could form B and T cells upon co-culture with OP9 and OP9- DL1 stroma, respectively. The use of p53 null ﬁbrob- lasts made reprogramming more efﬁcient and more rapid, shortening the transition from 8 to 5 days.87 It was\n",
      "\n",
      "equally to this manuscript.\n",
      "\n",
      "doi: 10.1111/nyas.12989\n",
      "\n",
      "24\n",
      "\n",
      "Ann. N.Y. Acad. Sci. 1370 (2016) 24–35 C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "Figure 1. Established TF reprogramming strategies. TF reprogramming has been shown to work across a variety of cell types. This ﬁgure shows the starting and end cell populations, as well as the TF cocktails used to derive each cell type of choice. Adapted from Ref. 25.\n",
      "\n",
      "The novel technology of induced pluripotent stem cells (iPSCs) from Yamanaka and colleagues instigated the entire ﬁeld of regenerative medicine. Using only four transcription factors (TFs), Oct3/4, Sox2, Klf4, and c-Myc,\n",
      "\n",
      "Cell Stem Cell 1, 55–70 (2007). 161. Park, I. H. et al. Reprogramming of human somatic\n",
      "\n",
      "cells to pluripotency with defined factors. Nature 451, 141–146 (2008).\n",
      "\n",
      "162. Hanna, J. et al. Direct cell reprogramming is a stochastic process amenable to acceleration. Nature 462, 595–601 (2009).\n",
      "\n",
      "163. Yang, J. et al. Stat3 activation is limiting for reprogramming to ground state pluripotency. Cell Stem Cell 7, 319–328 (2010).\n",
      "\n",
      "164. Silva, J. et al. Promotion of reprogramming to ground state pluripotency by signal inhibition. PLoS Biol. 6, e253 (2008).\n",
      "\n",
      "Acknowledgements The authors would like to thank all Yamanaka laboratory members, past and present, and are also grateful to Y. Miyake, R. Kato, E.\n",
      "\n",
      "that Drive Neuronal Fate and Reprogramming. Cell Stem . Cell ​ .\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "41. A. Ebrahimi, K. Sevinç, G. Gürhan Sevinç, A. P. Cribbs, M. Philpott, F. Uyulur, T. Morova, J. E. Dunford, S. Göklemez, Ş. Arı, U. Oppermann, T. T. Önder, Bromodomain inhibition of the coactivators CBP/EP300 facilitate cellular reprogramming. 519–528 (2019).\n",
      "\n",
      "Nat. Chem. Biol. ​\n",
      "\n",
      "15\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      ",\n",
      "\n",
      "​\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made\n",
      "\n",
      "available under a\n",
      "\n",
      "CC-BY\n",
      "\n",
      "cells. This work augments study of the hemangioblast as well as hematopoiesis in vitro.\n",
      "\n",
      "Rapid HSC reprogramming Rapid reprogramming of fetal and adult mouse into HSPCs using a TF cocktail ﬁbroblasts\n",
      "\n",
      "Ann. N.Y. Acad. Sci. 1370 (2016) 24–35 C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "ESC/iPSC\n",
      "\n",
      "Slukvin - GATA2 + ETV2 or - GATA2 + TAL1\n",
      "\n",
      "H\n",
      "\n",
      "Thomson - Gata2 - Lmo2 - Mycn - Pitx2 - Sox17 - Tal1\n",
      "\n",
      "M\n",
      "\n",
      "Daley - HoxB4\n",
      "\n",
      "M\n",
      "\n",
      "CD34+CD45+ Myeloid Progenitor\n",
      "\n",
      "Hemogenic Endothelium\n",
      "\n",
      "Endothelial cell\n",
      "\n",
      "Daley\n",
      "\n",
      "Moore - Gata2 - Gfi1b - cFos - Etv6\n",
      "\n",
      "M\n",
      "\n",
      "Lacaud - Erg - Gata2 - Lmo2 - Runx1c - Scl\n",
      "\n",
      "M\n",
      "\n",
      "Hemangioblast\n",
      "\n",
      "HSC\n",
      "\n",
      "Rafii - FOSB - GFI1 - RUNX1 - SPI1\n",
      "\n",
      "hemogenic induc- tion (Lis et al., 2017; Pereira et al., 2013; Sandler et al., 2014) and has recently been implicated during EHT from mouse ESCs (Obier et al., 2016). Our studies corroborate that direct cellular reprogramming informs the process of HSC speciﬁcation and of- fers tractable means to systematically interrogate pathways of human HSC formation. It will be interesting to address the role of other enriched pathways such as cytoskeletal regulation by Rho GTPase and the histamine H1 receptor pathway. Genes that are co-targeted during hematopoietic reprogramming by GATA2 and GFI1B are additional candidates for further study (e.g., ITGA2, PLCL1, ADRA1B, RHOJ, SEMA3A, and PLCB1; see\n",
      "\n",
      "and tran- scription to oppose the lineage reprogramming process. Finally, we will test whether the targeted inhibition of such fate stabilizers might enhance our ability to use reprogramming TFs to improve adult organ function post-injury.\n",
      "\n",
      "To uncover fate stabilizers, we employed a genome-wide TF siRNA screen in a cardiac reprogramming (CR) assay using mouse embryonic ﬁbroblasts, followed by the combinatorial evaluation of top-performing hits, and validated our ﬁndings in human primary cells (human ﬁbroblasts and adult endothelial cells) in reprogram- ming assays probing three different lineages (cardiac, neural, iPSC). This approach led us to identify a conserved set of 4 TFs (ATF7IP,\n",
      "Human: /force RAG What are the potential applications of cell reprogramming in regenerative medicine?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "through a plas- tic intermediate. Moreover, although the in vivo hematopoietic niche can provide all signals neces- sary for reprogramming, we cannot dissect the niche in this study, and thus cannot determine the ele- ments that support reprogramming. Furthermore, the inability to generate any hematopoietic cells other than macrophages and monocyte-like pro- genitors also limits future studies with this process unless they focus on these lineages.\n",
      "\n",
      "Hemogenic induction As previously stated, our laboratory became the ﬁrst to induce a hemogenic program in somatic cells via nonpluripotency TF overexpression in mouse embryonic ﬁbroblasts (MEFs). We ﬁrst performed a large screen by transducing\n",
      "\n",
      "application. Functionally, these cells only possess short-term engraftment potential, neces- sitating further studies into this reprogramming method.\n",
      "\n",
      "We have graphically summarized the above- discussed efforts in Figure 4 and present speciﬁc reprogramming strategies. TFs, starting cells, cul- ture conditions, media, and cytokines, as well as the outcomes in terms of lineage production and engraftment, are compiled in Table 1.\n",
      "\n",
      "32\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Perspectives on the various reprogramming strategies\n",
      "\n",
      "Obtaining HSCs de novo remains the holy grail of regenerative HSC biology because of the multiple issues associated with HSC transplant and HLA matching. Generating and studying these cells in\n",
      "\n",
      "C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "31\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "Table 1. Reprogramming strategies and materials\n",
      "\n",
      "comprising Erg, Gata2, Lmo2, Runx1c, and Scl was demonstrated by the group of George Lacaud. As in our studies, these reprogrammed cells appeared to travel through an HE intermediate and had an endothelial and hematopoietic gene expression pro- ﬁle. The generated progenitors formed erythroid, megakaryocytic, and myeloid cells, and could form B and T cells upon co-culture with OP9 and OP9- DL1 stroma, respectively. The use of p53 null ﬁbrob- lasts made reprogramming more efﬁcient and more rapid, shortening the transition from 8 to 5 days.87 It was\n",
      "\n",
      "equally to this manuscript.\n",
      "\n",
      "doi: 10.1111/nyas.12989\n",
      "\n",
      "24\n",
      "\n",
      "Ann. N.Y. Acad. Sci. 1370 (2016) 24–35 C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "Figure 1. Established TF reprogramming strategies. TF reprogramming has been shown to work across a variety of cell types. This ﬁgure shows the starting and end cell populations, as well as the TF cocktails used to derive each cell type of choice. Adapted from Ref. 25.\n",
      "\n",
      "The novel technology of induced pluripotent stem cells (iPSCs) from Yamanaka and colleagues instigated the entire ﬁeld of regenerative medicine. Using only four transcription factors (TFs), Oct3/4, Sox2, Klf4, and c-Myc,\n",
      "\n",
      "Cell Stem Cell 1, 55–70 (2007). 161. Park, I. H. et al. Reprogramming of human somatic\n",
      "\n",
      "cells to pluripotency with defined factors. Nature 451, 141–146 (2008).\n",
      "\n",
      "162. Hanna, J. et al. Direct cell reprogramming is a stochastic process amenable to acceleration. Nature 462, 595–601 (2009).\n",
      "\n",
      "163. Yang, J. et al. Stat3 activation is limiting for reprogramming to ground state pluripotency. Cell Stem Cell 7, 319–328 (2010).\n",
      "\n",
      "164. Silva, J. et al. Promotion of reprogramming to ground state pluripotency by signal inhibition. PLoS Biol. 6, e253 (2008).\n",
      "\n",
      "Acknowledgements The authors would like to thank all Yamanaka laboratory members, past and present, and are also grateful to Y. Miyake, R. Kato, E.\n",
      "\n",
      "that Drive Neuronal Fate and Reprogramming. Cell Stem . Cell ​ .\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "41. A. Ebrahimi, K. Sevinç, G. Gürhan Sevinç, A. P. Cribbs, M. Philpott, F. Uyulur, T. Morova, J. E. Dunford, S. Göklemez, Ş. Arı, U. Oppermann, T. T. Önder, Bromodomain inhibition of the coactivators CBP/EP300 facilitate cellular reprogramming. 519–528 (2019).\n",
      "\n",
      "Nat. Chem. Biol. ​\n",
      "\n",
      "15\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      ",\n",
      "\n",
      "​\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made\n",
      "\n",
      "available under a\n",
      "\n",
      "CC-BY\n",
      "\n",
      "cells. This work augments study of the hemangioblast as well as hematopoiesis in vitro.\n",
      "\n",
      "Rapid HSC reprogramming Rapid reprogramming of fetal and adult mouse into HSPCs using a TF cocktail ﬁbroblasts\n",
      "\n",
      "Ann. N.Y. Acad. Sci. 1370 (2016) 24–35 C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "ESC/iPSC\n",
      "\n",
      "Slukvin - GATA2 + ETV2 or - GATA2 + TAL1\n",
      "\n",
      "H\n",
      "\n",
      "Thomson - Gata2 - Lmo2 - Mycn - Pitx2 - Sox17 - Tal1\n",
      "\n",
      "M\n",
      "\n",
      "Daley - HoxB4\n",
      "\n",
      "M\n",
      "\n",
      "CD34+CD45+ Myeloid Progenitor\n",
      "\n",
      "Hemogenic Endothelium\n",
      "\n",
      "Endothelial cell\n",
      "\n",
      "Daley\n",
      "\n",
      "Moore - Gata2 - Gfi1b - cFos - Etv6\n",
      "\n",
      "M\n",
      "\n",
      "Lacaud - Erg - Gata2 - Lmo2 - Runx1c - Scl\n",
      "\n",
      "M\n",
      "\n",
      "Hemangioblast\n",
      "\n",
      "HSC\n",
      "\n",
      "Rafii - FOSB - GFI1 - RUNX1 - SPI1\n",
      "\n",
      "hemogenic induc- tion (Lis et al., 2017; Pereira et al., 2013; Sandler et al., 2014) and has recently been implicated during EHT from mouse ESCs (Obier et al., 2016). Our studies corroborate that direct cellular reprogramming informs the process of HSC speciﬁcation and of- fers tractable means to systematically interrogate pathways of human HSC formation. It will be interesting to address the role of other enriched pathways such as cytoskeletal regulation by Rho GTPase and the histamine H1 receptor pathway. Genes that are co-targeted during hematopoietic reprogramming by GATA2 and GFI1B are additional candidates for further study (e.g., ITGA2, PLCL1, ADRA1B, RHOJ, SEMA3A, and PLCB1; see\n",
      "\n",
      "and tran- scription to oppose the lineage reprogramming process. Finally, we will test whether the targeted inhibition of such fate stabilizers might enhance our ability to use reprogramming TFs to improve adult organ function post-injury.\n",
      "\n",
      "To uncover fate stabilizers, we employed a genome-wide TF siRNA screen in a cardiac reprogramming (CR) assay using mouse embryonic ﬁbroblasts, followed by the combinatorial evaluation of top-performing hits, and validated our ﬁndings in human primary cells (human ﬁbroblasts and adult endothelial cells) in reprogram- ming assays probing three different lineages (cardiac, neural, iPSC). This approach led us to identify a conserved set of 4 TFs (ATF7IP,\n",
      "Human: /force RAG What are the potential applications of cell reprogramming in regenerative medicine?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "through a plas- tic intermediate. Moreover, although the in vivo hematopoietic niche can provide all signals neces- sary for reprogramming, we cannot dissect the niche in this study, and thus cannot determine the ele- ments that support reprogramming. Furthermore, the inability to generate any hematopoietic cells other than macrophages and monocyte-like pro- genitors also limits future studies with this process unless they focus on these lineages.\n",
      "\n",
      "Hemogenic induction As previously stated, our laboratory became the ﬁrst to induce a hemogenic program in somatic cells via nonpluripotency TF overexpression in mouse embryonic ﬁbroblasts (MEFs). We ﬁrst performed a large screen by transducing\n",
      "\n",
      "application. Functionally, these cells only possess short-term engraftment potential, neces- sitating further studies into this reprogramming method.\n",
      "\n",
      "We have graphically summarized the above- discussed efforts in Figure 4 and present speciﬁc reprogramming strategies. TFs, starting cells, cul- ture conditions, media, and cytokines, as well as the outcomes in terms of lineage production and engraftment, are compiled in Table 1.\n",
      "\n",
      "32\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Perspectives on the various reprogramming strategies\n",
      "\n",
      "Obtaining HSCs de novo remains the holy grail of regenerative HSC biology because of the multiple issues associated with HSC transplant and HLA matching. Generating and studying these cells in\n",
      "\n",
      "capable of multilineage engraftment potential, however, limits their use.\n",
      "\n",
      "TF reprogramming somatic cells\n",
      "\n",
      "Building upon the framework established by other studies of cell fate conversion (Fig. 1), our labora- tory was the ﬁrst to attempt somatic cell program- ming into hematopoietic stem and progenitor cells (HSPCs) using nonpluripotent TFs.26 Several labo- ratories have followed and are now trying to directly reprogram somatic cells into HSCs via TF overex- pression. Most of these studies bypass a pluripo- tent state by not using pluripotency factors, which can generate plastic intermediates.72,73 Variations between each strategy, such as starting cell popula- tions, TF cocktails, or\n",
      "\n",
      "equally to this manuscript.\n",
      "\n",
      "doi: 10.1111/nyas.12989\n",
      "\n",
      "24\n",
      "\n",
      "Ann. N.Y. Acad. Sci. 1370 (2016) 24–35 C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "Figure 1. Established TF reprogramming strategies. TF reprogramming has been shown to work across a variety of cell types. This ﬁgure shows the starting and end cell populations, as well as the TF cocktails used to derive each cell type of choice. Adapted from Ref. 25.\n",
      "\n",
      "The novel technology of induced pluripotent stem cells (iPSCs) from Yamanaka and colleagues instigated the entire ﬁeld of regenerative medicine. Using only four transcription factors (TFs), Oct3/4, Sox2, Klf4, and c-Myc,\n",
      "\n",
      "Cell Stem Cell 1, 55–70 (2007). 161. Park, I. H. et al. Reprogramming of human somatic\n",
      "\n",
      "cells to pluripotency with defined factors. Nature 451, 141–146 (2008).\n",
      "\n",
      "162. Hanna, J. et al. Direct cell reprogramming is a stochastic process amenable to acceleration. Nature 462, 595–601 (2009).\n",
      "\n",
      "163. Yang, J. et al. Stat3 activation is limiting for reprogramming to ground state pluripotency. Cell Stem Cell 7, 319–328 (2010).\n",
      "\n",
      "164. Silva, J. et al. Promotion of reprogramming to ground state pluripotency by signal inhibition. PLoS Biol. 6, e253 (2008).\n",
      "\n",
      "Acknowledgements The authors would like to thank all Yamanaka laboratory members, past and present, and are also grateful to Y. Miyake, R. Kato, E.\n",
      "\n",
      "that Drive Neuronal Fate and Reprogramming. Cell Stem . Cell ​ .\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "41. A. Ebrahimi, K. Sevinç, G. Gürhan Sevinç, A. P. Cribbs, M. Philpott, F. Uyulur, T. Morova, J. E. Dunford, S. Göklemez, Ş. Arı, U. Oppermann, T. T. Önder, Bromodomain inhibition of the coactivators CBP/EP300 facilitate cellular reprogramming. 519–528 (2019).\n",
      "\n",
      "Nat. Chem. Biol. ​\n",
      "\n",
      "15\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      ",\n",
      "\n",
      "​\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made\n",
      "\n",
      "available under a\n",
      "\n",
      "CC-BY\n",
      "\n",
      "cells. This work augments study of the hemangioblast as well as hematopoiesis in vitro.\n",
      "\n",
      "Rapid HSC reprogramming Rapid reprogramming of fetal and adult mouse into HSPCs using a TF cocktail ﬁbroblasts\n",
      "\n",
      "Ann. N.Y. Acad. Sci. 1370 (2016) 24–35 C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "ESC/iPSC\n",
      "\n",
      "Slukvin - GATA2 + ETV2 or - GATA2 + TAL1\n",
      "\n",
      "H\n",
      "\n",
      "Thomson - Gata2 - Lmo2 - Mycn - Pitx2 - Sox17 - Tal1\n",
      "\n",
      "M\n",
      "\n",
      "Daley - HoxB4\n",
      "\n",
      "M\n",
      "\n",
      "CD34+CD45+ Myeloid Progenitor\n",
      "\n",
      "Hemogenic Endothelium\n",
      "\n",
      "Endothelial cell\n",
      "\n",
      "Daley\n",
      "\n",
      "Moore - Gata2 - Gfi1b - cFos - Etv6\n",
      "\n",
      "M\n",
      "\n",
      "Lacaud - Erg - Gata2 - Lmo2 - Runx1c - Scl\n",
      "\n",
      "M\n",
      "\n",
      "Hemangioblast\n",
      "\n",
      "HSC\n",
      "\n",
      "Rafii - FOSB - GFI1 - RUNX1 - SPI1\n",
      "\n",
      "hemogenic induc- tion (Lis et al., 2017; Pereira et al., 2013; Sandler et al., 2014) and has recently been implicated during EHT from mouse ESCs (Obier et al., 2016). Our studies corroborate that direct cellular reprogramming informs the process of HSC speciﬁcation and of- fers tractable means to systematically interrogate pathways of human HSC formation. It will be interesting to address the role of other enriched pathways such as cytoskeletal regulation by Rho GTPase and the histamine H1 receptor pathway. Genes that are co-targeted during hematopoietic reprogramming by GATA2 and GFI1B are additional candidates for further study (e.g., ITGA2, PLCL1, ADRA1B, RHOJ, SEMA3A, and PLCB1; see\n",
      "\n",
      "and tran- scription to oppose the lineage reprogramming process. Finally, we will test whether the targeted inhibition of such fate stabilizers might enhance our ability to use reprogramming TFs to improve adult organ function post-injury.\n",
      "\n",
      "To uncover fate stabilizers, we employed a genome-wide TF siRNA screen in a cardiac reprogramming (CR) assay using mouse embryonic ﬁbroblasts, followed by the combinatorial evaluation of top-performing hits, and validated our ﬁndings in human primary cells (human ﬁbroblasts and adult endothelial cells) in reprogram- ming assays probing three different lineages (cardiac, neural, iPSC). This approach led us to identify a conserved set of 4 TFs (ATF7IP,\n",
      "Human: /force RAG What are the potential applications of cell reprogramming in regenerative medicine?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "through a plas- tic intermediate. Moreover, although the in vivo hematopoietic niche can provide all signals neces- sary for reprogramming, we cannot dissect the niche in this study, and thus cannot determine the ele- ments that support reprogramming. Furthermore, the inability to generate any hematopoietic cells other than macrophages and monocyte-like pro- genitors also limits future studies with this process unless they focus on these lineages.\n",
      "\n",
      "Hemogenic induction As previously stated, our laboratory became the ﬁrst to induce a hemogenic program in somatic cells via nonpluripotency TF overexpression in mouse embryonic ﬁbroblasts (MEFs). We ﬁrst performed a large screen by transducing\n",
      "\n",
      "application. Functionally, these cells only possess short-term engraftment potential, neces- sitating further studies into this reprogramming method.\n",
      "\n",
      "We have graphically summarized the above- discussed efforts in Figure 4 and present speciﬁc reprogramming strategies. TFs, starting cells, cul- ture conditions, media, and cytokines, as well as the outcomes in terms of lineage production and engraftment, are compiled in Table 1.\n",
      "\n",
      "32\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Perspectives on the various reprogramming strategies\n",
      "\n",
      "Obtaining HSCs de novo remains the holy grail of regenerative HSC biology because of the multiple issues associated with HSC transplant and HLA matching. Generating and studying these cells in\n",
      "\n",
      "capable of multilineage engraftment potential, however, limits their use.\n",
      "\n",
      "TF reprogramming somatic cells\n",
      "\n",
      "Building upon the framework established by other studies of cell fate conversion (Fig. 1), our labora- tory was the ﬁrst to attempt somatic cell program- ming into hematopoietic stem and progenitor cells (HSPCs) using nonpluripotent TFs.26 Several labo- ratories have followed and are now trying to directly reprogram somatic cells into HSCs via TF overex- pression. Most of these studies bypass a pluripo- tent state by not using pluripotency factors, which can generate plastic intermediates.72,73 Variations between each strategy, such as starting cell popula- tions, TF cocktails, or\n",
      "\n",
      "C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "31\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "Table 1. Reprogramming strategies and materials\n",
      "\n",
      "comprising Erg, Gata2, Lmo2, Runx1c, and Scl was demonstrated by the group of George Lacaud. As in our studies, these reprogrammed cells appeared to travel through an HE intermediate and had an endothelial and hematopoietic gene expression pro- ﬁle. The generated progenitors formed erythroid, megakaryocytic, and myeloid cells, and could form B and T cells upon co-culture with OP9 and OP9- DL1 stroma, respectively. The use of p53 null ﬁbrob- lasts made reprogramming more efﬁcient and more rapid, shortening the transition from 8 to 5 days.87 It was\n",
      "\n",
      "Cell Stem Cell 1, 55–70 (2007). 161. Park, I. H. et al. Reprogramming of human somatic\n",
      "\n",
      "cells to pluripotency with defined factors. Nature 451, 141–146 (2008).\n",
      "\n",
      "162. Hanna, J. et al. Direct cell reprogramming is a stochastic process amenable to acceleration. Nature 462, 595–601 (2009).\n",
      "\n",
      "163. Yang, J. et al. Stat3 activation is limiting for reprogramming to ground state pluripotency. Cell Stem Cell 7, 319–328 (2010).\n",
      "\n",
      "164. Silva, J. et al. Promotion of reprogramming to ground state pluripotency by signal inhibition. PLoS Biol. 6, e253 (2008).\n",
      "\n",
      "Acknowledgements The authors would like to thank all Yamanaka laboratory members, past and present, and are also grateful to Y. Miyake, R. Kato, E.\n",
      "\n",
      "that Drive Neuronal Fate and Reprogramming. Cell Stem . Cell ​ .\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "41. A. Ebrahimi, K. Sevinç, G. Gürhan Sevinç, A. P. Cribbs, M. Philpott, F. Uyulur, T. Morova, J. E. Dunford, S. Göklemez, Ş. Arı, U. Oppermann, T. T. Önder, Bromodomain inhibition of the coactivators CBP/EP300 facilitate cellular reprogramming. 519–528 (2019).\n",
      "\n",
      "Nat. Chem. Biol. ​\n",
      "\n",
      "15\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      ",\n",
      "\n",
      "​\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made\n",
      "\n",
      "available under a\n",
      "\n",
      "CC-BY\n",
      "\n",
      "cells. This work augments study of the hemangioblast as well as hematopoiesis in vitro.\n",
      "\n",
      "Rapid HSC reprogramming Rapid reprogramming of fetal and adult mouse into HSPCs using a TF cocktail ﬁbroblasts\n",
      "\n",
      "Ann. N.Y. Acad. Sci. 1370 (2016) 24–35 C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "ESC/iPSC\n",
      "\n",
      "Slukvin - GATA2 + ETV2 or - GATA2 + TAL1\n",
      "\n",
      "H\n",
      "\n",
      "Thomson - Gata2 - Lmo2 - Mycn - Pitx2 - Sox17 - Tal1\n",
      "\n",
      "M\n",
      "\n",
      "Daley - HoxB4\n",
      "\n",
      "M\n",
      "\n",
      "CD34+CD45+ Myeloid Progenitor\n",
      "\n",
      "Hemogenic Endothelium\n",
      "\n",
      "Endothelial cell\n",
      "\n",
      "Daley\n",
      "\n",
      "Moore - Gata2 - Gfi1b - cFos - Etv6\n",
      "\n",
      "M\n",
      "\n",
      "Lacaud - Erg - Gata2 - Lmo2 - Runx1c - Scl\n",
      "\n",
      "M\n",
      "\n",
      "Hemangioblast\n",
      "\n",
      "HSC\n",
      "\n",
      "Rafii - FOSB - GFI1 - RUNX1 - SPI1\n",
      "\n",
      "hemogenic induc- tion (Lis et al., 2017; Pereira et al., 2013; Sandler et al., 2014) and has recently been implicated during EHT from mouse ESCs (Obier et al., 2016). Our studies corroborate that direct cellular reprogramming informs the process of HSC speciﬁcation and of- fers tractable means to systematically interrogate pathways of human HSC formation. It will be interesting to address the role of other enriched pathways such as cytoskeletal regulation by Rho GTPase and the histamine H1 receptor pathway. Genes that are co-targeted during hematopoietic reprogramming by GATA2 and GFI1B are additional candidates for further study (e.g., ITGA2, PLCL1, ADRA1B, RHOJ, SEMA3A, and PLCB1; see\n",
      "\n",
      "and tran- scription to oppose the lineage reprogramming process. Finally, we will test whether the targeted inhibition of such fate stabilizers might enhance our ability to use reprogramming TFs to improve adult organ function post-injury.\n",
      "\n",
      "To uncover fate stabilizers, we employed a genome-wide TF siRNA screen in a cardiac reprogramming (CR) assay using mouse embryonic ﬁbroblasts, followed by the combinatorial evaluation of top-performing hits, and validated our ﬁndings in human primary cells (human ﬁbroblasts and adult endothelial cells) in reprogram- ming assays probing three different lineages (cardiac, neural, iPSC). This approach led us to identify a conserved set of 4 TFs (ATF7IP,\n",
      "Human: /force RAG What are the potential applications of cell reprogramming in regenerative medicine?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "through a plas- tic intermediate. Moreover, although the in vivo hematopoietic niche can provide all signals neces- sary for reprogramming, we cannot dissect the niche in this study, and thus cannot determine the ele- ments that support reprogramming. Furthermore, the inability to generate any hematopoietic cells other than macrophages and monocyte-like pro- genitors also limits future studies with this process unless they focus on these lineages.\n",
      "\n",
      "Hemogenic induction As previously stated, our laboratory became the ﬁrst to induce a hemogenic program in somatic cells via nonpluripotency TF overexpression in mouse embryonic ﬁbroblasts (MEFs). We ﬁrst performed a large screen by transducing\n",
      "\n",
      "application. Functionally, these cells only possess short-term engraftment potential, neces- sitating further studies into this reprogramming method.\n",
      "\n",
      "We have graphically summarized the above- discussed efforts in Figure 4 and present speciﬁc reprogramming strategies. TFs, starting cells, cul- ture conditions, media, and cytokines, as well as the outcomes in terms of lineage production and engraftment, are compiled in Table 1.\n",
      "\n",
      "32\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Perspectives on the various reprogramming strategies\n",
      "\n",
      "Obtaining HSCs de novo remains the holy grail of regenerative HSC biology because of the multiple issues associated with HSC transplant and HLA matching. Generating and studying these cells in\n",
      "\n",
      "capable of multilineage engraftment potential, however, limits their use.\n",
      "\n",
      "TF reprogramming somatic cells\n",
      "\n",
      "Building upon the framework established by other studies of cell fate conversion (Fig. 1), our labora- tory was the ﬁrst to attempt somatic cell program- ming into hematopoietic stem and progenitor cells (HSPCs) using nonpluripotent TFs.26 Several labo- ratories have followed and are now trying to directly reprogram somatic cells into HSCs via TF overex- pression. Most of these studies bypass a pluripo- tent state by not using pluripotency factors, which can generate plastic intermediates.72,73 Variations between each strategy, such as starting cell popula- tions, TF cocktails, or\n",
      "\n",
      "C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "31\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "Table 1. Reprogramming strategies and materials\n",
      "\n",
      "comprising Erg, Gata2, Lmo2, Runx1c, and Scl was demonstrated by the group of George Lacaud. As in our studies, these reprogrammed cells appeared to travel through an HE intermediate and had an endothelial and hematopoietic gene expression pro- ﬁle. The generated progenitors formed erythroid, megakaryocytic, and myeloid cells, and could form B and T cells upon co-culture with OP9 and OP9- DL1 stroma, respectively. The use of p53 null ﬁbrob- lasts made reprogramming more efﬁcient and more rapid, shortening the transition from 8 to 5 days.87 It was\n",
      "\n",
      "equally to this manuscript.\n",
      "\n",
      "doi: 10.1111/nyas.12989\n",
      "\n",
      "24\n",
      "\n",
      "Ann. N.Y. Acad. Sci. 1370 (2016) 24–35 C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "Figure 1. Established TF reprogramming strategies. TF reprogramming has been shown to work across a variety of cell types. This ﬁgure shows the starting and end cell populations, as well as the TF cocktails used to derive each cell type of choice. Adapted from Ref. 25.\n",
      "\n",
      "The novel technology of induced pluripotent stem cells (iPSCs) from Yamanaka and colleagues instigated the entire ﬁeld of regenerative medicine. Using only four transcription factors (TFs), Oct3/4, Sox2, Klf4, and c-Myc,\n",
      "\n",
      "that Drive Neuronal Fate and Reprogramming. Cell Stem . Cell ​ .\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "41. A. Ebrahimi, K. Sevinç, G. Gürhan Sevinç, A. P. Cribbs, M. Philpott, F. Uyulur, T. Morova, J. E. Dunford, S. Göklemez, Ş. Arı, U. Oppermann, T. T. Önder, Bromodomain inhibition of the coactivators CBP/EP300 facilitate cellular reprogramming. 519–528 (2019).\n",
      "\n",
      "Nat. Chem. Biol. ​\n",
      "\n",
      "15\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      ",\n",
      "\n",
      "​\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made\n",
      "\n",
      "available under a\n",
      "\n",
      "CC-BY\n",
      "\n",
      "cells. This work augments study of the hemangioblast as well as hematopoiesis in vitro.\n",
      "\n",
      "Rapid HSC reprogramming Rapid reprogramming of fetal and adult mouse into HSPCs using a TF cocktail ﬁbroblasts\n",
      "\n",
      "Ann. N.Y. Acad. Sci. 1370 (2016) 24–35 C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "ESC/iPSC\n",
      "\n",
      "Slukvin - GATA2 + ETV2 or - GATA2 + TAL1\n",
      "\n",
      "H\n",
      "\n",
      "Thomson - Gata2 - Lmo2 - Mycn - Pitx2 - Sox17 - Tal1\n",
      "\n",
      "M\n",
      "\n",
      "Daley - HoxB4\n",
      "\n",
      "M\n",
      "\n",
      "CD34+CD45+ Myeloid Progenitor\n",
      "\n",
      "Hemogenic Endothelium\n",
      "\n",
      "Endothelial cell\n",
      "\n",
      "Daley\n",
      "\n",
      "Moore - Gata2 - Gfi1b - cFos - Etv6\n",
      "\n",
      "M\n",
      "\n",
      "Lacaud - Erg - Gata2 - Lmo2 - Runx1c - Scl\n",
      "\n",
      "M\n",
      "\n",
      "Hemangioblast\n",
      "\n",
      "HSC\n",
      "\n",
      "Rafii - FOSB - GFI1 - RUNX1 - SPI1\n",
      "\n",
      "hemogenic induc- tion (Lis et al., 2017; Pereira et al., 2013; Sandler et al., 2014) and has recently been implicated during EHT from mouse ESCs (Obier et al., 2016). Our studies corroborate that direct cellular reprogramming informs the process of HSC speciﬁcation and of- fers tractable means to systematically interrogate pathways of human HSC formation. It will be interesting to address the role of other enriched pathways such as cytoskeletal regulation by Rho GTPase and the histamine H1 receptor pathway. Genes that are co-targeted during hematopoietic reprogramming by GATA2 and GFI1B are additional candidates for further study (e.g., ITGA2, PLCL1, ADRA1B, RHOJ, SEMA3A, and PLCB1; see\n",
      "\n",
      "and tran- scription to oppose the lineage reprogramming process. Finally, we will test whether the targeted inhibition of such fate stabilizers might enhance our ability to use reprogramming TFs to improve adult organ function post-injury.\n",
      "\n",
      "To uncover fate stabilizers, we employed a genome-wide TF siRNA screen in a cardiac reprogramming (CR) assay using mouse embryonic ﬁbroblasts, followed by the combinatorial evaluation of top-performing hits, and validated our ﬁndings in human primary cells (human ﬁbroblasts and adult endothelial cells) in reprogram- ming assays probing three different lineages (cardiac, neural, iPSC). This approach led us to identify a conserved set of 4 TFs (ATF7IP,\n",
      "Human: /force RAG What are the potential applications of cell reprogramming in regenerative medicine?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "through a plas- tic intermediate. Moreover, although the in vivo hematopoietic niche can provide all signals neces- sary for reprogramming, we cannot dissect the niche in this study, and thus cannot determine the ele- ments that support reprogramming. Furthermore, the inability to generate any hematopoietic cells other than macrophages and monocyte-like pro- genitors also limits future studies with this process unless they focus on these lineages.\n",
      "\n",
      "Hemogenic induction As previously stated, our laboratory became the ﬁrst to induce a hemogenic program in somatic cells via nonpluripotency TF overexpression in mouse embryonic ﬁbroblasts (MEFs). We ﬁrst performed a large screen by transducing\n",
      "\n",
      "application. Functionally, these cells only possess short-term engraftment potential, neces- sitating further studies into this reprogramming method.\n",
      "\n",
      "We have graphically summarized the above- discussed efforts in Figure 4 and present speciﬁc reprogramming strategies. TFs, starting cells, cul- ture conditions, media, and cytokines, as well as the outcomes in terms of lineage production and engraftment, are compiled in Table 1.\n",
      "\n",
      "32\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Perspectives on the various reprogramming strategies\n",
      "\n",
      "Obtaining HSCs de novo remains the holy grail of regenerative HSC biology because of the multiple issues associated with HSC transplant and HLA matching. Generating and studying these cells in\n",
      "\n",
      "capable of multilineage engraftment potential, however, limits their use.\n",
      "\n",
      "TF reprogramming somatic cells\n",
      "\n",
      "Building upon the framework established by other studies of cell fate conversion (Fig. 1), our labora- tory was the ﬁrst to attempt somatic cell program- ming into hematopoietic stem and progenitor cells (HSPCs) using nonpluripotent TFs.26 Several labo- ratories have followed and are now trying to directly reprogram somatic cells into HSCs via TF overex- pression. Most of these studies bypass a pluripo- tent state by not using pluripotency factors, which can generate plastic intermediates.72,73 Variations between each strategy, such as starting cell popula- tions, TF cocktails, or\n",
      "\n",
      "C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "31\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "Table 1. Reprogramming strategies and materials\n",
      "\n",
      "comprising Erg, Gata2, Lmo2, Runx1c, and Scl was demonstrated by the group of George Lacaud. As in our studies, these reprogrammed cells appeared to travel through an HE intermediate and had an endothelial and hematopoietic gene expression pro- ﬁle. The generated progenitors formed erythroid, megakaryocytic, and myeloid cells, and could form B and T cells upon co-culture with OP9 and OP9- DL1 stroma, respectively. The use of p53 null ﬁbrob- lasts made reprogramming more efﬁcient and more rapid, shortening the transition from 8 to 5 days.87 It was\n",
      "\n",
      "equally to this manuscript.\n",
      "\n",
      "doi: 10.1111/nyas.12989\n",
      "\n",
      "24\n",
      "\n",
      "Ann. N.Y. Acad. Sci. 1370 (2016) 24–35 C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "Figure 1. Established TF reprogramming strategies. TF reprogramming has been shown to work across a variety of cell types. This ﬁgure shows the starting and end cell populations, as well as the TF cocktails used to derive each cell type of choice. Adapted from Ref. 25.\n",
      "\n",
      "The novel technology of induced pluripotent stem cells (iPSCs) from Yamanaka and colleagues instigated the entire ﬁeld of regenerative medicine. Using only four transcription factors (TFs), Oct3/4, Sox2, Klf4, and c-Myc,\n",
      "\n",
      "Cell Stem Cell 1, 55–70 (2007). 161. Park, I. H. et al. Reprogramming of human somatic\n",
      "\n",
      "cells to pluripotency with defined factors. Nature 451, 141–146 (2008).\n",
      "\n",
      "162. Hanna, J. et al. Direct cell reprogramming is a stochastic process amenable to acceleration. Nature 462, 595–601 (2009).\n",
      "\n",
      "163. Yang, J. et al. Stat3 activation is limiting for reprogramming to ground state pluripotency. Cell Stem Cell 7, 319–328 (2010).\n",
      "\n",
      "164. Silva, J. et al. Promotion of reprogramming to ground state pluripotency by signal inhibition. PLoS Biol. 6, e253 (2008).\n",
      "\n",
      "Acknowledgements The authors would like to thank all Yamanaka laboratory members, past and present, and are also grateful to Y. Miyake, R. Kato, E.\n",
      "\n",
      "cells. This work augments study of the hemangioblast as well as hematopoiesis in vitro.\n",
      "\n",
      "Rapid HSC reprogramming Rapid reprogramming of fetal and adult mouse into HSPCs using a TF cocktail ﬁbroblasts\n",
      "\n",
      "Ann. N.Y. Acad. Sci. 1370 (2016) 24–35 C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "ESC/iPSC\n",
      "\n",
      "Slukvin - GATA2 + ETV2 or - GATA2 + TAL1\n",
      "\n",
      "H\n",
      "\n",
      "Thomson - Gata2 - Lmo2 - Mycn - Pitx2 - Sox17 - Tal1\n",
      "\n",
      "M\n",
      "\n",
      "Daley - HoxB4\n",
      "\n",
      "M\n",
      "\n",
      "CD34+CD45+ Myeloid Progenitor\n",
      "\n",
      "Hemogenic Endothelium\n",
      "\n",
      "Endothelial cell\n",
      "\n",
      "Daley\n",
      "\n",
      "Moore - Gata2 - Gfi1b - cFos - Etv6\n",
      "\n",
      "M\n",
      "\n",
      "Lacaud - Erg - Gata2 - Lmo2 - Runx1c - Scl\n",
      "\n",
      "M\n",
      "\n",
      "Hemangioblast\n",
      "\n",
      "HSC\n",
      "\n",
      "Rafii - FOSB - GFI1 - RUNX1 - SPI1\n",
      "\n",
      "hemogenic induc- tion (Lis et al., 2017; Pereira et al., 2013; Sandler et al., 2014) and has recently been implicated during EHT from mouse ESCs (Obier et al., 2016). Our studies corroborate that direct cellular reprogramming informs the process of HSC speciﬁcation and of- fers tractable means to systematically interrogate pathways of human HSC formation. It will be interesting to address the role of other enriched pathways such as cytoskeletal regulation by Rho GTPase and the histamine H1 receptor pathway. Genes that are co-targeted during hematopoietic reprogramming by GATA2 and GFI1B are additional candidates for further study (e.g., ITGA2, PLCL1, ADRA1B, RHOJ, SEMA3A, and PLCB1; see\n",
      "\n",
      "and tran- scription to oppose the lineage reprogramming process. Finally, we will test whether the targeted inhibition of such fate stabilizers might enhance our ability to use reprogramming TFs to improve adult organ function post-injury.\n",
      "\n",
      "To uncover fate stabilizers, we employed a genome-wide TF siRNA screen in a cardiac reprogramming (CR) assay using mouse embryonic ﬁbroblasts, followed by the combinatorial evaluation of top-performing hits, and validated our ﬁndings in human primary cells (human ﬁbroblasts and adult endothelial cells) in reprogram- ming assays probing three different lineages (cardiac, neural, iPSC). This approach led us to identify a conserved set of 4 TFs (ATF7IP,\n",
      "Human: /force RAG What are the potential applications of cell reprogramming in regenerative medicine?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "through a plas- tic intermediate. Moreover, although the in vivo hematopoietic niche can provide all signals neces- sary for reprogramming, we cannot dissect the niche in this study, and thus cannot determine the ele- ments that support reprogramming. Furthermore, the inability to generate any hematopoietic cells other than macrophages and monocyte-like pro- genitors also limits future studies with this process unless they focus on these lineages.\n",
      "\n",
      "Hemogenic induction As previously stated, our laboratory became the ﬁrst to induce a hemogenic program in somatic cells via nonpluripotency TF overexpression in mouse embryonic ﬁbroblasts (MEFs). We ﬁrst performed a large screen by transducing\n",
      "\n",
      "application. Functionally, these cells only possess short-term engraftment potential, neces- sitating further studies into this reprogramming method.\n",
      "\n",
      "We have graphically summarized the above- discussed efforts in Figure 4 and present speciﬁc reprogramming strategies. TFs, starting cells, cul- ture conditions, media, and cytokines, as well as the outcomes in terms of lineage production and engraftment, are compiled in Table 1.\n",
      "\n",
      "32\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Perspectives on the various reprogramming strategies\n",
      "\n",
      "Obtaining HSCs de novo remains the holy grail of regenerative HSC biology because of the multiple issues associated with HSC transplant and HLA matching. Generating and studying these cells in\n",
      "\n",
      "capable of multilineage engraftment potential, however, limits their use.\n",
      "\n",
      "TF reprogramming somatic cells\n",
      "\n",
      "Building upon the framework established by other studies of cell fate conversion (Fig. 1), our labora- tory was the ﬁrst to attempt somatic cell program- ming into hematopoietic stem and progenitor cells (HSPCs) using nonpluripotent TFs.26 Several labo- ratories have followed and are now trying to directly reprogram somatic cells into HSCs via TF overex- pression. Most of these studies bypass a pluripo- tent state by not using pluripotency factors, which can generate plastic intermediates.72,73 Variations between each strategy, such as starting cell popula- tions, TF cocktails, or\n",
      "\n",
      "C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "31\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "Table 1. Reprogramming strategies and materials\n",
      "\n",
      "comprising Erg, Gata2, Lmo2, Runx1c, and Scl was demonstrated by the group of George Lacaud. As in our studies, these reprogrammed cells appeared to travel through an HE intermediate and had an endothelial and hematopoietic gene expression pro- ﬁle. The generated progenitors formed erythroid, megakaryocytic, and myeloid cells, and could form B and T cells upon co-culture with OP9 and OP9- DL1 stroma, respectively. The use of p53 null ﬁbrob- lasts made reprogramming more efﬁcient and more rapid, shortening the transition from 8 to 5 days.87 It was\n",
      "\n",
      "equally to this manuscript.\n",
      "\n",
      "doi: 10.1111/nyas.12989\n",
      "\n",
      "24\n",
      "\n",
      "Ann. N.Y. Acad. Sci. 1370 (2016) 24–35 C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "Figure 1. Established TF reprogramming strategies. TF reprogramming has been shown to work across a variety of cell types. This ﬁgure shows the starting and end cell populations, as well as the TF cocktails used to derive each cell type of choice. Adapted from Ref. 25.\n",
      "\n",
      "The novel technology of induced pluripotent stem cells (iPSCs) from Yamanaka and colleagues instigated the entire ﬁeld of regenerative medicine. Using only four transcription factors (TFs), Oct3/4, Sox2, Klf4, and c-Myc,\n",
      "\n",
      "Cell Stem Cell 1, 55–70 (2007). 161. Park, I. H. et al. Reprogramming of human somatic\n",
      "\n",
      "cells to pluripotency with defined factors. Nature 451, 141–146 (2008).\n",
      "\n",
      "162. Hanna, J. et al. Direct cell reprogramming is a stochastic process amenable to acceleration. Nature 462, 595–601 (2009).\n",
      "\n",
      "163. Yang, J. et al. Stat3 activation is limiting for reprogramming to ground state pluripotency. Cell Stem Cell 7, 319–328 (2010).\n",
      "\n",
      "164. Silva, J. et al. Promotion of reprogramming to ground state pluripotency by signal inhibition. PLoS Biol. 6, e253 (2008).\n",
      "\n",
      "Acknowledgements The authors would like to thank all Yamanaka laboratory members, past and present, and are also grateful to Y. Miyake, R. Kato, E.\n",
      "\n",
      "that Drive Neuronal Fate and Reprogramming. Cell Stem . Cell ​ .\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "41. A. Ebrahimi, K. Sevinç, G. Gürhan Sevinç, A. P. Cribbs, M. Philpott, F. Uyulur, T. Morova, J. E. Dunford, S. Göklemez, Ş. Arı, U. Oppermann, T. T. Önder, Bromodomain inhibition of the coactivators CBP/EP300 facilitate cellular reprogramming. 519–528 (2019).\n",
      "\n",
      "Nat. Chem. Biol. ​\n",
      "\n",
      "15\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      ",\n",
      "\n",
      "​\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made\n",
      "\n",
      "available under a\n",
      "\n",
      "CC-BY\n",
      "\n",
      "hemogenic induc- tion (Lis et al., 2017; Pereira et al., 2013; Sandler et al., 2014) and has recently been implicated during EHT from mouse ESCs (Obier et al., 2016). Our studies corroborate that direct cellular reprogramming informs the process of HSC speciﬁcation and of- fers tractable means to systematically interrogate pathways of human HSC formation. It will be interesting to address the role of other enriched pathways such as cytoskeletal regulation by Rho GTPase and the histamine H1 receptor pathway. Genes that are co-targeted during hematopoietic reprogramming by GATA2 and GFI1B are additional candidates for further study (e.g., ITGA2, PLCL1, ADRA1B, RHOJ, SEMA3A, and PLCB1; see\n",
      "\n",
      "and tran- scription to oppose the lineage reprogramming process. Finally, we will test whether the targeted inhibition of such fate stabilizers might enhance our ability to use reprogramming TFs to improve adult organ function post-injury.\n",
      "\n",
      "To uncover fate stabilizers, we employed a genome-wide TF siRNA screen in a cardiac reprogramming (CR) assay using mouse embryonic ﬁbroblasts, followed by the combinatorial evaluation of top-performing hits, and validated our ﬁndings in human primary cells (human ﬁbroblasts and adult endothelial cells) in reprogram- ming assays probing three different lineages (cardiac, neural, iPSC). This approach led us to identify a conserved set of 4 TFs (ATF7IP,\n",
      "Human: /force RAG What are the potential applications of cell reprogramming in regenerative medicine?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "through a plas- tic intermediate. Moreover, although the in vivo hematopoietic niche can provide all signals neces- sary for reprogramming, we cannot dissect the niche in this study, and thus cannot determine the ele- ments that support reprogramming. Furthermore, the inability to generate any hematopoietic cells other than macrophages and monocyte-like pro- genitors also limits future studies with this process unless they focus on these lineages.\n",
      "\n",
      "Hemogenic induction As previously stated, our laboratory became the ﬁrst to induce a hemogenic program in somatic cells via nonpluripotency TF overexpression in mouse embryonic ﬁbroblasts (MEFs). We ﬁrst performed a large screen by transducing\n",
      "\n",
      "application. Functionally, these cells only possess short-term engraftment potential, neces- sitating further studies into this reprogramming method.\n",
      "\n",
      "We have graphically summarized the above- discussed efforts in Figure 4 and present speciﬁc reprogramming strategies. TFs, starting cells, cul- ture conditions, media, and cytokines, as well as the outcomes in terms of lineage production and engraftment, are compiled in Table 1.\n",
      "\n",
      "32\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Perspectives on the various reprogramming strategies\n",
      "\n",
      "Obtaining HSCs de novo remains the holy grail of regenerative HSC biology because of the multiple issues associated with HSC transplant and HLA matching. Generating and studying these cells in\n",
      "\n",
      "capable of multilineage engraftment potential, however, limits their use.\n",
      "\n",
      "TF reprogramming somatic cells\n",
      "\n",
      "Building upon the framework established by other studies of cell fate conversion (Fig. 1), our labora- tory was the ﬁrst to attempt somatic cell program- ming into hematopoietic stem and progenitor cells (HSPCs) using nonpluripotent TFs.26 Several labo- ratories have followed and are now trying to directly reprogram somatic cells into HSCs via TF overex- pression. Most of these studies bypass a pluripo- tent state by not using pluripotency factors, which can generate plastic intermediates.72,73 Variations between each strategy, such as starting cell popula- tions, TF cocktails, or\n",
      "\n",
      "C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "31\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "Table 1. Reprogramming strategies and materials\n",
      "\n",
      "comprising Erg, Gata2, Lmo2, Runx1c, and Scl was demonstrated by the group of George Lacaud. As in our studies, these reprogrammed cells appeared to travel through an HE intermediate and had an endothelial and hematopoietic gene expression pro- ﬁle. The generated progenitors formed erythroid, megakaryocytic, and myeloid cells, and could form B and T cells upon co-culture with OP9 and OP9- DL1 stroma, respectively. The use of p53 null ﬁbrob- lasts made reprogramming more efﬁcient and more rapid, shortening the transition from 8 to 5 days.87 It was\n",
      "\n",
      "equally to this manuscript.\n",
      "\n",
      "doi: 10.1111/nyas.12989\n",
      "\n",
      "24\n",
      "\n",
      "Ann. N.Y. Acad. Sci. 1370 (2016) 24–35 C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "Figure 1. Established TF reprogramming strategies. TF reprogramming has been shown to work across a variety of cell types. This ﬁgure shows the starting and end cell populations, as well as the TF cocktails used to derive each cell type of choice. Adapted from Ref. 25.\n",
      "\n",
      "The novel technology of induced pluripotent stem cells (iPSCs) from Yamanaka and colleagues instigated the entire ﬁeld of regenerative medicine. Using only four transcription factors (TFs), Oct3/4, Sox2, Klf4, and c-Myc,\n",
      "\n",
      "Cell Stem Cell 1, 55–70 (2007). 161. Park, I. H. et al. Reprogramming of human somatic\n",
      "\n",
      "cells to pluripotency with defined factors. Nature 451, 141–146 (2008).\n",
      "\n",
      "162. Hanna, J. et al. Direct cell reprogramming is a stochastic process amenable to acceleration. Nature 462, 595–601 (2009).\n",
      "\n",
      "163. Yang, J. et al. Stat3 activation is limiting for reprogramming to ground state pluripotency. Cell Stem Cell 7, 319–328 (2010).\n",
      "\n",
      "164. Silva, J. et al. Promotion of reprogramming to ground state pluripotency by signal inhibition. PLoS Biol. 6, e253 (2008).\n",
      "\n",
      "Acknowledgements The authors would like to thank all Yamanaka laboratory members, past and present, and are also grateful to Y. Miyake, R. Kato, E.\n",
      "\n",
      "that Drive Neuronal Fate and Reprogramming. Cell Stem . Cell ​ .\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "41. A. Ebrahimi, K. Sevinç, G. Gürhan Sevinç, A. P. Cribbs, M. Philpott, F. Uyulur, T. Morova, J. E. Dunford, S. Göklemez, Ş. Arı, U. Oppermann, T. T. Önder, Bromodomain inhibition of the coactivators CBP/EP300 facilitate cellular reprogramming. 519–528 (2019).\n",
      "\n",
      "Nat. Chem. Biol. ​\n",
      "\n",
      "15\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      ",\n",
      "\n",
      "​\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made\n",
      "\n",
      "available under a\n",
      "\n",
      "CC-BY\n",
      "\n",
      "cells. This work augments study of the hemangioblast as well as hematopoiesis in vitro.\n",
      "\n",
      "Rapid HSC reprogramming Rapid reprogramming of fetal and adult mouse into HSPCs using a TF cocktail ﬁbroblasts\n",
      "\n",
      "Ann. N.Y. Acad. Sci. 1370 (2016) 24–35 C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "ESC/iPSC\n",
      "\n",
      "Slukvin - GATA2 + ETV2 or - GATA2 + TAL1\n",
      "\n",
      "H\n",
      "\n",
      "Thomson - Gata2 - Lmo2 - Mycn - Pitx2 - Sox17 - Tal1\n",
      "\n",
      "M\n",
      "\n",
      "Daley - HoxB4\n",
      "\n",
      "M\n",
      "\n",
      "CD34+CD45+ Myeloid Progenitor\n",
      "\n",
      "Hemogenic Endothelium\n",
      "\n",
      "Endothelial cell\n",
      "\n",
      "Daley\n",
      "\n",
      "Moore - Gata2 - Gfi1b - cFos - Etv6\n",
      "\n",
      "M\n",
      "\n",
      "Lacaud - Erg - Gata2 - Lmo2 - Runx1c - Scl\n",
      "\n",
      "M\n",
      "\n",
      "Hemangioblast\n",
      "\n",
      "HSC\n",
      "\n",
      "Rafii - FOSB - GFI1 - RUNX1 - SPI1\n",
      "\n",
      "and tran- scription to oppose the lineage reprogramming process. Finally, we will test whether the targeted inhibition of such fate stabilizers might enhance our ability to use reprogramming TFs to improve adult organ function post-injury.\n",
      "\n",
      "To uncover fate stabilizers, we employed a genome-wide TF siRNA screen in a cardiac reprogramming (CR) assay using mouse embryonic ﬁbroblasts, followed by the combinatorial evaluation of top-performing hits, and validated our ﬁndings in human primary cells (human ﬁbroblasts and adult endothelial cells) in reprogram- ming assays probing three different lineages (cardiac, neural, iPSC). This approach led us to identify a conserved set of 4 TFs (ATF7IP,\n",
      "Human: /force RAG What are the potential applications of cell reprogramming in regenerative medicine?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "through a plas- tic intermediate. Moreover, although the in vivo hematopoietic niche can provide all signals neces- sary for reprogramming, we cannot dissect the niche in this study, and thus cannot determine the ele- ments that support reprogramming. Furthermore, the inability to generate any hematopoietic cells other than macrophages and monocyte-like pro- genitors also limits future studies with this process unless they focus on these lineages.\n",
      "\n",
      "Hemogenic induction As previously stated, our laboratory became the ﬁrst to induce a hemogenic program in somatic cells via nonpluripotency TF overexpression in mouse embryonic ﬁbroblasts (MEFs). We ﬁrst performed a large screen by transducing\n",
      "\n",
      "application. Functionally, these cells only possess short-term engraftment potential, neces- sitating further studies into this reprogramming method.\n",
      "\n",
      "We have graphically summarized the above- discussed efforts in Figure 4 and present speciﬁc reprogramming strategies. TFs, starting cells, cul- ture conditions, media, and cytokines, as well as the outcomes in terms of lineage production and engraftment, are compiled in Table 1.\n",
      "\n",
      "32\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Perspectives on the various reprogramming strategies\n",
      "\n",
      "Obtaining HSCs de novo remains the holy grail of regenerative HSC biology because of the multiple issues associated with HSC transplant and HLA matching. Generating and studying these cells in\n",
      "\n",
      "capable of multilineage engraftment potential, however, limits their use.\n",
      "\n",
      "TF reprogramming somatic cells\n",
      "\n",
      "Building upon the framework established by other studies of cell fate conversion (Fig. 1), our labora- tory was the ﬁrst to attempt somatic cell program- ming into hematopoietic stem and progenitor cells (HSPCs) using nonpluripotent TFs.26 Several labo- ratories have followed and are now trying to directly reprogram somatic cells into HSCs via TF overex- pression. Most of these studies bypass a pluripo- tent state by not using pluripotency factors, which can generate plastic intermediates.72,73 Variations between each strategy, such as starting cell popula- tions, TF cocktails, or\n",
      "\n",
      "C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "31\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "Table 1. Reprogramming strategies and materials\n",
      "\n",
      "comprising Erg, Gata2, Lmo2, Runx1c, and Scl was demonstrated by the group of George Lacaud. As in our studies, these reprogrammed cells appeared to travel through an HE intermediate and had an endothelial and hematopoietic gene expression pro- ﬁle. The generated progenitors formed erythroid, megakaryocytic, and myeloid cells, and could form B and T cells upon co-culture with OP9 and OP9- DL1 stroma, respectively. The use of p53 null ﬁbrob- lasts made reprogramming more efﬁcient and more rapid, shortening the transition from 8 to 5 days.87 It was\n",
      "\n",
      "equally to this manuscript.\n",
      "\n",
      "doi: 10.1111/nyas.12989\n",
      "\n",
      "24\n",
      "\n",
      "Ann. N.Y. Acad. Sci. 1370 (2016) 24–35 C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "Figure 1. Established TF reprogramming strategies. TF reprogramming has been shown to work across a variety of cell types. This ﬁgure shows the starting and end cell populations, as well as the TF cocktails used to derive each cell type of choice. Adapted from Ref. 25.\n",
      "\n",
      "The novel technology of induced pluripotent stem cells (iPSCs) from Yamanaka and colleagues instigated the entire ﬁeld of regenerative medicine. Using only four transcription factors (TFs), Oct3/4, Sox2, Klf4, and c-Myc,\n",
      "\n",
      "Cell Stem Cell 1, 55–70 (2007). 161. Park, I. H. et al. Reprogramming of human somatic\n",
      "\n",
      "cells to pluripotency with defined factors. Nature 451, 141–146 (2008).\n",
      "\n",
      "162. Hanna, J. et al. Direct cell reprogramming is a stochastic process amenable to acceleration. Nature 462, 595–601 (2009).\n",
      "\n",
      "163. Yang, J. et al. Stat3 activation is limiting for reprogramming to ground state pluripotency. Cell Stem Cell 7, 319–328 (2010).\n",
      "\n",
      "164. Silva, J. et al. Promotion of reprogramming to ground state pluripotency by signal inhibition. PLoS Biol. 6, e253 (2008).\n",
      "\n",
      "Acknowledgements The authors would like to thank all Yamanaka laboratory members, past and present, and are also grateful to Y. Miyake, R. Kato, E.\n",
      "\n",
      "that Drive Neuronal Fate and Reprogramming. Cell Stem . Cell ​ .\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "41. A. Ebrahimi, K. Sevinç, G. Gürhan Sevinç, A. P. Cribbs, M. Philpott, F. Uyulur, T. Morova, J. E. Dunford, S. Göklemez, Ş. Arı, U. Oppermann, T. T. Önder, Bromodomain inhibition of the coactivators CBP/EP300 facilitate cellular reprogramming. 519–528 (2019).\n",
      "\n",
      "Nat. Chem. Biol. ​\n",
      "\n",
      "15\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      ",\n",
      "\n",
      "​\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made\n",
      "\n",
      "available under a\n",
      "\n",
      "CC-BY\n",
      "\n",
      "cells. This work augments study of the hemangioblast as well as hematopoiesis in vitro.\n",
      "\n",
      "Rapid HSC reprogramming Rapid reprogramming of fetal and adult mouse into HSPCs using a TF cocktail ﬁbroblasts\n",
      "\n",
      "Ann. N.Y. Acad. Sci. 1370 (2016) 24–35 C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "ESC/iPSC\n",
      "\n",
      "Slukvin - GATA2 + ETV2 or - GATA2 + TAL1\n",
      "\n",
      "H\n",
      "\n",
      "Thomson - Gata2 - Lmo2 - Mycn - Pitx2 - Sox17 - Tal1\n",
      "\n",
      "M\n",
      "\n",
      "Daley - HoxB4\n",
      "\n",
      "M\n",
      "\n",
      "CD34+CD45+ Myeloid Progenitor\n",
      "\n",
      "Hemogenic Endothelium\n",
      "\n",
      "Endothelial cell\n",
      "\n",
      "Daley\n",
      "\n",
      "Moore - Gata2 - Gfi1b - cFos - Etv6\n",
      "\n",
      "M\n",
      "\n",
      "Lacaud - Erg - Gata2 - Lmo2 - Runx1c - Scl\n",
      "\n",
      "M\n",
      "\n",
      "Hemangioblast\n",
      "\n",
      "HSC\n",
      "\n",
      "Rafii - FOSB - GFI1 - RUNX1 - SPI1\n",
      "\n",
      "hemogenic induc- tion (Lis et al., 2017; Pereira et al., 2013; Sandler et al., 2014) and has recently been implicated during EHT from mouse ESCs (Obier et al., 2016). Our studies corroborate that direct cellular reprogramming informs the process of HSC speciﬁcation and of- fers tractable means to systematically interrogate pathways of human HSC formation. It will be interesting to address the role of other enriched pathways such as cytoskeletal regulation by Rho GTPase and the histamine H1 receptor pathway. Genes that are co-targeted during hematopoietic reprogramming by GATA2 and GFI1B are additional candidates for further study (e.g., ITGA2, PLCL1, ADRA1B, RHOJ, SEMA3A, and PLCB1; see\n",
      "Human: /force RAG What are the potential applications of cell reprogramming in regenerative medicine?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "==================================================\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  How do transcription factors influence epigenetic modifications during the reprogramming process?\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fri Jun  7 23:22:18 2024 INFO RAG\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "RAG >> 7: \n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "the path toward pluripotency.\n",
      "\n",
      "Epigenetic modifiers. Cellular reprogramming is associated with epigenetic changes, and thus epigenetic modifiers have been studied with respect to their involvement in reprogramming (reviewed in REF. 65). Enzymes that regulate the post‑ translational modifications of histones (such as their acetylation or methylation), which largely affect transcription, are one\n",
      "\n",
      "example of such epigenetic modifiers66–73. These enzymes can affect (by upregulation or downregulation) the expression of pluripotency‑associated, as well as somatic, genes and thereby drive cell fate in a particular direction (more towards either pluripotency or differentiation). For example,\n",
      "\n",
      "et al. Rapid chromatin switch in the direct repro- gramming of ﬁbroblasts to neurons. Cell Rep. 20, 3236–3247 (2017).\n",
      "\n",
      "42. Dall’Agnese, A. et al. Transcription factor-directed re-wiring of\n",
      "\n",
      "chromatin architecture for somatic cell nuclear reprogramming toward trans-differentiation. Mol. Cell 76, 453–472 (2019). 43. Stone, N. R. et al. Context-speciﬁc transcription factor functions\n",
      "\n",
      "regulate epigenomic and transcriptional dynamics during cardiac reprogramming. Cell Stem Cell 25, 87–102 (2019).\n",
      "\n",
      "44. Nam, Y. J. et al. Reprogramming of human ﬁbroblasts toward a cardiac fate. Proc. Natl Acad. Sci. USA 110, 5588–5593 (2013). 45. Zhou, Y. et al. Single-cell transcriptomic analyses of cell fate\n",
      "\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "the transcrip- tion factors that are used to elicit cellular reprogramming appear to be needed only for the initial activation of the appropriate cellular networks. After initiation, the exog- enous reprogramming factors can be extinguished, while\n",
      "\n",
      "Box 2. Important issues to be addressed\n",
      "\n",
      "What is the influence of predicted high nucleosome occupancy and nucleosome position on defining the subset of target sites to which pioneer factors will bind in a cell?\n",
      "\n",
      "What are the structural features of DNA-binding domains that allow certain factors to have pioneer activity? How do domains outside of the DNA-binding domain affect pioneer activity and subsequent chromatin-opening events? How does a\n",
      "\n",
      "system: a status report. Trends Immunol. 35, 190–194.\n",
      "\n",
      "Souﬁ, A., Garcia, M.F., Jaroszewicz, A., Osman, N., Pellegrini, M., and Zaret, K.S. (2015). Pioneer transcription factors target partial DNA motifs on nucleo- somes to initiate reprogramming. Cell 161, 555–568.\n",
      "\n",
      "Spitz, F., and Furlong, E.E. (2012). Transcription factors: from enhancer bind- ing to developmental control. Nat. Rev. Genet. 13, 613–626.\n",
      "\n",
      "Stormo, G.D., and Zhao, Y. (2010). Determining the speciﬁcity of protein-DNA interactions. Nat. Rev. Genet. 11, 751–760.\n",
      "\n",
      "Stubbs, L., Sun, Y., and Caetano-Anolles, D. (2011). Function and evolution of C2H2 zinc ﬁnger arrays. Subcell. Biochem. 52, 75–94.\n",
      "\n",
      "Sur, I., and Taipale, J. (2016). The\n",
      "\n",
      "pre- ceded gene expression changes at all levels and provided an explana- tion for the sequential activation of core pluripotency genes during reprogramming. Together, our observations indicate that genome topology has an instructive role in implementing transcriptional programs relevant to cell-fate decisions in mammals.\n",
      "\n",
      "Results Transcription factors prime the epigenome for reprogramming. We exposed mouse bone marrow–derived pre-B cells to the myeloid TF C/EBPα to generate cells denoted Bα cells. Subsequent activation of OSKM has been found to induce reprogramming of nearly 100% of Bα cells into PSC-like cells within 4–8 d (refs 31,32). To obtain a high- resolution map of changes in gene\n",
      "\n",
      "J. et al. Targeted bisulfite sequencing reveals changes in DNA methylation associated with nuclear reprogramming. Nat. Biotechnol. 27, 353–360 (2009).\n",
      "\n",
      "88. Polo, J. M. et al. A molecular roadmap of\n",
      "\n",
      "reprogramming somatic cells into iPS cells. Cell 151, 1617–1632 (2012).\n",
      "\n",
      "89. Soufi, A., Donahue, G. & Zaret, K. S. Facilitators and impediments of the pluripotency reprogramming factors’ initial engagement with the genome. Cell 151, 994–1004 (2012).\n",
      "\n",
      "90. Sridharan, R. et al. Role of the murine reprogramming factors in the induction of pluripotency. Cell 136, 364–377 (2009).\n",
      "\n",
      "91. Feng, B., Ng, J. H., Heng, J. C. & Ng, H. H. Molecules that promote or enhance reprogramming of somatic cells to\n",
      "\n",
      "work from ref. 40, have shown that appropriate closing of AP-1 motif-enriched chromatin is required for efﬁcient reprogramming of MEFs into iPSCs. The second main region of interaction involves JUNB binding to reprogramming TF motifs-enriched (i.e., MEF2C, ASCL1, KLF4, and cMYC) closed chromatin (Fig. 3). In this context, our comparative chromatin accessibility analysis (siControl vs siAJSZ) during cardiac reprogramming of HDFs (Fig. 4), shows that under control conditions, pre-existing binding of JUNB to reprogramming TF (MEF2C) motif-enriched regions (domain 2), remains in a closed state even when MGT is overexpressed, thus contributing to limit repro- gramming TF access to target its DNA\n",
      "\n",
      "T. et al. Histone variants enriched in oocytes enhance reprogramming to induced pluripotent stem cells. Cell Stem Cell 14, 217–227 (2014).\n",
      "\n",
      "75. Pawlak, M. & Jaenisch, R. De novo DNA methylation\n",
      "\n",
      "by Dnmt3a and Dnmt3b is dispensable for nuclear reprogramming of somatic cells to a pluripotent state. Genes Dev. 25, 1035–1040 (2011).\n",
      "\n",
      "76. Hu, X. et al. Tet and TDG mediate DNA demethylation essential for mesenchymal‑to‑epithelial transition in somatic cell reprogramming. Cell Stem Cell 14, 512–522 (2014).\n",
      "\n",
      "Reprogramming of pancreatic β cells into induced pluripotent stem cells. Curr. Biol. 18, 890–894 (2008).\n",
      "\n",
      "82. Samavarchi‑Tehrani, P. et al. Functional genomics reveals a BMP‑driven\n",
      "Human: How do transcription factors influence epigenetic modifications during the reprogramming process?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "Transcription factors orchestrate dynamic interplay between genome topology and gene regulation during cell reprogramming\n",
      "\n",
      "Ralph Stadhouders  François Le Dily1,2,3, Javier Quilez  Yasmina Cuartero1,2,3, Jochen Hecht2,6, Guillaume J. Filion1,2, Miguel Beato1,2, Marc A. Marti-Renom  and Thomas Graf\n",
      "\n",
      "1,2,8*, Enrique Vidal\n",
      "\n",
      "1,2, François Serra\n",
      "\n",
      "1,2,3, Bruno Di Stefano\n",
      "\n",
      "1,2,4,  5, Clara Berenguer1,2,\n",
      "\n",
      "1,2, Antonio Gomez1,2, Samuel Collombet\n",
      "\n",
      "1,2*\n",
      "\n",
      "1,2,3,7*\n",
      "\n",
      "Chromosomal architecture is known to influence gene expression, yet its role in controlling cell fate remains poorly understood. Reprogramming of somatic cells into pluripotent stem cells (PSCs) by the transcription factors (TFs) OCT4, SOX2,\n",
      "\n",
      "et al. Rapid chromatin switch in the direct repro- gramming of ﬁbroblasts to neurons. Cell Rep. 20, 3236–3247 (2017).\n",
      "\n",
      "42. Dall’Agnese, A. et al. Transcription factor-directed re-wiring of\n",
      "\n",
      "chromatin architecture for somatic cell nuclear reprogramming toward trans-differentiation. Mol. Cell 76, 453–472 (2019). 43. Stone, N. R. et al. Context-speciﬁc transcription factor functions\n",
      "\n",
      "regulate epigenomic and transcriptional dynamics during cardiac reprogramming. Cell Stem Cell 25, 87–102 (2019).\n",
      "\n",
      "44. Nam, Y. J. et al. Reprogramming of human ﬁbroblasts toward a cardiac fate. Proc. Natl Acad. Sci. USA 110, 5588–5593 (2013). 45. Zhou, Y. et al. Single-cell transcriptomic analyses of cell fate\n",
      "\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "the transcrip- tion factors that are used to elicit cellular reprogramming appear to be needed only for the initial activation of the appropriate cellular networks. After initiation, the exog- enous reprogramming factors can be extinguished, while\n",
      "\n",
      "Box 2. Important issues to be addressed\n",
      "\n",
      "What is the influence of predicted high nucleosome occupancy and nucleosome position on defining the subset of target sites to which pioneer factors will bind in a cell?\n",
      "\n",
      "What are the structural features of DNA-binding domains that allow certain factors to have pioneer activity? How do domains outside of the DNA-binding domain affect pioneer activity and subsequent chromatin-opening events? How does a\n",
      "\n",
      "system: a status report. Trends Immunol. 35, 190–194.\n",
      "\n",
      "Souﬁ, A., Garcia, M.F., Jaroszewicz, A., Osman, N., Pellegrini, M., and Zaret, K.S. (2015). Pioneer transcription factors target partial DNA motifs on nucleo- somes to initiate reprogramming. Cell 161, 555–568.\n",
      "\n",
      "Spitz, F., and Furlong, E.E. (2012). Transcription factors: from enhancer bind- ing to developmental control. Nat. Rev. Genet. 13, 613–626.\n",
      "\n",
      "Stormo, G.D., and Zhao, Y. (2010). Determining the speciﬁcity of protein-DNA interactions. Nat. Rev. Genet. 11, 751–760.\n",
      "\n",
      "Stubbs, L., Sun, Y., and Caetano-Anolles, D. (2011). Function and evolution of C2H2 zinc ﬁnger arrays. Subcell. Biochem. 52, 75–94.\n",
      "\n",
      "Sur, I., and Taipale, J. (2016). The\n",
      "\n",
      "pre- ceded gene expression changes at all levels and provided an explana- tion for the sequential activation of core pluripotency genes during reprogramming. Together, our observations indicate that genome topology has an instructive role in implementing transcriptional programs relevant to cell-fate decisions in mammals.\n",
      "\n",
      "Results Transcription factors prime the epigenome for reprogramming. We exposed mouse bone marrow–derived pre-B cells to the myeloid TF C/EBPα to generate cells denoted Bα cells. Subsequent activation of OSKM has been found to induce reprogramming of nearly 100% of Bα cells into PSC-like cells within 4–8 d (refs 31,32). To obtain a high- resolution map of changes in gene\n",
      "\n",
      "J. et al. Targeted bisulfite sequencing reveals changes in DNA methylation associated with nuclear reprogramming. Nat. Biotechnol. 27, 353–360 (2009).\n",
      "\n",
      "88. Polo, J. M. et al. A molecular roadmap of\n",
      "\n",
      "reprogramming somatic cells into iPS cells. Cell 151, 1617–1632 (2012).\n",
      "\n",
      "89. Soufi, A., Donahue, G. & Zaret, K. S. Facilitators and impediments of the pluripotency reprogramming factors’ initial engagement with the genome. Cell 151, 994–1004 (2012).\n",
      "\n",
      "90. Sridharan, R. et al. Role of the murine reprogramming factors in the induction of pluripotency. Cell 136, 364–377 (2009).\n",
      "\n",
      "91. Feng, B., Ng, J. H., Heng, J. C. & Ng, H. H. Molecules that promote or enhance reprogramming of somatic cells to\n",
      "\n",
      "work from ref. 40, have shown that appropriate closing of AP-1 motif-enriched chromatin is required for efﬁcient reprogramming of MEFs into iPSCs. The second main region of interaction involves JUNB binding to reprogramming TF motifs-enriched (i.e., MEF2C, ASCL1, KLF4, and cMYC) closed chromatin (Fig. 3). In this context, our comparative chromatin accessibility analysis (siControl vs siAJSZ) during cardiac reprogramming of HDFs (Fig. 4), shows that under control conditions, pre-existing binding of JUNB to reprogramming TF (MEF2C) motif-enriched regions (domain 2), remains in a closed state even when MGT is overexpressed, thus contributing to limit repro- gramming TF access to target its DNA\n",
      "\n",
      "T. et al. Histone variants enriched in oocytes enhance reprogramming to induced pluripotent stem cells. Cell Stem Cell 14, 217–227 (2014).\n",
      "\n",
      "75. Pawlak, M. & Jaenisch, R. De novo DNA methylation\n",
      "\n",
      "by Dnmt3a and Dnmt3b is dispensable for nuclear reprogramming of somatic cells to a pluripotent state. Genes Dev. 25, 1035–1040 (2011).\n",
      "\n",
      "76. Hu, X. et al. Tet and TDG mediate DNA demethylation essential for mesenchymal‑to‑epithelial transition in somatic cell reprogramming. Cell Stem Cell 14, 512–522 (2014).\n",
      "\n",
      "Reprogramming of pancreatic β cells into induced pluripotent stem cells. Curr. Biol. 18, 890–894 (2008).\n",
      "\n",
      "82. Samavarchi‑Tehrani, P. et al. Functional genomics reveals a BMP‑driven\n",
      "Human: How do transcription factors influence epigenetic modifications during the reprogramming process?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "Transcription factors orchestrate dynamic interplay between genome topology and gene regulation during cell reprogramming\n",
      "\n",
      "Ralph Stadhouders  François Le Dily1,2,3, Javier Quilez  Yasmina Cuartero1,2,3, Jochen Hecht2,6, Guillaume J. Filion1,2, Miguel Beato1,2, Marc A. Marti-Renom  and Thomas Graf\n",
      "\n",
      "1,2,8*, Enrique Vidal\n",
      "\n",
      "1,2, François Serra\n",
      "\n",
      "1,2,3, Bruno Di Stefano\n",
      "\n",
      "1,2,4,  5, Clara Berenguer1,2,\n",
      "\n",
      "1,2, Antonio Gomez1,2, Samuel Collombet\n",
      "\n",
      "1,2*\n",
      "\n",
      "1,2,3,7*\n",
      "\n",
      "Chromosomal architecture is known to influence gene expression, yet its role in controlling cell fate remains poorly understood. Reprogramming of somatic cells into pluripotent stem cells (PSCs) by the transcription factors (TFs) OCT4, SOX2,\n",
      "\n",
      "the path toward pluripotency.\n",
      "\n",
      "Epigenetic modifiers. Cellular reprogramming is associated with epigenetic changes, and thus epigenetic modifiers have been studied with respect to their involvement in reprogramming (reviewed in REF. 65). Enzymes that regulate the post‑ translational modifications of histones (such as their acetylation or methylation), which largely affect transcription, are one\n",
      "\n",
      "example of such epigenetic modifiers66–73. These enzymes can affect (by upregulation or downregulation) the expression of pluripotency‑associated, as well as somatic, genes and thereby drive cell fate in a particular direction (more towards either pluripotency or differentiation). For example,\n",
      "\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "the transcrip- tion factors that are used to elicit cellular reprogramming appear to be needed only for the initial activation of the appropriate cellular networks. After initiation, the exog- enous reprogramming factors can be extinguished, while\n",
      "\n",
      "Box 2. Important issues to be addressed\n",
      "\n",
      "What is the influence of predicted high nucleosome occupancy and nucleosome position on defining the subset of target sites to which pioneer factors will bind in a cell?\n",
      "\n",
      "What are the structural features of DNA-binding domains that allow certain factors to have pioneer activity? How do domains outside of the DNA-binding domain affect pioneer activity and subsequent chromatin-opening events? How does a\n",
      "\n",
      "system: a status report. Trends Immunol. 35, 190–194.\n",
      "\n",
      "Souﬁ, A., Garcia, M.F., Jaroszewicz, A., Osman, N., Pellegrini, M., and Zaret, K.S. (2015). Pioneer transcription factors target partial DNA motifs on nucleo- somes to initiate reprogramming. Cell 161, 555–568.\n",
      "\n",
      "Spitz, F., and Furlong, E.E. (2012). Transcription factors: from enhancer bind- ing to developmental control. Nat. Rev. Genet. 13, 613–626.\n",
      "\n",
      "Stormo, G.D., and Zhao, Y. (2010). Determining the speciﬁcity of protein-DNA interactions. Nat. Rev. Genet. 11, 751–760.\n",
      "\n",
      "Stubbs, L., Sun, Y., and Caetano-Anolles, D. (2011). Function and evolution of C2H2 zinc ﬁnger arrays. Subcell. Biochem. 52, 75–94.\n",
      "\n",
      "Sur, I., and Taipale, J. (2016). The\n",
      "\n",
      "pre- ceded gene expression changes at all levels and provided an explana- tion for the sequential activation of core pluripotency genes during reprogramming. Together, our observations indicate that genome topology has an instructive role in implementing transcriptional programs relevant to cell-fate decisions in mammals.\n",
      "\n",
      "Results Transcription factors prime the epigenome for reprogramming. We exposed mouse bone marrow–derived pre-B cells to the myeloid TF C/EBPα to generate cells denoted Bα cells. Subsequent activation of OSKM has been found to induce reprogramming of nearly 100% of Bα cells into PSC-like cells within 4–8 d (refs 31,32). To obtain a high- resolution map of changes in gene\n",
      "\n",
      "J. et al. Targeted bisulfite sequencing reveals changes in DNA methylation associated with nuclear reprogramming. Nat. Biotechnol. 27, 353–360 (2009).\n",
      "\n",
      "88. Polo, J. M. et al. A molecular roadmap of\n",
      "\n",
      "reprogramming somatic cells into iPS cells. Cell 151, 1617–1632 (2012).\n",
      "\n",
      "89. Soufi, A., Donahue, G. & Zaret, K. S. Facilitators and impediments of the pluripotency reprogramming factors’ initial engagement with the genome. Cell 151, 994–1004 (2012).\n",
      "\n",
      "90. Sridharan, R. et al. Role of the murine reprogramming factors in the induction of pluripotency. Cell 136, 364–377 (2009).\n",
      "\n",
      "91. Feng, B., Ng, J. H., Heng, J. C. & Ng, H. H. Molecules that promote or enhance reprogramming of somatic cells to\n",
      "\n",
      "work from ref. 40, have shown that appropriate closing of AP-1 motif-enriched chromatin is required for efﬁcient reprogramming of MEFs into iPSCs. The second main region of interaction involves JUNB binding to reprogramming TF motifs-enriched (i.e., MEF2C, ASCL1, KLF4, and cMYC) closed chromatin (Fig. 3). In this context, our comparative chromatin accessibility analysis (siControl vs siAJSZ) during cardiac reprogramming of HDFs (Fig. 4), shows that under control conditions, pre-existing binding of JUNB to reprogramming TF (MEF2C) motif-enriched regions (domain 2), remains in a closed state even when MGT is overexpressed, thus contributing to limit repro- gramming TF access to target its DNA\n",
      "\n",
      "T. et al. Histone variants enriched in oocytes enhance reprogramming to induced pluripotent stem cells. Cell Stem Cell 14, 217–227 (2014).\n",
      "\n",
      "75. Pawlak, M. & Jaenisch, R. De novo DNA methylation\n",
      "\n",
      "by Dnmt3a and Dnmt3b is dispensable for nuclear reprogramming of somatic cells to a pluripotent state. Genes Dev. 25, 1035–1040 (2011).\n",
      "\n",
      "76. Hu, X. et al. Tet and TDG mediate DNA demethylation essential for mesenchymal‑to‑epithelial transition in somatic cell reprogramming. Cell Stem Cell 14, 512–522 (2014).\n",
      "\n",
      "Reprogramming of pancreatic β cells into induced pluripotent stem cells. Curr. Biol. 18, 890–894 (2008).\n",
      "\n",
      "82. Samavarchi‑Tehrani, P. et al. Functional genomics reveals a BMP‑driven\n",
      "Human: How do transcription factors influence epigenetic modifications during the reprogramming process?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "Transcription factors orchestrate dynamic interplay between genome topology and gene regulation during cell reprogramming\n",
      "\n",
      "Ralph Stadhouders  François Le Dily1,2,3, Javier Quilez  Yasmina Cuartero1,2,3, Jochen Hecht2,6, Guillaume J. Filion1,2, Miguel Beato1,2, Marc A. Marti-Renom  and Thomas Graf\n",
      "\n",
      "1,2,8*, Enrique Vidal\n",
      "\n",
      "1,2, François Serra\n",
      "\n",
      "1,2,3, Bruno Di Stefano\n",
      "\n",
      "1,2,4,  5, Clara Berenguer1,2,\n",
      "\n",
      "1,2, Antonio Gomez1,2, Samuel Collombet\n",
      "\n",
      "1,2*\n",
      "\n",
      "1,2,3,7*\n",
      "\n",
      "Chromosomal architecture is known to influence gene expression, yet its role in controlling cell fate remains poorly understood. Reprogramming of somatic cells into pluripotent stem cells (PSCs) by the transcription factors (TFs) OCT4, SOX2,\n",
      "\n",
      "the path toward pluripotency.\n",
      "\n",
      "Epigenetic modifiers. Cellular reprogramming is associated with epigenetic changes, and thus epigenetic modifiers have been studied with respect to their involvement in reprogramming (reviewed in REF. 65). Enzymes that regulate the post‑ translational modifications of histones (such as their acetylation or methylation), which largely affect transcription, are one\n",
      "\n",
      "example of such epigenetic modifiers66–73. These enzymes can affect (by upregulation or downregulation) the expression of pluripotency‑associated, as well as somatic, genes and thereby drive cell fate in a particular direction (more towards either pluripotency or differentiation). For example,\n",
      "\n",
      "et al. Rapid chromatin switch in the direct repro- gramming of ﬁbroblasts to neurons. Cell Rep. 20, 3236–3247 (2017).\n",
      "\n",
      "42. Dall’Agnese, A. et al. Transcription factor-directed re-wiring of\n",
      "\n",
      "chromatin architecture for somatic cell nuclear reprogramming toward trans-differentiation. Mol. Cell 76, 453–472 (2019). 43. Stone, N. R. et al. Context-speciﬁc transcription factor functions\n",
      "\n",
      "regulate epigenomic and transcriptional dynamics during cardiac reprogramming. Cell Stem Cell 25, 87–102 (2019).\n",
      "\n",
      "44. Nam, Y. J. et al. Reprogramming of human ﬁbroblasts toward a cardiac fate. Proc. Natl Acad. Sci. USA 110, 5588–5593 (2013). 45. Zhou, Y. et al. Single-cell transcriptomic analyses of cell fate\n",
      "\n",
      "the transcrip- tion factors that are used to elicit cellular reprogramming appear to be needed only for the initial activation of the appropriate cellular networks. After initiation, the exog- enous reprogramming factors can be extinguished, while\n",
      "\n",
      "Box 2. Important issues to be addressed\n",
      "\n",
      "What is the influence of predicted high nucleosome occupancy and nucleosome position on defining the subset of target sites to which pioneer factors will bind in a cell?\n",
      "\n",
      "What are the structural features of DNA-binding domains that allow certain factors to have pioneer activity? How do domains outside of the DNA-binding domain affect pioneer activity and subsequent chromatin-opening events? How does a\n",
      "\n",
      "system: a status report. Trends Immunol. 35, 190–194.\n",
      "\n",
      "Souﬁ, A., Garcia, M.F., Jaroszewicz, A., Osman, N., Pellegrini, M., and Zaret, K.S. (2015). Pioneer transcription factors target partial DNA motifs on nucleo- somes to initiate reprogramming. Cell 161, 555–568.\n",
      "\n",
      "Spitz, F., and Furlong, E.E. (2012). Transcription factors: from enhancer bind- ing to developmental control. Nat. Rev. Genet. 13, 613–626.\n",
      "\n",
      "Stormo, G.D., and Zhao, Y. (2010). Determining the speciﬁcity of protein-DNA interactions. Nat. Rev. Genet. 11, 751–760.\n",
      "\n",
      "Stubbs, L., Sun, Y., and Caetano-Anolles, D. (2011). Function and evolution of C2H2 zinc ﬁnger arrays. Subcell. Biochem. 52, 75–94.\n",
      "\n",
      "Sur, I., and Taipale, J. (2016). The\n",
      "\n",
      "pre- ceded gene expression changes at all levels and provided an explana- tion for the sequential activation of core pluripotency genes during reprogramming. Together, our observations indicate that genome topology has an instructive role in implementing transcriptional programs relevant to cell-fate decisions in mammals.\n",
      "\n",
      "Results Transcription factors prime the epigenome for reprogramming. We exposed mouse bone marrow–derived pre-B cells to the myeloid TF C/EBPα to generate cells denoted Bα cells. Subsequent activation of OSKM has been found to induce reprogramming of nearly 100% of Bα cells into PSC-like cells within 4–8 d (refs 31,32). To obtain a high- resolution map of changes in gene\n",
      "\n",
      "J. et al. Targeted bisulfite sequencing reveals changes in DNA methylation associated with nuclear reprogramming. Nat. Biotechnol. 27, 353–360 (2009).\n",
      "\n",
      "88. Polo, J. M. et al. A molecular roadmap of\n",
      "\n",
      "reprogramming somatic cells into iPS cells. Cell 151, 1617–1632 (2012).\n",
      "\n",
      "89. Soufi, A., Donahue, G. & Zaret, K. S. Facilitators and impediments of the pluripotency reprogramming factors’ initial engagement with the genome. Cell 151, 994–1004 (2012).\n",
      "\n",
      "90. Sridharan, R. et al. Role of the murine reprogramming factors in the induction of pluripotency. Cell 136, 364–377 (2009).\n",
      "\n",
      "91. Feng, B., Ng, J. H., Heng, J. C. & Ng, H. H. Molecules that promote or enhance reprogramming of somatic cells to\n",
      "\n",
      "work from ref. 40, have shown that appropriate closing of AP-1 motif-enriched chromatin is required for efﬁcient reprogramming of MEFs into iPSCs. The second main region of interaction involves JUNB binding to reprogramming TF motifs-enriched (i.e., MEF2C, ASCL1, KLF4, and cMYC) closed chromatin (Fig. 3). In this context, our comparative chromatin accessibility analysis (siControl vs siAJSZ) during cardiac reprogramming of HDFs (Fig. 4), shows that under control conditions, pre-existing binding of JUNB to reprogramming TF (MEF2C) motif-enriched regions (domain 2), remains in a closed state even when MGT is overexpressed, thus contributing to limit repro- gramming TF access to target its DNA\n",
      "\n",
      "T. et al. Histone variants enriched in oocytes enhance reprogramming to induced pluripotent stem cells. Cell Stem Cell 14, 217–227 (2014).\n",
      "\n",
      "75. Pawlak, M. & Jaenisch, R. De novo DNA methylation\n",
      "\n",
      "by Dnmt3a and Dnmt3b is dispensable for nuclear reprogramming of somatic cells to a pluripotent state. Genes Dev. 25, 1035–1040 (2011).\n",
      "\n",
      "76. Hu, X. et al. Tet and TDG mediate DNA demethylation essential for mesenchymal‑to‑epithelial transition in somatic cell reprogramming. Cell Stem Cell 14, 512–522 (2014).\n",
      "\n",
      "Reprogramming of pancreatic β cells into induced pluripotent stem cells. Curr. Biol. 18, 890–894 (2008).\n",
      "\n",
      "82. Samavarchi‑Tehrani, P. et al. Functional genomics reveals a BMP‑driven\n",
      "Human: How do transcription factors influence epigenetic modifications during the reprogramming process?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "Transcription factors orchestrate dynamic interplay between genome topology and gene regulation during cell reprogramming\n",
      "\n",
      "Ralph Stadhouders  François Le Dily1,2,3, Javier Quilez  Yasmina Cuartero1,2,3, Jochen Hecht2,6, Guillaume J. Filion1,2, Miguel Beato1,2, Marc A. Marti-Renom  and Thomas Graf\n",
      "\n",
      "1,2,8*, Enrique Vidal\n",
      "\n",
      "1,2, François Serra\n",
      "\n",
      "1,2,3, Bruno Di Stefano\n",
      "\n",
      "1,2,4,  5, Clara Berenguer1,2,\n",
      "\n",
      "1,2, Antonio Gomez1,2, Samuel Collombet\n",
      "\n",
      "1,2*\n",
      "\n",
      "1,2,3,7*\n",
      "\n",
      "Chromosomal architecture is known to influence gene expression, yet its role in controlling cell fate remains poorly understood. Reprogramming of somatic cells into pluripotent stem cells (PSCs) by the transcription factors (TFs) OCT4, SOX2,\n",
      "\n",
      "the path toward pluripotency.\n",
      "\n",
      "Epigenetic modifiers. Cellular reprogramming is associated with epigenetic changes, and thus epigenetic modifiers have been studied with respect to their involvement in reprogramming (reviewed in REF. 65). Enzymes that regulate the post‑ translational modifications of histones (such as their acetylation or methylation), which largely affect transcription, are one\n",
      "\n",
      "example of such epigenetic modifiers66–73. These enzymes can affect (by upregulation or downregulation) the expression of pluripotency‑associated, as well as somatic, genes and thereby drive cell fate in a particular direction (more towards either pluripotency or differentiation). For example,\n",
      "\n",
      "et al. Rapid chromatin switch in the direct repro- gramming of ﬁbroblasts to neurons. Cell Rep. 20, 3236–3247 (2017).\n",
      "\n",
      "42. Dall’Agnese, A. et al. Transcription factor-directed re-wiring of\n",
      "\n",
      "chromatin architecture for somatic cell nuclear reprogramming toward trans-differentiation. Mol. Cell 76, 453–472 (2019). 43. Stone, N. R. et al. Context-speciﬁc transcription factor functions\n",
      "\n",
      "regulate epigenomic and transcriptional dynamics during cardiac reprogramming. Cell Stem Cell 25, 87–102 (2019).\n",
      "\n",
      "44. Nam, Y. J. et al. Reprogramming of human ﬁbroblasts toward a cardiac fate. Proc. Natl Acad. Sci. USA 110, 5588–5593 (2013). 45. Zhou, Y. et al. Single-cell transcriptomic analyses of cell fate\n",
      "\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "system: a status report. Trends Immunol. 35, 190–194.\n",
      "\n",
      "Souﬁ, A., Garcia, M.F., Jaroszewicz, A., Osman, N., Pellegrini, M., and Zaret, K.S. (2015). Pioneer transcription factors target partial DNA motifs on nucleo- somes to initiate reprogramming. Cell 161, 555–568.\n",
      "\n",
      "Spitz, F., and Furlong, E.E. (2012). Transcription factors: from enhancer bind- ing to developmental control. Nat. Rev. Genet. 13, 613–626.\n",
      "\n",
      "Stormo, G.D., and Zhao, Y. (2010). Determining the speciﬁcity of protein-DNA interactions. Nat. Rev. Genet. 11, 751–760.\n",
      "\n",
      "Stubbs, L., Sun, Y., and Caetano-Anolles, D. (2011). Function and evolution of C2H2 zinc ﬁnger arrays. Subcell. Biochem. 52, 75–94.\n",
      "\n",
      "Sur, I., and Taipale, J. (2016). The\n",
      "\n",
      "pre- ceded gene expression changes at all levels and provided an explana- tion for the sequential activation of core pluripotency genes during reprogramming. Together, our observations indicate that genome topology has an instructive role in implementing transcriptional programs relevant to cell-fate decisions in mammals.\n",
      "\n",
      "Results Transcription factors prime the epigenome for reprogramming. We exposed mouse bone marrow–derived pre-B cells to the myeloid TF C/EBPα to generate cells denoted Bα cells. Subsequent activation of OSKM has been found to induce reprogramming of nearly 100% of Bα cells into PSC-like cells within 4–8 d (refs 31,32). To obtain a high- resolution map of changes in gene\n",
      "\n",
      "J. et al. Targeted bisulfite sequencing reveals changes in DNA methylation associated with nuclear reprogramming. Nat. Biotechnol. 27, 353–360 (2009).\n",
      "\n",
      "88. Polo, J. M. et al. A molecular roadmap of\n",
      "\n",
      "reprogramming somatic cells into iPS cells. Cell 151, 1617–1632 (2012).\n",
      "\n",
      "89. Soufi, A., Donahue, G. & Zaret, K. S. Facilitators and impediments of the pluripotency reprogramming factors’ initial engagement with the genome. Cell 151, 994–1004 (2012).\n",
      "\n",
      "90. Sridharan, R. et al. Role of the murine reprogramming factors in the induction of pluripotency. Cell 136, 364–377 (2009).\n",
      "\n",
      "91. Feng, B., Ng, J. H., Heng, J. C. & Ng, H. H. Molecules that promote or enhance reprogramming of somatic cells to\n",
      "\n",
      "work from ref. 40, have shown that appropriate closing of AP-1 motif-enriched chromatin is required for efﬁcient reprogramming of MEFs into iPSCs. The second main region of interaction involves JUNB binding to reprogramming TF motifs-enriched (i.e., MEF2C, ASCL1, KLF4, and cMYC) closed chromatin (Fig. 3). In this context, our comparative chromatin accessibility analysis (siControl vs siAJSZ) during cardiac reprogramming of HDFs (Fig. 4), shows that under control conditions, pre-existing binding of JUNB to reprogramming TF (MEF2C) motif-enriched regions (domain 2), remains in a closed state even when MGT is overexpressed, thus contributing to limit repro- gramming TF access to target its DNA\n",
      "\n",
      "T. et al. Histone variants enriched in oocytes enhance reprogramming to induced pluripotent stem cells. Cell Stem Cell 14, 217–227 (2014).\n",
      "\n",
      "75. Pawlak, M. & Jaenisch, R. De novo DNA methylation\n",
      "\n",
      "by Dnmt3a and Dnmt3b is dispensable for nuclear reprogramming of somatic cells to a pluripotent state. Genes Dev. 25, 1035–1040 (2011).\n",
      "\n",
      "76. Hu, X. et al. Tet and TDG mediate DNA demethylation essential for mesenchymal‑to‑epithelial transition in somatic cell reprogramming. Cell Stem Cell 14, 512–522 (2014).\n",
      "\n",
      "Reprogramming of pancreatic β cells into induced pluripotent stem cells. Curr. Biol. 18, 890–894 (2008).\n",
      "\n",
      "82. Samavarchi‑Tehrani, P. et al. Functional genomics reveals a BMP‑driven\n",
      "Human: How do transcription factors influence epigenetic modifications during the reprogramming process?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "Transcription factors orchestrate dynamic interplay between genome topology and gene regulation during cell reprogramming\n",
      "\n",
      "Ralph Stadhouders  François Le Dily1,2,3, Javier Quilez  Yasmina Cuartero1,2,3, Jochen Hecht2,6, Guillaume J. Filion1,2, Miguel Beato1,2, Marc A. Marti-Renom  and Thomas Graf\n",
      "\n",
      "1,2,8*, Enrique Vidal\n",
      "\n",
      "1,2, François Serra\n",
      "\n",
      "1,2,3, Bruno Di Stefano\n",
      "\n",
      "1,2,4,  5, Clara Berenguer1,2,\n",
      "\n",
      "1,2, Antonio Gomez1,2, Samuel Collombet\n",
      "\n",
      "1,2*\n",
      "\n",
      "1,2,3,7*\n",
      "\n",
      "Chromosomal architecture is known to influence gene expression, yet its role in controlling cell fate remains poorly understood. Reprogramming of somatic cells into pluripotent stem cells (PSCs) by the transcription factors (TFs) OCT4, SOX2,\n",
      "\n",
      "the path toward pluripotency.\n",
      "\n",
      "Epigenetic modifiers. Cellular reprogramming is associated with epigenetic changes, and thus epigenetic modifiers have been studied with respect to their involvement in reprogramming (reviewed in REF. 65). Enzymes that regulate the post‑ translational modifications of histones (such as their acetylation or methylation), which largely affect transcription, are one\n",
      "\n",
      "example of such epigenetic modifiers66–73. These enzymes can affect (by upregulation or downregulation) the expression of pluripotency‑associated, as well as somatic, genes and thereby drive cell fate in a particular direction (more towards either pluripotency or differentiation). For example,\n",
      "\n",
      "et al. Rapid chromatin switch in the direct repro- gramming of ﬁbroblasts to neurons. Cell Rep. 20, 3236–3247 (2017).\n",
      "\n",
      "42. Dall’Agnese, A. et al. Transcription factor-directed re-wiring of\n",
      "\n",
      "chromatin architecture for somatic cell nuclear reprogramming toward trans-differentiation. Mol. Cell 76, 453–472 (2019). 43. Stone, N. R. et al. Context-speciﬁc transcription factor functions\n",
      "\n",
      "regulate epigenomic and transcriptional dynamics during cardiac reprogramming. Cell Stem Cell 25, 87–102 (2019).\n",
      "\n",
      "44. Nam, Y. J. et al. Reprogramming of human ﬁbroblasts toward a cardiac fate. Proc. Natl Acad. Sci. USA 110, 5588–5593 (2013). 45. Zhou, Y. et al. Single-cell transcriptomic analyses of cell fate\n",
      "\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "the transcrip- tion factors that are used to elicit cellular reprogramming appear to be needed only for the initial activation of the appropriate cellular networks. After initiation, the exog- enous reprogramming factors can be extinguished, while\n",
      "\n",
      "Box 2. Important issues to be addressed\n",
      "\n",
      "What is the influence of predicted high nucleosome occupancy and nucleosome position on defining the subset of target sites to which pioneer factors will bind in a cell?\n",
      "\n",
      "What are the structural features of DNA-binding domains that allow certain factors to have pioneer activity? How do domains outside of the DNA-binding domain affect pioneer activity and subsequent chromatin-opening events? How does a\n",
      "\n",
      "pre- ceded gene expression changes at all levels and provided an explana- tion for the sequential activation of core pluripotency genes during reprogramming. Together, our observations indicate that genome topology has an instructive role in implementing transcriptional programs relevant to cell-fate decisions in mammals.\n",
      "\n",
      "Results Transcription factors prime the epigenome for reprogramming. We exposed mouse bone marrow–derived pre-B cells to the myeloid TF C/EBPα to generate cells denoted Bα cells. Subsequent activation of OSKM has been found to induce reprogramming of nearly 100% of Bα cells into PSC-like cells within 4–8 d (refs 31,32). To obtain a high- resolution map of changes in gene\n",
      "\n",
      "J. et al. Targeted bisulfite sequencing reveals changes in DNA methylation associated with nuclear reprogramming. Nat. Biotechnol. 27, 353–360 (2009).\n",
      "\n",
      "88. Polo, J. M. et al. A molecular roadmap of\n",
      "\n",
      "reprogramming somatic cells into iPS cells. Cell 151, 1617–1632 (2012).\n",
      "\n",
      "89. Soufi, A., Donahue, G. & Zaret, K. S. Facilitators and impediments of the pluripotency reprogramming factors’ initial engagement with the genome. Cell 151, 994–1004 (2012).\n",
      "\n",
      "90. Sridharan, R. et al. Role of the murine reprogramming factors in the induction of pluripotency. Cell 136, 364–377 (2009).\n",
      "\n",
      "91. Feng, B., Ng, J. H., Heng, J. C. & Ng, H. H. Molecules that promote or enhance reprogramming of somatic cells to\n",
      "\n",
      "work from ref. 40, have shown that appropriate closing of AP-1 motif-enriched chromatin is required for efﬁcient reprogramming of MEFs into iPSCs. The second main region of interaction involves JUNB binding to reprogramming TF motifs-enriched (i.e., MEF2C, ASCL1, KLF4, and cMYC) closed chromatin (Fig. 3). In this context, our comparative chromatin accessibility analysis (siControl vs siAJSZ) during cardiac reprogramming of HDFs (Fig. 4), shows that under control conditions, pre-existing binding of JUNB to reprogramming TF (MEF2C) motif-enriched regions (domain 2), remains in a closed state even when MGT is overexpressed, thus contributing to limit repro- gramming TF access to target its DNA\n",
      "\n",
      "T. et al. Histone variants enriched in oocytes enhance reprogramming to induced pluripotent stem cells. Cell Stem Cell 14, 217–227 (2014).\n",
      "\n",
      "75. Pawlak, M. & Jaenisch, R. De novo DNA methylation\n",
      "\n",
      "by Dnmt3a and Dnmt3b is dispensable for nuclear reprogramming of somatic cells to a pluripotent state. Genes Dev. 25, 1035–1040 (2011).\n",
      "\n",
      "76. Hu, X. et al. Tet and TDG mediate DNA demethylation essential for mesenchymal‑to‑epithelial transition in somatic cell reprogramming. Cell Stem Cell 14, 512–522 (2014).\n",
      "\n",
      "Reprogramming of pancreatic β cells into induced pluripotent stem cells. Curr. Biol. 18, 890–894 (2008).\n",
      "\n",
      "82. Samavarchi‑Tehrani, P. et al. Functional genomics reveals a BMP‑driven\n",
      "Human: How do transcription factors influence epigenetic modifications during the reprogramming process?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "Transcription factors orchestrate dynamic interplay between genome topology and gene regulation during cell reprogramming\n",
      "\n",
      "Ralph Stadhouders  François Le Dily1,2,3, Javier Quilez  Yasmina Cuartero1,2,3, Jochen Hecht2,6, Guillaume J. Filion1,2, Miguel Beato1,2, Marc A. Marti-Renom  and Thomas Graf\n",
      "\n",
      "1,2,8*, Enrique Vidal\n",
      "\n",
      "1,2, François Serra\n",
      "\n",
      "1,2,3, Bruno Di Stefano\n",
      "\n",
      "1,2,4,  5, Clara Berenguer1,2,\n",
      "\n",
      "1,2, Antonio Gomez1,2, Samuel Collombet\n",
      "\n",
      "1,2*\n",
      "\n",
      "1,2,3,7*\n",
      "\n",
      "Chromosomal architecture is known to influence gene expression, yet its role in controlling cell fate remains poorly understood. Reprogramming of somatic cells into pluripotent stem cells (PSCs) by the transcription factors (TFs) OCT4, SOX2,\n",
      "\n",
      "the path toward pluripotency.\n",
      "\n",
      "Epigenetic modifiers. Cellular reprogramming is associated with epigenetic changes, and thus epigenetic modifiers have been studied with respect to their involvement in reprogramming (reviewed in REF. 65). Enzymes that regulate the post‑ translational modifications of histones (such as their acetylation or methylation), which largely affect transcription, are one\n",
      "\n",
      "example of such epigenetic modifiers66–73. These enzymes can affect (by upregulation or downregulation) the expression of pluripotency‑associated, as well as somatic, genes and thereby drive cell fate in a particular direction (more towards either pluripotency or differentiation). For example,\n",
      "\n",
      "et al. Rapid chromatin switch in the direct repro- gramming of ﬁbroblasts to neurons. Cell Rep. 20, 3236–3247 (2017).\n",
      "\n",
      "42. Dall’Agnese, A. et al. Transcription factor-directed re-wiring of\n",
      "\n",
      "chromatin architecture for somatic cell nuclear reprogramming toward trans-differentiation. Mol. Cell 76, 453–472 (2019). 43. Stone, N. R. et al. Context-speciﬁc transcription factor functions\n",
      "\n",
      "regulate epigenomic and transcriptional dynamics during cardiac reprogramming. Cell Stem Cell 25, 87–102 (2019).\n",
      "\n",
      "44. Nam, Y. J. et al. Reprogramming of human ﬁbroblasts toward a cardiac fate. Proc. Natl Acad. Sci. USA 110, 5588–5593 (2013). 45. Zhou, Y. et al. Single-cell transcriptomic analyses of cell fate\n",
      "\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "the transcrip- tion factors that are used to elicit cellular reprogramming appear to be needed only for the initial activation of the appropriate cellular networks. After initiation, the exog- enous reprogramming factors can be extinguished, while\n",
      "\n",
      "Box 2. Important issues to be addressed\n",
      "\n",
      "What is the influence of predicted high nucleosome occupancy and nucleosome position on defining the subset of target sites to which pioneer factors will bind in a cell?\n",
      "\n",
      "What are the structural features of DNA-binding domains that allow certain factors to have pioneer activity? How do domains outside of the DNA-binding domain affect pioneer activity and subsequent chromatin-opening events? How does a\n",
      "\n",
      "system: a status report. Trends Immunol. 35, 190–194.\n",
      "\n",
      "Souﬁ, A., Garcia, M.F., Jaroszewicz, A., Osman, N., Pellegrini, M., and Zaret, K.S. (2015). Pioneer transcription factors target partial DNA motifs on nucleo- somes to initiate reprogramming. Cell 161, 555–568.\n",
      "\n",
      "Spitz, F., and Furlong, E.E. (2012). Transcription factors: from enhancer bind- ing to developmental control. Nat. Rev. Genet. 13, 613–626.\n",
      "\n",
      "Stormo, G.D., and Zhao, Y. (2010). Determining the speciﬁcity of protein-DNA interactions. Nat. Rev. Genet. 11, 751–760.\n",
      "\n",
      "Stubbs, L., Sun, Y., and Caetano-Anolles, D. (2011). Function and evolution of C2H2 zinc ﬁnger arrays. Subcell. Biochem. 52, 75–94.\n",
      "\n",
      "Sur, I., and Taipale, J. (2016). The\n",
      "\n",
      "J. et al. Targeted bisulfite sequencing reveals changes in DNA methylation associated with nuclear reprogramming. Nat. Biotechnol. 27, 353–360 (2009).\n",
      "\n",
      "88. Polo, J. M. et al. A molecular roadmap of\n",
      "\n",
      "reprogramming somatic cells into iPS cells. Cell 151, 1617–1632 (2012).\n",
      "\n",
      "89. Soufi, A., Donahue, G. & Zaret, K. S. Facilitators and impediments of the pluripotency reprogramming factors’ initial engagement with the genome. Cell 151, 994–1004 (2012).\n",
      "\n",
      "90. Sridharan, R. et al. Role of the murine reprogramming factors in the induction of pluripotency. Cell 136, 364–377 (2009).\n",
      "\n",
      "91. Feng, B., Ng, J. H., Heng, J. C. & Ng, H. H. Molecules that promote or enhance reprogramming of somatic cells to\n",
      "\n",
      "work from ref. 40, have shown that appropriate closing of AP-1 motif-enriched chromatin is required for efﬁcient reprogramming of MEFs into iPSCs. The second main region of interaction involves JUNB binding to reprogramming TF motifs-enriched (i.e., MEF2C, ASCL1, KLF4, and cMYC) closed chromatin (Fig. 3). In this context, our comparative chromatin accessibility analysis (siControl vs siAJSZ) during cardiac reprogramming of HDFs (Fig. 4), shows that under control conditions, pre-existing binding of JUNB to reprogramming TF (MEF2C) motif-enriched regions (domain 2), remains in a closed state even when MGT is overexpressed, thus contributing to limit repro- gramming TF access to target its DNA\n",
      "\n",
      "T. et al. Histone variants enriched in oocytes enhance reprogramming to induced pluripotent stem cells. Cell Stem Cell 14, 217–227 (2014).\n",
      "\n",
      "75. Pawlak, M. & Jaenisch, R. De novo DNA methylation\n",
      "\n",
      "by Dnmt3a and Dnmt3b is dispensable for nuclear reprogramming of somatic cells to a pluripotent state. Genes Dev. 25, 1035–1040 (2011).\n",
      "\n",
      "76. Hu, X. et al. Tet and TDG mediate DNA demethylation essential for mesenchymal‑to‑epithelial transition in somatic cell reprogramming. Cell Stem Cell 14, 512–522 (2014).\n",
      "\n",
      "Reprogramming of pancreatic β cells into induced pluripotent stem cells. Curr. Biol. 18, 890–894 (2008).\n",
      "\n",
      "82. Samavarchi‑Tehrani, P. et al. Functional genomics reveals a BMP‑driven\n",
      "Human: How do transcription factors influence epigenetic modifications during the reprogramming process?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "Transcription factors orchestrate dynamic interplay between genome topology and gene regulation during cell reprogramming\n",
      "\n",
      "Ralph Stadhouders  François Le Dily1,2,3, Javier Quilez  Yasmina Cuartero1,2,3, Jochen Hecht2,6, Guillaume J. Filion1,2, Miguel Beato1,2, Marc A. Marti-Renom  and Thomas Graf\n",
      "\n",
      "1,2,8*, Enrique Vidal\n",
      "\n",
      "1,2, François Serra\n",
      "\n",
      "1,2,3, Bruno Di Stefano\n",
      "\n",
      "1,2,4,  5, Clara Berenguer1,2,\n",
      "\n",
      "1,2, Antonio Gomez1,2, Samuel Collombet\n",
      "\n",
      "1,2*\n",
      "\n",
      "1,2,3,7*\n",
      "\n",
      "Chromosomal architecture is known to influence gene expression, yet its role in controlling cell fate remains poorly understood. Reprogramming of somatic cells into pluripotent stem cells (PSCs) by the transcription factors (TFs) OCT4, SOX2,\n",
      "\n",
      "the path toward pluripotency.\n",
      "\n",
      "Epigenetic modifiers. Cellular reprogramming is associated with epigenetic changes, and thus epigenetic modifiers have been studied with respect to their involvement in reprogramming (reviewed in REF. 65). Enzymes that regulate the post‑ translational modifications of histones (such as their acetylation or methylation), which largely affect transcription, are one\n",
      "\n",
      "example of such epigenetic modifiers66–73. These enzymes can affect (by upregulation or downregulation) the expression of pluripotency‑associated, as well as somatic, genes and thereby drive cell fate in a particular direction (more towards either pluripotency or differentiation). For example,\n",
      "\n",
      "et al. Rapid chromatin switch in the direct repro- gramming of ﬁbroblasts to neurons. Cell Rep. 20, 3236–3247 (2017).\n",
      "\n",
      "42. Dall’Agnese, A. et al. Transcription factor-directed re-wiring of\n",
      "\n",
      "chromatin architecture for somatic cell nuclear reprogramming toward trans-differentiation. Mol. Cell 76, 453–472 (2019). 43. Stone, N. R. et al. Context-speciﬁc transcription factor functions\n",
      "\n",
      "regulate epigenomic and transcriptional dynamics during cardiac reprogramming. Cell Stem Cell 25, 87–102 (2019).\n",
      "\n",
      "44. Nam, Y. J. et al. Reprogramming of human ﬁbroblasts toward a cardiac fate. Proc. Natl Acad. Sci. USA 110, 5588–5593 (2013). 45. Zhou, Y. et al. Single-cell transcriptomic analyses of cell fate\n",
      "\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "the transcrip- tion factors that are used to elicit cellular reprogramming appear to be needed only for the initial activation of the appropriate cellular networks. After initiation, the exog- enous reprogramming factors can be extinguished, while\n",
      "\n",
      "Box 2. Important issues to be addressed\n",
      "\n",
      "What is the influence of predicted high nucleosome occupancy and nucleosome position on defining the subset of target sites to which pioneer factors will bind in a cell?\n",
      "\n",
      "What are the structural features of DNA-binding domains that allow certain factors to have pioneer activity? How do domains outside of the DNA-binding domain affect pioneer activity and subsequent chromatin-opening events? How does a\n",
      "\n",
      "system: a status report. Trends Immunol. 35, 190–194.\n",
      "\n",
      "Souﬁ, A., Garcia, M.F., Jaroszewicz, A., Osman, N., Pellegrini, M., and Zaret, K.S. (2015). Pioneer transcription factors target partial DNA motifs on nucleo- somes to initiate reprogramming. Cell 161, 555–568.\n",
      "\n",
      "Spitz, F., and Furlong, E.E. (2012). Transcription factors: from enhancer bind- ing to developmental control. Nat. Rev. Genet. 13, 613–626.\n",
      "\n",
      "Stormo, G.D., and Zhao, Y. (2010). Determining the speciﬁcity of protein-DNA interactions. Nat. Rev. Genet. 11, 751–760.\n",
      "\n",
      "Stubbs, L., Sun, Y., and Caetano-Anolles, D. (2011). Function and evolution of C2H2 zinc ﬁnger arrays. Subcell. Biochem. 52, 75–94.\n",
      "\n",
      "Sur, I., and Taipale, J. (2016). The\n",
      "\n",
      "pre- ceded gene expression changes at all levels and provided an explana- tion for the sequential activation of core pluripotency genes during reprogramming. Together, our observations indicate that genome topology has an instructive role in implementing transcriptional programs relevant to cell-fate decisions in mammals.\n",
      "\n",
      "Results Transcription factors prime the epigenome for reprogramming. We exposed mouse bone marrow–derived pre-B cells to the myeloid TF C/EBPα to generate cells denoted Bα cells. Subsequent activation of OSKM has been found to induce reprogramming of nearly 100% of Bα cells into PSC-like cells within 4–8 d (refs 31,32). To obtain a high- resolution map of changes in gene\n",
      "\n",
      "work from ref. 40, have shown that appropriate closing of AP-1 motif-enriched chromatin is required for efﬁcient reprogramming of MEFs into iPSCs. The second main region of interaction involves JUNB binding to reprogramming TF motifs-enriched (i.e., MEF2C, ASCL1, KLF4, and cMYC) closed chromatin (Fig. 3). In this context, our comparative chromatin accessibility analysis (siControl vs siAJSZ) during cardiac reprogramming of HDFs (Fig. 4), shows that under control conditions, pre-existing binding of JUNB to reprogramming TF (MEF2C) motif-enriched regions (domain 2), remains in a closed state even when MGT is overexpressed, thus contributing to limit repro- gramming TF access to target its DNA\n",
      "\n",
      "T. et al. Histone variants enriched in oocytes enhance reprogramming to induced pluripotent stem cells. Cell Stem Cell 14, 217–227 (2014).\n",
      "\n",
      "75. Pawlak, M. & Jaenisch, R. De novo DNA methylation\n",
      "\n",
      "by Dnmt3a and Dnmt3b is dispensable for nuclear reprogramming of somatic cells to a pluripotent state. Genes Dev. 25, 1035–1040 (2011).\n",
      "\n",
      "76. Hu, X. et al. Tet and TDG mediate DNA demethylation essential for mesenchymal‑to‑epithelial transition in somatic cell reprogramming. Cell Stem Cell 14, 512–522 (2014).\n",
      "\n",
      "Reprogramming of pancreatic β cells into induced pluripotent stem cells. Curr. Biol. 18, 890–894 (2008).\n",
      "\n",
      "82. Samavarchi‑Tehrani, P. et al. Functional genomics reveals a BMP‑driven\n",
      "Human: How do transcription factors influence epigenetic modifications during the reprogramming process?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "Transcription factors orchestrate dynamic interplay between genome topology and gene regulation during cell reprogramming\n",
      "\n",
      "Ralph Stadhouders  François Le Dily1,2,3, Javier Quilez  Yasmina Cuartero1,2,3, Jochen Hecht2,6, Guillaume J. Filion1,2, Miguel Beato1,2, Marc A. Marti-Renom  and Thomas Graf\n",
      "\n",
      "1,2,8*, Enrique Vidal\n",
      "\n",
      "1,2, François Serra\n",
      "\n",
      "1,2,3, Bruno Di Stefano\n",
      "\n",
      "1,2,4,  5, Clara Berenguer1,2,\n",
      "\n",
      "1,2, Antonio Gomez1,2, Samuel Collombet\n",
      "\n",
      "1,2*\n",
      "\n",
      "1,2,3,7*\n",
      "\n",
      "Chromosomal architecture is known to influence gene expression, yet its role in controlling cell fate remains poorly understood. Reprogramming of somatic cells into pluripotent stem cells (PSCs) by the transcription factors (TFs) OCT4, SOX2,\n",
      "\n",
      "the path toward pluripotency.\n",
      "\n",
      "Epigenetic modifiers. Cellular reprogramming is associated with epigenetic changes, and thus epigenetic modifiers have been studied with respect to their involvement in reprogramming (reviewed in REF. 65). Enzymes that regulate the post‑ translational modifications of histones (such as their acetylation or methylation), which largely affect transcription, are one\n",
      "\n",
      "example of such epigenetic modifiers66–73. These enzymes can affect (by upregulation or downregulation) the expression of pluripotency‑associated, as well as somatic, genes and thereby drive cell fate in a particular direction (more towards either pluripotency or differentiation). For example,\n",
      "\n",
      "et al. Rapid chromatin switch in the direct repro- gramming of ﬁbroblasts to neurons. Cell Rep. 20, 3236–3247 (2017).\n",
      "\n",
      "42. Dall’Agnese, A. et al. Transcription factor-directed re-wiring of\n",
      "\n",
      "chromatin architecture for somatic cell nuclear reprogramming toward trans-differentiation. Mol. Cell 76, 453–472 (2019). 43. Stone, N. R. et al. Context-speciﬁc transcription factor functions\n",
      "\n",
      "regulate epigenomic and transcriptional dynamics during cardiac reprogramming. Cell Stem Cell 25, 87–102 (2019).\n",
      "\n",
      "44. Nam, Y. J. et al. Reprogramming of human ﬁbroblasts toward a cardiac fate. Proc. Natl Acad. Sci. USA 110, 5588–5593 (2013). 45. Zhou, Y. et al. Single-cell transcriptomic analyses of cell fate\n",
      "\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "the transcrip- tion factors that are used to elicit cellular reprogramming appear to be needed only for the initial activation of the appropriate cellular networks. After initiation, the exog- enous reprogramming factors can be extinguished, while\n",
      "\n",
      "Box 2. Important issues to be addressed\n",
      "\n",
      "What is the influence of predicted high nucleosome occupancy and nucleosome position on defining the subset of target sites to which pioneer factors will bind in a cell?\n",
      "\n",
      "What are the structural features of DNA-binding domains that allow certain factors to have pioneer activity? How do domains outside of the DNA-binding domain affect pioneer activity and subsequent chromatin-opening events? How does a\n",
      "\n",
      "system: a status report. Trends Immunol. 35, 190–194.\n",
      "\n",
      "Souﬁ, A., Garcia, M.F., Jaroszewicz, A., Osman, N., Pellegrini, M., and Zaret, K.S. (2015). Pioneer transcription factors target partial DNA motifs on nucleo- somes to initiate reprogramming. Cell 161, 555–568.\n",
      "\n",
      "Spitz, F., and Furlong, E.E. (2012). Transcription factors: from enhancer bind- ing to developmental control. Nat. Rev. Genet. 13, 613–626.\n",
      "\n",
      "Stormo, G.D., and Zhao, Y. (2010). Determining the speciﬁcity of protein-DNA interactions. Nat. Rev. Genet. 11, 751–760.\n",
      "\n",
      "Stubbs, L., Sun, Y., and Caetano-Anolles, D. (2011). Function and evolution of C2H2 zinc ﬁnger arrays. Subcell. Biochem. 52, 75–94.\n",
      "\n",
      "Sur, I., and Taipale, J. (2016). The\n",
      "\n",
      "pre- ceded gene expression changes at all levels and provided an explana- tion for the sequential activation of core pluripotency genes during reprogramming. Together, our observations indicate that genome topology has an instructive role in implementing transcriptional programs relevant to cell-fate decisions in mammals.\n",
      "\n",
      "Results Transcription factors prime the epigenome for reprogramming. We exposed mouse bone marrow–derived pre-B cells to the myeloid TF C/EBPα to generate cells denoted Bα cells. Subsequent activation of OSKM has been found to induce reprogramming of nearly 100% of Bα cells into PSC-like cells within 4–8 d (refs 31,32). To obtain a high- resolution map of changes in gene\n",
      "\n",
      "J. et al. Targeted bisulfite sequencing reveals changes in DNA methylation associated with nuclear reprogramming. Nat. Biotechnol. 27, 353–360 (2009).\n",
      "\n",
      "88. Polo, J. M. et al. A molecular roadmap of\n",
      "\n",
      "reprogramming somatic cells into iPS cells. Cell 151, 1617–1632 (2012).\n",
      "\n",
      "89. Soufi, A., Donahue, G. & Zaret, K. S. Facilitators and impediments of the pluripotency reprogramming factors’ initial engagement with the genome. Cell 151, 994–1004 (2012).\n",
      "\n",
      "90. Sridharan, R. et al. Role of the murine reprogramming factors in the induction of pluripotency. Cell 136, 364–377 (2009).\n",
      "\n",
      "91. Feng, B., Ng, J. H., Heng, J. C. & Ng, H. H. Molecules that promote or enhance reprogramming of somatic cells to\n",
      "\n",
      "T. et al. Histone variants enriched in oocytes enhance reprogramming to induced pluripotent stem cells. Cell Stem Cell 14, 217–227 (2014).\n",
      "\n",
      "75. Pawlak, M. & Jaenisch, R. De novo DNA methylation\n",
      "\n",
      "by Dnmt3a and Dnmt3b is dispensable for nuclear reprogramming of somatic cells to a pluripotent state. Genes Dev. 25, 1035–1040 (2011).\n",
      "\n",
      "76. Hu, X. et al. Tet and TDG mediate DNA demethylation essential for mesenchymal‑to‑epithelial transition in somatic cell reprogramming. Cell Stem Cell 14, 512–522 (2014).\n",
      "\n",
      "Reprogramming of pancreatic β cells into induced pluripotent stem cells. Curr. Biol. 18, 890–894 (2008).\n",
      "\n",
      "82. Samavarchi‑Tehrani, P. et al. Functional genomics reveals a BMP‑driven\n",
      "Human: How do transcription factors influence epigenetic modifications during the reprogramming process?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "Transcription factors orchestrate dynamic interplay between genome topology and gene regulation during cell reprogramming\n",
      "\n",
      "Ralph Stadhouders  François Le Dily1,2,3, Javier Quilez  Yasmina Cuartero1,2,3, Jochen Hecht2,6, Guillaume J. Filion1,2, Miguel Beato1,2, Marc A. Marti-Renom  and Thomas Graf\n",
      "\n",
      "1,2,8*, Enrique Vidal\n",
      "\n",
      "1,2, François Serra\n",
      "\n",
      "1,2,3, Bruno Di Stefano\n",
      "\n",
      "1,2,4,  5, Clara Berenguer1,2,\n",
      "\n",
      "1,2, Antonio Gomez1,2, Samuel Collombet\n",
      "\n",
      "1,2*\n",
      "\n",
      "1,2,3,7*\n",
      "\n",
      "Chromosomal architecture is known to influence gene expression, yet its role in controlling cell fate remains poorly understood. Reprogramming of somatic cells into pluripotent stem cells (PSCs) by the transcription factors (TFs) OCT4, SOX2,\n",
      "\n",
      "the path toward pluripotency.\n",
      "\n",
      "Epigenetic modifiers. Cellular reprogramming is associated with epigenetic changes, and thus epigenetic modifiers have been studied with respect to their involvement in reprogramming (reviewed in REF. 65). Enzymes that regulate the post‑ translational modifications of histones (such as their acetylation or methylation), which largely affect transcription, are one\n",
      "\n",
      "example of such epigenetic modifiers66–73. These enzymes can affect (by upregulation or downregulation) the expression of pluripotency‑associated, as well as somatic, genes and thereby drive cell fate in a particular direction (more towards either pluripotency or differentiation). For example,\n",
      "\n",
      "et al. Rapid chromatin switch in the direct repro- gramming of ﬁbroblasts to neurons. Cell Rep. 20, 3236–3247 (2017).\n",
      "\n",
      "42. Dall’Agnese, A. et al. Transcription factor-directed re-wiring of\n",
      "\n",
      "chromatin architecture for somatic cell nuclear reprogramming toward trans-differentiation. Mol. Cell 76, 453–472 (2019). 43. Stone, N. R. et al. Context-speciﬁc transcription factor functions\n",
      "\n",
      "regulate epigenomic and transcriptional dynamics during cardiac reprogramming. Cell Stem Cell 25, 87–102 (2019).\n",
      "\n",
      "44. Nam, Y. J. et al. Reprogramming of human ﬁbroblasts toward a cardiac fate. Proc. Natl Acad. Sci. USA 110, 5588–5593 (2013). 45. Zhou, Y. et al. Single-cell transcriptomic analyses of cell fate\n",
      "\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "the transcrip- tion factors that are used to elicit cellular reprogramming appear to be needed only for the initial activation of the appropriate cellular networks. After initiation, the exog- enous reprogramming factors can be extinguished, while\n",
      "\n",
      "Box 2. Important issues to be addressed\n",
      "\n",
      "What is the influence of predicted high nucleosome occupancy and nucleosome position on defining the subset of target sites to which pioneer factors will bind in a cell?\n",
      "\n",
      "What are the structural features of DNA-binding domains that allow certain factors to have pioneer activity? How do domains outside of the DNA-binding domain affect pioneer activity and subsequent chromatin-opening events? How does a\n",
      "\n",
      "system: a status report. Trends Immunol. 35, 190–194.\n",
      "\n",
      "Souﬁ, A., Garcia, M.F., Jaroszewicz, A., Osman, N., Pellegrini, M., and Zaret, K.S. (2015). Pioneer transcription factors target partial DNA motifs on nucleo- somes to initiate reprogramming. Cell 161, 555–568.\n",
      "\n",
      "Spitz, F., and Furlong, E.E. (2012). Transcription factors: from enhancer bind- ing to developmental control. Nat. Rev. Genet. 13, 613–626.\n",
      "\n",
      "Stormo, G.D., and Zhao, Y. (2010). Determining the speciﬁcity of protein-DNA interactions. Nat. Rev. Genet. 11, 751–760.\n",
      "\n",
      "Stubbs, L., Sun, Y., and Caetano-Anolles, D. (2011). Function and evolution of C2H2 zinc ﬁnger arrays. Subcell. Biochem. 52, 75–94.\n",
      "\n",
      "Sur, I., and Taipale, J. (2016). The\n",
      "\n",
      "pre- ceded gene expression changes at all levels and provided an explana- tion for the sequential activation of core pluripotency genes during reprogramming. Together, our observations indicate that genome topology has an instructive role in implementing transcriptional programs relevant to cell-fate decisions in mammals.\n",
      "\n",
      "Results Transcription factors prime the epigenome for reprogramming. We exposed mouse bone marrow–derived pre-B cells to the myeloid TF C/EBPα to generate cells denoted Bα cells. Subsequent activation of OSKM has been found to induce reprogramming of nearly 100% of Bα cells into PSC-like cells within 4–8 d (refs 31,32). To obtain a high- resolution map of changes in gene\n",
      "\n",
      "J. et al. Targeted bisulfite sequencing reveals changes in DNA methylation associated with nuclear reprogramming. Nat. Biotechnol. 27, 353–360 (2009).\n",
      "\n",
      "88. Polo, J. M. et al. A molecular roadmap of\n",
      "\n",
      "reprogramming somatic cells into iPS cells. Cell 151, 1617–1632 (2012).\n",
      "\n",
      "89. Soufi, A., Donahue, G. & Zaret, K. S. Facilitators and impediments of the pluripotency reprogramming factors’ initial engagement with the genome. Cell 151, 994–1004 (2012).\n",
      "\n",
      "90. Sridharan, R. et al. Role of the murine reprogramming factors in the induction of pluripotency. Cell 136, 364–377 (2009).\n",
      "\n",
      "91. Feng, B., Ng, J. H., Heng, J. C. & Ng, H. H. Molecules that promote or enhance reprogramming of somatic cells to\n",
      "\n",
      "work from ref. 40, have shown that appropriate closing of AP-1 motif-enriched chromatin is required for efﬁcient reprogramming of MEFs into iPSCs. The second main region of interaction involves JUNB binding to reprogramming TF motifs-enriched (i.e., MEF2C, ASCL1, KLF4, and cMYC) closed chromatin (Fig. 3). In this context, our comparative chromatin accessibility analysis (siControl vs siAJSZ) during cardiac reprogramming of HDFs (Fig. 4), shows that under control conditions, pre-existing binding of JUNB to reprogramming TF (MEF2C) motif-enriched regions (domain 2), remains in a closed state even when MGT is overexpressed, thus contributing to limit repro- gramming TF access to target its DNA\n",
      "Human: How do transcription factors influence epigenetic modifications during the reprogramming process?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "==================================================\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  What are the main challenges and limitations of using transcription factor-mediated reprogramming for therapeutic purposes?\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fri Jun  7 23:23:01 2024 INFO RAG\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "RAG >> 8: \n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "through a plas- tic intermediate. Moreover, although the in vivo hematopoietic niche can provide all signals neces- sary for reprogramming, we cannot dissect the niche in this study, and thus cannot determine the ele- ments that support reprogramming. Furthermore, the inability to generate any hematopoietic cells other than macrophages and monocyte-like pro- genitors also limits future studies with this process unless they focus on these lineages.\n",
      "\n",
      "Hemogenic induction As previously stated, our laboratory became the ﬁrst to induce a hemogenic program in somatic cells via nonpluripotency TF overexpression in mouse embryonic ﬁbroblasts (MEFs). We ﬁrst performed a large screen by transducing\n",
      "\n",
      "equally to this manuscript.\n",
      "\n",
      "doi: 10.1111/nyas.12989\n",
      "\n",
      "24\n",
      "\n",
      "Ann. N.Y. Acad. Sci. 1370 (2016) 24–35 C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "Figure 1. Established TF reprogramming strategies. TF reprogramming has been shown to work across a variety of cell types. This ﬁgure shows the starting and end cell populations, as well as the TF cocktails used to derive each cell type of choice. Adapted from Ref. 25.\n",
      "\n",
      "The novel technology of induced pluripotent stem cells (iPSCs) from Yamanaka and colleagues instigated the entire ﬁeld of regenerative medicine. Using only four transcription factors (TFs), Oct3/4, Sox2, Klf4, and c-Myc,\n",
      "\n",
      "that Drive Neuronal Fate and Reprogramming. Cell Stem . Cell ​ .\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "41. A. Ebrahimi, K. Sevinç, G. Gürhan Sevinç, A. P. Cribbs, M. Philpott, F. Uyulur, T. Morova, J. E. Dunford, S. Göklemez, Ş. Arı, U. Oppermann, T. T. Önder, Bromodomain inhibition of the coactivators CBP/EP300 facilitate cellular reprogramming. 519–528 (2019).\n",
      "\n",
      "Nat. Chem. Biol. ​\n",
      "\n",
      "15\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      ",\n",
      "\n",
      "​\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made\n",
      "\n",
      "available under a\n",
      "\n",
      "CC-BY\n",
      "\n",
      "Cell Stem Cell 1, 55–70 (2007). 161. Park, I. H. et al. Reprogramming of human somatic\n",
      "\n",
      "cells to pluripotency with defined factors. Nature 451, 141–146 (2008).\n",
      "\n",
      "162. Hanna, J. et al. Direct cell reprogramming is a stochastic process amenable to acceleration. Nature 462, 595–601 (2009).\n",
      "\n",
      "163. Yang, J. et al. Stat3 activation is limiting for reprogramming to ground state pluripotency. Cell Stem Cell 7, 319–328 (2010).\n",
      "\n",
      "164. Silva, J. et al. Promotion of reprogramming to ground state pluripotency by signal inhibition. PLoS Biol. 6, e253 (2008).\n",
      "\n",
      "Acknowledgements The authors would like to thank all Yamanaka laboratory members, past and present, and are also grateful to Y. Miyake, R. Kato, E.\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "\n",
      "to C.-F.P.), and Fundac¸ ~ao para a Ci^encia e Tecnologia (SFRH/BD/51968/2012 and PTDC/BIM-MED/0075/2014 to A.M.G. and C.-F.P.). The Knut and Alice Wallenberg Foundation is acknowledged for generous support.\n",
      "\n",
      "AUTHOR CONTRIBUTIONS\n",
      "\n",
      "C.-F.P., A.M.G, B.C., and M.D. conducted reprogramming experiments. A.M.G., C.-F.P., B.C., I.K., A.L., Z.W., L.F., A.M., and D.P. analyzed mRNA- seq and ChIP-seq datasets. C.-F.P., K.A.M., K.L., N.S., and B.K.C. performed in vivo experiments and isolated UCB cells. C.-F.P., A.M.G., K.A.M., and I.R.L. designed the experiments and wrote the paper.\n",
      "\n",
      "DECLARATION OF INTERESTS\n",
      "\n",
      "The authors declare no competing interests.\n",
      "\n",
      "Received: January 18, 2018 Revised: August 19,\n",
      "\n",
      "the other hand, the technology can also be used to our advantage by enabling the identiﬁcation of genes involved in cancers, such as hepatocellu- lar carcinoma.96 These concerns may be overcome using modiﬁed RNAs97,98 or self-replicating RNA.99 Both of these methods have been shown to efﬁciently reprogram ﬁbroblasts into iPSCs.\n",
      "\n",
      "The ﬁeld of reprogramming somatic cells into HSPCs is in its infancy. With each new study we learn more about the reprogramming process, as it appears that there are multiple ways emerging to bring the process to fruition, with the eventual cre- ation of bona ﬁde HSCs. Many obstacles remain, the foremost being the need for expanding knowledge of the mechanisms that\n",
      "\n",
      "more reprogramming factors that can ‘erase’ the ‘memories’ of somatic cells.\n",
      "\n",
      "T I M E L I N E\n",
      "\n",
      "A decade of transcription factor‑mediated reprogramming to pluripotency\n",
      "\n",
      "Kazutoshi Takahashi and Shinya Yamanaka\n",
      "\n",
      "Abstract | The past 10 years have seen great advances in our ability to manipulate cell fate, including the induction of pluripotency in vitro to generate induced pluripotent stem cells (iPSCs). This process proved to be remarkably simple from a technical perspective, only needing the host cell and a defined cocktail of transcription factors, with four factors — octamer-binding protein 3/4 (OCT3/4), SOX2, Krüppel-like factor 4 (KLF4) and MYC (collectively referred to as OSKM) —\n",
      "\n",
      "in the field of cellular reprogramming. iPSCs are promising for many applications, including drug discovery, disease modelling and regenerative medicine23 (FIG. 2).\n",
      "\n",
      "Somatic cell nuclear transfer (SCNT)\n",
      "\n",
      "+\n",
      "\n",
      "Somatic cell (nucleus)\n",
      "\n",
      "Enucleated egg\n",
      "\n",
      "SCNT- ES cell\n",
      "\n",
      "Cell fusion with pluripotent stem cell (PSC)\n",
      "\n",
      "In this Timeline article, we first introduce how reprogramming factors were discovered and how they led to the first report of iPSCs 10 years ago. We also describe other factors that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated\n",
      "Human: What are the main challenges and limitations of using transcription factor-mediated reprogramming for therapeutic purposes?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "of ﬁbroblasts into functional cardiomyocytes by deﬁned factors. Cell 142, 375–386 (2010). 9. Haridhasapavalan, K. K., Raina, K., Dey, C., Adhikari, P. & Thummer,\n",
      "\n",
      "6.\n",
      "\n",
      "7.\n",
      "\n",
      "8.\n",
      "\n",
      "R. P. An insight into reprogramming barriers to iPSC generation. Stem Cell Rev. Rep. 16, 56–81 (2020).\n",
      "\n",
      "10. Takahashi, K. & Yamanaka, S. A decade of transcription factor-\n",
      "\n",
      "mediated reprogramming to pluripotency. Nat. Rev. Mol. Cell Biol. 17, 183–193 (2016).\n",
      "\n",
      "11. Brumbaugh, J., Di Stefano, B. & Hochedlinger, K. Reprogramming: identifying the mechanisms that safeguard cell identity. Develop- ment https://doi.org/10.1242/dev.182170 (2019).\n",
      "\n",
      "12. Dal-Pra, S., Hodgkinson, C. P., Mirotsou, M., Kirste, I. & Dzau, V. J.\n",
      "\n",
      "equally to this manuscript.\n",
      "\n",
      "doi: 10.1111/nyas.12989\n",
      "\n",
      "24\n",
      "\n",
      "Ann. N.Y. Acad. Sci. 1370 (2016) 24–35 C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "Figure 1. Established TF reprogramming strategies. TF reprogramming has been shown to work across a variety of cell types. This ﬁgure shows the starting and end cell populations, as well as the TF cocktails used to derive each cell type of choice. Adapted from Ref. 25.\n",
      "\n",
      "The novel technology of induced pluripotent stem cells (iPSCs) from Yamanaka and colleagues instigated the entire ﬁeld of regenerative medicine. Using only four transcription factors (TFs), Oct3/4, Sox2, Klf4, and c-Myc,\n",
      "\n",
      "that Drive Neuronal Fate and Reprogramming. Cell Stem . Cell ​ .\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "41. A. Ebrahimi, K. Sevinç, G. Gürhan Sevinç, A. P. Cribbs, M. Philpott, F. Uyulur, T. Morova, J. E. Dunford, S. Göklemez, Ş. Arı, U. Oppermann, T. T. Önder, Bromodomain inhibition of the coactivators CBP/EP300 facilitate cellular reprogramming. 519–528 (2019).\n",
      "\n",
      "Nat. Chem. Biol. ​\n",
      "\n",
      "15\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      ",\n",
      "\n",
      "​\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made\n",
      "\n",
      "available under a\n",
      "\n",
      "CC-BY\n",
      "\n",
      "Cell Stem Cell 1, 55–70 (2007). 161. Park, I. H. et al. Reprogramming of human somatic\n",
      "\n",
      "cells to pluripotency with defined factors. Nature 451, 141–146 (2008).\n",
      "\n",
      "162. Hanna, J. et al. Direct cell reprogramming is a stochastic process amenable to acceleration. Nature 462, 595–601 (2009).\n",
      "\n",
      "163. Yang, J. et al. Stat3 activation is limiting for reprogramming to ground state pluripotency. Cell Stem Cell 7, 319–328 (2010).\n",
      "\n",
      "164. Silva, J. et al. Promotion of reprogramming to ground state pluripotency by signal inhibition. PLoS Biol. 6, e253 (2008).\n",
      "\n",
      "Acknowledgements The authors would like to thank all Yamanaka laboratory members, past and present, and are also grateful to Y. Miyake, R. Kato, E.\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "\n",
      "to C.-F.P.), and Fundac¸ ~ao para a Ci^encia e Tecnologia (SFRH/BD/51968/2012 and PTDC/BIM-MED/0075/2014 to A.M.G. and C.-F.P.). The Knut and Alice Wallenberg Foundation is acknowledged for generous support.\n",
      "\n",
      "AUTHOR CONTRIBUTIONS\n",
      "\n",
      "C.-F.P., A.M.G, B.C., and M.D. conducted reprogramming experiments. A.M.G., C.-F.P., B.C., I.K., A.L., Z.W., L.F., A.M., and D.P. analyzed mRNA- seq and ChIP-seq datasets. C.-F.P., K.A.M., K.L., N.S., and B.K.C. performed in vivo experiments and isolated UCB cells. C.-F.P., A.M.G., K.A.M., and I.R.L. designed the experiments and wrote the paper.\n",
      "\n",
      "DECLARATION OF INTERESTS\n",
      "\n",
      "The authors declare no competing interests.\n",
      "\n",
      "Received: January 18, 2018 Revised: August 19,\n",
      "\n",
      "the other hand, the technology can also be used to our advantage by enabling the identiﬁcation of genes involved in cancers, such as hepatocellu- lar carcinoma.96 These concerns may be overcome using modiﬁed RNAs97,98 or self-replicating RNA.99 Both of these methods have been shown to efﬁciently reprogram ﬁbroblasts into iPSCs.\n",
      "\n",
      "The ﬁeld of reprogramming somatic cells into HSPCs is in its infancy. With each new study we learn more about the reprogramming process, as it appears that there are multiple ways emerging to bring the process to fruition, with the eventual cre- ation of bona ﬁde HSCs. Many obstacles remain, the foremost being the need for expanding knowledge of the mechanisms that\n",
      "\n",
      "more reprogramming factors that can ‘erase’ the ‘memories’ of somatic cells.\n",
      "\n",
      "T I M E L I N E\n",
      "\n",
      "A decade of transcription factor‑mediated reprogramming to pluripotency\n",
      "\n",
      "Kazutoshi Takahashi and Shinya Yamanaka\n",
      "\n",
      "Abstract | The past 10 years have seen great advances in our ability to manipulate cell fate, including the induction of pluripotency in vitro to generate induced pluripotent stem cells (iPSCs). This process proved to be remarkably simple from a technical perspective, only needing the host cell and a defined cocktail of transcription factors, with four factors — octamer-binding protein 3/4 (OCT3/4), SOX2, Krüppel-like factor 4 (KLF4) and MYC (collectively referred to as OSKM) —\n",
      "\n",
      "in the field of cellular reprogramming. iPSCs are promising for many applications, including drug discovery, disease modelling and regenerative medicine23 (FIG. 2).\n",
      "\n",
      "Somatic cell nuclear transfer (SCNT)\n",
      "\n",
      "+\n",
      "\n",
      "Somatic cell (nucleus)\n",
      "\n",
      "Enucleated egg\n",
      "\n",
      "SCNT- ES cell\n",
      "\n",
      "Cell fusion with pluripotent stem cell (PSC)\n",
      "\n",
      "In this Timeline article, we first introduce how reprogramming factors were discovered and how they led to the first report of iPSCs 10 years ago. We also describe other factors that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated\n",
      "Human: What are the main challenges and limitations of using transcription factor-mediated reprogramming for therapeutic purposes?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "of ﬁbroblasts into functional cardiomyocytes by deﬁned factors. Cell 142, 375–386 (2010). 9. Haridhasapavalan, K. K., Raina, K., Dey, C., Adhikari, P. & Thummer,\n",
      "\n",
      "6.\n",
      "\n",
      "7.\n",
      "\n",
      "8.\n",
      "\n",
      "R. P. An insight into reprogramming barriers to iPSC generation. Stem Cell Rev. Rep. 16, 56–81 (2020).\n",
      "\n",
      "10. Takahashi, K. & Yamanaka, S. A decade of transcription factor-\n",
      "\n",
      "mediated reprogramming to pluripotency. Nat. Rev. Mol. Cell Biol. 17, 183–193 (2016).\n",
      "\n",
      "11. Brumbaugh, J., Di Stefano, B. & Hochedlinger, K. Reprogramming: identifying the mechanisms that safeguard cell identity. Develop- ment https://doi.org/10.1242/dev.182170 (2019).\n",
      "\n",
      "12. Dal-Pra, S., Hodgkinson, C. P., Mirotsou, M., Kirste, I. & Dzau, V. J.\n",
      "\n",
      "through a plas- tic intermediate. Moreover, although the in vivo hematopoietic niche can provide all signals neces- sary for reprogramming, we cannot dissect the niche in this study, and thus cannot determine the ele- ments that support reprogramming. Furthermore, the inability to generate any hematopoietic cells other than macrophages and monocyte-like pro- genitors also limits future studies with this process unless they focus on these lineages.\n",
      "\n",
      "Hemogenic induction As previously stated, our laboratory became the ﬁrst to induce a hemogenic program in somatic cells via nonpluripotency TF overexpression in mouse embryonic ﬁbroblasts (MEFs). We ﬁrst performed a large screen by transducing\n",
      "\n",
      "that Drive Neuronal Fate and Reprogramming. Cell Stem . Cell ​ .\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "41. A. Ebrahimi, K. Sevinç, G. Gürhan Sevinç, A. P. Cribbs, M. Philpott, F. Uyulur, T. Morova, J. E. Dunford, S. Göklemez, Ş. Arı, U. Oppermann, T. T. Önder, Bromodomain inhibition of the coactivators CBP/EP300 facilitate cellular reprogramming. 519–528 (2019).\n",
      "\n",
      "Nat. Chem. Biol. ​\n",
      "\n",
      "15\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      ",\n",
      "\n",
      "​\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made\n",
      "\n",
      "available under a\n",
      "\n",
      "CC-BY\n",
      "\n",
      "Cell Stem Cell 1, 55–70 (2007). 161. Park, I. H. et al. Reprogramming of human somatic\n",
      "\n",
      "cells to pluripotency with defined factors. Nature 451, 141–146 (2008).\n",
      "\n",
      "162. Hanna, J. et al. Direct cell reprogramming is a stochastic process amenable to acceleration. Nature 462, 595–601 (2009).\n",
      "\n",
      "163. Yang, J. et al. Stat3 activation is limiting for reprogramming to ground state pluripotency. Cell Stem Cell 7, 319–328 (2010).\n",
      "\n",
      "164. Silva, J. et al. Promotion of reprogramming to ground state pluripotency by signal inhibition. PLoS Biol. 6, e253 (2008).\n",
      "\n",
      "Acknowledgements The authors would like to thank all Yamanaka laboratory members, past and present, and are also grateful to Y. Miyake, R. Kato, E.\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "\n",
      "to C.-F.P.), and Fundac¸ ~ao para a Ci^encia e Tecnologia (SFRH/BD/51968/2012 and PTDC/BIM-MED/0075/2014 to A.M.G. and C.-F.P.). The Knut and Alice Wallenberg Foundation is acknowledged for generous support.\n",
      "\n",
      "AUTHOR CONTRIBUTIONS\n",
      "\n",
      "C.-F.P., A.M.G, B.C., and M.D. conducted reprogramming experiments. A.M.G., C.-F.P., B.C., I.K., A.L., Z.W., L.F., A.M., and D.P. analyzed mRNA- seq and ChIP-seq datasets. C.-F.P., K.A.M., K.L., N.S., and B.K.C. performed in vivo experiments and isolated UCB cells. C.-F.P., A.M.G., K.A.M., and I.R.L. designed the experiments and wrote the paper.\n",
      "\n",
      "DECLARATION OF INTERESTS\n",
      "\n",
      "The authors declare no competing interests.\n",
      "\n",
      "Received: January 18, 2018 Revised: August 19,\n",
      "\n",
      "the other hand, the technology can also be used to our advantage by enabling the identiﬁcation of genes involved in cancers, such as hepatocellu- lar carcinoma.96 These concerns may be overcome using modiﬁed RNAs97,98 or self-replicating RNA.99 Both of these methods have been shown to efﬁciently reprogram ﬁbroblasts into iPSCs.\n",
      "\n",
      "The ﬁeld of reprogramming somatic cells into HSPCs is in its infancy. With each new study we learn more about the reprogramming process, as it appears that there are multiple ways emerging to bring the process to fruition, with the eventual cre- ation of bona ﬁde HSCs. Many obstacles remain, the foremost being the need for expanding knowledge of the mechanisms that\n",
      "\n",
      "more reprogramming factors that can ‘erase’ the ‘memories’ of somatic cells.\n",
      "\n",
      "T I M E L I N E\n",
      "\n",
      "A decade of transcription factor‑mediated reprogramming to pluripotency\n",
      "\n",
      "Kazutoshi Takahashi and Shinya Yamanaka\n",
      "\n",
      "Abstract | The past 10 years have seen great advances in our ability to manipulate cell fate, including the induction of pluripotency in vitro to generate induced pluripotent stem cells (iPSCs). This process proved to be remarkably simple from a technical perspective, only needing the host cell and a defined cocktail of transcription factors, with four factors — octamer-binding protein 3/4 (OCT3/4), SOX2, Krüppel-like factor 4 (KLF4) and MYC (collectively referred to as OSKM) —\n",
      "\n",
      "in the field of cellular reprogramming. iPSCs are promising for many applications, including drug discovery, disease modelling and regenerative medicine23 (FIG. 2).\n",
      "\n",
      "Somatic cell nuclear transfer (SCNT)\n",
      "\n",
      "+\n",
      "\n",
      "Somatic cell (nucleus)\n",
      "\n",
      "Enucleated egg\n",
      "\n",
      "SCNT- ES cell\n",
      "\n",
      "Cell fusion with pluripotent stem cell (PSC)\n",
      "\n",
      "In this Timeline article, we first introduce how reprogramming factors were discovered and how they led to the first report of iPSCs 10 years ago. We also describe other factors that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated\n",
      "Human: What are the main challenges and limitations of using transcription factor-mediated reprogramming for therapeutic purposes?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "of ﬁbroblasts into functional cardiomyocytes by deﬁned factors. Cell 142, 375–386 (2010). 9. Haridhasapavalan, K. K., Raina, K., Dey, C., Adhikari, P. & Thummer,\n",
      "\n",
      "6.\n",
      "\n",
      "7.\n",
      "\n",
      "8.\n",
      "\n",
      "R. P. An insight into reprogramming barriers to iPSC generation. Stem Cell Rev. Rep. 16, 56–81 (2020).\n",
      "\n",
      "10. Takahashi, K. & Yamanaka, S. A decade of transcription factor-\n",
      "\n",
      "mediated reprogramming to pluripotency. Nat. Rev. Mol. Cell Biol. 17, 183–193 (2016).\n",
      "\n",
      "11. Brumbaugh, J., Di Stefano, B. & Hochedlinger, K. Reprogramming: identifying the mechanisms that safeguard cell identity. Develop- ment https://doi.org/10.1242/dev.182170 (2019).\n",
      "\n",
      "12. Dal-Pra, S., Hodgkinson, C. P., Mirotsou, M., Kirste, I. & Dzau, V. J.\n",
      "\n",
      "through a plas- tic intermediate. Moreover, although the in vivo hematopoietic niche can provide all signals neces- sary for reprogramming, we cannot dissect the niche in this study, and thus cannot determine the ele- ments that support reprogramming. Furthermore, the inability to generate any hematopoietic cells other than macrophages and monocyte-like pro- genitors also limits future studies with this process unless they focus on these lineages.\n",
      "\n",
      "Hemogenic induction As previously stated, our laboratory became the ﬁrst to induce a hemogenic program in somatic cells via nonpluripotency TF overexpression in mouse embryonic ﬁbroblasts (MEFs). We ﬁrst performed a large screen by transducing\n",
      "\n",
      "equally to this manuscript.\n",
      "\n",
      "doi: 10.1111/nyas.12989\n",
      "\n",
      "24\n",
      "\n",
      "Ann. N.Y. Acad. Sci. 1370 (2016) 24–35 C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "Figure 1. Established TF reprogramming strategies. TF reprogramming has been shown to work across a variety of cell types. This ﬁgure shows the starting and end cell populations, as well as the TF cocktails used to derive each cell type of choice. Adapted from Ref. 25.\n",
      "\n",
      "The novel technology of induced pluripotent stem cells (iPSCs) from Yamanaka and colleagues instigated the entire ﬁeld of regenerative medicine. Using only four transcription factors (TFs), Oct3/4, Sox2, Klf4, and c-Myc,\n",
      "\n",
      "Cell Stem Cell 1, 55–70 (2007). 161. Park, I. H. et al. Reprogramming of human somatic\n",
      "\n",
      "cells to pluripotency with defined factors. Nature 451, 141–146 (2008).\n",
      "\n",
      "162. Hanna, J. et al. Direct cell reprogramming is a stochastic process amenable to acceleration. Nature 462, 595–601 (2009).\n",
      "\n",
      "163. Yang, J. et al. Stat3 activation is limiting for reprogramming to ground state pluripotency. Cell Stem Cell 7, 319–328 (2010).\n",
      "\n",
      "164. Silva, J. et al. Promotion of reprogramming to ground state pluripotency by signal inhibition. PLoS Biol. 6, e253 (2008).\n",
      "\n",
      "Acknowledgements The authors would like to thank all Yamanaka laboratory members, past and present, and are also grateful to Y. Miyake, R. Kato, E.\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "\n",
      "to C.-F.P.), and Fundac¸ ~ao para a Ci^encia e Tecnologia (SFRH/BD/51968/2012 and PTDC/BIM-MED/0075/2014 to A.M.G. and C.-F.P.). The Knut and Alice Wallenberg Foundation is acknowledged for generous support.\n",
      "\n",
      "AUTHOR CONTRIBUTIONS\n",
      "\n",
      "C.-F.P., A.M.G, B.C., and M.D. conducted reprogramming experiments. A.M.G., C.-F.P., B.C., I.K., A.L., Z.W., L.F., A.M., and D.P. analyzed mRNA- seq and ChIP-seq datasets. C.-F.P., K.A.M., K.L., N.S., and B.K.C. performed in vivo experiments and isolated UCB cells. C.-F.P., A.M.G., K.A.M., and I.R.L. designed the experiments and wrote the paper.\n",
      "\n",
      "DECLARATION OF INTERESTS\n",
      "\n",
      "The authors declare no competing interests.\n",
      "\n",
      "Received: January 18, 2018 Revised: August 19,\n",
      "\n",
      "the other hand, the technology can also be used to our advantage by enabling the identiﬁcation of genes involved in cancers, such as hepatocellu- lar carcinoma.96 These concerns may be overcome using modiﬁed RNAs97,98 or self-replicating RNA.99 Both of these methods have been shown to efﬁciently reprogram ﬁbroblasts into iPSCs.\n",
      "\n",
      "The ﬁeld of reprogramming somatic cells into HSPCs is in its infancy. With each new study we learn more about the reprogramming process, as it appears that there are multiple ways emerging to bring the process to fruition, with the eventual cre- ation of bona ﬁde HSCs. Many obstacles remain, the foremost being the need for expanding knowledge of the mechanisms that\n",
      "\n",
      "more reprogramming factors that can ‘erase’ the ‘memories’ of somatic cells.\n",
      "\n",
      "T I M E L I N E\n",
      "\n",
      "A decade of transcription factor‑mediated reprogramming to pluripotency\n",
      "\n",
      "Kazutoshi Takahashi and Shinya Yamanaka\n",
      "\n",
      "Abstract | The past 10 years have seen great advances in our ability to manipulate cell fate, including the induction of pluripotency in vitro to generate induced pluripotent stem cells (iPSCs). This process proved to be remarkably simple from a technical perspective, only needing the host cell and a defined cocktail of transcription factors, with four factors — octamer-binding protein 3/4 (OCT3/4), SOX2, Krüppel-like factor 4 (KLF4) and MYC (collectively referred to as OSKM) —\n",
      "\n",
      "in the field of cellular reprogramming. iPSCs are promising for many applications, including drug discovery, disease modelling and regenerative medicine23 (FIG. 2).\n",
      "\n",
      "Somatic cell nuclear transfer (SCNT)\n",
      "\n",
      "+\n",
      "\n",
      "Somatic cell (nucleus)\n",
      "\n",
      "Enucleated egg\n",
      "\n",
      "SCNT- ES cell\n",
      "\n",
      "Cell fusion with pluripotent stem cell (PSC)\n",
      "\n",
      "In this Timeline article, we first introduce how reprogramming factors were discovered and how they led to the first report of iPSCs 10 years ago. We also describe other factors that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated\n",
      "Human: What are the main challenges and limitations of using transcription factor-mediated reprogramming for therapeutic purposes?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "of ﬁbroblasts into functional cardiomyocytes by deﬁned factors. Cell 142, 375–386 (2010). 9. Haridhasapavalan, K. K., Raina, K., Dey, C., Adhikari, P. & Thummer,\n",
      "\n",
      "6.\n",
      "\n",
      "7.\n",
      "\n",
      "8.\n",
      "\n",
      "R. P. An insight into reprogramming barriers to iPSC generation. Stem Cell Rev. Rep. 16, 56–81 (2020).\n",
      "\n",
      "10. Takahashi, K. & Yamanaka, S. A decade of transcription factor-\n",
      "\n",
      "mediated reprogramming to pluripotency. Nat. Rev. Mol. Cell Biol. 17, 183–193 (2016).\n",
      "\n",
      "11. Brumbaugh, J., Di Stefano, B. & Hochedlinger, K. Reprogramming: identifying the mechanisms that safeguard cell identity. Develop- ment https://doi.org/10.1242/dev.182170 (2019).\n",
      "\n",
      "12. Dal-Pra, S., Hodgkinson, C. P., Mirotsou, M., Kirste, I. & Dzau, V. J.\n",
      "\n",
      "through a plas- tic intermediate. Moreover, although the in vivo hematopoietic niche can provide all signals neces- sary for reprogramming, we cannot dissect the niche in this study, and thus cannot determine the ele- ments that support reprogramming. Furthermore, the inability to generate any hematopoietic cells other than macrophages and monocyte-like pro- genitors also limits future studies with this process unless they focus on these lineages.\n",
      "\n",
      "Hemogenic induction As previously stated, our laboratory became the ﬁrst to induce a hemogenic program in somatic cells via nonpluripotency TF overexpression in mouse embryonic ﬁbroblasts (MEFs). We ﬁrst performed a large screen by transducing\n",
      "\n",
      "equally to this manuscript.\n",
      "\n",
      "doi: 10.1111/nyas.12989\n",
      "\n",
      "24\n",
      "\n",
      "Ann. N.Y. Acad. Sci. 1370 (2016) 24–35 C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "Figure 1. Established TF reprogramming strategies. TF reprogramming has been shown to work across a variety of cell types. This ﬁgure shows the starting and end cell populations, as well as the TF cocktails used to derive each cell type of choice. Adapted from Ref. 25.\n",
      "\n",
      "The novel technology of induced pluripotent stem cells (iPSCs) from Yamanaka and colleagues instigated the entire ﬁeld of regenerative medicine. Using only four transcription factors (TFs), Oct3/4, Sox2, Klf4, and c-Myc,\n",
      "\n",
      "that Drive Neuronal Fate and Reprogramming. Cell Stem . Cell ​ .\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "41. A. Ebrahimi, K. Sevinç, G. Gürhan Sevinç, A. P. Cribbs, M. Philpott, F. Uyulur, T. Morova, J. E. Dunford, S. Göklemez, Ş. Arı, U. Oppermann, T. T. Önder, Bromodomain inhibition of the coactivators CBP/EP300 facilitate cellular reprogramming. 519–528 (2019).\n",
      "\n",
      "Nat. Chem. Biol. ​\n",
      "\n",
      "15\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      ",\n",
      "\n",
      "​\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made\n",
      "\n",
      "available under a\n",
      "\n",
      "CC-BY\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "\n",
      "to C.-F.P.), and Fundac¸ ~ao para a Ci^encia e Tecnologia (SFRH/BD/51968/2012 and PTDC/BIM-MED/0075/2014 to A.M.G. and C.-F.P.). The Knut and Alice Wallenberg Foundation is acknowledged for generous support.\n",
      "\n",
      "AUTHOR CONTRIBUTIONS\n",
      "\n",
      "C.-F.P., A.M.G, B.C., and M.D. conducted reprogramming experiments. A.M.G., C.-F.P., B.C., I.K., A.L., Z.W., L.F., A.M., and D.P. analyzed mRNA- seq and ChIP-seq datasets. C.-F.P., K.A.M., K.L., N.S., and B.K.C. performed in vivo experiments and isolated UCB cells. C.-F.P., A.M.G., K.A.M., and I.R.L. designed the experiments and wrote the paper.\n",
      "\n",
      "DECLARATION OF INTERESTS\n",
      "\n",
      "The authors declare no competing interests.\n",
      "\n",
      "Received: January 18, 2018 Revised: August 19,\n",
      "\n",
      "the other hand, the technology can also be used to our advantage by enabling the identiﬁcation of genes involved in cancers, such as hepatocellu- lar carcinoma.96 These concerns may be overcome using modiﬁed RNAs97,98 or self-replicating RNA.99 Both of these methods have been shown to efﬁciently reprogram ﬁbroblasts into iPSCs.\n",
      "\n",
      "The ﬁeld of reprogramming somatic cells into HSPCs is in its infancy. With each new study we learn more about the reprogramming process, as it appears that there are multiple ways emerging to bring the process to fruition, with the eventual cre- ation of bona ﬁde HSCs. Many obstacles remain, the foremost being the need for expanding knowledge of the mechanisms that\n",
      "\n",
      "more reprogramming factors that can ‘erase’ the ‘memories’ of somatic cells.\n",
      "\n",
      "T I M E L I N E\n",
      "\n",
      "A decade of transcription factor‑mediated reprogramming to pluripotency\n",
      "\n",
      "Kazutoshi Takahashi and Shinya Yamanaka\n",
      "\n",
      "Abstract | The past 10 years have seen great advances in our ability to manipulate cell fate, including the induction of pluripotency in vitro to generate induced pluripotent stem cells (iPSCs). This process proved to be remarkably simple from a technical perspective, only needing the host cell and a defined cocktail of transcription factors, with four factors — octamer-binding protein 3/4 (OCT3/4), SOX2, Krüppel-like factor 4 (KLF4) and MYC (collectively referred to as OSKM) —\n",
      "\n",
      "in the field of cellular reprogramming. iPSCs are promising for many applications, including drug discovery, disease modelling and regenerative medicine23 (FIG. 2).\n",
      "\n",
      "Somatic cell nuclear transfer (SCNT)\n",
      "\n",
      "+\n",
      "\n",
      "Somatic cell (nucleus)\n",
      "\n",
      "Enucleated egg\n",
      "\n",
      "SCNT- ES cell\n",
      "\n",
      "Cell fusion with pluripotent stem cell (PSC)\n",
      "\n",
      "In this Timeline article, we first introduce how reprogramming factors were discovered and how they led to the first report of iPSCs 10 years ago. We also describe other factors that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated\n",
      "Human: What are the main challenges and limitations of using transcription factor-mediated reprogramming for therapeutic purposes?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "of ﬁbroblasts into functional cardiomyocytes by deﬁned factors. Cell 142, 375–386 (2010). 9. Haridhasapavalan, K. K., Raina, K., Dey, C., Adhikari, P. & Thummer,\n",
      "\n",
      "6.\n",
      "\n",
      "7.\n",
      "\n",
      "8.\n",
      "\n",
      "R. P. An insight into reprogramming barriers to iPSC generation. Stem Cell Rev. Rep. 16, 56–81 (2020).\n",
      "\n",
      "10. Takahashi, K. & Yamanaka, S. A decade of transcription factor-\n",
      "\n",
      "mediated reprogramming to pluripotency. Nat. Rev. Mol. Cell Biol. 17, 183–193 (2016).\n",
      "\n",
      "11. Brumbaugh, J., Di Stefano, B. & Hochedlinger, K. Reprogramming: identifying the mechanisms that safeguard cell identity. Develop- ment https://doi.org/10.1242/dev.182170 (2019).\n",
      "\n",
      "12. Dal-Pra, S., Hodgkinson, C. P., Mirotsou, M., Kirste, I. & Dzau, V. J.\n",
      "\n",
      "through a plas- tic intermediate. Moreover, although the in vivo hematopoietic niche can provide all signals neces- sary for reprogramming, we cannot dissect the niche in this study, and thus cannot determine the ele- ments that support reprogramming. Furthermore, the inability to generate any hematopoietic cells other than macrophages and monocyte-like pro- genitors also limits future studies with this process unless they focus on these lineages.\n",
      "\n",
      "Hemogenic induction As previously stated, our laboratory became the ﬁrst to induce a hemogenic program in somatic cells via nonpluripotency TF overexpression in mouse embryonic ﬁbroblasts (MEFs). We ﬁrst performed a large screen by transducing\n",
      "\n",
      "equally to this manuscript.\n",
      "\n",
      "doi: 10.1111/nyas.12989\n",
      "\n",
      "24\n",
      "\n",
      "Ann. N.Y. Acad. Sci. 1370 (2016) 24–35 C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "Figure 1. Established TF reprogramming strategies. TF reprogramming has been shown to work across a variety of cell types. This ﬁgure shows the starting and end cell populations, as well as the TF cocktails used to derive each cell type of choice. Adapted from Ref. 25.\n",
      "\n",
      "The novel technology of induced pluripotent stem cells (iPSCs) from Yamanaka and colleagues instigated the entire ﬁeld of regenerative medicine. Using only four transcription factors (TFs), Oct3/4, Sox2, Klf4, and c-Myc,\n",
      "\n",
      "that Drive Neuronal Fate and Reprogramming. Cell Stem . Cell ​ .\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "41. A. Ebrahimi, K. Sevinç, G. Gürhan Sevinç, A. P. Cribbs, M. Philpott, F. Uyulur, T. Morova, J. E. Dunford, S. Göklemez, Ş. Arı, U. Oppermann, T. T. Önder, Bromodomain inhibition of the coactivators CBP/EP300 facilitate cellular reprogramming. 519–528 (2019).\n",
      "\n",
      "Nat. Chem. Biol. ​\n",
      "\n",
      "15\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      ",\n",
      "\n",
      "​\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made\n",
      "\n",
      "available under a\n",
      "\n",
      "CC-BY\n",
      "\n",
      "Cell Stem Cell 1, 55–70 (2007). 161. Park, I. H. et al. Reprogramming of human somatic\n",
      "\n",
      "cells to pluripotency with defined factors. Nature 451, 141–146 (2008).\n",
      "\n",
      "162. Hanna, J. et al. Direct cell reprogramming is a stochastic process amenable to acceleration. Nature 462, 595–601 (2009).\n",
      "\n",
      "163. Yang, J. et al. Stat3 activation is limiting for reprogramming to ground state pluripotency. Cell Stem Cell 7, 319–328 (2010).\n",
      "\n",
      "164. Silva, J. et al. Promotion of reprogramming to ground state pluripotency by signal inhibition. PLoS Biol. 6, e253 (2008).\n",
      "\n",
      "Acknowledgements The authors would like to thank all Yamanaka laboratory members, past and present, and are also grateful to Y. Miyake, R. Kato, E.\n",
      "\n",
      "to C.-F.P.), and Fundac¸ ~ao para a Ci^encia e Tecnologia (SFRH/BD/51968/2012 and PTDC/BIM-MED/0075/2014 to A.M.G. and C.-F.P.). The Knut and Alice Wallenberg Foundation is acknowledged for generous support.\n",
      "\n",
      "AUTHOR CONTRIBUTIONS\n",
      "\n",
      "C.-F.P., A.M.G, B.C., and M.D. conducted reprogramming experiments. A.M.G., C.-F.P., B.C., I.K., A.L., Z.W., L.F., A.M., and D.P. analyzed mRNA- seq and ChIP-seq datasets. C.-F.P., K.A.M., K.L., N.S., and B.K.C. performed in vivo experiments and isolated UCB cells. C.-F.P., A.M.G., K.A.M., and I.R.L. designed the experiments and wrote the paper.\n",
      "\n",
      "DECLARATION OF INTERESTS\n",
      "\n",
      "The authors declare no competing interests.\n",
      "\n",
      "Received: January 18, 2018 Revised: August 19,\n",
      "\n",
      "the other hand, the technology can also be used to our advantage by enabling the identiﬁcation of genes involved in cancers, such as hepatocellu- lar carcinoma.96 These concerns may be overcome using modiﬁed RNAs97,98 or self-replicating RNA.99 Both of these methods have been shown to efﬁciently reprogram ﬁbroblasts into iPSCs.\n",
      "\n",
      "The ﬁeld of reprogramming somatic cells into HSPCs is in its infancy. With each new study we learn more about the reprogramming process, as it appears that there are multiple ways emerging to bring the process to fruition, with the eventual cre- ation of bona ﬁde HSCs. Many obstacles remain, the foremost being the need for expanding knowledge of the mechanisms that\n",
      "\n",
      "more reprogramming factors that can ‘erase’ the ‘memories’ of somatic cells.\n",
      "\n",
      "T I M E L I N E\n",
      "\n",
      "A decade of transcription factor‑mediated reprogramming to pluripotency\n",
      "\n",
      "Kazutoshi Takahashi and Shinya Yamanaka\n",
      "\n",
      "Abstract | The past 10 years have seen great advances in our ability to manipulate cell fate, including the induction of pluripotency in vitro to generate induced pluripotent stem cells (iPSCs). This process proved to be remarkably simple from a technical perspective, only needing the host cell and a defined cocktail of transcription factors, with four factors — octamer-binding protein 3/4 (OCT3/4), SOX2, Krüppel-like factor 4 (KLF4) and MYC (collectively referred to as OSKM) —\n",
      "\n",
      "in the field of cellular reprogramming. iPSCs are promising for many applications, including drug discovery, disease modelling and regenerative medicine23 (FIG. 2).\n",
      "\n",
      "Somatic cell nuclear transfer (SCNT)\n",
      "\n",
      "+\n",
      "\n",
      "Somatic cell (nucleus)\n",
      "\n",
      "Enucleated egg\n",
      "\n",
      "SCNT- ES cell\n",
      "\n",
      "Cell fusion with pluripotent stem cell (PSC)\n",
      "\n",
      "In this Timeline article, we first introduce how reprogramming factors were discovered and how they led to the first report of iPSCs 10 years ago. We also describe other factors that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated\n",
      "Human: What are the main challenges and limitations of using transcription factor-mediated reprogramming for therapeutic purposes?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "of ﬁbroblasts into functional cardiomyocytes by deﬁned factors. Cell 142, 375–386 (2010). 9. Haridhasapavalan, K. K., Raina, K., Dey, C., Adhikari, P. & Thummer,\n",
      "\n",
      "6.\n",
      "\n",
      "7.\n",
      "\n",
      "8.\n",
      "\n",
      "R. P. An insight into reprogramming barriers to iPSC generation. Stem Cell Rev. Rep. 16, 56–81 (2020).\n",
      "\n",
      "10. Takahashi, K. & Yamanaka, S. A decade of transcription factor-\n",
      "\n",
      "mediated reprogramming to pluripotency. Nat. Rev. Mol. Cell Biol. 17, 183–193 (2016).\n",
      "\n",
      "11. Brumbaugh, J., Di Stefano, B. & Hochedlinger, K. Reprogramming: identifying the mechanisms that safeguard cell identity. Develop- ment https://doi.org/10.1242/dev.182170 (2019).\n",
      "\n",
      "12. Dal-Pra, S., Hodgkinson, C. P., Mirotsou, M., Kirste, I. & Dzau, V. J.\n",
      "\n",
      "through a plas- tic intermediate. Moreover, although the in vivo hematopoietic niche can provide all signals neces- sary for reprogramming, we cannot dissect the niche in this study, and thus cannot determine the ele- ments that support reprogramming. Furthermore, the inability to generate any hematopoietic cells other than macrophages and monocyte-like pro- genitors also limits future studies with this process unless they focus on these lineages.\n",
      "\n",
      "Hemogenic induction As previously stated, our laboratory became the ﬁrst to induce a hemogenic program in somatic cells via nonpluripotency TF overexpression in mouse embryonic ﬁbroblasts (MEFs). We ﬁrst performed a large screen by transducing\n",
      "\n",
      "equally to this manuscript.\n",
      "\n",
      "doi: 10.1111/nyas.12989\n",
      "\n",
      "24\n",
      "\n",
      "Ann. N.Y. Acad. Sci. 1370 (2016) 24–35 C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "Figure 1. Established TF reprogramming strategies. TF reprogramming has been shown to work across a variety of cell types. This ﬁgure shows the starting and end cell populations, as well as the TF cocktails used to derive each cell type of choice. Adapted from Ref. 25.\n",
      "\n",
      "The novel technology of induced pluripotent stem cells (iPSCs) from Yamanaka and colleagues instigated the entire ﬁeld of regenerative medicine. Using only four transcription factors (TFs), Oct3/4, Sox2, Klf4, and c-Myc,\n",
      "\n",
      "that Drive Neuronal Fate and Reprogramming. Cell Stem . Cell ​ .\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "41. A. Ebrahimi, K. Sevinç, G. Gürhan Sevinç, A. P. Cribbs, M. Philpott, F. Uyulur, T. Morova, J. E. Dunford, S. Göklemez, Ş. Arı, U. Oppermann, T. T. Önder, Bromodomain inhibition of the coactivators CBP/EP300 facilitate cellular reprogramming. 519–528 (2019).\n",
      "\n",
      "Nat. Chem. Biol. ​\n",
      "\n",
      "15\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      ",\n",
      "\n",
      "​\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made\n",
      "\n",
      "available under a\n",
      "\n",
      "CC-BY\n",
      "\n",
      "Cell Stem Cell 1, 55–70 (2007). 161. Park, I. H. et al. Reprogramming of human somatic\n",
      "\n",
      "cells to pluripotency with defined factors. Nature 451, 141–146 (2008).\n",
      "\n",
      "162. Hanna, J. et al. Direct cell reprogramming is a stochastic process amenable to acceleration. Nature 462, 595–601 (2009).\n",
      "\n",
      "163. Yang, J. et al. Stat3 activation is limiting for reprogramming to ground state pluripotency. Cell Stem Cell 7, 319–328 (2010).\n",
      "\n",
      "164. Silva, J. et al. Promotion of reprogramming to ground state pluripotency by signal inhibition. PLoS Biol. 6, e253 (2008).\n",
      "\n",
      "Acknowledgements The authors would like to thank all Yamanaka laboratory members, past and present, and are also grateful to Y. Miyake, R. Kato, E.\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "\n",
      "the other hand, the technology can also be used to our advantage by enabling the identiﬁcation of genes involved in cancers, such as hepatocellu- lar carcinoma.96 These concerns may be overcome using modiﬁed RNAs97,98 or self-replicating RNA.99 Both of these methods have been shown to efﬁciently reprogram ﬁbroblasts into iPSCs.\n",
      "\n",
      "The ﬁeld of reprogramming somatic cells into HSPCs is in its infancy. With each new study we learn more about the reprogramming process, as it appears that there are multiple ways emerging to bring the process to fruition, with the eventual cre- ation of bona ﬁde HSCs. Many obstacles remain, the foremost being the need for expanding knowledge of the mechanisms that\n",
      "\n",
      "more reprogramming factors that can ‘erase’ the ‘memories’ of somatic cells.\n",
      "\n",
      "T I M E L I N E\n",
      "\n",
      "A decade of transcription factor‑mediated reprogramming to pluripotency\n",
      "\n",
      "Kazutoshi Takahashi and Shinya Yamanaka\n",
      "\n",
      "Abstract | The past 10 years have seen great advances in our ability to manipulate cell fate, including the induction of pluripotency in vitro to generate induced pluripotent stem cells (iPSCs). This process proved to be remarkably simple from a technical perspective, only needing the host cell and a defined cocktail of transcription factors, with four factors — octamer-binding protein 3/4 (OCT3/4), SOX2, Krüppel-like factor 4 (KLF4) and MYC (collectively referred to as OSKM) —\n",
      "\n",
      "in the field of cellular reprogramming. iPSCs are promising for many applications, including drug discovery, disease modelling and regenerative medicine23 (FIG. 2).\n",
      "\n",
      "Somatic cell nuclear transfer (SCNT)\n",
      "\n",
      "+\n",
      "\n",
      "Somatic cell (nucleus)\n",
      "\n",
      "Enucleated egg\n",
      "\n",
      "SCNT- ES cell\n",
      "\n",
      "Cell fusion with pluripotent stem cell (PSC)\n",
      "\n",
      "In this Timeline article, we first introduce how reprogramming factors were discovered and how they led to the first report of iPSCs 10 years ago. We also describe other factors that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated\n",
      "Human: What are the main challenges and limitations of using transcription factor-mediated reprogramming for therapeutic purposes?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "of ﬁbroblasts into functional cardiomyocytes by deﬁned factors. Cell 142, 375–386 (2010). 9. Haridhasapavalan, K. K., Raina, K., Dey, C., Adhikari, P. & Thummer,\n",
      "\n",
      "6.\n",
      "\n",
      "7.\n",
      "\n",
      "8.\n",
      "\n",
      "R. P. An insight into reprogramming barriers to iPSC generation. Stem Cell Rev. Rep. 16, 56–81 (2020).\n",
      "\n",
      "10. Takahashi, K. & Yamanaka, S. A decade of transcription factor-\n",
      "\n",
      "mediated reprogramming to pluripotency. Nat. Rev. Mol. Cell Biol. 17, 183–193 (2016).\n",
      "\n",
      "11. Brumbaugh, J., Di Stefano, B. & Hochedlinger, K. Reprogramming: identifying the mechanisms that safeguard cell identity. Develop- ment https://doi.org/10.1242/dev.182170 (2019).\n",
      "\n",
      "12. Dal-Pra, S., Hodgkinson, C. P., Mirotsou, M., Kirste, I. & Dzau, V. J.\n",
      "\n",
      "through a plas- tic intermediate. Moreover, although the in vivo hematopoietic niche can provide all signals neces- sary for reprogramming, we cannot dissect the niche in this study, and thus cannot determine the ele- ments that support reprogramming. Furthermore, the inability to generate any hematopoietic cells other than macrophages and monocyte-like pro- genitors also limits future studies with this process unless they focus on these lineages.\n",
      "\n",
      "Hemogenic induction As previously stated, our laboratory became the ﬁrst to induce a hemogenic program in somatic cells via nonpluripotency TF overexpression in mouse embryonic ﬁbroblasts (MEFs). We ﬁrst performed a large screen by transducing\n",
      "\n",
      "equally to this manuscript.\n",
      "\n",
      "doi: 10.1111/nyas.12989\n",
      "\n",
      "24\n",
      "\n",
      "Ann. N.Y. Acad. Sci. 1370 (2016) 24–35 C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "Figure 1. Established TF reprogramming strategies. TF reprogramming has been shown to work across a variety of cell types. This ﬁgure shows the starting and end cell populations, as well as the TF cocktails used to derive each cell type of choice. Adapted from Ref. 25.\n",
      "\n",
      "The novel technology of induced pluripotent stem cells (iPSCs) from Yamanaka and colleagues instigated the entire ﬁeld of regenerative medicine. Using only four transcription factors (TFs), Oct3/4, Sox2, Klf4, and c-Myc,\n",
      "\n",
      "that Drive Neuronal Fate and Reprogramming. Cell Stem . Cell ​ .\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "41. A. Ebrahimi, K. Sevinç, G. Gürhan Sevinç, A. P. Cribbs, M. Philpott, F. Uyulur, T. Morova, J. E. Dunford, S. Göklemez, Ş. Arı, U. Oppermann, T. T. Önder, Bromodomain inhibition of the coactivators CBP/EP300 facilitate cellular reprogramming. 519–528 (2019).\n",
      "\n",
      "Nat. Chem. Biol. ​\n",
      "\n",
      "15\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      ",\n",
      "\n",
      "​\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made\n",
      "\n",
      "available under a\n",
      "\n",
      "CC-BY\n",
      "\n",
      "Cell Stem Cell 1, 55–70 (2007). 161. Park, I. H. et al. Reprogramming of human somatic\n",
      "\n",
      "cells to pluripotency with defined factors. Nature 451, 141–146 (2008).\n",
      "\n",
      "162. Hanna, J. et al. Direct cell reprogramming is a stochastic process amenable to acceleration. Nature 462, 595–601 (2009).\n",
      "\n",
      "163. Yang, J. et al. Stat3 activation is limiting for reprogramming to ground state pluripotency. Cell Stem Cell 7, 319–328 (2010).\n",
      "\n",
      "164. Silva, J. et al. Promotion of reprogramming to ground state pluripotency by signal inhibition. PLoS Biol. 6, e253 (2008).\n",
      "\n",
      "Acknowledgements The authors would like to thank all Yamanaka laboratory members, past and present, and are also grateful to Y. Miyake, R. Kato, E.\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "\n",
      "to C.-F.P.), and Fundac¸ ~ao para a Ci^encia e Tecnologia (SFRH/BD/51968/2012 and PTDC/BIM-MED/0075/2014 to A.M.G. and C.-F.P.). The Knut and Alice Wallenberg Foundation is acknowledged for generous support.\n",
      "\n",
      "AUTHOR CONTRIBUTIONS\n",
      "\n",
      "C.-F.P., A.M.G, B.C., and M.D. conducted reprogramming experiments. A.M.G., C.-F.P., B.C., I.K., A.L., Z.W., L.F., A.M., and D.P. analyzed mRNA- seq and ChIP-seq datasets. C.-F.P., K.A.M., K.L., N.S., and B.K.C. performed in vivo experiments and isolated UCB cells. C.-F.P., A.M.G., K.A.M., and I.R.L. designed the experiments and wrote the paper.\n",
      "\n",
      "DECLARATION OF INTERESTS\n",
      "\n",
      "The authors declare no competing interests.\n",
      "\n",
      "Received: January 18, 2018 Revised: August 19,\n",
      "\n",
      "more reprogramming factors that can ‘erase’ the ‘memories’ of somatic cells.\n",
      "\n",
      "T I M E L I N E\n",
      "\n",
      "A decade of transcription factor‑mediated reprogramming to pluripotency\n",
      "\n",
      "Kazutoshi Takahashi and Shinya Yamanaka\n",
      "\n",
      "Abstract | The past 10 years have seen great advances in our ability to manipulate cell fate, including the induction of pluripotency in vitro to generate induced pluripotent stem cells (iPSCs). This process proved to be remarkably simple from a technical perspective, only needing the host cell and a defined cocktail of transcription factors, with four factors — octamer-binding protein 3/4 (OCT3/4), SOX2, Krüppel-like factor 4 (KLF4) and MYC (collectively referred to as OSKM) —\n",
      "\n",
      "in the field of cellular reprogramming. iPSCs are promising for many applications, including drug discovery, disease modelling and regenerative medicine23 (FIG. 2).\n",
      "\n",
      "Somatic cell nuclear transfer (SCNT)\n",
      "\n",
      "+\n",
      "\n",
      "Somatic cell (nucleus)\n",
      "\n",
      "Enucleated egg\n",
      "\n",
      "SCNT- ES cell\n",
      "\n",
      "Cell fusion with pluripotent stem cell (PSC)\n",
      "\n",
      "In this Timeline article, we first introduce how reprogramming factors were discovered and how they led to the first report of iPSCs 10 years ago. We also describe other factors that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated\n",
      "Human: What are the main challenges and limitations of using transcription factor-mediated reprogramming for therapeutic purposes?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "of ﬁbroblasts into functional cardiomyocytes by deﬁned factors. Cell 142, 375–386 (2010). 9. Haridhasapavalan, K. K., Raina, K., Dey, C., Adhikari, P. & Thummer,\n",
      "\n",
      "6.\n",
      "\n",
      "7.\n",
      "\n",
      "8.\n",
      "\n",
      "R. P. An insight into reprogramming barriers to iPSC generation. Stem Cell Rev. Rep. 16, 56–81 (2020).\n",
      "\n",
      "10. Takahashi, K. & Yamanaka, S. A decade of transcription factor-\n",
      "\n",
      "mediated reprogramming to pluripotency. Nat. Rev. Mol. Cell Biol. 17, 183–193 (2016).\n",
      "\n",
      "11. Brumbaugh, J., Di Stefano, B. & Hochedlinger, K. Reprogramming: identifying the mechanisms that safeguard cell identity. Develop- ment https://doi.org/10.1242/dev.182170 (2019).\n",
      "\n",
      "12. Dal-Pra, S., Hodgkinson, C. P., Mirotsou, M., Kirste, I. & Dzau, V. J.\n",
      "\n",
      "through a plas- tic intermediate. Moreover, although the in vivo hematopoietic niche can provide all signals neces- sary for reprogramming, we cannot dissect the niche in this study, and thus cannot determine the ele- ments that support reprogramming. Furthermore, the inability to generate any hematopoietic cells other than macrophages and monocyte-like pro- genitors also limits future studies with this process unless they focus on these lineages.\n",
      "\n",
      "Hemogenic induction As previously stated, our laboratory became the ﬁrst to induce a hemogenic program in somatic cells via nonpluripotency TF overexpression in mouse embryonic ﬁbroblasts (MEFs). We ﬁrst performed a large screen by transducing\n",
      "\n",
      "equally to this manuscript.\n",
      "\n",
      "doi: 10.1111/nyas.12989\n",
      "\n",
      "24\n",
      "\n",
      "Ann. N.Y. Acad. Sci. 1370 (2016) 24–35 C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "Figure 1. Established TF reprogramming strategies. TF reprogramming has been shown to work across a variety of cell types. This ﬁgure shows the starting and end cell populations, as well as the TF cocktails used to derive each cell type of choice. Adapted from Ref. 25.\n",
      "\n",
      "The novel technology of induced pluripotent stem cells (iPSCs) from Yamanaka and colleagues instigated the entire ﬁeld of regenerative medicine. Using only four transcription factors (TFs), Oct3/4, Sox2, Klf4, and c-Myc,\n",
      "\n",
      "that Drive Neuronal Fate and Reprogramming. Cell Stem . Cell ​ .\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "41. A. Ebrahimi, K. Sevinç, G. Gürhan Sevinç, A. P. Cribbs, M. Philpott, F. Uyulur, T. Morova, J. E. Dunford, S. Göklemez, Ş. Arı, U. Oppermann, T. T. Önder, Bromodomain inhibition of the coactivators CBP/EP300 facilitate cellular reprogramming. 519–528 (2019).\n",
      "\n",
      "Nat. Chem. Biol. ​\n",
      "\n",
      "15\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      ",\n",
      "\n",
      "​\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made\n",
      "\n",
      "available under a\n",
      "\n",
      "CC-BY\n",
      "\n",
      "Cell Stem Cell 1, 55–70 (2007). 161. Park, I. H. et al. Reprogramming of human somatic\n",
      "\n",
      "cells to pluripotency with defined factors. Nature 451, 141–146 (2008).\n",
      "\n",
      "162. Hanna, J. et al. Direct cell reprogramming is a stochastic process amenable to acceleration. Nature 462, 595–601 (2009).\n",
      "\n",
      "163. Yang, J. et al. Stat3 activation is limiting for reprogramming to ground state pluripotency. Cell Stem Cell 7, 319–328 (2010).\n",
      "\n",
      "164. Silva, J. et al. Promotion of reprogramming to ground state pluripotency by signal inhibition. PLoS Biol. 6, e253 (2008).\n",
      "\n",
      "Acknowledgements The authors would like to thank all Yamanaka laboratory members, past and present, and are also grateful to Y. Miyake, R. Kato, E.\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "\n",
      "to C.-F.P.), and Fundac¸ ~ao para a Ci^encia e Tecnologia (SFRH/BD/51968/2012 and PTDC/BIM-MED/0075/2014 to A.M.G. and C.-F.P.). The Knut and Alice Wallenberg Foundation is acknowledged for generous support.\n",
      "\n",
      "AUTHOR CONTRIBUTIONS\n",
      "\n",
      "C.-F.P., A.M.G, B.C., and M.D. conducted reprogramming experiments. A.M.G., C.-F.P., B.C., I.K., A.L., Z.W., L.F., A.M., and D.P. analyzed mRNA- seq and ChIP-seq datasets. C.-F.P., K.A.M., K.L., N.S., and B.K.C. performed in vivo experiments and isolated UCB cells. C.-F.P., A.M.G., K.A.M., and I.R.L. designed the experiments and wrote the paper.\n",
      "\n",
      "DECLARATION OF INTERESTS\n",
      "\n",
      "The authors declare no competing interests.\n",
      "\n",
      "Received: January 18, 2018 Revised: August 19,\n",
      "\n",
      "the other hand, the technology can also be used to our advantage by enabling the identiﬁcation of genes involved in cancers, such as hepatocellu- lar carcinoma.96 These concerns may be overcome using modiﬁed RNAs97,98 or self-replicating RNA.99 Both of these methods have been shown to efﬁciently reprogram ﬁbroblasts into iPSCs.\n",
      "\n",
      "The ﬁeld of reprogramming somatic cells into HSPCs is in its infancy. With each new study we learn more about the reprogramming process, as it appears that there are multiple ways emerging to bring the process to fruition, with the eventual cre- ation of bona ﬁde HSCs. Many obstacles remain, the foremost being the need for expanding knowledge of the mechanisms that\n",
      "\n",
      "in the field of cellular reprogramming. iPSCs are promising for many applications, including drug discovery, disease modelling and regenerative medicine23 (FIG. 2).\n",
      "\n",
      "Somatic cell nuclear transfer (SCNT)\n",
      "\n",
      "+\n",
      "\n",
      "Somatic cell (nucleus)\n",
      "\n",
      "Enucleated egg\n",
      "\n",
      "SCNT- ES cell\n",
      "\n",
      "Cell fusion with pluripotent stem cell (PSC)\n",
      "\n",
      "In this Timeline article, we first introduce how reprogramming factors were discovered and how they led to the first report of iPSCs 10 years ago. We also describe other factors that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated\n",
      "Human: What are the main challenges and limitations of using transcription factor-mediated reprogramming for therapeutic purposes?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "of ﬁbroblasts into functional cardiomyocytes by deﬁned factors. Cell 142, 375–386 (2010). 9. Haridhasapavalan, K. K., Raina, K., Dey, C., Adhikari, P. & Thummer,\n",
      "\n",
      "6.\n",
      "\n",
      "7.\n",
      "\n",
      "8.\n",
      "\n",
      "R. P. An insight into reprogramming barriers to iPSC generation. Stem Cell Rev. Rep. 16, 56–81 (2020).\n",
      "\n",
      "10. Takahashi, K. & Yamanaka, S. A decade of transcription factor-\n",
      "\n",
      "mediated reprogramming to pluripotency. Nat. Rev. Mol. Cell Biol. 17, 183–193 (2016).\n",
      "\n",
      "11. Brumbaugh, J., Di Stefano, B. & Hochedlinger, K. Reprogramming: identifying the mechanisms that safeguard cell identity. Develop- ment https://doi.org/10.1242/dev.182170 (2019).\n",
      "\n",
      "12. Dal-Pra, S., Hodgkinson, C. P., Mirotsou, M., Kirste, I. & Dzau, V. J.\n",
      "\n",
      "through a plas- tic intermediate. Moreover, although the in vivo hematopoietic niche can provide all signals neces- sary for reprogramming, we cannot dissect the niche in this study, and thus cannot determine the ele- ments that support reprogramming. Furthermore, the inability to generate any hematopoietic cells other than macrophages and monocyte-like pro- genitors also limits future studies with this process unless they focus on these lineages.\n",
      "\n",
      "Hemogenic induction As previously stated, our laboratory became the ﬁrst to induce a hemogenic program in somatic cells via nonpluripotency TF overexpression in mouse embryonic ﬁbroblasts (MEFs). We ﬁrst performed a large screen by transducing\n",
      "\n",
      "equally to this manuscript.\n",
      "\n",
      "doi: 10.1111/nyas.12989\n",
      "\n",
      "24\n",
      "\n",
      "Ann. N.Y. Acad. Sci. 1370 (2016) 24–35 C(cid:3) 2016 New York Academy of Sciences.\n",
      "\n",
      "Daniel et al.\n",
      "\n",
      "Transcription factor induction of hematopoiesis\n",
      "\n",
      "Figure 1. Established TF reprogramming strategies. TF reprogramming has been shown to work across a variety of cell types. This ﬁgure shows the starting and end cell populations, as well as the TF cocktails used to derive each cell type of choice. Adapted from Ref. 25.\n",
      "\n",
      "The novel technology of induced pluripotent stem cells (iPSCs) from Yamanaka and colleagues instigated the entire ﬁeld of regenerative medicine. Using only four transcription factors (TFs), Oct3/4, Sox2, Klf4, and c-Myc,\n",
      "\n",
      "that Drive Neuronal Fate and Reprogramming. Cell Stem . Cell ​ .\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "41. A. Ebrahimi, K. Sevinç, G. Gürhan Sevinç, A. P. Cribbs, M. Philpott, F. Uyulur, T. Morova, J. E. Dunford, S. Göklemez, Ş. Arı, U. Oppermann, T. T. Önder, Bromodomain inhibition of the coactivators CBP/EP300 facilitate cellular reprogramming. 519–528 (2019).\n",
      "\n",
      "Nat. Chem. Biol. ​\n",
      "\n",
      "15\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      ",\n",
      "\n",
      "​\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made\n",
      "\n",
      "available under a\n",
      "\n",
      "CC-BY\n",
      "\n",
      "Cell Stem Cell 1, 55–70 (2007). 161. Park, I. H. et al. Reprogramming of human somatic\n",
      "\n",
      "cells to pluripotency with defined factors. Nature 451, 141–146 (2008).\n",
      "\n",
      "162. Hanna, J. et al. Direct cell reprogramming is a stochastic process amenable to acceleration. Nature 462, 595–601 (2009).\n",
      "\n",
      "163. Yang, J. et al. Stat3 activation is limiting for reprogramming to ground state pluripotency. Cell Stem Cell 7, 319–328 (2010).\n",
      "\n",
      "164. Silva, J. et al. Promotion of reprogramming to ground state pluripotency by signal inhibition. PLoS Biol. 6, e253 (2008).\n",
      "\n",
      "Acknowledgements The authors would like to thank all Yamanaka laboratory members, past and present, and are also grateful to Y. Miyake, R. Kato, E.\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "\n",
      "to C.-F.P.), and Fundac¸ ~ao para a Ci^encia e Tecnologia (SFRH/BD/51968/2012 and PTDC/BIM-MED/0075/2014 to A.M.G. and C.-F.P.). The Knut and Alice Wallenberg Foundation is acknowledged for generous support.\n",
      "\n",
      "AUTHOR CONTRIBUTIONS\n",
      "\n",
      "C.-F.P., A.M.G, B.C., and M.D. conducted reprogramming experiments. A.M.G., C.-F.P., B.C., I.K., A.L., Z.W., L.F., A.M., and D.P. analyzed mRNA- seq and ChIP-seq datasets. C.-F.P., K.A.M., K.L., N.S., and B.K.C. performed in vivo experiments and isolated UCB cells. C.-F.P., A.M.G., K.A.M., and I.R.L. designed the experiments and wrote the paper.\n",
      "\n",
      "DECLARATION OF INTERESTS\n",
      "\n",
      "The authors declare no competing interests.\n",
      "\n",
      "Received: January 18, 2018 Revised: August 19,\n",
      "\n",
      "the other hand, the technology can also be used to our advantage by enabling the identiﬁcation of genes involved in cancers, such as hepatocellu- lar carcinoma.96 These concerns may be overcome using modiﬁed RNAs97,98 or self-replicating RNA.99 Both of these methods have been shown to efﬁciently reprogram ﬁbroblasts into iPSCs.\n",
      "\n",
      "The ﬁeld of reprogramming somatic cells into HSPCs is in its infancy. With each new study we learn more about the reprogramming process, as it appears that there are multiple ways emerging to bring the process to fruition, with the eventual cre- ation of bona ﬁde HSCs. Many obstacles remain, the foremost being the need for expanding knowledge of the mechanisms that\n",
      "\n",
      "more reprogramming factors that can ‘erase’ the ‘memories’ of somatic cells.\n",
      "\n",
      "T I M E L I N E\n",
      "\n",
      "A decade of transcription factor‑mediated reprogramming to pluripotency\n",
      "\n",
      "Kazutoshi Takahashi and Shinya Yamanaka\n",
      "\n",
      "Abstract | The past 10 years have seen great advances in our ability to manipulate cell fate, including the induction of pluripotency in vitro to generate induced pluripotent stem cells (iPSCs). This process proved to be remarkably simple from a technical perspective, only needing the host cell and a defined cocktail of transcription factors, with four factors — octamer-binding protein 3/4 (OCT3/4), SOX2, Krüppel-like factor 4 (KLF4) and MYC (collectively referred to as OSKM) —\n",
      "Human: What are the main challenges and limitations of using transcription factor-mediated reprogramming for therapeutic purposes?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "==================================================\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  Can transcription factors alone efficiently reprogram cells, or are additional factors or conditions required for successful reprogramming?\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fri Jun  7 23:23:56 2024 INFO RAG\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "RAG >> 9: \n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the first iPSCs.\n",
      "\n",
      "Direct cell conversion\n",
      "\n",
      "+\n",
      "\n",
      "MYOD\n",
      "\n",
      "Fibroblast\n",
      "\n",
      "+\n",
      "\n",
      "GATA1\n",
      "\n",
      "Myeloblast\n",
      "\n",
      "+\n",
      "\n",
      "CEBP\n",
      "\n",
      "B lymphocyte\n",
      "\n",
      "Diﬀerentiated\n",
      "\n",
      "Transdiﬀerentiation\n",
      "\n",
      "Myoblast\n",
      "\n",
      "Megakaryocyte, eosinophil, erythrocyte\n",
      "\n",
      "Macrophage\n",
      "\n",
      "Identifying candidate reprogramming factors. According to previous experiments, genes that contribute to the identities of fertilized eggs or ES cells were thought to be, in large part, responsible for the reconstitution of specific transcriptional networks during reprogramming. Thus, in 2000, we\n",
      "\n",
      "of ﬁbroblasts into functional cardiomyocytes by deﬁned factors. Cell 142, 375–386 (2010). 9. Haridhasapavalan, K. K., Raina, K., Dey, C., Adhikari, P. & Thummer,\n",
      "\n",
      "6.\n",
      "\n",
      "7.\n",
      "\n",
      "8.\n",
      "\n",
      "R. P. An insight into reprogramming barriers to iPSC generation. Stem Cell Rev. Rep. 16, 56–81 (2020).\n",
      "\n",
      "10. Takahashi, K. & Yamanaka, S. A decade of transcription factor-\n",
      "\n",
      "mediated reprogramming to pluripotency. Nat. Rev. Mol. Cell Biol. 17, 183–193 (2016).\n",
      "\n",
      "11. Brumbaugh, J., Di Stefano, B. & Hochedlinger, K. Reprogramming: identifying the mechanisms that safeguard cell identity. Develop- ment https://doi.org/10.1242/dev.182170 (2019).\n",
      "\n",
      "12. Dal-Pra, S., Hodgkinson, C. P., Mirotsou, M., Kirste, I. & Dzau, V. J.\n",
      "\n",
      "Cell Stem Cell 1, 55–70 (2007). 161. Park, I. H. et al. Reprogramming of human somatic\n",
      "\n",
      "cells to pluripotency with defined factors. Nature 451, 141–146 (2008).\n",
      "\n",
      "162. Hanna, J. et al. Direct cell reprogramming is a stochastic process amenable to acceleration. Nature 462, 595–601 (2009).\n",
      "\n",
      "163. Yang, J. et al. Stat3 activation is limiting for reprogramming to ground state pluripotency. Cell Stem Cell 7, 319–328 (2010).\n",
      "\n",
      "164. Silva, J. et al. Promotion of reprogramming to ground state pluripotency by signal inhibition. PLoS Biol. 6, e253 (2008).\n",
      "\n",
      "Acknowledgements The authors would like to thank all Yamanaka laboratory members, past and present, and are also grateful to Y. Miyake, R. Kato, E.\n",
      "\n",
      "that Drive Neuronal Fate and Reprogramming. Cell Stem . Cell ​ .\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "41. A. Ebrahimi, K. Sevinç, G. Gürhan Sevinç, A. P. Cribbs, M. Philpott, F. Uyulur, T. Morova, J. E. Dunford, S. Göklemez, Ş. Arı, U. Oppermann, T. T. Önder, Bromodomain inhibition of the coactivators CBP/EP300 facilitate cellular reprogramming. 519–528 (2019).\n",
      "\n",
      "Nat. Chem. Biol. ​\n",
      "\n",
      "15\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      ",\n",
      "\n",
      "​\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made\n",
      "\n",
      "available under a\n",
      "\n",
      "CC-BY\n",
      "\n",
      "and adult ﬁbroblast cultures by deﬁned factors. Cell 126, 663–676 (2006). 4.\n",
      "\n",
      "5. Xu, J., Du, Y. & Deng, H. Direct lineage reprogramming: strategies,\n",
      "\n",
      "mechanisms, and applications. Cell Stem Cell 16, 119–134 (2015). Takahashi, K. et al. Induction of pluripotent stem cells from adult human ﬁbroblasts by deﬁned factors. Cell 131, 861–872 (2007). Vierbuchen, T. et al. Direct conversion of ﬁbroblasts to functional neurons by deﬁned factors. Nature 463, 1035–1041 (2010). Ieda, M. et al. Direct reprogramming of ﬁbroblasts into functional cardiomyocytes by deﬁned factors. Cell 142, 375–386 (2010). 9. Haridhasapavalan, K. K., Raina, K., Dey, C., Adhikari, P. & Thummer,\n",
      "\n",
      "6.\n",
      "\n",
      "7.\n",
      "\n",
      "8.\n",
      "\n",
      "R. P. An insight\n",
      "\n",
      "adult human ﬁbroblasts by deﬁned factors. Cell 131: 861–872.\n",
      "\n",
      "10. Morris, S.A. & G.Q. Daley. 2013. A blueprint for engineer- ing cell fate: current technologies to reprogram cell identity. Cell Res. 23: 33–48.\n",
      "\n",
      "11. Kelaini, S., A. Cochrane & A. Margariti. 2014. Direct repro- gramming of adult cells: avoiding the pluripotent state. Stem Cells Cloning 7: 19–29.\n",
      "\n",
      "12. Pereira, C.F., I.R. Lemischka & K. Moore. 2012. Reprogram- ming cell fates: insights from combinatorial approaches. Ann. N. Y. Acad. Sci. 1266: 7–17.\n",
      "\n",
      "13. Orkin, S.H. 1995. Transcription factors and hematopoietic development. J. Biol. Chem. 270: 4955–4958.\n",
      "\n",
      "14. Vierbuchen, T. et al. 2010. Direct conversion of ﬁbrob- lasts to\n",
      "\n",
      "more reprogramming factors that can ‘erase’ the ‘memories’ of somatic cells.\n",
      "\n",
      "T I M E L I N E\n",
      "\n",
      "A decade of transcription factor‑mediated reprogramming to pluripotency\n",
      "\n",
      "Kazutoshi Takahashi and Shinya Yamanaka\n",
      "\n",
      "Abstract | The past 10 years have seen great advances in our ability to manipulate cell fate, including the induction of pluripotency in vitro to generate induced pluripotent stem cells (iPSCs). This process proved to be remarkably simple from a technical perspective, only needing the host cell and a defined cocktail of transcription factors, with four factors — octamer-binding protein 3/4 (OCT3/4), SOX2, Krüppel-like factor 4 (KLF4) and MYC (collectively referred to as OSKM) —\n",
      "\n",
      "themselves58,59,61–63,161,162. Furthermore, it has been clearly shown that the culture environment is important for successful reprogramming, and modification of the culture medium can increase the reprogramming efficiency. For instance the LIF–STAT3 (leukaemia inhibitor factor–signal transducer and activator of transcription 3) pathway was shown to be an active player in induced mouse pluripotency163. Treatment with inhibitors of MEK (for example, PD0325901) and glycogen synthase kinase 3 (for example, CHIR99021) — which are termed 2i — allows mouse ES cells to undergo LIF-independent self-renewal and increases the homogeneity of the cell population, thus facilitating the efficiency of\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "Human: Can transcription factors alone efficiently reprogram cells, or are additional factors or conditions required for successful reprogramming?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "of ﬁbroblasts into functional cardiomyocytes by deﬁned factors. Cell 142, 375–386 (2010). 9. Haridhasapavalan, K. K., Raina, K., Dey, C., Adhikari, P. & Thummer,\n",
      "\n",
      "6.\n",
      "\n",
      "7.\n",
      "\n",
      "8.\n",
      "\n",
      "R. P. An insight into reprogramming barriers to iPSC generation. Stem Cell Rev. Rep. 16, 56–81 (2020).\n",
      "\n",
      "10. Takahashi, K. & Yamanaka, S. A decade of transcription factor-\n",
      "\n",
      "mediated reprogramming to pluripotency. Nat. Rev. Mol. Cell Biol. 17, 183–193 (2016).\n",
      "\n",
      "11. Brumbaugh, J., Di Stefano, B. & Hochedlinger, K. Reprogramming: identifying the mechanisms that safeguard cell identity. Develop- ment https://doi.org/10.1242/dev.182170 (2019).\n",
      "\n",
      "12. Dal-Pra, S., Hodgkinson, C. P., Mirotsou, M., Kirste, I. & Dzau, V. J.\n",
      "\n",
      "Cell Stem Cell 1, 55–70 (2007). 161. Park, I. H. et al. Reprogramming of human somatic\n",
      "\n",
      "cells to pluripotency with defined factors. Nature 451, 141–146 (2008).\n",
      "\n",
      "162. Hanna, J. et al. Direct cell reprogramming is a stochastic process amenable to acceleration. Nature 462, 595–601 (2009).\n",
      "\n",
      "163. Yang, J. et al. Stat3 activation is limiting for reprogramming to ground state pluripotency. Cell Stem Cell 7, 319–328 (2010).\n",
      "\n",
      "164. Silva, J. et al. Promotion of reprogramming to ground state pluripotency by signal inhibition. PLoS Biol. 6, e253 (2008).\n",
      "\n",
      "Acknowledgements The authors would like to thank all Yamanaka laboratory members, past and present, and are also grateful to Y. Miyake, R. Kato, E.\n",
      "\n",
      "that Drive Neuronal Fate and Reprogramming. Cell Stem . Cell ​ .\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "41. A. Ebrahimi, K. Sevinç, G. Gürhan Sevinç, A. P. Cribbs, M. Philpott, F. Uyulur, T. Morova, J. E. Dunford, S. Göklemez, Ş. Arı, U. Oppermann, T. T. Önder, Bromodomain inhibition of the coactivators CBP/EP300 facilitate cellular reprogramming. 519–528 (2019).\n",
      "\n",
      "Nat. Chem. Biol. ​\n",
      "\n",
      "15\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      ",\n",
      "\n",
      "​\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made\n",
      "\n",
      "available under a\n",
      "\n",
      "CC-BY\n",
      "\n",
      "and adult ﬁbroblast cultures by deﬁned factors. Cell 126, 663–676 (2006). 4.\n",
      "\n",
      "5. Xu, J., Du, Y. & Deng, H. Direct lineage reprogramming: strategies,\n",
      "\n",
      "mechanisms, and applications. Cell Stem Cell 16, 119–134 (2015). Takahashi, K. et al. Induction of pluripotent stem cells from adult human ﬁbroblasts by deﬁned factors. Cell 131, 861–872 (2007). Vierbuchen, T. et al. Direct conversion of ﬁbroblasts to functional neurons by deﬁned factors. Nature 463, 1035–1041 (2010). Ieda, M. et al. Direct reprogramming of ﬁbroblasts into functional cardiomyocytes by deﬁned factors. Cell 142, 375–386 (2010). 9. Haridhasapavalan, K. K., Raina, K., Dey, C., Adhikari, P. & Thummer,\n",
      "\n",
      "6.\n",
      "\n",
      "7.\n",
      "\n",
      "8.\n",
      "\n",
      "R. P. An insight\n",
      "\n",
      "adult human ﬁbroblasts by deﬁned factors. Cell 131: 861–872.\n",
      "\n",
      "10. Morris, S.A. & G.Q. Daley. 2013. A blueprint for engineer- ing cell fate: current technologies to reprogram cell identity. Cell Res. 23: 33–48.\n",
      "\n",
      "11. Kelaini, S., A. Cochrane & A. Margariti. 2014. Direct repro- gramming of adult cells: avoiding the pluripotent state. Stem Cells Cloning 7: 19–29.\n",
      "\n",
      "12. Pereira, C.F., I.R. Lemischka & K. Moore. 2012. Reprogram- ming cell fates: insights from combinatorial approaches. Ann. N. Y. Acad. Sci. 1266: 7–17.\n",
      "\n",
      "13. Orkin, S.H. 1995. Transcription factors and hematopoietic development. J. Biol. Chem. 270: 4955–4958.\n",
      "\n",
      "14. Vierbuchen, T. et al. 2010. Direct conversion of ﬁbrob- lasts to\n",
      "\n",
      "more reprogramming factors that can ‘erase’ the ‘memories’ of somatic cells.\n",
      "\n",
      "T I M E L I N E\n",
      "\n",
      "A decade of transcription factor‑mediated reprogramming to pluripotency\n",
      "\n",
      "Kazutoshi Takahashi and Shinya Yamanaka\n",
      "\n",
      "Abstract | The past 10 years have seen great advances in our ability to manipulate cell fate, including the induction of pluripotency in vitro to generate induced pluripotent stem cells (iPSCs). This process proved to be remarkably simple from a technical perspective, only needing the host cell and a defined cocktail of transcription factors, with four factors — octamer-binding protein 3/4 (OCT3/4), SOX2, Krüppel-like factor 4 (KLF4) and MYC (collectively referred to as OSKM) —\n",
      "\n",
      "themselves58,59,61–63,161,162. Furthermore, it has been clearly shown that the culture environment is important for successful reprogramming, and modification of the culture medium can increase the reprogramming efficiency. For instance the LIF–STAT3 (leukaemia inhibitor factor–signal transducer and activator of transcription 3) pathway was shown to be an active player in induced mouse pluripotency163. Treatment with inhibitors of MEK (for example, PD0325901) and glycogen synthase kinase 3 (for example, CHIR99021) — which are termed 2i — allows mouse ES cells to undergo LIF-independent self-renewal and increases the homogeneity of the cell population, thus facilitating the efficiency of\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "Human: Can transcription factors alone efficiently reprogram cells, or are additional factors or conditions required for successful reprogramming?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the first iPSCs.\n",
      "\n",
      "Direct cell conversion\n",
      "\n",
      "+\n",
      "\n",
      "MYOD\n",
      "\n",
      "Fibroblast\n",
      "\n",
      "+\n",
      "\n",
      "GATA1\n",
      "\n",
      "Myeloblast\n",
      "\n",
      "+\n",
      "\n",
      "CEBP\n",
      "\n",
      "B lymphocyte\n",
      "\n",
      "Diﬀerentiated\n",
      "\n",
      "Transdiﬀerentiation\n",
      "\n",
      "Myoblast\n",
      "\n",
      "Megakaryocyte, eosinophil, erythrocyte\n",
      "\n",
      "Macrophage\n",
      "\n",
      "Identifying candidate reprogramming factors. According to previous experiments, genes that contribute to the identities of fertilized eggs or ES cells were thought to be, in large part, responsible for the reconstitution of specific transcriptional networks during reprogramming. Thus, in 2000, we\n",
      "\n",
      "Cell Stem Cell 1, 55–70 (2007). 161. Park, I. H. et al. Reprogramming of human somatic\n",
      "\n",
      "cells to pluripotency with defined factors. Nature 451, 141–146 (2008).\n",
      "\n",
      "162. Hanna, J. et al. Direct cell reprogramming is a stochastic process amenable to acceleration. Nature 462, 595–601 (2009).\n",
      "\n",
      "163. Yang, J. et al. Stat3 activation is limiting for reprogramming to ground state pluripotency. Cell Stem Cell 7, 319–328 (2010).\n",
      "\n",
      "164. Silva, J. et al. Promotion of reprogramming to ground state pluripotency by signal inhibition. PLoS Biol. 6, e253 (2008).\n",
      "\n",
      "Acknowledgements The authors would like to thank all Yamanaka laboratory members, past and present, and are also grateful to Y. Miyake, R. Kato, E.\n",
      "\n",
      "that Drive Neuronal Fate and Reprogramming. Cell Stem . Cell ​ .\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "41. A. Ebrahimi, K. Sevinç, G. Gürhan Sevinç, A. P. Cribbs, M. Philpott, F. Uyulur, T. Morova, J. E. Dunford, S. Göklemez, Ş. Arı, U. Oppermann, T. T. Önder, Bromodomain inhibition of the coactivators CBP/EP300 facilitate cellular reprogramming. 519–528 (2019).\n",
      "\n",
      "Nat. Chem. Biol. ​\n",
      "\n",
      "15\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      ",\n",
      "\n",
      "​\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made\n",
      "\n",
      "available under a\n",
      "\n",
      "CC-BY\n",
      "\n",
      "and adult ﬁbroblast cultures by deﬁned factors. Cell 126, 663–676 (2006). 4.\n",
      "\n",
      "5. Xu, J., Du, Y. & Deng, H. Direct lineage reprogramming: strategies,\n",
      "\n",
      "mechanisms, and applications. Cell Stem Cell 16, 119–134 (2015). Takahashi, K. et al. Induction of pluripotent stem cells from adult human ﬁbroblasts by deﬁned factors. Cell 131, 861–872 (2007). Vierbuchen, T. et al. Direct conversion of ﬁbroblasts to functional neurons by deﬁned factors. Nature 463, 1035–1041 (2010). Ieda, M. et al. Direct reprogramming of ﬁbroblasts into functional cardiomyocytes by deﬁned factors. Cell 142, 375–386 (2010). 9. Haridhasapavalan, K. K., Raina, K., Dey, C., Adhikari, P. & Thummer,\n",
      "\n",
      "6.\n",
      "\n",
      "7.\n",
      "\n",
      "8.\n",
      "\n",
      "R. P. An insight\n",
      "\n",
      "adult human ﬁbroblasts by deﬁned factors. Cell 131: 861–872.\n",
      "\n",
      "10. Morris, S.A. & G.Q. Daley. 2013. A blueprint for engineer- ing cell fate: current technologies to reprogram cell identity. Cell Res. 23: 33–48.\n",
      "\n",
      "11. Kelaini, S., A. Cochrane & A. Margariti. 2014. Direct repro- gramming of adult cells: avoiding the pluripotent state. Stem Cells Cloning 7: 19–29.\n",
      "\n",
      "12. Pereira, C.F., I.R. Lemischka & K. Moore. 2012. Reprogram- ming cell fates: insights from combinatorial approaches. Ann. N. Y. Acad. Sci. 1266: 7–17.\n",
      "\n",
      "13. Orkin, S.H. 1995. Transcription factors and hematopoietic development. J. Biol. Chem. 270: 4955–4958.\n",
      "\n",
      "14. Vierbuchen, T. et al. 2010. Direct conversion of ﬁbrob- lasts to\n",
      "\n",
      "more reprogramming factors that can ‘erase’ the ‘memories’ of somatic cells.\n",
      "\n",
      "T I M E L I N E\n",
      "\n",
      "A decade of transcription factor‑mediated reprogramming to pluripotency\n",
      "\n",
      "Kazutoshi Takahashi and Shinya Yamanaka\n",
      "\n",
      "Abstract | The past 10 years have seen great advances in our ability to manipulate cell fate, including the induction of pluripotency in vitro to generate induced pluripotent stem cells (iPSCs). This process proved to be remarkably simple from a technical perspective, only needing the host cell and a defined cocktail of transcription factors, with four factors — octamer-binding protein 3/4 (OCT3/4), SOX2, Krüppel-like factor 4 (KLF4) and MYC (collectively referred to as OSKM) —\n",
      "\n",
      "themselves58,59,61–63,161,162. Furthermore, it has been clearly shown that the culture environment is important for successful reprogramming, and modification of the culture medium can increase the reprogramming efficiency. For instance the LIF–STAT3 (leukaemia inhibitor factor–signal transducer and activator of transcription 3) pathway was shown to be an active player in induced mouse pluripotency163. Treatment with inhibitors of MEK (for example, PD0325901) and glycogen synthase kinase 3 (for example, CHIR99021) — which are termed 2i — allows mouse ES cells to undergo LIF-independent self-renewal and increases the homogeneity of the cell population, thus facilitating the efficiency of\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "Human: Can transcription factors alone efficiently reprogram cells, or are additional factors or conditions required for successful reprogramming?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the first iPSCs.\n",
      "\n",
      "Direct cell conversion\n",
      "\n",
      "+\n",
      "\n",
      "MYOD\n",
      "\n",
      "Fibroblast\n",
      "\n",
      "+\n",
      "\n",
      "GATA1\n",
      "\n",
      "Myeloblast\n",
      "\n",
      "+\n",
      "\n",
      "CEBP\n",
      "\n",
      "B lymphocyte\n",
      "\n",
      "Diﬀerentiated\n",
      "\n",
      "Transdiﬀerentiation\n",
      "\n",
      "Myoblast\n",
      "\n",
      "Megakaryocyte, eosinophil, erythrocyte\n",
      "\n",
      "Macrophage\n",
      "\n",
      "Identifying candidate reprogramming factors. According to previous experiments, genes that contribute to the identities of fertilized eggs or ES cells were thought to be, in large part, responsible for the reconstitution of specific transcriptional networks during reprogramming. Thus, in 2000, we\n",
      "\n",
      "of ﬁbroblasts into functional cardiomyocytes by deﬁned factors. Cell 142, 375–386 (2010). 9. Haridhasapavalan, K. K., Raina, K., Dey, C., Adhikari, P. & Thummer,\n",
      "\n",
      "6.\n",
      "\n",
      "7.\n",
      "\n",
      "8.\n",
      "\n",
      "R. P. An insight into reprogramming barriers to iPSC generation. Stem Cell Rev. Rep. 16, 56–81 (2020).\n",
      "\n",
      "10. Takahashi, K. & Yamanaka, S. A decade of transcription factor-\n",
      "\n",
      "mediated reprogramming to pluripotency. Nat. Rev. Mol. Cell Biol. 17, 183–193 (2016).\n",
      "\n",
      "11. Brumbaugh, J., Di Stefano, B. & Hochedlinger, K. Reprogramming: identifying the mechanisms that safeguard cell identity. Develop- ment https://doi.org/10.1242/dev.182170 (2019).\n",
      "\n",
      "12. Dal-Pra, S., Hodgkinson, C. P., Mirotsou, M., Kirste, I. & Dzau, V. J.\n",
      "\n",
      "that Drive Neuronal Fate and Reprogramming. Cell Stem . Cell ​ .\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "41. A. Ebrahimi, K. Sevinç, G. Gürhan Sevinç, A. P. Cribbs, M. Philpott, F. Uyulur, T. Morova, J. E. Dunford, S. Göklemez, Ş. Arı, U. Oppermann, T. T. Önder, Bromodomain inhibition of the coactivators CBP/EP300 facilitate cellular reprogramming. 519–528 (2019).\n",
      "\n",
      "Nat. Chem. Biol. ​\n",
      "\n",
      "15\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      ",\n",
      "\n",
      "​\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made\n",
      "\n",
      "available under a\n",
      "\n",
      "CC-BY\n",
      "\n",
      "and adult ﬁbroblast cultures by deﬁned factors. Cell 126, 663–676 (2006). 4.\n",
      "\n",
      "5. Xu, J., Du, Y. & Deng, H. Direct lineage reprogramming: strategies,\n",
      "\n",
      "mechanisms, and applications. Cell Stem Cell 16, 119–134 (2015). Takahashi, K. et al. Induction of pluripotent stem cells from adult human ﬁbroblasts by deﬁned factors. Cell 131, 861–872 (2007). Vierbuchen, T. et al. Direct conversion of ﬁbroblasts to functional neurons by deﬁned factors. Nature 463, 1035–1041 (2010). Ieda, M. et al. Direct reprogramming of ﬁbroblasts into functional cardiomyocytes by deﬁned factors. Cell 142, 375–386 (2010). 9. Haridhasapavalan, K. K., Raina, K., Dey, C., Adhikari, P. & Thummer,\n",
      "\n",
      "6.\n",
      "\n",
      "7.\n",
      "\n",
      "8.\n",
      "\n",
      "R. P. An insight\n",
      "\n",
      "adult human ﬁbroblasts by deﬁned factors. Cell 131: 861–872.\n",
      "\n",
      "10. Morris, S.A. & G.Q. Daley. 2013. A blueprint for engineer- ing cell fate: current technologies to reprogram cell identity. Cell Res. 23: 33–48.\n",
      "\n",
      "11. Kelaini, S., A. Cochrane & A. Margariti. 2014. Direct repro- gramming of adult cells: avoiding the pluripotent state. Stem Cells Cloning 7: 19–29.\n",
      "\n",
      "12. Pereira, C.F., I.R. Lemischka & K. Moore. 2012. Reprogram- ming cell fates: insights from combinatorial approaches. Ann. N. Y. Acad. Sci. 1266: 7–17.\n",
      "\n",
      "13. Orkin, S.H. 1995. Transcription factors and hematopoietic development. J. Biol. Chem. 270: 4955–4958.\n",
      "\n",
      "14. Vierbuchen, T. et al. 2010. Direct conversion of ﬁbrob- lasts to\n",
      "\n",
      "more reprogramming factors that can ‘erase’ the ‘memories’ of somatic cells.\n",
      "\n",
      "T I M E L I N E\n",
      "\n",
      "A decade of transcription factor‑mediated reprogramming to pluripotency\n",
      "\n",
      "Kazutoshi Takahashi and Shinya Yamanaka\n",
      "\n",
      "Abstract | The past 10 years have seen great advances in our ability to manipulate cell fate, including the induction of pluripotency in vitro to generate induced pluripotent stem cells (iPSCs). This process proved to be remarkably simple from a technical perspective, only needing the host cell and a defined cocktail of transcription factors, with four factors — octamer-binding protein 3/4 (OCT3/4), SOX2, Krüppel-like factor 4 (KLF4) and MYC (collectively referred to as OSKM) —\n",
      "\n",
      "themselves58,59,61–63,161,162. Furthermore, it has been clearly shown that the culture environment is important for successful reprogramming, and modification of the culture medium can increase the reprogramming efficiency. For instance the LIF–STAT3 (leukaemia inhibitor factor–signal transducer and activator of transcription 3) pathway was shown to be an active player in induced mouse pluripotency163. Treatment with inhibitors of MEK (for example, PD0325901) and glycogen synthase kinase 3 (for example, CHIR99021) — which are termed 2i — allows mouse ES cells to undergo LIF-independent self-renewal and increases the homogeneity of the cell population, thus facilitating the efficiency of\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "Human: Can transcription factors alone efficiently reprogram cells, or are additional factors or conditions required for successful reprogramming?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the first iPSCs.\n",
      "\n",
      "Direct cell conversion\n",
      "\n",
      "+\n",
      "\n",
      "MYOD\n",
      "\n",
      "Fibroblast\n",
      "\n",
      "+\n",
      "\n",
      "GATA1\n",
      "\n",
      "Myeloblast\n",
      "\n",
      "+\n",
      "\n",
      "CEBP\n",
      "\n",
      "B lymphocyte\n",
      "\n",
      "Diﬀerentiated\n",
      "\n",
      "Transdiﬀerentiation\n",
      "\n",
      "Myoblast\n",
      "\n",
      "Megakaryocyte, eosinophil, erythrocyte\n",
      "\n",
      "Macrophage\n",
      "\n",
      "Identifying candidate reprogramming factors. According to previous experiments, genes that contribute to the identities of fertilized eggs or ES cells were thought to be, in large part, responsible for the reconstitution of specific transcriptional networks during reprogramming. Thus, in 2000, we\n",
      "\n",
      "of ﬁbroblasts into functional cardiomyocytes by deﬁned factors. Cell 142, 375–386 (2010). 9. Haridhasapavalan, K. K., Raina, K., Dey, C., Adhikari, P. & Thummer,\n",
      "\n",
      "6.\n",
      "\n",
      "7.\n",
      "\n",
      "8.\n",
      "\n",
      "R. P. An insight into reprogramming barriers to iPSC generation. Stem Cell Rev. Rep. 16, 56–81 (2020).\n",
      "\n",
      "10. Takahashi, K. & Yamanaka, S. A decade of transcription factor-\n",
      "\n",
      "mediated reprogramming to pluripotency. Nat. Rev. Mol. Cell Biol. 17, 183–193 (2016).\n",
      "\n",
      "11. Brumbaugh, J., Di Stefano, B. & Hochedlinger, K. Reprogramming: identifying the mechanisms that safeguard cell identity. Develop- ment https://doi.org/10.1242/dev.182170 (2019).\n",
      "\n",
      "12. Dal-Pra, S., Hodgkinson, C. P., Mirotsou, M., Kirste, I. & Dzau, V. J.\n",
      "\n",
      "Cell Stem Cell 1, 55–70 (2007). 161. Park, I. H. et al. Reprogramming of human somatic\n",
      "\n",
      "cells to pluripotency with defined factors. Nature 451, 141–146 (2008).\n",
      "\n",
      "162. Hanna, J. et al. Direct cell reprogramming is a stochastic process amenable to acceleration. Nature 462, 595–601 (2009).\n",
      "\n",
      "163. Yang, J. et al. Stat3 activation is limiting for reprogramming to ground state pluripotency. Cell Stem Cell 7, 319–328 (2010).\n",
      "\n",
      "164. Silva, J. et al. Promotion of reprogramming to ground state pluripotency by signal inhibition. PLoS Biol. 6, e253 (2008).\n",
      "\n",
      "Acknowledgements The authors would like to thank all Yamanaka laboratory members, past and present, and are also grateful to Y. Miyake, R. Kato, E.\n",
      "\n",
      "and adult ﬁbroblast cultures by deﬁned factors. Cell 126, 663–676 (2006). 4.\n",
      "\n",
      "5. Xu, J., Du, Y. & Deng, H. Direct lineage reprogramming: strategies,\n",
      "\n",
      "mechanisms, and applications. Cell Stem Cell 16, 119–134 (2015). Takahashi, K. et al. Induction of pluripotent stem cells from adult human ﬁbroblasts by deﬁned factors. Cell 131, 861–872 (2007). Vierbuchen, T. et al. Direct conversion of ﬁbroblasts to functional neurons by deﬁned factors. Nature 463, 1035–1041 (2010). Ieda, M. et al. Direct reprogramming of ﬁbroblasts into functional cardiomyocytes by deﬁned factors. Cell 142, 375–386 (2010). 9. Haridhasapavalan, K. K., Raina, K., Dey, C., Adhikari, P. & Thummer,\n",
      "\n",
      "6.\n",
      "\n",
      "7.\n",
      "\n",
      "8.\n",
      "\n",
      "R. P. An insight\n",
      "\n",
      "adult human ﬁbroblasts by deﬁned factors. Cell 131: 861–872.\n",
      "\n",
      "10. Morris, S.A. & G.Q. Daley. 2013. A blueprint for engineer- ing cell fate: current technologies to reprogram cell identity. Cell Res. 23: 33–48.\n",
      "\n",
      "11. Kelaini, S., A. Cochrane & A. Margariti. 2014. Direct repro- gramming of adult cells: avoiding the pluripotent state. Stem Cells Cloning 7: 19–29.\n",
      "\n",
      "12. Pereira, C.F., I.R. Lemischka & K. Moore. 2012. Reprogram- ming cell fates: insights from combinatorial approaches. Ann. N. Y. Acad. Sci. 1266: 7–17.\n",
      "\n",
      "13. Orkin, S.H. 1995. Transcription factors and hematopoietic development. J. Biol. Chem. 270: 4955–4958.\n",
      "\n",
      "14. Vierbuchen, T. et al. 2010. Direct conversion of ﬁbrob- lasts to\n",
      "\n",
      "more reprogramming factors that can ‘erase’ the ‘memories’ of somatic cells.\n",
      "\n",
      "T I M E L I N E\n",
      "\n",
      "A decade of transcription factor‑mediated reprogramming to pluripotency\n",
      "\n",
      "Kazutoshi Takahashi and Shinya Yamanaka\n",
      "\n",
      "Abstract | The past 10 years have seen great advances in our ability to manipulate cell fate, including the induction of pluripotency in vitro to generate induced pluripotent stem cells (iPSCs). This process proved to be remarkably simple from a technical perspective, only needing the host cell and a defined cocktail of transcription factors, with four factors — octamer-binding protein 3/4 (OCT3/4), SOX2, Krüppel-like factor 4 (KLF4) and MYC (collectively referred to as OSKM) —\n",
      "\n",
      "themselves58,59,61–63,161,162. Furthermore, it has been clearly shown that the culture environment is important for successful reprogramming, and modification of the culture medium can increase the reprogramming efficiency. For instance the LIF–STAT3 (leukaemia inhibitor factor–signal transducer and activator of transcription 3) pathway was shown to be an active player in induced mouse pluripotency163. Treatment with inhibitors of MEK (for example, PD0325901) and glycogen synthase kinase 3 (for example, CHIR99021) — which are termed 2i — allows mouse ES cells to undergo LIF-independent self-renewal and increases the homogeneity of the cell population, thus facilitating the efficiency of\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "Human: Can transcription factors alone efficiently reprogram cells, or are additional factors or conditions required for successful reprogramming?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the first iPSCs.\n",
      "\n",
      "Direct cell conversion\n",
      "\n",
      "+\n",
      "\n",
      "MYOD\n",
      "\n",
      "Fibroblast\n",
      "\n",
      "+\n",
      "\n",
      "GATA1\n",
      "\n",
      "Myeloblast\n",
      "\n",
      "+\n",
      "\n",
      "CEBP\n",
      "\n",
      "B lymphocyte\n",
      "\n",
      "Diﬀerentiated\n",
      "\n",
      "Transdiﬀerentiation\n",
      "\n",
      "Myoblast\n",
      "\n",
      "Megakaryocyte, eosinophil, erythrocyte\n",
      "\n",
      "Macrophage\n",
      "\n",
      "Identifying candidate reprogramming factors. According to previous experiments, genes that contribute to the identities of fertilized eggs or ES cells were thought to be, in large part, responsible for the reconstitution of specific transcriptional networks during reprogramming. Thus, in 2000, we\n",
      "\n",
      "of ﬁbroblasts into functional cardiomyocytes by deﬁned factors. Cell 142, 375–386 (2010). 9. Haridhasapavalan, K. K., Raina, K., Dey, C., Adhikari, P. & Thummer,\n",
      "\n",
      "6.\n",
      "\n",
      "7.\n",
      "\n",
      "8.\n",
      "\n",
      "R. P. An insight into reprogramming barriers to iPSC generation. Stem Cell Rev. Rep. 16, 56–81 (2020).\n",
      "\n",
      "10. Takahashi, K. & Yamanaka, S. A decade of transcription factor-\n",
      "\n",
      "mediated reprogramming to pluripotency. Nat. Rev. Mol. Cell Biol. 17, 183–193 (2016).\n",
      "\n",
      "11. Brumbaugh, J., Di Stefano, B. & Hochedlinger, K. Reprogramming: identifying the mechanisms that safeguard cell identity. Develop- ment https://doi.org/10.1242/dev.182170 (2019).\n",
      "\n",
      "12. Dal-Pra, S., Hodgkinson, C. P., Mirotsou, M., Kirste, I. & Dzau, V. J.\n",
      "\n",
      "Cell Stem Cell 1, 55–70 (2007). 161. Park, I. H. et al. Reprogramming of human somatic\n",
      "\n",
      "cells to pluripotency with defined factors. Nature 451, 141–146 (2008).\n",
      "\n",
      "162. Hanna, J. et al. Direct cell reprogramming is a stochastic process amenable to acceleration. Nature 462, 595–601 (2009).\n",
      "\n",
      "163. Yang, J. et al. Stat3 activation is limiting for reprogramming to ground state pluripotency. Cell Stem Cell 7, 319–328 (2010).\n",
      "\n",
      "164. Silva, J. et al. Promotion of reprogramming to ground state pluripotency by signal inhibition. PLoS Biol. 6, e253 (2008).\n",
      "\n",
      "Acknowledgements The authors would like to thank all Yamanaka laboratory members, past and present, and are also grateful to Y. Miyake, R. Kato, E.\n",
      "\n",
      "that Drive Neuronal Fate and Reprogramming. Cell Stem . Cell ​ .\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "41. A. Ebrahimi, K. Sevinç, G. Gürhan Sevinç, A. P. Cribbs, M. Philpott, F. Uyulur, T. Morova, J. E. Dunford, S. Göklemez, Ş. Arı, U. Oppermann, T. T. Önder, Bromodomain inhibition of the coactivators CBP/EP300 facilitate cellular reprogramming. 519–528 (2019).\n",
      "\n",
      "Nat. Chem. Biol. ​\n",
      "\n",
      "15\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      ",\n",
      "\n",
      "​\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made\n",
      "\n",
      "available under a\n",
      "\n",
      "CC-BY\n",
      "\n",
      "adult human ﬁbroblasts by deﬁned factors. Cell 131: 861–872.\n",
      "\n",
      "10. Morris, S.A. & G.Q. Daley. 2013. A blueprint for engineer- ing cell fate: current technologies to reprogram cell identity. Cell Res. 23: 33–48.\n",
      "\n",
      "11. Kelaini, S., A. Cochrane & A. Margariti. 2014. Direct repro- gramming of adult cells: avoiding the pluripotent state. Stem Cells Cloning 7: 19–29.\n",
      "\n",
      "12. Pereira, C.F., I.R. Lemischka & K. Moore. 2012. Reprogram- ming cell fates: insights from combinatorial approaches. Ann. N. Y. Acad. Sci. 1266: 7–17.\n",
      "\n",
      "13. Orkin, S.H. 1995. Transcription factors and hematopoietic development. J. Biol. Chem. 270: 4955–4958.\n",
      "\n",
      "14. Vierbuchen, T. et al. 2010. Direct conversion of ﬁbrob- lasts to\n",
      "\n",
      "more reprogramming factors that can ‘erase’ the ‘memories’ of somatic cells.\n",
      "\n",
      "T I M E L I N E\n",
      "\n",
      "A decade of transcription factor‑mediated reprogramming to pluripotency\n",
      "\n",
      "Kazutoshi Takahashi and Shinya Yamanaka\n",
      "\n",
      "Abstract | The past 10 years have seen great advances in our ability to manipulate cell fate, including the induction of pluripotency in vitro to generate induced pluripotent stem cells (iPSCs). This process proved to be remarkably simple from a technical perspective, only needing the host cell and a defined cocktail of transcription factors, with four factors — octamer-binding protein 3/4 (OCT3/4), SOX2, Krüppel-like factor 4 (KLF4) and MYC (collectively referred to as OSKM) —\n",
      "\n",
      "themselves58,59,61–63,161,162. Furthermore, it has been clearly shown that the culture environment is important for successful reprogramming, and modification of the culture medium can increase the reprogramming efficiency. For instance the LIF–STAT3 (leukaemia inhibitor factor–signal transducer and activator of transcription 3) pathway was shown to be an active player in induced mouse pluripotency163. Treatment with inhibitors of MEK (for example, PD0325901) and glycogen synthase kinase 3 (for example, CHIR99021) — which are termed 2i — allows mouse ES cells to undergo LIF-independent self-renewal and increases the homogeneity of the cell population, thus facilitating the efficiency of\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "Human: Can transcription factors alone efficiently reprogram cells, or are additional factors or conditions required for successful reprogramming?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the first iPSCs.\n",
      "\n",
      "Direct cell conversion\n",
      "\n",
      "+\n",
      "\n",
      "MYOD\n",
      "\n",
      "Fibroblast\n",
      "\n",
      "+\n",
      "\n",
      "GATA1\n",
      "\n",
      "Myeloblast\n",
      "\n",
      "+\n",
      "\n",
      "CEBP\n",
      "\n",
      "B lymphocyte\n",
      "\n",
      "Diﬀerentiated\n",
      "\n",
      "Transdiﬀerentiation\n",
      "\n",
      "Myoblast\n",
      "\n",
      "Megakaryocyte, eosinophil, erythrocyte\n",
      "\n",
      "Macrophage\n",
      "\n",
      "Identifying candidate reprogramming factors. According to previous experiments, genes that contribute to the identities of fertilized eggs or ES cells were thought to be, in large part, responsible for the reconstitution of specific transcriptional networks during reprogramming. Thus, in 2000, we\n",
      "\n",
      "of ﬁbroblasts into functional cardiomyocytes by deﬁned factors. Cell 142, 375–386 (2010). 9. Haridhasapavalan, K. K., Raina, K., Dey, C., Adhikari, P. & Thummer,\n",
      "\n",
      "6.\n",
      "\n",
      "7.\n",
      "\n",
      "8.\n",
      "\n",
      "R. P. An insight into reprogramming barriers to iPSC generation. Stem Cell Rev. Rep. 16, 56–81 (2020).\n",
      "\n",
      "10. Takahashi, K. & Yamanaka, S. A decade of transcription factor-\n",
      "\n",
      "mediated reprogramming to pluripotency. Nat. Rev. Mol. Cell Biol. 17, 183–193 (2016).\n",
      "\n",
      "11. Brumbaugh, J., Di Stefano, B. & Hochedlinger, K. Reprogramming: identifying the mechanisms that safeguard cell identity. Develop- ment https://doi.org/10.1242/dev.182170 (2019).\n",
      "\n",
      "12. Dal-Pra, S., Hodgkinson, C. P., Mirotsou, M., Kirste, I. & Dzau, V. J.\n",
      "\n",
      "Cell Stem Cell 1, 55–70 (2007). 161. Park, I. H. et al. Reprogramming of human somatic\n",
      "\n",
      "cells to pluripotency with defined factors. Nature 451, 141–146 (2008).\n",
      "\n",
      "162. Hanna, J. et al. Direct cell reprogramming is a stochastic process amenable to acceleration. Nature 462, 595–601 (2009).\n",
      "\n",
      "163. Yang, J. et al. Stat3 activation is limiting for reprogramming to ground state pluripotency. Cell Stem Cell 7, 319–328 (2010).\n",
      "\n",
      "164. Silva, J. et al. Promotion of reprogramming to ground state pluripotency by signal inhibition. PLoS Biol. 6, e253 (2008).\n",
      "\n",
      "Acknowledgements The authors would like to thank all Yamanaka laboratory members, past and present, and are also grateful to Y. Miyake, R. Kato, E.\n",
      "\n",
      "that Drive Neuronal Fate and Reprogramming. Cell Stem . Cell ​ .\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "41. A. Ebrahimi, K. Sevinç, G. Gürhan Sevinç, A. P. Cribbs, M. Philpott, F. Uyulur, T. Morova, J. E. Dunford, S. Göklemez, Ş. Arı, U. Oppermann, T. T. Önder, Bromodomain inhibition of the coactivators CBP/EP300 facilitate cellular reprogramming. 519–528 (2019).\n",
      "\n",
      "Nat. Chem. Biol. ​\n",
      "\n",
      "15\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      ",\n",
      "\n",
      "​\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made\n",
      "\n",
      "available under a\n",
      "\n",
      "CC-BY\n",
      "\n",
      "and adult ﬁbroblast cultures by deﬁned factors. Cell 126, 663–676 (2006). 4.\n",
      "\n",
      "5. Xu, J., Du, Y. & Deng, H. Direct lineage reprogramming: strategies,\n",
      "\n",
      "mechanisms, and applications. Cell Stem Cell 16, 119–134 (2015). Takahashi, K. et al. Induction of pluripotent stem cells from adult human ﬁbroblasts by deﬁned factors. Cell 131, 861–872 (2007). Vierbuchen, T. et al. Direct conversion of ﬁbroblasts to functional neurons by deﬁned factors. Nature 463, 1035–1041 (2010). Ieda, M. et al. Direct reprogramming of ﬁbroblasts into functional cardiomyocytes by deﬁned factors. Cell 142, 375–386 (2010). 9. Haridhasapavalan, K. K., Raina, K., Dey, C., Adhikari, P. & Thummer,\n",
      "\n",
      "6.\n",
      "\n",
      "7.\n",
      "\n",
      "8.\n",
      "\n",
      "R. P. An insight\n",
      "\n",
      "more reprogramming factors that can ‘erase’ the ‘memories’ of somatic cells.\n",
      "\n",
      "T I M E L I N E\n",
      "\n",
      "A decade of transcription factor‑mediated reprogramming to pluripotency\n",
      "\n",
      "Kazutoshi Takahashi and Shinya Yamanaka\n",
      "\n",
      "Abstract | The past 10 years have seen great advances in our ability to manipulate cell fate, including the induction of pluripotency in vitro to generate induced pluripotent stem cells (iPSCs). This process proved to be remarkably simple from a technical perspective, only needing the host cell and a defined cocktail of transcription factors, with four factors — octamer-binding protein 3/4 (OCT3/4), SOX2, Krüppel-like factor 4 (KLF4) and MYC (collectively referred to as OSKM) —\n",
      "\n",
      "themselves58,59,61–63,161,162. Furthermore, it has been clearly shown that the culture environment is important for successful reprogramming, and modification of the culture medium can increase the reprogramming efficiency. For instance the LIF–STAT3 (leukaemia inhibitor factor–signal transducer and activator of transcription 3) pathway was shown to be an active player in induced mouse pluripotency163. Treatment with inhibitors of MEK (for example, PD0325901) and glycogen synthase kinase 3 (for example, CHIR99021) — which are termed 2i — allows mouse ES cells to undergo LIF-independent self-renewal and increases the homogeneity of the cell population, thus facilitating the efficiency of\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "Human: Can transcription factors alone efficiently reprogram cells, or are additional factors or conditions required for successful reprogramming?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the first iPSCs.\n",
      "\n",
      "Direct cell conversion\n",
      "\n",
      "+\n",
      "\n",
      "MYOD\n",
      "\n",
      "Fibroblast\n",
      "\n",
      "+\n",
      "\n",
      "GATA1\n",
      "\n",
      "Myeloblast\n",
      "\n",
      "+\n",
      "\n",
      "CEBP\n",
      "\n",
      "B lymphocyte\n",
      "\n",
      "Diﬀerentiated\n",
      "\n",
      "Transdiﬀerentiation\n",
      "\n",
      "Myoblast\n",
      "\n",
      "Megakaryocyte, eosinophil, erythrocyte\n",
      "\n",
      "Macrophage\n",
      "\n",
      "Identifying candidate reprogramming factors. According to previous experiments, genes that contribute to the identities of fertilized eggs or ES cells were thought to be, in large part, responsible for the reconstitution of specific transcriptional networks during reprogramming. Thus, in 2000, we\n",
      "\n",
      "of ﬁbroblasts into functional cardiomyocytes by deﬁned factors. Cell 142, 375–386 (2010). 9. Haridhasapavalan, K. K., Raina, K., Dey, C., Adhikari, P. & Thummer,\n",
      "\n",
      "6.\n",
      "\n",
      "7.\n",
      "\n",
      "8.\n",
      "\n",
      "R. P. An insight into reprogramming barriers to iPSC generation. Stem Cell Rev. Rep. 16, 56–81 (2020).\n",
      "\n",
      "10. Takahashi, K. & Yamanaka, S. A decade of transcription factor-\n",
      "\n",
      "mediated reprogramming to pluripotency. Nat. Rev. Mol. Cell Biol. 17, 183–193 (2016).\n",
      "\n",
      "11. Brumbaugh, J., Di Stefano, B. & Hochedlinger, K. Reprogramming: identifying the mechanisms that safeguard cell identity. Develop- ment https://doi.org/10.1242/dev.182170 (2019).\n",
      "\n",
      "12. Dal-Pra, S., Hodgkinson, C. P., Mirotsou, M., Kirste, I. & Dzau, V. J.\n",
      "\n",
      "Cell Stem Cell 1, 55–70 (2007). 161. Park, I. H. et al. Reprogramming of human somatic\n",
      "\n",
      "cells to pluripotency with defined factors. Nature 451, 141–146 (2008).\n",
      "\n",
      "162. Hanna, J. et al. Direct cell reprogramming is a stochastic process amenable to acceleration. Nature 462, 595–601 (2009).\n",
      "\n",
      "163. Yang, J. et al. Stat3 activation is limiting for reprogramming to ground state pluripotency. Cell Stem Cell 7, 319–328 (2010).\n",
      "\n",
      "164. Silva, J. et al. Promotion of reprogramming to ground state pluripotency by signal inhibition. PLoS Biol. 6, e253 (2008).\n",
      "\n",
      "Acknowledgements The authors would like to thank all Yamanaka laboratory members, past and present, and are also grateful to Y. Miyake, R. Kato, E.\n",
      "\n",
      "that Drive Neuronal Fate and Reprogramming. Cell Stem . Cell ​ .\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "41. A. Ebrahimi, K. Sevinç, G. Gürhan Sevinç, A. P. Cribbs, M. Philpott, F. Uyulur, T. Morova, J. E. Dunford, S. Göklemez, Ş. Arı, U. Oppermann, T. T. Önder, Bromodomain inhibition of the coactivators CBP/EP300 facilitate cellular reprogramming. 519–528 (2019).\n",
      "\n",
      "Nat. Chem. Biol. ​\n",
      "\n",
      "15\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      ",\n",
      "\n",
      "​\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made\n",
      "\n",
      "available under a\n",
      "\n",
      "CC-BY\n",
      "\n",
      "and adult ﬁbroblast cultures by deﬁned factors. Cell 126, 663–676 (2006). 4.\n",
      "\n",
      "5. Xu, J., Du, Y. & Deng, H. Direct lineage reprogramming: strategies,\n",
      "\n",
      "mechanisms, and applications. Cell Stem Cell 16, 119–134 (2015). Takahashi, K. et al. Induction of pluripotent stem cells from adult human ﬁbroblasts by deﬁned factors. Cell 131, 861–872 (2007). Vierbuchen, T. et al. Direct conversion of ﬁbroblasts to functional neurons by deﬁned factors. Nature 463, 1035–1041 (2010). Ieda, M. et al. Direct reprogramming of ﬁbroblasts into functional cardiomyocytes by deﬁned factors. Cell 142, 375–386 (2010). 9. Haridhasapavalan, K. K., Raina, K., Dey, C., Adhikari, P. & Thummer,\n",
      "\n",
      "6.\n",
      "\n",
      "7.\n",
      "\n",
      "8.\n",
      "\n",
      "R. P. An insight\n",
      "\n",
      "adult human ﬁbroblasts by deﬁned factors. Cell 131: 861–872.\n",
      "\n",
      "10. Morris, S.A. & G.Q. Daley. 2013. A blueprint for engineer- ing cell fate: current technologies to reprogram cell identity. Cell Res. 23: 33–48.\n",
      "\n",
      "11. Kelaini, S., A. Cochrane & A. Margariti. 2014. Direct repro- gramming of adult cells: avoiding the pluripotent state. Stem Cells Cloning 7: 19–29.\n",
      "\n",
      "12. Pereira, C.F., I.R. Lemischka & K. Moore. 2012. Reprogram- ming cell fates: insights from combinatorial approaches. Ann. N. Y. Acad. Sci. 1266: 7–17.\n",
      "\n",
      "13. Orkin, S.H. 1995. Transcription factors and hematopoietic development. J. Biol. Chem. 270: 4955–4958.\n",
      "\n",
      "14. Vierbuchen, T. et al. 2010. Direct conversion of ﬁbrob- lasts to\n",
      "\n",
      "themselves58,59,61–63,161,162. Furthermore, it has been clearly shown that the culture environment is important for successful reprogramming, and modification of the culture medium can increase the reprogramming efficiency. For instance the LIF–STAT3 (leukaemia inhibitor factor–signal transducer and activator of transcription 3) pathway was shown to be an active player in induced mouse pluripotency163. Treatment with inhibitors of MEK (for example, PD0325901) and glycogen synthase kinase 3 (for example, CHIR99021) — which are termed 2i — allows mouse ES cells to undergo LIF-independent self-renewal and increases the homogeneity of the cell population, thus facilitating the efficiency of\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "Human: Can transcription factors alone efficiently reprogram cells, or are additional factors or conditions required for successful reprogramming?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the first iPSCs.\n",
      "\n",
      "Direct cell conversion\n",
      "\n",
      "+\n",
      "\n",
      "MYOD\n",
      "\n",
      "Fibroblast\n",
      "\n",
      "+\n",
      "\n",
      "GATA1\n",
      "\n",
      "Myeloblast\n",
      "\n",
      "+\n",
      "\n",
      "CEBP\n",
      "\n",
      "B lymphocyte\n",
      "\n",
      "Diﬀerentiated\n",
      "\n",
      "Transdiﬀerentiation\n",
      "\n",
      "Myoblast\n",
      "\n",
      "Megakaryocyte, eosinophil, erythrocyte\n",
      "\n",
      "Macrophage\n",
      "\n",
      "Identifying candidate reprogramming factors. According to previous experiments, genes that contribute to the identities of fertilized eggs or ES cells were thought to be, in large part, responsible for the reconstitution of specific transcriptional networks during reprogramming. Thus, in 2000, we\n",
      "\n",
      "of ﬁbroblasts into functional cardiomyocytes by deﬁned factors. Cell 142, 375–386 (2010). 9. Haridhasapavalan, K. K., Raina, K., Dey, C., Adhikari, P. & Thummer,\n",
      "\n",
      "6.\n",
      "\n",
      "7.\n",
      "\n",
      "8.\n",
      "\n",
      "R. P. An insight into reprogramming barriers to iPSC generation. Stem Cell Rev. Rep. 16, 56–81 (2020).\n",
      "\n",
      "10. Takahashi, K. & Yamanaka, S. A decade of transcription factor-\n",
      "\n",
      "mediated reprogramming to pluripotency. Nat. Rev. Mol. Cell Biol. 17, 183–193 (2016).\n",
      "\n",
      "11. Brumbaugh, J., Di Stefano, B. & Hochedlinger, K. Reprogramming: identifying the mechanisms that safeguard cell identity. Develop- ment https://doi.org/10.1242/dev.182170 (2019).\n",
      "\n",
      "12. Dal-Pra, S., Hodgkinson, C. P., Mirotsou, M., Kirste, I. & Dzau, V. J.\n",
      "\n",
      "Cell Stem Cell 1, 55–70 (2007). 161. Park, I. H. et al. Reprogramming of human somatic\n",
      "\n",
      "cells to pluripotency with defined factors. Nature 451, 141–146 (2008).\n",
      "\n",
      "162. Hanna, J. et al. Direct cell reprogramming is a stochastic process amenable to acceleration. Nature 462, 595–601 (2009).\n",
      "\n",
      "163. Yang, J. et al. Stat3 activation is limiting for reprogramming to ground state pluripotency. Cell Stem Cell 7, 319–328 (2010).\n",
      "\n",
      "164. Silva, J. et al. Promotion of reprogramming to ground state pluripotency by signal inhibition. PLoS Biol. 6, e253 (2008).\n",
      "\n",
      "Acknowledgements The authors would like to thank all Yamanaka laboratory members, past and present, and are also grateful to Y. Miyake, R. Kato, E.\n",
      "\n",
      "that Drive Neuronal Fate and Reprogramming. Cell Stem . Cell ​ .\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "41. A. Ebrahimi, K. Sevinç, G. Gürhan Sevinç, A. P. Cribbs, M. Philpott, F. Uyulur, T. Morova, J. E. Dunford, S. Göklemez, Ş. Arı, U. Oppermann, T. T. Önder, Bromodomain inhibition of the coactivators CBP/EP300 facilitate cellular reprogramming. 519–528 (2019).\n",
      "\n",
      "Nat. Chem. Biol. ​\n",
      "\n",
      "15\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      ",\n",
      "\n",
      "​\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made\n",
      "\n",
      "available under a\n",
      "\n",
      "CC-BY\n",
      "\n",
      "and adult ﬁbroblast cultures by deﬁned factors. Cell 126, 663–676 (2006). 4.\n",
      "\n",
      "5. Xu, J., Du, Y. & Deng, H. Direct lineage reprogramming: strategies,\n",
      "\n",
      "mechanisms, and applications. Cell Stem Cell 16, 119–134 (2015). Takahashi, K. et al. Induction of pluripotent stem cells from adult human ﬁbroblasts by deﬁned factors. Cell 131, 861–872 (2007). Vierbuchen, T. et al. Direct conversion of ﬁbroblasts to functional neurons by deﬁned factors. Nature 463, 1035–1041 (2010). Ieda, M. et al. Direct reprogramming of ﬁbroblasts into functional cardiomyocytes by deﬁned factors. Cell 142, 375–386 (2010). 9. Haridhasapavalan, K. K., Raina, K., Dey, C., Adhikari, P. & Thummer,\n",
      "\n",
      "6.\n",
      "\n",
      "7.\n",
      "\n",
      "8.\n",
      "\n",
      "R. P. An insight\n",
      "\n",
      "adult human ﬁbroblasts by deﬁned factors. Cell 131: 861–872.\n",
      "\n",
      "10. Morris, S.A. & G.Q. Daley. 2013. A blueprint for engineer- ing cell fate: current technologies to reprogram cell identity. Cell Res. 23: 33–48.\n",
      "\n",
      "11. Kelaini, S., A. Cochrane & A. Margariti. 2014. Direct repro- gramming of adult cells: avoiding the pluripotent state. Stem Cells Cloning 7: 19–29.\n",
      "\n",
      "12. Pereira, C.F., I.R. Lemischka & K. Moore. 2012. Reprogram- ming cell fates: insights from combinatorial approaches. Ann. N. Y. Acad. Sci. 1266: 7–17.\n",
      "\n",
      "13. Orkin, S.H. 1995. Transcription factors and hematopoietic development. J. Biol. Chem. 270: 4955–4958.\n",
      "\n",
      "14. Vierbuchen, T. et al. 2010. Direct conversion of ﬁbrob- lasts to\n",
      "\n",
      "more reprogramming factors that can ‘erase’ the ‘memories’ of somatic cells.\n",
      "\n",
      "T I M E L I N E\n",
      "\n",
      "A decade of transcription factor‑mediated reprogramming to pluripotency\n",
      "\n",
      "Kazutoshi Takahashi and Shinya Yamanaka\n",
      "\n",
      "Abstract | The past 10 years have seen great advances in our ability to manipulate cell fate, including the induction of pluripotency in vitro to generate induced pluripotent stem cells (iPSCs). This process proved to be remarkably simple from a technical perspective, only needing the host cell and a defined cocktail of transcription factors, with four factors — octamer-binding protein 3/4 (OCT3/4), SOX2, Krüppel-like factor 4 (KLF4) and MYC (collectively referred to as OSKM) —\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "Human: Can transcription factors alone efficiently reprogram cells, or are additional factors or conditions required for successful reprogramming?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the first iPSCs.\n",
      "\n",
      "Direct cell conversion\n",
      "\n",
      "+\n",
      "\n",
      "MYOD\n",
      "\n",
      "Fibroblast\n",
      "\n",
      "+\n",
      "\n",
      "GATA1\n",
      "\n",
      "Myeloblast\n",
      "\n",
      "+\n",
      "\n",
      "CEBP\n",
      "\n",
      "B lymphocyte\n",
      "\n",
      "Diﬀerentiated\n",
      "\n",
      "Transdiﬀerentiation\n",
      "\n",
      "Myoblast\n",
      "\n",
      "Megakaryocyte, eosinophil, erythrocyte\n",
      "\n",
      "Macrophage\n",
      "\n",
      "Identifying candidate reprogramming factors. According to previous experiments, genes that contribute to the identities of fertilized eggs or ES cells were thought to be, in large part, responsible for the reconstitution of specific transcriptional networks during reprogramming. Thus, in 2000, we\n",
      "\n",
      "of ﬁbroblasts into functional cardiomyocytes by deﬁned factors. Cell 142, 375–386 (2010). 9. Haridhasapavalan, K. K., Raina, K., Dey, C., Adhikari, P. & Thummer,\n",
      "\n",
      "6.\n",
      "\n",
      "7.\n",
      "\n",
      "8.\n",
      "\n",
      "R. P. An insight into reprogramming barriers to iPSC generation. Stem Cell Rev. Rep. 16, 56–81 (2020).\n",
      "\n",
      "10. Takahashi, K. & Yamanaka, S. A decade of transcription factor-\n",
      "\n",
      "mediated reprogramming to pluripotency. Nat. Rev. Mol. Cell Biol. 17, 183–193 (2016).\n",
      "\n",
      "11. Brumbaugh, J., Di Stefano, B. & Hochedlinger, K. Reprogramming: identifying the mechanisms that safeguard cell identity. Develop- ment https://doi.org/10.1242/dev.182170 (2019).\n",
      "\n",
      "12. Dal-Pra, S., Hodgkinson, C. P., Mirotsou, M., Kirste, I. & Dzau, V. J.\n",
      "\n",
      "Cell Stem Cell 1, 55–70 (2007). 161. Park, I. H. et al. Reprogramming of human somatic\n",
      "\n",
      "cells to pluripotency with defined factors. Nature 451, 141–146 (2008).\n",
      "\n",
      "162. Hanna, J. et al. Direct cell reprogramming is a stochastic process amenable to acceleration. Nature 462, 595–601 (2009).\n",
      "\n",
      "163. Yang, J. et al. Stat3 activation is limiting for reprogramming to ground state pluripotency. Cell Stem Cell 7, 319–328 (2010).\n",
      "\n",
      "164. Silva, J. et al. Promotion of reprogramming to ground state pluripotency by signal inhibition. PLoS Biol. 6, e253 (2008).\n",
      "\n",
      "Acknowledgements The authors would like to thank all Yamanaka laboratory members, past and present, and are also grateful to Y. Miyake, R. Kato, E.\n",
      "\n",
      "that Drive Neuronal Fate and Reprogramming. Cell Stem . Cell ​ .\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "41. A. Ebrahimi, K. Sevinç, G. Gürhan Sevinç, A. P. Cribbs, M. Philpott, F. Uyulur, T. Morova, J. E. Dunford, S. Göklemez, Ş. Arı, U. Oppermann, T. T. Önder, Bromodomain inhibition of the coactivators CBP/EP300 facilitate cellular reprogramming. 519–528 (2019).\n",
      "\n",
      "Nat. Chem. Biol. ​\n",
      "\n",
      "15\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      ",\n",
      "\n",
      "​\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made\n",
      "\n",
      "available under a\n",
      "\n",
      "CC-BY\n",
      "\n",
      "and adult ﬁbroblast cultures by deﬁned factors. Cell 126, 663–676 (2006). 4.\n",
      "\n",
      "5. Xu, J., Du, Y. & Deng, H. Direct lineage reprogramming: strategies,\n",
      "\n",
      "mechanisms, and applications. Cell Stem Cell 16, 119–134 (2015). Takahashi, K. et al. Induction of pluripotent stem cells from adult human ﬁbroblasts by deﬁned factors. Cell 131, 861–872 (2007). Vierbuchen, T. et al. Direct conversion of ﬁbroblasts to functional neurons by deﬁned factors. Nature 463, 1035–1041 (2010). Ieda, M. et al. Direct reprogramming of ﬁbroblasts into functional cardiomyocytes by deﬁned factors. Cell 142, 375–386 (2010). 9. Haridhasapavalan, K. K., Raina, K., Dey, C., Adhikari, P. & Thummer,\n",
      "\n",
      "6.\n",
      "\n",
      "7.\n",
      "\n",
      "8.\n",
      "\n",
      "R. P. An insight\n",
      "\n",
      "adult human ﬁbroblasts by deﬁned factors. Cell 131: 861–872.\n",
      "\n",
      "10. Morris, S.A. & G.Q. Daley. 2013. A blueprint for engineer- ing cell fate: current technologies to reprogram cell identity. Cell Res. 23: 33–48.\n",
      "\n",
      "11. Kelaini, S., A. Cochrane & A. Margariti. 2014. Direct repro- gramming of adult cells: avoiding the pluripotent state. Stem Cells Cloning 7: 19–29.\n",
      "\n",
      "12. Pereira, C.F., I.R. Lemischka & K. Moore. 2012. Reprogram- ming cell fates: insights from combinatorial approaches. Ann. N. Y. Acad. Sci. 1266: 7–17.\n",
      "\n",
      "13. Orkin, S.H. 1995. Transcription factors and hematopoietic development. J. Biol. Chem. 270: 4955–4958.\n",
      "\n",
      "14. Vierbuchen, T. et al. 2010. Direct conversion of ﬁbrob- lasts to\n",
      "\n",
      "more reprogramming factors that can ‘erase’ the ‘memories’ of somatic cells.\n",
      "\n",
      "T I M E L I N E\n",
      "\n",
      "A decade of transcription factor‑mediated reprogramming to pluripotency\n",
      "\n",
      "Kazutoshi Takahashi and Shinya Yamanaka\n",
      "\n",
      "Abstract | The past 10 years have seen great advances in our ability to manipulate cell fate, including the induction of pluripotency in vitro to generate induced pluripotent stem cells (iPSCs). This process proved to be remarkably simple from a technical perspective, only needing the host cell and a defined cocktail of transcription factors, with four factors — octamer-binding protein 3/4 (OCT3/4), SOX2, Krüppel-like factor 4 (KLF4) and MYC (collectively referred to as OSKM) —\n",
      "\n",
      "themselves58,59,61–63,161,162. Furthermore, it has been clearly shown that the culture environment is important for successful reprogramming, and modification of the culture medium can increase the reprogramming efficiency. For instance the LIF–STAT3 (leukaemia inhibitor factor–signal transducer and activator of transcription 3) pathway was shown to be an active player in induced mouse pluripotency163. Treatment with inhibitors of MEK (for example, PD0325901) and glycogen synthase kinase 3 (for example, CHIR99021) — which are termed 2i — allows mouse ES cells to undergo LIF-independent self-renewal and increases the homogeneity of the cell population, thus facilitating the efficiency of\n",
      "Human: Can transcription factors alone efficiently reprogram cells, or are additional factors or conditions required for successful reprogramming?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "==================================================\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  What role do non-coding RNAs play in transcription factor-mediated cell reprogramming?\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fri Jun  7 23:25:05 2024 INFO RAG\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "RAG >> 10: \n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "Transcription factors orchestrate dynamic interplay between genome topology and gene regulation during cell reprogramming\n",
      "\n",
      "Ralph Stadhouders  François Le Dily1,2,3, Javier Quilez  Yasmina Cuartero1,2,3, Jochen Hecht2,6, Guillaume J. Filion1,2, Miguel Beato1,2, Marc A. Marti-Renom  and Thomas Graf\n",
      "\n",
      "1,2,8*, Enrique Vidal\n",
      "\n",
      "1,2, François Serra\n",
      "\n",
      "1,2,3, Bruno Di Stefano\n",
      "\n",
      "1,2,4,  5, Clara Berenguer1,2,\n",
      "\n",
      "1,2, Antonio Gomez1,2, Samuel Collombet\n",
      "\n",
      "1,2*\n",
      "\n",
      "1,2,3,7*\n",
      "\n",
      "Chromosomal architecture is known to influence gene expression, yet its role in controlling cell fate remains poorly understood. Reprogramming of somatic cells into pluripotent stem cells (PSCs) by the transcription factors (TFs) OCT4, SOX2,\n",
      "\n",
      "promotes reprogramming to pluripotency. Cell Stem Cell 16, 289–301 (2015).\n",
      "\n",
      "21. Aguilo, F. et al. Coordination of m(6)A mRNA methylation and gene transcription by ZFP217 regulates pluripotency and reprogram- ming. Cell Stem Cell 17, 689–704 (2015).\n",
      "\n",
      "22. Brumbaugh, J. et al. Nudt21 controls cell fate by connecting alter- native polyadenylation to chromatin signaling. Cell 172, 106–120 (2018).\n",
      "\n",
      "23. Liu, Z. et al. Single-cell transcriptomics reconstructs fate conver- sion from ﬁbroblast to cardiomyocyte. Nature 551, 100–104 (2017).\n",
      "\n",
      "24. Olson, E. N. Gene regulatory networks in the evolution and devel- opment of the heart. Science 313, 1922–1927 (2006).\n",
      "\n",
      "25. Lambert, S. A. et al. The human\n",
      "\n",
      "et al. Rapid chromatin switch in the direct repro- gramming of ﬁbroblasts to neurons. Cell Rep. 20, 3236–3247 (2017).\n",
      "\n",
      "42. Dall’Agnese, A. et al. Transcription factor-directed re-wiring of\n",
      "\n",
      "chromatin architecture for somatic cell nuclear reprogramming toward trans-differentiation. Mol. Cell 76, 453–472 (2019). 43. Stone, N. R. et al. Context-speciﬁc transcription factor functions\n",
      "\n",
      "regulate epigenomic and transcriptional dynamics during cardiac reprogramming. Cell Stem Cell 25, 87–102 (2019).\n",
      "\n",
      "44. Nam, Y. J. et al. Reprogramming of human ﬁbroblasts toward a cardiac fate. Proc. Natl Acad. Sci. USA 110, 5588–5593 (2013). 45. Zhou, Y. et al. Single-cell transcriptomic analyses of cell fate\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "\n",
      "; , : ) (\n",
      "\n",
      "0 9 8 7 6 5 4 3 2 1\n",
      "\n",
      "; , : ) (\n",
      "\n",
      "0 9 8 7 6 5 4 3 2 1\n",
      "\n",
      "Article\n",
      "\n",
      "https://doi.org/10.1038/s41467-023-37256-8\n",
      "\n",
      "Conserved transcription factors promote cell fate stability and restrict reprogramming potential in differentiated cells\n",
      "\n",
      "Received: 7 August 2021\n",
      "\n",
      "Accepted: 8 March 2023\n",
      "\n",
      "Maria A. Missinato1,6, Sean Murphy2,6, Michaela Lynott Anaïs Kervadec1, Yu-Ling Chang1, Suraj Kannan2, Mafalda Loreti1, Christopher Lee 1, Prashila Amatya1, Hiroshi Tanaka1, Chun-Teng Huang3, 1, Chulan Kwon 2, Peter D. Adams Pier Lorenzo Puri Alessandra Sacco1, Peter Andersen 2 & Alexandre R. Colas\n",
      "\n",
      "1,6, Michael S. Yu1,\n",
      "\n",
      "4, Li Qian 5,\n",
      "\n",
      "1\n",
      "\n",
      "Check for updates\n",
      "\n",
      "Deﬁning the mechanisms safeguarding cell fate\n",
      "\n",
      "reveals the critical role of the TRN in pro- tecting cells against aberrant reprogramming, and demonstrates the vulnerability of donor cell’s TRNs,\n",
      "\n",
      "offering a novel strategy to induce transgene-free trans-differentiations.\n",
      "\n",
      "INTRODUCTION\n",
      "\n",
      "In any given cell, a set of transcription factors (TFs) is ex- pressed and works in concert to govern cellular homeosta- sis and function (1). The Transcriptional Regulatory Net- works (TRNs) in somatic cells are believed to be robust and act to prevent cells from aberrant plasticity (2). However, recent works have revealed that somatic cells in mammals are more plastic than previously thought (3). Particularly, Takahashi and Yamanaka demonstrated this\n",
      "\n",
      "(Figure 2F). To characterize the reprogramming process in more detail, we performed single-cell RNA-seq throughout induction. UCB Lin- CD34+ cells and non-transduced HDFs were proﬁled as con- trols. Genome-wide unsupervised hierarchical clustering shows that most single cells cluster according to sample group, with a clear separation of HDF from reprogrammed cells. Just 2 days after transgene activation, ﬁbroblasts show a dramatic transcrip- tional change. Importantly, endogenous expression of GFI1B and GATA2 is absent at day 2 but activated in reprogrammed cells (Figure S4F), suggesting that the reprogramming process\n",
      "\n",
      "(D) Gene list enrichment analysis with libraries from MGI mutant mouse\n",
      "\n",
      "insertion of SOX2, C‑MYC and TCL‑1A. Stem Cells Dev. 20, 169–180 (2010).\n",
      "\n",
      "NATURE REVIEWS | MOLECULAR CELL BIOLOGY\n",
      "\n",
      "VOLUME 17 | MARCH 2016 | 191\n",
      "\n",
      "© 2016 Macmillan Publishers Limited. All rights reserved\n",
      "\n",
      "52. Heng, J. C. et al. The nuclear receptor Nr5a2 can replace Oct4 in the reprogramming of murine somatic cells to pluripotent cells. Cell Stem Cell 6, 167–174 (2010).\n",
      "\n",
      "53. Judson, R. L., Babiarz, J. E., Venere, M. & Blelloch, R. Embryonic stem cell‑specific microRNAs promote induced pluripotency. Nat. Biotechnol. 27, 459–461 (2009).\n",
      "\n",
      "54. Worringer, K. A. et al. The let‑7/LIN‑41 pathway regulates reprogramming to human induced pluripotent stem cells by controlling expression of\n",
      "Human: What role do non-coding RNAs play in transcription factor-mediated cell reprogramming?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "Gurdon, J. B., Javed, K., Vodnala, M. & Garrett, N. Long-term asso- ciation of a transcription factor with its chromatin binding site can stabilize gene expression and cell fate commitment. Proc. Natl Acad. Sci. USA 117, 15075–15084 (2020).\n",
      "\n",
      "29. Muraoka, N. et al. MiR-133 promotes cardiac reprogramming by\n",
      "\n",
      "directly repressing Snai1 and silencing ﬁbroblast signatures. EMBO J. 33, 1565–1581 (2014).\n",
      "\n",
      "30. Liu, J. et al. The oncogene c-Jun impedes somatic cell reprogram- ming. Nat. Cell Biol. 17, 856–867 (2015).\n",
      "\n",
      "31. Popowski, M. et al. Bright/Arid3A acts as a barrier to somatic cell reprogramming through direct regulation of Oct4, Sox2, and Nanog. Stem Cell Rep. 2, 26–35 (2014).\n",
      "\n",
      "32. Vaseghi,\n",
      "\n",
      "Transcription factors orchestrate dynamic interplay between genome topology and gene regulation during cell reprogramming\n",
      "\n",
      "Ralph Stadhouders  François Le Dily1,2,3, Javier Quilez  Yasmina Cuartero1,2,3, Jochen Hecht2,6, Guillaume J. Filion1,2, Miguel Beato1,2, Marc A. Marti-Renom  and Thomas Graf\n",
      "\n",
      "1,2,8*, Enrique Vidal\n",
      "\n",
      "1,2, François Serra\n",
      "\n",
      "1,2,3, Bruno Di Stefano\n",
      "\n",
      "1,2,4,  5, Clara Berenguer1,2,\n",
      "\n",
      "1,2, Antonio Gomez1,2, Samuel Collombet\n",
      "\n",
      "1,2*\n",
      "\n",
      "1,2,3,7*\n",
      "\n",
      "Chromosomal architecture is known to influence gene expression, yet its role in controlling cell fate remains poorly understood. Reprogramming of somatic cells into pluripotent stem cells (PSCs) by the transcription factors (TFs) OCT4, SOX2,\n",
      "\n",
      "promotes reprogramming to pluripotency. Cell Stem Cell 16, 289–301 (2015).\n",
      "\n",
      "21. Aguilo, F. et al. Coordination of m(6)A mRNA methylation and gene transcription by ZFP217 regulates pluripotency and reprogram- ming. Cell Stem Cell 17, 689–704 (2015).\n",
      "\n",
      "22. Brumbaugh, J. et al. Nudt21 controls cell fate by connecting alter- native polyadenylation to chromatin signaling. Cell 172, 106–120 (2018).\n",
      "\n",
      "23. Liu, Z. et al. Single-cell transcriptomics reconstructs fate conver- sion from ﬁbroblast to cardiomyocyte. Nature 551, 100–104 (2017).\n",
      "\n",
      "24. Olson, E. N. Gene regulatory networks in the evolution and devel- opment of the heart. Science 313, 1922–1927 (2006).\n",
      "\n",
      "25. Lambert, S. A. et al. The human\n",
      "\n",
      "et al. Rapid chromatin switch in the direct repro- gramming of ﬁbroblasts to neurons. Cell Rep. 20, 3236–3247 (2017).\n",
      "\n",
      "42. Dall’Agnese, A. et al. Transcription factor-directed re-wiring of\n",
      "\n",
      "chromatin architecture for somatic cell nuclear reprogramming toward trans-differentiation. Mol. Cell 76, 453–472 (2019). 43. Stone, N. R. et al. Context-speciﬁc transcription factor functions\n",
      "\n",
      "regulate epigenomic and transcriptional dynamics during cardiac reprogramming. Cell Stem Cell 25, 87–102 (2019).\n",
      "\n",
      "44. Nam, Y. J. et al. Reprogramming of human ﬁbroblasts toward a cardiac fate. Proc. Natl Acad. Sci. USA 110, 5588–5593 (2013). 45. Zhou, Y. et al. Single-cell transcriptomic analyses of cell fate\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "\n",
      "; , : ) (\n",
      "\n",
      "0 9 8 7 6 5 4 3 2 1\n",
      "\n",
      "; , : ) (\n",
      "\n",
      "0 9 8 7 6 5 4 3 2 1\n",
      "\n",
      "Article\n",
      "\n",
      "https://doi.org/10.1038/s41467-023-37256-8\n",
      "\n",
      "Conserved transcription factors promote cell fate stability and restrict reprogramming potential in differentiated cells\n",
      "\n",
      "Received: 7 August 2021\n",
      "\n",
      "Accepted: 8 March 2023\n",
      "\n",
      "Maria A. Missinato1,6, Sean Murphy2,6, Michaela Lynott Anaïs Kervadec1, Yu-Ling Chang1, Suraj Kannan2, Mafalda Loreti1, Christopher Lee 1, Prashila Amatya1, Hiroshi Tanaka1, Chun-Teng Huang3, 1, Chulan Kwon 2, Peter D. Adams Pier Lorenzo Puri Alessandra Sacco1, Peter Andersen 2 & Alexandre R. Colas\n",
      "\n",
      "1,6, Michael S. Yu1,\n",
      "\n",
      "4, Li Qian 5,\n",
      "\n",
      "1\n",
      "\n",
      "Check for updates\n",
      "\n",
      "Deﬁning the mechanisms safeguarding cell fate\n",
      "\n",
      "reveals the critical role of the TRN in pro- tecting cells against aberrant reprogramming, and demonstrates the vulnerability of donor cell’s TRNs,\n",
      "\n",
      "offering a novel strategy to induce transgene-free trans-differentiations.\n",
      "\n",
      "INTRODUCTION\n",
      "\n",
      "In any given cell, a set of transcription factors (TFs) is ex- pressed and works in concert to govern cellular homeosta- sis and function (1). The Transcriptional Regulatory Net- works (TRNs) in somatic cells are believed to be robust and act to prevent cells from aberrant plasticity (2). However, recent works have revealed that somatic cells in mammals are more plastic than previously thought (3). Particularly, Takahashi and Yamanaka demonstrated this\n",
      "\n",
      "(Figure 2F). To characterize the reprogramming process in more detail, we performed single-cell RNA-seq throughout induction. UCB Lin- CD34+ cells and non-transduced HDFs were proﬁled as con- trols. Genome-wide unsupervised hierarchical clustering shows that most single cells cluster according to sample group, with a clear separation of HDF from reprogrammed cells. Just 2 days after transgene activation, ﬁbroblasts show a dramatic transcrip- tional change. Importantly, endogenous expression of GFI1B and GATA2 is absent at day 2 but activated in reprogrammed cells (Figure S4F), suggesting that the reprogramming process\n",
      "\n",
      "(D) Gene list enrichment analysis with libraries from MGI mutant mouse\n",
      "\n",
      "insertion of SOX2, C‑MYC and TCL‑1A. Stem Cells Dev. 20, 169–180 (2010).\n",
      "\n",
      "NATURE REVIEWS | MOLECULAR CELL BIOLOGY\n",
      "\n",
      "VOLUME 17 | MARCH 2016 | 191\n",
      "\n",
      "© 2016 Macmillan Publishers Limited. All rights reserved\n",
      "\n",
      "52. Heng, J. C. et al. The nuclear receptor Nr5a2 can replace Oct4 in the reprogramming of murine somatic cells to pluripotent cells. Cell Stem Cell 6, 167–174 (2010).\n",
      "\n",
      "53. Judson, R. L., Babiarz, J. E., Venere, M. & Blelloch, R. Embryonic stem cell‑specific microRNAs promote induced pluripotency. Nat. Biotechnol. 27, 459–461 (2009).\n",
      "\n",
      "54. Worringer, K. A. et al. The let‑7/LIN‑41 pathway regulates reprogramming to human induced pluripotent stem cells by controlling expression of\n",
      "Human: What role do non-coding RNAs play in transcription factor-mediated cell reprogramming?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "Gurdon, J. B., Javed, K., Vodnala, M. & Garrett, N. Long-term asso- ciation of a transcription factor with its chromatin binding site can stabilize gene expression and cell fate commitment. Proc. Natl Acad. Sci. USA 117, 15075–15084 (2020).\n",
      "\n",
      "29. Muraoka, N. et al. MiR-133 promotes cardiac reprogramming by\n",
      "\n",
      "directly repressing Snai1 and silencing ﬁbroblast signatures. EMBO J. 33, 1565–1581 (2014).\n",
      "\n",
      "30. Liu, J. et al. The oncogene c-Jun impedes somatic cell reprogram- ming. Nat. Cell Biol. 17, 856–867 (2015).\n",
      "\n",
      "31. Popowski, M. et al. Bright/Arid3A acts as a barrier to somatic cell reprogramming through direct regulation of Oct4, Sox2, and Nanog. Stem Cell Rep. 2, 26–35 (2014).\n",
      "\n",
      "32. Vaseghi,\n",
      "\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "promotes reprogramming to pluripotency. Cell Stem Cell 16, 289–301 (2015).\n",
      "\n",
      "21. Aguilo, F. et al. Coordination of m(6)A mRNA methylation and gene transcription by ZFP217 regulates pluripotency and reprogram- ming. Cell Stem Cell 17, 689–704 (2015).\n",
      "\n",
      "22. Brumbaugh, J. et al. Nudt21 controls cell fate by connecting alter- native polyadenylation to chromatin signaling. Cell 172, 106–120 (2018).\n",
      "\n",
      "23. Liu, Z. et al. Single-cell transcriptomics reconstructs fate conver- sion from ﬁbroblast to cardiomyocyte. Nature 551, 100–104 (2017).\n",
      "\n",
      "24. Olson, E. N. Gene regulatory networks in the evolution and devel- opment of the heart. Science 313, 1922–1927 (2006).\n",
      "\n",
      "25. Lambert, S. A. et al. The human\n",
      "\n",
      "et al. Rapid chromatin switch in the direct repro- gramming of ﬁbroblasts to neurons. Cell Rep. 20, 3236–3247 (2017).\n",
      "\n",
      "42. Dall’Agnese, A. et al. Transcription factor-directed re-wiring of\n",
      "\n",
      "chromatin architecture for somatic cell nuclear reprogramming toward trans-differentiation. Mol. Cell 76, 453–472 (2019). 43. Stone, N. R. et al. Context-speciﬁc transcription factor functions\n",
      "\n",
      "regulate epigenomic and transcriptional dynamics during cardiac reprogramming. Cell Stem Cell 25, 87–102 (2019).\n",
      "\n",
      "44. Nam, Y. J. et al. Reprogramming of human ﬁbroblasts toward a cardiac fate. Proc. Natl Acad. Sci. USA 110, 5588–5593 (2013). 45. Zhou, Y. et al. Single-cell transcriptomic analyses of cell fate\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "\n",
      "; , : ) (\n",
      "\n",
      "0 9 8 7 6 5 4 3 2 1\n",
      "\n",
      "; , : ) (\n",
      "\n",
      "0 9 8 7 6 5 4 3 2 1\n",
      "\n",
      "Article\n",
      "\n",
      "https://doi.org/10.1038/s41467-023-37256-8\n",
      "\n",
      "Conserved transcription factors promote cell fate stability and restrict reprogramming potential in differentiated cells\n",
      "\n",
      "Received: 7 August 2021\n",
      "\n",
      "Accepted: 8 March 2023\n",
      "\n",
      "Maria A. Missinato1,6, Sean Murphy2,6, Michaela Lynott Anaïs Kervadec1, Yu-Ling Chang1, Suraj Kannan2, Mafalda Loreti1, Christopher Lee 1, Prashila Amatya1, Hiroshi Tanaka1, Chun-Teng Huang3, 1, Chulan Kwon 2, Peter D. Adams Pier Lorenzo Puri Alessandra Sacco1, Peter Andersen 2 & Alexandre R. Colas\n",
      "\n",
      "1,6, Michael S. Yu1,\n",
      "\n",
      "4, Li Qian 5,\n",
      "\n",
      "1\n",
      "\n",
      "Check for updates\n",
      "\n",
      "Deﬁning the mechanisms safeguarding cell fate\n",
      "\n",
      "reveals the critical role of the TRN in pro- tecting cells against aberrant reprogramming, and demonstrates the vulnerability of donor cell’s TRNs,\n",
      "\n",
      "offering a novel strategy to induce transgene-free trans-differentiations.\n",
      "\n",
      "INTRODUCTION\n",
      "\n",
      "In any given cell, a set of transcription factors (TFs) is ex- pressed and works in concert to govern cellular homeosta- sis and function (1). The Transcriptional Regulatory Net- works (TRNs) in somatic cells are believed to be robust and act to prevent cells from aberrant plasticity (2). However, recent works have revealed that somatic cells in mammals are more plastic than previously thought (3). Particularly, Takahashi and Yamanaka demonstrated this\n",
      "\n",
      "(Figure 2F). To characterize the reprogramming process in more detail, we performed single-cell RNA-seq throughout induction. UCB Lin- CD34+ cells and non-transduced HDFs were proﬁled as con- trols. Genome-wide unsupervised hierarchical clustering shows that most single cells cluster according to sample group, with a clear separation of HDF from reprogrammed cells. Just 2 days after transgene activation, ﬁbroblasts show a dramatic transcrip- tional change. Importantly, endogenous expression of GFI1B and GATA2 is absent at day 2 but activated in reprogrammed cells (Figure S4F), suggesting that the reprogramming process\n",
      "\n",
      "(D) Gene list enrichment analysis with libraries from MGI mutant mouse\n",
      "\n",
      "insertion of SOX2, C‑MYC and TCL‑1A. Stem Cells Dev. 20, 169–180 (2010).\n",
      "\n",
      "NATURE REVIEWS | MOLECULAR CELL BIOLOGY\n",
      "\n",
      "VOLUME 17 | MARCH 2016 | 191\n",
      "\n",
      "© 2016 Macmillan Publishers Limited. All rights reserved\n",
      "\n",
      "52. Heng, J. C. et al. The nuclear receptor Nr5a2 can replace Oct4 in the reprogramming of murine somatic cells to pluripotent cells. Cell Stem Cell 6, 167–174 (2010).\n",
      "\n",
      "53. Judson, R. L., Babiarz, J. E., Venere, M. & Blelloch, R. Embryonic stem cell‑specific microRNAs promote induced pluripotency. Nat. Biotechnol. 27, 459–461 (2009).\n",
      "\n",
      "54. Worringer, K. A. et al. The let‑7/LIN‑41 pathway regulates reprogramming to human induced pluripotent stem cells by controlling expression of\n",
      "Human: What role do non-coding RNAs play in transcription factor-mediated cell reprogramming?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "Gurdon, J. B., Javed, K., Vodnala, M. & Garrett, N. Long-term asso- ciation of a transcription factor with its chromatin binding site can stabilize gene expression and cell fate commitment. Proc. Natl Acad. Sci. USA 117, 15075–15084 (2020).\n",
      "\n",
      "29. Muraoka, N. et al. MiR-133 promotes cardiac reprogramming by\n",
      "\n",
      "directly repressing Snai1 and silencing ﬁbroblast signatures. EMBO J. 33, 1565–1581 (2014).\n",
      "\n",
      "30. Liu, J. et al. The oncogene c-Jun impedes somatic cell reprogram- ming. Nat. Cell Biol. 17, 856–867 (2015).\n",
      "\n",
      "31. Popowski, M. et al. Bright/Arid3A acts as a barrier to somatic cell reprogramming through direct regulation of Oct4, Sox2, and Nanog. Stem Cell Rep. 2, 26–35 (2014).\n",
      "\n",
      "32. Vaseghi,\n",
      "\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "Transcription factors orchestrate dynamic interplay between genome topology and gene regulation during cell reprogramming\n",
      "\n",
      "Ralph Stadhouders  François Le Dily1,2,3, Javier Quilez  Yasmina Cuartero1,2,3, Jochen Hecht2,6, Guillaume J. Filion1,2, Miguel Beato1,2, Marc A. Marti-Renom  and Thomas Graf\n",
      "\n",
      "1,2,8*, Enrique Vidal\n",
      "\n",
      "1,2, François Serra\n",
      "\n",
      "1,2,3, Bruno Di Stefano\n",
      "\n",
      "1,2,4,  5, Clara Berenguer1,2,\n",
      "\n",
      "1,2, Antonio Gomez1,2, Samuel Collombet\n",
      "\n",
      "1,2*\n",
      "\n",
      "1,2,3,7*\n",
      "\n",
      "Chromosomal architecture is known to influence gene expression, yet its role in controlling cell fate remains poorly understood. Reprogramming of somatic cells into pluripotent stem cells (PSCs) by the transcription factors (TFs) OCT4, SOX2,\n",
      "\n",
      "et al. Rapid chromatin switch in the direct repro- gramming of ﬁbroblasts to neurons. Cell Rep. 20, 3236–3247 (2017).\n",
      "\n",
      "42. Dall’Agnese, A. et al. Transcription factor-directed re-wiring of\n",
      "\n",
      "chromatin architecture for somatic cell nuclear reprogramming toward trans-differentiation. Mol. Cell 76, 453–472 (2019). 43. Stone, N. R. et al. Context-speciﬁc transcription factor functions\n",
      "\n",
      "regulate epigenomic and transcriptional dynamics during cardiac reprogramming. Cell Stem Cell 25, 87–102 (2019).\n",
      "\n",
      "44. Nam, Y. J. et al. Reprogramming of human ﬁbroblasts toward a cardiac fate. Proc. Natl Acad. Sci. USA 110, 5588–5593 (2013). 45. Zhou, Y. et al. Single-cell transcriptomic analyses of cell fate\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "\n",
      "; , : ) (\n",
      "\n",
      "0 9 8 7 6 5 4 3 2 1\n",
      "\n",
      "; , : ) (\n",
      "\n",
      "0 9 8 7 6 5 4 3 2 1\n",
      "\n",
      "Article\n",
      "\n",
      "https://doi.org/10.1038/s41467-023-37256-8\n",
      "\n",
      "Conserved transcription factors promote cell fate stability and restrict reprogramming potential in differentiated cells\n",
      "\n",
      "Received: 7 August 2021\n",
      "\n",
      "Accepted: 8 March 2023\n",
      "\n",
      "Maria A. Missinato1,6, Sean Murphy2,6, Michaela Lynott Anaïs Kervadec1, Yu-Ling Chang1, Suraj Kannan2, Mafalda Loreti1, Christopher Lee 1, Prashila Amatya1, Hiroshi Tanaka1, Chun-Teng Huang3, 1, Chulan Kwon 2, Peter D. Adams Pier Lorenzo Puri Alessandra Sacco1, Peter Andersen 2 & Alexandre R. Colas\n",
      "\n",
      "1,6, Michael S. Yu1,\n",
      "\n",
      "4, Li Qian 5,\n",
      "\n",
      "1\n",
      "\n",
      "Check for updates\n",
      "\n",
      "Deﬁning the mechanisms safeguarding cell fate\n",
      "\n",
      "reveals the critical role of the TRN in pro- tecting cells against aberrant reprogramming, and demonstrates the vulnerability of donor cell’s TRNs,\n",
      "\n",
      "offering a novel strategy to induce transgene-free trans-differentiations.\n",
      "\n",
      "INTRODUCTION\n",
      "\n",
      "In any given cell, a set of transcription factors (TFs) is ex- pressed and works in concert to govern cellular homeosta- sis and function (1). The Transcriptional Regulatory Net- works (TRNs) in somatic cells are believed to be robust and act to prevent cells from aberrant plasticity (2). However, recent works have revealed that somatic cells in mammals are more plastic than previously thought (3). Particularly, Takahashi and Yamanaka demonstrated this\n",
      "\n",
      "(Figure 2F). To characterize the reprogramming process in more detail, we performed single-cell RNA-seq throughout induction. UCB Lin- CD34+ cells and non-transduced HDFs were proﬁled as con- trols. Genome-wide unsupervised hierarchical clustering shows that most single cells cluster according to sample group, with a clear separation of HDF from reprogrammed cells. Just 2 days after transgene activation, ﬁbroblasts show a dramatic transcrip- tional change. Importantly, endogenous expression of GFI1B and GATA2 is absent at day 2 but activated in reprogrammed cells (Figure S4F), suggesting that the reprogramming process\n",
      "\n",
      "(D) Gene list enrichment analysis with libraries from MGI mutant mouse\n",
      "\n",
      "insertion of SOX2, C‑MYC and TCL‑1A. Stem Cells Dev. 20, 169–180 (2010).\n",
      "\n",
      "NATURE REVIEWS | MOLECULAR CELL BIOLOGY\n",
      "\n",
      "VOLUME 17 | MARCH 2016 | 191\n",
      "\n",
      "© 2016 Macmillan Publishers Limited. All rights reserved\n",
      "\n",
      "52. Heng, J. C. et al. The nuclear receptor Nr5a2 can replace Oct4 in the reprogramming of murine somatic cells to pluripotent cells. Cell Stem Cell 6, 167–174 (2010).\n",
      "\n",
      "53. Judson, R. L., Babiarz, J. E., Venere, M. & Blelloch, R. Embryonic stem cell‑specific microRNAs promote induced pluripotency. Nat. Biotechnol. 27, 459–461 (2009).\n",
      "\n",
      "54. Worringer, K. A. et al. The let‑7/LIN‑41 pathway regulates reprogramming to human induced pluripotent stem cells by controlling expression of\n",
      "Human: What role do non-coding RNAs play in transcription factor-mediated cell reprogramming?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "Gurdon, J. B., Javed, K., Vodnala, M. & Garrett, N. Long-term asso- ciation of a transcription factor with its chromatin binding site can stabilize gene expression and cell fate commitment. Proc. Natl Acad. Sci. USA 117, 15075–15084 (2020).\n",
      "\n",
      "29. Muraoka, N. et al. MiR-133 promotes cardiac reprogramming by\n",
      "\n",
      "directly repressing Snai1 and silencing ﬁbroblast signatures. EMBO J. 33, 1565–1581 (2014).\n",
      "\n",
      "30. Liu, J. et al. The oncogene c-Jun impedes somatic cell reprogram- ming. Nat. Cell Biol. 17, 856–867 (2015).\n",
      "\n",
      "31. Popowski, M. et al. Bright/Arid3A acts as a barrier to somatic cell reprogramming through direct regulation of Oct4, Sox2, and Nanog. Stem Cell Rep. 2, 26–35 (2014).\n",
      "\n",
      "32. Vaseghi,\n",
      "\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "Transcription factors orchestrate dynamic interplay between genome topology and gene regulation during cell reprogramming\n",
      "\n",
      "Ralph Stadhouders  François Le Dily1,2,3, Javier Quilez  Yasmina Cuartero1,2,3, Jochen Hecht2,6, Guillaume J. Filion1,2, Miguel Beato1,2, Marc A. Marti-Renom  and Thomas Graf\n",
      "\n",
      "1,2,8*, Enrique Vidal\n",
      "\n",
      "1,2, François Serra\n",
      "\n",
      "1,2,3, Bruno Di Stefano\n",
      "\n",
      "1,2,4,  5, Clara Berenguer1,2,\n",
      "\n",
      "1,2, Antonio Gomez1,2, Samuel Collombet\n",
      "\n",
      "1,2*\n",
      "\n",
      "1,2,3,7*\n",
      "\n",
      "Chromosomal architecture is known to influence gene expression, yet its role in controlling cell fate remains poorly understood. Reprogramming of somatic cells into pluripotent stem cells (PSCs) by the transcription factors (TFs) OCT4, SOX2,\n",
      "\n",
      "promotes reprogramming to pluripotency. Cell Stem Cell 16, 289–301 (2015).\n",
      "\n",
      "21. Aguilo, F. et al. Coordination of m(6)A mRNA methylation and gene transcription by ZFP217 regulates pluripotency and reprogram- ming. Cell Stem Cell 17, 689–704 (2015).\n",
      "\n",
      "22. Brumbaugh, J. et al. Nudt21 controls cell fate by connecting alter- native polyadenylation to chromatin signaling. Cell 172, 106–120 (2018).\n",
      "\n",
      "23. Liu, Z. et al. Single-cell transcriptomics reconstructs fate conver- sion from ﬁbroblast to cardiomyocyte. Nature 551, 100–104 (2017).\n",
      "\n",
      "24. Olson, E. N. Gene regulatory networks in the evolution and devel- opment of the heart. Science 313, 1922–1927 (2006).\n",
      "\n",
      "25. Lambert, S. A. et al. The human\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "\n",
      "; , : ) (\n",
      "\n",
      "0 9 8 7 6 5 4 3 2 1\n",
      "\n",
      "; , : ) (\n",
      "\n",
      "0 9 8 7 6 5 4 3 2 1\n",
      "\n",
      "Article\n",
      "\n",
      "https://doi.org/10.1038/s41467-023-37256-8\n",
      "\n",
      "Conserved transcription factors promote cell fate stability and restrict reprogramming potential in differentiated cells\n",
      "\n",
      "Received: 7 August 2021\n",
      "\n",
      "Accepted: 8 March 2023\n",
      "\n",
      "Maria A. Missinato1,6, Sean Murphy2,6, Michaela Lynott Anaïs Kervadec1, Yu-Ling Chang1, Suraj Kannan2, Mafalda Loreti1, Christopher Lee 1, Prashila Amatya1, Hiroshi Tanaka1, Chun-Teng Huang3, 1, Chulan Kwon 2, Peter D. Adams Pier Lorenzo Puri Alessandra Sacco1, Peter Andersen 2 & Alexandre R. Colas\n",
      "\n",
      "1,6, Michael S. Yu1,\n",
      "\n",
      "4, Li Qian 5,\n",
      "\n",
      "1\n",
      "\n",
      "Check for updates\n",
      "\n",
      "Deﬁning the mechanisms safeguarding cell fate\n",
      "\n",
      "reveals the critical role of the TRN in pro- tecting cells against aberrant reprogramming, and demonstrates the vulnerability of donor cell’s TRNs,\n",
      "\n",
      "offering a novel strategy to induce transgene-free trans-differentiations.\n",
      "\n",
      "INTRODUCTION\n",
      "\n",
      "In any given cell, a set of transcription factors (TFs) is ex- pressed and works in concert to govern cellular homeosta- sis and function (1). The Transcriptional Regulatory Net- works (TRNs) in somatic cells are believed to be robust and act to prevent cells from aberrant plasticity (2). However, recent works have revealed that somatic cells in mammals are more plastic than previously thought (3). Particularly, Takahashi and Yamanaka demonstrated this\n",
      "\n",
      "(Figure 2F). To characterize the reprogramming process in more detail, we performed single-cell RNA-seq throughout induction. UCB Lin- CD34+ cells and non-transduced HDFs were proﬁled as con- trols. Genome-wide unsupervised hierarchical clustering shows that most single cells cluster according to sample group, with a clear separation of HDF from reprogrammed cells. Just 2 days after transgene activation, ﬁbroblasts show a dramatic transcrip- tional change. Importantly, endogenous expression of GFI1B and GATA2 is absent at day 2 but activated in reprogrammed cells (Figure S4F), suggesting that the reprogramming process\n",
      "\n",
      "(D) Gene list enrichment analysis with libraries from MGI mutant mouse\n",
      "\n",
      "insertion of SOX2, C‑MYC and TCL‑1A. Stem Cells Dev. 20, 169–180 (2010).\n",
      "\n",
      "NATURE REVIEWS | MOLECULAR CELL BIOLOGY\n",
      "\n",
      "VOLUME 17 | MARCH 2016 | 191\n",
      "\n",
      "© 2016 Macmillan Publishers Limited. All rights reserved\n",
      "\n",
      "52. Heng, J. C. et al. The nuclear receptor Nr5a2 can replace Oct4 in the reprogramming of murine somatic cells to pluripotent cells. Cell Stem Cell 6, 167–174 (2010).\n",
      "\n",
      "53. Judson, R. L., Babiarz, J. E., Venere, M. & Blelloch, R. Embryonic stem cell‑specific microRNAs promote induced pluripotency. Nat. Biotechnol. 27, 459–461 (2009).\n",
      "\n",
      "54. Worringer, K. A. et al. The let‑7/LIN‑41 pathway regulates reprogramming to human induced pluripotent stem cells by controlling expression of\n",
      "Human: What role do non-coding RNAs play in transcription factor-mediated cell reprogramming?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "Gurdon, J. B., Javed, K., Vodnala, M. & Garrett, N. Long-term asso- ciation of a transcription factor with its chromatin binding site can stabilize gene expression and cell fate commitment. Proc. Natl Acad. Sci. USA 117, 15075–15084 (2020).\n",
      "\n",
      "29. Muraoka, N. et al. MiR-133 promotes cardiac reprogramming by\n",
      "\n",
      "directly repressing Snai1 and silencing ﬁbroblast signatures. EMBO J. 33, 1565–1581 (2014).\n",
      "\n",
      "30. Liu, J. et al. The oncogene c-Jun impedes somatic cell reprogram- ming. Nat. Cell Biol. 17, 856–867 (2015).\n",
      "\n",
      "31. Popowski, M. et al. Bright/Arid3A acts as a barrier to somatic cell reprogramming through direct regulation of Oct4, Sox2, and Nanog. Stem Cell Rep. 2, 26–35 (2014).\n",
      "\n",
      "32. Vaseghi,\n",
      "\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "Transcription factors orchestrate dynamic interplay between genome topology and gene regulation during cell reprogramming\n",
      "\n",
      "Ralph Stadhouders  François Le Dily1,2,3, Javier Quilez  Yasmina Cuartero1,2,3, Jochen Hecht2,6, Guillaume J. Filion1,2, Miguel Beato1,2, Marc A. Marti-Renom  and Thomas Graf\n",
      "\n",
      "1,2,8*, Enrique Vidal\n",
      "\n",
      "1,2, François Serra\n",
      "\n",
      "1,2,3, Bruno Di Stefano\n",
      "\n",
      "1,2,4,  5, Clara Berenguer1,2,\n",
      "\n",
      "1,2, Antonio Gomez1,2, Samuel Collombet\n",
      "\n",
      "1,2*\n",
      "\n",
      "1,2,3,7*\n",
      "\n",
      "Chromosomal architecture is known to influence gene expression, yet its role in controlling cell fate remains poorly understood. Reprogramming of somatic cells into pluripotent stem cells (PSCs) by the transcription factors (TFs) OCT4, SOX2,\n",
      "\n",
      "promotes reprogramming to pluripotency. Cell Stem Cell 16, 289–301 (2015).\n",
      "\n",
      "21. Aguilo, F. et al. Coordination of m(6)A mRNA methylation and gene transcription by ZFP217 regulates pluripotency and reprogram- ming. Cell Stem Cell 17, 689–704 (2015).\n",
      "\n",
      "22. Brumbaugh, J. et al. Nudt21 controls cell fate by connecting alter- native polyadenylation to chromatin signaling. Cell 172, 106–120 (2018).\n",
      "\n",
      "23. Liu, Z. et al. Single-cell transcriptomics reconstructs fate conver- sion from ﬁbroblast to cardiomyocyte. Nature 551, 100–104 (2017).\n",
      "\n",
      "24. Olson, E. N. Gene regulatory networks in the evolution and devel- opment of the heart. Science 313, 1922–1927 (2006).\n",
      "\n",
      "25. Lambert, S. A. et al. The human\n",
      "\n",
      "et al. Rapid chromatin switch in the direct repro- gramming of ﬁbroblasts to neurons. Cell Rep. 20, 3236–3247 (2017).\n",
      "\n",
      "42. Dall’Agnese, A. et al. Transcription factor-directed re-wiring of\n",
      "\n",
      "chromatin architecture for somatic cell nuclear reprogramming toward trans-differentiation. Mol. Cell 76, 453–472 (2019). 43. Stone, N. R. et al. Context-speciﬁc transcription factor functions\n",
      "\n",
      "regulate epigenomic and transcriptional dynamics during cardiac reprogramming. Cell Stem Cell 25, 87–102 (2019).\n",
      "\n",
      "44. Nam, Y. J. et al. Reprogramming of human ﬁbroblasts toward a cardiac fate. Proc. Natl Acad. Sci. USA 110, 5588–5593 (2013). 45. Zhou, Y. et al. Single-cell transcriptomic analyses of cell fate\n",
      "\n",
      "; , : ) (\n",
      "\n",
      "0 9 8 7 6 5 4 3 2 1\n",
      "\n",
      "; , : ) (\n",
      "\n",
      "0 9 8 7 6 5 4 3 2 1\n",
      "\n",
      "Article\n",
      "\n",
      "https://doi.org/10.1038/s41467-023-37256-8\n",
      "\n",
      "Conserved transcription factors promote cell fate stability and restrict reprogramming potential in differentiated cells\n",
      "\n",
      "Received: 7 August 2021\n",
      "\n",
      "Accepted: 8 March 2023\n",
      "\n",
      "Maria A. Missinato1,6, Sean Murphy2,6, Michaela Lynott Anaïs Kervadec1, Yu-Ling Chang1, Suraj Kannan2, Mafalda Loreti1, Christopher Lee 1, Prashila Amatya1, Hiroshi Tanaka1, Chun-Teng Huang3, 1, Chulan Kwon 2, Peter D. Adams Pier Lorenzo Puri Alessandra Sacco1, Peter Andersen 2 & Alexandre R. Colas\n",
      "\n",
      "1,6, Michael S. Yu1,\n",
      "\n",
      "4, Li Qian 5,\n",
      "\n",
      "1\n",
      "\n",
      "Check for updates\n",
      "\n",
      "Deﬁning the mechanisms safeguarding cell fate\n",
      "\n",
      "reveals the critical role of the TRN in pro- tecting cells against aberrant reprogramming, and demonstrates the vulnerability of donor cell’s TRNs,\n",
      "\n",
      "offering a novel strategy to induce transgene-free trans-differentiations.\n",
      "\n",
      "INTRODUCTION\n",
      "\n",
      "In any given cell, a set of transcription factors (TFs) is ex- pressed and works in concert to govern cellular homeosta- sis and function (1). The Transcriptional Regulatory Net- works (TRNs) in somatic cells are believed to be robust and act to prevent cells from aberrant plasticity (2). However, recent works have revealed that somatic cells in mammals are more plastic than previously thought (3). Particularly, Takahashi and Yamanaka demonstrated this\n",
      "\n",
      "(Figure 2F). To characterize the reprogramming process in more detail, we performed single-cell RNA-seq throughout induction. UCB Lin- CD34+ cells and non-transduced HDFs were proﬁled as con- trols. Genome-wide unsupervised hierarchical clustering shows that most single cells cluster according to sample group, with a clear separation of HDF from reprogrammed cells. Just 2 days after transgene activation, ﬁbroblasts show a dramatic transcrip- tional change. Importantly, endogenous expression of GFI1B and GATA2 is absent at day 2 but activated in reprogrammed cells (Figure S4F), suggesting that the reprogramming process\n",
      "\n",
      "(D) Gene list enrichment analysis with libraries from MGI mutant mouse\n",
      "\n",
      "insertion of SOX2, C‑MYC and TCL‑1A. Stem Cells Dev. 20, 169–180 (2010).\n",
      "\n",
      "NATURE REVIEWS | MOLECULAR CELL BIOLOGY\n",
      "\n",
      "VOLUME 17 | MARCH 2016 | 191\n",
      "\n",
      "© 2016 Macmillan Publishers Limited. All rights reserved\n",
      "\n",
      "52. Heng, J. C. et al. The nuclear receptor Nr5a2 can replace Oct4 in the reprogramming of murine somatic cells to pluripotent cells. Cell Stem Cell 6, 167–174 (2010).\n",
      "\n",
      "53. Judson, R. L., Babiarz, J. E., Venere, M. & Blelloch, R. Embryonic stem cell‑specific microRNAs promote induced pluripotency. Nat. Biotechnol. 27, 459–461 (2009).\n",
      "\n",
      "54. Worringer, K. A. et al. The let‑7/LIN‑41 pathway regulates reprogramming to human induced pluripotent stem cells by controlling expression of\n",
      "Human: What role do non-coding RNAs play in transcription factor-mediated cell reprogramming?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "Gurdon, J. B., Javed, K., Vodnala, M. & Garrett, N. Long-term asso- ciation of a transcription factor with its chromatin binding site can stabilize gene expression and cell fate commitment. Proc. Natl Acad. Sci. USA 117, 15075–15084 (2020).\n",
      "\n",
      "29. Muraoka, N. et al. MiR-133 promotes cardiac reprogramming by\n",
      "\n",
      "directly repressing Snai1 and silencing ﬁbroblast signatures. EMBO J. 33, 1565–1581 (2014).\n",
      "\n",
      "30. Liu, J. et al. The oncogene c-Jun impedes somatic cell reprogram- ming. Nat. Cell Biol. 17, 856–867 (2015).\n",
      "\n",
      "31. Popowski, M. et al. Bright/Arid3A acts as a barrier to somatic cell reprogramming through direct regulation of Oct4, Sox2, and Nanog. Stem Cell Rep. 2, 26–35 (2014).\n",
      "\n",
      "32. Vaseghi,\n",
      "\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "Transcription factors orchestrate dynamic interplay between genome topology and gene regulation during cell reprogramming\n",
      "\n",
      "Ralph Stadhouders  François Le Dily1,2,3, Javier Quilez  Yasmina Cuartero1,2,3, Jochen Hecht2,6, Guillaume J. Filion1,2, Miguel Beato1,2, Marc A. Marti-Renom  and Thomas Graf\n",
      "\n",
      "1,2,8*, Enrique Vidal\n",
      "\n",
      "1,2, François Serra\n",
      "\n",
      "1,2,3, Bruno Di Stefano\n",
      "\n",
      "1,2,4,  5, Clara Berenguer1,2,\n",
      "\n",
      "1,2, Antonio Gomez1,2, Samuel Collombet\n",
      "\n",
      "1,2*\n",
      "\n",
      "1,2,3,7*\n",
      "\n",
      "Chromosomal architecture is known to influence gene expression, yet its role in controlling cell fate remains poorly understood. Reprogramming of somatic cells into pluripotent stem cells (PSCs) by the transcription factors (TFs) OCT4, SOX2,\n",
      "\n",
      "promotes reprogramming to pluripotency. Cell Stem Cell 16, 289–301 (2015).\n",
      "\n",
      "21. Aguilo, F. et al. Coordination of m(6)A mRNA methylation and gene transcription by ZFP217 regulates pluripotency and reprogram- ming. Cell Stem Cell 17, 689–704 (2015).\n",
      "\n",
      "22. Brumbaugh, J. et al. Nudt21 controls cell fate by connecting alter- native polyadenylation to chromatin signaling. Cell 172, 106–120 (2018).\n",
      "\n",
      "23. Liu, Z. et al. Single-cell transcriptomics reconstructs fate conver- sion from ﬁbroblast to cardiomyocyte. Nature 551, 100–104 (2017).\n",
      "\n",
      "24. Olson, E. N. Gene regulatory networks in the evolution and devel- opment of the heart. Science 313, 1922–1927 (2006).\n",
      "\n",
      "25. Lambert, S. A. et al. The human\n",
      "\n",
      "et al. Rapid chromatin switch in the direct repro- gramming of ﬁbroblasts to neurons. Cell Rep. 20, 3236–3247 (2017).\n",
      "\n",
      "42. Dall’Agnese, A. et al. Transcription factor-directed re-wiring of\n",
      "\n",
      "chromatin architecture for somatic cell nuclear reprogramming toward trans-differentiation. Mol. Cell 76, 453–472 (2019). 43. Stone, N. R. et al. Context-speciﬁc transcription factor functions\n",
      "\n",
      "regulate epigenomic and transcriptional dynamics during cardiac reprogramming. Cell Stem Cell 25, 87–102 (2019).\n",
      "\n",
      "44. Nam, Y. J. et al. Reprogramming of human ﬁbroblasts toward a cardiac fate. Proc. Natl Acad. Sci. USA 110, 5588–5593 (2013). 45. Zhou, Y. et al. Single-cell transcriptomic analyses of cell fate\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "\n",
      "reveals the critical role of the TRN in pro- tecting cells against aberrant reprogramming, and demonstrates the vulnerability of donor cell’s TRNs,\n",
      "\n",
      "offering a novel strategy to induce transgene-free trans-differentiations.\n",
      "\n",
      "INTRODUCTION\n",
      "\n",
      "In any given cell, a set of transcription factors (TFs) is ex- pressed and works in concert to govern cellular homeosta- sis and function (1). The Transcriptional Regulatory Net- works (TRNs) in somatic cells are believed to be robust and act to prevent cells from aberrant plasticity (2). However, recent works have revealed that somatic cells in mammals are more plastic than previously thought (3). Particularly, Takahashi and Yamanaka demonstrated this\n",
      "\n",
      "(Figure 2F). To characterize the reprogramming process in more detail, we performed single-cell RNA-seq throughout induction. UCB Lin- CD34+ cells and non-transduced HDFs were proﬁled as con- trols. Genome-wide unsupervised hierarchical clustering shows that most single cells cluster according to sample group, with a clear separation of HDF from reprogrammed cells. Just 2 days after transgene activation, ﬁbroblasts show a dramatic transcrip- tional change. Importantly, endogenous expression of GFI1B and GATA2 is absent at day 2 but activated in reprogrammed cells (Figure S4F), suggesting that the reprogramming process\n",
      "\n",
      "(D) Gene list enrichment analysis with libraries from MGI mutant mouse\n",
      "\n",
      "insertion of SOX2, C‑MYC and TCL‑1A. Stem Cells Dev. 20, 169–180 (2010).\n",
      "\n",
      "NATURE REVIEWS | MOLECULAR CELL BIOLOGY\n",
      "\n",
      "VOLUME 17 | MARCH 2016 | 191\n",
      "\n",
      "© 2016 Macmillan Publishers Limited. All rights reserved\n",
      "\n",
      "52. Heng, J. C. et al. The nuclear receptor Nr5a2 can replace Oct4 in the reprogramming of murine somatic cells to pluripotent cells. Cell Stem Cell 6, 167–174 (2010).\n",
      "\n",
      "53. Judson, R. L., Babiarz, J. E., Venere, M. & Blelloch, R. Embryonic stem cell‑specific microRNAs promote induced pluripotency. Nat. Biotechnol. 27, 459–461 (2009).\n",
      "\n",
      "54. Worringer, K. A. et al. The let‑7/LIN‑41 pathway regulates reprogramming to human induced pluripotent stem cells by controlling expression of\n",
      "Human: What role do non-coding RNAs play in transcription factor-mediated cell reprogramming?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "Gurdon, J. B., Javed, K., Vodnala, M. & Garrett, N. Long-term asso- ciation of a transcription factor with its chromatin binding site can stabilize gene expression and cell fate commitment. Proc. Natl Acad. Sci. USA 117, 15075–15084 (2020).\n",
      "\n",
      "29. Muraoka, N. et al. MiR-133 promotes cardiac reprogramming by\n",
      "\n",
      "directly repressing Snai1 and silencing ﬁbroblast signatures. EMBO J. 33, 1565–1581 (2014).\n",
      "\n",
      "30. Liu, J. et al. The oncogene c-Jun impedes somatic cell reprogram- ming. Nat. Cell Biol. 17, 856–867 (2015).\n",
      "\n",
      "31. Popowski, M. et al. Bright/Arid3A acts as a barrier to somatic cell reprogramming through direct regulation of Oct4, Sox2, and Nanog. Stem Cell Rep. 2, 26–35 (2014).\n",
      "\n",
      "32. Vaseghi,\n",
      "\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "Transcription factors orchestrate dynamic interplay between genome topology and gene regulation during cell reprogramming\n",
      "\n",
      "Ralph Stadhouders  François Le Dily1,2,3, Javier Quilez  Yasmina Cuartero1,2,3, Jochen Hecht2,6, Guillaume J. Filion1,2, Miguel Beato1,2, Marc A. Marti-Renom  and Thomas Graf\n",
      "\n",
      "1,2,8*, Enrique Vidal\n",
      "\n",
      "1,2, François Serra\n",
      "\n",
      "1,2,3, Bruno Di Stefano\n",
      "\n",
      "1,2,4,  5, Clara Berenguer1,2,\n",
      "\n",
      "1,2, Antonio Gomez1,2, Samuel Collombet\n",
      "\n",
      "1,2*\n",
      "\n",
      "1,2,3,7*\n",
      "\n",
      "Chromosomal architecture is known to influence gene expression, yet its role in controlling cell fate remains poorly understood. Reprogramming of somatic cells into pluripotent stem cells (PSCs) by the transcription factors (TFs) OCT4, SOX2,\n",
      "\n",
      "promotes reprogramming to pluripotency. Cell Stem Cell 16, 289–301 (2015).\n",
      "\n",
      "21. Aguilo, F. et al. Coordination of m(6)A mRNA methylation and gene transcription by ZFP217 regulates pluripotency and reprogram- ming. Cell Stem Cell 17, 689–704 (2015).\n",
      "\n",
      "22. Brumbaugh, J. et al. Nudt21 controls cell fate by connecting alter- native polyadenylation to chromatin signaling. Cell 172, 106–120 (2018).\n",
      "\n",
      "23. Liu, Z. et al. Single-cell transcriptomics reconstructs fate conver- sion from ﬁbroblast to cardiomyocyte. Nature 551, 100–104 (2017).\n",
      "\n",
      "24. Olson, E. N. Gene regulatory networks in the evolution and devel- opment of the heart. Science 313, 1922–1927 (2006).\n",
      "\n",
      "25. Lambert, S. A. et al. The human\n",
      "\n",
      "et al. Rapid chromatin switch in the direct repro- gramming of ﬁbroblasts to neurons. Cell Rep. 20, 3236–3247 (2017).\n",
      "\n",
      "42. Dall’Agnese, A. et al. Transcription factor-directed re-wiring of\n",
      "\n",
      "chromatin architecture for somatic cell nuclear reprogramming toward trans-differentiation. Mol. Cell 76, 453–472 (2019). 43. Stone, N. R. et al. Context-speciﬁc transcription factor functions\n",
      "\n",
      "regulate epigenomic and transcriptional dynamics during cardiac reprogramming. Cell Stem Cell 25, 87–102 (2019).\n",
      "\n",
      "44. Nam, Y. J. et al. Reprogramming of human ﬁbroblasts toward a cardiac fate. Proc. Natl Acad. Sci. USA 110, 5588–5593 (2013). 45. Zhou, Y. et al. Single-cell transcriptomic analyses of cell fate\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "\n",
      "; , : ) (\n",
      "\n",
      "0 9 8 7 6 5 4 3 2 1\n",
      "\n",
      "; , : ) (\n",
      "\n",
      "0 9 8 7 6 5 4 3 2 1\n",
      "\n",
      "Article\n",
      "\n",
      "https://doi.org/10.1038/s41467-023-37256-8\n",
      "\n",
      "Conserved transcription factors promote cell fate stability and restrict reprogramming potential in differentiated cells\n",
      "\n",
      "Received: 7 August 2021\n",
      "\n",
      "Accepted: 8 March 2023\n",
      "\n",
      "Maria A. Missinato1,6, Sean Murphy2,6, Michaela Lynott Anaïs Kervadec1, Yu-Ling Chang1, Suraj Kannan2, Mafalda Loreti1, Christopher Lee 1, Prashila Amatya1, Hiroshi Tanaka1, Chun-Teng Huang3, 1, Chulan Kwon 2, Peter D. Adams Pier Lorenzo Puri Alessandra Sacco1, Peter Andersen 2 & Alexandre R. Colas\n",
      "\n",
      "1,6, Michael S. Yu1,\n",
      "\n",
      "4, Li Qian 5,\n",
      "\n",
      "1\n",
      "\n",
      "Check for updates\n",
      "\n",
      "Deﬁning the mechanisms safeguarding cell fate\n",
      "\n",
      "(Figure 2F). To characterize the reprogramming process in more detail, we performed single-cell RNA-seq throughout induction. UCB Lin- CD34+ cells and non-transduced HDFs were proﬁled as con- trols. Genome-wide unsupervised hierarchical clustering shows that most single cells cluster according to sample group, with a clear separation of HDF from reprogrammed cells. Just 2 days after transgene activation, ﬁbroblasts show a dramatic transcrip- tional change. Importantly, endogenous expression of GFI1B and GATA2 is absent at day 2 but activated in reprogrammed cells (Figure S4F), suggesting that the reprogramming process\n",
      "\n",
      "(D) Gene list enrichment analysis with libraries from MGI mutant mouse\n",
      "\n",
      "insertion of SOX2, C‑MYC and TCL‑1A. Stem Cells Dev. 20, 169–180 (2010).\n",
      "\n",
      "NATURE REVIEWS | MOLECULAR CELL BIOLOGY\n",
      "\n",
      "VOLUME 17 | MARCH 2016 | 191\n",
      "\n",
      "© 2016 Macmillan Publishers Limited. All rights reserved\n",
      "\n",
      "52. Heng, J. C. et al. The nuclear receptor Nr5a2 can replace Oct4 in the reprogramming of murine somatic cells to pluripotent cells. Cell Stem Cell 6, 167–174 (2010).\n",
      "\n",
      "53. Judson, R. L., Babiarz, J. E., Venere, M. & Blelloch, R. Embryonic stem cell‑specific microRNAs promote induced pluripotency. Nat. Biotechnol. 27, 459–461 (2009).\n",
      "\n",
      "54. Worringer, K. A. et al. The let‑7/LIN‑41 pathway regulates reprogramming to human induced pluripotent stem cells by controlling expression of\n",
      "Human: What role do non-coding RNAs play in transcription factor-mediated cell reprogramming?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "Gurdon, J. B., Javed, K., Vodnala, M. & Garrett, N. Long-term asso- ciation of a transcription factor with its chromatin binding site can stabilize gene expression and cell fate commitment. Proc. Natl Acad. Sci. USA 117, 15075–15084 (2020).\n",
      "\n",
      "29. Muraoka, N. et al. MiR-133 promotes cardiac reprogramming by\n",
      "\n",
      "directly repressing Snai1 and silencing ﬁbroblast signatures. EMBO J. 33, 1565–1581 (2014).\n",
      "\n",
      "30. Liu, J. et al. The oncogene c-Jun impedes somatic cell reprogram- ming. Nat. Cell Biol. 17, 856–867 (2015).\n",
      "\n",
      "31. Popowski, M. et al. Bright/Arid3A acts as a barrier to somatic cell reprogramming through direct regulation of Oct4, Sox2, and Nanog. Stem Cell Rep. 2, 26–35 (2014).\n",
      "\n",
      "32. Vaseghi,\n",
      "\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "Transcription factors orchestrate dynamic interplay between genome topology and gene regulation during cell reprogramming\n",
      "\n",
      "Ralph Stadhouders  François Le Dily1,2,3, Javier Quilez  Yasmina Cuartero1,2,3, Jochen Hecht2,6, Guillaume J. Filion1,2, Miguel Beato1,2, Marc A. Marti-Renom  and Thomas Graf\n",
      "\n",
      "1,2,8*, Enrique Vidal\n",
      "\n",
      "1,2, François Serra\n",
      "\n",
      "1,2,3, Bruno Di Stefano\n",
      "\n",
      "1,2,4,  5, Clara Berenguer1,2,\n",
      "\n",
      "1,2, Antonio Gomez1,2, Samuel Collombet\n",
      "\n",
      "1,2*\n",
      "\n",
      "1,2,3,7*\n",
      "\n",
      "Chromosomal architecture is known to influence gene expression, yet its role in controlling cell fate remains poorly understood. Reprogramming of somatic cells into pluripotent stem cells (PSCs) by the transcription factors (TFs) OCT4, SOX2,\n",
      "\n",
      "promotes reprogramming to pluripotency. Cell Stem Cell 16, 289–301 (2015).\n",
      "\n",
      "21. Aguilo, F. et al. Coordination of m(6)A mRNA methylation and gene transcription by ZFP217 regulates pluripotency and reprogram- ming. Cell Stem Cell 17, 689–704 (2015).\n",
      "\n",
      "22. Brumbaugh, J. et al. Nudt21 controls cell fate by connecting alter- native polyadenylation to chromatin signaling. Cell 172, 106–120 (2018).\n",
      "\n",
      "23. Liu, Z. et al. Single-cell transcriptomics reconstructs fate conver- sion from ﬁbroblast to cardiomyocyte. Nature 551, 100–104 (2017).\n",
      "\n",
      "24. Olson, E. N. Gene regulatory networks in the evolution and devel- opment of the heart. Science 313, 1922–1927 (2006).\n",
      "\n",
      "25. Lambert, S. A. et al. The human\n",
      "\n",
      "et al. Rapid chromatin switch in the direct repro- gramming of ﬁbroblasts to neurons. Cell Rep. 20, 3236–3247 (2017).\n",
      "\n",
      "42. Dall’Agnese, A. et al. Transcription factor-directed re-wiring of\n",
      "\n",
      "chromatin architecture for somatic cell nuclear reprogramming toward trans-differentiation. Mol. Cell 76, 453–472 (2019). 43. Stone, N. R. et al. Context-speciﬁc transcription factor functions\n",
      "\n",
      "regulate epigenomic and transcriptional dynamics during cardiac reprogramming. Cell Stem Cell 25, 87–102 (2019).\n",
      "\n",
      "44. Nam, Y. J. et al. Reprogramming of human ﬁbroblasts toward a cardiac fate. Proc. Natl Acad. Sci. USA 110, 5588–5593 (2013). 45. Zhou, Y. et al. Single-cell transcriptomic analyses of cell fate\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "\n",
      "; , : ) (\n",
      "\n",
      "0 9 8 7 6 5 4 3 2 1\n",
      "\n",
      "; , : ) (\n",
      "\n",
      "0 9 8 7 6 5 4 3 2 1\n",
      "\n",
      "Article\n",
      "\n",
      "https://doi.org/10.1038/s41467-023-37256-8\n",
      "\n",
      "Conserved transcription factors promote cell fate stability and restrict reprogramming potential in differentiated cells\n",
      "\n",
      "Received: 7 August 2021\n",
      "\n",
      "Accepted: 8 March 2023\n",
      "\n",
      "Maria A. Missinato1,6, Sean Murphy2,6, Michaela Lynott Anaïs Kervadec1, Yu-Ling Chang1, Suraj Kannan2, Mafalda Loreti1, Christopher Lee 1, Prashila Amatya1, Hiroshi Tanaka1, Chun-Teng Huang3, 1, Chulan Kwon 2, Peter D. Adams Pier Lorenzo Puri Alessandra Sacco1, Peter Andersen 2 & Alexandre R. Colas\n",
      "\n",
      "1,6, Michael S. Yu1,\n",
      "\n",
      "4, Li Qian 5,\n",
      "\n",
      "1\n",
      "\n",
      "Check for updates\n",
      "\n",
      "Deﬁning the mechanisms safeguarding cell fate\n",
      "\n",
      "reveals the critical role of the TRN in pro- tecting cells against aberrant reprogramming, and demonstrates the vulnerability of donor cell’s TRNs,\n",
      "\n",
      "offering a novel strategy to induce transgene-free trans-differentiations.\n",
      "\n",
      "INTRODUCTION\n",
      "\n",
      "In any given cell, a set of transcription factors (TFs) is ex- pressed and works in concert to govern cellular homeosta- sis and function (1). The Transcriptional Regulatory Net- works (TRNs) in somatic cells are believed to be robust and act to prevent cells from aberrant plasticity (2). However, recent works have revealed that somatic cells in mammals are more plastic than previously thought (3). Particularly, Takahashi and Yamanaka demonstrated this\n",
      "\n",
      "insertion of SOX2, C‑MYC and TCL‑1A. Stem Cells Dev. 20, 169–180 (2010).\n",
      "\n",
      "NATURE REVIEWS | MOLECULAR CELL BIOLOGY\n",
      "\n",
      "VOLUME 17 | MARCH 2016 | 191\n",
      "\n",
      "© 2016 Macmillan Publishers Limited. All rights reserved\n",
      "\n",
      "52. Heng, J. C. et al. The nuclear receptor Nr5a2 can replace Oct4 in the reprogramming of murine somatic cells to pluripotent cells. Cell Stem Cell 6, 167–174 (2010).\n",
      "\n",
      "53. Judson, R. L., Babiarz, J. E., Venere, M. & Blelloch, R. Embryonic stem cell‑specific microRNAs promote induced pluripotency. Nat. Biotechnol. 27, 459–461 (2009).\n",
      "\n",
      "54. Worringer, K. A. et al. The let‑7/LIN‑41 pathway regulates reprogramming to human induced pluripotent stem cells by controlling expression of\n",
      "Human: What role do non-coding RNAs play in transcription factor-mediated cell reprogramming?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "Gurdon, J. B., Javed, K., Vodnala, M. & Garrett, N. Long-term asso- ciation of a transcription factor with its chromatin binding site can stabilize gene expression and cell fate commitment. Proc. Natl Acad. Sci. USA 117, 15075–15084 (2020).\n",
      "\n",
      "29. Muraoka, N. et al. MiR-133 promotes cardiac reprogramming by\n",
      "\n",
      "directly repressing Snai1 and silencing ﬁbroblast signatures. EMBO J. 33, 1565–1581 (2014).\n",
      "\n",
      "30. Liu, J. et al. The oncogene c-Jun impedes somatic cell reprogram- ming. Nat. Cell Biol. 17, 856–867 (2015).\n",
      "\n",
      "31. Popowski, M. et al. Bright/Arid3A acts as a barrier to somatic cell reprogramming through direct regulation of Oct4, Sox2, and Nanog. Stem Cell Rep. 2, 26–35 (2014).\n",
      "\n",
      "32. Vaseghi,\n",
      "\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "Transcription factors orchestrate dynamic interplay between genome topology and gene regulation during cell reprogramming\n",
      "\n",
      "Ralph Stadhouders  François Le Dily1,2,3, Javier Quilez  Yasmina Cuartero1,2,3, Jochen Hecht2,6, Guillaume J. Filion1,2, Miguel Beato1,2, Marc A. Marti-Renom  and Thomas Graf\n",
      "\n",
      "1,2,8*, Enrique Vidal\n",
      "\n",
      "1,2, François Serra\n",
      "\n",
      "1,2,3, Bruno Di Stefano\n",
      "\n",
      "1,2,4,  5, Clara Berenguer1,2,\n",
      "\n",
      "1,2, Antonio Gomez1,2, Samuel Collombet\n",
      "\n",
      "1,2*\n",
      "\n",
      "1,2,3,7*\n",
      "\n",
      "Chromosomal architecture is known to influence gene expression, yet its role in controlling cell fate remains poorly understood. Reprogramming of somatic cells into pluripotent stem cells (PSCs) by the transcription factors (TFs) OCT4, SOX2,\n",
      "\n",
      "promotes reprogramming to pluripotency. Cell Stem Cell 16, 289–301 (2015).\n",
      "\n",
      "21. Aguilo, F. et al. Coordination of m(6)A mRNA methylation and gene transcription by ZFP217 regulates pluripotency and reprogram- ming. Cell Stem Cell 17, 689–704 (2015).\n",
      "\n",
      "22. Brumbaugh, J. et al. Nudt21 controls cell fate by connecting alter- native polyadenylation to chromatin signaling. Cell 172, 106–120 (2018).\n",
      "\n",
      "23. Liu, Z. et al. Single-cell transcriptomics reconstructs fate conver- sion from ﬁbroblast to cardiomyocyte. Nature 551, 100–104 (2017).\n",
      "\n",
      "24. Olson, E. N. Gene regulatory networks in the evolution and devel- opment of the heart. Science 313, 1922–1927 (2006).\n",
      "\n",
      "25. Lambert, S. A. et al. The human\n",
      "\n",
      "et al. Rapid chromatin switch in the direct repro- gramming of ﬁbroblasts to neurons. Cell Rep. 20, 3236–3247 (2017).\n",
      "\n",
      "42. Dall’Agnese, A. et al. Transcription factor-directed re-wiring of\n",
      "\n",
      "chromatin architecture for somatic cell nuclear reprogramming toward trans-differentiation. Mol. Cell 76, 453–472 (2019). 43. Stone, N. R. et al. Context-speciﬁc transcription factor functions\n",
      "\n",
      "regulate epigenomic and transcriptional dynamics during cardiac reprogramming. Cell Stem Cell 25, 87–102 (2019).\n",
      "\n",
      "44. Nam, Y. J. et al. Reprogramming of human ﬁbroblasts toward a cardiac fate. Proc. Natl Acad. Sci. USA 110, 5588–5593 (2013). 45. Zhou, Y. et al. Single-cell transcriptomic analyses of cell fate\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "\n",
      "; , : ) (\n",
      "\n",
      "0 9 8 7 6 5 4 3 2 1\n",
      "\n",
      "; , : ) (\n",
      "\n",
      "0 9 8 7 6 5 4 3 2 1\n",
      "\n",
      "Article\n",
      "\n",
      "https://doi.org/10.1038/s41467-023-37256-8\n",
      "\n",
      "Conserved transcription factors promote cell fate stability and restrict reprogramming potential in differentiated cells\n",
      "\n",
      "Received: 7 August 2021\n",
      "\n",
      "Accepted: 8 March 2023\n",
      "\n",
      "Maria A. Missinato1,6, Sean Murphy2,6, Michaela Lynott Anaïs Kervadec1, Yu-Ling Chang1, Suraj Kannan2, Mafalda Loreti1, Christopher Lee 1, Prashila Amatya1, Hiroshi Tanaka1, Chun-Teng Huang3, 1, Chulan Kwon 2, Peter D. Adams Pier Lorenzo Puri Alessandra Sacco1, Peter Andersen 2 & Alexandre R. Colas\n",
      "\n",
      "1,6, Michael S. Yu1,\n",
      "\n",
      "4, Li Qian 5,\n",
      "\n",
      "1\n",
      "\n",
      "Check for updates\n",
      "\n",
      "Deﬁning the mechanisms safeguarding cell fate\n",
      "\n",
      "reveals the critical role of the TRN in pro- tecting cells against aberrant reprogramming, and demonstrates the vulnerability of donor cell’s TRNs,\n",
      "\n",
      "offering a novel strategy to induce transgene-free trans-differentiations.\n",
      "\n",
      "INTRODUCTION\n",
      "\n",
      "In any given cell, a set of transcription factors (TFs) is ex- pressed and works in concert to govern cellular homeosta- sis and function (1). The Transcriptional Regulatory Net- works (TRNs) in somatic cells are believed to be robust and act to prevent cells from aberrant plasticity (2). However, recent works have revealed that somatic cells in mammals are more plastic than previously thought (3). Particularly, Takahashi and Yamanaka demonstrated this\n",
      "\n",
      "(Figure 2F). To characterize the reprogramming process in more detail, we performed single-cell RNA-seq throughout induction. UCB Lin- CD34+ cells and non-transduced HDFs were proﬁled as con- trols. Genome-wide unsupervised hierarchical clustering shows that most single cells cluster according to sample group, with a clear separation of HDF from reprogrammed cells. Just 2 days after transgene activation, ﬁbroblasts show a dramatic transcrip- tional change. Importantly, endogenous expression of GFI1B and GATA2 is absent at day 2 but activated in reprogrammed cells (Figure S4F), suggesting that the reprogramming process\n",
      "\n",
      "(D) Gene list enrichment analysis with libraries from MGI mutant mouse\n",
      "Human: What role do non-coding RNAs play in transcription factor-mediated cell reprogramming?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "==================================================\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  What are the current methods for delivering transcription factors to cells for reprogramming purposes?\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fri Jun  7 23:26:02 2024 INFO RAG\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "RAG >> 11: \n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "acids. J. Proteomics 94, 89–109.\n",
      "\n",
      "Takahashi, K., and Yamanaka, S. (2016). A decade of transcription factor- mediated reprogramming to pluripotency. Nat. Rev. Mol. Cell Biol. 17, 183–193.\n",
      "\n",
      "Uhle´ n, M., Fagerberg, L., Hallstro¨ m, B.M., Lindskog, C., Oksvold, P., Mardino- glu, A., Sivertsson, A˚ ., Kampf, C., Sjo¨ stedt, E., Asplund, A., et al. (2015). Proteomics. Tissue-based map of the human proteome. Science 347, 1260419.\n",
      "\n",
      "Vaquerizas, J.M., Kummerfeld, S.K., Teichmann, S.A., and Luscombe, N.M. (2009). A census of human transcription factors: function, expression and evo- lution. Nat. Rev. Genet. 10, 252–263.\n",
      "\n",
      "Virbasius, C.A., Virbasius, J.V., and Scarpulla, R.C. (1993). NRF-1, an activator\n",
      "\n",
      "adult human ﬁbroblasts by deﬁned factors. Cell 131: 861–872.\n",
      "\n",
      "10. Morris, S.A. & G.Q. Daley. 2013. A blueprint for engineer- ing cell fate: current technologies to reprogram cell identity. Cell Res. 23: 33–48.\n",
      "\n",
      "11. Kelaini, S., A. Cochrane & A. Margariti. 2014. Direct repro- gramming of adult cells: avoiding the pluripotent state. Stem Cells Cloning 7: 19–29.\n",
      "\n",
      "12. Pereira, C.F., I.R. Lemischka & K. Moore. 2012. Reprogram- ming cell fates: insights from combinatorial approaches. Ann. N. Y. Acad. Sci. 1266: 7–17.\n",
      "\n",
      "13. Orkin, S.H. 1995. Transcription factors and hematopoietic development. J. Biol. Chem. 270: 4955–4958.\n",
      "\n",
      "14. Vierbuchen, T. et al. 2010. Direct conversion of ﬁbrob- lasts to\n",
      "\n",
      "system: a status report. Trends Immunol. 35, 190–194.\n",
      "\n",
      "Souﬁ, A., Garcia, M.F., Jaroszewicz, A., Osman, N., Pellegrini, M., and Zaret, K.S. (2015). Pioneer transcription factors target partial DNA motifs on nucleo- somes to initiate reprogramming. Cell 161, 555–568.\n",
      "\n",
      "Spitz, F., and Furlong, E.E. (2012). Transcription factors: from enhancer bind- ing to developmental control. Nat. Rev. Genet. 13, 613–626.\n",
      "\n",
      "Stormo, G.D., and Zhao, Y. (2010). Determining the speciﬁcity of protein-DNA interactions. Nat. Rev. Genet. 11, 751–760.\n",
      "\n",
      "Stubbs, L., Sun, Y., and Caetano-Anolles, D. (2011). Function and evolution of C2H2 zinc ﬁnger arrays. Subcell. Biochem. 52, 75–94.\n",
      "\n",
      "Sur, I., and Taipale, J. (2016). The\n",
      "\n",
      "in the field of cellular reprogramming. iPSCs are promising for many applications, including drug discovery, disease modelling and regenerative medicine23 (FIG. 2).\n",
      "\n",
      "Somatic cell nuclear transfer (SCNT)\n",
      "\n",
      "+\n",
      "\n",
      "Somatic cell (nucleus)\n",
      "\n",
      "Enucleated egg\n",
      "\n",
      "SCNT- ES cell\n",
      "\n",
      "Cell fusion with pluripotent stem cell (PSC)\n",
      "\n",
      "In this Timeline article, we first introduce how reprogramming factors were discovered and how they led to the first report of iPSCs 10 years ago. We also describe other factors that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated\n",
      "\n",
      "more reprogramming factors that can ‘erase’ the ‘memories’ of somatic cells.\n",
      "\n",
      "T I M E L I N E\n",
      "\n",
      "A decade of transcription factor‑mediated reprogramming to pluripotency\n",
      "\n",
      "Kazutoshi Takahashi and Shinya Yamanaka\n",
      "\n",
      "Abstract | The past 10 years have seen great advances in our ability to manipulate cell fate, including the induction of pluripotency in vitro to generate induced pluripotent stem cells (iPSCs). This process proved to be remarkably simple from a technical perspective, only needing the host cell and a defined cocktail of transcription factors, with four factors — octamer-binding protein 3/4 (OCT3/4), SOX2, Krüppel-like factor 4 (KLF4) and MYC (collectively referred to as OSKM) —\n",
      "\n",
      "Cahan, P. et al. CellNet: network biology applied to stem cell engineering. Cell 158, 903–915 (2014).\n",
      "\n",
      "10. Rackham, O. J. et al. A predictive computational framework for direct reprogramming between human cell types. Nat. Genet. 48, 331–335 (2016).\n",
      "\n",
      "11. D’Alessio, A. C. et al. A systematic approach to identify candidate trans-\n",
      "\n",
      "cription factors that control cell identity. Stem Cell Rep. 5, 763–775 (2015). 12. Lambert, S. A. et al. The human transcription factors. Cell 175, 598–599 (2018). 13. Nakatake, Y. et al. Generation and profiling of 2,135 human ESC lines for the systematic analyses of cell states perturbed by inducing single transcription factors. Cell Rep. 31, 107655 (2020).\n",
      "\n",
      "14.\n",
      "\n",
      "Cahan, P. et al. CellNet: network biology applied to stem cell engineering. Cell 158, 903–915 (2014).\n",
      "\n",
      "10. Rackham, O. J. et al. A predictive computational framework for direct reprogramming between human cell types. Nat. Genet. 48, 331–335 (2016).\n",
      "\n",
      "11. D’Alessio, A. C. et al. A systematic approach to identify candidate trans-\n",
      "\n",
      "cription factors that control cell identity. Stem Cell Rep. 5, 763–775 (2015). 12. Lambert, S. A. et al. The human transcription factors. Cell 175, 598–599 (2018). 13. Nakatake, Y. et al. Generation and profiling of 2,135 human ESC lines for the systematic analyses of cell states perturbed by inducing single transcription factors. Cell Rep. 31, 107655 (2020).\n",
      "\n",
      "14.\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "\n",
      "RC, Ifkovits JL, Pinto F, Kellam LD, Esteso P, Rentschler S, Christoforou N, Epstein JA, Gearhart JD. 2013. Optimiza- tion of direct fibroblast reprogramming to cardiomyocytes using calcium activity as a functional measure of success. J Mol Cell Cardiol 60: 97–106.\n",
      "\n",
      "Anderson JD, Widom J. 2001. Poly(dA–dT) promoter elements increase the equilibrium accessibility of nucleosomal DNA target sites. Mol Cell Biol 21: 3830–3839.\n",
      "\n",
      "Badve S, Turbin D, Thorat MA, Morimiya A, Nielsen TO, Perou CM, Dunn S, Huntsman DG, Nakshatri H. 2007. FOXA1 expression in breast cancer—correlation with\n",
      "\n",
      "Pioneer factors in cell reprogramming\n",
      "\n",
      "luminal subtype A and survival. Clin Cancer Res 13: 4415–4421.\n",
      "\n",
      "Ballare C,\n",
      "Human: What are the current methods for delivering transcription factors to cells for reprogramming purposes?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "adult human ﬁbroblasts by deﬁned factors. Cell 131: 861–872.\n",
      "\n",
      "10. Morris, S.A. & G.Q. Daley. 2013. A blueprint for engineer- ing cell fate: current technologies to reprogram cell identity. Cell Res. 23: 33–48.\n",
      "\n",
      "11. Kelaini, S., A. Cochrane & A. Margariti. 2014. Direct repro- gramming of adult cells: avoiding the pluripotent state. Stem Cells Cloning 7: 19–29.\n",
      "\n",
      "12. Pereira, C.F., I.R. Lemischka & K. Moore. 2012. Reprogram- ming cell fates: insights from combinatorial approaches. Ann. N. Y. Acad. Sci. 1266: 7–17.\n",
      "\n",
      "13. Orkin, S.H. 1995. Transcription factors and hematopoietic development. J. Biol. Chem. 270: 4955–4958.\n",
      "\n",
      "14. Vierbuchen, T. et al. 2010. Direct conversion of ﬁbrob- lasts to\n",
      "\n",
      "system: a status report. Trends Immunol. 35, 190–194.\n",
      "\n",
      "Souﬁ, A., Garcia, M.F., Jaroszewicz, A., Osman, N., Pellegrini, M., and Zaret, K.S. (2015). Pioneer transcription factors target partial DNA motifs on nucleo- somes to initiate reprogramming. Cell 161, 555–568.\n",
      "\n",
      "Spitz, F., and Furlong, E.E. (2012). Transcription factors: from enhancer bind- ing to developmental control. Nat. Rev. Genet. 13, 613–626.\n",
      "\n",
      "Stormo, G.D., and Zhao, Y. (2010). Determining the speciﬁcity of protein-DNA interactions. Nat. Rev. Genet. 11, 751–760.\n",
      "\n",
      "Stubbs, L., Sun, Y., and Caetano-Anolles, D. (2011). Function and evolution of C2H2 zinc ﬁnger arrays. Subcell. Biochem. 52, 75–94.\n",
      "\n",
      "Sur, I., and Taipale, J. (2016). The\n",
      "\n",
      "in the field of cellular reprogramming. iPSCs are promising for many applications, including drug discovery, disease modelling and regenerative medicine23 (FIG. 2).\n",
      "\n",
      "Somatic cell nuclear transfer (SCNT)\n",
      "\n",
      "+\n",
      "\n",
      "Somatic cell (nucleus)\n",
      "\n",
      "Enucleated egg\n",
      "\n",
      "SCNT- ES cell\n",
      "\n",
      "Cell fusion with pluripotent stem cell (PSC)\n",
      "\n",
      "In this Timeline article, we first introduce how reprogramming factors were discovered and how they led to the first report of iPSCs 10 years ago. We also describe other factors that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated\n",
      "\n",
      "more reprogramming factors that can ‘erase’ the ‘memories’ of somatic cells.\n",
      "\n",
      "T I M E L I N E\n",
      "\n",
      "A decade of transcription factor‑mediated reprogramming to pluripotency\n",
      "\n",
      "Kazutoshi Takahashi and Shinya Yamanaka\n",
      "\n",
      "Abstract | The past 10 years have seen great advances in our ability to manipulate cell fate, including the induction of pluripotency in vitro to generate induced pluripotent stem cells (iPSCs). This process proved to be remarkably simple from a technical perspective, only needing the host cell and a defined cocktail of transcription factors, with four factors — octamer-binding protein 3/4 (OCT3/4), SOX2, Krüppel-like factor 4 (KLF4) and MYC (collectively referred to as OSKM) —\n",
      "\n",
      "Cahan, P. et al. CellNet: network biology applied to stem cell engineering. Cell 158, 903–915 (2014).\n",
      "\n",
      "10. Rackham, O. J. et al. A predictive computational framework for direct reprogramming between human cell types. Nat. Genet. 48, 331–335 (2016).\n",
      "\n",
      "11. D’Alessio, A. C. et al. A systematic approach to identify candidate trans-\n",
      "\n",
      "cription factors that control cell identity. Stem Cell Rep. 5, 763–775 (2015). 12. Lambert, S. A. et al. The human transcription factors. Cell 175, 598–599 (2018). 13. Nakatake, Y. et al. Generation and profiling of 2,135 human ESC lines for the systematic analyses of cell states perturbed by inducing single transcription factors. Cell Rep. 31, 107655 (2020).\n",
      "\n",
      "14.\n",
      "\n",
      "Cahan, P. et al. CellNet: network biology applied to stem cell engineering. Cell 158, 903–915 (2014).\n",
      "\n",
      "10. Rackham, O. J. et al. A predictive computational framework for direct reprogramming between human cell types. Nat. Genet. 48, 331–335 (2016).\n",
      "\n",
      "11. D’Alessio, A. C. et al. A systematic approach to identify candidate trans-\n",
      "\n",
      "cription factors that control cell identity. Stem Cell Rep. 5, 763–775 (2015). 12. Lambert, S. A. et al. The human transcription factors. Cell 175, 598–599 (2018). 13. Nakatake, Y. et al. Generation and profiling of 2,135 human ESC lines for the systematic analyses of cell states perturbed by inducing single transcription factors. Cell Rep. 31, 107655 (2020).\n",
      "\n",
      "14.\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "\n",
      "RC, Ifkovits JL, Pinto F, Kellam LD, Esteso P, Rentschler S, Christoforou N, Epstein JA, Gearhart JD. 2013. Optimiza- tion of direct fibroblast reprogramming to cardiomyocytes using calcium activity as a functional measure of success. J Mol Cell Cardiol 60: 97–106.\n",
      "\n",
      "Anderson JD, Widom J. 2001. Poly(dA–dT) promoter elements increase the equilibrium accessibility of nucleosomal DNA target sites. Mol Cell Biol 21: 3830–3839.\n",
      "\n",
      "Badve S, Turbin D, Thorat MA, Morimiya A, Nielsen TO, Perou CM, Dunn S, Huntsman DG, Nakshatri H. 2007. FOXA1 expression in breast cancer—correlation with\n",
      "\n",
      "Pioneer factors in cell reprogramming\n",
      "\n",
      "luminal subtype A and survival. Clin Cancer Res 13: 4415–4421.\n",
      "\n",
      "Ballare C,\n",
      "Human: What are the current methods for delivering transcription factors to cells for reprogramming purposes?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "acids. J. Proteomics 94, 89–109.\n",
      "\n",
      "Takahashi, K., and Yamanaka, S. (2016). A decade of transcription factor- mediated reprogramming to pluripotency. Nat. Rev. Mol. Cell Biol. 17, 183–193.\n",
      "\n",
      "Uhle´ n, M., Fagerberg, L., Hallstro¨ m, B.M., Lindskog, C., Oksvold, P., Mardino- glu, A., Sivertsson, A˚ ., Kampf, C., Sjo¨ stedt, E., Asplund, A., et al. (2015). Proteomics. Tissue-based map of the human proteome. Science 347, 1260419.\n",
      "\n",
      "Vaquerizas, J.M., Kummerfeld, S.K., Teichmann, S.A., and Luscombe, N.M. (2009). A census of human transcription factors: function, expression and evo- lution. Nat. Rev. Genet. 10, 252–263.\n",
      "\n",
      "Virbasius, C.A., Virbasius, J.V., and Scarpulla, R.C. (1993). NRF-1, an activator\n",
      "\n",
      "system: a status report. Trends Immunol. 35, 190–194.\n",
      "\n",
      "Souﬁ, A., Garcia, M.F., Jaroszewicz, A., Osman, N., Pellegrini, M., and Zaret, K.S. (2015). Pioneer transcription factors target partial DNA motifs on nucleo- somes to initiate reprogramming. Cell 161, 555–568.\n",
      "\n",
      "Spitz, F., and Furlong, E.E. (2012). Transcription factors: from enhancer bind- ing to developmental control. Nat. Rev. Genet. 13, 613–626.\n",
      "\n",
      "Stormo, G.D., and Zhao, Y. (2010). Determining the speciﬁcity of protein-DNA interactions. Nat. Rev. Genet. 11, 751–760.\n",
      "\n",
      "Stubbs, L., Sun, Y., and Caetano-Anolles, D. (2011). Function and evolution of C2H2 zinc ﬁnger arrays. Subcell. Biochem. 52, 75–94.\n",
      "\n",
      "Sur, I., and Taipale, J. (2016). The\n",
      "\n",
      "in the field of cellular reprogramming. iPSCs are promising for many applications, including drug discovery, disease modelling and regenerative medicine23 (FIG. 2).\n",
      "\n",
      "Somatic cell nuclear transfer (SCNT)\n",
      "\n",
      "+\n",
      "\n",
      "Somatic cell (nucleus)\n",
      "\n",
      "Enucleated egg\n",
      "\n",
      "SCNT- ES cell\n",
      "\n",
      "Cell fusion with pluripotent stem cell (PSC)\n",
      "\n",
      "In this Timeline article, we first introduce how reprogramming factors were discovered and how they led to the first report of iPSCs 10 years ago. We also describe other factors that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated\n",
      "\n",
      "more reprogramming factors that can ‘erase’ the ‘memories’ of somatic cells.\n",
      "\n",
      "T I M E L I N E\n",
      "\n",
      "A decade of transcription factor‑mediated reprogramming to pluripotency\n",
      "\n",
      "Kazutoshi Takahashi and Shinya Yamanaka\n",
      "\n",
      "Abstract | The past 10 years have seen great advances in our ability to manipulate cell fate, including the induction of pluripotency in vitro to generate induced pluripotent stem cells (iPSCs). This process proved to be remarkably simple from a technical perspective, only needing the host cell and a defined cocktail of transcription factors, with four factors — octamer-binding protein 3/4 (OCT3/4), SOX2, Krüppel-like factor 4 (KLF4) and MYC (collectively referred to as OSKM) —\n",
      "\n",
      "Cahan, P. et al. CellNet: network biology applied to stem cell engineering. Cell 158, 903–915 (2014).\n",
      "\n",
      "10. Rackham, O. J. et al. A predictive computational framework for direct reprogramming between human cell types. Nat. Genet. 48, 331–335 (2016).\n",
      "\n",
      "11. D’Alessio, A. C. et al. A systematic approach to identify candidate trans-\n",
      "\n",
      "cription factors that control cell identity. Stem Cell Rep. 5, 763–775 (2015). 12. Lambert, S. A. et al. The human transcription factors. Cell 175, 598–599 (2018). 13. Nakatake, Y. et al. Generation and profiling of 2,135 human ESC lines for the systematic analyses of cell states perturbed by inducing single transcription factors. Cell Rep. 31, 107655 (2020).\n",
      "\n",
      "14.\n",
      "\n",
      "Cahan, P. et al. CellNet: network biology applied to stem cell engineering. Cell 158, 903–915 (2014).\n",
      "\n",
      "10. Rackham, O. J. et al. A predictive computational framework for direct reprogramming between human cell types. Nat. Genet. 48, 331–335 (2016).\n",
      "\n",
      "11. D’Alessio, A. C. et al. A systematic approach to identify candidate trans-\n",
      "\n",
      "cription factors that control cell identity. Stem Cell Rep. 5, 763–775 (2015). 12. Lambert, S. A. et al. The human transcription factors. Cell 175, 598–599 (2018). 13. Nakatake, Y. et al. Generation and profiling of 2,135 human ESC lines for the systematic analyses of cell states perturbed by inducing single transcription factors. Cell Rep. 31, 107655 (2020).\n",
      "\n",
      "14.\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "\n",
      "RC, Ifkovits JL, Pinto F, Kellam LD, Esteso P, Rentschler S, Christoforou N, Epstein JA, Gearhart JD. 2013. Optimiza- tion of direct fibroblast reprogramming to cardiomyocytes using calcium activity as a functional measure of success. J Mol Cell Cardiol 60: 97–106.\n",
      "\n",
      "Anderson JD, Widom J. 2001. Poly(dA–dT) promoter elements increase the equilibrium accessibility of nucleosomal DNA target sites. Mol Cell Biol 21: 3830–3839.\n",
      "\n",
      "Badve S, Turbin D, Thorat MA, Morimiya A, Nielsen TO, Perou CM, Dunn S, Huntsman DG, Nakshatri H. 2007. FOXA1 expression in breast cancer—correlation with\n",
      "\n",
      "Pioneer factors in cell reprogramming\n",
      "\n",
      "luminal subtype A and survival. Clin Cancer Res 13: 4415–4421.\n",
      "\n",
      "Ballare C,\n",
      "Human: What are the current methods for delivering transcription factors to cells for reprogramming purposes?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "acids. J. Proteomics 94, 89–109.\n",
      "\n",
      "Takahashi, K., and Yamanaka, S. (2016). A decade of transcription factor- mediated reprogramming to pluripotency. Nat. Rev. Mol. Cell Biol. 17, 183–193.\n",
      "\n",
      "Uhle´ n, M., Fagerberg, L., Hallstro¨ m, B.M., Lindskog, C., Oksvold, P., Mardino- glu, A., Sivertsson, A˚ ., Kampf, C., Sjo¨ stedt, E., Asplund, A., et al. (2015). Proteomics. Tissue-based map of the human proteome. Science 347, 1260419.\n",
      "\n",
      "Vaquerizas, J.M., Kummerfeld, S.K., Teichmann, S.A., and Luscombe, N.M. (2009). A census of human transcription factors: function, expression and evo- lution. Nat. Rev. Genet. 10, 252–263.\n",
      "\n",
      "Virbasius, C.A., Virbasius, J.V., and Scarpulla, R.C. (1993). NRF-1, an activator\n",
      "\n",
      "adult human ﬁbroblasts by deﬁned factors. Cell 131: 861–872.\n",
      "\n",
      "10. Morris, S.A. & G.Q. Daley. 2013. A blueprint for engineer- ing cell fate: current technologies to reprogram cell identity. Cell Res. 23: 33–48.\n",
      "\n",
      "11. Kelaini, S., A. Cochrane & A. Margariti. 2014. Direct repro- gramming of adult cells: avoiding the pluripotent state. Stem Cells Cloning 7: 19–29.\n",
      "\n",
      "12. Pereira, C.F., I.R. Lemischka & K. Moore. 2012. Reprogram- ming cell fates: insights from combinatorial approaches. Ann. N. Y. Acad. Sci. 1266: 7–17.\n",
      "\n",
      "13. Orkin, S.H. 1995. Transcription factors and hematopoietic development. J. Biol. Chem. 270: 4955–4958.\n",
      "\n",
      "14. Vierbuchen, T. et al. 2010. Direct conversion of ﬁbrob- lasts to\n",
      "\n",
      "in the field of cellular reprogramming. iPSCs are promising for many applications, including drug discovery, disease modelling and regenerative medicine23 (FIG. 2).\n",
      "\n",
      "Somatic cell nuclear transfer (SCNT)\n",
      "\n",
      "+\n",
      "\n",
      "Somatic cell (nucleus)\n",
      "\n",
      "Enucleated egg\n",
      "\n",
      "SCNT- ES cell\n",
      "\n",
      "Cell fusion with pluripotent stem cell (PSC)\n",
      "\n",
      "In this Timeline article, we first introduce how reprogramming factors were discovered and how they led to the first report of iPSCs 10 years ago. We also describe other factors that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated\n",
      "\n",
      "more reprogramming factors that can ‘erase’ the ‘memories’ of somatic cells.\n",
      "\n",
      "T I M E L I N E\n",
      "\n",
      "A decade of transcription factor‑mediated reprogramming to pluripotency\n",
      "\n",
      "Kazutoshi Takahashi and Shinya Yamanaka\n",
      "\n",
      "Abstract | The past 10 years have seen great advances in our ability to manipulate cell fate, including the induction of pluripotency in vitro to generate induced pluripotent stem cells (iPSCs). This process proved to be remarkably simple from a technical perspective, only needing the host cell and a defined cocktail of transcription factors, with four factors — octamer-binding protein 3/4 (OCT3/4), SOX2, Krüppel-like factor 4 (KLF4) and MYC (collectively referred to as OSKM) —\n",
      "\n",
      "Cahan, P. et al. CellNet: network biology applied to stem cell engineering. Cell 158, 903–915 (2014).\n",
      "\n",
      "10. Rackham, O. J. et al. A predictive computational framework for direct reprogramming between human cell types. Nat. Genet. 48, 331–335 (2016).\n",
      "\n",
      "11. D’Alessio, A. C. et al. A systematic approach to identify candidate trans-\n",
      "\n",
      "cription factors that control cell identity. Stem Cell Rep. 5, 763–775 (2015). 12. Lambert, S. A. et al. The human transcription factors. Cell 175, 598–599 (2018). 13. Nakatake, Y. et al. Generation and profiling of 2,135 human ESC lines for the systematic analyses of cell states perturbed by inducing single transcription factors. Cell Rep. 31, 107655 (2020).\n",
      "\n",
      "14.\n",
      "\n",
      "Cahan, P. et al. CellNet: network biology applied to stem cell engineering. Cell 158, 903–915 (2014).\n",
      "\n",
      "10. Rackham, O. J. et al. A predictive computational framework for direct reprogramming between human cell types. Nat. Genet. 48, 331–335 (2016).\n",
      "\n",
      "11. D’Alessio, A. C. et al. A systematic approach to identify candidate trans-\n",
      "\n",
      "cription factors that control cell identity. Stem Cell Rep. 5, 763–775 (2015). 12. Lambert, S. A. et al. The human transcription factors. Cell 175, 598–599 (2018). 13. Nakatake, Y. et al. Generation and profiling of 2,135 human ESC lines for the systematic analyses of cell states perturbed by inducing single transcription factors. Cell Rep. 31, 107655 (2020).\n",
      "\n",
      "14.\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "\n",
      "RC, Ifkovits JL, Pinto F, Kellam LD, Esteso P, Rentschler S, Christoforou N, Epstein JA, Gearhart JD. 2013. Optimiza- tion of direct fibroblast reprogramming to cardiomyocytes using calcium activity as a functional measure of success. J Mol Cell Cardiol 60: 97–106.\n",
      "\n",
      "Anderson JD, Widom J. 2001. Poly(dA–dT) promoter elements increase the equilibrium accessibility of nucleosomal DNA target sites. Mol Cell Biol 21: 3830–3839.\n",
      "\n",
      "Badve S, Turbin D, Thorat MA, Morimiya A, Nielsen TO, Perou CM, Dunn S, Huntsman DG, Nakshatri H. 2007. FOXA1 expression in breast cancer—correlation with\n",
      "\n",
      "Pioneer factors in cell reprogramming\n",
      "\n",
      "luminal subtype A and survival. Clin Cancer Res 13: 4415–4421.\n",
      "\n",
      "Ballare C,\n",
      "Human: What are the current methods for delivering transcription factors to cells for reprogramming purposes?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "acids. J. Proteomics 94, 89–109.\n",
      "\n",
      "Takahashi, K., and Yamanaka, S. (2016). A decade of transcription factor- mediated reprogramming to pluripotency. Nat. Rev. Mol. Cell Biol. 17, 183–193.\n",
      "\n",
      "Uhle´ n, M., Fagerberg, L., Hallstro¨ m, B.M., Lindskog, C., Oksvold, P., Mardino- glu, A., Sivertsson, A˚ ., Kampf, C., Sjo¨ stedt, E., Asplund, A., et al. (2015). Proteomics. Tissue-based map of the human proteome. Science 347, 1260419.\n",
      "\n",
      "Vaquerizas, J.M., Kummerfeld, S.K., Teichmann, S.A., and Luscombe, N.M. (2009). A census of human transcription factors: function, expression and evo- lution. Nat. Rev. Genet. 10, 252–263.\n",
      "\n",
      "Virbasius, C.A., Virbasius, J.V., and Scarpulla, R.C. (1993). NRF-1, an activator\n",
      "\n",
      "adult human ﬁbroblasts by deﬁned factors. Cell 131: 861–872.\n",
      "\n",
      "10. Morris, S.A. & G.Q. Daley. 2013. A blueprint for engineer- ing cell fate: current technologies to reprogram cell identity. Cell Res. 23: 33–48.\n",
      "\n",
      "11. Kelaini, S., A. Cochrane & A. Margariti. 2014. Direct repro- gramming of adult cells: avoiding the pluripotent state. Stem Cells Cloning 7: 19–29.\n",
      "\n",
      "12. Pereira, C.F., I.R. Lemischka & K. Moore. 2012. Reprogram- ming cell fates: insights from combinatorial approaches. Ann. N. Y. Acad. Sci. 1266: 7–17.\n",
      "\n",
      "13. Orkin, S.H. 1995. Transcription factors and hematopoietic development. J. Biol. Chem. 270: 4955–4958.\n",
      "\n",
      "14. Vierbuchen, T. et al. 2010. Direct conversion of ﬁbrob- lasts to\n",
      "\n",
      "system: a status report. Trends Immunol. 35, 190–194.\n",
      "\n",
      "Souﬁ, A., Garcia, M.F., Jaroszewicz, A., Osman, N., Pellegrini, M., and Zaret, K.S. (2015). Pioneer transcription factors target partial DNA motifs on nucleo- somes to initiate reprogramming. Cell 161, 555–568.\n",
      "\n",
      "Spitz, F., and Furlong, E.E. (2012). Transcription factors: from enhancer bind- ing to developmental control. Nat. Rev. Genet. 13, 613–626.\n",
      "\n",
      "Stormo, G.D., and Zhao, Y. (2010). Determining the speciﬁcity of protein-DNA interactions. Nat. Rev. Genet. 11, 751–760.\n",
      "\n",
      "Stubbs, L., Sun, Y., and Caetano-Anolles, D. (2011). Function and evolution of C2H2 zinc ﬁnger arrays. Subcell. Biochem. 52, 75–94.\n",
      "\n",
      "Sur, I., and Taipale, J. (2016). The\n",
      "\n",
      "more reprogramming factors that can ‘erase’ the ‘memories’ of somatic cells.\n",
      "\n",
      "T I M E L I N E\n",
      "\n",
      "A decade of transcription factor‑mediated reprogramming to pluripotency\n",
      "\n",
      "Kazutoshi Takahashi and Shinya Yamanaka\n",
      "\n",
      "Abstract | The past 10 years have seen great advances in our ability to manipulate cell fate, including the induction of pluripotency in vitro to generate induced pluripotent stem cells (iPSCs). This process proved to be remarkably simple from a technical perspective, only needing the host cell and a defined cocktail of transcription factors, with four factors — octamer-binding protein 3/4 (OCT3/4), SOX2, Krüppel-like factor 4 (KLF4) and MYC (collectively referred to as OSKM) —\n",
      "\n",
      "Cahan, P. et al. CellNet: network biology applied to stem cell engineering. Cell 158, 903–915 (2014).\n",
      "\n",
      "10. Rackham, O. J. et al. A predictive computational framework for direct reprogramming between human cell types. Nat. Genet. 48, 331–335 (2016).\n",
      "\n",
      "11. D’Alessio, A. C. et al. A systematic approach to identify candidate trans-\n",
      "\n",
      "cription factors that control cell identity. Stem Cell Rep. 5, 763–775 (2015). 12. Lambert, S. A. et al. The human transcription factors. Cell 175, 598–599 (2018). 13. Nakatake, Y. et al. Generation and profiling of 2,135 human ESC lines for the systematic analyses of cell states perturbed by inducing single transcription factors. Cell Rep. 31, 107655 (2020).\n",
      "\n",
      "14.\n",
      "\n",
      "Cahan, P. et al. CellNet: network biology applied to stem cell engineering. Cell 158, 903–915 (2014).\n",
      "\n",
      "10. Rackham, O. J. et al. A predictive computational framework for direct reprogramming between human cell types. Nat. Genet. 48, 331–335 (2016).\n",
      "\n",
      "11. D’Alessio, A. C. et al. A systematic approach to identify candidate trans-\n",
      "\n",
      "cription factors that control cell identity. Stem Cell Rep. 5, 763–775 (2015). 12. Lambert, S. A. et al. The human transcription factors. Cell 175, 598–599 (2018). 13. Nakatake, Y. et al. Generation and profiling of 2,135 human ESC lines for the systematic analyses of cell states perturbed by inducing single transcription factors. Cell Rep. 31, 107655 (2020).\n",
      "\n",
      "14.\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "\n",
      "RC, Ifkovits JL, Pinto F, Kellam LD, Esteso P, Rentschler S, Christoforou N, Epstein JA, Gearhart JD. 2013. Optimiza- tion of direct fibroblast reprogramming to cardiomyocytes using calcium activity as a functional measure of success. J Mol Cell Cardiol 60: 97–106.\n",
      "\n",
      "Anderson JD, Widom J. 2001. Poly(dA–dT) promoter elements increase the equilibrium accessibility of nucleosomal DNA target sites. Mol Cell Biol 21: 3830–3839.\n",
      "\n",
      "Badve S, Turbin D, Thorat MA, Morimiya A, Nielsen TO, Perou CM, Dunn S, Huntsman DG, Nakshatri H. 2007. FOXA1 expression in breast cancer—correlation with\n",
      "\n",
      "Pioneer factors in cell reprogramming\n",
      "\n",
      "luminal subtype A and survival. Clin Cancer Res 13: 4415–4421.\n",
      "\n",
      "Ballare C,\n",
      "Human: What are the current methods for delivering transcription factors to cells for reprogramming purposes?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "acids. J. Proteomics 94, 89–109.\n",
      "\n",
      "Takahashi, K., and Yamanaka, S. (2016). A decade of transcription factor- mediated reprogramming to pluripotency. Nat. Rev. Mol. Cell Biol. 17, 183–193.\n",
      "\n",
      "Uhle´ n, M., Fagerberg, L., Hallstro¨ m, B.M., Lindskog, C., Oksvold, P., Mardino- glu, A., Sivertsson, A˚ ., Kampf, C., Sjo¨ stedt, E., Asplund, A., et al. (2015). Proteomics. Tissue-based map of the human proteome. Science 347, 1260419.\n",
      "\n",
      "Vaquerizas, J.M., Kummerfeld, S.K., Teichmann, S.A., and Luscombe, N.M. (2009). A census of human transcription factors: function, expression and evo- lution. Nat. Rev. Genet. 10, 252–263.\n",
      "\n",
      "Virbasius, C.A., Virbasius, J.V., and Scarpulla, R.C. (1993). NRF-1, an activator\n",
      "\n",
      "adult human ﬁbroblasts by deﬁned factors. Cell 131: 861–872.\n",
      "\n",
      "10. Morris, S.A. & G.Q. Daley. 2013. A blueprint for engineer- ing cell fate: current technologies to reprogram cell identity. Cell Res. 23: 33–48.\n",
      "\n",
      "11. Kelaini, S., A. Cochrane & A. Margariti. 2014. Direct repro- gramming of adult cells: avoiding the pluripotent state. Stem Cells Cloning 7: 19–29.\n",
      "\n",
      "12. Pereira, C.F., I.R. Lemischka & K. Moore. 2012. Reprogram- ming cell fates: insights from combinatorial approaches. Ann. N. Y. Acad. Sci. 1266: 7–17.\n",
      "\n",
      "13. Orkin, S.H. 1995. Transcription factors and hematopoietic development. J. Biol. Chem. 270: 4955–4958.\n",
      "\n",
      "14. Vierbuchen, T. et al. 2010. Direct conversion of ﬁbrob- lasts to\n",
      "\n",
      "system: a status report. Trends Immunol. 35, 190–194.\n",
      "\n",
      "Souﬁ, A., Garcia, M.F., Jaroszewicz, A., Osman, N., Pellegrini, M., and Zaret, K.S. (2015). Pioneer transcription factors target partial DNA motifs on nucleo- somes to initiate reprogramming. Cell 161, 555–568.\n",
      "\n",
      "Spitz, F., and Furlong, E.E. (2012). Transcription factors: from enhancer bind- ing to developmental control. Nat. Rev. Genet. 13, 613–626.\n",
      "\n",
      "Stormo, G.D., and Zhao, Y. (2010). Determining the speciﬁcity of protein-DNA interactions. Nat. Rev. Genet. 11, 751–760.\n",
      "\n",
      "Stubbs, L., Sun, Y., and Caetano-Anolles, D. (2011). Function and evolution of C2H2 zinc ﬁnger arrays. Subcell. Biochem. 52, 75–94.\n",
      "\n",
      "Sur, I., and Taipale, J. (2016). The\n",
      "\n",
      "in the field of cellular reprogramming. iPSCs are promising for many applications, including drug discovery, disease modelling and regenerative medicine23 (FIG. 2).\n",
      "\n",
      "Somatic cell nuclear transfer (SCNT)\n",
      "\n",
      "+\n",
      "\n",
      "Somatic cell (nucleus)\n",
      "\n",
      "Enucleated egg\n",
      "\n",
      "SCNT- ES cell\n",
      "\n",
      "Cell fusion with pluripotent stem cell (PSC)\n",
      "\n",
      "In this Timeline article, we first introduce how reprogramming factors were discovered and how they led to the first report of iPSCs 10 years ago. We also describe other factors that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated\n",
      "\n",
      "Cahan, P. et al. CellNet: network biology applied to stem cell engineering. Cell 158, 903–915 (2014).\n",
      "\n",
      "10. Rackham, O. J. et al. A predictive computational framework for direct reprogramming between human cell types. Nat. Genet. 48, 331–335 (2016).\n",
      "\n",
      "11. D’Alessio, A. C. et al. A systematic approach to identify candidate trans-\n",
      "\n",
      "cription factors that control cell identity. Stem Cell Rep. 5, 763–775 (2015). 12. Lambert, S. A. et al. The human transcription factors. Cell 175, 598–599 (2018). 13. Nakatake, Y. et al. Generation and profiling of 2,135 human ESC lines for the systematic analyses of cell states perturbed by inducing single transcription factors. Cell Rep. 31, 107655 (2020).\n",
      "\n",
      "14.\n",
      "\n",
      "Cahan, P. et al. CellNet: network biology applied to stem cell engineering. Cell 158, 903–915 (2014).\n",
      "\n",
      "10. Rackham, O. J. et al. A predictive computational framework for direct reprogramming between human cell types. Nat. Genet. 48, 331–335 (2016).\n",
      "\n",
      "11. D’Alessio, A. C. et al. A systematic approach to identify candidate trans-\n",
      "\n",
      "cription factors that control cell identity. Stem Cell Rep. 5, 763–775 (2015). 12. Lambert, S. A. et al. The human transcription factors. Cell 175, 598–599 (2018). 13. Nakatake, Y. et al. Generation and profiling of 2,135 human ESC lines for the systematic analyses of cell states perturbed by inducing single transcription factors. Cell Rep. 31, 107655 (2020).\n",
      "\n",
      "14.\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "\n",
      "RC, Ifkovits JL, Pinto F, Kellam LD, Esteso P, Rentschler S, Christoforou N, Epstein JA, Gearhart JD. 2013. Optimiza- tion of direct fibroblast reprogramming to cardiomyocytes using calcium activity as a functional measure of success. J Mol Cell Cardiol 60: 97–106.\n",
      "\n",
      "Anderson JD, Widom J. 2001. Poly(dA–dT) promoter elements increase the equilibrium accessibility of nucleosomal DNA target sites. Mol Cell Biol 21: 3830–3839.\n",
      "\n",
      "Badve S, Turbin D, Thorat MA, Morimiya A, Nielsen TO, Perou CM, Dunn S, Huntsman DG, Nakshatri H. 2007. FOXA1 expression in breast cancer—correlation with\n",
      "\n",
      "Pioneer factors in cell reprogramming\n",
      "\n",
      "luminal subtype A and survival. Clin Cancer Res 13: 4415–4421.\n",
      "\n",
      "Ballare C,\n",
      "Human: What are the current methods for delivering transcription factors to cells for reprogramming purposes?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "acids. J. Proteomics 94, 89–109.\n",
      "\n",
      "Takahashi, K., and Yamanaka, S. (2016). A decade of transcription factor- mediated reprogramming to pluripotency. Nat. Rev. Mol. Cell Biol. 17, 183–193.\n",
      "\n",
      "Uhle´ n, M., Fagerberg, L., Hallstro¨ m, B.M., Lindskog, C., Oksvold, P., Mardino- glu, A., Sivertsson, A˚ ., Kampf, C., Sjo¨ stedt, E., Asplund, A., et al. (2015). Proteomics. Tissue-based map of the human proteome. Science 347, 1260419.\n",
      "\n",
      "Vaquerizas, J.M., Kummerfeld, S.K., Teichmann, S.A., and Luscombe, N.M. (2009). A census of human transcription factors: function, expression and evo- lution. Nat. Rev. Genet. 10, 252–263.\n",
      "\n",
      "Virbasius, C.A., Virbasius, J.V., and Scarpulla, R.C. (1993). NRF-1, an activator\n",
      "\n",
      "adult human ﬁbroblasts by deﬁned factors. Cell 131: 861–872.\n",
      "\n",
      "10. Morris, S.A. & G.Q. Daley. 2013. A blueprint for engineer- ing cell fate: current technologies to reprogram cell identity. Cell Res. 23: 33–48.\n",
      "\n",
      "11. Kelaini, S., A. Cochrane & A. Margariti. 2014. Direct repro- gramming of adult cells: avoiding the pluripotent state. Stem Cells Cloning 7: 19–29.\n",
      "\n",
      "12. Pereira, C.F., I.R. Lemischka & K. Moore. 2012. Reprogram- ming cell fates: insights from combinatorial approaches. Ann. N. Y. Acad. Sci. 1266: 7–17.\n",
      "\n",
      "13. Orkin, S.H. 1995. Transcription factors and hematopoietic development. J. Biol. Chem. 270: 4955–4958.\n",
      "\n",
      "14. Vierbuchen, T. et al. 2010. Direct conversion of ﬁbrob- lasts to\n",
      "\n",
      "system: a status report. Trends Immunol. 35, 190–194.\n",
      "\n",
      "Souﬁ, A., Garcia, M.F., Jaroszewicz, A., Osman, N., Pellegrini, M., and Zaret, K.S. (2015). Pioneer transcription factors target partial DNA motifs on nucleo- somes to initiate reprogramming. Cell 161, 555–568.\n",
      "\n",
      "Spitz, F., and Furlong, E.E. (2012). Transcription factors: from enhancer bind- ing to developmental control. Nat. Rev. Genet. 13, 613–626.\n",
      "\n",
      "Stormo, G.D., and Zhao, Y. (2010). Determining the speciﬁcity of protein-DNA interactions. Nat. Rev. Genet. 11, 751–760.\n",
      "\n",
      "Stubbs, L., Sun, Y., and Caetano-Anolles, D. (2011). Function and evolution of C2H2 zinc ﬁnger arrays. Subcell. Biochem. 52, 75–94.\n",
      "\n",
      "Sur, I., and Taipale, J. (2016). The\n",
      "\n",
      "in the field of cellular reprogramming. iPSCs are promising for many applications, including drug discovery, disease modelling and regenerative medicine23 (FIG. 2).\n",
      "\n",
      "Somatic cell nuclear transfer (SCNT)\n",
      "\n",
      "+\n",
      "\n",
      "Somatic cell (nucleus)\n",
      "\n",
      "Enucleated egg\n",
      "\n",
      "SCNT- ES cell\n",
      "\n",
      "Cell fusion with pluripotent stem cell (PSC)\n",
      "\n",
      "In this Timeline article, we first introduce how reprogramming factors were discovered and how they led to the first report of iPSCs 10 years ago. We also describe other factors that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated\n",
      "\n",
      "more reprogramming factors that can ‘erase’ the ‘memories’ of somatic cells.\n",
      "\n",
      "T I M E L I N E\n",
      "\n",
      "A decade of transcription factor‑mediated reprogramming to pluripotency\n",
      "\n",
      "Kazutoshi Takahashi and Shinya Yamanaka\n",
      "\n",
      "Abstract | The past 10 years have seen great advances in our ability to manipulate cell fate, including the induction of pluripotency in vitro to generate induced pluripotent stem cells (iPSCs). This process proved to be remarkably simple from a technical perspective, only needing the host cell and a defined cocktail of transcription factors, with four factors — octamer-binding protein 3/4 (OCT3/4), SOX2, Krüppel-like factor 4 (KLF4) and MYC (collectively referred to as OSKM) —\n",
      "\n",
      "Cahan, P. et al. CellNet: network biology applied to stem cell engineering. Cell 158, 903–915 (2014).\n",
      "\n",
      "10. Rackham, O. J. et al. A predictive computational framework for direct reprogramming between human cell types. Nat. Genet. 48, 331–335 (2016).\n",
      "\n",
      "11. D’Alessio, A. C. et al. A systematic approach to identify candidate trans-\n",
      "\n",
      "cription factors that control cell identity. Stem Cell Rep. 5, 763–775 (2015). 12. Lambert, S. A. et al. The human transcription factors. Cell 175, 598–599 (2018). 13. Nakatake, Y. et al. Generation and profiling of 2,135 human ESC lines for the systematic analyses of cell states perturbed by inducing single transcription factors. Cell Rep. 31, 107655 (2020).\n",
      "\n",
      "14.\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "\n",
      "RC, Ifkovits JL, Pinto F, Kellam LD, Esteso P, Rentschler S, Christoforou N, Epstein JA, Gearhart JD. 2013. Optimiza- tion of direct fibroblast reprogramming to cardiomyocytes using calcium activity as a functional measure of success. J Mol Cell Cardiol 60: 97–106.\n",
      "\n",
      "Anderson JD, Widom J. 2001. Poly(dA–dT) promoter elements increase the equilibrium accessibility of nucleosomal DNA target sites. Mol Cell Biol 21: 3830–3839.\n",
      "\n",
      "Badve S, Turbin D, Thorat MA, Morimiya A, Nielsen TO, Perou CM, Dunn S, Huntsman DG, Nakshatri H. 2007. FOXA1 expression in breast cancer—correlation with\n",
      "\n",
      "Pioneer factors in cell reprogramming\n",
      "\n",
      "luminal subtype A and survival. Clin Cancer Res 13: 4415–4421.\n",
      "\n",
      "Ballare C,\n",
      "Human: What are the current methods for delivering transcription factors to cells for reprogramming purposes?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "acids. J. Proteomics 94, 89–109.\n",
      "\n",
      "Takahashi, K., and Yamanaka, S. (2016). A decade of transcription factor- mediated reprogramming to pluripotency. Nat. Rev. Mol. Cell Biol. 17, 183–193.\n",
      "\n",
      "Uhle´ n, M., Fagerberg, L., Hallstro¨ m, B.M., Lindskog, C., Oksvold, P., Mardino- glu, A., Sivertsson, A˚ ., Kampf, C., Sjo¨ stedt, E., Asplund, A., et al. (2015). Proteomics. Tissue-based map of the human proteome. Science 347, 1260419.\n",
      "\n",
      "Vaquerizas, J.M., Kummerfeld, S.K., Teichmann, S.A., and Luscombe, N.M. (2009). A census of human transcription factors: function, expression and evo- lution. Nat. Rev. Genet. 10, 252–263.\n",
      "\n",
      "Virbasius, C.A., Virbasius, J.V., and Scarpulla, R.C. (1993). NRF-1, an activator\n",
      "\n",
      "adult human ﬁbroblasts by deﬁned factors. Cell 131: 861–872.\n",
      "\n",
      "10. Morris, S.A. & G.Q. Daley. 2013. A blueprint for engineer- ing cell fate: current technologies to reprogram cell identity. Cell Res. 23: 33–48.\n",
      "\n",
      "11. Kelaini, S., A. Cochrane & A. Margariti. 2014. Direct repro- gramming of adult cells: avoiding the pluripotent state. Stem Cells Cloning 7: 19–29.\n",
      "\n",
      "12. Pereira, C.F., I.R. Lemischka & K. Moore. 2012. Reprogram- ming cell fates: insights from combinatorial approaches. Ann. N. Y. Acad. Sci. 1266: 7–17.\n",
      "\n",
      "13. Orkin, S.H. 1995. Transcription factors and hematopoietic development. J. Biol. Chem. 270: 4955–4958.\n",
      "\n",
      "14. Vierbuchen, T. et al. 2010. Direct conversion of ﬁbrob- lasts to\n",
      "\n",
      "system: a status report. Trends Immunol. 35, 190–194.\n",
      "\n",
      "Souﬁ, A., Garcia, M.F., Jaroszewicz, A., Osman, N., Pellegrini, M., and Zaret, K.S. (2015). Pioneer transcription factors target partial DNA motifs on nucleo- somes to initiate reprogramming. Cell 161, 555–568.\n",
      "\n",
      "Spitz, F., and Furlong, E.E. (2012). Transcription factors: from enhancer bind- ing to developmental control. Nat. Rev. Genet. 13, 613–626.\n",
      "\n",
      "Stormo, G.D., and Zhao, Y. (2010). Determining the speciﬁcity of protein-DNA interactions. Nat. Rev. Genet. 11, 751–760.\n",
      "\n",
      "Stubbs, L., Sun, Y., and Caetano-Anolles, D. (2011). Function and evolution of C2H2 zinc ﬁnger arrays. Subcell. Biochem. 52, 75–94.\n",
      "\n",
      "Sur, I., and Taipale, J. (2016). The\n",
      "\n",
      "in the field of cellular reprogramming. iPSCs are promising for many applications, including drug discovery, disease modelling and regenerative medicine23 (FIG. 2).\n",
      "\n",
      "Somatic cell nuclear transfer (SCNT)\n",
      "\n",
      "+\n",
      "\n",
      "Somatic cell (nucleus)\n",
      "\n",
      "Enucleated egg\n",
      "\n",
      "SCNT- ES cell\n",
      "\n",
      "Cell fusion with pluripotent stem cell (PSC)\n",
      "\n",
      "In this Timeline article, we first introduce how reprogramming factors were discovered and how they led to the first report of iPSCs 10 years ago. We also describe other factors that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated\n",
      "\n",
      "more reprogramming factors that can ‘erase’ the ‘memories’ of somatic cells.\n",
      "\n",
      "T I M E L I N E\n",
      "\n",
      "A decade of transcription factor‑mediated reprogramming to pluripotency\n",
      "\n",
      "Kazutoshi Takahashi and Shinya Yamanaka\n",
      "\n",
      "Abstract | The past 10 years have seen great advances in our ability to manipulate cell fate, including the induction of pluripotency in vitro to generate induced pluripotent stem cells (iPSCs). This process proved to be remarkably simple from a technical perspective, only needing the host cell and a defined cocktail of transcription factors, with four factors — octamer-binding protein 3/4 (OCT3/4), SOX2, Krüppel-like factor 4 (KLF4) and MYC (collectively referred to as OSKM) —\n",
      "\n",
      "Cahan, P. et al. CellNet: network biology applied to stem cell engineering. Cell 158, 903–915 (2014).\n",
      "\n",
      "10. Rackham, O. J. et al. A predictive computational framework for direct reprogramming between human cell types. Nat. Genet. 48, 331–335 (2016).\n",
      "\n",
      "11. D’Alessio, A. C. et al. A systematic approach to identify candidate trans-\n",
      "\n",
      "cription factors that control cell identity. Stem Cell Rep. 5, 763–775 (2015). 12. Lambert, S. A. et al. The human transcription factors. Cell 175, 598–599 (2018). 13. Nakatake, Y. et al. Generation and profiling of 2,135 human ESC lines for the systematic analyses of cell states perturbed by inducing single transcription factors. Cell Rep. 31, 107655 (2020).\n",
      "\n",
      "14.\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "\n",
      "RC, Ifkovits JL, Pinto F, Kellam LD, Esteso P, Rentschler S, Christoforou N, Epstein JA, Gearhart JD. 2013. Optimiza- tion of direct fibroblast reprogramming to cardiomyocytes using calcium activity as a functional measure of success. J Mol Cell Cardiol 60: 97–106.\n",
      "\n",
      "Anderson JD, Widom J. 2001. Poly(dA–dT) promoter elements increase the equilibrium accessibility of nucleosomal DNA target sites. Mol Cell Biol 21: 3830–3839.\n",
      "\n",
      "Badve S, Turbin D, Thorat MA, Morimiya A, Nielsen TO, Perou CM, Dunn S, Huntsman DG, Nakshatri H. 2007. FOXA1 expression in breast cancer—correlation with\n",
      "\n",
      "Pioneer factors in cell reprogramming\n",
      "\n",
      "luminal subtype A and survival. Clin Cancer Res 13: 4415–4421.\n",
      "\n",
      "Ballare C,\n",
      "Human: What are the current methods for delivering transcription factors to cells for reprogramming purposes?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "acids. J. Proteomics 94, 89–109.\n",
      "\n",
      "Takahashi, K., and Yamanaka, S. (2016). A decade of transcription factor- mediated reprogramming to pluripotency. Nat. Rev. Mol. Cell Biol. 17, 183–193.\n",
      "\n",
      "Uhle´ n, M., Fagerberg, L., Hallstro¨ m, B.M., Lindskog, C., Oksvold, P., Mardino- glu, A., Sivertsson, A˚ ., Kampf, C., Sjo¨ stedt, E., Asplund, A., et al. (2015). Proteomics. Tissue-based map of the human proteome. Science 347, 1260419.\n",
      "\n",
      "Vaquerizas, J.M., Kummerfeld, S.K., Teichmann, S.A., and Luscombe, N.M. (2009). A census of human transcription factors: function, expression and evo- lution. Nat. Rev. Genet. 10, 252–263.\n",
      "\n",
      "Virbasius, C.A., Virbasius, J.V., and Scarpulla, R.C. (1993). NRF-1, an activator\n",
      "\n",
      "adult human ﬁbroblasts by deﬁned factors. Cell 131: 861–872.\n",
      "\n",
      "10. Morris, S.A. & G.Q. Daley. 2013. A blueprint for engineer- ing cell fate: current technologies to reprogram cell identity. Cell Res. 23: 33–48.\n",
      "\n",
      "11. Kelaini, S., A. Cochrane & A. Margariti. 2014. Direct repro- gramming of adult cells: avoiding the pluripotent state. Stem Cells Cloning 7: 19–29.\n",
      "\n",
      "12. Pereira, C.F., I.R. Lemischka & K. Moore. 2012. Reprogram- ming cell fates: insights from combinatorial approaches. Ann. N. Y. Acad. Sci. 1266: 7–17.\n",
      "\n",
      "13. Orkin, S.H. 1995. Transcription factors and hematopoietic development. J. Biol. Chem. 270: 4955–4958.\n",
      "\n",
      "14. Vierbuchen, T. et al. 2010. Direct conversion of ﬁbrob- lasts to\n",
      "\n",
      "system: a status report. Trends Immunol. 35, 190–194.\n",
      "\n",
      "Souﬁ, A., Garcia, M.F., Jaroszewicz, A., Osman, N., Pellegrini, M., and Zaret, K.S. (2015). Pioneer transcription factors target partial DNA motifs on nucleo- somes to initiate reprogramming. Cell 161, 555–568.\n",
      "\n",
      "Spitz, F., and Furlong, E.E. (2012). Transcription factors: from enhancer bind- ing to developmental control. Nat. Rev. Genet. 13, 613–626.\n",
      "\n",
      "Stormo, G.D., and Zhao, Y. (2010). Determining the speciﬁcity of protein-DNA interactions. Nat. Rev. Genet. 11, 751–760.\n",
      "\n",
      "Stubbs, L., Sun, Y., and Caetano-Anolles, D. (2011). Function and evolution of C2H2 zinc ﬁnger arrays. Subcell. Biochem. 52, 75–94.\n",
      "\n",
      "Sur, I., and Taipale, J. (2016). The\n",
      "\n",
      "in the field of cellular reprogramming. iPSCs are promising for many applications, including drug discovery, disease modelling and regenerative medicine23 (FIG. 2).\n",
      "\n",
      "Somatic cell nuclear transfer (SCNT)\n",
      "\n",
      "+\n",
      "\n",
      "Somatic cell (nucleus)\n",
      "\n",
      "Enucleated egg\n",
      "\n",
      "SCNT- ES cell\n",
      "\n",
      "Cell fusion with pluripotent stem cell (PSC)\n",
      "\n",
      "In this Timeline article, we first introduce how reprogramming factors were discovered and how they led to the first report of iPSCs 10 years ago. We also describe other factors that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated\n",
      "\n",
      "more reprogramming factors that can ‘erase’ the ‘memories’ of somatic cells.\n",
      "\n",
      "T I M E L I N E\n",
      "\n",
      "A decade of transcription factor‑mediated reprogramming to pluripotency\n",
      "\n",
      "Kazutoshi Takahashi and Shinya Yamanaka\n",
      "\n",
      "Abstract | The past 10 years have seen great advances in our ability to manipulate cell fate, including the induction of pluripotency in vitro to generate induced pluripotent stem cells (iPSCs). This process proved to be remarkably simple from a technical perspective, only needing the host cell and a defined cocktail of transcription factors, with four factors — octamer-binding protein 3/4 (OCT3/4), SOX2, Krüppel-like factor 4 (KLF4) and MYC (collectively referred to as OSKM) —\n",
      "\n",
      "Cahan, P. et al. CellNet: network biology applied to stem cell engineering. Cell 158, 903–915 (2014).\n",
      "\n",
      "10. Rackham, O. J. et al. A predictive computational framework for direct reprogramming between human cell types. Nat. Genet. 48, 331–335 (2016).\n",
      "\n",
      "11. D’Alessio, A. C. et al. A systematic approach to identify candidate trans-\n",
      "\n",
      "cription factors that control cell identity. Stem Cell Rep. 5, 763–775 (2015). 12. Lambert, S. A. et al. The human transcription factors. Cell 175, 598–599 (2018). 13. Nakatake, Y. et al. Generation and profiling of 2,135 human ESC lines for the systematic analyses of cell states perturbed by inducing single transcription factors. Cell Rep. 31, 107655 (2020).\n",
      "\n",
      "14.\n",
      "\n",
      "Cahan, P. et al. CellNet: network biology applied to stem cell engineering. Cell 158, 903–915 (2014).\n",
      "\n",
      "10. Rackham, O. J. et al. A predictive computational framework for direct reprogramming between human cell types. Nat. Genet. 48, 331–335 (2016).\n",
      "\n",
      "11. D’Alessio, A. C. et al. A systematic approach to identify candidate trans-\n",
      "\n",
      "cription factors that control cell identity. Stem Cell Rep. 5, 763–775 (2015). 12. Lambert, S. A. et al. The human transcription factors. Cell 175, 598–599 (2018). 13. Nakatake, Y. et al. Generation and profiling of 2,135 human ESC lines for the systematic analyses of cell states perturbed by inducing single transcription factors. Cell Rep. 31, 107655 (2020).\n",
      "\n",
      "14.\n",
      "\n",
      "RC, Ifkovits JL, Pinto F, Kellam LD, Esteso P, Rentschler S, Christoforou N, Epstein JA, Gearhart JD. 2013. Optimiza- tion of direct fibroblast reprogramming to cardiomyocytes using calcium activity as a functional measure of success. J Mol Cell Cardiol 60: 97–106.\n",
      "\n",
      "Anderson JD, Widom J. 2001. Poly(dA–dT) promoter elements increase the equilibrium accessibility of nucleosomal DNA target sites. Mol Cell Biol 21: 3830–3839.\n",
      "\n",
      "Badve S, Turbin D, Thorat MA, Morimiya A, Nielsen TO, Perou CM, Dunn S, Huntsman DG, Nakshatri H. 2007. FOXA1 expression in breast cancer—correlation with\n",
      "\n",
      "Pioneer factors in cell reprogramming\n",
      "\n",
      "luminal subtype A and survival. Clin Cancer Res 13: 4415–4421.\n",
      "\n",
      "Ballare C,\n",
      "Human: What are the current methods for delivering transcription factors to cells for reprogramming purposes?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "REVIEW\n",
      "\n",
      "Pioneer transcription factors in cell reprogramming\n",
      "\n",
      "Makiko Iwafuchi-Doi and Kenneth S. Zaret\n",
      "\n",
      "Institute for Regenerative Medicine, Department of Cell and Developmental Biology, Perelman School of Medicine, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA\n",
      "\n",
      "A subset of eukaryotic transcription factors possesses the remarkable ability to reprogram one type of cell into another. The transcription factors that reprogram cell fate are invariably those that are crucial for the initial cell programming in embryonic development. To elicit cell programming or reprogramming, transcription fac- tors must be able to engage genes that are developmen- tally silenced and\n",
      "\n",
      "acids. J. Proteomics 94, 89–109.\n",
      "\n",
      "Takahashi, K., and Yamanaka, S. (2016). A decade of transcription factor- mediated reprogramming to pluripotency. Nat. Rev. Mol. Cell Biol. 17, 183–193.\n",
      "\n",
      "Uhle´ n, M., Fagerberg, L., Hallstro¨ m, B.M., Lindskog, C., Oksvold, P., Mardino- glu, A., Sivertsson, A˚ ., Kampf, C., Sjo¨ stedt, E., Asplund, A., et al. (2015). Proteomics. Tissue-based map of the human proteome. Science 347, 1260419.\n",
      "\n",
      "Vaquerizas, J.M., Kummerfeld, S.K., Teichmann, S.A., and Luscombe, N.M. (2009). A census of human transcription factors: function, expression and evo- lution. Nat. Rev. Genet. 10, 252–263.\n",
      "\n",
      "Virbasius, C.A., Virbasius, J.V., and Scarpulla, R.C. (1993). NRF-1, an activator\n",
      "\n",
      "adult human ﬁbroblasts by deﬁned factors. Cell 131: 861–872.\n",
      "\n",
      "10. Morris, S.A. & G.Q. Daley. 2013. A blueprint for engineer- ing cell fate: current technologies to reprogram cell identity. Cell Res. 23: 33–48.\n",
      "\n",
      "11. Kelaini, S., A. Cochrane & A. Margariti. 2014. Direct repro- gramming of adult cells: avoiding the pluripotent state. Stem Cells Cloning 7: 19–29.\n",
      "\n",
      "12. Pereira, C.F., I.R. Lemischka & K. Moore. 2012. Reprogram- ming cell fates: insights from combinatorial approaches. Ann. N. Y. Acad. Sci. 1266: 7–17.\n",
      "\n",
      "13. Orkin, S.H. 1995. Transcription factors and hematopoietic development. J. Biol. Chem. 270: 4955–4958.\n",
      "\n",
      "14. Vierbuchen, T. et al. 2010. Direct conversion of ﬁbrob- lasts to\n",
      "\n",
      "system: a status report. Trends Immunol. 35, 190–194.\n",
      "\n",
      "Souﬁ, A., Garcia, M.F., Jaroszewicz, A., Osman, N., Pellegrini, M., and Zaret, K.S. (2015). Pioneer transcription factors target partial DNA motifs on nucleo- somes to initiate reprogramming. Cell 161, 555–568.\n",
      "\n",
      "Spitz, F., and Furlong, E.E. (2012). Transcription factors: from enhancer bind- ing to developmental control. Nat. Rev. Genet. 13, 613–626.\n",
      "\n",
      "Stormo, G.D., and Zhao, Y. (2010). Determining the speciﬁcity of protein-DNA interactions. Nat. Rev. Genet. 11, 751–760.\n",
      "\n",
      "Stubbs, L., Sun, Y., and Caetano-Anolles, D. (2011). Function and evolution of C2H2 zinc ﬁnger arrays. Subcell. Biochem. 52, 75–94.\n",
      "\n",
      "Sur, I., and Taipale, J. (2016). The\n",
      "\n",
      "in the field of cellular reprogramming. iPSCs are promising for many applications, including drug discovery, disease modelling and regenerative medicine23 (FIG. 2).\n",
      "\n",
      "Somatic cell nuclear transfer (SCNT)\n",
      "\n",
      "+\n",
      "\n",
      "Somatic cell (nucleus)\n",
      "\n",
      "Enucleated egg\n",
      "\n",
      "SCNT- ES cell\n",
      "\n",
      "Cell fusion with pluripotent stem cell (PSC)\n",
      "\n",
      "In this Timeline article, we first introduce how reprogramming factors were discovered and how they led to the first report of iPSCs 10 years ago. We also describe other factors that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated\n",
      "\n",
      "more reprogramming factors that can ‘erase’ the ‘memories’ of somatic cells.\n",
      "\n",
      "T I M E L I N E\n",
      "\n",
      "A decade of transcription factor‑mediated reprogramming to pluripotency\n",
      "\n",
      "Kazutoshi Takahashi and Shinya Yamanaka\n",
      "\n",
      "Abstract | The past 10 years have seen great advances in our ability to manipulate cell fate, including the induction of pluripotency in vitro to generate induced pluripotent stem cells (iPSCs). This process proved to be remarkably simple from a technical perspective, only needing the host cell and a defined cocktail of transcription factors, with four factors — octamer-binding protein 3/4 (OCT3/4), SOX2, Krüppel-like factor 4 (KLF4) and MYC (collectively referred to as OSKM) —\n",
      "\n",
      "Cahan, P. et al. CellNet: network biology applied to stem cell engineering. Cell 158, 903–915 (2014).\n",
      "\n",
      "10. Rackham, O. J. et al. A predictive computational framework for direct reprogramming between human cell types. Nat. Genet. 48, 331–335 (2016).\n",
      "\n",
      "11. D’Alessio, A. C. et al. A systematic approach to identify candidate trans-\n",
      "\n",
      "cription factors that control cell identity. Stem Cell Rep. 5, 763–775 (2015). 12. Lambert, S. A. et al. The human transcription factors. Cell 175, 598–599 (2018). 13. Nakatake, Y. et al. Generation and profiling of 2,135 human ESC lines for the systematic analyses of cell states perturbed by inducing single transcription factors. Cell Rep. 31, 107655 (2020).\n",
      "\n",
      "14.\n",
      "\n",
      "Cahan, P. et al. CellNet: network biology applied to stem cell engineering. Cell 158, 903–915 (2014).\n",
      "\n",
      "10. Rackham, O. J. et al. A predictive computational framework for direct reprogramming between human cell types. Nat. Genet. 48, 331–335 (2016).\n",
      "\n",
      "11. D’Alessio, A. C. et al. A systematic approach to identify candidate trans-\n",
      "\n",
      "cription factors that control cell identity. Stem Cell Rep. 5, 763–775 (2015). 12. Lambert, S. A. et al. The human transcription factors. Cell 175, 598–599 (2018). 13. Nakatake, Y. et al. Generation and profiling of 2,135 human ESC lines for the systematic analyses of cell states perturbed by inducing single transcription factors. Cell Rep. 31, 107655 (2020).\n",
      "\n",
      "14.\n",
      "\n",
      "that contribute to transcription factor‑mediated reprogramming and outline the current knowledge of such mechanisms. Finally, we briefly discuss future perspectives on transcription factor‑mediated reprogramming.\n",
      "\n",
      "+\n",
      "\n",
      "(diﬀerentiation) Development\n",
      "\n",
      "Pluripotent\n",
      "\n",
      "Intermediate (progenitor)\n",
      "\n",
      "Somatic cell\n",
      "\n",
      "PSC\n",
      "\n",
      "( r\n",
      "\n",
      "e\n",
      "\n",
      "j\n",
      "\n",
      "r\n",
      "\n",
      "v\n",
      "\n",
      "r\n",
      "\n",
      "a\n",
      "\n",
      "a\n",
      "\n",
      "t i\n",
      "\n",
      "m\n",
      "\n",
      "i\n",
      "\n",
      "o\n",
      "\n",
      "n\n",
      "\n",
      ")\n",
      "\n",
      "4n-PSC\n",
      "\n",
      "Induction of pluripotency Previous works, as described above, suggested the existence of reprogramming factors that can change cellular fate. If this is the case, then what are these reprogramming factors? In this section, we describe the seminal work that led us to the discovery of these factors and their use for the generation of the\n",
      "Human: What are the current methods for delivering transcription factors to cells for reprogramming purposes?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "==================================================\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  How does the overexpression or suppression of specific transcription factors affect the stability and functionality of reprogrammed cells?\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fri Jun  7 23:27:10 2024 INFO RAG\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "RAG >> 12: \n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "differentiation and maintenance were among the most frequently up-regulated (most responsive). We reasoned that one potential function of responsive genes may be to maintain cellular identity. We identified 3 responsive transcription factors in fibroblasts using P ​ and found that suppressing their expression led to enhanced reprogramming efficiency. We propose that responsiveness to perturbations is a property of factors that help maintain cellular identity.\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "Introduction\n",
      "\n",
      "Cells of a particular cell type likely maintain many of their specific behavioral properties across a range of conditions. The maintenance of these properties can be the result of the activity of certain\n",
      "\n",
      "reveals the critical role of the TRN in pro- tecting cells against aberrant reprogramming, and demonstrates the vulnerability of donor cell’s TRNs,\n",
      "\n",
      "offering a novel strategy to induce transgene-free trans-differentiations.\n",
      "\n",
      "INTRODUCTION\n",
      "\n",
      "In any given cell, a set of transcription factors (TFs) is ex- pressed and works in concert to govern cellular homeosta- sis and function (1). The Transcriptional Regulatory Net- works (TRNs) in somatic cells are believed to be robust and act to prevent cells from aberrant plasticity (2). However, recent works have revealed that somatic cells in mammals are more plastic than previously thought (3). Particularly, Takahashi and Yamanaka demonstrated this\n",
      "\n",
      "whose expression or activity is cell type- or tissue-specific. In principle, responsiveness could have identified a proper subset of the cell type-specific genes, a superset of the cell type-specific genes, a wholly distinct set of genes, or a partially overlapping subset. We found that responsiveness seemed to identify a superset of the cell type-specific genes, meaning that there were many genes whose expression was responsive but not particularly cell type-specific. When we explicitly tested whether these responsive-but-not-cell type-specific genes could affect maintenance, we found that, for fully half the factors we tested, suppression of the factor led to an increase in reprogramming\n",
      "\n",
      "whether these responsive-but-not-cell type-specific genes could affect maintenance, we found that, for fully half the factors we tested, suppression of the factor led to an increase in reprogramming efficiency. This validation rate suggests that these responsive 3 factors are indeed important for cell type maintenance and not merely falsely identified by P ​. The fact that these non-cell type-specific factors are important for maintenance suggests that cell type maintenance may involve factors that act as general barriers to cell type transformations in addition to factors specific to particular types. (Although it is possible that such factors may only act as barriers in particular cell\n",
      "\n",
      "Each experimental replicate was conducted using different vials of cells and different batches of media and feeder mouse embryonic fibroblasts (MEFs) on different days; knockdown samples are compared against controls in the same batch. Reprogramming rate variability was highest in the third replicate for unknown reasons.\n",
      "\n",
      "Supplementary Figure 12: Evaluating cardiac transdifferentiation after overexpression of iPSC-derived cardiac myocyte-responsive transcription factors in fibroblasts. A) Selection of 7 transcription factors found to be frequently up-regulated after perturbation in iPSC-derived cardiac myocytes (iPSC-CM; see Methods). B) Schematic of experimental workflow:\n",
      "\n",
      "by enhancing source cell type gene expression or repressing reprogramming TF-associated gene expression. How- ever, a systematic evaluation of TFs as fate stabilizers has not been conducted to date, and in this context, we will explore whether such factors might be evolutionarily conserved and play lineage and/or cell type-speciﬁc or independent roles. Mechanistically, we will evaluate how fate-stabilizing TFs binding to the DNA might con- tribute to regulate chromatin accessibility dynamics and tran- scription to oppose the lineage reprogramming process. Finally, we will test whether the targeted inhibition of such fate stabilizers might enhance our ability to use reprogramming TFs to\n",
      "\n",
      "that Drive Neuronal Fate and Reprogramming. Cell Stem . Cell ​ .\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "41. A. Ebrahimi, K. Sevinç, G. Gürhan Sevinç, A. P. Cribbs, M. Philpott, F. Uyulur, T. Morova, J. E. Dunford, S. Göklemez, Ş. Arı, U. Oppermann, T. T. Önder, Bromodomain inhibition of the coactivators CBP/EP300 facilitate cellular reprogramming. 519–528 (2019).\n",
      "\n",
      "Nat. Chem. Biol. ​\n",
      "\n",
      "15\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      ",\n",
      "\n",
      "​\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made\n",
      "\n",
      "available under a\n",
      "\n",
      "CC-BY\n",
      "\n",
      "Transcription factors orchestrate dynamic interplay between genome topology and gene regulation during cell reprogramming\n",
      "\n",
      "Ralph Stadhouders  François Le Dily1,2,3, Javier Quilez  Yasmina Cuartero1,2,3, Jochen Hecht2,6, Guillaume J. Filion1,2, Miguel Beato1,2, Marc A. Marti-Renom  and Thomas Graf\n",
      "\n",
      "1,2,8*, Enrique Vidal\n",
      "\n",
      "1,2, François Serra\n",
      "\n",
      "1,2,3, Bruno Di Stefano\n",
      "\n",
      "1,2,4,  5, Clara Berenguer1,2,\n",
      "\n",
      "1,2, Antonio Gomez1,2, Samuel Collombet\n",
      "\n",
      "1,2*\n",
      "\n",
      "1,2,3,7*\n",
      "\n",
      "Chromosomal architecture is known to influence gene expression, yet its role in controlling cell fate remains poorly understood. Reprogramming of somatic cells into pluripotent stem cells (PSCs) by the transcription factors (TFs) OCT4, SOX2,\n",
      "\n",
      "Hence we reasoned that responsiveness might also identify previously undescribed factors that maintain cellular identity. One way to test this hypothesis is in reprogramming from a source to target cell type: we reasoned that the downregulation of factors that are responsive in the source cell type may interfere with source cell type maintenance, manifesting as an enhancement of reprogramming to the target cell type. We tested this possibility by focusing on the induction of induced pluripotent stem cells from fibroblasts, i.e., fibroblast reprogramming.\n",
      "\n",
      "3 To test the hypothesis that P ​ can reveal genes that, if removed, would inhibit 3 ​ on isogenic dermal fibroblasts (GM00942, Coriell,\n",
      "Human: How does the overexpression or suppression of specific transcription factors affect the stability and functionality of reprogrammed cells?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "; , : ) (\n",
      "\n",
      "0 9 8 7 6 5 4 3 2 1\n",
      "\n",
      "; , : ) (\n",
      "\n",
      "0 9 8 7 6 5 4 3 2 1\n",
      "\n",
      "Article\n",
      "\n",
      "https://doi.org/10.1038/s41467-023-37256-8\n",
      "\n",
      "Conserved transcription factors promote cell fate stability and restrict reprogramming potential in differentiated cells\n",
      "\n",
      "Received: 7 August 2021\n",
      "\n",
      "Accepted: 8 March 2023\n",
      "\n",
      "Maria A. Missinato1,6, Sean Murphy2,6, Michaela Lynott Anaïs Kervadec1, Yu-Ling Chang1, Suraj Kannan2, Mafalda Loreti1, Christopher Lee 1, Prashila Amatya1, Hiroshi Tanaka1, Chun-Teng Huang3, 1, Chulan Kwon 2, Peter D. Adams Pier Lorenzo Puri Alessandra Sacco1, Peter Andersen 2 & Alexandre R. Colas\n",
      "\n",
      "1,6, Michael S. Yu1,\n",
      "\n",
      "4, Li Qian 5,\n",
      "\n",
      "1\n",
      "\n",
      "Check for updates\n",
      "\n",
      "Deﬁning the mechanisms safeguarding cell fate\n",
      "\n",
      "reveals the critical role of the TRN in pro- tecting cells against aberrant reprogramming, and demonstrates the vulnerability of donor cell’s TRNs,\n",
      "\n",
      "offering a novel strategy to induce transgene-free trans-differentiations.\n",
      "\n",
      "INTRODUCTION\n",
      "\n",
      "In any given cell, a set of transcription factors (TFs) is ex- pressed and works in concert to govern cellular homeosta- sis and function (1). The Transcriptional Regulatory Net- works (TRNs) in somatic cells are believed to be robust and act to prevent cells from aberrant plasticity (2). However, recent works have revealed that somatic cells in mammals are more plastic than previously thought (3). Particularly, Takahashi and Yamanaka demonstrated this\n",
      "\n",
      "whose expression or activity is cell type- or tissue-specific. In principle, responsiveness could have identified a proper subset of the cell type-specific genes, a superset of the cell type-specific genes, a wholly distinct set of genes, or a partially overlapping subset. We found that responsiveness seemed to identify a superset of the cell type-specific genes, meaning that there were many genes whose expression was responsive but not particularly cell type-specific. When we explicitly tested whether these responsive-but-not-cell type-specific genes could affect maintenance, we found that, for fully half the factors we tested, suppression of the factor led to an increase in reprogramming\n",
      "\n",
      "whether these responsive-but-not-cell type-specific genes could affect maintenance, we found that, for fully half the factors we tested, suppression of the factor led to an increase in reprogramming efficiency. This validation rate suggests that these responsive 3 factors are indeed important for cell type maintenance and not merely falsely identified by P ​. The fact that these non-cell type-specific factors are important for maintenance suggests that cell type maintenance may involve factors that act as general barriers to cell type transformations in addition to factors specific to particular types. (Although it is possible that such factors may only act as barriers in particular cell\n",
      "\n",
      "Each experimental replicate was conducted using different vials of cells and different batches of media and feeder mouse embryonic fibroblasts (MEFs) on different days; knockdown samples are compared against controls in the same batch. Reprogramming rate variability was highest in the third replicate for unknown reasons.\n",
      "\n",
      "Supplementary Figure 12: Evaluating cardiac transdifferentiation after overexpression of iPSC-derived cardiac myocyte-responsive transcription factors in fibroblasts. A) Selection of 7 transcription factors found to be frequently up-regulated after perturbation in iPSC-derived cardiac myocytes (iPSC-CM; see Methods). B) Schematic of experimental workflow:\n",
      "\n",
      "by enhancing source cell type gene expression or repressing reprogramming TF-associated gene expression. How- ever, a systematic evaluation of TFs as fate stabilizers has not been conducted to date, and in this context, we will explore whether such factors might be evolutionarily conserved and play lineage and/or cell type-speciﬁc or independent roles. Mechanistically, we will evaluate how fate-stabilizing TFs binding to the DNA might con- tribute to regulate chromatin accessibility dynamics and tran- scription to oppose the lineage reprogramming process. Finally, we will test whether the targeted inhibition of such fate stabilizers might enhance our ability to use reprogramming TFs to\n",
      "\n",
      "that Drive Neuronal Fate and Reprogramming. Cell Stem . Cell ​ .\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "41. A. Ebrahimi, K. Sevinç, G. Gürhan Sevinç, A. P. Cribbs, M. Philpott, F. Uyulur, T. Morova, J. E. Dunford, S. Göklemez, Ş. Arı, U. Oppermann, T. T. Önder, Bromodomain inhibition of the coactivators CBP/EP300 facilitate cellular reprogramming. 519–528 (2019).\n",
      "\n",
      "Nat. Chem. Biol. ​\n",
      "\n",
      "15\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      ",\n",
      "\n",
      "​\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made\n",
      "\n",
      "available under a\n",
      "\n",
      "CC-BY\n",
      "\n",
      "Transcription factors orchestrate dynamic interplay between genome topology and gene regulation during cell reprogramming\n",
      "\n",
      "Ralph Stadhouders  François Le Dily1,2,3, Javier Quilez  Yasmina Cuartero1,2,3, Jochen Hecht2,6, Guillaume J. Filion1,2, Miguel Beato1,2, Marc A. Marti-Renom  and Thomas Graf\n",
      "\n",
      "1,2,8*, Enrique Vidal\n",
      "\n",
      "1,2, François Serra\n",
      "\n",
      "1,2,3, Bruno Di Stefano\n",
      "\n",
      "1,2,4,  5, Clara Berenguer1,2,\n",
      "\n",
      "1,2, Antonio Gomez1,2, Samuel Collombet\n",
      "\n",
      "1,2*\n",
      "\n",
      "1,2,3,7*\n",
      "\n",
      "Chromosomal architecture is known to influence gene expression, yet its role in controlling cell fate remains poorly understood. Reprogramming of somatic cells into pluripotent stem cells (PSCs) by the transcription factors (TFs) OCT4, SOX2,\n",
      "\n",
      "Hence we reasoned that responsiveness might also identify previously undescribed factors that maintain cellular identity. One way to test this hypothesis is in reprogramming from a source to target cell type: we reasoned that the downregulation of factors that are responsive in the source cell type may interfere with source cell type maintenance, manifesting as an enhancement of reprogramming to the target cell type. We tested this possibility by focusing on the induction of induced pluripotent stem cells from fibroblasts, i.e., fibroblast reprogramming.\n",
      "\n",
      "3 To test the hypothesis that P ​ can reveal genes that, if removed, would inhibit 3 ​ on isogenic dermal fibroblasts (GM00942, Coriell,\n",
      "Human: How does the overexpression or suppression of specific transcription factors affect the stability and functionality of reprogrammed cells?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "; , : ) (\n",
      "\n",
      "0 9 8 7 6 5 4 3 2 1\n",
      "\n",
      "; , : ) (\n",
      "\n",
      "0 9 8 7 6 5 4 3 2 1\n",
      "\n",
      "Article\n",
      "\n",
      "https://doi.org/10.1038/s41467-023-37256-8\n",
      "\n",
      "Conserved transcription factors promote cell fate stability and restrict reprogramming potential in differentiated cells\n",
      "\n",
      "Received: 7 August 2021\n",
      "\n",
      "Accepted: 8 March 2023\n",
      "\n",
      "Maria A. Missinato1,6, Sean Murphy2,6, Michaela Lynott Anaïs Kervadec1, Yu-Ling Chang1, Suraj Kannan2, Mafalda Loreti1, Christopher Lee 1, Prashila Amatya1, Hiroshi Tanaka1, Chun-Teng Huang3, 1, Chulan Kwon 2, Peter D. Adams Pier Lorenzo Puri Alessandra Sacco1, Peter Andersen 2 & Alexandre R. Colas\n",
      "\n",
      "1,6, Michael S. Yu1,\n",
      "\n",
      "4, Li Qian 5,\n",
      "\n",
      "1\n",
      "\n",
      "Check for updates\n",
      "\n",
      "Deﬁning the mechanisms safeguarding cell fate\n",
      "\n",
      "differentiation and maintenance were among the most frequently up-regulated (most responsive). We reasoned that one potential function of responsive genes may be to maintain cellular identity. We identified 3 responsive transcription factors in fibroblasts using P ​ and found that suppressing their expression led to enhanced reprogramming efficiency. We propose that responsiveness to perturbations is a property of factors that help maintain cellular identity.\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "Introduction\n",
      "\n",
      "Cells of a particular cell type likely maintain many of their specific behavioral properties across a range of conditions. The maintenance of these properties can be the result of the activity of certain\n",
      "\n",
      "whose expression or activity is cell type- or tissue-specific. In principle, responsiveness could have identified a proper subset of the cell type-specific genes, a superset of the cell type-specific genes, a wholly distinct set of genes, or a partially overlapping subset. We found that responsiveness seemed to identify a superset of the cell type-specific genes, meaning that there were many genes whose expression was responsive but not particularly cell type-specific. When we explicitly tested whether these responsive-but-not-cell type-specific genes could affect maintenance, we found that, for fully half the factors we tested, suppression of the factor led to an increase in reprogramming\n",
      "\n",
      "whether these responsive-but-not-cell type-specific genes could affect maintenance, we found that, for fully half the factors we tested, suppression of the factor led to an increase in reprogramming efficiency. This validation rate suggests that these responsive 3 factors are indeed important for cell type maintenance and not merely falsely identified by P ​. The fact that these non-cell type-specific factors are important for maintenance suggests that cell type maintenance may involve factors that act as general barriers to cell type transformations in addition to factors specific to particular types. (Although it is possible that such factors may only act as barriers in particular cell\n",
      "\n",
      "Each experimental replicate was conducted using different vials of cells and different batches of media and feeder mouse embryonic fibroblasts (MEFs) on different days; knockdown samples are compared against controls in the same batch. Reprogramming rate variability was highest in the third replicate for unknown reasons.\n",
      "\n",
      "Supplementary Figure 12: Evaluating cardiac transdifferentiation after overexpression of iPSC-derived cardiac myocyte-responsive transcription factors in fibroblasts. A) Selection of 7 transcription factors found to be frequently up-regulated after perturbation in iPSC-derived cardiac myocytes (iPSC-CM; see Methods). B) Schematic of experimental workflow:\n",
      "\n",
      "by enhancing source cell type gene expression or repressing reprogramming TF-associated gene expression. How- ever, a systematic evaluation of TFs as fate stabilizers has not been conducted to date, and in this context, we will explore whether such factors might be evolutionarily conserved and play lineage and/or cell type-speciﬁc or independent roles. Mechanistically, we will evaluate how fate-stabilizing TFs binding to the DNA might con- tribute to regulate chromatin accessibility dynamics and tran- scription to oppose the lineage reprogramming process. Finally, we will test whether the targeted inhibition of such fate stabilizers might enhance our ability to use reprogramming TFs to\n",
      "\n",
      "that Drive Neuronal Fate and Reprogramming. Cell Stem . Cell ​ .\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "41. A. Ebrahimi, K. Sevinç, G. Gürhan Sevinç, A. P. Cribbs, M. Philpott, F. Uyulur, T. Morova, J. E. Dunford, S. Göklemez, Ş. Arı, U. Oppermann, T. T. Önder, Bromodomain inhibition of the coactivators CBP/EP300 facilitate cellular reprogramming. 519–528 (2019).\n",
      "\n",
      "Nat. Chem. Biol. ​\n",
      "\n",
      "15\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      ",\n",
      "\n",
      "​\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made\n",
      "\n",
      "available under a\n",
      "\n",
      "CC-BY\n",
      "\n",
      "Transcription factors orchestrate dynamic interplay between genome topology and gene regulation during cell reprogramming\n",
      "\n",
      "Ralph Stadhouders  François Le Dily1,2,3, Javier Quilez  Yasmina Cuartero1,2,3, Jochen Hecht2,6, Guillaume J. Filion1,2, Miguel Beato1,2, Marc A. Marti-Renom  and Thomas Graf\n",
      "\n",
      "1,2,8*, Enrique Vidal\n",
      "\n",
      "1,2, François Serra\n",
      "\n",
      "1,2,3, Bruno Di Stefano\n",
      "\n",
      "1,2,4,  5, Clara Berenguer1,2,\n",
      "\n",
      "1,2, Antonio Gomez1,2, Samuel Collombet\n",
      "\n",
      "1,2*\n",
      "\n",
      "1,2,3,7*\n",
      "\n",
      "Chromosomal architecture is known to influence gene expression, yet its role in controlling cell fate remains poorly understood. Reprogramming of somatic cells into pluripotent stem cells (PSCs) by the transcription factors (TFs) OCT4, SOX2,\n",
      "\n",
      "Hence we reasoned that responsiveness might also identify previously undescribed factors that maintain cellular identity. One way to test this hypothesis is in reprogramming from a source to target cell type: we reasoned that the downregulation of factors that are responsive in the source cell type may interfere with source cell type maintenance, manifesting as an enhancement of reprogramming to the target cell type. We tested this possibility by focusing on the induction of induced pluripotent stem cells from fibroblasts, i.e., fibroblast reprogramming.\n",
      "\n",
      "3 To test the hypothesis that P ​ can reveal genes that, if removed, would inhibit 3 ​ on isogenic dermal fibroblasts (GM00942, Coriell,\n",
      "Human: How does the overexpression or suppression of specific transcription factors affect the stability and functionality of reprogrammed cells?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "; , : ) (\n",
      "\n",
      "0 9 8 7 6 5 4 3 2 1\n",
      "\n",
      "; , : ) (\n",
      "\n",
      "0 9 8 7 6 5 4 3 2 1\n",
      "\n",
      "Article\n",
      "\n",
      "https://doi.org/10.1038/s41467-023-37256-8\n",
      "\n",
      "Conserved transcription factors promote cell fate stability and restrict reprogramming potential in differentiated cells\n",
      "\n",
      "Received: 7 August 2021\n",
      "\n",
      "Accepted: 8 March 2023\n",
      "\n",
      "Maria A. Missinato1,6, Sean Murphy2,6, Michaela Lynott Anaïs Kervadec1, Yu-Ling Chang1, Suraj Kannan2, Mafalda Loreti1, Christopher Lee 1, Prashila Amatya1, Hiroshi Tanaka1, Chun-Teng Huang3, 1, Chulan Kwon 2, Peter D. Adams Pier Lorenzo Puri Alessandra Sacco1, Peter Andersen 2 & Alexandre R. Colas\n",
      "\n",
      "1,6, Michael S. Yu1,\n",
      "\n",
      "4, Li Qian 5,\n",
      "\n",
      "1\n",
      "\n",
      "Check for updates\n",
      "\n",
      "Deﬁning the mechanisms safeguarding cell fate\n",
      "\n",
      "differentiation and maintenance were among the most frequently up-regulated (most responsive). We reasoned that one potential function of responsive genes may be to maintain cellular identity. We identified 3 responsive transcription factors in fibroblasts using P ​ and found that suppressing their expression led to enhanced reprogramming efficiency. We propose that responsiveness to perturbations is a property of factors that help maintain cellular identity.\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "Introduction\n",
      "\n",
      "Cells of a particular cell type likely maintain many of their specific behavioral properties across a range of conditions. The maintenance of these properties can be the result of the activity of certain\n",
      "\n",
      "reveals the critical role of the TRN in pro- tecting cells against aberrant reprogramming, and demonstrates the vulnerability of donor cell’s TRNs,\n",
      "\n",
      "offering a novel strategy to induce transgene-free trans-differentiations.\n",
      "\n",
      "INTRODUCTION\n",
      "\n",
      "In any given cell, a set of transcription factors (TFs) is ex- pressed and works in concert to govern cellular homeosta- sis and function (1). The Transcriptional Regulatory Net- works (TRNs) in somatic cells are believed to be robust and act to prevent cells from aberrant plasticity (2). However, recent works have revealed that somatic cells in mammals are more plastic than previously thought (3). Particularly, Takahashi and Yamanaka demonstrated this\n",
      "\n",
      "whether these responsive-but-not-cell type-specific genes could affect maintenance, we found that, for fully half the factors we tested, suppression of the factor led to an increase in reprogramming efficiency. This validation rate suggests that these responsive 3 factors are indeed important for cell type maintenance and not merely falsely identified by P ​. The fact that these non-cell type-specific factors are important for maintenance suggests that cell type maintenance may involve factors that act as general barriers to cell type transformations in addition to factors specific to particular types. (Although it is possible that such factors may only act as barriers in particular cell\n",
      "\n",
      "Each experimental replicate was conducted using different vials of cells and different batches of media and feeder mouse embryonic fibroblasts (MEFs) on different days; knockdown samples are compared against controls in the same batch. Reprogramming rate variability was highest in the third replicate for unknown reasons.\n",
      "\n",
      "Supplementary Figure 12: Evaluating cardiac transdifferentiation after overexpression of iPSC-derived cardiac myocyte-responsive transcription factors in fibroblasts. A) Selection of 7 transcription factors found to be frequently up-regulated after perturbation in iPSC-derived cardiac myocytes (iPSC-CM; see Methods). B) Schematic of experimental workflow:\n",
      "\n",
      "by enhancing source cell type gene expression or repressing reprogramming TF-associated gene expression. How- ever, a systematic evaluation of TFs as fate stabilizers has not been conducted to date, and in this context, we will explore whether such factors might be evolutionarily conserved and play lineage and/or cell type-speciﬁc or independent roles. Mechanistically, we will evaluate how fate-stabilizing TFs binding to the DNA might con- tribute to regulate chromatin accessibility dynamics and tran- scription to oppose the lineage reprogramming process. Finally, we will test whether the targeted inhibition of such fate stabilizers might enhance our ability to use reprogramming TFs to\n",
      "\n",
      "that Drive Neuronal Fate and Reprogramming. Cell Stem . Cell ​ .\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "41. A. Ebrahimi, K. Sevinç, G. Gürhan Sevinç, A. P. Cribbs, M. Philpott, F. Uyulur, T. Morova, J. E. Dunford, S. Göklemez, Ş. Arı, U. Oppermann, T. T. Önder, Bromodomain inhibition of the coactivators CBP/EP300 facilitate cellular reprogramming. 519–528 (2019).\n",
      "\n",
      "Nat. Chem. Biol. ​\n",
      "\n",
      "15\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      ",\n",
      "\n",
      "​\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made\n",
      "\n",
      "available under a\n",
      "\n",
      "CC-BY\n",
      "\n",
      "Transcription factors orchestrate dynamic interplay between genome topology and gene regulation during cell reprogramming\n",
      "\n",
      "Ralph Stadhouders  François Le Dily1,2,3, Javier Quilez  Yasmina Cuartero1,2,3, Jochen Hecht2,6, Guillaume J. Filion1,2, Miguel Beato1,2, Marc A. Marti-Renom  and Thomas Graf\n",
      "\n",
      "1,2,8*, Enrique Vidal\n",
      "\n",
      "1,2, François Serra\n",
      "\n",
      "1,2,3, Bruno Di Stefano\n",
      "\n",
      "1,2,4,  5, Clara Berenguer1,2,\n",
      "\n",
      "1,2, Antonio Gomez1,2, Samuel Collombet\n",
      "\n",
      "1,2*\n",
      "\n",
      "1,2,3,7*\n",
      "\n",
      "Chromosomal architecture is known to influence gene expression, yet its role in controlling cell fate remains poorly understood. Reprogramming of somatic cells into pluripotent stem cells (PSCs) by the transcription factors (TFs) OCT4, SOX2,\n",
      "\n",
      "Hence we reasoned that responsiveness might also identify previously undescribed factors that maintain cellular identity. One way to test this hypothesis is in reprogramming from a source to target cell type: we reasoned that the downregulation of factors that are responsive in the source cell type may interfere with source cell type maintenance, manifesting as an enhancement of reprogramming to the target cell type. We tested this possibility by focusing on the induction of induced pluripotent stem cells from fibroblasts, i.e., fibroblast reprogramming.\n",
      "\n",
      "3 To test the hypothesis that P ​ can reveal genes that, if removed, would inhibit 3 ​ on isogenic dermal fibroblasts (GM00942, Coriell,\n",
      "Human: How does the overexpression or suppression of specific transcription factors affect the stability and functionality of reprogrammed cells?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "; , : ) (\n",
      "\n",
      "0 9 8 7 6 5 4 3 2 1\n",
      "\n",
      "; , : ) (\n",
      "\n",
      "0 9 8 7 6 5 4 3 2 1\n",
      "\n",
      "Article\n",
      "\n",
      "https://doi.org/10.1038/s41467-023-37256-8\n",
      "\n",
      "Conserved transcription factors promote cell fate stability and restrict reprogramming potential in differentiated cells\n",
      "\n",
      "Received: 7 August 2021\n",
      "\n",
      "Accepted: 8 March 2023\n",
      "\n",
      "Maria A. Missinato1,6, Sean Murphy2,6, Michaela Lynott Anaïs Kervadec1, Yu-Ling Chang1, Suraj Kannan2, Mafalda Loreti1, Christopher Lee 1, Prashila Amatya1, Hiroshi Tanaka1, Chun-Teng Huang3, 1, Chulan Kwon 2, Peter D. Adams Pier Lorenzo Puri Alessandra Sacco1, Peter Andersen 2 & Alexandre R. Colas\n",
      "\n",
      "1,6, Michael S. Yu1,\n",
      "\n",
      "4, Li Qian 5,\n",
      "\n",
      "1\n",
      "\n",
      "Check for updates\n",
      "\n",
      "Deﬁning the mechanisms safeguarding cell fate\n",
      "\n",
      "differentiation and maintenance were among the most frequently up-regulated (most responsive). We reasoned that one potential function of responsive genes may be to maintain cellular identity. We identified 3 responsive transcription factors in fibroblasts using P ​ and found that suppressing their expression led to enhanced reprogramming efficiency. We propose that responsiveness to perturbations is a property of factors that help maintain cellular identity.\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "Introduction\n",
      "\n",
      "Cells of a particular cell type likely maintain many of their specific behavioral properties across a range of conditions. The maintenance of these properties can be the result of the activity of certain\n",
      "\n",
      "reveals the critical role of the TRN in pro- tecting cells against aberrant reprogramming, and demonstrates the vulnerability of donor cell’s TRNs,\n",
      "\n",
      "offering a novel strategy to induce transgene-free trans-differentiations.\n",
      "\n",
      "INTRODUCTION\n",
      "\n",
      "In any given cell, a set of transcription factors (TFs) is ex- pressed and works in concert to govern cellular homeosta- sis and function (1). The Transcriptional Regulatory Net- works (TRNs) in somatic cells are believed to be robust and act to prevent cells from aberrant plasticity (2). However, recent works have revealed that somatic cells in mammals are more plastic than previously thought (3). Particularly, Takahashi and Yamanaka demonstrated this\n",
      "\n",
      "whose expression or activity is cell type- or tissue-specific. In principle, responsiveness could have identified a proper subset of the cell type-specific genes, a superset of the cell type-specific genes, a wholly distinct set of genes, or a partially overlapping subset. We found that responsiveness seemed to identify a superset of the cell type-specific genes, meaning that there were many genes whose expression was responsive but not particularly cell type-specific. When we explicitly tested whether these responsive-but-not-cell type-specific genes could affect maintenance, we found that, for fully half the factors we tested, suppression of the factor led to an increase in reprogramming\n",
      "\n",
      "Each experimental replicate was conducted using different vials of cells and different batches of media and feeder mouse embryonic fibroblasts (MEFs) on different days; knockdown samples are compared against controls in the same batch. Reprogramming rate variability was highest in the third replicate for unknown reasons.\n",
      "\n",
      "Supplementary Figure 12: Evaluating cardiac transdifferentiation after overexpression of iPSC-derived cardiac myocyte-responsive transcription factors in fibroblasts. A) Selection of 7 transcription factors found to be frequently up-regulated after perturbation in iPSC-derived cardiac myocytes (iPSC-CM; see Methods). B) Schematic of experimental workflow:\n",
      "\n",
      "by enhancing source cell type gene expression or repressing reprogramming TF-associated gene expression. How- ever, a systematic evaluation of TFs as fate stabilizers has not been conducted to date, and in this context, we will explore whether such factors might be evolutionarily conserved and play lineage and/or cell type-speciﬁc or independent roles. Mechanistically, we will evaluate how fate-stabilizing TFs binding to the DNA might con- tribute to regulate chromatin accessibility dynamics and tran- scription to oppose the lineage reprogramming process. Finally, we will test whether the targeted inhibition of such fate stabilizers might enhance our ability to use reprogramming TFs to\n",
      "\n",
      "that Drive Neuronal Fate and Reprogramming. Cell Stem . Cell ​ .\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "41. A. Ebrahimi, K. Sevinç, G. Gürhan Sevinç, A. P. Cribbs, M. Philpott, F. Uyulur, T. Morova, J. E. Dunford, S. Göklemez, Ş. Arı, U. Oppermann, T. T. Önder, Bromodomain inhibition of the coactivators CBP/EP300 facilitate cellular reprogramming. 519–528 (2019).\n",
      "\n",
      "Nat. Chem. Biol. ​\n",
      "\n",
      "15\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      ",\n",
      "\n",
      "​\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made\n",
      "\n",
      "available under a\n",
      "\n",
      "CC-BY\n",
      "\n",
      "Transcription factors orchestrate dynamic interplay between genome topology and gene regulation during cell reprogramming\n",
      "\n",
      "Ralph Stadhouders  François Le Dily1,2,3, Javier Quilez  Yasmina Cuartero1,2,3, Jochen Hecht2,6, Guillaume J. Filion1,2, Miguel Beato1,2, Marc A. Marti-Renom  and Thomas Graf\n",
      "\n",
      "1,2,8*, Enrique Vidal\n",
      "\n",
      "1,2, François Serra\n",
      "\n",
      "1,2,3, Bruno Di Stefano\n",
      "\n",
      "1,2,4,  5, Clara Berenguer1,2,\n",
      "\n",
      "1,2, Antonio Gomez1,2, Samuel Collombet\n",
      "\n",
      "1,2*\n",
      "\n",
      "1,2,3,7*\n",
      "\n",
      "Chromosomal architecture is known to influence gene expression, yet its role in controlling cell fate remains poorly understood. Reprogramming of somatic cells into pluripotent stem cells (PSCs) by the transcription factors (TFs) OCT4, SOX2,\n",
      "\n",
      "Hence we reasoned that responsiveness might also identify previously undescribed factors that maintain cellular identity. One way to test this hypothesis is in reprogramming from a source to target cell type: we reasoned that the downregulation of factors that are responsive in the source cell type may interfere with source cell type maintenance, manifesting as an enhancement of reprogramming to the target cell type. We tested this possibility by focusing on the induction of induced pluripotent stem cells from fibroblasts, i.e., fibroblast reprogramming.\n",
      "\n",
      "3 To test the hypothesis that P ​ can reveal genes that, if removed, would inhibit 3 ​ on isogenic dermal fibroblasts (GM00942, Coriell,\n",
      "Human: How does the overexpression or suppression of specific transcription factors affect the stability and functionality of reprogrammed cells?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "; , : ) (\n",
      "\n",
      "0 9 8 7 6 5 4 3 2 1\n",
      "\n",
      "; , : ) (\n",
      "\n",
      "0 9 8 7 6 5 4 3 2 1\n",
      "\n",
      "Article\n",
      "\n",
      "https://doi.org/10.1038/s41467-023-37256-8\n",
      "\n",
      "Conserved transcription factors promote cell fate stability and restrict reprogramming potential in differentiated cells\n",
      "\n",
      "Received: 7 August 2021\n",
      "\n",
      "Accepted: 8 March 2023\n",
      "\n",
      "Maria A. Missinato1,6, Sean Murphy2,6, Michaela Lynott Anaïs Kervadec1, Yu-Ling Chang1, Suraj Kannan2, Mafalda Loreti1, Christopher Lee 1, Prashila Amatya1, Hiroshi Tanaka1, Chun-Teng Huang3, 1, Chulan Kwon 2, Peter D. Adams Pier Lorenzo Puri Alessandra Sacco1, Peter Andersen 2 & Alexandre R. Colas\n",
      "\n",
      "1,6, Michael S. Yu1,\n",
      "\n",
      "4, Li Qian 5,\n",
      "\n",
      "1\n",
      "\n",
      "Check for updates\n",
      "\n",
      "Deﬁning the mechanisms safeguarding cell fate\n",
      "\n",
      "differentiation and maintenance were among the most frequently up-regulated (most responsive). We reasoned that one potential function of responsive genes may be to maintain cellular identity. We identified 3 responsive transcription factors in fibroblasts using P ​ and found that suppressing their expression led to enhanced reprogramming efficiency. We propose that responsiveness to perturbations is a property of factors that help maintain cellular identity.\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "Introduction\n",
      "\n",
      "Cells of a particular cell type likely maintain many of their specific behavioral properties across a range of conditions. The maintenance of these properties can be the result of the activity of certain\n",
      "\n",
      "reveals the critical role of the TRN in pro- tecting cells against aberrant reprogramming, and demonstrates the vulnerability of donor cell’s TRNs,\n",
      "\n",
      "offering a novel strategy to induce transgene-free trans-differentiations.\n",
      "\n",
      "INTRODUCTION\n",
      "\n",
      "In any given cell, a set of transcription factors (TFs) is ex- pressed and works in concert to govern cellular homeosta- sis and function (1). The Transcriptional Regulatory Net- works (TRNs) in somatic cells are believed to be robust and act to prevent cells from aberrant plasticity (2). However, recent works have revealed that somatic cells in mammals are more plastic than previously thought (3). Particularly, Takahashi and Yamanaka demonstrated this\n",
      "\n",
      "whose expression or activity is cell type- or tissue-specific. In principle, responsiveness could have identified a proper subset of the cell type-specific genes, a superset of the cell type-specific genes, a wholly distinct set of genes, or a partially overlapping subset. We found that responsiveness seemed to identify a superset of the cell type-specific genes, meaning that there were many genes whose expression was responsive but not particularly cell type-specific. When we explicitly tested whether these responsive-but-not-cell type-specific genes could affect maintenance, we found that, for fully half the factors we tested, suppression of the factor led to an increase in reprogramming\n",
      "\n",
      "whether these responsive-but-not-cell type-specific genes could affect maintenance, we found that, for fully half the factors we tested, suppression of the factor led to an increase in reprogramming efficiency. This validation rate suggests that these responsive 3 factors are indeed important for cell type maintenance and not merely falsely identified by P ​. The fact that these non-cell type-specific factors are important for maintenance suggests that cell type maintenance may involve factors that act as general barriers to cell type transformations in addition to factors specific to particular types. (Although it is possible that such factors may only act as barriers in particular cell\n",
      "\n",
      "by enhancing source cell type gene expression or repressing reprogramming TF-associated gene expression. How- ever, a systematic evaluation of TFs as fate stabilizers has not been conducted to date, and in this context, we will explore whether such factors might be evolutionarily conserved and play lineage and/or cell type-speciﬁc or independent roles. Mechanistically, we will evaluate how fate-stabilizing TFs binding to the DNA might con- tribute to regulate chromatin accessibility dynamics and tran- scription to oppose the lineage reprogramming process. Finally, we will test whether the targeted inhibition of such fate stabilizers might enhance our ability to use reprogramming TFs to\n",
      "\n",
      "that Drive Neuronal Fate and Reprogramming. Cell Stem . Cell ​ .\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "41. A. Ebrahimi, K. Sevinç, G. Gürhan Sevinç, A. P. Cribbs, M. Philpott, F. Uyulur, T. Morova, J. E. Dunford, S. Göklemez, Ş. Arı, U. Oppermann, T. T. Önder, Bromodomain inhibition of the coactivators CBP/EP300 facilitate cellular reprogramming. 519–528 (2019).\n",
      "\n",
      "Nat. Chem. Biol. ​\n",
      "\n",
      "15\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      ",\n",
      "\n",
      "​\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made\n",
      "\n",
      "available under a\n",
      "\n",
      "CC-BY\n",
      "\n",
      "Transcription factors orchestrate dynamic interplay between genome topology and gene regulation during cell reprogramming\n",
      "\n",
      "Ralph Stadhouders  François Le Dily1,2,3, Javier Quilez  Yasmina Cuartero1,2,3, Jochen Hecht2,6, Guillaume J. Filion1,2, Miguel Beato1,2, Marc A. Marti-Renom  and Thomas Graf\n",
      "\n",
      "1,2,8*, Enrique Vidal\n",
      "\n",
      "1,2, François Serra\n",
      "\n",
      "1,2,3, Bruno Di Stefano\n",
      "\n",
      "1,2,4,  5, Clara Berenguer1,2,\n",
      "\n",
      "1,2, Antonio Gomez1,2, Samuel Collombet\n",
      "\n",
      "1,2*\n",
      "\n",
      "1,2,3,7*\n",
      "\n",
      "Chromosomal architecture is known to influence gene expression, yet its role in controlling cell fate remains poorly understood. Reprogramming of somatic cells into pluripotent stem cells (PSCs) by the transcription factors (TFs) OCT4, SOX2,\n",
      "\n",
      "Hence we reasoned that responsiveness might also identify previously undescribed factors that maintain cellular identity. One way to test this hypothesis is in reprogramming from a source to target cell type: we reasoned that the downregulation of factors that are responsive in the source cell type may interfere with source cell type maintenance, manifesting as an enhancement of reprogramming to the target cell type. We tested this possibility by focusing on the induction of induced pluripotent stem cells from fibroblasts, i.e., fibroblast reprogramming.\n",
      "\n",
      "3 To test the hypothesis that P ​ can reveal genes that, if removed, would inhibit 3 ​ on isogenic dermal fibroblasts (GM00942, Coriell,\n",
      "Human: How does the overexpression or suppression of specific transcription factors affect the stability and functionality of reprogrammed cells?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "; , : ) (\n",
      "\n",
      "0 9 8 7 6 5 4 3 2 1\n",
      "\n",
      "; , : ) (\n",
      "\n",
      "0 9 8 7 6 5 4 3 2 1\n",
      "\n",
      "Article\n",
      "\n",
      "https://doi.org/10.1038/s41467-023-37256-8\n",
      "\n",
      "Conserved transcription factors promote cell fate stability and restrict reprogramming potential in differentiated cells\n",
      "\n",
      "Received: 7 August 2021\n",
      "\n",
      "Accepted: 8 March 2023\n",
      "\n",
      "Maria A. Missinato1,6, Sean Murphy2,6, Michaela Lynott Anaïs Kervadec1, Yu-Ling Chang1, Suraj Kannan2, Mafalda Loreti1, Christopher Lee 1, Prashila Amatya1, Hiroshi Tanaka1, Chun-Teng Huang3, 1, Chulan Kwon 2, Peter D. Adams Pier Lorenzo Puri Alessandra Sacco1, Peter Andersen 2 & Alexandre R. Colas\n",
      "\n",
      "1,6, Michael S. Yu1,\n",
      "\n",
      "4, Li Qian 5,\n",
      "\n",
      "1\n",
      "\n",
      "Check for updates\n",
      "\n",
      "Deﬁning the mechanisms safeguarding cell fate\n",
      "\n",
      "differentiation and maintenance were among the most frequently up-regulated (most responsive). We reasoned that one potential function of responsive genes may be to maintain cellular identity. We identified 3 responsive transcription factors in fibroblasts using P ​ and found that suppressing their expression led to enhanced reprogramming efficiency. We propose that responsiveness to perturbations is a property of factors that help maintain cellular identity.\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "Introduction\n",
      "\n",
      "Cells of a particular cell type likely maintain many of their specific behavioral properties across a range of conditions. The maintenance of these properties can be the result of the activity of certain\n",
      "\n",
      "reveals the critical role of the TRN in pro- tecting cells against aberrant reprogramming, and demonstrates the vulnerability of donor cell’s TRNs,\n",
      "\n",
      "offering a novel strategy to induce transgene-free trans-differentiations.\n",
      "\n",
      "INTRODUCTION\n",
      "\n",
      "In any given cell, a set of transcription factors (TFs) is ex- pressed and works in concert to govern cellular homeosta- sis and function (1). The Transcriptional Regulatory Net- works (TRNs) in somatic cells are believed to be robust and act to prevent cells from aberrant plasticity (2). However, recent works have revealed that somatic cells in mammals are more plastic than previously thought (3). Particularly, Takahashi and Yamanaka demonstrated this\n",
      "\n",
      "whose expression or activity is cell type- or tissue-specific. In principle, responsiveness could have identified a proper subset of the cell type-specific genes, a superset of the cell type-specific genes, a wholly distinct set of genes, or a partially overlapping subset. We found that responsiveness seemed to identify a superset of the cell type-specific genes, meaning that there were many genes whose expression was responsive but not particularly cell type-specific. When we explicitly tested whether these responsive-but-not-cell type-specific genes could affect maintenance, we found that, for fully half the factors we tested, suppression of the factor led to an increase in reprogramming\n",
      "\n",
      "whether these responsive-but-not-cell type-specific genes could affect maintenance, we found that, for fully half the factors we tested, suppression of the factor led to an increase in reprogramming efficiency. This validation rate suggests that these responsive 3 factors are indeed important for cell type maintenance and not merely falsely identified by P ​. The fact that these non-cell type-specific factors are important for maintenance suggests that cell type maintenance may involve factors that act as general barriers to cell type transformations in addition to factors specific to particular types. (Although it is possible that such factors may only act as barriers in particular cell\n",
      "\n",
      "Each experimental replicate was conducted using different vials of cells and different batches of media and feeder mouse embryonic fibroblasts (MEFs) on different days; knockdown samples are compared against controls in the same batch. Reprogramming rate variability was highest in the third replicate for unknown reasons.\n",
      "\n",
      "Supplementary Figure 12: Evaluating cardiac transdifferentiation after overexpression of iPSC-derived cardiac myocyte-responsive transcription factors in fibroblasts. A) Selection of 7 transcription factors found to be frequently up-regulated after perturbation in iPSC-derived cardiac myocytes (iPSC-CM; see Methods). B) Schematic of experimental workflow:\n",
      "\n",
      "that Drive Neuronal Fate and Reprogramming. Cell Stem . Cell ​ .\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "41. A. Ebrahimi, K. Sevinç, G. Gürhan Sevinç, A. P. Cribbs, M. Philpott, F. Uyulur, T. Morova, J. E. Dunford, S. Göklemez, Ş. Arı, U. Oppermann, T. T. Önder, Bromodomain inhibition of the coactivators CBP/EP300 facilitate cellular reprogramming. 519–528 (2019).\n",
      "\n",
      "Nat. Chem. Biol. ​\n",
      "\n",
      "15\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      ",\n",
      "\n",
      "​\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made\n",
      "\n",
      "available under a\n",
      "\n",
      "CC-BY\n",
      "\n",
      "Transcription factors orchestrate dynamic interplay between genome topology and gene regulation during cell reprogramming\n",
      "\n",
      "Ralph Stadhouders  François Le Dily1,2,3, Javier Quilez  Yasmina Cuartero1,2,3, Jochen Hecht2,6, Guillaume J. Filion1,2, Miguel Beato1,2, Marc A. Marti-Renom  and Thomas Graf\n",
      "\n",
      "1,2,8*, Enrique Vidal\n",
      "\n",
      "1,2, François Serra\n",
      "\n",
      "1,2,3, Bruno Di Stefano\n",
      "\n",
      "1,2,4,  5, Clara Berenguer1,2,\n",
      "\n",
      "1,2, Antonio Gomez1,2, Samuel Collombet\n",
      "\n",
      "1,2*\n",
      "\n",
      "1,2,3,7*\n",
      "\n",
      "Chromosomal architecture is known to influence gene expression, yet its role in controlling cell fate remains poorly understood. Reprogramming of somatic cells into pluripotent stem cells (PSCs) by the transcription factors (TFs) OCT4, SOX2,\n",
      "\n",
      "Hence we reasoned that responsiveness might also identify previously undescribed factors that maintain cellular identity. One way to test this hypothesis is in reprogramming from a source to target cell type: we reasoned that the downregulation of factors that are responsive in the source cell type may interfere with source cell type maintenance, manifesting as an enhancement of reprogramming to the target cell type. We tested this possibility by focusing on the induction of induced pluripotent stem cells from fibroblasts, i.e., fibroblast reprogramming.\n",
      "\n",
      "3 To test the hypothesis that P ​ can reveal genes that, if removed, would inhibit 3 ​ on isogenic dermal fibroblasts (GM00942, Coriell,\n",
      "Human: How does the overexpression or suppression of specific transcription factors affect the stability and functionality of reprogrammed cells?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "; , : ) (\n",
      "\n",
      "0 9 8 7 6 5 4 3 2 1\n",
      "\n",
      "; , : ) (\n",
      "\n",
      "0 9 8 7 6 5 4 3 2 1\n",
      "\n",
      "Article\n",
      "\n",
      "https://doi.org/10.1038/s41467-023-37256-8\n",
      "\n",
      "Conserved transcription factors promote cell fate stability and restrict reprogramming potential in differentiated cells\n",
      "\n",
      "Received: 7 August 2021\n",
      "\n",
      "Accepted: 8 March 2023\n",
      "\n",
      "Maria A. Missinato1,6, Sean Murphy2,6, Michaela Lynott Anaïs Kervadec1, Yu-Ling Chang1, Suraj Kannan2, Mafalda Loreti1, Christopher Lee 1, Prashila Amatya1, Hiroshi Tanaka1, Chun-Teng Huang3, 1, Chulan Kwon 2, Peter D. Adams Pier Lorenzo Puri Alessandra Sacco1, Peter Andersen 2 & Alexandre R. Colas\n",
      "\n",
      "1,6, Michael S. Yu1,\n",
      "\n",
      "4, Li Qian 5,\n",
      "\n",
      "1\n",
      "\n",
      "Check for updates\n",
      "\n",
      "Deﬁning the mechanisms safeguarding cell fate\n",
      "\n",
      "differentiation and maintenance were among the most frequently up-regulated (most responsive). We reasoned that one potential function of responsive genes may be to maintain cellular identity. We identified 3 responsive transcription factors in fibroblasts using P ​ and found that suppressing their expression led to enhanced reprogramming efficiency. We propose that responsiveness to perturbations is a property of factors that help maintain cellular identity.\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "Introduction\n",
      "\n",
      "Cells of a particular cell type likely maintain many of their specific behavioral properties across a range of conditions. The maintenance of these properties can be the result of the activity of certain\n",
      "\n",
      "reveals the critical role of the TRN in pro- tecting cells against aberrant reprogramming, and demonstrates the vulnerability of donor cell’s TRNs,\n",
      "\n",
      "offering a novel strategy to induce transgene-free trans-differentiations.\n",
      "\n",
      "INTRODUCTION\n",
      "\n",
      "In any given cell, a set of transcription factors (TFs) is ex- pressed and works in concert to govern cellular homeosta- sis and function (1). The Transcriptional Regulatory Net- works (TRNs) in somatic cells are believed to be robust and act to prevent cells from aberrant plasticity (2). However, recent works have revealed that somatic cells in mammals are more plastic than previously thought (3). Particularly, Takahashi and Yamanaka demonstrated this\n",
      "\n",
      "whose expression or activity is cell type- or tissue-specific. In principle, responsiveness could have identified a proper subset of the cell type-specific genes, a superset of the cell type-specific genes, a wholly distinct set of genes, or a partially overlapping subset. We found that responsiveness seemed to identify a superset of the cell type-specific genes, meaning that there were many genes whose expression was responsive but not particularly cell type-specific. When we explicitly tested whether these responsive-but-not-cell type-specific genes could affect maintenance, we found that, for fully half the factors we tested, suppression of the factor led to an increase in reprogramming\n",
      "\n",
      "whether these responsive-but-not-cell type-specific genes could affect maintenance, we found that, for fully half the factors we tested, suppression of the factor led to an increase in reprogramming efficiency. This validation rate suggests that these responsive 3 factors are indeed important for cell type maintenance and not merely falsely identified by P ​. The fact that these non-cell type-specific factors are important for maintenance suggests that cell type maintenance may involve factors that act as general barriers to cell type transformations in addition to factors specific to particular types. (Although it is possible that such factors may only act as barriers in particular cell\n",
      "\n",
      "Each experimental replicate was conducted using different vials of cells and different batches of media and feeder mouse embryonic fibroblasts (MEFs) on different days; knockdown samples are compared against controls in the same batch. Reprogramming rate variability was highest in the third replicate for unknown reasons.\n",
      "\n",
      "Supplementary Figure 12: Evaluating cardiac transdifferentiation after overexpression of iPSC-derived cardiac myocyte-responsive transcription factors in fibroblasts. A) Selection of 7 transcription factors found to be frequently up-regulated after perturbation in iPSC-derived cardiac myocytes (iPSC-CM; see Methods). B) Schematic of experimental workflow:\n",
      "\n",
      "by enhancing source cell type gene expression or repressing reprogramming TF-associated gene expression. How- ever, a systematic evaluation of TFs as fate stabilizers has not been conducted to date, and in this context, we will explore whether such factors might be evolutionarily conserved and play lineage and/or cell type-speciﬁc or independent roles. Mechanistically, we will evaluate how fate-stabilizing TFs binding to the DNA might con- tribute to regulate chromatin accessibility dynamics and tran- scription to oppose the lineage reprogramming process. Finally, we will test whether the targeted inhibition of such fate stabilizers might enhance our ability to use reprogramming TFs to\n",
      "\n",
      "Transcription factors orchestrate dynamic interplay between genome topology and gene regulation during cell reprogramming\n",
      "\n",
      "Ralph Stadhouders  François Le Dily1,2,3, Javier Quilez  Yasmina Cuartero1,2,3, Jochen Hecht2,6, Guillaume J. Filion1,2, Miguel Beato1,2, Marc A. Marti-Renom  and Thomas Graf\n",
      "\n",
      "1,2,8*, Enrique Vidal\n",
      "\n",
      "1,2, François Serra\n",
      "\n",
      "1,2,3, Bruno Di Stefano\n",
      "\n",
      "1,2,4,  5, Clara Berenguer1,2,\n",
      "\n",
      "1,2, Antonio Gomez1,2, Samuel Collombet\n",
      "\n",
      "1,2*\n",
      "\n",
      "1,2,3,7*\n",
      "\n",
      "Chromosomal architecture is known to influence gene expression, yet its role in controlling cell fate remains poorly understood. Reprogramming of somatic cells into pluripotent stem cells (PSCs) by the transcription factors (TFs) OCT4, SOX2,\n",
      "\n",
      "Hence we reasoned that responsiveness might also identify previously undescribed factors that maintain cellular identity. One way to test this hypothesis is in reprogramming from a source to target cell type: we reasoned that the downregulation of factors that are responsive in the source cell type may interfere with source cell type maintenance, manifesting as an enhancement of reprogramming to the target cell type. We tested this possibility by focusing on the induction of induced pluripotent stem cells from fibroblasts, i.e., fibroblast reprogramming.\n",
      "\n",
      "3 To test the hypothesis that P ​ can reveal genes that, if removed, would inhibit 3 ​ on isogenic dermal fibroblasts (GM00942, Coriell,\n",
      "Human: How does the overexpression or suppression of specific transcription factors affect the stability and functionality of reprogrammed cells?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "; , : ) (\n",
      "\n",
      "0 9 8 7 6 5 4 3 2 1\n",
      "\n",
      "; , : ) (\n",
      "\n",
      "0 9 8 7 6 5 4 3 2 1\n",
      "\n",
      "Article\n",
      "\n",
      "https://doi.org/10.1038/s41467-023-37256-8\n",
      "\n",
      "Conserved transcription factors promote cell fate stability and restrict reprogramming potential in differentiated cells\n",
      "\n",
      "Received: 7 August 2021\n",
      "\n",
      "Accepted: 8 March 2023\n",
      "\n",
      "Maria A. Missinato1,6, Sean Murphy2,6, Michaela Lynott Anaïs Kervadec1, Yu-Ling Chang1, Suraj Kannan2, Mafalda Loreti1, Christopher Lee 1, Prashila Amatya1, Hiroshi Tanaka1, Chun-Teng Huang3, 1, Chulan Kwon 2, Peter D. Adams Pier Lorenzo Puri Alessandra Sacco1, Peter Andersen 2 & Alexandre R. Colas\n",
      "\n",
      "1,6, Michael S. Yu1,\n",
      "\n",
      "4, Li Qian 5,\n",
      "\n",
      "1\n",
      "\n",
      "Check for updates\n",
      "\n",
      "Deﬁning the mechanisms safeguarding cell fate\n",
      "\n",
      "differentiation and maintenance were among the most frequently up-regulated (most responsive). We reasoned that one potential function of responsive genes may be to maintain cellular identity. We identified 3 responsive transcription factors in fibroblasts using P ​ and found that suppressing their expression led to enhanced reprogramming efficiency. We propose that responsiveness to perturbations is a property of factors that help maintain cellular identity.\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "Introduction\n",
      "\n",
      "Cells of a particular cell type likely maintain many of their specific behavioral properties across a range of conditions. The maintenance of these properties can be the result of the activity of certain\n",
      "\n",
      "reveals the critical role of the TRN in pro- tecting cells against aberrant reprogramming, and demonstrates the vulnerability of donor cell’s TRNs,\n",
      "\n",
      "offering a novel strategy to induce transgene-free trans-differentiations.\n",
      "\n",
      "INTRODUCTION\n",
      "\n",
      "In any given cell, a set of transcription factors (TFs) is ex- pressed and works in concert to govern cellular homeosta- sis and function (1). The Transcriptional Regulatory Net- works (TRNs) in somatic cells are believed to be robust and act to prevent cells from aberrant plasticity (2). However, recent works have revealed that somatic cells in mammals are more plastic than previously thought (3). Particularly, Takahashi and Yamanaka demonstrated this\n",
      "\n",
      "whose expression or activity is cell type- or tissue-specific. In principle, responsiveness could have identified a proper subset of the cell type-specific genes, a superset of the cell type-specific genes, a wholly distinct set of genes, or a partially overlapping subset. We found that responsiveness seemed to identify a superset of the cell type-specific genes, meaning that there were many genes whose expression was responsive but not particularly cell type-specific. When we explicitly tested whether these responsive-but-not-cell type-specific genes could affect maintenance, we found that, for fully half the factors we tested, suppression of the factor led to an increase in reprogramming\n",
      "\n",
      "whether these responsive-but-not-cell type-specific genes could affect maintenance, we found that, for fully half the factors we tested, suppression of the factor led to an increase in reprogramming efficiency. This validation rate suggests that these responsive 3 factors are indeed important for cell type maintenance and not merely falsely identified by P ​. The fact that these non-cell type-specific factors are important for maintenance suggests that cell type maintenance may involve factors that act as general barriers to cell type transformations in addition to factors specific to particular types. (Although it is possible that such factors may only act as barriers in particular cell\n",
      "\n",
      "Each experimental replicate was conducted using different vials of cells and different batches of media and feeder mouse embryonic fibroblasts (MEFs) on different days; knockdown samples are compared against controls in the same batch. Reprogramming rate variability was highest in the third replicate for unknown reasons.\n",
      "\n",
      "Supplementary Figure 12: Evaluating cardiac transdifferentiation after overexpression of iPSC-derived cardiac myocyte-responsive transcription factors in fibroblasts. A) Selection of 7 transcription factors found to be frequently up-regulated after perturbation in iPSC-derived cardiac myocytes (iPSC-CM; see Methods). B) Schematic of experimental workflow:\n",
      "\n",
      "by enhancing source cell type gene expression or repressing reprogramming TF-associated gene expression. How- ever, a systematic evaluation of TFs as fate stabilizers has not been conducted to date, and in this context, we will explore whether such factors might be evolutionarily conserved and play lineage and/or cell type-speciﬁc or independent roles. Mechanistically, we will evaluate how fate-stabilizing TFs binding to the DNA might con- tribute to regulate chromatin accessibility dynamics and tran- scription to oppose the lineage reprogramming process. Finally, we will test whether the targeted inhibition of such fate stabilizers might enhance our ability to use reprogramming TFs to\n",
      "\n",
      "that Drive Neuronal Fate and Reprogramming. Cell Stem . Cell ​ .\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "41. A. Ebrahimi, K. Sevinç, G. Gürhan Sevinç, A. P. Cribbs, M. Philpott, F. Uyulur, T. Morova, J. E. Dunford, S. Göklemez, Ş. Arı, U. Oppermann, T. T. Önder, Bromodomain inhibition of the coactivators CBP/EP300 facilitate cellular reprogramming. 519–528 (2019).\n",
      "\n",
      "Nat. Chem. Biol. ​\n",
      "\n",
      "15\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      ",\n",
      "\n",
      "​\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made\n",
      "\n",
      "available under a\n",
      "\n",
      "CC-BY\n",
      "\n",
      "Hence we reasoned that responsiveness might also identify previously undescribed factors that maintain cellular identity. One way to test this hypothesis is in reprogramming from a source to target cell type: we reasoned that the downregulation of factors that are responsive in the source cell type may interfere with source cell type maintenance, manifesting as an enhancement of reprogramming to the target cell type. We tested this possibility by focusing on the induction of induced pluripotent stem cells from fibroblasts, i.e., fibroblast reprogramming.\n",
      "\n",
      "3 To test the hypothesis that P ​ can reveal genes that, if removed, would inhibit 3 ​ on isogenic dermal fibroblasts (GM00942, Coriell,\n",
      "Human: How does the overexpression or suppression of specific transcription factors affect the stability and functionality of reprogrammed cells?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "; , : ) (\n",
      "\n",
      "0 9 8 7 6 5 4 3 2 1\n",
      "\n",
      "; , : ) (\n",
      "\n",
      "0 9 8 7 6 5 4 3 2 1\n",
      "\n",
      "Article\n",
      "\n",
      "https://doi.org/10.1038/s41467-023-37256-8\n",
      "\n",
      "Conserved transcription factors promote cell fate stability and restrict reprogramming potential in differentiated cells\n",
      "\n",
      "Received: 7 August 2021\n",
      "\n",
      "Accepted: 8 March 2023\n",
      "\n",
      "Maria A. Missinato1,6, Sean Murphy2,6, Michaela Lynott Anaïs Kervadec1, Yu-Ling Chang1, Suraj Kannan2, Mafalda Loreti1, Christopher Lee 1, Prashila Amatya1, Hiroshi Tanaka1, Chun-Teng Huang3, 1, Chulan Kwon 2, Peter D. Adams Pier Lorenzo Puri Alessandra Sacco1, Peter Andersen 2 & Alexandre R. Colas\n",
      "\n",
      "1,6, Michael S. Yu1,\n",
      "\n",
      "4, Li Qian 5,\n",
      "\n",
      "1\n",
      "\n",
      "Check for updates\n",
      "\n",
      "Deﬁning the mechanisms safeguarding cell fate\n",
      "\n",
      "differentiation and maintenance were among the most frequently up-regulated (most responsive). We reasoned that one potential function of responsive genes may be to maintain cellular identity. We identified 3 responsive transcription factors in fibroblasts using P ​ and found that suppressing their expression led to enhanced reprogramming efficiency. We propose that responsiveness to perturbations is a property of factors that help maintain cellular identity.\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "Introduction\n",
      "\n",
      "Cells of a particular cell type likely maintain many of their specific behavioral properties across a range of conditions. The maintenance of these properties can be the result of the activity of certain\n",
      "\n",
      "reveals the critical role of the TRN in pro- tecting cells against aberrant reprogramming, and demonstrates the vulnerability of donor cell’s TRNs,\n",
      "\n",
      "offering a novel strategy to induce transgene-free trans-differentiations.\n",
      "\n",
      "INTRODUCTION\n",
      "\n",
      "In any given cell, a set of transcription factors (TFs) is ex- pressed and works in concert to govern cellular homeosta- sis and function (1). The Transcriptional Regulatory Net- works (TRNs) in somatic cells are believed to be robust and act to prevent cells from aberrant plasticity (2). However, recent works have revealed that somatic cells in mammals are more plastic than previously thought (3). Particularly, Takahashi and Yamanaka demonstrated this\n",
      "\n",
      "whose expression or activity is cell type- or tissue-specific. In principle, responsiveness could have identified a proper subset of the cell type-specific genes, a superset of the cell type-specific genes, a wholly distinct set of genes, or a partially overlapping subset. We found that responsiveness seemed to identify a superset of the cell type-specific genes, meaning that there were many genes whose expression was responsive but not particularly cell type-specific. When we explicitly tested whether these responsive-but-not-cell type-specific genes could affect maintenance, we found that, for fully half the factors we tested, suppression of the factor led to an increase in reprogramming\n",
      "\n",
      "whether these responsive-but-not-cell type-specific genes could affect maintenance, we found that, for fully half the factors we tested, suppression of the factor led to an increase in reprogramming efficiency. This validation rate suggests that these responsive 3 factors are indeed important for cell type maintenance and not merely falsely identified by P ​. The fact that these non-cell type-specific factors are important for maintenance suggests that cell type maintenance may involve factors that act as general barriers to cell type transformations in addition to factors specific to particular types. (Although it is possible that such factors may only act as barriers in particular cell\n",
      "\n",
      "Each experimental replicate was conducted using different vials of cells and different batches of media and feeder mouse embryonic fibroblasts (MEFs) on different days; knockdown samples are compared against controls in the same batch. Reprogramming rate variability was highest in the third replicate for unknown reasons.\n",
      "\n",
      "Supplementary Figure 12: Evaluating cardiac transdifferentiation after overexpression of iPSC-derived cardiac myocyte-responsive transcription factors in fibroblasts. A) Selection of 7 transcription factors found to be frequently up-regulated after perturbation in iPSC-derived cardiac myocytes (iPSC-CM; see Methods). B) Schematic of experimental workflow:\n",
      "\n",
      "by enhancing source cell type gene expression or repressing reprogramming TF-associated gene expression. How- ever, a systematic evaluation of TFs as fate stabilizers has not been conducted to date, and in this context, we will explore whether such factors might be evolutionarily conserved and play lineage and/or cell type-speciﬁc or independent roles. Mechanistically, we will evaluate how fate-stabilizing TFs binding to the DNA might con- tribute to regulate chromatin accessibility dynamics and tran- scription to oppose the lineage reprogramming process. Finally, we will test whether the targeted inhibition of such fate stabilizers might enhance our ability to use reprogramming TFs to\n",
      "\n",
      "that Drive Neuronal Fate and Reprogramming. Cell Stem . Cell ​ .\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      "41. A. Ebrahimi, K. Sevinç, G. Gürhan Sevinç, A. P. Cribbs, M. Philpott, F. Uyulur, T. Morova, J. E. Dunford, S. Göklemez, Ş. Arı, U. Oppermann, T. T. Önder, Bromodomain inhibition of the coactivators CBP/EP300 facilitate cellular reprogramming. 519–528 (2019).\n",
      "\n",
      "Nat. Chem. Biol. ​\n",
      "\n",
      "15\n",
      "\n",
      "​\n",
      "\n",
      "​\n",
      "\n",
      ",\n",
      "\n",
      "​\n",
      "\n",
      "bioRxiv preprint\n",
      "\n",
      "doi:\n",
      "\n",
      "https://doi.org/10.1101/2020.06.11.147207\n",
      "\n",
      ";\n",
      "\n",
      "this version posted June 30, 2020.\n",
      "\n",
      "The copyright holder for this preprint\n",
      "\n",
      "(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made\n",
      "\n",
      "available under a\n",
      "\n",
      "CC-BY\n",
      "\n",
      "Transcription factors orchestrate dynamic interplay between genome topology and gene regulation during cell reprogramming\n",
      "\n",
      "Ralph Stadhouders  François Le Dily1,2,3, Javier Quilez  Yasmina Cuartero1,2,3, Jochen Hecht2,6, Guillaume J. Filion1,2, Miguel Beato1,2, Marc A. Marti-Renom  and Thomas Graf\n",
      "\n",
      "1,2,8*, Enrique Vidal\n",
      "\n",
      "1,2, François Serra\n",
      "\n",
      "1,2,3, Bruno Di Stefano\n",
      "\n",
      "1,2,4,  5, Clara Berenguer1,2,\n",
      "\n",
      "1,2, Antonio Gomez1,2, Samuel Collombet\n",
      "\n",
      "1,2*\n",
      "\n",
      "1,2,3,7*\n",
      "\n",
      "Chromosomal architecture is known to influence gene expression, yet its role in controlling cell fate remains poorly understood. Reprogramming of somatic cells into pluripotent stem cells (PSCs) by the transcription factors (TFs) OCT4, SOX2,\n",
      "Human: How does the overexpression or suppression of specific transcription factors affect the stability and functionality of reprogrammed cells?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "==================================================\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  q\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Thanks for chatting today! I hope to talk soon, and don't forget that a record of this conversation is available at: /home/jpic/BRAD/2024-06-07-23:15:59.212259.json\n"
     ]
    }
   ],
   "source": [
    "# what transcription factors are involved in cell reprogramming of skeletal muscle?\n",
    "brad.chat()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ee577fe-1ce3-4ac8-b1d9-118c51daa242",
   "metadata": {},
   "source": [
    "# Analyze Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "56d09776-1972-481f-996f-4efc2ec5f807",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b4ce7691-b65e-4392-a5b2-07fd0087d96a",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('/home/jpic/BRAD/EXP-out-2024-06-07-23:16:06.829689.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "00409744-305e-4564-82df-4066ddc1ab36",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAqgAAAFjCAYAAADxf/m2AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAABcSAAAXEgFnn9JSAABKvElEQVR4nO3deVxUVf8H8M8AwzLDrggqIIpL7mjq456FuFtqKvlUbplZWpqmZg8lT1pqamqp5ePavuGuaZq7ZZZmmvsGKpqiAsIMywhzfn/4m+uMDMgwA/cOfN6vF6/X5Z5z53znzIX7nXvvOVclhBAgIiIiIlIIF7kDICIiIiIyxwSViIiIiBSFCSoRERERKQoTVCIiIiJSFCaoRERERKQoTFCJiIiISFGYoBIRERGRojBBJSIiIiJFYYJKRERERIrCBJWIiIiIFIUJKhEREREpChNUIiIiIlIUJqhEREREpChMUImIiIhIUZigEhERAUhKSoJKpYJKpUJSUlKZt9+pUyeoVCrEx8fbVFYW5O4bqniYoBIVIT4+XvqnbPpxcXGBr68vQkND0bZtW4wePRoJCQkwGAxyh0s2WrVqFeLj47F79267X2vo0KEF9hWVSgWNRoOaNWsiNjYWP/30k/1Bl1Pz589HfHw8/vrrL7tf6+bNm3jvvffQvn17VK5cGWq1GpUrV0bDhg3Ru3dvzJo1C7/++qv9QROAe8lrfHy8bMkzlU9ucgdA5CyCg4Ol5ezsbFy7dg1Xr17FgQMHsHjxYlSqVAnTpk3DqFGjoFKpZIyUimvVqlXYs2cPgHtnqBzBxcUFQUFB0u+pqalISkpCUlISvv/+e7zwwgtYunQp95EHzJ8/H5cuXUJERASioqJK/Do7duzAwIEDkZqaKq3TarW4e/cuTp48iZMnT2LTpk0AACGExbZqtRr16tWTlstaeHg46tWrh8qVK5d52w9TVN8kJSXhv//9LwAwSSWHYYJKVEzXr1+3+D0/Px8nT57E9u3bsXDhQiQmJuKVV17B/v378eWXXzIBqaDCwsIsLoHm5eXh8OHDGDNmDA4dOoTly5ejWbNmGD16tHxBllOXL19Gnz59oNPpEBERgXfeeQd9+/aFv78/AODOnTs4ePAgNmzYgK+//rrA9tWrV8fp06fLOOr7Pv/8c9nafhi5+4YqHl7iJyohV1dXNG7cGOPHj8fx48fxzDPPAAC+/vprzJw5U+boSCnc3Nzwr3/9C5s3b0ZgYCAAYOHChTJHVT4tWbIEOp0O7u7u2LNnD4YNGyYlpwDg5+eHLl26YOHChbh69ap8gRLRQzFBJXIAjUaDzz77DM2aNQMAzJw50+ISo7nr169j4sSJaNiwIby9vaHVatGwYUNMmjQJN27ceGhb27ZtwzPPPIMaNWrAy8sLgYGBaNKkCV599VUcOHDAoq7pHtqiLl/v3r1bul/yQQ9uv2HDBkRHR6NSpUrw9fVF27ZtsW7dOottvvjiC7Rr1w4BAQHw9vZGx44dsWPHjoe+r927d2PQoEEIDw+Hp6cn/Pz80KpVK3zwwQfQ6/VWtzHd9zl06FAAQEJCAjp16oTAwEBoNBpERUVhwYIFMBqNFtutWrUKKpVKurz/3//+t8C9o44eCFKlShV07doVAHD69GnodDrpfZv3/5EjR/Dss88iNDQUarW6wGdXkv3nwQEuly5dwosvvij1dWRkJOLi4iz6+fjx43juuecQFhYGT09P1KlTB9OnT8fdu3ettmE+iMdgMGDmzJlo0qQJtFotAgICEBMTgy1bthTYzrSPXbp0CQAwbNiwAp9FcZnuX42KikJ4eHiRdb28vAqsK2og0IOf07FjxzBo0CBUq1YNXl5eqF+/PubMmYO8vDxpm19++QV9+vRB1apV4enpiUaNGmHRokUFbi0wKelAqJSUFKxYsQL9+vVD/fr14efnBy8vL9SuXRsjRozAiRMnCt3W/G9ICIFly5ahffv2qFSpElQqFVatWlVk30RERODxxx+Xfn/wszP9bbZu3RoqlQqvvPJKke9lx44d0r3+Fy9etKkfqJwRRFSoqVOnCgCiuH8qP/zwg1R/+fLlBcp3794t/P39pToajUZotVrp94CAALFv3z6rr63X68WAAQOkugCEj4+P8PDwkH5v2rSp1fgfe+yxQmPetWtXoe/RfPt33nlHABAuLi7Cz8/PIo5PPvlEGI1GMWTIEAFAuLm5CR8fH6nc1dVVbNq0yWr7d+/eFSNGjLB4PW9vb+Hq6ir9Xq9ePZGUlFRgW1N7Q4YMEaNHj5biM+9jAGLw4MEW23377bciODhYqNVqAUBotVoRHBxs8XP58uVC+8waUyw1atQotM7EiROlmK5evSqEsOz/hIQEKSZfX1/h6elp8dmVdP9JTEyU6qxevVp6DV9fX4t+7tChgzAYDGLTpk1Co9EIAMLPz0+oVCqpTmxsrNX39thjjwkAYsqUKaJDhw7SfvDgZzF16lSL7WbPni2Cg4OFi4uLFNODn0Vx9ejRQwAQoaGhwmg0Fns7a/2UmJhoUWb+Of3444/C09PTav8888wzQgghli5dKlxdXYVKpSrw9zJ58mSr7Zv68ME+eliZad8z/fj6+go3Nzfpdw8PD5GQkGC1TdO2gwcPFv3795f+hgICAoSLi4tYuXJlkX3TokULERAQIJU9+Nm99tprQgghVq5cKcWm1+sL/QxiY2MFABETE1NoHaoYmKASFcHWBDUzM1M64D+YFF2+fFk6WDdo0EDs379fKtu7d6+oV6+eACACAwNFcnJygdceOHCgdPCYPHmyuHLlihBCCKPRKJKTk8VXX30lRo0aZTV+exNUPz8/4erqKqZPny7S09OFEEIkJyeLrl27SonyO++8Izw9PcWnn34qHYDOnj0rWrRoIQCI8PBwkZ+fX6CNsWPHSge2xYsXi9u3bwshhDAYDGLXrl2iWbNmAoBo3rx5ge1NB9eAgADh7u4uPvzwQ3Hnzh0hhBC3bt2ySHx37NhRoO2iDvq2Kk6CavqCoVKppD4y739vb2/Ro0cPcerUKWmbs2fPCiHs23/Mkwt/f38RHR0tTpw4IYQQIisrS3z00UfSfhsXFyf8/PxEbGys9KUgMzNT/Oc//5FeY/v27QXem6kv/fz8hIeHh/j0009Fdna2FLsp+QEg1q9fX2D7GjVqCABSQlQS8fHxUhvjx48XOp3Opu2Lm6D6+/uL2NhYcenSJSGEEBkZGWLKlClS+YwZM4RarRavvvqquHHjhhBCiNTUVDF06FDpb/jMmTMF2i9pghofHy/i4uLEkSNHpPecn58vjh8/Lp599lnpS5jpS5E5037r7e0t3NzcxJw5c6S/oczMTHHt2jWb+qYwWVlZUiK7YsUKq3Vu3rwp3N3dpS9rVLExQSUqgq0JqhBC1KlTRwAQ7dq1s1g/atQoKZn6559/Cmx35coV4evrKwCI0aNHW5T9/PPPUhyLFy+2OX57E1QAYvr06QXK79y5Y3EG78svvyxQ5/z581L5g2f3/v77b6FSqYRGoxHHjh2zGl9GRoYIDQ0VAMTatWstyszPHBWW2Dz66KMCgBgxYkSBsrJMUJOSkqS+Mj/Tbd7/rVq1Enl5eVa3t2f/MU8uGjZsKHJycgps//zzz0t1YmJirJ6BNJ0ZfeGFFwqUmfoSsH71ID8/X3Ts2FFKsB/kiAT15s2bolq1alIcWq1WdOvWTbz99tti3bp1UrJYmOImYQ/rn8L2t7y8PBERESEAiGnTphUoL2mC+jA9e/YstE3zv6GPPvqo0NewN0EVQohx48YJAKJ169ZWy+fMmSN9WTUYDA9/Y1Su8R5UIgczDYQxvwdVCIHvv/8eADBq1CiEhIQU2C40NBSjRo0CAHz77bcWZStWrAAANGzYEC+//HKpxF0UT09PjBs3rsB6X19ftGnTBsC9KXL+/e9/F6gTGRmJ2rVrA7h335655cuXQwiBnj17onHjxlbb9vHxQZ8+fQCg0HlEw8LCMHjwYKtlTz75pNW2y8rNmzexfv16xMTESPd4jh8/3mrdiRMnwtXVtcB6e/cfc6+//jo8PDwKrDfdHwsAb775ptV7P011iurLsLAwDBs2rMB6FxcXxMXFAQBOnjyJv//+u9DXKKnKlStj//79iImJAQDo9Xps3boV06ZNQ58+fRAcHIwWLVpg1apVBe5LtsXkyZOL7B8AmDJlSoFyV1dXdO7cGUDZ7o89e/YEAOzfv7/QOgEBAXjppZdKNQ7T/vnbb79Z/fyXLVsGABg+fLgs03yRsjBBJXIwYWUARGJiopSwmg5Q1pgOrLdv30ZiYqK03jSpeO/evR0ZarE1aNAAWq3WaplpftgWLVoUOqDFVCctLc1ivemAuWXLFoSEhBT6s3LlSgCQBtI8qGXLlnBxsf7vrFq1agBQ6KA1R7t06ZLFIJEqVaqgT58+OHfuHFQqFSZPnlxoMt2uXTur6+3df8y1atXK6nrzeX5btmxZZJ0HP0dzpoE+1nTs2BFubvdmNzx06FChr2GPmjVrYtu2bTh58iRmzpyJp556ymLA1OHDhzFs2DB0794dOTk5JWrjYX0YGBiIWrVqFVmnqD4siaNHj+KVV15BkyZN4OvrCxcXF2kfNA1MSk5OLnT7li1bwt3d3aExPahevXrSgKqlS5dalO3btw+nT5+GSqXCiBEjSjUOcg6cB5XIwUwHnkqVKknrUlJSpOXq1asXum1oaKjFNjVr1gRwfw7WGjVqODTW4vLx8Sm0zJRwFKfOgyPAr127BgDQ6XTSqPaiZGVllTi+wkafP4y1s5UAEBsbiwULFhRY/+BE/Z6enggODkarVq0wdOhQPProo4W2VaVKFavr7d1/zBXWV6Z+Kk6dovqyqPg8PDxQqVIl3Lhxw+I9lYb69eujfv360u/Xr1/H+vXrMWPGDFy6dAnbtm1DXFwc5syZY/NrP6x/SnN/tGbhwoUYO3asdFZYpVLBz89POlOenZ2NjIyMQmfDAArf9xxt1KhR2LVrF7744gvMmjVLmk3hf//7H4B7X7IKS+6pYuEZVCIH0ul00tQokZGRVusUd9oc83qm5fI2+X9+fj6Ae9NyiXv3xBf544hHktrqxo0bVn/u3LljtX5YWBiuX78u/SQlJeHgwYP4+OOPi0xOAVi9vP+gkuw/ZUmp+2hISAheeuklHDx4UErGVqxYYdelfiU4deoUxo0bB6PRiAEDBuD3339HTk4O0tLSpH3www8/BGD96o5JcfY9R+jbty9CQkKQnp6OH374AQCQnp6OhIQEAMDIkSPLJA5SPiaoRA60detWKekyn7/S/OzElStXCt3e/BKc+Vk401k8W+fmNJ2tKepSZmGJVlkwva/SuB/RUQpLlk3zQ5YFe/efslTUZeTc3Fzcvn0bQNmdsXtQcHAwnnrqKQD3rnbcvHlTljgcJSEhAfn5+ahfvz6+/fZbq5fqH3wKnpzUajWGDx8O4P5l/i+++AI5OTkICQmR7hknYoJK5CAGgwHvv/8+gHtPrDEN7AHu3RdnGjxV1KT1P//8M4B7tweYX55t27YtAGDjxo02xRQQEACg6KTm4MGDNr2mI5nuudy8eXOxLvE7mum+1aLOLCmBvftPWdqzZ0+h/blv3z5pIvsWLVpYlJXlZ+Ht7S0tWxsw5kxMf9tNmzYt9D5s035RWszbLc7nN3LkSLi4uGD//v04deqUlKgOGzaMg6NIwgSVyAGys7MxdOhQHDlyBMC9Ebzmj1hUqVSIjY0FcO9xjNbOaFy7dg1LliwBAAwaNMii7IUXXgAAnDhxAp988kmx42ratKn02r/99luB8pSUlAKDFcrSiy++CJVKhfT0dEycOLHIunfv3nV4Euvr6wvg3iVGJbN3/ylLly9fxmeffVZgvdFolL7A1a9fv8CsDY74LPbt21fofcomOp0Oa9asAXAv8Tf/O3VGfn5+AO5dhbCWHG7ZsqXUb40xfXZA8T6/GjVqoHv37gDu3ZP6999/Q6VS4cUXXyytEMkJMUElKiGj0Yjjx4/jww8/RMOGDfHNN98AAJ5//nlMmjSpQP233noL/v7+SE1NRefOnaWR+cC9RyJ27twZ6enpCAwMxJtvvmmx7eOPP45nnnkGADBmzBhMmTJFupQqhMC1a9ewbNkyKZE1adu2rTSwaujQoTh06BCEEDAajdi9ezc6deok6z14UVFR0vRVn376KQYMGIC//vpLOtDm5+fj6NGjmDZtGiIjI6VHWTpKo0aNAAA//vij4p/Nbs/+U5b8/Pzw8ssvY+nSpdKtJVeuXMGgQYOwa9cuAMB7771XYDvTZ5GQkFDiEe4LFixAeHg4Xn31Vfz888/IyMiQyjIyMvD999+jbdu20mwQEyZMKFE7StKtWzcA9768jh49WprtQa/XY8mSJejfv7/FgM3SULduXem2gmXLlhXrLKppyqm9e/cCuDc4Sq6z/qRMTFCJisl82qOAgACo1Wo0btwYEyZMQGJiIipXroxPP/0Un3/+udWBIqGhoVi3bh38/Pxw4sQJtGvXDt7e3vD29kb79u1x6tQp+Pv7Y926dVZHQi9fvhz9+vWD0WjEzJkzERYWJj1zu3r16njxxRdx+PBhi21cXFywZMkSqNVqnDlzBi1btpSe3/74448jLy8PixYtKrU+K47Zs2dLSWpCQgKaNWsGjUaDypUrw9PTE1FRUXjnnXdw5coVhw/AGTJkCDw9PXH+/HmEh4cjJCQEERERiIiIKPJeSjnYu/+UlVdeeQUtWrTAyJEj4evri8DAQISHh0vzuMbFxaFv374Fths5ciRUKhV+/fVXBAUFoVq1atJnUVxqtRq3b9/GwoULERMTAz8/P/j6+sLHxwd+fn6IjY3F33//DRcXF0yaNOmhz4V3BtHR0dKX108++QSVKlVCQEAA/Pz8MGrUKNSvXx/x8fGlGoNGo8Hzzz8PAJg0aRK8vb1Ro0YNRERE4I033rC6TY8ePSxmJeHgKHoQE1SiYjKN3k5JSUFeXh5CQkLQunVrvPzyy0hISMDVq1cfOtH1Y489htOnT2PChAmoX78+jEYjhBCoX78+3njjDZw6dQodOnSwuq1Go8Hq1auxadMm9O3bF9WqVUNOTg68vb3RpEkTvPbaa9JULea6du2Kffv2oVevXggICEB+fj7CwsLw5ptv4vDhw4VOo1RWXF1dMW/ePPz5558YOXIk6tWrB1dXV9y5cwcBAQFo164d4uPj8ddffxU6T2hJ1alTB7t27cKTTz6JoKAg3L59G5cuXcKlS5ekeyWVxJ79p6y4u7tjx44deP/991GvXj3k5ubCz88P0dHR2Lx5M6ZNm2Z1u44dO2Lz5s3o3Lkz/Pz8cOPGDemzKK4vvvgCO3fuxJQpUxAdHY3Q0FAYDAbk5OQgICAALVu2xOuvv44jR45g1qxZip1xwFZfffUV5s+fjyZNmsDDwwP5+flo3LgxZsyYgV9++cXintvSsmjRIsTHx0tnwi9fvoxLly7h1q1bVuu7uLigX79+AMDBUWSVSih9dAARESlep06dsGfPHkydOrXUz9hR+dCkSRP8/fffmDJlinR/MpEJz6ASERFRmdq9e7d0uwUv75M1TFCJiIiozKSkpEj3nffv39+m+4yp4uCjTomIiKjUPfPMM/jll19w/fp15OXlwcfHBzNnzpQ7LFIonkElIiKiUnf9+nUkJydDq9Wic+fO2L17N6eWokJxkBQRERERKQrPoBIRERGRojBBJSIiIiJFYYJKRERERIrCBJWIiIiIFIUJKhEREREpCudBdaCQkBDo9XqEh4fLHQoRERGRrC5fvgytVovr16/bvC3PoDqQXq/H3bt35Q6DiIiISHZ3796FXq8v0bY8g+pApjOnJ06ckDkSIiIiInk1bNiwxNvyDCoRERERKQoTVCIiIiJSFCaoRERERKQoTFCJiIiISFGYoBIRERGRojBBJSIiIiJFYYJKRERERIrCeVCJiMqQEAJZWVlyh+EQGo0GKpVK7jCIqBxigkpEVIaysrLg7e0tdxgOodPpoNVq5Q6DiMohXuInIiIiIkXhGVQiIpk8NXsT3Dy85A7DJnm52Vg/sZfcYRBROccElYhIJm4eXk6XoBIRlQVe4iciIiIiRWGCSkRERESKwgSViIiIiBSFCSoRERERKQoTVCIiIiJSFCaoRERERKQoTFCJiIiISFGYoBIRERGRojBBJSIiIiJFYYJKRERERIrCBJWIiIiIFIUJKhEREREpChNUIiIiIlIUJqhEREREpChMUImIiIhIUZigEhEREZGiMEElIiIiIkVhgkpEREREisIElYiIiIgUhQkqERERESkKE1QiIiIiUhQ3uQMgIqpIhBDScvrtW3B195QxGtvl5WZLyzqdTsZI7KfRaKBSqeQOg4isYIJKRFSGsrKypOUd/42VMRL7hYSEyB2CXXQ6HbRardxhEJEVvMRPRERERIrCM6hERGVIo9FIy1P+PQDuauf6N6zLzsHc79cCAHrN2gAPL+c6A5mXm431E3vJHQYRPYRz/WckInJy5vc8PhIRBk93dxmjsd0dnV5adnP3gpuHl4zREFF5xUv8RERERKQoTFCJiIiISFGYoBIRERGRojBBJSIiIiJF4SApIqIyZD5Rf1ZuDvKN+TJGY7ssQ47Zb6LQekRE9mCCSkRUhswn6n957mIZI7GfMd8odwhEVE7xEj8RERERKQrPoBIRlSHzifr/eG8zNO7ONY/obV0aOv73aQCAiyvPcRBR6XDKBDUrKwvbtm3Dxo0b8ccffyApKQn5+fmoXbs2nn76aYwfPx7e3t5Wt/3888+xcOFCnDx5Eu7u7mjdujXi4uLQtm3bMn4XRFQRmU/UX9knEFoPTRG1lU718CpERCXglF9/v/76a/Tt2xcrVqyA0WhEt27d0KFDByQmJmLq1Klo2bIlUlJSCmw3fvx4DBkyBMePH0fnzp3RqlUrbN++HR07dsTatWtleCdERERE9CCnTFDd3d3x8ssv4+zZszh+/Di+//57bN26FWfOnEGzZs1w+vRpjBs3zmKbnTt3Yt68eahUqRKOHj2KdevWYevWrdi7dy9cXV0xbNgwpKWlyfOGiIiIiEjilAnq4MGDsXjxYtSpU8difdWqVbFo0SIAwJo1a2AwGKSyuXPnAgDi4uIstmvTpg1GjRqFO3fuYMWKFWUQPREREREVxSkT1KI0bdoUAJCbm4vbt28DAHJycrBjxw4AQP/+/QtsY1q3cePGMoqSiIiIiArjlIOkinLx4kUAgFqtRmBgIADg9OnTyM3NRVBQEEJDQwts07x5cwDAsWPHyi5QIiInZP6gAUO2XsZISibP7EED5u+FiJSl3CWoCxYsAAB069YNHh4eAIDLly8DgNXkFAC0Wi38/f2RlpaGzMxM+Pj4FNlGw4YNra6/cOECIiMjSxo6EZHiZd+9n+D9+FYf+QJxgKysrEJnfCEieZWrS/w//vgjli9fDrVajWnTpknrdTodAMv5Bx+k1Wot6hIRERGRPMrNGdRTp07hueeegxACs2fPlu5FBe5fxjGff/BBtlzqOXHihNX1hZ1ZJSIqL7zUntLy+Rkd4aNRyxiN7fS5RtSatAtA0SctiEhe5SJBTU5ORrdu3ZCWlobx48dj7NixFuWmS/Z6feH3S5mej83LPUREhTP/oh/k6w5fjYeM0dhOn5snLRd10oKI5OX0l/hv3bqFmJgYXL58GcOGDcOcOXMK1AkPDwdwL5G1Rq/XIz09Hf7+/g+9/5SIiIiISpdTJ6iZmZno3r07Tp8+jX79+mHp0qVWvxHXq1cPHh4euHnzptUk9c8//wQANGnSpNRjJiIiIqKiOW2Cmpubi6eeegqHDh1C165d8c0338DV1dVqXS8vLzzxxBMAgISEhALlpnW9evUqvYCJiIiIqFicMkHNz8/HoEGDsGvXLnTo0AFr1qyBu7t7kduMHz8eADB9+nScO3dOWn/gwAEsWbIEvr6+eOGFF0o1biIiIiJ6OKccJLVw4UKsXbsWAFC5cmW88sorVuvNmTMHlStXBgB07twZY8eOxYIFCxAVFYWYmBgYDAZs374dRqMRX331lTSxPxERERHJxykT1LS0NGnZlKhaEx8fLyWoADB//nxERUVh4cKF2L59O9RqNaKjoxEXF4f27duXasxEREREVDxOmaDGx8cjPj6+RNsOHToUQ4cOdWg8REREROQ4TnkPKhERERGVX0xQiYiIiEhRmKASERERkaIwQSUiIiIiRWGCSkRERESKwgSViIiIiBSFCSoRERERKQoTVCIiIiJSFCaoRERERKQoTFCJiIiISFGYoBIRERGRojBBJSIiIiJFYYJKRERERIrCBJWIiIiIFIUJKhEREREpChNUIiIiIlIUJqhEREREpChMUImIiIhIUZigEhEREZGiMEElIiIiIkVhgkpEREREisIElYiIiIgUhQkqERERESmKmyNf7PLly/D29kZgYGCR9dLS0pCZmYnw8HBHNk9EpHhCCGlZl62HMd8oYzS2y8zWScs3bt+BXq+WMRrbZRnypWXzz4KIlMWhCWrNmjUxdOhQLF++vMh6kyZNwsqVK5GXl+fI5omIFC8rK0tabjy5s4yR2K/uO3/IHYJdsrKy4O3tLXcYRGSFQy/xCyGK/Y2U31yJiIiIyBqHnkEtrlu3bsHLy0uOpomIZKXRaKTlDz74AB4eHjJGY7uMjAy8/fbbAIBvxiyExsO5/pfn3M1F7EevALD8LIhIWexOUPfu3Wvx+/Xr1wusM8nLy8OZM2ewdetWNGrUyN6miYicjkqlkpZ9fX2dLkE11yqyGXw0znWJXJ97/xYL88+CiJTF7gS1U6dOFn/kP/30E3766adC6wshoFKpMGHCBHubJiIiIqJyyO4EdfDgwVKC+tlnnyEyMhLt2rWzWtfd3R3VqlVD79690bx5c3ubJiIiIqJyyO4EddWqVdLyZ599hvbt22PFihX2viwRERERVVAOHSRlNDrXfH5EREREpDx8khQRERERKYrDp5nKzc3FN998g7179+Kff/5Bbm6u1XoqlQo7duxwdPNERERE5OQcmqBevXoV0dHROHfu3EMn4uf0HkRERERkjUMT1IkTJ+Ls2bNo27YtJkyYgLp16/IxckRERERkE4cmqD/99BPCw8Px888/w9PT05EvTUREREQVhEMHSeXm5qJly5ZMTomIiIioxByaoDZu3BjJycmOfEkiIiIiqmAcmqBOnjwZf/zxB/bs2ePIlyUiIiKiCsSh96A2b94cEyZMQO/evTF+/HjExMQgNDS00BH74eHhjmyeiIiIiMoBhyaoERERUKlUEEJg2rRpmDZtWqF1VSoV8vLyHNk8EREREZUDDk1QO3bsyPlNiYiIiMguDk1Qd+/e7ciXIyIiIqIKyKGDpIiIiIiI7MUElYiIiIgUxaGX+N99991i11WpVHj77bcd2TwRERERlQMOTVDj4+OlUfzWmAZQCSGYoBIRERGRVQ5NUFeuXGl1vdFoxJUrV/DTTz/hwIEDGD16NFq0aOHIpomIiIionHBogjpkyJAiy9955x3MmDED7733HkaOHOnIpomIiIionCjzQVJTpkxBaGgo3nrrrbJumoiIiIicgCyj+Bs3boz9+/fL0TQRERERKZwsCeqFCxf4mFMiIiIisqpME9T09HRMmDABf/31F1q1alWWTRMRERGRk3DoIKlatWoVWqbT6XD79m0IIeDl5YUZM2Y4smkiIiIiKiccmqAmJSUVWqZWqxEWFobHHnsMkydPRoMGDRzZNBERERGVEw5NUI1GoyNfjoiIiIgqIFkGSRERERERFcahZ1CtyczMhEqlgre3d2k3RUSkeOaPgs7KykJ+fr6M0dhOr9dLy5nZOhkjKZksQ7a0XNhjuYlIfqWSoG7duhUfffQRfvnlF+h09/6BabVadOjQAa+++iq6detWGs0SESleVlaWtDx+/HgZI7Ff1Ftd5Q7BLllZWTx5QqRQDr/EP378ePTs2RNbt25FZmYmfH194evrC51Ohy1btqBnz55O/0+ZiIiIiEqPQ8+gfvfdd5g/fz6qVKmCuLg4PP/88/Dz8wMAZGRk4IsvvsD06dOxYMECtG7dGgMHDnRk80REiqfRaKTl9Rt6w8ur1O+0cqjU1Gw8E7sVADBjxgyL9+MMcnNzMWnSJABwutiJKhKH/mdcvHgxPD09sXfvXtStW9eizNfXF6NHj0ZMTAyioqKwePFiJqhEVOGoVCppOTDQ0+kSVHO+vr5Ol+Tl5uZKy+afBREpi0Mv8R89ehRPPPFEgeTUXN26dfHEE0/gr7/+cmTTRERERFROODRBNRgM0Gq1D62n1WphMBgc2TQRERERlRMOTVAjIyOxZ88ei1GqD8rKysKePXsQGRnpyKaJiIiIqJxwaII6cOBApKSkoF+/frh48WKB8gsXLqBfv364efMmYmNjHdk0EREREZUTDr07/4033sD69euxbds21KtXD61atUJERARUKhUSExPx+++/Iz8/Hy1atMCECRMc2TQRFZMQosirHEpnHr9Go3G6gS7mE91zongiIuscmqB6eXlh9+7dmDJlClasWIEDBw7gwIEDFuXDhw/HjBkz4OXl5cimiaiYODm5cuTk5EOjUcsdBhGR4jh8fhNvb298/PHHmDVrFg4fPoxr164BAKpVq4ZHH33U6aYkISIiIqKyZXeCunPnTiQnJ6NFixZo0KCBtF6j0aBDhw4WdU+ePIlDhw4hLCwMjz/+uL1NE5Gdnpq9CW4eznU1IyczDZv/0x8A0GvWRnh4OdeX3tzsLGya3FvuMIiIFM2uBPXKlSvo2bMnwsLCcPjw4YfWDwsLQ9++fZGcnIxz586hWrVqJW778OHD2L59O37//XccPHgQ165dg4eHB3Jycorc7vPPP8fChQtx8uRJuLu7o3Xr1oiLi0Pbtm1LHAuRs3Lz8HK6BNUtN/v+srun08Wfn2+UOwQiIsWzaxT/smXLYDAY8MEHH8DHx+eh9X18fDB79mxkZ2dj+fLl9jSNadOmYcqUKVi7dq10G8HDjB8/HkOGDMHx48fRuXNntGrVCtu3b0fHjh2xdu1au+IhIiIiIsewK0Hdvn07goKC0KdPn2Jv8+STTyI4OBhbtmyxp2m0adMG77zzDjZu3Ijr168/tP7OnTsxb948VKpUCUePHsW6deuwdetW7N27F66urhg2bBjS0tLsiomIiIiI7GdXgnr69Gm0bNnS5u1atGiBM2fO2NM0Jk+ejP/+97/o1asXgoODH1p/7ty5AIC4uDjUqVNHWt+mTRuMGjUKd+7cwYoVK+yKiYiIiIjsZ1eCqtfr4efnZ/N2fn5+0Ol09jRtk5ycHOzYsQMA0L9//wLlpnUbN24ss5iIiIiIyDq7EtSAgADcuHHD5u1u3LiBgIAAe5q2yenTp5Gbm4ugoCCEhoYWKG/evDkA4NixY2UWExERERFZZ9co/gYNGuC3335DdnZ2sSfez8rKwoEDB9CqVSt7mrbJ5cuXAcBqcgoAWq0W/v7+SEtLQ2Zm5kMHfDVs2NDq+gsXLiAyMtK+YImIiIgqOLvOoPbu3Rt6vR7Tp08v9jbTp09HdnY2evcuu3kATbcTFPWQAK1Wa1GXiIiIiORh1xnUl156CbNnz8bMmTPh5eWFt956Cy4u1nNeo9GI9957DzNnzkRISAheeukle5q2iel510U9s9uWZ2KfOHHC6vrCzqwSERERUfHZlaBqNBqsWbMG0dHRmDp1KpYuXYoBAwagefPmCAoKAgDcvHkTf/75J3744QckJyfD09MTq1evLtNHnpou2ev1+kLrZGVlAQCfUU5EREQkM7sfddq6dWscOHAAzz33HI4fP4558+YVqGM6O9mwYUN8+eWXaNq0qb3N2iQ8PBwAkJycbLVcr9cjPT0d/v7+xXrgABERERGVHrsTVABo0qQJjh07hp9++gmbN2/GkSNHcPv2bQghULlyZURFRaFnz57o1q2bI5qzWb169eDh4YGbN28iOTm5wGCpP//8E8C990FERERE8nJIgmrStWtXdO3a1ZEv6RBeXl544oknsGXLFiQkJGDcuHEW5QkJCQCAXr16yRAdEREREZmzaxS/Mxk/fjyAe7MInDt3Tlp/4MABLFmyBL6+vnjhhRfkCo+IiIiI/p9Dz6CWpc2bN2PatGkW6wwGA1q3bi39/vbbb6Nnz54AgM6dO2Ps2LFYsGABoqKiEBMTA4PBgO3bt8NoNOKrr75CYGBgmb4HIjmYz1iRrcuEm+GujNHYLkdvPhVc8WffICIi5+G0CerNmzdx8OBBi3VCCIt1N2/etCifP38+oqKisHDhQmzfvh1qtRrR0dGIi4tD+/btyyRuIrmZZqwAgB/f6iNfIA5gzDfKHQIREZUCp01Qhw4diqFDh5bZdkRERERUNpw2QSWikjGfg7jPkj5w83CufwM5GTnY9NomAICLa4W5jZ6IqEJxriMTEdnN/Ilq3sHeUHuqZYzGdm6e5v+2Cn86HBEROS+efiAiIiIiRWGCSkRERESKwgSViIiIiBSFCSoRERERKQoHSRGRUzF/0ECOPgPG/HwZo7GdIUcvLZu/Fyob5n2ekpICrVYrYzQlp9VqodFoLAY9EpUnTFCJyKnk595PSLfGPS1jJPbLyXGu5Lo8MBgM0nKtWrVkjMR+Op3OaRNsoofhJX4iIiIiUhSeQSUip+Lq4Sot917cGx4aDxmjsV1uVi42vrIRAODp6fqQ2lSazs/oCB+Nc80DrM81otakXXKHQVTqmKASkVN58EEDnlpPGaOxnVp/PyHi/YPyCvJ1h6+TfcHR5+bJHQJRmeAlfiIiIiJSFCaoRERERKQoTFCJiIiISFGYoBIRERGRojBBJSIiIiJFYYJKRERERIrCBJWIiIiIFIUJKhEREREpChNUIiIiIlIUJqhEREREpChMUImIiIhIUZigEhEREZGiuMkdAJGzEUIgKysLKSkpcodSIllZWfd/EfLFUWJmMRv0BvniKCHzmIVwvg/APGaLfclJmMcsnPAPwLz/U1JSoNVqZYym5LRaLTQaDVQqldyhkEIxQSWyUVZWFry9veUOwyHyjflQQy13GDbJN+ZLy6sHr5YxEvvl5OQ/vJLCmMc8ceJEGSOxnzHf+RLULMP9/q9Vq5aMkdhPp9M5bYJNpY+X+ImIiIhIUXgGlcgOfZb0gZuHc/0Z5WTkYNNrm+QOo8RcPVyl5W29eiPQ01PGaGynMxjQacN6AICnp+tDaiuPeczr1vWERusuYzS2S0vLQezALQAAF1fnvrx8fkZH+Gic6wqIPteIWpN2yR0GOQHnOrISKYx3sDfUns51gHDzdO4/e/N71gI9PVHZy0vGaGzn4XL/wpUz3n9nHnNAgCe03s6VoJpTwfn631yQrzt8NR5yh2ETfW6e3CGQk+AlfiIiIiJSFCaoRERERKQoTFCJiIiISFGYoBIRERGRojj3aIkKyjRRfFJSktyh2CUiIoITNcuhHE10n5GdDTcnm2xdn+vcE/Wb0+lz5Q7BZnrd/ZhvZ+hhuHtXxmhsdzvz/v7jjPHrc+/P4+rs+z+VLiaoTqg8TRTPiZrLXnma6P6Jzc47XRbgnBP1G/ON0nLPHs7d/7WmHJA7BLs4e/zl6VhGjsdL/ERERESkKDyD6uTWd+sOLzfn+hiz8/Lw1NYtcodRYTn7RPeZubl4fOMGAMD0F56Hn5Odgc+6m4PJn6wC4PwT9T/xw4/wCAiQMRrb5aSmYtfAngCAma+ugpeHRuaIbJOhS8fbn44EAMwYsxIaT+fa/3MNOZj00WAAgEbjXH1PZcu5MhsqoI6/PzRq55ooPsvJ7pkqb8rTRPdVAvzh5+1cB2h9Tra07Iz3X5vH7BEQAM/AyjJGY5/Q4JrQejnXJeb0zFRpOSykltPFn5Pr3Ps/lR1e4iciIiIiRWGCSkRERESKwgSViIiIiBSFCSoRERERKQoHSRFVYBk5Tj7RvYxxEBFR6WGCSlTBmE+0/sQm555o3Wh0vonuiYjo4XiJn4iIiIgUhWdQiSoYF9f730vfGzEYvk42WXZGVhb+s+xzAIC7mv/CiIjKI/53J6rAgvz9nG6iew/3+w+m4ETfRETlEy/xExEREZGiMEElIiIiIkVhgkpEREREisIElYiIiIgUhYOkiIjKktnTBbL0Bni4O9d5Ap0+V+4QyIkJcf8PICUlBVqtcw3SNNFqtdBoNByoWYqYoBIRlaF84/0DdJ8+m2WMxH75Zg99ICqO3Ls50nKtWrVkjMR+Op3OaRNsZ+BcX92JiIiIqNzjGVQiojLk6nL/kuATP/wIj4AAGaOxXW5aGnYO6AEAcPX0lDkacmbnZ3SEj0b98IoKos81otakXXKHUSEwQSUiKktmt6x5BATAM7CyfLHYifffkT2CfN3hq/GQOwyb6HPz5A6hwuAlfiIiIiJSFCaoRERERKQoTFCJiIiISFF4DyqRjczn8cvV5yI/P1/GaGxn0BukZfP3QlQRmO/yt1NvIcszS75gSiBDly53CHYx/59zOyMLhrt3ZYzGdrqc+/eg3rhxA97e3jJGU3LOMI8rE1QiG2Vl3T+g/fDsDzJGYj/DXd7wTxVLfk62tDz2w6dljMR+zjgPrcF8HtQpv8oYif0iIyPlDsEuSp/HlZf4iYiIiEhReAaVyEYajUZa3tenL7zcnOvPSGcwoNOG9QAAd7VzxU5kL/O5W//18VKoPTVF1FYeQ0YGfp/wMgDA1dX5zjG5q+/3/8xXV8HLw7n6P1N/B3GfvAgAODejA3w17jJHZBtnmseVRyciG5nfs1PJ0xMatXNNNO3hcv+gpuT7j4hKg8Xfb9NH4eHtI2M0tstJvSV3CHYx7//Q4JrQejnXPZzpmanSchVfD87jWoqc7+sXEREREZVrTFCJiIiISFGYoBIRERGRojBBJSIiIiJF4SApKnPmEzWfPHnSYlS8MzCfB5WIqCyZP2ggU5/pdHOh6vSZcodgF2d/0IA+9/6DZZT+oBYmqFTmss2evNSqVSsZI7FfvtG5Dg5E5NzMHzTw8oxeMkZiP2dLroHy9aCBrKwsRT8Ji5f4iYiIiEhReAaVZLW2S1do3Z1rouO0nBzE/rxd7jCIqAIyf9DA7rd/gNbJHjSQqktH1xnPAuCDBuSQa8jBpI8GA4Dib6+rcAlqTk4OZsyYgW+++QaXL19GYGAgunXrhnfffRehoaFyh1fh1Pb3h4+Hc010fCs7++GViIhKgflE92GB1eCjUe4lWms07l5yh2AXZ3/QQE7u/eOX0h/U4nxfX+yQk5OD6OhovPvuu9DpdHjqqacQFhaGlStXonnz5rhw4YLcIRIRERFVeBUqQX3//ffx66+/ok2bNjh79iy+++47HDx4EHPnzsXNmzcxfPhwuUMkIiIiqvAqTIJ69+5dfPzxxwCARYsWWYxcGz9+PJo0aYK9e/fi8OHDcoVIRERERKhACer+/fuRnp6OyMhINGvWrEB5//79AQAbN24s69CIiIiIyEyFGSR19OhRAEDz5s2tlpvWm+o5i+upqfBydZU7DJuk5uZKyzdSU6FTq2WMxnbOHr/ebGLptDt3kO9kE01nmD0oIe1OBnKzc4uorTxZhvvx5mfnIC/buR78kJ+dY7acjTwn+/9TnuJPzUhzuoni03R3pOXbqbeQ5elc+3+GLl1adsb4cw05D6+kECqh9EcJOMj48eMxb948vP766/jwww8LlB89ehRRUVFo3rz5Qy/zN2zY0Or606dPQ61WIzIy0iExF8ZoNOL06dOl2gYRERGVX4888ghcXEr3QvqFCxegVquRmWn7E8QqzBlUnU4HoPB5v7RarUW9klCpVFCXwdk0FxcXNGjQoFh1TTMTlHbSTAWx7+XDvpcP+14+7Hv5sO+tU6vVUn5lqwqToJpOFBc275ctJ5JPnDjhkJjKgulsrzPFXF6w7+XDvpcP+14+7Hv5sO8dr8IMkvLx8QEA6PV6q+VZ/39fm5KfS0tERERUEVSYBDU8PBwAkJycbLXctN5Uj4iIiIjkUWES1KZNmwIA/vzzT6vlpvVNmjQps5iIiIiIqKAKk6C2a9cOfn5+uHDhAo4cOVKgPCEhAQDQq1evsg6NiIiIiMxUmATV3d0dY8aMAQCMGTPG4l7UDz/8EMeOHUP79u3RsmVLuUIkIiIiIlSgeVABICcnB506dcLBgwdRtWpVdOjQAZcuXcLBgwdRqVIl/Pbbb6hdu7bcYRIRERFVaBUqQQWA7OxszJgxA19//TWuXLmCgIAAdOvWDdOmTUNYWJjc4RERERFVeBUuQSUiIiIiZasw96ASERERkXNggkpEREREisIElYiIiIgUhQkqERERESkKE1QiIiIiUhQmqAqUk5ODqVOnom7duvD09ES1atUwfPhwJCcnF/s1Vq1aBZVK9dCfzz//vMjXMRgMaNCgAVQqFTw9Pe19a4qnhL7PyMjAO++8g0aNGkGr1cLPzw+NGjXC6NGjodPpHPVWFUfuvj916hQGDx6MsLAwqNVq+Pr6om3btvjf//4Ho9HoyLeqOI7oe5OtW7eie/fuqFy5MtRqNapUqYJevXphx44dhW5jNBoxf/58NG7cGF5eXggKCsKAAQNw8uRJe96WU5Cr7+/evYtt27ZhzJgxePTRRxEYGAgvLy/Ur18fb7zxBm7evOmIt6docu/35irasbY4OM2UwuTk5CA6Ohq//vqr9DCBpKQk/P777wgKCsKBAwcQGRn50NfZv38/li1bZrXszp07WLduHQDgwoULqFWrVqGvEx8fj3fffRdCCHh4eCAnJ6dE78sZKKHvz549i86dO+PKlSuoWbMmHn30UeTm5uLMmTM4e/Ysrly5gtDQULvfq9LI3ff79+9Hly5dkJ2djYYNG6JBgwZITU3Fvn37YDAYEBsbi2+//dYh71VpHNX3wL2n8k2YMAEqlQrt2rVD9erVcfHiRfzxxx8AgE8++QSjRo2y2EYIgYEDByIhIQH+/v6Ijo7GrVu3sHfvXnh6emLXrl3417/+5fD3rQRy9v3PP/+MmJgYAEBkZCSaNm2Ku3fv4sCBA7h16xZCQkKwe/du1KtXz/FvXAHk3u8fVJGOtcUmSFHefvttAUC0adNGZGZmSuvnzp0rAIiOHTva3cbixYsFANGuXbsi6508eVK4u7uLkSNHCgDCw8PD7raVTO6+1+l0ombNmsLFxUUsWrRIGI1Gi/K///5b6PV6u2NQIrn7vlmzZgKA+OCDDyzWnz17VlSuXFkAEDt37rQ7BiVyVN+npKQId3d34e7uLvbt22dRlpCQIFQqldBoNBZtCCHE8uXLBQBRp04dcf36dYttAIjIyEhx9+5dO96hcsnZ9zt27BCDBg0Sf/75p0X99PR00bVrVymu8kru/d5cRTvWFhcTVAUxGAzC399fACjwT0MIIZo0aSIAiEOHDtnVTtu2bQUA8emnnxZax2g0ivbt24sqVaqI1NTUcv9Ho4S+N/3DnDBhgl1tOBu5+z4zM1MAEBqNRuTn5xfYbuzYsQKAmDVrll3tK5Ej+37jxo0CgOjWrZvV8qZNmwoA4uDBgxbrGzRoIACItWvXFtjmySefFABEQkJC8d6QE1FC3xfm2rVrAoAAIJKSkoq1jTNRUt9XtGOtLXgPqoLs378f6enpiIyMRLNmzQqU9+/fHwCwcePGEreRmJiIX3/9Fe7u7hg4cGCh9ZYsWYL9+/dj7ty5CAgIKHF7zkLuvjcajVi2bBlUKhVef/31ErfhjOTue7VaDRcXF6hUqiJfIzAwsMTtK5Uj+97Dw6NYbZr3Y2JiIk6ePAkvLy/07NnTrvadjdx9X5SqVasiKCgIAHDt2rVibeNMlNT3Fe1YawsmqApy9OhRAEDz5s2tlpvWm+qVxJdffgkA6NmzZ6F/DP/88w/efPNNPPHEE3juuedK3JYzkbvvT548iX/++QcNGjRA9erV8dNPP+GNN97Ayy+/jDlz5iAxMbHE7Sqd3H3v4eGBDh06QK/XY+7cuRZl586dw1dffQU/Pz/06dOnxO0rlSP7vmXLlvDz88POnTuxf/9+i7I1a9bg2LFjaNu2LWrXrl2g/UaNGkGtVtvVvrORu++Lkp6ejrS0NABASEhIsbZxJkrp+4p4rLWFm9wB0H2XL18GgEIHwZjWm+qVxFdffQUAeP755wutM2bMGOTk5OCTTz4pcTvORu6+P3HiBACgZs2a6NOnD9avX29R/tZbb2H27NkYO3ZsidtXKrn7Hrg3iCEmJgaTJk3CZ599hoYNGyI1NRV79+5F7dq1sXLlSlSuXLnE7SuVI/ve398fy5Ytw7PPPouOHTtKg0USExPxxx9/oFu3bli1alWpte9s5O77oixatAh5eXlo3LgxatasWeztnIVS+r4iHmttwQRVQUxTCGk0GqvlWq3Wop6tfv/9d5w5cwYBAQFWL6cBwPr167FmzRpp6o2KQu6+N52t2Lp1KwDggw8+wHPPPQchBL744gvExcVh3LhxqFu3Lrp3716iGJRK7r4HgPr162P//v3o168fjhw5In1hUKvViImJKZcHacDxfd+/f38EBgYiNjbW4mxScHAwnnjiCVSqVKlU23cmcvd9YY4cOYLp06cDAGbNmlWsbZyNEvq+oh5rbcFL/Aoi/n/Gr8LuhRN2zghmuswZGxsLd3f3AuWZmZkYM2YM6tSpgylTptjVlrORu+/z8/MBAHl5eZg4cSImTpyIqlWrolq1apg8eTLGjRsHAHjvvffsikOJ5O57ANi5cyeaNWuGvLw87Ny5ExkZGUhMTMTkyZPx8ccfo3379khNTbUrDiVydN/PnTsXMTEx6NixI44dOwadTodjx46hTZs2mDhxImJjY21qvzyTu++tuX79Ovr164ecnByMGzeu3H0ZNpG77yvysdYWTFAVxMfHBwCg1+utlmdlZQEAvL29bX7tvLw8fPfddwAKv8z51ltvITk5GZ988kmxb/wuL+Tue1P7ADB8+PAC5aZ1v/32G3Jzc22OQcnk7vu0tDQMGDAAeXl52LJlCx5//HH4+PggIiIC06ZNw+jRo3H27FnMmTPH5vaVzpF9v2fPHrzxxhuIiorCDz/8gMaNG0Or1aJx48ZISEhAs2bNsHr1amzbtq3Y7ZvWl+SzVzq5+/5Bd+7cQffu3ZGUlIQBAwYUuB+7PJG77yvysdYWTFAVJDw8HAAKfYqFab2pni22bduGlJQU1KpVC23btrVaZ+PGjfD09MS0adPQqVMnix/g3pMuTL+fP3/e5hiUTO6+j4iIkJZr1KhRaHl+fj5u375tcwxKJnffb9q0CampqWjdujWqV69eoNw06n/37t02t690jux709O5+vXrBxcXy0OLq6sr+vXrB8CyH0vzs1c6ufveXHZ2Nnr37o2//voLXbp0wZdfflngdcoTufu+Ih9rbcF7UBWkadOmAIA///zTarlpfZMmTWx+bdNlzoeNFMzJycGePXuslgkhpLLydk+Y3H3fpEkTuLq6Ij8/H6mpqQgODrYoN09Ky9vZJLn73nQw8vX1tVpuWl8eL/E7su9L0o+m9o8fP467d+8WGMlvz2evdHL3vUleXh4GDBiAffv2oW3btlizZk2ht8KUF0ro+4p6rLVF+f2K5ITatWsHPz8/XLhwAUeOHClQnpCQAADo1auXTa+r0+mkUeFFHaiTkpIg7j28ocAPcG86HtPvUVFRNsWgdHL3vb+/Pzp06AAA2LVrV4Fy07fvyMjIQv8ROiu5+940jc6RI0eke4HNmR5XaH6Wu7xwZN+b+vHQoUNWy631Y82aNVG/fn1kZ2dj8+bNdrXvbOTue+BeIjR06FBs3rwZUVFR2Lx5szRAqDyTu+8r8rHWJqX+KACyyX/+8x8BQLRt21bodDppvenxa+3bt7eo//HHH4t69eqJN998s9DX/OyzzwQA0bp16xLHhQrwdAu5+37r1q0CgKhVq5a4ePGitP78+fMiIiJCABCzZ88uwTtTPjn7/tq1a8LDw0MAEFOmTLF4mtTp06dFtWrVBACxdOnSEr47ZXNU369Zs0YAEK6urmLDhg0WZevWrRMuLi7CxcVFnD592qJs6dKl0qNOb9y4Ia1fvXq1ACBq1qwpDAaDo96uosjd96+++qoAIB555BGRkpLi4HenbHL3fWEqwrG2uJigKkx2drb417/+JQCIqlWrioEDB0q/V6pUSZw7d86i/tSpUwUAMWTIkEJfMyYmRgAQixYtKnFcFeGPRgl9P2HCBAFAeHt7i65du4ouXboIrVYrAIju3buLvLw8e96iYsnd9x9//LFQqVTSF4Snn35adOrUSUpce/ToUW6fB++ovjcajWLAgAHSIzJbtGghBgwYIFq0aCGte++99wq0n5+fL/r27SsAiICAANG/f3/RqVMnoVKphKenp/jll19K8+3LSs6+X7dunVQWExMjhgwZYvXn1KlTpd0NspB7vy9MRTjWFhcTVAXKysoSb7/9toiMjBTu7u4iODhYDBkyRFy+fLlA3YcdqK9duyZcXV2FWq0Wt27dKnFMFeWPRgl9/80334i2bdsKb29v4eXlJZo1ayYWLFhQbhMkE7n7fteuXaJPnz4iJCREuLm5CV9fX9G6dWuxaNGicvvFwMRRfW80GsXy5ctFx44dhb+/v3BzcxOVK1cWPXr0EFu2bCm0/by8PDF37lzRsGFD4enpKSpVqiT69esnjh8/7si3qUhy9f3KlSulBKqon127dpXCu1YGufd7ayrKsbY4VELYOckgEREREZEDcZAUERERESkKE1QiIiIiUhQmqERERESkKExQiYiIiEhRmKASERERkaIwQSUiIiIiRWGCSkRERESKwgSViIiIiBSFCSoRERERKQoTVCIiIiJSFCaoRERERKQoTFCJiIqwfft29OnTByEhIXB3d0elSpXQoEEDPPvss1i6dCkMBoNUNyIiAiqVqtRj6tSpE1QqFZKSkizWl1X7ADB06FCoVCrs3r27TNojooqFCSoRUSGmTp2KLl26YP369QgKCkLv3r0RHR0NtVqNb775BiNHjkRqaqrcYSpKWSbJRFR+qYQQQu4giIiU5tChQ2jZsiXc3d2xdu1a9OjRw6L86tWrWLp0KcaOHYuAgAAAwIULF3D37l088sgjpRrb5cuXkZWVhcjISKjVaml9REQELl26hLL4t/7PP//gzp07CA8Ph0ajkSUGIiq/3OQOgIhIidauXQsAGDhwYIHkFACqV6+O+Ph4i3WRkZFlERrCw8PLpJ2iVK1aFVWrVpU7DCIqp3iJn4jIips3bwIAgoKCir2NtcvbSUlJUKlU6NSpE/R6PcaPH4+wsDB4eXmhefPm2Lhxo1T3hx9+QKtWraDVahEcHIzXXnsN2dnZBdop7B7UwmzevBnDhw9H/fr14evrC61Wi6ZNm+L9999Hbm5ugfqrVq2CSqVCfHw8zp49i2eeeQbBwcFwcXHBunXrABS8B3X37t1QqVS4dOkSAEClUkk/ERERAIBGjRpBpVLh7NmzVuNMSkqCi4sL6tSpwzOwRBUcE1QiIitCQ0MBAKtXr5aSVXsYDAZER0fjiy++QFRUFFq3bo2jR4+ib9+++PnnnzFv3jz8+9//hpubG7p06YL8/Hx8/PHHGDFihN1tv/DCC/jhhx/g5+eHbt26oUOHDrhy5Qr+85//oEePHsjPz7e63ZkzZ9CyZUv8/vvvePzxxxETE2NxS4G5kJAQDBkyBFqtFgAwZMgQ6ad///4AgJdeegkAsGzZMquvsXz5cgghMGLECN7HSlTRCSIiKuD8+fPC09NTABC+vr5i8ODBYunSpeL48ePCaDRa3aZGjRriwX+riYmJAoAAIDp16iRSU1OlspUrVwoAonbt2iIwMFDs3btXKrt69aqoUqWKACAuXLhg8ZqPPfaYACASExMf2r4QQqxdu1bodDqLdRkZGaJXr14CgPjss88sykxxARBjxowReXl5BV5zyJAhAoDYtWtXsWIQQoj09HSh0WhElSpVhMFgsCjLy8sT1atXF25ubuL69etWtyeiioNnUImIrIiMjMT69etRrVo1ZGRk4PPPP8eLL76IRo0aISQkBJMmTUJ6enqxX8/V1RVLly6VBlQBwODBgxEUFITz589jzJgx6NChg1RWrVo1PPvsswCAvXv32vVe+vTpI53ZNPHx8cG8efMAAOvXr7e6XVBQEGbNmgVXV1e72jfx8/NDbGwsUlJSsGHDBouyLVu24OrVq3jyyScRHBzskPaIyHlxkBQRUSG6dOmCixcvYsOGDdi+fTsOHjyI48ePIyUlBbNnz8batWvx66+/Fus+1YiICNSuXdtinYuLC2rUqIGbN28iJiamwDamQVf//POP3e/l3Llz+PHHH3H+/Hno9XoYjUbpPs9z585Z3aZz584WI/QdYdSoUVi5ciWWLl2Kp59+Wlq/dOlSAMCLL77o0PaIyDkxQSUiKoKHhwcGDBiAAQMGALg3eGrVqlWIj4/H+fPn8dZbb0nJVVGqV69udb3pzKa1clOZtYFMxSWEwBtvvIF58+YVOvAoMzPT6vrSmC2gVatWaNasGbZv345Lly6hRo0a+Oeff/Djjz8iPDwcXbp0cXibROR8eImfiMgGQUFBmDhxImbNmgXg3gj54njYoJ/SGhT03Xff4cMPP0T16tWRkJCAq1evwmAwQAghJb6FJa6enp6lEtNLL70Eo9GIFStWAABWrlyJvLw8vPDCC3Bx4WGJiJigEhGVSKdOnQAAt27dkjeQhzDN5/rJJ5/g6aefRrVq1aSR+BcvXpQlpmeffRY+Pj5YsWIF8vLysHz5cri4uGD48OGyxENEysMElYjIisLOKppcuHABwL3BTEqWlpYGAAgLCytQ9v333zu8PXd3dwBAXl5eoXW8vb3x73//G8nJyZg4cSIuXryI7t27S1N7ERExQSUisuLtt9/GpEmTkJiYWKDs3LlzmDBhAgCgX79+ZR2aTerWrQsA+N///meRdO/btw+zZ892eHumhP3MmTNF1hs1ahQAYP78+QA4OIqILDFBJSKyQqfTYfbs2YiMjET9+vXRr18/xMbGok2bNnjkkUdw4cIFPProo5g6darcoRbptddeg1arxeLFi9GoUSMMGjQIHTt2xGOPPSYliY705JNPAgCio6MxaNAgjBgxAm+++WaBelFRUWjVqhWAe49N7dmzp8NjISLnxQSViMiKuLg4fP7559LTnfbs2YM1a9bg/PnzeOyxx7Bo0SL8+uuv8PPzkzvUItWtWxd//PEHevfujVu3bmHDhg3Q6XRYsmRJqZxBfe211xAXFwdvb2+sXr0ay5cvx7fffmu1bnR0NABg2LBhcHPjpDJEdJ9KPOxGKyIiIgcTQuCRRx7BuXPncP78edSqVUvukIhIQXgGlYiIylxCQgLOnj2LHj16MDklogJ4BpWIiMrMiBEjkJ6ejk2bNiEvLw+///47mjdvLndYRKQwTFCJiKjMqFQquLm5oW7dupg2bZriZ0EgInnwrnQiIiozPCdCRMXBe1CJiIiISFGYoBIRERGRojBBJSIiIiJFYYJKRERERIrCBJWIiIiIFIUJKhEREREpChNUIiIiIlIUJqhEREREpChMUImIiIhIUZigEhEREZGiMEElIiIiIkVhgkpEREREisIElYiIiIgUhQkqERERESnK/wGGXYFaU9xH8wAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 750x300 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.figure(dpi=150, figsize=(5,2))\n",
    "sns.histplot(data=df, x='hiddenScore', hue='prompt', multiple='stack')\n",
    "plt.title('Document-Prompt Similarity')\n",
    "plt.xlabel('Similarity')\n",
    "plt.legend([], [], frameon=False)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "47d1d051-89b4-4f00-ad23-031fb063ea02",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "64"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['hidden'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "57d381a7-c6de-46e9-a3ce-6855f7209492",
   "metadata": {},
   "outputs": [],
   "source": [
    "import bert_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "d8be1731-1b8c-4da1-8fa8-822f6173e913",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "be236af3a454472088687bef1bab02f6",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "tokenizer_config.json:   0%|          | 0.00/25.0 [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "09bc9089148c4de7bee5ba1424335ecc",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "config.json:   0%|          | 0.00/482 [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6d6417e21cdd45448f938085b741b3e7",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "vocab.json:   0%|          | 0.00/899k [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "62ebf2daa11c474395d86aba117113d8",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e2d4c0762f0841a196f71163f5277707",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "691be7bba7b64797bb77f9bac7f756b5",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "model.safetensors:   0%|          | 0.00/1.42G [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tensor([0.8405])\n"
     ]
    }
   ],
   "source": [
    "from bert_score import BERTScorer\n",
    "scorer = BERTScorer(lang=\"en\")#, rescale_with_baseline=True)\n",
    "cand = [\"Myod and Myf5 are important transcription factors for myogenic reprogramming as they are crucial for skeletal muscle formation during embryonic development. They can access muscle-specific promoters and specify the skeletal muscle lineage by recruiting histone-modifying and chromatin-remodeling complexes to reset the cellular chromatin structure and transcriptional program. Chromatin context plays an instructive role in their activity by providing an environment for their binding. Other factors, such as those discovered through recent studies on chromatin-associated complexes, may also contribute to myogenic reprogramming, but they are not mentioned in the provided context.\"]\n",
    "refs = [\"Myf5. As noted above, these transcription factors are necessary and sufficient for skeletal muscle formation, and the instruction to express either of them effectively speciﬁes the skeletal muscle lineage. In addition, auto-regulation and cross- regulation exists among the myogenic bHLH proteins, such that a transient embryonically instructed induction results in stable expression of these factors. The convergence of multiple embryonic signals on this simple binary decision has been referred to as a ‘nodal point’ in muscle cell speciﬁcation. The identiﬁcation of other bHLH proteins that regulate complex programs of differentiation\"]\n",
    "P, R, F1 = scorer.score(cand, refs)\n",
    "print(F1)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
