{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b5645027-5ee8-4a54-8062-f9858dad8864",
   "metadata": {},
   "source": [
    "# Planning Larger Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f7cb860d-0647-4dd4-a35b-931a580a9c61",
   "metadata": {},
   "outputs": [],
   "source": [
    "from BRAD import brad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0333a24d-e684-46f3-8d56-eba58f4231be",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import os\n",
    "from langchain.document_loaders import DirectoryLoader\n",
    "from langchain.document_loaders import UnstructuredPDFLoader\n",
    "from langchain.vectorstores import Chroma\n",
    "from langchain.embeddings import HuggingFaceEmbeddings\n",
    "import chromadb\n",
    "from langchain_community.document_loaders import PyPDFLoader\n",
    "from langchain_text_splitters import RecursiveCharacterTextSplitter\n",
    "from langchain import PromptTemplate, LLMChain\n",
    "from langchain.memory import ConversationBufferMemory\n",
    "from langchain.chains import ConversationChain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a9adf57d-c10b-4abe-82b0-33a2cac0b4b1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Tue 18 Jun 2024 06:18:02 PM EDT INFO PyTorch version 2.0.1 available.\n",
      "Tue 18 Jun 2024 06:18:02 PM EDT INFO Use pytorch device_name: cpu\n",
      "Tue 18 Jun 2024 06:18:02 PM EDT INFO Load pretrained SentenceTransformer: BAAI/bge-base-en-v1.5\n",
      "/home/jpic/.local/lib/python3.11/site-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.\n",
      "  warnings.warn(\n",
      "Tue 18 Jun 2024 06:18:04 PM EDT INFO Anonymized telemetry enabled. See                     https://docs.trychroma.com/telemetry for more information.\n",
      "Tue 18 Jun 2024 06:18:04 PM EDT INFO Collection DigitalLibrary is not created.\n"
     ]
    }
   ],
   "source": [
    "# Load the database\n",
    "persist_directory = '/nfs/turbo/umms-indikar/shared/projects/RAG/databases/DigitalLibrary-10-June-2024/'\n",
    "embeddings_model = HuggingFaceEmbeddings(model_name='BAAI/bge-base-en-v1.5')\n",
    "db_name = \"DigitalLibrary\"\n",
    "_client_settings = chromadb.PersistentClient(path=(persist_directory + db_name))\n",
    "vectordb = Chroma(persist_directory=persist_directory, embedding_function=embeddings_model, client=_client_settings, collection_name=db_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1d9608e-39da-4584-bd70-dfceb81f6772",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome to RAG! The chat log from this conversation will be saved to /home/jpic/BRAD/2024-06-18_18-20-20/log.json. How can I help?\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Tue 18 Jun 2024 06:20:20 PM EDT INFO local\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  /force PLANNER make a 2 step process using: Step 1 Method is Code to run the hwgFirstOrderInteractions code on PCNA, and step 2 method is database to load in the Gene column from the output file created in step 1 and query it against ENRICHR. Use suitable file names and paths\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Tue 18 Jun 2024 06:20:28 PM EDT INFO PLANNER\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "RAG >> 1: \n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3m**INSTRUCTIONS:**\n",
      "You are planning a bioinformatics analysis pipeline to address a user's query. Your task is to outline a multi-step workflow using the available methods listed below. Each method should be used appropriately to ensure a thorough analysis. For each step in the pipeline, explain the goal of the process as well as provide a prompt to the chatbot that will execute that step. If information is passed between stages, such as from literature or code to databases, indicate the dependencies of steps in the prompt.\n",
      "\n",
      "**Available Methods:**\n",
      "1. **Retrieval Augmented Generation**(RAG): Look up literature and documents from a text database.\n",
      "2. **Web Search for New Literature**(SCRAPE): Search platforms like arXiv, bioRxiv, and PubMed for the latest research.\n",
      "3. **Bioinformatics Databases**(DATABASE): Utilize databases such as Gene Ontology and Enrichr to perform gene set enrichment analyses.\n",
      "4. **Run Codes**(CODE): Execute bioinformatics pipelines. Utilize prebuilt pipelines or develop new ones as needed.\n",
      "5. **Write a Report**(WRITE): Synthesize and summarize information. This can include summarizing database searches, code pipeline results, or creating a final report to encapsulate the entire analysis.\n",
      "\n",
      "Current conversation: \n",
      "\n",
      "\n",
      "User Query: make a 2 step process using: Step 1 Method is Code to run the hwgFirstOrderInteractions code on PCNA, and step 2 method is database to load in the Gene column from the output file created in step 1 and query it against ENRICHR. Use suitable file names and paths\n",
      "\n",
      "Plan:\n",
      "**Step 1 (method, eg. RAG)**:\n",
      "Prompt: [Description of action for chatbot to do Step 1, i.e. do X, Y, Z]\n",
      "\n",
      "**Step 2 (method, eg. CODE)**:\n",
      "Prompt: [Description of action for chatbot to do Step 2, i.e. do A, B, C. Use information from Step X]\n",
      "\n",
      "**Step 3 (method, eg. DATABASE)**:\n",
      "Prompt: [Description of action for chatbot to do Step 3, i.e. do D, E, F. Use information from Step Y]\n",
      "...\n",
      "\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      " **Step 1 (method: CODE)**:\n",
      "Prompt: Run the hwgFirstOrderInteractions code on PCNA. Save the output file in a location accessible to the next step with a name such as \"PCNA_hwgFirstOrderInteractions_output.tsv\".\n",
      "\n",
      "**Step 2 (method: DATABASE)**:\n",
      "Prompt: Load the Gene column from the output file created in Step 1, i.e. \"PCNA_hwgFirstOrderInteractions_output.tsv\", into Enrichr. Perform gene set enrichment analysis using the desired tool and save the results in a file named \"PCNA_gene_enrichment_results.txt\".\n",
      "\n",
      "\n",
      "Do you want to proceed with this plan? [Y/N/edit]\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  Y\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Tue 18 Jun 2024 06:20:31 PM EDT INFO CODE\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "==================================================\n",
      "RAG >> 2: CODER\n",
      "input_variables=['user_query'] template=\"You must select which code to run to help a user.\\n\\n**Available Scripts**\\nScript Name: hwgFirstOrderInteractions, \\t Description: Find the first order interactions of a gene in the Hardwired Genome\\n\\n\\n**User Query**\\n{user_query}\\n\\n**Task**\\nBased on the user's query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\\n\\n**Response Template**\\nSCRIPT: <selected script>\\nREASON: <reasoning why the script fits the user prompt>\\n\"\n",
      "FIRST LLM CALL\n",
      "Warning! the type doesnt exist\n",
      "scriptPath=/home/jpic/RAG-DEV/tutorials/planner/python/\n",
      "pyPath=/home/jpic/RAG-DEV/tutorials/planner/python/\n",
      "ALL SCRIPTS FOUND. BUILDING TEMPLATE\n",
      "input_variables=['history', 'input'] template='Current conversation:\\n{history}\\n\\n**PYTHON SCRIPT**\\nYou must run this python script:\\n/home/jpic/RAG-DEV/tutorials/planner/python/hwgFirstOrderInteractions.py\\n\\n**PYTHON SCRIPT DOCUMENTATION**:\\nThis is the doc string of this python script:\\n\\'\\'\\'\\nFind the first order interactions of a gene in the Hardwired Genome\\n\\nUsage:\\n    python <path/to/script/>hwgFirstOrderInteractions.py <gene name> <output file>\\n\\nTypically the output file name should be S<Step number>-HWG-<gene name>.csv\\n\\nThis script uses the Hardwired Genome construction. Note: Cooper recently found a bug in how the data are integrated for the Hardwired Genome.\\nThe code here should side step that by using a different download from Ensembl (GRCh38.p13.tsv) for mapping between gene names and ENSG IDs, but\\nthe issue may persist.\\n\\nArguments:\\n    gene name: this can either be a gene name in its symbol or common name or it can be a ENSG ID\\n\\'\\'\\'\\n\\n\\n**CALL PYTHON SCRIPTS FROM PYTHON**:\\nUse the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\\n\\nTo call a Python script `example_script.py` which has no arguments:\\n    \\n```\\nExecute: subprocess.call([sys.executable, \\'<full path to script/> example_script.py\\'])\\n```\\n\\nTo call a Python script `example_script.py` with one argument:\\n\\n```\\nExecute: subprocess.call([sys.executable, \\'<full path to script/> example_script.py\\', \\'arg1\\'])\\n```\\n\\nTo call a Python script `example_script.py` with two arguments:\\n```\\nExecute: subprocess.call([sys.executable, \\'<full path to script/> example_script.py\\', \\'arg1\\', \\'arg2\\'])\\n```\\nQuery:{input}\\n\\n**OUTPUT**\\nIf output or output files are created, all output files should be named: /home/jpic/BRAD/2024-06-18_18-20-20/<output file>\\n\\n**INSTRUCTIONS**\\n1. Given the user query and the documentation, identify each of the arguments found in the user\\'s query that should be passed to the Python script.\\n2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\\n3. The last line of your response should say \"Execute: <Python code to execute>\"\\n4. Format the response/output as:\\n    Arguments: \\n    Python Code Explanation: <2 sentences maximum>\\n    Execute: <your code here>\\n\\n**IMPORTANT**\\nThe code to execute from your response must be formatted as:\\n    Execute: subprocess.call([sys.executable, \\'<path to python script>\\', \\'<argument 1>\\', \\'<argument 2>\\', ..., \\'<argument n>\\'])\\nThis output should be exactly one line and no longer. Stop the response after this line.\\n'\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation:\n",
      "\n",
      "\n",
      "**PYTHON SCRIPT**\n",
      "You must run this python script:\n",
      "/home/jpic/RAG-DEV/tutorials/planner/python/hwgFirstOrderInteractions.py\n",
      "\n",
      "**PYTHON SCRIPT DOCUMENTATION**:\n",
      "This is the doc string of this python script:\n",
      "'''\n",
      "Find the first order interactions of a gene in the Hardwired Genome\n",
      "\n",
      "Usage:\n",
      "    python <path/to/script/>hwgFirstOrderInteractions.py <gene name> <output file>\n",
      "\n",
      "Typically the output file name should be S<Step number>-HWG-<gene name>.csv\n",
      "\n",
      "This script uses the Hardwired Genome construction. Note: Cooper recently found a bug in how the data are integrated for the Hardwired Genome.\n",
      "The code here should side step that by using a different download from Ensembl (GRCh38.p13.tsv) for mapping between gene names and ENSG IDs, but\n",
      "the issue may persist.\n",
      "\n",
      "Arguments:\n",
      "    gene name: this can either be a gene name in its symbol or common name or it can be a ENSG ID\n",
      "'''\n",
      "\n",
      "\n",
      "**CALL PYTHON SCRIPTS FROM PYTHON**:\n",
      "Use the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\n",
      "\n",
      "To call a Python script `example_script.py` which has no arguments:\n",
      "    \n",
      "```\n",
      "Execute: subprocess.call([sys.executable, '<full path to script/> example_script.py'])\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with one argument:\n",
      "\n",
      "```\n",
      "Execute: subprocess.call([sys.executable, '<full path to script/> example_script.py', 'arg1'])\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with two arguments:\n",
      "```\n",
      "Execute: subprocess.call([sys.executable, '<full path to script/> example_script.py', 'arg1', 'arg2'])\n",
      "```\n",
      "Query:Run the hwgFirstOrderInteractions code on PCNA. Save the output file in a location accessible to the next step with a name such as \"PCNA_hwgFirstOrderInteractions_output.tsv\".\n",
      "\n",
      "**OUTPUT**\n",
      "If output or output files are created, all output files should be named: /home/jpic/BRAD/2024-06-18_18-20-20/<output file>\n",
      "\n",
      "**INSTRUCTIONS**\n",
      "1. Given the user query and the documentation, identify each of the arguments found in the user's query that should be passed to the Python script.\n",
      "2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\n",
      "3. The last line of your response should say \"Execute: <Python code to execute>\"\n",
      "4. Format the response/output as:\n",
      "    Arguments: \n",
      "    Python Code Explanation: <2 sentences maximum>\n",
      "    Execute: <your code here>\n",
      "\n",
      "**IMPORTANT**\n",
      "The code to execute from your response must be formatted as:\n",
      "    Execute: subprocess.call([sys.executable, '<path to python script>', '<argument 1>', '<argument 2>', ..., '<argument n>'])\n",
      "This output should be exactly one line and no longer. Stop the response after this line.\n",
      "\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "LLM OUTPUT PARSER\n",
      " Arguments: The user query specifies the gene name \"PCNA\" and the desired output file name \"PCNA\\_hwgFirstOrderInteractions\\_output.tsv\".\n",
      "\n",
      "Python Code Explanation: Run the hwgFirstOrderInteractions script using the specified argument for the gene name and output file name.\n",
      "\n",
      "Execute: subprocess.call([sys.executable, '/home/jpic/RAG-DEV/tutorials/planner/python/hwgFirstOrderInteractions.py', 'PCNA', '/home/jpic/BRAD/2024-06-18_18-20-20/PCNA_hwgFirstOrderInteractions_output.tsv'])\n",
      "[' Arguments: The user query specifies the gene name \"PCNA\" and the desired output file name \"PCNA\\\\_hwgFirstOrderInteractions\\\\_output.tsv\".\\n\\nPython Code Explanation: Run the hwgFirstOrderInteractions script using the specified argument for the gene name and output file name.\\n\\n', \" subprocess.call([sys.executable, '/home/jpic/RAG-DEV/tutorials/planner/python/hwgFirstOrderInteractions.py', 'PCNA', '/home/jpic/BRAD/2024-06-18_18-20-20/PCNA_hwgFirstOrderInteractions_output.tsv'])\"]\n",
      "Executing Code!!\n",
      "EVAL\n",
      "subprocess.call([sys.executable, '/home/jpic/RAG-DEV/tutorials/planner/python/hwgFirstOrderInteractions.py', 'PCNA', '/home/jpic/BRAD/2024-06-18_18-20-20/PCNA_hwgFirstOrderInteractions_output.tsv'])\n",
      "END EVAL CHECK\n",
      "******************************\n",
      "Finding 1st Order Interactions\n",
      "******************************\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Tue 18 Jun 2024 06:20:34 PM EDT INFO DATABASE\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Debug: PYTHON code executed successfully.\n",
      "==================================================\n",
      "==================================================\n",
      "RAG >> 3: \n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation:\n",
      "Human: make a 2 step process using: Step 1 Method is Code to run the hwgFirstOrderInteractions code on PCNA, and step 2 method is database to load in the Gene column from the output file created in step 1 and query it against ENRICHR. Use suitable file names and paths\n",
      "BRAD:  **Step 1 (method: CODE)**:\n",
      "Prompt: Run the hwgFirstOrderInteractions code on PCNA. Save the output file in a location accessible to the next step with a name such as \"PCNA_hwgFirstOrderInteractions_output.tsv\".\n",
      "\n",
      "**Step 2 (method: DATABASE)**:\n",
      "Prompt: Load the Gene column from the output file created in Step 1, i.e. \"PCNA_hwgFirstOrderInteractions_output.tsv\", into Enrichr. Perform gene set enrichment analysis using the desired tool and save the results in a file named \"PCNA_gene_enrichment_results.txt\".\n",
      "    \n",
      "GENEONTOLOGY: The Gene Ontology (GO) is an initiative to unify the representation of gene and gene product attributes across all species via the aims: 1) maintain and develop its controlled vocabulary of gene and gene product attributes; 2) annotate genes and gene products, and assimilate and disseminate annotation data; and 3) provide tools for easy access to all aspects of the data provided by the project, and to enable functional interpretation of experimental data using the GO.\n",
      "    \n",
      "ENRICHR: is a tool used to lookup sets of genes and their functional association. ENRICHR has access to many gene-set libraries including Allen_Brain_Atlas_up, ENCODE_Histone_Modifications_2015, Enrichr_Libraries_Most_Popular_Genes, FANTOM6_lncRNA_KD_DEGs, GO_Biological_Process_2023, GTEx, Human_Gene_Atlas, KEGG, REACTOME, Transcription_Factor_PPIs, WikiPathways and many others databases.\n",
      "    \n",
      "Available Tables:\n",
      "\n",
      "\n",
      "Query:Load the Gene column from the output file created in Step 1, i.e. \"PCNA_hwgFirstOrderInteractions_output.tsv\", into Enrichr. Perform gene set enrichment analysis using the desired tool and save the results in a file named \"PCNA_gene_enrichment_results.txt\".\n",
      "\n",
      "**INSTRUCTIONS**\n",
      "1. From the query, decide if GENEONTOLOGY or ENRICHR should be used\n",
      "2. Identify genes in the users input that should be searched. propose genes with similar names, correct the users spelling, or make small modifications to the list, but do not propose genes that are not found in the humans input.\n",
      "3. Indicate if genes need to be loaded from a file: True/False\n",
      "\n",
      "database: <ENRICHR or GENEONTOLOGY>\n",
      "genes: <List of genes separated by commas in query or None if code is required>\n",
      "load: <True/False>\n",
      "\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      " database: ENRICHR\n",
      "genes: None (we will load the genes from the file created in the first step)\n",
      "load: True\n",
      " database: ENRICHR\n",
      "genes: None (we will load the genes from the file created in the first step)\n",
      "load: True\n",
      "{'database': 'ENRICHR', 'genes': ['None (we will load the genes from the file created in the first step)'], 'load': 'True'}\n",
      "['log.json', 'S1-PCNA_hwgFirstOrderInteractions_output.tsv']\n",
      "**INSTRUCTIONS**\n",
      "You are a digital assistant responsible for determining which file we should load based on a users prompt. You many choose either the files listed under AVAILABLE FILES or found in the USER QUERY. Based upon the user query, specify which if any fields should be loaded from the table. Rather than specifying a table name, the user might specify ot load data from a particular step. In this case, files are saved with the prefix `S<number>-<File Name>` to denote which step to load them from.\n",
      "\n",
      "**EXAMPLES**\n",
      "\n",
      "Here the files are selected by name:\n",
      "```\n",
      "Available Files:\n",
      "S1-PYTHON.tsv\n",
      "S2-ENRICHR.csv\n",
      "\n",
      "User Query:\n",
      "Use the pathways output from enricher for the this phase of anlaysis.\n",
      "\n",
      "Output:\n",
      "File: S2-ENRICHR.csv\n",
      "Fields: pathways\n",
      "```\n",
      "\n",
      "Here the files are selected by Step:\n",
      "```\n",
      "Available Files:\n",
      "S1-PYTHON.tsv\n",
      "S2-ENRICHR.csv\n",
      "\n",
      "User Query:\n",
      "Use the Genes identified in Step 1 as input to the codes.\n",
      "\n",
      "Output:\n",
      "File: S1-PYTHON.tsv\n",
      "Fields: Genes\n",
      "```\n",
      "\n",
      "**AVAILABLE FILES**\n",
      "log.json\n",
      "S1-PCNA_hwgFirstOrderInteractions_output.tsv\n",
      "\n",
      "**USER QUERY**\n",
      "{user_query}\n",
      "\n",
      "**OUTPUT**\n",
      "File: <which file you selected>\n",
      "Fields: <which fields in the file are important>\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Tue 18 Jun 2024 06:20:36 PM EDT INFO Performing Enichr analysis using database GO_Biological_Process_2021.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File=S1-PCNA_hwgFirstOrderInteractions_output.tsv\n",
      "Fields=Genes\n",
      "['CUTC', 'CDC5L', 'PIMREG', 'RBMX', 'TRMT11', 'ANP32B', 'CENPM', 'ATRX', 'SLF1', 'RPA4', 'CCT5', 'CCT8', 'PIK3C3', 'CDKN2B', 'COL4A1', 'TPX2', 'XRCC6', 'CHEK2', 'USP18', 'COL4A5', 'INCENP', 'CCNO', 'RPS10-NUDT3', 'HSPH1', 'PRPS2', 'DSTN', 'RANGAP1', 'PACSIN3', 'L3MBTL1', 'WDHD1', 'NUP155', 'SHMT2', 'DHX15', 'CNOT7', 'MIS12', 'UHRF1', 'AP5Z1', 'UBR7', 'CDCA2', 'MAGOHB', 'REV3L', 'APEX2', 'CDC45', 'PI4KA', 'GPD1', 'ANG', 'NNT', 'MOCS1', 'CKS2', 'PPCDC', 'CCNB3', 'TTK', 'DIAPH3', 'RNPS1', 'FTSJ1', 'SUV39H2', 'PRTFDC1', 'POLR2E', 'SRSF3', 'MCM3', 'COL21A1', 'METTL15', 'RNF8', 'TFDP1', 'EXOSC8', 'CDC123', 'TEX30', 'UBE2C', 'TYMS', 'CCNY', 'PSMD10', 'RACGAP1', 'CCNA1', 'CDKAL1', 'MX2', 'PAPOLB', 'DSN1', 'RPUSD2', 'AAR2', 'PDS5B', 'DONSON', 'MGME1', 'PIAS2', 'SNU13', 'SETBP1', 'RPL36A-HNRNPH2', 'SNRPD3', 'WRN', 'SMC1A', 'EIF2S2', 'WARS1', 'YWHAB', 'NDC80', 'ANP32A', 'MCM5', 'RAD54B', 'HMGN1', 'IAH1', 'VBP1', 'CA2', 'MAGED1', 'PSMG2', 'E2F1', 'DEPDC1B', 'E2F5', 'NUDT5', 'ADAT2', 'ESCO1', 'AFG3L2', 'RTCB', 'MRPS6', 'CCDC33', 'PAPOLG', 'PGAM1', 'KPNA5', 'RBBP7', 'BORA', 'CHAF1B', 'AARS2', 'UBE2D1', 'MCAT', 'HIRA', 'N6AMT1', 'SNRPC', 'RAD21L1', 'E2F8', 'RAD23B', 'KIN', 'UBL4A', 'POLE2', 'MSH2', 'PGK2', 'PRMT3', 'CENPQ', 'CMTM5', 'KPNA3', 'PDSS1', 'PNMA6A', 'BMS1', 'RPS12', 'CCNH', 'ELP3', 'DHTKD1', 'POLD3', 'BCCIP', 'LDHAL6B', 'BCAT1', 'CDADC1', 'SMARCA2', 'DDX11', 'SMC1B', 'FOXO3', 'CSTF1', 'CKAP2', 'SNX5', 'COX5A', 'LMNB1', 'MAP1LC3A', 'DNTT', 'UFD1', 'CENPA', 'TSPEAR', 'ESCO2', 'RBBP8', 'MGMT', 'ITM2A', 'TXNDC5', 'NUDT15', 'DEK', 'POLN', 'BCL2L1', 'POLR2F', 'SPAG6', 'MT-ATP6', 'HDAC2', nan, 'CARS2', 'PDS5A', 'RIOK2', 'MRPS28', 'PRPS1L1', 'TFDP3', 'OXA1L', 'HSPA13', 'H3-5', 'MSH3', 'SUB1', 'OAT', 'FAM83D', 'HMGB1', 'H2AB3', 'BUB3', 'INTS13', 'HACD3', 'SLC2A13', 'PMS2', 'PCNA', 'RAD50', 'PHF11', 'PSAT1', 'DPM1', 'MT-CYB', 'TRUB1', 'RBL1', 'DSCC1', 'SKA1', 'MAT1A', 'PHF21B', 'EIF3M', 'RPL36A', 'MBD1', 'MNAT1', 'RAD51', 'MCM8', 'RANBP1', 'UBQLN2', 'ASMTL', 'SNRPB2', 'RNASEH2B', 'AURKA', 'MPHOSPH8', 'TFAM', 'MTERF3', 'RFC3', 'PRPS1', 'TXNRD2', 'NONO', 'SUGT1', 'ANK3', 'MKKS', 'RAB9B', 'GTF2A2', 'BRCA2', 'GEN1', 'RPL36AL', 'XPNPEP1', 'AHCY', 'RPS7', 'AIRE', 'HMGA1', 'NOP56', 'PSMA4', 'CDCA4', 'PDXK', 'HMGB3', 'RAD1', 'CHD6', 'UBR5', 'MRPL13', 'RB1', 'OSGEP', 'PFKM', 'PSMC6', 'RBX1', 'RCN2', 'EP300', 'CCNB2', 'TRDMT1', 'MOCS3', 'HOGA1', 'VIRMA', 'APEX1', 'EIF2AK1', 'SLC6A8', 'ITM2B', 'HDDC2', 'POLI', 'ASPG', 'CRLS1', 'EEF1E1', 'SLC27A5', 'UBL7', 'PRIM2', 'RAB33A', 'ACP2', 'MTRF1', 'TGDS', 'UBE2J1', 'YWHAQ', 'PFKL', 'LIG4', 'GADD45G', 'RPA3', 'PSMB5', 'SUMO4', 'MTG2', 'PRPF19', 'MAPRE1', 'GNPNAT1', 'ASB8', 'MCIDAS', 'NUBPL', 'TESMIN', 'SMC6', 'COL6A2', 'LRR1', 'ATAD2B', 'MTFR2', 'BABAM1', 'DHX32', 'MX1', 'CCNQ', 'CSNK1E', 'NUP50', 'SEC61A2', 'UBQLN1', 'PRDX3', 'AIMP2', 'POLA1', 'GRPEL1', 'PSMA7', 'MPI', 'SETDB2', 'EEF1G', 'PFDN4', 'CCAR1', 'SLC27A2', 'RPS4X', 'MRPL39', 'HSPA12B', 'CTPS2', 'MAFB', 'RRAS2', 'AGAP9', 'ERCC6L', 'NANOG', 'RANBP6', 'SEC11A', 'HSCB', 'ZWINT', 'GART', 'IPO5', 'XRCC3', 'RPL3', 'RBPJL', 'PRMT5', 'MLH3', 'SUPT16H', 'PRMT8', 'CCNK', 'RPAP3', 'HTRA1', 'RPL17-C18orf32', 'SUMO3', 'PDSS2', 'C1QBP', 'ZNF367', 'THNSL1', 'PRMT2', 'CSE1L', 'CSNK2A1', 'BRIX1', 'TIMELESS', 'TSPYL4', 'MIPEP', 'CSNK1A1L', 'ENO4', 'AK3', 'TRMU', 'MRPL15', 'CAPRIN1', 'COL9A3', 'SEPHS1', 'WDR4', 'PSMG1', 'ING1', 'EIF2AK4', 'SIVA1', 'SPO11', 'MIS18A', 'APC', 'RTF2', 'SNRPD1', 'CENPK', 'MKI67', 'PBK', 'POGLUT2', 'GTPBP3', 'SCML2', 'DDX39A', 'SKA3', 'TSPYL5', 'COL17A1', 'TXN', 'DHX33', 'CUL4A', 'AK7', 'CDC25B', 'ARHGAP4', 'POLL', 'COL19A1', 'HAUS6', 'ATP5F1C', 'TOP1', 'COL4A2', 'RIDA', 'MRPL16', 'ERCC5', 'DHX35', 'ZFAND4', 'ARMC1', 'APMAP', 'UHRF2', 'RPP38', 'COL6A1', 'NUDT12', 'DNA2', 'LDHC', 'ADSL', 'UROS', 'NUS1', 'CPSF3', 'CCNT1', 'MYBL2']\n",
      "The following table was generated by quering the gene list against GO_Biological_Process_2021:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "</style>\n",
       "<table id=\"T_97fb8\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th id=\"T_97fb8_level0_col0\" class=\"col_heading level0 col0\" >rank</th>\n",
       "      <th id=\"T_97fb8_level0_col1\" class=\"col_heading level0 col1\" >path_name</th>\n",
       "      <th id=\"T_97fb8_level0_col2\" class=\"col_heading level0 col2\" >p_val</th>\n",
       "      <th id=\"T_97fb8_level0_col3\" class=\"col_heading level0 col3\" >z_score</th>\n",
       "      <th id=\"T_97fb8_level0_col4\" class=\"col_heading level0 col4\" >combined_score</th>\n",
       "      <th id=\"T_97fb8_level0_col5\" class=\"col_heading level0 col5\" >overlapping_genes</th>\n",
       "      <th id=\"T_97fb8_level0_col6\" class=\"col_heading level0 col6\" >adj_p_val</th>\n",
       "      <th id=\"T_97fb8_level0_col7\" class=\"col_heading level0 col7\" >database</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_97fb8_level0_row0\" class=\"row_heading level0 row0\" >0</th>\n",
       "      <td id=\"T_97fb8_row0_col0\" class=\"data row0 col0\" >1</td>\n",
       "      <td id=\"T_97fb8_row0_col1\" class=\"data row0 col1\" >DNA metabolic process (GO:0006259)</td>\n",
       "      <td id=\"T_97fb8_row0_col2\" class=\"data row0 col2\" >0.000000</td>\n",
       "      <td id=\"T_97fb8_row0_col3\" class=\"data row0 col3\" >9.904695</td>\n",
       "      <td id=\"T_97fb8_row0_col4\" class=\"data row0 col4\" >582.320047</td>\n",
       "      <td id=\"T_97fb8_row0_col5\" class=\"data row0 col5\" >['PRIM2', 'PCNA', 'UHRF2', 'MCM8', 'UHRF1', 'HMGB3', 'PDS5B', 'HMGB1', 'TYMS', 'PDS5A', 'POLD3', 'WRN', 'CDC45', 'RBBP8', 'POLI', 'POLL', 'NUDT15', 'GEN1', 'WDHD1', 'XRCC6', 'RFC3', 'MGMT', 'XRCC3', 'LIG4', 'DNTT', 'CSNK1E', 'SMC1A', 'POLA1', 'RAD50', 'RAD51', 'MSH2', 'MSH3', 'APEX1', 'RPA3', 'POLE2', 'RPA4', 'TIMELESS', 'MCM3', 'RAD1', 'MCM5', 'ERCC5', 'DNA2', 'TOP1', 'MNAT1']</td>\n",
       "      <td id=\"T_97fb8_row0_col6\" class=\"data row0 col6\" >0.000000</td>\n",
       "      <td id=\"T_97fb8_row0_col7\" class=\"data row0 col7\" >GO_Biological_Process_2021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_97fb8_level0_row1\" class=\"row_heading level0 row1\" >1</th>\n",
       "      <td id=\"T_97fb8_row1_col0\" class=\"data row1 col0\" >2</td>\n",
       "      <td id=\"T_97fb8_row1_col1\" class=\"data row1 col1\" >DNA repair (GO:0006281)</td>\n",
       "      <td id=\"T_97fb8_row1_col2\" class=\"data row1 col2\" >0.000000</td>\n",
       "      <td id=\"T_97fb8_row1_col3\" class=\"data row1 col3\" >7.781802</td>\n",
       "      <td id=\"T_97fb8_row1_col4\" class=\"data row1 col4\" >350.042919</td>\n",
       "      <td id=\"T_97fb8_row1_col5\" class=\"data row1 col5\" >['PCNA', 'RNF8', 'PDS5B', 'BRCA2', 'PDS5A', 'BABAM1', 'POLD3', 'WRN', 'CHEK2', 'RBBP8', 'PMS2', 'POLI', 'POLL', 'WDHD1', 'MGME1', 'RFC3', 'POLN', 'MGMT', 'XRCC3', 'APEX2', 'LIG4', 'CSNK1E', 'SMC1A', 'RAD23B', 'MLH3', 'POLA1', 'RAD50', 'RAD51', 'MSH2', 'MSH3', 'APEX1', 'RPA3', 'POLE2', 'RPA4', 'TIMELESS', 'RAD1', 'ERCC5', 'DNA2', 'MNAT1']</td>\n",
       "      <td id=\"T_97fb8_row1_col6\" class=\"data row1 col6\" >0.000000</td>\n",
       "      <td id=\"T_97fb8_row1_col7\" class=\"data row1 col7\" >GO_Biological_Process_2021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_97fb8_level0_row2\" class=\"row_heading level0 row2\" >2</th>\n",
       "      <td id=\"T_97fb8_row2_col0\" class=\"data row2 col0\" >3</td>\n",
       "      <td id=\"T_97fb8_row2_col1\" class=\"data row2 col1\" >double-strand break repair (GO:0006302)</td>\n",
       "      <td id=\"T_97fb8_row2_col2\" class=\"data row2 col2\" >0.000000</td>\n",
       "      <td id=\"T_97fb8_row2_col3\" class=\"data row2 col3\" >9.930600</td>\n",
       "      <td id=\"T_97fb8_row2_col4\" class=\"data row2 col4\" >370.620304</td>\n",
       "      <td id=\"T_97fb8_row2_col5\" class=\"data row2 col5\" >['AP5Z1', 'MCM8', 'RNF8', 'SMC6', 'HMGB1', 'BRCA2', 'PRPF19', 'RAD54B', 'BABAM1', 'WRN', 'CHEK2', 'REV3L', 'RBBP8', 'POLL', 'GEN1', 'XRCC6', 'POLN', 'XRCC3', 'LIG4', 'DNTT', 'POLA1', 'RAD50', 'RAD51', 'RPA3', 'RPA4', 'ERCC5', 'DNA2']</td>\n",
       "      <td id=\"T_97fb8_row2_col6\" class=\"data row2 col6\" >0.000000</td>\n",
       "      <td id=\"T_97fb8_row2_col7\" class=\"data row2 col7\" >GO_Biological_Process_2021</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x150ba8658fd0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n"
     ]
    }
   ],
   "source": [
    "brad.chat(ragvectordb = vectordb)"
   ]
  },
  {
   "cell_type": "raw",
   "id": "e92833ca-e796-467a-888d-ebe8d5a576a0",
   "metadata": {},
   "source": [
    "/force PLANNER make a 2 step process using: Step 1 Method is Code to run the hwgFirstOrderInteractions code on PCNA, and step 2 method is database to load in the Gene column from the output file created in step 1 and query it against ENRICHR. Use suitable file names and paths"
   ]
  },
  {
   "cell_type": "raw",
   "id": "78989e5c-3917-4279-89fb-36c6bdac9310",
   "metadata": {},
   "source": [
    "/force PLANNER make a 2 step process using: Step 1 Method is Code to run the hwgFirstOrderInteractions code on PCNA, and step 2 method is database to load in the Gene column from the output file created in step 1 and query it against ENRICHR. Use suitable file names and paths"
   ]
  },
  {
   "cell_type": "raw",
   "id": "67f6fd69-8af2-4586-878c-50036072656b",
   "metadata": {},
   "source": [
    "/force PLANNER make a 2 step process using: Step 1 Method is Code to run the hwgFirstOrderInteractions code on PCNA, and step 2 method is database to load in the Gene column from the output file created in step 1 and query it against ENRICHR. Use suitable file names for saving and loading"
   ]
  },
  {
   "cell_type": "raw",
   "id": "7a0bea15-dd33-4988-8066-1005c59ef0a4",
   "metadata": {},
   "source": [
    "/force PLANNER make a 2 step process using: Step 1 Method is Code to run the hwgFirstOrderInteractions code on PCNA, and step 2 method is database to take the output of step 1 and load it to query the genes against ENRICHR"
   ]
  },
  {
   "cell_type": "raw",
   "id": "40625c32-3a23-4caa-b139-cbc08adc9bf7",
   "metadata": {},
   "source": [
    "/force PLANNER make a three step process to find the latest research on MYOD. Start by looking up the gene in erichr"
   ]
  },
  {
   "cell_type": "raw",
   "id": "5e48c8a2-073e-4ef1-b02d-867aa191c6f1",
   "metadata": {},
   "source": [
    "/force PLANNER design an analysis pipeline to find the first order interactions of PCNA in the hardwired genome (run the hwgFirstOrderInteractions). First run the hwgFirstOrderInteractions code and then take the outputs and put them into ENRICHR"
   ]
  },
  {
   "cell_type": "raw",
   "id": "16dcb516-22c1-49fb-ac95-23d40dea843f",
   "metadata": {},
   "source": [
    "/force DATABASE search MYOD, CMYC, PCNA, and CTD1 in ENRICHR"
   ]
  },
  {
   "cell_type": "raw",
   "id": "522ad48c-92eb-479c-a974-2d3a548b4151",
   "metadata": {},
   "source": [
    "/force PLANNER plan a two step process where first you run the hwg code and second you query the output genes from hwg codes into enrichr"
   ]
  },
  {
   "cell_type": "raw",
   "id": "eb0d5a2a-9941-4519-8b29-ff5133411852",
   "metadata": {},
   "source": [
    "/force PLANNER design an analysis pipeline to find the first order interactions of MYOD in the hardwired genome. Write a report about the role of the MYOD gene and which pathways, cell types, and genes are related to MYOD. The report should be structured with 1. an introduction on MYOD and its role in cellular processes. 2. details of MYOD based on the hardwired genome analysis pipeline and searching the neighboring genes in enrichr. and 3. a conclussion pointing to recent literature studying MYOD."
   ]
  },
  {
   "cell_type": "raw",
   "id": "7d5e37f4-1d55-4d88-a127-d5f143c8cf68",
   "metadata": {},
   "source": [
    "I would like steps 7 - 10 to be condensed to a single writing step. The prompt for the new step should contain the instructions to write the different sections specified in the current steps 7-10"
   ]
  },
  {
   "cell_type": "raw",
   "id": "8fa7ccf5-14b6-4671-a258-6f4e6502170f",
   "metadata": {},
   "source": [
    "/force PLANNER design an analysis pipeline to find the first order interactions of PCNA in the hardwired genome. Write a report about the role of the PCNA gene and which pathways, cell types, and genes are related to PCNA."
   ]
  },
  {
   "cell_type": "raw",
   "id": "25062031-73a1-47d8-9f28-97026601a1e9",
   "metadata": {},
   "source": [
    "/force PLANNER design an analysis pipeline to find the first order interactions of MYOD in the hardwired genome. Write a report about the role of the MYOD gene and which pathways, cell types, and genes are related to MYOD."
   ]
  },
  {
   "cell_type": "raw",
   "id": "97d6aa41-2a87-48d7-8a3d-d8a4cfa3d043",
   "metadata": {},
   "source": [
    "/force PLANNER design an analysis pipeline to find the first order interactions of MYOD in the hardwired genome. Write a small report about which pathways and cell types these genes are implicated in "
   ]
  },
  {
   "cell_type": "raw",
   "id": "233d47c7-8584-4a4d-a0f5-bb4e53cc7739",
   "metadata": {},
   "source": [
    "/force PLANNER design an analysis pipeline to find the first order interactions of MYOD in the hardwired genome from HWG.mat. Write a small report about which pathways and cell types these genes are implicated in "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d205ed67-2b4e-494d-b8cc-982aa33cdcc4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "You must select which code to run to help a user.\n",
      "\n",
      "**Available Scripts**\n",
      "Script Name: hwgFirstOrderInteractions, \t Description: \n",
      "\n",
      "\n",
      "**User Query**\n",
      "{user_query}\n",
      "\n",
      "**Task**\n",
      "Based on the user's query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\n",
      "\n",
      "**Response Template**\n",
      "SCRIPT: <selected script>\n",
      "REASON: <reasoning why the script fits the user prompt>\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(\"You must select which code to run to help a user.\\n\\n**Available Scripts**\\nScript Name: hwgFirstOrderInteractions, \\t Description: \\n\\n\\n**User Query**\\n{user_query}\\n\\n**Task**\\nBased on the user's query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\\n\\n**Response Template**\\nSCRIPT: <selected script>\\nREASON: <reasoning why the script fits the user prompt>\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1964016f-c6f7-4c18-bc69-7955fe0b758f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "''"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from BRAD.pythonCaller import get_py_description\n",
    "get_py_description('/home/jpic/RAG-DEV/tutorials/planner/python/hwgFirstOrderInteractions.py')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c11d01a4-8878-410c-b761-e0d235655981",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current conversation:\n",
      "{history}\n",
      "\n",
      "**PYTHON SCRIPT**\n",
      "You must run this python script: \n",
      "/home/jpic/RAG-DEV/tutorials/planner/python/hwgFirstOrderInteractions.py\n",
      "\n",
      "**PYTHON SCRIPT DOCUMENTATION**:\n",
      "This is the doc string of this python script:\n",
      "'''Find the first order interactions of a gene in the Hardwired Genome\n",
      "\n",
      "Usage:\n",
      "    python <path/to/script/>hwgFirstOrderInteractions.py <gene name>\n",
      "\n",
      "This script uses the Hardwired Genome construction. Note: Cooper recently found a bug in how the data are integrated for the Hardwired Genome.\n",
      "The code here should side step that by using a different download from Ensembl (GRCh38.p13.tsv) for mapping between gene names and ENSG IDs, but\n",
      "the issue may persist.\n",
      "\n",
      "Arguments:\n",
      "    gene name: this can either be a gene name in its symbol or common name or it can be a ENSG ID\n",
      "'''\n",
      "\n",
      "\n",
      "**CALL PYTHON SCRIPTS FROM PYTHON**:\n",
      "Use the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\n",
      "\n",
      "To call a Python script `example_script.py` which has no arguments:\n",
      "    \n",
      "```\n",
      "Execute: subprocess.call([sys.executable, '<full path to script/> example_script.py'])\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with one argument:\n",
      "\n",
      "```\n",
      "Execute: subprocess.call([sys.executable, '<full path to script/> example_script.py', 'arg1'])\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with two arguments:\n",
      "```\n",
      "Execute: subprocess.call([sys.executable, '<full path to script/> example_script.py', 'arg1', 'arg2'])\n",
      "```\n",
      "Query:{input}\n",
      "\n",
      "\n",
      "**INSTRUCTIONS**\n",
      "1. Given the user query and the documentation, identify each of the arguments found in the user's query that should be passed to the Python script.\n",
      "2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\n",
      "3. The last line of your response should say \"Execute: <Python code to execute>\"\n",
      "4. Format the response/output as:\n",
      "    Arguments: \n",
      "    Python Code Explanation: <2 sentences maximum>\n",
      "    Execute: <your code here>\n",
      "\n",
      "**IMPORTANT**\n",
      "The code to execute from your response must be formatted as:\n",
      "    Execute: subprocess.call([sys.executable, '<path to python script>', '<argument 1>', '<argument 2>', ..., '<argument n>'])\n",
      "This output should be exactly one line and no longer. Stop the response after this line.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print('Current conversation:\\n{history}\\n\\n**PYTHON SCRIPT**\\nYou must run this python script: \\n/home/jpic/RAG-DEV/tutorials/planner/python/hwgFirstOrderInteractions.py\\n\\n**PYTHON SCRIPT DOCUMENTATION**:\\nThis is the doc string of this python script:\\n\\'\\'\\'Find the first order interactions of a gene in the Hardwired Genome\\n\\nUsage:\\n    python <path/to/script/>hwgFirstOrderInteractions.py <gene name>\\n\\nThis script uses the Hardwired Genome construction. Note: Cooper recently found a bug in how the data are integrated for the Hardwired Genome.\\nThe code here should side step that by using a different download from Ensembl (GRCh38.p13.tsv) for mapping between gene names and ENSG IDs, but\\nthe issue may persist.\\n\\nArguments:\\n    gene name: this can either be a gene name in its symbol or common name or it can be a ENSG ID\\n\\'\\'\\'\\n\\n\\n**CALL PYTHON SCRIPTS FROM PYTHON**:\\nUse the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\\n\\nTo call a Python script `example_script.py` which has no arguments:\\n    \\n```\\nExecute: subprocess.call([sys.executable, \\'<full path to script/> example_script.py\\'])\\n```\\n\\nTo call a Python script `example_script.py` with one argument:\\n\\n```\\nExecute: subprocess.call([sys.executable, \\'<full path to script/> example_script.py\\', \\'arg1\\'])\\n```\\n\\nTo call a Python script `example_script.py` with two arguments:\\n```\\nExecute: subprocess.call([sys.executable, \\'<full path to script/> example_script.py\\', \\'arg1\\', \\'arg2\\'])\\n```\\nQuery:{input}\\n\\n\\n**INSTRUCTIONS**\\n1. Given the user query and the documentation, identify each of the arguments found in the user\\'s query that should be passed to the Python script.\\n2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\\n3. The last line of your response should say \"Execute: <Python code to execute>\"\\n4. Format the response/output as:\\n    Arguments: \\n    Python Code Explanation: <2 sentences maximum>\\n    Execute: <your code here>\\n\\n**IMPORTANT**\\nThe code to execute from your response must be formatted as:\\n    Execute: subprocess.call([sys.executable, \\'<path to python script>\\', \\'<argument 1>\\', \\'<argument 2>\\', ..., \\'<argument n>\\'])\\nThis output should be exactly one line and no longer. Stop the response after this line.\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73a585c4-e1a0-4264-98d5-9c61911e84b8",
   "metadata": {},
   "source": [
    "# Write a python script to use"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "21c3d231-14a7-49d2-8d88-2cf204a214cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import scipy.io as sio\n",
    "import numpy as np\n",
    "from scipy.sparse import csc_matrix\n",
    "\n",
    "geneName = 'ENSG00000129152'\n",
    "\n",
    "HWG = sio.loadmat('/home/jpic/RAG-DEV/tutorials/planner/HWG.mat')            # Load Hardwired Genome A matrix\n",
    "A = HWG['HWG']['A']\n",
    "A = A[0,0].todense()\n",
    "df = pd.read_csv('/home/jpic/RAG-DEV/tutorials/planner/geneIndexTable.csv')  # Load gene index table\n",
    "df2 = pd.read_csv('GRCh38.p13.tsv', delimiter='\\t')                          # Load GRCH38 to map between gene names and ENSG IDs\n",
    "if df[df['Stable ID'].isin([geneName])].shape[0] == 0:\n",
    "    temp0 = df2[df2['Gene name'] == geneName]                                    # Find ENSG of Gene Name\n",
    "    ENSG1 = temp0['Gene stable ID'].values[0]\n",
    "else:\n",
    "    ENSG1 = geneName\n",
    "idx = df[df['Stable ID'] == ENSG1].index[0]                                  # Find position of target gene in HWG\n",
    "idxs = np.where(A[idx,:] == 1)[1]                                            # Find first order interactions in network\n",
    "firstOrderEnsg = list(df.iloc[idxs]['Stable ID'].values)                     # Get ENSG of first order interactions\n",
    "firstOrderGenes = list(set(df2[df2['Gene stable ID'].isin(firstOrderEnsg)]['Gene name'].values)) # Get gene names of first order contacts\n",
    "dfSave = pd.DataFrame({'Genes':firstOrderGenes})\n",
    "dfSave.to_csv('HWG_first_order_interactions.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "dcdfbafb-5bc9-4664-9e44-9c77b4c8084f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['RORA',\n",
       " 'CDH2',\n",
       " 'HDAC8',\n",
       " 'MSGN1',\n",
       " 'EBF4',\n",
       " 'TUBA8',\n",
       " 'ADCK1',\n",
       " 'SOX11',\n",
       " 'OXCT1',\n",
       " 'TERT',\n",
       " 'FGF3',\n",
       " 'VGLL2',\n",
       " 'RUNX1',\n",
       " 'IRX1',\n",
       " 'E2F5',\n",
       " 'FST',\n",
       " 'HOXC9',\n",
       " 'ROCK1',\n",
       " 'STAU1',\n",
       " 'SIM2',\n",
       " 'HSP90AB1',\n",
       " 'ZNF219',\n",
       " 'COL19A1',\n",
       " 'FRMD6',\n",
       " 'TCF15',\n",
       " 'LRP5',\n",
       " 'SMARCA2',\n",
       " 'SOX17',\n",
       " 'SGCG',\n",
       " 'CSRP3',\n",
       " 'SMTN',\n",
       " 'TCF7',\n",
       " 'ZNF143',\n",
       " 'CLOCK',\n",
       " 'EYA1',\n",
       " 'POU3F2',\n",
       " 'MKX',\n",
       " 'TNMD',\n",
       " 'CEBPG',\n",
       " 'SLC38A2',\n",
       " 'ANXA11',\n",
       " 'SIX4',\n",
       " 'GABARAPL1',\n",
       " 'BDNF',\n",
       " 'BMP5',\n",
       " 'YY2',\n",
       " 'PBX4',\n",
       " 'SLC8A3',\n",
       " 'RBL1',\n",
       " 'PVALB',\n",
       " 'MEF2B',\n",
       " 'HSPB3',\n",
       " 'HDAC10',\n",
       " 'KRT5',\n",
       " 'DBX1',\n",
       " 'MKI67',\n",
       " 'SALL4',\n",
       " 'YY1',\n",
       " 'TSPAN9',\n",
       " 'TBX1',\n",
       " 'JDP2',\n",
       " 'THBS1',\n",
       " 'ZBED4',\n",
       " 'MAPK12',\n",
       " 'CAPN3',\n",
       " 'JARID2',\n",
       " 'MYCN',\n",
       " 'NCOA3',\n",
       " 'SOX7',\n",
       " 'TBX18',\n",
       " 'NEO1',\n",
       " 'ARID3A',\n",
       " 'ACTC1',\n",
       " 'FOSL2',\n",
       " 'HOXC10',\n",
       " 'ATG5',\n",
       " 'LMNB1',\n",
       " 'CHGA',\n",
       " 'FOS',\n",
       " 'TFE3',\n",
       " 'INSR',\n",
       " 'BMP4',\n",
       " 'FOXO4',\n",
       " 'LMO7',\n",
       " 'SIX6',\n",
       " 'SETD3',\n",
       " 'TGFB3',\n",
       " 'SLC1A1',\n",
       " 'DACH1',\n",
       " 'FOXG1',\n",
       " 'DCD',\n",
       " 'SDC4',\n",
       " 'ANKRD2',\n",
       " 'PTF1A',\n",
       " 'GDF5',\n",
       " 'ACSS3',\n",
       " 'FBXO32',\n",
       " 'SULF1',\n",
       " 'UHRF1',\n",
       " 'TCF4',\n",
       " 'TPM4',\n",
       " 'MYO18B',\n",
       " 'PKNOX1',\n",
       " 'NKX3-1',\n",
       " 'PAX5',\n",
       " 'CCDC6',\n",
       " 'HEY2',\n",
       " 'EP300',\n",
       " 'PCDH8',\n",
       " 'CAT',\n",
       " 'CNTF',\n",
       " 'KLHL31',\n",
       " 'ZIC3',\n",
       " 'SMARCD1',\n",
       " 'H3-5',\n",
       " 'EYA2',\n",
       " 'CEBPA',\n",
       " 'MSX1',\n",
       " 'ALDH1A2',\n",
       " 'BMP6',\n",
       " 'ZNRF3',\n",
       " 'SNAI2',\n",
       " 'LARGE1',\n",
       " 'KPNA3',\n",
       " 'FOXA2',\n",
       " 'MAPK1',\n",
       " 'NANOG',\n",
       " 'TNNC2',\n",
       " 'SERGEF',\n",
       " 'WNT7B',\n",
       " 'HEY1',\n",
       " 'HDAC2',\n",
       " 'GDI1',\n",
       " 'DGCR8',\n",
       " 'TRAF7',\n",
       " 'RHOXF1',\n",
       " 'COL4A1',\n",
       " 'GDF7',\n",
       " 'CYP26A1',\n",
       " 'DEK',\n",
       " 'MMP9',\n",
       " 'SOX21',\n",
       " 'GJA1',\n",
       " 'NKX2-2',\n",
       " 'VGLL1',\n",
       " 'MZF1',\n",
       " 'MYH7',\n",
       " 'MGMT',\n",
       " 'KLF5',\n",
       " 'NKX3-2',\n",
       " 'SON',\n",
       " 'AR',\n",
       " 'CEBPD',\n",
       " 'KLF12',\n",
       " 'MYLK4',\n",
       " 'NFIC',\n",
       " 'HMGB3',\n",
       " 'PINX1',\n",
       " 'TRIM27',\n",
       " 'KLF8',\n",
       " 'COL4A2',\n",
       " 'SOX4',\n",
       " 'TNFSF11',\n",
       " 'PML',\n",
       " 'DTNA',\n",
       " 'FGF16',\n",
       " 'SS18',\n",
       " 'PCDH17',\n",
       " 'GATA6',\n",
       " 'S100B',\n",
       " 'TET1',\n",
       " 'AKAP4',\n",
       " 'MMUT',\n",
       " 'TSHZ3',\n",
       " 'MYLK2',\n",
       " 'SMAD7',\n",
       " 'ASH2L',\n",
       " 'SCD',\n",
       " 'FOXO1',\n",
       " 'CABIN1',\n",
       " 'NKX6-2',\n",
       " 'IRS2',\n",
       " 'FOXC1',\n",
       " 'MEGF10',\n",
       " 'ZIC2',\n",
       " 'LATS2',\n",
       " 'CSPG4',\n",
       " 'LDHC',\n",
       " 'JAG1',\n",
       " 'PMEPA1',\n",
       " 'FOXK1',\n",
       " 'DCT',\n",
       " 'AXIN1',\n",
       " 'TFAM',\n",
       " 'SMAD9',\n",
       " nan,\n",
       " 'HOXC8',\n",
       " 'TRDN',\n",
       " 'LDLRAD4',\n",
       " 'DIO3',\n",
       " 'FOXO3',\n",
       " 'SNAI1',\n",
       " 'ID4',\n",
       " 'HIRA',\n",
       " 'PPARGC1A',\n",
       " 'TIMP3',\n",
       " 'ETS2',\n",
       " 'MYOF',\n",
       " 'CCNA1',\n",
       " 'WNT8B',\n",
       " 'CHRNA4',\n",
       " 'SMAD2',\n",
       " 'FOXF2',\n",
       " 'BMP7',\n",
       " 'MEIS2',\n",
       " 'SMPX',\n",
       " 'POLA1',\n",
       " 'LRRC30',\n",
       " 'EBF2',\n",
       " 'NFATC1',\n",
       " 'GDF3',\n",
       " 'BMP2',\n",
       " 'ITGA5',\n",
       " 'CDKN2B',\n",
       " 'ITGB1BP2',\n",
       " 'KLF6',\n",
       " 'FKBP1A',\n",
       " 'DLL4',\n",
       " 'EIF4EBP1',\n",
       " 'SMARCB1',\n",
       " 'HSPH1',\n",
       " 'MBNL2',\n",
       " 'CNN2',\n",
       " 'MYL12B',\n",
       " 'PRMT5',\n",
       " 'BNIP2',\n",
       " 'ERG',\n",
       " 'SUV39H2',\n",
       " 'EEF1A2',\n",
       " 'TMOD1',\n",
       " 'RUNX2',\n",
       " 'PACSIN3',\n",
       " 'GATA5',\n",
       " 'EWSR1',\n",
       " 'ID1',\n",
       " 'DYRK1A',\n",
       " 'ANK3',\n",
       " 'GDF2',\n",
       " 'NEUROG1',\n",
       " 'MYL9',\n",
       " 'ZFAND5',\n",
       " 'IRX4',\n",
       " 'DKK1',\n",
       " 'LARP6',\n",
       " 'ALDH1A1',\n",
       " 'ACP1',\n",
       " 'ADD3',\n",
       " 'SMAD6',\n",
       " 'FOXB1',\n",
       " 'CKMT2',\n",
       " 'COL6A1',\n",
       " 'RIPPLY1',\n",
       " 'ETV1',\n",
       " 'SLC2A4RG',\n",
       " 'LPIN3',\n",
       " 'MYL12A',\n",
       " 'LPL',\n",
       " 'MCAT',\n",
       " 'CDX2',\n",
       " 'CIDEA',\n",
       " 'HOXC11',\n",
       " 'CCNT1',\n",
       " 'RBM24',\n",
       " 'MYOM1',\n",
       " 'TMTC1',\n",
       " 'LGR4',\n",
       " 'CTNNBL1',\n",
       " 'GDNF',\n",
       " 'MEOX2',\n",
       " 'E2F8',\n",
       " 'BCL2L1',\n",
       " 'FGF13',\n",
       " 'PITX1',\n",
       " 'FGF14',\n",
       " 'CDC37',\n",
       " 'LIF',\n",
       " 'FGF9',\n",
       " 'DAPK1',\n",
       " 'OLIG2',\n",
       " 'NFATC2',\n",
       " 'LIN28B',\n",
       " 'ELOVL2',\n",
       " 'PDX1',\n",
       " 'STUB1',\n",
       " 'POSTN',\n",
       " 'PRDM1',\n",
       " 'DNTT',\n",
       " 'SOX3',\n",
       " 'WT1',\n",
       " 'NRIP1',\n",
       " 'E2F1',\n",
       " 'SOD1',\n",
       " 'SIX1',\n",
       " 'IGFBP6',\n",
       " 'SOX1',\n",
       " 'MYBL2',\n",
       " 'APBA1',\n",
       " 'SMCHD1',\n",
       " 'ISL1',\n",
       " 'FGF10',\n",
       " 'NR2F1',\n",
       " 'SMAD4',\n",
       " 'TEAD4',\n",
       " 'RB1',\n",
       " 'MAPK11',\n",
       " 'UNC45A',\n",
       " 'CMYA5',\n",
       " 'ZDHHC8',\n",
       " 'TEF',\n",
       " 'TRAF6',\n",
       " 'XRCC6',\n",
       " 'THBS4',\n",
       " 'SNW1',\n",
       " 'UNCX',\n",
       " 'MYLIP',\n",
       " 'CDH17',\n",
       " 'TWSG1']"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "firstOrderGenes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "cb38694d-c5ce-4a81-80f3-a394eed3dbca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Stable ID</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Transcription Factor</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6001</th>\n",
       "      <td>ENSG00000129152</td>\n",
       "      <td>MUC20-OT1</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>17719571</td>\n",
       "      <td>17722136</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Stable ID  Gene Name  Transcription Factor  Chromosome     Start  \\\n",
       "6001  ENSG00000129152  MUC20-OT1                     1          11  17719571   \n",
       "\n",
       "           End  \n",
       "6001  17722136  "
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['Stable ID'] == 'ENSG00000129152']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "c8c87527-12b4-4771-a374-89361f438107",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ELP3',\n",
       " 'MRPL16',\n",
       " 'MKKS',\n",
       " 'MIPEP',\n",
       " 'EIF2AK4',\n",
       " 'E2F5',\n",
       " 'RTCB',\n",
       " 'PPCDC',\n",
       " 'SNRPB2',\n",
       " 'CCNB3',\n",
       " 'PSMB5',\n",
       " 'RPL3',\n",
       " 'ARHGAP4',\n",
       " 'POLI',\n",
       " 'GADD45G',\n",
       " 'RPS10-NUDT3',\n",
       " 'CDC45',\n",
       " 'SLC2A13',\n",
       " 'BRIX1',\n",
       " 'COL19A1',\n",
       " 'SMARCA2',\n",
       " 'DSTN',\n",
       " 'SEC11A',\n",
       " 'GNPNAT1',\n",
       " 'ASMTL',\n",
       " 'SNRPD3',\n",
       " 'SRSF3',\n",
       " 'NUS1',\n",
       " 'DPM1',\n",
       " 'DDX39A',\n",
       " 'SMC1B',\n",
       " 'AARS2',\n",
       " 'NUP155',\n",
       " 'CCNO',\n",
       " 'CRLS1',\n",
       " 'POLD3',\n",
       " 'EIF3M',\n",
       " 'UBE2D1',\n",
       " 'ANP32A',\n",
       " 'GART',\n",
       " 'PMS2',\n",
       " 'FAM83D',\n",
       " 'TPX2',\n",
       " 'HMGA1',\n",
       " 'DHX35',\n",
       " 'ATRX',\n",
       " 'RBL1',\n",
       " 'EIF2S2',\n",
       " 'N6AMT1',\n",
       " 'TXNDC5',\n",
       " 'NUBPL',\n",
       " 'PRIM2',\n",
       " 'RPS7',\n",
       " 'DHX32',\n",
       " 'TRUB1',\n",
       " 'PRMT8',\n",
       " 'SLC27A5',\n",
       " 'PIAS2',\n",
       " 'MKI67',\n",
       " 'SMC6',\n",
       " 'GRPEL1',\n",
       " 'ZNF367',\n",
       " 'CENPQ',\n",
       " 'MIS12',\n",
       " 'KIN',\n",
       " 'TRDMT1',\n",
       " 'METTL15',\n",
       " 'TFDP1',\n",
       " 'TYMS',\n",
       " 'CDC25B',\n",
       " 'WDHD1',\n",
       " 'PSMG2',\n",
       " 'UBL7',\n",
       " 'MRPS28',\n",
       " 'BORA',\n",
       " 'TRMU',\n",
       " 'REV3L',\n",
       " 'CDCA2',\n",
       " 'TXN',\n",
       " 'TIMELESS',\n",
       " 'DDX11',\n",
       " 'AFG3L2',\n",
       " 'PIMREG',\n",
       " 'LMNB1',\n",
       " 'RBMX',\n",
       " 'APEX2',\n",
       " 'PDSS2',\n",
       " 'HOGA1',\n",
       " 'MTG2',\n",
       " 'UBE2C',\n",
       " 'SETBP1',\n",
       " 'IAH1',\n",
       " 'DHX33',\n",
       " 'PDXK',\n",
       " 'CCT8',\n",
       " 'RNASEH2B',\n",
       " 'ZFAND4',\n",
       " 'AAR2',\n",
       " 'OSGEP',\n",
       " 'MCM5',\n",
       " 'PRDX3',\n",
       " 'ARMC1',\n",
       " 'ITM2B',\n",
       " 'PSMC6',\n",
       " 'MTRF1',\n",
       " 'ZWINT',\n",
       " 'TGDS',\n",
       " 'PI4KA',\n",
       " 'MPI',\n",
       " 'UBQLN2',\n",
       " 'APEX1',\n",
       " 'CUL4A',\n",
       " 'PNMA6A',\n",
       " 'CDCA4',\n",
       " 'ATP5F1C',\n",
       " 'SNRPD1',\n",
       " 'ANP32B',\n",
       " 'UHRF1',\n",
       " 'PBK',\n",
       " 'DIAPH3',\n",
       " 'CHD6',\n",
       " 'COL4A5',\n",
       " 'CMTM5',\n",
       " 'BABAM1',\n",
       " 'BCAT1',\n",
       " 'RACGAP1',\n",
       " 'CCNB2',\n",
       " 'EP300',\n",
       " 'COL21A1',\n",
       " 'IPO5',\n",
       " 'CCAR1',\n",
       " 'NUDT12',\n",
       " 'H3-5',\n",
       " 'POGLUT2',\n",
       " 'CKAP2',\n",
       " 'DEPDC1B',\n",
       " 'TESMIN',\n",
       " 'UBE2J1',\n",
       " 'MTFR2',\n",
       " 'ESCO1',\n",
       " 'UROS',\n",
       " 'RPUSD2',\n",
       " 'SUPT16H',\n",
       " 'PRPF19',\n",
       " 'YWHAQ',\n",
       " 'PDS5A',\n",
       " 'KPNA3',\n",
       " 'YWHAB',\n",
       " 'MRPS6',\n",
       " 'RPL17-C18orf32',\n",
       " 'LRR1',\n",
       " 'NANOG',\n",
       " 'MCIDAS',\n",
       " 'HDAC2',\n",
       " 'AIRE',\n",
       " 'MCM3',\n",
       " 'AGAP9',\n",
       " 'L3MBTL1',\n",
       " 'COL4A1',\n",
       " 'WDR4',\n",
       " 'HTRA1',\n",
       " 'SETDB2',\n",
       " 'NONO',\n",
       " 'ITM2A',\n",
       " 'SPAG6',\n",
       " 'DEK',\n",
       " 'CUTC',\n",
       " 'TSPYL5',\n",
       " 'RANBP1',\n",
       " 'RRAS2',\n",
       " 'MAPRE1',\n",
       " 'MGMT',\n",
       " 'CSNK1E',\n",
       " 'PGAM1',\n",
       " 'CCDC33',\n",
       " 'RPS4X',\n",
       " 'CCNK',\n",
       " 'PRPS2',\n",
       " 'PFKL',\n",
       " 'PRMT3',\n",
       " 'ERCC6L',\n",
       " 'PSMG1',\n",
       " 'HMGB3',\n",
       " 'TSPEAR',\n",
       " 'SMC1A',\n",
       " 'COL4A2',\n",
       " 'CNOT7',\n",
       " 'COL17A1',\n",
       " 'POLL',\n",
       " 'COL6A2',\n",
       " 'RAD50',\n",
       " 'EEF1G',\n",
       " 'MRPL13',\n",
       " 'DSCC1',\n",
       " 'INTS13',\n",
       " 'CSE1L',\n",
       " 'PSAT1',\n",
       " 'SKA1',\n",
       " 'TXNRD2',\n",
       " 'RPAP3',\n",
       " 'MIS18A',\n",
       " 'SUGT1',\n",
       " 'MT-CYB',\n",
       " 'NNT',\n",
       " 'CENPM',\n",
       " 'ASB8',\n",
       " 'SIVA1',\n",
       " 'UBQLN1',\n",
       " 'AHCY',\n",
       " 'CDC123',\n",
       " 'CDKAL1',\n",
       " 'CCNH',\n",
       " 'PRTFDC1',\n",
       " 'SUB1',\n",
       " 'LDHC',\n",
       " 'VIRMA',\n",
       " 'CPSF3',\n",
       " 'CA2',\n",
       " 'CSNK1A1L',\n",
       " 'RTF2',\n",
       " 'UBR7',\n",
       " 'MPHOSPH8',\n",
       " 'SUMO4',\n",
       " 'HSCB',\n",
       " 'AURKA',\n",
       " 'TFAM',\n",
       " 'APMAP',\n",
       " nan,\n",
       " 'MOCS3',\n",
       " 'MOCS1',\n",
       " 'RPL36A',\n",
       " 'DNA2',\n",
       " 'PSMD10',\n",
       " 'PFDN4',\n",
       " 'RIOK2',\n",
       " 'PAPOLG',\n",
       " 'SCML2',\n",
       " 'ATAD2B',\n",
       " 'FOXO3',\n",
       " 'NUDT15',\n",
       " 'RPA3',\n",
       " 'HIRA',\n",
       " 'MSH3',\n",
       " 'ENO4',\n",
       " 'CCNA1',\n",
       " 'SHMT2',\n",
       " 'ADAT2',\n",
       " 'RBBP7',\n",
       " 'CDADC1',\n",
       " 'RNF8',\n",
       " 'BCCIP',\n",
       " 'MCM8',\n",
       " 'POLA1',\n",
       " 'NDC80',\n",
       " 'HACD3',\n",
       " 'DSN1',\n",
       " 'POLE2',\n",
       " 'EXOSC8',\n",
       " 'UBL4A',\n",
       " 'CAPRIN1',\n",
       " 'CDKN2B',\n",
       " 'WRN',\n",
       " 'TOP1',\n",
       " 'UBR5',\n",
       " 'DHTKD1',\n",
       " 'HSPH1',\n",
       " 'SNRPC',\n",
       " 'MAT1A',\n",
       " 'NOP56',\n",
       " 'PRMT5',\n",
       " 'PSMA7',\n",
       " 'SUV39H2',\n",
       " 'NUDT5',\n",
       " 'MBD1',\n",
       " 'MAGOHB',\n",
       " 'CCNY',\n",
       " 'PACSIN3',\n",
       " 'DONSON',\n",
       " 'RFC3',\n",
       " 'ACP2',\n",
       " 'RPL36AL',\n",
       " 'MAP1LC3A',\n",
       " 'ANG',\n",
       " 'RIDA',\n",
       " 'ANK3',\n",
       " 'CTPS2',\n",
       " 'C1QBP',\n",
       " 'PHF11',\n",
       " 'PSMA4',\n",
       " 'INCENP',\n",
       " 'RPP38',\n",
       " 'HSPA12B',\n",
       " 'CENPA',\n",
       " 'PRPS1L1',\n",
       " 'COX5A',\n",
       " 'POLR2F',\n",
       " 'RPL36A-HNRNPH2',\n",
       " 'EIF2AK1',\n",
       " 'USP18',\n",
       " 'CSTF1',\n",
       " 'RAD51',\n",
       " 'RBX1',\n",
       " 'XPNPEP1',\n",
       " 'COL9A3',\n",
       " 'SLF1',\n",
       " 'CSNK2A1',\n",
       " 'MAGED1',\n",
       " 'TTK',\n",
       " 'VBP1',\n",
       " 'MLH3',\n",
       " 'APC',\n",
       " 'H2AB3',\n",
       " 'COL6A1',\n",
       " 'TEX30',\n",
       " 'RAD1',\n",
       " 'RPA4',\n",
       " 'RAD21L1',\n",
       " 'ESCO2',\n",
       " 'ERCC5',\n",
       " 'GTF2A2',\n",
       " 'HAUS6',\n",
       " 'MX1',\n",
       " 'OXA1L',\n",
       " 'HSPA13',\n",
       " 'GTPBP3',\n",
       " 'MCAT',\n",
       " 'DHX15',\n",
       " 'HDDC2',\n",
       " 'ING1',\n",
       " 'POLR2E',\n",
       " 'RANBP6',\n",
       " 'PFKM',\n",
       " 'RCN2',\n",
       " 'GEN1',\n",
       " 'CCNT1',\n",
       " 'AK7',\n",
       " 'RAD54B',\n",
       " 'PCNA',\n",
       " 'ASPG',\n",
       " 'UHRF2',\n",
       " 'BCL2L1',\n",
       " 'E2F8',\n",
       " 'BUB3',\n",
       " 'LIG4',\n",
       " 'RANGAP1',\n",
       " 'CKS2',\n",
       " 'PAPOLB',\n",
       " 'FTSJ1',\n",
       " 'MTERF3',\n",
       " 'SKA3',\n",
       " 'SEC61A2',\n",
       " 'RBBP8',\n",
       " 'CENPK',\n",
       " 'MRPL15',\n",
       " 'SUMO3',\n",
       " 'CCNQ',\n",
       " 'SEPHS1',\n",
       " 'RAD23B',\n",
       " 'RAB9B',\n",
       " 'RAB33A',\n",
       " 'PRPS1',\n",
       " 'SLC27A2',\n",
       " 'DNTT',\n",
       " 'OAT',\n",
       " 'MGME1',\n",
       " 'E2F1',\n",
       " 'SLC6A8',\n",
       " 'EEF1E1',\n",
       " 'CCT5',\n",
       " 'AP5Z1',\n",
       " 'SPO11',\n",
       " 'HMGN1',\n",
       " 'RBPJL',\n",
       " 'MSH2',\n",
       " 'MYBL2',\n",
       " 'CDC5L',\n",
       " 'MNAT1',\n",
       " 'GPD1',\n",
       " 'BRCA2',\n",
       " 'MX2',\n",
       " 'POLN',\n",
       " 'BMS1',\n",
       " 'RB1',\n",
       " 'TSPYL4',\n",
       " 'ADSL',\n",
       " 'THNSL1',\n",
       " 'TFDP3',\n",
       " 'RNPS1',\n",
       " 'SNX5',\n",
       " 'NUP50',\n",
       " 'PHF21B',\n",
       " 'HMGB1',\n",
       " 'MAFB',\n",
       " 'CHEK2',\n",
       " 'TRMT11',\n",
       " 'PDSS1',\n",
       " 'XRCC6',\n",
       " 'AK3',\n",
       " 'WARS1',\n",
       " 'PDS5B',\n",
       " 'SNU13',\n",
       " 'XRCC3',\n",
       " 'KPNA5',\n",
       " 'UFD1',\n",
       " 'CHAF1B',\n",
       " 'CARS2',\n",
       " 'LDHAL6B',\n",
       " 'AIMP2',\n",
       " 'RPS12',\n",
       " 'PIK3C3',\n",
       " 'PGK2',\n",
       " 'MRPL39',\n",
       " 'MT-ATP6',\n",
       " 'PRMT2']"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "firstOrderGenes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "0b95bf43-8c07-4a96-bbcf-608695b5f85c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "414\n"
     ]
    }
   ],
   "source": [
    "firstOrderGenes = list(set(df2[df2['Gene stable ID'].isin(firstOrderEnsg)]['Gene name'].values))\n",
    "print(len(firstOrderGenes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "6916cd11-867e-4b02-abe7-f35559842700",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6507"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "idx = df[df['Stable ID'] == ENSG1].index[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "2ef13b8e-887a-4183-b12f-d11d0b648fb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "idx = df[df['Stable ID'] == ENSG1].index[0]\n",
    "idxs = np.where(A[idx,:] == 1)[1]\n",
    "firstOrderEnsg = list(df.iloc[idxs]['Stable ID'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "82a94368-adcd-4f33-8eb3-6be12188815a",
   "metadata": {},
   "outputs": [],
   "source": [
    "idxs = np.where(A[6507,:] == 1)[1]\n",
    "firstOrderEnsg = list(df.iloc[idxs]['Stable ID'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "d83152eb-7493-473c-b5da-b944fa46dc05",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on _ArrayFunctionDispatcher in module numpy:\n",
      "\n",
      "where(...)\n",
      "    where(condition, [x, y], /)\n",
      "    \n",
      "    Return elements chosen from `x` or `y` depending on `condition`.\n",
      "    \n",
      "    .. note::\n",
      "        When only `condition` is provided, this function is a shorthand for\n",
      "        ``np.asarray(condition).nonzero()``. Using `nonzero` directly should be\n",
      "        preferred, as it behaves correctly for subclasses. The rest of this\n",
      "        documentation covers only the case where all three arguments are\n",
      "        provided.\n",
      "    \n",
      "    Parameters\n",
      "    ----------\n",
      "    condition : array_like, bool\n",
      "        Where True, yield `x`, otherwise yield `y`.\n",
      "    x, y : array_like\n",
      "        Values from which to choose. `x`, `y` and `condition` need to be\n",
      "        broadcastable to some shape.\n",
      "    \n",
      "    Returns\n",
      "    -------\n",
      "    out : ndarray\n",
      "        An array with elements from `x` where `condition` is True, and elements\n",
      "        from `y` elsewhere.\n",
      "    \n",
      "    See Also\n",
      "    --------\n",
      "    choose\n",
      "    nonzero : The function that is called when x and y are omitted\n",
      "    \n",
      "    Notes\n",
      "    -----\n",
      "    If all the arrays are 1-D, `where` is equivalent to::\n",
      "    \n",
      "        [xv if c else yv\n",
      "         for c, xv, yv in zip(condition, x, y)]\n",
      "    \n",
      "    Examples\n",
      "    --------\n",
      "    >>> a = np.arange(10)\n",
      "    >>> a\n",
      "    array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])\n",
      "    >>> np.where(a < 5, a, 10*a)\n",
      "    array([ 0,  1,  2,  3,  4, 50, 60, 70, 80, 90])\n",
      "    \n",
      "    This can be used on multidimensional arrays too:\n",
      "    \n",
      "    >>> np.where([[True, False], [True, True]],\n",
      "    ...          [[1, 2], [3, 4]],\n",
      "    ...          [[9, 8], [7, 6]])\n",
      "    array([[1, 8],\n",
      "           [3, 4]])\n",
      "    \n",
      "    The shapes of x, y, and the condition are broadcast together:\n",
      "    \n",
      "    >>> x, y = np.ogrid[:3, :4]\n",
      "    >>> np.where(x < y, x, 10 + y)  # both x and 10+y are broadcast\n",
      "    array([[10,  0,  0,  0],\n",
      "           [10, 11,  1,  1],\n",
      "           [10, 11, 12,  2]])\n",
      "    \n",
      "    >>> a = np.array([[0, 1, 2],\n",
      "    ...               [0, 2, 4],\n",
      "    ...               [0, 3, 6]])\n",
      "    >>> np.where(a < 4, a, -1)  # -1 is broadcast\n",
      "    array([[ 0,  1,  2],\n",
      "           [ 0,  2, -1],\n",
      "           [ 0,  3, -1]])\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(np.where)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "48f4d525-5b3a-46a4-92d0-1d3ea698c13a",
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "tuple indices must be integers or slices, not tuple",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[43], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m A \u001b[38;5;241m=\u001b[39m \u001b[43mA\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m,\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241m.\u001b[39mtodense()\n",
      "\u001b[0;31mTypeError\u001b[0m: tuple indices must be integers or slices, not tuple"
     ]
    }
   ],
   "source": [
    "A = A[0,0].todense()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "c4776e82-3460-49d7-a4f9-77f21546bcbc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Protein stable ID</th>\n",
       "      <th>Gene start (bp)</th>\n",
       "      <th>Gene end (bp)</th>\n",
       "      <th>Chromosome/scaffold name</th>\n",
       "      <th>Gene name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>50403</th>\n",
       "      <td>ENSG00000132646</td>\n",
       "      <td>ENSP00000368438</td>\n",
       "      <td>5114953</td>\n",
       "      <td>5126626</td>\n",
       "      <td>20</td>\n",
       "      <td>PCNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50404</th>\n",
       "      <td>ENSG00000132646</td>\n",
       "      <td>ENSP00000368458</td>\n",
       "      <td>5114953</td>\n",
       "      <td>5126626</td>\n",
       "      <td>20</td>\n",
       "      <td>PCNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Gene stable ID Protein stable ID  Gene start (bp)  Gene end (bp)  \\\n",
       "50403  ENSG00000132646   ENSP00000368438          5114953        5126626   \n",
       "50404  ENSG00000132646   ENSP00000368458          5114953        5126626   \n",
       "\n",
       "      Chromosome/scaffold name Gene name  \n",
       "50403                       20      PCNA  \n",
       "50404                       20      PCNA  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2 = pd.read_csv('GRCh38.p13.tsv', delimiter='\\t')\n",
    "df2[df2['Gene name'] == geneName]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "330bf723-801d-4f70-9209-dff4c813ace2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Stable ID</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Transcription Factor</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Stable ID, Gene Name, Transcription Factor, Chromosome, Start, End]\n",
       "Index: []"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['Gene Name'] == 'PCNA']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ce876d4c-d90e-46ee-addc-477b5bb82619",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Stable ID</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Transcription Factor</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>MT-RNR1</td>\n",
       "      <td>0</td>\n",
       "      <td>23</td>\n",
       "      <td>100627108</td>\n",
       "      <td>100639991</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>MT-TV</td>\n",
       "      <td>0</td>\n",
       "      <td>23</td>\n",
       "      <td>100584936</td>\n",
       "      <td>100599885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>MT-RNR2</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>50934867</td>\n",
       "      <td>50958555</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>MT-TL1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>169849631</td>\n",
       "      <td>169894267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>MT-ND1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>169662007</td>\n",
       "      <td>169854080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19211</th>\n",
       "      <td>ENSG00000287363</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>42155406</td>\n",
       "      <td>42180056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19212</th>\n",
       "      <td>ENSG00000288602</td>\n",
       "      <td>LINC02311</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19213</th>\n",
       "      <td>ENSG00000288674</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19214</th>\n",
       "      <td>ENSG00000288722</td>\n",
       "      <td>EMX2OS</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19215</th>\n",
       "      <td>ENSG00000288859</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19216 rows  6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Stable ID  Gene Name  Transcription Factor  Chromosome  \\\n",
       "0      ENSG00000000003    MT-RNR1                     0          23   \n",
       "1      ENSG00000000005      MT-TV                     0          23   \n",
       "2      ENSG00000000419    MT-RNR2                     0          20   \n",
       "3      ENSG00000000457     MT-TL1                     0           1   \n",
       "4      ENSG00000000460     MT-ND1                     0           1   \n",
       "...                ...        ...                   ...         ...   \n",
       "19211  ENSG00000287363        NaN                     0           6   \n",
       "19212  ENSG00000288602  LINC02311                     0           0   \n",
       "19213  ENSG00000288674        NaN                     0           0   \n",
       "19214  ENSG00000288722     EMX2OS                     0           0   \n",
       "19215  ENSG00000288859        NaN                     0           0   \n",
       "\n",
       "           Start        End  \n",
       "0      100627108  100639991  \n",
       "1      100584936  100599885  \n",
       "2       50934867   50958555  \n",
       "3      169849631  169894267  \n",
       "4      169662007  169854080  \n",
       "...          ...        ...  \n",
       "19211   42155406   42180056  \n",
       "19212          0          0  \n",
       "19213          0          0  \n",
       "19214          0          0  \n",
       "19215          0          0  \n",
       "\n",
       "[19216 rows x 6 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "10d22dab-3f8c-41fa-9629-892ae3b425ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s0</th>\n",
       "      <th>s1</th>\n",
       "      <th>s2</th>\n",
       "      <th>arr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b''</td>\n",
       "      <td>b'MCOS'</td>\n",
       "      <td>b'table'</td>\n",
       "      <td>[[3707764736], [2], [1], [1], [1], [2]]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    s0       s1        s2                                      arr\n",
       "0  b''  b'MCOS'  b'table'  [[3707764736], [2], [1], [1], [1], [2]]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "pd.DataFrame(HWG['HWG']['geneIndexTable'][0][0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "727bafa2-0bfb-4949-a8c7-e61d0b143876",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[(array([[0]], dtype=uint8), <19216x19216 sparse matrix of type '<class 'numpy.uint8'>'\n",
       "        \twith 11750266 stored elements in Compressed Sparse Column format>, <19216x1621 sparse matrix of type '<class 'numpy.uint8'>'\n",
       "        \twith 881627 stored elements in Compressed Sparse Column format>, <1621x1621 sparse matrix of type '<class 'numpy.uint8'>'\n",
       "        \twith 234808 stored elements in Compressed Sparse Column format>, MatlabOpaque([(b'', b'MCOS', b'table', array([[3707764736],\n",
       "                             [         2],\n",
       "                             [         1],\n",
       "                             [         1],\n",
       "                             [         1],\n",
       "                             [         2]], dtype=uint32))         ],\n",
       "                     dtype=[('s0', 'O'), ('s1', 'O'), ('s2', 'O'), ('arr', 'O')]), MatlabOpaque([(b'', b'MCOS', b'table', array([[3707764736],\n",
       "                             [         2],\n",
       "                             [         1],\n",
       "                             [         1],\n",
       "                             [     18979],\n",
       "                             [         2]], dtype=uint32))         ],\n",
       "                     dtype=[('s0', 'O'), ('s1', 'O'), ('s2', 'O'), ('arr', 'O')]))                                                            ]],\n",
       "      dtype=[('thresh', 'O'), ('A', 'O'), ('B', 'O'), ('C', 'O'), ('geneIndexTable', 'O'), ('TFIndexTable', 'O')])"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HWG['HWG']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36f1cdd3-7a20-4ad2-a69d-706abd3ca1d5",
   "metadata": {},
   "source": [
    "# REPORTER"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "813dd728-03d0-47c4-8dcb-680d4e22da6d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1805a9d2-894c-42e8-a9bb-fd5045f6ab16",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab6d4882-81a4-422a-9ad6-6645c564f8a2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd9932b1-8156-4cda-ad0e-75d3402bafcb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "3415fc4d-7d87-4b96-95aa-f6b7c7feb46e",
   "metadata": {},
   "source": [
    "# DEBUG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "326c4767-bc1f-4d62-8f00-d439f0291e0c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Perform gene set enrichment analysis on the identified interactions from Step 2 using databases like Gene Ontology and Enrichr. This will provide information on biological pathways and cell types associated with these interactions.'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "XX = \"/force DATABASE Perform gene set enrichment analysis on the identified interactions from Step 2 using databases like Gene Ontology and Enrichr. This will provide information on biological pathways and cell types associated with these interactions.\"\n",
    "\" \".join(XX.split(' ')[2:]).strip()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7ee58c30-e7ac-41bf-bbf7-5448f23f38c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "from BRAD.matlabCaller import find_matlab_files, get_matlab_description\n",
    "from BRAD.pythonCaller import find_py_files, get_py_description"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "89d689d4-263c-48a9-90d7-dca0336f11ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['animal_info', 'example', 'helloWorldPYTHON']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "find_py_files('/home/jpic/RAG-DEV/tutorials/running-codes/python/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ea1a7695-8668-4cd0-b79e-19ade4f3072d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'This script prints hello world to the terminal. No arguments are required.'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_py_description('/home/jpic/RAG-DEV/tutorials/running-codes/python/helloWorldPYTHON.py')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b46e3f06-c310-4caa-b6c5-b890154fd85b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['matlabPipeline', 'myFunc', 'helloWorldMATLAB', 'saveplot', 'qualityControl']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "find_matlab_files('/home/jpic/RAG-DEV/tutorials/running-codes/matlab/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8d43c2f8-78df-4a60-a9b0-7e3a7d908a4a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "' qualityControl - Constructs and runs a bioinformatics quality control pipeline.'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_matlab_description('/home/jpic/RAG-DEV/tutorials/running-codes/matlab/qualityControl.m')"
   ]
  },
  {
   "cell_type": "raw",
   "id": "f1ab7fd5-53ae-466c-9c43-55e0dc3c08fa",
   "metadata": {},
   "source": [
    "/force CODE say hello world using python!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4dd54f73-6d6c-4add-9e82-96b9e7324026",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome to RAG! The chat log from this conversation will be saved to /home/jpic/BRAD/2024-06-16-16:50:05.318931.json. How can I help?\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter your NVIDIA API key:  \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Would you like to use a database with BRAD [Y/N]?\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      " N\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sun 16 Jun 2024 04:50:09 PM EDT INFO local\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  /force CODE say hello world using python!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sun 16 Jun 2024 04:50:14 PM EDT INFO CODE\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "RAG >> 1: CODER\n",
      "FIRST LLM CALL\n",
      "Warning! the type doesnt exist\n",
      "scriptPath=/home/jpic/RAG-DEV/tutorials/running-codes/python/\n",
      "pyPath=/home/jpic/RAG-DEV/tutorials/running-codes/python/\n",
      "ALL SCRIPTS FOUND. BUILDING TEMPLATE\n",
      "input_variables=['history', 'input'] template='Current conversation:\\n{history}\\n\\n**PYTHON SCRIPT**\\nYou must run this python script: \\n/home/jpic/RAG-DEV/tutorials/running-codes/python/helloWorldPYTHON.py\\n\\n**PYTHON SCRIPT DOCUMENTATION**:\\nThis is the doc string of this python script:\\n\"\"\"\\nThis script prints hello world to the terminal. No arguments are required.\\n\\nExample usage:\\n    subprocess.call([sys.executable, \\'python/helloWorldPYTHON.py\\'])\\n\"\"\"\\n\\n\\n**CALL PYTHON SCRIPTS FROM PYTHON**:\\nUse the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\\n\\nTo call a Python script `example_script.py` which has no arguments:\\n    \\n```\\nExecute: subprocess.call([sys.executable, \\'<full path to script/> example_script.py\\'])\\n```\\n\\nTo call a Python script `example_script.py` with one argument:\\n\\n```\\nExecute: subprocess.call([sys.executable, \\'<full path to script/> example_script.py\\', \\'arg1\\'])\\n```\\n\\nTo call a Python script `example_script.py` with two arguments:\\n```\\nExecute: subprocess.call([sys.executable, \\'<full path to script/> example_script.py\\', \\'arg1\\', \\'arg2\\'])\\n```\\nQuery:{input}\\n\\n\\n**INSTRUCTIONS**\\n1. Given the user query and the documentation, identify each of the arguments found in the user\\'s query that should be passed to the Python script.\\n2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\\n3. The last line of your response should say \"Execute: <Python code to execute>\"\\n4. Format the response/output as:\\n    Arguments: \\n    Python Code Explanation: <2 sentences maximum>\\n    Execute: <your code here>\\n\\n**IMPORTANT**\\nThe code to execute from your response must be formatted as:\\n    Execute: subprocess.call([sys.executable, \\'<path to python script>\\', \\'<argument 1>\\', \\'<argument 2>\\', ..., \\'<argument n>\\'])\\nThis output should be exactly one line and no longer. Stop the response after this line.\\n'\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation:\n",
      "\n",
      "\n",
      "**PYTHON SCRIPT**\n",
      "You must run this python script: \n",
      "/home/jpic/RAG-DEV/tutorials/running-codes/python/helloWorldPYTHON.py\n",
      "\n",
      "**PYTHON SCRIPT DOCUMENTATION**:\n",
      "This is the doc string of this python script:\n",
      "\"\"\"\n",
      "This script prints hello world to the terminal. No arguments are required.\n",
      "\n",
      "Example usage:\n",
      "    subprocess.call([sys.executable, 'python/helloWorldPYTHON.py'])\n",
      "\"\"\"\n",
      "\n",
      "\n",
      "**CALL PYTHON SCRIPTS FROM PYTHON**:\n",
      "Use the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\n",
      "\n",
      "To call a Python script `example_script.py` which has no arguments:\n",
      "    \n",
      "```\n",
      "Execute: subprocess.call([sys.executable, '<full path to script/> example_script.py'])\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with one argument:\n",
      "\n",
      "```\n",
      "Execute: subprocess.call([sys.executable, '<full path to script/> example_script.py', 'arg1'])\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with two arguments:\n",
      "```\n",
      "Execute: subprocess.call([sys.executable, '<full path to script/> example_script.py', 'arg1', 'arg2'])\n",
      "```\n",
      "Query:/force CODE say hello world using python!\n",
      "\n",
      "\n",
      "**INSTRUCTIONS**\n",
      "1. Given the user query and the documentation, identify each of the arguments found in the user's query that should be passed to the Python script.\n",
      "2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\n",
      "3. The last line of your response should say \"Execute: <Python code to execute>\"\n",
      "4. Format the response/output as:\n",
      "    Arguments: \n",
      "    Python Code Explanation: <2 sentences maximum>\n",
      "    Execute: <your code here>\n",
      "\n",
      "**IMPORTANT**\n",
      "The code to execute from your response must be formatted as:\n",
      "    Execute: subprocess.call([sys.executable, '<path to python script>', '<argument 1>', '<argument 2>', ..., '<argument n>'])\n",
      "This output should be exactly one line and no longer. Stop the response after this line.\n",
      "\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "LLM OUTPUT PARSER\n",
      " Arguments: The user query does not provide any arguments to be passed to the Python script. The script documentation indicates no arguments are required.\n",
      "\n",
      "Python Code Explanation: We will execute the \"helloWorldPYTHON.py\" script with no arguments.\n",
      "\n",
      "Execute: subprocess.call([sys.executable, '/home/jpic/RAG-DEV/tutorials/running-codes/python/helloWorldPYTHON.py'])\n",
      "[' Arguments: The user query does not provide any arguments to be passed to the Python script. The script documentation indicates no arguments are required.\\n\\nPython Code Explanation: We will execute the \"helloWorldPYTHON.py\" script with no arguments.\\n\\n', \" subprocess.call([sys.executable, '/home/jpic/RAG-DEV/tutorials/running-codes/python/helloWorldPYTHON.py'])\"]\n",
      "Executing Code!!\n",
      "EVAL\n",
      "subprocess.call([sys.executable, '/home/jpic/RAG-DEV/tutorials/running-codes/python/helloWorldPYTHON.py'])\n",
      "END EVAL CHECK\n",
      "Hello World from a Python Script!\n",
      "Debug: PYTHON code executed successfully.\n",
      "==================================================\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  /force CODE say hello world using matlab\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sun 16 Jun 2024 04:50:26 PM EDT INFO CODE\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "RAG >> 2: CODER\n",
      "FIRST LLM CALL\n",
      "Warning! the type doesnt exist\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...\n",
      "To disable this warning, you can either:\n",
      "\t- Avoid using `tokenizers` before the fork if possible\n",
      "\t- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Matlab Engine On\n",
      "ALL SCRIPTS FOUND. BUILDING TEMPLATE\n",
      "input_variables=['history', 'input'] template='Current conversation:\\n{history}\\n\\n**MATLAB SCRIPT**\\nYou must run this python script: \\nhelloWorldMATLAB.m\\n\\n**MATLAB FUNCTION DOCUMENTATION**:\\nfunction output = helloWorldMATLAB()\\n% Print \"Hello World\"\\n%\\n% Author: Joshua Pickard\\n%         jpic@umich.edu\\n% Date: June 14, 2024\\n\\n**CALL MATLAB FUNCTIONS FROM PYTHON**:\\nUse MATLAB Engine API for Python to call any MATLAB function on the MATLAB path. Some examples to call a few common matlab functions include:\\n\\nTo call the matlab function myFnc which has no arguments:\\n```\\nExecute: eng.myFnc()\\n```\\n\\nTo use the is prime function which requires an input argument:\\n```\\nExecute: tf = eng.isprime(37)\\n```\\n\\nTo determine the greatest common denominator of two numbers, use the gcd function. Set nargout to return the three output arguments from gcd.\\n```\\nExecute: t = eng.gcd(100.0,80.0,nargout=3)\\n```\\n\\nQuery:{input}\\n\\n**INSTRUCTIONS**\\n1. Given the user query and the documentation, identify each of the arguments found in the users query that should be passed to the matlab function.\\n2. Using the matlab enging eng, provide the one line of code to execute the desired matlab commands. Assume all functions are added to the path and eng already exist.\\n3. The last line of your response should say \"Execute: <MATLAB codes to execute>\"\\n4. Format the response/output as:\\nArguments: \\nMATLAB Code Explanation: <2 sentences maximum>\\nExecute: <your code here>\\n\\n**IMPORTANT**\\nThe code to execute from your response must be formatted as:\\n    Execute: eng.<function name>(<arguments>)>\\nThis output should be exactly one line and no longer. Stop the response after this line.\\n'\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation:\n",
      "Human: /force CODE say hello world using python!\n",
      "BRAD:  Arguments: The user query does not provide any arguments to be passed to the Python script. The script documentation indicates no arguments are required.\n",
      "\n",
      "Python Code Explanation: We will execute the \"helloWorldPYTHON.py\" script with no arguments.\n",
      "\n",
      "Execute: subprocess.call([sys.executable, '/home/jpic/RAG-DEV/tutorials/running-codes/python/helloWorldPYTHON.py'])\n",
      "\n",
      "**MATLAB SCRIPT**\n",
      "You must run this python script: \n",
      "helloWorldMATLAB.m\n",
      "\n",
      "**MATLAB FUNCTION DOCUMENTATION**:\n",
      "function output = helloWorldMATLAB()\n",
      "% Print \"Hello World\"\n",
      "%\n",
      "% Author: Joshua Pickard\n",
      "%         jpic@umich.edu\n",
      "% Date: June 14, 2024\n",
      "\n",
      "**CALL MATLAB FUNCTIONS FROM PYTHON**:\n",
      "Use MATLAB Engine API for Python to call any MATLAB function on the MATLAB path. Some examples to call a few common matlab functions include:\n",
      "\n",
      "To call the matlab function myFnc which has no arguments:\n",
      "```\n",
      "Execute: eng.myFnc()\n",
      "```\n",
      "\n",
      "To use the is prime function which requires an input argument:\n",
      "```\n",
      "Execute: tf = eng.isprime(37)\n",
      "```\n",
      "\n",
      "To determine the greatest common denominator of two numbers, use the gcd function. Set nargout to return the three output arguments from gcd.\n",
      "```\n",
      "Execute: t = eng.gcd(100.0,80.0,nargout=3)\n",
      "```\n",
      "\n",
      "Query:/force CODE say hello world using matlab\n",
      "\n",
      "**INSTRUCTIONS**\n",
      "1. Given the user query and the documentation, identify each of the arguments found in the users query that should be passed to the matlab function.\n",
      "2. Using the matlab enging eng, provide the one line of code to execute the desired matlab commands. Assume all functions are added to the path and eng already exist.\n",
      "3. The last line of your response should say \"Execute: <MATLAB codes to execute>\"\n",
      "4. Format the response/output as:\n",
      "Arguments: \n",
      "MATLAB Code Explanation: <2 sentences maximum>\n",
      "Execute: <your code here>\n",
      "\n",
      "**IMPORTANT**\n",
      "The code to execute from your response must be formatted as:\n",
      "    Execute: eng.<function name>(<arguments>)>\n",
      "This output should be exactly one line and no longer. Stop the response after this line.\n",
      "\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "LLM OUTPUT PARSER\n",
      " Arguments: No arguments are required for the \"helloWorldMATLAB\" function.\n",
      "\n",
      "MATLAB Code Explanation: We will execute the \"helloWorldMATLAB.m\" function with no arguments.\n",
      "\n",
      "Execute: eng.helloWorldMATLAB()\n",
      "[' Arguments: No arguments are required for the \"helloWorldMATLAB\" function.\\n\\nMATLAB Code Explanation: We will execute the \"helloWorldMATLAB.m\" function with no arguments.\\n\\n', ' eng.helloWorldMATLAB()']\n",
      "Executing Code!!\n",
      "EVAL\n",
      "eng.helloWorldMATLAB()\n",
      "END EVAL CHECK\n",
      "Hello World! This message is generated from BRAD running the helloWorldMATLAB function!\n",
      "Debug: MATLAB code executed successfully.\n",
      "==================================================\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  /force CODE say hello world using matlab\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sun 16 Jun 2024 04:50:46 PM EDT INFO CODE\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "RAG >> 3: CODER\n",
      "FIRST LLM CALL\n",
      "Warning! the type doesnt exist\n",
      "Matlab Engine On\n",
      "ALL SCRIPTS FOUND. BUILDING TEMPLATE\n",
      "input_variables=['history', 'input'] template='Current conversation:\\n{history}\\n\\n**MATLAB SCRIPT**\\nYou must run this python script: \\nhelloWorldMATLAB.m\\n\\n**MATLAB FUNCTION DOCUMENTATION**:\\nfunction output = helloWorldMATLAB()\\n% Print \"Hello World\"\\n%\\n% Author: Joshua Pickard\\n%         jpic@umich.edu\\n% Date: June 14, 2024\\n\\n**CALL MATLAB FUNCTIONS FROM PYTHON**:\\nUse MATLAB Engine API for Python to call any MATLAB function on the MATLAB path. Some examples to call a few common matlab functions include:\\n\\nTo call the matlab function myFnc which has no arguments:\\n```\\nExecute: eng.myFnc()\\n```\\n\\nTo use the is prime function which requires an input argument:\\n```\\nExecute: tf = eng.isprime(37)\\n```\\n\\nTo determine the greatest common denominator of two numbers, use the gcd function. Set nargout to return the three output arguments from gcd.\\n```\\nExecute: t = eng.gcd(100.0,80.0,nargout=3)\\n```\\n\\nQuery:{input}\\n\\n**INSTRUCTIONS**\\n1. Given the user query and the documentation, identify each of the arguments found in the users query that should be passed to the matlab function.\\n2. Using the matlab enging eng, provide the one line of code to execute the desired matlab commands. Assume all functions are added to the path and eng already exist.\\n3. The last line of your response should say \"Execute: <MATLAB codes to execute>\"\\n4. Format the response/output as:\\nArguments: \\nMATLAB Code Explanation: <2 sentences maximum>\\nExecute: <your code here>\\n\\n**IMPORTANT**\\nThe code to execute from your response must be formatted as:\\n    Execute: eng.<function name>(<arguments>)>\\nThis output should be exactly one line and no longer. Stop the response after this line.\\n'\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation:\n",
      "Human: /force CODE say hello world using python!\n",
      "BRAD:  Arguments: The user query does not provide any arguments to be passed to the Python script. The script documentation indicates no arguments are required.\n",
      "\n",
      "Python Code Explanation: We will execute the \"helloWorldPYTHON.py\" script with no arguments.\n",
      "\n",
      "Execute: subprocess.call([sys.executable, '/home/jpic/RAG-DEV/tutorials/running-codes/python/helloWorldPYTHON.py'])\n",
      "Human: /force CODE say hello world using matlab\n",
      "BRAD:  Arguments: No arguments are required for the \"helloWorldMATLAB\" function.\n",
      "\n",
      "MATLAB Code Explanation: We will execute the \"helloWorldMATLAB.m\" function with no arguments.\n",
      "\n",
      "Execute: eng.helloWorldMATLAB()\n",
      "\n",
      "**MATLAB SCRIPT**\n",
      "You must run this python script: \n",
      "helloWorldMATLAB.m\n",
      "\n",
      "**MATLAB FUNCTION DOCUMENTATION**:\n",
      "function output = helloWorldMATLAB()\n",
      "% Print \"Hello World\"\n",
      "%\n",
      "% Author: Joshua Pickard\n",
      "%         jpic@umich.edu\n",
      "% Date: June 14, 2024\n",
      "\n",
      "**CALL MATLAB FUNCTIONS FROM PYTHON**:\n",
      "Use MATLAB Engine API for Python to call any MATLAB function on the MATLAB path. Some examples to call a few common matlab functions include:\n",
      "\n",
      "To call the matlab function myFnc which has no arguments:\n",
      "```\n",
      "Execute: eng.myFnc()\n",
      "```\n",
      "\n",
      "To use the is prime function which requires an input argument:\n",
      "```\n",
      "Execute: tf = eng.isprime(37)\n",
      "```\n",
      "\n",
      "To determine the greatest common denominator of two numbers, use the gcd function. Set nargout to return the three output arguments from gcd.\n",
      "```\n",
      "Execute: t = eng.gcd(100.0,80.0,nargout=3)\n",
      "```\n",
      "\n",
      "Query:/force CODE say hello world using matlab\n",
      "\n",
      "**INSTRUCTIONS**\n",
      "1. Given the user query and the documentation, identify each of the arguments found in the users query that should be passed to the matlab function.\n",
      "2. Using the matlab enging eng, provide the one line of code to execute the desired matlab commands. Assume all functions are added to the path and eng already exist.\n",
      "3. The last line of your response should say \"Execute: <MATLAB codes to execute>\"\n",
      "4. Format the response/output as:\n",
      "Arguments: \n",
      "MATLAB Code Explanation: <2 sentences maximum>\n",
      "Execute: <your code here>\n",
      "\n",
      "**IMPORTANT**\n",
      "The code to execute from your response must be formatted as:\n",
      "    Execute: eng.<function name>(<arguments>)>\n",
      "This output should be exactly one line and no longer. Stop the response after this line.\n",
      "\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "LLM OUTPUT PARSER\n",
      " Arguments: No arguments are required for the \"helloWorldMATLAB\" function.\n",
      "\n",
      "MATLAB Code Explanation: We will execute the \"helloWorldMATLAB\" function with no arguments.\n",
      "\n",
      "Execute: eng.helloWorldMATLAB()\n",
      "[' Arguments: No arguments are required for the \"helloWorldMATLAB\" function.\\n\\nMATLAB Code Explanation: We will execute the \"helloWorldMATLAB\" function with no arguments.\\n\\n', ' eng.helloWorldMATLAB()']\n",
      "Executing Code!!\n",
      "EVAL\n",
      "eng.helloWorldMATLAB()\n",
      "END EVAL CHECK\n",
      "Hello World! This message is generated from BRAD running the helloWorldMATLAB function!\n",
      "Debug: MATLAB code executed successfully.\n",
      "==================================================\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  /force CODE say hello world using python!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sun 16 Jun 2024 04:50:55 PM EDT INFO CODE\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "RAG >> 4: CODER\n",
      "FIRST LLM CALL\n",
      "Warning! the type doesnt exist\n",
      "scriptPath=/home/jpic/RAG-DEV/tutorials/running-codes/python/\n",
      "pyPath=/home/jpic/RAG-DEV/tutorials/running-codes/python/\n",
      "ALL SCRIPTS FOUND. BUILDING TEMPLATE\n",
      "input_variables=['history', 'input'] template='Current conversation:\\n{history}\\n\\n**PYTHON SCRIPT**\\nYou must run this python script: \\n/home/jpic/RAG-DEV/tutorials/running-codes/python/helloWorldPYTHON.py\\n\\n**PYTHON SCRIPT DOCUMENTATION**:\\nThis is the doc string of this python script:\\n\"\"\"\\nThis script prints hello world to the terminal. No arguments are required.\\n\\nExample usage:\\n    subprocess.call([sys.executable, \\'python/helloWorldPYTHON.py\\'])\\n\"\"\"\\n\\n\\n**CALL PYTHON SCRIPTS FROM PYTHON**:\\nUse the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\\n\\nTo call a Python script `example_script.py` which has no arguments:\\n    \\n```\\nExecute: subprocess.call([sys.executable, \\'<full path to script/> example_script.py\\'])\\n```\\n\\nTo call a Python script `example_script.py` with one argument:\\n\\n```\\nExecute: subprocess.call([sys.executable, \\'<full path to script/> example_script.py\\', \\'arg1\\'])\\n```\\n\\nTo call a Python script `example_script.py` with two arguments:\\n```\\nExecute: subprocess.call([sys.executable, \\'<full path to script/> example_script.py\\', \\'arg1\\', \\'arg2\\'])\\n```\\nQuery:{input}\\n\\n\\n**INSTRUCTIONS**\\n1. Given the user query and the documentation, identify each of the arguments found in the user\\'s query that should be passed to the Python script.\\n2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\\n3. The last line of your response should say \"Execute: <Python code to execute>\"\\n4. Format the response/output as:\\n    Arguments: \\n    Python Code Explanation: <2 sentences maximum>\\n    Execute: <your code here>\\n\\n**IMPORTANT**\\nThe code to execute from your response must be formatted as:\\n    Execute: subprocess.call([sys.executable, \\'<path to python script>\\', \\'<argument 1>\\', \\'<argument 2>\\', ..., \\'<argument n>\\'])\\nThis output should be exactly one line and no longer. Stop the response after this line.\\n'\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation:\n",
      "Human: /force CODE say hello world using python!\n",
      "BRAD:  Arguments: The user query does not provide any arguments to be passed to the Python script. The script documentation indicates no arguments are required.\n",
      "\n",
      "Python Code Explanation: We will execute the \"helloWorldPYTHON.py\" script with no arguments.\n",
      "\n",
      "Execute: subprocess.call([sys.executable, '/home/jpic/RAG-DEV/tutorials/running-codes/python/helloWorldPYTHON.py'])\n",
      "Human: /force CODE say hello world using matlab\n",
      "BRAD:  Arguments: No arguments are required for the \"helloWorldMATLAB\" function.\n",
      "\n",
      "MATLAB Code Explanation: We will execute the \"helloWorldMATLAB.m\" function with no arguments.\n",
      "\n",
      "Execute: eng.helloWorldMATLAB()\n",
      "Human: /force CODE say hello world using matlab\n",
      "BRAD:  Arguments: No arguments are required for the \"helloWorldMATLAB\" function.\n",
      "\n",
      "MATLAB Code Explanation: We will execute the \"helloWorldMATLAB\" function with no arguments.\n",
      "\n",
      "Execute: eng.helloWorldMATLAB()\n",
      "\n",
      "**PYTHON SCRIPT**\n",
      "You must run this python script: \n",
      "/home/jpic/RAG-DEV/tutorials/running-codes/python/helloWorldPYTHON.py\n",
      "\n",
      "**PYTHON SCRIPT DOCUMENTATION**:\n",
      "This is the doc string of this python script:\n",
      "\"\"\"\n",
      "This script prints hello world to the terminal. No arguments are required.\n",
      "\n",
      "Example usage:\n",
      "    subprocess.call([sys.executable, 'python/helloWorldPYTHON.py'])\n",
      "\"\"\"\n",
      "\n",
      "\n",
      "**CALL PYTHON SCRIPTS FROM PYTHON**:\n",
      "Use the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\n",
      "\n",
      "To call a Python script `example_script.py` which has no arguments:\n",
      "    \n",
      "```\n",
      "Execute: subprocess.call([sys.executable, '<full path to script/> example_script.py'])\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with one argument:\n",
      "\n",
      "```\n",
      "Execute: subprocess.call([sys.executable, '<full path to script/> example_script.py', 'arg1'])\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with two arguments:\n",
      "```\n",
      "Execute: subprocess.call([sys.executable, '<full path to script/> example_script.py', 'arg1', 'arg2'])\n",
      "```\n",
      "Query:/force CODE say hello world using python!\n",
      "\n",
      "\n",
      "**INSTRUCTIONS**\n",
      "1. Given the user query and the documentation, identify each of the arguments found in the user's query that should be passed to the Python script.\n",
      "2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\n",
      "3. The last line of your response should say \"Execute: <Python code to execute>\"\n",
      "4. Format the response/output as:\n",
      "    Arguments: \n",
      "    Python Code Explanation: <2 sentences maximum>\n",
      "    Execute: <your code here>\n",
      "\n",
      "**IMPORTANT**\n",
      "The code to execute from your response must be formatted as:\n",
      "    Execute: subprocess.call([sys.executable, '<path to python script>', '<argument 1>', '<argument 2>', ..., '<argument n>'])\n",
      "This output should be exactly one line and no longer. Stop the response after this line.\n",
      "\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "LLM OUTPUT PARSER\n",
      " Arguments: No arguments are required for the \"helloWorldPYTHON.py\" script according to the documentation and the user query.\n",
      "\n",
      "Python Code Explanation: We will execute the \"helloWorldPYTHON.py\" script with no arguments.\n",
      "\n",
      "Execute: subprocess.call([sys.executable, '/home/jpic/RAG-DEV/tutorials/running-codes/python/helloWorldPYTHON.py'])\n",
      "[' Arguments: No arguments are required for the \"helloWorldPYTHON.py\" script according to the documentation and the user query.\\n\\nPython Code Explanation: We will execute the \"helloWorldPYTHON.py\" script with no arguments.\\n\\n', \" subprocess.call([sys.executable, '/home/jpic/RAG-DEV/tutorials/running-codes/python/helloWorldPYTHON.py'])\"]\n",
      "Executing Code!!\n",
      "EVAL\n",
      "subprocess.call([sys.executable, '/home/jpic/RAG-DEV/tutorials/running-codes/python/helloWorldPYTHON.py'])\n",
      "END EVAL CHECK\n",
      "Hello World from a Python Script!\n",
      "Debug: PYTHON code executed successfully.\n",
      "==================================================\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  q\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Thanks for chatting today! I hope to talk soon, and don't forget that a record of this conversation is available at: /home/jpic/BRAD/2024-06-16-16:50:05.318931.json\n"
     ]
    }
   ],
   "source": [
    "from BRAD import brad\n",
    "brad.chat()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0638b74c-cdc3-4336-90de-7d6ef15014dc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78878573-5b4e-437a-a112-671aa57fc425",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9247861d-8023-4c0c-88ef-484fac25932e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "094373c3-6dc9-46cb-97dc-0891ad1f4cbf",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9563c151-cef6-4efc-8175-5c71bf6a563f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1dda72ea-7694-40b4-be8e-141e793cbbe6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "e229f04e-ec7c-4b13-a2aa-a1d4f0e6fe5d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "**Available Scripts**\n",
      "Script Name: animal_info, \t Description: This script provides information about different animals based on user input.\n",
      "Script Name: example, \t Description: This is the top-level docstring for the example script.\n",
      "Script Name: helloWorldPYTHON, \t Description: This script prints hello world to the terminal. No arguments are required.\n",
      "Script Name: matlabPipeline, \t Description:  matlabPipeline - A sample function to demonstrate argument handling\n",
      "Script Name: myFunc, \t Description:  myFunc - A simple function to multiply each element of the input array by 2.\n",
      "Script Name: helloWorldMATLAB, \t Description:  Print \"Hello World\"\n",
      "Script Name: saveplot, \t Description: SAVEPLOT Save the plot to a file\n",
      "Script Name: qualityControl, \t Description:  qualityControl - Constructs and runs a bioinformatics quality control pipeline.\n",
      "\n",
      "\n",
      "**User Query**\n",
      "{user_query}\n",
      "\n",
      "**Task**\n",
      "Based on the user's query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\n",
      "\n",
      "**Response Template**\n",
      "SCRIPT: <selected script>\n",
      "REASON: <reasoning why the script fits the user prompt>\n",
      "\n"
     ]
    }
   ],
   "source": [
    "scriptPurpose = {'animal_info': 'This script provides information about different animals based on user input.', 'example': 'This is the top-level docstring for the example script.', 'helloWorldPYTHON': 'This script prints hello world to the terminal. No arguments are required.', 'matlabPipeline': ' matlabPipeline - A sample function to demonstrate argument handling', 'myFunc': ' myFunc - A simple function to multiply each element of the input array by 2.', 'helloWorldMATLAB': ' Print \"Hello World\"', 'saveplot': 'SAVEPLOT Save the plot to a file', 'qualityControl': ' qualityControl - Constructs and runs a bioinformatics quality control pipeline.'}\n",
    "script_list = \"\"\n",
    "for script in scriptPurpose.keys():\n",
    "    script_list += \"Script Name: \" + script + \", \\t Description: \" + scriptPurpose[script] + '\\n'\n",
    "template = scriptSelectorTemplate()\n",
    "template = template.format(script_list=script_list)\n",
    "print(template)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "f9602f0a-5b9f-47cd-a591-5b605859bef3",
   "metadata": {},
   "outputs": [],
   "source": [
    "from BRAD import llms\n",
    "from langchain.memory import ConversationBufferMemory\n",
    "llm = llms.load_nvidia()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "7ba080f9-6672-4e25-8820-95f7dfaa8f45",
   "metadata": {},
   "outputs": [],
   "source": [
    "template = scriptSelectorTemplate()\n",
    "template = template.format(script_list=script_list)\n",
    "PROMPT = PromptTemplate(input_variables=[\"user_query\"], template=template)\n",
    "chain = PROMPT | llm\n",
    "res = chain.invoke(\"use python to say hello\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "e99d5376-f427-4822-891d-f92605fb8a3d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SCRIPT: helloWorldPYTHON\n",
      "REASON: The user query \"use python to say hello\" is asking for a simple Python script to print a greeting message to the terminal. The script named \"helloWorldPYTHON\" matches this query as its primary function is to print the message \"Hello World\" to the terminal in Python. However, since the user explicitly requested \"hello\" in their query, it is a more accurate match for the script named \"helloWorldPYTHON\" that explicitly prints the message \"hello\" instead. Therefore, \"helloWorldPYTHON\" is the best script to run to help the user.\n"
     ]
    }
   ],
   "source": [
    "print(res.content.strip())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "a8fb5916-fc2b-40fd-bda3-baac442f6d07",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'helloWorldPYTHON'"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scriptName = res.content.strip().split('\\n')[0].split(':')[1].strip()\n",
    "scriptName"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "2c4848ee-99d1-4cd5-a429-093d1a2e0754",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValidationError",
     "evalue": "1 validation error for ConversationChain\n__root__\n  Got unexpected prompt input variables. The prompt expects ['history', 'user_query'], but got ['history'] as inputs from memory, and input as the normal input key. (type=value_error)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValidationError\u001b[0m                           Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[33], line 4\u001b[0m\n\u001b[1;32m      2\u001b[0m template \u001b[38;5;241m=\u001b[39m template\u001b[38;5;241m.\u001b[39mformat(script_list\u001b[38;5;241m=\u001b[39mscript_list)\n\u001b[1;32m      3\u001b[0m PROMPT \u001b[38;5;241m=\u001b[39m PromptTemplate(input_variables\u001b[38;5;241m=\u001b[39m[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhistory\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124muser_query\u001b[39m\u001b[38;5;124m\"\u001b[39m], template\u001b[38;5;241m=\u001b[39mtemplate)\n\u001b[0;32m----> 4\u001b[0m conversation \u001b[38;5;241m=\u001b[39m \u001b[43mConversationChain\u001b[49m\u001b[43m(\u001b[49m\u001b[43mprompt\u001b[49m\u001b[43m  \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mPROMPT\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m      5\u001b[0m \u001b[43m                                 \u001b[49m\u001b[43mllm\u001b[49m\u001b[43m     \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mllm\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m      6\u001b[0m \u001b[43m                                 \u001b[49m\u001b[43mverbose\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m      7\u001b[0m \u001b[43m                                 \u001b[49m\u001b[43mmemory\u001b[49m\u001b[43m  \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mConversationBufferMemory\u001b[49m\u001b[43m(\u001b[49m\u001b[43mai_prefix\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mBRAD\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m      8\u001b[0m \u001b[43m                                \u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/langchain_core/_api/deprecation.py:183\u001b[0m, in \u001b[0;36mdeprecated.<locals>.deprecate.<locals>.finalize.<locals>.warn_if_direct_instance\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    181\u001b[0m     warned \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[1;32m    182\u001b[0m     emit_warning()\n\u001b[0;32m--> 183\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mwrapped\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/pydantic/v1/main.py:341\u001b[0m, in \u001b[0;36mBaseModel.__init__\u001b[0;34m(__pydantic_self__, **data)\u001b[0m\n\u001b[1;32m    339\u001b[0m values, fields_set, validation_error \u001b[38;5;241m=\u001b[39m validate_model(__pydantic_self__\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__class__\u001b[39m, data)\n\u001b[1;32m    340\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m validation_error:\n\u001b[0;32m--> 341\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m validation_error\n\u001b[1;32m    342\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    343\u001b[0m     object_setattr(__pydantic_self__, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m__dict__\u001b[39m\u001b[38;5;124m'\u001b[39m, values)\n",
      "\u001b[0;31mValidationError\u001b[0m: 1 validation error for ConversationChain\n__root__\n  Got unexpected prompt input variables. The prompt expects ['history', 'user_query'], but got ['history'] as inputs from memory, and input as the normal input key. (type=value_error)"
     ]
    }
   ],
   "source": [
    "template = scriptSelectorTemplate()\n",
    "template = template.format(script_list=script_list)\n",
    "PROMPT = PromptTemplate(input_variables=[\"history\", \"user_query\"], template=template)\n",
    "conversation = ConversationChain(prompt  = PROMPT,\n",
    "                                 llm     = llm,\n",
    "                                 verbose = True,\n",
    "                                 memory  = ConversationBufferMemory(ai_prefix=\"BRAD\"),\n",
    "                                )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6bc932e1-2256-4188-8598-8ebcbbcb0cb5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "**Available Scripts**\n",
      "{script_list}\n",
      "\n",
      "**User Query**\n",
      "{user_query}\n",
      "\n",
      "**Task**\n",
      "Based on the user's query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\n",
      "\n",
      "**Response Template**\n",
      "SCRIPT: <selected script>\n",
      "REASON: <reasoning why the script fits the user prompt>\n",
      "\n"
     ]
    }
   ],
   "source": [
    "template = scriptSelectorTemplate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "52f8896c-69e1-48d8-8d68-466f543490b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def scriptSelectorTemplate():\n",
    "    template=\"\"\"You must select which code to run to help a user.\n",
    "\n",
    "**Available Scripts**\n",
    "{script_list}\n",
    "\n",
    "**User Query**\n",
    "{{user_query}}\n",
    "\n",
    "**Task**\n",
    "Based on the user's query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\n",
    "\n",
    "**Response Template**\n",
    "SCRIPT: <selected script>\n",
    "REASON: <reasoning why the script fits the user prompt>\n",
    "\"\"\"\n",
    "    return template"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb0d08aa-5ff9-4a00-883d-4776fba40abd",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
