{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b5645027-5ee8-4a54-8062-f9858dad8864",
   "metadata": {},
   "source": [
    "# Planning Larger Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f7cb860d-0647-4dd4-a35b-931a580a9c61",
   "metadata": {},
   "outputs": [],
   "source": [
    "from BRAD import brad\n",
    "import importlib\n",
    "from BRAD import planner\n",
    "importlib.reload(planner)\n",
    "from BRAD.planner import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0333a24d-e684-46f3-8d56-eba58f4231be",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import os\n",
    "from langchain.document_loaders import DirectoryLoader\n",
    "from langchain.document_loaders import UnstructuredPDFLoader\n",
    "from langchain.vectorstores import Chroma\n",
    "from langchain.embeddings import HuggingFaceEmbeddings\n",
    "import chromadb\n",
    "from langchain_community.document_loaders import PyPDFLoader\n",
    "from langchain_text_splitters import RecursiveCharacterTextSplitter\n",
    "from langchain import PromptTemplate, LLMChain\n",
    "from langchain.memory import ConversationBufferMemory\n",
    "from langchain.chains import ConversationChain\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a9adf57d-c10b-4abe-82b0-33a2cac0b4b1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sun 16 Jun 2024 11:09:27 PM EDT INFO PyTorch version 2.0.1 available.\n",
      "Sun 16 Jun 2024 11:09:30 PM EDT INFO Use pytorch device_name: cpu\n",
      "Sun 16 Jun 2024 11:09:30 PM EDT INFO Load pretrained SentenceTransformer: BAAI/bge-base-en-v1.5\n",
      "/home/jpic/.local/lib/python3.11/site-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.\n",
      "  warnings.warn(\n",
      "Sun 16 Jun 2024 11:09:44 PM EDT INFO Anonymized telemetry enabled. See                     https://docs.trychroma.com/telemetry for more information.\n",
      "Sun 16 Jun 2024 11:09:45 PM EDT INFO Collection DigitalLibrary is not created.\n"
     ]
    }
   ],
   "source": [
    "# Load the database\n",
    "persist_directory = '/nfs/turbo/umms-indikar/shared/projects/RAG/databases/DigitalLibrary-10-June-2024/'\n",
    "embeddings_model = HuggingFaceEmbeddings(model_name='BAAI/bge-base-en-v1.5')\n",
    "db_name = \"DigitalLibrary\"\n",
    "_client_settings = chromadb.PersistentClient(path=(persist_directory + db_name))\n",
    "vectordb = Chroma(persist_directory=persist_directory, embedding_function=embeddings_model, client=_client_settings, collection_name=db_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c1d9608e-39da-4584-bd70-dfceb81f6772",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome to RAG! The chat log from this conversation will be saved to /home/jpic/BRAD/2024-06-16-23:11:48.232963.json. How can I help?\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sun 16 Jun 2024 11:11:48 PM EDT INFO local\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  /force PLANNER design an analysis pipeline to find the first order interactions of MYOD in the hardwired genome. Write a small report about which pathways and cell types these genes are implicated in\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sun 16 Jun 2024 11:12:19 PM EDT INFO PLANNER\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "RAG >> 1: \n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3m\n",
      "You are planning a bioinformatics analysis pipeline to address a user's query. Your task is to outline a multi-step workflow using the available methods listed below. Each method should be used appropriately to ensure a thorough analysis.\n",
      "For each step in the pipeline, explain the goal of the process as well as provide a prompt to the chatbot that will execute that step. You can use one method per step and the pipeline can be between 3-12 steps.\n",
      "\n",
      "Available Methods:\n",
      "1. **Retrieval Augmented Generation**(RAG): Look up literature and documents from a text database.\n",
      "2. **Web Search for New Literature**(SCRAPE): Search platforms like arXiv, bioRxiv, and PubMed for the latest research.\n",
      "3. **Bioinformatics Databases**(DATABASE): Utilize databases such as Gene Ontology and Enrichr to perform gene set enrichment analyses.\n",
      "4. **Run Codes**(CODE): Execute bioinformatics pipelines. Utilize prebuilt pipelines or develop new ones as needed.\n",
      "5. **Write a Report**(WRITE): Synthesize and summarize information. This can include summarizing database searches, code pipeline results, or creating a final report to encapsulate the entire analysis.\n",
      "\n",
      "Based on the user's query, create a detailed plan outlining the steps of the analysis. Ensure that each step is clearly defined and makes use of the appropriate method(s) listed above. Clarify which steps are dependent on one another, such as if information from a CODE step is used by DATABASE step, if information from a RAG step is used by a SCRAPE step, or if all previous steps are needed in a WRITE step, or any other dependencies.\n",
      "\n",
      "Current conversation: \n",
      "\n",
      "\n",
      "User Query: design an analysis pipeline to find the first order interactions of MYOD in the hardwired genome. Write a small report about which pathways and cell types these genes are implicated in\n",
      "\n",
      "Plan:\n",
      "**Step 1 (method, eg. RAG)**:\n",
      "Prompt: [Description of action for chatbot to do Step 1, i.e. do X, Y, Z]\n",
      "\n",
      "**Step 2 (method, eg. CODE)**:\n",
      "Prompt: [Description of action for chatbot to do Step 2, i.e. do X, Y, Z]\n",
      "\n",
      "**Step 3 (method, eg. DATABASE)**:\n",
      "Prompt: [Description of action for chatbot to do Step 3, i.e. do X, Y, Z]\n",
      "...\n",
      "\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sun 16 Jun 2024 11:12:22 PM EDT INFO RAG\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      " **Step 1 (method: RAG)**:\n",
      "Prompt: Retrieve information about first order interactions of MYOD from literature and documents in various databases. This may include interaction databases and papers that discuss such interactions.\n",
      "\n",
      "**Step 2 (method: CODE)**:\n",
      "Prompt: Execute a bioinformatics pipeline to identify the genes directly interacting with MYOD from the retrieved data. This may involve using tools such as STRING for protein-protein interaction analysis or other relevant pipelines.\n",
      "\n",
      "**Step 3 (method: DATABASE)**:\n",
      "Prompt: Utilize Gene Ontology and Enrichr databases to perform gene set enrichment analyses on the identified genes. This will help to determine which specific cell types and pathways these genes are implicated in.\n",
      "\n",
      "**Step 4 (method: SCRAPE)**:\n",
      "Prompt: Search platforms like arXiv, bioRxiv, and PubMed for the latest research on these gene interactions and their association with specific cell types and pathways. This information will be used to enhance and validate the findings from the previous database analysis.\n",
      "\n",
      "**Step 5 (method: CODE)**:\n",
      "Prompt: Dry run existing bioinformatics pipelines or develop new ones, as needed, to analyze the association of the identified genes with specific cell types and pathways using the data from the previous steps.\n",
      "\n",
      "**Step 6 (method: DATABASE)**:\n",
      "Prompt: Utilize multiple databases, such as PathwayCommondb and Reactome, to determine which cellular pathways these interacting genes are involved in.\n",
      "\n",
      "**Step 7 (method: WRITE)**:\n",
      "Prompt: Summarize the findings from each step into a final report. This should include an overview of the methods used, the stress-specific genes identified and their interactions, and any cell types or pathways implicated in the first order interactions of MYOD.\n",
      "\n",
      "\n",
      "==================================================\n",
      "==================================================\n",
      "RAG >> 2: \n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "interaction evidence directly from the source\n",
      "publications,andtheyarecoordinatingtheireffortsthrough\n",
      "theIMExconsortium ( 19). They provide highly valuable\n",
      "addedservicessuchascuratingmetadata,maintainingcom-mon name spaces and devising ontologies and standards.\n",
      "Asecondsourceofprotein–proteininteractioninformation\n",
      "is provided by computational prediction efforts, some ofwhicharehostedbydedicateddatabases,e.g.( 20,21).Lastly,\n",
      "a third class of databases is dedicated to protein interac-\n",
      "tions at the widest scope, integrating both primary as wellas predicted interactions, often including annotated path-\n",
      "way knowledge, text-mining results, inter-organism trans-\n",
      "fersorotheraccessoryinformation.TheSTRINGdatabase(‘Search Tool for Retrieval of Interacting Genes /Proteins’)\n",
      "belongs to this latter class, along with GeneMania ( 22),\n",
      "FunCoup ( 23), I2D (24), ConsensusPathDb ( 25), IMP\n",
      "(26) and HumanNet ( 27)––most of which have recently\n",
      "been reviewedand benchmarked in ( 7).\n",
      "STRING is one of the\n",
      "\n",
      "high-throughput text-mining as well as on a hi-\n",
      "erarchical clustering of the association network it-\n",
      "self. The STRING resource is available online at\n",
      "https://string-db.org/ .\n",
      "INTRODUCTION\n",
      "While an impressive amount of structural and functional\n",
      "information on individual proteins has been amassed ( 1–\n",
      "3), our knowledge about their interactions remains more\n",
      "fragmented. Some interactions are quite well documentedand understood, for example in the context of three-\n",
      "dimensional reconstructions of large cellular machineries\n",
      "(4–6), while others are only hinted at so far, through in-\n",
      "direct evidence such as genetic observations or statistical\n",
      "predictions. Furthermore, the space of potential protein–protein interactions is much larger, and also more context-\n",
      "dependent,thanthespaceofintrinsicmolecularfunctionof\n",
      "individual molecules. Interactions may not only be\n",
      "\n",
      "D608NucleicAcidsResearch,2019,Vol.47,Databaseissue\n",
      "tein,orwhentwoenzymesexchangeaspecificsubstratevia\n",
      "diffusion.\n",
      "Arguably, the common denominator of the various\n",
      "forms of protein–protein associations is informationflow––biologically meaningful interfaces have evolved to\n",
      "allowtheflowofinformationthroughthecell,andtheyare\n",
      "ultimately essential for implementing a functional system.Hence, it is desirable to collect and integrate all types of\n",
      "protein–protein interactions under one framework; this\n",
      "then provides support for data analysis pipelines in diverseareas, ranging from disease module identification ( 7,8)t o\n",
      "biomarker discovery ( 9–11) and allows manual browsing,\n",
      "adhocdiscovery and annotation.\n",
      "Protein–protein interactions can be collected from a\n",
      "number of online databases (reviewed in ( 12,13)), as well\n",
      "as from individual high-throughput efforts, e.g. ( 14). Pri-\n",
      "maryinteractiondatabases( 3,15–18)arejointlyannotating\n",
      "experimental interaction evidence directly from the\n",
      "\n",
      "not be mapped to PubMed. We also removed\n",
      "highly unspecific articles that mention more than 200 rel-\n",
      "evant biomedical entities such as proteins, chemicals, dis-eases or tissues. The final corpus consists of 28 579 637\n",
      "scientific publications, of which 2 106 542 are available as\n",
      "full-text articles and the remainder as abstracts. While thetext-miningpipelineitselfhasremainedunchanged(lastde-\n",
      "s c r i b e di n( 29)), its dictionary of gene and protein names\n",
      "has been updated to the new set of genomes and the stop-word list improved to increase precision, especially for hu-\n",
      "man proteins.\n",
      "NEW ENRICHMENT DETECTION MODE\n",
      "For users that query the STRING database with a set of\n",
      "proteins (as opposed to a single query protein only), the\n",
      "website computes a functional enrichment analysis in thebackground; this can then be inspected and browsed by\n",
      "the user, and includes interactive projections of the results\n",
      "ontotheuser’sproteinnetwork.Thisfunctionalityhasbeenavailable since version 9.1, and is based on\n",
      "\n",
      "interactome (BCI):  http://amdec-bioinfo.cu-genome.\n",
      "org/html/BCellInteractome.html\n",
      "Biochemical Genetic and Genomics knowledgebase (BIGG):  \n",
      "http://bigg.ucsd.edu\n",
      "Biological General Repository for Interaction Datasets \n",
      "(BioGRID):  http://thebiogrid.or g\n",
      "Biomolecular Interaction Network Database (BIND):   \n",
      "http://www.bind.ca\n",
      "CBS prediction server:  http://www.cbs.dtu.dk/services\n",
      "Database of Interacting Proteins (DIP):   \n",
      "http://dip.doe-mbi.ucla.edu/dip\n",
      "Human Protein Reference Database (HPRD):   \n",
      "http://www.hprd.or g\n",
      "JASPAR:  http://jaspar.genereg.net\n",
      "Kyoto Encyclopedia of Genes and Genomes (KEGG):   \n",
      "http://www.genome.jp/kegg\n",
      "microRNA:  http://www.microrna.or g\n",
      "miRBase:  http://www.mirbase.or g\n",
      "miRDB:  http://mirdb.or g\n",
      "miRecords:  http://mirecords.biolead.or g Molecular Interaction database (MINT):   \n",
      "http://mint.bio.uniroma2.it/mint\n",
      "Munich Information Center for Protein Sequence (MIPS) \n",
      "Mammalian Protein–Protein Interaction Database:\n",
      "\n",
      "https://www.orthodb.org/170\n",
      "gProfiler https://biit.cs.ut.ee/gprofiler/orth171\n",
      "Databases of protein–protein interactionsThe Human Protein Atlas https://www.proteinatlas.org/\n",
      "172\n",
      "STRING https://string-db.org/173\n",
      "BioGRID https://thebiogrid.org/174\n",
      "PICKLE http://www.pickle.gr175\n",
      "APID http://apid.dep.usal.es176\n",
      "IntAct https://www.ebi.ac.uk/intact177\n",
      "Pathway Commons http://www.pathwaycommons.org/178\n",
      "Ligand–receptor interactionsIUPHAR/BPS Guide to Pharmacology https://www.guidetopharmacology.org/\n",
      "179\n",
      "Resources for cell–cell interactions from \n",
      "the Bader laboratoryhttp://baderlab.org/CellCellInteractions32\n",
      "Compendium of ligand–receptor pairs in the literaturehttps://github.com/LewisLabUCSD/\n",
      "Ligand-Receptor-Pairshttps://doi.org/10.1038/\n",
      "s41576-020-00292-x\n",
      "www.nature.com/nrgReviews\n",
      "78 | February 2021 | volume 22\n",
      "\n",
      "functional \n",
      "profiling of gene lists from large- scale experiments. \n",
      "Nucleic Acids Res. 35, W193–W200 (2007).\n",
      "172. Uhlén, M. et al. Tissue- based map of the human \n",
      "proteome. Science 347, 1260419 (2015).\n",
      "173. Szklarczyk, D. et al. STRING v11: protein–protein \n",
      "association networks with increased coverage, \n",
      "supporting functional discovery in genome- wide \n",
      "experimental datasets. Nucleic Acids Res. 47, \n",
      "D607–D613 (2019).\n",
      "174. Stark, C. et al. BioGRID: a general repository for \n",
      "interaction datasets. Nucleic Acids Res. 34, \n",
      "D535–D539 (2006).\n",
      "175. Gioutlakis, A., Klapa, M. I. & Moschonas, N. K. PICKLE \n",
      "2.0: a human protein- protein interaction meta- database \n",
      "employing data integration via genetic information \n",
      "ontology. PLoS ONE 12, e0186039 (2017).\n",
      "176. Prieto, C. & De Las Rivas, J. APID: agile protein \n",
      "interaction DataAnalyzer. Nucleic Acids Res. 34, \n",
      "W298–W302 (2006).\n",
      "177. Hermjakob, H. et al. IntAct: an open source molecular \n",
      "interaction database. Nucleic Acids Res. 32, \n",
      "D452–D455\n",
      "\n",
      "Association\n",
      "NetworkTo construct the transcriptional regulatory network for each tran-\n",
      "scription factor, an extensive database of the interactions between\n",
      "the genes is necessary. Proteins are the major product of gene\n",
      "expression, and the interactions between expressed proteins canalter the physiology of the cell and in turn have a knock-on effect\n",
      "on the transcriptional regulation. Thus, we used the STRING\n",
      "database [ 67], consisting of direct (physical) and indirect (func-\n",
      "tional) associations, in order to provide interactions beyond the\n",
      "transcriptional level.\n",
      "Each interaction in the STRING database is given seven evi-\n",
      "dence scores based on a variety of information, including (1) actualexperiments, (2) curated biological pathways, (3) automated text\n",
      "mining of literature, and (4) gene coexpression data. The remainingthree genome-based evidence scores are less relevant to human\n",
      "cells, involving the proximity of genes in the genome neighbor-\n",
      "hood, association of pairs of proteins with\n",
      "\n",
      "systems, which capturemany of their essential characteristics. In thenetwork, molecules are represented by nodes,and their interactions are represented by edg-es (or arrows). The cell can be viewed as anoverlay of at least three types of networks,which describes protein-protein, protein-DNA, and protein-metabolite interactions. In-herent in this description is suppression ofdetail: many different mechanisms of tran-scription regulation, for example, may be de-scribed by a single type of arrow. Further-more, the interactions can be of differentstrengths, so there should be numbers orweights on each arrow (3). Whenever two or\n",
      "more arrows converge on a node, an inputfunction needs to be specified (for example,AND or OR gates) (4, 5). At present, many of\n",
      "the connections, numbers and input functionsare not known. However, something can stillbe learned even from the very incompletenetworks currently available (6–8). First, thenetwork description allows application oftools and concepts (9)\n",
      "\n",
      "by TFs\n",
      "and enhance gene expression downstream of the genome. The en-hancer data were subdivided based on the tissue; therefore, we\n",
      "extracted data on specific tissues (e.g. liver tissue), which were\n",
      "manually curated.\n",
      "2.5 Molecular interaction data\n",
      "We obtained gene–gene association data from ‘ Homo sapiens ’ from\n",
      "the STRING database ( Szklarczyk et al. , 2019 ). Several types of\n",
      "interaction exist such as conserved neighborhood, co-occurrence, fu-sion, coexpression, experiment, database and text mining interac-\n",
      "tions. In the present study, we used gene–gene associations based on\n",
      "coexpression associations, coexpression; accordingly, our resultsincluded 6 080 249 associations among 18 201 genes.\n",
      "2.6 Cell ontology data\n",
      "To create cell ontology tree, we obtained cell ontology data from the\n",
      "National Center for Biomedical Ontology (NCBO) Bioportal ( Diehl\n",
      "et al. , 2016 ), in which ontology terms are assigned according to the\n",
      "principles of the Open Biological and Biomedical Ontologies\n",
      "Foundry (\n",
      "Human: Retrieve information about first order interactions of MYOD from literature and documents in various databases. This may include interaction databases and papers that discuss such interactions.\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sun 16 Jun 2024 11:12:24 PM EDT INFO CODE\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      " The STRING database is a resource for protein-protein interaction information. It provides various types of interactions including experimental, curated, and predicted interactions. The database also integrates interaction data from different sources and provides added value services such as curating metadata, maintaining common name spaces, and devising ontologies and standards. In addition to STRING, there are other databases that provide protein interaction information at a wider scope, integrating both primary and predicted interactions. Some examples of these databases include GeneMania, FunCoup, I2D, ConsensusPathDb, IMP, and HumanNet. STRING uses text-mining and hierarchical clustering to identify protein interactions. It also provides a functional enrichment analysis feature for users who query the database with a set of proteins. The database is available online at <https://string-db.org/>.\n",
      "==================================================\n",
      "==================================================\n",
      "RAG >> 3: CODER\n",
      "input_variables=['user_query'] template='You must select which code to run to help a user.\\n\\n**Available Scripts**\\nScript Name: animal_info, \\t Description: This script provides information about different animals based on user input.\\nScript Name: hwgFirstOrderInteractions, \\t Description: \\nScript Name: example, \\t Description: This is the top-level docstring for the example script.\\nScript Name: helloWorldPYTHON, \\t Description: This script prints hello world to the terminal. No arguments are required.\\nScript Name: matlabPipeline, \\t Description:  matlabPipeline - A sample function to demonstrate argument handling\\nScript Name: myFunc, \\t Description:  myFunc - A simple function to multiply each element of the input array by 2.\\nScript Name: helloWorldMATLAB, \\t Description:  Print \"Hello World\"\\nScript Name: saveplot, \\t Description: SAVEPLOT Save the plot to a file\\nScript Name: qualityControl, \\t Description:  qualityControl - Constructs and runs a bioinformatics quality control pipeline.\\n\\n\\n**User Query**\\n{user_query}\\n\\n**Task**\\nBased on the user\\'s query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\\n\\n**Response Template**\\nSCRIPT: <selected script>\\nREASON: <reasoning why the script fits the user prompt>\\n'\n",
      "FIRST LLM CALL\n",
      "Warning! the type doesnt exist\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...\n",
      "To disable this warning, you can either:\n",
      "\t- Avoid using `tokenizers` before the fork if possible\n",
      "\t- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Matlab Engine On\n",
      "ALL SCRIPTS FOUND. BUILDING TEMPLATE\n",
      "input_variables=['history', 'input'] template='Current conversation:\\n{history}\\n\\n**MATLAB SCRIPT**\\nYou must run this python script: \\nqualityControl.m\\n\\n**MATLAB FUNCTION DOCUMENTATION**:\\nfunction output = qualityControl(fileName)\\n% qualityControl - Constructs and runs a bioinformatics quality control pipeline.\\n%\\n% BRAD\\'s Example to call this function:\\n%    Execute: eng.qualityControl(\"SRR005164_1_50.fastq\");\\n%\\n% This function creates a bioinformatics quality control pipeline using the\\n% MATLAB bioinformatics toolbox. It reads a FASTQ file, applies sequence\\n% filtering, and generates a quality control plot.\\n%\\n% The pipeline consists of the following steps:\\n% 1. FileChooser block to select the input FASTQ file.\\n% 2. SeqFilter block to filter the sequences based on quality thresholds.\\n% 3. UserFunction block to generate a quality control plot of the filtered sequences.\\n%\\n% The blocks are connected in sequence and executed in the specified order.\\n%\\n% Reference: https://www.mathworks.com/help/bioinfo/ref/bioinfo.pipeline.pipeline.run.html\\n%\\n% Auth: Joshua Pickard\\n%       jpic@umich.edu\\n% Date: June 14, 2024\\n\\n**CALL MATLAB FUNCTIONS FROM PYTHON**:\\nUse MATLAB® Engine API for Python® to call any MATLAB function on the MATLAB path. Some examples to call a few common matlab functions include:\\n\\nTo call the matlab function myFnc which has no arguments:\\n```\\nExecute: eng.myFnc()\\n```\\n\\nTo use the is prime function which requires an input argument:\\n```\\nExecute: tf = eng.isprime(37)\\n```\\n\\nTo determine the greatest common denominator of two numbers, use the gcd function. Set nargout to return the three output arguments from gcd.\\n```\\nExecute: t = eng.gcd(100.0,80.0,nargout=3)\\n```\\n\\nQuery:{input}\\n\\n**INSTRUCTIONS**\\n1. Given the user query and the documentation, identify each of the arguments found in the users query that should be passed to the matlab function.\\n2. Using the matlab enging eng, provide the one line of code to execute the desired matlab commands. Assume all functions are added to the path and eng already exist.\\n3. The last line of your response should say \"Execute: <MATLAB codes to execute>\"\\n4. Format the response/output as:\\nArguments: \\nMATLAB Code Explanation: <2 sentences maximum>\\nExecute: <your code here>\\n\\n**IMPORTANT**\\nThe code to execute from your response must be formatted as:\\n    Execute: eng.<function name>(<arguments>)>\\nThis output should be exactly one line and no longer. Stop the response after this line.\\n'\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation:\n",
      "Human: design an analysis pipeline to find the first order interactions of MYOD in the hardwired genome. Write a small report about which pathways and cell types these genes are implicated in\n",
      "BRAD:  **Step 1 (method: RAG)**:\n",
      "Prompt: Retrieve information about first order interactions of MYOD from literature and documents in various databases. This may include interaction databases and papers that discuss such interactions.\n",
      "\n",
      "**Step 2 (method: CODE)**:\n",
      "Prompt: Execute a bioinformatics pipeline to identify the genes directly interacting with MYOD from the retrieved data. This may involve using tools such as STRING for protein-protein interaction analysis or other relevant pipelines.\n",
      "\n",
      "**Step 3 (method: DATABASE)**:\n",
      "Prompt: Utilize Gene Ontology and Enrichr databases to perform gene set enrichment analyses on the identified genes. This will help to determine which specific cell types and pathways these genes are implicated in.\n",
      "\n",
      "**Step 4 (method: SCRAPE)**:\n",
      "Prompt: Search platforms like arXiv, bioRxiv, and PubMed for the latest research on these gene interactions and their association with specific cell types and pathways. This information will be used to enhance and validate the findings from the previous database analysis.\n",
      "\n",
      "**Step 5 (method: CODE)**:\n",
      "Prompt: Dry run existing bioinformatics pipelines or develop new ones, as needed, to analyze the association of the identified genes with specific cell types and pathways using the data from the previous steps.\n",
      "\n",
      "**Step 6 (method: DATABASE)**:\n",
      "Prompt: Utilize multiple databases, such as PathwayCommondb and Reactome, to determine which cellular pathways these interacting genes are involved in.\n",
      "\n",
      "**Step 7 (method: WRITE)**:\n",
      "Prompt: Summarize the findings from each step into a final report. This should include an overview of the methods used, the stress-specific genes identified and their interactions, and any cell types or pathways implicated in the first order interactions of MYOD.\n",
      "\n",
      "**MATLAB SCRIPT**\n",
      "You must run this python script: \n",
      "qualityControl.m\n",
      "\n",
      "**MATLAB FUNCTION DOCUMENTATION**:\n",
      "function output = qualityControl(fileName)\n",
      "% qualityControl - Constructs and runs a bioinformatics quality control pipeline.\n",
      "%\n",
      "% BRAD's Example to call this function:\n",
      "%    Execute: eng.qualityControl(\"SRR005164_1_50.fastq\");\n",
      "%\n",
      "% This function creates a bioinformatics quality control pipeline using the\n",
      "% MATLAB bioinformatics toolbox. It reads a FASTQ file, applies sequence\n",
      "% filtering, and generates a quality control plot.\n",
      "%\n",
      "% The pipeline consists of the following steps:\n",
      "% 1. FileChooser block to select the input FASTQ file.\n",
      "% 2. SeqFilter block to filter the sequences based on quality thresholds.\n",
      "% 3. UserFunction block to generate a quality control plot of the filtered sequences.\n",
      "%\n",
      "% The blocks are connected in sequence and executed in the specified order.\n",
      "%\n",
      "% Reference: https://www.mathworks.com/help/bioinfo/ref/bioinfo.pipeline.pipeline.run.html\n",
      "%\n",
      "% Auth: Joshua Pickard\n",
      "%       jpic@umich.edu\n",
      "% Date: June 14, 2024\n",
      "\n",
      "**CALL MATLAB FUNCTIONS FROM PYTHON**:\n",
      "Use MATLAB® Engine API for Python® to call any MATLAB function on the MATLAB path. Some examples to call a few common matlab functions include:\n",
      "\n",
      "To call the matlab function myFnc which has no arguments:\n",
      "```\n",
      "Execute: eng.myFnc()\n",
      "```\n",
      "\n",
      "To use the is prime function which requires an input argument:\n",
      "```\n",
      "Execute: tf = eng.isprime(37)\n",
      "```\n",
      "\n",
      "To determine the greatest common denominator of two numbers, use the gcd function. Set nargout to return the three output arguments from gcd.\n",
      "```\n",
      "Execute: t = eng.gcd(100.0,80.0,nargout=3)\n",
      "```\n",
      "\n",
      "Query:Execute a bioinformatics pipeline to identify the genes directly interacting with MYOD from the retrieved data. This may involve using tools such as STRING for protein-protein interaction analysis or other relevant pipelines.\n",
      "\n",
      "**INSTRUCTIONS**\n",
      "1. Given the user query and the documentation, identify each of the arguments found in the users query that should be passed to the matlab function.\n",
      "2. Using the matlab enging eng, provide the one line of code to execute the desired matlab commands. Assume all functions are added to the path and eng already exist.\n",
      "3. The last line of your response should say \"Execute: <MATLAB codes to execute>\"\n",
      "4. Format the response/output as:\n",
      "Arguments: \n",
      "MATLAB Code Explanation: <2 sentences maximum>\n",
      "Execute: <your code here>\n",
      "\n",
      "**IMPORTANT**\n",
      "The code to execute from your response must be formatted as:\n",
      "    Execute: eng.<function name>(<arguments>)>\n",
      "This output should be exactly one line and no longer. Stop the response after this line.\n",
      "\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sun 16 Jun 2024 11:12:51 PM EDT INFO DATABASE\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "LLM OUTPUT PARSER\n",
      " Arguments: \"fileName\" (name of the data file containing the interactions of MYOD with other genes)\n",
      "MATLAB Code Explanation: Reads the input data file and applies bioinformatics pipeline for identification of genes interacting with MYOD using STRING or other relevant pipelines.\n",
      "Execute: eng.qualityControl(\"MYOD\\_interactions.txt\"); % Assuming the input file is named \"MYOD\\_interactions.txt\" and is in the current working directory. Replace \"qualityControl\" function with the appropriate function name for protein-protein interaction analysis.\n",
      "[' Arguments: \"fileName\" (name of the data file containing the interactions of MYOD with other genes)\\nMATLAB Code Explanation: Reads the input data file and applies bioinformatics pipeline for identification of genes interacting with MYOD using STRING or other relevant pipelines.\\n', ' eng.qualityControl(\"MYOD\\\\_interactions.txt\"); % Assuming the input file is named \"MYOD\\\\_interactions.txt\" and is in the current working directory. Replace \"qualityControl\" function with the appropriate function name for protein-protein interaction analysis.']\n",
      "Executing Code!!\n",
      "EVAL\n",
      "eng.qualityControl(\"MYOD\\_interactions.txt\"); % Assuming the input file is named \"MYOD\\_interactions.txt\" and is in the current working directory. Replace \"qualityControl\" function with the appropriate function name for protein-protein interaction analysis.\n",
      "END EVAL CHECK\n",
      "Debug: Syntax error in the MATLAB code: invalid syntax (<string>, line 1)\n",
      "==================================================\n",
      "==================================================\n",
      "RAG >> 4: \n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation:\n",
      "Human: design an analysis pipeline to find the first order interactions of MYOD in the hardwired genome. Write a small report about which pathways and cell types these genes are implicated in\n",
      "BRAD:  **Step 1 (method: RAG)**:\n",
      "Prompt: Retrieve information about first order interactions of MYOD from literature and documents in various databases. This may include interaction databases and papers that discuss such interactions.\n",
      "\n",
      "**Step 2 (method: CODE)**:\n",
      "Prompt: Execute a bioinformatics pipeline to identify the genes directly interacting with MYOD from the retrieved data. This may involve using tools such as STRING for protein-protein interaction analysis or other relevant pipelines.\n",
      "\n",
      "**Step 3 (method: DATABASE)**:\n",
      "Prompt: Utilize Gene Ontology and Enrichr databases to perform gene set enrichment analyses on the identified genes. This will help to determine which specific cell types and pathways these genes are implicated in.\n",
      "\n",
      "**Step 4 (method: SCRAPE)**:\n",
      "Prompt: Search platforms like arXiv, bioRxiv, and PubMed for the latest research on these gene interactions and their association with specific cell types and pathways. This information will be used to enhance and validate the findings from the previous database analysis.\n",
      "\n",
      "**Step 5 (method: CODE)**:\n",
      "Prompt: Dry run existing bioinformatics pipelines or develop new ones, as needed, to analyze the association of the identified genes with specific cell types and pathways using the data from the previous steps.\n",
      "\n",
      "**Step 6 (method: DATABASE)**:\n",
      "Prompt: Utilize multiple databases, such as PathwayCommondb and Reactome, to determine which cellular pathways these interacting genes are involved in.\n",
      "\n",
      "**Step 7 (method: WRITE)**:\n",
      "Prompt: Summarize the findings from each step into a final report. This should include an overview of the methods used, the stress-specific genes identified and their interactions, and any cell types or pathways implicated in the first order interactions of MYOD.\n",
      "Human: Execute a bioinformatics pipeline to identify the genes directly interacting with MYOD from the retrieved data. This may involve using tools such as STRING for protein-protein interaction analysis or other relevant pipelines.\n",
      "BRAD:  Arguments: \"fileName\" (name of the data file containing the interactions of MYOD with other genes)\n",
      "MATLAB Code Explanation: Reads the input data file and applies bioinformatics pipeline for identification of genes interacting with MYOD using STRING or other relevant pipelines.\n",
      "Execute: eng.qualityControl(\"MYOD\\_interactions.txt\"); % Assuming the input file is named \"MYOD\\_interactions.txt\" and is in the current working directory. Replace \"qualityControl\" function with the appropriate function name for protein-protein interaction analysis.\n",
      "    \n",
      "    GENEONTOLOGY: The Gene Ontology (GO) is an initiative to unify the representation of gene and gene product attributes across all species via the aims: 1) maintain and develop its controlled vocabulary of gene and gene product attributes; 2) annotate genes and gene products, and assimilate and disseminate annotation data; and 3) provide tools for easy access to all aspects of the data provided by the project, and to enable functional interpretation of experimental data using the GO.\n",
      "    \n",
      "    ENRICHR: is a tool used to lookup sets of genes and their functional association. ENRICHR has access to many gene-set libraries including Allen_Brain_Atlas_up, ENCODE_Histone_Modifications_2015, Enrichr_Libraries_Most_Popular_Genes, FANTOM6_lncRNA_KD_DEGs, GO_Biological_Process_2023, GTEx, Human_Gene_Atlas, KEGG, REACTOME, Transcription_Factor_PPIs, WikiPathways and many others databases.\n",
      "    \n",
      "    Available Tables:\n",
      "    \n",
      "    \n",
      "    Query:Utilize Gene Ontology and Enrichr databases to perform gene set enrichment analyses on the identified genes. This will help to determine which specific cell types and pathways these genes are implicated in.\n",
      "    \n",
      "    **INSTRUCTIONS**\n",
      "    1. From the query, decide if GENEONTOLOGY or ENRICHR should be used\n",
      "    2. Identify genes in the users input that should be searched. propose genes with similar names, correct the users spelling, or make small modifications to the list, but do not propose genes that are not found in the humans input.\n",
      "    3. If the user wants to extract genes from a table (python dataframe), provide the necessary code to get the genes into a python list, otherwise, say None.\n",
      "    4. If there is code required, identify the name of the table, otherwise say None\n",
      "    5. If there is code required, identify the row or column name of the table, otherwise say None\n",
      "    6. If there is code required, specify if it is a row or column\n",
      "    Format your output as follows with no additional information:\n",
      "    \n",
      "    database: <ENRICHR or GENEONTOLOGY>\n",
      "    genes: <List of genes separated by commas in query or None if code is required>\n",
      "    code: <True or None>\n",
      "    table: <Table Name>\n",
      "    mode name: <Row or Column Name>\n",
      "    mode: <Row or Column>\n",
      "    \u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sun 16 Jun 2024 11:12:53 PM EDT INFO Performing Enichr analysis using database GO_Biological_Process_2021.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      " database: ENRICHR\n",
      "genes: None\n",
      "code: None\n",
      "table: None\n",
      "mode name: None\n",
      "mode: None\n",
      "\n",
      "Explanation: The first step in the query asks to identify genes that interact directly with MYOD. To accomplish this task, a bioinformatics pipeline will be employed. No specific genes are mentioned in the query, so no modifications are needed. There is no need to extract genes from a table as the interaction data for MYOD should be provided in the input data file. Thus, the genes, code, table, and mode name/mode are all set to None. The database to be used is ENRICHR, as it is mentioned in the query for performing gene set enrichment analysis.\n",
      " database: ENRICHR\n",
      "genes: None\n",
      "code: None\n",
      "table: None\n",
      "mode name: None\n",
      "mode: None\n",
      "\n",
      "Explanation: The first step in the query asks to identify genes that interact directly with MYOD. To accomplish this task, a bioinformatics pipeline will be employed. No specific genes are mentioned in the query, so no modifications are needed. There is no need to extract genes from a table as the interaction data for MYOD should be provided in the input data file. Thus, the genes, code, table, and mode name/mode are all set to None. The database to be used is ENRICHR, as it is mentioned in the query for performing gene set enrichment analysis.\n",
      "None\n",
      "{'database': 'ENRICHR', 'genes': ['None'], 'code': 'None'}\n",
      "['None']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sun 16 Jun 2024 11:12:53 PM EDT ERROR \n",
      "            No Enrichr results were found for genes ['None'] and database GO_Biological_Process_2021. \n",
      "\n",
      "            If the genes are Ensembl IDs, please set argument 'ensembl=True' (for terminal, add flag: [--ensembl]).\n",
      "            \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The following table was generated by quering the gene list against GO_Biological_Process_2021:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "</style>\n",
       "<table id=\"T_5bb02\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th id=\"T_5bb02_level0_col0\" class=\"col_heading level0 col0\" >rank</th>\n",
       "      <th id=\"T_5bb02_level0_col1\" class=\"col_heading level0 col1\" >path_name</th>\n",
       "      <th id=\"T_5bb02_level0_col2\" class=\"col_heading level0 col2\" >p_val</th>\n",
       "      <th id=\"T_5bb02_level0_col3\" class=\"col_heading level0 col3\" >z_score</th>\n",
       "      <th id=\"T_5bb02_level0_col4\" class=\"col_heading level0 col4\" >combined_score</th>\n",
       "      <th id=\"T_5bb02_level0_col5\" class=\"col_heading level0 col5\" >overlapping_genes</th>\n",
       "      <th id=\"T_5bb02_level0_col6\" class=\"col_heading level0 col6\" >adj_p_val</th>\n",
       "      <th id=\"T_5bb02_level0_col7\" class=\"col_heading level0 col7\" >database</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x1506ad8e6a90>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sun 16 Jun 2024 11:12:54 PM EDT INFO SCRAPE\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The table has been saved to: RAG-gget-GO_Biological_Process_2021-2024-06-1623:12:54.136051.csv\n",
      "==================================================\n",
      "==================================================\n",
      "RAG >> 5: \n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation:\n",
      "Human: design an analysis pipeline to find the first order interactions of MYOD in the hardwired genome. Write a small report about which pathways and cell types these genes are implicated in\n",
      "BRAD:  **Step 1 (method: RAG)**:\n",
      "Prompt: Retrieve information about first order interactions of MYOD from literature and documents in various databases. This may include interaction databases and papers that discuss such interactions.\n",
      "\n",
      "**Step 2 (method: CODE)**:\n",
      "Prompt: Execute a bioinformatics pipeline to identify the genes directly interacting with MYOD from the retrieved data. This may involve using tools such as STRING for protein-protein interaction analysis or other relevant pipelines.\n",
      "\n",
      "**Step 3 (method: DATABASE)**:\n",
      "Prompt: Utilize Gene Ontology and Enrichr databases to perform gene set enrichment analyses on the identified genes. This will help to determine which specific cell types and pathways these genes are implicated in.\n",
      "\n",
      "**Step 4 (method: SCRAPE)**:\n",
      "Prompt: Search platforms like arXiv, bioRxiv, and PubMed for the latest research on these gene interactions and their association with specific cell types and pathways. This information will be used to enhance and validate the findings from the previous database analysis.\n",
      "\n",
      "**Step 5 (method: CODE)**:\n",
      "Prompt: Dry run existing bioinformatics pipelines or develop new ones, as needed, to analyze the association of the identified genes with specific cell types and pathways using the data from the previous steps.\n",
      "\n",
      "**Step 6 (method: DATABASE)**:\n",
      "Prompt: Utilize multiple databases, such as PathwayCommondb and Reactome, to determine which cellular pathways these interacting genes are involved in.\n",
      "\n",
      "**Step 7 (method: WRITE)**:\n",
      "Prompt: Summarize the findings from each step into a final report. This should include an overview of the methods used, the stress-specific genes identified and their interactions, and any cell types or pathways implicated in the first order interactions of MYOD.\n",
      "Human: Execute a bioinformatics pipeline to identify the genes directly interacting with MYOD from the retrieved data. This may involve using tools such as STRING for protein-protein interaction analysis or other relevant pipelines.\n",
      "BRAD:  Arguments: \"fileName\" (name of the data file containing the interactions of MYOD with other genes)\n",
      "MATLAB Code Explanation: Reads the input data file and applies bioinformatics pipeline for identification of genes interacting with MYOD using STRING or other relevant pipelines.\n",
      "Execute: eng.qualityControl(\"MYOD\\_interactions.txt\"); % Assuming the input file is named \"MYOD\\_interactions.txt\" and is in the current working directory. Replace \"qualityControl\" function with the appropriate function name for protein-protein interaction analysis.\n",
      "Human: Utilize Gene Ontology and Enrichr databases to perform gene set enrichment analyses on the identified genes. This will help to determine which specific cell types and pathways these genes are implicated in.\n",
      "BRAD:  database: ENRICHR\n",
      "genes: None\n",
      "code: None\n",
      "table: None\n",
      "mode name: None\n",
      "mode: None\n",
      "\n",
      "Explanation: The first step in the query asks to identify genes that interact directly with MYOD. To accomplish this task, a bioinformatics pipeline will be employed. No specific genes are mentioned in the query, so no modifications are needed. There is no need to extract genes from a table as the interaction data for MYOD should be provided in the input data file. Thus, the genes, code, table, and mode name/mode are all set to None. The database to be used is ENRICHR, as it is mentioned in the query for performing gene set enrichment analysis.\n",
      "    \n",
      "    Query:Search platforms like arXiv, bioRxiv, and PubMed for the latest research on these gene interactions and their association with specific cell types and pathways. This information will be used to enhance and validate the findings from the previous database analysis.\n",
      "    \n",
      "    From the query, decide if ARXIV, PUBMED, or BIORXIV should be searched, and propose no more than 10 search terms for this query and database. Separate each term with a comma, and provide no extra information/explination for either the database or search terms. Format your output as follows with no additions:\n",
      "    \n",
      "    Database: <ARXIV, PUBMED, or BIORXIV>\n",
      "    Search Terms: <improved search terms>\n",
      "    \u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "searching on PUBMED...\n",
      "Search Terms: ['\"MYOD protein-protein interactions', 'MYOD genetic interactions', 'MYOD cellular pathways', 'MYOD gene expression', 'MYOD cell types', 'MYOD gene regulation', 'MYOD signaling pathways', 'MYOD molecular mechanism', 'MYOD chemical interactions', 'MYOD disease associations\"']\n",
      "Directory '/home/jpic/RAG-DEV/tutorials/planner/specialized_docs' created successfully\n",
      "20641378 could not be gathered.\n",
      "12705589 could not be gathered.\n",
      "pdf collection complete!\n",
      "Directory '/home/jpic/RAG-DEV/tutorials/planner/specialized_docs' created successfully\n",
      "23406902 could not be gathered.\n",
      "pdf collection complete!\n",
      "Directory '/home/jpic/RAG-DEV/tutorials/planner/specialized_docs' created successfully\n",
      "pdf collection complete!\n",
      "Directory '/home/jpic/RAG-DEV/tutorials/planner/specialized_docs' created successfully\n",
      "pdf collection complete!\n",
      "Directory '/home/jpic/RAG-DEV/tutorials/planner/specialized_docs' created successfully\n",
      "pdf collection complete!\n",
      "Directory '/home/jpic/RAG-DEV/tutorials/planner/specialized_docs' created successfully\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sun 16 Jun 2024 11:15:02 PM EDT INFO CODE\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error occurred while searching on PUBMED: 'NoneType' object has no attribute 'attrs'\n",
      "==================================================\n",
      "==================================================\n",
      "RAG >> 6: CODER\n",
      "input_variables=['user_query'] template='You must select which code to run to help a user.\\n\\n**Available Scripts**\\nScript Name: animal_info, \\t Description: This script provides information about different animals based on user input.\\nScript Name: hwgFirstOrderInteractions, \\t Description: \\nScript Name: example, \\t Description: This is the top-level docstring for the example script.\\nScript Name: helloWorldPYTHON, \\t Description: This script prints hello world to the terminal. No arguments are required.\\nScript Name: matlabPipeline, \\t Description:  matlabPipeline - A sample function to demonstrate argument handling\\nScript Name: myFunc, \\t Description:  myFunc - A simple function to multiply each element of the input array by 2.\\nScript Name: helloWorldMATLAB, \\t Description:  Print \"Hello World\"\\nScript Name: saveplot, \\t Description: SAVEPLOT Save the plot to a file\\nScript Name: qualityControl, \\t Description:  qualityControl - Constructs and runs a bioinformatics quality control pipeline.\\n\\n\\n**User Query**\\n{user_query}\\n\\n**Task**\\nBased on the user\\'s query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\\n\\n**Response Template**\\nSCRIPT: <selected script>\\nREASON: <reasoning why the script fits the user prompt>\\n'\n",
      "FIRST LLM CALL\n",
      "Warning! the type doesnt exist\n",
      "Matlab Engine On\n",
      "ALL SCRIPTS FOUND. BUILDING TEMPLATE\n",
      "input_variables=['history', 'input'] template='Current conversation:\\n{history}\\n\\n**MATLAB SCRIPT**\\nYou must run this python script: \\nqualityControl.m\\n\\n**MATLAB FUNCTION DOCUMENTATION**:\\nfunction output = qualityControl(fileName)\\n% qualityControl - Constructs and runs a bioinformatics quality control pipeline.\\n%\\n% BRAD\\'s Example to call this function:\\n%    Execute: eng.qualityControl(\"SRR005164_1_50.fastq\");\\n%\\n% This function creates a bioinformatics quality control pipeline using the\\n% MATLAB bioinformatics toolbox. It reads a FASTQ file, applies sequence\\n% filtering, and generates a quality control plot.\\n%\\n% The pipeline consists of the following steps:\\n% 1. FileChooser block to select the input FASTQ file.\\n% 2. SeqFilter block to filter the sequences based on quality thresholds.\\n% 3. UserFunction block to generate a quality control plot of the filtered sequences.\\n%\\n% The blocks are connected in sequence and executed in the specified order.\\n%\\n% Reference: https://www.mathworks.com/help/bioinfo/ref/bioinfo.pipeline.pipeline.run.html\\n%\\n% Auth: Joshua Pickard\\n%       jpic@umich.edu\\n% Date: June 14, 2024\\n\\n**CALL MATLAB FUNCTIONS FROM PYTHON**:\\nUse MATLAB® Engine API for Python® to call any MATLAB function on the MATLAB path. Some examples to call a few common matlab functions include:\\n\\nTo call the matlab function myFnc which has no arguments:\\n```\\nExecute: eng.myFnc()\\n```\\n\\nTo use the is prime function which requires an input argument:\\n```\\nExecute: tf = eng.isprime(37)\\n```\\n\\nTo determine the greatest common denominator of two numbers, use the gcd function. Set nargout to return the three output arguments from gcd.\\n```\\nExecute: t = eng.gcd(100.0,80.0,nargout=3)\\n```\\n\\nQuery:{input}\\n\\n**INSTRUCTIONS**\\n1. Given the user query and the documentation, identify each of the arguments found in the users query that should be passed to the matlab function.\\n2. Using the matlab enging eng, provide the one line of code to execute the desired matlab commands. Assume all functions are added to the path and eng already exist.\\n3. The last line of your response should say \"Execute: <MATLAB codes to execute>\"\\n4. Format the response/output as:\\nArguments: \\nMATLAB Code Explanation: <2 sentences maximum>\\nExecute: <your code here>\\n\\n**IMPORTANT**\\nThe code to execute from your response must be formatted as:\\n    Execute: eng.<function name>(<arguments>)>\\nThis output should be exactly one line and no longer. Stop the response after this line.\\n'\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation:\n",
      "Human: design an analysis pipeline to find the first order interactions of MYOD in the hardwired genome. Write a small report about which pathways and cell types these genes are implicated in\n",
      "BRAD:  **Step 1 (method: RAG)**:\n",
      "Prompt: Retrieve information about first order interactions of MYOD from literature and documents in various databases. This may include interaction databases and papers that discuss such interactions.\n",
      "\n",
      "**Step 2 (method: CODE)**:\n",
      "Prompt: Execute a bioinformatics pipeline to identify the genes directly interacting with MYOD from the retrieved data. This may involve using tools such as STRING for protein-protein interaction analysis or other relevant pipelines.\n",
      "\n",
      "**Step 3 (method: DATABASE)**:\n",
      "Prompt: Utilize Gene Ontology and Enrichr databases to perform gene set enrichment analyses on the identified genes. This will help to determine which specific cell types and pathways these genes are implicated in.\n",
      "\n",
      "**Step 4 (method: SCRAPE)**:\n",
      "Prompt: Search platforms like arXiv, bioRxiv, and PubMed for the latest research on these gene interactions and their association with specific cell types and pathways. This information will be used to enhance and validate the findings from the previous database analysis.\n",
      "\n",
      "**Step 5 (method: CODE)**:\n",
      "Prompt: Dry run existing bioinformatics pipelines or develop new ones, as needed, to analyze the association of the identified genes with specific cell types and pathways using the data from the previous steps.\n",
      "\n",
      "**Step 6 (method: DATABASE)**:\n",
      "Prompt: Utilize multiple databases, such as PathwayCommondb and Reactome, to determine which cellular pathways these interacting genes are involved in.\n",
      "\n",
      "**Step 7 (method: WRITE)**:\n",
      "Prompt: Summarize the findings from each step into a final report. This should include an overview of the methods used, the stress-specific genes identified and their interactions, and any cell types or pathways implicated in the first order interactions of MYOD.\n",
      "Human: Execute a bioinformatics pipeline to identify the genes directly interacting with MYOD from the retrieved data. This may involve using tools such as STRING for protein-protein interaction analysis or other relevant pipelines.\n",
      "BRAD:  Arguments: \"fileName\" (name of the data file containing the interactions of MYOD with other genes)\n",
      "MATLAB Code Explanation: Reads the input data file and applies bioinformatics pipeline for identification of genes interacting with MYOD using STRING or other relevant pipelines.\n",
      "Execute: eng.qualityControl(\"MYOD\\_interactions.txt\"); % Assuming the input file is named \"MYOD\\_interactions.txt\" and is in the current working directory. Replace \"qualityControl\" function with the appropriate function name for protein-protein interaction analysis.\n",
      "Human: Utilize Gene Ontology and Enrichr databases to perform gene set enrichment analyses on the identified genes. This will help to determine which specific cell types and pathways these genes are implicated in.\n",
      "BRAD:  database: ENRICHR\n",
      "genes: None\n",
      "code: None\n",
      "table: None\n",
      "mode name: None\n",
      "mode: None\n",
      "\n",
      "Explanation: The first step in the query asks to identify genes that interact directly with MYOD. To accomplish this task, a bioinformatics pipeline will be employed. No specific genes are mentioned in the query, so no modifications are needed. There is no need to extract genes from a table as the interaction data for MYOD should be provided in the input data file. Thus, the genes, code, table, and mode name/mode are all set to None. The database to be used is ENRICHR, as it is mentioned in the query for performing gene set enrichment analysis.\n",
      "Human: Search platforms like arXiv, bioRxiv, and PubMed for the latest research on these gene interactions and their association with specific cell types and pathways. This information will be used to enhance and validate the findings from the previous database analysis.\n",
      "BRAD:  Database: PubMed\n",
      "Search Terms: \"MYOD protein-protein interactions, MYOD genetic interactions, MYOD cellular pathways, MYOD gene expression, MYOD cell types, MYOD gene regulation, MYOD signaling pathways, MYOD molecular mechanism, MYOD chemical interactions, MYOD disease associations\"\n",
      "\n",
      "**MATLAB SCRIPT**\n",
      "You must run this python script: \n",
      "qualityControl.m\n",
      "\n",
      "**MATLAB FUNCTION DOCUMENTATION**:\n",
      "function output = qualityControl(fileName)\n",
      "% qualityControl - Constructs and runs a bioinformatics quality control pipeline.\n",
      "%\n",
      "% BRAD's Example to call this function:\n",
      "%    Execute: eng.qualityControl(\"SRR005164_1_50.fastq\");\n",
      "%\n",
      "% This function creates a bioinformatics quality control pipeline using the\n",
      "% MATLAB bioinformatics toolbox. It reads a FASTQ file, applies sequence\n",
      "% filtering, and generates a quality control plot.\n",
      "%\n",
      "% The pipeline consists of the following steps:\n",
      "% 1. FileChooser block to select the input FASTQ file.\n",
      "% 2. SeqFilter block to filter the sequences based on quality thresholds.\n",
      "% 3. UserFunction block to generate a quality control plot of the filtered sequences.\n",
      "%\n",
      "% The blocks are connected in sequence and executed in the specified order.\n",
      "%\n",
      "% Reference: https://www.mathworks.com/help/bioinfo/ref/bioinfo.pipeline.pipeline.run.html\n",
      "%\n",
      "% Auth: Joshua Pickard\n",
      "%       jpic@umich.edu\n",
      "% Date: June 14, 2024\n",
      "\n",
      "**CALL MATLAB FUNCTIONS FROM PYTHON**:\n",
      "Use MATLAB® Engine API for Python® to call any MATLAB function on the MATLAB path. Some examples to call a few common matlab functions include:\n",
      "\n",
      "To call the matlab function myFnc which has no arguments:\n",
      "```\n",
      "Execute: eng.myFnc()\n",
      "```\n",
      "\n",
      "To use the is prime function which requires an input argument:\n",
      "```\n",
      "Execute: tf = eng.isprime(37)\n",
      "```\n",
      "\n",
      "To determine the greatest common denominator of two numbers, use the gcd function. Set nargout to return the three output arguments from gcd.\n",
      "```\n",
      "Execute: t = eng.gcd(100.0,80.0,nargout=3)\n",
      "```\n",
      "\n",
      "Query:Dry run existing bioinformatics pipelines or develop new ones, as needed, to analyze the association of the identified genes with specific cell types and pathways using the data from the previous steps.\n",
      "\n",
      "**INSTRUCTIONS**\n",
      "1. Given the user query and the documentation, identify each of the arguments found in the users query that should be passed to the matlab function.\n",
      "2. Using the matlab enging eng, provide the one line of code to execute the desired matlab commands. Assume all functions are added to the path and eng already exist.\n",
      "3. The last line of your response should say \"Execute: <MATLAB codes to execute>\"\n",
      "4. Format the response/output as:\n",
      "Arguments: \n",
      "MATLAB Code Explanation: <2 sentences maximum>\n",
      "Execute: <your code here>\n",
      "\n",
      "**IMPORTANT**\n",
      "The code to execute from your response must be formatted as:\n",
      "    Execute: eng.<function name>(<arguments>)>\n",
      "This output should be exactly one line and no longer. Stop the response after this line.\n",
      "\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sun 16 Jun 2024 11:15:04 PM EDT INFO DATABASE\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "LLM OUTPUT PARSER\n",
      " Arguments: \"fileName\" (name of the input file for the bioinformatics pipeline)\n",
      "MATLAB Code Explanation: Reads the input file and performs bioinformatics analysis on the identified genes and their association with specific cell types and pathways.\n",
      "Execute: eng.pipeline(\"MYOD\\_bioinformatics\\_pipeline.m\",\"-InputFile\",fileName)\n",
      "\n",
      "**Note:** Replace \"MYOD\\_bioinformatics\\_pipeline.m\" with the actual name of your MATLAB pipeline script.\n",
      "[' Arguments: \"fileName\" (name of the input file for the bioinformatics pipeline)\\nMATLAB Code Explanation: Reads the input file and performs bioinformatics analysis on the identified genes and their association with specific cell types and pathways.\\n', ' eng.pipeline(\"MYOD\\\\_bioinformatics\\\\_pipeline.m\",\"-InputFile\",fileName)\\n\\n**Note:** Replace \"MYOD\\\\_bioinformatics\\\\_pipeline.m\" with the actual name of your MATLAB pipeline script.']\n",
      "Executing Code!!\n",
      "EVAL\n",
      "eng.pipeline(\"MYOD\\_bioinformatics\\_pipeline.m\",\"-InputFile\",fileName)\n",
      "\n",
      "**Note:** Replace \"MYOD\\_bioinformatics\\_pipeline.m\" with the actual name of your MATLAB pipeline script.\n",
      "END EVAL CHECK\n",
      "Debug: Syntax error in the MATLAB code: invalid syntax (<string>, line 3)\n",
      "==================================================\n",
      "==================================================\n",
      "RAG >> 7: \n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation:\n",
      "Human: design an analysis pipeline to find the first order interactions of MYOD in the hardwired genome. Write a small report about which pathways and cell types these genes are implicated in\n",
      "BRAD:  **Step 1 (method: RAG)**:\n",
      "Prompt: Retrieve information about first order interactions of MYOD from literature and documents in various databases. This may include interaction databases and papers that discuss such interactions.\n",
      "\n",
      "**Step 2 (method: CODE)**:\n",
      "Prompt: Execute a bioinformatics pipeline to identify the genes directly interacting with MYOD from the retrieved data. This may involve using tools such as STRING for protein-protein interaction analysis or other relevant pipelines.\n",
      "\n",
      "**Step 3 (method: DATABASE)**:\n",
      "Prompt: Utilize Gene Ontology and Enrichr databases to perform gene set enrichment analyses on the identified genes. This will help to determine which specific cell types and pathways these genes are implicated in.\n",
      "\n",
      "**Step 4 (method: SCRAPE)**:\n",
      "Prompt: Search platforms like arXiv, bioRxiv, and PubMed for the latest research on these gene interactions and their association with specific cell types and pathways. This information will be used to enhance and validate the findings from the previous database analysis.\n",
      "\n",
      "**Step 5 (method: CODE)**:\n",
      "Prompt: Dry run existing bioinformatics pipelines or develop new ones, as needed, to analyze the association of the identified genes with specific cell types and pathways using the data from the previous steps.\n",
      "\n",
      "**Step 6 (method: DATABASE)**:\n",
      "Prompt: Utilize multiple databases, such as PathwayCommondb and Reactome, to determine which cellular pathways these interacting genes are involved in.\n",
      "\n",
      "**Step 7 (method: WRITE)**:\n",
      "Prompt: Summarize the findings from each step into a final report. This should include an overview of the methods used, the stress-specific genes identified and their interactions, and any cell types or pathways implicated in the first order interactions of MYOD.\n",
      "Human: Execute a bioinformatics pipeline to identify the genes directly interacting with MYOD from the retrieved data. This may involve using tools such as STRING for protein-protein interaction analysis or other relevant pipelines.\n",
      "BRAD:  Arguments: \"fileName\" (name of the data file containing the interactions of MYOD with other genes)\n",
      "MATLAB Code Explanation: Reads the input data file and applies bioinformatics pipeline for identification of genes interacting with MYOD using STRING or other relevant pipelines.\n",
      "Execute: eng.qualityControl(\"MYOD\\_interactions.txt\"); % Assuming the input file is named \"MYOD\\_interactions.txt\" and is in the current working directory. Replace \"qualityControl\" function with the appropriate function name for protein-protein interaction analysis.\n",
      "Human: Utilize Gene Ontology and Enrichr databases to perform gene set enrichment analyses on the identified genes. This will help to determine which specific cell types and pathways these genes are implicated in.\n",
      "BRAD:  database: ENRICHR\n",
      "genes: None\n",
      "code: None\n",
      "table: None\n",
      "mode name: None\n",
      "mode: None\n",
      "\n",
      "Explanation: The first step in the query asks to identify genes that interact directly with MYOD. To accomplish this task, a bioinformatics pipeline will be employed. No specific genes are mentioned in the query, so no modifications are needed. There is no need to extract genes from a table as the interaction data for MYOD should be provided in the input data file. Thus, the genes, code, table, and mode name/mode are all set to None. The database to be used is ENRICHR, as it is mentioned in the query for performing gene set enrichment analysis.\n",
      "Human: Search platforms like arXiv, bioRxiv, and PubMed for the latest research on these gene interactions and their association with specific cell types and pathways. This information will be used to enhance and validate the findings from the previous database analysis.\n",
      "BRAD:  Database: PubMed\n",
      "Search Terms: \"MYOD protein-protein interactions, MYOD genetic interactions, MYOD cellular pathways, MYOD gene expression, MYOD cell types, MYOD gene regulation, MYOD signaling pathways, MYOD molecular mechanism, MYOD chemical interactions, MYOD disease associations\"\n",
      "Human: Dry run existing bioinformatics pipelines or develop new ones, as needed, to analyze the association of the identified genes with specific cell types and pathways using the data from the previous steps.\n",
      "BRAD:  Arguments: \"fileName\" (name of the input file for the bioinformatics pipeline)\n",
      "MATLAB Code Explanation: Reads the input file and performs bioinformatics analysis on the identified genes and their association with specific cell types and pathways.\n",
      "Execute: eng.pipeline(\"MYOD\\_bioinformatics\\_pipeline.m\",\"-InputFile\",fileName)\n",
      "\n",
      "**Note:** Replace \"MYOD\\_bioinformatics\\_pipeline.m\" with the actual name of your MATLAB pipeline script.\n",
      "    \n",
      "    GENEONTOLOGY: The Gene Ontology (GO) is an initiative to unify the representation of gene and gene product attributes across all species via the aims: 1) maintain and develop its controlled vocabulary of gene and gene product attributes; 2) annotate genes and gene products, and assimilate and disseminate annotation data; and 3) provide tools for easy access to all aspects of the data provided by the project, and to enable functional interpretation of experimental data using the GO.\n",
      "    \n",
      "    ENRICHR: is a tool used to lookup sets of genes and their functional association. ENRICHR has access to many gene-set libraries including Allen_Brain_Atlas_up, ENCODE_Histone_Modifications_2015, Enrichr_Libraries_Most_Popular_Genes, FANTOM6_lncRNA_KD_DEGs, GO_Biological_Process_2023, GTEx, Human_Gene_Atlas, KEGG, REACTOME, Transcription_Factor_PPIs, WikiPathways and many others databases.\n",
      "    \n",
      "    Available Tables:\n",
      "    \n",
      "    \n",
      "    Query:Utilize multiple databases, such as PathwayCommondb and Reactome, to determine which cellular pathways these interacting genes are involved in.\n",
      "    \n",
      "    **INSTRUCTIONS**\n",
      "    1. From the query, decide if GENEONTOLOGY or ENRICHR should be used\n",
      "    2. Identify genes in the users input that should be searched. propose genes with similar names, correct the users spelling, or make small modifications to the list, but do not propose genes that are not found in the humans input.\n",
      "    3. If the user wants to extract genes from a table (python dataframe), provide the necessary code to get the genes into a python list, otherwise, say None.\n",
      "    4. If there is code required, identify the name of the table, otherwise say None\n",
      "    5. If there is code required, identify the row or column name of the table, otherwise say None\n",
      "    6. If there is code required, specify if it is a row or column\n",
      "    Format your output as follows with no additional information:\n",
      "    \n",
      "    database: <ENRICHR or GENEONTOLOGY>\n",
      "    genes: <List of genes separated by commas in query or None if code is required>\n",
      "    code: <True or None>\n",
      "    table: <Table Name>\n",
      "    mode name: <Row or Column Name>\n",
      "    mode: <Row or Column>\n",
      "    \u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sun 16 Jun 2024 11:15:06 PM EDT INFO Performing Enichr analysis using database GO_Biological_Process_2021.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      " database: ENRICHR\n",
      "genes: None\n",
      "code: None\n",
      "table: None\n",
      "mode name: None\n",
      "mode: None\n",
      "\n",
      "Explanation: In this step, the goal is to utilize multiple databases, such as PathwayCommondb and Reactome, to determine which cellular pathways these interacting genes are involved in. As the query specifically mentions using ENRICHR for gene set enrichment analysis, and it also supports pathway analysis, ENRICHR will be the database to be used. No specific genes are mentioned in the query, so no modifications to the gene list are required. No code is needed to extract genes from a table, as the focus is on using databases for analysis, not extracting data from a table for input.\n",
      " database: ENRICHR\n",
      "genes: None\n",
      "code: None\n",
      "table: None\n",
      "mode name: None\n",
      "mode: None\n",
      "\n",
      "Explanation: In this step, the goal is to utilize multiple databases, such as PathwayCommondb and Reactome, to determine which cellular pathways these interacting genes are involved in. As the query specifically mentions using ENRICHR for gene set enrichment analysis, and it also supports pathway analysis, ENRICHR will be the database to be used. No specific genes are mentioned in the query, so no modifications to the gene list are required. No code is needed to extract genes from a table, as the focus is on using databases for analysis, not extracting data from a table for input.\n",
      "None\n",
      "{'database': 'ENRICHR', 'genes': ['None'], 'code': 'None'}\n",
      "['None']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sun 16 Jun 2024 11:15:06 PM EDT ERROR \n",
      "            No Enrichr results were found for genes ['None'] and database GO_Biological_Process_2021. \n",
      "\n",
      "            If the genes are Ensembl IDs, please set argument 'ensembl=True' (for terminal, add flag: [--ensembl]).\n",
      "            \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The following table was generated by quering the gene list against GO_Biological_Process_2021:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "</style>\n",
       "<table id=\"T_5c55d\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th id=\"T_5c55d_level0_col0\" class=\"col_heading level0 col0\" >rank</th>\n",
       "      <th id=\"T_5c55d_level0_col1\" class=\"col_heading level0 col1\" >path_name</th>\n",
       "      <th id=\"T_5c55d_level0_col2\" class=\"col_heading level0 col2\" >p_val</th>\n",
       "      <th id=\"T_5c55d_level0_col3\" class=\"col_heading level0 col3\" >z_score</th>\n",
       "      <th id=\"T_5c55d_level0_col4\" class=\"col_heading level0 col4\" >combined_score</th>\n",
       "      <th id=\"T_5c55d_level0_col5\" class=\"col_heading level0 col5\" >overlapping_genes</th>\n",
       "      <th id=\"T_5c55d_level0_col6\" class=\"col_heading level0 col6\" >adj_p_val</th>\n",
       "      <th id=\"T_5c55d_level0_col7\" class=\"col_heading level0 col7\" >database</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x1506ac53d490>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The table has been saved to: RAG-gget-GO_Biological_Process_2021-2024-06-1623:15:06.547314.csv\n",
      "==================================================\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'WRITE'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[5], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mbrad\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mchat\u001b[49m\u001b[43m(\u001b[49m\u001b[43mragvectordb\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mvectordb\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/RAG-DEV/BRAD/brad.py:445\u001b[0m, in \u001b[0;36mchat\u001b[0;34m(model_path, persist_directory, llm, ragvectordb, embeddings_model)\u001b[0m\n\u001b[1;32m    443\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    444\u001b[0m     route \u001b[38;5;241m=\u001b[39m chatstatus[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mprompt\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39msplit(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m \u001b[39m\u001b[38;5;124m'\u001b[39m)[\u001b[38;5;241m1\u001b[39m]            \u001b[38;5;66;03m# use the forced router\u001b[39;00m\n\u001b[0;32m--> 445\u001b[0m     \u001b[43mbuildRoutes\u001b[49m\u001b[43m(\u001b[49m\u001b[43mchatstatus\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mprompt\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    446\u001b[0m     chatstatus[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mprompt\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m \u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mjoin(chatstatus[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mprompt\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39msplit(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m \u001b[39m\u001b[38;5;124m'\u001b[39m)[\u001b[38;5;241m2\u001b[39m:])\u001b[38;5;241m.\u001b[39mstrip()\n\u001b[1;32m    448\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m==================================================\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "File \u001b[0;32m~/RAG-DEV/BRAD/router.py:140\u001b[0m, in \u001b[0;36mbuildRoutes\u001b[0;34m(prompt)\u001b[0m\n\u001b[1;32m    127\u001b[0m     i \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n\u001b[1;32m    128\u001b[0m paths \u001b[38;5;241m=\u001b[39m {\n\u001b[1;32m    129\u001b[0m     \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mDATABASE\u001b[39m\u001b[38;5;124m'\u001b[39m: getRouterPath(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124menrichr.txt\u001b[39m\u001b[38;5;124m'\u001b[39m),\n\u001b[1;32m    130\u001b[0m     \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mSCRAPE\u001b[39m\u001b[38;5;124m'\u001b[39m  : getRouterPath(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mscrape.txt\u001b[39m\u001b[38;5;124m'\u001b[39m),\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    138\u001b[0m     \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mCODE\u001b[39m\u001b[38;5;124m'\u001b[39m    : getRouterPath(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcode.txt\u001b[39m\u001b[38;5;124m'\u001b[39m),\n\u001b[1;32m    139\u001b[0m }\n\u001b[0;32m--> 140\u001b[0m filepath \u001b[38;5;241m=\u001b[39m \u001b[43mpaths\u001b[49m\u001b[43m[\u001b[49m\u001b[43mroute\u001b[49m\u001b[43m]\u001b[49m\n\u001b[1;32m    141\u001b[0m add_sentence(filepath, rebuiltPrompt)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'WRITE'"
     ]
    }
   ],
   "source": [
    "brad.chat(ragvectordb = vectordb)"
   ]
  },
  {
   "cell_type": "raw",
   "id": "97d6aa41-2a87-48d7-8a3d-d8a4cfa3d043",
   "metadata": {},
   "source": [
    "/force PLANNER design an analysis pipeline to find the first order interactions of MYOD in the hardwired genome. Write a small report about which pathways and cell types these genes are implicated in "
   ]
  },
  {
   "cell_type": "raw",
   "id": "233d47c7-8584-4a4d-a0f5-bb4e53cc7739",
   "metadata": {},
   "source": [
    "/force PLANNER design an analysis pipeline to find the first order interactions of MYOD in the hardwired genome from HWG.mat. Write a small report about which pathways and cell types these genes are implicated in "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d205ed67-2b4e-494d-b8cc-982aa33cdcc4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "73a585c4-e1a0-4264-98d5-9c61911e84b8",
   "metadata": {},
   "source": [
    "# Write a python script to use"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "21c3d231-14a7-49d2-8d88-2cf204a214cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import scipy.io as sio\n",
    "import numpy as np\n",
    "from scipy.sparse import csc_matrix\n",
    "\n",
    "geneName = 'ENSG00000129152'\n",
    "\n",
    "HWG = sio.loadmat('/home/jpic/RAG-DEV/tutorials/planner/HWG.mat')            # Load Hardwired Genome A matrix\n",
    "A = HWG['HWG']['A']\n",
    "A = A[0,0].todense()\n",
    "df = pd.read_csv('/home/jpic/RAG-DEV/tutorials/planner/geneIndexTable.csv')  # Load gene index table\n",
    "df2 = pd.read_csv('GRCh38.p13.tsv', delimiter='\\t')                          # Load GRCH38 to map between gene names and ENSG IDs\n",
    "if df[df['Stable ID'].isin([geneName])].shape[0] == 0:\n",
    "    temp0 = df2[df2['Gene name'] == geneName]                                    # Find ENSG of Gene Name\n",
    "    ENSG1 = temp0['Gene stable ID'].values[0]\n",
    "else:\n",
    "    ENSG1 = geneName\n",
    "idx = df[df['Stable ID'] == ENSG1].index[0]                                  # Find position of target gene in HWG\n",
    "idxs = np.where(A[idx,:] == 1)[1]                                            # Find first order interactions in network\n",
    "firstOrderEnsg = list(df.iloc[idxs]['Stable ID'].values)                     # Get ENSG of first order interactions\n",
    "firstOrderGenes = list(set(df2[df2['Gene stable ID'].isin(firstOrderEnsg)]['Gene name'].values)) # Get gene names of first order contacts\n",
    "dfSave = pd.DataFrame({'Genes':firstOrderGenes})\n",
    "dfSave.to_csv('HWG_first_order_interactions.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "dcdfbafb-5bc9-4664-9e44-9c77b4c8084f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['RORA',\n",
       " 'CDH2',\n",
       " 'HDAC8',\n",
       " 'MSGN1',\n",
       " 'EBF4',\n",
       " 'TUBA8',\n",
       " 'ADCK1',\n",
       " 'SOX11',\n",
       " 'OXCT1',\n",
       " 'TERT',\n",
       " 'FGF3',\n",
       " 'VGLL2',\n",
       " 'RUNX1',\n",
       " 'IRX1',\n",
       " 'E2F5',\n",
       " 'FST',\n",
       " 'HOXC9',\n",
       " 'ROCK1',\n",
       " 'STAU1',\n",
       " 'SIM2',\n",
       " 'HSP90AB1',\n",
       " 'ZNF219',\n",
       " 'COL19A1',\n",
       " 'FRMD6',\n",
       " 'TCF15',\n",
       " 'LRP5',\n",
       " 'SMARCA2',\n",
       " 'SOX17',\n",
       " 'SGCG',\n",
       " 'CSRP3',\n",
       " 'SMTN',\n",
       " 'TCF7',\n",
       " 'ZNF143',\n",
       " 'CLOCK',\n",
       " 'EYA1',\n",
       " 'POU3F2',\n",
       " 'MKX',\n",
       " 'TNMD',\n",
       " 'CEBPG',\n",
       " 'SLC38A2',\n",
       " 'ANXA11',\n",
       " 'SIX4',\n",
       " 'GABARAPL1',\n",
       " 'BDNF',\n",
       " 'BMP5',\n",
       " 'YY2',\n",
       " 'PBX4',\n",
       " 'SLC8A3',\n",
       " 'RBL1',\n",
       " 'PVALB',\n",
       " 'MEF2B',\n",
       " 'HSPB3',\n",
       " 'HDAC10',\n",
       " 'KRT5',\n",
       " 'DBX1',\n",
       " 'MKI67',\n",
       " 'SALL4',\n",
       " 'YY1',\n",
       " 'TSPAN9',\n",
       " 'TBX1',\n",
       " 'JDP2',\n",
       " 'THBS1',\n",
       " 'ZBED4',\n",
       " 'MAPK12',\n",
       " 'CAPN3',\n",
       " 'JARID2',\n",
       " 'MYCN',\n",
       " 'NCOA3',\n",
       " 'SOX7',\n",
       " 'TBX18',\n",
       " 'NEO1',\n",
       " 'ARID3A',\n",
       " 'ACTC1',\n",
       " 'FOSL2',\n",
       " 'HOXC10',\n",
       " 'ATG5',\n",
       " 'LMNB1',\n",
       " 'CHGA',\n",
       " 'FOS',\n",
       " 'TFE3',\n",
       " 'INSR',\n",
       " 'BMP4',\n",
       " 'FOXO4',\n",
       " 'LMO7',\n",
       " 'SIX6',\n",
       " 'SETD3',\n",
       " 'TGFB3',\n",
       " 'SLC1A1',\n",
       " 'DACH1',\n",
       " 'FOXG1',\n",
       " 'DCD',\n",
       " 'SDC4',\n",
       " 'ANKRD2',\n",
       " 'PTF1A',\n",
       " 'GDF5',\n",
       " 'ACSS3',\n",
       " 'FBXO32',\n",
       " 'SULF1',\n",
       " 'UHRF1',\n",
       " 'TCF4',\n",
       " 'TPM4',\n",
       " 'MYO18B',\n",
       " 'PKNOX1',\n",
       " 'NKX3-1',\n",
       " 'PAX5',\n",
       " 'CCDC6',\n",
       " 'HEY2',\n",
       " 'EP300',\n",
       " 'PCDH8',\n",
       " 'CAT',\n",
       " 'CNTF',\n",
       " 'KLHL31',\n",
       " 'ZIC3',\n",
       " 'SMARCD1',\n",
       " 'H3-5',\n",
       " 'EYA2',\n",
       " 'CEBPA',\n",
       " 'MSX1',\n",
       " 'ALDH1A2',\n",
       " 'BMP6',\n",
       " 'ZNRF3',\n",
       " 'SNAI2',\n",
       " 'LARGE1',\n",
       " 'KPNA3',\n",
       " 'FOXA2',\n",
       " 'MAPK1',\n",
       " 'NANOG',\n",
       " 'TNNC2',\n",
       " 'SERGEF',\n",
       " 'WNT7B',\n",
       " 'HEY1',\n",
       " 'HDAC2',\n",
       " 'GDI1',\n",
       " 'DGCR8',\n",
       " 'TRAF7',\n",
       " 'RHOXF1',\n",
       " 'COL4A1',\n",
       " 'GDF7',\n",
       " 'CYP26A1',\n",
       " 'DEK',\n",
       " 'MMP9',\n",
       " 'SOX21',\n",
       " 'GJA1',\n",
       " 'NKX2-2',\n",
       " 'VGLL1',\n",
       " 'MZF1',\n",
       " 'MYH7',\n",
       " 'MGMT',\n",
       " 'KLF5',\n",
       " 'NKX3-2',\n",
       " 'SON',\n",
       " 'AR',\n",
       " 'CEBPD',\n",
       " 'KLF12',\n",
       " 'MYLK4',\n",
       " 'NFIC',\n",
       " 'HMGB3',\n",
       " 'PINX1',\n",
       " 'TRIM27',\n",
       " 'KLF8',\n",
       " 'COL4A2',\n",
       " 'SOX4',\n",
       " 'TNFSF11',\n",
       " 'PML',\n",
       " 'DTNA',\n",
       " 'FGF16',\n",
       " 'SS18',\n",
       " 'PCDH17',\n",
       " 'GATA6',\n",
       " 'S100B',\n",
       " 'TET1',\n",
       " 'AKAP4',\n",
       " 'MMUT',\n",
       " 'TSHZ3',\n",
       " 'MYLK2',\n",
       " 'SMAD7',\n",
       " 'ASH2L',\n",
       " 'SCD',\n",
       " 'FOXO1',\n",
       " 'CABIN1',\n",
       " 'NKX6-2',\n",
       " 'IRS2',\n",
       " 'FOXC1',\n",
       " 'MEGF10',\n",
       " 'ZIC2',\n",
       " 'LATS2',\n",
       " 'CSPG4',\n",
       " 'LDHC',\n",
       " 'JAG1',\n",
       " 'PMEPA1',\n",
       " 'FOXK1',\n",
       " 'DCT',\n",
       " 'AXIN1',\n",
       " 'TFAM',\n",
       " 'SMAD9',\n",
       " nan,\n",
       " 'HOXC8',\n",
       " 'TRDN',\n",
       " 'LDLRAD4',\n",
       " 'DIO3',\n",
       " 'FOXO3',\n",
       " 'SNAI1',\n",
       " 'ID4',\n",
       " 'HIRA',\n",
       " 'PPARGC1A',\n",
       " 'TIMP3',\n",
       " 'ETS2',\n",
       " 'MYOF',\n",
       " 'CCNA1',\n",
       " 'WNT8B',\n",
       " 'CHRNA4',\n",
       " 'SMAD2',\n",
       " 'FOXF2',\n",
       " 'BMP7',\n",
       " 'MEIS2',\n",
       " 'SMPX',\n",
       " 'POLA1',\n",
       " 'LRRC30',\n",
       " 'EBF2',\n",
       " 'NFATC1',\n",
       " 'GDF3',\n",
       " 'BMP2',\n",
       " 'ITGA5',\n",
       " 'CDKN2B',\n",
       " 'ITGB1BP2',\n",
       " 'KLF6',\n",
       " 'FKBP1A',\n",
       " 'DLL4',\n",
       " 'EIF4EBP1',\n",
       " 'SMARCB1',\n",
       " 'HSPH1',\n",
       " 'MBNL2',\n",
       " 'CNN2',\n",
       " 'MYL12B',\n",
       " 'PRMT5',\n",
       " 'BNIP2',\n",
       " 'ERG',\n",
       " 'SUV39H2',\n",
       " 'EEF1A2',\n",
       " 'TMOD1',\n",
       " 'RUNX2',\n",
       " 'PACSIN3',\n",
       " 'GATA5',\n",
       " 'EWSR1',\n",
       " 'ID1',\n",
       " 'DYRK1A',\n",
       " 'ANK3',\n",
       " 'GDF2',\n",
       " 'NEUROG1',\n",
       " 'MYL9',\n",
       " 'ZFAND5',\n",
       " 'IRX4',\n",
       " 'DKK1',\n",
       " 'LARP6',\n",
       " 'ALDH1A1',\n",
       " 'ACP1',\n",
       " 'ADD3',\n",
       " 'SMAD6',\n",
       " 'FOXB1',\n",
       " 'CKMT2',\n",
       " 'COL6A1',\n",
       " 'RIPPLY1',\n",
       " 'ETV1',\n",
       " 'SLC2A4RG',\n",
       " 'LPIN3',\n",
       " 'MYL12A',\n",
       " 'LPL',\n",
       " 'MCAT',\n",
       " 'CDX2',\n",
       " 'CIDEA',\n",
       " 'HOXC11',\n",
       " 'CCNT1',\n",
       " 'RBM24',\n",
       " 'MYOM1',\n",
       " 'TMTC1',\n",
       " 'LGR4',\n",
       " 'CTNNBL1',\n",
       " 'GDNF',\n",
       " 'MEOX2',\n",
       " 'E2F8',\n",
       " 'BCL2L1',\n",
       " 'FGF13',\n",
       " 'PITX1',\n",
       " 'FGF14',\n",
       " 'CDC37',\n",
       " 'LIF',\n",
       " 'FGF9',\n",
       " 'DAPK1',\n",
       " 'OLIG2',\n",
       " 'NFATC2',\n",
       " 'LIN28B',\n",
       " 'ELOVL2',\n",
       " 'PDX1',\n",
       " 'STUB1',\n",
       " 'POSTN',\n",
       " 'PRDM1',\n",
       " 'DNTT',\n",
       " 'SOX3',\n",
       " 'WT1',\n",
       " 'NRIP1',\n",
       " 'E2F1',\n",
       " 'SOD1',\n",
       " 'SIX1',\n",
       " 'IGFBP6',\n",
       " 'SOX1',\n",
       " 'MYBL2',\n",
       " 'APBA1',\n",
       " 'SMCHD1',\n",
       " 'ISL1',\n",
       " 'FGF10',\n",
       " 'NR2F1',\n",
       " 'SMAD4',\n",
       " 'TEAD4',\n",
       " 'RB1',\n",
       " 'MAPK11',\n",
       " 'UNC45A',\n",
       " 'CMYA5',\n",
       " 'ZDHHC8',\n",
       " 'TEF',\n",
       " 'TRAF6',\n",
       " 'XRCC6',\n",
       " 'THBS4',\n",
       " 'SNW1',\n",
       " 'UNCX',\n",
       " 'MYLIP',\n",
       " 'CDH17',\n",
       " 'TWSG1']"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "firstOrderGenes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "cb38694d-c5ce-4a81-80f3-a394eed3dbca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Stable ID</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Transcription Factor</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6001</th>\n",
       "      <td>ENSG00000129152</td>\n",
       "      <td>MUC20-OT1</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>17719571</td>\n",
       "      <td>17722136</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Stable ID  Gene Name  Transcription Factor  Chromosome     Start  \\\n",
       "6001  ENSG00000129152  MUC20-OT1                     1          11  17719571   \n",
       "\n",
       "           End  \n",
       "6001  17722136  "
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['Stable ID'] == 'ENSG00000129152']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "c8c87527-12b4-4771-a374-89361f438107",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ELP3',\n",
       " 'MRPL16',\n",
       " 'MKKS',\n",
       " 'MIPEP',\n",
       " 'EIF2AK4',\n",
       " 'E2F5',\n",
       " 'RTCB',\n",
       " 'PPCDC',\n",
       " 'SNRPB2',\n",
       " 'CCNB3',\n",
       " 'PSMB5',\n",
       " 'RPL3',\n",
       " 'ARHGAP4',\n",
       " 'POLI',\n",
       " 'GADD45G',\n",
       " 'RPS10-NUDT3',\n",
       " 'CDC45',\n",
       " 'SLC2A13',\n",
       " 'BRIX1',\n",
       " 'COL19A1',\n",
       " 'SMARCA2',\n",
       " 'DSTN',\n",
       " 'SEC11A',\n",
       " 'GNPNAT1',\n",
       " 'ASMTL',\n",
       " 'SNRPD3',\n",
       " 'SRSF3',\n",
       " 'NUS1',\n",
       " 'DPM1',\n",
       " 'DDX39A',\n",
       " 'SMC1B',\n",
       " 'AARS2',\n",
       " 'NUP155',\n",
       " 'CCNO',\n",
       " 'CRLS1',\n",
       " 'POLD3',\n",
       " 'EIF3M',\n",
       " 'UBE2D1',\n",
       " 'ANP32A',\n",
       " 'GART',\n",
       " 'PMS2',\n",
       " 'FAM83D',\n",
       " 'TPX2',\n",
       " 'HMGA1',\n",
       " 'DHX35',\n",
       " 'ATRX',\n",
       " 'RBL1',\n",
       " 'EIF2S2',\n",
       " 'N6AMT1',\n",
       " 'TXNDC5',\n",
       " 'NUBPL',\n",
       " 'PRIM2',\n",
       " 'RPS7',\n",
       " 'DHX32',\n",
       " 'TRUB1',\n",
       " 'PRMT8',\n",
       " 'SLC27A5',\n",
       " 'PIAS2',\n",
       " 'MKI67',\n",
       " 'SMC6',\n",
       " 'GRPEL1',\n",
       " 'ZNF367',\n",
       " 'CENPQ',\n",
       " 'MIS12',\n",
       " 'KIN',\n",
       " 'TRDMT1',\n",
       " 'METTL15',\n",
       " 'TFDP1',\n",
       " 'TYMS',\n",
       " 'CDC25B',\n",
       " 'WDHD1',\n",
       " 'PSMG2',\n",
       " 'UBL7',\n",
       " 'MRPS28',\n",
       " 'BORA',\n",
       " 'TRMU',\n",
       " 'REV3L',\n",
       " 'CDCA2',\n",
       " 'TXN',\n",
       " 'TIMELESS',\n",
       " 'DDX11',\n",
       " 'AFG3L2',\n",
       " 'PIMREG',\n",
       " 'LMNB1',\n",
       " 'RBMX',\n",
       " 'APEX2',\n",
       " 'PDSS2',\n",
       " 'HOGA1',\n",
       " 'MTG2',\n",
       " 'UBE2C',\n",
       " 'SETBP1',\n",
       " 'IAH1',\n",
       " 'DHX33',\n",
       " 'PDXK',\n",
       " 'CCT8',\n",
       " 'RNASEH2B',\n",
       " 'ZFAND4',\n",
       " 'AAR2',\n",
       " 'OSGEP',\n",
       " 'MCM5',\n",
       " 'PRDX3',\n",
       " 'ARMC1',\n",
       " 'ITM2B',\n",
       " 'PSMC6',\n",
       " 'MTRF1',\n",
       " 'ZWINT',\n",
       " 'TGDS',\n",
       " 'PI4KA',\n",
       " 'MPI',\n",
       " 'UBQLN2',\n",
       " 'APEX1',\n",
       " 'CUL4A',\n",
       " 'PNMA6A',\n",
       " 'CDCA4',\n",
       " 'ATP5F1C',\n",
       " 'SNRPD1',\n",
       " 'ANP32B',\n",
       " 'UHRF1',\n",
       " 'PBK',\n",
       " 'DIAPH3',\n",
       " 'CHD6',\n",
       " 'COL4A5',\n",
       " 'CMTM5',\n",
       " 'BABAM1',\n",
       " 'BCAT1',\n",
       " 'RACGAP1',\n",
       " 'CCNB2',\n",
       " 'EP300',\n",
       " 'COL21A1',\n",
       " 'IPO5',\n",
       " 'CCAR1',\n",
       " 'NUDT12',\n",
       " 'H3-5',\n",
       " 'POGLUT2',\n",
       " 'CKAP2',\n",
       " 'DEPDC1B',\n",
       " 'TESMIN',\n",
       " 'UBE2J1',\n",
       " 'MTFR2',\n",
       " 'ESCO1',\n",
       " 'UROS',\n",
       " 'RPUSD2',\n",
       " 'SUPT16H',\n",
       " 'PRPF19',\n",
       " 'YWHAQ',\n",
       " 'PDS5A',\n",
       " 'KPNA3',\n",
       " 'YWHAB',\n",
       " 'MRPS6',\n",
       " 'RPL17-C18orf32',\n",
       " 'LRR1',\n",
       " 'NANOG',\n",
       " 'MCIDAS',\n",
       " 'HDAC2',\n",
       " 'AIRE',\n",
       " 'MCM3',\n",
       " 'AGAP9',\n",
       " 'L3MBTL1',\n",
       " 'COL4A1',\n",
       " 'WDR4',\n",
       " 'HTRA1',\n",
       " 'SETDB2',\n",
       " 'NONO',\n",
       " 'ITM2A',\n",
       " 'SPAG6',\n",
       " 'DEK',\n",
       " 'CUTC',\n",
       " 'TSPYL5',\n",
       " 'RANBP1',\n",
       " 'RRAS2',\n",
       " 'MAPRE1',\n",
       " 'MGMT',\n",
       " 'CSNK1E',\n",
       " 'PGAM1',\n",
       " 'CCDC33',\n",
       " 'RPS4X',\n",
       " 'CCNK',\n",
       " 'PRPS2',\n",
       " 'PFKL',\n",
       " 'PRMT3',\n",
       " 'ERCC6L',\n",
       " 'PSMG1',\n",
       " 'HMGB3',\n",
       " 'TSPEAR',\n",
       " 'SMC1A',\n",
       " 'COL4A2',\n",
       " 'CNOT7',\n",
       " 'COL17A1',\n",
       " 'POLL',\n",
       " 'COL6A2',\n",
       " 'RAD50',\n",
       " 'EEF1G',\n",
       " 'MRPL13',\n",
       " 'DSCC1',\n",
       " 'INTS13',\n",
       " 'CSE1L',\n",
       " 'PSAT1',\n",
       " 'SKA1',\n",
       " 'TXNRD2',\n",
       " 'RPAP3',\n",
       " 'MIS18A',\n",
       " 'SUGT1',\n",
       " 'MT-CYB',\n",
       " 'NNT',\n",
       " 'CENPM',\n",
       " 'ASB8',\n",
       " 'SIVA1',\n",
       " 'UBQLN1',\n",
       " 'AHCY',\n",
       " 'CDC123',\n",
       " 'CDKAL1',\n",
       " 'CCNH',\n",
       " 'PRTFDC1',\n",
       " 'SUB1',\n",
       " 'LDHC',\n",
       " 'VIRMA',\n",
       " 'CPSF3',\n",
       " 'CA2',\n",
       " 'CSNK1A1L',\n",
       " 'RTF2',\n",
       " 'UBR7',\n",
       " 'MPHOSPH8',\n",
       " 'SUMO4',\n",
       " 'HSCB',\n",
       " 'AURKA',\n",
       " 'TFAM',\n",
       " 'APMAP',\n",
       " nan,\n",
       " 'MOCS3',\n",
       " 'MOCS1',\n",
       " 'RPL36A',\n",
       " 'DNA2',\n",
       " 'PSMD10',\n",
       " 'PFDN4',\n",
       " 'RIOK2',\n",
       " 'PAPOLG',\n",
       " 'SCML2',\n",
       " 'ATAD2B',\n",
       " 'FOXO3',\n",
       " 'NUDT15',\n",
       " 'RPA3',\n",
       " 'HIRA',\n",
       " 'MSH3',\n",
       " 'ENO4',\n",
       " 'CCNA1',\n",
       " 'SHMT2',\n",
       " 'ADAT2',\n",
       " 'RBBP7',\n",
       " 'CDADC1',\n",
       " 'RNF8',\n",
       " 'BCCIP',\n",
       " 'MCM8',\n",
       " 'POLA1',\n",
       " 'NDC80',\n",
       " 'HACD3',\n",
       " 'DSN1',\n",
       " 'POLE2',\n",
       " 'EXOSC8',\n",
       " 'UBL4A',\n",
       " 'CAPRIN1',\n",
       " 'CDKN2B',\n",
       " 'WRN',\n",
       " 'TOP1',\n",
       " 'UBR5',\n",
       " 'DHTKD1',\n",
       " 'HSPH1',\n",
       " 'SNRPC',\n",
       " 'MAT1A',\n",
       " 'NOP56',\n",
       " 'PRMT5',\n",
       " 'PSMA7',\n",
       " 'SUV39H2',\n",
       " 'NUDT5',\n",
       " 'MBD1',\n",
       " 'MAGOHB',\n",
       " 'CCNY',\n",
       " 'PACSIN3',\n",
       " 'DONSON',\n",
       " 'RFC3',\n",
       " 'ACP2',\n",
       " 'RPL36AL',\n",
       " 'MAP1LC3A',\n",
       " 'ANG',\n",
       " 'RIDA',\n",
       " 'ANK3',\n",
       " 'CTPS2',\n",
       " 'C1QBP',\n",
       " 'PHF11',\n",
       " 'PSMA4',\n",
       " 'INCENP',\n",
       " 'RPP38',\n",
       " 'HSPA12B',\n",
       " 'CENPA',\n",
       " 'PRPS1L1',\n",
       " 'COX5A',\n",
       " 'POLR2F',\n",
       " 'RPL36A-HNRNPH2',\n",
       " 'EIF2AK1',\n",
       " 'USP18',\n",
       " 'CSTF1',\n",
       " 'RAD51',\n",
       " 'RBX1',\n",
       " 'XPNPEP1',\n",
       " 'COL9A3',\n",
       " 'SLF1',\n",
       " 'CSNK2A1',\n",
       " 'MAGED1',\n",
       " 'TTK',\n",
       " 'VBP1',\n",
       " 'MLH3',\n",
       " 'APC',\n",
       " 'H2AB3',\n",
       " 'COL6A1',\n",
       " 'TEX30',\n",
       " 'RAD1',\n",
       " 'RPA4',\n",
       " 'RAD21L1',\n",
       " 'ESCO2',\n",
       " 'ERCC5',\n",
       " 'GTF2A2',\n",
       " 'HAUS6',\n",
       " 'MX1',\n",
       " 'OXA1L',\n",
       " 'HSPA13',\n",
       " 'GTPBP3',\n",
       " 'MCAT',\n",
       " 'DHX15',\n",
       " 'HDDC2',\n",
       " 'ING1',\n",
       " 'POLR2E',\n",
       " 'RANBP6',\n",
       " 'PFKM',\n",
       " 'RCN2',\n",
       " 'GEN1',\n",
       " 'CCNT1',\n",
       " 'AK7',\n",
       " 'RAD54B',\n",
       " 'PCNA',\n",
       " 'ASPG',\n",
       " 'UHRF2',\n",
       " 'BCL2L1',\n",
       " 'E2F8',\n",
       " 'BUB3',\n",
       " 'LIG4',\n",
       " 'RANGAP1',\n",
       " 'CKS2',\n",
       " 'PAPOLB',\n",
       " 'FTSJ1',\n",
       " 'MTERF3',\n",
       " 'SKA3',\n",
       " 'SEC61A2',\n",
       " 'RBBP8',\n",
       " 'CENPK',\n",
       " 'MRPL15',\n",
       " 'SUMO3',\n",
       " 'CCNQ',\n",
       " 'SEPHS1',\n",
       " 'RAD23B',\n",
       " 'RAB9B',\n",
       " 'RAB33A',\n",
       " 'PRPS1',\n",
       " 'SLC27A2',\n",
       " 'DNTT',\n",
       " 'OAT',\n",
       " 'MGME1',\n",
       " 'E2F1',\n",
       " 'SLC6A8',\n",
       " 'EEF1E1',\n",
       " 'CCT5',\n",
       " 'AP5Z1',\n",
       " 'SPO11',\n",
       " 'HMGN1',\n",
       " 'RBPJL',\n",
       " 'MSH2',\n",
       " 'MYBL2',\n",
       " 'CDC5L',\n",
       " 'MNAT1',\n",
       " 'GPD1',\n",
       " 'BRCA2',\n",
       " 'MX2',\n",
       " 'POLN',\n",
       " 'BMS1',\n",
       " 'RB1',\n",
       " 'TSPYL4',\n",
       " 'ADSL',\n",
       " 'THNSL1',\n",
       " 'TFDP3',\n",
       " 'RNPS1',\n",
       " 'SNX5',\n",
       " 'NUP50',\n",
       " 'PHF21B',\n",
       " 'HMGB1',\n",
       " 'MAFB',\n",
       " 'CHEK2',\n",
       " 'TRMT11',\n",
       " 'PDSS1',\n",
       " 'XRCC6',\n",
       " 'AK3',\n",
       " 'WARS1',\n",
       " 'PDS5B',\n",
       " 'SNU13',\n",
       " 'XRCC3',\n",
       " 'KPNA5',\n",
       " 'UFD1',\n",
       " 'CHAF1B',\n",
       " 'CARS2',\n",
       " 'LDHAL6B',\n",
       " 'AIMP2',\n",
       " 'RPS12',\n",
       " 'PIK3C3',\n",
       " 'PGK2',\n",
       " 'MRPL39',\n",
       " 'MT-ATP6',\n",
       " 'PRMT2']"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "firstOrderGenes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "0b95bf43-8c07-4a96-bbcf-608695b5f85c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "414\n"
     ]
    }
   ],
   "source": [
    "firstOrderGenes = list(set(df2[df2['Gene stable ID'].isin(firstOrderEnsg)]['Gene name'].values))\n",
    "print(len(firstOrderGenes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "6916cd11-867e-4b02-abe7-f35559842700",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6507"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "idx = df[df['Stable ID'] == ENSG1].index[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "2ef13b8e-887a-4183-b12f-d11d0b648fb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "idx = df[df['Stable ID'] == ENSG1].index[0]\n",
    "idxs = np.where(A[idx,:] == 1)[1]\n",
    "firstOrderEnsg = list(df.iloc[idxs]['Stable ID'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "82a94368-adcd-4f33-8eb3-6be12188815a",
   "metadata": {},
   "outputs": [],
   "source": [
    "idxs = np.where(A[6507,:] == 1)[1]\n",
    "firstOrderEnsg = list(df.iloc[idxs]['Stable ID'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "d83152eb-7493-473c-b5da-b944fa46dc05",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on _ArrayFunctionDispatcher in module numpy:\n",
      "\n",
      "where(...)\n",
      "    where(condition, [x, y], /)\n",
      "    \n",
      "    Return elements chosen from `x` or `y` depending on `condition`.\n",
      "    \n",
      "    .. note::\n",
      "        When only `condition` is provided, this function is a shorthand for\n",
      "        ``np.asarray(condition).nonzero()``. Using `nonzero` directly should be\n",
      "        preferred, as it behaves correctly for subclasses. The rest of this\n",
      "        documentation covers only the case where all three arguments are\n",
      "        provided.\n",
      "    \n",
      "    Parameters\n",
      "    ----------\n",
      "    condition : array_like, bool\n",
      "        Where True, yield `x`, otherwise yield `y`.\n",
      "    x, y : array_like\n",
      "        Values from which to choose. `x`, `y` and `condition` need to be\n",
      "        broadcastable to some shape.\n",
      "    \n",
      "    Returns\n",
      "    -------\n",
      "    out : ndarray\n",
      "        An array with elements from `x` where `condition` is True, and elements\n",
      "        from `y` elsewhere.\n",
      "    \n",
      "    See Also\n",
      "    --------\n",
      "    choose\n",
      "    nonzero : The function that is called when x and y are omitted\n",
      "    \n",
      "    Notes\n",
      "    -----\n",
      "    If all the arrays are 1-D, `where` is equivalent to::\n",
      "    \n",
      "        [xv if c else yv\n",
      "         for c, xv, yv in zip(condition, x, y)]\n",
      "    \n",
      "    Examples\n",
      "    --------\n",
      "    >>> a = np.arange(10)\n",
      "    >>> a\n",
      "    array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])\n",
      "    >>> np.where(a < 5, a, 10*a)\n",
      "    array([ 0,  1,  2,  3,  4, 50, 60, 70, 80, 90])\n",
      "    \n",
      "    This can be used on multidimensional arrays too:\n",
      "    \n",
      "    >>> np.where([[True, False], [True, True]],\n",
      "    ...          [[1, 2], [3, 4]],\n",
      "    ...          [[9, 8], [7, 6]])\n",
      "    array([[1, 8],\n",
      "           [3, 4]])\n",
      "    \n",
      "    The shapes of x, y, and the condition are broadcast together:\n",
      "    \n",
      "    >>> x, y = np.ogrid[:3, :4]\n",
      "    >>> np.where(x < y, x, 10 + y)  # both x and 10+y are broadcast\n",
      "    array([[10,  0,  0,  0],\n",
      "           [10, 11,  1,  1],\n",
      "           [10, 11, 12,  2]])\n",
      "    \n",
      "    >>> a = np.array([[0, 1, 2],\n",
      "    ...               [0, 2, 4],\n",
      "    ...               [0, 3, 6]])\n",
      "    >>> np.where(a < 4, a, -1)  # -1 is broadcast\n",
      "    array([[ 0,  1,  2],\n",
      "           [ 0,  2, -1],\n",
      "           [ 0,  3, -1]])\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(np.where)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "48f4d525-5b3a-46a4-92d0-1d3ea698c13a",
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "tuple indices must be integers or slices, not tuple",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[43], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m A \u001b[38;5;241m=\u001b[39m \u001b[43mA\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m,\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241m.\u001b[39mtodense()\n",
      "\u001b[0;31mTypeError\u001b[0m: tuple indices must be integers or slices, not tuple"
     ]
    }
   ],
   "source": [
    "A = A[0,0].todense()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "c4776e82-3460-49d7-a4f9-77f21546bcbc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Protein stable ID</th>\n",
       "      <th>Gene start (bp)</th>\n",
       "      <th>Gene end (bp)</th>\n",
       "      <th>Chromosome/scaffold name</th>\n",
       "      <th>Gene name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>50403</th>\n",
       "      <td>ENSG00000132646</td>\n",
       "      <td>ENSP00000368438</td>\n",
       "      <td>5114953</td>\n",
       "      <td>5126626</td>\n",
       "      <td>20</td>\n",
       "      <td>PCNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50404</th>\n",
       "      <td>ENSG00000132646</td>\n",
       "      <td>ENSP00000368458</td>\n",
       "      <td>5114953</td>\n",
       "      <td>5126626</td>\n",
       "      <td>20</td>\n",
       "      <td>PCNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Gene stable ID Protein stable ID  Gene start (bp)  Gene end (bp)  \\\n",
       "50403  ENSG00000132646   ENSP00000368438          5114953        5126626   \n",
       "50404  ENSG00000132646   ENSP00000368458          5114953        5126626   \n",
       "\n",
       "      Chromosome/scaffold name Gene name  \n",
       "50403                       20      PCNA  \n",
       "50404                       20      PCNA  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2 = pd.read_csv('GRCh38.p13.tsv', delimiter='\\t')\n",
    "df2[df2['Gene name'] == geneName]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "330bf723-801d-4f70-9209-dff4c813ace2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Stable ID</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Transcription Factor</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Stable ID, Gene Name, Transcription Factor, Chromosome, Start, End]\n",
       "Index: []"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['Gene Name'] == 'PCNA']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ce876d4c-d90e-46ee-addc-477b5bb82619",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Stable ID</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Transcription Factor</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>MT-RNR1</td>\n",
       "      <td>0</td>\n",
       "      <td>23</td>\n",
       "      <td>100627108</td>\n",
       "      <td>100639991</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>MT-TV</td>\n",
       "      <td>0</td>\n",
       "      <td>23</td>\n",
       "      <td>100584936</td>\n",
       "      <td>100599885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>MT-RNR2</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>50934867</td>\n",
       "      <td>50958555</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>MT-TL1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>169849631</td>\n",
       "      <td>169894267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>MT-ND1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>169662007</td>\n",
       "      <td>169854080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19211</th>\n",
       "      <td>ENSG00000287363</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>42155406</td>\n",
       "      <td>42180056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19212</th>\n",
       "      <td>ENSG00000288602</td>\n",
       "      <td>LINC02311</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19213</th>\n",
       "      <td>ENSG00000288674</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19214</th>\n",
       "      <td>ENSG00000288722</td>\n",
       "      <td>EMX2OS</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19215</th>\n",
       "      <td>ENSG00000288859</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19216 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Stable ID  Gene Name  Transcription Factor  Chromosome  \\\n",
       "0      ENSG00000000003    MT-RNR1                     0          23   \n",
       "1      ENSG00000000005      MT-TV                     0          23   \n",
       "2      ENSG00000000419    MT-RNR2                     0          20   \n",
       "3      ENSG00000000457     MT-TL1                     0           1   \n",
       "4      ENSG00000000460     MT-ND1                     0           1   \n",
       "...                ...        ...                   ...         ...   \n",
       "19211  ENSG00000287363        NaN                     0           6   \n",
       "19212  ENSG00000288602  LINC02311                     0           0   \n",
       "19213  ENSG00000288674        NaN                     0           0   \n",
       "19214  ENSG00000288722     EMX2OS                     0           0   \n",
       "19215  ENSG00000288859        NaN                     0           0   \n",
       "\n",
       "           Start        End  \n",
       "0      100627108  100639991  \n",
       "1      100584936  100599885  \n",
       "2       50934867   50958555  \n",
       "3      169849631  169894267  \n",
       "4      169662007  169854080  \n",
       "...          ...        ...  \n",
       "19211   42155406   42180056  \n",
       "19212          0          0  \n",
       "19213          0          0  \n",
       "19214          0          0  \n",
       "19215          0          0  \n",
       "\n",
       "[19216 rows x 6 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "10d22dab-3f8c-41fa-9629-892ae3b425ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s0</th>\n",
       "      <th>s1</th>\n",
       "      <th>s2</th>\n",
       "      <th>arr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b''</td>\n",
       "      <td>b'MCOS'</td>\n",
       "      <td>b'table'</td>\n",
       "      <td>[[3707764736], [2], [1], [1], [1], [2]]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    s0       s1        s2                                      arr\n",
       "0  b''  b'MCOS'  b'table'  [[3707764736], [2], [1], [1], [1], [2]]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "pd.DataFrame(HWG['HWG']['geneIndexTable'][0][0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "727bafa2-0bfb-4949-a8c7-e61d0b143876",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[(array([[0]], dtype=uint8), <19216x19216 sparse matrix of type '<class 'numpy.uint8'>'\n",
       "        \twith 11750266 stored elements in Compressed Sparse Column format>, <19216x1621 sparse matrix of type '<class 'numpy.uint8'>'\n",
       "        \twith 881627 stored elements in Compressed Sparse Column format>, <1621x1621 sparse matrix of type '<class 'numpy.uint8'>'\n",
       "        \twith 234808 stored elements in Compressed Sparse Column format>, MatlabOpaque([(b'', b'MCOS', b'table', array([[3707764736],\n",
       "                             [         2],\n",
       "                             [         1],\n",
       "                             [         1],\n",
       "                             [         1],\n",
       "                             [         2]], dtype=uint32))         ],\n",
       "                     dtype=[('s0', 'O'), ('s1', 'O'), ('s2', 'O'), ('arr', 'O')]), MatlabOpaque([(b'', b'MCOS', b'table', array([[3707764736],\n",
       "                             [         2],\n",
       "                             [         1],\n",
       "                             [         1],\n",
       "                             [     18979],\n",
       "                             [         2]], dtype=uint32))         ],\n",
       "                     dtype=[('s0', 'O'), ('s1', 'O'), ('s2', 'O'), ('arr', 'O')]))                                                            ]],\n",
       "      dtype=[('thresh', 'O'), ('A', 'O'), ('B', 'O'), ('C', 'O'), ('geneIndexTable', 'O'), ('TFIndexTable', 'O')])"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HWG['HWG']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3415fc4d-7d87-4b96-95aa-f6b7c7feb46e",
   "metadata": {},
   "source": [
    "# DEBUG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "326c4767-bc1f-4d62-8f00-d439f0291e0c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Perform gene set enrichment analysis on the identified interactions from Step 2 using databases like Gene Ontology and Enrichr. This will provide information on biological pathways and cell types associated with these interactions.'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "XX = \"/force DATABASE Perform gene set enrichment analysis on the identified interactions from Step 2 using databases like Gene Ontology and Enrichr. This will provide information on biological pathways and cell types associated with these interactions.\"\n",
    "\" \".join(XX.split(' ')[2:]).strip()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7ee58c30-e7ac-41bf-bbf7-5448f23f38c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "from BRAD.matlabCaller import find_matlab_files, get_matlab_description\n",
    "from BRAD.pythonCaller import find_py_files, get_py_description"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "89d689d4-263c-48a9-90d7-dca0336f11ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['animal_info', 'example', 'helloWorldPYTHON']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "find_py_files('/home/jpic/RAG-DEV/tutorials/running-codes/python/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ea1a7695-8668-4cd0-b79e-19ade4f3072d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'This script prints hello world to the terminal. No arguments are required.'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_py_description('/home/jpic/RAG-DEV/tutorials/running-codes/python/helloWorldPYTHON.py')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b46e3f06-c310-4caa-b6c5-b890154fd85b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['matlabPipeline', 'myFunc', 'helloWorldMATLAB', 'saveplot', 'qualityControl']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "find_matlab_files('/home/jpic/RAG-DEV/tutorials/running-codes/matlab/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8d43c2f8-78df-4a60-a9b0-7e3a7d908a4a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "' qualityControl - Constructs and runs a bioinformatics quality control pipeline.'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_matlab_description('/home/jpic/RAG-DEV/tutorials/running-codes/matlab/qualityControl.m')"
   ]
  },
  {
   "cell_type": "raw",
   "id": "f1ab7fd5-53ae-466c-9c43-55e0dc3c08fa",
   "metadata": {},
   "source": [
    "/force CODE say hello world using python!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4dd54f73-6d6c-4add-9e82-96b9e7324026",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome to RAG! The chat log from this conversation will be saved to /home/jpic/BRAD/2024-06-16-16:50:05.318931.json. How can I help?\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter your NVIDIA API key:  ········\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Would you like to use a database with BRAD [Y/N]?\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      " N\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sun 16 Jun 2024 04:50:09 PM EDT INFO local\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  /force CODE say hello world using python!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sun 16 Jun 2024 04:50:14 PM EDT INFO CODE\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "RAG >> 1: CODER\n",
      "FIRST LLM CALL\n",
      "Warning! the type doesnt exist\n",
      "scriptPath=/home/jpic/RAG-DEV/tutorials/running-codes/python/\n",
      "pyPath=/home/jpic/RAG-DEV/tutorials/running-codes/python/\n",
      "ALL SCRIPTS FOUND. BUILDING TEMPLATE\n",
      "input_variables=['history', 'input'] template='Current conversation:\\n{history}\\n\\n**PYTHON SCRIPT**\\nYou must run this python script: \\n/home/jpic/RAG-DEV/tutorials/running-codes/python/helloWorldPYTHON.py\\n\\n**PYTHON SCRIPT DOCUMENTATION**:\\nThis is the doc string of this python script:\\n\"\"\"\\nThis script prints hello world to the terminal. No arguments are required.\\n\\nExample usage:\\n    subprocess.call([sys.executable, \\'python/helloWorldPYTHON.py\\'])\\n\"\"\"\\n\\n\\n**CALL PYTHON SCRIPTS FROM PYTHON**:\\nUse the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\\n\\nTo call a Python script `example_script.py` which has no arguments:\\n    \\n```\\nExecute: subprocess.call([sys.executable, \\'<full path to script/> example_script.py\\'])\\n```\\n\\nTo call a Python script `example_script.py` with one argument:\\n\\n```\\nExecute: subprocess.call([sys.executable, \\'<full path to script/> example_script.py\\', \\'arg1\\'])\\n```\\n\\nTo call a Python script `example_script.py` with two arguments:\\n```\\nExecute: subprocess.call([sys.executable, \\'<full path to script/> example_script.py\\', \\'arg1\\', \\'arg2\\'])\\n```\\nQuery:{input}\\n\\n\\n**INSTRUCTIONS**\\n1. Given the user query and the documentation, identify each of the arguments found in the user\\'s query that should be passed to the Python script.\\n2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\\n3. The last line of your response should say \"Execute: <Python code to execute>\"\\n4. Format the response/output as:\\n    Arguments: \\n    Python Code Explanation: <2 sentences maximum>\\n    Execute: <your code here>\\n\\n**IMPORTANT**\\nThe code to execute from your response must be formatted as:\\n    Execute: subprocess.call([sys.executable, \\'<path to python script>\\', \\'<argument 1>\\', \\'<argument 2>\\', ..., \\'<argument n>\\'])\\nThis output should be exactly one line and no longer. Stop the response after this line.\\n'\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation:\n",
      "\n",
      "\n",
      "**PYTHON SCRIPT**\n",
      "You must run this python script: \n",
      "/home/jpic/RAG-DEV/tutorials/running-codes/python/helloWorldPYTHON.py\n",
      "\n",
      "**PYTHON SCRIPT DOCUMENTATION**:\n",
      "This is the doc string of this python script:\n",
      "\"\"\"\n",
      "This script prints hello world to the terminal. No arguments are required.\n",
      "\n",
      "Example usage:\n",
      "    subprocess.call([sys.executable, 'python/helloWorldPYTHON.py'])\n",
      "\"\"\"\n",
      "\n",
      "\n",
      "**CALL PYTHON SCRIPTS FROM PYTHON**:\n",
      "Use the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\n",
      "\n",
      "To call a Python script `example_script.py` which has no arguments:\n",
      "    \n",
      "```\n",
      "Execute: subprocess.call([sys.executable, '<full path to script/> example_script.py'])\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with one argument:\n",
      "\n",
      "```\n",
      "Execute: subprocess.call([sys.executable, '<full path to script/> example_script.py', 'arg1'])\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with two arguments:\n",
      "```\n",
      "Execute: subprocess.call([sys.executable, '<full path to script/> example_script.py', 'arg1', 'arg2'])\n",
      "```\n",
      "Query:/force CODE say hello world using python!\n",
      "\n",
      "\n",
      "**INSTRUCTIONS**\n",
      "1. Given the user query and the documentation, identify each of the arguments found in the user's query that should be passed to the Python script.\n",
      "2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\n",
      "3. The last line of your response should say \"Execute: <Python code to execute>\"\n",
      "4. Format the response/output as:\n",
      "    Arguments: \n",
      "    Python Code Explanation: <2 sentences maximum>\n",
      "    Execute: <your code here>\n",
      "\n",
      "**IMPORTANT**\n",
      "The code to execute from your response must be formatted as:\n",
      "    Execute: subprocess.call([sys.executable, '<path to python script>', '<argument 1>', '<argument 2>', ..., '<argument n>'])\n",
      "This output should be exactly one line and no longer. Stop the response after this line.\n",
      "\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "LLM OUTPUT PARSER\n",
      " Arguments: The user query does not provide any arguments to be passed to the Python script. The script documentation indicates no arguments are required.\n",
      "\n",
      "Python Code Explanation: We will execute the \"helloWorldPYTHON.py\" script with no arguments.\n",
      "\n",
      "Execute: subprocess.call([sys.executable, '/home/jpic/RAG-DEV/tutorials/running-codes/python/helloWorldPYTHON.py'])\n",
      "[' Arguments: The user query does not provide any arguments to be passed to the Python script. The script documentation indicates no arguments are required.\\n\\nPython Code Explanation: We will execute the \"helloWorldPYTHON.py\" script with no arguments.\\n\\n', \" subprocess.call([sys.executable, '/home/jpic/RAG-DEV/tutorials/running-codes/python/helloWorldPYTHON.py'])\"]\n",
      "Executing Code!!\n",
      "EVAL\n",
      "subprocess.call([sys.executable, '/home/jpic/RAG-DEV/tutorials/running-codes/python/helloWorldPYTHON.py'])\n",
      "END EVAL CHECK\n",
      "Hello World from a Python Script!\n",
      "Debug: PYTHON code executed successfully.\n",
      "==================================================\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  /force CODE say hello world using matlab\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sun 16 Jun 2024 04:50:26 PM EDT INFO CODE\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "RAG >> 2: CODER\n",
      "FIRST LLM CALL\n",
      "Warning! the type doesnt exist\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...\n",
      "To disable this warning, you can either:\n",
      "\t- Avoid using `tokenizers` before the fork if possible\n",
      "\t- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Matlab Engine On\n",
      "ALL SCRIPTS FOUND. BUILDING TEMPLATE\n",
      "input_variables=['history', 'input'] template='Current conversation:\\n{history}\\n\\n**MATLAB SCRIPT**\\nYou must run this python script: \\nhelloWorldMATLAB.m\\n\\n**MATLAB FUNCTION DOCUMENTATION**:\\nfunction output = helloWorldMATLAB()\\n% Print \"Hello World\"\\n%\\n% Author: Joshua Pickard\\n%         jpic@umich.edu\\n% Date: June 14, 2024\\n\\n**CALL MATLAB FUNCTIONS FROM PYTHON**:\\nUse MATLAB® Engine API for Python® to call any MATLAB function on the MATLAB path. Some examples to call a few common matlab functions include:\\n\\nTo call the matlab function myFnc which has no arguments:\\n```\\nExecute: eng.myFnc()\\n```\\n\\nTo use the is prime function which requires an input argument:\\n```\\nExecute: tf = eng.isprime(37)\\n```\\n\\nTo determine the greatest common denominator of two numbers, use the gcd function. Set nargout to return the three output arguments from gcd.\\n```\\nExecute: t = eng.gcd(100.0,80.0,nargout=3)\\n```\\n\\nQuery:{input}\\n\\n**INSTRUCTIONS**\\n1. Given the user query and the documentation, identify each of the arguments found in the users query that should be passed to the matlab function.\\n2. Using the matlab enging eng, provide the one line of code to execute the desired matlab commands. Assume all functions are added to the path and eng already exist.\\n3. The last line of your response should say \"Execute: <MATLAB codes to execute>\"\\n4. Format the response/output as:\\nArguments: \\nMATLAB Code Explanation: <2 sentences maximum>\\nExecute: <your code here>\\n\\n**IMPORTANT**\\nThe code to execute from your response must be formatted as:\\n    Execute: eng.<function name>(<arguments>)>\\nThis output should be exactly one line and no longer. Stop the response after this line.\\n'\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation:\n",
      "Human: /force CODE say hello world using python!\n",
      "BRAD:  Arguments: The user query does not provide any arguments to be passed to the Python script. The script documentation indicates no arguments are required.\n",
      "\n",
      "Python Code Explanation: We will execute the \"helloWorldPYTHON.py\" script with no arguments.\n",
      "\n",
      "Execute: subprocess.call([sys.executable, '/home/jpic/RAG-DEV/tutorials/running-codes/python/helloWorldPYTHON.py'])\n",
      "\n",
      "**MATLAB SCRIPT**\n",
      "You must run this python script: \n",
      "helloWorldMATLAB.m\n",
      "\n",
      "**MATLAB FUNCTION DOCUMENTATION**:\n",
      "function output = helloWorldMATLAB()\n",
      "% Print \"Hello World\"\n",
      "%\n",
      "% Author: Joshua Pickard\n",
      "%         jpic@umich.edu\n",
      "% Date: June 14, 2024\n",
      "\n",
      "**CALL MATLAB FUNCTIONS FROM PYTHON**:\n",
      "Use MATLAB® Engine API for Python® to call any MATLAB function on the MATLAB path. Some examples to call a few common matlab functions include:\n",
      "\n",
      "To call the matlab function myFnc which has no arguments:\n",
      "```\n",
      "Execute: eng.myFnc()\n",
      "```\n",
      "\n",
      "To use the is prime function which requires an input argument:\n",
      "```\n",
      "Execute: tf = eng.isprime(37)\n",
      "```\n",
      "\n",
      "To determine the greatest common denominator of two numbers, use the gcd function. Set nargout to return the three output arguments from gcd.\n",
      "```\n",
      "Execute: t = eng.gcd(100.0,80.0,nargout=3)\n",
      "```\n",
      "\n",
      "Query:/force CODE say hello world using matlab\n",
      "\n",
      "**INSTRUCTIONS**\n",
      "1. Given the user query and the documentation, identify each of the arguments found in the users query that should be passed to the matlab function.\n",
      "2. Using the matlab enging eng, provide the one line of code to execute the desired matlab commands. Assume all functions are added to the path and eng already exist.\n",
      "3. The last line of your response should say \"Execute: <MATLAB codes to execute>\"\n",
      "4. Format the response/output as:\n",
      "Arguments: \n",
      "MATLAB Code Explanation: <2 sentences maximum>\n",
      "Execute: <your code here>\n",
      "\n",
      "**IMPORTANT**\n",
      "The code to execute from your response must be formatted as:\n",
      "    Execute: eng.<function name>(<arguments>)>\n",
      "This output should be exactly one line and no longer. Stop the response after this line.\n",
      "\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "LLM OUTPUT PARSER\n",
      " Arguments: No arguments are required for the \"helloWorldMATLAB\" function.\n",
      "\n",
      "MATLAB Code Explanation: We will execute the \"helloWorldMATLAB.m\" function with no arguments.\n",
      "\n",
      "Execute: eng.helloWorldMATLAB()\n",
      "[' Arguments: No arguments are required for the \"helloWorldMATLAB\" function.\\n\\nMATLAB Code Explanation: We will execute the \"helloWorldMATLAB.m\" function with no arguments.\\n\\n', ' eng.helloWorldMATLAB()']\n",
      "Executing Code!!\n",
      "EVAL\n",
      "eng.helloWorldMATLAB()\n",
      "END EVAL CHECK\n",
      "Hello World! This message is generated from BRAD running the helloWorldMATLAB function!\n",
      "Debug: MATLAB code executed successfully.\n",
      "==================================================\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  /force CODE say hello world using matlab\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sun 16 Jun 2024 04:50:46 PM EDT INFO CODE\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "RAG >> 3: CODER\n",
      "FIRST LLM CALL\n",
      "Warning! the type doesnt exist\n",
      "Matlab Engine On\n",
      "ALL SCRIPTS FOUND. BUILDING TEMPLATE\n",
      "input_variables=['history', 'input'] template='Current conversation:\\n{history}\\n\\n**MATLAB SCRIPT**\\nYou must run this python script: \\nhelloWorldMATLAB.m\\n\\n**MATLAB FUNCTION DOCUMENTATION**:\\nfunction output = helloWorldMATLAB()\\n% Print \"Hello World\"\\n%\\n% Author: Joshua Pickard\\n%         jpic@umich.edu\\n% Date: June 14, 2024\\n\\n**CALL MATLAB FUNCTIONS FROM PYTHON**:\\nUse MATLAB® Engine API for Python® to call any MATLAB function on the MATLAB path. Some examples to call a few common matlab functions include:\\n\\nTo call the matlab function myFnc which has no arguments:\\n```\\nExecute: eng.myFnc()\\n```\\n\\nTo use the is prime function which requires an input argument:\\n```\\nExecute: tf = eng.isprime(37)\\n```\\n\\nTo determine the greatest common denominator of two numbers, use the gcd function. Set nargout to return the three output arguments from gcd.\\n```\\nExecute: t = eng.gcd(100.0,80.0,nargout=3)\\n```\\n\\nQuery:{input}\\n\\n**INSTRUCTIONS**\\n1. Given the user query and the documentation, identify each of the arguments found in the users query that should be passed to the matlab function.\\n2. Using the matlab enging eng, provide the one line of code to execute the desired matlab commands. Assume all functions are added to the path and eng already exist.\\n3. The last line of your response should say \"Execute: <MATLAB codes to execute>\"\\n4. Format the response/output as:\\nArguments: \\nMATLAB Code Explanation: <2 sentences maximum>\\nExecute: <your code here>\\n\\n**IMPORTANT**\\nThe code to execute from your response must be formatted as:\\n    Execute: eng.<function name>(<arguments>)>\\nThis output should be exactly one line and no longer. Stop the response after this line.\\n'\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation:\n",
      "Human: /force CODE say hello world using python!\n",
      "BRAD:  Arguments: The user query does not provide any arguments to be passed to the Python script. The script documentation indicates no arguments are required.\n",
      "\n",
      "Python Code Explanation: We will execute the \"helloWorldPYTHON.py\" script with no arguments.\n",
      "\n",
      "Execute: subprocess.call([sys.executable, '/home/jpic/RAG-DEV/tutorials/running-codes/python/helloWorldPYTHON.py'])\n",
      "Human: /force CODE say hello world using matlab\n",
      "BRAD:  Arguments: No arguments are required for the \"helloWorldMATLAB\" function.\n",
      "\n",
      "MATLAB Code Explanation: We will execute the \"helloWorldMATLAB.m\" function with no arguments.\n",
      "\n",
      "Execute: eng.helloWorldMATLAB()\n",
      "\n",
      "**MATLAB SCRIPT**\n",
      "You must run this python script: \n",
      "helloWorldMATLAB.m\n",
      "\n",
      "**MATLAB FUNCTION DOCUMENTATION**:\n",
      "function output = helloWorldMATLAB()\n",
      "% Print \"Hello World\"\n",
      "%\n",
      "% Author: Joshua Pickard\n",
      "%         jpic@umich.edu\n",
      "% Date: June 14, 2024\n",
      "\n",
      "**CALL MATLAB FUNCTIONS FROM PYTHON**:\n",
      "Use MATLAB® Engine API for Python® to call any MATLAB function on the MATLAB path. Some examples to call a few common matlab functions include:\n",
      "\n",
      "To call the matlab function myFnc which has no arguments:\n",
      "```\n",
      "Execute: eng.myFnc()\n",
      "```\n",
      "\n",
      "To use the is prime function which requires an input argument:\n",
      "```\n",
      "Execute: tf = eng.isprime(37)\n",
      "```\n",
      "\n",
      "To determine the greatest common denominator of two numbers, use the gcd function. Set nargout to return the three output arguments from gcd.\n",
      "```\n",
      "Execute: t = eng.gcd(100.0,80.0,nargout=3)\n",
      "```\n",
      "\n",
      "Query:/force CODE say hello world using matlab\n",
      "\n",
      "**INSTRUCTIONS**\n",
      "1. Given the user query and the documentation, identify each of the arguments found in the users query that should be passed to the matlab function.\n",
      "2. Using the matlab enging eng, provide the one line of code to execute the desired matlab commands. Assume all functions are added to the path and eng already exist.\n",
      "3. The last line of your response should say \"Execute: <MATLAB codes to execute>\"\n",
      "4. Format the response/output as:\n",
      "Arguments: \n",
      "MATLAB Code Explanation: <2 sentences maximum>\n",
      "Execute: <your code here>\n",
      "\n",
      "**IMPORTANT**\n",
      "The code to execute from your response must be formatted as:\n",
      "    Execute: eng.<function name>(<arguments>)>\n",
      "This output should be exactly one line and no longer. Stop the response after this line.\n",
      "\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "LLM OUTPUT PARSER\n",
      " Arguments: No arguments are required for the \"helloWorldMATLAB\" function.\n",
      "\n",
      "MATLAB Code Explanation: We will execute the \"helloWorldMATLAB\" function with no arguments.\n",
      "\n",
      "Execute: eng.helloWorldMATLAB()\n",
      "[' Arguments: No arguments are required for the \"helloWorldMATLAB\" function.\\n\\nMATLAB Code Explanation: We will execute the \"helloWorldMATLAB\" function with no arguments.\\n\\n', ' eng.helloWorldMATLAB()']\n",
      "Executing Code!!\n",
      "EVAL\n",
      "eng.helloWorldMATLAB()\n",
      "END EVAL CHECK\n",
      "Hello World! This message is generated from BRAD running the helloWorldMATLAB function!\n",
      "Debug: MATLAB code executed successfully.\n",
      "==================================================\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  /force CODE say hello world using python!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sun 16 Jun 2024 04:50:55 PM EDT INFO CODE\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "RAG >> 4: CODER\n",
      "FIRST LLM CALL\n",
      "Warning! the type doesnt exist\n",
      "scriptPath=/home/jpic/RAG-DEV/tutorials/running-codes/python/\n",
      "pyPath=/home/jpic/RAG-DEV/tutorials/running-codes/python/\n",
      "ALL SCRIPTS FOUND. BUILDING TEMPLATE\n",
      "input_variables=['history', 'input'] template='Current conversation:\\n{history}\\n\\n**PYTHON SCRIPT**\\nYou must run this python script: \\n/home/jpic/RAG-DEV/tutorials/running-codes/python/helloWorldPYTHON.py\\n\\n**PYTHON SCRIPT DOCUMENTATION**:\\nThis is the doc string of this python script:\\n\"\"\"\\nThis script prints hello world to the terminal. No arguments are required.\\n\\nExample usage:\\n    subprocess.call([sys.executable, \\'python/helloWorldPYTHON.py\\'])\\n\"\"\"\\n\\n\\n**CALL PYTHON SCRIPTS FROM PYTHON**:\\nUse the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\\n\\nTo call a Python script `example_script.py` which has no arguments:\\n    \\n```\\nExecute: subprocess.call([sys.executable, \\'<full path to script/> example_script.py\\'])\\n```\\n\\nTo call a Python script `example_script.py` with one argument:\\n\\n```\\nExecute: subprocess.call([sys.executable, \\'<full path to script/> example_script.py\\', \\'arg1\\'])\\n```\\n\\nTo call a Python script `example_script.py` with two arguments:\\n```\\nExecute: subprocess.call([sys.executable, \\'<full path to script/> example_script.py\\', \\'arg1\\', \\'arg2\\'])\\n```\\nQuery:{input}\\n\\n\\n**INSTRUCTIONS**\\n1. Given the user query and the documentation, identify each of the arguments found in the user\\'s query that should be passed to the Python script.\\n2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\\n3. The last line of your response should say \"Execute: <Python code to execute>\"\\n4. Format the response/output as:\\n    Arguments: \\n    Python Code Explanation: <2 sentences maximum>\\n    Execute: <your code here>\\n\\n**IMPORTANT**\\nThe code to execute from your response must be formatted as:\\n    Execute: subprocess.call([sys.executable, \\'<path to python script>\\', \\'<argument 1>\\', \\'<argument 2>\\', ..., \\'<argument n>\\'])\\nThis output should be exactly one line and no longer. Stop the response after this line.\\n'\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation:\n",
      "Human: /force CODE say hello world using python!\n",
      "BRAD:  Arguments: The user query does not provide any arguments to be passed to the Python script. The script documentation indicates no arguments are required.\n",
      "\n",
      "Python Code Explanation: We will execute the \"helloWorldPYTHON.py\" script with no arguments.\n",
      "\n",
      "Execute: subprocess.call([sys.executable, '/home/jpic/RAG-DEV/tutorials/running-codes/python/helloWorldPYTHON.py'])\n",
      "Human: /force CODE say hello world using matlab\n",
      "BRAD:  Arguments: No arguments are required for the \"helloWorldMATLAB\" function.\n",
      "\n",
      "MATLAB Code Explanation: We will execute the \"helloWorldMATLAB.m\" function with no arguments.\n",
      "\n",
      "Execute: eng.helloWorldMATLAB()\n",
      "Human: /force CODE say hello world using matlab\n",
      "BRAD:  Arguments: No arguments are required for the \"helloWorldMATLAB\" function.\n",
      "\n",
      "MATLAB Code Explanation: We will execute the \"helloWorldMATLAB\" function with no arguments.\n",
      "\n",
      "Execute: eng.helloWorldMATLAB()\n",
      "\n",
      "**PYTHON SCRIPT**\n",
      "You must run this python script: \n",
      "/home/jpic/RAG-DEV/tutorials/running-codes/python/helloWorldPYTHON.py\n",
      "\n",
      "**PYTHON SCRIPT DOCUMENTATION**:\n",
      "This is the doc string of this python script:\n",
      "\"\"\"\n",
      "This script prints hello world to the terminal. No arguments are required.\n",
      "\n",
      "Example usage:\n",
      "    subprocess.call([sys.executable, 'python/helloWorldPYTHON.py'])\n",
      "\"\"\"\n",
      "\n",
      "\n",
      "**CALL PYTHON SCRIPTS FROM PYTHON**:\n",
      "Use the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\n",
      "\n",
      "To call a Python script `example_script.py` which has no arguments:\n",
      "    \n",
      "```\n",
      "Execute: subprocess.call([sys.executable, '<full path to script/> example_script.py'])\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with one argument:\n",
      "\n",
      "```\n",
      "Execute: subprocess.call([sys.executable, '<full path to script/> example_script.py', 'arg1'])\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with two arguments:\n",
      "```\n",
      "Execute: subprocess.call([sys.executable, '<full path to script/> example_script.py', 'arg1', 'arg2'])\n",
      "```\n",
      "Query:/force CODE say hello world using python!\n",
      "\n",
      "\n",
      "**INSTRUCTIONS**\n",
      "1. Given the user query and the documentation, identify each of the arguments found in the user's query that should be passed to the Python script.\n",
      "2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\n",
      "3. The last line of your response should say \"Execute: <Python code to execute>\"\n",
      "4. Format the response/output as:\n",
      "    Arguments: \n",
      "    Python Code Explanation: <2 sentences maximum>\n",
      "    Execute: <your code here>\n",
      "\n",
      "**IMPORTANT**\n",
      "The code to execute from your response must be formatted as:\n",
      "    Execute: subprocess.call([sys.executable, '<path to python script>', '<argument 1>', '<argument 2>', ..., '<argument n>'])\n",
      "This output should be exactly one line and no longer. Stop the response after this line.\n",
      "\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "LLM OUTPUT PARSER\n",
      " Arguments: No arguments are required for the \"helloWorldPYTHON.py\" script according to the documentation and the user query.\n",
      "\n",
      "Python Code Explanation: We will execute the \"helloWorldPYTHON.py\" script with no arguments.\n",
      "\n",
      "Execute: subprocess.call([sys.executable, '/home/jpic/RAG-DEV/tutorials/running-codes/python/helloWorldPYTHON.py'])\n",
      "[' Arguments: No arguments are required for the \"helloWorldPYTHON.py\" script according to the documentation and the user query.\\n\\nPython Code Explanation: We will execute the \"helloWorldPYTHON.py\" script with no arguments.\\n\\n', \" subprocess.call([sys.executable, '/home/jpic/RAG-DEV/tutorials/running-codes/python/helloWorldPYTHON.py'])\"]\n",
      "Executing Code!!\n",
      "EVAL\n",
      "subprocess.call([sys.executable, '/home/jpic/RAG-DEV/tutorials/running-codes/python/helloWorldPYTHON.py'])\n",
      "END EVAL CHECK\n",
      "Hello World from a Python Script!\n",
      "Debug: PYTHON code executed successfully.\n",
      "==================================================\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  q\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Thanks for chatting today! I hope to talk soon, and don't forget that a record of this conversation is available at: /home/jpic/BRAD/2024-06-16-16:50:05.318931.json\n"
     ]
    }
   ],
   "source": [
    "from BRAD import brad\n",
    "brad.chat()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0638b74c-cdc3-4336-90de-7d6ef15014dc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78878573-5b4e-437a-a112-671aa57fc425",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9247861d-8023-4c0c-88ef-484fac25932e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "094373c3-6dc9-46cb-97dc-0891ad1f4cbf",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9563c151-cef6-4efc-8175-5c71bf6a563f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1dda72ea-7694-40b4-be8e-141e793cbbe6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "e229f04e-ec7c-4b13-a2aa-a1d4f0e6fe5d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "**Available Scripts**\n",
      "Script Name: animal_info, \t Description: This script provides information about different animals based on user input.\n",
      "Script Name: example, \t Description: This is the top-level docstring for the example script.\n",
      "Script Name: helloWorldPYTHON, \t Description: This script prints hello world to the terminal. No arguments are required.\n",
      "Script Name: matlabPipeline, \t Description:  matlabPipeline - A sample function to demonstrate argument handling\n",
      "Script Name: myFunc, \t Description:  myFunc - A simple function to multiply each element of the input array by 2.\n",
      "Script Name: helloWorldMATLAB, \t Description:  Print \"Hello World\"\n",
      "Script Name: saveplot, \t Description: SAVEPLOT Save the plot to a file\n",
      "Script Name: qualityControl, \t Description:  qualityControl - Constructs and runs a bioinformatics quality control pipeline.\n",
      "\n",
      "\n",
      "**User Query**\n",
      "{user_query}\n",
      "\n",
      "**Task**\n",
      "Based on the user's query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\n",
      "\n",
      "**Response Template**\n",
      "SCRIPT: <selected script>\n",
      "REASON: <reasoning why the script fits the user prompt>\n",
      "\n"
     ]
    }
   ],
   "source": [
    "scriptPurpose = {'animal_info': 'This script provides information about different animals based on user input.', 'example': 'This is the top-level docstring for the example script.', 'helloWorldPYTHON': 'This script prints hello world to the terminal. No arguments are required.', 'matlabPipeline': ' matlabPipeline - A sample function to demonstrate argument handling', 'myFunc': ' myFunc - A simple function to multiply each element of the input array by 2.', 'helloWorldMATLAB': ' Print \"Hello World\"', 'saveplot': 'SAVEPLOT Save the plot to a file', 'qualityControl': ' qualityControl - Constructs and runs a bioinformatics quality control pipeline.'}\n",
    "script_list = \"\"\n",
    "for script in scriptPurpose.keys():\n",
    "    script_list += \"Script Name: \" + script + \", \\t Description: \" + scriptPurpose[script] + '\\n'\n",
    "template = scriptSelectorTemplate()\n",
    "template = template.format(script_list=script_list)\n",
    "print(template)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "f9602f0a-5b9f-47cd-a591-5b605859bef3",
   "metadata": {},
   "outputs": [],
   "source": [
    "from BRAD import llms\n",
    "from langchain.memory import ConversationBufferMemory\n",
    "llm = llms.load_nvidia()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "7ba080f9-6672-4e25-8820-95f7dfaa8f45",
   "metadata": {},
   "outputs": [],
   "source": [
    "template = scriptSelectorTemplate()\n",
    "template = template.format(script_list=script_list)\n",
    "PROMPT = PromptTemplate(input_variables=[\"user_query\"], template=template)\n",
    "chain = PROMPT | llm\n",
    "res = chain.invoke(\"use python to say hello\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "e99d5376-f427-4822-891d-f92605fb8a3d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SCRIPT: helloWorldPYTHON\n",
      "REASON: The user query \"use python to say hello\" is asking for a simple Python script to print a greeting message to the terminal. The script named \"helloWorldPYTHON\" matches this query as its primary function is to print the message \"Hello World\" to the terminal in Python. However, since the user explicitly requested \"hello\" in their query, it is a more accurate match for the script named \"helloWorldPYTHON\" that explicitly prints the message \"hello\" instead. Therefore, \"helloWorldPYTHON\" is the best script to run to help the user.\n"
     ]
    }
   ],
   "source": [
    "print(res.content.strip())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "a8fb5916-fc2b-40fd-bda3-baac442f6d07",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'helloWorldPYTHON'"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scriptName = res.content.strip().split('\\n')[0].split(':')[1].strip()\n",
    "scriptName"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "2c4848ee-99d1-4cd5-a429-093d1a2e0754",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValidationError",
     "evalue": "1 validation error for ConversationChain\n__root__\n  Got unexpected prompt input variables. The prompt expects ['history', 'user_query'], but got ['history'] as inputs from memory, and input as the normal input key. (type=value_error)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValidationError\u001b[0m                           Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[33], line 4\u001b[0m\n\u001b[1;32m      2\u001b[0m template \u001b[38;5;241m=\u001b[39m template\u001b[38;5;241m.\u001b[39mformat(script_list\u001b[38;5;241m=\u001b[39mscript_list)\n\u001b[1;32m      3\u001b[0m PROMPT \u001b[38;5;241m=\u001b[39m PromptTemplate(input_variables\u001b[38;5;241m=\u001b[39m[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhistory\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124muser_query\u001b[39m\u001b[38;5;124m\"\u001b[39m], template\u001b[38;5;241m=\u001b[39mtemplate)\n\u001b[0;32m----> 4\u001b[0m conversation \u001b[38;5;241m=\u001b[39m \u001b[43mConversationChain\u001b[49m\u001b[43m(\u001b[49m\u001b[43mprompt\u001b[49m\u001b[43m  \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mPROMPT\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m      5\u001b[0m \u001b[43m                                 \u001b[49m\u001b[43mllm\u001b[49m\u001b[43m     \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mllm\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m      6\u001b[0m \u001b[43m                                 \u001b[49m\u001b[43mverbose\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m      7\u001b[0m \u001b[43m                                 \u001b[49m\u001b[43mmemory\u001b[49m\u001b[43m  \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mConversationBufferMemory\u001b[49m\u001b[43m(\u001b[49m\u001b[43mai_prefix\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mBRAD\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m      8\u001b[0m \u001b[43m                                \u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/langchain_core/_api/deprecation.py:183\u001b[0m, in \u001b[0;36mdeprecated.<locals>.deprecate.<locals>.finalize.<locals>.warn_if_direct_instance\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    181\u001b[0m     warned \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[1;32m    182\u001b[0m     emit_warning()\n\u001b[0;32m--> 183\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mwrapped\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/pydantic/v1/main.py:341\u001b[0m, in \u001b[0;36mBaseModel.__init__\u001b[0;34m(__pydantic_self__, **data)\u001b[0m\n\u001b[1;32m    339\u001b[0m values, fields_set, validation_error \u001b[38;5;241m=\u001b[39m validate_model(__pydantic_self__\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__class__\u001b[39m, data)\n\u001b[1;32m    340\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m validation_error:\n\u001b[0;32m--> 341\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m validation_error\n\u001b[1;32m    342\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    343\u001b[0m     object_setattr(__pydantic_self__, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m__dict__\u001b[39m\u001b[38;5;124m'\u001b[39m, values)\n",
      "\u001b[0;31mValidationError\u001b[0m: 1 validation error for ConversationChain\n__root__\n  Got unexpected prompt input variables. The prompt expects ['history', 'user_query'], but got ['history'] as inputs from memory, and input as the normal input key. (type=value_error)"
     ]
    }
   ],
   "source": [
    "template = scriptSelectorTemplate()\n",
    "template = template.format(script_list=script_list)\n",
    "PROMPT = PromptTemplate(input_variables=[\"history\", \"user_query\"], template=template)\n",
    "conversation = ConversationChain(prompt  = PROMPT,\n",
    "                                 llm     = llm,\n",
    "                                 verbose = True,\n",
    "                                 memory  = ConversationBufferMemory(ai_prefix=\"BRAD\"),\n",
    "                                )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6bc932e1-2256-4188-8598-8ebcbbcb0cb5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "**Available Scripts**\n",
      "{script_list}\n",
      "\n",
      "**User Query**\n",
      "{user_query}\n",
      "\n",
      "**Task**\n",
      "Based on the user's query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\n",
      "\n",
      "**Response Template**\n",
      "SCRIPT: <selected script>\n",
      "REASON: <reasoning why the script fits the user prompt>\n",
      "\n"
     ]
    }
   ],
   "source": [
    "template = scriptSelectorTemplate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "52f8896c-69e1-48d8-8d68-466f543490b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def scriptSelectorTemplate():\n",
    "    template=\"\"\"You must select which code to run to help a user.\n",
    "\n",
    "**Available Scripts**\n",
    "{script_list}\n",
    "\n",
    "**User Query**\n",
    "{{user_query}}\n",
    "\n",
    "**Task**\n",
    "Based on the user's query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\n",
    "\n",
    "**Response Template**\n",
    "SCRIPT: <selected script>\n",
    "REASON: <reasoning why the script fits the user prompt>\n",
    "\"\"\"\n",
    "    return template"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb0d08aa-5ff9-4a00-883d-4776fba40abd",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
