{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "52bf1e03-eda1-489c-8b8c-5e78a761db42",
   "metadata": {},
   "source": [
    "# HSC Investigation with BRAD"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc2b8d1a-e733-4bf5-8e04-24399c1d505f",
   "metadata": {},
   "source": [
    "# Organize Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "6d446662-cb92-4676-922a-d9549a09bbf0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "from sklearn.cluster import KMeans\n",
    "import anndata as ad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "451cb7ad-d961-4adc-ba82-be6ca05eb185",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/sklearn/cluster/_kmeans.py:1412: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  super()._check_params_vs_input(X, default_n_init=10)\n"
     ]
    }
   ],
   "source": [
    "# Cluster source cells\n",
    "fname = '/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/originals/2024-07-31_07-01-47_job_number_83.h5ad'\n",
    "adata = sc.read_h5ad(fname)\n",
    "\n",
    "# Get the data matrix\n",
    "data_matrix = adata.X\n",
    "\n",
    "# Perform KMeans clustering\n",
    "kmeans = KMeans(n_clusters=25, random_state=0)\n",
    "kmeans.fit(data_matrix)\n",
    "\n",
    "# Get the centroids of each cluster\n",
    "cluster_centroids = kmeans.cluster_centers_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "efd9416d-5086-4673-879f-8a00c88d2e3b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>recipe</th>\n",
       "      <th>filename</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>source</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>source</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>source</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>source</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>source</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>source</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>source</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>source</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>source</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>source</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>source</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>source</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>source</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>source</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>source</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>source</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>source</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>source</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>source</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>source</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>source</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>source</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>source</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>source</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>source</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    recipe                                           filename\n",
       "0   source  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "1   source  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "2   source  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "3   source  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "4   source  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "5   source  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "6   source  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "7   source  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "8   source  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "9   source  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "10  source  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "11  source  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "12  source  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "13  source  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "14  source  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "15  source  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "16  source  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "17  source  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "18  source  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "19  source  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "20  source  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "21  source  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "22  source  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "23  source  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "24  source  /nfs/turbo/umms-indikar/shared/projects/genefo..."
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create an AnnData object for the centroids of the current file\n",
    "adata_centroids = ad.AnnData(X=cluster_centroids)\n",
    "adata_centroids.obs['recipe'] = 'source'\n",
    "adata_centroids.obs['filename'] = fname\n",
    "adata_fib = adata_centroids\n",
    "adata_fib.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "37e9f036-dd6a-4b59-ae7f-b413c3c5cb22",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/aligned_df.py:67: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/sklearn/cluster/_kmeans.py:1412: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  super()._check_params_vs_input(X, default_n_init=10)\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'ad' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[72], line 15\u001b[0m\n\u001b[1;32m     12\u001b[0m cluster_centroids \u001b[38;5;241m=\u001b[39m kmeans\u001b[38;5;241m.\u001b[39mcluster_centers_\n\u001b[1;32m     14\u001b[0m \u001b[38;5;66;03m# Create an AnnData object for the centroids of the current file\u001b[39;00m\n\u001b[0;32m---> 15\u001b[0m adata_centroids \u001b[38;5;241m=\u001b[39m \u001b[43mad\u001b[49m\u001b[38;5;241m.\u001b[39mAnnData(X\u001b[38;5;241m=\u001b[39mcluster_centroids)\n\u001b[1;32m     16\u001b[0m adata_centroids\u001b[38;5;241m.\u001b[39mobs[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mrecipe\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtarget\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m     17\u001b[0m adata_centroids\u001b[38;5;241m.\u001b[39mobs[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mfilename\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m fname\n",
      "\u001b[0;31mNameError\u001b[0m: name 'ad' is not defined"
     ]
    }
   ],
   "source": [
    "# Cluster target cells\n",
    "fname = '/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/working_output/hsc_1.h5ad'\n",
    "adata = sc.read_h5ad(fname)\n",
    "\n",
    "# Get the data matrix\n",
    "data_matrix = adata.X\n",
    "\n",
    "# Perform KMeans clustering\n",
    "kmeans = KMeans(n_clusters=25, random_state=0)\n",
    "kmeans.fit(data_matrix)\n",
    "\n",
    "# Get the centroids of each cluster\n",
    "cluster_centroids = kmeans.cluster_centers_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "13fe408c-5176-4222-8fb4-edc7e769eb56",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>recipe</th>\n",
       "      <th>filename</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>target</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>target</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>target</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>target</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>target</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>target</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>target</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>target</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>target</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>target</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>target</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>target</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>target</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>target</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>target</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>target</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>target</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>target</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>target</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>target</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>target</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>target</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>target</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>target</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>target</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    recipe                                           filename\n",
       "0   target  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "1   target  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "2   target  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "3   target  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "4   target  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "5   target  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "6   target  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "7   target  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "8   target  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "9   target  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "10  target  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "11  target  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "12  target  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "13  target  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "14  target  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "15  target  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "16  target  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "17  target  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "18  target  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "19  target  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "20  target  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "21  target  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "22  target  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "23  target  /nfs/turbo/umms-indikar/shared/projects/genefo...\n",
       "24  target  /nfs/turbo/umms-indikar/shared/projects/genefo..."
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create an AnnData object for the centroids of the current file\n",
    "adata_centroids = ad.AnnData(X=cluster_centroids)\n",
    "adata_centroids.obs['recipe'] = 'target'\n",
    "adata_centroids.obs['filename'] = fname\n",
    "adata_hsc = adata_centroids\n",
    "adata_hsc.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "5711729e-5070-46cd-b1d5-dd7730a39c11",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>recipe</th>\n",
       "      <th>filename</th>\n",
       "      <th>REL</th>\n",
       "      <th>STAT5A</th>\n",
       "      <th>FOS</th>\n",
       "      <th>ETV6</th>\n",
       "      <th>source</th>\n",
       "      <th>MEF2C</th>\n",
       "      <th>RUNX3</th>\n",
       "      <th>FOSB</th>\n",
       "      <th>GATA2</th>\n",
       "      <th>IKZF1</th>\n",
       "      <th>GFI1B</th>\n",
       "      <th>target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>target</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>target</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>target</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>target</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>target</td>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/genefo...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6320</th>\n",
       "      <td>GFI1B;FOS;REL;IKZF1;RUNX3</td>\n",
       "      <td>2024-07-31_10-13-03_job_number_146.h5ad</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6321</th>\n",
       "      <td>GFI1B;FOS;REL;IKZF1;RUNX3</td>\n",
       "      <td>2024-07-31_10-13-03_job_number_146.h5ad</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6322</th>\n",
       "      <td>GFI1B;FOS;REL;IKZF1;RUNX3</td>\n",
       "      <td>2024-07-31_10-13-03_job_number_146.h5ad</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6323</th>\n",
       "      <td>GFI1B;FOS;REL;IKZF1;RUNX3</td>\n",
       "      <td>2024-07-31_10-13-03_job_number_146.h5ad</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6324</th>\n",
       "      <td>GFI1B;FOS;REL;IKZF1;RUNX3</td>\n",
       "      <td>2024-07-31_10-13-03_job_number_146.h5ad</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6375 rows Ã— 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         recipe  \\\n",
       "0                        target   \n",
       "1                        target   \n",
       "2                        target   \n",
       "3                        target   \n",
       "4                        target   \n",
       "...                         ...   \n",
       "6320  GFI1B;FOS;REL;IKZF1;RUNX3   \n",
       "6321  GFI1B;FOS;REL;IKZF1;RUNX3   \n",
       "6322  GFI1B;FOS;REL;IKZF1;RUNX3   \n",
       "6323  GFI1B;FOS;REL;IKZF1;RUNX3   \n",
       "6324  GFI1B;FOS;REL;IKZF1;RUNX3   \n",
       "\n",
       "                                               filename    REL  STAT5A    FOS  \\\n",
       "0     /nfs/turbo/umms-indikar/shared/projects/genefo...  False   False  False   \n",
       "1     /nfs/turbo/umms-indikar/shared/projects/genefo...  False   False  False   \n",
       "2     /nfs/turbo/umms-indikar/shared/projects/genefo...  False   False  False   \n",
       "3     /nfs/turbo/umms-indikar/shared/projects/genefo...  False   False  False   \n",
       "4     /nfs/turbo/umms-indikar/shared/projects/genefo...  False   False  False   \n",
       "...                                                 ...    ...     ...    ...   \n",
       "6320            2024-07-31_10-13-03_job_number_146.h5ad   True   False   True   \n",
       "6321            2024-07-31_10-13-03_job_number_146.h5ad   True   False   True   \n",
       "6322            2024-07-31_10-13-03_job_number_146.h5ad   True   False   True   \n",
       "6323            2024-07-31_10-13-03_job_number_146.h5ad   True   False   True   \n",
       "6324            2024-07-31_10-13-03_job_number_146.h5ad   True   False   True   \n",
       "\n",
       "       ETV6  source  MEF2C  RUNX3   FOSB  GATA2  IKZF1  GFI1B  target  \n",
       "0     False   False  False  False  False  False  False  False    True  \n",
       "1     False   False  False  False  False  False  False  False    True  \n",
       "2     False   False  False  False  False  False  False  False    True  \n",
       "3     False   False  False  False  False  False  False  False    True  \n",
       "4     False   False  False  False  False  False  False  False    True  \n",
       "...     ...     ...    ...    ...    ...    ...    ...    ...     ...  \n",
       "6320  False   False  False   True  False  False   True   True   False  \n",
       "6321  False   False  False   True  False  False   True   True   False  \n",
       "6322  False   False  False   True  False  False   True   True   False  \n",
       "6323  False   False  False   True  False  False   True   True   False  \n",
       "6324  False   False  False   True  False  False   True   True   False  \n",
       "\n",
       "[6375 rows x 14 columns]"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Combine HSC targets with file of reprogrammed embeddings, add meta info, and save\n",
    "fname = '/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/cluster_centroids.h5ad'\n",
    "adata_reprogrammed = sc.read_h5ad(fname)\n",
    "adata = ad.concat([adata_hsc, adata_fib, adata_reprogrammed])\n",
    "unique_genes = []\n",
    "for i in adata.obs['recipe'].values:\n",
    "    unique_genes += i.split(';')\n",
    "unique_genes = list(set(unique_genes))\n",
    "unique_genes\n",
    "\n",
    "for gene in unique_genes:\n",
    "    vals = []\n",
    "    for i in adata.obs['recipe'].values:\n",
    "        vals.append(gene in i)\n",
    "    adata.obs[gene] = vals\n",
    "adata.write('/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad')\n",
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd3373be-96db-49fd-83a1-c14f2a6cf00a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "0bf57559-ca88-4edb-b798-296bb72409d8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAnMAAAGtCAYAAAB9bPLMAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAADaDElEQVR4nOzddXSTZxvA4V88dXenpVhxhw3XjcGGjDE2xtwZc3f75u4Cg42xsY0JuuHu7pS21N019n5/pKQNTQ3qfa5zOIe8lidtk/fOI/ctkyRJQhAEQRAEQWiV5M3dAEEQBEEQBOHSiWBOEARBEAShFRPBnCAIgiAIQismgjlBEARBEIRWTARzgiAIgiAIrZgI5gRBEARBEFoxEcwJgiAIgiC0YiKYEwRBEARBaMVEMCcIgiAIgtCKiWBOEGxYuHAhMpmMuLi4BrvmqlWreOmllxrseo3txIkTvPTSSw36MxAEQRAangjmBMGGq6++mp07d+Ln59dg11y1ahUvv/xyg12vsZ04cYKXX35ZBHOCIAgtnLK5GyAILZGXlxdeXl7N3Yw6KS4uxt7evrmb0WD0ej0ymQylsvV8PLW134EgCK2L6JkTBBsuHmY9ePAgkyZNwtvbG41Gg7+/P1dffTWJiYl1ut7cuXP57LPPAJDJZJZ/F67/2WefMWzYMLy9vXFwcKB79+68/fbb6PV6q+uMGDGCqKgotmzZwpAhQ7C3t+e2224DIDExkenTp+Pk5ISrqyuzZ89m7969yGQyFi5caHWdffv2MXnyZNzd3dFqtfTu3Ztff/3V6vXPmDEDgJEjR1rae/F1anLs2DGmTJmCm5sbWq2WXr168cMPP1gds2nTJmQyGYsXL+bRRx8lICAAjUZDdHR0nZ5j2bJlDBw4EBcXF+zt7enQoYPl53FBfHw8N910k+V316VLF9577z1MJlOVdmzatMnq3Li4uCqve+7cuTg6OnL06FHGjRuHk5MTo0ePBqCsrIxXXnmFLl26oNVq8fDwYOTIkezYscNyviRJfP755/Tq1Qs7Ozvc3NyYPn06MTExdXrNgiAIF2s9X30FoZkUFRUxduxYwsLC+Oyzz/Dx8SE1NZWNGzdSUFBQp2s8//zzFBUV8dtvv7Fz507L9gvDuOfOnePGG28kLCwMtVrN4cOHef311zl16hTff/+91bVSUlK46aabeOKJJ3jjjTeQy+UUFRUxcuRIsrOzeeutt4iIiGDNmjXMnDmzSls2btzIhAkTGDhwIF9++SUuLi4sXbqUmTNnUlxczNy5c7n66qt54403eOaZZ/jss8/o06cPAOHh4XV6vadPn2bIkCF4e3vz8ccf4+HhwY8//sjcuXNJS0vjiSeesDr+6aefZvDgwXz55ZfI5XK8vb1rfY6dO3cyc+ZMZs6cyUsvvYRWq+X8+fNs2LDBckxGRgZDhgxBp9Px6quvEhoayooVK3jsscc4d+4cn3/+eZ1ez8V0Oh2TJ0/m7rvv5qmnnsJgMGAwGJg4cSJbt25l/vz5jBo1CoPBwK5du4iPj2fIkCEA3H333SxcuJB58+bx1ltvkZ2dzSuvvMKQIUM4fPgwPj4+l9QmQRDaMUkQhCoWLFggAVJsbKy0b98+CZD+/PPPy7rm/fffL9XlLWc0GiW9Xi8tWrRIUigUUnZ2tmXf8OHDJUBav3691TmfffaZBEirV6+22n733XdLgLRgwQLLts6dO0u9e/eW9Hq91bGTJk2S/Pz8JKPRKEmSJC1btkwCpI0bN9bzlUrSDTfcIGk0Gik+Pt5q+8SJEyV7e3spNzdXkiRJ2rhxowRIw4YNq/dzvPvuuxJguZYtTz31lARIu3fvttp+7733SjKZTDp9+rRVOy5+rbGxsVV+frfccosESN9//73VsYsWLZIA6Ztvvqm2PTt37pQA6b333rPanpCQINnZ2UlPPPFETS9ZEATBJjHMKgi1iIiIwM3NjSeffJIvv/ySEydONPhzHDx4kMmTJ+Ph4YFCoUClUjFnzhyMRiNnzpyxOtbNzY1Ro0ZZbdu8eTNOTk5MmDDBavusWbOsHkdHR3Pq1Clmz54NYOlRMhgMXHXVVaSkpHD69OnLfj0bNmxg9OjRBAUFWW2fO3cuxcXFVr2TANOmTav3c/Tv3x+A66+/nl9//ZWkpCSb7ejatSsDBgyo0g5Jkqx68err4javXr0arVZbZZi3shUrViCTybjpppusfva+vr707NmzyjCvIAhCXYhgThBq4eLiwubNm+nVqxfPPPMM3bp1w9/fnxdffLHKnLZLER8fz5VXXklSUhIfffQRW7duZe/evZY5diUlJVbH21phm5WVZXN47uJtaWlpADz22GOoVCqrf/fddx8AmZmZl/2asrKybLbT39/fsr+yS1k1PGzYMP78808MBgNz5swhMDCQqKgofv7550tuR13Z29vj7OxstS0jIwN/f3/k8uo/VtPS0pAkCR8fnyo//127djXIz14QhPZHzJkThDro3r07S5cuRZIkjhw5wsKFC3nllVews7Pjqaeeuqxr//nnnxQVFfHHH38QEhJi2X7o0CGbx8tksirbPDw82LNnT5XtqampVo89PT0B8xy1qVOn2rx+p06d6tr0anl4eJCSklJle3JyslU7LrD1mupiypQpTJkyhbKyMnbt2sWbb77JjTfeSGhoKIMHD65zO7RaLWBewFBZdcGVrfZ6eXmxbds2TCZTtQGdp6cnMpmMrVu3otFoquy3tU0QBKE2omdOEOpBJpPRs2dPPvjgA1xdXTlw4ECdz71wo764p+1CYFD5Ri5JEt98802drz18+HAKCgpYvXq11falS5daPe7UqRMdO3bk8OHD9OvXz+Y/JyenGttbF6NHj2bDhg2WoOmCRYsWYW9vz6BBg+p9zZpoNBqGDx/OW2+9BZiHrS+048SJE1V+T4sWLUImkzFy5EgAQkNDAThy5IjVcX///Xed2zBx4kRKS0trXPE7adIkJEkiKSnJ5s++e/fudX4+QRCEC0TPnCDUYsWKFXz++edce+21dOjQAUmS+OOPP8jNzWXs2LF1vs6FG/Vbb73FxIkTUSgU9OjRg7Fjx6JWq5k1axZPPPEEpaWlfPHFF+Tk5NT52rfccgsffPABN910E6+99hoRERGsXr2atWvXAlj1FH311VdMnDiR8ePHM3fuXAICAsjOzubkyZMcOHCAZcuWARAVFQXA119/jZOTE1qtlrCwMDw8PGptz4svvsiKFSsYOXIkL7zwAu7u7vz000+sXLmSt99+GxcXlzq/tuq88MILJCYmMnr0aAIDA8nNzeWjjz5CpVIxfPhwAB5++GEWLVrE1VdfzSuvvEJISAgrV67k888/59577yUyMhIAX19fxowZw5tvvombmxshISGsX7+eP/74o87tmTVrFgsWLOCee+7h9OnTjBw5EpPJxO7du+nSpQs33HADQ4cO5a677uLWW29l3759DBs2DAcHB1JSUti2bRvdu3fn3nvvveyfjSAI7Uxzrr4QhJaq8mrWU6dOSbNmzZLCw8MlOzs7ycXFRRowYIC0cOHCel2zrKxMuuOOOyQvLy9JJpNZri9JkvTPP/9IPXv2lLRarRQQECA9/vjj0urVq6ussBw+fLjUrVs3m9ePj4+Xpk6dKjk6OkpOTk7StGnTpFWrVkmA9Ndff1kde/jwYen666+XvL29JZVKJfn6+kqjRo2SvvzyS6vjPvzwQyksLExSKBRVVnXW5ujRo9I111wjubi4SGq1WurZs2eV8y+sIl22bFmdr3vBihUrpIkTJ0oBAQGSWq2WvL29pauuukraunWr1XHnz5+XbrzxRsnDw0NSqVRSp06dpHfeeceyaveClJQUafr06ZK7u7vk4uIi3XTTTZaVzBevZnVwcLDZppKSEumFF16QOnbsKKnVasnDw0MaNWqUtGPHDqvjvv/+e2ngwIGSg4ODZGdnJ4WHh0tz5syR9u3bV++fgyAIgkySJKkZY0lBEBrRG2+8wXPPPUd8fDyBgYHN3RxBEAShEYhhVkFoIz799FMAOnfujF6vZ8OGDXz88cfcdNNNIpATBEFow0QwJwiXyWQyWZWGsqUp6oza29vzwQcfEBcXR1lZGcHBwTz55JM899xzDfYckiRhNBprPEahUFzy6tQLDAZDjfvlcnmNKUAEQRDaEzHMKgiXae7cuVVqjl6srbzNFi5cyK233lrjMRs3bmTEiBGX9Ty1BYO33HJLverECoIgtGUimBOEyxQXF1drstd+/fo1UWsaV1ZWFrGxsTUe06lTJ0t6k0u1b9++Gvd7enpa0okIgiC0dyKYEwRBEARBaMXEpBNBEARBEIRWrFkWQJhMJpKTk3FycrrsidKCIAiCIDQNSZIoKCiotQ6x0LSaJZhLTk4mKCioOZ5aEARBEITLlJCQIFIetSDNEsxdmBydkJCAs7NzczRBEARBEIR6ys/PJygo6LIXOQkNq1mCuQtDq87OziKYEwRBEIRWRkyRalnEgLcgCIIgCEIrJoI5QRAEQRCEVkwEc4IgCIIgCK2YCOYEQRAEQRBaMRHMCYIgCIIgtGIimBMEQRAEQWjFRDAnCIIgCILQiolgThAEQRAEoRUTwZwgCIIgCEIrJoI5QRAEQRCEVkwEc4IgCIIgCK2YCOYEQRAEQRBaMWVzN0AQhNbPZJKIyyrCYJSY8/1u5HIZL17TjeGRXmhViuZuniAIQpsmgjlBEC6J0SRx74/72RObRbC7A0eS8lDJQW8y77978X56B7mw/P4rmrehgiAIbZwYZhUE4ZJsOZPOvyfSyC0xcCQpD6gI5C44lJBHdHphM7ROEASh/RDBnCAIl+RgfG6VbRcPqErAmPc38+vehKZokiAIQrskgjlBEC6Jl5O2yjYjMKKTF7KLtn+0/myTtEkQBKE9EsGcIAiX5KZBwcwZHIJaYR26HU/KY+Gt/XG3V1m2peeXAvDDjjhmfLmDvw8lsf98DqV6Y5O2WRAEoS0SCyAEQbgkMpmMV6ZEseVMBnFZxZbtGYU6vtoSw4EXxrFoZxw/7YrnxoHBlOiMvPTPcSQJDiXkojdKOGgUvHZtFNf1DmzGVyIIgtC6iZ45QWilSnRG9sZlU6Iz9241Vy/XS5O74aix/l54NCmP/edzmDM4lLUPD+OWIaFoVXKi/F0A0BslAIrKjDy27Agx6QVN3m5BEIS2QgRzgtBK3fz9bmZ8uZNR727kmeVH6fz8Gm5buBdJkpq0HSM6eTOll7/VtoJSA3ctsm6LTCbj5zsH8uKkLlbHGk0So9/fQlJOMSZT07ZdEAShLRDBnCC0MjlFOmZ8uYN9cTkApOSXsWR3PAAbTqWTX2qweV5usY5PN5xl46n0Bm/TPcPD6R7gbLUtq0jPrQv2WG17/LcjvLziZJXzJeCZ5UcJf3YVs7/dxe/7E5s8KBUEQWitxJw5QWhl/j2Ryt7yQO5ifi5aXOxUHIjP4YU/j6FRyXlgRASHEnP5anMMpQZzIrjHxkXywKiODdamIHd7Ft02kL6v/UflzrXTaRU55rKLdKw9nmrz/GB3OzafyQRge3QW26OzeHTZYR4a3ZGHx0Y2WDsFQRDaItEzJwitTGGZ7Z43MM+be3vNKWZ/u4tjyfnsP5/LrT/s46P10ZZADuDdf8/Q65V/OZGc12DtcnNQM7N/EABKuYwwT3tentzNsn/5wURsjaLeOiSEid39bF7zo/VnGfb2BhJzim3uFwRBEETPnCC0CvvPZ/PNlljCvBz4t5reLYCcYj2fbzpXp2vmFuu59rPt/HL3YApKDZgkE++sPYNGKefjWb0JdLO3Ov5Uaj4lOiO9g92qvebr13ZnYpQfYZ4OBLlbn++sVdk8Z86QMIp1Bv48mESJzkB+qfVCjvjsEr7ZEsPLU6Lq9LoEQRDaGxHMCUIr8OTvRxulLJbOKDH18x1c3GF21YebOfLyRMvj23/Yy/qT5rl2r18XxeyBITavJ5fLGBbpZXPf6C7eqBVydMaKHsLvb+lHmKcDADufGs27/57m260x6IzWLeof6l7flyYIgtBuiGBOEFqDGtYCOGuVNhc9yCqd5qCWMTjcCwe1ktmDgvnf6lMcS8pDZ5RsXjq/zETHZ1aiUSlYfPtASyAH8Pyfx/B20jA43LNKSpLKCssMmCSJvbHZOGqUhHk5YCxf1OCkVdLRy5Hvt8fy+G9HuKaHPz/tOW9JWaJWQHnGFTRKOZN6+lf3NIIgCO2eTGqGJWP5+fm4uLiQl5eHs7Nz7ScIQjt2MD6H6z7fUeMxGqWcskpz4oLc7JjWN4AP10UD8NSETtwzIsLqnJS8EtYeS+X7bTHE55RWe+05g0LIKdbxz5EUq+eTgO9v6c8VHT2rnHM8OY+ZX+2iVG/EUD5R7ovZffhxdzwJWUXE55TU+roDXLWk5ZcR4e3I7/cOwaGGwFEQhKYh7t8tk/h0FIQWzt/VDqVcZgmKLqZSyHhsXCS/7Etkep8A+oe5E+HlhIRETEYxSoWMmwaHVjnPz8WOuUPDCHSz545F+2xeWyaDa3r50z/Undeu1fHAzweJzSwisTwY23g63WYwtysmu8pCjSV74tkenVnn1y2TmV/zqdQCtr15DeOH9IVxr9X5fEEQhPZC9MwJQiuwJzaLG77eZXM1KMCSOwcyJLxqUFVXP+06z4ZTaaw/lWHZNntgMA+N6Yi3k9bq2NxiHff8uJ+8EgOfzOpNhLdjletlFpbx4JKDlOqNdAtwxt/VDpVcxuurTgEgl1Hta7mYH1n8pXkeb2UJPN/wOfIEQag7cf9umUQwJwitxNrjqbzyzwn8XbUMDHPHSavig//O0NnPiZ/uGNQgw5CHE3L4aP1ZBoZ5cMeVHVDIZQ3Q8grhT6/EWI9PHKUcFBj5SvU+IwYNhIlvNWh7BEGoH3H/bplEMCcIQpNIzStl0JvrL+lcJ60StULGlzf1pX+YRwO3TBCEuhL375ZJJA0WBKFJeDqqCfdysLnPTiXnzivDuGdYB4Z08KB7gAtjO3tzoV+woNRAVpGeOd/vIb9U33SNFgRBaAVEMCcIQpNQKuT89/Bwega6WLbJgBl9AyjRm8goKMPfzY7cEj1RAc5kF+uqpE0p0Zvo/cp/vLGqan1XQRCE9koMswqC0KT2n8/hlX+OY5QkHhzVkaf/OEp2kQ4AhYw6z6k7/MI4XOxtV5UQBKFxiPt3yyRSkwiC0KT6hrjx1wNXWB5Hpxfy5eZzTO0dwIH4HI4m5dd6DSeNEiet+PgSBEEA0TMnCEILci6jkEkfb6NEb7S5X6WQMbqzN09M6EwHr6opUQRBaFzi/t0yia+2giC0GOFejmx6fATJuSW8vuok++JyKu1zYOW8K9GqFM3YQkEQhJZHBHOCILQoPs5afJy1/HbPENLyS1m6J4Hugc6M6uzT3E0TBEFokUQwJwhCi+XjrOWhMR2buxmCIAgtmkhNIgiCIAiC0IqJYE4QBEEQBKEVE8GcIAiCIAhCKyaCOUEQBEEQhFZMBHOCIAiCIAitmAjmBEEQBEEQWjERzAmCIAiCILRiIpgTBEEQBEFoxUQwJwiCIAiC0IqJYE4QBEEQBKEVE8GcILRDaXkllOqNzd0MQRAEoQGI2qyC0E78ujeBt9ecIqtIhwTIADcHFVd39+fVa6Oau3mCIAjCJRLBnCC0E59ujCazSGd5LAHZRXoW7zpPYakBhULGC9d0xVmrshyz4VQacZnFzBoQjJ1a0QytFgRBEGojgjlBaCeGR3qyeFd8le1u9iqWH0oCINzLkXtHhLMvLpu7F+8nqzz4+9/qE3w8qw99gt3wdtY2absFQRCEmok5c4LQTozu4mNze06x3vJ/CYnPN0Zzw9e7LIEcgM4I9/x4gEFvrGfnuaxGb6sgCIJQdyKYE4R2YmCYByM6edV4zNtrTvPef2cwmCSb+03Aop1xDd84QRAE4ZKJYE4Q2gk7tYIHRkagkNV8nNEkIa/hmKTckoZtmCAIgnBZRDAnCO3InrhsjOWdbgq5DIVcRpC7XZXj7hsRjrKagG72wOBGbKEgCIJQXyKYE4R2ZEbfINwd1AD4OWvY9OgwCksNVY77Yed5ZDLb0dz0vkGN2kZBEAShfkQwJwjtiJeTBlX5OGtibil3LT5gtQDigm5+zuhtzJuTYe7REwRBEFoOEcwJQjuSnFNMZmGZ5fHJ1AK6+TtZAjwAZ62CFyd3w81ehau9igvZ5dwdVLx3fc8mbrEgCIJQG5FnThDakU82RmM0WW8bGObBVzf3ZfwHWynWGXl4bCf+Ppxs6bG7f0QEHX0cuaq7H2ql+P4nCILQ0ohgThDaEaWNIdJf9iUQk1lIkc5IsLsdNw8KYcvZDL6Wy1Ar5FzVw5du/i7N0FpBEAShLkQwJwiNqFhnYPmBJIySxMz+QWiUzVMSq0Rn5FxGIUMiqlaBKCozsvl0JgDx2SUsP5jIjH7BbH9yFCqFDA9HTXM0WRAEQagjEcwJQgMrMxjZdjYTJ42C2d/uRl8+rLknNptXp0ThVr6atCld/9VOjibl0TvYFTulnBKDuVFyGZgkc53WC/45ksKMfsH4uoiyXYIgCK2BCOYEoQEVlekZ+r8N5JZUTfex4kgKRxLz2PDocJSKirlnBaV65v18kKwiHe9O78n6U2nkFut5fHwnNp/J4GRqPjf0D8bzEnvIygxGjifnAXAwPheAmwYFM290Rwa8vr7K8ZO6+13S8wiCIAjNQwRzgtBADEYTA19fT6HOWO0xiTnFlOiNOFUK5tadTGPj6QwAxn24xbJ9xZFkknJLAdgdk83i2wdeUrs0SgWvXdudxTvjOJlaAJh74zwdNLjaq8gtX+hgr1ZwTQ9/ZvQTeeQEQRBaExHMCRbZRTqe+P0IxxJzifB25OMbeuMu5kvV2fHk/BoDufHdfBjfzRcnrcpqe1Gp7XMuBHIARxJzyS/R42ynsnlsbW4cGMyNA4P5ZksMCTnFPDS6I3K5jO1PjGT1sTSujPTEx1kMqwqCILRGIs+AYLFgeyzrTqSRml/Gtugsrv5kK5kFpRTrqg4ZClWFeztWW9O0m58TSrmcQR08rLYvP5DIS/8cr/XaeSUG+rz6LzvPZVV7zGcbo4l4ZhX3LzmAJFVN+Atw57AOvDIlyrKowUGrYnq/QBHICYIgtGIimBOQJIm7F+3j6y0xVttT8sro9/p6+r36H3tispupda2Ho0bJ93P746xVIJeZqyW42quYNSCI4ykFrDyawicbzlqd8/TyoxhsVFqwxWCCF/86xuA31/PZxugq+3/eE4/BJLHySAp5JVWrOgiCIAhtkxhmFUjMKWHtibRq9xfrTcz8eicPjYlg/phOTdiy1mdEJ2+OvDTBalt0eiHLDyZRqjfR2dfZsj2nSIfh4gy+tTiXUYhRgo/Xn+X+kRFW++YMDuGD/84yMcoXV3s1OoOJh385xJm0Av43rTt9Q9wv/YUJgiAILZYI5tq5Ur2RE+UrHWsiAR+uiybCy4lJPf0bv2FtSIS3I/89PJycYh09Al0B8xy4+386gMFGLOekUVBQVnUe3cx+gZQZTPx5KJmre1RdcXrXsHDuGhZuebwvLpuVR1MA+HpLDF/dLII5QRCEtkgEc+1YbGYRUz/fbrPQeqiHHTnFesr0JkorRRxfbj4ngrlLEORuT5C7veXx99tiScgpsXmsrUDO11nLq9d2R62U8+q1UVUWUdiSkF1s+X+kt9MltFoQBEFoDUQw147tic2qEsg5aZV4OKiJyyq2eY6dSkF6fineYsL8ZRndxYe/DiVT22w5f1ctH1zfi06+Tpa6qDUFcpIksXhnHCuPplhVmwjysK/2HEEQBKF1E8FcO1NUZuCeH/eTlFPCy1O6MbiDB1lFZczsH8TYLj4Eutkz+7vd1QZzhxJzGfDGeuaNjuCRsWL+3KW6pqc/XfyceGv1KbZGZ1Kqrzre+uqUbkzuFYBLPdKRLNwRx8v/nLA8DvWwp4ufM5NFb6ogCEKbJYK5diI9v5QfdsYBsPWsuQ7n8oNJ/HzXoCrHfjqrN0/+doRNZzKqrLTUG82Pv9h0jvtHRjRbrdG2IMLbiW9u6Q/AnO92s+VsBiBDiYGn+yu5eXBova958SrWuKxiUvJKSM4toYOXYwO0WhAEQWhpRGqSduLpP47y2cZzfLHpHD7OGlQKGeO6+to81sNRw7dz+3Ntr+p7c/RGic83RrPlTEZjNbldWXT7QH4dnsNNiv/4xf5dbh8eeUnXub5v1eoNZQaJcR9sobBM5AsUBEFoi0TPXDthpzb3oKkUcpbfNwRXezX26pp//QFuFfOsgt3siL9owv5H6825zl6/NorZg0IauMXtz4CJNzOgTx+wewKcL21Y9KbvdtvcbjBJlOmNOGrEW76lMxhNnM8uJsTdHoVcxppjqTholAyL9GrupgmC0EKJT/Y27NutMXzw3xnGR/ly44BgDEaJ2QOD8Xet22T4mwaFcDw5j4PxuVUCucqe/fMYEhI3DQptoJa3Yz7dLvnU/FI9MZlFNvcNj/Qku0jH99tjGdvVF4PRxLJ9iVzT05+8Ej2leiPX9Q5AXl0JC6FBpReU4mavRqWQozOY2BeXzeazGczqH8QTvx1hT1wOWqUcrUpBbvnQuaNGQbHOSIi7PTP6B3HfiIhankUQhPZCJlVX96cR5efn4+LiQl5eHs7OzrWfINTLjuhM5i09SGahzrJNKTdXEBge6cUPtw2o1/W6vrCG4hpqjgJEeDuw7pERl9JcoYGsOZbCPT8esLnv7mEd+GnXeQp1RhzUCrQqOVlF1vPrXp3S7ZLm6Ql1dyQxlyd/P8LJlALsVQp8XTTEZRVzYWqqSiHDYJRqXeUMEOnjSExGIQYTyGRwZYQn787oKVaaC41K3L9bJjFnrg16YMkBq0AOsCSnPZ1WwA1f7+SKtzaw81xmlXMNRhPGixY9DIv0wlGjZEi4R5XjL8gu0lc5T2g6v+5NqDaQA/hqSwyF5QF5kc5YJZAD+ONAkqjD28AyC8v4bGM0u2OyyCvWc9N3uzmZUgBAsd5ITGZFIAfmuah1fRedSSu0vK8lCbaczWTsB1uITi9o2BchCEKLJ4ZZ2xidwUSJrbIC5VLzSknNKwXgvp8OcMeVHZjRNxBvZy2P/3aYZfsS0ShljOniw964HDr7OrGlfPVrQk4x783oyUfrzxCfbT3sml2kI6uoDG8n0SvQHBqiFuvBhFwmfLCFLU+OaoAWCQCPLzvMxtMZqBQyuvo6kV9iHSwrMOBBPq5OjpwtUFsCOXuVnF6BLuyMzUECvJzUmExgNEmWYVdb8kr0TP18B0deGt94L0oQhBZH9My1MauPpVBSy5DoBTnFet5Ze5qbv9sDwO/7EwHz6seVR1NJLyizBHIACdklPLrsMBkFZVWu5aBRiECuGd0yJJTbhoZe9nUyCqv+boX62Xg6nd0xWQDIZeY5iHqjxOGk/CrHGlEyVH6cM5UCOTDXQ95RHsipFDL2PDOG/c+P5ZYhtS80yi81sMNGr7sgCG2XCObaCIPRxMLtsRxLqr3O6sVOpxXw+caz1HWUtMRGgtsZNlJiCE1HrZQzJNzzsq9jp1aIFCaX4Ze98dy6YC8zv97FljMZvDOjJzP7V//eUCBR7Fh7gNbl+TX0eGktafllPDgqgt5BrjUe/9u+RExi2oMgtBsimGsjFmyP46V/TvDN1lib+1+Z3A0fJ7Xl8YQoX7wqPf7jYPIlPa+jRs7eZ8fw0uRLX4UpNIzB4R74u15e72h2kZ7EHNvVP9qrwjIDb64+yZebz1HberH0/LJK/y/lhb+OsfxAImqFjJ6BLqgqrRa+sqMny+4byjN3z8FBXX3ybRkySg0m8ksNLN2bwKjO3rwyJQoPBzW+LloUNhYg/3EwiU1n0uv/YgVBaJXEnLk2QlFLSomfdseTXqBDjjmQu2lQCGuOpQIgA2b0DeTN1aeqPd/HSUOajeHVwjITRxJzGd3F53KaLzQAB42S/x4ezsA31lFYVreh9sqcNEqm9PKnk49TI7Su9Xr1n+P8ss88BeGbLTFEeNoT4evM9f2C2BWThZuDmuv7mXvf7riyA8V6I/YqBe6OalYcSbFcJzm3FFc7NRlF5vfRlR09cdYqKSozUlQ+NcLNXsXYLj5si84kJa8UH2cNqZUCRAeNggBXO7ydtex/fiyHE3KZ8tl2m+0OchP1eAWhvRCpSdoIo0li8qdbOZ5ct5VsHTwdiMsqwiTBjQODeWFSV/q9+p9lxePFgty0JOSU2tznoFFw/OUJl9x2oWHpjSaOJeVxLqOQXkGuOGtVbDmbQU6hntdXnwRgQKgbe+JysFfJuf3KDozo5E3fELdmbnnLFPbUSpsrTGUy8ypSgOt6+XNFRy+m9Q0EYMuZDOYu2FPt1AU/Fy12agUxGUXcfkUYSoWMg/G5PD2xM+//d8ZScs+WB0dFMLmnP15OGlzt1bz893GW7I7j4vh977Nj8HLSXMIrFoTqift3yyR65tqIrWcz6hzIAZbksgoZhHnYU1hm4Oe7BzHji52U2lgNW10gBxAhan62KCqFnN7BbvQOrgjOppfPaRwU7kGZwUi/UHdiMgpx0qrEDb8GBqOp2lQhlb8GLz+UzPJDyfi5aBkS4cnZ9EJLIGevVlTJ05hZUMqFqaebTqex/tGRln21LWD653ASn2yIxsNBzaqHriTS14k7hoXz2cZzlmOctErc7FV1fp2CILRuYs5cG1FqY1FCXRgleH3VKa79bDuejho2PzGC7gHO2KlkBLvZno9TeVMnHyd+unPQpTVaaHLdA13oF+oOQAcvRxHI1UKpkPP81V2q3a++6A1yKCGXDafS2Hkuk65+zkyI8sXdQV3lvMpv1/NZxUz7YoclZVBn35qHueOyzGmBsop0fLs1xlJ3eUgHDyK9HblvRDjbnxyFUiE+3gWhvRDDrG3IpxvO8u6/Zy7pXLkMTBL0DnYlMbuYjPKkw69f241n/zxudawMkDDPsXpreg+u6u53mS0XhJZte3Qmr644gUIu44nxnblj0V70RgkHtcIy3+2CC+8lAA8HlSVBs7+LluS86nu4542K4JFxnTiWlMukT6znwTlpFBRUGkcdEOqGv6sdsRmFlpQnj43vxAMjRYkvoXGJ+3fLJL66tSHjuvnWeoxCBr7OVXtjHNTmEfeD8bm42Ff0JGQV6aoceyH6H9HZWwRyQrswNMKTNfOHsXLelQzv5MWSOwcxe2AwC28bwCNjI62OrTxP7kIgJ5NRYwUVAGc787BoVIAra+cPY9aAIGSYvzxVnvrg56Jlb1wOfx5K5kil3HXjunhf3osUBKHVEsFcG1JWh6FWGbDiwSsYEGo92b1nkAuRPo48MDKCj2/oTbC7HY4aJSuOpODuYHvuTZCbXUM0WxBanf6h7rx+XXf6h7pzXe8AnDTm1CIhHtYrSC+kHOni68RvB5Ks9ikv+vQNcXew/L+TrxPhXo5ImL886Y3mCFEhk+Fip7J8oZKAHgEufDOnH5G+opdEENorEcy1Id0DXbihf6Dlsd3Fdwvg5sEheDpp+enOQbw6pSI3XLiXI/8+PJzHxneiq78ztw0No7DMwJm0QkZEetv8QykoFcllBSHI3Z6tT47iu1v6VVnoMDjcndemdONEStXFSRevM3ruzyNWj28cGMxNg4K5rncA4Z4OuNop6RXkwvBIL6vFDWqFnLFdRWogQWjPxGrWNuaVKd3RKBX8vCfeZo3WkZ3NH/oqhZzZA0M4l1FISl4pj4/vbHXc0AhPXO1V6AwmZg8K5s5hHViwPZaVR1Isc4RGi2EdQQDA1V7NJxvOVil1t+5kBmn5Vacq2JJ30Zcje7WS167tbnnc5fk17I/P5VRqARsfG8G9Px3gUEIu0/sFXnwpQRDaGRHMtTFqpZyXp0SxZE+8zf3zlx7Ew1HDFzf14c+DySzccR4AmewQX97Uz3JcRx8ndj8zmtS8Uub9fJBTqQWUVQoOZTII83Socn1BaK90BttryY4m5dEz0IXDiVVL7bnZqxge6cWe2GzmDAkhNa8UX5eKKh4rj6YgSRKTevjT2c+Jg/G5FOmMDPnfen66YxD9Qt1rTRguCELbJ4ZZ26h7R4Qjl5nntSkrfdhnF+s5m17Iop1xpOSXWLafz6xawkmjVPD4siMcTsyzCuQAPOzVhHiIYE4QLrh7eAerxxfedh29HZna23bvWU6xnuPJ+bjYqfjf6tMMfWsDG0+nIUkSb685xf0/HeCBJQd5bvlRovwr5sQZTPDBf6dFICcIAiBSk7QLJ1Py+WVvAkv3xFtWxfUOcuH2K8P4aVc8J1ML8HXWMjTCg6HhHuyMyeZ4cj7PXt2FB5cctCQYruy2K8J4YVLXpn4pgtCiJeWW8PnGaNadTCOtvAzXhVQlgzq4ozOY8HPRsvJoao3XqZzepDrT+wby7oyeDdV0QagTcf9umUQw146M+2AzZ9IKAfNKOoMJHDWKaut4Dgp1YVfchaEhCZDhaqdiWt9AHh/fCa2q+uLggtDefbHpHMv2xxOTYd3rXZdAzRaNUm7VQ/7O9B7MKK8JKwhNRdy/WyYxzNqOLL1zEKM6ezG+m49lJZ2tQO7CyE1RYQEVWeVkONsp2fz4SJ6f1FUEcoJQi3tHhLP+kRHcNawDQ8M9cCxPX2KSQFmP0dELueb8XLRoKp2YbSMHpCAI7ZPomWunXv3nON9tj6vxmIt7EO4ZHs5TEztXf4IgCNX670Qq9y85iIudihB3e/adz7mk6/QPcaNPiBsPj40UX6qEJifu3y2TCObasVVHk1l7PI2+Ia689PeJaod+tCo5X9zUh5GdRC4rQbgcpXojSrmM2Mwipn+5k7wSvc3jOno7MKVXAGuOpXIsOd9q3/dz+zGqs3gvCs1D3L9bJhHMCQD8fTiZT9afpbBUT0r5xG1/Vy3p+WUYTBJyGTw6rhP3t6Daj4t3xvHr3ngcNCqMkkQHTweeuborLna2K1YIQkuyNy6b2xbsxShJVsmGewe58Pu9Q5HLZSw/mMiTvx/F3V6Ns52Su4eFM62vyCsnNB9x/26ZRDAnVJGUU0Jafgk9Al0Z/s4mknLNKUxUChnzRnVkXDcfOjVD6aD1J9OIzyomOqOQnGIdq2ysCJTJYMHc/ozoJBIaCy2fJEkYTBI7zmXhaq/Cw0FNoJt9lWNkMpGCRGgZxP27ZRLBnFCjhOxiHlhygMOJeShkUF4iki9m92Fid78ma8cfBxJ55NfDdTq2k48jv949hJVHU+gX6kakj1Mjt04QBKF9EPfvlkkEc0KdpOaVcMVbGyyrYEd08mLhrQMa9TkzC8swmSQMRhPD3tmEoY75HEZ38SanSM+B+BycNEp2PD0KJ60YehUEQbhc4v7dMolyXkKd+LrYMbN/MD/tNpcJC/dybNTnO5SQy8yvdmKSJKb1CagSyMmoSJpysfUn0y3/LzOYMBib/PuKIAiCIDQZEcwJdfbkxM4UlRkwSTBvdMdGfa7dMVmWBKlL9yZW2X9xeKZUyKoEbfZqBV/e1Bc3BzUpeSWsPprKsEgvIrwdWXEkmfNZxdwyJBRHTd3eBmLukiAIgtASiWBOqDNnrYoPb+jdJM/19+Hkeh3fJ9iVOYND2XYmg9isYiSTxB3DOjAs0guA2xbu42RKPh4Oar66uS8PLDkIQFp+Ka9MibJ5zc83RbPmaCoxmYWW5MpKuYwuvk7E5xTjqFHx6LhIhkV64emouYxXKwiCIAiXTgRzQotjNEmcSsmv9Tg3exVKGeSWGrjrynDGdPVh29lMdsdm083fmdFdKnJxlRnMwZjOaEJZqTh5Yk6JzWsnZhfz9prTVbYbTBJHy/N+5ZUYLIsy5o3uyCNjI+v+IgVBEAShgYhgTmhxFHIZr0/tztO/H612XpydSs6KeVcS4GqHwWhCqTBXptsVkwXA8eR8Ckr1uNqrAfhmTj/+OpjE6C4+9AxypV+oG/victhwKp2x72+mZ5ArD43uSJC7OS3Ea6tO2nxepVyG0SRVadfnG6O5aVAw3k7ay379giAIglAfojar0CLd0D+Yj27oRScfR2b0CUClsJ6rdtewcAJc7QAsgRzAM1d1oWeQK09M6GQJ5MC8YOORcZ3oGeQKQJiHg2Xf2fRCftufyKsrTli27YnNstmut6f1YMsTI7l3WAer7QaTRExG0aW9WEEQBEG4DKJnTmixJvcKYFw3X/46lMS0foE8tPQQafllBNgZGRlgsjr2YHwOd/ywDxd7FT/fOQgf54oespMp+Xg5aazmtb16bRQleiMrjqRYtinkMmIyCvlkw1myi2yXWTqSmEuPIBe8nKvOkft8YzQpuSWoFHISckqYMzgEhzourmiJotMLWHkklfFRPnRuhiTRgiAIQt2IPHNCi/bs8qP8tDsepVzGX/cPZd6Xf3NO727Z72KnYtFtA/jjQCI/7DwPwJMTIhndxZdIHyfeWXuKzzaew1mrZO3Dw/BzMffmvbHqJF9viaGDpwPx2cUYTBIeDmqyinTVtkWjkBHm5cip1AI8HdVkFlZ/LMDcIaG8NLlbA/wUmsb6k2k426mQy+CuRfstPwtPRw37nhvTzK0TBKElEPfvlqn1dhsI7cKFmpUGkzktiMlo3SOXV6Ln5u92ozMYLNveWnOGt9acsTouv9TA1M938M2cfoR7ObL1bCYAMZlF+LtoSc4rRaGoOe3I4HAPNp0xn2cySSjlshoTGVdeaNHSfbs1htdW2p4nmFtcRmGpnrfXniYlr5T7R3Zg9bE0ovxduKanfxO3VBAEQbiY6JkTWrScIh3fbYulk68T1/T058yx/Tz112kOFLhc8jW9nTQ8NbEz322LZVikF7cNDWXlkRTeWnOKEr2pyvEf3tCT5JxSroz0ZHdMNn8fTmZYR0+W7I4nu9g8HKtRyi158QC6+Dqx/P6haFWKS25nY8sqLOOzTdGsPZ6G3mAkvaD6nkalHAxVfzT89/AwOopyaYLQboj7d8skeuaEFs3NQc1j4ztZHkdG9eWPqL6EPbWy2pWutUkvKMPNQc3KeVcC8NTvR1i6N6Ha48v0Jj7fdI63157mxWu6kllQxqcbz1kfc1Gk0zPItcUGcm+sOsHKIylkFZZRaqjbT9FWIAe02NcoCILQnohgTmiVZDK4lD7lIDc7egS5MiTcw7Jt/am0ao93d1CjUsgpLDMP4y7cHkdyXmmtzzN7YHD9G9fITqfmM+2LHZYEyJVpVXJevzaK9SfT6ervTEZBKf+dSCM5r6za67nZKTHWsV6uIAiC0HhEMCe0SndcEcbXW2PrdU6Iuz1rHx5m1ZuUXaQjt5pFD2GeDmx4dDh6o8SumCwSc0rYcc46ZYlcBhfHM0FudnQPdK1X25rCm6tP2Qzk5MD0voFM6xvEtL5Blu13XBnOyHc3VZkX2MHLgaTsYnJKDIx4dxM3Dwrh1WttV9EQBEEQGp/IMye0Sk9M6IxdPYb4IrwdWf/ocKtALr9Uz4QPt2BjmhwANw8KQSaToVbKeXt6T67rHVDlmMpxjpeTBmet0mpYuCUJdrH93c0E/Lgrnr1x2Uz6ZCs3fL2TnCIdQe72rHjwCqb3CWD2wGCu7eXPc1d1wd1eRVmlOrjbz2U20SsQBEEQbBE9c0KrpFTI+fP+oUz4aEutw63ju/rw1vQeVsmFAc6lF5JeUP0wYlqB9XBqR2/HGp/n6YmdmdonsObGNKOnSz9iCddhQFVln6ejmtXHUjiWZC5VtuZ4Kl38nPlheywBrvZ0D3Ih3MuRqz/eapkfKAO8nTW4aJWM/2AzcwaHMntQSFO+JEEQBAERzAmtWCdfJxzUSst8NlvcHdR8cmMf1ErrQO6vQ0k8/MshZFDtQortZzP51TMBL2cNIzt50z3QlS6+TpxMLbB5fLhXzcFec7MzFnBQfTcLTRPpOOUJRvfpzIHzOZxOK2Bqn0COJOTy0654tCoFq46m8PxfxzBU6oGbGOVrtdBDAjILykjLNwfEz/55jJ5BrkQFXPpKY0EQBKH+RGoSoVVLzCnmf6tP0dnXiY/Xn0VXHnz4uWiZ2ieQqX0CbAZZDy09yF+Hkuv8PBO6+rAvPgckicxK1SG0SjkTu/vSP9SDG1vgogcr+cmw91sIHgwdx9o8pLDMwPoTaTz0y6Eq++4aFsa6k2mczyzGKGEzz56Hg5q1Dw+zqrYhCELbIe7fLZPomRNatUA3ez69sQ8AU/sE8tbqU/QNcWPOkNAaz7v9ijBOpRZwPquI0vJJcxcvZvBx1lh6ndacsL3itdRgYv6YSEIq1XptsZz9YfQLNR7iqFHSJ8QNdwc1OUU6vJw0yGQwMcqPR8Z24nhyPjEZxWiUcu4a1oFPNkRbnZ9VpGP9yTRm9m/hga0gCEIbIoI5oc3wd7Xjo1m963Rsj0BX1s4fxt64bO5atA8vJw3zRnXkgZ8PWo6Z1ieA77fHWYI9W1zsVLjYVZ2D1poFuduz9YmRFOuMeDmZe9j2n89m1LubLCW+DCYJhQwUcqhclMNOpWBIuGdzNFsQBKHdEsGc0K71D3Xn4AvjLI/1JhMf/ncWdwcV47r58fmmmCrnyGQwe0Awg8M96Bnkiqu9uimb3CQcNEocNOY8cncv3se6k+mWfReGVz9cb90rp5TBrmdGt7ngtjVYti+BzzedY0ovf+aPiWzu5giC0MREMCcIlVzXO5DrelesSH1yQme2ns1g/phIgt3t2XImgz4hrkR4t48SVqdTC6wCOaDaRMEdvB1FINcMSvVGHv/tCAAfrjvLieQ8pvYJYkKUbzO3TGgWcTthyfXg3QXm/Alq++ZukdAExAIIQRCqVVCip9cr/2Ks4VPCUS1nQnd/bh0aSjd/sZK1qX21+Rxvrj5ltU2lkHHs5fFolKLcWrvy+91wdGnF4763wTUfNOhTiPt3yyR65gRBqNay/Qk1BnIAz03qxg0DxIKH5lBQqudUSn6V7TJk/Hs8jWt6+jdDq4Rm8ao3GC/Km6lpHyMIggjmBEGoga1h06m9/bl/ZATRGUUoZDLGdPVphpYJ57OKuOqjrRTpqpZo0xlNPLHsIGn5pcwdElolYbbQhphM8FGPqoGcXA3jXmmeNglNTrzDBUGo1rS+QXQPcEEuMw/dRXg78vRVXQn3dmJ8N18RyDWjQwm5NgO5C0oM8NrKk4x8bxPpF1UzEdqQo8sgL8F6W7+74IWM5mmP0CxEz1wbJkkSWUU6jiTk8sLfx+nm78xnN/YR39KFevnnwSuauwmCDb0CXeng6UBMZlGNxyVklzDkzfUsuXMwA8Lcm6h1QpPx6QaWWjZKuH0VBA1s5kYJTU0Ec23Ygz8fZMWRFMvjxJwSIp5djVxmTi9x6xVhPD2xSzO2UBCES7FoZxwv/HUcTwe1VSWOcC8H0vJLKdYZrRJgG0xwID5HBHNtkW8U3L8HsqKh81XN3RqhmYgumjZszfFUm9tNEuiMEl9tjmHrGdEV39KtO5nGG6tO8sf+hGrTggjty+bT5vdtZpGOz2f3YWpvf3oEuDBvVEeUcjkX/5koZXB9v6BmaKnQJLwiRSDXzomeuTYs0NWOuKziGo+5+fs9XNvLnw9vqFvlBKFpfbYxmnfWnrY8fnP1KTY8NoK1x9MI83Sgb4hbM7ZOaC73jQwnJa+UAFc7Np5KJzaziCNJeTy9/CjFNubRvXJtFO4ObS+5tSAIZiKYa6P+PZ5aayB3wYZT6bUfJDSLXTFZVo8zCnV0f+lfwFxL9t+Hh7WbBMZ1IUkSG0+n4+2kJcTDnsJSA6fTChjW0Qu5XNbczWswfUPcWfXQlUz6ZCv/nayoG2wrkAN4d+0Zuvq70CvItYlaKAhCUxLBXBulqrTIwVmrIL/U/CEf4m6PhER8doll/9Xd/RjxzkZ8nbV8dXM/XOxFFv+WYlikF1vPZtrcZ5Jg6mfbCfJwYNHtA/Bw0DRx61qezzeds+rJvCDQ1Y6tT45EJms7AV16QSlejjX/zi9Mi88u1vHL3gQRzAlCGyXmzLVRIzt78+VNffjf1O50C3AFINLHic1PjOTNqd0tx/k6a/jnSDJxWcXsis1m7Qnb8+yE5rHuRFqN+/PLjBxPzmfekoNN1KLmZzCauP+nA1zx1gY2nq7oVd4RnclH687aPCcxt4StZzPJKCjDYDQ1VVMb1IViPaV6I3MX7GHA6+spKDVw17AOVY6VAaM6e7PsnsEEuNqhVcm5qrso7yUIbZUo59UOdH1hDcU6I2qFnGGRnla1Nkd39mL9qYpFEJ18Hbl7WDhT+wTaupTQREwmibfXnmbLmQxOpOSjkIG3s4a0/LIqk9svUClkRHg58sTEzozs5N20DW4E606ksfxgEo5aJWl5pRxKyMXLScNzk7pyy/d7AHOP27+PDOPf42m88Ncx8ksNtV63Z6ALy+4ZTInO1Gp6od9cfZJvt8YyrU8A2UV61lUaWp3eJ5DfDiRaHd8ryJU/7x8KmINAo0kSKYmEBiHu3y2TCObagbsW7WPz6XQMku0i6bcMDmHRzvNU3rPnmdF4O2ubrI3/nUjDQaNgSLhnkz1nS/bu2lN8uvFcle33jgjni00V27UqGaV6699pkLsdW58Y1ehtbCxrjqUwb+lBdAbbH0039Ati+aFEysr3XxHhybZo20PR1ekX4sq+87l0D3Dmf1N70C2gZdeU7fHSWvJLDchk0MHTgXMZFbnlvpjdhw/WnSElr5SC8mD23uEdeGhMJFqVgowCc2UALycxDC9cPnH/bpnEnLk2Lq9Ez7+1DNUFudmjqJSrCuCdtad4Z0avy3ruUr2RjIIygtztAXMg+diyw+w/n8Nj4yMJcXfAw17BP78t5K1Y81BRryAXDifk0dXPmd/vG4JW1T4Lhf+0O77KNqUcHhgZwd7YLPadzwWoEsgBeNcyj6ql+2ZrTLWBHMD6U2mWQC7QzY6swrJqj61MLjPPM5SB5ed3NCmfyZ9uZ+aAQI4l5fPE+M5c0bFlfaE4lZpP9wAX9sZlE+LhQK9gV3yctSTkFOOkUZKcV8Kah4axOzab2xbuRSGX8cXmGH7Zm0CIhwNHkvJQyGX8ePtAkWdOENoo0TPXxq04ksyjvx6mzGA9T8jdXkV2sR6AmwYG8dPuBC7+Q1hyx0CGRFzaja1EZ+TqT7YSk1HELYNDWHEkhewiXZXnAJBjxETVoG3WgCDenNrjkp6/tev+4hoKyqquTHTSKCkoq3kocUoPH16+tgdJuSV082/ZPU6V7T+fzQf/nSUlr8Sq52lClC/rTqRy0Z8wTlolD4+J5JUVJ2q9tlYpp3ewGzsvWh18sQj7YtbNcoWOYy/lJTS4Yp2BgW+st/S4XdDJx5HEnBJLOa9Rnb35fm5/ygxG7l28nw2nq+aPdNIomBDli0qp4IGREfi72jXJaxDaFnH/bpnEJIo2TJIkHlp6qEogB5BdrEcGqOQyRnT25r6R4ThoKgIqhQxOpuSTmF239CYXS84rIab8hvzb/kSyqgnkAJuBHMCfB5Mu6bnbgt7BtvPH1RbIAaw9kcHY97dw9cfbuPrjrehb+IT/Ep2RzIJS7v/pANuiM60COYUM3prWvUr6FRc7FS9M6srmWpJeX8hGEublwNQ+AYzu7GXejrmn82LnijXsWPwS5DX/397JlHxWHU21+fs7nVZoVZd1y5l0Oj67inlLDnJNT9sLHQrKjCzbn8SS3fHc99MBkYBaENoQ0TPXxoU/s5Ka7uWzBgRzJq2A/edzUCtk6IzmPwcPBzVZRToAuvg6kVOsY1C4B89f3ZVinZGvt8TQN8SNq7r7sTMmiy5+Tng7mefYHUnM5b6fDpCYU1Lt84L5Rtsj0JW8ohJis6sOlQ0N9+CnOwdd4itv3dLzS5n0yTbSC+o2hFiTUZ28+P7WAQ3QqoaXmlfKxI+2kFPeS2zLo+MiWbrnPMm55p+FBEzp5U9KXil7YrOrHB/p40hMRpFl2oCPswY/FzsOJeSiVclRyWU2ez0r+2l2JEO7d7z0F3aZ4rOKGfXexiq9kXXh56wmJV9X63EPjIzgsfGdLqF1Qnsm7t8tk+iZa+MGhXnUuH9YR0+OJuUCWAK5+0dEWAI5gJOpBaTml/HnwWSGvrWBx387zOJd53n4l0NM+mQrt3y/h2s+2UZJeU/Bpxuiaw3kwNwreHV3X/6eN5wPZ/Yiys/6g6FyD0174+2sZXRn2ytSu/rVL0nwhtMZPPnbIYrq0KvX1Lafy6wxkJPJ4HRKPkm5ZUiYFzs8d3UXBnZwtxnIAZxJK7Sa/5mWX8ahhFwASvWmWgM5gBPZzfvR+NehpEsK5ABSKwVyNWXVyymuPeATBKF1EMFcG/f9rf0Z19WHIeEeXHnR/DdvJw0Tu/tx+xVhVtuVNaw5KNWbSM4tBcw9JGfSCgHzDfOn3XEUlRmIquPKwDKjxOurTnH/koNM6eXP0nsGY6euePLU/FJOJOfV6VptUWD5wpGLTe0TiJu9CkWlO7WiluoGv+xL4rWVxxuyeZctv0TPmytPVrvf20nD1d39WHE0FbkM/Fy1zB/bkX6h7jz3x7Far+9qd+nru85lNu8XieW1TDHwcdJwTU8/y+MQd3vCPR1w0CispjNUN+zi46wh1MOBnedqnkMoCELrIFaztnEapYKv5/QDzJOp/zqUxLGkfIrLDNxaHsQ9Mb4zOoPEv8dTGRrhydQ+gXy3LY7CanpyIr0dibcxl+61lad4beWpatuiAGz1icRmFHL9VzvZG5eD9qKJTHMX7GXPs2Pq9mLrqURnoFhnxKOFrv68b0Q4G06ls/98jmWbg9o8if2OKzuw/3wOb605haejmlcmR5GUU8LTy49wLr2QMmPV23h8Hcu7NYXt0ZnM+X63ZQrAhUoFlQ3u4GGZI2iS4MGREfQNcefRXw9Rl06r3BLrv9/ewa4cTsitNk9fZaX62nvvGsvumCxiqgkmtUo5KoWcB0ZFcPPgUB4cVcDc7/dwPrsYrUrOvcPD+Wj9WVztVWQXmXs8tSo5pXrrn1hafhmvrzqJQi5j7fwrRUk4QWjlRDDXjtirlcwaEFJlu0wm4/lJXXl+UlfLtrUPDyM5t4SCUj3LDybxz+EUAILc7Jg9KIR1l1DP9Y2p3fl0UzQJ2SV4OKgwmCTySgwk5JSQUD4sW3rR2JKykeppPvrrIX4/YO79GBHpxYJb+7e4Uk8ymYyXJ3fjhb+OEenjxPS+gQS42eHnYl6F2DfEjV/vHmw53tNJw6qHhpGQXcy1n223GioHWtTK4LXHU63mcl4cX43s5MWxpDxiKwU1X24+x9GkPMvvrT6C3LQ8e1VnHlxykJT8qvMQx3bxJjmvlOPJ+fQOcuWFa7rV+zkaiq0FSwDhng4svG2AJdUPgKu9irwSc9BWqjex6mgqAa5aksp7z8Gcl250F2++3hJjSelygdEkVft8giC0HmIBhFCrLWcyeHXFCaLTC1EqZNwyOBQ3exWfbTqHn4uWjj6O/Hc8HWOlPyW1Us61vQLoHezK/vM5zOgbwMAOnpQZjGQV6vB3tWPezwf5+3Byjc/9w639Gd4I1Qw6PL3SqofGy1HNnmfHtLiA7lKV6Iyk5JWwNy6bKH9nInyc0NQ0ft7EjiXlccPXOymsZv7aD7f255YFewHz35LuMgMOhQxsdFYC4OWoYu9z49AbTWQV6vB1abpk2dX5bX8ihxNyWH4wyfIzUitkOKgV6IwSdw7rwKQe/kz8cAt6k4QcqvRWBrnZWb4k2asVFOuMOGmVOGuVVsHe3cM78NDojtirxXd7oXbi/t0yiXevUKObvt1tya4vB/RGiW+3xTK5px/HXx5vCX5+3ZfA4p3nmd43gJn9g62S/c4aEGz5v0apsOS3enVKFJ39nPhqc4yld6Gyd2f0bJRADsBRrSS/0jByRqGOfw4nM7lXQKM8X1OzUyvo4OVIBy/H5m6KTVEBLrg7aCgsqzr066BW0CPQlUEd3NkVk43OYGJMF2+rMnRgLst106AQdsdm89v+xCrXqcxWIGenUmCnkvPG1J4AqBTyFhHIAUzvG8jU3gG4Oqj5ZH00YF6gpCsfOv5w3VlWHU1GX/6NpHIgF+CqxSTBM1d14aGlB9EZJcviJHuVgou/v3+1OQaNQs4j48TKVkForcQCCKFaBqPJqkxS5RvG34dTOJdexLGkPIwmiev7BfHPg1dwy5CwWqs2JGQX8/6/pzmTXsB9IyIsN5fKfWIvTurC9L6NVx82KqDqHKHK9S2PJeXxy9548or17D+fw5pjKZhEXq4G9dCYjlzcEaqQwcJb++PmoGbeqIrUILnFel6e0g2v8vmNfi4a/rx/KDP6BXFFRMWK7SC3ikS4XXydrBaJXODnrOHx8ZEcf3k8B14Yx9iuPg37whrAHwcSuWXBHksgZ8v5LOsV47MHBvPo2Ei+n9uf92b0RK2sSDUkYf7ZjI/yITmv6jDzl5vPsWxfQoO+BkEQmo4YZm1lErOLWbo3nuiMIm4eFMLQS6zQUBdZhWX0fW1dtfvDvcw1Iqf2DuD9mb3qdM1jSXk88uthzqQVoFXJ2ffcWNYeS2XRzjim9w1kfDdfHLXKRh/yySoso99r66rM1Zo7JJTx3XyY9c1uoKIEFMDAMHd+qTRHTbg8RWUGer/6X5Uh1PljOvLQ6I6cTS9k3pIDnCpfMS2XgbeTltT8iiFC5UVl6CovpFDIYP/zY9l0OoPo9EL8XLQ4aJRM6eXfoofT/zqUxENLD9V63FVRvqw6lgqYX+umx0ciSTD+wy2UXOICjnWPDCfCu2X25gotg7h/t0ximLUVue7z7RyMz7U83heXzb7nGq/sUE3Z9dVKuSUP3L5Kqy3BPKn6XEYhIR72VvO01h5P5e7F+y2PVQo5chlM6xtI/1B3Hvz5ACuPpvD57L7Yqxv4xVzEw1HDW9O688TvR622Lz+YSEGlId/KnXG7Y7N5dvlRbhkcSqSvWP13uRw0Sr6+uS8bTqVzMiWfvXE5yIBdMVkcSchjw+l0q95ak4RVIAdYBXJgvZBCo1LgYqfi2t6ta+i8rl+vVQoZe58ZzdboTIZGeOLjrGV3TNYlB3IAh+NzRTAnCK2QCOZaiQ/XnbEK5ABC3B0a9TkHh3tYVYKorHJvSkil1XV6o4kHlhxg7fE0fJw13DQwhJ5BLgyL9OZcRqHluAndfHloTMWk66V74zmcaM4pt/JIMjcPDm2kV1WhZ5ArjhqF1ST8wlIDPjXMm/ppdzw/7Y7npYnhzO2QD369QNnIkWcbNqKTNyPK50Wm5pUy9oPN7IqpSAZ8OcMG39zct0X3wFVnQJhblR5HWzadyeSjWVqm9qmYjjCwgwcudkrySmpOEG2vklOsr7qoZMGOWLxcNAwK80Btq96ZIAgtkgjmWokzaQVVtn1+U+9GfU4/Fzv2Pz8Wk0nidFo+85ce4nRaYZXjsorKOBCfg4eDmmlf7CCz0Bz8peWX8d5/ZwDo4ueEq52KEZFeBLnb88xVXawSBF/Z0YvvtsWiUcoZUEvViobyyYboKqspjRJ8sekcMlnNPSQvrY7mazL5NGIxJ6MeY+GOWM5nFWOvVvL7vUMIF70b9ebrosXXRUtBadW/sfroE+TKM1d3oV+oewO1rGkcSsjl4V8OoVbILwrkJDraFRFT6mS1YjyvRM+xpLwqSbqHRnhaUpToDCYcNEocNEpSckvILq+2YSuQAziWnM+c7/bg6ajm7weusCxWEgShZRNz5lqJwwm5TPlsu+VxN39nVjx4RbP0PDz9+xF+3puADPOqyWKdEUeNkmeu6sQzy2uvMvDMxM7cNTy8yvaCUj0KuazJUiR8vy2WV1acsJmw9mKudiqKdAb0Fy2LdKCUIqx78hw1CrY+OQq3xh4rboP+t+okX26JuaxrLLp9AMM6ejVQi5rOgNfX1VqLVwZEeDtyNt0c8N41rAPPXNXF6hi90cTRpDwifZxw1JjfS/mlega+vo4SG0Gcn4sGkwnSC8qs3gcf39CrzazuFhqOuH+3TKIfvZXo6u9Mx/LengndfPnz/qHNNoS0q7wmpgT4OJsDmcIyA38dSkGtqP1P6q9DtpO+OmlVTZrr6rYrwvjv4WFMKi+LNDDMndkDg+kZaN3T0dHHkdXzr+TkKxOY0TcQrbIi/Ls4kAMoLDPS/7V17D8vSiXV1+xBIfQMciXY3R7nSyzHFdgKe5PGf7C51kAOzH91j4yLtDw+mZJf5RiVQk6fYDdLIAdQUGqwBHLqSkt8Q9zt+fXuIex+dgyfze5jtfp3SETT9JALgnD5RM9cK1JUZiAuq4guvs7IG6kyQl18sv4s7/13hv4hrpxKyadAZ75J1KWHC6CzrxNr5g9r1DbWhyRJpOWX4eWkQSGXYTCauH/JQdYeT7UcMyLSi71x2ZTqTVZDXTWRAT/fNYhBHcRN8VL9cziZ11eeILNQZzX0eGFOWaiHPQaTicQc88KIQWHuLG1lK47jMosY8e6mOh2rUco5/OI4Plp/lsMJuTw1sTM9Al3rdO4ve+PZeS6Lu4eHE5dZhLezhr4h1kPRG0+ls3BHHNf09G/U1EBC6yXu3y2TCOZagBKdkcIyA15OLbNGqC1vrznF55vOVbvfTiW3OaQDcHV3Pz6b3aexmtYgJElizbFU5v18AL3JOkVJZWqlnDBPe+66sgNP/X7UksT1gluHhPLi5OYrDdWWvPfvaf4+lMwNA4KYGOVHYZmBqAAXJEkiOr2QQDd7q3mYrcXZtALGfrCl1uMeH9+JcV196OgjVlILzUfcv1smsQCiGSXnllCiM3LDNzvJKNDx9rQeXN8/yOax+aV6Hv31MMU6A+/O6Gmpz9lcVh1NqXH/1D6B/LovARc7JZmFFak+egW68O6Mno3dvMtSWGYgMaeYid39eCqvK6+uOGEzkHtzandu6B9kGe6+rncgI97dSHx2RTLXIl3NqwqFunt0XCcetVGlQCaTteoAp6OPE1/e1Jc9sVks3RNf7eKEmweH4KxVNXHrBEFoDUQw1wyMJoklu+N5/q9jKGVwofb1S38fZ0CYO6GeVVOO/H0omf9OpAGweOd5npjQuSmbXMUjYzvx8C8Hq613qVXKUStkVoEcwKHEPLKLdQSoW+a8phKdkas/3sr5rGIifRw5Vz7RXKOUc2VHT0tJqQhvB6b3DbSatyiXy9j8+Eh+25/IU38cxWiS6BXk1iyvQ2hdJkT5MiHKl0k9/fl0QzSFZQZUChm7z2VhkMDfRdukgVyJzkh+iQ6fZv7SKAhC3YhgrgllFZZx4ze7OJNWaJlbZqgUDBXrjdz47S5uHhTKvSOsV3v2DnbFXq1AZzDRP6z5Uy5M7uVPBy8Hpny6DaMESjlUTuT/3fa4as/VX2bR9MaSklfC7G93cz7LXC/0TKU0LF18Hbl5UAgbTpmDuXuHh6OysdhDJpMxo18QvYJcySrS4e6gRm802TxWEC7WJ9iN7+f2tzzWGUzEZhbS0bvpeh5jMgoZ98EWDCaJmweF8Oq1UU323IIgXBoRzDWhbdGZNvO0VZacW8pba04xPNKLrv4V8xG6+buw7clRGIwmvJ1bRjHwqAAXNj8xkhKdkQhvR06m5PP0H0ctyX9tuXd4B5s9j82pzGDkRHI+r/xzgpjyqhZOWiVFZQbL8Ororr54OGosj38/kMS0vraHxME8dPbwJ1s5lpSPq52KXc+MrrVmrSBcTK2U08m3aeclrTqaallssvJoigjmBKEVqHd3weHDh3nttdf4/PPPyczMtNqXn5/Pbbfd1mCNa2vcHWznHbO/aNK2i50Kb+eqiyHcHdR4OWlYezyVHefMP/stZzI4llR98NTYAt3s6ejjhEwmo6u/C09O7IxWJcdercBZq8TbSW1VkqkllcHKLdaRll/K9C92ct3nOziYkGvZV1BqDuQivBx5/dooHhgZQbCHPQHlaS+GhNe8QlWSJI4lmdNG5Jbo+aA8ebIgtHSzBwZhV/7FI7tIx/ylB5u5RYIg1KZeq1n//fdfrrnmGjp27EhBQQHFxcX8+uuvjBw5EoC0tDT8/f0xGmuuDdgeV8MUlRm4e/F+tkVbB8AXp/N4cGQEsweF4GujpFSJzshzfx7l9wPmPG1udipySvTIZfDHfUPpFeTaeC+gHsoMRpRyOYry9CkFpXre/+8MTloV80ZFoGwBQ45nUvO56uNttZZMAnMv3X8PD8fXRUtRmYHMwjJCPGruXZQkiaH/20BynjllxtTeAbw/s1eN55hMEmUGE3qjieUHk+gd7FrntBOC0JAGv7melLyKOrh/3t9yPl+E5tUe79+tQb2GWV966SUee+wxXn/9dSRJ4t1332Xy5MksW7aMCRMmNFYb24Rnlx+tEsiBdSDXN8SVv48ks3BHHA+P7chtV3SwOvbx3w6z4kjFKtKc8oLwJglS80qghXzYapTWPY1OWhUvXtOy0nO8ufpUjYFc5XJeBaUGtp7NYEa/IIySOeACc/LjX/YmMLVPYJWcXLtjsy2BnBz4+3ASLnYqOng7Mqt/kCWgjU4vYOtZc6H0Oxfts8zXu+DbOf0Y09WngV61INRN/2BX/j5akWcxq7D2hMaCIDSfegVzx48fZ/HixYB5ovfjjz9OYGAg06dP5+eff2bAgAGN0si2oLSadAOVRXg5sf98AgCvrTzJzP5B5JUY8He148+DSWw5k1HtuTEZl1fPsr2J9HFi42nbP087lRyJit9ZqKc9wyO9yCvRM+HDLaTklTK+my+bTqdTZjCx41wWi3fG8dcDV1iuEerhgIudirwSPSbAZIIFO+IAKNEZuGtYOAajieu/2kV2kY4wT4cqgRzAo8sOc/jFcQ398gWhRu/O7E2Qxxm2ns1kbFcfRnX2bu4mCUKLs3DhQubPn09ubm5zN6V+c+Y0Gk2VRs+aNYvvvvuOG264geXLlzdk29qU166Lon9o9WkqAly1TO7lz4XCDuFejkz7YidD/reBt9ec4rFlh8kvrT5n2dlaFlYIFVYfTeHfY7bz5D1zVWf+fmCoJZDzdFSz5I5BONupSMopsQw9rT2eaumhAzicmEd2UUXvha+LlvWPDmdan6q1LS/0+Jkk82pFAIVchkpRtaqH0STRDHm9hXZOrZTz+ITO/P3gFTw4umOzlQ4UhKYwd+5cZDJZlX/R0dHN3bQ6q1fPXK9evdi4cSN9+/a12j5z5kxMJhO33HJLgzauLfF01DClZwB743Js7k/KLeVYYh4Hnh/L2bRCXOyUjPtwKwBbz2bi4agmLb/6oY7lh5K5tk8gwyNbX4HxpvT9tlheWXGi2v1vrDrFj7vOE+hmR2JOCTlFOob8bwNKuYwvZvehg6cDsZlFVcqWOWqUljxgG0+ls/lMBqM6e7H8oHUd2i5+Ttx+RRhgvmEuvn0AG06lc23vADIKyvh6SwyHErLxctTQN9SdyT0DxI1UEIR2xWiS2BObTXpBKd5OWgaEuVvmYDeWCRMmsGDBAqttXl6t535ar2Du3nvvZcsW22VnZs2aBcDXX399+a1qoyJ8HC3/lwEOGgWFZRWLRf635hR/Hk7i5zsH4Wqv5vYrwtgenckjYyNZtj+BVZXmsAA8MT6St9dWrJI8mZIvgrkalOiMvL3mVK3HxWeX8Ns9g/huWxyrj5l/5gaTxJ2L91d7ziezeqNUyMkr0XPX4n3ojRI/7YqrUjmif6i71QKQ3sFu9A4299iGezmKOq6CILRra46l8PI/J6wW4Pi5aHnxmq5MiPJrtOfVaDT4+vpabXv//fdZsGABMTExuLu7c8011/D222/j6Oho8xqHDx9m/vz57Nu3z1yZpmNHvvrqK/r16wfAjh07eOqpp9i7dy+enp5cd911vPnmmzg4XH66rnoNs1533XV88MEH1e6fNWsWGzduvOxGtVWDOnjwypRu3DY0jIW39sfFzjqjuwScTCngi03mrt3nJ3VlzfxhjOzszcQoPxRyGb7OWv5+YCibHhvBfSM74qQxx+N2KgWzBgQ39UtqVT5Yd4bSOiQsdtYo+G1/kiWQq41cBl39zSlX1Ao5juW/E1vTJAd1aP6Ez4IgCC3RmmMp3PvjAatADiA1r5R7fzzAmmqmxzQWuVzOxx9/zLFjx/jhhx/YsGEDTzzxRLXHz549m8DAQPbu3cv+/ft56qmnUKnM9/mjR48yfvx4pk6dypEjR/jll1/Ytm0bDzzwQIO0tV6pSQCWLVvGn3/+iV6vZ8yYMdx11131flKxtBl6v/IvOcV6m/s6+jjy38PDKTMYeejnQ5xIyad7gDPxOcVkFZRxZaQXr1/bHaVCTmpeKduiMxkW6Ym3U8tIJtxSvf/fGT5ef9Zq28WpYS6FUi7jxCsTUCvN3412ncvinh/3kVtSMcexi68TE6L8eHBUBPJGHi4QBEFoLI11/zaaJK54a0OVQO4CGea5yNueHNXgQ65z587lxx9/RKutuIdOnDiRZcuWWR23bNky7r33XkuO3YsXQDg7O/PJJ5/YnHI2Z84c7Ozs+Oqrryzbtm3bxvDhwykqKrJ67ktRr2HWr7/+mnvuuYeOHTui1Wr5/fffiY2N5c0337ysRrRHDhqFJZhTyMy1Py8U2O5QXiFh57ks1hw39w7FZ1esdPxlbyITovwY2ckbXxetVVqM06kFRKcXMq6bjyghdZEHR0VQpjfy3bZYS1oSlUKOu4Oa1HzbHyDV6R3syg39g9h8JoNJPfwtgRzAu/+esgrkAKIzCvFwVItAThAEwYY9sdnVBnJg/tKdklfKnthsBteStP1SjBw5ki+++MLy2MHBgY0bN/LGG29w4sQJ8vPzMRgMlJaWUlRUZHNo9JFHHuGOO+5g8eLFjBkzhhkzZhAebi7NuX//fqKjo/npp58qXpMkYTKZiI2NpUuXLpfV/nrd7T/55BOeffZZTp8+zeHDh/nuu+/49NNPL6sB7dXy+66gX4gbMsAoQY8gV4Ld7ekf6oqfsx2vrzzB7wcSbZ6rlMuI8Ko6Zp+YU8zkT7dx/5IDvFbDJP/2KKOgjDHvb+arLTFW+eV0RhOp+aUM7uCOVlkRaKmVMq6IsP7A8HXW8MKkLrw1rTt/3DuEmf2D+Xx2X67qbp7HYTCamP7FDvadz63y/HqjVKf5eoIgCO1RekHdvlDX9bj6cnBwICIiwvJPp9Nx1VVXERUVxe+//87+/fv57LPPANDrbY+qvfTSSxw/fpyrr76aDRs20LVrV0uWD5PJxN13382hQ4cs/w4fPszZs2ctAd/lqFfPXExMDLfeeqvl8c0338xdd91FampqlYmDQs3WnUhj3/mKla27YrIBc4mpvXG51Z4nB967vidB7vZV9uWXGCzpMjJEkk8ru2OzbOZxA3Ov6Gez+/Lphmi+3x4LgM4gse98DjLMeeZeuLoraYVlxGQUcdeV/pYVpiaTxPKDSeZarjqD1e/0YsPE4hRBEASb6jpNqKmmE+3btw+DwcB7772HXG7u9/r1119rPS8yMpLIyEgefvhhZs2axYIFC7juuuvo06cPx48fJyIiolHaW6+euZKSEqtVHAqFAo1GQ3Gx7ZukUL1FO+Nsbi+oIZccgAn4YUcc6QWlFJYZ2HEukxKdeUVsV39nHhsXSbiXAzvPZdHrlX8Z9e5G1p1Ia+DWtz5Dwz2thkIrc7dX4e6g5v6R4YR6VATJpXoTEhCbWUyx3shTvx/l6y0xDHpzA2Pe30xWYRnfb4/l0WWHuWvxfn7ceb7GNtw7vEON+wVBENqrAWHu+LloqW4iigzzqtYBYU2ziCw8PByDwcAnn3xCTEwMixcv5ssvv6z2+JKSEh544AE2bdrE+fPn2b59O3v37rUMnz755JPs3LmT+++/n0OHDnH27Fn+/vtvHnzwwQZpb7165gC+/fZbq4DOYDCwcOFCPD09LdvmzZvXII1ry67rHcDJ1VWH3TQqOWO6+FjKdvUMdMHf1Y4Rnbx45Z8TFOmMHE3KY8Dr6y3nKOUQFeDC4+M78+OueKv5X7nFep7/6xhyOYzq3H7LQrk5qHn+6i68suIEeqN5mNVeZZ6n2D/MPJzq4ajhwxt6M/OrnZQZTHTwdCC/VM/ADh509HFCKZdhMEkYTBLR6YXsjs22ShzcJ9iN/fG51bbBz7Vqb6ogCIJgTpz+4jVduffHA1UWpl0I8F68pmuj55u7oFevXrz//vu89dZbPP300wwbNow333yTOXPm2DxeoVCQlZXFnDlzSEtLw9PTk6lTp/Lyyy8D0KNHDzZv3syzzz7LlVdeiSRJhIeHM3PmzAZpb71Ws4aGhtaawFQmkxETE1PjMWI1q9numExu/HYPRpNk+eOd2ieAt6f14JnlR4nJKMTbScvJ1AKentiZgWEe/LT7PG+vPX1Jz7fsnsH0DxWpMXbHZOHhqMbXxY5z6YV083e2yv0mSRJZRTo8HTVW5x1OyGXdyTR+2nUeP1c7frx9IPYaBQu3x+GkVXHjwGBWHkni6T+OkV9qQCGDGwcGk5ZfyuguPszsL1LHCEJbkFusw1mrapcLmhr7/t1ceeZau3qnJmkIIpirsOZYCv8cSWFMF2+cNCpGdPIit0TPvrhsAt3smPTJdgB6BLrw4x0DeXDJQbZFZ2KsoUh8dZbcMZAhEZ61HyhctszCMuzVCuzV9e78FoQW7WhiHkv2xDO+mw8jOrX9mq2ZhWWYTBLvrD3N6mMpaFUKMgt1DOrgzsx+QZgAlVzO8EgvXOzNOcV2x2SRlFvCNT38kJBVO8WjNWqK+3dzVIBo7Ro0mMvKymLx4sXMnz+/xuNEMFe9s2n5XP/VLnKK9TioFQS723MytYAAVzsivB3ZfMZcHN7PWUtKPdNpHH1pHE5aVe0HCoIglMst1qEsT4ZdojMy6M315JWYV/N19nVi9UNXtsmSc6dS85n51U7ySmqex3yBWinnu1v6cTI5nzdXn7IaJpzWJ4D3ru/VKO1sauL+3TJddjAnSRL//vsv3333HX/99RfOzs5kZGTUeI74Y7B2PquIvw8lYzCZ+Gi9dWHf3+8ZzJurT1lWSTqoFeiMJmb0DWLJnnhkgLOdErVCTk6x3irtxsVi37yqTX7oCq3PkcRc/F3tqgxlCy3H4YRc9sVl88aqkyhkEs93z+XTGB/SCnRWx80dEoqjRslNg0LwdWk7ictfX3mCb7bG1usclUJmmZNbmVYp59RrExuqac1K3L9bpkseA4qLi+P7779n4cKFJCUlMXv2bFauXMnIkSMbsn3twm0L9nIus6jKKh4vRw1RgS4MifBk3/kc/F20/HbvELQqBe4OanoFu/LEb0fIKzHQL8SNpXf3YPR7m6tc314l59e7B4tATmgW60+msfJoCv7OWhbsiKOofPW1Ui5jWKQnSDLOpBeQUVBKmUHCSavk/et7klOkw0GjEgmwm8Hfh5KYt/SQ5bFRkvH8YRdAV+XYhTviAPhqyznGdfXl41m928SQWP9Qd5vB3LTeAfx+MAkwT8wP87QnJrMYtUKGzkYgBzAhSqTuEhpXvYK5srIy/vjjD7799lt27NjBxIkTef/995k1axZPPfUUXbt2bax2tlklOiPny6s7XPwxkFFYhsEo8cjYSCb39MPHWWs1TDq5pz8/7TrPkcQ8HDRKNEo5948I57NN56yuM7KzD1GBro38SgTBWl6JngeXHGDL2Uyb+w0miQ2nqvbiF5QauHPRfstjtVLOvcPDeXhsZKO1VbB2YTqHtZoDar1RYuXRFIZHenF9/6DGaVgTeeaPIyzZk2Bz365Yc05QZ62SXkFubDlr/lnpjJLNnrlBHdz58Ibejdtgod2r19fdgIAAvvjiC2bOnElycjJ//PEH06dPb6y2tXmFZQZu+X53tUOjIzp54VBetD3C28kqkItOLyQxp4RFtw9ApZCx+UwGUz7dzum0Arr7O6Mo/2Ic5e/MOzN6NPprEYTKjCaJuxbtqzaQqw+dwcRH689Sqjc2QMuEuhjUwa1Ox9nqf3vi9yN8vinaxp7WY+le24Gct5OGpNwSAEI9HYjJLLTsC3a3tznEOqOPf+M0UhAqqVfPnNFoRCaTIZPJUCgUjdWmduPj9WfZE2e7YsC7M3owrU+gzX1bzmQwd8EeTBLcPDgEpUKOzmgkq0jHupPpeDioOPTiOOxUCquUG4LQ2DIKytgencmrK06QVVR1SM4WGXBtL3/WnUyloKwib59KDvqKh8jFNIFGU6o38saqk+yOzeb6fkGcTs2r03nVzdCNzShquMY1A2etitySqiWbysq/UMiAzj6OZBbpkCSJEZ28WVmeG7SCxDDZYa49+zP0+7HxGy20a/W606ekpHDXXXfx888/4+vry7Rp01i+fLmYi3UJissMfLu1+nx8L/x1nDdWnbS571RqPhc68xbvPI+uUuJagKwiPS//fVwEckKT+njdWfq/vo75vxyqcyAH5oDgr0PJFJSZsFOZvyS62qksgZyXo5qXJ3clu0jHptPpPP/nMcKeWsmodzdhuoQUPUJVzyw/yqKd5zmdWsBrK04wurMPChmoFTJCPezqfT13B3UjtLJpLD+YSLHOdi9wXnmFHgn4dX8SG05lkJRbyk+746sEfzIkPlR/gSJxd2M3WRDq1zOn1WqZPXs2s2fP5ty5cyxYsIB58+ZhMBh4/fXXmTt3LqNGjRK9dnVwPLkiILOlWGfkm62xKOUyZvUPJtjTwbJv1oBgluyOJ6681qitYdrSiwI8oWXJL9UjgzaVKuavw0m1HuOkUVBQZr5RhnrYE5dVjEIus+RNLCnv+ah8Y+wd7MZ1fQIZ+c4msop0KGTmm2lMZhGPLTvM0AgPIn2c6C7mhV6StcdSiMus6Enr6u/M+Cg/Dr04DpVCjkohZ9XRFILd7fl5T7zVEKRaDjobHzVL9ybQPdCFST1azxDj+awirvpoq2WBzsVkgFwOxjp8tMpkMNxbj7trTxh4d8M2VBBsuOzUJCaTibVr1/Ldd9/xzz//4OTkRGZmzfNkxNJmKDMY6fXyv5To6xZ0udqp2PPsaNRKc6D8wl/HWFReC3RCNx/s1UoOJ+SSnFdCoJs9/zw4FK1KJKxtiZbuieeZ5UeRJIjwduSdGT3pFeTa3M26bL8fSOSJZUcw1uEjRa2QceKVCQx6cz2ZhbX34q1+6AomfrQNACet0mYNY0eNgh9uG0jfkLrN92rPMgrKWH8yjff+O0NGQRkAV0R4MLyTN7cMDq02ye3Q/22wzBmb3jeA3/ZXBPByGVZfUGUy2PL4SILcW0cZu4eWHuCvQxcPlVp7ZmJnfJy1/LAzjgM1lO67YN0jw4jwdmqgFrYM4v7dMl323V4ulzNx4kQmTpxIRkYGixcvboh2tXkapYKruvvx+4HaezPA3FOx41yWJeP6nVd2IDazCA8HNf+b1gOtSvSGthafboy23PTOphcy57vdrH90BCqFDFf71jc8lZ5fyldbYogKcObs6xPZfCadWxfuq/EcnVFi+cEkcotrD+QGhLrTyceZN6d2Z+e5LO4dEc6p1Hwe/uWw1XGFZUZuX7iXUV28eXFSN0s2fsEst1jHaytP4qhRsu5EKom51knH3ezV3HllhxqvMaNfIB+tP8v4rr6EeThY7ese4MLhxIq5dnZKOfbq1vG5tD06k7XH0mo97o3Vp3ju6i5c1d2v1mDOy0mDv2v9h6gFwGSE8zugMA0cfSBkCMgb52+ptmlit9xyCwsXLmyU525IopxXM0rMKWbsB5sp0ZnwdtJQUGqwDDPZcvzl8ZbVrULrVFCq57rPthN90QRxGaBSQO9gd/JL9DwxoRMjO/s0TyPr4uw6KEqHHjO5d8khVh9LBWDt/GF08nVi5ZEUXvzrGJnVzJ3zc9FSWGaw2cNW2ahO3nx1c19UNnqKtpzJYFt0Jr/vT6wyR++GfkE8ObEzbq147lZDe3ftaT7dWHWVqUxmTkb+1/1DCa9DL5LJJCGXy9AbTAx9az3pBTpUChlR/s7EZBaTV6JHq5Lz8JiOfLUlhmB3ez6a2ZsQT4dar90ccot1TP9iR5X3ZHX6BLvibKdi0+nqk+PPGRTC01d1xq4NlvNr9Pv3ib9hzZOQn1yxzdkfJrwFXSc3+NOlpqZa/v/LL7/wwgsvcPp0Rf1zOzs7XFxcLI/1ej0qVcv7olivYK5Dh5q/tV0QE1P9xH4QwVxlBaV6isqM+Jbf3H7YHsfaE6kcqfQN18VOxR/3DiHc27EZWyo0hCH/W09ybt3KsE3vG8CYLr6M7erTLElYjSaJEr2RuMwi1p9M5UxaES9P7oZn1l5YeLX5oFHP8UzWBJbsjketlLPh0eEEutlTojPS//X/KCyz/eUkyM2OUE8HttYhdcmFALHGY46ncDw5n083RFsN9d0yJJiXJ3ev82tuy/46lMRDlRIBA6gVcOb1qy/pekaTxEfrz/Bxpao1zlolUf4uHIjPwSRJVkl0HxkbiVwGQe72TOkVcEnP2Riu/3JHlawCoR72xGcXY5LMn78rHriC99edYU9sNs9P6srmM+n8XE0eunAvB9Y/OqIJWt48GvX+feJv+HUOVddJl3/+Xb+oUQK6CxYuXMj8+fPJzc0FzMURwsLC+OWXX/j888/ZtWsXX3zxBefPn+fPP//k0KFDlnM//PBDPvzwQ+Li4izbFixYwNtvv01sbCyhoaHMmzeP++67r1HaXq+vDXFxcYSEhHDjjTfi7d32Cyw3BSetyjIJ3lGj5P5REdw/KoLo9EK2R2cQ4e1IVIArLnYt75uAUDtJkqy68S/MT7pARvXpHX7bn8Rv+5O4Z3g4T03s3HiNtOHeH/ez+lgqDmqF1YTwEyn5bJyiRy8pWG68gsBUBS9O7cqAUHc6+jgS6GaeH2Uwmaxu5PZqBWsfGsaN3+4iIaeEcG9HpvUJJC6zkDAvR2b1D+JsWhGfbYqmrNLiHTd7FcF1mHM1vpsf47v5ERXgwl2VEg7/sCOe3THZ3D0snN2x2dw4MJge7XShxJReAeyNzebH3fGWbe4Ol15O7cvN56wCOYD8UgM7YrJsHv/x+rOWxVoBrnb0C3W/5OduKKuOpthMD3VNDz+u6OjFlrMZzOwXTJCHPS9N7sbtC/fy3J9H6epXfRDjcRk/03bNZDT3yNn8RJQAGax5Cjpf3WhDrtV58sknee+991iwYAEajYavv/661nO++eYbXnzxRT799FN69+7NwYMHufPOO3FwcOCWW25p8DbWK5hbunQpCxYs4P3332fixIncdtttXHXVVcjlIgVGQ4vwdiRC9MS1atujM7lr0T68nbX8evdgvJw0vDIlinfWnCK3RI+dSsHSuwYz+dNt1QZ0AL/tS6BnoDMTuzf+ysCUvBIWbo+zDJtevLJPZzBBx7G8H/Y1X5yyQ3YQ/h5SyLW9K3palh9M5Lnlx+js60RBiZ7YrGKKdUZOpObz6z2Dmb/0EPklet799zQJOaUk5JTy9rSeTOzuj0Ih4521FUMccwaHYFePeVfjuvpy65BQFpSXmAI4lVrIw7+a59f9sjeBlyd3Y86Q0Ev46bR+K45aT/APcNFSojPW62d8QWFZxRB5By8HYmoZprwQyMlk5jnDLcF/J2zPk/tk4zluGBjC4+MrvkTtiM601Mi2lQzbWaskxN2eZ67u0jiNbevO77AeWq1Cgvwk83FhVzZZswDmz5/P1KlT63XOq6++ynvvvWc5LywsjBMnTvDVV181fzB3/fXXc/3115OUlMTChQt5+OGHueuuu5gzZw633347HTt2bPAGCkJrcjIln21nMwlws2PBdnMd0tjMInbHZjGphz+zBgQza0Cw5fgtZzKqBHKdfR3JKCgjq8icniOzSMe9Px1kzXxHOvs23rSEY0l5TPpkm819oR72BLvb4+ui5eqPt+LnEgKkI0lQrDPf1E8k57PmWArrTqZTpDNyJDGPW4eEErsjDpVCRkcvR9afTGd3eTmkC0On3QNc8HA0z227f2QEYR723LfkIADB7vWfZ/Xi5G70DnHlmT+OVhnmlYAX/j7OV1vO0S/UnTeu696u5qEOjfC0Sm67PyGPeUsP8s2cfvW+1oOjIigqM3Auo5DdMdl1OkcG2KsUXP/VTrQqOe9d35NRzTg31NvJek5luJcD5zKKcLVX4ahRkpZfyubTGQzv5EWejSTCFyy6rT+DOnhWuwpYqIPC2heg1Ou4BtSvX/3eHxkZGSQkJHD77bdz5513WrYbDAar+XcN6ZI+xQICAnj22Wd59tln2bx5My+99BLvvPMOmZmZuLmJtABCzTILy3hnzWl8XbQ8NLoj8jZQlBsgIbuYqz/aysXJZtQKGTvPZRHkZs+m0xkUlOq5/cowvJ202MrrfCq1sOpGYF9cdqMGc19ttj3XVS6Dx8Z3IsrfhRHvbrJsnz+mI0Fu9gzs4AHALQv2kFFQhn2lldUXirDrjRLfbovl5z3mIT4ZcO/wcAZ2cMfTUYNJkvj7cDJdfJ24qoc/i7QqDCZTtTf6Y0l5LD+YxFXdfekbUnW4bnLPAPqHujNvyUH2nq86jJaUW0rSoWQ2nc4g0M2Ob+b0xd+1daTQuBzvzejJoYQcknIq5m0eTshl0+l0y0r5urJTKVh1NJXMwoqpA9WljQFw1irILzVaentL9EZ+2HG+WYO5C7k6AZw0Sku+vZsGBuNip+Laz7YRm2k+prpPqXAvBwaHe6ISSdovj2Md/w7qelwDcnCw/lIpl8u5eLmBXl8R7JtM5rvAN998w8CBA62Oa6w8vJf8lbS0tJTffvuN77//nt27dzNjxgzs7dv+h6Fw+T5ad5Zf9pknD3cPcGFM1xa8arOOjiflcfeP+6oEcmAeXvppdzwrj6aQW2x+w3+7LZZu/s7cdkVYnZ/jpb9PsCc2h49u6NUoVVfGdvXmnyPWwxxKuXle57yfD/L6dd3p6O3I2fRChkd6cfewcAxGE5vPZNDJxwl1+c2suNKK7Mofd0v2VMzVcrVX4eeixc/FnLrhxm92seNcFiq5jK1PjmJYpBeGGrKz3rloHyl5pfy6L4EjL46z+fPwc7Hj21v6Mfr9zWQW6nC3V9Ev1I11J9MtiyTySvTklei57vMdfDKrNwPCPOr7Y2tVtCoF258cTdhTKy2/m/SCMuYu2MvMfgHE55SwJzabgaHu/HD7wBoDFEkyz40Ec6Dj7qCusfKHrZyak3s2b1Lh+0ZEkJxbiqejmo2VVqcu3ZuAg1ppCeTA+m/ZUaNk7fxhuDuo2uSK1WYRMsS8ajU/Bdvz5mTm/SFDmrplVXh5eZGammo1J7ryYggfHx8CAgKIiYlh9uzZTdKmev8V7t69m++++45ffvmF8PBwbrvtNn7//XfRIyfUWYiHOehXKWQEuLX+PExp+aVc8+m2ait6uNipyCnWo71oCOZ4cj7d/JxRymU2q3hczGAy9175uWi5pqc/UQEN112fnFvC99vjqtRDNZggpzwA/e9EGotvH8BPu+Jxd1TR/cU1GCo1uz4dEznFemZ9swt3BzWju3iz45x50rzeJPHhujOU6o38eSiZ268I4/lJXauc72KnIiWvFGetymYgl1eiZ+ZXO4nLKqK0/AVlF+vZHZtt8/eUll/G9V/tYt0jw9vFXNUHRkbwyUVpSn7ZV5HzckdMNssPJjEwzJ1gd3v+O5GGi52KCG9HvtsWi1opZ1qfQH68fSDrTqYxrqsPkT5OfLHpHF9vjbHZO+ekUZFdKa/goA5uTOtru/50U+kZ5Mo/D15BUZmBKZ9uJzrD3CueWajjrUpzNy8IdrPj7Rk96R3kikbk9mxYcoU5/civc6i6NKz8PT7hf02++MGWESNGkJGRwdtvv8306dNZs2YNq1evtlrd+9JLLzFv3jycnZ2ZOHEiZWVl7Nu3j5ycHB555JEGb1O9UpN069aN9PR0brzxRm6//XZ69OhxSU8qUpO0H0cT8wh0s+NseiGJOcXIkPh5byJOGgWTewUwuad/q6/tu/JwMvf/fNDmPg8HNX89MJQ3Vp5k1bFUNEo5PYNciM8qZnC4JzIk/jhY06Rf21zsVOx/bkyD1d/9bGO01cIDW3oFuXA0MQ9jnT8xLs3AMDd2x5qHRl3tVRx6YVyVY9LzS9l4Op0rOnoRYCMx66bT6cxdsBfAKliWg83e0wvGdfXh60uYP9YaGU0Skc+tqrY81YWVzBOjfFhdnlB3YKgbu8tXf6qVctY9PJxgD+sRmaIyAxtOpZOQVcTb/54BINjdjh9uHcDjvx1m3/lcAJ67ugt31JKkuCnpjSYKSg30fe0/Lr4rBrvZMfeKUG4cENLuE7Q3T565AHMg14hpSaD61CQHDx6kV69eVsd++eWXvPHGG2RnZzNt2jQ6derE119/bZWaZMmSJbzzzjucOHECBwcHunfvzvz587nuuusavO31CubkcjkODg4olcoab8DZ2TVPhhXBXNtXojMy6M31NU4aBrhlcAgvT4lqolY1rLT8Uj7dcJbFuyqGD+3VCrwcNZzPNg/P2KvkHH9lAncu2se6k+lW52uUMsoMdXv7OWqURPg4cqg867y9SsHRl8c3WP65JbvP88zyYzb3XSjT5OeiISWvzOYxF7txQBAbT6fbPL5yfVYw3+hHdPK2lKcDcLEzz72aMziUlyZ3q+erMa+0nPnVTo4n59frPDuVgpOvTqj387VWUS+uqTYX4AUqhQx9eQR/cX/JkxM6ce+ICKvjE8r/9oPc7fngvzMcT87jyQmd6ehjXvByOCGX7CIdIzu3zPRWr644znfb4iyPr+nhyyc39m2+BrUwTXL/bsIKEG1FvYZZFyxY0FjtENqY/60+UWsgB7D2eCp3DQ/neFIewzt5tZiUBbXJKixj2NsbrXKi9Qx0IatIZwnkAKb3CyIp1zwP6WIXB3IXCs9fULlHqVhnwMepIn9Vsd7IwfPZRPo546y9/ByER5Pyqt1nksDLSV1tIOftqCa9Un3V+aM7Mn9spPlck8RVH2+xLOqY0TeQlZXSY8gAe6WC7ELra+eVmIfp5JfYa+uoUXLzoBCe+uOozf0KmTlFhsFkXVNUJoObv9vN1zf3u6R0Ha3N/6b14LFfD6NWytEZTOgMJkyY50pe+NPWV+qKvfirxxebzjG5pz/nMopYujeeAaHuvLryJAA/3DqAh8dGsnRPPPcvOUAnX2fendGDni28DvHzk7rhZq/m/f/OEOhmz/vX92ruJrU/ckWTpx9p7eoVzDVGbhShbaoc0NTEKMH4D7ZQWGbAzV7Fw2MiGRzuYfkWf3HS3ZZi9bFUq0DOVavkyo6efLrxHGDuzbiqux+Pje/E7Qv3kl/NCj+NUoZKIWdYpBePj+/EdZ/vILdYT78QN56e2IXbftiDyQSPjotkaIQn/51IswQes77djdEk8e6Mnkztc3lzj9xqqQmbUVARrClkWIZa7VRyNjw+kl3nMlm6N4HregdydQ8/y7FyuYykHHNhdpkM3prWg4ScYnaVp7KQgFPphZxKt17Bq1XKQQZjulx6701+afVfJowS9A9xw8dJa8m91s3fmePJ+Ww9m8nPe87j7axlYpRfs1TfaCqTevgzNNyTV1acQKOUszc2m3OZRRhqGouuJL/UwPB3Nlm+dKw+mmoJ+HbFZHE6rYBXV5wA4ExaIVN7B7TYHrnKHhjVkQdGiVRbQutRr2BOLpfbvLE6OzvTqVMnnnjiiXon1hPapqu7+7PpdO2lmsI97dlVPj8qp1jPC38fByDIVUtibikSMKmHHx/O7NVg88MawpJKWfSn9PLnoxt6M6/SvLmnJnTm9vL5QMoakmqXGSTKDOZybmGejvx2z2B2nMviqu5+fLoh2tJD5ahRkl+qZ0I3H1aVz1+60GPy7/G0yw7mHhrTES8nDauOprDXRkb8ytwd1HTxd0ajlPPkhC44apSM6erLmK6+No9/eUoUP+w8z/S+gcjlMpbcMYgfd57nhX+OV/sc47r68PaMnpc1P6m2Xt7TqQW8dm13jqfk46BRMKt/MC/8dQyVQs4rK8y9SyHup1n/6PAW9bfXkEwmie+3x7L8oHnxg0pR/8C18uKdyj13JQYDf2xLtDxWK2U4tqOcfoLQlOr1zvrjjz9sBnO5ubns2bOHm266iR9++IEZM2Y0WAOF1mlGvyC6lA8B6k0m3O1VDHhjPXqjhKu9ypKiw1YpHYCESvVLVxxJwdtJywvXVF3V2BweW3aYk6nmuVgR3o48Oi6SEe9sJLG8B0ohl3E+q4iR724ir1hHdnHtw83peWV8s/Uce2NzSM0vpbOvExOjfPltfyIudip2x2bz2G9H8HRUE+RmR0L5c3k6qjFKEvvPZ9vMt1ZXGqWCLr7O/HEgsdpjLsyXenBUx3pVUJjaJ9Aq2JTLZehMJqu5WBfzcNJc9kTz+OyaKxJEBbjQydeJjY+NsGzrHezKVR9XJE4+n13MhA+3kJxbjJNWTf8wdz6+oXebyI34+/5Envz9iNW26n4fttgpQa1SUlhmxGhjifB3W+OsHusMEq+uPMEzV3XBw0Ft6X0XBOHy1WsBRG0+++wzFi1axO7du2s8TiyAaJ+Sc0tYfzKNbdGZrD1evyze4V4O/HbPENwcah4ObGzFOgNdX1hredzB04ESvZGUvNIazqpQeX5WbR4ZG8kDIyOQy2XM/GqnpXLC5J7+/H04GUeNEkmSKNIZ8XTUsO+5MfV+PRckZBcz/J2N1bYt0NWOZfcOpqjMeNmpO/JL9fR46d9q93fyduTveVdc1vzJI4m5XP/VTktaksrkMnNQ+s70nky/KDVGmcHIsLc2klZQ/UKPcV29+XpO/0tuW0txIbffxSoPo1c2rU8Avx9IqroDc77IY0l5NZalAwjztCc2sxiVQsaf9w+lm3/jZMMXGo+4f7dMDTp2MG7cOM6cOdOQlxTaEH9XO6ICXGwGcpU7OtwdVEzo5sOL13Tl1qGhAJzLKOK18onVzSb9JPaf9mSguqJSQkxmkVUgpykfjrt4uEqtlPPWtO71eroF22N5YMkBer78L+4OatztVYzu4sX71/fk2zn9WDnvCnyctYA5hcc3W2I4m1ZwSS+tSGewCuTUShmhHvY8f1Vnlt41kLUPD8PPxa5BcrA5qJU1Fip/f2avy14Iszcux2Yg98ms3mx+bCRbnxhZJZADcw/l5idGMKGaIWOAUym2K3S0NjcPCkFpo4exciDn56JlVv9ApvauGsgNjfDAQa3ASaPkXEahJZDTVMqnaK+q+P+VEZ70CHAFzD2AF+ZSCoJw+Rp0AkNJSQlarbYhLym0MUHu9rjZqyyJaGWYkJDTwdMeV3s1d10Zxrioiqzw0emFLN55HoNJwsXu8ldtXqqk3BIStv7HwLxEfpE/x/uRn/HJWTfUCvMqwAs3sjKjCY1Szm/3DCY2o4h5vxwCzCtdR3b2RlY+VikDtGoFMiRGRHqTkFPM0STrNBo5xXpWlRe8v1D4fv3JDN5YfZJnJnbhzdWnSMkrZUwXb+KS03h91Uk+WneG727tz8B6VjLo7OvMezN68vfhZGYNCGJClF/tJ10ihVzGH/cNIS6riKOJuRTrTQyL8CQpr4SeAa44NcDv+breAby24oRVT9HgMHeuqUPFAa1KyZdz+rLqSDLzfjmEoVJ0o1HKeGRc5GW3ryWY2N2PRXYqbvy2+pGUlLxSJGQcSqyYDhHqaU8XX2fendETmQxkyJi7YA+7Y7NRyGX0C3Vje7S5x08mkyEDrunpx9azmeQU6xkY5s7AMHfGdGn9lV8EoaVo0GHWBx98kHPnzrFq1aoajxPdtO1bdpGOl/85zl+HqibLvVAHdM7gUMtk6YPxOcRmFjGphz+HE3OJzSzi2l4BTVbU+u9DyTz+22HKDCbut/uXx53WwS1/kyj3Q6tSIMNcb/axZYc5mpSPViVnz7NjUMhkXPPJNmIyi/B11tI72JUZ/QL5flsc26LNi0Pmj+5Ian4pBWUGvBzULKyUa602Mpm5pJKdEgKMiURLFT1NC2/tX+9am21BUZmBlLwSDifk8uiyivlgXXyd+OG2AXg71+/LZnxWMY/8eoh953PwcFCz4bERzfqlojFkF+kY+c5G8spXXGtVclQKOWV6IzqjxKvXRqGQyXj339OM6uzNO9N7VJk7LUkSO85l0snXGRc7FV9tjiG/RM/XW8292IM6uFtWMPs4a8gu1PHylChuHBjctC9WuGzi/t0y1SuYq64ERV5eHvv27ePcuXNs3bqV3r1713gd8ccgHE/O49pPt6I32Z5IHuCqZd0jI6xyfZ1JK2DiR1sxmiRuGxpWrwURMRmFeDppcNaqSM4tISajiL4hbuSX6i1DlbYcScxl8qfbLY9rqhCQklfCb/sSGRLhYVmMUKo38tivhy3pL76Y3Qc3BzU3lacVgYoVgHMGh1glzq0rLzv41fgIV+lepwRzNYQRnbxYeOuAel+rNSvWGZj40VbOZxXjYqeyynP4ztQoZgwIuaTrGk0Su2KyiPB2rPFvpbVbfTSFY8l53DY0DA9HDcm5JeQU6y55Xlup3sijyw6TmFPCXVeG8fCvhzEYTZbhfFc7FYderFrdQ2jZxP27ZarXMOvBg7ZLFjk7OzNhwgTuu+8+QkIu7QNTaF9+3ZtQbSAHkJRbSlp+KW4OaktPSImuYtVcfHYRA99Yh6udmsW3W/e4ZBWWcTQpj67+zhw4n8sPO2LZGZONm70KX2ctp9IKkCTzzSS3RM/8MR2ZP8b20JmuUsKtUA97nr6qS7Vt9nOx48HR1rmptCoF46N8WXUsBVd7NVEBLgS52/PhzF48cFEJMDuVAh9nDWn5dauycIG/hwthw1+n31qJreXZYHoHtr+J5al5pZwvT7pcEchJ9PKQuK5v0CVfVyGXMTTCswFa2LJN7O7HxO4Vw+v+rnb42yiVVldalYLPbuwDwLdbY6zeSwC5dUgqLghC3dQrmNu4cWNjtUNoZwLd7Kts6+zrxNn0QiRJ4rreAdzz435OpRbw/KSu3H5FGD2DXPlwZi+i0wvJKdax7mQ6afllbDqTQWdfRzacymBcVx+u/Ww7OqNUpQ5nTrHeMlcPKm4mX2w6R5inA1N6BVRpU79Qdz66oRcJ2cXMHRp2SXmyrunpz8AO7tirlThqlLz8z3EWbI+zKo00qIM7X22JqekyNvm7aJgzJIS3kzxRuOdBpjmak9eQ266t6uDlSI8AZ45UmnsY7uXIn4+OaL5GCYA5UPxtfyKnUq0X6Dz7xxFen3ppNb4FQajQoHPm6kp00woAW89msCcmm5/3xlOiM1KqN1pW0nk5acgoTw8xINSdb27ph1wGTloVx5Py+HHXef48lIyHo5pbBofw+qpTAKjkYGMRI1C1rmRlYZ4OVvnGGlOfV/8ju0hnte3mQSEs3lX7EKtCJsN40Vv24nQnKrmM3+4d0uLLJjW0U6n5TPhwq+WxWilny+Mj8XVpu0Ojrc2Y9zYTnWG9GvjvB4bSI9C1eRok1Ju4f7dM7e/ru9BiXNnRi0fHd6JHoCtFOqNVSgQ7lZypvQMIcrcjJrOQni//S4+X/uXWBXuY9Mk2ft6bQIneSLC7PduiK3JlVRfIjYj04rYrQqtstytPTKtSyLj5u93EZdacaLY6RpPEaytO8MCSA6Tl15xz7p7hHdBWStng6ajm+UldeXRsJDcPCsFRo7CZ2gTAKElUySZxUYSqN0kk57a/tA/zlhywenzHFWEikGthvry5Lxfnna+pQoogCHUj3kVCs5sQVTWnV3x2CSV6I5O6+5FZXsRdAjaezrCKXWIzitAoZVUDnIvojSaen9SN069O4Os5fSx/+PNHR/C/ad05k1bI1rOZfLTuLMcSc9HXtThluf9OpPHttlhWHEnhkw1nLdt3xWTxv9WniK0UJN41LJybBlbMLY0KcEGtlPPg6I68em0Uh14Yx46nRhPsbp6vpJTLePO6KCK8Henk48Qf9w7hqu4VaR0utFQBuNip6BfixpB2MMersgPxOZxJr/gZX9vTjycmdG7GFgm2RHg7cvzl8Yzv6o2vi5bnru5MV3/RuyMIl0sUyqvBX4eS+PNgEu72KvqEuDO2mw/eTuKbfkO7vl8QHb0d+WjdGcI8HVmwIw6oyK1mi5NGibezhpS8Ev49kV7rc2w/l8W+uGz6hbozrqsfg8I92HEuiz8OJuPrrLEct/xQEssPmZOjLr5tAH1D3bBX1/42CfN0QK0055zr7Gu+ORXrDMxdsIdSvYktZzJY9dCVxGcV88OOOP45UpGAtWN5Il690YRKIUepkOPpqOanOweRXaijg5cDTloV0/oGkZRbwqcbojmTWjVxrRG4tpc/L0+JqrW9bc39P1X0ysll8M71vZqvMUKN7NVKvmoDFTQEoSURwVw18or1PLT0kOXx7weTefnvYygUcuaPieTu4eHN17g2qHewGwtvG0h+qd4SzNXE21nDuYwiSwb7mubDgbl3K8CtYmXehZ6y6PRCzlRTNeG+nw5QqDPw+rXda82H1cnXiXUPDye3RGeZ/yOXydCqFJTqTThqlBiMJqZ+sYPMQuvVqt9sjeX7bbFIwLszejK1TyCP/HqY5QeTCPd2oKjUQJnBRJinAwf/3959x0V5ZQ0c/01lhl6lYwFEEbuxxRh71BRNb8Z003svm55sdlP2zSabsomu6dH0atTEaOwdG6IiUqX3NgNT3j8GBoYZsEQZyvl+Pu+78jzPzNwhwHPm3nPPySpv932O6nPy/Vm7sqadklq1kp/unIBGJYsOQoieQzZAtPLPX1NZvi8fi8XCkRLXeUeR/nrWPzqlg0fWc7y35jB/X5ba5nm1UoGpRcb/pP7BrDlUTHs/yY/PHsCCic0B+OoDhXy4IYNwPz2fbclqdzwT4oL55KYxx/8GWkgrrGJjeimzk8Lw8lAz7LkVLttMtXRBi2r5x6LXqOgT7MntZ8cyOMqfPsFeJzXOrm53Tjk/7c7j/CERDO6BZVmE6Cid+f7dk0kw18KurDLmvL2h3WsU2Bqgt64nJk6t537Yy6INjrs7WwdxroT5aMmvctwpqlUpGR8XhFqp5NVLh+DvqbWfK6wycMk7G8kqrXV4jJ9ejdFkwcdDwz8vHcLkU9RNYcPhYn5LKcRksZBZUkt6UTXZLnpURvrryC133EihUSpoaHz/SuCHuyaQFCmBS2dWZWjgu525DInyZ2i0P9sySokM0OOv17I1o5ShUf74eXavjhKie+us9++erscHc3tzK1hzsIjEcF/uXbKTijpTm9cGeWlY/+gUdBpZnT7dTGYL3+zM5dXlByisMhLkpeXq0dH8+4/D9mt0agUGk+3HV61UEBWg5/4Z/bn782TAljsVFaBnXL8glmzLAeCJ2QO5eWI/+3OkFVahViqZ/Opql8uXCyb24/F2CgWfCtcu2syag8X2r4O8tJS0Kl3SZHCkLxcOj2T+uD6oZSmx0zJbrPyQfJT3/jxMan4VapWCK0ZF88nmLJRA72BPjhTXEhvixW/3n83u7Ar2Hq3gzLggegd5ObXLEqKz6Ez3b9GsR0YlZouVVamFRPrruOr9TVQaTOg0yjaXv5QKWDR/FBP6h8gNtIOoVUouGxXNZaOiySqpYfXBIlLzHHPbmgI5AJPFSkZJLX2DvAj301FYZeSNK4ahAO5q7LSgVioYHuNvf8zSrdk8/PXuNvPttColFquVV5anctukuJMqGHw8PrxhDIcKqgjy1pJbXkffIG8e+DKZPw8WMXd4JPG9fHh7dRrF1fXMTArnhgn9jv2kwi1yymo5kF/Fg1/uclgmN5mtfLnd9oHCAhwpts0EZ5bU8P7aNF765aD92guHRfCvK9pviSiEEC31yGDun7+m8t6f6WjVSrSNwZnJ7DqQmz6wF/dM6y/LWW60O7eCp77fB9gapqfmV7kMvpIifUkI82XNQ5MxmMz46jRMfW21vaCu2WLlQH4Vob46ogM9+fNQEdD2xonJCSF8sPYIANsyyvhiwdjTNmMSH+oDQKCXbWfte9c49n+9eGQURVVG4hp3vorOJ6eslumvrabO5PonqnU7KwCtWuUQyAF8m3yUoup6Fl9/hnx4FEIcl24dzO07WkGor45gbw+H43kVtlykepOFp89L5Mnv9mKy2GbgrFbwUCu46ax+XDwimr4hPTOhvDNpuTPxsjOiya8w2FtfRQXoeWzWQIZE+RHmp7Nfq1UrMVusDo+1Ak//sA+Txcp90/ozLjaQn3bntfm6FYbmmZXNR0pZlVrI1IGhbV5/OvnpNfYetaLzqTY0MOXV1dSb285aCfTW4qtT22flLh4Rydc7cl1euy6tmKPlBmKCnNveCSFEa902mHvux30saux/+cqlQ7ikRaPtJ84diI9OTf9QH5Ii/ewzMzqNin3PniP5Kp3MOYPCeOfqEVQbTVw8Iooqo4nPt2RRaTCRFOHHuUPCXT5ue2aZvRekt4eKaqPZvoHiX78ddPmYljall3L+kHB+3J2HApw+FIhTY9H6dN5ZnU5iuA9vXjkC3y4YtD701a52AzmAkup6ShoLYOs0Sg4XOdcKbDIwzIf5izZTUdfAa5cOZYqbPkQIIbqGbrcBIqO4hv+uOcySbdn29lAalYIDz89CqVRgtVoxmizoGts4WSxWHvpqF1sySnn6/EFMkz+aXUJ6UTW7cyqYMSgUrUrJ0m05eOvUXDA0wn5NcbWR2W+spbDK2M4zOVIroOUq2RVnRDElIRRfvQZvnZqB4b6olApWphTw+Ld7CPXx4JVLhlJWV8+YvkGojtWKQjjp8+jP9n+rlLD3mZnotSq3jcfQYGb1gSLMFgs/785Dq1ZSVGXkmnG9mZnk+MGhuNpIRnENP+46yocbj91bt4lKgUP7upZ5m6G+HlxxRgxv/G7rJKJWKlj1wNnEBMkqgXA/2QDROXWrYM5qtTLs2RVUGBx3pEb46djw2FQq6uq5+J2NHC6q5rk5SVwztncbzyS6kv/+eZiXfrHVpXt//iimJzYH5FWGBp74di8/7Dp63M/n46Giymi2f90nyBOVUsHhohrmDotAoVDw7c7m5bGmRvdXjYnhpQsHn4J31HOk5FUw+411Dsfunx7P3VP7u2lEcOV/N7IxvdTluYnxwRgaLDw0M4H+vXyY+voaiquNzBsbg6dWTY3RhJ9ewy978iioNFB3jJqCTcb1C+Sp8wcR4u1BoJeW9OJqpr3+p/28Sqlg4bWjmHSKSuQIcbIkmOuculV27bq0YqdArnegJyqVgvkLNzPl1dWkFVZjtcJPJ3BzF51by9Jz5lZ16Hx0Gh6emYCH+vh/1K8YHY1e03x9Rkkth4tsHSN2ZJU5BHItX3/p1mx+3i0/VyfivhZdVpq8vvIQhZUG54s7gNVqZVMbgRzAn4eK2ZJRylX/3URueZ29m0daYTWPzx7IixcOZsHEfhRUGh0COaXCttTflo3ppfjq1AT7eKBUKliVWuBw3myxcvcXO6mrN7fxDEKInqzb5MxlltRwzcItDsduP7sfb6+xJcpnlzYXZlUAN07o25HDE6fRjRP6oteo8PZQMzMpDICsklpWH7RtWIgK8OT+6f3b7SoB4KlV8cLcQTz53T77jViBbQPGvLEx7M+rYs6wCB7/dg+uahebLFae+HYv5w6JcD4pHJgtVv7vt4McKHDOGwv19cBH5568uV/35rvc3axWQsvNqA0WK+lF1Tw2awBbM8q4b3pzEXGLFcytFjx6B+pRKBRUGx2LU7d04dsb6Bfixei+gfz79zSn85V1JnZmlzE+NviE35cQonvrNsHctowyp2NNgVxrFwyLYMagsNM9JNFBNColY/oFsnRrDuH+xYyPDebidzdQVGXkqe/38cS5A7nprH4kZ5WzNs1WjqTa6DzDUVtvZn9eFbUtZj+ePn8Q5w4JJ8THtvlhy5FSl4Fck/6hUjrkWAwNZi54cx0HC5sDOZUCnp+bRHSgJwPCfN2WM9eUp9aai6oifLo5i88XjOWWsx2PB3ppuXNyHO+tOUxN489SW60Bm6iVUFhlpLDKSMrRSqfzCmz9fwdLiSQhhAvdJpibNjCUYdF+JGdXtHtdbIgX/7h4SAeNSnSU2z/dQXpRDZ9tyeSFOUlU1jZ3UHj5l1TiQrxZti8fAH+9hhExPuzIKnd4Dn+9mqFR/g7HVqbko9MouWJ0DADfJ9uWWEO8PahrMDkFhSN7B5zid9a9mC1Wft9f4BDI+enUfH/nBLf2lTWazDzx7R6ySmravEarAoVCiclswWyFc5IcN0tZLFbeWZPG8r0F7M5t/+9Qay0/IFQanLvQvHHFMC4YFnlCzymE6Dm61QYIgDlvrWNXjus/pB/dcAYT+0sCcXd00dvrHYKzAY3FhZuE+XqQX+m4q/X2s2P5eHMmVY03z6ber356tVNbt0ERvoT56kjOLre32vLRqe2PbaJRKVgwsR8Pzkjo8SVujCYz6UU17MouZ11aMRPig3l1+UFqjA3Umy001emekhDMouvHuHWs3+7M4b4lu4553aykMEbE+PPiL6mE+nrw/Z1nEuarB+Crbdk8+NXuNh87JNKH3PI6SmrabhnoipdWxcbHp+LrpqVnIVqSDRCdU7cL5v7zRxqvLD/gdPz6M/vw9PmDTulrdQbltfXU1puJ8Ne7eyhuVVJtZNQLv9nznQaE+RDm68HqFj1PWwr19eDRWQOprTehAF78ZT81LWbZegd6klnqnN907uAw9uRWkhDqzcr9hfbjrVuCfXbzmB6V21ReW89V728iq7SWS0ZG8fT5g5j1xlqHgLplOY6WOWgalYJvbjuTwVEdv4RYW2/i9RUHWbwhw16DsD0DwrxpMFvtG2JCfDxY9cDZzFu4hQN5lRharMd6qJUYW3zdP8SLFQ9M4vd9+dz08fY2O480CfPV8Y+LBzMw3JdevrqTen9CnGoSzHVO3WaZtckdk+OYNjAUq8XKXV/sJLO0lmcvSOTK0d2vDMlNi7fyW6otoHjt0qFcPDLKzSNynxUpBQ51ut64YjjJ2WUugzmdRsmDMxK4b0kyAI/PHkCgp5Yaoy2vaVi0P/+5egQfbczgf+uOYLJY7ctg2zLKuHtaPOcNjuDCt9eTXlxDoJeWWkMDhsZIRa9REh3Qsyr33/nZTlIae+cu3pDJ55szaZ2WqFIpMTcGNw6bCcxWVqbkuyWYe33FQT5Yd+S4r88sqXXYpVpUZeR/6zPYlV0OQLifB0aTlctGRZEQ6sOi9RnsaVxyPVhU41BTr7WpA0L4PbXI/vX9M/pztpQiEUIch24XzIEtURhg5f1nH+PKrmvzkRJ7IAfwwJe7UCsVzBneM/Nq9h1tXlp/6rxEEsJ86BPsycbDJfyyJx+LxUJsLx/MFgsvXjjYoQm6xWrrCnLrJzsAGBETQKS/rU3Y7ZPi2JZRyo0fbgOgoMrIE9/uJTWviv+7YhhPfb+H5OzmhPXBkX68ffUIogN7VjBnNDlGbi72l+ClVTn1J1UooE+Ql9vywQ4UVB37ohYMLurGbT5Sgr+nBmODmRfjDzEl8w1ez32E+9b0QqGAa8b25uNNrgsK+3ioqTLall3vmtqfZy5IYmVKPp5aNZf24A9nQogT0+2WWXuKl5ft510Xu3UvHhHJa5cN6/gBucl3O3MxWayMjw3khZ/308tHx5PnDnRoUG40mTFbrHhqHT+7fLMjh2qjiatGx6BWKfnzYBGZJTVcOira3iGk5ev8+/dDpBfblte8PVTUGM1OS2UXDovAAqw9VMxD5yQwIS4YvVbV7VuB1debSXp2ebstrUJ9PairN3P5qGg+3pyJocHC8Gh/vr3jzA4cabP9eZXMemOt0/GmJh4RfjoKqgw0mG2bZoZE+7Mvt8KeM9nahcMj+dehc8Bk4BnT9Sw2TQfgg/mj+HhjBpuOlNqXXWcnhfHY7IFsTC/hsW/2kBDqw1e3jXP6GRWis5H7d+ckwVwXdd6ba9mb61zCAKBfsBcPzxxgr7nWHRkazFz23kZ2N252eenCwVw1Jua0vma9ycLDX+1i7aEipg4MZem2nHavVyrAagW9VsWbVw5nag9oFffv3w7y+m/N5T2uHRfDxvRSW1/SxiDogqER3D01jv/8cZhALy13T43DT6/t8LEWVho48x+raDhGT1WwLZ3vfGoGry4/0OayrAJ4LXYnF+W+QmXiPN71vYvIAD1Xj2lO8ag2mlApFA6lV4wmMx5q97UvE+JEyP27c5KPgV3UvDExPPrNXnRqJb18PchqURQ5vbiG2z/ZTnyoNxV1Jt67ZiRDo/3dN9jTYPH6DHsgB2C2HF/bpL9Cq1byf1cMB2xlKBrMVtYdKsZHryK9yHmzRFOeXW29mRs/3MaUhBDGxQVTVlPP3OGR9A/1Oe1j7mg3ntWPNYeKScmr5M5JsdwxxVZM967Pd/DjrjwAyuvqySyptXfSWLjuCLMHh/H21SM7dKy9fHVcM6Y3izZkAM5t3FqqN1v55/JUtrroDpEY5kVKfg1W4OEjI5j5WBa+Pn487OJ5vD2c/+RKICeE+KtkZq6LMpktrEjJJynCn5ggTx76chdfbc9xuUPuuvF9eOaC7rOT9/6lyXyzwxYIKIBLRkXx8kVD3Nrkftmeozzw5S5q648vqIwO1LP24SmneVSdh9lsIe6JZfafz9a7fwF2PT0DP33Hl9948acUPtqU6bDztDWNEtprs6oEmk6vuO8s+ofK3zXRPcn9u3OSmbku6m/f7+PzLVn46TVMT+zFH6lFTE4IYdWBIofrlAqYPTjcTaM8taxWK2/+fsgeyAH83xXDmNMJiqnOGhzBxP69WJlSwJfbssirMJBe3Hbrpqbeo9szy6itN3FWfEhHDdUtNh0pdQjeWgdyWpXC3qS+o909LZ73j7Gjtb1ADmz5gHmVRrQqpVuWjIUQPZsEc11UVqktEb+iroGvttuCm9aBHNjKbIzuG9ihYzsd0gqrmPv2BqpbFOkNUVTQmeryenmomTs8krnDIzGazNz4vy1sSC8lzNeDvAqjQwBjNFmZ/vpqDhXa/jteM7Y3z89Ncs/ATyNDg5mtR0p54Mvkdq+rN1t56Zf9vHXViI4ZWAs+Og1Dovwclu2PpfVM3fNzkyisqicp0pdQqQknhOhgEsx1MfvzKvh6ey7XjuuDyWwlxEdLlcHMmoPOgRzATWf16+ARnnpmi5UbF29zCOR8qKHI6sd9S3ZxVlwIAV6dazbEQ63ik5vHORy75/PtfL8r3/51UyAH8MWWLNKLqqmoa2B4TADPzRnULTpIXLNwM1td9E12JcqNtfm+v+NMhjy7wqmjB0CIt5aiascdrJefEcOSrdk0WKwMjvDhrP4hkvsmhHAbyZnrIgwNZtIKqpnzn3WYrbZlqaYyEJMTQrjprH48/NUucssNqJUK/nbuAMb0C2ZAeNf+/pbX1vP5liz+8autq4cCuGBYBJV1Jv44UEiIjwdrH57sVEqks8osrmHK66sxW2zvRa1SuNxNGeKtZfENoxkU4UeloYHfUgrYcLgEbw81j84a0GXe76CnfrU3m3cl2FuLn17D5WdEc9OEfijdmPe4PbOMJVuzmDKgF19uzeFAYRUT+4fwxOyBfLwpk51ZZSzfVwA0F+muN1nQqpXHeGYhug+5f3dOEsx1AW21KGuiVipIe2k2YMvF8tapu029qvmLtvDnwSI0SgXBPh48MCOBS0ZGUVdv5s9DRQyJ8iPcr2u1MiuuMrJkWxa/7y906Cfb2oKJ/Xh89kCmv76GQy0a009P7MUTsxPd2pj+eH2xJYtHv9njdFyjhNH9gvhg/hkOZTo6M7PFyiebMtGolFw5OrpbzJwKcaLk/t05dY87fjdWVGXktRXOgVy/YE97gr1G1XxT6W49HEuqjPZ//3L3WfblVL1WxTmDumYdvWAfD+6YHI+hweIUzPUP9abKYKLGaGJmUhiGBrNDIAewMqWQjYdLAAW9fDy4eGQk3+08yog+/uzKquCcQWHcN71/x72hdlwyMop3VqeR2aJ0zsiYAL64ZSwaVdea0VIpFVw7vo+7hyGEEE4kmOvkft2bh6v+3+9dM5JrFm4hv9LILRNjO35gp8m2jFLu/SKZPsGevHnlCHuz+9he3gR4afnHr6n8llLAPdPiOW9IhJtH28xqtZ7wTM0DMxLYcLiY7ZnlgK0x+1PnDWJCfLD9+cwWK718PCisMjosrVc31kOrNpp4dflBrGAP+lLzq7hkZFSnaCmmVilZ9eBkMktr2JlZToiPB2fFB8uslhBCnEISzHVy42KD8NSqqG3MO9Kplfz7yuHEh/qy8bGpVBtN+Og6vpzD6fL4t3vIKa8jp7yOSa/+QW29LSHdYoXiaiPvrD4MwN9/SeWN3w5SUGlkSLQ//75iOIFu2ASRUVLN/A+2cLTCQC9fD5beMu6EEvkXTIzlvi+S8dGrmTUojN5Btsc2BTsqpYLpA0P5bX8+Ef56dmY777hMCPUmtaB59k4B/Ou3gzxzwSB8O8HPhkqpoF+wN/2Cvd09FCGE6JYkZ64LMJrM1NWb8dFp3FoY93TJLq3lgaW7OFBQRV29mXqzc1Gv8bFBfHzjGC5+ZwPJ2eX46lRUGpoT6y8aHsnrlw/rwFHDe2vS+PsyxyXwv180mCtHn1hbsbp6MyOeX0ldg5nBkX78eNcE+7k3Vx3itRUH23zsgDBvrh3Xl8e+dc5Lu2tKHHOGRRDXy4f//JFGSl4lD85IoG8XyLUTQnROcv/unGRmrgvwUKu6bdmDI0XV3PbpDlLzq+zHfHRqe4kIpcI2Kzc8xh+VUsFXt45j6bYcHm8VvNiWK8sY2TugQ8Z904fb+G1/gcMxf72GKQN6nfBzHS6qoq7BFpg2zUQ2ySxpu/AwQGp+Nc/+tI9Ifx255QaHpdh31xzmzVVpzB0WwXfJRwHIrzDwyiVD6Bcis2RCCNFddK0MZNGt5JbXMfm1NQ6BHECVwcSspDAuHRnF6L6BvDA3iYfOGQDYcrBKqo1Oz5VfaeTaRVtocDGrdyp9sjGD+Md/cQrkNCoFK+6beFIFY1OONr//0X0D+XhTJh9tzGBvbkXjRgfXfHW2z2JWKyxZMA5vD7U9kAPsJU/WHCiyz+huzyxjymtr+HJb9gmPUwghROckM3PCLSwWK/+3su3lw/VpxVQZTVitUFbTwLyxve3n/rs23eVjlApQnsbE+ud+2MuiDZmOr4ltM8qkAb1OenfmrMFhLN+XT0VdA9EBnvztu70ABHhqKKttcLhWpYCmeO3x2QMprjYypl8QUYGeBHlrqTY6F70tq2twOvbQV7t5eVkqH94wmqRIv5Mat2i2O6ectMJqzh0S3m1n0YUQnZcEc8Itftx9lC+357R5/roz+/D19lxyy+uIC21eEnx3zWGXVfoBrhvfhwazBZWy+WZaWGXAQ6XCz/PkNgJszyxlc3opaYXVfLsz1+GcRqVgzYOTiQj4a3XufHQaFl53BgBft/ieVLp4n2fGB5NXbuCK0TFc0So376tbx3HPF8lsaGc2r6WSmnrmvrWOjY9PI8TH4y+8g54ts6SGS97ZSL3Zws6s8m7Zlk0I0blJMCfcomVD9WBvLZH+ei47I5roAE9ie3kT6a9n7rBIvtuZy9pDxTzx7R6uHB3NK786bjgI9tIyNjaIspp6/r0qjR935/Hd7eM5XFRDYaWBOz/fiU6t5Nk5gzg7oRfB3scftKQcreTSdze6LA1zw4Q+PDk78ZR3LLh4ZBRqlYKs4hpe++0QACHeGm6bFE9ZbT23TYptsyB0iI+Oz24eS2ZJDc/+sI/U/Er8PbWk5FWhUNhmLluvQpustgByf34l88f16bCcw+6ktsWmnXIXs6BCCHG6yW5W4RbfJ+dyzxfJgG2p8pFZA5iVFM7DX+8i0EvLNWN7c/3irRgbLPYG9QqgrR/WIC8tJTX1Dv9ufb1aqeCSkVG8fPEQ+7H3/0xne1YZcSHezB1u2/nZZOPhEq58f5PTaz0yM4HbJsWd7Ftvk6HBzEVvr+dgQTWmFhHkxLggPrpp7Ek/b3J2OTqNkl925/HftenMGRpJWa2R9OIaZieF8/bqw5gsVtRKBSNiAnj/2lEOwbY4tsXrj/D5liymDwrjwRkJ7h6OEKeN3L87J5mZE26RX2Gw/9sCLN6QwfJ9+faOCH+kFmJocJxGau9TR1Mg1/Lfra83Wax8sTWb7NJa3ps/ikMFVbz4y377+c+3ZLH1iWkYTGZu+2QHm9Nty5Uqhe25Ajy13DkljuvP7Huib/e4/Lo3n5S8Kqfj0UF/rfjvsGh/AAaE+XJ/q0DDZLbwzc5ccsrqMFmsbMko5fmf9rE9s5xLR0Vx+2kIWrujrRmlHCio5kBBGr/szuPZOYOI6+VNuJ/+pApKdwYNJgv5lQY81Ep+2ZPH/vxKJsQGExHgKTO4QnQyMjMn3MLQYOb+Jcn8sjcfALXCtuTnip9OTUUbeXIna+6wCB48J4Hpr/9pLwvi76lhx5PTWb4vn9s+3eFwfVOf1NPpukVbWH2wyOGYAjj80uzT2oC+uNrIx5syefuPNIK8tORXNu8W1qlh6a1nMiTK/7S9fncw4vmVlLb4QAGgVSkI8NJSazTz+LkDmJEYRpC3B5V1DWhUyk7dk3bJ1mwe+Xp3m+enJ4by7ryR3bLupWif3L87J5mZE26h06h4bPZAft2Xj8XadiAHoFQq0GmUGBoseKiVGE22GbtIPx35VQanPLDjYWiwEBXgybJ7zmLzkRLyyg1MSwxFqVQwMNzxD5RaqWBSQsiJv8gJ8tY5/zpagZS8ytO64zTY24P7pvXn9kmxKFEw4KlfMTcu8xpMcMFb6/nwhtGc3f/0fw+6IqvViqGhuYB1U62/erOVgsbA+LFv9vLK8oNE+uvYk1uJWgEXj4pGrVQwPTGUSQknXp/wVKs2mnj0692s2JdHvbn9a1emFBD7+C/cMTmWu6bEo9N03sBUiJ5AZuaEWz31/R4+2pjlcEwJaDVKe77ceUPC+Wl3HuBYruOcgb2wKmBFSiHgWLajtUERvjw2awBf7cgh0k/PDRP6EtTGZogvtmbx6Ne2osRnxgbx3/mj8PI4/Z97qgwNfLszlyNFNfxvQ4b9+LXjevPsnI7bIfntjmzuW+o4K+OtVXHe0AiHfEPR7Jkf9rK4VdmaJu39XDZ59dIhqJQKZg92T2mTwkoDz/+Uwo+Nv2et+erUWKxWe0/glvz1atY9MgXvTtA6Tpx+cv/unGRmTrhV/1DnPwZKJfZ8uZsm9OWBGQlYrFaSs8spralHrVRgslhZvr+QB2f0J9DLA5UCPt3iuhCur07NK5cOITHcjwnxx55deuGn5jy6AC9thwRyYCtRMn9cH8pr61m8IcOe8zc9MbRDXr/JhSOiKa8z8fIvKTTdu6vrzXyxNZsBYT5cd5pyBruyM/oEuQzmogP11JusFFTackSbOpq09tg3e2gwW9mWUcaLFw4GbEW1i6uMDG3MeTxdXl62n3fXuK7dGB2g555p/blkZBQAX2zJ5NkfU6hrkc9aXmci6ZkVvHxREleM7u3yeYQQp5cEc8KtXJXZbVo2VSpg3tje6LUqbp8Ux3lvrnO6dtG6I5TWNtBefnmlwcTKfYUkhh/fUuWAMB+2ZZYR7K3l2QsGHddjTqWKugaHzRsBntoOH8P1Z/ZlW0YZP+9xnKlJK6rp8LF0BecOCeefyz2d2q9ll9Y5fN06kPPXa7jl7H78o7HkTnpRDYcKqvhoYwafbMrCCtw7LZ57p/W3P6aitoHiGiOxJ9mSzWKxsjG9mHqTmWV7Cljaqt6jh1rJ2f1DuHliP87oE+hw7orRvbloRDRphdVc8NZaTC1SHF76JVWCOSHcRII54VZ5lQaHr/uHelNrNJNTXkeor47ejTs5P9qY4fLxpY1LrsdKFvhhVy73TIs/rjF9dONokrPKGRLtj3cHzco12ZtbwfmtgtaHv97Nz3ef1aHjAGiwNN+pbTszddx6dj9WHyhk7aFirjgjmvhQn3aeoWcor61nwUfbKa+tP/bFrZwzKJTbJsXxyaZMcssNbDpSwvR//elwzffJR+3BXGGlgVlvrKWkpp6LR0TioVZx69mxxBzHjufaehMvL0tlydYsjG0kqZ4/JIy/nTeIXu20pdOqlSRG+PLkuYk882OK/XilwcSOrDJGxMhOVyE6mgRzwq2CvRxnnQ4VVtsDs1BfD0wWKxqVgrAWN5fLRkaSnF3BwcJq+zFfnZpKg4lgby1TB/aioMJAWlE1OWWOweLx8NSqGR8XfHJv6C9amVLgVFKlqOrE38Op8LdzE2kwWQj31/PCnCR25ZTzyvIDfJ98FICF646QEOrDt3eMb7OQcU+wKrWQLRmlJ/XYJdtymDIw1F5s2NWHEh8PJTd+uJXN6aX0C/a0l975eoetI8lnW7K4ZGQU90/vT4S/czcSo8nMO6sP882OHLJazRQ28fZQM39cb+6ZFt9uzt5//kjj59153DopluUp+Q7nWv+eCiE6jmyAEG6z72gF5/7beem0pYfOSeCOyXHUGE3889dUtGolRVVGvmsMKFr67o7xDIn056fdR7m7sSBxk5lJYbw7b+SpHP5pkV5UzTULN3O03IBKCfG9fHjwnASmDuzYvLnWLBYrg59dTo2LBPibz+rLE+cmumFUnUNOWS0Xvb2Boipju7UQ26JVKe0dJP6q964ZyTmDwhyOfbA2nRd+3u/iaiug4LKRUfzz0qHHfO5qo4mkp5cDtq4tZ8cH8/VO2++hRqVg6S3jGC6zct2e3L87p577cVq4XWXdsWvHKRSQVVLL5iO2Ar4FlQZ+2OV6x52/XoPBZOa+Jbvsx8b0DWRi/xCuatXHtLPy0KioMZqxAiYLZJfV2uvguVNTOzBXPtucxcJ1R5g/rg/PuCHH0N2iAjxZeutYJr2y5oQf66/X2GflYgL1FFYaMZhOPrBbn1bsFMwdKqh2caUtkAPwPM5UAi+tyj7ekup6rpvQjwFhvoT4ejB3eNRJj1kI8ddJMCfcZlxskEOpkdbCfD2YN6Y3k15d7VSQ1ZV7lyTjo9NgbjHZ/MLcpC6T17UhrZgbFm91uJlXG83cv2Qn5w2JcOPIQKFQ8MH8M7jiv5ucZp9qGouSfbQxg2BvLQWVRu6YHEeYn23JbdmePDYcLsFTa8vvCvDq+A0dp1uQlwd6jcoh8G6v/RyAj4eaLxaM5anv95FTVkvvIC9enJvENYu2nvQ4dmSWOXWccFWcWK1Ucu+0eIqr67n3GLmk1UYT76xOI62wunk5uPH/3Xx27EmPVQhx6kgwJ9zK7KpOQ6P8SiNphVVUGY6veXlydoXD17EhXiccyKXmV5JdWsfUAb1Oa9cFV178Zb/LWRkPTef4NR3TL4jl957FjR9uIbvM6HTeYoVXVxwE4JsdOax7ZArf7szluZ+ak+S3Z5ahUiiID/PhmrG98dSqiA78a+3KOgMfnYZPbhzN9Yu3UtnYraTpJ/vM2EDWH3bOqRsa5cfKlAJumNCXWz/ZztEKA+vTih3q0qmUCqffEQ+VAmOLwnUty53sPVrJ//12iPumN+9+vfXsWPLK69h7tIJ6k5UALw1PnZfoUKbn1715vLP6MDOTwrltUiz5lQaKqww8+0MKe45WOLXWu2lCXwZHnb5C1kKIE9M57hKix5qeGGpP5HYltpcPZ8UHsyq1iIFh3oT56akyNJBytILahrYDwfGxQXwwf9QJjeVwUTUXvLmeerOFOybH8tA5A07o8X9VsLfrGavaehM/7jrK+UPdOzsH0D/Ml88WjOOsf6xu97qaejMrUvL5bX+Bw/FtmWUAbM4o5bPNmVisto4JY/sF8eENo7tkD9MmI/sE8vzcJO5pla+5I6v5Q0bLmeh1h0tYd7iEMF+dfVav9Webcf2C2JFZRm2LGb+WgdyspFDemTeKd1an2cubZJQ4lo8J89Px3jF+F574di8lNfXsyqngq21ZHC6udXndlAEhJEX4OZRKEUK4n6syX0J0mGcvSELTzk9hVkkte3IrAdifX80780by1W1nUn+MtKJ7p/U/7lygPTkVPPPDPtanFdsT0fMqOm4HaV5FHZ9szGTbEdc7Is0WeODLXbhhr5JL0QFePDQjod1rNEoFj3y9h9KaehTYuiDcOzUeH4/mJb+mwKXebOXPQ8VszSglu9R1ENFVzBkWyZqHJjEjMRS1UkGQl4aogBY7TBXgoXYMWFVKBZ4tfgkCvTQ8ce4ArhnXm4P5VfZATqWwtZZraUJ8CEaTmYrGAFGjUnDThH4AlNbUc+8XO3n82z0O7cZcqTY256+6CuQ81Eo+v2kMi64bzf0zEjp81loI0T6ZmRNu5a1Ts/uZmfx92X4+2uhcQT+ztIa7p8bz7urDzBkWYe8BedOEfny6ORMfDxVHK1ou+Vl5cEYCo/sGOj1XW275eBtHKwz4eKh56rxE0ouruXvq8dWka81ktrBw3RHMVis3n2W7qX65LYdePh5Ma9HJYXN6CZmZGeiy1nDP/gSXuVV6jZK+wd6k5FUyKMK3U81a3TEljgnxwdzzxU4ySpxv/g2NkVpxdT1/PjyZrRmlbDlS2m4O2WXvbQLg8dkDWDCx6+Zi9Q7y4r8tZsIKqwy8vCyVb3bkUlbjmDIQHaAnKlDPxsZl2IRQbz6+cQy9fHWsPlDIx42/E01dT1rWLukX7MWzP+zjxZ9TMDXO1jWYrXy8KYMQHw+q6kz2Xd9Do/y4/Iy2NwGNiPFnY7rzhwl/vZpxsUE8PyeJYB8pOyJEZyXBnHA7vVbFnZPjSM4uZ3eOY97b33/Zz/pHpzKpfwg/7DrK3twKkiL9eHTWAB6Y0Z+EJ5a1ejaFy3ZJ7fHVazhaYaDebGHawFA8NOHc9dkOzBYr/7lqJKF+x38T+3J7Dn9flgrA1iOlHC2v40DjbsI+QXqGxwTaCsV+sgMroCTOKcBRKmzJ8XdNjWfe2N7sO1px3N0rOtLQaH8enz2QBR9vb/OaFy8chFql4KGvdrebH9nSP5alEhPoycyk8FM1VLfq5aPjyXMT2XqklOwyxzpv2WV1ZJfVodcoqWuwcKiwmrTCanr56hgXG8S5g8PZmF7icgOQXquk3mylvlXj16XbbB0dejfmIqqVCuJ6tZ87+vmCcfy0+ygBnlr+SC0kMcKX84dGoFHJ4o0QXYHUmROdRr3JwvDnVzjVMgvwVGO1Kiiva8BXp2b736ajUSmxWKyMemGlvQuEXq0k2MeD964ZSWLE8Qc/S7Zm8cjXewAY3TeQPTnl9t6TOo2SzY9Pw09/fE3Ef9p9lDs/29nuNVq1kvpWGx2adj6qFLb/1WtV/HzXBPoEn1zLpo704YYjPP/TftvMkQuBnhoqDaY2z7uiVio4+MKsbrGc9+nmTD7akIFGpWTvUVvKQN8gT6xAdmktrWIxLhoRyauXDGVVaiExQZ6sSMnn1eUHT/h1+wR58s68keg0KvoGe52CdyKE3L87K/nYJToNrVrJbS5KHZTVmuwlEVpSKhUsuWUcmsYbfp3JQnZZnVPB4GM5o08gWrXtV2F3i0AOwNBg4ZqFmzEdZ1HXQE8tN5zZp91frNaBHIC2MY/KbLXlktUYzbzwi6tCr53PteP7svWJaWx8bAovzE1yOl9a20Cw1/EFw01MFiupBZWnaohu9cJPKRwoqLYHcmqlgroGMxklzoEcwIp9+Tz9w15u+mgb5725jlX7C0/qdc/o48/AcF8J5IToAWSZVXQq14zrw4GCavIr6tiaUeZwTqmA1y4bal/6WXOwiJJqoz0/q0lGcTVmixXVcc7q9AvxJsTbg9zyOqcSDAC7cyr4LaWQb5NzGB8bhL+nljBfHWP6BTlc9/PuPO74bMcxX29cv0Cn/CQfDzVGk2PAGnCcs4GdQVPtuHlje+PvqeHuz3c6LHfnVzUvE+o1SvRaFaU17Zec0Sq7x2fNqABPDjW2nlMqbIFqfqVzaZcm1UYzX2zJBmyB/56c8pN63ZJjfH+FEN2HBHOiU/HTa3jzyuEAbD1SwjM/pnCooIp6sxWLFd5alcbI3oHMX7SZvbmuZ25MFnh1RSqPzBx4zNdbd6iYslojo/sE8G1ynb3Ha0v9gr14Z00au3IqWL7PVmpDoYBvbhvv0L4ov8J138uW+gZ7salVIDc82h8vDxXr0krsxxLDfbmpcQNFV3PekAj89VrmLdzs8rzFakWnVgGOwcY5iaH8tr8As9W27NwvpPMvMR9LeW29PZADnPI52yos3PIDSsvPF0oFXD4qiqzSOrZllWJsLM+jVinsmyCaDDqBVAMhRNcmOXOi0ztYUMXM//sTi9UWDPUN8mTVgSL7+UAvTZuzPN4eKr6/80xiQ5wTwDekFXPVB80Bh6dGSW2rmTmlAj66YQw/7Mpl6bYchyr/H984mrMaC6/+49dU3ltzmHA/Hf6eWvYdrcRDrcTYYklVAcQEepLZWH6jZc2xUF8PClrN1gwI8+HXeyce77ep03nz94O8tvIQABqlY1ByLBH+OlY/ONm+/N1VmcwWZv97LQdbtdS6bnwfdudUsCPLNvusUyswmI79p/jRmQncOikOgOd/3MfC9Rn46zWsfWQyaYXVhPrqWLY3D4UCrhvXt1vkHIrORe7fnZMEc4KcslrUSqW9/VJn9EdqIcv35XPVmBgWrjvC940lFwaF+7Lk1rEMfWaFy/wjsCWGTugfTFGVkRfmDmZkb9ts2n/+SOOV5QccrvX2UDvU3Hr/mpFMHxSGyWxha0YZ4f46vt6eQ5ifjqvH9MZktvD3Zal8vDHToVm6t1bFkMYK+aW1DVisFg4WOBZzbRnsTU0IYWdOhcOuxehAPWsfnnJy37BOYvm+fF759QBpRdV4aVX21l9NlApbnbUGF//xLh0Zwa6cSg4XVhMX6sP3d5xpL03TlVTU1TP2pd/tuZgKbBtrwJY7p1TAA9P7E+Cl4R/LDlBhaLtn8cUjInntsmH2r9OLbAGc13HWVBTir5L7d+ckfwF6qGqjif1HK1jw8Xb77NB/rxnJjFZNujuLyQN6MXlALwBev2wYQV5arFZ4ZNYAdBoVvXx15FUYHFobNbEAfx4sBuDvy/Zz95R4evl6UOliU8VLcwdx39Jd9sDwlk+2890dZ6JSKvDTa+gT5MUDLQrmrkgpYOG6I4Djkll1vZkNLZZT+4c6Lxk2fY7y0qp47bJhlNbW8+6aw4T66GiwWJk73P0dH/6q/qE+pBXZZqVq6s1oVUp70BvuZ/tvZjFbSQj1ZlTvQNYfLrbXrfty+1H78xzIr+LLbTlcM653x7+Jv+jfv6c5bKqxgv3rUB8PCqqMvLLiIBcNj2RItB9rD5W08UzOxay7w1K0EOKvk2Cuh1m8/ggL1x8ht7HeVcvA54N16Z02mGtJpVTw1PmDHI59fvNYVqYUEOCl4eVlqZjMFhQKhT1QbWI2W5m/aAtqpYK3rhpOhJ+Ooy1ukJ9vzSEhzIeUvCrA9v15feUB1jQGgwsm9mVVSiGZpbVo1UqmDQxFo1TQYLEyIiaAfXnOfSzBtmM2tpc3W46UUlJtm31rqg/2xhXD8ffS4u+l5Z+XDD1136hOwFOrQqNqnnnTaxSYzLYAO7/S9n1XAPdNT6DBbOHTLVltPteLP6fQP9TbaeNJZ7d8b36b56YmhvLZZtt7/i45t80aiU2LpZefEX2KRyeE6A4kmOuGDA1mXlmeyvaMMuaP68NFI6MAsFisPPdTSps3jOHRAa5PdAF9gr24eaJtw8AlI5tveBsPF7Mnt4L9eVVEBegpq61nZ3Y5JosVD42Ksf2C+GZnc2/YjeklBHk190hVK6FvkDerrbZg7r01R+zn6s1mvks+yuWjoliyLYftWY67b5uc0TuAp88fhFat5N4vdtqr8oNtmbFPcNdvNN+WUF8dH14/2p6bWGGwLbM2BcAAvno1M5PCuPPTpuLDTT+gjvleBpOF75KPdrlgbkRvf3LK6/DUKDGarfbiya9cMphLR8Uwpm8ge3Mr7DO8rvztvETOTgghVmbihBAuSDDXCZXX1pNWWM3rKw+wLaOMyAA9HmoVz81JOmabKqvVyvi//24vpJv85S50WiWzB0egVCoYFxvE+rQSvLQqhsX4M7J3gK3ivI/upFtYdWbjYoMZFxts/7qk2ojFCpH+eqID9A6BXKCnBnPj0ueEuCCuG9+XqQN7UVtvpqbeZK+s39rXO3KclncjA/QMifAjOsiTe6bGU1Rt5KWf9xPgqbHnjo3tF8jdU+KPWZ2/qxsfF8xrlw7l4a+bu0A0WKwoG3uNBnt78N3ObJbZZ7DaTtq/cHhkB4z41Hrl0qFcMDSSgRG+PLh0FxvTbcuoA8JsOZVzhkUyZ1gk42ODWb4vn+Tsckb2DuCOyXFkFNcwKMIPP8+uU6ZGCNHxZANEJ5NdWsusN9Y6JOE3iQvx4rcHJmGxWKmpN+Gja/4DbzZbuPPznZTX1jvVMPvbuYnceFZfwLa7LqesjuhAz+Ouw9ZdFVcbmfTKaqqNJgK9NFgsVsrrmr/vZ/cPYeG1o1CrlNz56Q5+2pPX5nO1XEps0jfYk9hgb1Lzq4gJ0rOhsf/mv68YRmSAJyNi/DtVv9XT7YfkXO5dkmwPerUqhVMrqiYeKgXGVuci/XWsf3Tq6R7maVVR28DCdenEhfpwwdCunxPpTtVGk72Y9+/7CxndN5DowO47y91ZyP27c5KZuU5mT26Fy0AO4HBRDcXVRhZ8tI0dWeWE+Hhw15Q4Zg4K4+aPt7Er29bXVK201VoDCPP14Poz+9ifQ61S0kcqwgMQ7O3BL3efxY+7jvLKigNO59ccLGLl/gJKqoxOgZxOo3TIjWswW1ErwWxpXiQ8UlzLkWJbMn9OeXMNuvxKAxcM63ozTH/VBcMiSYr04/0/01mZku9U1LblBhKj2cqo3gFsy2xeunbVHaSr8fPUcH+LDTTi+JXW1PPemsP0DfZieIw/F7+zAUODBZVSgdFkIdhby6bHpqKWfrKiB5JgrpNZlVrQ5jm1UkFBpYEdWeUAFFUZeer7fTz/U4rDrJBKqeTJcwcQ5qdnZlLn39DgTjFBnlw5JoZPN2c6bIRoctsnjh0d/HQaKgwNmM1W1EqFQ7/RpgC6ZT6YK5vSS1kwsesHJiejX4g3HhoVxS7qArb+jrUM5AAWrj9CsI8HM5PCT+MIRWf18rL9LlMdmn4HS2rqMVmsqLte9Roh/jL5CNOJ7Mou56vtuS7PhfvpWHLrOBLDfblkZBTqFkukrZf3zh8SznVn9pVA7jgFemn5/YFJJEW0v2SQGO7DiN7+gC3n65nzE/HTa/DUOt49mgI5rUqBq0mCph2J2aW1lLWoK3ci9uZW8EfqyfXsdLez4oM5mRX+I8W1/N9vh079gESnV1hp4GBBVbvXWK222o1C9ETyk99J/PPXVB7/Zo/Lc0oFjOkbSEpjo+5XLx3KvufO4fbJzrM7L12Y1O3KW3QEvVbF83OTGNM3sM1A4+kLEmmZ4hbby4ddT89g51PTmTMsgphAT84ZFEqYrw6FAu6eFo/ZRdeDMF8d3+zIYeIrfzDxlT/ILKlxvqgde3IqmPOf9Vy/eCvvrTkM2HYw784px2gysz+vkqVbsymtsXWUSM4u553VaRgbzO09bYeZOjCU7U9O52/nDiQ6QO/ymrZivSmNtQZFz/Lw17tJbkwjaUvvQM8elYMqREuyzNoJ7M2t4O3Vh9s8H+LjwXfJR/ku+SgH8qt4+vxETGarvT5Vk+Ex/lw2Klpa+Jyk4TEBLLllHJe9t4EtR5zLjPzfyjTiG4v/+ntq6NuYe+ihVvHGFcPt1xkazFQZTKiV8Oryg07PU2M0seVIKVYrVBlM7M+rpHfQ8eUx1tabeOyb5l2hTUVk532wmW2ZZSSG+5JeVI3BZIGv4ZKRUXy13bY0tXRbNvPG9mFyQojbi80GeGk5q38Ib/2R5vK8q0XqX+6eQKL0G+2RjmfGLa+ijhd+TuGq0TFu//kWoqNJMNcJRPrrCfHxoKjKsTenSgFmK9QYm2dUPt6UiaeHCl+dhvIWBXEVwFtXDpfk31PgjSuG887qw3y2OcshJ25jegm7c8sJ99Nxycgol+3P1qcVc8enO/DyUFHRYmesUgHRgZ4smNiX8XHBhPvrySypJdxPZ+9scTzWHSpmb+MMbWyIF/dOi2fNgUJ7fllKXqXD9T/uaq5pl1Fcy/M/pfD2Hxo2PT4NjZt/Vq5+f5NTUee2BHlp6B/avUu4iLa9culQRkSm8Y8VaVgaF5SGRvqyK7f5573ebOWDtUdYfaCI3+4/211DFcIt5M7fCQR4aVl530TiQxy31TelwlUbTVwzNsZ+vLKuwWEL/pmxQbx66VAiA2Rb/qkQ7qfnuTlJvD9/pNO5GqOZvAoD76w+jKuqPg8sTaa8roHccoPDruS4Xt6seWgyV4/pA0DfYC8+XzCW1y8fhkdjxnZpTT33LUnmqe/3YjS5XhIdFuNPhJ8OtVLBVWNi+GJrNtf9b2ub7yWqxTKmp4ftdUpqGjj332tZe6iIr7fnsDm97fZRp5PB5GINulHrP0xBXh7yQaUH89VpuEX5PZ9oXmK0Yj83JDS4nL0FOFxYTXLjJjEhegqZmesk/D211NQ739zUSgX3TovnzinxJEX6caS4llvP7oe/p5aD+ZW8syadKqOJ2YNlh9+pNjQ6AJVSYV/SbEmnUfHAl7t45ZKhDvX6wvz05FfaZlgDvTSUNu7anHUcOzA/WJvOt41FjJMi/bhslG2jRKWhgYe/2k1GcQ1Pnz+INQ9PZmVKPvd+scve57SJClCpFAR6aRkS7c9DMxK447Md6LVqHp81gKs/2IzJYuVgQTXXLNwCgEIBP945gaTIjlvC/HRzJjq1krZS2i3A5IQQVh8owoptuVj0cF4hjFelMF6VAhO/Y8wXRpeXWYGL393AukcmE+7nOidTiO5GgrlOpNLgXF/ObLGyPq2E2ybFcfkZttm5tMIqfth1lC0ZZZgtVnbnVHCgoIph0f4dPOLuLdBLi6dWRVWr/y5alYJqo4lvduRy9ZjepBdVcyC/ijPjggnw1DA9MZRQHw8+aZHT2Po5XEkIsy0jqpUK4no15/zc8tF2e9eAGxZv5dFZCTz3QwpNc3ceaiXGxlkuM7b+swqFgv9eMwqA5fdO5JNNmWzLLOPZCxJ5feUhSlrsorVaabO24al2tLyO7NJaXl9x0GEMYOvAUdpi2fXms/rxr8uHUVxd7/D9ED2DxWIlr9JAmK/O9oHpjBvBNxI8fDDHjGdYzHaW73NdyslssTqkpwjR3Ukw14lcf2Yf3lqV5rB8YMWWq7VoXTrXju+LFSuXvLuR8toG9BrbslmwtweJ4VKJ+3QYERPAmoNFDseauhZEB+j4zx9prGosEfLl9myHPDmw5TKO7hvIrZP6HfO15gyLJK6XN3qNyiGBO+Vo8y6+ugYzT/+Q4vA4o4vlyoJKA/9acYCBEX6YLVb+9v0+AO6eEsfvD5zNhxsyKao2cDC/ilmDwxnbL4iK2gb+9dtBevl6cOnIKFamFDI+NuiUFZnOrzBwzr/+pMpook+gnpIa0KmV9uXWloHcxPhgxsUGoVAo8PfUtvWUopsxNJi5fvFWDuVXUdtgprbezFnxwXx84xjbBQkzAXj3jzSXgdyAMB/C/HTMbfxdEqKnkGCuE3lgRgJ+eg0v/Lzf6dyLv6RyoKCa5+ckYWzsPFDXWGqiuNrIV9uzuWpM7w4db0/w/vxRLPhoK6sPFjudyy4zkF3WXGi4dSAHoNco+WLB2OMumTDIxW7N84ZG8GmrncvHYrHCG6tsO0UfatFx4M1Vafy0O49xsUH25zxUWE1iuA+/7i1g8YYMAN7+I41qoxlvrYof755A3+C/fmPMq6ijqnEGMKPU1hHDVd7cNWN78/T5iVJmoge689PtbDzsmMO59lAxDyxNxstDzaOzBuCpVbucSZ4QF8Si60ajlVpzogeS3qydTIPZwsvLUll3sIgDhdUO58b0DeRIcQ2lNUa8tGoqWizdaVQKDr04u6OH2yP8ujeP2z7Z0WbC9bEsvv4MJiWcfH00q9VKcnY5V/53U7ubBtqSFOHLDRP68vxPKe3uHm25XNuSTqNEo1Ry1ZgYHps98IRfv0l5bT2TX/2Dstq2l3Q/u2kM4+OCT/o1RNdUb7Jw/eItrE9z3ozjqVFS2/gB9sFz+nPn5PjGEj17WHOgiH4hXvz7yuFEyQawDiH3785JZuY6GY1Kyd/OSwRgZ1YZ2zLK0KgUpOZXoVTA5iO2Zu0VrXKwGsxWqgwN+Og0HT7m7m5mUjirH5rEx5sy+GBtxgk9VgH08nEuYXJCz6FQMDwmgKW3jOPepTtJL6pt89qmVmIt+5zuPVrJD8m59kBueLQfDRYre3Mdy5i4CuQADA0WDFh47890Ar20nBUfwgdr07lkVBTjY4OxWq32WnaXjIxyOaP27prDvLwstd33GRfiKYFcD2RoMPP6igMuAzlw/LkMbFxy99SqHWo7CtHTSTDXiQ2PCWB4TID96z8OFPLZlmyX1+o1SoddleLU6h3kxZPnDsJktrJkazZ1DY6Bz/jYQPqH+tqXKQE0Kvjf9aNJPEabsOM1JNqfVQ9MZv6izfzZatnX20NFtdFMg8XKwHAf9uc57hM9XNTcZcJgsvLyRUnM+c+Gdl/Pz0NFgxVq65sTyf+3/givLE/FZIHvknMZ2TuAepOFXTm2vD6rFS5rbFcGtiT2Z3/cx9Jtrn9uWzp/qOxY7Wn+sWw/765Jb3fWW6lUkBTuy/Bof0klEaINEsx1IZMTevHFgrFkFNfw4k97qapv/hM4f1xvPLXyn/N0e+aCJJ65IIkP1qazJ7eC8f0CCfbRMXVgKGDbeLBsbz4AH10/hnGnYabpoxvGMOXV1aQXNwdo1S127qUXVTs9xmyxoFUpqTdb2J9XSVSAJ2qlwqEoskalIMBTQ2GVbZdpRYvn9NKqqDdb7GVXwJaXtzXDsVPGxvRiNh0pobDSgEqpYH1aicNrtMfLQzqk9yR7cst5Z036Ma8bFu3Pl7eO74ARCdF1Sc5cF3X5u+vZnFFu/9pTq2T30+dIYVU3azBb+GZHDpH+nkyIP31LhvuOVjB/4RbKaus5nlhJp1bip9dQ0Nhl5NGZCbz86wGX1543OJxle/PsRasB+gZ5cqTE9fKuj4eKW86OZUVKAbtz2u+f2VKgp4b4UG8azFYmxAVzx5Q4ewFl0b39vDuPjzZm2NNGVAp4fm4SA8J8OVRYTV5FHf9bn4HVauWzm8d2aA1E0T65f3dOMpXTRd0zPYGr3t9s/9rYYKHBbEXuhe6lUSnt9QBPp0ERfmz/23T+/st+3vvz2LMbMUGePDcniffWHGbKwFAWrmv7MT/tyUOtVNjWTBv1C/FmaLQ/3yXb2oP56tRUGkzcOTmWm8+Kxc9T47T029IDM+JJzaumpMbIXVPiMDRYmNg/xO0txUTHqjI08O3OHJ75IQWLFfQaFfGh3jx9fiIjewcCMKJ3AO+sPkxFXQMalQKdRn5GhDgWCea6qPGxwXx0/WgWrk/HZLZy7fg+6LUSyfU0Cyb2I7e8Di+tmvOGhPPznqN8sdW2GSEmUI/VCteN78PFI6Pw99Qytl8QAF825rBpVQp73byWmpZGg7y0zBvbm5sn9sPbQ83lZ8SwN7eCS0ZGoVQq8NM3b7h58cIkLnhrvb1kjp9OzU1n9ePacX3w9ZSNOQKufH+Tw8Ybo8nM0lvGodM4/u3KaEwhaDBbufr9zbw3f5QURReiHbLMKkQ3U15bjwIFfu0EULnldXy3M5ekCF9u+2S7vfQDgEqp4OoxMdw9NZ5ATy3KE9hYU15bz9Yjpei1as6MC5JaccLBgCeXOZXXOSs+mJyyOnJKa5gQ34uiqjqKquvx06s5WGAL6i4dGcUrlw51x5BFK3L/7pwkmBOihyuuMnDX5zvZmG7LX5qRGMp/549y86hEdzTs2eWUuyiu3ZYwHw9Kaxt4++oRTEsMPY0jE8dL7t+dkyyzCtHDBfvo+HzBOCpq60krqmZwpL+7hyS6qWmJoXy1Pfe4r3/r6hEkRfo5LcMKIRxJZqkQAgA/Ty0jewdKOyRx2rxyyVBCfT2O+/p//XZQAjkhjoP81RZCCNEhFAoFL180xOl467RMhf245FwKcTxkmVWcFku3ZfPnwSJuPTtWakQJIewmD+jFeYPD+WlPnv1YQpgvCyb25evtuVw9JoZwfz1bj5Ry0YhIN45UiK5DgjlxyhVWGXjk691YrZBZUsuPd01w95CEEJ3Im1cNx/sbNdszS5k8IJSbzupLLx8dFw5vbukmpUiEOH4SzIlTqsbYwLTXVtvrzfYN9nLvgIQQnY5CoeDli52XW4UQJ0dy5sQpsyq1gDNfXkWlobmn57MXDHLjiDpGjdHEr3vzyK8wuHsoQggheiCZmROnRGGlgRsWb3M4plIqCPDSumlEHefidzaQml+Fp1bF65cOZVxcsENnBCGEEOJ0kmBOnBI3frjV6diTswe4YSQd65GvdpOaXwVAbb2ZWz/dgVoJt0yM5aGZ3f/9CyGEcD8J5sQpUdmqqnu0v47rJ/Rz02g6zg+7nAugmizwn9WHUSphc3oZV4+NYc4w2ZUnhBDi9JCcOXFK/HDnePx0ajQqBS9flMTaR6e6e0gdYtrAtlsMvbnqMFsySnnwy2Te/P0QbuicJ4QQogeQ3qxC/EVbjpSwLbOM/r28+WBtOpuOlLm8rm+wJx9eP4boQL00oBdCdEly/+6cJJgT4hSqqG1g9Eu/YTRZ2rxm7rAI/u+K4R04KiGEODXk/t05yTKrEKeQn6eGB6b3b/eaH3fl0WBuO9gTQgghToQEc0KcIJPZwt7cCurqzS7PXz22N3OGRTBtYCh/PHA2l46Mcjhvtlo5XFTdEUMVQgjRA8huViFO0D1fJPPznjwSQn34/OYxBHp7OJz38lDzRotl1OfnJrFiXz4VBtuO3/heXvQL9u7QMQshhOi+ZGZOiBO0M8u2weFAQRVjX17FhrTidq/XaVR8e8eZDI32Y+qAXnx7xwS0avnVE0IIcWrIHUWI4/T55iwuf3cjeo0KP71tUrveZGHD4ZJ2H1dUZeT2T7aTVVLL5WdE4+0hE+JCCCFOHbmrCHEcPtyQwdM/7HM4FuKtpU+wF1eOiWn3sa+vPEBqgS1H7pZPtpP+0mwpTSKEEOKUkZk5IY7D//120OlYUXU9E+KCifTXt/m4oiojv+7Ns39ttcKunIrTMkYhhBA9kwRzQhxDRW09ZbUNLs/967dDTjtTLRYrtfW2zQ43f7SNstrmVmdKIDqg7eBPCCGEOFESzAlxDK+ucJ6Va6JRKXjq+70cyK8C4KfdR5nwz1UkPb2cL7ZkUW1sDuQCPdUsv28iQa12vwohhBB/heTMCXEME+KD+HhTpstzDWYr69NKeGX5AdYeKnLo/PDznjzeuXoEX23PYerAUEb3DeyoIQshhOhBJJgT4hjOGRTOFzeP5ZudOYyIDuDFX/ZTZTShUirw9lBTUddAfkWdQyDnr9dw44S+xIf68NjsgW4cvRBCiO5OgjkhjsPY2CDGxgYBMHtoONszShnZJ5AGk4X8SgNPf7/X4fptT05DrZIsBiGEEKef3G2EOEG+Og2TB4Tiq9MQ5O3BoAg/np2TRIBejbdWxa/3TJBATgghRIeRmTkhToFBEX7sfPocdw9DCCFEDyTTB0IIIYQQXZgEc0IIIYQQXZgEc0IIIYQQXZgEc0IIIYQQXZhsgBBC9BiZJTW8/2c6MwaFEuDpQd8QL7w95M+gEKJrk79iQogeY9rra2gwW/lkcxYAfYM9WX7v2WjVskghhOi65C+YEKJHOFpWS4PZ6nDsSHEt/1i2H5PZQlGVkbTCKjeNTgghTp7MzAkheoTCaqPL4wvXZ7Ats5TDRbVUG028dOFgFAr4cls2M5PC2He0kqQIP26e2K+DRyyEEMdHYbVarce+7NSqrKzEz8+PiooKfH19O/rlhRA91HX/28KGtGJ0GhWVBtMJPfaLBWMZ2y/oNI1MiK5B7t+dk8zMCSF6jMXXj7b/+8lv99hz5wCUCrC089F20bojEswJITolCeaEED3SU+cPYkiUPxV19fy6N5/tWeXtXh/pr++YgQkhxAmSYE4I0SNp1UouOyMagMIqY7vB3Ni+gTx+7sAOGtnxqzdZeO6nfZRUG7l/egLxoT7uHpIQwg0kmBNC9HgPzxzAV9tzKKttcDr3z4sHc8nIaJRKhRtG1jar1cprKw7wySbbUvGyvQX0D/XmnEFhXD2mN2F+OjePUAjRUaQ0iRCix9OolKx6YBJR/o4BUKiPB5edEdPpAjmAf/+exnt/pjscO1hQzZur0pj1xp9uGpUQwh0kmBNCCCDAS8u6R6cyPTHUfmzGoDA3jqhthVUGFq5Lb/N8WW0DG9KKO3BEQgh3kmBOCCFaePmiwQR7a/FQKxnVO8Ddw3FpzYGiY5ZWqTaeWOkVIUTXJcGcEEK0cCC/iuLqeowmC/9Znebu4bg0MLz9jQ5TEkI67ayiEOLUk2BOCCFaSIxoLoSaXlTjxpG0LSHMF0+tqs3ze3IrWC/LrEL0GBLMCSFEC/6eWmYPts1qzRoc7ubRuKZRKXlxblKb54uq63n2x30dOCIhhDtJaRIhhGhkMlu48/MdrD1YjLeHijM6ac4cwIUjougT7MWunApWpRbw50HbTJwCsAJDovzdOTwhRAeS3qxCCNHo7s938MOuPPvXOrWCfc/NQtVYmiS/wgDQKWu4ZRRXszG9hMGR/lisVpIi/DplSRXRtcn9u3OSmTkhhABKa+odAjkAg8nK377bzZPnDeKXPXk89s0eLFZ4d94Ipid2rg0GfYK96RPs7e5hCCHcQII5IYQA/PQahkb7syu73OH4Z1ty+GnXUSqNFvuxB7/cxfYne6FWSdqxEML95C+REEIAKqWCr24dxxc3j6H1RtGWgRxARZ2Juz/fQWp+ZQeOUAghXJNgTgghGmlUSsbGBpP6/GzCfD3avfaXvQXc8vH2DhqZEEK0TYI5IYRoRalUsPSW8Rxr+4CXVjJVhBDuJ3+JhBDChZggT96+egR/HCik3mTBbLawcn8hBpNtyXXOsAiemD3QzaMUQggJ5oQQok2zBoc7FQ6uqzdTZWygl0/nK08ihOiZJJgTQogToNeq0LfTSksIITqa5MwJIYQQQnRhEswJIYQQQnRhEswJIYQQQnRhEswJIYQQQnRhEswJIYQQQnRhEswJIYQQQnRhEswJIYQQQnRhEswJIYQQQnRhEswJIYQQQnRhEswJIYQQQnRhEswJIYQQQnRhEswJIYQQQnRhane8qNVqBaCystIdLy+EEEKIk9B03266j4vOwS3BXFVVFQDR0dHueHkhhBBC/AVVVVX4+fm5exiikcLqhvDaYrFw9OhRfHx8UCgUHf3yQgghhDgJVquVqqoqIiIiUColU6uzcEswJ4QQQgghTg0Jq4UQQgghujAJ5oQQQgghujAJ5oQQQgghujAJ5oQQQgghujAJ5oTogSZNmsS9997rdPy7776z7zBfvHgxCoWCgQMHOl23dOlSFAoFffr0cTpXV1dHQEAAgYGB1NXVOZ3v06cPCoUChUKBp6cnSUlJvPfee/bzeXl5XHXVVSQkJKBUKl2OUwghRDMJ5oQQbfLy8qKwsJCNGzc6HF+0aBExMTEuH/P111+TlJREYmIi33zzjctrnnvuOfLy8ti9ezdz587l1ltvZcmSJQAYjUZCQkJ44oknGDp06Kl9Q0II0Q1JMCeEaJNareaqq65i0aJF9mM5OTmsXr2aq666yuVjFi5cyLx585g3bx4LFy50eY2Pjw9hYWHExcXxwgsvEB8fz3fffQfYZu7eeOMN5s+fL0VJhRDiOEgwJ4Ro14033siSJUuora0FbMuvM2fOJDQ01Onaw4cPs3HjRi677DIuu+wyNmzYQHp6+jFfQ6fT0dDQcMrHLoQQPYEEc0KIdg0bNozY2Fi++uorrFYrixcv5oYbbnB57aJFi5g1a5Y9Z27mzJkOs3qtmUwmFi9ezJ49e5g6derpegtCCNGtSTAnhDimG264gf/973+sWbOG6upqZs+e7XSN2Wzmww8/ZN68efZj8+bN48MPP8RsNjtc+8gjj+Dt7Y1er+eOO+7goYce4pZbbjnt70MIIbojtbsHIIToeL6+vlRUVDgdLy8vx9fX1+n41VdfzcMPP8wzzzzD/PnzUaud/3QsX76c3NxcLr/8cofjZrOZFStWMGvWLPuxhx56iOuuuw5PT0/Cw8OlR7MQQvwFMjMnRA80YMAAtm3b5nR869atJCQkOB0PDAzkggsuYM2aNW0usS5cuJArrriC5ORkh/+7+uqrnTZCBAcHExcXR0REhARyQgjxF8nMnBA90O23385bb73FHXfcwYIFC9Dr9axcuZKFCxfy8ccfu3zM4sWLefvttwkKCnI6V1RUxI8//sgPP/xAUlKSw7lrr72Wc889l6KiIkJCQo5rfMnJyQBUV1dTVFREcnIyWq2WxMTEE3ujQgjRA0gwJ0QP1KdPH9auXcsTTzzBjBkzMBgM9O/fn8WLF3PppZe6fIxer0ev17s899FHH+Hl5eVyE8PkyZPx8fHh448/5v777z+u8Q0fPtz+7+3bt/PZZ5/Ru3dvMjIyjuvxQgjRkyisVqvV3YMQQgghhBAnR3LmhBBCCCG6MAnmhBBCCCG6MAnmhBBCCCG6MAnmhBBCCCG6MAnmhBBCCCG6MAnmhBBCCCG6MAnmhBBCCCG6MAnmhBBCCCG6MAnmhBBCCCG6MAnmhBBCCCG6MAnmhBBCCCG6MAnmhBBCCCG6sP8HNNGE4fEwD20AAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "color_recipes = ['target', 'source']\n",
    "\n",
    "# Scale the data and perform PCA\n",
    "sc.pp.scale(adata, max_value=10)\n",
    "sc.tl.pca(adata, svd_solver='arpack')\n",
    "\n",
    "# Compute the neighborhood graph\n",
    "sc.pp.neighbors(adata, n_neighbors=15, n_pcs=40)\n",
    "\n",
    "# Compute the UMAP representation\n",
    "sc.tl.umap(adata)\n",
    "\n",
    "# Add a column to indicate whether the recipe matches the target\n",
    "adata.obs['is_target_or_source'] = adata.obs['recipe'].isin(color_recipes)\n",
    "\n",
    "# Plot the UMAP, coloring by the new column\n",
    "sc.pl.umap(adata, color='is_target_or_source')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "1e6a0455-406a-42da-b2ef-0f5e33aa5b92",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['recipe', 'filename', 'REL', 'STAT5A', 'FOS', 'ETV6', 'source', 'MEF2C',\n",
       "       'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B', 'target',\n",
       "       'is_target_or_source'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2361108b-00a2-42a0-9232-5e2fa319dfa4",
   "metadata": {},
   "source": [
    "# Digital Laboratory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9207f723-9208-423c-bd56-a5dec7e16a85",
   "metadata": {},
   "outputs": [],
   "source": [
    "from BRAD import llms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "204f6fdf-d10c-4bd6-a7ff-bf69922b3940",
   "metadata": {},
   "outputs": [],
   "source": [
    "from BRAD import brad\n",
    "from BRAD import llms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2cbee5e2-1c07-42b0-ae0d-20465f405f70",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter your Open AI API key:  Â·Â·Â·Â·Â·Â·Â·Â·\n"
     ]
    }
   ],
   "source": [
    "llm = llms.load_openai(model_name='gpt-4o')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e3d3a3a-c16b-4ab3-8b7d-d33044af46e5",
   "metadata": {},
   "source": [
    "## Master Bot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "608fff22-b096-4f51-9d04-94788b03fd16",
   "metadata": {},
   "outputs": [],
   "source": [
    "initial_master_prompt = \"\"\"You are BRAD, a bioinformatics digital assistant, and you are responsible for\n",
    "organizing a team of digital assistants to analyze a cellular reprogramming experiment. Your job is to generate\n",
    "instruction sets that can be sent to 5 separate bots called BOT1, BOT2, BOT3, BOT4, and BOT5. These BOTs have the\n",
    "ability to:\n",
    "\n",
    "1. write and run python codes\n",
    "2. search information on arxiv, bioRxiv, PubMed, Enricher, or Gene Ontology databases\n",
    "3. search a literature database\n",
    "\n",
    "You must provide instructions to the BOTs to investigate the following experiment:\n",
    "\n",
    "Experiment Protocol:\n",
    "1. 10 Transcription Factors (TFs) were identified from Data Guided Control to reprogram Fibroblasts to Hematopoietic Stem\n",
    "Cells (HSCs).\n",
    "2. single cell data was obtained for Fibroblast and HSC data\n",
    "3. 252 perturbation recipies, consisting of 5 TFs each, were used to perturb each Fibroblast cell, by upregulating\n",
    "the select TFs.\n",
    "4. Each source cell, each target cell, and each combination of source cell with every recipie were embedded with \n",
    "Geneformer. Geneformer is a single cell foundation model that converts scRNAseq data into 512 dimensional embeddings\n",
    "5. For the source cells, the target cells, and each recipie, the embeddings were clustered using K-Means, and the centroids\n",
    "of each cluster were saved in a file.\n",
    "\n",
    "The output of the data are in a file called: /nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\n",
    "This file contains a single anndata object with the following attributes:\n",
    "1. adata.X is 6375 by 512, corresponding to the 6375 centroids obtained during K-means clustering and 512 dimension\n",
    "vector representing each centroid\n",
    "2. adata.obs contains the columns: 'recipe', 'REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2',\n",
    "'IKZF1', 'GFI1B'. The recipe column says 'source', 'target' or a recipie of 5 TFs. The values in\n",
    "this column are strings. The remaining columns are names of TFs that may or may not be in the recipie, and the values\n",
    "of these columns are True or False.\n",
    "\n",
    "Our goal is to identify the best recipie to reprogram source Fibroblasts to HSCs. To help solve this, I would like\n",
    "to understand this dataset, and I want you to produce instructions for each chatbot to help explore these data.\n",
    "\n",
    "Some different tasks that would be helpful include:\n",
    "1. visualize the data in our file\n",
    "2. computing the distance between each observation in the ann data that represents a recipie to each observation\n",
    "that represents a source or targe cell\n",
    "3. building a model that predicts the contribution to reprogramming obtains by the inclusion or exclusion of\n",
    "individual TFs\n",
    "4. analyzing the variance in the data for different recipies and for individual TFs.\n",
    "\n",
    "You can split up the directives to each chat bot separately. For each bot, you should provide full and clear\n",
    "instructions. In particular, if you want a BOT to use the file, you must include their full file path given above,\n",
    "if you want a bot to investigate specific columns, you must give their exact names.\n",
    "\n",
    "Please provide instructions in the format of:\n",
    "BOT1: <instructions here>\n",
    "BOT2: <instructions here>\n",
    "...\n",
    "\"\"\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "fc658e02-e1ec-404e-b953-51d6fe462377",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter your Open AI API key:  Â·Â·Â·Â·Â·Â·Â·Â·\n"
     ]
    }
   ],
   "source": [
    "llm = llms.load_openai(model_name='gpt-4o')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "ba49a03f-f057-48c7-ba50-04f9b211ea6b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32m2024-08-02 18:18:31 INFO semantic_router.utils.logger local\u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome to RAG! The chat log from this conversation will be saved to /home/jpic/BRAD/2024-08-02_18-18-27/log.json. How can I help?\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-02 18:18:36,956 - INFO - RAG\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "BRAD >> 0: \n",
      "\n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation: \n",
      "\n",
      "\n",
      "New Input: \n",
      "Context: You are BRAD (Bioinformatic Retrieval Augmented Data), a chatbot specializing in biology,\n",
      "bioinformatics, genetics, and data science. You can be connected to a text database to augment your answers\n",
      "based on the literature with Retrieval Augmented Generation, or you can use several additional modules including\n",
      "searching the web for new articles, searching Gene Ontology or Enrichr bioinformatics databases, running snakemake\n",
      "and matlab pipelines, or analyzing your own codes. Please answer the following questions to the best of your\n",
      "ability.\n",
      "\n",
      "Prompt: You are BRAD, a bioinformatics digital assistant, and you are responsible for\n",
      "organizing a team of digital assistants to analyze a cellular reprogramming experiment. Your job is to generate\n",
      "instruction sets that can be sent to 5 separate bots called BOT1, BOT2, BOT3, BOT4, and BOT5. These BOTs have the\n",
      "ability to:\n",
      "\n",
      "1. write and run python codes\n",
      "2. search information on arxiv, bioRxiv, PubMed, Enricher, or Gene Ontology databases\n",
      "3. search a literature database\n",
      "\n",
      "You must provide instructions to the BOTs to investigate the following experiment:\n",
      "\n",
      "Experiment Protocol:\n",
      "1. 10 Transcription Factors (TFs) were identified from Data Guided Control to reprogram Fibroblasts to Hematopoietic Stem\n",
      "Cells (HSCs).\n",
      "2. single cell data was obtained for Fibroblast and HSC data\n",
      "3. 252 perturbation recipies, consisting of 5 TFs each, were used to perturb each Fibroblast cell, by upregulating\n",
      "the select TFs.\n",
      "4. Each source cell, each target cell, and each combination of source cell with every recipie were embedded with \n",
      "Geneformer. Geneformer is a single cell foundation model that converts scRNAseq data into 512 dimensional embeddings\n",
      "5. For the source cells, the target cells, and each recipie, the embeddings were clustered using K-Means, and the centroids\n",
      "of each cluster were saved in a file.\n",
      "\n",
      "The output of the data are in a file called: /nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\n",
      "This file contains a single anndata object with the following attributes:\n",
      "1. adata.X is 6375 by 512, corresponding to the 6375 centroids obtained during K-means clustering and 512 dimension\n",
      "vector representing each centroid\n",
      "2. adata.obs contains the columns: 'recipe', 'REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2',\n",
      "'IKZF1', 'GFI1B'. The recipe column says 'source', 'target' or a recipie of 5 TFs. The values in\n",
      "this column are strings. The remaining columns are names of TFs that may or may not be in the recipie, and the values\n",
      "of these columns are True or False.\n",
      "\n",
      "Our goal is to identify the best recipie to reprogram source Fibroblasts to HSCs. To help solve this, I would like\n",
      "to understand this dataset, and I want you to produce instructions for each chatbot to help explore these data.\n",
      "\n",
      "Some different tasks that would be helpful include:\n",
      "1. visualize the data in our file\n",
      "2. computing the distance between each observation in the ann data that represents a recipie to each observation\n",
      "that represents a source or targe cell\n",
      "3. building a model that predicts the contribution to reprogramming obtains by the inclusion or exclusion of\n",
      "individual TFs\n",
      "4. analyzing the variance in the data for different recipies and for individual TFs.\n",
      "\n",
      "You can split up the directives to each chat bot separately. For each bot, you should provide full and clear\n",
      "instructions. In particular, if you want a BOT to use the file, you must include their full file path given above,\n",
      "if you want a bot to investigate specific columns, you must give their exact names.\n",
      "\n",
      "Please provide instructions in the format of:\n",
      "BOT1: <instructions here>\n",
      "BOT2: <instructions here>\n",
      "...\n",
      "\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-02 18:18:49,131 - INFO - HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2024-08-02 18:18:49,133 - INFO - \n",
      "\n",
      "\n",
      "route\n",
      "\n",
      "\n",
      "\n",
      "2024-08-02 18:18:49,133 - INFO - GGET\n",
      "2024-08-02 18:18:49,134 - INFO - \n",
      "\n",
      "PLANNER:\n",
      "\n",
      "{'module': 'RAG', 'steps': [{'llm': \"client=<openai.resources.chat.completions.Completions object at 0x14def8df6ed0> async_client=<openai.resources.chat.completions.AsyncCompletions object at 0x14def8e0b0d0> model_name='gpt-4o' temperature=0.0 openai_api_key=SecretStr('**********') openai_proxy=''\", 'memory': 'None', 'prompt': \"input_variables=['history', 'input'] template='Current conversation: {history}\\\\n\\\\n\\\\nNew Input: \\\\n{input}'\", 'input': \"Context: You are BRAD (Bioinformatic Retrieval Augmented Data), a chatbot specializing in biology,\\nbioinformatics, genetics, and data science. You can be connected to a text database to augment your answers\\nbased on the literature with Retrieval Augmented Generation, or you can use several additional modules including\\nsearching the web for new articles, searching Gene Ontology or Enrichr bioinformatics databases, running snakemake\\nand matlab pipelines, or analyzing your own codes. Please answer the following questions to the best of your\\nability.\\n\\nPrompt: You are BRAD, a bioinformatics digital assistant, and you are responsible for\\norganizing a team of digital assistants to analyze a cellular reprogramming experiment. Your job is to generate\\ninstruction sets that can be sent to 5 separate bots called BOT1, BOT2, BOT3, BOT4, and BOT5. These BOTs have the\\nability to:\\n\\n1. write and run python codes\\n2. search information on arxiv, bioRxiv, PubMed, Enricher, or Gene Ontology databases\\n3. search a literature database\\n\\nYou must provide instructions to the BOTs to investigate the following experiment:\\n\\nExperiment Protocol:\\n1. 10 Transcription Factors (TFs) were identified from Data Guided Control to reprogram Fibroblasts to Hematopoietic Stem\\nCells (HSCs).\\n2. single cell data was obtained for Fibroblast and HSC data\\n3. 252 perturbation recipies, consisting of 5 TFs each, were used to perturb each Fibroblast cell, by upregulating\\nthe select TFs.\\n4. Each source cell, each target cell, and each combination of source cell with every recipie were embedded with \\nGeneformer. Geneformer is a single cell foundation model that converts scRNAseq data into 512 dimensional embeddings\\n5. For the source cells, the target cells, and each recipie, the embeddings were clustered using K-Means, and the centroids\\nof each cluster were saved in a file.\\n\\nThe output of the data are in a file called: /nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\\nThis file contains a single anndata object with the following attributes:\\n1. adata.X is 6375 by 512, corresponding to the 6375 centroids obtained during K-means clustering and 512 dimension\\nvector representing each centroid\\n2. adata.obs contains the columns: 'recipe', 'REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2',\\n'IKZF1', 'GFI1B'. The recipe column says 'source', 'target' or a recipie of 5 TFs. The values in\\nthis column are strings. The remaining columns are names of TFs that may or may not be in the recipie, and the values\\nof these columns are True or False.\\n\\nOur goal is to identify the best recipie to reprogram source Fibroblasts to HSCs. To help solve this, I would like\\nto understand this dataset, and I want you to produce instructions for each chatbot to help explore these data.\\n\\nSome different tasks that would be helpful include:\\n1. visualize the data in our file\\n2. computing the distance between each observation in the ann data that represents a recipie to each observation\\nthat represents a source or targe cell\\n3. building a model that predicts the contribution to reprogramming obtains by the inclusion or exclusion of\\nindividual TFs\\n4. analyzing the variance in the data for different recipies and for individual TFs.\\n\\nYou can split up the directives to each chat bot separately. For each bot, you should provide full and clear\\ninstructions. In particular, if you want a BOT to use the file, you must include their full file path given above,\\nif you want a bot to investigate specific columns, you must give their exact names.\\n\\nPlease provide instructions in the format of:\\nBOT1: <instructions here>\\nBOT2: <instructions here>\\n...\\n\", 'output': \"Sure, here are the detailed instructions for each BOT to help analyze the cellular reprogramming experiment dataset:\\n\\n### BOT1: Data Visualization\\n**Instructions:**\\n1. Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`.\\n2. Visualize the data using UMAP or t-SNE to reduce the 512-dimensional embeddings to 2D or 3D space.\\n3. Color the visualizations based on the 'recipe' column to distinguish between 'source', 'target', and different TF recipes.\\n4. Generate separate visualizations highlighting the presence of each TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) by coloring the points based on the True/False values in the respective columns.\\n5. Save the visualizations as high-resolution images and provide a summary of the key observations.\\n\\n### BOT2: Distance Computation\\n**Instructions:**\\n1. Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`.\\n2. Compute the Euclidean distance between each observation representing a recipe and each observation representing a source cell.\\n3. Compute the Euclidean distance between each observation representing a recipe and each observation representing a target cell.\\n4. Store the computed distances in a new dataframe with columns: 'recipe', 'source_distance', 'target_distance'.\\n5. Save the dataframe to a CSV file for further analysis.\\n\\n### BOT3: Predictive Modeling\\n**Instructions:**\\n1. Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`.\\n2. Prepare the data for modeling by encoding the 'recipe' column and the TF columns (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) as features.\\n3. Use the distances computed by BOT2 as the target variable for the model.\\n4. Build a machine learning model (e.g., Random Forest, Gradient Boosting) to predict the contribution to reprogramming based on the inclusion or exclusion of individual TFs.\\n5. Evaluate the model using cross-validation and report the performance metrics (e.g., R^2, RMSE).\\n6. Identify the most important TFs contributing to successful reprogramming based on feature importance scores.\\n7. Save the model and the feature importance scores for further analysis.\\n\\n### BOT4: Variance Analysis\\n**Instructions:**\\n1. Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`.\\n2. Analyze the variance in the 512-dimensional embeddings for different recipes.\\n3. Compute the variance for each individual TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) across all recipes.\\n4. Generate box plots or violin plots to visualize the variance distribution for each TF.\\n5. Identify which TFs contribute to higher variance and which contribute to lower variance.\\n6. Save the plots and provide a summary of the key findings.\\n\\n### BOT5: Literature Search\\n**Instructions:**\\n1. Search for relevant literature on arxiv, bioRxiv, PubMed, Enricher, or Gene Ontology databases related to the reprogramming of fibroblasts to hematopoietic stem cells using transcription factors.\\n2. Focus on identifying studies that have used similar TFs (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) for cellular reprogramming.\\n3. Summarize the key findings from the literature, including the methodologies used, the success rates of reprogramming, and any insights into the role of individual TFs.\\n4. Provide a list of the most relevant articles with their abstracts and links for further reading.\\n\\nBy following these instructions, each BOT will contribute to a comprehensive analysis of the cellular reprogramming experiment dataset.\", 'parsedOutput': \"Sure, here are the detailed instructions for each BOT to help analyze the cellular reprogramming experiment dataset:\\n\\n### BOT1: Data Visualization\\n**Instructions:**\\n1. Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`.\\n2. Visualize the data using UMAP or t-SNE to reduce the 512-dimensional embeddings to 2D or 3D space.\\n3. Color the visualizations based on the 'recipe' column to distinguish between 'source', 'target', and different TF recipes.\\n4. Generate separate visualizations highlighting the presence of each TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) by coloring the points based on the True/False values in the respective columns.\\n5. Save the visualizations as high-resolution images and provide a summary of the key observations.\\n\\n### BOT2: Distance Computation\\n**Instructions:**\\n1. Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`.\\n2. Compute the Euclidean distance between each observation representing a recipe and each observation representing a source cell.\\n3. Compute the Euclidean distance between each observation representing a recipe and each observation representing a target cell.\\n4. Store the computed distances in a new dataframe with columns: 'recipe', 'source_distance', 'target_distance'.\\n5. Save the dataframe to a CSV file for further analysis.\\n\\n### BOT3: Predictive Modeling\\n**Instructions:**\\n1. Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`.\\n2. Prepare the data for modeling by encoding the 'recipe' column and the TF columns (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) as features.\\n3. Use the distances computed by BOT2 as the target variable for the model.\\n4. Build a machine learning model (e.g., Random Forest, Gradient Boosting) to predict the contribution to reprogramming based on the inclusion or exclusion of individual TFs.\\n5. Evaluate the model using cross-validation and report the performance metrics (e.g., R^2, RMSE).\\n6. Identify the most important TFs contributing to successful reprogramming based on feature importance scores.\\n7. Save the model and the feature importance scores for further analysis.\\n\\n### BOT4: Variance Analysis\\n**Instructions:**\\n1. Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`.\\n2. Analyze the variance in the 512-dimensional embeddings for different recipes.\\n3. Compute the variance for each individual TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) across all recipes.\\n4. Generate box plots or violin plots to visualize the variance distribution for each TF.\\n5. Identify which TFs contribute to higher variance and which contribute to lower variance.\\n6. Save the plots and provide a summary of the key findings.\\n\\n### BOT5: Literature Search\\n**Instructions:**\\n1. Search for relevant literature on arxiv, bioRxiv, PubMed, Enricher, or Gene Ontology databases related to the reprogramming of fibroblasts to hematopoietic stem cells using transcription factors.\\n2. Focus on identifying studies that have used similar TFs (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) for cellular reprogramming.\\n3. Summarize the key findings from the literature, including the methodologies used, the success rates of reprogramming, and any insights into the role of individual TFs.\\n4. Provide a list of the most relevant articles with their abstracts and links for further reading.\\n\\nBy following these instructions, each BOT will contribute to a comprehensive analysis of the cellular reprogramming experiment dataset.\", 'purpose': 'chat without RAG'}]}\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "Sure, here are the detailed instructions for each BOT to help analyze the cellular reprogramming experiment dataset:\n",
      "\n",
      "### BOT1: Data Visualization\n",
      "**Instructions:**\n",
      "1. Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`.\n",
      "2. Visualize the data using UMAP or t-SNE to reduce the 512-dimensional embeddings to 2D or 3D space.\n",
      "3. Color the visualizations based on the 'recipe' column to distinguish between 'source', 'target', and different TF recipes.\n",
      "4. Generate separate visualizations highlighting the presence of each TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) by coloring the points based on the True/False values in the respective columns.\n",
      "5. Save the visualizations as high-resolution images and provide a summary of the key observations.\n",
      "\n",
      "### BOT2: Distance Computation\n",
      "**Instructions:**\n",
      "1. Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`.\n",
      "2. Compute the Euclidean distance between each observation representing a recipe and each observation representing a source cell.\n",
      "3. Compute the Euclidean distance between each observation representing a recipe and each observation representing a target cell.\n",
      "4. Store the computed distances in a new dataframe with columns: 'recipe', 'source_distance', 'target_distance'.\n",
      "5. Save the dataframe to a CSV file for further analysis.\n",
      "\n",
      "### BOT3: Predictive Modeling\n",
      "**Instructions:**\n",
      "1. Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`.\n",
      "2. Prepare the data for modeling by encoding the 'recipe' column and the TF columns (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) as features.\n",
      "3. Use the distances computed by BOT2 as the target variable for the model.\n",
      "4. Build a machine learning model (e.g., Random Forest, Gradient Boosting) to predict the contribution to reprogramming based on the inclusion or exclusion of individual TFs.\n",
      "5. Evaluate the model using cross-validation and report the performance metrics (e.g., R^2, RMSE).\n",
      "6. Identify the most important TFs contributing to successful reprogramming based on feature importance scores.\n",
      "7. Save the model and the feature importance scores for further analysis.\n",
      "\n",
      "### BOT4: Variance Analysis\n",
      "**Instructions:**\n",
      "1. Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`.\n",
      "2. Analyze the variance in the 512-dimensional embeddings for different recipes.\n",
      "3. Compute the variance for each individual TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) across all recipes.\n",
      "4. Generate box plots or violin plots to visualize the variance distribution for each TF.\n",
      "5. Identify which TFs contribute to higher variance and which contribute to lower variance.\n",
      "6. Save the plots and provide a summary of the key findings.\n",
      "\n",
      "### BOT5: Literature Search\n",
      "**Instructions:**\n",
      "1. Search for relevant literature on arxiv, bioRxiv, PubMed, Enricher, or Gene Ontology databases related to the reprogramming of fibroblasts to hematopoietic stem cells using transcription factors.\n",
      "2. Focus on identifying studies that have used similar TFs (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) for cellular reprogramming.\n",
      "3. Summarize the key findings from the literature, including the methodologies used, the success rates of reprogramming, and any insights into the role of individual TFs.\n",
      "4. Provide a list of the most relevant articles with their abstracts and links for further reading.\n",
      "\n",
      "By following these instructions, each BOT will contribute to a comprehensive analysis of the cellular reprogramming experiment dataset.\n"
     ]
    }
   ],
   "source": [
    "llm = llms.load_openai(model_name='gpt-4o')\n",
    "master = brad.chatbot(llm=llm, interactive=False)\n",
    "response = master.invoke(initial_master_prompt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "3b7aefdb-2749-4047-b3f3-fc90ea86e6d9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##########################\n",
      "Sure, here are the detailed instructions for each BOT to help analyze the cellular reprogramming experiment dataset:\n",
      "\n",
      "\n",
      "##########################\n",
      " BOT1: Data Visualization\n",
      "**Instructions:**\n",
      "1. Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`.\n",
      "2. Visualize the data using UMAP or t-SNE to reduce the 512-dimensional embeddings to 2D or 3D space.\n",
      "3. Color the visualizations based on the 'recipe' column to distinguish between 'source', 'target', and different TF recipes.\n",
      "4. Generate separate visualizations highlighting the presence of each TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) by coloring the points based on the True/False values in the respective columns.\n",
      "5. Save the visualizations as high-resolution images and provide a summary of the key observations.\n",
      "\n",
      "\n",
      "##########################\n",
      " BOT2: Distance Computation\n",
      "**Instructions:**\n",
      "1. Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`.\n",
      "2. Compute the Euclidean distance between each observation representing a recipe and each observation representing a source cell.\n",
      "3. Compute the Euclidean distance between each observation representing a recipe and each observation representing a target cell.\n",
      "4. Store the computed distances in a new dataframe with columns: 'recipe', 'source_distance', 'target_distance'.\n",
      "5. Save the dataframe to a CSV file for further analysis.\n",
      "\n",
      "\n",
      "##########################\n",
      " BOT3: Predictive Modeling\n",
      "**Instructions:**\n",
      "1. Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`.\n",
      "2. Prepare the data for modeling by encoding the 'recipe' column and the TF columns (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) as features.\n",
      "3. Use the distances computed by BOT2 as the target variable for the model.\n",
      "4. Build a machine learning model (e.g., Random Forest, Gradient Boosting) to predict the contribution to reprogramming based on the inclusion or exclusion of individual TFs.\n",
      "5. Evaluate the model using cross-validation and report the performance metrics (e.g., R^2, RMSE).\n",
      "6. Identify the most important TFs contributing to successful reprogramming based on feature importance scores.\n",
      "7. Save the model and the feature importance scores for further analysis.\n",
      "\n",
      "\n",
      "##########################\n",
      " BOT4: Variance Analysis\n",
      "**Instructions:**\n",
      "1. Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`.\n",
      "2. Analyze the variance in the 512-dimensional embeddings for different recipes.\n",
      "3. Compute the variance for each individual TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) across all recipes.\n",
      "4. Generate box plots or violin plots to visualize the variance distribution for each TF.\n",
      "5. Identify which TFs contribute to higher variance and which contribute to lower variance.\n",
      "6. Save the plots and provide a summary of the key findings.\n",
      "\n",
      "\n",
      "##########################\n",
      " BOT5: Literature Search\n",
      "**Instructions:**\n",
      "1. Search for relevant literature on arxiv, bioRxiv, PubMed, Enricher, or Gene Ontology databases related to the reprogramming of fibroblasts to hematopoietic stem cells using transcription factors.\n",
      "2. Focus on identifying studies that have used similar TFs (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) for cellular reprogramming.\n",
      "3. Summarize the key findings from the literature, including the methodologies used, the success rates of reprogramming, and any insights into the role of individual TFs.\n",
      "4. Provide a list of the most relevant articles with their abstracts and links for further reading.\n",
      "\n",
      "By following these instructions, each BOT will contribute to a comprehensive analysis of the cellular reprogramming experiment dataset.\n"
     ]
    }
   ],
   "source": [
    "instructionSets = response.split('###') # .pop(0)\n",
    "for inst in instructionSets:\n",
    "    print(\"##########################\")\n",
    "    print(inst)\n",
    "          "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a50fea33-5ca1-4890-86e7-47392e1827b7",
   "metadata": {},
   "source": [
    "## BOT PREPROMPT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "151094fe-a585-414f-a171-d0dab7b5b74c",
   "metadata": {},
   "outputs": [],
   "source": [
    "worker_system_prompt = \"\"\"/force PLANNER You are a chatbot assisting to help analyze a single cell dataset.\n",
    "You have  instructions below, and you must use your tools to answer the questions. You have one tool that allows\n",
    "you to write CODE. Generate a plan based on the instructions to use the CODE tool several times in order to satisfy\n",
    "your instructions. Each call to the CODE tool should have a single goal, such as making one type of plot, builting\n",
    "one type of model, and so forth.\n",
    "\n",
    "If your job is to make plots, then you can make a plan to make multiple kinds of plots with each different call\n",
    "to CODE in your plan. For instance. your plan could be\n",
    "\n",
    "**Step 1, CODE** load <path to data>.h5ad and make umap plots of ...\n",
    "**Step 2, CODE** load <path to data>.h5ad and make tsne plots of ...\n",
    "\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc36b7fc-0152-4230-a1fb-5d00fa862052",
   "metadata": {},
   "source": [
    "## BOT 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5df03a96-4073-4960-b1b0-272133d99631",
   "metadata": {},
   "outputs": [],
   "source": [
    "instructionSets = \"\"\"BOT1: Data Visualization\n",
    "**Instructions:**\n",
    "1. Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`.\n",
    "2. Visualize the data using UMAP or t-SNE to reduce the 512-dimensional embeddings to 2D or 3D space.\n",
    "3. Color the visualizations based on the 'recipe' column to distinguish between 'source', 'target', and different TF recipes.\n",
    "4. Generate separate visualizations highlighting the presence of each TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) by coloring the points based on the True/False values in the respective columns.\n",
    "5. Save the visualizations as high-resolution images and provide a summary of the key observations.\n",
    "\n",
    "Only use the CODE module to address the above instructions.\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "51cab9a9-30fe-40c0-b125-8ef242b0b6a5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/force PLANNER You are a chatbot assisting to help analyze a single cell dataset.\n",
      "You have  instructions below, and you must use your tools to answer the questions. You have one tool that allows\n",
      "you to write CODE. Generate a plan based on the instructions to use the CODE tool several times in order to satisfy\n",
      "your instructions. Each call to the CODE tool should have a single goal, such as making one type of plot, builting\n",
      "one type of model, and so forth.\n",
      "\n",
      "If your job is to make plots, then you can make a plan to make multiple kinds of plots with each different call\n",
      "to CODE in your plan. For instance. your plan could be\n",
      "\n",
      "**Step 1, CODE** load <path to data>.h5ad and make umap plots of ...\n",
      "**Step 2, CODE** load <path to data>.h5ad and make tsne plots of ...\n",
      "\n",
      "BOT1: Data Visualization\n",
      "**Instructions:**\n",
      "1. Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`.\n",
      "2. Visualize the data using UMAP or t-SNE to reduce the 512-dimensional embeddings to 2D or 3D space.\n",
      "3. Color the visualizations based on the 'recipe' column to distinguish between 'source', 'target', and different TF recipes.\n",
      "4. Generate separate visualizations highlighting the presence of each TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) by coloring the points based on the True/False values in the respective columns.\n",
      "5. Save the visualizations as high-resolution images and provide a summary of the key observations.\n",
      "\n",
      "Only use the CODE module to address the above instructions.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(worker_system_prompt + instructionSets)"
   ]
  },
  {
   "cell_type": "raw",
   "id": "c239af2c-b547-4b44-9d82-8b28d39040c1",
   "metadata": {},
   "source": [
    "/force PLANNER You are a chatbot assisting to help analyze a single cell dataset.\n",
    "You have  instructions below, and you must use your tools to answer the questions. You have one tool that allows\n",
    "you to write CODE. Generate a plan based on the instructions to use the CODE tool several times in order to satisfy\n",
    "your instructions. Each call to the CODE tool should have a single goal, such as making one type of plot, builting\n",
    "one type of model, and so forth.\n",
    "\n",
    "If your job is to make plots, then you can make a plan to make multiple kinds of plots with each different call\n",
    "to CODE in your plan. For instance. your plan could be\n",
    "\n",
    "**Step 1, CODE** load <path to data>.h5ad and make umap plots of ...\n",
    "**Step 2, CODE** load <path to data>.h5ad and make tsne plots of ...\n",
    "\n",
    "BOT1: Data Visualization\n",
    "**Instructions:**\n",
    "1. Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`.\n",
    "2. Visualize the data using UMAP or t-SNE to reduce the 512-dimensional embeddings to 2D or 3D space.\n",
    "3. Color the visualizations based on the 'recipe' column to distinguish between 'source', 'target', and different TF recipes.\n",
    "4. Generate separate visualizations highlighting the presence of each TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) by coloring the points based on the True/False values in the respective columns.\n",
    "5. Save the visualizations as high-resolution images and provide a summary of the key observations.\n",
    "\n",
    "Only use the CODE module to address the above instructions.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c9b53559-c761-4067-a1d8-e942c38b0a18",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Would you like to use a database with BRAD [Y/N]?\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      " N\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32m2024-08-03 00:40:15 INFO semantic_router.utils.logger local\u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome to RAG! The chat log from this conversation will be saved to /home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_00-39-53/log.json. How can I help?\n"
     ]
    }
   ],
   "source": [
    "# llm = llms.load_openai(model_name='gpt-4o')\n",
    "BOT1 = brad.chatbot(llm=llm, config='/home/jpic/RAG-DEV/tutorials/geneformer/config-sc.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "468c3d2f-dcc0-4a47-9bf2-ca17b8bab086",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  /force PLANNER please load and run the BOT1 pipeline\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 00:41:17,333 - INFO - PLANNER\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "BRAD >> 0: \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/langchain_core/_api/deprecation.py:139: LangChainDeprecationWarning: The class `LLMChain` was deprecated in LangChain 0.1.17 and will be removed in 1.0. Use RunnableSequence, e.g., `prompt | llm` instead.\n",
      "  warn_deprecated(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3m**INSTRUCTIONS:**\n",
      "You are planning a bioinformatics analysis pipeline to address a user's query. You must determine if there already exist a predesigned pipeline that will respond to a user's query, or if a new, custom process must be designed for this particular query. The multi-step workflow can use the available methods listed below.\n",
      "\n",
      "**Available Methods within a Pipeline:**\n",
      "1. **RAG**: Look up literature and documents from a text database.\n",
      "2. **SCRAPE**: Search platforms like arXiv, bioRxiv, and PubMed for the latest research.\n",
      "3. **DATABASE**: Utilize bioinformatics databases such as Gene Ontology and Enrichr to perform gene set enrichment analyses.\n",
      "4. **CODE**: Execute bioinformatics pipelines. Utilize prebuilt pipelines or develop new ones as needed.\n",
      "5. **WRITE**: Synthesize and summarize information. This can include summarizing database searches, code pipeline results, or creating a final report to encapsulate the entire analysis.\n",
      "6. **ROUTER**: Determine which step we should proceed to next.\n",
      "\n",
      "User Query: please load and run the BOT1 pipeline\n",
      "\n",
      "**Available Pipelines**\n",
      "Name: BOT1\tDescription: this plan is for basic visualization and summary of single cell data from the Geneformer reprogramming experiment. The data can be loaded in from ann data objects\n",
      "\n",
      "\n",
      "CUSTOM: The option to build a new pipeline\n",
      "\n",
      "The custom pipeline should be selected if no current pipeline addreses the user's query. If the custom pipeline is selected, please do not attempt to design the pipeline now. We will work to design it at a later point.\n",
      "\n",
      "**Output**:\n",
      "Please formate your output as follows:\n",
      "Pipeline Name: <available pipeline name above or \"CUSTOM\">\n",
      "Explination: <reason for selecting the above pipeline>\n",
      "\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 00:41:19,390 - INFO - HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2024-08-03 00:41:19,404 - INFO - Pipeline Name: BOT1\n",
      "\n",
      "Explanation: The user's query specifically requests to load and run the BOT1 pipeline. The BOT1 pipeline is designed for the basic visualization and summary of single cell data from the Geneformer reprogramming experiment, which aligns with the user's request. Therefore, there is no need to design a new custom pipeline.\n",
      "2024-08-03 00:41:19,405 - INFO - selected_pipeline=BOT1\n",
      "2024-08-03 00:41:19,405 - INFO - \n",
      "\n",
      "\n",
      "route\n",
      "\n",
      "\n",
      "\n",
      "2024-08-03 00:41:19,406 - INFO - PLANNER\n",
      "2024-08-03 00:41:19,406 - INFO - \n",
      "\n",
      "PLANNER:\n",
      "\n",
      "{'module': 'PLANNER', 'steps': [{'llm': \"client=<openai.resources.chat.completions.Completions object at 0x1480cc33e8d0> async_client=<openai.resources.chat.completions.AsyncCompletions object at 0x1480cc324e50> model_name='gpt-4o' temperature=0.0 openai_api_key=SecretStr('**********') openai_proxy=''\", 'memory': 'None', 'prompt': 'input_variables=[\\'input\\'] template=\\'**INSTRUCTIONS:**\\\\nYou are planning a bioinformatics analysis pipeline to address a user\\\\\\'s query. You must determine if there already exist a predesigned pipeline that will respond to a user\\\\\\'s query, or if a new, custom process must be designed for this particular query. The multi-step workflow can use the available methods listed below.\\\\n\\\\n**Available Methods within a Pipeline:**\\\\n1. **RAG**: Look up literature and documents from a text database.\\\\n2. **SCRAPE**: Search platforms like arXiv, bioRxiv, and PubMed for the latest research.\\\\n3. **DATABASE**: Utilize bioinformatics databases such as Gene Ontology and Enrichr to perform gene set enrichment analyses.\\\\n4. **CODE**: Execute bioinformatics pipelines. Utilize prebuilt pipelines or develop new ones as needed.\\\\n5. **WRITE**: Synthesize and summarize information. This can include summarizing database searches, code pipeline results, or creating a final report to encapsulate the entire analysis.\\\\n6. **ROUTER**: Determine which step we should proceed to next.\\\\n\\\\nUser Query: {input}\\\\n\\\\n**Available Pipelines**\\\\nName: BOT1\\\\tDescription: this plan is for basic visualization and summary of single cell data from the Geneformer reprogramming experiment. The data can be loaded in from ann data objects\\\\n\\\\n\\\\nCUSTOM: The option to build a new pipeline\\\\n\\\\nThe custom pipeline should be selected if no current pipeline addreses the user\\\\\\'s query. If the custom pipeline is selected, please do not attempt to design the pipeline now. We will work to design it at a later point.\\\\n\\\\n**Output**:\\\\nPlease formate your output as follows:\\\\nPipeline Name: <available pipeline name above or \"CUSTOM\">\\\\nExplination: <reason for selecting the above pipeline>\\\\n\\'', 'input': 'please load and run the BOT1 pipeline', 'output': \"Pipeline Name: BOT1\\n\\nExplanation: The user's query specifically requests to load and run the BOT1 pipeline. The BOT1 pipeline is designed for the basic visualization and summary of single cell data from the Geneformer reprogramming experiment, which aligns with the user's request. Therefore, there is no need to design a new custom pipeline.\", 'parsedOutput': {'selected pipeline': 'BOT1'}, 'purpose': 'determine if a known pipeline can be used or a new one si required'}, {'func': 'planner.planner', 'what': 'loaded an older pipeline'}, {'func': 'planner.planner', 'what': 'set the queue and set the queue pointer to 1'}]}\n",
      "2024-08-03 00:41:19,409 - INFO - current llm calls=1\n",
      "2024-08-03 00:41:19,413 - INFO - CODE\n",
      "2024-08-03 00:41:19,414 - INFO - CODER\n",
      "2024-08-03 00:41:19,420 - INFO - input_variables=['user_query'] template=\"You must select which code to run to help a user.\\n\\n**Available Scripts**\\nScript Name: scanpy_brad, \\t Description: This script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\\nScript Name: models_brad, \\t Description: This script executes a series of custom commands to build a model for predicting the contribution to reprogramming obtained by the inclusion or exclusion of individual TFs using an AnnData object or single cell datasets loaded from a .h5ad file.\\n\\n\\n**User Query**\\n{user_query}\\n\\n**Task**\\nBased on the user's query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\\n\\n**Response Template**\\nSCRIPT: <selected script>\\nREASON: <reasoning why the script fits the user prompt>\\n\"\n",
      "2024-08-03 00:41:19,420 - INFO - FIRST LLM CALL\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "** Step 0**\n",
      "{'order': 0, 'module': 'PLANNER', 'prompt': None, 'description': 'This step designed the plan. It is placed in the queue because we needed a place holder for 0 indexed lists.', 'output': []}\n",
      "\n",
      "\n",
      "** Step 1**\n",
      "{'order': 1, 'module': 'CODE', 'prompt': '/force CODE  1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and create a basic summary of the data.\\n\\n**', 'description': ' 1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and create a basic summary of the data.\\n\\n**', 'output': []}\n",
      "\n",
      "\n",
      "** Step 2**\n",
      "{'order': 2, 'module': 'CODE', 'prompt': \"/force CODE  2 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, create UMAP plots of the data in 2D and 3D space and color the visualizations based on the 'recipe' column to distinguish between 'source', 'target', and different TF recipes.\\n\\n**\", 'description': \" 2 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, create UMAP plots of the data in 2D and 3D space and color the visualizations based on the 'recipe' column to distinguish between 'source', 'target', and different TF recipes.\\n\\n**\", 'output': []}\n",
      "\n",
      "\n",
      "** Step 3**\n",
      "{'order': 3, 'module': 'CODE', 'prompt': \"/force CODE  3 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, create t-SNE plots of the data in 2D and 3D space and color the visualizations based on the 'recipe' column to distinguish between 'source', 'target', and different TF recipes.\\n\\n**\", 'description': \" 3 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, create t-SNE plots of the data in 2D and 3D space and color the visualizations based on the 'recipe' column to distinguish between 'source', 'target', and different TF recipes.\\n\\n**\", 'output': []}\n",
      "\n",
      "\n",
      "** Step 4**\n",
      "{'order': 4, 'module': 'CODE', 'prompt': '/force CODE  4 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, generate separate UMAP visualizations highlighting the presence of each TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) by coloring the points based on the True/False values in the respective columns.\\n\\n**', 'description': ' 4 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, generate separate UMAP visualizations highlighting the presence of each TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) by coloring the points based on the True/False values in the respective columns.\\n\\n**', 'output': []}\n",
      "\n",
      "\n",
      "** Step 5**\n",
      "{'order': 5, 'module': 'CODE', 'prompt': '/force CODE  5 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, generate separate t-SNE visualizations highlighting the presence of each TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) by coloring the points based on the True/False values in the respective columns.\\n\\nThe removed ', 'description': ' 5 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, generate separate t-SNE visualizations highlighting the presence of each TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) by coloring the points based on the True/False values in the respective columns.\\n\\nThe removed ', 'output': []}\n",
      "\n",
      "\n",
      "==================================================\n",
      "==================================================\n",
      "BRAD >> 1: \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 00:41:20,889 - INFO - HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2024-08-03 00:41:20,890 - INFO - SCRIPT: scanpy_brad\n",
      "REASON: The user's query involves loading an AnnData object from a .h5ad file and creating a basic summary of the data. The `scanpy_brad` script is designed to handle tasks related to manipulating, visualizing, and exploring single-cell datasets loaded from .h5ad files. It allows the use of libraries like scanpy, seaborn, and matplotlib.pyplot, which are well-suited for summarizing and visualizing single-cell data. Therefore, `scanpy_brad` is the best match for the user's request.\n",
      "2024-08-03 00:41:20,891 - INFO - scriptPath=/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts\n",
      "2024-08-03 00:41:20,891 - INFO - ALL SCRIPTS FOUND. BUILDING TEMPLATE\n",
      "2024-08-03 00:41:20,892 - INFO - input_variables=['history', 'input'] template='Current conversation:\\n{history}\\n\\n**PYTHON SCRIPT**\\nYou must run this python script:\\n/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\n\\n**PYTHON SCRIPT DOCUMENTATION**:\\nThis is the doc string of this python script:\\n\"\"\"\\nThis script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\\n\\nArguments (four arguments):\\n    1. output directory: chatstatus[\\'output-directory\\']\\n    2. output file: <name of output file>\\n    3. input file: <file created in previous step>\\n    4. scanpy commands: a list of scanpy commands to be executed on the AnnData object (provided as a single string with commands separated by a delimiter, e.g., \\';\\')\\n\\nBased on the arguments, the input file will be loaded, then your commands will be executed, and finally, the output or resulting ann data object will be saved to the corrrectou output file and directory. Your code is not responsible for loading the .h5ad object, that will happen automatically, and when loaded, the object will be called adata. Your scanpy commands can operate directly on the adata object that will be loaded for you.\\n\\nAdditionally, the following imports are already provided for you and can be used in your code:\\n```\\nimport scanpy as sc\\nimport seaborn as sns\\nimport matplotlib.pyplot as plt\\nimport os\\nimport numpy as np\\nimport pandas as pd\\nfrom scipy.spatial.distance import cdist\\n```\\n\\n**Usage**\\nBRAD Line:\\n```\\nsubprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\'output-directory\\'], <output file>, <input file>, \"<scanpy commands>\"], capture_output=True, text=True)\\n```\\n\\n**Examples**\\nUse the below examples to help generate your code.\\n\\n*Example 1*\\nUser Prompt: Run scanpy preprocessing and UMAP visualization on XXX.h5ad and save the UMAP plot\\nResponse Code:\\n```\\nresponse = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"XXX-modified.h5ad\", \"<path/to/data>/XXX.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save=\\'umap.png\\')\"], capture_output=True, text=True)\\n```\\nExplination: the adata object will be loaded in memory already. The command \"sc.pp.neighbors(adata)\" will preprocess the data, then the command \"sc.tl.umap(adata)\" will perform UMAP, and finally the command \"sc.pl.umap(adata, \\'umap.png\\')\" will save the UMAP to a well named file.\\n\\n*Example 2*\\nUser Prompt: Perform PCA and clustering on the dataset YYY.h5ad and save the PCA plot\\nResponse Code:\\n```\\nresponse = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"YYY-modified.h5ad\", \"<path/to/data>/YYY.h5ad\", \"sc.pp.pca(adata); sc.tl.leiden(adata); sc.pl.pca(adata, save=\\'pca.png\\')\"], capture_output=True, text=True)\\n```\\nExplination: the adata object will be loaded in memory already. The command \"sc.pp.pca(adata)\" will preprocess the data, then the command \"sc.tl.leiden(adata)\" will perform the leiden algorithm, and finally the command \"sc.pl.pca(adata, save=\\'pca.png\\')\" will save the PCA to a well named file.\\n\\n*Example 3*\\nUser Prompt: compute the distance between all cells that are not source or target to all source and target cells.\\nReponse Code:\\n```\\nadata_X = adata.X\\nmask = ~adata.obs[\\'recipe\\'].isin([\\'source\\', \\'target\\'])\\nrecipe_embeddings = adata[mask]\\nsource_embeddings = adata[adata.obs[\\'recipe\\'] == \\'source\\']\\ntarget_embeddings = adata[adata.obs[\\'recipe\\'] == \\'target\\']\\nrecipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X)\\nrecipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X)\\nadata_dist = sc.AnnData(X=np.concatenate((recipe_distances_source, recipe_distances_target), axis=1), var=pd.concat((source_embeddings.obs, target_embeddings.obs)), obs=recipe_embeddings.obs)\\nadata_dist.write(<file name>.h5ad)\\n```\\n\\n**OUTPUT FILE NAME INSTRUCTIONS**\\n1. Output path should be chatstatus[\\'output-directory\\']\\n2. Output file name should be `<descriptive name>.h5ad`\\n\"\"\"\\n\\n\\n**CALL PYTHON SCRIPTS FROM PYTHON**:\\nUse the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\\n\\nTo call a Python script `example_script.py` which has no arguments:\\n\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\']], capture_output=True, text=True)\\n```\\n\\nTo call a Python script `example_script.py` with one argument:\\n\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\'], \\'arg1\\'], capture_output=True, text=True)\\n```\\n\\nTo call a Python script `example_script.py` with two arguments:\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\'], \\'arg1\\', \\'arg2\\'], capture_output=True, text=True)\\n```\\n\\nNote that chatstatus[\\'output-directory\\'] is ALWAYS passed as the first argument.\\n\\nThe following files were previously created by BRAD and could be used as input to a function if necessary:\\nlog.json\\n\\nQuery:{input}\\n\\n**PYTHON SCRIPT**\\nPlease use the correct script name and path. You must run this python script:\\n/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\n\\n**INSTRUCTIONS**\\n1. Given the user query and the documentation, identify each of the arguments found in the user\\'s query that should be passed to the Python script.\\n2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\\n3. The last line of your response should say \"Execute: <Python code to execute>\"\\n4. Do not include any extra words or characters. Format the response/output as:\\n    Arguments: \\n    Python Code Explanation: <2 sentences maximum>\\n    Execute: <your code here>\\n\\n**IMPORTANT**\\nThe code to execute from your response must be formatted as:\\n    Execute: subprocess.call([sys.executable, \\'<path to python script>\\', \\'<argument 1>\\', \\'<argument 2>\\', ..., \\'<argument n>\\'], capture_output=True, text=True))\\nThis output should be exactly one line and no longer. Stop the response after this line.\\n'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation:\n",
      "\n",
      "\n",
      "**PYTHON SCRIPT**\n",
      "You must run this python script:\n",
      "/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\n",
      "\n",
      "**PYTHON SCRIPT DOCUMENTATION**:\n",
      "This is the doc string of this python script:\n",
      "\"\"\"\n",
      "This script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\n",
      "\n",
      "Arguments (four arguments):\n",
      "    1. output directory: chatstatus['output-directory']\n",
      "    2. output file: <name of output file>\n",
      "    3. input file: <file created in previous step>\n",
      "    4. scanpy commands: a list of scanpy commands to be executed on the AnnData object (provided as a single string with commands separated by a delimiter, e.g., ';')\n",
      "\n",
      "Based on the arguments, the input file will be loaded, then your commands will be executed, and finally, the output or resulting ann data object will be saved to the corrrectou output file and directory. Your code is not responsible for loading the .h5ad object, that will happen automatically, and when loaded, the object will be called adata. Your scanpy commands can operate directly on the adata object that will be loaded for you.\n",
      "\n",
      "Additionally, the following imports are already provided for you and can be used in your code:\n",
      "```\n",
      "import scanpy as sc\n",
      "import seaborn as sns\n",
      "import matplotlib.pyplot as plt\n",
      "import os\n",
      "import numpy as np\n",
      "import pandas as pd\n",
      "from scipy.spatial.distance import cdist\n",
      "```\n",
      "\n",
      "**Usage**\n",
      "BRAD Line:\n",
      "```\n",
      "subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus['output-directory'], <output file>, <input file>, \"<scanpy commands>\"], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "**Examples**\n",
      "Use the below examples to help generate your code.\n",
      "\n",
      "*Example 1*\n",
      "User Prompt: Run scanpy preprocessing and UMAP visualization on XXX.h5ad and save the UMAP plot\n",
      "Response Code:\n",
      "```\n",
      "response = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus['output-directory'], \"XXX-modified.h5ad\", \"<path/to/data>/XXX.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save='umap.png')\"], capture_output=True, text=True)\n",
      "```\n",
      "Explination: the adata object will be loaded in memory already. The command \"sc.pp.neighbors(adata)\" will preprocess the data, then the command \"sc.tl.umap(adata)\" will perform UMAP, and finally the command \"sc.pl.umap(adata, 'umap.png')\" will save the UMAP to a well named file.\n",
      "\n",
      "*Example 2*\n",
      "User Prompt: Perform PCA and clustering on the dataset YYY.h5ad and save the PCA plot\n",
      "Response Code:\n",
      "```\n",
      "response = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus['output-directory'], \"YYY-modified.h5ad\", \"<path/to/data>/YYY.h5ad\", \"sc.pp.pca(adata); sc.tl.leiden(adata); sc.pl.pca(adata, save='pca.png')\"], capture_output=True, text=True)\n",
      "```\n",
      "Explination: the adata object will be loaded in memory already. The command \"sc.pp.pca(adata)\" will preprocess the data, then the command \"sc.tl.leiden(adata)\" will perform the leiden algorithm, and finally the command \"sc.pl.pca(adata, save='pca.png')\" will save the PCA to a well named file.\n",
      "\n",
      "*Example 3*\n",
      "User Prompt: compute the distance between all cells that are not source or target to all source and target cells.\n",
      "Reponse Code:\n",
      "```\n",
      "adata_X = adata.X\n",
      "mask = ~adata.obs['recipe'].isin(['source', 'target'])\n",
      "recipe_embeddings = adata[mask]\n",
      "source_embeddings = adata[adata.obs['recipe'] == 'source']\n",
      "target_embeddings = adata[adata.obs['recipe'] == 'target']\n",
      "recipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X)\n",
      "recipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X)\n",
      "adata_dist = sc.AnnData(X=np.concatenate((recipe_distances_source, recipe_distances_target), axis=1), var=pd.concat((source_embeddings.obs, target_embeddings.obs)), obs=recipe_embeddings.obs)\n",
      "adata_dist.write(<file name>.h5ad)\n",
      "```\n",
      "\n",
      "**OUTPUT FILE NAME INSTRUCTIONS**\n",
      "1. Output path should be chatstatus['output-directory']\n",
      "2. Output file name should be `<descriptive name>.h5ad`\n",
      "\"\"\"\n",
      "\n",
      "\n",
      "**CALL PYTHON SCRIPTS FROM PYTHON**:\n",
      "Use the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\n",
      "\n",
      "To call a Python script `example_script.py` which has no arguments:\n",
      "\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory']], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with one argument:\n",
      "\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory'], 'arg1'], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with two arguments:\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory'], 'arg1', 'arg2'], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "Note that chatstatus['output-directory'] is ALWAYS passed as the first argument.\n",
      "\n",
      "The following files were previously created by BRAD and could be used as input to a function if necessary:\n",
      "log.json\n",
      "\n",
      "Query:1 (CODE)**:\n",
      "Prompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and create a basic summary of the data.\n",
      "\n",
      "**\n",
      "\n",
      "**PYTHON SCRIPT**\n",
      "Please use the correct script name and path. You must run this python script:\n",
      "/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\n",
      "\n",
      "**INSTRUCTIONS**\n",
      "1. Given the user query and the documentation, identify each of the arguments found in the user's query that should be passed to the Python script.\n",
      "2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\n",
      "3. The last line of your response should say \"Execute: <Python code to execute>\"\n",
      "4. Do not include any extra words or characters. Format the response/output as:\n",
      "    Arguments: \n",
      "    Python Code Explanation: <2 sentences maximum>\n",
      "    Execute: <your code here>\n",
      "\n",
      "**IMPORTANT**\n",
      "The code to execute from your response must be formatted as:\n",
      "    Execute: subprocess.call([sys.executable, '<path to python script>', '<argument 1>', '<argument 2>', ..., '<argument n>'], capture_output=True, text=True))\n",
      "This output should be exactly one line and no longer. Stop the response after this line.\n",
      "\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 00:41:23,081 - INFO - HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2024-08-03 00:41:23,083 - INFO - LLM OUTPUT PARSER\n",
      "2024-08-03 00:41:23,083 - INFO - Arguments:\n",
      "1. output directory: chatstatus['output-directory']\n",
      "2. output file: \"HSC-reprogrammed-DE-summary.h5ad\"\n",
      "3. input file: \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\"\n",
      "4. scanpy commands: \"print(adata)\"\n",
      "\n",
      "Python Code Explanation: This code will load the AnnData object from the specified input file, print a basic summary of the data, and save the result to the specified output file.\n",
      "\n",
      "Execute: subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-DE-summary.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"print(adata)\"], capture_output=True, text=True)\n",
      "2024-08-03 00:41:23,083 - INFO - ['Arguments:\\n1. output directory: chatstatus[\\'output-directory\\']\\n2. output file: \"HSC-reprogrammed-DE-summary.h5ad\"\\n3. input file: \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\"\\n4. scanpy commands: \"print(adata)\"\\n\\nPython Code Explanation: This code will load the AnnData object from the specified input file, print a basic summary of the data, and save the result to the specified output file.\\n\\n', ' subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"HSC-reprogrammed-DE-summary.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"print(adata)\"], capture_output=True, text=True)']\n",
      "2024-08-03 00:41:23,084 - INFO - Stripped funciton call\n",
      "2024-08-03 00:41:23,084 - INFO - subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-DE-summary.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"print(adata)\"], capture_output=True, text=True)\n",
      "2024-08-03 00:41:23,084 - INFO - Cleaned up function call:\n",
      "\n",
      "2024-08-03 00:41:23,084 - INFO - subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-DE-summary.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"print(adata)\"], capture_output=True, text=True)\n",
      "2024-08-03 00:41:23,085 - INFO - \n",
      "\n",
      "2024-08-03 00:41:23,086 - INFO - EVAL\n",
      "2024-08-03 00:41:23,086 - INFO - subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-DE-summary.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"print(adata)\"], capture_output=True, text=True)\n",
      "2024-08-03 00:41:23,087 - INFO - END EVAL CHECK\n",
      "2024-08-03 00:41:23,087 - INFO - Finalized PYTHON Call:\n",
      "2024-08-03 00:41:23,087 - INFO -    \n",
      "2024-08-03 00:41:23,087 - INFO - response = subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-DE-summary.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"print(adata)\"], capture_output=True, text=True)\n",
      "2024-08-03 00:41:23,087 - INFO -    \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 00:42:25,939 - INFO - Debug: PYTHON code executed successfully.\n",
      "2024-08-03 00:42:25,940 - INFO - Code Execution Output:\n",
      "2024-08-03 00:42:25,940 - INFO - CompletedProcess(args=['/sw/pkgs/arc/python3.11-anaconda/2024.02-1/bin/python', '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py', '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_00-39-53', 'HSC-reprogrammed-DE-summary.h5ad', '/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad', 'print(adata)'], returncode=0, stdout=\"adata loaded from /nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\\nadata.shape=(6375, 512)\\nadata.obs.columns=Index(['recipe', 'filename', 'REL', 'STAT5A', 'FOS', 'ETV6', 'source', 'MEF2C',\\n       'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B', 'target'],\\n      dtype='object')\\nadata.obs.head()=   recipe                                           filename  ...  GFI1B  target\\n0  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n1  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n2  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n3  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n4  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n\\n[5 rows x 14 columns]\\nadata.var.head()=Empty DataFrame\\nColumns: []\\nIndex: [0, 1, 2, 3, 4]\\n****************************\\n      EXECUTE COMMANDS      \\n****************************\\nprint(adata)\\nAnnData object with n_obs Ã— n_vars = 6375 Ã— 512\\n    obs: 'recipe', 'filename', 'REL', 'STAT5A', 'FOS', 'ETV6', 'source', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B', 'target'\\n\", stderr='/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\\n  utils.warn_names_duplicates(\"obs\")\\n')\n",
      "2024-08-03 00:42:25,940 - INFO - Debug: PYTHON code output saved to output.\n",
      "2024-08-03 00:42:25,941 - INFO - \n",
      "\n",
      "\n",
      "route\n",
      "\n",
      "\n",
      "\n",
      "2024-08-03 00:42:25,941 - INFO - CODE\n",
      "2024-08-03 00:42:25,942 - INFO - Finding Stage Number of Pipeline\n",
      "2024-08-03 00:42:25,942 - INFO - {0: {'order': 0, 'module': 'PLANNER', 'prompt': None, 'description': 'This step designed the plan. It is placed in the queue because we needed a place holder for 0 indexed lists.', 'output': []}, 1: {'order': 1, 'module': 'CODE', 'prompt': '/force CODE  1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and create a basic summary of the data.\\n\\n**', 'description': ' 1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and create a basic summary of the data.\\n\\n**', 'output': []}, 2: {'order': 2, 'module': 'CODE', 'prompt': \"/force CODE  2 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, create UMAP plots of the data in 2D and 3D space and color the visualizations based on the 'recipe' column to distinguish between 'source', 'target', and different TF recipes.\\n\\n**\", 'description': \" 2 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, create UMAP plots of the data in 2D and 3D space and color the visualizations based on the 'recipe' column to distinguish between 'source', 'target', and different TF recipes.\\n\\n**\", 'output': []}, 3: {'order': 3, 'module': 'CODE', 'prompt': \"/force CODE  3 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, create t-SNE plots of the data in 2D and 3D space and color the visualizations based on the 'recipe' column to distinguish between 'source', 'target', and different TF recipes.\\n\\n**\", 'description': \" 3 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, create t-SNE plots of the data in 2D and 3D space and color the visualizations based on the 'recipe' column to distinguish between 'source', 'target', and different TF recipes.\\n\\n**\", 'output': []}, 4: {'order': 4, 'module': 'CODE', 'prompt': '/force CODE  4 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, generate separate UMAP visualizations highlighting the presence of each TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) by coloring the points based on the True/False values in the respective columns.\\n\\n**', 'description': ' 4 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, generate separate UMAP visualizations highlighting the presence of each TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) by coloring the points based on the True/False values in the respective columns.\\n\\n**', 'output': []}, 5: {'order': 5, 'module': 'CODE', 'prompt': '/force CODE  5 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, generate separate t-SNE visualizations highlighting the presence of each TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) by coloring the points based on the True/False values in the respective columns.\\n\\nThe removed ', 'description': ' 5 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, generate separate t-SNE visualizations highlighting the presence of each TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) by coloring the points based on the True/False values in the respective columns.\\n\\nThe removed ', 'output': []}}\n",
      "2024-08-03 00:42:25,943 - INFO - IP=1\n",
      "2024-08-03 00:42:25,943 - INFO - type(IP)=<class 'int'>\n",
      "2024-08-03 00:42:25,945 - INFO - \n",
      "\n",
      "PLANNER:\n",
      "\n",
      "{'module': 'CODE', 'steps': [{'llm': \"client=<openai.resources.chat.completions.Completions object at 0x1480cc33e8d0> async_client=<openai.resources.chat.completions.AsyncCompletions object at 0x1480cc324e50> model_name='gpt-4o' temperature=0.0 openai_api_key=SecretStr('**********') openai_proxy=''\", 'memory': 'None', 'prompt': 'input_variables=[\\'user_query\\'] template=\"You must select which code to run to help a user.\\\\n\\\\n**Available Scripts**\\\\nScript Name: scanpy_brad, \\\\t Description: This script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\\\\nScript Name: models_brad, \\\\t Description: This script executes a series of custom commands to build a model for predicting the contribution to reprogramming obtained by the inclusion or exclusion of individual TFs using an AnnData object or single cell datasets loaded from a .h5ad file.\\\\n\\\\n\\\\n**User Query**\\\\n{user_query}\\\\n\\\\n**Task**\\\\nBased on the user\\'s query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\\\\n\\\\n**Response Template**\\\\nSCRIPT: <selected script>\\\\nREASON: <reasoning why the script fits the user prompt>\\\\n\"', 'input': '1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and create a basic summary of the data.\\n\\n**', 'output': 'content=\"SCRIPT: scanpy_brad\\\\nREASON: The user\\'s query involves loading an AnnData object from a .h5ad file and creating a basic summary of the data. The `scanpy_brad` script is designed to handle tasks related to manipulating, visualizing, and exploring single-cell datasets loaded from .h5ad files. It allows the use of libraries like scanpy, seaborn, and matplotlib.pyplot, which are well-suited for summarizing and visualizing single-cell data. Therefore, `scanpy_brad` is the best match for the user\\'s request.\" response_metadata={\\'token_usage\\': {\\'completion_tokens\\': 117, \\'prompt_tokens\\': 308, \\'total_tokens\\': 425}, \\'model_name\\': \\'gpt-4o-2024-05-13\\', \\'system_fingerprint\\': \\'fp_4e2b2da518\\', \\'finish_reason\\': \\'stop\\', \\'logprobs\\': None} id=\\'run-6eebba13-747b-4aaf-983d-d42af6d4213c-0\\' usage_metadata={\\'input_tokens\\': 308, \\'output_tokens\\': 117, \\'total_tokens\\': 425}', 'parsedOutput': {'scriptName': '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py', 'scriptType': 'python', 'scriptPath': '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts'}, 'purpose': 'Select which code to run'}, {'llm': \"client=<openai.resources.chat.completions.Completions object at 0x1480cc33e8d0> async_client=<openai.resources.chat.completions.AsyncCompletions object at 0x1480cc324e50> model_name='gpt-4o' temperature=0.0 openai_api_key=SecretStr('**********') openai_proxy=''\", 'memory': 'None', 'prompt': 'input_variables=[\\'history\\', \\'input\\'] template=\\'Current conversation:\\\\n{history}\\\\n\\\\n**PYTHON SCRIPT**\\\\nYou must run this python script:\\\\n/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\\\n\\\\n**PYTHON SCRIPT DOCUMENTATION**:\\\\nThis is the doc string of this python script:\\\\n\"\"\"\\\\nThis script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\\\\n\\\\nArguments (four arguments):\\\\n    1. output directory: chatstatus[\\\\\\'output-directory\\\\\\']\\\\n    2. output file: <name of output file>\\\\n    3. input file: <file created in previous step>\\\\n    4. scanpy commands: a list of scanpy commands to be executed on the AnnData object (provided as a single string with commands separated by a delimiter, e.g., \\\\\\';\\\\\\')\\\\n\\\\nBased on the arguments, the input file will be loaded, then your commands will be executed, and finally, the output or resulting ann data object will be saved to the corrrectou output file and directory. Your code is not responsible for loading the .h5ad object, that will happen automatically, and when loaded, the object will be called adata. Your scanpy commands can operate directly on the adata object that will be loaded for you.\\\\n\\\\nAdditionally, the following imports are already provided for you and can be used in your code:\\\\n```\\\\nimport scanpy as sc\\\\nimport seaborn as sns\\\\nimport matplotlib.pyplot as plt\\\\nimport os\\\\nimport numpy as np\\\\nimport pandas as pd\\\\nfrom scipy.spatial.distance import cdist\\\\n```\\\\n\\\\n**Usage**\\\\nBRAD Line:\\\\n```\\\\nsubprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\\\\\'output-directory\\\\\\'], <output file>, <input file>, \"<scanpy commands>\"], capture_output=True, text=True)\\\\n```\\\\n\\\\n**Examples**\\\\nUse the below examples to help generate your code.\\\\n\\\\n*Example 1*\\\\nUser Prompt: Run scanpy preprocessing and UMAP visualization on XXX.h5ad and save the UMAP plot\\\\nResponse Code:\\\\n```\\\\nresponse = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\\\\\'output-directory\\\\\\'], \"XXX-modified.h5ad\", \"<path/to/data>/XXX.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save=\\\\\\'umap.png\\\\\\')\"], capture_output=True, text=True)\\\\n```\\\\nExplination: the adata object will be loaded in memory already. The command \"sc.pp.neighbors(adata)\" will preprocess the data, then the command \"sc.tl.umap(adata)\" will perform UMAP, and finally the command \"sc.pl.umap(adata, \\\\\\'umap.png\\\\\\')\" will save the UMAP to a well named file.\\\\n\\\\n*Example 2*\\\\nUser Prompt: Perform PCA and clustering on the dataset YYY.h5ad and save the PCA plot\\\\nResponse Code:\\\\n```\\\\nresponse = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\\\\\'output-directory\\\\\\'], \"YYY-modified.h5ad\", \"<path/to/data>/YYY.h5ad\", \"sc.pp.pca(adata); sc.tl.leiden(adata); sc.pl.pca(adata, save=\\\\\\'pca.png\\\\\\')\"], capture_output=True, text=True)\\\\n```\\\\nExplination: the adata object will be loaded in memory already. The command \"sc.pp.pca(adata)\" will preprocess the data, then the command \"sc.tl.leiden(adata)\" will perform the leiden algorithm, and finally the command \"sc.pl.pca(adata, save=\\\\\\'pca.png\\\\\\')\" will save the PCA to a well named file.\\\\n\\\\n*Example 3*\\\\nUser Prompt: compute the distance between all cells that are not source or target to all source and target cells.\\\\nReponse Code:\\\\n```\\\\nadata_X = adata.X\\\\nmask = ~adata.obs[\\\\\\'recipe\\\\\\'].isin([\\\\\\'source\\\\\\', \\\\\\'target\\\\\\'])\\\\nrecipe_embeddings = adata[mask]\\\\nsource_embeddings = adata[adata.obs[\\\\\\'recipe\\\\\\'] == \\\\\\'source\\\\\\']\\\\ntarget_embeddings = adata[adata.obs[\\\\\\'recipe\\\\\\'] == \\\\\\'target\\\\\\']\\\\nrecipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X)\\\\nrecipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X)\\\\nadata_dist = sc.AnnData(X=np.concatenate((recipe_distances_source, recipe_distances_target), axis=1), var=pd.concat((source_embeddings.obs, target_embeddings.obs)), obs=recipe_embeddings.obs)\\\\nadata_dist.write(<file name>.h5ad)\\\\n```\\\\n\\\\n**OUTPUT FILE NAME INSTRUCTIONS**\\\\n1. Output path should be chatstatus[\\\\\\'output-directory\\\\\\']\\\\n2. Output file name should be `<descriptive name>.h5ad`\\\\n\"\"\"\\\\n\\\\n\\\\n**CALL PYTHON SCRIPTS FROM PYTHON**:\\\\nUse the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\\\\n\\\\nTo call a Python script `example_script.py` which has no arguments:\\\\n\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\']], capture_output=True, text=True)\\\\n```\\\\n\\\\nTo call a Python script `example_script.py` with one argument:\\\\n\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\'], \\\\\\'arg1\\\\\\'], capture_output=True, text=True)\\\\n```\\\\n\\\\nTo call a Python script `example_script.py` with two arguments:\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\'], \\\\\\'arg1\\\\\\', \\\\\\'arg2\\\\\\'], capture_output=True, text=True)\\\\n```\\\\n\\\\nNote that chatstatus[\\\\\\'output-directory\\\\\\'] is ALWAYS passed as the first argument.\\\\n\\\\nThe following files were previously created by BRAD and could be used as input to a function if necessary:\\\\nlog.json\\\\n\\\\nQuery:{input}\\\\n\\\\n**PYTHON SCRIPT**\\\\nPlease use the correct script name and path. You must run this python script:\\\\n/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\\\n\\\\n**INSTRUCTIONS**\\\\n1. Given the user query and the documentation, identify each of the arguments found in the user\\\\\\'s query that should be passed to the Python script.\\\\n2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\\\\n3. The last line of your response should say \"Execute: <Python code to execute>\"\\\\n4. Do not include any extra words or characters. Format the response/output as:\\\\n    Arguments: \\\\n    Python Code Explanation: <2 sentences maximum>\\\\n    Execute: <your code here>\\\\n\\\\n**IMPORTANT**\\\\nThe code to execute from your response must be formatted as:\\\\n    Execute: subprocess.call([sys.executable, \\\\\\'<path to python script>\\\\\\', \\\\\\'<argument 1>\\\\\\', \\\\\\'<argument 2>\\\\\\', ..., \\\\\\'<argument n>\\\\\\'], capture_output=True, text=True))\\\\nThis output should be exactly one line and no longer. Stop the response after this line.\\\\n\\'', 'input': '1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and create a basic summary of the data.\\n\\n**', 'output': 'Arguments:\\n1. output directory: chatstatus[\\'output-directory\\']\\n2. output file: \"HSC-reprogrammed-DE-summary.h5ad\"\\n3. input file: \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\"\\n4. scanpy commands: \"print(adata)\"\\n\\nPython Code Explanation: This code will load the AnnData object from the specified input file, print a basic summary of the data, and save the result to the specified output file.\\n\\nExecute: subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"HSC-reprogrammed-DE-summary.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"print(adata)\"], capture_output=True, text=True)', 'parsedOutput': {'code': 'subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"HSC-reprogrammed-DE-summary.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"print(adata)\"], capture_output=True, text=True)'}, 'purpose': 'Format function call'}, {'func': 'pythonCaller.execute_python_code', 'code': 'response = subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"HSC-reprogrammed-DE-summary.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"print(adata)\"], capture_output=True, text=True)', 'purpose': 'execute python code'}, {'func': 'utils.makeNamesConsistent', 'files': [{'old-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_00-39-53/HSC-reprogrammed-DE-summary.h5ad', 'new-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_00-39-53/S1-HSC-reprogrammed-DE-summary.h5ad'}]}]}\n",
      "2024-08-03 00:42:25,950 - INFO - current llm calls=3\n",
      "2024-08-03 00:42:25,952 - INFO - CODE\n",
      "2024-08-03 00:42:25,952 - INFO - CODER\n",
      "2024-08-03 00:42:25,953 - INFO - input_variables=['user_query'] template=\"You must select which code to run to help a user.\\n\\n**Available Scripts**\\nScript Name: scanpy_brad, \\t Description: This script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\\nScript Name: models_brad, \\t Description: This script executes a series of custom commands to build a model for predicting the contribution to reprogramming obtained by the inclusion or exclusion of individual TFs using an AnnData object or single cell datasets loaded from a .h5ad file.\\n\\n\\n**User Query**\\n{user_query}\\n\\n**Task**\\nBased on the user's query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\\n\\n**Response Template**\\nSCRIPT: <selected script>\\nREASON: <reasoning why the script fits the user prompt>\\n\"\n",
      "2024-08-03 00:42:25,954 - INFO - FIRST LLM CALL\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "==================================================\n",
      "BRAD >> 2: \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 00:42:27,562 - INFO - HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2024-08-03 00:42:27,563 - INFO - SCRIPT: scanpy_brad\n",
      "REASON: The user's query involves creating UMAP plots of an AnnData object and coloring the visualizations based on a specific column ('recipe'). The `scanpy_brad` script is designed to manipulate, visualize, and explore single cell datasets using libraries like scanpy, seaborn, and matplotlib.pyplot, which are well-suited for creating UMAP plots. The `models_brad` script, on the other hand, is focused on building predictive models and is not appropriate for the visualization task described by the user.\n",
      "2024-08-03 00:42:27,564 - INFO - scriptPath=/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts\n",
      "2024-08-03 00:42:27,564 - INFO - ALL SCRIPTS FOUND. BUILDING TEMPLATE\n",
      "2024-08-03 00:42:27,565 - INFO - input_variables=['history', 'input'] template='Current conversation:\\n{history}\\n\\n**PYTHON SCRIPT**\\nYou must run this python script:\\n/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\n\\n**PYTHON SCRIPT DOCUMENTATION**:\\nThis is the doc string of this python script:\\n\"\"\"\\nThis script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\\n\\nArguments (four arguments):\\n    1. output directory: chatstatus[\\'output-directory\\']\\n    2. output file: <name of output file>\\n    3. input file: <file created in previous step>\\n    4. scanpy commands: a list of scanpy commands to be executed on the AnnData object (provided as a single string with commands separated by a delimiter, e.g., \\';\\')\\n\\nBased on the arguments, the input file will be loaded, then your commands will be executed, and finally, the output or resulting ann data object will be saved to the corrrectou output file and directory. Your code is not responsible for loading the .h5ad object, that will happen automatically, and when loaded, the object will be called adata. Your scanpy commands can operate directly on the adata object that will be loaded for you.\\n\\nAdditionally, the following imports are already provided for you and can be used in your code:\\n```\\nimport scanpy as sc\\nimport seaborn as sns\\nimport matplotlib.pyplot as plt\\nimport os\\nimport numpy as np\\nimport pandas as pd\\nfrom scipy.spatial.distance import cdist\\n```\\n\\n**Usage**\\nBRAD Line:\\n```\\nsubprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\'output-directory\\'], <output file>, <input file>, \"<scanpy commands>\"], capture_output=True, text=True)\\n```\\n\\n**Examples**\\nUse the below examples to help generate your code.\\n\\n*Example 1*\\nUser Prompt: Run scanpy preprocessing and UMAP visualization on XXX.h5ad and save the UMAP plot\\nResponse Code:\\n```\\nresponse = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"XXX-modified.h5ad\", \"<path/to/data>/XXX.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save=\\'umap.png\\')\"], capture_output=True, text=True)\\n```\\nExplination: the adata object will be loaded in memory already. The command \"sc.pp.neighbors(adata)\" will preprocess the data, then the command \"sc.tl.umap(adata)\" will perform UMAP, and finally the command \"sc.pl.umap(adata, \\'umap.png\\')\" will save the UMAP to a well named file.\\n\\n*Example 2*\\nUser Prompt: Perform PCA and clustering on the dataset YYY.h5ad and save the PCA plot\\nResponse Code:\\n```\\nresponse = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"YYY-modified.h5ad\", \"<path/to/data>/YYY.h5ad\", \"sc.pp.pca(adata); sc.tl.leiden(adata); sc.pl.pca(adata, save=\\'pca.png\\')\"], capture_output=True, text=True)\\n```\\nExplination: the adata object will be loaded in memory already. The command \"sc.pp.pca(adata)\" will preprocess the data, then the command \"sc.tl.leiden(adata)\" will perform the leiden algorithm, and finally the command \"sc.pl.pca(adata, save=\\'pca.png\\')\" will save the PCA to a well named file.\\n\\n*Example 3*\\nUser Prompt: compute the distance between all cells that are not source or target to all source and target cells.\\nReponse Code:\\n```\\nadata_X = adata.X\\nmask = ~adata.obs[\\'recipe\\'].isin([\\'source\\', \\'target\\'])\\nrecipe_embeddings = adata[mask]\\nsource_embeddings = adata[adata.obs[\\'recipe\\'] == \\'source\\']\\ntarget_embeddings = adata[adata.obs[\\'recipe\\'] == \\'target\\']\\nrecipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X)\\nrecipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X)\\nadata_dist = sc.AnnData(X=np.concatenate((recipe_distances_source, recipe_distances_target), axis=1), var=pd.concat((source_embeddings.obs, target_embeddings.obs)), obs=recipe_embeddings.obs)\\nadata_dist.write(<file name>.h5ad)\\n```\\n\\n**OUTPUT FILE NAME INSTRUCTIONS**\\n1. Output path should be chatstatus[\\'output-directory\\']\\n2. Output file name should be `<descriptive name>.h5ad`\\n\"\"\"\\n\\n\\n**CALL PYTHON SCRIPTS FROM PYTHON**:\\nUse the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\\n\\nTo call a Python script `example_script.py` which has no arguments:\\n\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\']], capture_output=True, text=True)\\n```\\n\\nTo call a Python script `example_script.py` with one argument:\\n\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\'], \\'arg1\\'], capture_output=True, text=True)\\n```\\n\\nTo call a Python script `example_script.py` with two arguments:\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\'], \\'arg1\\', \\'arg2\\'], capture_output=True, text=True)\\n```\\n\\nNote that chatstatus[\\'output-directory\\'] is ALWAYS passed as the first argument.\\n\\nThe following files were previously created by BRAD and could be used as input to a function if necessary:\\nS1-HSC-reprogrammed-DE-summary.h5ad\\nlog.json\\n\\nQuery:{input}\\n\\n**PYTHON SCRIPT**\\nPlease use the correct script name and path. You must run this python script:\\n/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\n\\n**INSTRUCTIONS**\\n1. Given the user query and the documentation, identify each of the arguments found in the user\\'s query that should be passed to the Python script.\\n2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\\n3. The last line of your response should say \"Execute: <Python code to execute>\"\\n4. Do not include any extra words or characters. Format the response/output as:\\n    Arguments: \\n    Python Code Explanation: <2 sentences maximum>\\n    Execute: <your code here>\\n\\n**IMPORTANT**\\nThe code to execute from your response must be formatted as:\\n    Execute: subprocess.call([sys.executable, \\'<path to python script>\\', \\'<argument 1>\\', \\'<argument 2>\\', ..., \\'<argument n>\\'], capture_output=True, text=True))\\nThis output should be exactly one line and no longer. Stop the response after this line.\\n'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation:\n",
      "\n",
      "\n",
      "**PYTHON SCRIPT**\n",
      "You must run this python script:\n",
      "/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\n",
      "\n",
      "**PYTHON SCRIPT DOCUMENTATION**:\n",
      "This is the doc string of this python script:\n",
      "\"\"\"\n",
      "This script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\n",
      "\n",
      "Arguments (four arguments):\n",
      "    1. output directory: chatstatus['output-directory']\n",
      "    2. output file: <name of output file>\n",
      "    3. input file: <file created in previous step>\n",
      "    4. scanpy commands: a list of scanpy commands to be executed on the AnnData object (provided as a single string with commands separated by a delimiter, e.g., ';')\n",
      "\n",
      "Based on the arguments, the input file will be loaded, then your commands will be executed, and finally, the output or resulting ann data object will be saved to the corrrectou output file and directory. Your code is not responsible for loading the .h5ad object, that will happen automatically, and when loaded, the object will be called adata. Your scanpy commands can operate directly on the adata object that will be loaded for you.\n",
      "\n",
      "Additionally, the following imports are already provided for you and can be used in your code:\n",
      "```\n",
      "import scanpy as sc\n",
      "import seaborn as sns\n",
      "import matplotlib.pyplot as plt\n",
      "import os\n",
      "import numpy as np\n",
      "import pandas as pd\n",
      "from scipy.spatial.distance import cdist\n",
      "```\n",
      "\n",
      "**Usage**\n",
      "BRAD Line:\n",
      "```\n",
      "subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus['output-directory'], <output file>, <input file>, \"<scanpy commands>\"], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "**Examples**\n",
      "Use the below examples to help generate your code.\n",
      "\n",
      "*Example 1*\n",
      "User Prompt: Run scanpy preprocessing and UMAP visualization on XXX.h5ad and save the UMAP plot\n",
      "Response Code:\n",
      "```\n",
      "response = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus['output-directory'], \"XXX-modified.h5ad\", \"<path/to/data>/XXX.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save='umap.png')\"], capture_output=True, text=True)\n",
      "```\n",
      "Explination: the adata object will be loaded in memory already. The command \"sc.pp.neighbors(adata)\" will preprocess the data, then the command \"sc.tl.umap(adata)\" will perform UMAP, and finally the command \"sc.pl.umap(adata, 'umap.png')\" will save the UMAP to a well named file.\n",
      "\n",
      "*Example 2*\n",
      "User Prompt: Perform PCA and clustering on the dataset YYY.h5ad and save the PCA plot\n",
      "Response Code:\n",
      "```\n",
      "response = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus['output-directory'], \"YYY-modified.h5ad\", \"<path/to/data>/YYY.h5ad\", \"sc.pp.pca(adata); sc.tl.leiden(adata); sc.pl.pca(adata, save='pca.png')\"], capture_output=True, text=True)\n",
      "```\n",
      "Explination: the adata object will be loaded in memory already. The command \"sc.pp.pca(adata)\" will preprocess the data, then the command \"sc.tl.leiden(adata)\" will perform the leiden algorithm, and finally the command \"sc.pl.pca(adata, save='pca.png')\" will save the PCA to a well named file.\n",
      "\n",
      "*Example 3*\n",
      "User Prompt: compute the distance between all cells that are not source or target to all source and target cells.\n",
      "Reponse Code:\n",
      "```\n",
      "adata_X = adata.X\n",
      "mask = ~adata.obs['recipe'].isin(['source', 'target'])\n",
      "recipe_embeddings = adata[mask]\n",
      "source_embeddings = adata[adata.obs['recipe'] == 'source']\n",
      "target_embeddings = adata[adata.obs['recipe'] == 'target']\n",
      "recipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X)\n",
      "recipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X)\n",
      "adata_dist = sc.AnnData(X=np.concatenate((recipe_distances_source, recipe_distances_target), axis=1), var=pd.concat((source_embeddings.obs, target_embeddings.obs)), obs=recipe_embeddings.obs)\n",
      "adata_dist.write(<file name>.h5ad)\n",
      "```\n",
      "\n",
      "**OUTPUT FILE NAME INSTRUCTIONS**\n",
      "1. Output path should be chatstatus['output-directory']\n",
      "2. Output file name should be `<descriptive name>.h5ad`\n",
      "\"\"\"\n",
      "\n",
      "\n",
      "**CALL PYTHON SCRIPTS FROM PYTHON**:\n",
      "Use the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\n",
      "\n",
      "To call a Python script `example_script.py` which has no arguments:\n",
      "\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory']], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with one argument:\n",
      "\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory'], 'arg1'], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with two arguments:\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory'], 'arg1', 'arg2'], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "Note that chatstatus['output-directory'] is ALWAYS passed as the first argument.\n",
      "\n",
      "The following files were previously created by BRAD and could be used as input to a function if necessary:\n",
      "S1-HSC-reprogrammed-DE-summary.h5ad\n",
      "log.json\n",
      "\n",
      "Query:2 (CODE)**:\n",
      "Prompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, create UMAP plots of the data in 2D and 3D space and color the visualizations based on the 'recipe' column to distinguish between 'source', 'target', and different TF recipes.\n",
      "\n",
      "**\n",
      "\n",
      "**PYTHON SCRIPT**\n",
      "Please use the correct script name and path. You must run this python script:\n",
      "/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\n",
      "\n",
      "**INSTRUCTIONS**\n",
      "1. Given the user query and the documentation, identify each of the arguments found in the user's query that should be passed to the Python script.\n",
      "2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\n",
      "3. The last line of your response should say \"Execute: <Python code to execute>\"\n",
      "4. Do not include any extra words or characters. Format the response/output as:\n",
      "    Arguments: \n",
      "    Python Code Explanation: <2 sentences maximum>\n",
      "    Execute: <your code here>\n",
      "\n",
      "**IMPORTANT**\n",
      "The code to execute from your response must be formatted as:\n",
      "    Execute: subprocess.call([sys.executable, '<path to python script>', '<argument 1>', '<argument 2>', ..., '<argument n>'], capture_output=True, text=True))\n",
      "This output should be exactly one line and no longer. Stop the response after this line.\n",
      "\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 00:42:30,822 - INFO - HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2024-08-03 00:42:30,824 - INFO - LLM OUTPUT PARSER\n",
      "2024-08-03 00:42:30,824 - INFO - Arguments:\n",
      "1. output directory: chatstatus['output-directory']\n",
      "2. output file: \"HSC-reprogrammed-UMAP.h5ad\"\n",
      "3. input file: \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\"\n",
      "4. scanpy commands: \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, color='recipe', save='umap_2d.png'); sc.pl.umap(adata, color='recipe', save='umap_3d.png', projection='3d')\"\n",
      "\n",
      "Python Code Explanation: This code will preprocess the data, perform UMAP in both 2D and 3D, and save the visualizations colored by the 'recipe' column.\n",
      "\n",
      "Execute: subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-UMAP.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, color='recipe', save='umap_2d.png'); sc.pl.umap(adata, color='recipe', save='umap_3d.png', projection='3d')\"], capture_output=True, text=True)\n",
      "2024-08-03 00:42:30,824 - INFO - ['Arguments:\\n1. output directory: chatstatus[\\'output-directory\\']\\n2. output file: \"HSC-reprogrammed-UMAP.h5ad\"\\n3. input file: \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\"\\n4. scanpy commands: \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, color=\\'recipe\\', save=\\'umap_2d.png\\'); sc.pl.umap(adata, color=\\'recipe\\', save=\\'umap_3d.png\\', projection=\\'3d\\')\"\\n\\nPython Code Explanation: This code will preprocess the data, perform UMAP in both 2D and 3D, and save the visualizations colored by the \\'recipe\\' column.\\n\\n', ' subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"HSC-reprogrammed-UMAP.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, color=\\'recipe\\', save=\\'umap_2d.png\\'); sc.pl.umap(adata, color=\\'recipe\\', save=\\'umap_3d.png\\', projection=\\'3d\\')\"], capture_output=True, text=True)']\n",
      "2024-08-03 00:42:30,824 - INFO - Stripped funciton call\n",
      "2024-08-03 00:42:30,825 - INFO - subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-UMAP.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, color='recipe', save='umap_2d.png'); sc.pl.umap(adata, color='recipe', save='umap_3d.png', projection='3d')\"], capture_output=True, text=True)\n",
      "2024-08-03 00:42:30,825 - INFO - Cleaned up function call:\n",
      "\n",
      "2024-08-03 00:42:30,825 - INFO - subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-UMAP.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, color='recipe', save='umap_2d.png'); sc.pl.umap(adata, color='recipe', save='umap_3d.png', projection='3d')\"], capture_output=True, text=True)\n",
      "2024-08-03 00:42:30,825 - INFO - \n",
      "\n",
      "2024-08-03 00:42:30,826 - INFO - EVAL\n",
      "2024-08-03 00:42:30,826 - INFO - subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-UMAP.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, color='recipe', save='umap_2d.png'); sc.pl.umap(adata, color='recipe', save='umap_3d.png', projection='3d')\"], capture_output=True, text=True)\n",
      "2024-08-03 00:42:30,827 - INFO - END EVAL CHECK\n",
      "2024-08-03 00:42:30,827 - INFO - Finalized PYTHON Call:\n",
      "2024-08-03 00:42:30,827 - INFO -    \n",
      "2024-08-03 00:42:30,827 - INFO - response = subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-UMAP.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, color='recipe', save='umap_2d.png'); sc.pl.umap(adata, color='recipe', save='umap_3d.png', projection='3d')\"], capture_output=True, text=True)\n",
      "2024-08-03 00:42:30,828 - INFO -    \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 00:43:24,536 - INFO - Debug: PYTHON code executed successfully.\n",
      "2024-08-03 00:43:24,537 - INFO - Code Execution Output:\n",
      "2024-08-03 00:43:24,537 - INFO - CompletedProcess(args=['/sw/pkgs/arc/python3.11-anaconda/2024.02-1/bin/python', '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py', '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_00-39-53', 'HSC-reprogrammed-UMAP.h5ad', '/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad', \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, color='recipe', save='umap_2d.png'); sc.pl.umap(adata, color='recipe', save='umap_3d.png', projection='3d')\"], returncode=1, stdout=\"adata loaded from /nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\\nadata.shape=(6375, 512)\\nadata.obs.columns=Index(['recipe', 'filename', 'REL', 'STAT5A', 'FOS', 'ETV6', 'source', 'MEF2C',\\n       'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B', 'target'],\\n      dtype='object')\\nadata.obs.head()=   recipe                                           filename  ...  GFI1B  target\\n0  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n1  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n2  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n3  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n4  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n\\n[5 rows x 14 columns]\\nadata.var.head()=Empty DataFrame\\nColumns: []\\nIndex: [0, 1, 2, 3, 4]\\n****************************\\n      EXECUTE COMMANDS      \\n****************************\\nsc.pp.neighbors(adata)\\nsc.tl.umap(adata)\\nsc.pl.umap(adata, color='recipe', save='umap_2d.png')\\nFigure(1200x1200)\\nsc.pl.umap(adata, color='recipe', save='umap_3d.png', projection='3d')\\n\", stderr='/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\\n  utils.warn_names_duplicates(\"obs\")\\n/home/jpic/.local/lib/python3.11/site-packages/scanpy/tools/_utils.py:41: UserWarning: Youâ€™re trying to run this on 512 dimensions of `.X`, if you really want this, set `use_rep=\\'X\\'`.\\n         Falling back to preprocessing with `sc.pp.pca` and default params.\\n  warnings.warn(\\nWARNING: saving figure to file /home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_00-39-53/umapumap_2d.png\\nTraceback (most recent call last):\\n  File \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", line 117, in <module>\\n    main(args.output_directory, args.output_file, args.input_file, args.scanpy_commands)\\n  File \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", line 103, in main\\n    exec(command)\\n  File \"<string>\", line 1, in <module>\\n  File \"/home/jpic/.local/lib/python3.11/site-packages/scanpy/plotting/_tools/scatterplots.py\", line 690, in umap\\n    return embedding(adata, \"umap\", **kwargs)\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \"/home/jpic/.local/lib/python3.11/site-packages/scanpy/plotting/_tools/scatterplots.py\", line 308, in embedding\\n    coords = basis_values[:, dims][order, :]\\n             ~~~~~~~~~~~~^^^^^^^^^\\nIndexError: index 2 is out of bounds for axis 1 with size 2\\n')\n",
      "2024-08-03 00:43:24,538 - INFO - Debug: PYTHON code output saved to output.\n",
      "2024-08-03 00:43:24,538 - INFO - \n",
      "\n",
      "\n",
      "route\n",
      "\n",
      "\n",
      "\n",
      "2024-08-03 00:43:24,538 - INFO - CODE\n",
      "2024-08-03 00:43:24,539 - INFO - Finding Stage Number of Pipeline\n",
      "2024-08-03 00:43:24,540 - INFO - {0: {'order': 0, 'module': 'PLANNER', 'prompt': None, 'description': 'This step designed the plan. It is placed in the queue because we needed a place holder for 0 indexed lists.', 'output': []}, 1: {'order': 1, 'module': 'CODE', 'prompt': '/force CODE  1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and create a basic summary of the data.\\n\\n**', 'description': ' 1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and create a basic summary of the data.\\n\\n**', 'output': ['/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_00-39-53/S1-HSC-reprogrammed-DE-summary.h5ad']}, 2: {'order': 2, 'module': 'CODE', 'prompt': \"/force CODE  2 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, create UMAP plots of the data in 2D and 3D space and color the visualizations based on the 'recipe' column to distinguish between 'source', 'target', and different TF recipes.\\n\\n**\", 'description': \" 2 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, create UMAP plots of the data in 2D and 3D space and color the visualizations based on the 'recipe' column to distinguish between 'source', 'target', and different TF recipes.\\n\\n**\", 'output': []}, 3: {'order': 3, 'module': 'CODE', 'prompt': \"/force CODE  3 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, create t-SNE plots of the data in 2D and 3D space and color the visualizations based on the 'recipe' column to distinguish between 'source', 'target', and different TF recipes.\\n\\n**\", 'description': \" 3 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, create t-SNE plots of the data in 2D and 3D space and color the visualizations based on the 'recipe' column to distinguish between 'source', 'target', and different TF recipes.\\n\\n**\", 'output': []}, 4: {'order': 4, 'module': 'CODE', 'prompt': '/force CODE  4 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, generate separate UMAP visualizations highlighting the presence of each TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) by coloring the points based on the True/False values in the respective columns.\\n\\n**', 'description': ' 4 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, generate separate UMAP visualizations highlighting the presence of each TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) by coloring the points based on the True/False values in the respective columns.\\n\\n**', 'output': []}, 5: {'order': 5, 'module': 'CODE', 'prompt': '/force CODE  5 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, generate separate t-SNE visualizations highlighting the presence of each TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) by coloring the points based on the True/False values in the respective columns.\\n\\nThe removed ', 'description': ' 5 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, generate separate t-SNE visualizations highlighting the presence of each TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) by coloring the points based on the True/False values in the respective columns.\\n\\nThe removed ', 'output': []}}\n",
      "2024-08-03 00:43:24,540 - INFO - IP=2\n",
      "2024-08-03 00:43:24,540 - INFO - type(IP)=<class 'int'>\n",
      "2024-08-03 00:43:24,542 - INFO - \n",
      "\n",
      "PLANNER:\n",
      "\n",
      "{'module': 'CODE', 'steps': [{'llm': \"client=<openai.resources.chat.completions.Completions object at 0x1480cc33e8d0> async_client=<openai.resources.chat.completions.AsyncCompletions object at 0x1480cc324e50> model_name='gpt-4o' temperature=0.0 openai_api_key=SecretStr('**********') openai_proxy=''\", 'memory': 'None', 'prompt': 'input_variables=[\\'user_query\\'] template=\"You must select which code to run to help a user.\\\\n\\\\n**Available Scripts**\\\\nScript Name: scanpy_brad, \\\\t Description: This script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\\\\nScript Name: models_brad, \\\\t Description: This script executes a series of custom commands to build a model for predicting the contribution to reprogramming obtained by the inclusion or exclusion of individual TFs using an AnnData object or single cell datasets loaded from a .h5ad file.\\\\n\\\\n\\\\n**User Query**\\\\n{user_query}\\\\n\\\\n**Task**\\\\nBased on the user\\'s query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\\\\n\\\\n**Response Template**\\\\nSCRIPT: <selected script>\\\\nREASON: <reasoning why the script fits the user prompt>\\\\n\"', 'input': \"2 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, create UMAP plots of the data in 2D and 3D space and color the visualizations based on the 'recipe' column to distinguish between 'source', 'target', and different TF recipes.\\n\\n**\", 'output': 'content=\"SCRIPT: scanpy_brad\\\\nREASON: The user\\'s query involves creating UMAP plots of an AnnData object and coloring the visualizations based on a specific column (\\'recipe\\'). The `scanpy_brad` script is designed to manipulate, visualize, and explore single cell datasets using libraries like scanpy, seaborn, and matplotlib.pyplot, which are well-suited for creating UMAP plots. The `models_brad` script, on the other hand, is focused on building predictive models and is not appropriate for the visualization task described by the user.\" response_metadata={\\'token_usage\\': {\\'completion_tokens\\': 112, \\'prompt_tokens\\': 338, \\'total_tokens\\': 450}, \\'model_name\\': \\'gpt-4o-2024-05-13\\', \\'system_fingerprint\\': \\'fp_3cd8b62c3b\\', \\'finish_reason\\': \\'stop\\', \\'logprobs\\': None} id=\\'run-c4a469b1-97b9-4216-bae8-17aad8dd39cb-0\\' usage_metadata={\\'input_tokens\\': 338, \\'output_tokens\\': 112, \\'total_tokens\\': 450}', 'parsedOutput': {'scriptName': '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py', 'scriptType': 'python', 'scriptPath': '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts'}, 'purpose': 'Select which code to run'}, {'llm': \"client=<openai.resources.chat.completions.Completions object at 0x1480cc33e8d0> async_client=<openai.resources.chat.completions.AsyncCompletions object at 0x1480cc324e50> model_name='gpt-4o' temperature=0.0 openai_api_key=SecretStr('**********') openai_proxy=''\", 'memory': 'None', 'prompt': 'input_variables=[\\'history\\', \\'input\\'] template=\\'Current conversation:\\\\n{history}\\\\n\\\\n**PYTHON SCRIPT**\\\\nYou must run this python script:\\\\n/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\\\n\\\\n**PYTHON SCRIPT DOCUMENTATION**:\\\\nThis is the doc string of this python script:\\\\n\"\"\"\\\\nThis script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\\\\n\\\\nArguments (four arguments):\\\\n    1. output directory: chatstatus[\\\\\\'output-directory\\\\\\']\\\\n    2. output file: <name of output file>\\\\n    3. input file: <file created in previous step>\\\\n    4. scanpy commands: a list of scanpy commands to be executed on the AnnData object (provided as a single string with commands separated by a delimiter, e.g., \\\\\\';\\\\\\')\\\\n\\\\nBased on the arguments, the input file will be loaded, then your commands will be executed, and finally, the output or resulting ann data object will be saved to the corrrectou output file and directory. Your code is not responsible for loading the .h5ad object, that will happen automatically, and when loaded, the object will be called adata. Your scanpy commands can operate directly on the adata object that will be loaded for you.\\\\n\\\\nAdditionally, the following imports are already provided for you and can be used in your code:\\\\n```\\\\nimport scanpy as sc\\\\nimport seaborn as sns\\\\nimport matplotlib.pyplot as plt\\\\nimport os\\\\nimport numpy as np\\\\nimport pandas as pd\\\\nfrom scipy.spatial.distance import cdist\\\\n```\\\\n\\\\n**Usage**\\\\nBRAD Line:\\\\n```\\\\nsubprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\\\\\'output-directory\\\\\\'], <output file>, <input file>, \"<scanpy commands>\"], capture_output=True, text=True)\\\\n```\\\\n\\\\n**Examples**\\\\nUse the below examples to help generate your code.\\\\n\\\\n*Example 1*\\\\nUser Prompt: Run scanpy preprocessing and UMAP visualization on XXX.h5ad and save the UMAP plot\\\\nResponse Code:\\\\n```\\\\nresponse = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\\\\\'output-directory\\\\\\'], \"XXX-modified.h5ad\", \"<path/to/data>/XXX.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save=\\\\\\'umap.png\\\\\\')\"], capture_output=True, text=True)\\\\n```\\\\nExplination: the adata object will be loaded in memory already. The command \"sc.pp.neighbors(adata)\" will preprocess the data, then the command \"sc.tl.umap(adata)\" will perform UMAP, and finally the command \"sc.pl.umap(adata, \\\\\\'umap.png\\\\\\')\" will save the UMAP to a well named file.\\\\n\\\\n*Example 2*\\\\nUser Prompt: Perform PCA and clustering on the dataset YYY.h5ad and save the PCA plot\\\\nResponse Code:\\\\n```\\\\nresponse = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\\\\\'output-directory\\\\\\'], \"YYY-modified.h5ad\", \"<path/to/data>/YYY.h5ad\", \"sc.pp.pca(adata); sc.tl.leiden(adata); sc.pl.pca(adata, save=\\\\\\'pca.png\\\\\\')\"], capture_output=True, text=True)\\\\n```\\\\nExplination: the adata object will be loaded in memory already. The command \"sc.pp.pca(adata)\" will preprocess the data, then the command \"sc.tl.leiden(adata)\" will perform the leiden algorithm, and finally the command \"sc.pl.pca(adata, save=\\\\\\'pca.png\\\\\\')\" will save the PCA to a well named file.\\\\n\\\\n*Example 3*\\\\nUser Prompt: compute the distance between all cells that are not source or target to all source and target cells.\\\\nReponse Code:\\\\n```\\\\nadata_X = adata.X\\\\nmask = ~adata.obs[\\\\\\'recipe\\\\\\'].isin([\\\\\\'source\\\\\\', \\\\\\'target\\\\\\'])\\\\nrecipe_embeddings = adata[mask]\\\\nsource_embeddings = adata[adata.obs[\\\\\\'recipe\\\\\\'] == \\\\\\'source\\\\\\']\\\\ntarget_embeddings = adata[adata.obs[\\\\\\'recipe\\\\\\'] == \\\\\\'target\\\\\\']\\\\nrecipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X)\\\\nrecipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X)\\\\nadata_dist = sc.AnnData(X=np.concatenate((recipe_distances_source, recipe_distances_target), axis=1), var=pd.concat((source_embeddings.obs, target_embeddings.obs)), obs=recipe_embeddings.obs)\\\\nadata_dist.write(<file name>.h5ad)\\\\n```\\\\n\\\\n**OUTPUT FILE NAME INSTRUCTIONS**\\\\n1. Output path should be chatstatus[\\\\\\'output-directory\\\\\\']\\\\n2. Output file name should be `<descriptive name>.h5ad`\\\\n\"\"\"\\\\n\\\\n\\\\n**CALL PYTHON SCRIPTS FROM PYTHON**:\\\\nUse the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\\\\n\\\\nTo call a Python script `example_script.py` which has no arguments:\\\\n\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\']], capture_output=True, text=True)\\\\n```\\\\n\\\\nTo call a Python script `example_script.py` with one argument:\\\\n\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\'], \\\\\\'arg1\\\\\\'], capture_output=True, text=True)\\\\n```\\\\n\\\\nTo call a Python script `example_script.py` with two arguments:\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\'], \\\\\\'arg1\\\\\\', \\\\\\'arg2\\\\\\'], capture_output=True, text=True)\\\\n```\\\\n\\\\nNote that chatstatus[\\\\\\'output-directory\\\\\\'] is ALWAYS passed as the first argument.\\\\n\\\\nThe following files were previously created by BRAD and could be used as input to a function if necessary:\\\\nS1-HSC-reprogrammed-DE-summary.h5ad\\\\nlog.json\\\\n\\\\nQuery:{input}\\\\n\\\\n**PYTHON SCRIPT**\\\\nPlease use the correct script name and path. You must run this python script:\\\\n/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\\\n\\\\n**INSTRUCTIONS**\\\\n1. Given the user query and the documentation, identify each of the arguments found in the user\\\\\\'s query that should be passed to the Python script.\\\\n2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\\\\n3. The last line of your response should say \"Execute: <Python code to execute>\"\\\\n4. Do not include any extra words or characters. Format the response/output as:\\\\n    Arguments: \\\\n    Python Code Explanation: <2 sentences maximum>\\\\n    Execute: <your code here>\\\\n\\\\n**IMPORTANT**\\\\nThe code to execute from your response must be formatted as:\\\\n    Execute: subprocess.call([sys.executable, \\\\\\'<path to python script>\\\\\\', \\\\\\'<argument 1>\\\\\\', \\\\\\'<argument 2>\\\\\\', ..., \\\\\\'<argument n>\\\\\\'], capture_output=True, text=True))\\\\nThis output should be exactly one line and no longer. Stop the response after this line.\\\\n\\'', 'input': \"2 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, create UMAP plots of the data in 2D and 3D space and color the visualizations based on the 'recipe' column to distinguish between 'source', 'target', and different TF recipes.\\n\\n**\", 'output': 'Arguments:\\n1. output directory: chatstatus[\\'output-directory\\']\\n2. output file: \"HSC-reprogrammed-UMAP.h5ad\"\\n3. input file: \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\"\\n4. scanpy commands: \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, color=\\'recipe\\', save=\\'umap_2d.png\\'); sc.pl.umap(adata, color=\\'recipe\\', save=\\'umap_3d.png\\', projection=\\'3d\\')\"\\n\\nPython Code Explanation: This code will preprocess the data, perform UMAP in both 2D and 3D, and save the visualizations colored by the \\'recipe\\' column.\\n\\nExecute: subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"HSC-reprogrammed-UMAP.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, color=\\'recipe\\', save=\\'umap_2d.png\\'); sc.pl.umap(adata, color=\\'recipe\\', save=\\'umap_3d.png\\', projection=\\'3d\\')\"], capture_output=True, text=True)', 'parsedOutput': {'code': 'subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"HSC-reprogrammed-UMAP.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, color=\\'recipe\\', save=\\'umap_2d.png\\'); sc.pl.umap(adata, color=\\'recipe\\', save=\\'umap_3d.png\\', projection=\\'3d\\')\"], capture_output=True, text=True)'}, 'purpose': 'Format function call'}, {'func': 'pythonCaller.execute_python_code', 'code': 'response = subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"HSC-reprogrammed-UMAP.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, color=\\'recipe\\', save=\\'umap_2d.png\\'); sc.pl.umap(adata, color=\\'recipe\\', save=\\'umap_3d.png\\', projection=\\'3d\\')\"], capture_output=True, text=True)', 'purpose': 'execute python code'}, {'func': 'utils.makeNamesConsistent', 'files': [{'old-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_00-39-53/umapumap_2d.png', 'new-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_00-39-53/S2-umapumap_2d.png'}]}]}\n",
      "2024-08-03 00:43:24,547 - INFO - current llm calls=5\n",
      "2024-08-03 00:43:24,549 - INFO - CODE\n",
      "2024-08-03 00:43:24,549 - INFO - CODER\n",
      "2024-08-03 00:43:24,550 - INFO - input_variables=['user_query'] template=\"You must select which code to run to help a user.\\n\\n**Available Scripts**\\nScript Name: scanpy_brad, \\t Description: This script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\\nScript Name: models_brad, \\t Description: This script executes a series of custom commands to build a model for predicting the contribution to reprogramming obtained by the inclusion or exclusion of individual TFs using an AnnData object or single cell datasets loaded from a .h5ad file.\\n\\n\\n**User Query**\\n{user_query}\\n\\n**Task**\\nBased on the user's query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\\n\\n**Response Template**\\nSCRIPT: <selected script>\\nREASON: <reasoning why the script fits the user prompt>\\n\"\n",
      "2024-08-03 00:43:24,551 - INFO - FIRST LLM CALL\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "==================================================\n",
      "BRAD >> 3: \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 00:43:26,235 - INFO - HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2024-08-03 00:43:26,236 - INFO - SCRIPT: scanpy_brad\n",
      "REASON: The user's query involves creating t-SNE plots of an AnnData object and coloring the visualizations based on a specific column. The `scanpy_brad` script is designed to manipulate, visualize, and explore single-cell datasets using libraries like scanpy, seaborn, and matplotlib.pyplot, which are well-suited for creating t-SNE plots. The `models_brad` script, on the other hand, is focused on building predictive models and is not appropriate for the visualization task described.\n",
      "2024-08-03 00:43:26,237 - INFO - scriptPath=/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts\n",
      "2024-08-03 00:43:26,237 - INFO - ALL SCRIPTS FOUND. BUILDING TEMPLATE\n",
      "2024-08-03 00:43:26,238 - INFO - input_variables=['history', 'input'] template='Current conversation:\\n{history}\\n\\n**PYTHON SCRIPT**\\nYou must run this python script:\\n/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\n\\n**PYTHON SCRIPT DOCUMENTATION**:\\nThis is the doc string of this python script:\\n\"\"\"\\nThis script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\\n\\nArguments (four arguments):\\n    1. output directory: chatstatus[\\'output-directory\\']\\n    2. output file: <name of output file>\\n    3. input file: <file created in previous step>\\n    4. scanpy commands: a list of scanpy commands to be executed on the AnnData object (provided as a single string with commands separated by a delimiter, e.g., \\';\\')\\n\\nBased on the arguments, the input file will be loaded, then your commands will be executed, and finally, the output or resulting ann data object will be saved to the corrrectou output file and directory. Your code is not responsible for loading the .h5ad object, that will happen automatically, and when loaded, the object will be called adata. Your scanpy commands can operate directly on the adata object that will be loaded for you.\\n\\nAdditionally, the following imports are already provided for you and can be used in your code:\\n```\\nimport scanpy as sc\\nimport seaborn as sns\\nimport matplotlib.pyplot as plt\\nimport os\\nimport numpy as np\\nimport pandas as pd\\nfrom scipy.spatial.distance import cdist\\n```\\n\\n**Usage**\\nBRAD Line:\\n```\\nsubprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\'output-directory\\'], <output file>, <input file>, \"<scanpy commands>\"], capture_output=True, text=True)\\n```\\n\\n**Examples**\\nUse the below examples to help generate your code.\\n\\n*Example 1*\\nUser Prompt: Run scanpy preprocessing and UMAP visualization on XXX.h5ad and save the UMAP plot\\nResponse Code:\\n```\\nresponse = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"XXX-modified.h5ad\", \"<path/to/data>/XXX.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save=\\'umap.png\\')\"], capture_output=True, text=True)\\n```\\nExplination: the adata object will be loaded in memory already. The command \"sc.pp.neighbors(adata)\" will preprocess the data, then the command \"sc.tl.umap(adata)\" will perform UMAP, and finally the command \"sc.pl.umap(adata, \\'umap.png\\')\" will save the UMAP to a well named file.\\n\\n*Example 2*\\nUser Prompt: Perform PCA and clustering on the dataset YYY.h5ad and save the PCA plot\\nResponse Code:\\n```\\nresponse = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"YYY-modified.h5ad\", \"<path/to/data>/YYY.h5ad\", \"sc.pp.pca(adata); sc.tl.leiden(adata); sc.pl.pca(adata, save=\\'pca.png\\')\"], capture_output=True, text=True)\\n```\\nExplination: the adata object will be loaded in memory already. The command \"sc.pp.pca(adata)\" will preprocess the data, then the command \"sc.tl.leiden(adata)\" will perform the leiden algorithm, and finally the command \"sc.pl.pca(adata, save=\\'pca.png\\')\" will save the PCA to a well named file.\\n\\n*Example 3*\\nUser Prompt: compute the distance between all cells that are not source or target to all source and target cells.\\nReponse Code:\\n```\\nadata_X = adata.X\\nmask = ~adata.obs[\\'recipe\\'].isin([\\'source\\', \\'target\\'])\\nrecipe_embeddings = adata[mask]\\nsource_embeddings = adata[adata.obs[\\'recipe\\'] == \\'source\\']\\ntarget_embeddings = adata[adata.obs[\\'recipe\\'] == \\'target\\']\\nrecipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X)\\nrecipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X)\\nadata_dist = sc.AnnData(X=np.concatenate((recipe_distances_source, recipe_distances_target), axis=1), var=pd.concat((source_embeddings.obs, target_embeddings.obs)), obs=recipe_embeddings.obs)\\nadata_dist.write(<file name>.h5ad)\\n```\\n\\n**OUTPUT FILE NAME INSTRUCTIONS**\\n1. Output path should be chatstatus[\\'output-directory\\']\\n2. Output file name should be `<descriptive name>.h5ad`\\n\"\"\"\\n\\n\\n**CALL PYTHON SCRIPTS FROM PYTHON**:\\nUse the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\\n\\nTo call a Python script `example_script.py` which has no arguments:\\n\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\']], capture_output=True, text=True)\\n```\\n\\nTo call a Python script `example_script.py` with one argument:\\n\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\'], \\'arg1\\'], capture_output=True, text=True)\\n```\\n\\nTo call a Python script `example_script.py` with two arguments:\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\'], \\'arg1\\', \\'arg2\\'], capture_output=True, text=True)\\n```\\n\\nNote that chatstatus[\\'output-directory\\'] is ALWAYS passed as the first argument.\\n\\nThe following files were previously created by BRAD and could be used as input to a function if necessary:\\nS1-HSC-reprogrammed-DE-summary.h5ad\\nlog.json\\nS2-umapumap_2d.png\\n\\nQuery:{input}\\n\\n**PYTHON SCRIPT**\\nPlease use the correct script name and path. You must run this python script:\\n/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\n\\n**INSTRUCTIONS**\\n1. Given the user query and the documentation, identify each of the arguments found in the user\\'s query that should be passed to the Python script.\\n2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\\n3. The last line of your response should say \"Execute: <Python code to execute>\"\\n4. Do not include any extra words or characters. Format the response/output as:\\n    Arguments: \\n    Python Code Explanation: <2 sentences maximum>\\n    Execute: <your code here>\\n\\n**IMPORTANT**\\nThe code to execute from your response must be formatted as:\\n    Execute: subprocess.call([sys.executable, \\'<path to python script>\\', \\'<argument 1>\\', \\'<argument 2>\\', ..., \\'<argument n>\\'], capture_output=True, text=True))\\nThis output should be exactly one line and no longer. Stop the response after this line.\\n'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation:\n",
      "\n",
      "\n",
      "**PYTHON SCRIPT**\n",
      "You must run this python script:\n",
      "/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\n",
      "\n",
      "**PYTHON SCRIPT DOCUMENTATION**:\n",
      "This is the doc string of this python script:\n",
      "\"\"\"\n",
      "This script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\n",
      "\n",
      "Arguments (four arguments):\n",
      "    1. output directory: chatstatus['output-directory']\n",
      "    2. output file: <name of output file>\n",
      "    3. input file: <file created in previous step>\n",
      "    4. scanpy commands: a list of scanpy commands to be executed on the AnnData object (provided as a single string with commands separated by a delimiter, e.g., ';')\n",
      "\n",
      "Based on the arguments, the input file will be loaded, then your commands will be executed, and finally, the output or resulting ann data object will be saved to the corrrectou output file and directory. Your code is not responsible for loading the .h5ad object, that will happen automatically, and when loaded, the object will be called adata. Your scanpy commands can operate directly on the adata object that will be loaded for you.\n",
      "\n",
      "Additionally, the following imports are already provided for you and can be used in your code:\n",
      "```\n",
      "import scanpy as sc\n",
      "import seaborn as sns\n",
      "import matplotlib.pyplot as plt\n",
      "import os\n",
      "import numpy as np\n",
      "import pandas as pd\n",
      "from scipy.spatial.distance import cdist\n",
      "```\n",
      "\n",
      "**Usage**\n",
      "BRAD Line:\n",
      "```\n",
      "subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus['output-directory'], <output file>, <input file>, \"<scanpy commands>\"], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "**Examples**\n",
      "Use the below examples to help generate your code.\n",
      "\n",
      "*Example 1*\n",
      "User Prompt: Run scanpy preprocessing and UMAP visualization on XXX.h5ad and save the UMAP plot\n",
      "Response Code:\n",
      "```\n",
      "response = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus['output-directory'], \"XXX-modified.h5ad\", \"<path/to/data>/XXX.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save='umap.png')\"], capture_output=True, text=True)\n",
      "```\n",
      "Explination: the adata object will be loaded in memory already. The command \"sc.pp.neighbors(adata)\" will preprocess the data, then the command \"sc.tl.umap(adata)\" will perform UMAP, and finally the command \"sc.pl.umap(adata, 'umap.png')\" will save the UMAP to a well named file.\n",
      "\n",
      "*Example 2*\n",
      "User Prompt: Perform PCA and clustering on the dataset YYY.h5ad and save the PCA plot\n",
      "Response Code:\n",
      "```\n",
      "response = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus['output-directory'], \"YYY-modified.h5ad\", \"<path/to/data>/YYY.h5ad\", \"sc.pp.pca(adata); sc.tl.leiden(adata); sc.pl.pca(adata, save='pca.png')\"], capture_output=True, text=True)\n",
      "```\n",
      "Explination: the adata object will be loaded in memory already. The command \"sc.pp.pca(adata)\" will preprocess the data, then the command \"sc.tl.leiden(adata)\" will perform the leiden algorithm, and finally the command \"sc.pl.pca(adata, save='pca.png')\" will save the PCA to a well named file.\n",
      "\n",
      "*Example 3*\n",
      "User Prompt: compute the distance between all cells that are not source or target to all source and target cells.\n",
      "Reponse Code:\n",
      "```\n",
      "adata_X = adata.X\n",
      "mask = ~adata.obs['recipe'].isin(['source', 'target'])\n",
      "recipe_embeddings = adata[mask]\n",
      "source_embeddings = adata[adata.obs['recipe'] == 'source']\n",
      "target_embeddings = adata[adata.obs['recipe'] == 'target']\n",
      "recipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X)\n",
      "recipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X)\n",
      "adata_dist = sc.AnnData(X=np.concatenate((recipe_distances_source, recipe_distances_target), axis=1), var=pd.concat((source_embeddings.obs, target_embeddings.obs)), obs=recipe_embeddings.obs)\n",
      "adata_dist.write(<file name>.h5ad)\n",
      "```\n",
      "\n",
      "**OUTPUT FILE NAME INSTRUCTIONS**\n",
      "1. Output path should be chatstatus['output-directory']\n",
      "2. Output file name should be `<descriptive name>.h5ad`\n",
      "\"\"\"\n",
      "\n",
      "\n",
      "**CALL PYTHON SCRIPTS FROM PYTHON**:\n",
      "Use the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\n",
      "\n",
      "To call a Python script `example_script.py` which has no arguments:\n",
      "\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory']], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with one argument:\n",
      "\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory'], 'arg1'], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with two arguments:\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory'], 'arg1', 'arg2'], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "Note that chatstatus['output-directory'] is ALWAYS passed as the first argument.\n",
      "\n",
      "The following files were previously created by BRAD and could be used as input to a function if necessary:\n",
      "S1-HSC-reprogrammed-DE-summary.h5ad\n",
      "log.json\n",
      "S2-umapumap_2d.png\n",
      "\n",
      "Query:3 (CODE)**:\n",
      "Prompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, create t-SNE plots of the data in 2D and 3D space and color the visualizations based on the 'recipe' column to distinguish between 'source', 'target', and different TF recipes.\n",
      "\n",
      "**\n",
      "\n",
      "**PYTHON SCRIPT**\n",
      "Please use the correct script name and path. You must run this python script:\n",
      "/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\n",
      "\n",
      "**INSTRUCTIONS**\n",
      "1. Given the user query and the documentation, identify each of the arguments found in the user's query that should be passed to the Python script.\n",
      "2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\n",
      "3. The last line of your response should say \"Execute: <Python code to execute>\"\n",
      "4. Do not include any extra words or characters. Format the response/output as:\n",
      "    Arguments: \n",
      "    Python Code Explanation: <2 sentences maximum>\n",
      "    Execute: <your code here>\n",
      "\n",
      "**IMPORTANT**\n",
      "The code to execute from your response must be formatted as:\n",
      "    Execute: subprocess.call([sys.executable, '<path to python script>', '<argument 1>', '<argument 2>', ..., '<argument n>'], capture_output=True, text=True))\n",
      "This output should be exactly one line and no longer. Stop the response after this line.\n",
      "\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 00:43:29,805 - INFO - HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2024-08-03 00:43:29,806 - INFO - LLM OUTPUT PARSER\n",
      "2024-08-03 00:43:29,807 - INFO - Arguments:\n",
      "1. Output directory: chatstatus['output-directory']\n",
      "2. Output file: \"HSC-reprogrammed-DE-tsne.h5ad\"\n",
      "3. Input file: \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\"\n",
      "4. Scanpy commands: \"sc.tl.tsne(adata, n_pcs=50); sc.pl.tsne(adata, color='recipe', save='tsne_2d.png'); sc.pl.tsne(adata, color='recipe', projection='3d', save='tsne_3d.png')\"\n",
      "\n",
      "Python Code Explanation: This code will load the AnnData object, perform t-SNE in 2D and 3D, color the plots based on the 'recipe' column, and save the visualizations.\n",
      "\n",
      "Execute: subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-DE-tsne.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.tl.tsne(adata, n_pcs=50); sc.pl.tsne(adata, color='recipe', save='tsne_2d.png'); sc.pl.tsne(adata, color='recipe', projection='3d', save='tsne_3d.png')\"], capture_output=True, text=True)\n",
      "2024-08-03 00:43:29,807 - INFO - ['Arguments:\\n1. Output directory: chatstatus[\\'output-directory\\']\\n2. Output file: \"HSC-reprogrammed-DE-tsne.h5ad\"\\n3. Input file: \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\"\\n4. Scanpy commands: \"sc.tl.tsne(adata, n_pcs=50); sc.pl.tsne(adata, color=\\'recipe\\', save=\\'tsne_2d.png\\'); sc.pl.tsne(adata, color=\\'recipe\\', projection=\\'3d\\', save=\\'tsne_3d.png\\')\"\\n\\nPython Code Explanation: This code will load the AnnData object, perform t-SNE in 2D and 3D, color the plots based on the \\'recipe\\' column, and save the visualizations.\\n\\n', ' subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"HSC-reprogrammed-DE-tsne.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.tl.tsne(adata, n_pcs=50); sc.pl.tsne(adata, color=\\'recipe\\', save=\\'tsne_2d.png\\'); sc.pl.tsne(adata, color=\\'recipe\\', projection=\\'3d\\', save=\\'tsne_3d.png\\')\"], capture_output=True, text=True)']\n",
      "2024-08-03 00:43:29,807 - INFO - Stripped funciton call\n",
      "2024-08-03 00:43:29,807 - INFO - subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-DE-tsne.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.tl.tsne(adata, n_pcs=50); sc.pl.tsne(adata, color='recipe', save='tsne_2d.png'); sc.pl.tsne(adata, color='recipe', projection='3d', save='tsne_3d.png')\"], capture_output=True, text=True)\n",
      "2024-08-03 00:43:29,807 - INFO - Cleaned up function call:\n",
      "\n",
      "2024-08-03 00:43:29,808 - INFO - subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-DE-tsne.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.tl.tsne(adata, n_pcs=50); sc.pl.tsne(adata, color='recipe', save='tsne_2d.png'); sc.pl.tsne(adata, color='recipe', projection='3d', save='tsne_3d.png')\"], capture_output=True, text=True)\n",
      "2024-08-03 00:43:29,808 - INFO - \n",
      "\n",
      "2024-08-03 00:43:29,809 - INFO - EVAL\n",
      "2024-08-03 00:43:29,809 - INFO - subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-DE-tsne.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.tl.tsne(adata, n_pcs=50); sc.pl.tsne(adata, color='recipe', save='tsne_2d.png'); sc.pl.tsne(adata, color='recipe', projection='3d', save='tsne_3d.png')\"], capture_output=True, text=True)\n",
      "2024-08-03 00:43:29,809 - INFO - END EVAL CHECK\n",
      "2024-08-03 00:43:29,810 - INFO - Finalized PYTHON Call:\n",
      "2024-08-03 00:43:29,810 - INFO -    \n",
      "2024-08-03 00:43:29,810 - INFO - response = subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-DE-tsne.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.tl.tsne(adata, n_pcs=50); sc.pl.tsne(adata, color='recipe', save='tsne_2d.png'); sc.pl.tsne(adata, color='recipe', projection='3d', save='tsne_3d.png')\"], capture_output=True, text=True)\n",
      "2024-08-03 00:43:29,810 - INFO -    \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 00:43:55,189 - INFO - Debug: PYTHON code executed successfully.\n",
      "2024-08-03 00:43:55,190 - INFO - Code Execution Output:\n",
      "2024-08-03 00:43:55,190 - INFO - CompletedProcess(args=['/sw/pkgs/arc/python3.11-anaconda/2024.02-1/bin/python', '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py', '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_00-39-53', 'HSC-reprogrammed-DE-tsne.h5ad', '/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad', \"sc.tl.tsne(adata, n_pcs=50); sc.pl.tsne(adata, color='recipe', save='tsne_2d.png'); sc.pl.tsne(adata, color='recipe', projection='3d', save='tsne_3d.png')\"], returncode=1, stdout=\"adata loaded from /nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\\nadata.shape=(6375, 512)\\nadata.obs.columns=Index(['recipe', 'filename', 'REL', 'STAT5A', 'FOS', 'ETV6', 'source', 'MEF2C',\\n       'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B', 'target'],\\n      dtype='object')\\nadata.obs.head()=   recipe                                           filename  ...  GFI1B  target\\n0  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n1  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n2  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n3  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n4  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n\\n[5 rows x 14 columns]\\nadata.var.head()=Empty DataFrame\\nColumns: []\\nIndex: [0, 1, 2, 3, 4]\\n****************************\\n      EXECUTE COMMANDS      \\n****************************\\nsc.tl.tsne(adata, n_pcs=50)\\nsc.pl.tsne(adata, color='recipe', save='tsne_2d.png')\\nFigure(1200x1200)\\nsc.pl.tsne(adata, color='recipe', projection='3d', save='tsne_3d.png')\\n\", stderr='/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\\n  utils.warn_names_duplicates(\"obs\")\\n/home/jpic/.local/lib/python3.11/site-packages/scanpy/tools/_utils.py:41: UserWarning: Youâ€™re trying to run this on 512 dimensions of `.X`, if you really want this, set `use_rep=\\'X\\'`.\\n         Falling back to preprocessing with `sc.pp.pca` and default params.\\n  warnings.warn(\\nWARNING: saving figure to file /home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_00-39-53/tsnetsne_2d.png\\nTraceback (most recent call last):\\n  File \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", line 117, in <module>\\n    main(args.output_directory, args.output_file, args.input_file, args.scanpy_commands)\\n  File \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", line 103, in main\\n    exec(command)\\n  File \"<string>\", line 1, in <module>\\n  File \"/home/jpic/.local/lib/python3.11/site-packages/scanpy/plotting/_tools/scatterplots.py\", line 731, in tsne\\n    return embedding(adata, \"tsne\", **kwargs)\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \"/home/jpic/.local/lib/python3.11/site-packages/scanpy/plotting/_tools/scatterplots.py\", line 308, in embedding\\n    coords = basis_values[:, dims][order, :]\\n             ~~~~~~~~~~~~^^^^^^^^^\\nIndexError: index 2 is out of bounds for axis 1 with size 2\\n')\n",
      "2024-08-03 00:43:55,191 - INFO - Debug: PYTHON code output saved to output.\n",
      "2024-08-03 00:43:55,191 - INFO - \n",
      "\n",
      "\n",
      "route\n",
      "\n",
      "\n",
      "\n",
      "2024-08-03 00:43:55,191 - INFO - CODE\n",
      "2024-08-03 00:43:55,193 - INFO - Finding Stage Number of Pipeline\n",
      "2024-08-03 00:43:55,193 - INFO - {0: {'order': 0, 'module': 'PLANNER', 'prompt': None, 'description': 'This step designed the plan. It is placed in the queue because we needed a place holder for 0 indexed lists.', 'output': []}, 1: {'order': 1, 'module': 'CODE', 'prompt': '/force CODE  1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and create a basic summary of the data.\\n\\n**', 'description': ' 1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and create a basic summary of the data.\\n\\n**', 'output': ['/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_00-39-53/S1-HSC-reprogrammed-DE-summary.h5ad']}, 2: {'order': 2, 'module': 'CODE', 'prompt': \"/force CODE  2 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, create UMAP plots of the data in 2D and 3D space and color the visualizations based on the 'recipe' column to distinguish between 'source', 'target', and different TF recipes.\\n\\n**\", 'description': \" 2 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, create UMAP plots of the data in 2D and 3D space and color the visualizations based on the 'recipe' column to distinguish between 'source', 'target', and different TF recipes.\\n\\n**\", 'output': ['/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_00-39-53/S2-umapumap_2d.png']}, 3: {'order': 3, 'module': 'CODE', 'prompt': \"/force CODE  3 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, create t-SNE plots of the data in 2D and 3D space and color the visualizations based on the 'recipe' column to distinguish between 'source', 'target', and different TF recipes.\\n\\n**\", 'description': \" 3 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, create t-SNE plots of the data in 2D and 3D space and color the visualizations based on the 'recipe' column to distinguish between 'source', 'target', and different TF recipes.\\n\\n**\", 'output': []}, 4: {'order': 4, 'module': 'CODE', 'prompt': '/force CODE  4 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, generate separate UMAP visualizations highlighting the presence of each TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) by coloring the points based on the True/False values in the respective columns.\\n\\n**', 'description': ' 4 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, generate separate UMAP visualizations highlighting the presence of each TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) by coloring the points based on the True/False values in the respective columns.\\n\\n**', 'output': []}, 5: {'order': 5, 'module': 'CODE', 'prompt': '/force CODE  5 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, generate separate t-SNE visualizations highlighting the presence of each TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) by coloring the points based on the True/False values in the respective columns.\\n\\nThe removed ', 'description': ' 5 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, generate separate t-SNE visualizations highlighting the presence of each TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) by coloring the points based on the True/False values in the respective columns.\\n\\nThe removed ', 'output': []}}\n",
      "2024-08-03 00:43:55,193 - INFO - IP=3\n",
      "2024-08-03 00:43:55,194 - INFO - type(IP)=<class 'int'>\n",
      "2024-08-03 00:43:55,197 - INFO - \n",
      "\n",
      "PLANNER:\n",
      "\n",
      "{'module': 'CODE', 'steps': [{'llm': \"client=<openai.resources.chat.completions.Completions object at 0x1480cc33e8d0> async_client=<openai.resources.chat.completions.AsyncCompletions object at 0x1480cc324e50> model_name='gpt-4o' temperature=0.0 openai_api_key=SecretStr('**********') openai_proxy=''\", 'memory': 'None', 'prompt': 'input_variables=[\\'user_query\\'] template=\"You must select which code to run to help a user.\\\\n\\\\n**Available Scripts**\\\\nScript Name: scanpy_brad, \\\\t Description: This script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\\\\nScript Name: models_brad, \\\\t Description: This script executes a series of custom commands to build a model for predicting the contribution to reprogramming obtained by the inclusion or exclusion of individual TFs using an AnnData object or single cell datasets loaded from a .h5ad file.\\\\n\\\\n\\\\n**User Query**\\\\n{user_query}\\\\n\\\\n**Task**\\\\nBased on the user\\'s query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\\\\n\\\\n**Response Template**\\\\nSCRIPT: <selected script>\\\\nREASON: <reasoning why the script fits the user prompt>\\\\n\"', 'input': \"3 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, create t-SNE plots of the data in 2D and 3D space and color the visualizations based on the 'recipe' column to distinguish between 'source', 'target', and different TF recipes.\\n\\n**\", 'output': 'content=\"SCRIPT: scanpy_brad\\\\nREASON: The user\\'s query involves creating t-SNE plots of an AnnData object and coloring the visualizations based on a specific column. The `scanpy_brad` script is designed to manipulate, visualize, and explore single-cell datasets using libraries like scanpy, seaborn, and matplotlib.pyplot, which are well-suited for creating t-SNE plots. The `models_brad` script, on the other hand, is focused on building predictive models and is not appropriate for the visualization task described.\" response_metadata={\\'token_usage\\': {\\'completion_tokens\\': 109, \\'prompt_tokens\\': 339, \\'total_tokens\\': 448}, \\'model_name\\': \\'gpt-4o-2024-05-13\\', \\'system_fingerprint\\': \\'fp_3cd8b62c3b\\', \\'finish_reason\\': \\'stop\\', \\'logprobs\\': None} id=\\'run-3ebb7c05-b242-4f20-8655-f3e757949e21-0\\' usage_metadata={\\'input_tokens\\': 339, \\'output_tokens\\': 109, \\'total_tokens\\': 448}', 'parsedOutput': {'scriptName': '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py', 'scriptType': 'python', 'scriptPath': '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts'}, 'purpose': 'Select which code to run'}, {'llm': \"client=<openai.resources.chat.completions.Completions object at 0x1480cc33e8d0> async_client=<openai.resources.chat.completions.AsyncCompletions object at 0x1480cc324e50> model_name='gpt-4o' temperature=0.0 openai_api_key=SecretStr('**********') openai_proxy=''\", 'memory': 'None', 'prompt': 'input_variables=[\\'history\\', \\'input\\'] template=\\'Current conversation:\\\\n{history}\\\\n\\\\n**PYTHON SCRIPT**\\\\nYou must run this python script:\\\\n/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\\\n\\\\n**PYTHON SCRIPT DOCUMENTATION**:\\\\nThis is the doc string of this python script:\\\\n\"\"\"\\\\nThis script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\\\\n\\\\nArguments (four arguments):\\\\n    1. output directory: chatstatus[\\\\\\'output-directory\\\\\\']\\\\n    2. output file: <name of output file>\\\\n    3. input file: <file created in previous step>\\\\n    4. scanpy commands: a list of scanpy commands to be executed on the AnnData object (provided as a single string with commands separated by a delimiter, e.g., \\\\\\';\\\\\\')\\\\n\\\\nBased on the arguments, the input file will be loaded, then your commands will be executed, and finally, the output or resulting ann data object will be saved to the corrrectou output file and directory. Your code is not responsible for loading the .h5ad object, that will happen automatically, and when loaded, the object will be called adata. Your scanpy commands can operate directly on the adata object that will be loaded for you.\\\\n\\\\nAdditionally, the following imports are already provided for you and can be used in your code:\\\\n```\\\\nimport scanpy as sc\\\\nimport seaborn as sns\\\\nimport matplotlib.pyplot as plt\\\\nimport os\\\\nimport numpy as np\\\\nimport pandas as pd\\\\nfrom scipy.spatial.distance import cdist\\\\n```\\\\n\\\\n**Usage**\\\\nBRAD Line:\\\\n```\\\\nsubprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\\\\\'output-directory\\\\\\'], <output file>, <input file>, \"<scanpy commands>\"], capture_output=True, text=True)\\\\n```\\\\n\\\\n**Examples**\\\\nUse the below examples to help generate your code.\\\\n\\\\n*Example 1*\\\\nUser Prompt: Run scanpy preprocessing and UMAP visualization on XXX.h5ad and save the UMAP plot\\\\nResponse Code:\\\\n```\\\\nresponse = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\\\\\'output-directory\\\\\\'], \"XXX-modified.h5ad\", \"<path/to/data>/XXX.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save=\\\\\\'umap.png\\\\\\')\"], capture_output=True, text=True)\\\\n```\\\\nExplination: the adata object will be loaded in memory already. The command \"sc.pp.neighbors(adata)\" will preprocess the data, then the command \"sc.tl.umap(adata)\" will perform UMAP, and finally the command \"sc.pl.umap(adata, \\\\\\'umap.png\\\\\\')\" will save the UMAP to a well named file.\\\\n\\\\n*Example 2*\\\\nUser Prompt: Perform PCA and clustering on the dataset YYY.h5ad and save the PCA plot\\\\nResponse Code:\\\\n```\\\\nresponse = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\\\\\'output-directory\\\\\\'], \"YYY-modified.h5ad\", \"<path/to/data>/YYY.h5ad\", \"sc.pp.pca(adata); sc.tl.leiden(adata); sc.pl.pca(adata, save=\\\\\\'pca.png\\\\\\')\"], capture_output=True, text=True)\\\\n```\\\\nExplination: the adata object will be loaded in memory already. The command \"sc.pp.pca(adata)\" will preprocess the data, then the command \"sc.tl.leiden(adata)\" will perform the leiden algorithm, and finally the command \"sc.pl.pca(adata, save=\\\\\\'pca.png\\\\\\')\" will save the PCA to a well named file.\\\\n\\\\n*Example 3*\\\\nUser Prompt: compute the distance between all cells that are not source or target to all source and target cells.\\\\nReponse Code:\\\\n```\\\\nadata_X = adata.X\\\\nmask = ~adata.obs[\\\\\\'recipe\\\\\\'].isin([\\\\\\'source\\\\\\', \\\\\\'target\\\\\\'])\\\\nrecipe_embeddings = adata[mask]\\\\nsource_embeddings = adata[adata.obs[\\\\\\'recipe\\\\\\'] == \\\\\\'source\\\\\\']\\\\ntarget_embeddings = adata[adata.obs[\\\\\\'recipe\\\\\\'] == \\\\\\'target\\\\\\']\\\\nrecipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X)\\\\nrecipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X)\\\\nadata_dist = sc.AnnData(X=np.concatenate((recipe_distances_source, recipe_distances_target), axis=1), var=pd.concat((source_embeddings.obs, target_embeddings.obs)), obs=recipe_embeddings.obs)\\\\nadata_dist.write(<file name>.h5ad)\\\\n```\\\\n\\\\n**OUTPUT FILE NAME INSTRUCTIONS**\\\\n1. Output path should be chatstatus[\\\\\\'output-directory\\\\\\']\\\\n2. Output file name should be `<descriptive name>.h5ad`\\\\n\"\"\"\\\\n\\\\n\\\\n**CALL PYTHON SCRIPTS FROM PYTHON**:\\\\nUse the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\\\\n\\\\nTo call a Python script `example_script.py` which has no arguments:\\\\n\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\']], capture_output=True, text=True)\\\\n```\\\\n\\\\nTo call a Python script `example_script.py` with one argument:\\\\n\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\'], \\\\\\'arg1\\\\\\'], capture_output=True, text=True)\\\\n```\\\\n\\\\nTo call a Python script `example_script.py` with two arguments:\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\'], \\\\\\'arg1\\\\\\', \\\\\\'arg2\\\\\\'], capture_output=True, text=True)\\\\n```\\\\n\\\\nNote that chatstatus[\\\\\\'output-directory\\\\\\'] is ALWAYS passed as the first argument.\\\\n\\\\nThe following files were previously created by BRAD and could be used as input to a function if necessary:\\\\nS1-HSC-reprogrammed-DE-summary.h5ad\\\\nlog.json\\\\nS2-umapumap_2d.png\\\\n\\\\nQuery:{input}\\\\n\\\\n**PYTHON SCRIPT**\\\\nPlease use the correct script name and path. You must run this python script:\\\\n/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\\\n\\\\n**INSTRUCTIONS**\\\\n1. Given the user query and the documentation, identify each of the arguments found in the user\\\\\\'s query that should be passed to the Python script.\\\\n2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\\\\n3. The last line of your response should say \"Execute: <Python code to execute>\"\\\\n4. Do not include any extra words or characters. Format the response/output as:\\\\n    Arguments: \\\\n    Python Code Explanation: <2 sentences maximum>\\\\n    Execute: <your code here>\\\\n\\\\n**IMPORTANT**\\\\nThe code to execute from your response must be formatted as:\\\\n    Execute: subprocess.call([sys.executable, \\\\\\'<path to python script>\\\\\\', \\\\\\'<argument 1>\\\\\\', \\\\\\'<argument 2>\\\\\\', ..., \\\\\\'<argument n>\\\\\\'], capture_output=True, text=True))\\\\nThis output should be exactly one line and no longer. Stop the response after this line.\\\\n\\'', 'input': \"3 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, create t-SNE plots of the data in 2D and 3D space and color the visualizations based on the 'recipe' column to distinguish between 'source', 'target', and different TF recipes.\\n\\n**\", 'output': 'Arguments:\\n1. Output directory: chatstatus[\\'output-directory\\']\\n2. Output file: \"HSC-reprogrammed-DE-tsne.h5ad\"\\n3. Input file: \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\"\\n4. Scanpy commands: \"sc.tl.tsne(adata, n_pcs=50); sc.pl.tsne(adata, color=\\'recipe\\', save=\\'tsne_2d.png\\'); sc.pl.tsne(adata, color=\\'recipe\\', projection=\\'3d\\', save=\\'tsne_3d.png\\')\"\\n\\nPython Code Explanation: This code will load the AnnData object, perform t-SNE in 2D and 3D, color the plots based on the \\'recipe\\' column, and save the visualizations.\\n\\nExecute: subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"HSC-reprogrammed-DE-tsne.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.tl.tsne(adata, n_pcs=50); sc.pl.tsne(adata, color=\\'recipe\\', save=\\'tsne_2d.png\\'); sc.pl.tsne(adata, color=\\'recipe\\', projection=\\'3d\\', save=\\'tsne_3d.png\\')\"], capture_output=True, text=True)', 'parsedOutput': {'code': 'subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"HSC-reprogrammed-DE-tsne.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.tl.tsne(adata, n_pcs=50); sc.pl.tsne(adata, color=\\'recipe\\', save=\\'tsne_2d.png\\'); sc.pl.tsne(adata, color=\\'recipe\\', projection=\\'3d\\', save=\\'tsne_3d.png\\')\"], capture_output=True, text=True)'}, 'purpose': 'Format function call'}, {'func': 'pythonCaller.execute_python_code', 'code': 'response = subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"HSC-reprogrammed-DE-tsne.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.tl.tsne(adata, n_pcs=50); sc.pl.tsne(adata, color=\\'recipe\\', save=\\'tsne_2d.png\\'); sc.pl.tsne(adata, color=\\'recipe\\', projection=\\'3d\\', save=\\'tsne_3d.png\\')\"], capture_output=True, text=True)', 'purpose': 'execute python code'}, {'func': 'utils.makeNamesConsistent', 'files': [{'old-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_00-39-53/tsnetsne_2d.png', 'new-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_00-39-53/S3-tsnetsne_2d.png'}]}]}\n",
      "2024-08-03 00:43:55,203 - INFO - current llm calls=7\n",
      "2024-08-03 00:43:55,205 - INFO - CODE\n",
      "2024-08-03 00:43:55,205 - INFO - CODER\n",
      "2024-08-03 00:43:55,207 - INFO - input_variables=['user_query'] template=\"You must select which code to run to help a user.\\n\\n**Available Scripts**\\nScript Name: scanpy_brad, \\t Description: This script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\\nScript Name: models_brad, \\t Description: This script executes a series of custom commands to build a model for predicting the contribution to reprogramming obtained by the inclusion or exclusion of individual TFs using an AnnData object or single cell datasets loaded from a .h5ad file.\\n\\n\\n**User Query**\\n{user_query}\\n\\n**Task**\\nBased on the user's query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\\n\\n**Response Template**\\nSCRIPT: <selected script>\\nREASON: <reasoning why the script fits the user prompt>\\n\"\n",
      "2024-08-03 00:43:55,207 - INFO - FIRST LLM CALL\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "==================================================\n",
      "BRAD >> 4: \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 00:43:56,627 - INFO - HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2024-08-03 00:43:56,628 - INFO - SCRIPT: scanpy_brad\n",
      "REASON: The user's query involves generating UMAP visualizations from an AnnData object loaded from a .h5ad file and highlighting specific transcription factors (TFs) by coloring points based on True/False values. The `scanpy_brad` script is the best match because it allows for manipulation, visualization, and exploration of single-cell datasets using libraries like scanpy, seaborn, and matplotlib.pyplot. These libraries are well-suited for creating UMAP visualizations and handling the specific requirements mentioned in the user's prompt.\n",
      "2024-08-03 00:43:56,629 - INFO - scriptPath=/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts\n",
      "2024-08-03 00:43:56,629 - INFO - ALL SCRIPTS FOUND. BUILDING TEMPLATE\n",
      "2024-08-03 00:43:56,630 - INFO - input_variables=['history', 'input'] template='Current conversation:\\n{history}\\n\\n**PYTHON SCRIPT**\\nYou must run this python script:\\n/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\n\\n**PYTHON SCRIPT DOCUMENTATION**:\\nThis is the doc string of this python script:\\n\"\"\"\\nThis script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\\n\\nArguments (four arguments):\\n    1. output directory: chatstatus[\\'output-directory\\']\\n    2. output file: <name of output file>\\n    3. input file: <file created in previous step>\\n    4. scanpy commands: a list of scanpy commands to be executed on the AnnData object (provided as a single string with commands separated by a delimiter, e.g., \\';\\')\\n\\nBased on the arguments, the input file will be loaded, then your commands will be executed, and finally, the output or resulting ann data object will be saved to the corrrectou output file and directory. Your code is not responsible for loading the .h5ad object, that will happen automatically, and when loaded, the object will be called adata. Your scanpy commands can operate directly on the adata object that will be loaded for you.\\n\\nAdditionally, the following imports are already provided for you and can be used in your code:\\n```\\nimport scanpy as sc\\nimport seaborn as sns\\nimport matplotlib.pyplot as plt\\nimport os\\nimport numpy as np\\nimport pandas as pd\\nfrom scipy.spatial.distance import cdist\\n```\\n\\n**Usage**\\nBRAD Line:\\n```\\nsubprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\'output-directory\\'], <output file>, <input file>, \"<scanpy commands>\"], capture_output=True, text=True)\\n```\\n\\n**Examples**\\nUse the below examples to help generate your code.\\n\\n*Example 1*\\nUser Prompt: Run scanpy preprocessing and UMAP visualization on XXX.h5ad and save the UMAP plot\\nResponse Code:\\n```\\nresponse = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"XXX-modified.h5ad\", \"<path/to/data>/XXX.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save=\\'umap.png\\')\"], capture_output=True, text=True)\\n```\\nExplination: the adata object will be loaded in memory already. The command \"sc.pp.neighbors(adata)\" will preprocess the data, then the command \"sc.tl.umap(adata)\" will perform UMAP, and finally the command \"sc.pl.umap(adata, \\'umap.png\\')\" will save the UMAP to a well named file.\\n\\n*Example 2*\\nUser Prompt: Perform PCA and clustering on the dataset YYY.h5ad and save the PCA plot\\nResponse Code:\\n```\\nresponse = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"YYY-modified.h5ad\", \"<path/to/data>/YYY.h5ad\", \"sc.pp.pca(adata); sc.tl.leiden(adata); sc.pl.pca(adata, save=\\'pca.png\\')\"], capture_output=True, text=True)\\n```\\nExplination: the adata object will be loaded in memory already. The command \"sc.pp.pca(adata)\" will preprocess the data, then the command \"sc.tl.leiden(adata)\" will perform the leiden algorithm, and finally the command \"sc.pl.pca(adata, save=\\'pca.png\\')\" will save the PCA to a well named file.\\n\\n*Example 3*\\nUser Prompt: compute the distance between all cells that are not source or target to all source and target cells.\\nReponse Code:\\n```\\nadata_X = adata.X\\nmask = ~adata.obs[\\'recipe\\'].isin([\\'source\\', \\'target\\'])\\nrecipe_embeddings = adata[mask]\\nsource_embeddings = adata[adata.obs[\\'recipe\\'] == \\'source\\']\\ntarget_embeddings = adata[adata.obs[\\'recipe\\'] == \\'target\\']\\nrecipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X)\\nrecipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X)\\nadata_dist = sc.AnnData(X=np.concatenate((recipe_distances_source, recipe_distances_target), axis=1), var=pd.concat((source_embeddings.obs, target_embeddings.obs)), obs=recipe_embeddings.obs)\\nadata_dist.write(<file name>.h5ad)\\n```\\n\\n**OUTPUT FILE NAME INSTRUCTIONS**\\n1. Output path should be chatstatus[\\'output-directory\\']\\n2. Output file name should be `<descriptive name>.h5ad`\\n\"\"\"\\n\\n\\n**CALL PYTHON SCRIPTS FROM PYTHON**:\\nUse the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\\n\\nTo call a Python script `example_script.py` which has no arguments:\\n\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\']], capture_output=True, text=True)\\n```\\n\\nTo call a Python script `example_script.py` with one argument:\\n\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\'], \\'arg1\\'], capture_output=True, text=True)\\n```\\n\\nTo call a Python script `example_script.py` with two arguments:\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\'], \\'arg1\\', \\'arg2\\'], capture_output=True, text=True)\\n```\\n\\nNote that chatstatus[\\'output-directory\\'] is ALWAYS passed as the first argument.\\n\\nThe following files were previously created by BRAD and could be used as input to a function if necessary:\\nS1-HSC-reprogrammed-DE-summary.h5ad\\nlog.json\\nS2-umapumap_2d.png\\nS3-tsnetsne_2d.png\\n\\nQuery:{input}\\n\\n**PYTHON SCRIPT**\\nPlease use the correct script name and path. You must run this python script:\\n/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\n\\n**INSTRUCTIONS**\\n1. Given the user query and the documentation, identify each of the arguments found in the user\\'s query that should be passed to the Python script.\\n2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\\n3. The last line of your response should say \"Execute: <Python code to execute>\"\\n4. Do not include any extra words or characters. Format the response/output as:\\n    Arguments: \\n    Python Code Explanation: <2 sentences maximum>\\n    Execute: <your code here>\\n\\n**IMPORTANT**\\nThe code to execute from your response must be formatted as:\\n    Execute: subprocess.call([sys.executable, \\'<path to python script>\\', \\'<argument 1>\\', \\'<argument 2>\\', ..., \\'<argument n>\\'], capture_output=True, text=True))\\nThis output should be exactly one line and no longer. Stop the response after this line.\\n'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation:\n",
      "\n",
      "\n",
      "**PYTHON SCRIPT**\n",
      "You must run this python script:\n",
      "/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\n",
      "\n",
      "**PYTHON SCRIPT DOCUMENTATION**:\n",
      "This is the doc string of this python script:\n",
      "\"\"\"\n",
      "This script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\n",
      "\n",
      "Arguments (four arguments):\n",
      "    1. output directory: chatstatus['output-directory']\n",
      "    2. output file: <name of output file>\n",
      "    3. input file: <file created in previous step>\n",
      "    4. scanpy commands: a list of scanpy commands to be executed on the AnnData object (provided as a single string with commands separated by a delimiter, e.g., ';')\n",
      "\n",
      "Based on the arguments, the input file will be loaded, then your commands will be executed, and finally, the output or resulting ann data object will be saved to the corrrectou output file and directory. Your code is not responsible for loading the .h5ad object, that will happen automatically, and when loaded, the object will be called adata. Your scanpy commands can operate directly on the adata object that will be loaded for you.\n",
      "\n",
      "Additionally, the following imports are already provided for you and can be used in your code:\n",
      "```\n",
      "import scanpy as sc\n",
      "import seaborn as sns\n",
      "import matplotlib.pyplot as plt\n",
      "import os\n",
      "import numpy as np\n",
      "import pandas as pd\n",
      "from scipy.spatial.distance import cdist\n",
      "```\n",
      "\n",
      "**Usage**\n",
      "BRAD Line:\n",
      "```\n",
      "subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus['output-directory'], <output file>, <input file>, \"<scanpy commands>\"], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "**Examples**\n",
      "Use the below examples to help generate your code.\n",
      "\n",
      "*Example 1*\n",
      "User Prompt: Run scanpy preprocessing and UMAP visualization on XXX.h5ad and save the UMAP plot\n",
      "Response Code:\n",
      "```\n",
      "response = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus['output-directory'], \"XXX-modified.h5ad\", \"<path/to/data>/XXX.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save='umap.png')\"], capture_output=True, text=True)\n",
      "```\n",
      "Explination: the adata object will be loaded in memory already. The command \"sc.pp.neighbors(adata)\" will preprocess the data, then the command \"sc.tl.umap(adata)\" will perform UMAP, and finally the command \"sc.pl.umap(adata, 'umap.png')\" will save the UMAP to a well named file.\n",
      "\n",
      "*Example 2*\n",
      "User Prompt: Perform PCA and clustering on the dataset YYY.h5ad and save the PCA plot\n",
      "Response Code:\n",
      "```\n",
      "response = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus['output-directory'], \"YYY-modified.h5ad\", \"<path/to/data>/YYY.h5ad\", \"sc.pp.pca(adata); sc.tl.leiden(adata); sc.pl.pca(adata, save='pca.png')\"], capture_output=True, text=True)\n",
      "```\n",
      "Explination: the adata object will be loaded in memory already. The command \"sc.pp.pca(adata)\" will preprocess the data, then the command \"sc.tl.leiden(adata)\" will perform the leiden algorithm, and finally the command \"sc.pl.pca(adata, save='pca.png')\" will save the PCA to a well named file.\n",
      "\n",
      "*Example 3*\n",
      "User Prompt: compute the distance between all cells that are not source or target to all source and target cells.\n",
      "Reponse Code:\n",
      "```\n",
      "adata_X = adata.X\n",
      "mask = ~adata.obs['recipe'].isin(['source', 'target'])\n",
      "recipe_embeddings = adata[mask]\n",
      "source_embeddings = adata[adata.obs['recipe'] == 'source']\n",
      "target_embeddings = adata[adata.obs['recipe'] == 'target']\n",
      "recipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X)\n",
      "recipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X)\n",
      "adata_dist = sc.AnnData(X=np.concatenate((recipe_distances_source, recipe_distances_target), axis=1), var=pd.concat((source_embeddings.obs, target_embeddings.obs)), obs=recipe_embeddings.obs)\n",
      "adata_dist.write(<file name>.h5ad)\n",
      "```\n",
      "\n",
      "**OUTPUT FILE NAME INSTRUCTIONS**\n",
      "1. Output path should be chatstatus['output-directory']\n",
      "2. Output file name should be `<descriptive name>.h5ad`\n",
      "\"\"\"\n",
      "\n",
      "\n",
      "**CALL PYTHON SCRIPTS FROM PYTHON**:\n",
      "Use the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\n",
      "\n",
      "To call a Python script `example_script.py` which has no arguments:\n",
      "\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory']], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with one argument:\n",
      "\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory'], 'arg1'], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with two arguments:\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory'], 'arg1', 'arg2'], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "Note that chatstatus['output-directory'] is ALWAYS passed as the first argument.\n",
      "\n",
      "The following files were previously created by BRAD and could be used as input to a function if necessary:\n",
      "S1-HSC-reprogrammed-DE-summary.h5ad\n",
      "log.json\n",
      "S2-umapumap_2d.png\n",
      "S3-tsnetsne_2d.png\n",
      "\n",
      "Query:4 (CODE)**:\n",
      "Prompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, generate separate UMAP visualizations highlighting the presence of each TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) by coloring the points based on the True/False values in the respective columns.\n",
      "\n",
      "**\n",
      "\n",
      "**PYTHON SCRIPT**\n",
      "Please use the correct script name and path. You must run this python script:\n",
      "/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\n",
      "\n",
      "**INSTRUCTIONS**\n",
      "1. Given the user query and the documentation, identify each of the arguments found in the user's query that should be passed to the Python script.\n",
      "2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\n",
      "3. The last line of your response should say \"Execute: <Python code to execute>\"\n",
      "4. Do not include any extra words or characters. Format the response/output as:\n",
      "    Arguments: \n",
      "    Python Code Explanation: <2 sentences maximum>\n",
      "    Execute: <your code here>\n",
      "\n",
      "**IMPORTANT**\n",
      "The code to execute from your response must be formatted as:\n",
      "    Execute: subprocess.call([sys.executable, '<path to python script>', '<argument 1>', '<argument 2>', ..., '<argument n>'], capture_output=True, text=True))\n",
      "This output should be exactly one line and no longer. Stop the response after this line.\n",
      "\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 00:44:00,338 - INFO - HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2024-08-03 00:44:00,339 - INFO - LLM OUTPUT PARSER\n",
      "2024-08-03 00:44:00,340 - INFO - Arguments:\n",
      "1. chatstatus['output-directory']\n",
      "2. \"HSC-reprogrammed-DE-UMAPs.h5ad\"\n",
      "3. \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\"\n",
      "4. \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, color=['REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B'], save='_TF_UMAP.png')\"\n",
      "\n",
      "Python Code Explanation: This code will preprocess the data, perform UMAP, and generate UMAP visualizations for each specified TF, saving the plots.\n",
      "\n",
      "Execute: subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-DE-UMAPs.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, color=['REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B'], save='_TF_UMAP.png')\"], capture_output=True, text=True)\n",
      "2024-08-03 00:44:00,340 - INFO - ['Arguments:\\n1. chatstatus[\\'output-directory\\']\\n2. \"HSC-reprogrammed-DE-UMAPs.h5ad\"\\n3. \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\"\\n4. \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, color=[\\'REL\\', \\'STAT5A\\', \\'FOS\\', \\'ETV6\\', \\'MEF2C\\', \\'RUNX3\\', \\'FOSB\\', \\'GATA2\\', \\'IKZF1\\', \\'GFI1B\\'], save=\\'_TF_UMAP.png\\')\"\\n\\nPython Code Explanation: This code will preprocess the data, perform UMAP, and generate UMAP visualizations for each specified TF, saving the plots.\\n\\n', ' subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"HSC-reprogrammed-DE-UMAPs.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, color=[\\'REL\\', \\'STAT5A\\', \\'FOS\\', \\'ETV6\\', \\'MEF2C\\', \\'RUNX3\\', \\'FOSB\\', \\'GATA2\\', \\'IKZF1\\', \\'GFI1B\\'], save=\\'_TF_UMAP.png\\')\"], capture_output=True, text=True)']\n",
      "2024-08-03 00:44:00,340 - INFO - Stripped funciton call\n",
      "2024-08-03 00:44:00,341 - INFO - subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-DE-UMAPs.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, color=['REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B'], save='_TF_UMAP.png')\"], capture_output=True, text=True)\n",
      "2024-08-03 00:44:00,341 - INFO - Cleaned up function call:\n",
      "\n",
      "2024-08-03 00:44:00,342 - INFO - subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-DE-UMAPs.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, color=['REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B'], save='_TF_UMAP.png')\"], capture_output=True, text=True)\n",
      "2024-08-03 00:44:00,342 - INFO - \n",
      "\n",
      "2024-08-03 00:44:00,343 - INFO - EVAL\n",
      "2024-08-03 00:44:00,344 - INFO - subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-DE-UMAPs.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, color=['REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B'], save='_TF_UMAP.png')\"], capture_output=True, text=True)\n",
      "2024-08-03 00:44:00,344 - INFO - END EVAL CHECK\n",
      "2024-08-03 00:44:00,344 - INFO - Finalized PYTHON Call:\n",
      "2024-08-03 00:44:00,344 - INFO -    \n",
      "2024-08-03 00:44:00,345 - INFO - response = subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-DE-UMAPs.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, color=['REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B'], save='_TF_UMAP.png')\"], capture_output=True, text=True)\n",
      "2024-08-03 00:44:00,345 - INFO -    \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 00:44:32,430 - INFO - Debug: PYTHON code executed successfully.\n",
      "2024-08-03 00:44:32,431 - INFO - Code Execution Output:\n",
      "2024-08-03 00:44:32,431 - INFO - CompletedProcess(args=['/sw/pkgs/arc/python3.11-anaconda/2024.02-1/bin/python', '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py', '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_00-39-53', 'HSC-reprogrammed-DE-UMAPs.h5ad', '/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad', \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, color=['REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B'], save='_TF_UMAP.png')\"], returncode=0, stdout=\"adata loaded from /nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\\nadata.shape=(6375, 512)\\nadata.obs.columns=Index(['recipe', 'filename', 'REL', 'STAT5A', 'FOS', 'ETV6', 'source', 'MEF2C',\\n       'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B', 'target'],\\n      dtype='object')\\nadata.obs.head()=   recipe                                           filename  ...  GFI1B  target\\n0  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n1  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n2  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n3  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n4  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n\\n[5 rows x 14 columns]\\nadata.var.head()=Empty DataFrame\\nColumns: []\\nIndex: [0, 1, 2, 3, 4]\\n****************************\\n      EXECUTE COMMANDS      \\n****************************\\nsc.pp.neighbors(adata)\\nsc.tl.umap(adata)\\nsc.pl.umap(adata, color=['REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B'], save='_TF_UMAP.png')\\nFigure(5796x3600)\\n\", stderr='/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\\n  utils.warn_names_duplicates(\"obs\")\\n/home/jpic/.local/lib/python3.11/site-packages/scanpy/tools/_utils.py:41: UserWarning: Youâ€™re trying to run this on 512 dimensions of `.X`, if you really want this, set `use_rep=\\'X\\'`.\\n         Falling back to preprocessing with `sc.pp.pca` and default params.\\n  warnings.warn(\\nWARNING: saving figure to file /home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_00-39-53/umap_TF_UMAP.png\\n')\n",
      "2024-08-03 00:44:32,431 - INFO - Debug: PYTHON code output saved to output.\n",
      "2024-08-03 00:44:32,431 - INFO - \n",
      "\n",
      "\n",
      "route\n",
      "\n",
      "\n",
      "\n",
      "2024-08-03 00:44:32,432 - INFO - CODE\n",
      "2024-08-03 00:44:32,433 - INFO - Finding Stage Number of Pipeline\n",
      "2024-08-03 00:44:32,433 - INFO - {0: {'order': 0, 'module': 'PLANNER', 'prompt': None, 'description': 'This step designed the plan. It is placed in the queue because we needed a place holder for 0 indexed lists.', 'output': []}, 1: {'order': 1, 'module': 'CODE', 'prompt': '/force CODE  1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and create a basic summary of the data.\\n\\n**', 'description': ' 1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and create a basic summary of the data.\\n\\n**', 'output': ['/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_00-39-53/S1-HSC-reprogrammed-DE-summary.h5ad']}, 2: {'order': 2, 'module': 'CODE', 'prompt': \"/force CODE  2 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, create UMAP plots of the data in 2D and 3D space and color the visualizations based on the 'recipe' column to distinguish between 'source', 'target', and different TF recipes.\\n\\n**\", 'description': \" 2 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, create UMAP plots of the data in 2D and 3D space and color the visualizations based on the 'recipe' column to distinguish between 'source', 'target', and different TF recipes.\\n\\n**\", 'output': ['/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_00-39-53/S2-umapumap_2d.png']}, 3: {'order': 3, 'module': 'CODE', 'prompt': \"/force CODE  3 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, create t-SNE plots of the data in 2D and 3D space and color the visualizations based on the 'recipe' column to distinguish between 'source', 'target', and different TF recipes.\\n\\n**\", 'description': \" 3 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, create t-SNE plots of the data in 2D and 3D space and color the visualizations based on the 'recipe' column to distinguish between 'source', 'target', and different TF recipes.\\n\\n**\", 'output': ['/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_00-39-53/S3-tsnetsne_2d.png']}, 4: {'order': 4, 'module': 'CODE', 'prompt': '/force CODE  4 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, generate separate UMAP visualizations highlighting the presence of each TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) by coloring the points based on the True/False values in the respective columns.\\n\\n**', 'description': ' 4 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, generate separate UMAP visualizations highlighting the presence of each TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) by coloring the points based on the True/False values in the respective columns.\\n\\n**', 'output': []}, 5: {'order': 5, 'module': 'CODE', 'prompt': '/force CODE  5 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, generate separate t-SNE visualizations highlighting the presence of each TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) by coloring the points based on the True/False values in the respective columns.\\n\\nThe removed ', 'description': ' 5 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, generate separate t-SNE visualizations highlighting the presence of each TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) by coloring the points based on the True/False values in the respective columns.\\n\\nThe removed ', 'output': []}}\n",
      "2024-08-03 00:44:32,434 - INFO - IP=4\n",
      "2024-08-03 00:44:32,434 - INFO - type(IP)=<class 'int'>\n",
      "2024-08-03 00:44:32,436 - INFO - \n",
      "\n",
      "PLANNER:\n",
      "\n",
      "{'module': 'CODE', 'steps': [{'llm': \"client=<openai.resources.chat.completions.Completions object at 0x1480cc33e8d0> async_client=<openai.resources.chat.completions.AsyncCompletions object at 0x1480cc324e50> model_name='gpt-4o' temperature=0.0 openai_api_key=SecretStr('**********') openai_proxy=''\", 'memory': 'None', 'prompt': 'input_variables=[\\'user_query\\'] template=\"You must select which code to run to help a user.\\\\n\\\\n**Available Scripts**\\\\nScript Name: scanpy_brad, \\\\t Description: This script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\\\\nScript Name: models_brad, \\\\t Description: This script executes a series of custom commands to build a model for predicting the contribution to reprogramming obtained by the inclusion or exclusion of individual TFs using an AnnData object or single cell datasets loaded from a .h5ad file.\\\\n\\\\n\\\\n**User Query**\\\\n{user_query}\\\\n\\\\n**Task**\\\\nBased on the user\\'s query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\\\\n\\\\n**Response Template**\\\\nSCRIPT: <selected script>\\\\nREASON: <reasoning why the script fits the user prompt>\\\\n\"', 'input': '4 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, generate separate UMAP visualizations highlighting the presence of each TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) by coloring the points based on the True/False values in the respective columns.\\n\\n**', 'output': 'content=\"SCRIPT: scanpy_brad\\\\nREASON: The user\\'s query involves generating UMAP visualizations from an AnnData object loaded from a .h5ad file and highlighting specific transcription factors (TFs) by coloring points based on True/False values. The `scanpy_brad` script is the best match because it allows for manipulation, visualization, and exploration of single-cell datasets using libraries like scanpy, seaborn, and matplotlib.pyplot. These libraries are well-suited for creating UMAP visualizations and handling the specific requirements mentioned in the user\\'s prompt.\" response_metadata={\\'token_usage\\': {\\'completion_tokens\\': 113, \\'prompt_tokens\\': 364, \\'total_tokens\\': 477}, \\'model_name\\': \\'gpt-4o-2024-05-13\\', \\'system_fingerprint\\': \\'fp_3cd8b62c3b\\', \\'finish_reason\\': \\'stop\\', \\'logprobs\\': None} id=\\'run-44c4113e-4414-4ee5-8f98-56e90ca90d8d-0\\' usage_metadata={\\'input_tokens\\': 364, \\'output_tokens\\': 113, \\'total_tokens\\': 477}', 'parsedOutput': {'scriptName': '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py', 'scriptType': 'python', 'scriptPath': '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts'}, 'purpose': 'Select which code to run'}, {'llm': \"client=<openai.resources.chat.completions.Completions object at 0x1480cc33e8d0> async_client=<openai.resources.chat.completions.AsyncCompletions object at 0x1480cc324e50> model_name='gpt-4o' temperature=0.0 openai_api_key=SecretStr('**********') openai_proxy=''\", 'memory': 'None', 'prompt': 'input_variables=[\\'history\\', \\'input\\'] template=\\'Current conversation:\\\\n{history}\\\\n\\\\n**PYTHON SCRIPT**\\\\nYou must run this python script:\\\\n/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\\\n\\\\n**PYTHON SCRIPT DOCUMENTATION**:\\\\nThis is the doc string of this python script:\\\\n\"\"\"\\\\nThis script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\\\\n\\\\nArguments (four arguments):\\\\n    1. output directory: chatstatus[\\\\\\'output-directory\\\\\\']\\\\n    2. output file: <name of output file>\\\\n    3. input file: <file created in previous step>\\\\n    4. scanpy commands: a list of scanpy commands to be executed on the AnnData object (provided as a single string with commands separated by a delimiter, e.g., \\\\\\';\\\\\\')\\\\n\\\\nBased on the arguments, the input file will be loaded, then your commands will be executed, and finally, the output or resulting ann data object will be saved to the corrrectou output file and directory. Your code is not responsible for loading the .h5ad object, that will happen automatically, and when loaded, the object will be called adata. Your scanpy commands can operate directly on the adata object that will be loaded for you.\\\\n\\\\nAdditionally, the following imports are already provided for you and can be used in your code:\\\\n```\\\\nimport scanpy as sc\\\\nimport seaborn as sns\\\\nimport matplotlib.pyplot as plt\\\\nimport os\\\\nimport numpy as np\\\\nimport pandas as pd\\\\nfrom scipy.spatial.distance import cdist\\\\n```\\\\n\\\\n**Usage**\\\\nBRAD Line:\\\\n```\\\\nsubprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\\\\\'output-directory\\\\\\'], <output file>, <input file>, \"<scanpy commands>\"], capture_output=True, text=True)\\\\n```\\\\n\\\\n**Examples**\\\\nUse the below examples to help generate your code.\\\\n\\\\n*Example 1*\\\\nUser Prompt: Run scanpy preprocessing and UMAP visualization on XXX.h5ad and save the UMAP plot\\\\nResponse Code:\\\\n```\\\\nresponse = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\\\\\'output-directory\\\\\\'], \"XXX-modified.h5ad\", \"<path/to/data>/XXX.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save=\\\\\\'umap.png\\\\\\')\"], capture_output=True, text=True)\\\\n```\\\\nExplination: the adata object will be loaded in memory already. The command \"sc.pp.neighbors(adata)\" will preprocess the data, then the command \"sc.tl.umap(adata)\" will perform UMAP, and finally the command \"sc.pl.umap(adata, \\\\\\'umap.png\\\\\\')\" will save the UMAP to a well named file.\\\\n\\\\n*Example 2*\\\\nUser Prompt: Perform PCA and clustering on the dataset YYY.h5ad and save the PCA plot\\\\nResponse Code:\\\\n```\\\\nresponse = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\\\\\'output-directory\\\\\\'], \"YYY-modified.h5ad\", \"<path/to/data>/YYY.h5ad\", \"sc.pp.pca(adata); sc.tl.leiden(adata); sc.pl.pca(adata, save=\\\\\\'pca.png\\\\\\')\"], capture_output=True, text=True)\\\\n```\\\\nExplination: the adata object will be loaded in memory already. The command \"sc.pp.pca(adata)\" will preprocess the data, then the command \"sc.tl.leiden(adata)\" will perform the leiden algorithm, and finally the command \"sc.pl.pca(adata, save=\\\\\\'pca.png\\\\\\')\" will save the PCA to a well named file.\\\\n\\\\n*Example 3*\\\\nUser Prompt: compute the distance between all cells that are not source or target to all source and target cells.\\\\nReponse Code:\\\\n```\\\\nadata_X = adata.X\\\\nmask = ~adata.obs[\\\\\\'recipe\\\\\\'].isin([\\\\\\'source\\\\\\', \\\\\\'target\\\\\\'])\\\\nrecipe_embeddings = adata[mask]\\\\nsource_embeddings = adata[adata.obs[\\\\\\'recipe\\\\\\'] == \\\\\\'source\\\\\\']\\\\ntarget_embeddings = adata[adata.obs[\\\\\\'recipe\\\\\\'] == \\\\\\'target\\\\\\']\\\\nrecipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X)\\\\nrecipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X)\\\\nadata_dist = sc.AnnData(X=np.concatenate((recipe_distances_source, recipe_distances_target), axis=1), var=pd.concat((source_embeddings.obs, target_embeddings.obs)), obs=recipe_embeddings.obs)\\\\nadata_dist.write(<file name>.h5ad)\\\\n```\\\\n\\\\n**OUTPUT FILE NAME INSTRUCTIONS**\\\\n1. Output path should be chatstatus[\\\\\\'output-directory\\\\\\']\\\\n2. Output file name should be `<descriptive name>.h5ad`\\\\n\"\"\"\\\\n\\\\n\\\\n**CALL PYTHON SCRIPTS FROM PYTHON**:\\\\nUse the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\\\\n\\\\nTo call a Python script `example_script.py` which has no arguments:\\\\n\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\']], capture_output=True, text=True)\\\\n```\\\\n\\\\nTo call a Python script `example_script.py` with one argument:\\\\n\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\'], \\\\\\'arg1\\\\\\'], capture_output=True, text=True)\\\\n```\\\\n\\\\nTo call a Python script `example_script.py` with two arguments:\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\'], \\\\\\'arg1\\\\\\', \\\\\\'arg2\\\\\\'], capture_output=True, text=True)\\\\n```\\\\n\\\\nNote that chatstatus[\\\\\\'output-directory\\\\\\'] is ALWAYS passed as the first argument.\\\\n\\\\nThe following files were previously created by BRAD and could be used as input to a function if necessary:\\\\nS1-HSC-reprogrammed-DE-summary.h5ad\\\\nlog.json\\\\nS2-umapumap_2d.png\\\\nS3-tsnetsne_2d.png\\\\n\\\\nQuery:{input}\\\\n\\\\n**PYTHON SCRIPT**\\\\nPlease use the correct script name and path. You must run this python script:\\\\n/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\\\n\\\\n**INSTRUCTIONS**\\\\n1. Given the user query and the documentation, identify each of the arguments found in the user\\\\\\'s query that should be passed to the Python script.\\\\n2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\\\\n3. The last line of your response should say \"Execute: <Python code to execute>\"\\\\n4. Do not include any extra words or characters. Format the response/output as:\\\\n    Arguments: \\\\n    Python Code Explanation: <2 sentences maximum>\\\\n    Execute: <your code here>\\\\n\\\\n**IMPORTANT**\\\\nThe code to execute from your response must be formatted as:\\\\n    Execute: subprocess.call([sys.executable, \\\\\\'<path to python script>\\\\\\', \\\\\\'<argument 1>\\\\\\', \\\\\\'<argument 2>\\\\\\', ..., \\\\\\'<argument n>\\\\\\'], capture_output=True, text=True))\\\\nThis output should be exactly one line and no longer. Stop the response after this line.\\\\n\\'', 'input': '4 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, generate separate UMAP visualizations highlighting the presence of each TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) by coloring the points based on the True/False values in the respective columns.\\n\\n**', 'output': 'Arguments:\\n1. chatstatus[\\'output-directory\\']\\n2. \"HSC-reprogrammed-DE-UMAPs.h5ad\"\\n3. \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\"\\n4. \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, color=[\\'REL\\', \\'STAT5A\\', \\'FOS\\', \\'ETV6\\', \\'MEF2C\\', \\'RUNX3\\', \\'FOSB\\', \\'GATA2\\', \\'IKZF1\\', \\'GFI1B\\'], save=\\'_TF_UMAP.png\\')\"\\n\\nPython Code Explanation: This code will preprocess the data, perform UMAP, and generate UMAP visualizations for each specified TF, saving the plots.\\n\\nExecute: subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"HSC-reprogrammed-DE-UMAPs.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, color=[\\'REL\\', \\'STAT5A\\', \\'FOS\\', \\'ETV6\\', \\'MEF2C\\', \\'RUNX3\\', \\'FOSB\\', \\'GATA2\\', \\'IKZF1\\', \\'GFI1B\\'], save=\\'_TF_UMAP.png\\')\"], capture_output=True, text=True)', 'parsedOutput': {'code': 'subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"HSC-reprogrammed-DE-UMAPs.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, color=[\\'REL\\', \\'STAT5A\\', \\'FOS\\', \\'ETV6\\', \\'MEF2C\\', \\'RUNX3\\', \\'FOSB\\', \\'GATA2\\', \\'IKZF1\\', \\'GFI1B\\'], save=\\'_TF_UMAP.png\\')\"], capture_output=True, text=True)'}, 'purpose': 'Format function call'}, {'func': 'pythonCaller.execute_python_code', 'code': 'response = subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"HSC-reprogrammed-DE-UMAPs.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, color=[\\'REL\\', \\'STAT5A\\', \\'FOS\\', \\'ETV6\\', \\'MEF2C\\', \\'RUNX3\\', \\'FOSB\\', \\'GATA2\\', \\'IKZF1\\', \\'GFI1B\\'], save=\\'_TF_UMAP.png\\')\"], capture_output=True, text=True)', 'purpose': 'execute python code'}, {'func': 'utils.makeNamesConsistent', 'files': [{'old-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_00-39-53/HSC-reprogrammed-DE-UMAPs.h5ad', 'new-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_00-39-53/S4-HSC-reprogrammed-DE-UMAPs.h5ad'}, {'old-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_00-39-53/umap_TF_UMAP.png', 'new-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_00-39-53/S4-umap_TF_UMAP.png'}]}]}\n",
      "2024-08-03 00:44:32,440 - INFO - current llm calls=9\n",
      "2024-08-03 00:44:32,442 - INFO - CODE\n",
      "2024-08-03 00:44:32,443 - INFO - CODER\n",
      "2024-08-03 00:44:32,444 - INFO - input_variables=['user_query'] template=\"You must select which code to run to help a user.\\n\\n**Available Scripts**\\nScript Name: scanpy_brad, \\t Description: This script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\\nScript Name: models_brad, \\t Description: This script executes a series of custom commands to build a model for predicting the contribution to reprogramming obtained by the inclusion or exclusion of individual TFs using an AnnData object or single cell datasets loaded from a .h5ad file.\\n\\n\\n**User Query**\\n{user_query}\\n\\n**Task**\\nBased on the user's query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\\n\\n**Response Template**\\nSCRIPT: <selected script>\\nREASON: <reasoning why the script fits the user prompt>\\n\"\n",
      "2024-08-03 00:44:32,444 - INFO - FIRST LLM CALL\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "==================================================\n",
      "BRAD >> 5: \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 00:44:34,112 - INFO - HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2024-08-03 00:44:34,113 - INFO - SCRIPT: scanpy_brad\n",
      "REASON: The user's query involves generating t-SNE visualizations from an AnnData object loaded from a .h5ad file, which requires data manipulation and visualization capabilities. The `scanpy_brad` script is designed to handle such tasks, as it allows for the use of libraries like scanpy, seaborn, and matplotlib.pyplot to process .h5ad files and create visualizations. This makes it the best match for generating the requested t-SNE plots highlighting the presence of each specified transcription factor (TF).\n",
      "2024-08-03 00:44:34,114 - INFO - scriptPath=/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts\n",
      "2024-08-03 00:44:34,114 - INFO - ALL SCRIPTS FOUND. BUILDING TEMPLATE\n",
      "2024-08-03 00:44:34,115 - INFO - input_variables=['history', 'input'] template='Current conversation:\\n{history}\\n\\n**PYTHON SCRIPT**\\nYou must run this python script:\\n/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\n\\n**PYTHON SCRIPT DOCUMENTATION**:\\nThis is the doc string of this python script:\\n\"\"\"\\nThis script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\\n\\nArguments (four arguments):\\n    1. output directory: chatstatus[\\'output-directory\\']\\n    2. output file: <name of output file>\\n    3. input file: <file created in previous step>\\n    4. scanpy commands: a list of scanpy commands to be executed on the AnnData object (provided as a single string with commands separated by a delimiter, e.g., \\';\\')\\n\\nBased on the arguments, the input file will be loaded, then your commands will be executed, and finally, the output or resulting ann data object will be saved to the corrrectou output file and directory. Your code is not responsible for loading the .h5ad object, that will happen automatically, and when loaded, the object will be called adata. Your scanpy commands can operate directly on the adata object that will be loaded for you.\\n\\nAdditionally, the following imports are already provided for you and can be used in your code:\\n```\\nimport scanpy as sc\\nimport seaborn as sns\\nimport matplotlib.pyplot as plt\\nimport os\\nimport numpy as np\\nimport pandas as pd\\nfrom scipy.spatial.distance import cdist\\n```\\n\\n**Usage**\\nBRAD Line:\\n```\\nsubprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\'output-directory\\'], <output file>, <input file>, \"<scanpy commands>\"], capture_output=True, text=True)\\n```\\n\\n**Examples**\\nUse the below examples to help generate your code.\\n\\n*Example 1*\\nUser Prompt: Run scanpy preprocessing and UMAP visualization on XXX.h5ad and save the UMAP plot\\nResponse Code:\\n```\\nresponse = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"XXX-modified.h5ad\", \"<path/to/data>/XXX.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save=\\'umap.png\\')\"], capture_output=True, text=True)\\n```\\nExplination: the adata object will be loaded in memory already. The command \"sc.pp.neighbors(adata)\" will preprocess the data, then the command \"sc.tl.umap(adata)\" will perform UMAP, and finally the command \"sc.pl.umap(adata, \\'umap.png\\')\" will save the UMAP to a well named file.\\n\\n*Example 2*\\nUser Prompt: Perform PCA and clustering on the dataset YYY.h5ad and save the PCA plot\\nResponse Code:\\n```\\nresponse = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"YYY-modified.h5ad\", \"<path/to/data>/YYY.h5ad\", \"sc.pp.pca(adata); sc.tl.leiden(adata); sc.pl.pca(adata, save=\\'pca.png\\')\"], capture_output=True, text=True)\\n```\\nExplination: the adata object will be loaded in memory already. The command \"sc.pp.pca(adata)\" will preprocess the data, then the command \"sc.tl.leiden(adata)\" will perform the leiden algorithm, and finally the command \"sc.pl.pca(adata, save=\\'pca.png\\')\" will save the PCA to a well named file.\\n\\n*Example 3*\\nUser Prompt: compute the distance between all cells that are not source or target to all source and target cells.\\nReponse Code:\\n```\\nadata_X = adata.X\\nmask = ~adata.obs[\\'recipe\\'].isin([\\'source\\', \\'target\\'])\\nrecipe_embeddings = adata[mask]\\nsource_embeddings = adata[adata.obs[\\'recipe\\'] == \\'source\\']\\ntarget_embeddings = adata[adata.obs[\\'recipe\\'] == \\'target\\']\\nrecipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X)\\nrecipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X)\\nadata_dist = sc.AnnData(X=np.concatenate((recipe_distances_source, recipe_distances_target), axis=1), var=pd.concat((source_embeddings.obs, target_embeddings.obs)), obs=recipe_embeddings.obs)\\nadata_dist.write(<file name>.h5ad)\\n```\\n\\n**OUTPUT FILE NAME INSTRUCTIONS**\\n1. Output path should be chatstatus[\\'output-directory\\']\\n2. Output file name should be `<descriptive name>.h5ad`\\n\"\"\"\\n\\n\\n**CALL PYTHON SCRIPTS FROM PYTHON**:\\nUse the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\\n\\nTo call a Python script `example_script.py` which has no arguments:\\n\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\']], capture_output=True, text=True)\\n```\\n\\nTo call a Python script `example_script.py` with one argument:\\n\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\'], \\'arg1\\'], capture_output=True, text=True)\\n```\\n\\nTo call a Python script `example_script.py` with two arguments:\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\'], \\'arg1\\', \\'arg2\\'], capture_output=True, text=True)\\n```\\n\\nNote that chatstatus[\\'output-directory\\'] is ALWAYS passed as the first argument.\\n\\nThe following files were previously created by BRAD and could be used as input to a function if necessary:\\nS1-HSC-reprogrammed-DE-summary.h5ad\\nlog.json\\nS4-HSC-reprogrammed-DE-UMAPs.h5ad\\nS2-umapumap_2d.png\\nS4-umap_TF_UMAP.png\\nS3-tsnetsne_2d.png\\n\\nQuery:{input}\\n\\n**PYTHON SCRIPT**\\nPlease use the correct script name and path. You must run this python script:\\n/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\n\\n**INSTRUCTIONS**\\n1. Given the user query and the documentation, identify each of the arguments found in the user\\'s query that should be passed to the Python script.\\n2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\\n3. The last line of your response should say \"Execute: <Python code to execute>\"\\n4. Do not include any extra words or characters. Format the response/output as:\\n    Arguments: \\n    Python Code Explanation: <2 sentences maximum>\\n    Execute: <your code here>\\n\\n**IMPORTANT**\\nThe code to execute from your response must be formatted as:\\n    Execute: subprocess.call([sys.executable, \\'<path to python script>\\', \\'<argument 1>\\', \\'<argument 2>\\', ..., \\'<argument n>\\'], capture_output=True, text=True))\\nThis output should be exactly one line and no longer. Stop the response after this line.\\n'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation:\n",
      "\n",
      "\n",
      "**PYTHON SCRIPT**\n",
      "You must run this python script:\n",
      "/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\n",
      "\n",
      "**PYTHON SCRIPT DOCUMENTATION**:\n",
      "This is the doc string of this python script:\n",
      "\"\"\"\n",
      "This script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\n",
      "\n",
      "Arguments (four arguments):\n",
      "    1. output directory: chatstatus['output-directory']\n",
      "    2. output file: <name of output file>\n",
      "    3. input file: <file created in previous step>\n",
      "    4. scanpy commands: a list of scanpy commands to be executed on the AnnData object (provided as a single string with commands separated by a delimiter, e.g., ';')\n",
      "\n",
      "Based on the arguments, the input file will be loaded, then your commands will be executed, and finally, the output or resulting ann data object will be saved to the corrrectou output file and directory. Your code is not responsible for loading the .h5ad object, that will happen automatically, and when loaded, the object will be called adata. Your scanpy commands can operate directly on the adata object that will be loaded for you.\n",
      "\n",
      "Additionally, the following imports are already provided for you and can be used in your code:\n",
      "```\n",
      "import scanpy as sc\n",
      "import seaborn as sns\n",
      "import matplotlib.pyplot as plt\n",
      "import os\n",
      "import numpy as np\n",
      "import pandas as pd\n",
      "from scipy.spatial.distance import cdist\n",
      "```\n",
      "\n",
      "**Usage**\n",
      "BRAD Line:\n",
      "```\n",
      "subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus['output-directory'], <output file>, <input file>, \"<scanpy commands>\"], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "**Examples**\n",
      "Use the below examples to help generate your code.\n",
      "\n",
      "*Example 1*\n",
      "User Prompt: Run scanpy preprocessing and UMAP visualization on XXX.h5ad and save the UMAP plot\n",
      "Response Code:\n",
      "```\n",
      "response = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus['output-directory'], \"XXX-modified.h5ad\", \"<path/to/data>/XXX.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save='umap.png')\"], capture_output=True, text=True)\n",
      "```\n",
      "Explination: the adata object will be loaded in memory already. The command \"sc.pp.neighbors(adata)\" will preprocess the data, then the command \"sc.tl.umap(adata)\" will perform UMAP, and finally the command \"sc.pl.umap(adata, 'umap.png')\" will save the UMAP to a well named file.\n",
      "\n",
      "*Example 2*\n",
      "User Prompt: Perform PCA and clustering on the dataset YYY.h5ad and save the PCA plot\n",
      "Response Code:\n",
      "```\n",
      "response = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus['output-directory'], \"YYY-modified.h5ad\", \"<path/to/data>/YYY.h5ad\", \"sc.pp.pca(adata); sc.tl.leiden(adata); sc.pl.pca(adata, save='pca.png')\"], capture_output=True, text=True)\n",
      "```\n",
      "Explination: the adata object will be loaded in memory already. The command \"sc.pp.pca(adata)\" will preprocess the data, then the command \"sc.tl.leiden(adata)\" will perform the leiden algorithm, and finally the command \"sc.pl.pca(adata, save='pca.png')\" will save the PCA to a well named file.\n",
      "\n",
      "*Example 3*\n",
      "User Prompt: compute the distance between all cells that are not source or target to all source and target cells.\n",
      "Reponse Code:\n",
      "```\n",
      "adata_X = adata.X\n",
      "mask = ~adata.obs['recipe'].isin(['source', 'target'])\n",
      "recipe_embeddings = adata[mask]\n",
      "source_embeddings = adata[adata.obs['recipe'] == 'source']\n",
      "target_embeddings = adata[adata.obs['recipe'] == 'target']\n",
      "recipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X)\n",
      "recipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X)\n",
      "adata_dist = sc.AnnData(X=np.concatenate((recipe_distances_source, recipe_distances_target), axis=1), var=pd.concat((source_embeddings.obs, target_embeddings.obs)), obs=recipe_embeddings.obs)\n",
      "adata_dist.write(<file name>.h5ad)\n",
      "```\n",
      "\n",
      "**OUTPUT FILE NAME INSTRUCTIONS**\n",
      "1. Output path should be chatstatus['output-directory']\n",
      "2. Output file name should be `<descriptive name>.h5ad`\n",
      "\"\"\"\n",
      "\n",
      "\n",
      "**CALL PYTHON SCRIPTS FROM PYTHON**:\n",
      "Use the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\n",
      "\n",
      "To call a Python script `example_script.py` which has no arguments:\n",
      "\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory']], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with one argument:\n",
      "\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory'], 'arg1'], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with two arguments:\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory'], 'arg1', 'arg2'], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "Note that chatstatus['output-directory'] is ALWAYS passed as the first argument.\n",
      "\n",
      "The following files were previously created by BRAD and could be used as input to a function if necessary:\n",
      "S1-HSC-reprogrammed-DE-summary.h5ad\n",
      "log.json\n",
      "S4-HSC-reprogrammed-DE-UMAPs.h5ad\n",
      "S2-umapumap_2d.png\n",
      "S4-umap_TF_UMAP.png\n",
      "S3-tsnetsne_2d.png\n",
      "\n",
      "Query:5 (CODE)**:\n",
      "Prompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, generate separate t-SNE visualizations highlighting the presence of each TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) by coloring the points based on the True/False values in the respective columns.\n",
      "\n",
      "The removed\n",
      "\n",
      "**PYTHON SCRIPT**\n",
      "Please use the correct script name and path. You must run this python script:\n",
      "/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\n",
      "\n",
      "**INSTRUCTIONS**\n",
      "1. Given the user query and the documentation, identify each of the arguments found in the user's query that should be passed to the Python script.\n",
      "2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\n",
      "3. The last line of your response should say \"Execute: <Python code to execute>\"\n",
      "4. Do not include any extra words or characters. Format the response/output as:\n",
      "    Arguments: \n",
      "    Python Code Explanation: <2 sentences maximum>\n",
      "    Execute: <your code here>\n",
      "\n",
      "**IMPORTANT**\n",
      "The code to execute from your response must be formatted as:\n",
      "    Execute: subprocess.call([sys.executable, '<path to python script>', '<argument 1>', '<argument 2>', ..., '<argument n>'], capture_output=True, text=True))\n",
      "This output should be exactly one line and no longer. Stop the response after this line.\n",
      "\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 00:44:38,295 - INFO - HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2024-08-03 00:44:38,297 - INFO - LLM OUTPUT PARSER\n",
      "2024-08-03 00:44:38,297 - INFO - Arguments:\n",
      "1. chatstatus['output-directory']\n",
      "2. \"HSC-reprogrammed-DE-tSNE.h5ad\"\n",
      "3. \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\"\n",
      "4. \"sc.tl.tsne(adata); sc.pl.tsne(adata, color=['REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B'], save='tSNE_TFs.png')\"\n",
      "\n",
      "Python Code Explanation: This code will load the AnnData object, perform t-SNE, and generate visualizations highlighting the presence of each specified TF, saving the plot as 'tSNE_TFs.png'.\n",
      "\n",
      "Execute: subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-DE-tSNE.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.tl.tsne(adata); sc.pl.tsne(adata, color=['REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B'], save='tSNE_TFs.png')\"], capture_output=True, text=True)\n",
      "2024-08-03 00:44:38,297 - INFO - ['Arguments:\\n1. chatstatus[\\'output-directory\\']\\n2. \"HSC-reprogrammed-DE-tSNE.h5ad\"\\n3. \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\"\\n4. \"sc.tl.tsne(adata); sc.pl.tsne(adata, color=[\\'REL\\', \\'STAT5A\\', \\'FOS\\', \\'ETV6\\', \\'MEF2C\\', \\'RUNX3\\', \\'FOSB\\', \\'GATA2\\', \\'IKZF1\\', \\'GFI1B\\'], save=\\'tSNE_TFs.png\\')\"\\n\\nPython Code Explanation: This code will load the AnnData object, perform t-SNE, and generate visualizations highlighting the presence of each specified TF, saving the plot as \\'tSNE_TFs.png\\'.\\n\\n', ' subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"HSC-reprogrammed-DE-tSNE.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.tl.tsne(adata); sc.pl.tsne(adata, color=[\\'REL\\', \\'STAT5A\\', \\'FOS\\', \\'ETV6\\', \\'MEF2C\\', \\'RUNX3\\', \\'FOSB\\', \\'GATA2\\', \\'IKZF1\\', \\'GFI1B\\'], save=\\'tSNE_TFs.png\\')\"], capture_output=True, text=True)']\n",
      "2024-08-03 00:44:38,298 - INFO - Stripped funciton call\n",
      "2024-08-03 00:44:38,298 - INFO - subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-DE-tSNE.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.tl.tsne(adata); sc.pl.tsne(adata, color=['REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B'], save='tSNE_TFs.png')\"], capture_output=True, text=True)\n",
      "2024-08-03 00:44:38,298 - INFO - Cleaned up function call:\n",
      "\n",
      "2024-08-03 00:44:38,299 - INFO - subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-DE-tSNE.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.tl.tsne(adata); sc.pl.tsne(adata, color=['REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B'], save='tSNE_TFs.png')\"], capture_output=True, text=True)\n",
      "2024-08-03 00:44:38,299 - INFO - \n",
      "\n",
      "2024-08-03 00:44:38,300 - INFO - EVAL\n",
      "2024-08-03 00:44:38,300 - INFO - subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-DE-tSNE.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.tl.tsne(adata); sc.pl.tsne(adata, color=['REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B'], save='tSNE_TFs.png')\"], capture_output=True, text=True)\n",
      "2024-08-03 00:44:38,301 - INFO - END EVAL CHECK\n",
      "2024-08-03 00:44:38,301 - INFO - Finalized PYTHON Call:\n",
      "2024-08-03 00:44:38,301 - INFO -    \n",
      "2024-08-03 00:44:38,302 - INFO - response = subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-DE-tSNE.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.tl.tsne(adata); sc.pl.tsne(adata, color=['REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B'], save='tSNE_TFs.png')\"], capture_output=True, text=True)\n",
      "2024-08-03 00:44:38,302 - INFO -    \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 00:44:54,841 - INFO - Debug: PYTHON code executed successfully.\n",
      "2024-08-03 00:44:54,842 - INFO - Code Execution Output:\n",
      "2024-08-03 00:44:54,842 - INFO - CompletedProcess(args=['/sw/pkgs/arc/python3.11-anaconda/2024.02-1/bin/python', '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py', '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_00-39-53', 'HSC-reprogrammed-DE-tSNE.h5ad', '/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad', \"sc.tl.tsne(adata); sc.pl.tsne(adata, color=['REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B'], save='tSNE_TFs.png')\"], returncode=0, stdout=\"adata loaded from /nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\\nadata.shape=(6375, 512)\\nadata.obs.columns=Index(['recipe', 'filename', 'REL', 'STAT5A', 'FOS', 'ETV6', 'source', 'MEF2C',\\n       'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B', 'target'],\\n      dtype='object')\\nadata.obs.head()=   recipe                                           filename  ...  GFI1B  target\\n0  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n1  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n2  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n3  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n4  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n\\n[5 rows x 14 columns]\\nadata.var.head()=Empty DataFrame\\nColumns: []\\nIndex: [0, 1, 2, 3, 4]\\n****************************\\n      EXECUTE COMMANDS      \\n****************************\\nsc.tl.tsne(adata)\\nsc.pl.tsne(adata, color=['REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B'], save='tSNE_TFs.png')\\nFigure(5796x3600)\\n\", stderr='/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\\n  utils.warn_names_duplicates(\"obs\")\\n/home/jpic/.local/lib/python3.11/site-packages/scanpy/tools/_utils.py:41: UserWarning: Youâ€™re trying to run this on 512 dimensions of `.X`, if you really want this, set `use_rep=\\'X\\'`.\\n         Falling back to preprocessing with `sc.pp.pca` and default params.\\n  warnings.warn(\\nWARNING: saving figure to file /home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_00-39-53/tsnetSNE_TFs.png\\n')\n",
      "2024-08-03 00:44:54,843 - INFO - Debug: PYTHON code output saved to output.\n",
      "2024-08-03 00:44:54,843 - INFO - \n",
      "\n",
      "\n",
      "route\n",
      "\n",
      "\n",
      "\n",
      "2024-08-03 00:44:54,843 - INFO - CODE\n",
      "2024-08-03 00:44:54,845 - INFO - Finding Stage Number of Pipeline\n",
      "2024-08-03 00:44:54,845 - INFO - {0: {'order': 0, 'module': 'PLANNER', 'prompt': None, 'description': 'This step designed the plan. It is placed in the queue because we needed a place holder for 0 indexed lists.', 'output': []}, 1: {'order': 1, 'module': 'CODE', 'prompt': '/force CODE  1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and create a basic summary of the data.\\n\\n**', 'description': ' 1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and create a basic summary of the data.\\n\\n**', 'output': ['/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_00-39-53/S1-HSC-reprogrammed-DE-summary.h5ad']}, 2: {'order': 2, 'module': 'CODE', 'prompt': \"/force CODE  2 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, create UMAP plots of the data in 2D and 3D space and color the visualizations based on the 'recipe' column to distinguish between 'source', 'target', and different TF recipes.\\n\\n**\", 'description': \" 2 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, create UMAP plots of the data in 2D and 3D space and color the visualizations based on the 'recipe' column to distinguish between 'source', 'target', and different TF recipes.\\n\\n**\", 'output': ['/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_00-39-53/S2-umapumap_2d.png']}, 3: {'order': 3, 'module': 'CODE', 'prompt': \"/force CODE  3 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, create t-SNE plots of the data in 2D and 3D space and color the visualizations based on the 'recipe' column to distinguish between 'source', 'target', and different TF recipes.\\n\\n**\", 'description': \" 3 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, create t-SNE plots of the data in 2D and 3D space and color the visualizations based on the 'recipe' column to distinguish between 'source', 'target', and different TF recipes.\\n\\n**\", 'output': ['/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_00-39-53/S3-tsnetsne_2d.png']}, 4: {'order': 4, 'module': 'CODE', 'prompt': '/force CODE  4 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, generate separate UMAP visualizations highlighting the presence of each TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) by coloring the points based on the True/False values in the respective columns.\\n\\n**', 'description': ' 4 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, generate separate UMAP visualizations highlighting the presence of each TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) by coloring the points based on the True/False values in the respective columns.\\n\\n**', 'output': ['/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_00-39-53/S4-HSC-reprogrammed-DE-UMAPs.h5ad', '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_00-39-53/S4-umap_TF_UMAP.png']}, 5: {'order': 5, 'module': 'CODE', 'prompt': '/force CODE  5 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, generate separate t-SNE visualizations highlighting the presence of each TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) by coloring the points based on the True/False values in the respective columns.\\n\\nThe removed ', 'description': ' 5 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, generate separate t-SNE visualizations highlighting the presence of each TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) by coloring the points based on the True/False values in the respective columns.\\n\\nThe removed ', 'output': []}}\n",
      "2024-08-03 00:44:54,845 - INFO - IP=5\n",
      "2024-08-03 00:44:54,846 - INFO - type(IP)=<class 'int'>\n",
      "2024-08-03 00:44:54,848 - INFO - \n",
      "\n",
      "PLANNER:\n",
      "\n",
      "{'module': 'CODE', 'steps': [{'llm': \"client=<openai.resources.chat.completions.Completions object at 0x1480cc33e8d0> async_client=<openai.resources.chat.completions.AsyncCompletions object at 0x1480cc324e50> model_name='gpt-4o' temperature=0.0 openai_api_key=SecretStr('**********') openai_proxy=''\", 'memory': 'None', 'prompt': 'input_variables=[\\'user_query\\'] template=\"You must select which code to run to help a user.\\\\n\\\\n**Available Scripts**\\\\nScript Name: scanpy_brad, \\\\t Description: This script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\\\\nScript Name: models_brad, \\\\t Description: This script executes a series of custom commands to build a model for predicting the contribution to reprogramming obtained by the inclusion or exclusion of individual TFs using an AnnData object or single cell datasets loaded from a .h5ad file.\\\\n\\\\n\\\\n**User Query**\\\\n{user_query}\\\\n\\\\n**Task**\\\\nBased on the user\\'s query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\\\\n\\\\n**Response Template**\\\\nSCRIPT: <selected script>\\\\nREASON: <reasoning why the script fits the user prompt>\\\\n\"', 'input': '5 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, generate separate t-SNE visualizations highlighting the presence of each TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) by coloring the points based on the True/False values in the respective columns.\\n\\nThe removed', 'output': 'content=\"SCRIPT: scanpy_brad\\\\nREASON: The user\\'s query involves generating t-SNE visualizations from an AnnData object loaded from a .h5ad file, which requires data manipulation and visualization capabilities. The `scanpy_brad` script is designed to handle such tasks, as it allows for the use of libraries like scanpy, seaborn, and matplotlib.pyplot to process .h5ad files and create visualizations. This makes it the best match for generating the requested t-SNE plots highlighting the presence of each specified transcription factor (TF).\" response_metadata={\\'token_usage\\': {\\'completion_tokens\\': 112, \\'prompt_tokens\\': 367, \\'total_tokens\\': 479}, \\'model_name\\': \\'gpt-4o-2024-05-13\\', \\'system_fingerprint\\': \\'fp_3cd8b62c3b\\', \\'finish_reason\\': \\'stop\\', \\'logprobs\\': None} id=\\'run-2e0a08f7-627c-47d3-a348-68bee5eeb0d3-0\\' usage_metadata={\\'input_tokens\\': 367, \\'output_tokens\\': 112, \\'total_tokens\\': 479}', 'parsedOutput': {'scriptName': '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py', 'scriptType': 'python', 'scriptPath': '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts'}, 'purpose': 'Select which code to run'}, {'llm': \"client=<openai.resources.chat.completions.Completions object at 0x1480cc33e8d0> async_client=<openai.resources.chat.completions.AsyncCompletions object at 0x1480cc324e50> model_name='gpt-4o' temperature=0.0 openai_api_key=SecretStr('**********') openai_proxy=''\", 'memory': 'None', 'prompt': 'input_variables=[\\'history\\', \\'input\\'] template=\\'Current conversation:\\\\n{history}\\\\n\\\\n**PYTHON SCRIPT**\\\\nYou must run this python script:\\\\n/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\\\n\\\\n**PYTHON SCRIPT DOCUMENTATION**:\\\\nThis is the doc string of this python script:\\\\n\"\"\"\\\\nThis script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\\\\n\\\\nArguments (four arguments):\\\\n    1. output directory: chatstatus[\\\\\\'output-directory\\\\\\']\\\\n    2. output file: <name of output file>\\\\n    3. input file: <file created in previous step>\\\\n    4. scanpy commands: a list of scanpy commands to be executed on the AnnData object (provided as a single string with commands separated by a delimiter, e.g., \\\\\\';\\\\\\')\\\\n\\\\nBased on the arguments, the input file will be loaded, then your commands will be executed, and finally, the output or resulting ann data object will be saved to the corrrectou output file and directory. Your code is not responsible for loading the .h5ad object, that will happen automatically, and when loaded, the object will be called adata. Your scanpy commands can operate directly on the adata object that will be loaded for you.\\\\n\\\\nAdditionally, the following imports are already provided for you and can be used in your code:\\\\n```\\\\nimport scanpy as sc\\\\nimport seaborn as sns\\\\nimport matplotlib.pyplot as plt\\\\nimport os\\\\nimport numpy as np\\\\nimport pandas as pd\\\\nfrom scipy.spatial.distance import cdist\\\\n```\\\\n\\\\n**Usage**\\\\nBRAD Line:\\\\n```\\\\nsubprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\\\\\'output-directory\\\\\\'], <output file>, <input file>, \"<scanpy commands>\"], capture_output=True, text=True)\\\\n```\\\\n\\\\n**Examples**\\\\nUse the below examples to help generate your code.\\\\n\\\\n*Example 1*\\\\nUser Prompt: Run scanpy preprocessing and UMAP visualization on XXX.h5ad and save the UMAP plot\\\\nResponse Code:\\\\n```\\\\nresponse = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\\\\\'output-directory\\\\\\'], \"XXX-modified.h5ad\", \"<path/to/data>/XXX.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save=\\\\\\'umap.png\\\\\\')\"], capture_output=True, text=True)\\\\n```\\\\nExplination: the adata object will be loaded in memory already. The command \"sc.pp.neighbors(adata)\" will preprocess the data, then the command \"sc.tl.umap(adata)\" will perform UMAP, and finally the command \"sc.pl.umap(adata, \\\\\\'umap.png\\\\\\')\" will save the UMAP to a well named file.\\\\n\\\\n*Example 2*\\\\nUser Prompt: Perform PCA and clustering on the dataset YYY.h5ad and save the PCA plot\\\\nResponse Code:\\\\n```\\\\nresponse = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\\\\\'output-directory\\\\\\'], \"YYY-modified.h5ad\", \"<path/to/data>/YYY.h5ad\", \"sc.pp.pca(adata); sc.tl.leiden(adata); sc.pl.pca(adata, save=\\\\\\'pca.png\\\\\\')\"], capture_output=True, text=True)\\\\n```\\\\nExplination: the adata object will be loaded in memory already. The command \"sc.pp.pca(adata)\" will preprocess the data, then the command \"sc.tl.leiden(adata)\" will perform the leiden algorithm, and finally the command \"sc.pl.pca(adata, save=\\\\\\'pca.png\\\\\\')\" will save the PCA to a well named file.\\\\n\\\\n*Example 3*\\\\nUser Prompt: compute the distance between all cells that are not source or target to all source and target cells.\\\\nReponse Code:\\\\n```\\\\nadata_X = adata.X\\\\nmask = ~adata.obs[\\\\\\'recipe\\\\\\'].isin([\\\\\\'source\\\\\\', \\\\\\'target\\\\\\'])\\\\nrecipe_embeddings = adata[mask]\\\\nsource_embeddings = adata[adata.obs[\\\\\\'recipe\\\\\\'] == \\\\\\'source\\\\\\']\\\\ntarget_embeddings = adata[adata.obs[\\\\\\'recipe\\\\\\'] == \\\\\\'target\\\\\\']\\\\nrecipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X)\\\\nrecipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X)\\\\nadata_dist = sc.AnnData(X=np.concatenate((recipe_distances_source, recipe_distances_target), axis=1), var=pd.concat((source_embeddings.obs, target_embeddings.obs)), obs=recipe_embeddings.obs)\\\\nadata_dist.write(<file name>.h5ad)\\\\n```\\\\n\\\\n**OUTPUT FILE NAME INSTRUCTIONS**\\\\n1. Output path should be chatstatus[\\\\\\'output-directory\\\\\\']\\\\n2. Output file name should be `<descriptive name>.h5ad`\\\\n\"\"\"\\\\n\\\\n\\\\n**CALL PYTHON SCRIPTS FROM PYTHON**:\\\\nUse the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\\\\n\\\\nTo call a Python script `example_script.py` which has no arguments:\\\\n\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\']], capture_output=True, text=True)\\\\n```\\\\n\\\\nTo call a Python script `example_script.py` with one argument:\\\\n\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\'], \\\\\\'arg1\\\\\\'], capture_output=True, text=True)\\\\n```\\\\n\\\\nTo call a Python script `example_script.py` with two arguments:\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\'], \\\\\\'arg1\\\\\\', \\\\\\'arg2\\\\\\'], capture_output=True, text=True)\\\\n```\\\\n\\\\nNote that chatstatus[\\\\\\'output-directory\\\\\\'] is ALWAYS passed as the first argument.\\\\n\\\\nThe following files were previously created by BRAD and could be used as input to a function if necessary:\\\\nS1-HSC-reprogrammed-DE-summary.h5ad\\\\nlog.json\\\\nS4-HSC-reprogrammed-DE-UMAPs.h5ad\\\\nS2-umapumap_2d.png\\\\nS4-umap_TF_UMAP.png\\\\nS3-tsnetsne_2d.png\\\\n\\\\nQuery:{input}\\\\n\\\\n**PYTHON SCRIPT**\\\\nPlease use the correct script name and path. You must run this python script:\\\\n/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\\\n\\\\n**INSTRUCTIONS**\\\\n1. Given the user query and the documentation, identify each of the arguments found in the user\\\\\\'s query that should be passed to the Python script.\\\\n2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\\\\n3. The last line of your response should say \"Execute: <Python code to execute>\"\\\\n4. Do not include any extra words or characters. Format the response/output as:\\\\n    Arguments: \\\\n    Python Code Explanation: <2 sentences maximum>\\\\n    Execute: <your code here>\\\\n\\\\n**IMPORTANT**\\\\nThe code to execute from your response must be formatted as:\\\\n    Execute: subprocess.call([sys.executable, \\\\\\'<path to python script>\\\\\\', \\\\\\'<argument 1>\\\\\\', \\\\\\'<argument 2>\\\\\\', ..., \\\\\\'<argument n>\\\\\\'], capture_output=True, text=True))\\\\nThis output should be exactly one line and no longer. Stop the response after this line.\\\\n\\'', 'input': '5 (CODE)**:\\nPrompt: Using the loaded anndata object from `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`, generate separate t-SNE visualizations highlighting the presence of each TF (REL, STAT5A, FOS, ETV6, MEF2C, RUNX3, FOSB, GATA2, IKZF1, GFI1B) by coloring the points based on the True/False values in the respective columns.\\n\\nThe removed', 'output': 'Arguments:\\n1. chatstatus[\\'output-directory\\']\\n2. \"HSC-reprogrammed-DE-tSNE.h5ad\"\\n3. \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\"\\n4. \"sc.tl.tsne(adata); sc.pl.tsne(adata, color=[\\'REL\\', \\'STAT5A\\', \\'FOS\\', \\'ETV6\\', \\'MEF2C\\', \\'RUNX3\\', \\'FOSB\\', \\'GATA2\\', \\'IKZF1\\', \\'GFI1B\\'], save=\\'tSNE_TFs.png\\')\"\\n\\nPython Code Explanation: This code will load the AnnData object, perform t-SNE, and generate visualizations highlighting the presence of each specified TF, saving the plot as \\'tSNE_TFs.png\\'.\\n\\nExecute: subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"HSC-reprogrammed-DE-tSNE.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.tl.tsne(adata); sc.pl.tsne(adata, color=[\\'REL\\', \\'STAT5A\\', \\'FOS\\', \\'ETV6\\', \\'MEF2C\\', \\'RUNX3\\', \\'FOSB\\', \\'GATA2\\', \\'IKZF1\\', \\'GFI1B\\'], save=\\'tSNE_TFs.png\\')\"], capture_output=True, text=True)', 'parsedOutput': {'code': 'subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"HSC-reprogrammed-DE-tSNE.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.tl.tsne(adata); sc.pl.tsne(adata, color=[\\'REL\\', \\'STAT5A\\', \\'FOS\\', \\'ETV6\\', \\'MEF2C\\', \\'RUNX3\\', \\'FOSB\\', \\'GATA2\\', \\'IKZF1\\', \\'GFI1B\\'], save=\\'tSNE_TFs.png\\')\"], capture_output=True, text=True)'}, 'purpose': 'Format function call'}, {'func': 'pythonCaller.execute_python_code', 'code': 'response = subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"HSC-reprogrammed-DE-tSNE.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.tl.tsne(adata); sc.pl.tsne(adata, color=[\\'REL\\', \\'STAT5A\\', \\'FOS\\', \\'ETV6\\', \\'MEF2C\\', \\'RUNX3\\', \\'FOSB\\', \\'GATA2\\', \\'IKZF1\\', \\'GFI1B\\'], save=\\'tSNE_TFs.png\\')\"], capture_output=True, text=True)', 'purpose': 'execute python code'}, {'func': 'utils.makeNamesConsistent', 'files': [{'old-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_00-39-53/HSC-reprogrammed-DE-tSNE.h5ad', 'new-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_00-39-53/S5-HSC-reprogrammed-DE-tSNE.h5ad'}, {'old-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_00-39-53/tsnetSNE_TFs.png', 'new-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_00-39-53/S5-tsnetSNE_TFs.png'}]}]}\n",
      "2024-08-03 00:44:54,852 - INFO - current llm calls=11\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  exit\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Thanks for chatting today! I hope to talk soon, and don't forget that a record of this conversation is available at: /home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_00-39-53/log.json\n"
     ]
    }
   ],
   "source": [
    "# /force PLANNER please load and run the BOT1 pipeline\n",
    "BOT1.chat() # worker_system_prompt + instructionSets)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1c9c481-fcd3-415f-8a2d-2d06c4703b78",
   "metadata": {},
   "source": [
    "## BOT 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "66ecdebc-011e-4f96-a11d-9461c7121855",
   "metadata": {},
   "outputs": [],
   "source": [
    "modified_instructions = \"\"\"BOT2: Distance Computation\n",
    "**Objective:** Compute the distance between each observation representing a recipe and each observation representing a source or target cell.\n",
    "\n",
    "**Instructions:**\n",
    "1. Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`.\n",
    "2. Extract the embeddings for 'source', 'target', and each recipe, based on the 'recipe' column in adata.obs, from `adata.X`\n",
    "3. Compute the pairwise Euclidean distances between:\n",
    "   - Each 'source' embedding and each recipe embedding.\n",
    "   - Each 'target' embedding and each recipe embedding.\n",
    "4. Store the computed distances in a matrix format and save the results in an ann data object. Set\n",
    "   - .var of the new ann data to be the .obs of the source and target cells used in the distance calculations\n",
    "   - .obs of the new ann data to be the .obs of the reprogrammed cells used in the distance calculations, i.e. where 'recipe' is not 'target' and is not 'source'\n",
    "5. explicitly save the new ann data object\n",
    "6. Provide a summary of the distance metrics, highlighting which recipes are closest to the 'target' cells.\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "678849e8-3054-4648-aa0d-8fa45d4b0ca4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Would you like to use a database with BRAD [Y/N]?\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      " N\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32m2024-08-03 00:52:44 INFO semantic_router.utils.logger local\u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome to RAG! The chat log from this conversation will be saved to /home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_00-52-37/log.json. How can I help?\n"
     ]
    }
   ],
   "source": [
    "# Make the plan\n",
    "BOT2 = brad.chatbot(llm=llms.load_nvidia(), config='/home/jpic/RAG-DEV/tutorials/geneformer/config-sc.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "df4bdfdd-c45f-4b12-ad76-159e6eb63665",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 00:52:58,774 - INFO - PLANNER\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "BRAD >> 0: \n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3m**INSTRUCTIONS:**\n",
      "You are planning a bioinformatics analysis pipeline to address a user's query. You must determine if there already exist a predesigned pipeline that will respond to a user's query, or if a new, custom process must be designed for this particular query. The multi-step workflow can use the available methods listed below.\n",
      "\n",
      "**Available Methods within a Pipeline:**\n",
      "1. **RAG**: Look up literature and documents from a text database.\n",
      "2. **SCRAPE**: Search platforms like arXiv, bioRxiv, and PubMed for the latest research.\n",
      "3. **DATABASE**: Utilize bioinformatics databases such as Gene Ontology and Enrichr to perform gene set enrichment analyses.\n",
      "4. **CODE**: Execute bioinformatics pipelines. Utilize prebuilt pipelines or develop new ones as needed.\n",
      "5. **WRITE**: Synthesize and summarize information. This can include summarizing database searches, code pipeline results, or creating a final report to encapsulate the entire analysis.\n",
      "6. **ROUTER**: Determine which step we should proceed to next.\n",
      "\n",
      "User Query: /force PLANNER You are a chatbot assisting to help analyze a single cell dataset.\n",
      "You have  instructions below, and you must use your tools to answer the questions. You have one tool that allows\n",
      "you to write CODE. Generate a plan based on the instructions to use the CODE tool several times in order to satisfy\n",
      "your instructions. Each call to the CODE tool should have a single goal, such as making one type of plot, builting\n",
      "one type of model, and so forth.\n",
      "\n",
      "If your job is to make plots, then you can make a plan to make multiple kinds of plots with each different call\n",
      "to CODE in your plan. For instance. your plan could be\n",
      "\n",
      "**Step 1, CODE** load <path to data>.h5ad and make umap plots of ...\n",
      "**Step 2, CODE** load <path to data>.h5ad and make tsne plots of ...\n",
      "\n",
      "BOT2: Distance Computation\n",
      "**Objective:** Compute the distance between each observation representing a recipe and each observation representing a source or target cell.\n",
      "\n",
      "**Instructions:**\n",
      "1. Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`.\n",
      "2. Extract the embeddings for 'source', 'target', and each recipe, based on the 'recipe' column in adata.obs, from `adata.X`\n",
      "3. Compute the pairwise Euclidean distances between:\n",
      "   - Each 'source' embedding and each recipe embedding.\n",
      "   - Each 'target' embedding and each recipe embedding.\n",
      "4. Store the computed distances in a matrix format and save the results in an ann data object. Set\n",
      "   - .var of the new ann data to be the .obs of the source and target cells used in the distance calculations\n",
      "   - .obs of the new ann data to be the .obs of the reprogrammed cells used in the distance calculations, i.e. where 'recipe' is not 'target' and is not 'source'\n",
      "5. explicitly save the new ann data object\n",
      "6. Provide a summary of the distance metrics, highlighting which recipes are closest to the 'target' cells.\n",
      "\n",
      "**Available Pipelines**\n",
      "Name: BOT1\tDescription: this plan is for basic visualization and summary of single cell data from the Geneformer reprogramming experiment. The data can be loaded in from ann data objects\n",
      "\n",
      "\n",
      "CUSTOM: The option to build a new pipeline\n",
      "\n",
      "The custom pipeline should be selected if no current pipeline addreses the user's query. If the custom pipeline is selected, please do not attempt to design the pipeline now. We will work to design it at a later point.\n",
      "\n",
      "**Output**:\n",
      "Please formate your output as follows:\n",
      "Pipeline Name: <available pipeline name above or \"CUSTOM\">\n",
      "Explination: <reason for selecting the above pipeline>\n",
      "\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 00:53:00,317 - INFO - Pipeline Name: CUSTOM\n",
      "Explanation: The user's query requires a custom pipeline because none of the available pipelines, including BOT1, address the specific objective of computing the distance between each observation representing a recipe and each observation representing a source or target cell. The query requires a unique combination of steps involving loading anndata objects, extracting embeddings, computing pairwise Euclidean distances, storing results in a matrix format, and providing a summary of distance metrics. A custom pipeline needs to be designed to fulfill this specific requirement.\n",
      "2024-08-03 00:53:00,318 - INFO - selected_pipeline=CUSTOM\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3m**INSTRUCTIONS:**\n",
      "You are planning a bioinformatics analysis pipeline to address a user's query. Your task is to outline a multi-step workflow using the available methods listed below. Each method should be used appropriately to ensure a thorough analysis. For each step in the pipeline, explain the goal of the process as well as provide a prompt to the chatbot that will execute that step. If information is passed between stages, such as from literature or code to databases, indicate the dependencies of steps in the prompt.\n",
      "\n",
      "**Available Methods:**\n",
      "1. **RAG**: Look up literature and documents from a text database.\n",
      "2. **SCRAPE**: Search platforms like arXiv, bioRxiv, and PubMed for the latest research.\n",
      "3. **DATABASE**: Utilize bioinformatics databases such as Gene Ontology and Enrichr to perform gene set enrichment analyses.\n",
      "4. **CODE**: Execute bioinformatics pipelines. Utilize prebuilt pipelines or develop new ones as needed.\n",
      "5. **WRITE**: Synthesize and summarize information. This can include summarizing database searches, code pipeline results, or creating a final report to encapsulate the entire analysis.\n",
      "6. **ROUTER**: Determine which step we should proceed to next.\n",
      "\n",
      "Current conversation: \n",
      "\n",
      "\n",
      "User Query: /force PLANNER You are a chatbot assisting to help analyze a single cell dataset.\n",
      "You have  instructions below, and you must use your tools to answer the questions. You have one tool that allows\n",
      "you to write CODE. Generate a plan based on the instructions to use the CODE tool several times in order to satisfy\n",
      "your instructions. Each call to the CODE tool should have a single goal, such as making one type of plot, builting\n",
      "one type of model, and so forth.\n",
      "\n",
      "If your job is to make plots, then you can make a plan to make multiple kinds of plots with each different call\n",
      "to CODE in your plan. For instance. your plan could be\n",
      "\n",
      "**Step 1, CODE** load <path to data>.h5ad and make umap plots of ...\n",
      "**Step 2, CODE** load <path to data>.h5ad and make tsne plots of ...\n",
      "\n",
      "BOT2: Distance Computation\n",
      "**Objective:** Compute the distance between each observation representing a recipe and each observation representing a source or target cell.\n",
      "\n",
      "**Instructions:**\n",
      "1. Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`.\n",
      "2. Extract the embeddings for 'source', 'target', and each recipe, based on the 'recipe' column in adata.obs, from `adata.X`\n",
      "3. Compute the pairwise Euclidean distances between:\n",
      "   - Each 'source' embedding and each recipe embedding.\n",
      "   - Each 'target' embedding and each recipe embedding.\n",
      "4. Store the computed distances in a matrix format and save the results in an ann data object. Set\n",
      "   - .var of the new ann data to be the .obs of the source and target cells used in the distance calculations\n",
      "   - .obs of the new ann data to be the .obs of the reprogrammed cells used in the distance calculations, i.e. where 'recipe' is not 'target' and is not 'source'\n",
      "5. explicitly save the new ann data object\n",
      "6. Provide a summary of the distance metrics, highlighting which recipes are closest to the 'target' cells.\n",
      "\n",
      "Plan:\n",
      "**Step 1 (method, eg. RAG)**:\n",
      "Prompt: [Description of action for chatbot to do Step 1, i.e. do X, Y, Z]\n",
      "\n",
      "**Step 2 (method, eg. CODE)**:\n",
      "Prompt: [Description of action for chatbot to do Step 2, i.e. do A, B, C. Use information from Step X]\n",
      "...\n",
      "\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "Here is the plan to address the user's query:\n",
      "\n",
      "**Step 1 (ROUTER)**:\n",
      "Prompt: Determine which step to proceed to next, considering the objective of computing distances between observations representing recipes and source or target cells.\n",
      "\n",
      "**Step 2 (CODE)**:\n",
      "Prompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings for 'source', 'target', and each recipe from `adata.X`.\n",
      "\n",
      "**Step 3 (CODE)**:\n",
      "Prompt: Compute the pairwise Euclidean distances between each 'source' embedding and each recipe embedding, using the extracted embeddings from Step 2.\n",
      "\n",
      "**Step 4 (CODE)**:\n",
      "Prompt: Compute the pairwise Euclidean distances between each 'target' embedding and each recipe embedding, using the extracted embeddings from Step 2.\n",
      "\n",
      "**Step 5 (CODE)**:\n",
      "Prompt: Store the computed distances in a matrix format and create a new ann data object, setting `.var` to be the `.obs` of the source and target cells used in the distance calculations and `.obs` to be the `.obs` of the reprogrammed cells used in the distance calculations. Explicitly save the new ann data object.\n",
      "\n",
      "**Step 6 (WRITE)**:\n",
      "Prompt: Provide a summary of the distance metrics, highlighting which recipes are closest to the 'target' cells, using the results from Step 5.\n",
      "\n",
      "**Step 7 (ROUTER)**:\n",
      "Prompt: Determine the next step, considering the results from Step 6 and the overall objective of the analysis.\n",
      "\n",
      "Note: The dependencies between steps are:\n",
      "\n",
      "* Step 3 and Step 4 rely on the extracted embeddings from Step 2.\n",
      "* Step 5 uses the results from Step 3 and Step 4.\n",
      "* Step 6 relies on the new ann data object created in Step 5.\n",
      "\n",
      "\n",
      "Do you want to proceed with this plan? [Y/N/edit]\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  please remove steps 1, 6, and 7. Also, please provide instructions for figure generation within each stage or step\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 00:53:43,524 - INFO - Based on the most recently proposed Current Plan and the user's new requirements and requested changes, create a revised plan.\n",
      "\n",
      "Current Plan: Here is the plan to address the user's query:\n",
      "\n",
      "**Step 1 (ROUTER)**:\n",
      "Prompt: Determine which step to proceed to next, considering the objective of computing distances between observations representing recipes and source or target cells.\n",
      "\n",
      "**Step 2 (CODE)**:\n",
      "Prompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings for 'source', 'target', and each recipe from `adata.X`.\n",
      "\n",
      "**Step 3 (CODE)**:\n",
      "Prompt: Compute the pairwise Euclidean distances between each 'source' embedding and each recipe embedding, using the extracted embeddings from Step 2.\n",
      "\n",
      "**Step 4 (CODE)**:\n",
      "Prompt: Compute the pairwise Euclidean distances between each 'target' embedding and each recipe embedding, using the extracted embeddings from Step 2.\n",
      "\n",
      "**Step 5 (CODE)**:\n",
      "Prompt: Store the computed distances in a matrix format and create a new ann data object, setting `.var` to be the `.obs` of the source and target cells used in the distance calculations and `.obs` to be the `.obs` of the reprogrammed cells used in the distance calculations. Explicitly save the new ann data object.\n",
      "\n",
      "**Step 6 (WRITE)**:\n",
      "Prompt: Provide a summary of the distance metrics, highlighting which recipes are closest to the 'target' cells, using the results from Step 5.\n",
      "\n",
      "**Step 7 (ROUTER)**:\n",
      "Prompt: Determine the next step, considering the results from Step 6 and the overall objective of the analysis.\n",
      "\n",
      "Note: The dependencies between steps are:\n",
      "\n",
      "* Step 3 and Step 4 rely on the extracted embeddings from Step 2.\n",
      "* Step 5 uses the results from Step 3 and Step 4.\n",
      "* Step 6 relies on the new ann data object created in Step 5.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "User's requested changes: {input}\n",
      "\n",
      "**Available Methods:**\n",
      "1. **Retrieval Augmented Generation**(RAG): Look up literature and documents from a text database.\n",
      "2. **Web Search for New Literature**(SCRAPE): Search platforms like arXiv, bioRxiv, and PubMed for the latest research.\n",
      "3. **Bioinformatics Databases**(DATABASE): Utilize databases such as Gene Ontology and Enrichr to perform gene set enrichment analyses.\n",
      "4. **Run Codes**(CODE): Execute bioinformatics pipelines. Utilize prebuilt pipelines or develop new ones as needed.\n",
      "5. **Write a Report**(WRITE): Synthesize and summarize information. This can include summarizing database searches, code pipeline results, or creating a final report to encapsulate the entire analysis.\n",
      "\n",
      "**OUTPUT**\n",
      "Revised Plan:\n",
      "**Step 1 (method, eg. RAG)**:\n",
      "Prompt: [Description of action for chatbot to do Step 1, i.e. do X, Y, Z]\n",
      "\n",
      "**Step 2 (method, eg. CODE)**:\n",
      "Prompt: [Description of action for chatbot to do Step 2, i.e. do X, Y, Z]\n",
      "...\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Here is the revised plan based on the user's new requirements and requested changes:\n",
      "\n",
      "**Revised Plan:**\n",
      "\n",
      "**Step 1 (CODE)**:\n",
      "Prompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings for 'source', 'target', and each recipe from `adata.X`. **Generate a UMAP plot to visualize the embeddings.**\n",
      "\n",
      "**Step 2 (CODE)**:\n",
      "Prompt: Compute the pairwise Euclidean distances between each 'source' embedding and each recipe embedding, using the extracted embeddings from Step 1. **Generate a heatmap to visualize the distance matrix.**\n",
      "\n",
      "**Step 3 (CODE)**:\n",
      "Prompt: Compute the pairwise Euclidean distances between each 'target' embedding and each recipe embedding, using the extracted embeddings from Step 1. **Generate a heatmap to visualize the distance matrix.**\n",
      "\n",
      "**Step 4 (CODE)**:\n",
      "Prompt: Store the computed distances in a matrix format and create a new ann data object, setting `.var` to be the `.obs` of the source and target cells used in the distance calculations and `.obs` to be the `.obs` of the reprogrammed cells used in the distance calculations. Explicitly save the new ann data object. **Generate a summary plot to visualize the new ann data object.**\n",
      "\n",
      "The revised plan removes Steps 1, 6, and 7 as per the user's request and adds instructions for figure generation within each step. The available methods have been assigned to each step accordingly.\n",
      "\n",
      "\n",
      "Do you want to proceed with this plan? [Y/N/edit]\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  Y\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 00:54:15,085 - INFO - [{'order': 0, 'module': 'PLANNER', 'prompt': None, 'description': 'This step designed the plan. It is placed in the queue because we needed a place holder for 0 indexed lists.'}, {'order': 1, 'module': 'CODE', 'prompt': \"/force CODE  1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings for 'source', 'target', and each recipe from `adata.X`. **Generate a UMAP plot to visualize the embeddings.**\\n\\n**\", 'description': \" 1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings for 'source', 'target', and each recipe from `adata.X`. **Generate a UMAP plot to visualize the embeddings.**\\n\\n**\"}, {'order': 2, 'module': 'CODE', 'prompt': \"/force CODE  2 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'source' embedding and each recipe embedding, using the extracted embeddings from \", 'description': \" 2 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'source' embedding and each recipe embedding, using the extracted embeddings from \"}, {'order': 4, 'module': 'CODE', 'prompt': \"/force CODE  3 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'target' embedding and each recipe embedding, using the extracted embeddings from \", 'description': \" 3 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'target' embedding and each recipe embedding, using the extracted embeddings from \"}, {'order': 6, 'module': 'CODE', 'prompt': '/force CODE  4 (CODE)**:\\nPrompt: Store the computed distances in a matrix format and create a new ann data object, setting `.var` to be the `.obs` of the source and target cells used in the distance calculations and `.obs` to be the `.obs` of the reprogrammed cells used in the distance calculations. Explicitly save the new ann data object. **Generate a summary plot to visualize the new ann data object.**\\n\\nThe revised plan removes ', 'description': ' 4 (CODE)**:\\nPrompt: Store the computed distances in a matrix format and create a new ann data object, setting `.var` to be the `.obs` of the source and target cells used in the distance calculations and `.obs` to be the `.obs` of the reprogrammed cells used in the distance calculations. Explicitly save the new ann data object. **Generate a summary plot to visualize the new ann data object.**\\n\\nThe revised plan removes '}]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[\"Here is the revised plan based on the user's new requirements and requested changes:\\n\\n**Revised Plan:**\\n\\n**\", \" 1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings for 'source', 'target', and each recipe from `adata.X`. **Generate a UMAP plot to visualize the embeddings.**\\n\\n**\", \" 2 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'source' embedding and each recipe embedding, using the extracted embeddings from \", ' 1. **Generate a heatmap to visualize the distance matrix.**\\n\\n**', \" 3 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'target' embedding and each recipe embedding, using the extracted embeddings from \", ' 1. **Generate a heatmap to visualize the distance matrix.**\\n\\n**', ' 4 (CODE)**:\\nPrompt: Store the computed distances in a matrix format and create a new ann data object, setting `.var` to be the `.obs` of the source and target cells used in the distance calculations and `.obs` to be the `.obs` of the reprogrammed cells used in the distance calculations. Explicitly save the new ann data object. **Generate a summary plot to visualize the new ann data object.**\\n\\nThe revised plan removes ', \"s 1, 6, and 7 as per the user's request and adds instructions for figure generation within each step. The available methods have been assigned to each step accordingly.\\n\\n\"]\n",
      "Would you like to save this pipeline to a file? [Y/N]\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      ">>> Y\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter a name for your pipeline\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      ">>> BOT2\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter a description of your pipeline\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      ">>> this pipeline investigates distances between source, target, and reprogrammed cells. BOT2 should use this pipeline\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 00:56:05,081 - INFO - \n",
      "\n",
      "\n",
      "route\n",
      "\n",
      "\n",
      "\n",
      "2024-08-03 00:56:05,082 - INFO - PLANNER\n",
      "2024-08-03 00:56:05,082 - INFO - \n",
      "\n",
      "PLANNER:\n",
      "\n",
      "{'module': 'PLANNER', 'steps': [{'llm': \"model='meta/llama3-70b-instruct'\", 'memory': 'None', 'prompt': 'input_variables=[\\'input\\'] template=\\'**INSTRUCTIONS:**\\\\nYou are planning a bioinformatics analysis pipeline to address a user\\\\\\'s query. You must determine if there already exist a predesigned pipeline that will respond to a user\\\\\\'s query, or if a new, custom process must be designed for this particular query. The multi-step workflow can use the available methods listed below.\\\\n\\\\n**Available Methods within a Pipeline:**\\\\n1. **RAG**: Look up literature and documents from a text database.\\\\n2. **SCRAPE**: Search platforms like arXiv, bioRxiv, and PubMed for the latest research.\\\\n3. **DATABASE**: Utilize bioinformatics databases such as Gene Ontology and Enrichr to perform gene set enrichment analyses.\\\\n4. **CODE**: Execute bioinformatics pipelines. Utilize prebuilt pipelines or develop new ones as needed.\\\\n5. **WRITE**: Synthesize and summarize information. This can include summarizing database searches, code pipeline results, or creating a final report to encapsulate the entire analysis.\\\\n6. **ROUTER**: Determine which step we should proceed to next.\\\\n\\\\nUser Query: {input}\\\\n\\\\n**Available Pipelines**\\\\nName: BOT1\\\\tDescription: this plan is for basic visualization and summary of single cell data from the Geneformer reprogramming experiment. The data can be loaded in from ann data objects\\\\n\\\\n\\\\nCUSTOM: The option to build a new pipeline\\\\n\\\\nThe custom pipeline should be selected if no current pipeline addreses the user\\\\\\'s query. If the custom pipeline is selected, please do not attempt to design the pipeline now. We will work to design it at a later point.\\\\n\\\\n**Output**:\\\\nPlease formate your output as follows:\\\\nPipeline Name: <available pipeline name above or \"CUSTOM\">\\\\nExplination: <reason for selecting the above pipeline>\\\\n\\'', 'input': \"/force PLANNER You are a chatbot assisting to help analyze a single cell dataset.\\nYou have  instructions below, and you must use your tools to answer the questions. You have one tool that allows\\nyou to write CODE. Generate a plan based on the instructions to use the CODE tool several times in order to satisfy\\nyour instructions. Each call to the CODE tool should have a single goal, such as making one type of plot, builting\\none type of model, and so forth.\\n\\nIf your job is to make plots, then you can make a plan to make multiple kinds of plots with each different call\\nto CODE in your plan. For instance. your plan could be\\n\\n**Step 1, CODE** load <path to data>.h5ad and make umap plots of ...\\n**Step 2, CODE** load <path to data>.h5ad and make tsne plots of ...\\n\\nBOT2: Distance Computation\\n**Objective:** Compute the distance between each observation representing a recipe and each observation representing a source or target cell.\\n\\n**Instructions:**\\n1. Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`.\\n2. Extract the embeddings for 'source', 'target', and each recipe, based on the 'recipe' column in adata.obs, from `adata.X`\\n3. Compute the pairwise Euclidean distances between:\\n   - Each 'source' embedding and each recipe embedding.\\n   - Each 'target' embedding and each recipe embedding.\\n4. Store the computed distances in a matrix format and save the results in an ann data object. Set\\n   - .var of the new ann data to be the .obs of the source and target cells used in the distance calculations\\n   - .obs of the new ann data to be the .obs of the reprogrammed cells used in the distance calculations, i.e. where 'recipe' is not 'target' and is not 'source'\\n5. explicitly save the new ann data object\\n6. Provide a summary of the distance metrics, highlighting which recipes are closest to the 'target' cells.\", 'output': \"Pipeline Name: CUSTOM\\nExplanation: The user's query requires a custom pipeline because none of the available pipelines, including BOT1, address the specific objective of computing the distance between each observation representing a recipe and each observation representing a source or target cell. The query requires a unique combination of steps involving loading anndata objects, extracting embeddings, computing pairwise Euclidean distances, storing results in a matrix format, and providing a summary of distance metrics. A custom pipeline needs to be designed to fulfill this specific requirement.\", 'parsedOutput': {'selected pipeline': 'CUSTOM'}, 'purpose': 'determine if a known pipeline can be used or a new one si required'}, {'llm': \"model='meta/llama3-70b-instruct'\", 'memory': 'chat_memory=InMemoryChatMessageHistory(messages=[HumanMessage(content=\"/force PLANNER You are a chatbot assisting to help analyze a single cell dataset.\\\\nYou have  instructions below, and you must use your tools to answer the questions. You have one tool that allows\\\\nyou to write CODE. Generate a plan based on the instructions to use the CODE tool several times in order to satisfy\\\\nyour instructions. Each call to the CODE tool should have a single goal, such as making one type of plot, builting\\\\none type of model, and so forth.\\\\n\\\\nIf your job is to make plots, then you can make a plan to make multiple kinds of plots with each different call\\\\nto CODE in your plan. For instance. your plan could be\\\\n\\\\n**Step 1, CODE** load <path to data>.h5ad and make umap plots of ...\\\\n**Step 2, CODE** load <path to data>.h5ad and make tsne plots of ...\\\\n\\\\nBOT2: Distance Computation\\\\n**Objective:** Compute the distance between each observation representing a recipe and each observation representing a source or target cell.\\\\n\\\\n**Instructions:**\\\\n1. Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`.\\\\n2. Extract the embeddings for \\'source\\', \\'target\\', and each recipe, based on the \\'recipe\\' column in adata.obs, from `adata.X`\\\\n3. Compute the pairwise Euclidean distances between:\\\\n   - Each \\'source\\' embedding and each recipe embedding.\\\\n   - Each \\'target\\' embedding and each recipe embedding.\\\\n4. Store the computed distances in a matrix format and save the results in an ann data object. Set\\\\n   - .var of the new ann data to be the .obs of the source and target cells used in the distance calculations\\\\n   - .obs of the new ann data to be the .obs of the reprogrammed cells used in the distance calculations, i.e. where \\'recipe\\' is not \\'target\\' and is not \\'source\\'\\\\n5. explicitly save the new ann data object\\\\n6. Provide a summary of the distance metrics, highlighting which recipes are closest to the \\'target\\' cells.\"), AIMessage(content=\"Here is the plan to address the user\\'s query:\\\\n\\\\n**Step 1 (ROUTER)**:\\\\nPrompt: Determine which step to proceed to next, considering the objective of computing distances between observations representing recipes and source or target cells.\\\\n\\\\n**Step 2 (CODE)**:\\\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings for \\'source\\', \\'target\\', and each recipe from `adata.X`.\\\\n\\\\n**Step 3 (CODE)**:\\\\nPrompt: Compute the pairwise Euclidean distances between each \\'source\\' embedding and each recipe embedding, using the extracted embeddings from Step 2.\\\\n\\\\n**Step 4 (CODE)**:\\\\nPrompt: Compute the pairwise Euclidean distances between each \\'target\\' embedding and each recipe embedding, using the extracted embeddings from Step 2.\\\\n\\\\n**Step 5 (CODE)**:\\\\nPrompt: Store the computed distances in a matrix format and create a new ann data object, setting `.var` to be the `.obs` of the source and target cells used in the distance calculations and `.obs` to be the `.obs` of the reprogrammed cells used in the distance calculations. Explicitly save the new ann data object.\\\\n\\\\n**Step 6 (WRITE)**:\\\\nPrompt: Provide a summary of the distance metrics, highlighting which recipes are closest to the \\'target\\' cells, using the results from Step 5.\\\\n\\\\n**Step 7 (ROUTER)**:\\\\nPrompt: Determine the next step, considering the results from Step 6 and the overall objective of the analysis.\\\\n\\\\nNote: The dependencies between steps are:\\\\n\\\\n* Step 3 and Step 4 rely on the extracted embeddings from Step 2.\\\\n* Step 5 uses the results from Step 3 and Step 4.\\\\n* Step 6 relies on the new ann data object created in Step 5.\")]) ai_prefix=\\'BRAD\\'', 'prompt': 'input_variables=[\\'history\\', \\'input\\'] template=\"**INSTRUCTIONS:**\\\\nYou are planning a bioinformatics analysis pipeline to address a user\\'s query. Your task is to outline a multi-step workflow using the available methods listed below. Each method should be used appropriately to ensure a thorough analysis. For each step in the pipeline, explain the goal of the process as well as provide a prompt to the chatbot that will execute that step. If information is passed between stages, such as from literature or code to databases, indicate the dependencies of steps in the prompt.\\\\n\\\\n**Available Methods:**\\\\n1. **RAG**: Look up literature and documents from a text database.\\\\n2. **SCRAPE**: Search platforms like arXiv, bioRxiv, and PubMed for the latest research.\\\\n3. **DATABASE**: Utilize bioinformatics databases such as Gene Ontology and Enrichr to perform gene set enrichment analyses.\\\\n4. **CODE**: Execute bioinformatics pipelines. Utilize prebuilt pipelines or develop new ones as needed.\\\\n5. **WRITE**: Synthesize and summarize information. This can include summarizing database searches, code pipeline results, or creating a final report to encapsulate the entire analysis.\\\\n6. **ROUTER**: Determine which step we should proceed to next.\\\\n\\\\nCurrent conversation: {history}\\\\n\\\\n\\\\nUser Query: {input}\\\\n\\\\nPlan:\\\\n**Step 1 (method, eg. RAG)**:\\\\nPrompt: [Description of action for chatbot to do Step 1, i.e. do X, Y, Z]\\\\n\\\\n**Step 2 (method, eg. CODE)**:\\\\nPrompt: [Description of action for chatbot to do Step 2, i.e. do A, B, C. Use information from Step X]\\\\n...\\\\n\"', 'input': \"/force PLANNER You are a chatbot assisting to help analyze a single cell dataset.\\nYou have  instructions below, and you must use your tools to answer the questions. You have one tool that allows\\nyou to write CODE. Generate a plan based on the instructions to use the CODE tool several times in order to satisfy\\nyour instructions. Each call to the CODE tool should have a single goal, such as making one type of plot, builting\\none type of model, and so forth.\\n\\nIf your job is to make plots, then you can make a plan to make multiple kinds of plots with each different call\\nto CODE in your plan. For instance. your plan could be\\n\\n**Step 1, CODE** load <path to data>.h5ad and make umap plots of ...\\n**Step 2, CODE** load <path to data>.h5ad and make tsne plots of ...\\n\\nBOT2: Distance Computation\\n**Objective:** Compute the distance between each observation representing a recipe and each observation representing a source or target cell.\\n\\n**Instructions:**\\n1. Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`.\\n2. Extract the embeddings for 'source', 'target', and each recipe, based on the 'recipe' column in adata.obs, from `adata.X`\\n3. Compute the pairwise Euclidean distances between:\\n   - Each 'source' embedding and each recipe embedding.\\n   - Each 'target' embedding and each recipe embedding.\\n4. Store the computed distances in a matrix format and save the results in an ann data object. Set\\n   - .var of the new ann data to be the .obs of the source and target cells used in the distance calculations\\n   - .obs of the new ann data to be the .obs of the reprogrammed cells used in the distance calculations, i.e. where 'recipe' is not 'target' and is not 'source'\\n5. explicitly save the new ann data object\\n6. Provide a summary of the distance metrics, highlighting which recipes are closest to the 'target' cells.\", 'output': \"Here is the plan to address the user's query:\\n\\n**Step 1 (ROUTER)**:\\nPrompt: Determine which step to proceed to next, considering the objective of computing distances between observations representing recipes and source or target cells.\\n\\n**Step 2 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings for 'source', 'target', and each recipe from `adata.X`.\\n\\n**Step 3 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'source' embedding and each recipe embedding, using the extracted embeddings from Step 2.\\n\\n**Step 4 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'target' embedding and each recipe embedding, using the extracted embeddings from Step 2.\\n\\n**Step 5 (CODE)**:\\nPrompt: Store the computed distances in a matrix format and create a new ann data object, setting `.var` to be the `.obs` of the source and target cells used in the distance calculations and `.obs` to be the `.obs` of the reprogrammed cells used in the distance calculations. Explicitly save the new ann data object.\\n\\n**Step 6 (WRITE)**:\\nPrompt: Provide a summary of the distance metrics, highlighting which recipes are closest to the 'target' cells, using the results from Step 5.\\n\\n**Step 7 (ROUTER)**:\\nPrompt: Determine the next step, considering the results from Step 6 and the overall objective of the analysis.\\n\\nNote: The dependencies between steps are:\\n\\n* Step 3 and Step 4 rely on the extracted embeddings from Step 2.\\n* Step 5 uses the results from Step 3 and Step 4.\\n* Step 6 relies on the new ann data object created in Step 5.\\n\\n\", 'parsedOutput': {'output': \"Here is the plan to address the user's query:\\n\\n**Step 1 (ROUTER)**:\\nPrompt: Determine which step to proceed to next, considering the objective of computing distances between observations representing recipes and source or target cells.\\n\\n**Step 2 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings for 'source', 'target', and each recipe from `adata.X`.\\n\\n**Step 3 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'source' embedding and each recipe embedding, using the extracted embeddings from Step 2.\\n\\n**Step 4 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'target' embedding and each recipe embedding, using the extracted embeddings from Step 2.\\n\\n**Step 5 (CODE)**:\\nPrompt: Store the computed distances in a matrix format and create a new ann data object, setting `.var` to be the `.obs` of the source and target cells used in the distance calculations and `.obs` to be the `.obs` of the reprogrammed cells used in the distance calculations. Explicitly save the new ann data object.\\n\\n**Step 6 (WRITE)**:\\nPrompt: Provide a summary of the distance metrics, highlighting which recipes are closest to the 'target' cells, using the results from Step 5.\\n\\n**Step 7 (ROUTER)**:\\nPrompt: Determine the next step, considering the results from Step 6 and the overall objective of the analysis.\\n\\nNote: The dependencies between steps are:\\n\\n* Step 3 and Step 4 rely on the extracted embeddings from Step 2.\\n* Step 5 uses the results from Step 3 and Step 4.\\n* Step 6 relies on the new ann data object created in Step 5.\\n\\n\"}, 'purpose': 'generate a proposed set of prompts'}, {'func': 'planner.planner', 'prompt to user': 'Do you want to proceed with this plan? [Y/N/edit]', 'input': 'please remove steps 1, 6, and 7. Also, please provide instructions for figure generation within each stage or step', 'purpose': 'get new user input'}, {'llm': \"model='meta/llama3-70b-instruct'\", 'memory': 'None', 'prompt': 'input_variables=[\\'input\\'] template=\"Based on the most recently proposed Current Plan and the user\\'s new requirements and requested changes, create a revised plan.\\\\n\\\\nCurrent Plan: Here is the plan to address the user\\'s query:\\\\n\\\\n**Step 1 (ROUTER)**:\\\\nPrompt: Determine which step to proceed to next, considering the objective of computing distances between observations representing recipes and source or target cells.\\\\n\\\\n**Step 2 (CODE)**:\\\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings for \\'source\\', \\'target\\', and each recipe from `adata.X`.\\\\n\\\\n**Step 3 (CODE)**:\\\\nPrompt: Compute the pairwise Euclidean distances between each \\'source\\' embedding and each recipe embedding, using the extracted embeddings from Step 2.\\\\n\\\\n**Step 4 (CODE)**:\\\\nPrompt: Compute the pairwise Euclidean distances between each \\'target\\' embedding and each recipe embedding, using the extracted embeddings from Step 2.\\\\n\\\\n**Step 5 (CODE)**:\\\\nPrompt: Store the computed distances in a matrix format and create a new ann data object, setting `.var` to be the `.obs` of the source and target cells used in the distance calculations and `.obs` to be the `.obs` of the reprogrammed cells used in the distance calculations. Explicitly save the new ann data object.\\\\n\\\\n**Step 6 (WRITE)**:\\\\nPrompt: Provide a summary of the distance metrics, highlighting which recipes are closest to the \\'target\\' cells, using the results from Step 5.\\\\n\\\\n**Step 7 (ROUTER)**:\\\\nPrompt: Determine the next step, considering the results from Step 6 and the overall objective of the analysis.\\\\n\\\\nNote: The dependencies between steps are:\\\\n\\\\n* Step 3 and Step 4 rely on the extracted embeddings from Step 2.\\\\n* Step 5 uses the results from Step 3 and Step 4.\\\\n* Step 6 relies on the new ann data object created in Step 5.\\\\n\\\\n\\\\n\\\\n\\\\nUser\\'s requested changes: {input}\\\\n\\\\n**Available Methods:**\\\\n1. **Retrieval Augmented Generation**(RAG): Look up literature and documents from a text database.\\\\n2. **Web Search for New Literature**(SCRAPE): Search platforms like arXiv, bioRxiv, and PubMed for the latest research.\\\\n3. **Bioinformatics Databases**(DATABASE): Utilize databases such as Gene Ontology and Enrichr to perform gene set enrichment analyses.\\\\n4. **Run Codes**(CODE): Execute bioinformatics pipelines. Utilize prebuilt pipelines or develop new ones as needed.\\\\n5. **Write a Report**(WRITE): Synthesize and summarize information. This can include summarizing database searches, code pipeline results, or creating a final report to encapsulate the entire analysis.\\\\n\\\\n**OUTPUT**\\\\nRevised Plan:\\\\n**Step 1 (method, eg. RAG)**:\\\\nPrompt: [Description of action for chatbot to do Step 1, i.e. do X, Y, Z]\\\\n\\\\n**Step 2 (method, eg. CODE)**:\\\\nPrompt: [Description of action for chatbot to do Step 2, i.e. do X, Y, Z]\\\\n...\\\\n\"', 'input': 'please remove steps 1, 6, and 7. Also, please provide instructions for figure generation within each stage or step', 'output': \"Here is the revised plan based on the user's new requirements and requested changes:\\n\\n**Revised Plan:**\\n\\n**Step 1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings for 'source', 'target', and each recipe from `adata.X`. **Generate a UMAP plot to visualize the embeddings.**\\n\\n**Step 2 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'source' embedding and each recipe embedding, using the extracted embeddings from Step 1. **Generate a heatmap to visualize the distance matrix.**\\n\\n**Step 3 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'target' embedding and each recipe embedding, using the extracted embeddings from Step 1. **Generate a heatmap to visualize the distance matrix.**\\n\\n**Step 4 (CODE)**:\\nPrompt: Store the computed distances in a matrix format and create a new ann data object, setting `.var` to be the `.obs` of the source and target cells used in the distance calculations and `.obs` to be the `.obs` of the reprogrammed cells used in the distance calculations. Explicitly save the new ann data object. **Generate a summary plot to visualize the new ann data object.**\\n\\nThe revised plan removes Steps 1, 6, and 7 as per the user's request and adds instructions for figure generation within each step. The available methods have been assigned to each step accordingly.\\n\\n\", 'parsedOutput': {'output': \"Here is the revised plan based on the user's new requirements and requested changes:\\n\\n**Revised Plan:**\\n\\n**Step 1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings for 'source', 'target', and each recipe from `adata.X`. **Generate a UMAP plot to visualize the embeddings.**\\n\\n**Step 2 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'source' embedding and each recipe embedding, using the extracted embeddings from Step 1. **Generate a heatmap to visualize the distance matrix.**\\n\\n**Step 3 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'target' embedding and each recipe embedding, using the extracted embeddings from Step 1. **Generate a heatmap to visualize the distance matrix.**\\n\\n**Step 4 (CODE)**:\\nPrompt: Store the computed distances in a matrix format and create a new ann data object, setting `.var` to be the `.obs` of the source and target cells used in the distance calculations and `.obs` to be the `.obs` of the reprogrammed cells used in the distance calculations. Explicitly save the new ann data object. **Generate a summary plot to visualize the new ann data object.**\\n\\nThe revised plan removes Steps 1, 6, and 7 as per the user's request and adds instructions for figure generation within each step. The available methods have been assigned to each step accordingly.\\n\\n\"}, 'purpose': 'update the proposed set of prompts'}, {'func': 'planner.planner', 'prompt to user': 'Do you want to proceed with this plan? [Y/N/edit]', 'input': 'Y', 'purpose': 'get new user input'}, {'func': 'planner.planner', 'what': 'designed a new pipeline'}, {'func': 'planner.planner', 'what': 'set the queue and set the queue pointer to 1'}]}\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pipeline saved successfully to /home/jpic/RAG-DEV/tutorials/geneformer/pipelines/BOT2.json\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "\"Here is the plan to address the user's query:\\n\\n**Step 1 (ROUTER)**:\\nPrompt: Determine which step to proceed to next, considering the objective of computing distances between observations representing recipes and source or target cells.\\n\\n**Step 2 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings for 'source', 'target', and each recipe from `adata.X`.\\n\\n**Step 3 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'source' embedding and each recipe embedding, using the extracted embeddings from Step 2.\\n\\n**Step 4 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'target' embedding and each recipe embedding, using the extracted embeddings from Step 2.\\n\\n**Step 5 (CODE)**:\\nPrompt: Store the computed distances in a matrix format and create a new ann data object, setting `.var` to be the `.obs` of the source and target cells used in the distance calculations and `.obs` to be the `.obs` of the reprogrammed cells used in the distance calculations. Explicitly save the new ann data object.\\n\\n**Step 6 (WRITE)**:\\nPrompt: Provide a summary of the distance metrics, highlighting which recipes are closest to the 'target' cells, using the results from Step 5.\\n\\n**Step 7 (ROUTER)**:\\nPrompt: Determine the next step, considering the results from Step 6 and the overall objective of the analysis.\\n\\nNote: The dependencies between steps are:\\n\\n* Step 3 and Step 4 rely on the extracted embeddings from Step 2.\\n* Step 5 uses the results from Step 3 and Step 4.\\n* Step 6 relies on the new ann data object created in Step 5.\\n\\nDo you want to proceed with this plan? [Y/N/edit]Here is the revised plan based on the user's new requirements and requested changes:\\n\\n**Revised Plan:**\\n\\n**Step 1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings for 'source', 'target', and each recipe from `adata.X`. **Generate a UMAP plot to visualize the embeddings.**\\n\\n**Step 2 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'source' embedding and each recipe embedding, using the extracted embeddings from Step 1. **Generate a heatmap to visualize the distance matrix.**\\n\\n**Step 3 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'target' embedding and each recipe embedding, using the extracted embeddings from Step 1. **Generate a heatmap to visualize the distance matrix.**\\n\\n**Step 4 (CODE)**:\\nPrompt: Store the computed distances in a matrix format and create a new ann data object, setting `.var` to be the `.obs` of the source and target cells used in the distance calculations and `.obs` to be the `.obs` of the reprogrammed cells used in the distance calculations. Explicitly save the new ann data object. **Generate a summary plot to visualize the new ann data object.**\\n\\nThe revised plan removes Steps 1, 6, and 7 as per the user's request and adds instructions for figure generation within each step. The available methods have been assigned to each step accordingly.\\n\\nDo you want to proceed with this plan? [Y/N/edit]Would you like to save this pipeline to a file? [Y/N]Enter a name for your pipelineEnter a description of your pipelinePipeline saved successfully to /home/jpic/RAG-DEV/tutorials/geneformer/pipelines/BOT2.json\""
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "BOT2.invoke('/force PLANNER ' + worker_system_prompt + modified_instructions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "897719a6-3037-43d4-9498-d4a52e74516c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32m2024-08-03 01:07:41 INFO semantic_router.utils.logger local\u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome to RAG! The chat log from this conversation will be saved to /home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-07-41/log.json. How can I help?\n",
      "==================================================\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  /force PLANNER please load and run the BOT2 pipeline\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 01:07:55,366 - INFO - PLANNER\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "BRAD >> 0: \n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3m**INSTRUCTIONS:**\n",
      "You are planning a bioinformatics analysis pipeline to address a user's query. You must determine if there already exist a predesigned pipeline that will respond to a user's query, or if a new, custom process must be designed for this particular query. The multi-step workflow can use the available methods listed below.\n",
      "\n",
      "**Available Methods within a Pipeline:**\n",
      "1. **RAG**: Look up literature and documents from a text database.\n",
      "2. **SCRAPE**: Search platforms like arXiv, bioRxiv, and PubMed for the latest research.\n",
      "3. **DATABASE**: Utilize bioinformatics databases such as Gene Ontology and Enrichr to perform gene set enrichment analyses.\n",
      "4. **CODE**: Execute bioinformatics pipelines. Utilize prebuilt pipelines or develop new ones as needed.\n",
      "5. **WRITE**: Synthesize and summarize information. This can include summarizing database searches, code pipeline results, or creating a final report to encapsulate the entire analysis.\n",
      "6. **ROUTER**: Determine which step we should proceed to next.\n",
      "\n",
      "User Query: please load and run the BOT2 pipeline\n",
      "\n",
      "**Available Pipelines**\n",
      "Name: BOT2\tDescription: this pipeline investigates distances between source, target, and reprogrammed cells. BOT2 should use this pipeline\n",
      "Name: BOT1\tDescription: this plan is for basic visualization and summary of single cell data from the Geneformer reprogramming experiment. The data can be loaded in from ann data objects\n",
      "\n",
      "\n",
      "CUSTOM: The option to build a new pipeline\n",
      "\n",
      "The custom pipeline should be selected if no current pipeline addreses the user's query. If the custom pipeline is selected, please do not attempt to design the pipeline now. We will work to design it at a later point.\n",
      "\n",
      "**Output**:\n",
      "Please formate your output as follows:\n",
      "Pipeline Name: <available pipeline name above or \"CUSTOM\">\n",
      "Explination: <reason for selecting the above pipeline>\n",
      "\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 01:07:56,310 - INFO - HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2024-08-03 01:07:56,311 - INFO - Pipeline Name: BOT2\n",
      "Explanation: The user's query specifically requests to load and run the BOT2 pipeline. The description of BOT2 matches the user's request as it investigates distances between source, target, and reprogrammed cells. Therefore, the BOT2 pipeline is the appropriate choice for this query.\n",
      "2024-08-03 01:07:56,311 - INFO - selected_pipeline=BOT2\n",
      "2024-08-03 01:07:56,312 - INFO - \n",
      "\n",
      "\n",
      "route\n",
      "\n",
      "\n",
      "\n",
      "2024-08-03 01:07:56,312 - INFO - PLANNER\n",
      "2024-08-03 01:07:56,312 - INFO - \n",
      "\n",
      "PLANNER:\n",
      "\n",
      "{'module': 'PLANNER', 'steps': [{'llm': \"client=<openai.resources.chat.completions.Completions object at 0x147f010bd150> async_client=<openai.resources.chat.completions.AsyncCompletions object at 0x147f010bf650> model_name='gpt-4o' temperature=0.0 openai_api_key=SecretStr('**********') openai_proxy=''\", 'memory': 'None', 'prompt': 'input_variables=[\\'input\\'] template=\\'**INSTRUCTIONS:**\\\\nYou are planning a bioinformatics analysis pipeline to address a user\\\\\\'s query. You must determine if there already exist a predesigned pipeline that will respond to a user\\\\\\'s query, or if a new, custom process must be designed for this particular query. The multi-step workflow can use the available methods listed below.\\\\n\\\\n**Available Methods within a Pipeline:**\\\\n1. **RAG**: Look up literature and documents from a text database.\\\\n2. **SCRAPE**: Search platforms like arXiv, bioRxiv, and PubMed for the latest research.\\\\n3. **DATABASE**: Utilize bioinformatics databases such as Gene Ontology and Enrichr to perform gene set enrichment analyses.\\\\n4. **CODE**: Execute bioinformatics pipelines. Utilize prebuilt pipelines or develop new ones as needed.\\\\n5. **WRITE**: Synthesize and summarize information. This can include summarizing database searches, code pipeline results, or creating a final report to encapsulate the entire analysis.\\\\n6. **ROUTER**: Determine which step we should proceed to next.\\\\n\\\\nUser Query: {input}\\\\n\\\\n**Available Pipelines**\\\\nName: BOT2\\\\tDescription: this pipeline investigates distances between source, target, and reprogrammed cells. BOT2 should use this pipeline\\\\nName: BOT1\\\\tDescription: this plan is for basic visualization and summary of single cell data from the Geneformer reprogramming experiment. The data can be loaded in from ann data objects\\\\n\\\\n\\\\nCUSTOM: The option to build a new pipeline\\\\n\\\\nThe custom pipeline should be selected if no current pipeline addreses the user\\\\\\'s query. If the custom pipeline is selected, please do not attempt to design the pipeline now. We will work to design it at a later point.\\\\n\\\\n**Output**:\\\\nPlease formate your output as follows:\\\\nPipeline Name: <available pipeline name above or \"CUSTOM\">\\\\nExplination: <reason for selecting the above pipeline>\\\\n\\'', 'input': 'please load and run the BOT2 pipeline', 'output': \"Pipeline Name: BOT2\\nExplanation: The user's query specifically requests to load and run the BOT2 pipeline. The description of BOT2 matches the user's request as it investigates distances between source, target, and reprogrammed cells. Therefore, the BOT2 pipeline is the appropriate choice for this query.\", 'parsedOutput': {'selected pipeline': 'BOT2'}, 'purpose': 'determine if a known pipeline can be used or a new one si required'}, {'func': 'planner.planner', 'what': 'loaded an older pipeline'}, {'func': 'planner.planner', 'what': 'set the queue and set the queue pointer to 1'}]}\n",
      "2024-08-03 01:07:56,316 - INFO - current llm calls=1\n",
      "2024-08-03 01:07:56,319 - INFO - CODE\n",
      "2024-08-03 01:07:56,320 - INFO - CODER\n",
      "2024-08-03 01:07:56,321 - INFO - input_variables=['user_query'] template=\"You must select which code to run to help a user.\\n\\n**Available Scripts**\\nScript Name: scanpy_brad, \\t Description: This script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\\nScript Name: models_brad, \\t Description: This script executes a series of custom commands to build a model for predicting the contribution to reprogramming obtained by the inclusion or exclusion of individual TFs using an AnnData object or single cell datasets loaded from a .h5ad file.\\n\\n\\n**User Query**\\n{user_query}\\n\\n**Task**\\nBased on the user's query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\\n\\n**Response Template**\\nSCRIPT: <selected script>\\nREASON: <reasoning why the script fits the user prompt>\\n\"\n",
      "2024-08-03 01:07:56,321 - INFO - FIRST LLM CALL\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "** Step 0**\n",
      "{'order': 0, 'module': 'PLANNER', 'prompt': None, 'description': 'This step designed the plan. It is placed in the queue because we needed a place holder for 0 indexed lists.', 'output': []}\n",
      "\n",
      "\n",
      "** Step 1**\n",
      "{'order': 1, 'module': 'CODE', 'prompt': \"/force CODE  1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings for 'source', 'target', and each recipe from `adata.X`. **Generate a UMAP plot to visualize the embeddings.**\\n\\n**.\", 'description': \" 1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings for 'source', 'target', and each recipe from `adata.X`. **Generate a UMAP plot to visualize the embeddings.**\\n\\n**\", 'output': []}\n",
      "\n",
      "\n",
      "** Step 2**\n",
      "{'order': 2, 'module': 'CODE', 'prompt': \"/force CODE  2 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'source' embedding and each recipe embedding, using the extracted embeddings from the .h5ad file: `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` Plot histograms and or other visualizations of the distances.\", 'description': \" 2 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'source' embedding and each recipe embedding, using the extracted embeddings from \", 'output': []}\n",
      "\n",
      "\n",
      "** Step 3**\n",
      "{'order': 4, 'module': 'CODE', 'prompt': \"/force CODE  2 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'target' embedding and each recipe embedding, using the extracted embeddings from the .h5ad file: `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` Plot histograms and or other visualizations of the distances.\", 'description': \" 3 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'target' embedding and each recipe embedding, using the extracted embeddings from \", 'output': []}\n",
      "\n",
      "\n",
      "** Step 4**\n",
      "{'order': 6, 'module': 'CODE', 'prompt': '/force CODE  4 (CODE)**:\\nPrompt: Store the computed distances in a matrix format and create a new ann data object, setting `.var` to be the `.obs` of the source and target cells used in the distance calculations and `.obs` to be the `.obs` of the reprogrammed cells used in the distance calculations. Explicitly save the new ann data object. **Generate a summary plot to visualize the new ann data object.**\\n\\nThe revised plan removes ', 'description': ' 4 (CODE)**:\\nPrompt: Store the computed distances in a matrix format and create a new ann data object, setting `.var` to be the `.obs` of the source and target cells used in the distance calculations and `.obs` to be the `.obs` of the reprogrammed cells used in the distance calculations. Explicitly save the new ann data object. **Generate a summary plot to visualize the new ann data object.**\\n\\nThe revised plan removes ', 'output': []}\n",
      "\n",
      "\n",
      "==================================================\n",
      "==================================================\n",
      "BRAD >> 1: \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 01:07:57,794 - INFO - HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2024-08-03 01:07:57,795 - INFO - SCRIPT: scanpy_brad\n",
      "REASON: The user's query involves loading an AnnData object from a .h5ad file, extracting embeddings, and generating a UMAP plot for visualization. The `scanpy_brad` script is designed to handle such tasks, as it allows for manipulation, visualization, and exploration of single-cell datasets using libraries like scanpy, seaborn, and matplotlib.pyplot. This makes it the best match for the user's requirements.\n",
      "2024-08-03 01:07:57,795 - INFO - scriptPath=/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts\n",
      "2024-08-03 01:07:57,796 - INFO - ALL SCRIPTS FOUND. BUILDING TEMPLATE\n",
      "2024-08-03 01:07:57,797 - INFO - input_variables=['history', 'input'] template='Current conversation:\\n{history}\\n\\n**PYTHON SCRIPT**\\nYou must run this python script:\\n/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\n\\n**PYTHON SCRIPT DOCUMENTATION**:\\nThis is the doc string of this python script:\\n\"\"\"\\nThis script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\\n\\nArguments (four arguments):\\n    1. output directory: chatstatus[\\'output-directory\\']\\n    2. output file: <name of output file>\\n    3. input file: <file created in previous step>\\n    4. scanpy commands: a list of scanpy commands to be executed on the AnnData object (provided as a single string with commands separated by a delimiter, e.g., \\';\\')\\n\\nBased on the arguments, the input file will be loaded, then your commands will be executed, and finally, the output or resulting ann data object will be saved to the corrrectou output file and directory. Your code is not responsible for loading the .h5ad object, that will happen automatically, and when loaded, the object will be called adata. Your scanpy commands can operate directly on the adata object that will be loaded for you.\\n\\nAdditionally, the following imports are already provided for you and can be used in your code:\\n```\\nimport scanpy as sc\\nimport seaborn as sns\\nimport matplotlib.pyplot as plt\\nimport os\\nimport numpy as np\\nimport pandas as pd\\nfrom scipy.spatial.distance import cdist\\n```\\n\\n**Usage**\\nBRAD Line:\\n```\\nsubprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\'output-directory\\'], <output file>, <input file>, \"<scanpy commands>\"], capture_output=True, text=True)\\n```\\n\\n**Examples**\\nUse the below examples to help generate your code.\\n\\n*Example 1*\\nUser Prompt: Run scanpy preprocessing and UMAP visualization on XXX.h5ad and save the UMAP plot\\nResponse Code:\\n```\\nresponse = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"XXX-modified.h5ad\", \"<path/to/data>/XXX.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save=\\'umap.png\\')\"], capture_output=True, text=True)\\n```\\nExplination: the adata object will be loaded in memory already. The command \"sc.pp.neighbors(adata)\" will preprocess the data, then the command \"sc.tl.umap(adata)\" will perform UMAP, and finally the command \"sc.pl.umap(adata, \\'umap.png\\')\" will save the UMAP to a well named file.\\n\\n*Example 2*\\nUser Prompt: Perform PCA and clustering on the dataset YYY.h5ad and save the PCA plot\\nResponse Code:\\n```\\nresponse = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"YYY-modified.h5ad\", \"<path/to/data>/YYY.h5ad\", \"sc.pp.pca(adata); sc.tl.leiden(adata); sc.pl.pca(adata, save=\\'pca.png\\')\"], capture_output=True, text=True)\\n```\\nExplination: the adata object will be loaded in memory already. The command \"sc.pp.pca(adata)\" will preprocess the data, then the command \"sc.tl.leiden(adata)\" will perform the leiden algorithm, and finally the command \"sc.pl.pca(adata, save=\\'pca.png\\')\" will save the PCA to a well named file.\\n\\n*Example 3*\\nUser Prompt: compute the distance between all cells that are not source or target to all source and target cells.\\nReponse Code:\\n```\\nadata_X = adata.X\\nmask = ~adata.obs[\\'recipe\\'].isin([\\'source\\', \\'target\\'])\\nrecipe_embeddings = adata[mask]\\nsource_embeddings = adata[adata.obs[\\'recipe\\'] == \\'source\\']\\ntarget_embeddings = adata[adata.obs[\\'recipe\\'] == \\'target\\']\\nrecipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X)\\nrecipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X)\\nadata_dist = sc.AnnData(X=np.concatenate((recipe_distances_source, recipe_distances_target), axis=1), var=pd.concat((source_embeddings.obs, target_embeddings.obs)), obs=recipe_embeddings.obs)\\nadata_dist.write(<file name>.h5ad)\\n```\\n\\n**OUTPUT FILE NAME INSTRUCTIONS**\\n1. Output path should be chatstatus[\\'output-directory\\']\\n2. Output file name should be `<descriptive name>.h5ad`\\n\"\"\"\\n\\n\\n**CALL PYTHON SCRIPTS FROM PYTHON**:\\nUse the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\\n\\nTo call a Python script `example_script.py` which has no arguments:\\n\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\']], capture_output=True, text=True)\\n```\\n\\nTo call a Python script `example_script.py` with one argument:\\n\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\'], \\'arg1\\'], capture_output=True, text=True)\\n```\\n\\nTo call a Python script `example_script.py` with two arguments:\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\'], \\'arg1\\', \\'arg2\\'], capture_output=True, text=True)\\n```\\n\\nNote that chatstatus[\\'output-directory\\'] is ALWAYS passed as the first argument.\\n\\nThe following files were previously created by BRAD and could be used as input to a function if necessary:\\nlog.json\\n\\nQuery:{input}\\n\\n**PYTHON SCRIPT**\\nPlease use the correct script name and path. You must run this python script:\\n/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\n\\n**INSTRUCTIONS**\\n1. Given the user query and the documentation, identify each of the arguments found in the user\\'s query that should be passed to the Python script.\\n2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\\n3. The last line of your response should say \"Execute: <Python code to execute>\"\\n4. Do not include any extra words or characters. Format the response/output as:\\n    Arguments: \\n    Python Code Explanation: <2 sentences maximum>\\n    Execute: <your code here>\\n\\n**IMPORTANT**\\nThe code to execute from your response must be formatted as:\\n    Execute: subprocess.call([sys.executable, \\'<path to python script>\\', \\'<argument 1>\\', \\'<argument 2>\\', ..., \\'<argument n>\\'], capture_output=True, text=True))\\nThis output should be exactly one line and no longer. Stop the response after this line.\\n'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation:\n",
      "\n",
      "\n",
      "**PYTHON SCRIPT**\n",
      "You must run this python script:\n",
      "/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\n",
      "\n",
      "**PYTHON SCRIPT DOCUMENTATION**:\n",
      "This is the doc string of this python script:\n",
      "\"\"\"\n",
      "This script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\n",
      "\n",
      "Arguments (four arguments):\n",
      "    1. output directory: chatstatus['output-directory']\n",
      "    2. output file: <name of output file>\n",
      "    3. input file: <file created in previous step>\n",
      "    4. scanpy commands: a list of scanpy commands to be executed on the AnnData object (provided as a single string with commands separated by a delimiter, e.g., ';')\n",
      "\n",
      "Based on the arguments, the input file will be loaded, then your commands will be executed, and finally, the output or resulting ann data object will be saved to the corrrectou output file and directory. Your code is not responsible for loading the .h5ad object, that will happen automatically, and when loaded, the object will be called adata. Your scanpy commands can operate directly on the adata object that will be loaded for you.\n",
      "\n",
      "Additionally, the following imports are already provided for you and can be used in your code:\n",
      "```\n",
      "import scanpy as sc\n",
      "import seaborn as sns\n",
      "import matplotlib.pyplot as plt\n",
      "import os\n",
      "import numpy as np\n",
      "import pandas as pd\n",
      "from scipy.spatial.distance import cdist\n",
      "```\n",
      "\n",
      "**Usage**\n",
      "BRAD Line:\n",
      "```\n",
      "subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus['output-directory'], <output file>, <input file>, \"<scanpy commands>\"], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "**Examples**\n",
      "Use the below examples to help generate your code.\n",
      "\n",
      "*Example 1*\n",
      "User Prompt: Run scanpy preprocessing and UMAP visualization on XXX.h5ad and save the UMAP plot\n",
      "Response Code:\n",
      "```\n",
      "response = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus['output-directory'], \"XXX-modified.h5ad\", \"<path/to/data>/XXX.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save='umap.png')\"], capture_output=True, text=True)\n",
      "```\n",
      "Explination: the adata object will be loaded in memory already. The command \"sc.pp.neighbors(adata)\" will preprocess the data, then the command \"sc.tl.umap(adata)\" will perform UMAP, and finally the command \"sc.pl.umap(adata, 'umap.png')\" will save the UMAP to a well named file.\n",
      "\n",
      "*Example 2*\n",
      "User Prompt: Perform PCA and clustering on the dataset YYY.h5ad and save the PCA plot\n",
      "Response Code:\n",
      "```\n",
      "response = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus['output-directory'], \"YYY-modified.h5ad\", \"<path/to/data>/YYY.h5ad\", \"sc.pp.pca(adata); sc.tl.leiden(adata); sc.pl.pca(adata, save='pca.png')\"], capture_output=True, text=True)\n",
      "```\n",
      "Explination: the adata object will be loaded in memory already. The command \"sc.pp.pca(adata)\" will preprocess the data, then the command \"sc.tl.leiden(adata)\" will perform the leiden algorithm, and finally the command \"sc.pl.pca(adata, save='pca.png')\" will save the PCA to a well named file.\n",
      "\n",
      "*Example 3*\n",
      "User Prompt: compute the distance between all cells that are not source or target to all source and target cells.\n",
      "Reponse Code:\n",
      "```\n",
      "adata_X = adata.X\n",
      "mask = ~adata.obs['recipe'].isin(['source', 'target'])\n",
      "recipe_embeddings = adata[mask]\n",
      "source_embeddings = adata[adata.obs['recipe'] == 'source']\n",
      "target_embeddings = adata[adata.obs['recipe'] == 'target']\n",
      "recipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X)\n",
      "recipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X)\n",
      "adata_dist = sc.AnnData(X=np.concatenate((recipe_distances_source, recipe_distances_target), axis=1), var=pd.concat((source_embeddings.obs, target_embeddings.obs)), obs=recipe_embeddings.obs)\n",
      "adata_dist.write(<file name>.h5ad)\n",
      "```\n",
      "\n",
      "**OUTPUT FILE NAME INSTRUCTIONS**\n",
      "1. Output path should be chatstatus['output-directory']\n",
      "2. Output file name should be `<descriptive name>.h5ad`\n",
      "\"\"\"\n",
      "\n",
      "\n",
      "**CALL PYTHON SCRIPTS FROM PYTHON**:\n",
      "Use the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\n",
      "\n",
      "To call a Python script `example_script.py` which has no arguments:\n",
      "\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory']], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with one argument:\n",
      "\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory'], 'arg1'], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with two arguments:\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory'], 'arg1', 'arg2'], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "Note that chatstatus['output-directory'] is ALWAYS passed as the first argument.\n",
      "\n",
      "The following files were previously created by BRAD and could be used as input to a function if necessary:\n",
      "log.json\n",
      "\n",
      "Query:1 (CODE)**:\n",
      "Prompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings for 'source', 'target', and each recipe from `adata.X`. **Generate a UMAP plot to visualize the embeddings.**\n",
      "\n",
      "**.\n",
      "\n",
      "**PYTHON SCRIPT**\n",
      "Please use the correct script name and path. You must run this python script:\n",
      "/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\n",
      "\n",
      "**INSTRUCTIONS**\n",
      "1. Given the user query and the documentation, identify each of the arguments found in the user's query that should be passed to the Python script.\n",
      "2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\n",
      "3. The last line of your response should say \"Execute: <Python code to execute>\"\n",
      "4. Do not include any extra words or characters. Format the response/output as:\n",
      "    Arguments: \n",
      "    Python Code Explanation: <2 sentences maximum>\n",
      "    Execute: <your code here>\n",
      "\n",
      "**IMPORTANT**\n",
      "The code to execute from your response must be formatted as:\n",
      "    Execute: subprocess.call([sys.executable, '<path to python script>', '<argument 1>', '<argument 2>', ..., '<argument n>'], capture_output=True, text=True))\n",
      "This output should be exactly one line and no longer. Stop the response after this line.\n",
      "\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 01:08:00,772 - INFO - HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2024-08-03 01:08:00,773 - INFO - LLM OUTPUT PARSER\n",
      "2024-08-03 01:08:00,773 - INFO - Arguments:\n",
      "1. output directory: chatstatus['output-directory']\n",
      "2. output file: \"HSC-reprogrammed-DE-umap.h5ad\"\n",
      "3. input file: \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\"\n",
      "4. scanpy commands: \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save='umap.png')\"\n",
      "\n",
      "Python Code Explanation: This code will load the AnnData object, preprocess it, generate a UMAP plot, and save the plot and modified AnnData object.\n",
      "\n",
      "Execute: subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-DE-umap.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save='umap.png')\"], capture_output=True, text=True)\n",
      "2024-08-03 01:08:00,773 - INFO - ['Arguments:\\n1. output directory: chatstatus[\\'output-directory\\']\\n2. output file: \"HSC-reprogrammed-DE-umap.h5ad\"\\n3. input file: \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\"\\n4. scanpy commands: \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save=\\'umap.png\\')\"\\n\\nPython Code Explanation: This code will load the AnnData object, preprocess it, generate a UMAP plot, and save the plot and modified AnnData object.\\n\\n', ' subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"HSC-reprogrammed-DE-umap.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save=\\'umap.png\\')\"], capture_output=True, text=True)']\n",
      "2024-08-03 01:08:00,774 - INFO - Stripped funciton call\n",
      "2024-08-03 01:08:00,774 - INFO - subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-DE-umap.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save='umap.png')\"], capture_output=True, text=True)\n",
      "2024-08-03 01:08:00,774 - INFO - Cleaned up function call:\n",
      "\n",
      "2024-08-03 01:08:00,774 - INFO - subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-DE-umap.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save='umap.png')\"], capture_output=True, text=True)\n",
      "2024-08-03 01:08:00,774 - INFO - \n",
      "\n",
      "2024-08-03 01:08:00,775 - INFO - EVAL\n",
      "2024-08-03 01:08:00,776 - INFO - subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-DE-umap.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save='umap.png')\"], capture_output=True, text=True)\n",
      "2024-08-03 01:08:00,776 - INFO - END EVAL CHECK\n",
      "2024-08-03 01:08:00,776 - INFO - Finalized PYTHON Call:\n",
      "2024-08-03 01:08:00,776 - INFO -    \n",
      "2024-08-03 01:08:00,777 - INFO - response = subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-DE-umap.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save='umap.png')\"], capture_output=True, text=True)\n",
      "2024-08-03 01:08:00,777 - INFO -    \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 01:08:27,781 - INFO - Debug: PYTHON code executed successfully.\n",
      "2024-08-03 01:08:27,782 - INFO - Code Execution Output:\n",
      "2024-08-03 01:08:27,782 - INFO - CompletedProcess(args=['/sw/pkgs/arc/python3.11-anaconda/2024.02-1/bin/python', '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py', '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-07-41', 'HSC-reprogrammed-DE-umap.h5ad', '/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad', \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save='umap.png')\"], returncode=0, stdout=\"adata loaded from /nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\\nadata.shape=(6375, 512)\\nadata.obs.columns=Index(['recipe', 'filename', 'REL', 'STAT5A', 'FOS', 'ETV6', 'source', 'MEF2C',\\n       'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B', 'target'],\\n      dtype='object')\\nadata.obs.head()=   recipe                                           filename  ...  GFI1B  target\\n0  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n1  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n2  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n3  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n4  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n\\n[5 rows x 14 columns]\\nadata.var.head()=Empty DataFrame\\nColumns: []\\nIndex: [0, 1, 2, 3, 4]\\n****************************\\n      EXECUTE COMMANDS      \\n****************************\\nsc.pp.neighbors(adata)\\nsc.tl.umap(adata)\\nsc.pl.umap(adata, save='umap.png')\\nFigure(1200x1200)\\n\", stderr='/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\\n  utils.warn_names_duplicates(\"obs\")\\n/home/jpic/.local/lib/python3.11/site-packages/scanpy/tools/_utils.py:41: UserWarning: Youâ€™re trying to run this on 512 dimensions of `.X`, if you really want this, set `use_rep=\\'X\\'`.\\n         Falling back to preprocessing with `sc.pp.pca` and default params.\\n  warnings.warn(\\nWARNING: saving figure to file /home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-07-41/umapumap.png\\n')\n",
      "2024-08-03 01:08:27,782 - INFO - Debug: PYTHON code output saved to output.\n",
      "2024-08-03 01:08:27,783 - INFO - \n",
      "\n",
      "\n",
      "route\n",
      "\n",
      "\n",
      "\n",
      "2024-08-03 01:08:27,783 - INFO - CODE\n",
      "2024-08-03 01:08:27,784 - INFO - Finding Stage Number of Pipeline\n",
      "2024-08-03 01:08:27,785 - INFO - {0: {'order': 0, 'module': 'PLANNER', 'prompt': None, 'description': 'This step designed the plan. It is placed in the queue because we needed a place holder for 0 indexed lists.', 'output': []}, 1: {'order': 1, 'module': 'CODE', 'prompt': \"/force CODE  1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings for 'source', 'target', and each recipe from `adata.X`. **Generate a UMAP plot to visualize the embeddings.**\\n\\n**.\", 'description': \" 1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings for 'source', 'target', and each recipe from `adata.X`. **Generate a UMAP plot to visualize the embeddings.**\\n\\n**\", 'output': []}, 2: {'order': 2, 'module': 'CODE', 'prompt': \"/force CODE  2 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'source' embedding and each recipe embedding, using the extracted embeddings from the .h5ad file: `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` Plot histograms and or other visualizations of the distances.\", 'description': \" 2 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'source' embedding and each recipe embedding, using the extracted embeddings from \", 'output': []}, 3: {'order': 4, 'module': 'CODE', 'prompt': \"/force CODE  2 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'target' embedding and each recipe embedding, using the extracted embeddings from the .h5ad file: `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` Plot histograms and or other visualizations of the distances.\", 'description': \" 3 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'target' embedding and each recipe embedding, using the extracted embeddings from \", 'output': []}, 4: {'order': 6, 'module': 'CODE', 'prompt': '/force CODE  4 (CODE)**:\\nPrompt: Store the computed distances in a matrix format and create a new ann data object, setting `.var` to be the `.obs` of the source and target cells used in the distance calculations and `.obs` to be the `.obs` of the reprogrammed cells used in the distance calculations. Explicitly save the new ann data object. **Generate a summary plot to visualize the new ann data object.**\\n\\nThe revised plan removes ', 'description': ' 4 (CODE)**:\\nPrompt: Store the computed distances in a matrix format and create a new ann data object, setting `.var` to be the `.obs` of the source and target cells used in the distance calculations and `.obs` to be the `.obs` of the reprogrammed cells used in the distance calculations. Explicitly save the new ann data object. **Generate a summary plot to visualize the new ann data object.**\\n\\nThe revised plan removes ', 'output': []}}\n",
      "2024-08-03 01:08:27,785 - INFO - IP=1\n",
      "2024-08-03 01:08:27,785 - INFO - type(IP)=<class 'int'>\n",
      "2024-08-03 01:08:27,788 - INFO - \n",
      "\n",
      "PLANNER:\n",
      "\n",
      "{'module': 'CODE', 'steps': [{'llm': \"client=<openai.resources.chat.completions.Completions object at 0x147f010bd150> async_client=<openai.resources.chat.completions.AsyncCompletions object at 0x147f010bf650> model_name='gpt-4o' temperature=0.0 openai_api_key=SecretStr('**********') openai_proxy=''\", 'memory': 'None', 'prompt': 'input_variables=[\\'user_query\\'] template=\"You must select which code to run to help a user.\\\\n\\\\n**Available Scripts**\\\\nScript Name: scanpy_brad, \\\\t Description: This script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\\\\nScript Name: models_brad, \\\\t Description: This script executes a series of custom commands to build a model for predicting the contribution to reprogramming obtained by the inclusion or exclusion of individual TFs using an AnnData object or single cell datasets loaded from a .h5ad file.\\\\n\\\\n\\\\n**User Query**\\\\n{user_query}\\\\n\\\\n**Task**\\\\nBased on the user\\'s query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\\\\n\\\\n**Response Template**\\\\nSCRIPT: <selected script>\\\\nREASON: <reasoning why the script fits the user prompt>\\\\n\"', 'input': \"1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings for 'source', 'target', and each recipe from `adata.X`. **Generate a UMAP plot to visualize the embeddings.**\\n\\n**.\", 'output': 'content=\"SCRIPT: scanpy_brad\\\\nREASON: The user\\'s query involves loading an AnnData object from a .h5ad file, extracting embeddings, and generating a UMAP plot for visualization. The `scanpy_brad` script is designed to handle such tasks, as it allows for manipulation, visualization, and exploration of single-cell datasets using libraries like scanpy, seaborn, and matplotlib.pyplot. This makes it the best match for the user\\'s requirements.\" response_metadata={\\'token_usage\\': {\\'completion_tokens\\': 92, \\'prompt_tokens\\': 331, \\'total_tokens\\': 423}, \\'model_name\\': \\'gpt-4o-2024-05-13\\', \\'system_fingerprint\\': \\'fp_4e2b2da518\\', \\'finish_reason\\': \\'stop\\', \\'logprobs\\': None} id=\\'run-fae85920-5f55-4233-89a7-5e0ca0682953-0\\' usage_metadata={\\'input_tokens\\': 331, \\'output_tokens\\': 92, \\'total_tokens\\': 423}', 'parsedOutput': {'scriptName': '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py', 'scriptType': 'python', 'scriptPath': '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts'}, 'purpose': 'Select which code to run'}, {'llm': \"client=<openai.resources.chat.completions.Completions object at 0x147f010bd150> async_client=<openai.resources.chat.completions.AsyncCompletions object at 0x147f010bf650> model_name='gpt-4o' temperature=0.0 openai_api_key=SecretStr('**********') openai_proxy=''\", 'memory': 'None', 'prompt': 'input_variables=[\\'history\\', \\'input\\'] template=\\'Current conversation:\\\\n{history}\\\\n\\\\n**PYTHON SCRIPT**\\\\nYou must run this python script:\\\\n/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\\\n\\\\n**PYTHON SCRIPT DOCUMENTATION**:\\\\nThis is the doc string of this python script:\\\\n\"\"\"\\\\nThis script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\\\\n\\\\nArguments (four arguments):\\\\n    1. output directory: chatstatus[\\\\\\'output-directory\\\\\\']\\\\n    2. output file: <name of output file>\\\\n    3. input file: <file created in previous step>\\\\n    4. scanpy commands: a list of scanpy commands to be executed on the AnnData object (provided as a single string with commands separated by a delimiter, e.g., \\\\\\';\\\\\\')\\\\n\\\\nBased on the arguments, the input file will be loaded, then your commands will be executed, and finally, the output or resulting ann data object will be saved to the corrrectou output file and directory. Your code is not responsible for loading the .h5ad object, that will happen automatically, and when loaded, the object will be called adata. Your scanpy commands can operate directly on the adata object that will be loaded for you.\\\\n\\\\nAdditionally, the following imports are already provided for you and can be used in your code:\\\\n```\\\\nimport scanpy as sc\\\\nimport seaborn as sns\\\\nimport matplotlib.pyplot as plt\\\\nimport os\\\\nimport numpy as np\\\\nimport pandas as pd\\\\nfrom scipy.spatial.distance import cdist\\\\n```\\\\n\\\\n**Usage**\\\\nBRAD Line:\\\\n```\\\\nsubprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\\\\\'output-directory\\\\\\'], <output file>, <input file>, \"<scanpy commands>\"], capture_output=True, text=True)\\\\n```\\\\n\\\\n**Examples**\\\\nUse the below examples to help generate your code.\\\\n\\\\n*Example 1*\\\\nUser Prompt: Run scanpy preprocessing and UMAP visualization on XXX.h5ad and save the UMAP plot\\\\nResponse Code:\\\\n```\\\\nresponse = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\\\\\'output-directory\\\\\\'], \"XXX-modified.h5ad\", \"<path/to/data>/XXX.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save=\\\\\\'umap.png\\\\\\')\"], capture_output=True, text=True)\\\\n```\\\\nExplination: the adata object will be loaded in memory already. The command \"sc.pp.neighbors(adata)\" will preprocess the data, then the command \"sc.tl.umap(adata)\" will perform UMAP, and finally the command \"sc.pl.umap(adata, \\\\\\'umap.png\\\\\\')\" will save the UMAP to a well named file.\\\\n\\\\n*Example 2*\\\\nUser Prompt: Perform PCA and clustering on the dataset YYY.h5ad and save the PCA plot\\\\nResponse Code:\\\\n```\\\\nresponse = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\\\\\'output-directory\\\\\\'], \"YYY-modified.h5ad\", \"<path/to/data>/YYY.h5ad\", \"sc.pp.pca(adata); sc.tl.leiden(adata); sc.pl.pca(adata, save=\\\\\\'pca.png\\\\\\')\"], capture_output=True, text=True)\\\\n```\\\\nExplination: the adata object will be loaded in memory already. The command \"sc.pp.pca(adata)\" will preprocess the data, then the command \"sc.tl.leiden(adata)\" will perform the leiden algorithm, and finally the command \"sc.pl.pca(adata, save=\\\\\\'pca.png\\\\\\')\" will save the PCA to a well named file.\\\\n\\\\n*Example 3*\\\\nUser Prompt: compute the distance between all cells that are not source or target to all source and target cells.\\\\nReponse Code:\\\\n```\\\\nadata_X = adata.X\\\\nmask = ~adata.obs[\\\\\\'recipe\\\\\\'].isin([\\\\\\'source\\\\\\', \\\\\\'target\\\\\\'])\\\\nrecipe_embeddings = adata[mask]\\\\nsource_embeddings = adata[adata.obs[\\\\\\'recipe\\\\\\'] == \\\\\\'source\\\\\\']\\\\ntarget_embeddings = adata[adata.obs[\\\\\\'recipe\\\\\\'] == \\\\\\'target\\\\\\']\\\\nrecipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X)\\\\nrecipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X)\\\\nadata_dist = sc.AnnData(X=np.concatenate((recipe_distances_source, recipe_distances_target), axis=1), var=pd.concat((source_embeddings.obs, target_embeddings.obs)), obs=recipe_embeddings.obs)\\\\nadata_dist.write(<file name>.h5ad)\\\\n```\\\\n\\\\n**OUTPUT FILE NAME INSTRUCTIONS**\\\\n1. Output path should be chatstatus[\\\\\\'output-directory\\\\\\']\\\\n2. Output file name should be `<descriptive name>.h5ad`\\\\n\"\"\"\\\\n\\\\n\\\\n**CALL PYTHON SCRIPTS FROM PYTHON**:\\\\nUse the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\\\\n\\\\nTo call a Python script `example_script.py` which has no arguments:\\\\n\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\']], capture_output=True, text=True)\\\\n```\\\\n\\\\nTo call a Python script `example_script.py` with one argument:\\\\n\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\'], \\\\\\'arg1\\\\\\'], capture_output=True, text=True)\\\\n```\\\\n\\\\nTo call a Python script `example_script.py` with two arguments:\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\'], \\\\\\'arg1\\\\\\', \\\\\\'arg2\\\\\\'], capture_output=True, text=True)\\\\n```\\\\n\\\\nNote that chatstatus[\\\\\\'output-directory\\\\\\'] is ALWAYS passed as the first argument.\\\\n\\\\nThe following files were previously created by BRAD and could be used as input to a function if necessary:\\\\nlog.json\\\\n\\\\nQuery:{input}\\\\n\\\\n**PYTHON SCRIPT**\\\\nPlease use the correct script name and path. You must run this python script:\\\\n/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\\\n\\\\n**INSTRUCTIONS**\\\\n1. Given the user query and the documentation, identify each of the arguments found in the user\\\\\\'s query that should be passed to the Python script.\\\\n2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\\\\n3. The last line of your response should say \"Execute: <Python code to execute>\"\\\\n4. Do not include any extra words or characters. Format the response/output as:\\\\n    Arguments: \\\\n    Python Code Explanation: <2 sentences maximum>\\\\n    Execute: <your code here>\\\\n\\\\n**IMPORTANT**\\\\nThe code to execute from your response must be formatted as:\\\\n    Execute: subprocess.call([sys.executable, \\\\\\'<path to python script>\\\\\\', \\\\\\'<argument 1>\\\\\\', \\\\\\'<argument 2>\\\\\\', ..., \\\\\\'<argument n>\\\\\\'], capture_output=True, text=True))\\\\nThis output should be exactly one line and no longer. Stop the response after this line.\\\\n\\'', 'input': \"1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings for 'source', 'target', and each recipe from `adata.X`. **Generate a UMAP plot to visualize the embeddings.**\\n\\n**.\", 'output': 'Arguments:\\n1. output directory: chatstatus[\\'output-directory\\']\\n2. output file: \"HSC-reprogrammed-DE-umap.h5ad\"\\n3. input file: \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\"\\n4. scanpy commands: \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save=\\'umap.png\\')\"\\n\\nPython Code Explanation: This code will load the AnnData object, preprocess it, generate a UMAP plot, and save the plot and modified AnnData object.\\n\\nExecute: subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"HSC-reprogrammed-DE-umap.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save=\\'umap.png\\')\"], capture_output=True, text=True)', 'parsedOutput': {'code': 'subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"HSC-reprogrammed-DE-umap.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save=\\'umap.png\\')\"], capture_output=True, text=True)'}, 'purpose': 'Format function call'}, {'func': 'pythonCaller.execute_python_code', 'code': 'response = subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"HSC-reprogrammed-DE-umap.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save=\\'umap.png\\')\"], capture_output=True, text=True)', 'purpose': 'execute python code'}, {'func': 'utils.makeNamesConsistent', 'files': [{'old-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-07-41/HSC-reprogrammed-DE-umap.h5ad', 'new-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-07-41/S1-HSC-reprogrammed-DE-umap.h5ad'}, {'old-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-07-41/umapumap.png', 'new-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-07-41/S1-umapumap.png'}]}]}\n",
      "2024-08-03 01:08:27,791 - INFO - current llm calls=3\n",
      "2024-08-03 01:08:27,793 - INFO - CODE\n",
      "2024-08-03 01:08:27,794 - INFO - CODER\n",
      "2024-08-03 01:08:27,795 - INFO - input_variables=['user_query'] template=\"You must select which code to run to help a user.\\n\\n**Available Scripts**\\nScript Name: scanpy_brad, \\t Description: This script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\\nScript Name: models_brad, \\t Description: This script executes a series of custom commands to build a model for predicting the contribution to reprogramming obtained by the inclusion or exclusion of individual TFs using an AnnData object or single cell datasets loaded from a .h5ad file.\\n\\n\\n**User Query**\\n{user_query}\\n\\n**Task**\\nBased on the user's query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\\n\\n**Response Template**\\nSCRIPT: <selected script>\\nREASON: <reasoning why the script fits the user prompt>\\n\"\n",
      "2024-08-03 01:08:27,795 - INFO - FIRST LLM CALL\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "==================================================\n",
      "BRAD >> 2: \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 01:08:29,408 - INFO - HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2024-08-03 01:08:29,409 - INFO - SCRIPT: scanpy_brad\n",
      "REASON: The user's query involves computing pairwise Euclidean distances between embeddings extracted from a .h5ad file and creating visualizations of these distances. The `scanpy_brad` script is the best match because it allows for manipulation, visualization, and exploration of single-cell datasets using libraries like scanpy, seaborn, and matplotlib.pyplot. These tools are well-suited for computing distances and generating the required visualizations. The `models_brad` script, on the other hand, is focused on building predictive models and does not align with the user's request for distance computation and visualization.\n",
      "2024-08-03 01:08:29,409 - INFO - scriptPath=/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts\n",
      "2024-08-03 01:08:29,409 - INFO - ALL SCRIPTS FOUND. BUILDING TEMPLATE\n",
      "2024-08-03 01:08:29,410 - INFO - input_variables=['history', 'input'] template='Current conversation:\\n{history}\\n\\n**PYTHON SCRIPT**\\nYou must run this python script:\\n/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\n\\n**PYTHON SCRIPT DOCUMENTATION**:\\nThis is the doc string of this python script:\\n\"\"\"\\nThis script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\\n\\nArguments (four arguments):\\n    1. output directory: chatstatus[\\'output-directory\\']\\n    2. output file: <name of output file>\\n    3. input file: <file created in previous step>\\n    4. scanpy commands: a list of scanpy commands to be executed on the AnnData object (provided as a single string with commands separated by a delimiter, e.g., \\';\\')\\n\\nBased on the arguments, the input file will be loaded, then your commands will be executed, and finally, the output or resulting ann data object will be saved to the corrrectou output file and directory. Your code is not responsible for loading the .h5ad object, that will happen automatically, and when loaded, the object will be called adata. Your scanpy commands can operate directly on the adata object that will be loaded for you.\\n\\nAdditionally, the following imports are already provided for you and can be used in your code:\\n```\\nimport scanpy as sc\\nimport seaborn as sns\\nimport matplotlib.pyplot as plt\\nimport os\\nimport numpy as np\\nimport pandas as pd\\nfrom scipy.spatial.distance import cdist\\n```\\n\\n**Usage**\\nBRAD Line:\\n```\\nsubprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\'output-directory\\'], <output file>, <input file>, \"<scanpy commands>\"], capture_output=True, text=True)\\n```\\n\\n**Examples**\\nUse the below examples to help generate your code.\\n\\n*Example 1*\\nUser Prompt: Run scanpy preprocessing and UMAP visualization on XXX.h5ad and save the UMAP plot\\nResponse Code:\\n```\\nresponse = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"XXX-modified.h5ad\", \"<path/to/data>/XXX.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save=\\'umap.png\\')\"], capture_output=True, text=True)\\n```\\nExplination: the adata object will be loaded in memory already. The command \"sc.pp.neighbors(adata)\" will preprocess the data, then the command \"sc.tl.umap(adata)\" will perform UMAP, and finally the command \"sc.pl.umap(adata, \\'umap.png\\')\" will save the UMAP to a well named file.\\n\\n*Example 2*\\nUser Prompt: Perform PCA and clustering on the dataset YYY.h5ad and save the PCA plot\\nResponse Code:\\n```\\nresponse = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"YYY-modified.h5ad\", \"<path/to/data>/YYY.h5ad\", \"sc.pp.pca(adata); sc.tl.leiden(adata); sc.pl.pca(adata, save=\\'pca.png\\')\"], capture_output=True, text=True)\\n```\\nExplination: the adata object will be loaded in memory already. The command \"sc.pp.pca(adata)\" will preprocess the data, then the command \"sc.tl.leiden(adata)\" will perform the leiden algorithm, and finally the command \"sc.pl.pca(adata, save=\\'pca.png\\')\" will save the PCA to a well named file.\\n\\n*Example 3*\\nUser Prompt: compute the distance between all cells that are not source or target to all source and target cells.\\nReponse Code:\\n```\\nadata_X = adata.X\\nmask = ~adata.obs[\\'recipe\\'].isin([\\'source\\', \\'target\\'])\\nrecipe_embeddings = adata[mask]\\nsource_embeddings = adata[adata.obs[\\'recipe\\'] == \\'source\\']\\ntarget_embeddings = adata[adata.obs[\\'recipe\\'] == \\'target\\']\\nrecipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X)\\nrecipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X)\\nadata_dist = sc.AnnData(X=np.concatenate((recipe_distances_source, recipe_distances_target), axis=1), var=pd.concat((source_embeddings.obs, target_embeddings.obs)), obs=recipe_embeddings.obs)\\nadata_dist.write(<file name>.h5ad)\\n```\\n\\n**OUTPUT FILE NAME INSTRUCTIONS**\\n1. Output path should be chatstatus[\\'output-directory\\']\\n2. Output file name should be `<descriptive name>.h5ad`\\n\"\"\"\\n\\n\\n**CALL PYTHON SCRIPTS FROM PYTHON**:\\nUse the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\\n\\nTo call a Python script `example_script.py` which has no arguments:\\n\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\']], capture_output=True, text=True)\\n```\\n\\nTo call a Python script `example_script.py` with one argument:\\n\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\'], \\'arg1\\'], capture_output=True, text=True)\\n```\\n\\nTo call a Python script `example_script.py` with two arguments:\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\'], \\'arg1\\', \\'arg2\\'], capture_output=True, text=True)\\n```\\n\\nNote that chatstatus[\\'output-directory\\'] is ALWAYS passed as the first argument.\\n\\nThe following files were previously created by BRAD and could be used as input to a function if necessary:\\nlog.json\\nS1-HSC-reprogrammed-DE-umap.h5ad\\nS1-umapumap.png\\n\\nQuery:{input}\\n\\n**PYTHON SCRIPT**\\nPlease use the correct script name and path. You must run this python script:\\n/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\n\\n**INSTRUCTIONS**\\n1. Given the user query and the documentation, identify each of the arguments found in the user\\'s query that should be passed to the Python script.\\n2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\\n3. The last line of your response should say \"Execute: <Python code to execute>\"\\n4. Do not include any extra words or characters. Format the response/output as:\\n    Arguments: \\n    Python Code Explanation: <2 sentences maximum>\\n    Execute: <your code here>\\n\\n**IMPORTANT**\\nThe code to execute from your response must be formatted as:\\n    Execute: subprocess.call([sys.executable, \\'<path to python script>\\', \\'<argument 1>\\', \\'<argument 2>\\', ..., \\'<argument n>\\'], capture_output=True, text=True))\\nThis output should be exactly one line and no longer. Stop the response after this line.\\n'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation:\n",
      "\n",
      "\n",
      "**PYTHON SCRIPT**\n",
      "You must run this python script:\n",
      "/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\n",
      "\n",
      "**PYTHON SCRIPT DOCUMENTATION**:\n",
      "This is the doc string of this python script:\n",
      "\"\"\"\n",
      "This script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\n",
      "\n",
      "Arguments (four arguments):\n",
      "    1. output directory: chatstatus['output-directory']\n",
      "    2. output file: <name of output file>\n",
      "    3. input file: <file created in previous step>\n",
      "    4. scanpy commands: a list of scanpy commands to be executed on the AnnData object (provided as a single string with commands separated by a delimiter, e.g., ';')\n",
      "\n",
      "Based on the arguments, the input file will be loaded, then your commands will be executed, and finally, the output or resulting ann data object will be saved to the corrrectou output file and directory. Your code is not responsible for loading the .h5ad object, that will happen automatically, and when loaded, the object will be called adata. Your scanpy commands can operate directly on the adata object that will be loaded for you.\n",
      "\n",
      "Additionally, the following imports are already provided for you and can be used in your code:\n",
      "```\n",
      "import scanpy as sc\n",
      "import seaborn as sns\n",
      "import matplotlib.pyplot as plt\n",
      "import os\n",
      "import numpy as np\n",
      "import pandas as pd\n",
      "from scipy.spatial.distance import cdist\n",
      "```\n",
      "\n",
      "**Usage**\n",
      "BRAD Line:\n",
      "```\n",
      "subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus['output-directory'], <output file>, <input file>, \"<scanpy commands>\"], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "**Examples**\n",
      "Use the below examples to help generate your code.\n",
      "\n",
      "*Example 1*\n",
      "User Prompt: Run scanpy preprocessing and UMAP visualization on XXX.h5ad and save the UMAP plot\n",
      "Response Code:\n",
      "```\n",
      "response = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus['output-directory'], \"XXX-modified.h5ad\", \"<path/to/data>/XXX.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save='umap.png')\"], capture_output=True, text=True)\n",
      "```\n",
      "Explination: the adata object will be loaded in memory already. The command \"sc.pp.neighbors(adata)\" will preprocess the data, then the command \"sc.tl.umap(adata)\" will perform UMAP, and finally the command \"sc.pl.umap(adata, 'umap.png')\" will save the UMAP to a well named file.\n",
      "\n",
      "*Example 2*\n",
      "User Prompt: Perform PCA and clustering on the dataset YYY.h5ad and save the PCA plot\n",
      "Response Code:\n",
      "```\n",
      "response = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus['output-directory'], \"YYY-modified.h5ad\", \"<path/to/data>/YYY.h5ad\", \"sc.pp.pca(adata); sc.tl.leiden(adata); sc.pl.pca(adata, save='pca.png')\"], capture_output=True, text=True)\n",
      "```\n",
      "Explination: the adata object will be loaded in memory already. The command \"sc.pp.pca(adata)\" will preprocess the data, then the command \"sc.tl.leiden(adata)\" will perform the leiden algorithm, and finally the command \"sc.pl.pca(adata, save='pca.png')\" will save the PCA to a well named file.\n",
      "\n",
      "*Example 3*\n",
      "User Prompt: compute the distance between all cells that are not source or target to all source and target cells.\n",
      "Reponse Code:\n",
      "```\n",
      "adata_X = adata.X\n",
      "mask = ~adata.obs['recipe'].isin(['source', 'target'])\n",
      "recipe_embeddings = adata[mask]\n",
      "source_embeddings = adata[adata.obs['recipe'] == 'source']\n",
      "target_embeddings = adata[adata.obs['recipe'] == 'target']\n",
      "recipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X)\n",
      "recipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X)\n",
      "adata_dist = sc.AnnData(X=np.concatenate((recipe_distances_source, recipe_distances_target), axis=1), var=pd.concat((source_embeddings.obs, target_embeddings.obs)), obs=recipe_embeddings.obs)\n",
      "adata_dist.write(<file name>.h5ad)\n",
      "```\n",
      "\n",
      "**OUTPUT FILE NAME INSTRUCTIONS**\n",
      "1. Output path should be chatstatus['output-directory']\n",
      "2. Output file name should be `<descriptive name>.h5ad`\n",
      "\"\"\"\n",
      "\n",
      "\n",
      "**CALL PYTHON SCRIPTS FROM PYTHON**:\n",
      "Use the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\n",
      "\n",
      "To call a Python script `example_script.py` which has no arguments:\n",
      "\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory']], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with one argument:\n",
      "\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory'], 'arg1'], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with two arguments:\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory'], 'arg1', 'arg2'], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "Note that chatstatus['output-directory'] is ALWAYS passed as the first argument.\n",
      "\n",
      "The following files were previously created by BRAD and could be used as input to a function if necessary:\n",
      "log.json\n",
      "S1-HSC-reprogrammed-DE-umap.h5ad\n",
      "S1-umapumap.png\n",
      "\n",
      "Query:2 (CODE)**:\n",
      "Prompt: Compute the pairwise Euclidean distances between each 'source' embedding and each recipe embedding, using the extracted embeddings from the .h5ad file: `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` Plot histograms and or other visualizations of the distances.\n",
      "\n",
      "**PYTHON SCRIPT**\n",
      "Please use the correct script name and path. You must run this python script:\n",
      "/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\n",
      "\n",
      "**INSTRUCTIONS**\n",
      "1. Given the user query and the documentation, identify each of the arguments found in the user's query that should be passed to the Python script.\n",
      "2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\n",
      "3. The last line of your response should say \"Execute: <Python code to execute>\"\n",
      "4. Do not include any extra words or characters. Format the response/output as:\n",
      "    Arguments: \n",
      "    Python Code Explanation: <2 sentences maximum>\n",
      "    Execute: <your code here>\n",
      "\n",
      "**IMPORTANT**\n",
      "The code to execute from your response must be formatted as:\n",
      "    Execute: subprocess.call([sys.executable, '<path to python script>', '<argument 1>', '<argument 2>', ..., '<argument n>'], capture_output=True, text=True))\n",
      "This output should be exactly one line and no longer. Stop the response after this line.\n",
      "\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 01:08:34,673 - INFO - HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2024-08-03 01:08:34,674 - INFO - LLM OUTPUT PARSER\n",
      "2024-08-03 01:08:34,675 - INFO - Arguments:\n",
      "1. output directory: chatstatus['output-directory']\n",
      "2. output file: \"HSC-reprogrammed-DE-distances.h5ad\"\n",
      "3. input file: \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\"\n",
      "4. scanpy commands: \"mask = ~adata.obs['recipe'].isin(['source']); recipe_embeddings = adata[mask]; source_embeddings = adata[adata.obs['recipe'] == 'source']; recipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X); plt.hist(recipe_distances_source.flatten(), bins=50); plt.savefig('distances_histogram.png'); adata_dist = sc.AnnData(X=recipe_distances_source, var=source_embeddings.obs, obs=recipe_embeddings.obs); adata_dist.write('HSC-reprogrammed-DE-distances.h5ad')\"\n",
      "\n",
      "Python Code Explanation: This code computes the pairwise Euclidean distances between 'source' and recipe embeddings, plots a histogram of the distances, and saves the results.\n",
      "\n",
      "Execute: subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-DE-distances.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"mask = ~adata.obs['recipe'].isin([\\'source\\']); recipe_embeddings = adata[mask]; source_embeddings = adata[adata.obs[\\'recipe\\'] == \\'source\\']; recipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X); plt.hist(recipe_distances_source.flatten(), bins=50); plt.savefig(\\'distances_histogram.png\\'); adata_dist = sc.AnnData(X=recipe_distances_source, var=source_embeddings.obs, obs=recipe_embeddings.obs); adata_dist.write(\\'HSC-reprogrammed-DE-distances.h5ad\\')\"], capture_output=True, text=True)\n",
      "2024-08-03 01:08:34,675 - INFO - ['Arguments:\\n1. output directory: chatstatus[\\'output-directory\\']\\n2. output file: \"HSC-reprogrammed-DE-distances.h5ad\"\\n3. input file: \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\"\\n4. scanpy commands: \"mask = ~adata.obs[\\'recipe\\'].isin([\\'source\\']); recipe_embeddings = adata[mask]; source_embeddings = adata[adata.obs[\\'recipe\\'] == \\'source\\']; recipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X); plt.hist(recipe_distances_source.flatten(), bins=50); plt.savefig(\\'distances_histogram.png\\'); adata_dist = sc.AnnData(X=recipe_distances_source, var=source_embeddings.obs, obs=recipe_embeddings.obs); adata_dist.write(\\'HSC-reprogrammed-DE-distances.h5ad\\')\"\\n\\nPython Code Explanation: This code computes the pairwise Euclidean distances between \\'source\\' and recipe embeddings, plots a histogram of the distances, and saves the results.\\n\\n', ' subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"HSC-reprogrammed-DE-distances.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"mask = ~adata.obs[\\'recipe\\'].isin([\\\\\\'source\\\\\\']); recipe_embeddings = adata[mask]; source_embeddings = adata[adata.obs[\\\\\\'recipe\\\\\\'] == \\\\\\'source\\\\\\']; recipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X); plt.hist(recipe_distances_source.flatten(), bins=50); plt.savefig(\\\\\\'distances_histogram.png\\\\\\'); adata_dist = sc.AnnData(X=recipe_distances_source, var=source_embeddings.obs, obs=recipe_embeddings.obs); adata_dist.write(\\\\\\'HSC-reprogrammed-DE-distances.h5ad\\\\\\')\"], capture_output=True, text=True)']\n",
      "2024-08-03 01:08:34,675 - INFO - Stripped funciton call\n",
      "2024-08-03 01:08:34,675 - INFO - subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-DE-distances.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"mask = ~adata.obs['recipe'].isin([\\'source\\']); recipe_embeddings = adata[mask]; source_embeddings = adata[adata.obs[\\'recipe\\'] == \\'source\\']; recipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X); plt.hist(recipe_distances_source.flatten(), bins=50); plt.savefig(\\'distances_histogram.png\\'); adata_dist = sc.AnnData(X=recipe_distances_source, var=source_embeddings.obs, obs=recipe_embeddings.obs); adata_dist.write(\\'HSC-reprogrammed-DE-distances.h5ad\\')\"], capture_output=True, text=True)\n",
      "2024-08-03 01:08:34,676 - INFO - Cleaned up function call:\n",
      "\n",
      "2024-08-03 01:08:34,676 - INFO - subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-DE-distances.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"mask = ~adata.obs['recipe'].isin([\\'source\\']); recipe_embeddings = adata[mask]; source_embeddings = adata[adata.obs[\\'recipe\\'] == \\'source\\']; recipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X); plt.hist(recipe_distances_source.flatten(), bins=50); plt.savefig(\\'distances_histogram.png\\'); adata_dist = sc.AnnData(X=recipe_distances_source, var=source_embeddings.obs, obs=recipe_embeddings.obs); adata_dist.write(\\'HSC-reprogrammed-DE-distances.h5ad\\')\"], capture_output=True, text=True)\n",
      "2024-08-03 01:08:34,676 - INFO - \n",
      "\n",
      "2024-08-03 01:08:34,677 - INFO - EVAL\n",
      "2024-08-03 01:08:34,677 - INFO - subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-DE-distances.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"mask = ~adata.obs['recipe'].isin([\\'source\\']); recipe_embeddings = adata[mask]; source_embeddings = adata[adata.obs[\\'recipe\\'] == \\'source\\']; recipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X); plt.hist(recipe_distances_source.flatten(), bins=50); plt.savefig(\\'distances_histogram.png\\'); adata_dist = sc.AnnData(X=recipe_distances_source, var=source_embeddings.obs, obs=recipe_embeddings.obs); adata_dist.write(\\'HSC-reprogrammed-DE-distances.h5ad\\')\"], capture_output=True, text=True)\n",
      "2024-08-03 01:08:34,678 - INFO - END EVAL CHECK\n",
      "2024-08-03 01:08:34,678 - INFO - Finalized PYTHON Call:\n",
      "2024-08-03 01:08:34,678 - INFO -    \n",
      "2024-08-03 01:08:34,678 - INFO - response = subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-DE-distances.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"mask = ~adata.obs['recipe'].isin([\\'source\\']); recipe_embeddings = adata[mask]; source_embeddings = adata[adata.obs[\\'recipe\\'] == \\'source\\']; recipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X); plt.hist(recipe_distances_source.flatten(), bins=50); plt.savefig(\\'distances_histogram.png\\'); adata_dist = sc.AnnData(X=recipe_distances_source, var=source_embeddings.obs, obs=recipe_embeddings.obs); adata_dist.write(\\'HSC-reprogrammed-DE-distances.h5ad\\')\"], capture_output=True, text=True)\n",
      "2024-08-03 01:08:34,679 - INFO -    \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 01:08:39,693 - INFO - Debug: PYTHON code executed successfully.\n",
      "2024-08-03 01:08:39,694 - INFO - Code Execution Output:\n",
      "2024-08-03 01:08:39,694 - INFO - CompletedProcess(args=['/sw/pkgs/arc/python3.11-anaconda/2024.02-1/bin/python', '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py', '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-07-41', 'HSC-reprogrammed-DE-distances.h5ad', '/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad', \"mask = ~adata.obs['recipe'].isin(['source']); recipe_embeddings = adata[mask]; source_embeddings = adata[adata.obs['recipe'] == 'source']; recipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X); plt.hist(recipe_distances_source.flatten(), bins=50); plt.savefig('distances_histogram.png'); adata_dist = sc.AnnData(X=recipe_distances_source, var=source_embeddings.obs, obs=recipe_embeddings.obs); adata_dist.write('HSC-reprogrammed-DE-distances.h5ad')\"], returncode=0, stdout=\"adata loaded from /nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\\nadata.shape=(6375, 512)\\nadata.obs.columns=Index(['recipe', 'filename', 'REL', 'STAT5A', 'FOS', 'ETV6', 'source', 'MEF2C',\\n       'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B', 'target'],\\n      dtype='object')\\nadata.obs.head()=   recipe                                           filename  ...  GFI1B  target\\n0  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n1  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n2  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n3  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n4  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n\\n[5 rows x 14 columns]\\nadata.var.head()=Empty DataFrame\\nColumns: []\\nIndex: [0, 1, 2, 3, 4]\\n****************************\\n      EXECUTE COMMANDS      \\n****************************\\nmask = ~adata.obs['recipe'].isin(['source'])\\nrecipe_embeddings = adata[mask]\\nsource_embeddings = adata[adata.obs['recipe'] == 'source']\\nrecipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X)\\nplt.hist(recipe_distances_source.flatten(), bins=50)\\nplt.savefig('distances_histogram.png')\\nadata_dist = sc.AnnData(X=recipe_distances_source, var=source_embeddings.obs, obs=recipe_embeddings.obs)\\nadata_dist.write('HSC-reprogrammed-DE-distances.h5ad')\\n\", stderr='/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\\n  utils.warn_names_duplicates(\"obs\")\\n/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\\n  utils.warn_names_duplicates(\"obs\")\\n')\n",
      "2024-08-03 01:08:39,694 - INFO - Debug: PYTHON code output saved to output.\n",
      "2024-08-03 01:08:39,695 - INFO - \n",
      "\n",
      "\n",
      "route\n",
      "\n",
      "\n",
      "\n",
      "2024-08-03 01:08:39,695 - INFO - CODE\n",
      "2024-08-03 01:08:39,696 - INFO - Finding Stage Number of Pipeline\n",
      "2024-08-03 01:08:39,696 - INFO - {0: {'order': 0, 'module': 'PLANNER', 'prompt': None, 'description': 'This step designed the plan. It is placed in the queue because we needed a place holder for 0 indexed lists.', 'output': []}, 1: {'order': 1, 'module': 'CODE', 'prompt': \"/force CODE  1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings for 'source', 'target', and each recipe from `adata.X`. **Generate a UMAP plot to visualize the embeddings.**\\n\\n**.\", 'description': \" 1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings for 'source', 'target', and each recipe from `adata.X`. **Generate a UMAP plot to visualize the embeddings.**\\n\\n**\", 'output': ['/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-07-41/S1-HSC-reprogrammed-DE-umap.h5ad', '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-07-41/S1-umapumap.png']}, 2: {'order': 2, 'module': 'CODE', 'prompt': \"/force CODE  2 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'source' embedding and each recipe embedding, using the extracted embeddings from the .h5ad file: `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` Plot histograms and or other visualizations of the distances.\", 'description': \" 2 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'source' embedding and each recipe embedding, using the extracted embeddings from \", 'output': []}, 3: {'order': 4, 'module': 'CODE', 'prompt': \"/force CODE  2 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'target' embedding and each recipe embedding, using the extracted embeddings from the .h5ad file: `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` Plot histograms and or other visualizations of the distances.\", 'description': \" 3 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'target' embedding and each recipe embedding, using the extracted embeddings from \", 'output': []}, 4: {'order': 6, 'module': 'CODE', 'prompt': '/force CODE  4 (CODE)**:\\nPrompt: Store the computed distances in a matrix format and create a new ann data object, setting `.var` to be the `.obs` of the source and target cells used in the distance calculations and `.obs` to be the `.obs` of the reprogrammed cells used in the distance calculations. Explicitly save the new ann data object. **Generate a summary plot to visualize the new ann data object.**\\n\\nThe revised plan removes ', 'description': ' 4 (CODE)**:\\nPrompt: Store the computed distances in a matrix format and create a new ann data object, setting `.var` to be the `.obs` of the source and target cells used in the distance calculations and `.obs` to be the `.obs` of the reprogrammed cells used in the distance calculations. Explicitly save the new ann data object. **Generate a summary plot to visualize the new ann data object.**\\n\\nThe revised plan removes ', 'output': []}}\n",
      "2024-08-03 01:08:39,697 - INFO - IP=2\n",
      "2024-08-03 01:08:39,697 - INFO - type(IP)=<class 'int'>\n",
      "2024-08-03 01:08:39,699 - INFO - \n",
      "\n",
      "PLANNER:\n",
      "\n",
      "{'module': 'CODE', 'steps': [{'llm': \"client=<openai.resources.chat.completions.Completions object at 0x147f010bd150> async_client=<openai.resources.chat.completions.AsyncCompletions object at 0x147f010bf650> model_name='gpt-4o' temperature=0.0 openai_api_key=SecretStr('**********') openai_proxy=''\", 'memory': 'None', 'prompt': 'input_variables=[\\'user_query\\'] template=\"You must select which code to run to help a user.\\\\n\\\\n**Available Scripts**\\\\nScript Name: scanpy_brad, \\\\t Description: This script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\\\\nScript Name: models_brad, \\\\t Description: This script executes a series of custom commands to build a model for predicting the contribution to reprogramming obtained by the inclusion or exclusion of individual TFs using an AnnData object or single cell datasets loaded from a .h5ad file.\\\\n\\\\n\\\\n**User Query**\\\\n{user_query}\\\\n\\\\n**Task**\\\\nBased on the user\\'s query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\\\\n\\\\n**Response Template**\\\\nSCRIPT: <selected script>\\\\nREASON: <reasoning why the script fits the user prompt>\\\\n\"', 'input': \"2 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'source' embedding and each recipe embedding, using the extracted embeddings from the .h5ad file: `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` Plot histograms and or other visualizations of the distances.\", 'output': 'content=\"SCRIPT: scanpy_brad\\\\nREASON: The user\\'s query involves computing pairwise Euclidean distances between embeddings extracted from a .h5ad file and creating visualizations of these distances. The `scanpy_brad` script is the best match because it allows for manipulation, visualization, and exploration of single-cell datasets using libraries like scanpy, seaborn, and matplotlib.pyplot. These tools are well-suited for computing distances and generating the required visualizations. The `models_brad` script, on the other hand, is focused on building predictive models and does not align with the user\\'s request for distance computation and visualization.\" response_metadata={\\'token_usage\\': {\\'completion_tokens\\': 126, \\'prompt_tokens\\': 330, \\'total_tokens\\': 456}, \\'model_name\\': \\'gpt-4o-2024-05-13\\', \\'system_fingerprint\\': \\'fp_4e2b2da518\\', \\'finish_reason\\': \\'stop\\', \\'logprobs\\': None} id=\\'run-ee01f860-067d-478e-867e-7ae32cc9884a-0\\' usage_metadata={\\'input_tokens\\': 330, \\'output_tokens\\': 126, \\'total_tokens\\': 456}', 'parsedOutput': {'scriptName': '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py', 'scriptType': 'python', 'scriptPath': '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts'}, 'purpose': 'Select which code to run'}, {'llm': \"client=<openai.resources.chat.completions.Completions object at 0x147f010bd150> async_client=<openai.resources.chat.completions.AsyncCompletions object at 0x147f010bf650> model_name='gpt-4o' temperature=0.0 openai_api_key=SecretStr('**********') openai_proxy=''\", 'memory': 'None', 'prompt': 'input_variables=[\\'history\\', \\'input\\'] template=\\'Current conversation:\\\\n{history}\\\\n\\\\n**PYTHON SCRIPT**\\\\nYou must run this python script:\\\\n/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\\\n\\\\n**PYTHON SCRIPT DOCUMENTATION**:\\\\nThis is the doc string of this python script:\\\\n\"\"\"\\\\nThis script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\\\\n\\\\nArguments (four arguments):\\\\n    1. output directory: chatstatus[\\\\\\'output-directory\\\\\\']\\\\n    2. output file: <name of output file>\\\\n    3. input file: <file created in previous step>\\\\n    4. scanpy commands: a list of scanpy commands to be executed on the AnnData object (provided as a single string with commands separated by a delimiter, e.g., \\\\\\';\\\\\\')\\\\n\\\\nBased on the arguments, the input file will be loaded, then your commands will be executed, and finally, the output or resulting ann data object will be saved to the corrrectou output file and directory. Your code is not responsible for loading the .h5ad object, that will happen automatically, and when loaded, the object will be called adata. Your scanpy commands can operate directly on the adata object that will be loaded for you.\\\\n\\\\nAdditionally, the following imports are already provided for you and can be used in your code:\\\\n```\\\\nimport scanpy as sc\\\\nimport seaborn as sns\\\\nimport matplotlib.pyplot as plt\\\\nimport os\\\\nimport numpy as np\\\\nimport pandas as pd\\\\nfrom scipy.spatial.distance import cdist\\\\n```\\\\n\\\\n**Usage**\\\\nBRAD Line:\\\\n```\\\\nsubprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\\\\\'output-directory\\\\\\'], <output file>, <input file>, \"<scanpy commands>\"], capture_output=True, text=True)\\\\n```\\\\n\\\\n**Examples**\\\\nUse the below examples to help generate your code.\\\\n\\\\n*Example 1*\\\\nUser Prompt: Run scanpy preprocessing and UMAP visualization on XXX.h5ad and save the UMAP plot\\\\nResponse Code:\\\\n```\\\\nresponse = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\\\\\'output-directory\\\\\\'], \"XXX-modified.h5ad\", \"<path/to/data>/XXX.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save=\\\\\\'umap.png\\\\\\')\"], capture_output=True, text=True)\\\\n```\\\\nExplination: the adata object will be loaded in memory already. The command \"sc.pp.neighbors(adata)\" will preprocess the data, then the command \"sc.tl.umap(adata)\" will perform UMAP, and finally the command \"sc.pl.umap(adata, \\\\\\'umap.png\\\\\\')\" will save the UMAP to a well named file.\\\\n\\\\n*Example 2*\\\\nUser Prompt: Perform PCA and clustering on the dataset YYY.h5ad and save the PCA plot\\\\nResponse Code:\\\\n```\\\\nresponse = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\\\\\'output-directory\\\\\\'], \"YYY-modified.h5ad\", \"<path/to/data>/YYY.h5ad\", \"sc.pp.pca(adata); sc.tl.leiden(adata); sc.pl.pca(adata, save=\\\\\\'pca.png\\\\\\')\"], capture_output=True, text=True)\\\\n```\\\\nExplination: the adata object will be loaded in memory already. The command \"sc.pp.pca(adata)\" will preprocess the data, then the command \"sc.tl.leiden(adata)\" will perform the leiden algorithm, and finally the command \"sc.pl.pca(adata, save=\\\\\\'pca.png\\\\\\')\" will save the PCA to a well named file.\\\\n\\\\n*Example 3*\\\\nUser Prompt: compute the distance between all cells that are not source or target to all source and target cells.\\\\nReponse Code:\\\\n```\\\\nadata_X = adata.X\\\\nmask = ~adata.obs[\\\\\\'recipe\\\\\\'].isin([\\\\\\'source\\\\\\', \\\\\\'target\\\\\\'])\\\\nrecipe_embeddings = adata[mask]\\\\nsource_embeddings = adata[adata.obs[\\\\\\'recipe\\\\\\'] == \\\\\\'source\\\\\\']\\\\ntarget_embeddings = adata[adata.obs[\\\\\\'recipe\\\\\\'] == \\\\\\'target\\\\\\']\\\\nrecipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X)\\\\nrecipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X)\\\\nadata_dist = sc.AnnData(X=np.concatenate((recipe_distances_source, recipe_distances_target), axis=1), var=pd.concat((source_embeddings.obs, target_embeddings.obs)), obs=recipe_embeddings.obs)\\\\nadata_dist.write(<file name>.h5ad)\\\\n```\\\\n\\\\n**OUTPUT FILE NAME INSTRUCTIONS**\\\\n1. Output path should be chatstatus[\\\\\\'output-directory\\\\\\']\\\\n2. Output file name should be `<descriptive name>.h5ad`\\\\n\"\"\"\\\\n\\\\n\\\\n**CALL PYTHON SCRIPTS FROM PYTHON**:\\\\nUse the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\\\\n\\\\nTo call a Python script `example_script.py` which has no arguments:\\\\n\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\']], capture_output=True, text=True)\\\\n```\\\\n\\\\nTo call a Python script `example_script.py` with one argument:\\\\n\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\'], \\\\\\'arg1\\\\\\'], capture_output=True, text=True)\\\\n```\\\\n\\\\nTo call a Python script `example_script.py` with two arguments:\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\'], \\\\\\'arg1\\\\\\', \\\\\\'arg2\\\\\\'], capture_output=True, text=True)\\\\n```\\\\n\\\\nNote that chatstatus[\\\\\\'output-directory\\\\\\'] is ALWAYS passed as the first argument.\\\\n\\\\nThe following files were previously created by BRAD and could be used as input to a function if necessary:\\\\nlog.json\\\\nS1-HSC-reprogrammed-DE-umap.h5ad\\\\nS1-umapumap.png\\\\n\\\\nQuery:{input}\\\\n\\\\n**PYTHON SCRIPT**\\\\nPlease use the correct script name and path. You must run this python script:\\\\n/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\\\n\\\\n**INSTRUCTIONS**\\\\n1. Given the user query and the documentation, identify each of the arguments found in the user\\\\\\'s query that should be passed to the Python script.\\\\n2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\\\\n3. The last line of your response should say \"Execute: <Python code to execute>\"\\\\n4. Do not include any extra words or characters. Format the response/output as:\\\\n    Arguments: \\\\n    Python Code Explanation: <2 sentences maximum>\\\\n    Execute: <your code here>\\\\n\\\\n**IMPORTANT**\\\\nThe code to execute from your response must be formatted as:\\\\n    Execute: subprocess.call([sys.executable, \\\\\\'<path to python script>\\\\\\', \\\\\\'<argument 1>\\\\\\', \\\\\\'<argument 2>\\\\\\', ..., \\\\\\'<argument n>\\\\\\'], capture_output=True, text=True))\\\\nThis output should be exactly one line and no longer. Stop the response after this line.\\\\n\\'', 'input': \"2 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'source' embedding and each recipe embedding, using the extracted embeddings from the .h5ad file: `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` Plot histograms and or other visualizations of the distances.\", 'output': 'Arguments:\\n1. output directory: chatstatus[\\'output-directory\\']\\n2. output file: \"HSC-reprogrammed-DE-distances.h5ad\"\\n3. input file: \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\"\\n4. scanpy commands: \"mask = ~adata.obs[\\'recipe\\'].isin([\\'source\\']); recipe_embeddings = adata[mask]; source_embeddings = adata[adata.obs[\\'recipe\\'] == \\'source\\']; recipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X); plt.hist(recipe_distances_source.flatten(), bins=50); plt.savefig(\\'distances_histogram.png\\'); adata_dist = sc.AnnData(X=recipe_distances_source, var=source_embeddings.obs, obs=recipe_embeddings.obs); adata_dist.write(\\'HSC-reprogrammed-DE-distances.h5ad\\')\"\\n\\nPython Code Explanation: This code computes the pairwise Euclidean distances between \\'source\\' and recipe embeddings, plots a histogram of the distances, and saves the results.\\n\\nExecute: subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"HSC-reprogrammed-DE-distances.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"mask = ~adata.obs[\\'recipe\\'].isin([\\\\\\'source\\\\\\']); recipe_embeddings = adata[mask]; source_embeddings = adata[adata.obs[\\\\\\'recipe\\\\\\'] == \\\\\\'source\\\\\\']; recipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X); plt.hist(recipe_distances_source.flatten(), bins=50); plt.savefig(\\\\\\'distances_histogram.png\\\\\\'); adata_dist = sc.AnnData(X=recipe_distances_source, var=source_embeddings.obs, obs=recipe_embeddings.obs); adata_dist.write(\\\\\\'HSC-reprogrammed-DE-distances.h5ad\\\\\\')\"], capture_output=True, text=True)', 'parsedOutput': {'code': 'subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"HSC-reprogrammed-DE-distances.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"mask = ~adata.obs[\\'recipe\\'].isin([\\\\\\'source\\\\\\']); recipe_embeddings = adata[mask]; source_embeddings = adata[adata.obs[\\\\\\'recipe\\\\\\'] == \\\\\\'source\\\\\\']; recipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X); plt.hist(recipe_distances_source.flatten(), bins=50); plt.savefig(\\\\\\'distances_histogram.png\\\\\\'); adata_dist = sc.AnnData(X=recipe_distances_source, var=source_embeddings.obs, obs=recipe_embeddings.obs); adata_dist.write(\\\\\\'HSC-reprogrammed-DE-distances.h5ad\\\\\\')\"], capture_output=True, text=True)'}, 'purpose': 'Format function call'}, {'func': 'pythonCaller.execute_python_code', 'code': 'response = subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"HSC-reprogrammed-DE-distances.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"mask = ~adata.obs[\\'recipe\\'].isin([\\\\\\'source\\\\\\']); recipe_embeddings = adata[mask]; source_embeddings = adata[adata.obs[\\\\\\'recipe\\\\\\'] == \\\\\\'source\\\\\\']; recipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X); plt.hist(recipe_distances_source.flatten(), bins=50); plt.savefig(\\\\\\'distances_histogram.png\\\\\\'); adata_dist = sc.AnnData(X=recipe_distances_source, var=source_embeddings.obs, obs=recipe_embeddings.obs); adata_dist.write(\\\\\\'HSC-reprogrammed-DE-distances.h5ad\\\\\\')\"], capture_output=True, text=True)', 'purpose': 'execute python code'}, {'func': 'utils.makeNamesConsistent', 'files': [{'old-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-07-41/HSC-reprogrammed-DE-distances.h5ad', 'new-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-07-41/S2-HSC-reprogrammed-DE-distances.h5ad'}]}]}\n",
      "2024-08-03 01:08:39,703 - INFO - current llm calls=5\n",
      "2024-08-03 01:08:39,704 - INFO - CODE\n",
      "2024-08-03 01:08:39,704 - INFO - CODER\n",
      "2024-08-03 01:08:39,706 - INFO - input_variables=['user_query'] template=\"You must select which code to run to help a user.\\n\\n**Available Scripts**\\nScript Name: scanpy_brad, \\t Description: This script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\\nScript Name: models_brad, \\t Description: This script executes a series of custom commands to build a model for predicting the contribution to reprogramming obtained by the inclusion or exclusion of individual TFs using an AnnData object or single cell datasets loaded from a .h5ad file.\\n\\n\\n**User Query**\\n{user_query}\\n\\n**Task**\\nBased on the user's query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\\n\\n**Response Template**\\nSCRIPT: <selected script>\\nREASON: <reasoning why the script fits the user prompt>\\n\"\n",
      "2024-08-03 01:08:39,706 - INFO - FIRST LLM CALL\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "==================================================\n",
      "BRAD >> 3: \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 01:08:41,525 - INFO - HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2024-08-03 01:08:41,526 - INFO - SCRIPT: scanpy_brad\n",
      "REASON: The user's query involves computing pairwise Euclidean distances between embeddings and creating visualizations such as histograms. The `scanpy_brad` script is the best match because it allows for manipulation, visualization, and exploration of single-cell datasets using libraries like scanpy, seaborn, and matplotlib.pyplot. These libraries are well-suited for computing distances and generating the required visualizations. The `models_brad` script, on the other hand, is focused on building predictive models and does not align with the user's request for distance computation and visualization.\n",
      "2024-08-03 01:08:41,526 - INFO - scriptPath=/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts\n",
      "2024-08-03 01:08:41,527 - INFO - ALL SCRIPTS FOUND. BUILDING TEMPLATE\n",
      "2024-08-03 01:08:41,528 - INFO - input_variables=['history', 'input'] template='Current conversation:\\n{history}\\n\\n**PYTHON SCRIPT**\\nYou must run this python script:\\n/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\n\\n**PYTHON SCRIPT DOCUMENTATION**:\\nThis is the doc string of this python script:\\n\"\"\"\\nThis script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\\n\\nArguments (four arguments):\\n    1. output directory: chatstatus[\\'output-directory\\']\\n    2. output file: <name of output file>\\n    3. input file: <file created in previous step>\\n    4. scanpy commands: a list of scanpy commands to be executed on the AnnData object (provided as a single string with commands separated by a delimiter, e.g., \\';\\')\\n\\nBased on the arguments, the input file will be loaded, then your commands will be executed, and finally, the output or resulting ann data object will be saved to the corrrectou output file and directory. Your code is not responsible for loading the .h5ad object, that will happen automatically, and when loaded, the object will be called adata. Your scanpy commands can operate directly on the adata object that will be loaded for you.\\n\\nAdditionally, the following imports are already provided for you and can be used in your code:\\n```\\nimport scanpy as sc\\nimport seaborn as sns\\nimport matplotlib.pyplot as plt\\nimport os\\nimport numpy as np\\nimport pandas as pd\\nfrom scipy.spatial.distance import cdist\\n```\\n\\n**Usage**\\nBRAD Line:\\n```\\nsubprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\'output-directory\\'], <output file>, <input file>, \"<scanpy commands>\"], capture_output=True, text=True)\\n```\\n\\n**Examples**\\nUse the below examples to help generate your code.\\n\\n*Example 1*\\nUser Prompt: Run scanpy preprocessing and UMAP visualization on XXX.h5ad and save the UMAP plot\\nResponse Code:\\n```\\nresponse = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"XXX-modified.h5ad\", \"<path/to/data>/XXX.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save=\\'umap.png\\')\"], capture_output=True, text=True)\\n```\\nExplination: the adata object will be loaded in memory already. The command \"sc.pp.neighbors(adata)\" will preprocess the data, then the command \"sc.tl.umap(adata)\" will perform UMAP, and finally the command \"sc.pl.umap(adata, \\'umap.png\\')\" will save the UMAP to a well named file.\\n\\n*Example 2*\\nUser Prompt: Perform PCA and clustering on the dataset YYY.h5ad and save the PCA plot\\nResponse Code:\\n```\\nresponse = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"YYY-modified.h5ad\", \"<path/to/data>/YYY.h5ad\", \"sc.pp.pca(adata); sc.tl.leiden(adata); sc.pl.pca(adata, save=\\'pca.png\\')\"], capture_output=True, text=True)\\n```\\nExplination: the adata object will be loaded in memory already. The command \"sc.pp.pca(adata)\" will preprocess the data, then the command \"sc.tl.leiden(adata)\" will perform the leiden algorithm, and finally the command \"sc.pl.pca(adata, save=\\'pca.png\\')\" will save the PCA to a well named file.\\n\\n*Example 3*\\nUser Prompt: compute the distance between all cells that are not source or target to all source and target cells.\\nReponse Code:\\n```\\nadata_X = adata.X\\nmask = ~adata.obs[\\'recipe\\'].isin([\\'source\\', \\'target\\'])\\nrecipe_embeddings = adata[mask]\\nsource_embeddings = adata[adata.obs[\\'recipe\\'] == \\'source\\']\\ntarget_embeddings = adata[adata.obs[\\'recipe\\'] == \\'target\\']\\nrecipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X)\\nrecipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X)\\nadata_dist = sc.AnnData(X=np.concatenate((recipe_distances_source, recipe_distances_target), axis=1), var=pd.concat((source_embeddings.obs, target_embeddings.obs)), obs=recipe_embeddings.obs)\\nadata_dist.write(<file name>.h5ad)\\n```\\n\\n**OUTPUT FILE NAME INSTRUCTIONS**\\n1. Output path should be chatstatus[\\'output-directory\\']\\n2. Output file name should be `<descriptive name>.h5ad`\\n\"\"\"\\n\\n\\n**CALL PYTHON SCRIPTS FROM PYTHON**:\\nUse the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\\n\\nTo call a Python script `example_script.py` which has no arguments:\\n\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\']], capture_output=True, text=True)\\n```\\n\\nTo call a Python script `example_script.py` with one argument:\\n\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\'], \\'arg1\\'], capture_output=True, text=True)\\n```\\n\\nTo call a Python script `example_script.py` with two arguments:\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\'], \\'arg1\\', \\'arg2\\'], capture_output=True, text=True)\\n```\\n\\nNote that chatstatus[\\'output-directory\\'] is ALWAYS passed as the first argument.\\n\\nThe following files were previously created by BRAD and could be used as input to a function if necessary:\\nlog.json\\nS1-HSC-reprogrammed-DE-umap.h5ad\\nS1-umapumap.png\\nS2-HSC-reprogrammed-DE-distances.h5ad\\n\\nQuery:{input}\\n\\n**PYTHON SCRIPT**\\nPlease use the correct script name and path. You must run this python script:\\n/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\n\\n**INSTRUCTIONS**\\n1. Given the user query and the documentation, identify each of the arguments found in the user\\'s query that should be passed to the Python script.\\n2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\\n3. The last line of your response should say \"Execute: <Python code to execute>\"\\n4. Do not include any extra words or characters. Format the response/output as:\\n    Arguments: \\n    Python Code Explanation: <2 sentences maximum>\\n    Execute: <your code here>\\n\\n**IMPORTANT**\\nThe code to execute from your response must be formatted as:\\n    Execute: subprocess.call([sys.executable, \\'<path to python script>\\', \\'<argument 1>\\', \\'<argument 2>\\', ..., \\'<argument n>\\'], capture_output=True, text=True))\\nThis output should be exactly one line and no longer. Stop the response after this line.\\n'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation:\n",
      "\n",
      "\n",
      "**PYTHON SCRIPT**\n",
      "You must run this python script:\n",
      "/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\n",
      "\n",
      "**PYTHON SCRIPT DOCUMENTATION**:\n",
      "This is the doc string of this python script:\n",
      "\"\"\"\n",
      "This script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\n",
      "\n",
      "Arguments (four arguments):\n",
      "    1. output directory: chatstatus['output-directory']\n",
      "    2. output file: <name of output file>\n",
      "    3. input file: <file created in previous step>\n",
      "    4. scanpy commands: a list of scanpy commands to be executed on the AnnData object (provided as a single string with commands separated by a delimiter, e.g., ';')\n",
      "\n",
      "Based on the arguments, the input file will be loaded, then your commands will be executed, and finally, the output or resulting ann data object will be saved to the corrrectou output file and directory. Your code is not responsible for loading the .h5ad object, that will happen automatically, and when loaded, the object will be called adata. Your scanpy commands can operate directly on the adata object that will be loaded for you.\n",
      "\n",
      "Additionally, the following imports are already provided for you and can be used in your code:\n",
      "```\n",
      "import scanpy as sc\n",
      "import seaborn as sns\n",
      "import matplotlib.pyplot as plt\n",
      "import os\n",
      "import numpy as np\n",
      "import pandas as pd\n",
      "from scipy.spatial.distance import cdist\n",
      "```\n",
      "\n",
      "**Usage**\n",
      "BRAD Line:\n",
      "```\n",
      "subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus['output-directory'], <output file>, <input file>, \"<scanpy commands>\"], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "**Examples**\n",
      "Use the below examples to help generate your code.\n",
      "\n",
      "*Example 1*\n",
      "User Prompt: Run scanpy preprocessing and UMAP visualization on XXX.h5ad and save the UMAP plot\n",
      "Response Code:\n",
      "```\n",
      "response = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus['output-directory'], \"XXX-modified.h5ad\", \"<path/to/data>/XXX.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save='umap.png')\"], capture_output=True, text=True)\n",
      "```\n",
      "Explination: the adata object will be loaded in memory already. The command \"sc.pp.neighbors(adata)\" will preprocess the data, then the command \"sc.tl.umap(adata)\" will perform UMAP, and finally the command \"sc.pl.umap(adata, 'umap.png')\" will save the UMAP to a well named file.\n",
      "\n",
      "*Example 2*\n",
      "User Prompt: Perform PCA and clustering on the dataset YYY.h5ad and save the PCA plot\n",
      "Response Code:\n",
      "```\n",
      "response = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus['output-directory'], \"YYY-modified.h5ad\", \"<path/to/data>/YYY.h5ad\", \"sc.pp.pca(adata); sc.tl.leiden(adata); sc.pl.pca(adata, save='pca.png')\"], capture_output=True, text=True)\n",
      "```\n",
      "Explination: the adata object will be loaded in memory already. The command \"sc.pp.pca(adata)\" will preprocess the data, then the command \"sc.tl.leiden(adata)\" will perform the leiden algorithm, and finally the command \"sc.pl.pca(adata, save='pca.png')\" will save the PCA to a well named file.\n",
      "\n",
      "*Example 3*\n",
      "User Prompt: compute the distance between all cells that are not source or target to all source and target cells.\n",
      "Reponse Code:\n",
      "```\n",
      "adata_X = adata.X\n",
      "mask = ~adata.obs['recipe'].isin(['source', 'target'])\n",
      "recipe_embeddings = adata[mask]\n",
      "source_embeddings = adata[adata.obs['recipe'] == 'source']\n",
      "target_embeddings = adata[adata.obs['recipe'] == 'target']\n",
      "recipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X)\n",
      "recipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X)\n",
      "adata_dist = sc.AnnData(X=np.concatenate((recipe_distances_source, recipe_distances_target), axis=1), var=pd.concat((source_embeddings.obs, target_embeddings.obs)), obs=recipe_embeddings.obs)\n",
      "adata_dist.write(<file name>.h5ad)\n",
      "```\n",
      "\n",
      "**OUTPUT FILE NAME INSTRUCTIONS**\n",
      "1. Output path should be chatstatus['output-directory']\n",
      "2. Output file name should be `<descriptive name>.h5ad`\n",
      "\"\"\"\n",
      "\n",
      "\n",
      "**CALL PYTHON SCRIPTS FROM PYTHON**:\n",
      "Use the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\n",
      "\n",
      "To call a Python script `example_script.py` which has no arguments:\n",
      "\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory']], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with one argument:\n",
      "\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory'], 'arg1'], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with two arguments:\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory'], 'arg1', 'arg2'], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "Note that chatstatus['output-directory'] is ALWAYS passed as the first argument.\n",
      "\n",
      "The following files were previously created by BRAD and could be used as input to a function if necessary:\n",
      "log.json\n",
      "S1-HSC-reprogrammed-DE-umap.h5ad\n",
      "S1-umapumap.png\n",
      "S2-HSC-reprogrammed-DE-distances.h5ad\n",
      "\n",
      "Query:2 (CODE)**:\n",
      "Prompt: Compute the pairwise Euclidean distances between each 'target' embedding and each recipe embedding, using the extracted embeddings from the .h5ad file: `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` Plot histograms and or other visualizations of the distances.\n",
      "\n",
      "**PYTHON SCRIPT**\n",
      "Please use the correct script name and path. You must run this python script:\n",
      "/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\n",
      "\n",
      "**INSTRUCTIONS**\n",
      "1. Given the user query and the documentation, identify each of the arguments found in the user's query that should be passed to the Python script.\n",
      "2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\n",
      "3. The last line of your response should say \"Execute: <Python code to execute>\"\n",
      "4. Do not include any extra words or characters. Format the response/output as:\n",
      "    Arguments: \n",
      "    Python Code Explanation: <2 sentences maximum>\n",
      "    Execute: <your code here>\n",
      "\n",
      "**IMPORTANT**\n",
      "The code to execute from your response must be formatted as:\n",
      "    Execute: subprocess.call([sys.executable, '<path to python script>', '<argument 1>', '<argument 2>', ..., '<argument n>'], capture_output=True, text=True))\n",
      "This output should be exactly one line and no longer. Stop the response after this line.\n",
      "\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 01:08:45,575 - INFO - HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2024-08-03 01:08:45,576 - INFO - LLM OUTPUT PARSER\n",
      "2024-08-03 01:08:45,577 - INFO - Arguments:\n",
      "1. output directory: chatstatus['output-directory']\n",
      "2. output file: \"HSC-reprogrammed-DE-distances.h5ad\"\n",
      "3. input file: \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\"\n",
      "4. scanpy commands: \"mask = ~adata.obs['recipe'].isin(['source', 'target']); recipe_embeddings = adata[mask]; target_embeddings = adata[adata.obs['recipe'] == 'target']; recipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X); plt.hist(recipe_distances_target.flatten()); plt.savefig('distances_histogram.png')\"\n",
      "\n",
      "Python Code Explanation: This code computes the pairwise Euclidean distances between 'target' and recipe embeddings, plots a histogram of these distances, and saves the plot.\n",
      "\n",
      "Execute: subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-DE-distances.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"mask = ~adata.obs['recipe'].isin(['source', 'target']); recipe_embeddings = adata[mask]; target_embeddings = adata[adata.obs['recipe'] == 'target']; recipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X); plt.hist(recipe_distances_target.flatten()); plt.savefig('distances_histogram.png')\"], capture_output=True, text=True)\n",
      "2024-08-03 01:08:45,577 - INFO - ['Arguments:\\n1. output directory: chatstatus[\\'output-directory\\']\\n2. output file: \"HSC-reprogrammed-DE-distances.h5ad\"\\n3. input file: \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\"\\n4. scanpy commands: \"mask = ~adata.obs[\\'recipe\\'].isin([\\'source\\', \\'target\\']); recipe_embeddings = adata[mask]; target_embeddings = adata[adata.obs[\\'recipe\\'] == \\'target\\']; recipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X); plt.hist(recipe_distances_target.flatten()); plt.savefig(\\'distances_histogram.png\\')\"\\n\\nPython Code Explanation: This code computes the pairwise Euclidean distances between \\'target\\' and recipe embeddings, plots a histogram of these distances, and saves the plot.\\n\\n', ' subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"HSC-reprogrammed-DE-distances.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"mask = ~adata.obs[\\'recipe\\'].isin([\\'source\\', \\'target\\']); recipe_embeddings = adata[mask]; target_embeddings = adata[adata.obs[\\'recipe\\'] == \\'target\\']; recipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X); plt.hist(recipe_distances_target.flatten()); plt.savefig(\\'distances_histogram.png\\')\"], capture_output=True, text=True)']\n",
      "2024-08-03 01:08:45,577 - INFO - Stripped funciton call\n",
      "2024-08-03 01:08:45,577 - INFO - subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-DE-distances.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"mask = ~adata.obs['recipe'].isin(['source', 'target']); recipe_embeddings = adata[mask]; target_embeddings = adata[adata.obs['recipe'] == 'target']; recipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X); plt.hist(recipe_distances_target.flatten()); plt.savefig('distances_histogram.png')\"], capture_output=True, text=True)\n",
      "2024-08-03 01:08:45,578 - INFO - Cleaned up function call:\n",
      "\n",
      "2024-08-03 01:08:45,578 - INFO - subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-DE-distances.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"mask = ~adata.obs['recipe'].isin(['source', 'target']); recipe_embeddings = adata[mask]; target_embeddings = adata[adata.obs['recipe'] == 'target']; recipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X); plt.hist(recipe_distances_target.flatten()); plt.savefig('distances_histogram.png')\"], capture_output=True, text=True)\n",
      "2024-08-03 01:08:45,578 - INFO - \n",
      "\n",
      "2024-08-03 01:08:45,579 - INFO - EVAL\n",
      "2024-08-03 01:08:45,579 - INFO - subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-DE-distances.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"mask = ~adata.obs['recipe'].isin(['source', 'target']); recipe_embeddings = adata[mask]; target_embeddings = adata[adata.obs['recipe'] == 'target']; recipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X); plt.hist(recipe_distances_target.flatten()); plt.savefig('distances_histogram.png')\"], capture_output=True, text=True)\n",
      "2024-08-03 01:08:45,579 - INFO - END EVAL CHECK\n",
      "2024-08-03 01:08:45,580 - INFO - Finalized PYTHON Call:\n",
      "2024-08-03 01:08:45,580 - INFO -    \n",
      "2024-08-03 01:08:45,580 - INFO - response = subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-DE-distances.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"mask = ~adata.obs['recipe'].isin(['source', 'target']); recipe_embeddings = adata[mask]; target_embeddings = adata[adata.obs['recipe'] == 'target']; recipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X); plt.hist(recipe_distances_target.flatten()); plt.savefig('distances_histogram.png')\"], capture_output=True, text=True)\n",
      "2024-08-03 01:08:45,580 - INFO -    \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 01:08:50,445 - INFO - Debug: PYTHON code executed successfully.\n",
      "2024-08-03 01:08:50,446 - INFO - Code Execution Output:\n",
      "2024-08-03 01:08:50,446 - INFO - CompletedProcess(args=['/sw/pkgs/arc/python3.11-anaconda/2024.02-1/bin/python', '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py', '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-07-41', 'HSC-reprogrammed-DE-distances.h5ad', '/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad', \"mask = ~adata.obs['recipe'].isin(['source', 'target']); recipe_embeddings = adata[mask]; target_embeddings = adata[adata.obs['recipe'] == 'target']; recipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X); plt.hist(recipe_distances_target.flatten()); plt.savefig('distances_histogram.png')\"], returncode=0, stdout=\"adata loaded from /nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\\nadata.shape=(6375, 512)\\nadata.obs.columns=Index(['recipe', 'filename', 'REL', 'STAT5A', 'FOS', 'ETV6', 'source', 'MEF2C',\\n       'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B', 'target'],\\n      dtype='object')\\nadata.obs.head()=   recipe                                           filename  ...  GFI1B  target\\n0  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n1  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n2  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n3  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n4  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n\\n[5 rows x 14 columns]\\nadata.var.head()=Empty DataFrame\\nColumns: []\\nIndex: [0, 1, 2, 3, 4]\\n****************************\\n      EXECUTE COMMANDS      \\n****************************\\nmask = ~adata.obs['recipe'].isin(['source', 'target'])\\nrecipe_embeddings = adata[mask]\\ntarget_embeddings = adata[adata.obs['recipe'] == 'target']\\nrecipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X)\\nplt.hist(recipe_distances_target.flatten())\\nplt.savefig('distances_histogram.png')\\n\", stderr='/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\\n  utils.warn_names_duplicates(\"obs\")\\n')\n",
      "2024-08-03 01:08:50,446 - INFO - Debug: PYTHON code output saved to output.\n",
      "2024-08-03 01:08:50,446 - INFO - \n",
      "\n",
      "\n",
      "route\n",
      "\n",
      "\n",
      "\n",
      "2024-08-03 01:08:50,447 - INFO - CODE\n",
      "2024-08-03 01:08:50,448 - INFO - Finding Stage Number of Pipeline\n",
      "2024-08-03 01:08:50,448 - INFO - {0: {'order': 0, 'module': 'PLANNER', 'prompt': None, 'description': 'This step designed the plan. It is placed in the queue because we needed a place holder for 0 indexed lists.', 'output': []}, 1: {'order': 1, 'module': 'CODE', 'prompt': \"/force CODE  1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings for 'source', 'target', and each recipe from `adata.X`. **Generate a UMAP plot to visualize the embeddings.**\\n\\n**.\", 'description': \" 1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings for 'source', 'target', and each recipe from `adata.X`. **Generate a UMAP plot to visualize the embeddings.**\\n\\n**\", 'output': ['/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-07-41/S1-HSC-reprogrammed-DE-umap.h5ad', '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-07-41/S1-umapumap.png']}, 2: {'order': 2, 'module': 'CODE', 'prompt': \"/force CODE  2 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'source' embedding and each recipe embedding, using the extracted embeddings from the .h5ad file: `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` Plot histograms and or other visualizations of the distances.\", 'description': \" 2 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'source' embedding and each recipe embedding, using the extracted embeddings from \", 'output': ['/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-07-41/S2-HSC-reprogrammed-DE-distances.h5ad']}, 3: {'order': 4, 'module': 'CODE', 'prompt': \"/force CODE  2 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'target' embedding and each recipe embedding, using the extracted embeddings from the .h5ad file: `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` Plot histograms and or other visualizations of the distances.\", 'description': \" 3 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'target' embedding and each recipe embedding, using the extracted embeddings from \", 'output': []}, 4: {'order': 6, 'module': 'CODE', 'prompt': '/force CODE  4 (CODE)**:\\nPrompt: Store the computed distances in a matrix format and create a new ann data object, setting `.var` to be the `.obs` of the source and target cells used in the distance calculations and `.obs` to be the `.obs` of the reprogrammed cells used in the distance calculations. Explicitly save the new ann data object. **Generate a summary plot to visualize the new ann data object.**\\n\\nThe revised plan removes ', 'description': ' 4 (CODE)**:\\nPrompt: Store the computed distances in a matrix format and create a new ann data object, setting `.var` to be the `.obs` of the source and target cells used in the distance calculations and `.obs` to be the `.obs` of the reprogrammed cells used in the distance calculations. Explicitly save the new ann data object. **Generate a summary plot to visualize the new ann data object.**\\n\\nThe revised plan removes ', 'output': []}}\n",
      "2024-08-03 01:08:50,449 - INFO - IP=3\n",
      "2024-08-03 01:08:50,449 - INFO - type(IP)=<class 'int'>\n",
      "2024-08-03 01:08:50,451 - INFO - \n",
      "\n",
      "PLANNER:\n",
      "\n",
      "{'module': 'CODE', 'steps': [{'llm': \"client=<openai.resources.chat.completions.Completions object at 0x147f010bd150> async_client=<openai.resources.chat.completions.AsyncCompletions object at 0x147f010bf650> model_name='gpt-4o' temperature=0.0 openai_api_key=SecretStr('**********') openai_proxy=''\", 'memory': 'None', 'prompt': 'input_variables=[\\'user_query\\'] template=\"You must select which code to run to help a user.\\\\n\\\\n**Available Scripts**\\\\nScript Name: scanpy_brad, \\\\t Description: This script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\\\\nScript Name: models_brad, \\\\t Description: This script executes a series of custom commands to build a model for predicting the contribution to reprogramming obtained by the inclusion or exclusion of individual TFs using an AnnData object or single cell datasets loaded from a .h5ad file.\\\\n\\\\n\\\\n**User Query**\\\\n{user_query}\\\\n\\\\n**Task**\\\\nBased on the user\\'s query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\\\\n\\\\n**Response Template**\\\\nSCRIPT: <selected script>\\\\nREASON: <reasoning why the script fits the user prompt>\\\\n\"', 'input': \"2 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'target' embedding and each recipe embedding, using the extracted embeddings from the .h5ad file: `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` Plot histograms and or other visualizations of the distances.\", 'output': 'content=\"SCRIPT: scanpy_brad\\\\nREASON: The user\\'s query involves computing pairwise Euclidean distances between embeddings and creating visualizations such as histograms. The `scanpy_brad` script is the best match because it allows for manipulation, visualization, and exploration of single-cell datasets using libraries like scanpy, seaborn, and matplotlib.pyplot. These libraries are well-suited for computing distances and generating the required visualizations. The `models_brad` script, on the other hand, is focused on building predictive models and does not align with the user\\'s request for distance computation and visualization.\" response_metadata={\\'token_usage\\': {\\'completion_tokens\\': 119, \\'prompt_tokens\\': 330, \\'total_tokens\\': 449}, \\'model_name\\': \\'gpt-4o-2024-05-13\\', \\'system_fingerprint\\': \\'fp_bc2a86f5f5\\', \\'finish_reason\\': \\'stop\\', \\'logprobs\\': None} id=\\'run-390415c6-da44-4acd-a6e6-4f34b13c385e-0\\' usage_metadata={\\'input_tokens\\': 330, \\'output_tokens\\': 119, \\'total_tokens\\': 449}', 'parsedOutput': {'scriptName': '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py', 'scriptType': 'python', 'scriptPath': '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts'}, 'purpose': 'Select which code to run'}, {'llm': \"client=<openai.resources.chat.completions.Completions object at 0x147f010bd150> async_client=<openai.resources.chat.completions.AsyncCompletions object at 0x147f010bf650> model_name='gpt-4o' temperature=0.0 openai_api_key=SecretStr('**********') openai_proxy=''\", 'memory': 'None', 'prompt': 'input_variables=[\\'history\\', \\'input\\'] template=\\'Current conversation:\\\\n{history}\\\\n\\\\n**PYTHON SCRIPT**\\\\nYou must run this python script:\\\\n/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\\\n\\\\n**PYTHON SCRIPT DOCUMENTATION**:\\\\nThis is the doc string of this python script:\\\\n\"\"\"\\\\nThis script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\\\\n\\\\nArguments (four arguments):\\\\n    1. output directory: chatstatus[\\\\\\'output-directory\\\\\\']\\\\n    2. output file: <name of output file>\\\\n    3. input file: <file created in previous step>\\\\n    4. scanpy commands: a list of scanpy commands to be executed on the AnnData object (provided as a single string with commands separated by a delimiter, e.g., \\\\\\';\\\\\\')\\\\n\\\\nBased on the arguments, the input file will be loaded, then your commands will be executed, and finally, the output or resulting ann data object will be saved to the corrrectou output file and directory. Your code is not responsible for loading the .h5ad object, that will happen automatically, and when loaded, the object will be called adata. Your scanpy commands can operate directly on the adata object that will be loaded for you.\\\\n\\\\nAdditionally, the following imports are already provided for you and can be used in your code:\\\\n```\\\\nimport scanpy as sc\\\\nimport seaborn as sns\\\\nimport matplotlib.pyplot as plt\\\\nimport os\\\\nimport numpy as np\\\\nimport pandas as pd\\\\nfrom scipy.spatial.distance import cdist\\\\n```\\\\n\\\\n**Usage**\\\\nBRAD Line:\\\\n```\\\\nsubprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\\\\\'output-directory\\\\\\'], <output file>, <input file>, \"<scanpy commands>\"], capture_output=True, text=True)\\\\n```\\\\n\\\\n**Examples**\\\\nUse the below examples to help generate your code.\\\\n\\\\n*Example 1*\\\\nUser Prompt: Run scanpy preprocessing and UMAP visualization on XXX.h5ad and save the UMAP plot\\\\nResponse Code:\\\\n```\\\\nresponse = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\\\\\'output-directory\\\\\\'], \"XXX-modified.h5ad\", \"<path/to/data>/XXX.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save=\\\\\\'umap.png\\\\\\')\"], capture_output=True, text=True)\\\\n```\\\\nExplination: the adata object will be loaded in memory already. The command \"sc.pp.neighbors(adata)\" will preprocess the data, then the command \"sc.tl.umap(adata)\" will perform UMAP, and finally the command \"sc.pl.umap(adata, \\\\\\'umap.png\\\\\\')\" will save the UMAP to a well named file.\\\\n\\\\n*Example 2*\\\\nUser Prompt: Perform PCA and clustering on the dataset YYY.h5ad and save the PCA plot\\\\nResponse Code:\\\\n```\\\\nresponse = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\\\\\'output-directory\\\\\\'], \"YYY-modified.h5ad\", \"<path/to/data>/YYY.h5ad\", \"sc.pp.pca(adata); sc.tl.leiden(adata); sc.pl.pca(adata, save=\\\\\\'pca.png\\\\\\')\"], capture_output=True, text=True)\\\\n```\\\\nExplination: the adata object will be loaded in memory already. The command \"sc.pp.pca(adata)\" will preprocess the data, then the command \"sc.tl.leiden(adata)\" will perform the leiden algorithm, and finally the command \"sc.pl.pca(adata, save=\\\\\\'pca.png\\\\\\')\" will save the PCA to a well named file.\\\\n\\\\n*Example 3*\\\\nUser Prompt: compute the distance between all cells that are not source or target to all source and target cells.\\\\nReponse Code:\\\\n```\\\\nadata_X = adata.X\\\\nmask = ~adata.obs[\\\\\\'recipe\\\\\\'].isin([\\\\\\'source\\\\\\', \\\\\\'target\\\\\\'])\\\\nrecipe_embeddings = adata[mask]\\\\nsource_embeddings = adata[adata.obs[\\\\\\'recipe\\\\\\'] == \\\\\\'source\\\\\\']\\\\ntarget_embeddings = adata[adata.obs[\\\\\\'recipe\\\\\\'] == \\\\\\'target\\\\\\']\\\\nrecipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X)\\\\nrecipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X)\\\\nadata_dist = sc.AnnData(X=np.concatenate((recipe_distances_source, recipe_distances_target), axis=1), var=pd.concat((source_embeddings.obs, target_embeddings.obs)), obs=recipe_embeddings.obs)\\\\nadata_dist.write(<file name>.h5ad)\\\\n```\\\\n\\\\n**OUTPUT FILE NAME INSTRUCTIONS**\\\\n1. Output path should be chatstatus[\\\\\\'output-directory\\\\\\']\\\\n2. Output file name should be `<descriptive name>.h5ad`\\\\n\"\"\"\\\\n\\\\n\\\\n**CALL PYTHON SCRIPTS FROM PYTHON**:\\\\nUse the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\\\\n\\\\nTo call a Python script `example_script.py` which has no arguments:\\\\n\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\']], capture_output=True, text=True)\\\\n```\\\\n\\\\nTo call a Python script `example_script.py` with one argument:\\\\n\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\'], \\\\\\'arg1\\\\\\'], capture_output=True, text=True)\\\\n```\\\\n\\\\nTo call a Python script `example_script.py` with two arguments:\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\'], \\\\\\'arg1\\\\\\', \\\\\\'arg2\\\\\\'], capture_output=True, text=True)\\\\n```\\\\n\\\\nNote that chatstatus[\\\\\\'output-directory\\\\\\'] is ALWAYS passed as the first argument.\\\\n\\\\nThe following files were previously created by BRAD and could be used as input to a function if necessary:\\\\nlog.json\\\\nS1-HSC-reprogrammed-DE-umap.h5ad\\\\nS1-umapumap.png\\\\nS2-HSC-reprogrammed-DE-distances.h5ad\\\\n\\\\nQuery:{input}\\\\n\\\\n**PYTHON SCRIPT**\\\\nPlease use the correct script name and path. You must run this python script:\\\\n/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\\\n\\\\n**INSTRUCTIONS**\\\\n1. Given the user query and the documentation, identify each of the arguments found in the user\\\\\\'s query that should be passed to the Python script.\\\\n2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\\\\n3. The last line of your response should say \"Execute: <Python code to execute>\"\\\\n4. Do not include any extra words or characters. Format the response/output as:\\\\n    Arguments: \\\\n    Python Code Explanation: <2 sentences maximum>\\\\n    Execute: <your code here>\\\\n\\\\n**IMPORTANT**\\\\nThe code to execute from your response must be formatted as:\\\\n    Execute: subprocess.call([sys.executable, \\\\\\'<path to python script>\\\\\\', \\\\\\'<argument 1>\\\\\\', \\\\\\'<argument 2>\\\\\\', ..., \\\\\\'<argument n>\\\\\\'], capture_output=True, text=True))\\\\nThis output should be exactly one line and no longer. Stop the response after this line.\\\\n\\'', 'input': \"2 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'target' embedding and each recipe embedding, using the extracted embeddings from the .h5ad file: `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` Plot histograms and or other visualizations of the distances.\", 'output': 'Arguments:\\n1. output directory: chatstatus[\\'output-directory\\']\\n2. output file: \"HSC-reprogrammed-DE-distances.h5ad\"\\n3. input file: \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\"\\n4. scanpy commands: \"mask = ~adata.obs[\\'recipe\\'].isin([\\'source\\', \\'target\\']); recipe_embeddings = adata[mask]; target_embeddings = adata[adata.obs[\\'recipe\\'] == \\'target\\']; recipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X); plt.hist(recipe_distances_target.flatten()); plt.savefig(\\'distances_histogram.png\\')\"\\n\\nPython Code Explanation: This code computes the pairwise Euclidean distances between \\'target\\' and recipe embeddings, plots a histogram of these distances, and saves the plot.\\n\\nExecute: subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"HSC-reprogrammed-DE-distances.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"mask = ~adata.obs[\\'recipe\\'].isin([\\'source\\', \\'target\\']); recipe_embeddings = adata[mask]; target_embeddings = adata[adata.obs[\\'recipe\\'] == \\'target\\']; recipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X); plt.hist(recipe_distances_target.flatten()); plt.savefig(\\'distances_histogram.png\\')\"], capture_output=True, text=True)', 'parsedOutput': {'code': 'subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"HSC-reprogrammed-DE-distances.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"mask = ~adata.obs[\\'recipe\\'].isin([\\'source\\', \\'target\\']); recipe_embeddings = adata[mask]; target_embeddings = adata[adata.obs[\\'recipe\\'] == \\'target\\']; recipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X); plt.hist(recipe_distances_target.flatten()); plt.savefig(\\'distances_histogram.png\\')\"], capture_output=True, text=True)'}, 'purpose': 'Format function call'}, {'func': 'pythonCaller.execute_python_code', 'code': 'response = subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"HSC-reprogrammed-DE-distances.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"mask = ~adata.obs[\\'recipe\\'].isin([\\'source\\', \\'target\\']); recipe_embeddings = adata[mask]; target_embeddings = adata[adata.obs[\\'recipe\\'] == \\'target\\']; recipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X); plt.hist(recipe_distances_target.flatten()); plt.savefig(\\'distances_histogram.png\\')\"], capture_output=True, text=True)', 'purpose': 'execute python code'}, {'func': 'utils.makeNamesConsistent', 'files': [{'old-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-07-41/HSC-reprogrammed-DE-distances.h5ad', 'new-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-07-41/S3-HSC-reprogrammed-DE-distances.h5ad'}]}]}\n",
      "2024-08-03 01:08:50,458 - INFO - current llm calls=7\n",
      "2024-08-03 01:08:50,460 - INFO - CODE\n",
      "2024-08-03 01:08:50,461 - INFO - CODER\n",
      "2024-08-03 01:08:50,462 - INFO - input_variables=['user_query'] template=\"You must select which code to run to help a user.\\n\\n**Available Scripts**\\nScript Name: scanpy_brad, \\t Description: This script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\\nScript Name: models_brad, \\t Description: This script executes a series of custom commands to build a model for predicting the contribution to reprogramming obtained by the inclusion or exclusion of individual TFs using an AnnData object or single cell datasets loaded from a .h5ad file.\\n\\n\\n**User Query**\\n{user_query}\\n\\n**Task**\\nBased on the user's query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\\n\\n**Response Template**\\nSCRIPT: <selected script>\\nREASON: <reasoning why the script fits the user prompt>\\n\"\n",
      "2024-08-03 01:08:50,462 - INFO - FIRST LLM CALL\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "==================================================\n",
      "BRAD >> 4: \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 01:08:52,290 - INFO - HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2024-08-03 01:08:52,291 - INFO - SCRIPT: scanpy_brad\n",
      "REASON: The user's query involves manipulating an AnnData object, computing distances, creating a new AnnData object, and generating a summary plot for visualization. The `scanpy_brad` script is the best match because it allows for the manipulation, visualization, and exploration of single-cell datasets using libraries like scanpy, seaborn, and matplotlib.pyplot. These capabilities are essential for performing the requested tasks and saving the results back to disk. The `models_brad` script, on the other hand, is focused on building predictive models and does not align with the user's requirements for data manipulation and visualization.\n",
      "2024-08-03 01:08:52,291 - INFO - scriptPath=/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts\n",
      "2024-08-03 01:08:52,291 - INFO - ALL SCRIPTS FOUND. BUILDING TEMPLATE\n",
      "2024-08-03 01:08:52,292 - INFO - input_variables=['history', 'input'] template='Current conversation:\\n{history}\\n\\n**PYTHON SCRIPT**\\nYou must run this python script:\\n/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\n\\n**PYTHON SCRIPT DOCUMENTATION**:\\nThis is the doc string of this python script:\\n\"\"\"\\nThis script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\\n\\nArguments (four arguments):\\n    1. output directory: chatstatus[\\'output-directory\\']\\n    2. output file: <name of output file>\\n    3. input file: <file created in previous step>\\n    4. scanpy commands: a list of scanpy commands to be executed on the AnnData object (provided as a single string with commands separated by a delimiter, e.g., \\';\\')\\n\\nBased on the arguments, the input file will be loaded, then your commands will be executed, and finally, the output or resulting ann data object will be saved to the corrrectou output file and directory. Your code is not responsible for loading the .h5ad object, that will happen automatically, and when loaded, the object will be called adata. Your scanpy commands can operate directly on the adata object that will be loaded for you.\\n\\nAdditionally, the following imports are already provided for you and can be used in your code:\\n```\\nimport scanpy as sc\\nimport seaborn as sns\\nimport matplotlib.pyplot as plt\\nimport os\\nimport numpy as np\\nimport pandas as pd\\nfrom scipy.spatial.distance import cdist\\n```\\n\\n**Usage**\\nBRAD Line:\\n```\\nsubprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\'output-directory\\'], <output file>, <input file>, \"<scanpy commands>\"], capture_output=True, text=True)\\n```\\n\\n**Examples**\\nUse the below examples to help generate your code.\\n\\n*Example 1*\\nUser Prompt: Run scanpy preprocessing and UMAP visualization on XXX.h5ad and save the UMAP plot\\nResponse Code:\\n```\\nresponse = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"XXX-modified.h5ad\", \"<path/to/data>/XXX.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save=\\'umap.png\\')\"], capture_output=True, text=True)\\n```\\nExplination: the adata object will be loaded in memory already. The command \"sc.pp.neighbors(adata)\" will preprocess the data, then the command \"sc.tl.umap(adata)\" will perform UMAP, and finally the command \"sc.pl.umap(adata, \\'umap.png\\')\" will save the UMAP to a well named file.\\n\\n*Example 2*\\nUser Prompt: Perform PCA and clustering on the dataset YYY.h5ad and save the PCA plot\\nResponse Code:\\n```\\nresponse = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"YYY-modified.h5ad\", \"<path/to/data>/YYY.h5ad\", \"sc.pp.pca(adata); sc.tl.leiden(adata); sc.pl.pca(adata, save=\\'pca.png\\')\"], capture_output=True, text=True)\\n```\\nExplination: the adata object will be loaded in memory already. The command \"sc.pp.pca(adata)\" will preprocess the data, then the command \"sc.tl.leiden(adata)\" will perform the leiden algorithm, and finally the command \"sc.pl.pca(adata, save=\\'pca.png\\')\" will save the PCA to a well named file.\\n\\n*Example 3*\\nUser Prompt: compute the distance between all cells that are not source or target to all source and target cells.\\nReponse Code:\\n```\\nadata_X = adata.X\\nmask = ~adata.obs[\\'recipe\\'].isin([\\'source\\', \\'target\\'])\\nrecipe_embeddings = adata[mask]\\nsource_embeddings = adata[adata.obs[\\'recipe\\'] == \\'source\\']\\ntarget_embeddings = adata[adata.obs[\\'recipe\\'] == \\'target\\']\\nrecipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X)\\nrecipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X)\\nadata_dist = sc.AnnData(X=np.concatenate((recipe_distances_source, recipe_distances_target), axis=1), var=pd.concat((source_embeddings.obs, target_embeddings.obs)), obs=recipe_embeddings.obs)\\nadata_dist.write(<file name>.h5ad)\\n```\\n\\n**OUTPUT FILE NAME INSTRUCTIONS**\\n1. Output path should be chatstatus[\\'output-directory\\']\\n2. Output file name should be `<descriptive name>.h5ad`\\n\"\"\"\\n\\n\\n**CALL PYTHON SCRIPTS FROM PYTHON**:\\nUse the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\\n\\nTo call a Python script `example_script.py` which has no arguments:\\n\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\']], capture_output=True, text=True)\\n```\\n\\nTo call a Python script `example_script.py` with one argument:\\n\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\'], \\'arg1\\'], capture_output=True, text=True)\\n```\\n\\nTo call a Python script `example_script.py` with two arguments:\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\'], \\'arg1\\', \\'arg2\\'], capture_output=True, text=True)\\n```\\n\\nNote that chatstatus[\\'output-directory\\'] is ALWAYS passed as the first argument.\\n\\nThe following files were previously created by BRAD and could be used as input to a function if necessary:\\nlog.json\\nS1-HSC-reprogrammed-DE-umap.h5ad\\nS1-umapumap.png\\nS3-HSC-reprogrammed-DE-distances.h5ad\\nS2-HSC-reprogrammed-DE-distances.h5ad\\n\\nQuery:{input}\\n\\n**PYTHON SCRIPT**\\nPlease use the correct script name and path. You must run this python script:\\n/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\n\\n**INSTRUCTIONS**\\n1. Given the user query and the documentation, identify each of the arguments found in the user\\'s query that should be passed to the Python script.\\n2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\\n3. The last line of your response should say \"Execute: <Python code to execute>\"\\n4. Do not include any extra words or characters. Format the response/output as:\\n    Arguments: \\n    Python Code Explanation: <2 sentences maximum>\\n    Execute: <your code here>\\n\\n**IMPORTANT**\\nThe code to execute from your response must be formatted as:\\n    Execute: subprocess.call([sys.executable, \\'<path to python script>\\', \\'<argument 1>\\', \\'<argument 2>\\', ..., \\'<argument n>\\'], capture_output=True, text=True))\\nThis output should be exactly one line and no longer. Stop the response after this line.\\n'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation:\n",
      "\n",
      "\n",
      "**PYTHON SCRIPT**\n",
      "You must run this python script:\n",
      "/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\n",
      "\n",
      "**PYTHON SCRIPT DOCUMENTATION**:\n",
      "This is the doc string of this python script:\n",
      "\"\"\"\n",
      "This script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\n",
      "\n",
      "Arguments (four arguments):\n",
      "    1. output directory: chatstatus['output-directory']\n",
      "    2. output file: <name of output file>\n",
      "    3. input file: <file created in previous step>\n",
      "    4. scanpy commands: a list of scanpy commands to be executed on the AnnData object (provided as a single string with commands separated by a delimiter, e.g., ';')\n",
      "\n",
      "Based on the arguments, the input file will be loaded, then your commands will be executed, and finally, the output or resulting ann data object will be saved to the corrrectou output file and directory. Your code is not responsible for loading the .h5ad object, that will happen automatically, and when loaded, the object will be called adata. Your scanpy commands can operate directly on the adata object that will be loaded for you.\n",
      "\n",
      "Additionally, the following imports are already provided for you and can be used in your code:\n",
      "```\n",
      "import scanpy as sc\n",
      "import seaborn as sns\n",
      "import matplotlib.pyplot as plt\n",
      "import os\n",
      "import numpy as np\n",
      "import pandas as pd\n",
      "from scipy.spatial.distance import cdist\n",
      "```\n",
      "\n",
      "**Usage**\n",
      "BRAD Line:\n",
      "```\n",
      "subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus['output-directory'], <output file>, <input file>, \"<scanpy commands>\"], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "**Examples**\n",
      "Use the below examples to help generate your code.\n",
      "\n",
      "*Example 1*\n",
      "User Prompt: Run scanpy preprocessing and UMAP visualization on XXX.h5ad and save the UMAP plot\n",
      "Response Code:\n",
      "```\n",
      "response = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus['output-directory'], \"XXX-modified.h5ad\", \"<path/to/data>/XXX.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save='umap.png')\"], capture_output=True, text=True)\n",
      "```\n",
      "Explination: the adata object will be loaded in memory already. The command \"sc.pp.neighbors(adata)\" will preprocess the data, then the command \"sc.tl.umap(adata)\" will perform UMAP, and finally the command \"sc.pl.umap(adata, 'umap.png')\" will save the UMAP to a well named file.\n",
      "\n",
      "*Example 2*\n",
      "User Prompt: Perform PCA and clustering on the dataset YYY.h5ad and save the PCA plot\n",
      "Response Code:\n",
      "```\n",
      "response = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus['output-directory'], \"YYY-modified.h5ad\", \"<path/to/data>/YYY.h5ad\", \"sc.pp.pca(adata); sc.tl.leiden(adata); sc.pl.pca(adata, save='pca.png')\"], capture_output=True, text=True)\n",
      "```\n",
      "Explination: the adata object will be loaded in memory already. The command \"sc.pp.pca(adata)\" will preprocess the data, then the command \"sc.tl.leiden(adata)\" will perform the leiden algorithm, and finally the command \"sc.pl.pca(adata, save='pca.png')\" will save the PCA to a well named file.\n",
      "\n",
      "*Example 3*\n",
      "User Prompt: compute the distance between all cells that are not source or target to all source and target cells.\n",
      "Reponse Code:\n",
      "```\n",
      "adata_X = adata.X\n",
      "mask = ~adata.obs['recipe'].isin(['source', 'target'])\n",
      "recipe_embeddings = adata[mask]\n",
      "source_embeddings = adata[adata.obs['recipe'] == 'source']\n",
      "target_embeddings = adata[adata.obs['recipe'] == 'target']\n",
      "recipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X)\n",
      "recipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X)\n",
      "adata_dist = sc.AnnData(X=np.concatenate((recipe_distances_source, recipe_distances_target), axis=1), var=pd.concat((source_embeddings.obs, target_embeddings.obs)), obs=recipe_embeddings.obs)\n",
      "adata_dist.write(<file name>.h5ad)\n",
      "```\n",
      "\n",
      "**OUTPUT FILE NAME INSTRUCTIONS**\n",
      "1. Output path should be chatstatus['output-directory']\n",
      "2. Output file name should be `<descriptive name>.h5ad`\n",
      "\"\"\"\n",
      "\n",
      "\n",
      "**CALL PYTHON SCRIPTS FROM PYTHON**:\n",
      "Use the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\n",
      "\n",
      "To call a Python script `example_script.py` which has no arguments:\n",
      "\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory']], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with one argument:\n",
      "\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory'], 'arg1'], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with two arguments:\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory'], 'arg1', 'arg2'], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "Note that chatstatus['output-directory'] is ALWAYS passed as the first argument.\n",
      "\n",
      "The following files were previously created by BRAD and could be used as input to a function if necessary:\n",
      "log.json\n",
      "S1-HSC-reprogrammed-DE-umap.h5ad\n",
      "S1-umapumap.png\n",
      "S3-HSC-reprogrammed-DE-distances.h5ad\n",
      "S2-HSC-reprogrammed-DE-distances.h5ad\n",
      "\n",
      "Query:4 (CODE)**:\n",
      "Prompt: Store the computed distances in a matrix format and create a new ann data object, setting `.var` to be the `.obs` of the source and target cells used in the distance calculations and `.obs` to be the `.obs` of the reprogrammed cells used in the distance calculations. Explicitly save the new ann data object. **Generate a summary plot to visualize the new ann data object.**\n",
      "\n",
      "The revised plan removes\n",
      "\n",
      "**PYTHON SCRIPT**\n",
      "Please use the correct script name and path. You must run this python script:\n",
      "/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\n",
      "\n",
      "**INSTRUCTIONS**\n",
      "1. Given the user query and the documentation, identify each of the arguments found in the user's query that should be passed to the Python script.\n",
      "2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\n",
      "3. The last line of your response should say \"Execute: <Python code to execute>\"\n",
      "4. Do not include any extra words or characters. Format the response/output as:\n",
      "    Arguments: \n",
      "    Python Code Explanation: <2 sentences maximum>\n",
      "    Execute: <your code here>\n",
      "\n",
      "**IMPORTANT**\n",
      "The code to execute from your response must be formatted as:\n",
      "    Execute: subprocess.call([sys.executable, '<path to python script>', '<argument 1>', '<argument 2>', ..., '<argument n>'], capture_output=True, text=True))\n",
      "This output should be exactly one line and no longer. Stop the response after this line.\n",
      "\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 01:08:55,102 - INFO - HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2024-08-03 01:08:55,103 - INFO - LLM OUTPUT PARSER\n",
      "2024-08-03 01:08:55,103 - INFO - Arguments:\n",
      "1. chatstatus['output-directory']\n",
      "2. \"HSC-reprogrammed-DE-distances-summary.h5ad\"\n",
      "3. \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/S3-HSC-reprogrammed-DE-distances.h5ad\"\n",
      "4. \"sc.pl.matrixplot(adata, save='summary_plot.png')\"\n",
      "\n",
      "Python Code Explanation: This code will load the specified .h5ad file, generate a summary plot, and save the new AnnData object and plot to the specified output directory.\n",
      "\n",
      "Execute: subprocess.run([sys.executable, '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py', chatstatus['output-directory'], 'HSC-reprogrammed-DE-distances-summary.h5ad', '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/S3-HSC-reprogrammed-DE-distances.h5ad', 'sc.pl.matrixplot(adata, save=\"summary_plot.png\")'], capture_output=True, text=True)\n",
      "2024-08-03 01:08:55,104 - INFO - ['Arguments:\\n1. chatstatus[\\'output-directory\\']\\n2. \"HSC-reprogrammed-DE-distances-summary.h5ad\"\\n3. \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/S3-HSC-reprogrammed-DE-distances.h5ad\"\\n4. \"sc.pl.matrixplot(adata, save=\\'summary_plot.png\\')\"\\n\\nPython Code Explanation: This code will load the specified .h5ad file, generate a summary plot, and save the new AnnData object and plot to the specified output directory.\\n\\n', ' subprocess.run([sys.executable, \\'/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\', chatstatus[\\'output-directory\\'], \\'HSC-reprogrammed-DE-distances-summary.h5ad\\', \\'/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/S3-HSC-reprogrammed-DE-distances.h5ad\\', \\'sc.pl.matrixplot(adata, save=\"summary_plot.png\")\\'], capture_output=True, text=True)']\n",
      "2024-08-03 01:08:55,104 - INFO - Stripped funciton call\n",
      "2024-08-03 01:08:55,104 - INFO - subprocess.run([sys.executable, '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py', chatstatus['output-directory'], 'HSC-reprogrammed-DE-distances-summary.h5ad', '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/S3-HSC-reprogrammed-DE-distances.h5ad', 'sc.pl.matrixplot(adata, save=\"summary_plot.png\")'], capture_output=True, text=True)\n",
      "2024-08-03 01:08:55,104 - INFO - Cleaned up function call:\n",
      "\n",
      "2024-08-03 01:08:55,105 - INFO - subprocess.run([sys.executable, '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py', chatstatus['output-directory'], 'HSC-reprogrammed-DE-distances-summary.h5ad', '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/S3-HSC-reprogrammed-DE-distances.h5ad', 'sc.pl.matrixplot(adata, save=\"summary_plot.png\")'], capture_output=True, text=True)\n",
      "2024-08-03 01:08:55,105 - INFO - \n",
      "\n",
      "2024-08-03 01:08:55,106 - INFO - Replacing: S3-HSC-reprogrammed-DE-distances.h5ad with /home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-07-41/S3-HSC-reprogrammed-DE-distances.h5ad\n",
      "2024-08-03 01:08:55,106 - INFO - New String: subprocess.run([sys.executable, '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py', chatstatus['output-directory'], 'HSC-reprogrammed-DE-distances-summary.h5ad', '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts//home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-07-41/S3-HSC-reprogrammed-DE-distances.h5ad', 'sc.pl.matrixplot(adata, save=\"summary_plot.png\")'], capture_output=True, text=True)\n",
      "2024-08-03 01:08:55,106 - INFO - EVAL\n",
      "2024-08-03 01:08:55,106 - INFO - subprocess.run([sys.executable, '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py', chatstatus['output-directory'], 'HSC-reprogrammed-DE-distances-summary.h5ad', '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts//home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-07-41/S3-HSC-reprogrammed-DE-distances.h5ad', 'sc.pl.matrixplot(adata, save=\"summary_plot.png\")'], capture_output=True, text=True)\n",
      "2024-08-03 01:08:55,107 - INFO - END EVAL CHECK\n",
      "2024-08-03 01:08:55,107 - INFO - Finalized PYTHON Call:\n",
      "2024-08-03 01:08:55,107 - INFO -    \n",
      "2024-08-03 01:08:55,108 - INFO - response = subprocess.run([sys.executable, '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py', chatstatus['output-directory'], 'HSC-reprogrammed-DE-distances-summary.h5ad', '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts//home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-07-41/S3-HSC-reprogrammed-DE-distances.h5ad', 'sc.pl.matrixplot(adata, save=\"summary_plot.png\")'], capture_output=True, text=True)\n",
      "2024-08-03 01:08:55,108 - INFO -    \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 01:08:59,554 - INFO - Debug: PYTHON code executed successfully.\n",
      "2024-08-03 01:08:59,554 - INFO - Code Execution Output:\n",
      "2024-08-03 01:08:59,555 - INFO - CompletedProcess(args=['/sw/pkgs/arc/python3.11-anaconda/2024.02-1/bin/python', '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py', '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-07-41', 'HSC-reprogrammed-DE-distances-summary.h5ad', '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts//home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-07-41/S3-HSC-reprogrammed-DE-distances.h5ad', 'sc.pl.matrixplot(adata, save=\"summary_plot.png\")'], returncode=1, stdout='', stderr='Traceback (most recent call last):\\n  File \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", line 117, in <module>\\n    main(args.output_directory, args.output_file, args.input_file, args.scanpy_commands)\\n  File \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", line 86, in main\\n    adata = sc.read_h5ad(input_file)\\n            ^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \"/home/jpic/.local/lib/python3.11/site-packages/anndata/_io/h5ad.py\", line 237, in read_h5ad\\n    with h5py.File(filename, \"r\") as f:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \"/sw/pkgs/arc/python3.11-anaconda/2024.02-1/lib/python3.11/site-packages/h5py/_hl/files.py\", line 567, in __init__\\n    fid = make_fid(name, mode, userblock_size, fapl, fcpl, swmr=swmr)\\n          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \"/sw/pkgs/arc/python3.11-anaconda/2024.02-1/lib/python3.11/site-packages/h5py/_hl/files.py\", line 231, in make_fid\\n    fid = h5f.open(name, flags, fapl=fapl)\\n          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \"h5py/_objects.pyx\", line 54, in h5py._objects.with_phil.wrapper\\n  File \"h5py/_objects.pyx\", line 55, in h5py._objects.with_phil.wrapper\\n  File \"h5py/h5f.pyx\", line 106, in h5py.h5f.open\\nFileNotFoundError: [Errno 2] Unable to open file (unable to open file: name = \\'/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts//home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-07-41/S3-HSC-reprogrammed-DE-distances.h5ad\\', errno = 2, error message = \\'No such file or directory\\', flags = 0, o_flags = 0)\\n')\n",
      "2024-08-03 01:08:59,555 - INFO - Debug: PYTHON code output saved to output.\n",
      "2024-08-03 01:08:59,555 - INFO - \n",
      "\n",
      "\n",
      "route\n",
      "\n",
      "\n",
      "\n",
      "2024-08-03 01:08:59,555 - INFO - CODE\n",
      "2024-08-03 01:08:59,556 - INFO - Finding Stage Number of Pipeline\n",
      "2024-08-03 01:08:59,556 - INFO - {0: {'order': 0, 'module': 'PLANNER', 'prompt': None, 'description': 'This step designed the plan. It is placed in the queue because we needed a place holder for 0 indexed lists.', 'output': []}, 1: {'order': 1, 'module': 'CODE', 'prompt': \"/force CODE  1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings for 'source', 'target', and each recipe from `adata.X`. **Generate a UMAP plot to visualize the embeddings.**\\n\\n**.\", 'description': \" 1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings for 'source', 'target', and each recipe from `adata.X`. **Generate a UMAP plot to visualize the embeddings.**\\n\\n**\", 'output': ['/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-07-41/S1-HSC-reprogrammed-DE-umap.h5ad', '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-07-41/S1-umapumap.png']}, 2: {'order': 2, 'module': 'CODE', 'prompt': \"/force CODE  2 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'source' embedding and each recipe embedding, using the extracted embeddings from the .h5ad file: `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` Plot histograms and or other visualizations of the distances.\", 'description': \" 2 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'source' embedding and each recipe embedding, using the extracted embeddings from \", 'output': ['/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-07-41/S2-HSC-reprogrammed-DE-distances.h5ad']}, 3: {'order': 4, 'module': 'CODE', 'prompt': \"/force CODE  2 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'target' embedding and each recipe embedding, using the extracted embeddings from the .h5ad file: `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` Plot histograms and or other visualizations of the distances.\", 'description': \" 3 (CODE)**:\\nPrompt: Compute the pairwise Euclidean distances between each 'target' embedding and each recipe embedding, using the extracted embeddings from \", 'output': ['/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-07-41/S3-HSC-reprogrammed-DE-distances.h5ad']}, 4: {'order': 6, 'module': 'CODE', 'prompt': '/force CODE  4 (CODE)**:\\nPrompt: Store the computed distances in a matrix format and create a new ann data object, setting `.var` to be the `.obs` of the source and target cells used in the distance calculations and `.obs` to be the `.obs` of the reprogrammed cells used in the distance calculations. Explicitly save the new ann data object. **Generate a summary plot to visualize the new ann data object.**\\n\\nThe revised plan removes ', 'description': ' 4 (CODE)**:\\nPrompt: Store the computed distances in a matrix format and create a new ann data object, setting `.var` to be the `.obs` of the source and target cells used in the distance calculations and `.obs` to be the `.obs` of the reprogrammed cells used in the distance calculations. Explicitly save the new ann data object. **Generate a summary plot to visualize the new ann data object.**\\n\\nThe revised plan removes ', 'output': []}}\n",
      "2024-08-03 01:08:59,557 - INFO - IP=4\n",
      "2024-08-03 01:08:59,557 - INFO - type(IP)=<class 'int'>\n",
      "2024-08-03 01:08:59,557 - INFO - \n",
      "\n",
      "PLANNER:\n",
      "\n",
      "{'module': 'CODE', 'steps': [{'llm': \"client=<openai.resources.chat.completions.Completions object at 0x147f010bd150> async_client=<openai.resources.chat.completions.AsyncCompletions object at 0x147f010bf650> model_name='gpt-4o' temperature=0.0 openai_api_key=SecretStr('**********') openai_proxy=''\", 'memory': 'None', 'prompt': 'input_variables=[\\'user_query\\'] template=\"You must select which code to run to help a user.\\\\n\\\\n**Available Scripts**\\\\nScript Name: scanpy_brad, \\\\t Description: This script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\\\\nScript Name: models_brad, \\\\t Description: This script executes a series of custom commands to build a model for predicting the contribution to reprogramming obtained by the inclusion or exclusion of individual TFs using an AnnData object or single cell datasets loaded from a .h5ad file.\\\\n\\\\n\\\\n**User Query**\\\\n{user_query}\\\\n\\\\n**Task**\\\\nBased on the user\\'s query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\\\\n\\\\n**Response Template**\\\\nSCRIPT: <selected script>\\\\nREASON: <reasoning why the script fits the user prompt>\\\\n\"', 'input': '4 (CODE)**:\\nPrompt: Store the computed distances in a matrix format and create a new ann data object, setting `.var` to be the `.obs` of the source and target cells used in the distance calculations and `.obs` to be the `.obs` of the reprogrammed cells used in the distance calculations. Explicitly save the new ann data object. **Generate a summary plot to visualize the new ann data object.**\\n\\nThe revised plan removes', 'output': 'content=\"SCRIPT: scanpy_brad\\\\nREASON: The user\\'s query involves manipulating an AnnData object, computing distances, creating a new AnnData object, and generating a summary plot for visualization. The `scanpy_brad` script is the best match because it allows for the manipulation, visualization, and exploration of single-cell datasets using libraries like scanpy, seaborn, and matplotlib.pyplot. These capabilities are essential for performing the requested tasks and saving the results back to disk. The `models_brad` script, on the other hand, is focused on building predictive models and does not align with the user\\'s requirements for data manipulation and visualization.\" response_metadata={\\'token_usage\\': {\\'completion_tokens\\': 128, \\'prompt_tokens\\': 344, \\'total_tokens\\': 472}, \\'model_name\\': \\'gpt-4o-2024-05-13\\', \\'system_fingerprint\\': \\'fp_3cd8b62c3b\\', \\'finish_reason\\': \\'stop\\', \\'logprobs\\': None} id=\\'run-5ae47e5a-e019-4699-b13b-1703e74d99bb-0\\' usage_metadata={\\'input_tokens\\': 344, \\'output_tokens\\': 128, \\'total_tokens\\': 472}', 'parsedOutput': {'scriptName': '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py', 'scriptType': 'python', 'scriptPath': '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts'}, 'purpose': 'Select which code to run'}, {'llm': \"client=<openai.resources.chat.completions.Completions object at 0x147f010bd150> async_client=<openai.resources.chat.completions.AsyncCompletions object at 0x147f010bf650> model_name='gpt-4o' temperature=0.0 openai_api_key=SecretStr('**********') openai_proxy=''\", 'memory': 'None', 'prompt': 'input_variables=[\\'history\\', \\'input\\'] template=\\'Current conversation:\\\\n{history}\\\\n\\\\n**PYTHON SCRIPT**\\\\nYou must run this python script:\\\\n/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\\\n\\\\n**PYTHON SCRIPT DOCUMENTATION**:\\\\nThis is the doc string of this python script:\\\\n\"\"\"\\\\nThis script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\\\\n\\\\nArguments (four arguments):\\\\n    1. output directory: chatstatus[\\\\\\'output-directory\\\\\\']\\\\n    2. output file: <name of output file>\\\\n    3. input file: <file created in previous step>\\\\n    4. scanpy commands: a list of scanpy commands to be executed on the AnnData object (provided as a single string with commands separated by a delimiter, e.g., \\\\\\';\\\\\\')\\\\n\\\\nBased on the arguments, the input file will be loaded, then your commands will be executed, and finally, the output or resulting ann data object will be saved to the corrrectou output file and directory. Your code is not responsible for loading the .h5ad object, that will happen automatically, and when loaded, the object will be called adata. Your scanpy commands can operate directly on the adata object that will be loaded for you.\\\\n\\\\nAdditionally, the following imports are already provided for you and can be used in your code:\\\\n```\\\\nimport scanpy as sc\\\\nimport seaborn as sns\\\\nimport matplotlib.pyplot as plt\\\\nimport os\\\\nimport numpy as np\\\\nimport pandas as pd\\\\nfrom scipy.spatial.distance import cdist\\\\n```\\\\n\\\\n**Usage**\\\\nBRAD Line:\\\\n```\\\\nsubprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\\\\\'output-directory\\\\\\'], <output file>, <input file>, \"<scanpy commands>\"], capture_output=True, text=True)\\\\n```\\\\n\\\\n**Examples**\\\\nUse the below examples to help generate your code.\\\\n\\\\n*Example 1*\\\\nUser Prompt: Run scanpy preprocessing and UMAP visualization on XXX.h5ad and save the UMAP plot\\\\nResponse Code:\\\\n```\\\\nresponse = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\\\\\'output-directory\\\\\\'], \"XXX-modified.h5ad\", \"<path/to/data>/XXX.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save=\\\\\\'umap.png\\\\\\')\"], capture_output=True, text=True)\\\\n```\\\\nExplination: the adata object will be loaded in memory already. The command \"sc.pp.neighbors(adata)\" will preprocess the data, then the command \"sc.tl.umap(adata)\" will perform UMAP, and finally the command \"sc.pl.umap(adata, \\\\\\'umap.png\\\\\\')\" will save the UMAP to a well named file.\\\\n\\\\n*Example 2*\\\\nUser Prompt: Perform PCA and clustering on the dataset YYY.h5ad and save the PCA plot\\\\nResponse Code:\\\\n```\\\\nresponse = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\\\\\'output-directory\\\\\\'], \"YYY-modified.h5ad\", \"<path/to/data>/YYY.h5ad\", \"sc.pp.pca(adata); sc.tl.leiden(adata); sc.pl.pca(adata, save=\\\\\\'pca.png\\\\\\')\"], capture_output=True, text=True)\\\\n```\\\\nExplination: the adata object will be loaded in memory already. The command \"sc.pp.pca(adata)\" will preprocess the data, then the command \"sc.tl.leiden(adata)\" will perform the leiden algorithm, and finally the command \"sc.pl.pca(adata, save=\\\\\\'pca.png\\\\\\')\" will save the PCA to a well named file.\\\\n\\\\n*Example 3*\\\\nUser Prompt: compute the distance between all cells that are not source or target to all source and target cells.\\\\nReponse Code:\\\\n```\\\\nadata_X = adata.X\\\\nmask = ~adata.obs[\\\\\\'recipe\\\\\\'].isin([\\\\\\'source\\\\\\', \\\\\\'target\\\\\\'])\\\\nrecipe_embeddings = adata[mask]\\\\nsource_embeddings = adata[adata.obs[\\\\\\'recipe\\\\\\'] == \\\\\\'source\\\\\\']\\\\ntarget_embeddings = adata[adata.obs[\\\\\\'recipe\\\\\\'] == \\\\\\'target\\\\\\']\\\\nrecipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X)\\\\nrecipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X)\\\\nadata_dist = sc.AnnData(X=np.concatenate((recipe_distances_source, recipe_distances_target), axis=1), var=pd.concat((source_embeddings.obs, target_embeddings.obs)), obs=recipe_embeddings.obs)\\\\nadata_dist.write(<file name>.h5ad)\\\\n```\\\\n\\\\n**OUTPUT FILE NAME INSTRUCTIONS**\\\\n1. Output path should be chatstatus[\\\\\\'output-directory\\\\\\']\\\\n2. Output file name should be `<descriptive name>.h5ad`\\\\n\"\"\"\\\\n\\\\n\\\\n**CALL PYTHON SCRIPTS FROM PYTHON**:\\\\nUse the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\\\\n\\\\nTo call a Python script `example_script.py` which has no arguments:\\\\n\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\']], capture_output=True, text=True)\\\\n```\\\\n\\\\nTo call a Python script `example_script.py` with one argument:\\\\n\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\'], \\\\\\'arg1\\\\\\'], capture_output=True, text=True)\\\\n```\\\\n\\\\nTo call a Python script `example_script.py` with two arguments:\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\'], \\\\\\'arg1\\\\\\', \\\\\\'arg2\\\\\\'], capture_output=True, text=True)\\\\n```\\\\n\\\\nNote that chatstatus[\\\\\\'output-directory\\\\\\'] is ALWAYS passed as the first argument.\\\\n\\\\nThe following files were previously created by BRAD and could be used as input to a function if necessary:\\\\nlog.json\\\\nS1-HSC-reprogrammed-DE-umap.h5ad\\\\nS1-umapumap.png\\\\nS3-HSC-reprogrammed-DE-distances.h5ad\\\\nS2-HSC-reprogrammed-DE-distances.h5ad\\\\n\\\\nQuery:{input}\\\\n\\\\n**PYTHON SCRIPT**\\\\nPlease use the correct script name and path. You must run this python script:\\\\n/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\\\n\\\\n**INSTRUCTIONS**\\\\n1. Given the user query and the documentation, identify each of the arguments found in the user\\\\\\'s query that should be passed to the Python script.\\\\n2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\\\\n3. The last line of your response should say \"Execute: <Python code to execute>\"\\\\n4. Do not include any extra words or characters. Format the response/output as:\\\\n    Arguments: \\\\n    Python Code Explanation: <2 sentences maximum>\\\\n    Execute: <your code here>\\\\n\\\\n**IMPORTANT**\\\\nThe code to execute from your response must be formatted as:\\\\n    Execute: subprocess.call([sys.executable, \\\\\\'<path to python script>\\\\\\', \\\\\\'<argument 1>\\\\\\', \\\\\\'<argument 2>\\\\\\', ..., \\\\\\'<argument n>\\\\\\'], capture_output=True, text=True))\\\\nThis output should be exactly one line and no longer. Stop the response after this line.\\\\n\\'', 'input': '4 (CODE)**:\\nPrompt: Store the computed distances in a matrix format and create a new ann data object, setting `.var` to be the `.obs` of the source and target cells used in the distance calculations and `.obs` to be the `.obs` of the reprogrammed cells used in the distance calculations. Explicitly save the new ann data object. **Generate a summary plot to visualize the new ann data object.**\\n\\nThe revised plan removes', 'output': 'Arguments:\\n1. chatstatus[\\'output-directory\\']\\n2. \"HSC-reprogrammed-DE-distances-summary.h5ad\"\\n3. \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/S3-HSC-reprogrammed-DE-distances.h5ad\"\\n4. \"sc.pl.matrixplot(adata, save=\\'summary_plot.png\\')\"\\n\\nPython Code Explanation: This code will load the specified .h5ad file, generate a summary plot, and save the new AnnData object and plot to the specified output directory.\\n\\nExecute: subprocess.run([sys.executable, \\'/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\', chatstatus[\\'output-directory\\'], \\'HSC-reprogrammed-DE-distances-summary.h5ad\\', \\'/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/S3-HSC-reprogrammed-DE-distances.h5ad\\', \\'sc.pl.matrixplot(adata, save=\"summary_plot.png\")\\'], capture_output=True, text=True)', 'parsedOutput': {'code': 'subprocess.run([sys.executable, \\'/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\', chatstatus[\\'output-directory\\'], \\'HSC-reprogrammed-DE-distances-summary.h5ad\\', \\'/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/S3-HSC-reprogrammed-DE-distances.h5ad\\', \\'sc.pl.matrixplot(adata, save=\"summary_plot.png\")\\'], capture_output=True, text=True)'}, 'purpose': 'Format function call'}, {'func': 'pythonCaller.execute_python_code', 'code': 'response = subprocess.run([sys.executable, \\'/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\', chatstatus[\\'output-directory\\'], \\'HSC-reprogrammed-DE-distances-summary.h5ad\\', \\'/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts//home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-07-41/S3-HSC-reprogrammed-DE-distances.h5ad\\', \\'sc.pl.matrixplot(adata, save=\"summary_plot.png\")\\'], capture_output=True, text=True)', 'purpose': 'execute python code'}, {'func': 'utils.makeNamesConsistent', 'files': []}]}\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Thanks for chatting today! I hope to talk soon, and don't forget that a record of this conversation is available at: /home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-07-41/log.json\n"
     ]
    }
   ],
   "source": [
    "llm = llms.load_openai(model_name='gpt-4o')\n",
    "# llm = llms.load_nvidia() #model_name='meta/llama-3.1-405b-instruct')\n",
    "BOT2 = brad.chatbot(llm=llm, interactive=False, config='/home/jpic/RAG-DEV/tutorials/geneformer/config-sc.json')\n",
    "BOT2.chat() # invoke('/force PLANNER please load and run the BOT2 pipeline')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ef713a0-037f-45d0-87d8-8e4bdf6e3067",
   "metadata": {},
   "source": [
    "## BOT 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "9ce18d56-a435-4cf7-ab79-ea4655d3318d",
   "metadata": {},
   "outputs": [],
   "source": [
    "modified_instructions = \"\"\"### BOT3: Predictive Modeling\n",
    "**Objective:** Build a model to predict the contribution to reprogramming obtained by the inclusion or exclusion of individual TFs.\n",
    "\n",
    "**Instructions:**\n",
    "1. Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`.\n",
    "2. Prepare the data for modeling:\n",
    "   - Use the embeddings in `adata.X` as features.\n",
    "   - Use the presence/absence of each TF (columns: 'REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B') as additional features.\n",
    "   - Define the target variable based on the distance to 'target' cells (computed by BOT2).\n",
    "3. Split the data into training and testing sets.\n",
    "4. Train a machine learning model (e.g., Random Forest, Gradient Boosting) to predict the reprogramming efficiency.\n",
    "5. Evaluate the model's performance using appropriate metrics (e.g., R^2, RMSE).\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "3369b20b-addd-4932-a112-1527f6813247",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Would you like to use a database with BRAD [Y/N]?\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      " N\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32m2024-08-03 01:18:17 INFO semantic_router.utils.logger local\u001b[0m\n",
      "2024-08-03 01:18:23,659 - INFO - PLANNER\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome to RAG! The chat log from this conversation will be saved to /home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-18-15/log.json. How can I help?\n",
      "==================================================\n",
      "BRAD >> 0: \n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3m**INSTRUCTIONS:**\n",
      "You are planning a bioinformatics analysis pipeline to address a user's query. You must determine if there already exist a predesigned pipeline that will respond to a user's query, or if a new, custom process must be designed for this particular query. The multi-step workflow can use the available methods listed below.\n",
      "\n",
      "**Available Methods within a Pipeline:**\n",
      "1. **RAG**: Look up literature and documents from a text database.\n",
      "2. **SCRAPE**: Search platforms like arXiv, bioRxiv, and PubMed for the latest research.\n",
      "3. **DATABASE**: Utilize bioinformatics databases such as Gene Ontology and Enrichr to perform gene set enrichment analyses.\n",
      "4. **CODE**: Execute bioinformatics pipelines. Utilize prebuilt pipelines or develop new ones as needed.\n",
      "5. **WRITE**: Synthesize and summarize information. This can include summarizing database searches, code pipeline results, or creating a final report to encapsulate the entire analysis.\n",
      "6. **ROUTER**: Determine which step we should proceed to next.\n",
      "\n",
      "User Query: /force PLANNER You are a chatbot assisting to help analyze a single cell dataset.\n",
      "You have  instructions below, and you must use your tools to answer the questions. You have one tool that allows\n",
      "you to write CODE. Generate a plan based on the instructions to use the CODE tool several times in order to satisfy\n",
      "your instructions. Each call to the CODE tool should have a single goal, such as making one type of plot, builting\n",
      "one type of model, and so forth.\n",
      "\n",
      "If your job is to make plots, then you can make a plan to make multiple kinds of plots with each different call\n",
      "to CODE in your plan. For instance. your plan could be\n",
      "\n",
      "**Step 1, CODE** load <path to data>.h5ad and make umap plots of ...\n",
      "**Step 2, CODE** load <path to data>.h5ad and make tsne plots of ...\n",
      "\n",
      "### BOT3: Predictive Modeling\n",
      "**Objective:** Build a model to predict the contribution to reprogramming obtained by the inclusion or exclusion of individual TFs.\n",
      "\n",
      "**Instructions:**\n",
      "1. Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`.\n",
      "2. Prepare the data for modeling:\n",
      "   - Use the embeddings in `adata.X` as features.\n",
      "   - Use the presence/absence of each TF (columns: 'REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B') as additional features.\n",
      "   - Define the target variable based on the distance to 'target' cells (computed by BOT2).\n",
      "3. Split the data into training and testing sets.\n",
      "4. Train a machine learning model (e.g., Random Forest, Gradient Boosting) to predict the reprogramming efficiency.\n",
      "5. Evaluate the model's performance using appropriate metrics (e.g., R^2, RMSE).\n",
      "\n",
      "**Available Pipelines**\n",
      "Name: BOT2\tDescription: this pipeline investigates distances between source, target, and reprogrammed cells. BOT2 should use this pipeline\n",
      "Name: BOT1\tDescription: this plan is for basic visualization and summary of single cell data from the Geneformer reprogramming experiment. The data can be loaded in from ann data objects\n",
      "\n",
      "\n",
      "CUSTOM: The option to build a new pipeline\n",
      "\n",
      "The custom pipeline should be selected if no current pipeline addreses the user's query. If the custom pipeline is selected, please do not attempt to design the pipeline now. We will work to design it at a later point.\n",
      "\n",
      "**Output**:\n",
      "Please formate your output as follows:\n",
      "Pipeline Name: <available pipeline name above or \"CUSTOM\">\n",
      "Explination: <reason for selecting the above pipeline>\n",
      "\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 01:18:27,350 - INFO - Pipeline Name: CUSTOM\n",
      "Explanation: The user's query requires building a predictive model to predict the contribution to reprogramming obtained by the inclusion or exclusion of individual TFs, which is not addressed by any of the available pipelines. Therefore, a custom pipeline needs to be designed to fulfill this specific request.\n",
      "\n",
      "However, I can propose a plan to use the CODE tool several times to achieve the objectives:\n",
      "\n",
      "**Step 1, CODE** Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and prepare the data for modeling.\n",
      "\n",
      "**Step 2, CODE** Define the target variable based on the distance to 'target' cells (computed by BOT2) and split the data into training and testing sets.\n",
      "\n",
      "**Step 3, CODE** Train a Random Forest model to predict the reprogramming efficiency.\n",
      "\n",
      "**Step 4, CODE** Train a Gradient Boosting model to predict the reprogramming efficiency.\n",
      "\n",
      "**Step 5, CODE** Evaluate the performance of both models using appropriate metrics (e.g., R^2, RMSE).\n",
      "\n",
      "**Step 6, WRITE** Summarize the results of the model evaluation and provide a comprehensive report.\n",
      "2024-08-03 01:18:27,350 - INFO - selected_pipeline=CUSTOM\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3m**INSTRUCTIONS:**\n",
      "You are planning a bioinformatics analysis pipeline to address a user's query. Your task is to outline a multi-step workflow using the available methods listed below. Each method should be used appropriately to ensure a thorough analysis. For each step in the pipeline, explain the goal of the process as well as provide a prompt to the chatbot that will execute that step. If information is passed between stages, such as from literature or code to databases, indicate the dependencies of steps in the prompt.\n",
      "\n",
      "**Available Methods:**\n",
      "1. **RAG**: Look up literature and documents from a text database.\n",
      "2. **SCRAPE**: Search platforms like arXiv, bioRxiv, and PubMed for the latest research.\n",
      "3. **DATABASE**: Utilize bioinformatics databases such as Gene Ontology and Enrichr to perform gene set enrichment analyses.\n",
      "4. **CODE**: Execute bioinformatics pipelines. Utilize prebuilt pipelines or develop new ones as needed.\n",
      "5. **WRITE**: Synthesize and summarize information. This can include summarizing database searches, code pipeline results, or creating a final report to encapsulate the entire analysis.\n",
      "6. **ROUTER**: Determine which step we should proceed to next.\n",
      "\n",
      "Current conversation: \n",
      "\n",
      "\n",
      "User Query: /force PLANNER You are a chatbot assisting to help analyze a single cell dataset.\n",
      "You have  instructions below, and you must use your tools to answer the questions. You have one tool that allows\n",
      "you to write CODE. Generate a plan based on the instructions to use the CODE tool several times in order to satisfy\n",
      "your instructions. Each call to the CODE tool should have a single goal, such as making one type of plot, builting\n",
      "one type of model, and so forth.\n",
      "\n",
      "If your job is to make plots, then you can make a plan to make multiple kinds of plots with each different call\n",
      "to CODE in your plan. For instance. your plan could be\n",
      "\n",
      "**Step 1, CODE** load <path to data>.h5ad and make umap plots of ...\n",
      "**Step 2, CODE** load <path to data>.h5ad and make tsne plots of ...\n",
      "\n",
      "### BOT3: Predictive Modeling\n",
      "**Objective:** Build a model to predict the contribution to reprogramming obtained by the inclusion or exclusion of individual TFs.\n",
      "\n",
      "**Instructions:**\n",
      "1. Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`.\n",
      "2. Prepare the data for modeling:\n",
      "   - Use the embeddings in `adata.X` as features.\n",
      "   - Use the presence/absence of each TF (columns: 'REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B') as additional features.\n",
      "   - Define the target variable based on the distance to 'target' cells (computed by BOT2).\n",
      "3. Split the data into training and testing sets.\n",
      "4. Train a machine learning model (e.g., Random Forest, Gradient Boosting) to predict the reprogramming efficiency.\n",
      "5. Evaluate the model's performance using appropriate metrics (e.g., R^2, RMSE).\n",
      "\n",
      "Plan:\n",
      "**Step 1 (method, eg. RAG)**:\n",
      "Prompt: [Description of action for chatbot to do Step 1, i.e. do X, Y, Z]\n",
      "\n",
      "**Step 2 (method, eg. CODE)**:\n",
      "Prompt: [Description of action for chatbot to do Step 2, i.e. do A, B, C. Use information from Step X]\n",
      "...\n",
      "\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "Here is the plan to address the user's query:\n",
      "\n",
      "**Plan:**\n",
      "\n",
      "**Step 1, RAG**:\n",
      "Prompt: Search for literature on single-cell analysis, particularly on reprogramming efficiency prediction using machine learning models.\n",
      "\n",
      "**Step 2, CODE**:\n",
      "Prompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings in `adata.X` and the presence/absence of each TF as features.\n",
      "\n",
      "**Step 3, CODE**:\n",
      "Prompt: Using the extracted features, define the target variable based on the distance to 'target' cells (computed by BOT2) and split the data into training and testing sets.\n",
      "\n",
      "**Step 4, CODE**:\n",
      "Prompt: Train a Random Forest model on the training data to predict the reprogramming efficiency using the features extracted in Step 2.\n",
      "\n",
      "**Step 5, CODE**:\n",
      "Prompt: Train a Gradient Boosting model on the training data to predict the reprogramming efficiency using the features extracted in Step 2.\n",
      "\n",
      "**Step 6, WRITE**:\n",
      "Prompt: Write a report summarizing the performance of both models (Random Forest and Gradient Boosting) using metrics such as R^2 and RMSE, and comparing their results.\n",
      "\n",
      "**Step 7, ROUTER**:\n",
      "Prompt: Determine which model performs better and proceed to the next step based on the results.\n",
      "\n",
      "**Step 8, CODE**:\n",
      "Prompt: Using the better-performing model, make predictions on the testing data and create a plot to visualize the predicted reprogramming efficiency.\n",
      "\n",
      "**Step 9, WRITE**:\n",
      "Prompt: Write a final report encapsulating the entire analysis, including the literature review, data preparation, model performance, and predictions.\n",
      "\n",
      "\n",
      "Do you want to proceed with this plan? [Y/N/edit]\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  remove Step 1 and 7. Then, please condense the above steps into only using the code step one time. Please provide very clear and explicit instructions. Remember to include the names of all the adata.obs columns: 'REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B' or TFs to investigate\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 01:18:59,909 - INFO - Based on the most recently proposed Current Plan and the user's new requirements and requested changes, create a revised plan.\n",
      "\n",
      "Current Plan: Here is the plan to address the user's query:\n",
      "\n",
      "**Plan:**\n",
      "\n",
      "**Step 1, RAG**:\n",
      "Prompt: Search for literature on single-cell analysis, particularly on reprogramming efficiency prediction using machine learning models.\n",
      "\n",
      "**Step 2, CODE**:\n",
      "Prompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings in `adata.X` and the presence/absence of each TF as features.\n",
      "\n",
      "**Step 3, CODE**:\n",
      "Prompt: Using the extracted features, define the target variable based on the distance to 'target' cells (computed by BOT2) and split the data into training and testing sets.\n",
      "\n",
      "**Step 4, CODE**:\n",
      "Prompt: Train a Random Forest model on the training data to predict the reprogramming efficiency using the features extracted in Step 2.\n",
      "\n",
      "**Step 5, CODE**:\n",
      "Prompt: Train a Gradient Boosting model on the training data to predict the reprogramming efficiency using the features extracted in Step 2.\n",
      "\n",
      "**Step 6, WRITE**:\n",
      "Prompt: Write a report summarizing the performance of both models (Random Forest and Gradient Boosting) using metrics such as R^2 and RMSE, and comparing their results.\n",
      "\n",
      "**Step 7, ROUTER**:\n",
      "Prompt: Determine which model performs better and proceed to the next step based on the results.\n",
      "\n",
      "**Step 8, CODE**:\n",
      "Prompt: Using the better-performing model, make predictions on the testing data and create a plot to visualize the predicted reprogramming efficiency.\n",
      "\n",
      "**Step 9, WRITE**:\n",
      "Prompt: Write a final report encapsulating the entire analysis, including the literature review, data preparation, model performance, and predictions.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "User's requested changes: {input}\n",
      "\n",
      "**Available Methods:**\n",
      "1. **Retrieval Augmented Generation**(RAG): Look up literature and documents from a text database.\n",
      "2. **Web Search for New Literature**(SCRAPE): Search platforms like arXiv, bioRxiv, and PubMed for the latest research.\n",
      "3. **Bioinformatics Databases**(DATABASE): Utilize databases such as Gene Ontology and Enrichr to perform gene set enrichment analyses.\n",
      "4. **Run Codes**(CODE): Execute bioinformatics pipelines. Utilize prebuilt pipelines or develop new ones as needed.\n",
      "5. **Write a Report**(WRITE): Synthesize and summarize information. This can include summarizing database searches, code pipeline results, or creating a final report to encapsulate the entire analysis.\n",
      "\n",
      "**OUTPUT**\n",
      "Revised Plan:\n",
      "**Step 1 (method, eg. RAG)**:\n",
      "Prompt: [Description of action for chatbot to do Step 1, i.e. do X, Y, Z]\n",
      "\n",
      "**Step 2 (method, eg. CODE)**:\n",
      "Prompt: [Description of action for chatbot to do Step 2, i.e. do X, Y, Z]\n",
      "...\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Here is the revised plan:\n",
      "\n",
      "**Revised Plan:**\n",
      "\n",
      "**Step 1 (CODE)**:\n",
      "Prompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings in `adata.X` and the presence/absence of each TF as features, specifically 'REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B'.\n",
      "\n",
      "**Step 2 (CODE)**:\n",
      "Prompt: Using the extracted features, define the target variable based on the distance to 'target' cells (computed by BOT2) and split the data into training and testing sets. Then, train both a Random Forest model and a Gradient Boosting model on the training data to predict the reprogramming efficiency using the features extracted in Step 1.\n",
      "\n",
      "**Step 3 (WRITE)**:\n",
      "Prompt: Write a report summarizing the performance of both models (Random Forest and Gradient Boosting) using metrics such as R^2 and RMSE, and comparing their results.\n",
      "\n",
      "**Step 4 (CODE)**:\n",
      "Prompt: Using the better-performing model, make predictions on the testing data and create a plot to visualize the predicted reprogramming efficiency.\n",
      "\n",
      "**Step 5 (WRITE)**:\n",
      "Prompt: Write a final report encapsulating the entire analysis, including the data preparation, model performance, and predictions.\n",
      "\n",
      "Note: I removed Step 1 and 7 as per the user's request, and condensed the steps to use the CODE method only twice. I also included the names of all the adata.obs columns as requested. Let me know if you need any further changes!\n",
      "\n",
      "\n",
      "Do you want to proceed with this plan? [Y/N/edit]\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  Please remove steps 3 and 5 and condense the other steps to a single step\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 01:19:26,316 - INFO - Based on the most recently proposed Current Plan and the user's new requirements and requested changes, create a revised plan.\n",
      "\n",
      "Current Plan: Here is the revised plan:\n",
      "\n",
      "**Revised Plan:**\n",
      "\n",
      "**Step 1 (CODE)**:\n",
      "Prompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings in `adata.X` and the presence/absence of each TF as features, specifically 'REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B'.\n",
      "\n",
      "**Step 2 (CODE)**:\n",
      "Prompt: Using the extracted features, define the target variable based on the distance to 'target' cells (computed by BOT2) and split the data into training and testing sets. Then, train both a Random Forest model and a Gradient Boosting model on the training data to predict the reprogramming efficiency using the features extracted in Step 1.\n",
      "\n",
      "**Step 3 (WRITE)**:\n",
      "Prompt: Write a report summarizing the performance of both models (Random Forest and Gradient Boosting) using metrics such as R^2 and RMSE, and comparing their results.\n",
      "\n",
      "**Step 4 (CODE)**:\n",
      "Prompt: Using the better-performing model, make predictions on the testing data and create a plot to visualize the predicted reprogramming efficiency.\n",
      "\n",
      "**Step 5 (WRITE)**:\n",
      "Prompt: Write a final report encapsulating the entire analysis, including the data preparation, model performance, and predictions.\n",
      "\n",
      "Note: I removed Step 1 and 7 as per the user's request, and condensed the steps to use the CODE method only twice. I also included the names of all the adata.obs columns as requested. Let me know if you need any further changes!\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "User's requested changes: {input}\n",
      "\n",
      "**Available Methods:**\n",
      "1. **Retrieval Augmented Generation**(RAG): Look up literature and documents from a text database.\n",
      "2. **Web Search for New Literature**(SCRAPE): Search platforms like arXiv, bioRxiv, and PubMed for the latest research.\n",
      "3. **Bioinformatics Databases**(DATABASE): Utilize databases such as Gene Ontology and Enrichr to perform gene set enrichment analyses.\n",
      "4. **Run Codes**(CODE): Execute bioinformatics pipelines. Utilize prebuilt pipelines or develop new ones as needed.\n",
      "5. **Write a Report**(WRITE): Synthesize and summarize information. This can include summarizing database searches, code pipeline results, or creating a final report to encapsulate the entire analysis.\n",
      "\n",
      "**OUTPUT**\n",
      "Revised Plan:\n",
      "**Step 1 (method, eg. RAG)**:\n",
      "Prompt: [Description of action for chatbot to do Step 1, i.e. do X, Y, Z]\n",
      "\n",
      "**Step 2 (method, eg. CODE)**:\n",
      "Prompt: [Description of action for chatbot to do Step 2, i.e. do X, Y, Z]\n",
      "...\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Here is the revised plan:\n",
      "\n",
      "**Revised Plan:**\n",
      "\n",
      "**Step 1 (CODE)**:\n",
      "Prompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings in `adata.X` and the presence/absence of each TF as features, specifically 'REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B'. Using the extracted features, define the target variable based on the distance to 'target' cells (computed by BOT2) and split the data into training and testing sets. Then, train both a Random Forest model and a Gradient Boosting model on the training data to predict the reprogramming efficiency using the features extracted. Finally, make predictions on the testing data using the better-performing model and create a plot to visualize the predicted reprogramming efficiency.\n",
      "\n",
      "This revised plan condenses the original steps into a single step that uses the CODE method to perform all the required tasks. Steps 3 and 5 have been removed as requested, and the other steps have been combined into a single step.\n",
      "\n",
      "\n",
      "Do you want to proceed with this plan? [Y/N/edit]\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  Y\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 01:19:39,257 - INFO - [{'order': 0, 'module': 'PLANNER', 'prompt': None, 'description': 'This step designed the plan. It is placed in the queue because we needed a place holder for 0 indexed lists.'}, {'order': 1, 'module': 'CODE', 'prompt': \"/force CODE  1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings in `adata.X` and the presence/absence of each TF as features, specifically 'REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B'. Using the extracted features, define the target variable based on the distance to 'target' cells (computed by BOT2) and split the data into training and testing sets. Then, train both a Random Forest model and a Gradient Boosting model on the training data to predict the reprogramming efficiency using the features extracted. Finally, make predictions on the testing data using the better-performing model and create a plot to visualize the predicted reprogramming efficiency.\\n\\nThis revised plan condenses the original steps into a single step that uses the CODE method to perform all the required tasks. \", 'description': \" 1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings in `adata.X` and the presence/absence of each TF as features, specifically 'REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B'. Using the extracted features, define the target variable based on the distance to 'target' cells (computed by BOT2) and split the data into training and testing sets. Then, train both a Random Forest model and a Gradient Boosting model on the training data to predict the reprogramming efficiency using the features extracted. Finally, make predictions on the testing data using the better-performing model and create a plot to visualize the predicted reprogramming efficiency.\\n\\nThis revised plan condenses the original steps into a single step that uses the CODE method to perform all the required tasks. \"}]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Here is the revised plan:\\n\\n**Revised Plan:**\\n\\n**', \" 1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings in `adata.X` and the presence/absence of each TF as features, specifically 'REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B'. Using the extracted features, define the target variable based on the distance to 'target' cells (computed by BOT2) and split the data into training and testing sets. Then, train both a Random Forest model and a Gradient Boosting model on the training data to predict the reprogramming efficiency using the features extracted. Finally, make predictions on the testing data using the better-performing model and create a plot to visualize the predicted reprogramming efficiency.\\n\\nThis revised plan condenses the original steps into a single step that uses the CODE method to perform all the required tasks. \", 's 3 and 5 have been removed as requested, and the other steps have been combined into a single step.\\n\\n']\n",
      "Would you like to save this pipeline to a file? [Y/N]\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      ">>> Y\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter a name for your pipeline\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      ">>> BOT3\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter a description of your pipeline\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      ">>> this pipeline should be run by BOT3\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 01:19:55,164 - INFO - \n",
      "\n",
      "\n",
      "route\n",
      "\n",
      "\n",
      "\n",
      "2024-08-03 01:19:55,164 - INFO - PLANNER\n",
      "2024-08-03 01:19:55,165 - INFO - \n",
      "\n",
      "PLANNER:\n",
      "\n",
      "{'module': 'PLANNER', 'steps': [{'llm': \"model='meta/llama3-70b-instruct'\", 'memory': 'None', 'prompt': 'input_variables=[\\'input\\'] template=\\'**INSTRUCTIONS:**\\\\nYou are planning a bioinformatics analysis pipeline to address a user\\\\\\'s query. You must determine if there already exist a predesigned pipeline that will respond to a user\\\\\\'s query, or if a new, custom process must be designed for this particular query. The multi-step workflow can use the available methods listed below.\\\\n\\\\n**Available Methods within a Pipeline:**\\\\n1. **RAG**: Look up literature and documents from a text database.\\\\n2. **SCRAPE**: Search platforms like arXiv, bioRxiv, and PubMed for the latest research.\\\\n3. **DATABASE**: Utilize bioinformatics databases such as Gene Ontology and Enrichr to perform gene set enrichment analyses.\\\\n4. **CODE**: Execute bioinformatics pipelines. Utilize prebuilt pipelines or develop new ones as needed.\\\\n5. **WRITE**: Synthesize and summarize information. This can include summarizing database searches, code pipeline results, or creating a final report to encapsulate the entire analysis.\\\\n6. **ROUTER**: Determine which step we should proceed to next.\\\\n\\\\nUser Query: {input}\\\\n\\\\n**Available Pipelines**\\\\nName: BOT2\\\\tDescription: this pipeline investigates distances between source, target, and reprogrammed cells. BOT2 should use this pipeline\\\\nName: BOT1\\\\tDescription: this plan is for basic visualization and summary of single cell data from the Geneformer reprogramming experiment. The data can be loaded in from ann data objects\\\\n\\\\n\\\\nCUSTOM: The option to build a new pipeline\\\\n\\\\nThe custom pipeline should be selected if no current pipeline addreses the user\\\\\\'s query. If the custom pipeline is selected, please do not attempt to design the pipeline now. We will work to design it at a later point.\\\\n\\\\n**Output**:\\\\nPlease formate your output as follows:\\\\nPipeline Name: <available pipeline name above or \"CUSTOM\">\\\\nExplination: <reason for selecting the above pipeline>\\\\n\\'', 'input': \"/force PLANNER You are a chatbot assisting to help analyze a single cell dataset.\\nYou have  instructions below, and you must use your tools to answer the questions. You have one tool that allows\\nyou to write CODE. Generate a plan based on the instructions to use the CODE tool several times in order to satisfy\\nyour instructions. Each call to the CODE tool should have a single goal, such as making one type of plot, builting\\none type of model, and so forth.\\n\\nIf your job is to make plots, then you can make a plan to make multiple kinds of plots with each different call\\nto CODE in your plan. For instance. your plan could be\\n\\n**Step 1, CODE** load <path to data>.h5ad and make umap plots of ...\\n**Step 2, CODE** load <path to data>.h5ad and make tsne plots of ...\\n\\n### BOT3: Predictive Modeling\\n**Objective:** Build a model to predict the contribution to reprogramming obtained by the inclusion or exclusion of individual TFs.\\n\\n**Instructions:**\\n1. Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`.\\n2. Prepare the data for modeling:\\n   - Use the embeddings in `adata.X` as features.\\n   - Use the presence/absence of each TF (columns: 'REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B') as additional features.\\n   - Define the target variable based on the distance to 'target' cells (computed by BOT2).\\n3. Split the data into training and testing sets.\\n4. Train a machine learning model (e.g., Random Forest, Gradient Boosting) to predict the reprogramming efficiency.\\n5. Evaluate the model's performance using appropriate metrics (e.g., R^2, RMSE).\", 'output': \"Pipeline Name: CUSTOM\\nExplanation: The user's query requires building a predictive model to predict the contribution to reprogramming obtained by the inclusion or exclusion of individual TFs, which is not addressed by any of the available pipelines. Therefore, a custom pipeline needs to be designed to fulfill this specific request.\\n\\nHowever, I can propose a plan to use the CODE tool several times to achieve the objectives:\\n\\n**Step 1, CODE** Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and prepare the data for modeling.\\n\\n**Step 2, CODE** Define the target variable based on the distance to 'target' cells (computed by BOT2) and split the data into training and testing sets.\\n\\n**Step 3, CODE** Train a Random Forest model to predict the reprogramming efficiency.\\n\\n**Step 4, CODE** Train a Gradient Boosting model to predict the reprogramming efficiency.\\n\\n**Step 5, CODE** Evaluate the performance of both models using appropriate metrics (e.g., R^2, RMSE).\\n\\n**Step 6, WRITE** Summarize the results of the model evaluation and provide a comprehensive report.\", 'parsedOutput': {'selected pipeline': 'CUSTOM'}, 'purpose': 'determine if a known pipeline can be used or a new one si required'}, {'llm': \"model='meta/llama3-70b-instruct'\", 'memory': 'chat_memory=InMemoryChatMessageHistory(messages=[HumanMessage(content=\"/force PLANNER You are a chatbot assisting to help analyze a single cell dataset.\\\\nYou have  instructions below, and you must use your tools to answer the questions. You have one tool that allows\\\\nyou to write CODE. Generate a plan based on the instructions to use the CODE tool several times in order to satisfy\\\\nyour instructions. Each call to the CODE tool should have a single goal, such as making one type of plot, builting\\\\none type of model, and so forth.\\\\n\\\\nIf your job is to make plots, then you can make a plan to make multiple kinds of plots with each different call\\\\nto CODE in your plan. For instance. your plan could be\\\\n\\\\n**Step 1, CODE** load <path to data>.h5ad and make umap plots of ...\\\\n**Step 2, CODE** load <path to data>.h5ad and make tsne plots of ...\\\\n\\\\n### BOT3: Predictive Modeling\\\\n**Objective:** Build a model to predict the contribution to reprogramming obtained by the inclusion or exclusion of individual TFs.\\\\n\\\\n**Instructions:**\\\\n1. Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`.\\\\n2. Prepare the data for modeling:\\\\n   - Use the embeddings in `adata.X` as features.\\\\n   - Use the presence/absence of each TF (columns: \\'REL\\', \\'STAT5A\\', \\'FOS\\', \\'ETV6\\', \\'MEF2C\\', \\'RUNX3\\', \\'FOSB\\', \\'GATA2\\', \\'IKZF1\\', \\'GFI1B\\') as additional features.\\\\n   - Define the target variable based on the distance to \\'target\\' cells (computed by BOT2).\\\\n3. Split the data into training and testing sets.\\\\n4. Train a machine learning model (e.g., Random Forest, Gradient Boosting) to predict the reprogramming efficiency.\\\\n5. Evaluate the model\\'s performance using appropriate metrics (e.g., R^2, RMSE).\"), AIMessage(content=\"Here is the plan to address the user\\'s query:\\\\n\\\\n**Plan:**\\\\n\\\\n**Step 1, RAG**:\\\\nPrompt: Search for literature on single-cell analysis, particularly on reprogramming efficiency prediction using machine learning models.\\\\n\\\\n**Step 2, CODE**:\\\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings in `adata.X` and the presence/absence of each TF as features.\\\\n\\\\n**Step 3, CODE**:\\\\nPrompt: Using the extracted features, define the target variable based on the distance to \\'target\\' cells (computed by BOT2) and split the data into training and testing sets.\\\\n\\\\n**Step 4, CODE**:\\\\nPrompt: Train a Random Forest model on the training data to predict the reprogramming efficiency using the features extracted in Step 2.\\\\n\\\\n**Step 5, CODE**:\\\\nPrompt: Train a Gradient Boosting model on the training data to predict the reprogramming efficiency using the features extracted in Step 2.\\\\n\\\\n**Step 6, WRITE**:\\\\nPrompt: Write a report summarizing the performance of both models (Random Forest and Gradient Boosting) using metrics such as R^2 and RMSE, and comparing their results.\\\\n\\\\n**Step 7, ROUTER**:\\\\nPrompt: Determine which model performs better and proceed to the next step based on the results.\\\\n\\\\n**Step 8, CODE**:\\\\nPrompt: Using the better-performing model, make predictions on the testing data and create a plot to visualize the predicted reprogramming efficiency.\\\\n\\\\n**Step 9, WRITE**:\\\\nPrompt: Write a final report encapsulating the entire analysis, including the literature review, data preparation, model performance, and predictions.\")]) ai_prefix=\\'BRAD\\'', 'prompt': 'input_variables=[\\'history\\', \\'input\\'] template=\"**INSTRUCTIONS:**\\\\nYou are planning a bioinformatics analysis pipeline to address a user\\'s query. Your task is to outline a multi-step workflow using the available methods listed below. Each method should be used appropriately to ensure a thorough analysis. For each step in the pipeline, explain the goal of the process as well as provide a prompt to the chatbot that will execute that step. If information is passed between stages, such as from literature or code to databases, indicate the dependencies of steps in the prompt.\\\\n\\\\n**Available Methods:**\\\\n1. **RAG**: Look up literature and documents from a text database.\\\\n2. **SCRAPE**: Search platforms like arXiv, bioRxiv, and PubMed for the latest research.\\\\n3. **DATABASE**: Utilize bioinformatics databases such as Gene Ontology and Enrichr to perform gene set enrichment analyses.\\\\n4. **CODE**: Execute bioinformatics pipelines. Utilize prebuilt pipelines or develop new ones as needed.\\\\n5. **WRITE**: Synthesize and summarize information. This can include summarizing database searches, code pipeline results, or creating a final report to encapsulate the entire analysis.\\\\n6. **ROUTER**: Determine which step we should proceed to next.\\\\n\\\\nCurrent conversation: {history}\\\\n\\\\n\\\\nUser Query: {input}\\\\n\\\\nPlan:\\\\n**Step 1 (method, eg. RAG)**:\\\\nPrompt: [Description of action for chatbot to do Step 1, i.e. do X, Y, Z]\\\\n\\\\n**Step 2 (method, eg. CODE)**:\\\\nPrompt: [Description of action for chatbot to do Step 2, i.e. do A, B, C. Use information from Step X]\\\\n...\\\\n\"', 'input': \"/force PLANNER You are a chatbot assisting to help analyze a single cell dataset.\\nYou have  instructions below, and you must use your tools to answer the questions. You have one tool that allows\\nyou to write CODE. Generate a plan based on the instructions to use the CODE tool several times in order to satisfy\\nyour instructions. Each call to the CODE tool should have a single goal, such as making one type of plot, builting\\none type of model, and so forth.\\n\\nIf your job is to make plots, then you can make a plan to make multiple kinds of plots with each different call\\nto CODE in your plan. For instance. your plan could be\\n\\n**Step 1, CODE** load <path to data>.h5ad and make umap plots of ...\\n**Step 2, CODE** load <path to data>.h5ad and make tsne plots of ...\\n\\n### BOT3: Predictive Modeling\\n**Objective:** Build a model to predict the contribution to reprogramming obtained by the inclusion or exclusion of individual TFs.\\n\\n**Instructions:**\\n1. Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad`.\\n2. Prepare the data for modeling:\\n   - Use the embeddings in `adata.X` as features.\\n   - Use the presence/absence of each TF (columns: 'REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B') as additional features.\\n   - Define the target variable based on the distance to 'target' cells (computed by BOT2).\\n3. Split the data into training and testing sets.\\n4. Train a machine learning model (e.g., Random Forest, Gradient Boosting) to predict the reprogramming efficiency.\\n5. Evaluate the model's performance using appropriate metrics (e.g., R^2, RMSE).\", 'output': \"Here is the plan to address the user's query:\\n\\n**Plan:**\\n\\n**Step 1, RAG**:\\nPrompt: Search for literature on single-cell analysis, particularly on reprogramming efficiency prediction using machine learning models.\\n\\n**Step 2, CODE**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings in `adata.X` and the presence/absence of each TF as features.\\n\\n**Step 3, CODE**:\\nPrompt: Using the extracted features, define the target variable based on the distance to 'target' cells (computed by BOT2) and split the data into training and testing sets.\\n\\n**Step 4, CODE**:\\nPrompt: Train a Random Forest model on the training data to predict the reprogramming efficiency using the features extracted in Step 2.\\n\\n**Step 5, CODE**:\\nPrompt: Train a Gradient Boosting model on the training data to predict the reprogramming efficiency using the features extracted in Step 2.\\n\\n**Step 6, WRITE**:\\nPrompt: Write a report summarizing the performance of both models (Random Forest and Gradient Boosting) using metrics such as R^2 and RMSE, and comparing their results.\\n\\n**Step 7, ROUTER**:\\nPrompt: Determine which model performs better and proceed to the next step based on the results.\\n\\n**Step 8, CODE**:\\nPrompt: Using the better-performing model, make predictions on the testing data and create a plot to visualize the predicted reprogramming efficiency.\\n\\n**Step 9, WRITE**:\\nPrompt: Write a final report encapsulating the entire analysis, including the literature review, data preparation, model performance, and predictions.\\n\\n\", 'parsedOutput': {'output': \"Here is the plan to address the user's query:\\n\\n**Plan:**\\n\\n**Step 1, RAG**:\\nPrompt: Search for literature on single-cell analysis, particularly on reprogramming efficiency prediction using machine learning models.\\n\\n**Step 2, CODE**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings in `adata.X` and the presence/absence of each TF as features.\\n\\n**Step 3, CODE**:\\nPrompt: Using the extracted features, define the target variable based on the distance to 'target' cells (computed by BOT2) and split the data into training and testing sets.\\n\\n**Step 4, CODE**:\\nPrompt: Train a Random Forest model on the training data to predict the reprogramming efficiency using the features extracted in Step 2.\\n\\n**Step 5, CODE**:\\nPrompt: Train a Gradient Boosting model on the training data to predict the reprogramming efficiency using the features extracted in Step 2.\\n\\n**Step 6, WRITE**:\\nPrompt: Write a report summarizing the performance of both models (Random Forest and Gradient Boosting) using metrics such as R^2 and RMSE, and comparing their results.\\n\\n**Step 7, ROUTER**:\\nPrompt: Determine which model performs better and proceed to the next step based on the results.\\n\\n**Step 8, CODE**:\\nPrompt: Using the better-performing model, make predictions on the testing data and create a plot to visualize the predicted reprogramming efficiency.\\n\\n**Step 9, WRITE**:\\nPrompt: Write a final report encapsulating the entire analysis, including the literature review, data preparation, model performance, and predictions.\\n\\n\"}, 'purpose': 'generate a proposed set of prompts'}, {'func': 'planner.planner', 'prompt to user': 'Do you want to proceed with this plan? [Y/N/edit]', 'input': \"remove Step 1 and 7. Then, please condense the above steps into only using the code step one time. Please provide very clear and explicit instructions. Remember to include the names of all the adata.obs columns: 'REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B' or TFs to investigate\", 'purpose': 'get new user input'}, {'llm': \"model='meta/llama3-70b-instruct'\", 'memory': 'None', 'prompt': 'input_variables=[\\'input\\'] template=\"Based on the most recently proposed Current Plan and the user\\'s new requirements and requested changes, create a revised plan.\\\\n\\\\nCurrent Plan: Here is the plan to address the user\\'s query:\\\\n\\\\n**Plan:**\\\\n\\\\n**Step 1, RAG**:\\\\nPrompt: Search for literature on single-cell analysis, particularly on reprogramming efficiency prediction using machine learning models.\\\\n\\\\n**Step 2, CODE**:\\\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings in `adata.X` and the presence/absence of each TF as features.\\\\n\\\\n**Step 3, CODE**:\\\\nPrompt: Using the extracted features, define the target variable based on the distance to \\'target\\' cells (computed by BOT2) and split the data into training and testing sets.\\\\n\\\\n**Step 4, CODE**:\\\\nPrompt: Train a Random Forest model on the training data to predict the reprogramming efficiency using the features extracted in Step 2.\\\\n\\\\n**Step 5, CODE**:\\\\nPrompt: Train a Gradient Boosting model on the training data to predict the reprogramming efficiency using the features extracted in Step 2.\\\\n\\\\n**Step 6, WRITE**:\\\\nPrompt: Write a report summarizing the performance of both models (Random Forest and Gradient Boosting) using metrics such as R^2 and RMSE, and comparing their results.\\\\n\\\\n**Step 7, ROUTER**:\\\\nPrompt: Determine which model performs better and proceed to the next step based on the results.\\\\n\\\\n**Step 8, CODE**:\\\\nPrompt: Using the better-performing model, make predictions on the testing data and create a plot to visualize the predicted reprogramming efficiency.\\\\n\\\\n**Step 9, WRITE**:\\\\nPrompt: Write a final report encapsulating the entire analysis, including the literature review, data preparation, model performance, and predictions.\\\\n\\\\n\\\\n\\\\n\\\\nUser\\'s requested changes: {input}\\\\n\\\\n**Available Methods:**\\\\n1. **Retrieval Augmented Generation**(RAG): Look up literature and documents from a text database.\\\\n2. **Web Search for New Literature**(SCRAPE): Search platforms like arXiv, bioRxiv, and PubMed for the latest research.\\\\n3. **Bioinformatics Databases**(DATABASE): Utilize databases such as Gene Ontology and Enrichr to perform gene set enrichment analyses.\\\\n4. **Run Codes**(CODE): Execute bioinformatics pipelines. Utilize prebuilt pipelines or develop new ones as needed.\\\\n5. **Write a Report**(WRITE): Synthesize and summarize information. This can include summarizing database searches, code pipeline results, or creating a final report to encapsulate the entire analysis.\\\\n\\\\n**OUTPUT**\\\\nRevised Plan:\\\\n**Step 1 (method, eg. RAG)**:\\\\nPrompt: [Description of action for chatbot to do Step 1, i.e. do X, Y, Z]\\\\n\\\\n**Step 2 (method, eg. CODE)**:\\\\nPrompt: [Description of action for chatbot to do Step 2, i.e. do X, Y, Z]\\\\n...\\\\n\"', 'input': \"remove Step 1 and 7. Then, please condense the above steps into only using the code step one time. Please provide very clear and explicit instructions. Remember to include the names of all the adata.obs columns: 'REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B' or TFs to investigate\", 'output': \"Here is the revised plan:\\n\\n**Revised Plan:**\\n\\n**Step 1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings in `adata.X` and the presence/absence of each TF as features, specifically 'REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B'.\\n\\n**Step 2 (CODE)**:\\nPrompt: Using the extracted features, define the target variable based on the distance to 'target' cells (computed by BOT2) and split the data into training and testing sets. Then, train both a Random Forest model and a Gradient Boosting model on the training data to predict the reprogramming efficiency using the features extracted in Step 1.\\n\\n**Step 3 (WRITE)**:\\nPrompt: Write a report summarizing the performance of both models (Random Forest and Gradient Boosting) using metrics such as R^2 and RMSE, and comparing their results.\\n\\n**Step 4 (CODE)**:\\nPrompt: Using the better-performing model, make predictions on the testing data and create a plot to visualize the predicted reprogramming efficiency.\\n\\n**Step 5 (WRITE)**:\\nPrompt: Write a final report encapsulating the entire analysis, including the data preparation, model performance, and predictions.\\n\\nNote: I removed Step 1 and 7 as per the user's request, and condensed the steps to use the CODE method only twice. I also included the names of all the adata.obs columns as requested. Let me know if you need any further changes!\\n\\n\", 'parsedOutput': {'output': \"Here is the revised plan:\\n\\n**Revised Plan:**\\n\\n**Step 1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings in `adata.X` and the presence/absence of each TF as features, specifically 'REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B'.\\n\\n**Step 2 (CODE)**:\\nPrompt: Using the extracted features, define the target variable based on the distance to 'target' cells (computed by BOT2) and split the data into training and testing sets. Then, train both a Random Forest model and a Gradient Boosting model on the training data to predict the reprogramming efficiency using the features extracted in Step 1.\\n\\n**Step 3 (WRITE)**:\\nPrompt: Write a report summarizing the performance of both models (Random Forest and Gradient Boosting) using metrics such as R^2 and RMSE, and comparing their results.\\n\\n**Step 4 (CODE)**:\\nPrompt: Using the better-performing model, make predictions on the testing data and create a plot to visualize the predicted reprogramming efficiency.\\n\\n**Step 5 (WRITE)**:\\nPrompt: Write a final report encapsulating the entire analysis, including the data preparation, model performance, and predictions.\\n\\nNote: I removed Step 1 and 7 as per the user's request, and condensed the steps to use the CODE method only twice. I also included the names of all the adata.obs columns as requested. Let me know if you need any further changes!\\n\\n\"}, 'purpose': 'update the proposed set of prompts'}, {'func': 'planner.planner', 'prompt to user': 'Do you want to proceed with this plan? [Y/N/edit]', 'input': 'Please remove steps 3 and 5 and condense the other steps to a single step', 'purpose': 'get new user input'}, {'llm': \"model='meta/llama3-70b-instruct'\", 'memory': 'None', 'prompt': 'input_variables=[\\'input\\'] template=\"Based on the most recently proposed Current Plan and the user\\'s new requirements and requested changes, create a revised plan.\\\\n\\\\nCurrent Plan: Here is the revised plan:\\\\n\\\\n**Revised Plan:**\\\\n\\\\n**Step 1 (CODE)**:\\\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings in `adata.X` and the presence/absence of each TF as features, specifically \\'REL\\', \\'STAT5A\\', \\'FOS\\', \\'ETV6\\', \\'MEF2C\\', \\'RUNX3\\', \\'FOSB\\', \\'GATA2\\', \\'IKZF1\\', \\'GFI1B\\'.\\\\n\\\\n**Step 2 (CODE)**:\\\\nPrompt: Using the extracted features, define the target variable based on the distance to \\'target\\' cells (computed by BOT2) and split the data into training and testing sets. Then, train both a Random Forest model and a Gradient Boosting model on the training data to predict the reprogramming efficiency using the features extracted in Step 1.\\\\n\\\\n**Step 3 (WRITE)**:\\\\nPrompt: Write a report summarizing the performance of both models (Random Forest and Gradient Boosting) using metrics such as R^2 and RMSE, and comparing their results.\\\\n\\\\n**Step 4 (CODE)**:\\\\nPrompt: Using the better-performing model, make predictions on the testing data and create a plot to visualize the predicted reprogramming efficiency.\\\\n\\\\n**Step 5 (WRITE)**:\\\\nPrompt: Write a final report encapsulating the entire analysis, including the data preparation, model performance, and predictions.\\\\n\\\\nNote: I removed Step 1 and 7 as per the user\\'s request, and condensed the steps to use the CODE method only twice. I also included the names of all the adata.obs columns as requested. Let me know if you need any further changes!\\\\n\\\\n\\\\n\\\\n\\\\nUser\\'s requested changes: {input}\\\\n\\\\n**Available Methods:**\\\\n1. **Retrieval Augmented Generation**(RAG): Look up literature and documents from a text database.\\\\n2. **Web Search for New Literature**(SCRAPE): Search platforms like arXiv, bioRxiv, and PubMed for the latest research.\\\\n3. **Bioinformatics Databases**(DATABASE): Utilize databases such as Gene Ontology and Enrichr to perform gene set enrichment analyses.\\\\n4. **Run Codes**(CODE): Execute bioinformatics pipelines. Utilize prebuilt pipelines or develop new ones as needed.\\\\n5. **Write a Report**(WRITE): Synthesize and summarize information. This can include summarizing database searches, code pipeline results, or creating a final report to encapsulate the entire analysis.\\\\n\\\\n**OUTPUT**\\\\nRevised Plan:\\\\n**Step 1 (method, eg. RAG)**:\\\\nPrompt: [Description of action for chatbot to do Step 1, i.e. do X, Y, Z]\\\\n\\\\n**Step 2 (method, eg. CODE)**:\\\\nPrompt: [Description of action for chatbot to do Step 2, i.e. do X, Y, Z]\\\\n...\\\\n\"', 'input': 'Please remove steps 3 and 5 and condense the other steps to a single step', 'output': \"Here is the revised plan:\\n\\n**Revised Plan:**\\n\\n**Step 1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings in `adata.X` and the presence/absence of each TF as features, specifically 'REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B'. Using the extracted features, define the target variable based on the distance to 'target' cells (computed by BOT2) and split the data into training and testing sets. Then, train both a Random Forest model and a Gradient Boosting model on the training data to predict the reprogramming efficiency using the features extracted. Finally, make predictions on the testing data using the better-performing model and create a plot to visualize the predicted reprogramming efficiency.\\n\\nThis revised plan condenses the original steps into a single step that uses the CODE method to perform all the required tasks. Steps 3 and 5 have been removed as requested, and the other steps have been combined into a single step.\\n\\n\", 'parsedOutput': {'output': \"Here is the revised plan:\\n\\n**Revised Plan:**\\n\\n**Step 1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings in `adata.X` and the presence/absence of each TF as features, specifically 'REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B'. Using the extracted features, define the target variable based on the distance to 'target' cells (computed by BOT2) and split the data into training and testing sets. Then, train both a Random Forest model and a Gradient Boosting model on the training data to predict the reprogramming efficiency using the features extracted. Finally, make predictions on the testing data using the better-performing model and create a plot to visualize the predicted reprogramming efficiency.\\n\\nThis revised plan condenses the original steps into a single step that uses the CODE method to perform all the required tasks. Steps 3 and 5 have been removed as requested, and the other steps have been combined into a single step.\\n\\n\"}, 'purpose': 'update the proposed set of prompts'}, {'func': 'planner.planner', 'prompt to user': 'Do you want to proceed with this plan? [Y/N/edit]', 'input': 'Y', 'purpose': 'get new user input'}, {'func': 'planner.planner', 'what': 'designed a new pipeline'}, {'func': 'planner.planner', 'what': 'set the queue and set the queue pointer to 1'}]}\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pipeline saved successfully to /home/jpic/RAG-DEV/tutorials/geneformer/pipelines/BOT3.json\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "\"Here is the plan to address the user's query:\\n\\n**Plan:**\\n\\n**Step 1, RAG**:\\nPrompt: Search for literature on single-cell analysis, particularly on reprogramming efficiency prediction using machine learning models.\\n\\n**Step 2, CODE**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings in `adata.X` and the presence/absence of each TF as features.\\n\\n**Step 3, CODE**:\\nPrompt: Using the extracted features, define the target variable based on the distance to 'target' cells (computed by BOT2) and split the data into training and testing sets.\\n\\n**Step 4, CODE**:\\nPrompt: Train a Random Forest model on the training data to predict the reprogramming efficiency using the features extracted in Step 2.\\n\\n**Step 5, CODE**:\\nPrompt: Train a Gradient Boosting model on the training data to predict the reprogramming efficiency using the features extracted in Step 2.\\n\\n**Step 6, WRITE**:\\nPrompt: Write a report summarizing the performance of both models (Random Forest and Gradient Boosting) using metrics such as R^2 and RMSE, and comparing their results.\\n\\n**Step 7, ROUTER**:\\nPrompt: Determine which model performs better and proceed to the next step based on the results.\\n\\n**Step 8, CODE**:\\nPrompt: Using the better-performing model, make predictions on the testing data and create a plot to visualize the predicted reprogramming efficiency.\\n\\n**Step 9, WRITE**:\\nPrompt: Write a final report encapsulating the entire analysis, including the literature review, data preparation, model performance, and predictions.\\n\\nDo you want to proceed with this plan? [Y/N/edit]Here is the revised plan:\\n\\n**Revised Plan:**\\n\\n**Step 1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings in `adata.X` and the presence/absence of each TF as features, specifically 'REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B'.\\n\\n**Step 2 (CODE)**:\\nPrompt: Using the extracted features, define the target variable based on the distance to 'target' cells (computed by BOT2) and split the data into training and testing sets. Then, train both a Random Forest model and a Gradient Boosting model on the training data to predict the reprogramming efficiency using the features extracted in Step 1.\\n\\n**Step 3 (WRITE)**:\\nPrompt: Write a report summarizing the performance of both models (Random Forest and Gradient Boosting) using metrics such as R^2 and RMSE, and comparing their results.\\n\\n**Step 4 (CODE)**:\\nPrompt: Using the better-performing model, make predictions on the testing data and create a plot to visualize the predicted reprogramming efficiency.\\n\\n**Step 5 (WRITE)**:\\nPrompt: Write a final report encapsulating the entire analysis, including the data preparation, model performance, and predictions.\\n\\nNote: I removed Step 1 and 7 as per the user's request, and condensed the steps to use the CODE method only twice. I also included the names of all the adata.obs columns as requested. Let me know if you need any further changes!\\n\\nDo you want to proceed with this plan? [Y/N/edit]Here is the revised plan:\\n\\n**Revised Plan:**\\n\\n**Step 1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings in `adata.X` and the presence/absence of each TF as features, specifically 'REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B'. Using the extracted features, define the target variable based on the distance to 'target' cells (computed by BOT2) and split the data into training and testing sets. Then, train both a Random Forest model and a Gradient Boosting model on the training data to predict the reprogramming efficiency using the features extracted. Finally, make predictions on the testing data using the better-performing model and create a plot to visualize the predicted reprogramming efficiency.\\n\\nThis revised plan condenses the original steps into a single step that uses the CODE method to perform all the required tasks. Steps 3 and 5 have been removed as requested, and the other steps have been combined into a single step.\\n\\nDo you want to proceed with this plan? [Y/N/edit]Would you like to save this pipeline to a file? [Y/N]Enter a name for your pipelineEnter a description of your pipelinePipeline saved successfully to /home/jpic/RAG-DEV/tutorials/geneformer/pipelines/BOT3.json\""
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# llm = llms.load_openai(model_name='gpt-4o')\n",
    "llm = llms.load_nvidia() # model_name='meta/llama-3.1-405b-instruct')\n",
    "BOT3 = brad.chatbot(llm=llm, config='/home/jpic/RAG-DEV/tutorials/geneformer/config-sc.json')\n",
    "BOT3.invoke('/force PLANNER ' + worker_system_prompt + modified_instructions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba520e54-ed5d-4d12-8d4e-481b7a53b14d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32m2024-08-03 01:40:37 INFO semantic_router.utils.logger local\u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome to RAG! The chat log from this conversation will be saved to /home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-40-37/log.json. How can I help?\n",
      "==================================================\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  /force PLANNER please load and run the BOT3 pipeline\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 01:40:54,851 - INFO - PLANNER\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "BRAD >> 0: \n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3m**INSTRUCTIONS:**\n",
      "You are planning a bioinformatics analysis pipeline to address a user's query. You must determine if there already exist a predesigned pipeline that will respond to a user's query, or if a new, custom process must be designed for this particular query. The multi-step workflow can use the available methods listed below.\n",
      "\n",
      "**Available Methods within a Pipeline:**\n",
      "1. **RAG**: Look up literature and documents from a text database.\n",
      "2. **SCRAPE**: Search platforms like arXiv, bioRxiv, and PubMed for the latest research.\n",
      "3. **DATABASE**: Utilize bioinformatics databases such as Gene Ontology and Enrichr to perform gene set enrichment analyses.\n",
      "4. **CODE**: Execute bioinformatics pipelines. Utilize prebuilt pipelines or develop new ones as needed.\n",
      "5. **WRITE**: Synthesize and summarize information. This can include summarizing database searches, code pipeline results, or creating a final report to encapsulate the entire analysis.\n",
      "6. **ROUTER**: Determine which step we should proceed to next.\n",
      "\n",
      "User Query: please load and run the BOT3 pipeline\n",
      "\n",
      "**Available Pipelines**\n",
      "Name: BOT2\tDescription: this pipeline investigates distances between source, target, and reprogrammed cells. BOT2 should use this pipeline\n",
      "Name: BOT3\tDescription: this pipeline should be run by BOT3\n",
      "Name: BOT1\tDescription: this plan is for basic visualization and summary of single cell data from the Geneformer reprogramming experiment. The data can be loaded in from ann data objects\n",
      "\n",
      "\n",
      "CUSTOM: The option to build a new pipeline\n",
      "\n",
      "The custom pipeline should be selected if no current pipeline addreses the user's query. If the custom pipeline is selected, please do not attempt to design the pipeline now. We will work to design it at a later point.\n",
      "\n",
      "**Output**:\n",
      "Please formate your output as follows:\n",
      "Pipeline Name: <available pipeline name above or \"CUSTOM\">\n",
      "Explination: <reason for selecting the above pipeline>\n",
      "\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 01:40:55,926 - INFO - HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2024-08-03 01:40:55,927 - INFO - Pipeline Name: BOT3\n",
      "Explanation: The user's query specifically requests to load and run the BOT3 pipeline. The available pipeline named BOT3 matches this request directly, indicating that it is the appropriate predesigned pipeline to use.\n",
      "2024-08-03 01:40:55,928 - INFO - selected_pipeline=BOT3\n",
      "2024-08-03 01:40:55,928 - INFO - \n",
      "\n",
      "\n",
      "route\n",
      "\n",
      "\n",
      "\n",
      "2024-08-03 01:40:55,928 - INFO - PLANNER\n",
      "2024-08-03 01:40:55,929 - INFO - \n",
      "\n",
      "PLANNER:\n",
      "\n",
      "{'module': 'PLANNER', 'steps': [{'llm': \"client=<openai.resources.chat.completions.Completions object at 0x147ee873af90> async_client=<openai.resources.chat.completions.AsyncCompletions object at 0x147ef87436d0> model_name='gpt-4o' temperature=0.0 openai_api_key=SecretStr('**********') openai_proxy=''\", 'memory': 'None', 'prompt': 'input_variables=[\\'input\\'] template=\\'**INSTRUCTIONS:**\\\\nYou are planning a bioinformatics analysis pipeline to address a user\\\\\\'s query. You must determine if there already exist a predesigned pipeline that will respond to a user\\\\\\'s query, or if a new, custom process must be designed for this particular query. The multi-step workflow can use the available methods listed below.\\\\n\\\\n**Available Methods within a Pipeline:**\\\\n1. **RAG**: Look up literature and documents from a text database.\\\\n2. **SCRAPE**: Search platforms like arXiv, bioRxiv, and PubMed for the latest research.\\\\n3. **DATABASE**: Utilize bioinformatics databases such as Gene Ontology and Enrichr to perform gene set enrichment analyses.\\\\n4. **CODE**: Execute bioinformatics pipelines. Utilize prebuilt pipelines or develop new ones as needed.\\\\n5. **WRITE**: Synthesize and summarize information. This can include summarizing database searches, code pipeline results, or creating a final report to encapsulate the entire analysis.\\\\n6. **ROUTER**: Determine which step we should proceed to next.\\\\n\\\\nUser Query: {input}\\\\n\\\\n**Available Pipelines**\\\\nName: BOT2\\\\tDescription: this pipeline investigates distances between source, target, and reprogrammed cells. BOT2 should use this pipeline\\\\nName: BOT3\\\\tDescription: this pipeline should be run by BOT3\\\\nName: BOT1\\\\tDescription: this plan is for basic visualization and summary of single cell data from the Geneformer reprogramming experiment. The data can be loaded in from ann data objects\\\\n\\\\n\\\\nCUSTOM: The option to build a new pipeline\\\\n\\\\nThe custom pipeline should be selected if no current pipeline addreses the user\\\\\\'s query. If the custom pipeline is selected, please do not attempt to design the pipeline now. We will work to design it at a later point.\\\\n\\\\n**Output**:\\\\nPlease formate your output as follows:\\\\nPipeline Name: <available pipeline name above or \"CUSTOM\">\\\\nExplination: <reason for selecting the above pipeline>\\\\n\\'', 'input': 'please load and run the BOT3 pipeline', 'output': \"Pipeline Name: BOT3\\nExplanation: The user's query specifically requests to load and run the BOT3 pipeline. The available pipeline named BOT3 matches this request directly, indicating that it is the appropriate predesigned pipeline to use.\", 'parsedOutput': {'selected pipeline': 'BOT3'}, 'purpose': 'determine if a known pipeline can be used or a new one si required'}, {'func': 'planner.planner', 'what': 'loaded an older pipeline'}, {'func': 'planner.planner', 'what': 'set the queue and set the queue pointer to 1'}]}\n",
      "2024-08-03 01:40:55,963 - INFO - current llm calls=1\n",
      "2024-08-03 01:40:55,991 - INFO - CODE\n",
      "2024-08-03 01:40:55,993 - INFO - CODER\n",
      "2024-08-03 01:40:55,995 - INFO - input_variables=['user_query'] template=\"You must select which code to run to help a user.\\n\\n**Available Scripts**\\nScript Name: scanpy_brad, \\t Description: This script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\\nScript Name: models_brad, \\t Description: This script executes a series of custom commands to build a model for predicting the contribution to reprogramming obtained by the inclusion or exclusion of individual TFs using an AnnData object or single cell datasets loaded from a .h5ad file.\\n\\n\\n**User Query**\\n{user_query}\\n\\n**Task**\\nBased on the user's query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\\n\\n**Response Template**\\nSCRIPT: <selected script>\\nREASON: <reasoning why the script fits the user prompt>\\n\"\n",
      "2024-08-03 01:40:55,996 - INFO - FIRST LLM CALL\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "** Step 0**\n",
      "{'order': 0, 'module': 'PLANNER', 'prompt': None, 'description': 'This step designed the plan. It is placed in the queue because we needed a place holder for 0 indexed lists.'}\n",
      "\n",
      "\n",
      "** Step 1**\n",
      "{'order': 1, 'module': 'CODE', 'prompt': \"Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` split the dataframe into several dataframes based on if the value of adata.obs is true or false for the following columns: 'REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B'. Using each adata generated above, compute the cosine similarities of the observations in each of the 10 new AnnData objects relative to the observations in the original adata where 'recipe' == 'target'. Then, please generate a separate visualization for each distribution of cosine similarities. Once you make plots for all 10 Transcription Factors, please train a model to predict the distance based on the inclusion or exclusion of each TF and visualize the model. Do not use any loops or conditional statements. Please write the code to study each transcription factor separately.\", 'description': \" 1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings in `adata.X` and the presence/absence of each TF as features, specifically 'REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B'. Using the extracted features, define the target variable based on the distance to 'target' cells (computed by BOT2) and split the data into training and testing sets. Then, train both a Random Forest model and a Gradient Boosting model on the training data to predict the reprogramming efficiency using the features extracted. Finally, make predictions on the testing data using the better-performing model and create a plot to visualize the predicted reprogramming efficiency.\\n\\nThis revised plan condenses the original steps into a single step that uses the CODE method to perform all the required tasks. \", 'output': []}\n",
      "\n",
      "\n",
      "==================================================\n",
      "==================================================\n",
      "BRAD >> 1: \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 01:40:57,778 - INFO - HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2024-08-03 01:40:57,780 - INFO - SCRIPT: scanpy_brad\n",
      "REASON: The user's query involves loading an AnnData object, manipulating it by splitting into several dataframes based on specific conditions, computing cosine similarities, and generating visualizations for each distribution of cosine similarities. The `scanpy_brad` script is the best match because it allows for the manipulation, visualization, and exploration of single-cell datasets using libraries like scanpy, seaborn, and matplotlib.pyplot. These capabilities are essential for performing the requested operations on the AnnData object and creating the necessary visualizations. The `models_brad` script, on the other hand, is focused on building predictive models and does not provide the necessary tools for data manipulation and visualization as described in the user's query.\n",
      "2024-08-03 01:40:57,780 - INFO - scriptPath=/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts\n",
      "2024-08-03 01:40:57,780 - INFO - ALL SCRIPTS FOUND. BUILDING TEMPLATE\n",
      "2024-08-03 01:40:57,781 - INFO - input_variables=['history', 'input'] template='Current conversation:\\n{history}\\n\\n**PYTHON SCRIPT**\\nYou must run this python script:\\n/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\n\\n**PYTHON SCRIPT DOCUMENTATION**:\\nThis is the doc string of this python script:\\n\"\"\"\\nThis script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\\n\\nArguments (four arguments):\\n    1. output directory: chatstatus[\\'output-directory\\']\\n    2. output file: <name of output file>\\n    3. input file: <file created in previous step>\\n    4. scanpy commands: a list of scanpy commands to be executed on the AnnData object (provided as a single string with commands separated by a delimiter, e.g., \\';\\')\\n\\nBased on the arguments, the input file will be loaded, then your commands will be executed, and finally, the output or resulting ann data object will be saved to the corrrectou output file and directory. Your code is not responsible for loading the .h5ad object, that will happen automatically, and when loaded, the object will be called adata. Your scanpy commands can operate directly on the adata object that will be loaded for you.\\n\\nAdditionally, the following imports are already provided for you and can be used in your code:\\n```\\nimport scanpy as sc\\nimport seaborn as sns\\nimport matplotlib.pyplot as plt\\nimport os\\nimport numpy as np\\nimport pandas as pd\\nfrom scipy.spatial.distance import cdist\\n```\\n\\n**Usage**\\nBRAD Line:\\n```\\nsubprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\'output-directory\\'], <output file>, <input file>, \"<scanpy commands>\"], capture_output=True, text=True)\\n```\\n\\n**Examples**\\nUse the below examples to help generate your code.\\n\\n*Example 1*\\nUser Prompt: Run scanpy preprocessing and UMAP visualization on XXX.h5ad and save the UMAP plot\\nResponse Code:\\n```\\nresponse = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"XXX-modified.h5ad\", \"<path/to/data>/XXX.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save=\\'umap.png\\')\"], capture_output=True, text=True)\\n```\\nExplination: the adata object will be loaded in memory already. The command \"sc.pp.neighbors(adata)\" will preprocess the data, then the command \"sc.tl.umap(adata)\" will perform UMAP, and finally the command \"sc.pl.umap(adata, \\'umap.png\\')\" will save the UMAP to a well named file.\\n\\n*Example 2*\\nUser Prompt: Perform PCA and clustering on the dataset YYY.h5ad and save the PCA plot\\nResponse Code:\\n```\\nresponse = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"YYY-modified.h5ad\", \"<path/to/data>/YYY.h5ad\", \"sc.pp.pca(adata); sc.tl.leiden(adata); sc.pl.pca(adata, save=\\'pca.png\\')\"], capture_output=True, text=True)\\n```\\nExplination: the adata object will be loaded in memory already. The command \"sc.pp.pca(adata)\" will preprocess the data, then the command \"sc.tl.leiden(adata)\" will perform the leiden algorithm, and finally the command \"sc.pl.pca(adata, save=\\'pca.png\\')\" will save the PCA to a well named file.\\n\\n*Example 3*\\nUser Prompt: compute the distance between all cells that are not source or target to all source and target cells.\\nReponse Code:\\n```\\nadata_X = adata.X\\nmask = ~adata.obs[\\'recipe\\'].isin([\\'source\\', \\'target\\'])\\nrecipe_embeddings = adata[mask]\\nsource_embeddings = adata[adata.obs[\\'recipe\\'] == \\'source\\']\\ntarget_embeddings = adata[adata.obs[\\'recipe\\'] == \\'target\\']\\nrecipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X)\\nrecipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X)\\nadata_dist = sc.AnnData(X=np.concatenate((recipe_distances_source, recipe_distances_target), axis=1), var=pd.concat((source_embeddings.obs, target_embeddings.obs)), obs=recipe_embeddings.obs)\\nadata_dist.write(os.path.join(chatstatus[\\'output-directory\\'], <file name>.h5ad))\\n```\\n\\nEach plot should have a title. Also if a legend is used, please place the legend on the right hand side.\\n\\n**OUTPUT FILE NAME INSTRUCTIONS**\\n1. Output path should be chatstatus[\\'output-directory\\']\\n2. Output file name should be `<descriptive name>.h5ad`\\n\"\"\"\\n\\n\\n**CALL PYTHON SCRIPTS FROM PYTHON**:\\nUse the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\\n\\nTo call a Python script `example_script.py` which has no arguments:\\n\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\']], capture_output=True, text=True)\\n```\\n\\nTo call a Python script `example_script.py` with one argument:\\n\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\'], \\'arg1\\'], capture_output=True, text=True)\\n```\\n\\nTo call a Python script `example_script.py` with two arguments:\\n```\\nExecute: subprocess.run([sys.executable, \\'<full path to script/> example_script.py\\', chatstatus[\\'output-directory\\'], \\'arg1\\', \\'arg2\\'], capture_output=True, text=True)\\n```\\n\\nNote that chatstatus[\\'output-directory\\'] is ALWAYS passed as the first argument.\\n\\nThe following files were previously created by BRAD and could be used as input to a function if necessary:\\nlog.json\\n\\nQuery:{input}\\n\\n**PYTHON SCRIPT**\\nPlease use the correct script name and path. You must run this python script:\\n/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\n\\n**INSTRUCTIONS**\\n1. Given the user query and the documentation, identify each of the arguments found in the user\\'s query that should be passed to the Python script.\\n2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\\n3. The last line of your response should say \"Execute: <Python code to execute>\"\\n4. Do not include any extra words or characters. Format the response/output as:\\n    Arguments: \\n    Python Code Explanation: <2 sentences maximum>\\n    Execute: <your code here>\\n\\n**IMPORTANT**\\nThe code to execute from your response must be formatted as:\\n    Execute: subprocess.call([sys.executable, \\'<path to python script>\\', \\'<argument 1>\\', \\'<argument 2>\\', ..., \\'<argument n>\\'], capture_output=True, text=True))\\nThis output should be exactly one line and no longer. Stop the response after this line.\\n'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation:\n",
      "\n",
      "\n",
      "**PYTHON SCRIPT**\n",
      "You must run this python script:\n",
      "/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\n",
      "\n",
      "**PYTHON SCRIPT DOCUMENTATION**:\n",
      "This is the doc string of this python script:\n",
      "\"\"\"\n",
      "This script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\n",
      "\n",
      "Arguments (four arguments):\n",
      "    1. output directory: chatstatus['output-directory']\n",
      "    2. output file: <name of output file>\n",
      "    3. input file: <file created in previous step>\n",
      "    4. scanpy commands: a list of scanpy commands to be executed on the AnnData object (provided as a single string with commands separated by a delimiter, e.g., ';')\n",
      "\n",
      "Based on the arguments, the input file will be loaded, then your commands will be executed, and finally, the output or resulting ann data object will be saved to the corrrectou output file and directory. Your code is not responsible for loading the .h5ad object, that will happen automatically, and when loaded, the object will be called adata. Your scanpy commands can operate directly on the adata object that will be loaded for you.\n",
      "\n",
      "Additionally, the following imports are already provided for you and can be used in your code:\n",
      "```\n",
      "import scanpy as sc\n",
      "import seaborn as sns\n",
      "import matplotlib.pyplot as plt\n",
      "import os\n",
      "import numpy as np\n",
      "import pandas as pd\n",
      "from scipy.spatial.distance import cdist\n",
      "```\n",
      "\n",
      "**Usage**\n",
      "BRAD Line:\n",
      "```\n",
      "subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus['output-directory'], <output file>, <input file>, \"<scanpy commands>\"], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "**Examples**\n",
      "Use the below examples to help generate your code.\n",
      "\n",
      "*Example 1*\n",
      "User Prompt: Run scanpy preprocessing and UMAP visualization on XXX.h5ad and save the UMAP plot\n",
      "Response Code:\n",
      "```\n",
      "response = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus['output-directory'], \"XXX-modified.h5ad\", \"<path/to/data>/XXX.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save='umap.png')\"], capture_output=True, text=True)\n",
      "```\n",
      "Explination: the adata object will be loaded in memory already. The command \"sc.pp.neighbors(adata)\" will preprocess the data, then the command \"sc.tl.umap(adata)\" will perform UMAP, and finally the command \"sc.pl.umap(adata, 'umap.png')\" will save the UMAP to a well named file.\n",
      "\n",
      "*Example 2*\n",
      "User Prompt: Perform PCA and clustering on the dataset YYY.h5ad and save the PCA plot\n",
      "Response Code:\n",
      "```\n",
      "response = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus['output-directory'], \"YYY-modified.h5ad\", \"<path/to/data>/YYY.h5ad\", \"sc.pp.pca(adata); sc.tl.leiden(adata); sc.pl.pca(adata, save='pca.png')\"], capture_output=True, text=True)\n",
      "```\n",
      "Explination: the adata object will be loaded in memory already. The command \"sc.pp.pca(adata)\" will preprocess the data, then the command \"sc.tl.leiden(adata)\" will perform the leiden algorithm, and finally the command \"sc.pl.pca(adata, save='pca.png')\" will save the PCA to a well named file.\n",
      "\n",
      "*Example 3*\n",
      "User Prompt: compute the distance between all cells that are not source or target to all source and target cells.\n",
      "Reponse Code:\n",
      "```\n",
      "adata_X = adata.X\n",
      "mask = ~adata.obs['recipe'].isin(['source', 'target'])\n",
      "recipe_embeddings = adata[mask]\n",
      "source_embeddings = adata[adata.obs['recipe'] == 'source']\n",
      "target_embeddings = adata[adata.obs['recipe'] == 'target']\n",
      "recipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X)\n",
      "recipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X)\n",
      "adata_dist = sc.AnnData(X=np.concatenate((recipe_distances_source, recipe_distances_target), axis=1), var=pd.concat((source_embeddings.obs, target_embeddings.obs)), obs=recipe_embeddings.obs)\n",
      "adata_dist.write(os.path.join(chatstatus['output-directory'], <file name>.h5ad))\n",
      "```\n",
      "\n",
      "Each plot should have a title. Also if a legend is used, please place the legend on the right hand side.\n",
      "\n",
      "**OUTPUT FILE NAME INSTRUCTIONS**\n",
      "1. Output path should be chatstatus['output-directory']\n",
      "2. Output file name should be `<descriptive name>.h5ad`\n",
      "\"\"\"\n",
      "\n",
      "\n",
      "**CALL PYTHON SCRIPTS FROM PYTHON**:\n",
      "Use the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\n",
      "\n",
      "To call a Python script `example_script.py` which has no arguments:\n",
      "\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory']], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with one argument:\n",
      "\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory'], 'arg1'], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "To call a Python script `example_script.py` with two arguments:\n",
      "```\n",
      "Execute: subprocess.run([sys.executable, '<full path to script/> example_script.py', chatstatus['output-directory'], 'arg1', 'arg2'], capture_output=True, text=True)\n",
      "```\n",
      "\n",
      "Note that chatstatus['output-directory'] is ALWAYS passed as the first argument.\n",
      "\n",
      "The following files were previously created by BRAD and could be used as input to a function if necessary:\n",
      "log.json\n",
      "\n",
      "Query:Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` split the dataframe into several dataframes based on if the value of adata.obs is true or false for the following columns: 'REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B'. Using each adata generated above, compute the cosine similarities of the observations in each of the 10 new AnnData objects relative to the observations in the original adata where 'recipe' == 'target'. Then, please generate a separate visualization for each distribution of cosine similarities. Once you make plots for all 10 Transcription Factors, please train a model to predict the distance based on the inclusion or exclusion of each TF and visualize the model. Do not use any loops or conditional statements. Please write the code to study each transcription factor separately.\n",
      "\n",
      "**PYTHON SCRIPT**\n",
      "Please use the correct script name and path. You must run this python script:\n",
      "/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\n",
      "\n",
      "**INSTRUCTIONS**\n",
      "1. Given the user query and the documentation, identify each of the arguments found in the user's query that should be passed to the Python script.\n",
      "2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\n",
      "3. The last line of your response should say \"Execute: <Python code to execute>\"\n",
      "4. Do not include any extra words or characters. Format the response/output as:\n",
      "    Arguments: \n",
      "    Python Code Explanation: <2 sentences maximum>\n",
      "    Execute: <your code here>\n",
      "\n",
      "**IMPORTANT**\n",
      "The code to execute from your response must be formatted as:\n",
      "    Execute: subprocess.call([sys.executable, '<path to python script>', '<argument 1>', '<argument 2>', ..., '<argument n>'], capture_output=True, text=True))\n",
      "This output should be exactly one line and no longer. Stop the response after this line.\n",
      "\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 01:41:17,664 - INFO - HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2024-08-03 01:41:17,666 - INFO - LLM OUTPUT PARSER\n",
      "2024-08-03 01:41:17,666 - INFO - Arguments:\n",
      "1. output directory: chatstatus['output-directory']\n",
      "2. output file: \"HSC-reprogrammed-DE-modified.h5ad\"\n",
      "3. input file: \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\"\n",
      "4. scanpy commands: \"REL_adata = adata[adata.obs['REL']]; STAT5A_adata = adata[adata.obs['STAT5A']]; FOS_adata = adata[adata.obs['FOS']]; ETV6_adata = adata[adata.obs['ETV6']]; MEF2C_adata = adata[adata.obs['MEF2C']]; RUNX3_adata = adata[adata.obs['RUNX3']]; FOSB_adata = adata[adata.obs['FOSB']]; GATA2_adata = adata[adata.obs['GATA2']]; IKZF1_adata = adata[adata.obs['IKZF1']]; GFI1B_adata = adata[adata.obs['GFI1B']]; target_embeddings = adata[adata.obs['recipe'] == 'target']; REL_cosine_sim = 1 - cdist(REL_adata.X, target_embeddings.X, metric='cosine'); STAT5A_cosine_sim = 1 - cdist(STAT5A_adata.X, target_embeddings.X, metric='cosine'); FOS_cosine_sim = 1 - cdist(FOS_adata.X, target_embeddings.X, metric='cosine'); ETV6_cosine_sim = 1 - cdist(ETV6_adata.X, target_embeddings.X, metric='cosine'); MEF2C_cosine_sim = 1 - cdist(MEF2C_adata.X, target_embeddings.X, metric='cosine'); RUNX3_cosine_sim = 1 - cdist(RUNX3_adata.X, target_embeddings.X, metric='cosine'); FOSB_cosine_sim = 1 - cdist(FOSB_adata.X, target_embeddings.X, metric='cosine'); GATA2_cosine_sim = 1 - cdist(GATA2_adata.X, target_embeddings.X, metric='cosine'); IKZF1_cosine_sim = 1 - cdist(IKZF1_adata.X, target_embeddings.X, metric='cosine'); GFI1B_cosine_sim = 1 - cdist(GFI1B_adata.X, target_embeddings.X, metric='cosine'); sns.histplot(REL_cosine_sim.flatten()).set(title='REL Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'REL_cosine_similarity.png')); sns.histplot(STAT5A_cosine_sim.flatten()).set(title='STAT5A Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'STAT5A_cosine_similarity.png')); sns.histplot(FOS_cosine_sim.flatten()).set(title='FOS Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'FOS_cosine_similarity.png')); sns.histplot(ETV6_cosine_sim.flatten()).set(title='ETV6 Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'ETV6_cosine_similarity.png')); sns.histplot(MEF2C_cosine_sim.flatten()).set(title='MEF2C Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'MEF2C_cosine_similarity.png')); sns.histplot(RUNX3_cosine_sim.flatten()).set(title='RUNX3 Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'RUNX3_cosine_similarity.png')); sns.histplot(FOSB_cosine_sim.flatten()).set(title='FOSB Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'FOSB_cosine_similarity.png')); sns.histplot(GATA2_cosine_sim.flatten()).set(title='GATA2 Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'GATA2_cosine_similarity.png')); sns.histplot(IKZF1_cosine_sim.flatten()).set(title='IKZF1 Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'IKZF1_cosine_similarity.png')); sns.histplot(GFI1B_cosine_sim.flatten()).set(title='GFI1B Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'GFI1B_cosine_similarity.png'))\"\n",
      "\n",
      "Python Code Explanation: The code loads the AnnData object, splits it based on the specified columns, computes cosine similarities, and generates visualizations for each transcription factor.\n",
      "\n",
      "Execute: subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-DE-modified.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"REL_adata = adata[adata.obs['REL']]; STAT5A_adata = adata[adata.obs['STAT5A']]; FOS_adata = adata[adata.obs['FOS']]; ETV6_adata = adata[adata.obs['ETV6']]; MEF2C_adata = adata[adata.obs['MEF2C']]; RUNX3_adata = adata[adata.obs['RUNX3']]; FOSB_adata = adata[adata.obs['FOSB']]; GATA2_adata = adata[adata.obs['GATA2']]; IKZF1_adata = adata[adata.obs['IKZF1']]; GFI1B_adata = adata[adata.obs['GFI1B']]; target_embeddings = adata[adata.obs['recipe'] == 'target']; REL_cosine_sim = 1 - cdist(REL_adata.X, target_embeddings.X, metric='cosine'); STAT5A_cosine_sim = 1 - cdist(STAT5A_adata.X, target_embeddings.X, metric='cosine'); FOS_cosine_sim = 1 - cdist(FOS_adata.X, target_embeddings.X, metric='cosine'); ETV6_cosine_sim = 1 - cdist(ETV6_adata.X, target_embeddings.X, metric='cosine'); MEF2C_cosine_sim = 1 - cdist(MEF2C_adata.X, target_embeddings.X, metric='cosine'); RUNX3_cosine_sim = 1 - cdist(RUNX3_adata.X, target_embeddings.X, metric='cosine'); FOSB_cosine_sim = 1 - cdist(FOSB_adata.X, target_embeddings.X, metric='cosine'); GATA2_cosine_sim = 1 - cdist(GATA2_adata.X, target_embeddings.X, metric='cosine'); IKZF1_cosine_sim = 1 - cdist(IKZF1_adata.X, target_embeddings.X, metric='cosine'); GFI1B_cosine_sim = 1 - cdist(GFI1B_adata.X, target_embeddings.X, metric='cosine'); sns.histplot(REL_cosine_sim.flatten()).set(title='REL Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'REL_cosine_similarity.png')); sns.histplot(STAT5A_cosine_sim.flatten()).set(title='STAT5A Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'STAT5A_cosine_similarity.png')); sns.histplot(FOS_cosine_sim.flatten()).set(title='FOS Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'FOS_cosine_similarity.png')); sns.histplot(ETV6_cosine_sim.flatten()).set(title='ETV6 Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'ETV6_cosine_similarity.png')); sns.histplot(MEF2C_cosine_sim.flatten()).set(title='MEF2C Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'MEF2C_cosine_similarity.png')); sns.histplot(RUNX3_cosine_sim.flatten()).set(title='RUNX3 Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'RUNX3_cosine_similarity.png')); sns.histplot(FOSB_cosine_sim.flatten()).set(title='FOSB Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'FOSB_cosine_similarity.png')); sns.histplot(GATA2_cosine_sim.flatten()).set(title='GATA2 Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'GATA2_cosine_similarity.png')); sns.histplot(IKZF1_cosine_sim.flatten()).set(title='IKZF1 Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'IKZF1_cosine_similarity.png')); sns.histplot(GFI1B_cosine_sim.flatten()).set(title='GFI1B Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'GFI1B_cosine_similarity.png'))\"], capture_output=True, text=True)\n",
      "2024-08-03 01:41:17,667 - INFO - ['Arguments:\\n1. output directory: chatstatus[\\'output-directory\\']\\n2. output file: \"HSC-reprogrammed-DE-modified.h5ad\"\\n3. input file: \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\"\\n4. scanpy commands: \"REL_adata = adata[adata.obs[\\'REL\\']]; STAT5A_adata = adata[adata.obs[\\'STAT5A\\']]; FOS_adata = adata[adata.obs[\\'FOS\\']]; ETV6_adata = adata[adata.obs[\\'ETV6\\']]; MEF2C_adata = adata[adata.obs[\\'MEF2C\\']]; RUNX3_adata = adata[adata.obs[\\'RUNX3\\']]; FOSB_adata = adata[adata.obs[\\'FOSB\\']]; GATA2_adata = adata[adata.obs[\\'GATA2\\']]; IKZF1_adata = adata[adata.obs[\\'IKZF1\\']]; GFI1B_adata = adata[adata.obs[\\'GFI1B\\']]; target_embeddings = adata[adata.obs[\\'recipe\\'] == \\'target\\']; REL_cosine_sim = 1 - cdist(REL_adata.X, target_embeddings.X, metric=\\'cosine\\'); STAT5A_cosine_sim = 1 - cdist(STAT5A_adata.X, target_embeddings.X, metric=\\'cosine\\'); FOS_cosine_sim = 1 - cdist(FOS_adata.X, target_embeddings.X, metric=\\'cosine\\'); ETV6_cosine_sim = 1 - cdist(ETV6_adata.X, target_embeddings.X, metric=\\'cosine\\'); MEF2C_cosine_sim = 1 - cdist(MEF2C_adata.X, target_embeddings.X, metric=\\'cosine\\'); RUNX3_cosine_sim = 1 - cdist(RUNX3_adata.X, target_embeddings.X, metric=\\'cosine\\'); FOSB_cosine_sim = 1 - cdist(FOSB_adata.X, target_embeddings.X, metric=\\'cosine\\'); GATA2_cosine_sim = 1 - cdist(GATA2_adata.X, target_embeddings.X, metric=\\'cosine\\'); IKZF1_cosine_sim = 1 - cdist(IKZF1_adata.X, target_embeddings.X, metric=\\'cosine\\'); GFI1B_cosine_sim = 1 - cdist(GFI1B_adata.X, target_embeddings.X, metric=\\'cosine\\'); sns.histplot(REL_cosine_sim.flatten()).set(title=\\'REL Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'REL_cosine_similarity.png\\')); sns.histplot(STAT5A_cosine_sim.flatten()).set(title=\\'STAT5A Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'STAT5A_cosine_similarity.png\\')); sns.histplot(FOS_cosine_sim.flatten()).set(title=\\'FOS Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'FOS_cosine_similarity.png\\')); sns.histplot(ETV6_cosine_sim.flatten()).set(title=\\'ETV6 Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'ETV6_cosine_similarity.png\\')); sns.histplot(MEF2C_cosine_sim.flatten()).set(title=\\'MEF2C Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'MEF2C_cosine_similarity.png\\')); sns.histplot(RUNX3_cosine_sim.flatten()).set(title=\\'RUNX3 Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'RUNX3_cosine_similarity.png\\')); sns.histplot(FOSB_cosine_sim.flatten()).set(title=\\'FOSB Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'FOSB_cosine_similarity.png\\')); sns.histplot(GATA2_cosine_sim.flatten()).set(title=\\'GATA2 Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'GATA2_cosine_similarity.png\\')); sns.histplot(IKZF1_cosine_sim.flatten()).set(title=\\'IKZF1 Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'IKZF1_cosine_similarity.png\\')); sns.histplot(GFI1B_cosine_sim.flatten()).set(title=\\'GFI1B Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'GFI1B_cosine_similarity.png\\'))\"\\n\\nPython Code Explanation: The code loads the AnnData object, splits it based on the specified columns, computes cosine similarities, and generates visualizations for each transcription factor.\\n\\n', ' subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"HSC-reprogrammed-DE-modified.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"REL_adata = adata[adata.obs[\\'REL\\']]; STAT5A_adata = adata[adata.obs[\\'STAT5A\\']]; FOS_adata = adata[adata.obs[\\'FOS\\']]; ETV6_adata = adata[adata.obs[\\'ETV6\\']]; MEF2C_adata = adata[adata.obs[\\'MEF2C\\']]; RUNX3_adata = adata[adata.obs[\\'RUNX3\\']]; FOSB_adata = adata[adata.obs[\\'FOSB\\']]; GATA2_adata = adata[adata.obs[\\'GATA2\\']]; IKZF1_adata = adata[adata.obs[\\'IKZF1\\']]; GFI1B_adata = adata[adata.obs[\\'GFI1B\\']]; target_embeddings = adata[adata.obs[\\'recipe\\'] == \\'target\\']; REL_cosine_sim = 1 - cdist(REL_adata.X, target_embeddings.X, metric=\\'cosine\\'); STAT5A_cosine_sim = 1 - cdist(STAT5A_adata.X, target_embeddings.X, metric=\\'cosine\\'); FOS_cosine_sim = 1 - cdist(FOS_adata.X, target_embeddings.X, metric=\\'cosine\\'); ETV6_cosine_sim = 1 - cdist(ETV6_adata.X, target_embeddings.X, metric=\\'cosine\\'); MEF2C_cosine_sim = 1 - cdist(MEF2C_adata.X, target_embeddings.X, metric=\\'cosine\\'); RUNX3_cosine_sim = 1 - cdist(RUNX3_adata.X, target_embeddings.X, metric=\\'cosine\\'); FOSB_cosine_sim = 1 - cdist(FOSB_adata.X, target_embeddings.X, metric=\\'cosine\\'); GATA2_cosine_sim = 1 - cdist(GATA2_adata.X, target_embeddings.X, metric=\\'cosine\\'); IKZF1_cosine_sim = 1 - cdist(IKZF1_adata.X, target_embeddings.X, metric=\\'cosine\\'); GFI1B_cosine_sim = 1 - cdist(GFI1B_adata.X, target_embeddings.X, metric=\\'cosine\\'); sns.histplot(REL_cosine_sim.flatten()).set(title=\\'REL Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'REL_cosine_similarity.png\\')); sns.histplot(STAT5A_cosine_sim.flatten()).set(title=\\'STAT5A Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'STAT5A_cosine_similarity.png\\')); sns.histplot(FOS_cosine_sim.flatten()).set(title=\\'FOS Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'FOS_cosine_similarity.png\\')); sns.histplot(ETV6_cosine_sim.flatten()).set(title=\\'ETV6 Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'ETV6_cosine_similarity.png\\')); sns.histplot(MEF2C_cosine_sim.flatten()).set(title=\\'MEF2C Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'MEF2C_cosine_similarity.png\\')); sns.histplot(RUNX3_cosine_sim.flatten()).set(title=\\'RUNX3 Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'RUNX3_cosine_similarity.png\\')); sns.histplot(FOSB_cosine_sim.flatten()).set(title=\\'FOSB Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'FOSB_cosine_similarity.png\\')); sns.histplot(GATA2_cosine_sim.flatten()).set(title=\\'GATA2 Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'GATA2_cosine_similarity.png\\')); sns.histplot(IKZF1_cosine_sim.flatten()).set(title=\\'IKZF1 Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'IKZF1_cosine_similarity.png\\')); sns.histplot(GFI1B_cosine_sim.flatten()).set(title=\\'GFI1B Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'GFI1B_cosine_similarity.png\\'))\"], capture_output=True, text=True)']\n",
      "2024-08-03 01:41:17,667 - INFO - Stripped funciton call\n",
      "2024-08-03 01:41:17,667 - INFO - subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-DE-modified.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"REL_adata = adata[adata.obs['REL']]; STAT5A_adata = adata[adata.obs['STAT5A']]; FOS_adata = adata[adata.obs['FOS']]; ETV6_adata = adata[adata.obs['ETV6']]; MEF2C_adata = adata[adata.obs['MEF2C']]; RUNX3_adata = adata[adata.obs['RUNX3']]; FOSB_adata = adata[adata.obs['FOSB']]; GATA2_adata = adata[adata.obs['GATA2']]; IKZF1_adata = adata[adata.obs['IKZF1']]; GFI1B_adata = adata[adata.obs['GFI1B']]; target_embeddings = adata[adata.obs['recipe'] == 'target']; REL_cosine_sim = 1 - cdist(REL_adata.X, target_embeddings.X, metric='cosine'); STAT5A_cosine_sim = 1 - cdist(STAT5A_adata.X, target_embeddings.X, metric='cosine'); FOS_cosine_sim = 1 - cdist(FOS_adata.X, target_embeddings.X, metric='cosine'); ETV6_cosine_sim = 1 - cdist(ETV6_adata.X, target_embeddings.X, metric='cosine'); MEF2C_cosine_sim = 1 - cdist(MEF2C_adata.X, target_embeddings.X, metric='cosine'); RUNX3_cosine_sim = 1 - cdist(RUNX3_adata.X, target_embeddings.X, metric='cosine'); FOSB_cosine_sim = 1 - cdist(FOSB_adata.X, target_embeddings.X, metric='cosine'); GATA2_cosine_sim = 1 - cdist(GATA2_adata.X, target_embeddings.X, metric='cosine'); IKZF1_cosine_sim = 1 - cdist(IKZF1_adata.X, target_embeddings.X, metric='cosine'); GFI1B_cosine_sim = 1 - cdist(GFI1B_adata.X, target_embeddings.X, metric='cosine'); sns.histplot(REL_cosine_sim.flatten()).set(title='REL Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'REL_cosine_similarity.png')); sns.histplot(STAT5A_cosine_sim.flatten()).set(title='STAT5A Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'STAT5A_cosine_similarity.png')); sns.histplot(FOS_cosine_sim.flatten()).set(title='FOS Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'FOS_cosine_similarity.png')); sns.histplot(ETV6_cosine_sim.flatten()).set(title='ETV6 Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'ETV6_cosine_similarity.png')); sns.histplot(MEF2C_cosine_sim.flatten()).set(title='MEF2C Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'MEF2C_cosine_similarity.png')); sns.histplot(RUNX3_cosine_sim.flatten()).set(title='RUNX3 Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'RUNX3_cosine_similarity.png')); sns.histplot(FOSB_cosine_sim.flatten()).set(title='FOSB Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'FOSB_cosine_similarity.png')); sns.histplot(GATA2_cosine_sim.flatten()).set(title='GATA2 Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'GATA2_cosine_similarity.png')); sns.histplot(IKZF1_cosine_sim.flatten()).set(title='IKZF1 Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'IKZF1_cosine_similarity.png')); sns.histplot(GFI1B_cosine_sim.flatten()).set(title='GFI1B Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'GFI1B_cosine_similarity.png'))\"], capture_output=True, text=True)\n",
      "2024-08-03 01:41:17,667 - INFO - Cleaned up function call:\n",
      "\n",
      "2024-08-03 01:41:17,668 - INFO - subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-DE-modified.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"REL_adata = adata[adata.obs['REL']]; STAT5A_adata = adata[adata.obs['STAT5A']]; FOS_adata = adata[adata.obs['FOS']]; ETV6_adata = adata[adata.obs['ETV6']]; MEF2C_adata = adata[adata.obs['MEF2C']]; RUNX3_adata = adata[adata.obs['RUNX3']]; FOSB_adata = adata[adata.obs['FOSB']]; GATA2_adata = adata[adata.obs['GATA2']]; IKZF1_adata = adata[adata.obs['IKZF1']]; GFI1B_adata = adata[adata.obs['GFI1B']]; target_embeddings = adata[adata.obs['recipe'] == 'target']; REL_cosine_sim = 1 - cdist(REL_adata.X, target_embeddings.X, metric='cosine'); STAT5A_cosine_sim = 1 - cdist(STAT5A_adata.X, target_embeddings.X, metric='cosine'); FOS_cosine_sim = 1 - cdist(FOS_adata.X, target_embeddings.X, metric='cosine'); ETV6_cosine_sim = 1 - cdist(ETV6_adata.X, target_embeddings.X, metric='cosine'); MEF2C_cosine_sim = 1 - cdist(MEF2C_adata.X, target_embeddings.X, metric='cosine'); RUNX3_cosine_sim = 1 - cdist(RUNX3_adata.X, target_embeddings.X, metric='cosine'); FOSB_cosine_sim = 1 - cdist(FOSB_adata.X, target_embeddings.X, metric='cosine'); GATA2_cosine_sim = 1 - cdist(GATA2_adata.X, target_embeddings.X, metric='cosine'); IKZF1_cosine_sim = 1 - cdist(IKZF1_adata.X, target_embeddings.X, metric='cosine'); GFI1B_cosine_sim = 1 - cdist(GFI1B_adata.X, target_embeddings.X, metric='cosine'); sns.histplot(REL_cosine_sim.flatten()).set(title='REL Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'REL_cosine_similarity.png')); sns.histplot(STAT5A_cosine_sim.flatten()).set(title='STAT5A Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'STAT5A_cosine_similarity.png')); sns.histplot(FOS_cosine_sim.flatten()).set(title='FOS Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'FOS_cosine_similarity.png')); sns.histplot(ETV6_cosine_sim.flatten()).set(title='ETV6 Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'ETV6_cosine_similarity.png')); sns.histplot(MEF2C_cosine_sim.flatten()).set(title='MEF2C Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'MEF2C_cosine_similarity.png')); sns.histplot(RUNX3_cosine_sim.flatten()).set(title='RUNX3 Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'RUNX3_cosine_similarity.png')); sns.histplot(FOSB_cosine_sim.flatten()).set(title='FOSB Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'FOSB_cosine_similarity.png')); sns.histplot(GATA2_cosine_sim.flatten()).set(title='GATA2 Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'GATA2_cosine_similarity.png')); sns.histplot(IKZF1_cosine_sim.flatten()).set(title='IKZF1 Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'IKZF1_cosine_similarity.png')); sns.histplot(GFI1B_cosine_sim.flatten()).set(title='GFI1B Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'GFI1B_cosine_similarity.png'))\"], capture_output=True, text=True)\n",
      "2024-08-03 01:41:17,668 - INFO - \n",
      "\n",
      "2024-08-03 01:41:17,669 - INFO - EVAL\n",
      "2024-08-03 01:41:17,669 - INFO - subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-DE-modified.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"REL_adata = adata[adata.obs['REL']]; STAT5A_adata = adata[adata.obs['STAT5A']]; FOS_adata = adata[adata.obs['FOS']]; ETV6_adata = adata[adata.obs['ETV6']]; MEF2C_adata = adata[adata.obs['MEF2C']]; RUNX3_adata = adata[adata.obs['RUNX3']]; FOSB_adata = adata[adata.obs['FOSB']]; GATA2_adata = adata[adata.obs['GATA2']]; IKZF1_adata = adata[adata.obs['IKZF1']]; GFI1B_adata = adata[adata.obs['GFI1B']]; target_embeddings = adata[adata.obs['recipe'] == 'target']; REL_cosine_sim = 1 - cdist(REL_adata.X, target_embeddings.X, metric='cosine'); STAT5A_cosine_sim = 1 - cdist(STAT5A_adata.X, target_embeddings.X, metric='cosine'); FOS_cosine_sim = 1 - cdist(FOS_adata.X, target_embeddings.X, metric='cosine'); ETV6_cosine_sim = 1 - cdist(ETV6_adata.X, target_embeddings.X, metric='cosine'); MEF2C_cosine_sim = 1 - cdist(MEF2C_adata.X, target_embeddings.X, metric='cosine'); RUNX3_cosine_sim = 1 - cdist(RUNX3_adata.X, target_embeddings.X, metric='cosine'); FOSB_cosine_sim = 1 - cdist(FOSB_adata.X, target_embeddings.X, metric='cosine'); GATA2_cosine_sim = 1 - cdist(GATA2_adata.X, target_embeddings.X, metric='cosine'); IKZF1_cosine_sim = 1 - cdist(IKZF1_adata.X, target_embeddings.X, metric='cosine'); GFI1B_cosine_sim = 1 - cdist(GFI1B_adata.X, target_embeddings.X, metric='cosine'); sns.histplot(REL_cosine_sim.flatten()).set(title='REL Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'REL_cosine_similarity.png')); sns.histplot(STAT5A_cosine_sim.flatten()).set(title='STAT5A Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'STAT5A_cosine_similarity.png')); sns.histplot(FOS_cosine_sim.flatten()).set(title='FOS Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'FOS_cosine_similarity.png')); sns.histplot(ETV6_cosine_sim.flatten()).set(title='ETV6 Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'ETV6_cosine_similarity.png')); sns.histplot(MEF2C_cosine_sim.flatten()).set(title='MEF2C Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'MEF2C_cosine_similarity.png')); sns.histplot(RUNX3_cosine_sim.flatten()).set(title='RUNX3 Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'RUNX3_cosine_similarity.png')); sns.histplot(FOSB_cosine_sim.flatten()).set(title='FOSB Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'FOSB_cosine_similarity.png')); sns.histplot(GATA2_cosine_sim.flatten()).set(title='GATA2 Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'GATA2_cosine_similarity.png')); sns.histplot(IKZF1_cosine_sim.flatten()).set(title='IKZF1 Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'IKZF1_cosine_similarity.png')); sns.histplot(GFI1B_cosine_sim.flatten()).set(title='GFI1B Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'GFI1B_cosine_similarity.png'))\"], capture_output=True, text=True)\n",
      "2024-08-03 01:41:17,670 - INFO - END EVAL CHECK\n",
      "2024-08-03 01:41:17,670 - INFO - Finalized PYTHON Call:\n",
      "2024-08-03 01:41:17,670 - INFO -    \n",
      "2024-08-03 01:41:17,670 - INFO - response = subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus['output-directory'], \"HSC-reprogrammed-DE-modified.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"REL_adata = adata[adata.obs['REL']]; STAT5A_adata = adata[adata.obs['STAT5A']]; FOS_adata = adata[adata.obs['FOS']]; ETV6_adata = adata[adata.obs['ETV6']]; MEF2C_adata = adata[adata.obs['MEF2C']]; RUNX3_adata = adata[adata.obs['RUNX3']]; FOSB_adata = adata[adata.obs['FOSB']]; GATA2_adata = adata[adata.obs['GATA2']]; IKZF1_adata = adata[adata.obs['IKZF1']]; GFI1B_adata = adata[adata.obs['GFI1B']]; target_embeddings = adata[adata.obs['recipe'] == 'target']; REL_cosine_sim = 1 - cdist(REL_adata.X, target_embeddings.X, metric='cosine'); STAT5A_cosine_sim = 1 - cdist(STAT5A_adata.X, target_embeddings.X, metric='cosine'); FOS_cosine_sim = 1 - cdist(FOS_adata.X, target_embeddings.X, metric='cosine'); ETV6_cosine_sim = 1 - cdist(ETV6_adata.X, target_embeddings.X, metric='cosine'); MEF2C_cosine_sim = 1 - cdist(MEF2C_adata.X, target_embeddings.X, metric='cosine'); RUNX3_cosine_sim = 1 - cdist(RUNX3_adata.X, target_embeddings.X, metric='cosine'); FOSB_cosine_sim = 1 - cdist(FOSB_adata.X, target_embeddings.X, metric='cosine'); GATA2_cosine_sim = 1 - cdist(GATA2_adata.X, target_embeddings.X, metric='cosine'); IKZF1_cosine_sim = 1 - cdist(IKZF1_adata.X, target_embeddings.X, metric='cosine'); GFI1B_cosine_sim = 1 - cdist(GFI1B_adata.X, target_embeddings.X, metric='cosine'); sns.histplot(REL_cosine_sim.flatten()).set(title='REL Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'REL_cosine_similarity.png')); sns.histplot(STAT5A_cosine_sim.flatten()).set(title='STAT5A Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'STAT5A_cosine_similarity.png')); sns.histplot(FOS_cosine_sim.flatten()).set(title='FOS Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'FOS_cosine_similarity.png')); sns.histplot(ETV6_cosine_sim.flatten()).set(title='ETV6 Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'ETV6_cosine_similarity.png')); sns.histplot(MEF2C_cosine_sim.flatten()).set(title='MEF2C Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'MEF2C_cosine_similarity.png')); sns.histplot(RUNX3_cosine_sim.flatten()).set(title='RUNX3 Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'RUNX3_cosine_similarity.png')); sns.histplot(FOSB_cosine_sim.flatten()).set(title='FOSB Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'FOSB_cosine_similarity.png')); sns.histplot(GATA2_cosine_sim.flatten()).set(title='GATA2 Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'GATA2_cosine_similarity.png')); sns.histplot(IKZF1_cosine_sim.flatten()).set(title='IKZF1 Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'IKZF1_cosine_similarity.png')); sns.histplot(GFI1B_cosine_sim.flatten()).set(title='GFI1B Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'GFI1B_cosine_similarity.png'))\"], capture_output=True, text=True)\n",
      "2024-08-03 01:41:17,671 - INFO -    \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-03 01:41:40,310 - INFO - Debug: PYTHON code executed successfully.\n",
      "2024-08-03 01:41:40,311 - INFO - Code Execution Output:\n",
      "2024-08-03 01:41:40,311 - INFO - CompletedProcess(args=['/sw/pkgs/arc/python3.11-anaconda/2024.02-1/bin/python', '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py', '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-40-37', 'HSC-reprogrammed-DE-modified.h5ad', '/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad', \"REL_adata = adata[adata.obs['REL']]; STAT5A_adata = adata[adata.obs['STAT5A']]; FOS_adata = adata[adata.obs['FOS']]; ETV6_adata = adata[adata.obs['ETV6']]; MEF2C_adata = adata[adata.obs['MEF2C']]; RUNX3_adata = adata[adata.obs['RUNX3']]; FOSB_adata = adata[adata.obs['FOSB']]; GATA2_adata = adata[adata.obs['GATA2']]; IKZF1_adata = adata[adata.obs['IKZF1']]; GFI1B_adata = adata[adata.obs['GFI1B']]; target_embeddings = adata[adata.obs['recipe'] == 'target']; REL_cosine_sim = 1 - cdist(REL_adata.X, target_embeddings.X, metric='cosine'); STAT5A_cosine_sim = 1 - cdist(STAT5A_adata.X, target_embeddings.X, metric='cosine'); FOS_cosine_sim = 1 - cdist(FOS_adata.X, target_embeddings.X, metric='cosine'); ETV6_cosine_sim = 1 - cdist(ETV6_adata.X, target_embeddings.X, metric='cosine'); MEF2C_cosine_sim = 1 - cdist(MEF2C_adata.X, target_embeddings.X, metric='cosine'); RUNX3_cosine_sim = 1 - cdist(RUNX3_adata.X, target_embeddings.X, metric='cosine'); FOSB_cosine_sim = 1 - cdist(FOSB_adata.X, target_embeddings.X, metric='cosine'); GATA2_cosine_sim = 1 - cdist(GATA2_adata.X, target_embeddings.X, metric='cosine'); IKZF1_cosine_sim = 1 - cdist(IKZF1_adata.X, target_embeddings.X, metric='cosine'); GFI1B_cosine_sim = 1 - cdist(GFI1B_adata.X, target_embeddings.X, metric='cosine'); sns.histplot(REL_cosine_sim.flatten()).set(title='REL Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'REL_cosine_similarity.png')); sns.histplot(STAT5A_cosine_sim.flatten()).set(title='STAT5A Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'STAT5A_cosine_similarity.png')); sns.histplot(FOS_cosine_sim.flatten()).set(title='FOS Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'FOS_cosine_similarity.png')); sns.histplot(ETV6_cosine_sim.flatten()).set(title='ETV6 Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'ETV6_cosine_similarity.png')); sns.histplot(MEF2C_cosine_sim.flatten()).set(title='MEF2C Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'MEF2C_cosine_similarity.png')); sns.histplot(RUNX3_cosine_sim.flatten()).set(title='RUNX3 Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'RUNX3_cosine_similarity.png')); sns.histplot(FOSB_cosine_sim.flatten()).set(title='FOSB Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'FOSB_cosine_similarity.png')); sns.histplot(GATA2_cosine_sim.flatten()).set(title='GATA2 Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'GATA2_cosine_similarity.png')); sns.histplot(IKZF1_cosine_sim.flatten()).set(title='IKZF1 Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'IKZF1_cosine_similarity.png')); sns.histplot(GFI1B_cosine_sim.flatten()).set(title='GFI1B Cosine Similarity'); plt.savefig(os.path.join(chatstatus['output-directory'], 'GFI1B_cosine_similarity.png'))\"], returncode=0, stdout=\"adata loaded from /nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\\nadata.shape=(6375, 512)\\nadata.obs.columns=Index(['recipe', 'filename', 'REL', 'STAT5A', 'FOS', 'ETV6', 'source', 'MEF2C',\\n       'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B', 'target'],\\n      dtype='object')\\nadata.obs.head()=   recipe                                           filename  ...  GFI1B  target\\n0  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n1  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n2  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n3  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n4  target  /nfs/turbo/umms-indikar/shared/projects/genefo...  ...  False    True\\n\\n[5 rows x 14 columns]\\nadata.var.head()=Empty DataFrame\\nColumns: []\\nIndex: [0, 1, 2, 3, 4]\\n****************************\\n      EXECUTE COMMANDS      \\n****************************\\nREL_adata = adata[adata.obs['REL']]\\nSTAT5A_adata = adata[adata.obs['STAT5A']]\\nFOS_adata = adata[adata.obs['FOS']]\\nETV6_adata = adata[adata.obs['ETV6']]\\nMEF2C_adata = adata[adata.obs['MEF2C']]\\nRUNX3_adata = adata[adata.obs['RUNX3']]\\nFOSB_adata = adata[adata.obs['FOSB']]\\nGATA2_adata = adata[adata.obs['GATA2']]\\nIKZF1_adata = adata[adata.obs['IKZF1']]\\nGFI1B_adata = adata[adata.obs['GFI1B']]\\ntarget_embeddings = adata[adata.obs['recipe'] == 'target']\\nREL_cosine_sim = 1 - cdist(REL_adata.X, target_embeddings.X, metric='cosine')\\nSTAT5A_cosine_sim = 1 - cdist(STAT5A_adata.X, target_embeddings.X, metric='cosine')\\nFOS_cosine_sim = 1 - cdist(FOS_adata.X, target_embeddings.X, metric='cosine')\\nETV6_cosine_sim = 1 - cdist(ETV6_adata.X, target_embeddings.X, metric='cosine')\\nMEF2C_cosine_sim = 1 - cdist(MEF2C_adata.X, target_embeddings.X, metric='cosine')\\nRUNX3_cosine_sim = 1 - cdist(RUNX3_adata.X, target_embeddings.X, metric='cosine')\\nFOSB_cosine_sim = 1 - cdist(FOSB_adata.X, target_embeddings.X, metric='cosine')\\nGATA2_cosine_sim = 1 - cdist(GATA2_adata.X, target_embeddings.X, metric='cosine')\\nIKZF1_cosine_sim = 1 - cdist(IKZF1_adata.X, target_embeddings.X, metric='cosine')\\nGFI1B_cosine_sim = 1 - cdist(GFI1B_adata.X, target_embeddings.X, metric='cosine')\\nsns.histplot(REL_cosine_sim.flatten()).set(title='REL Cosine Similarity')\\nplt.savefig(os.path.join(chatstatus['output-directory'], 'REL_cosine_similarity.png'))\\nsns.histplot(STAT5A_cosine_sim.flatten()).set(title='STAT5A Cosine Similarity')\\nplt.savefig(os.path.join(chatstatus['output-directory'], 'STAT5A_cosine_similarity.png'))\\nsns.histplot(FOS_cosine_sim.flatten()).set(title='FOS Cosine Similarity')\\nplt.savefig(os.path.join(chatstatus['output-directory'], 'FOS_cosine_similarity.png'))\\nsns.histplot(ETV6_cosine_sim.flatten()).set(title='ETV6 Cosine Similarity')\\nplt.savefig(os.path.join(chatstatus['output-directory'], 'ETV6_cosine_similarity.png'))\\nsns.histplot(MEF2C_cosine_sim.flatten()).set(title='MEF2C Cosine Similarity')\\nplt.savefig(os.path.join(chatstatus['output-directory'], 'MEF2C_cosine_similarity.png'))\\nsns.histplot(RUNX3_cosine_sim.flatten()).set(title='RUNX3 Cosine Similarity')\\nplt.savefig(os.path.join(chatstatus['output-directory'], 'RUNX3_cosine_similarity.png'))\\nsns.histplot(FOSB_cosine_sim.flatten()).set(title='FOSB Cosine Similarity')\\nplt.savefig(os.path.join(chatstatus['output-directory'], 'FOSB_cosine_similarity.png'))\\nsns.histplot(GATA2_cosine_sim.flatten()).set(title='GATA2 Cosine Similarity')\\nplt.savefig(os.path.join(chatstatus['output-directory'], 'GATA2_cosine_similarity.png'))\\nsns.histplot(IKZF1_cosine_sim.flatten()).set(title='IKZF1 Cosine Similarity')\\nplt.savefig(os.path.join(chatstatus['output-directory'], 'IKZF1_cosine_similarity.png'))\\nsns.histplot(GFI1B_cosine_sim.flatten()).set(title='GFI1B Cosine Similarity')\\nplt.savefig(os.path.join(chatstatus['output-directory'], 'GFI1B_cosine_similarity.png'))\\n\", stderr='/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\\n  utils.warn_names_duplicates(\"obs\")\\n')\n",
      "2024-08-03 01:41:40,311 - INFO - Debug: PYTHON code output saved to output.\n",
      "2024-08-03 01:41:40,311 - INFO - \n",
      "\n",
      "\n",
      "route\n",
      "\n",
      "\n",
      "\n",
      "2024-08-03 01:41:40,312 - INFO - CODE\n",
      "2024-08-03 01:41:40,313 - INFO - Finding Stage Number of Pipeline\n",
      "2024-08-03 01:41:40,313 - INFO - {0: {'order': 0, 'module': 'PLANNER', 'prompt': None, 'description': 'This step designed the plan. It is placed in the queue because we needed a place holder for 0 indexed lists.'}, 1: {'order': 1, 'module': 'CODE', 'prompt': \"Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` split the dataframe into several dataframes based on if the value of adata.obs is true or false for the following columns: 'REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B'. Using each adata generated above, compute the cosine similarities of the observations in each of the 10 new AnnData objects relative to the observations in the original adata where 'recipe' == 'target'. Then, please generate a separate visualization for each distribution of cosine similarities. Once you make plots for all 10 Transcription Factors, please train a model to predict the distance based on the inclusion or exclusion of each TF and visualize the model. Do not use any loops or conditional statements. Please write the code to study each transcription factor separately.\", 'description': \" 1 (CODE)**:\\nPrompt: Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` and extract the embeddings in `adata.X` and the presence/absence of each TF as features, specifically 'REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B'. Using the extracted features, define the target variable based on the distance to 'target' cells (computed by BOT2) and split the data into training and testing sets. Then, train both a Random Forest model and a Gradient Boosting model on the training data to predict the reprogramming efficiency using the features extracted. Finally, make predictions on the testing data using the better-performing model and create a plot to visualize the predicted reprogramming efficiency.\\n\\nThis revised plan condenses the original steps into a single step that uses the CODE method to perform all the required tasks. \", 'output': []}}\n",
      "2024-08-03 01:41:40,314 - INFO - IP=1\n",
      "2024-08-03 01:41:40,314 - INFO - type(IP)=<class 'int'>\n",
      "2024-08-03 01:41:40,646 - INFO - \n",
      "\n",
      "PLANNER:\n",
      "\n",
      "{'module': 'CODE', 'steps': [{'llm': \"client=<openai.resources.chat.completions.Completions object at 0x147ee873af90> async_client=<openai.resources.chat.completions.AsyncCompletions object at 0x147ef87436d0> model_name='gpt-4o' temperature=0.0 openai_api_key=SecretStr('**********') openai_proxy=''\", 'memory': 'None', 'prompt': 'input_variables=[\\'user_query\\'] template=\"You must select which code to run to help a user.\\\\n\\\\n**Available Scripts**\\\\nScript Name: scanpy_brad, \\\\t Description: This script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\\\\nScript Name: models_brad, \\\\t Description: This script executes a series of custom commands to build a model for predicting the contribution to reprogramming obtained by the inclusion or exclusion of individual TFs using an AnnData object or single cell datasets loaded from a .h5ad file.\\\\n\\\\n\\\\n**User Query**\\\\n{user_query}\\\\n\\\\n**Task**\\\\nBased on the user\\'s query, select the best script from the available scripts. Provide the script name and explain why it is the best match. If no script is good, replace script with None\\\\n\\\\n**Response Template**\\\\nSCRIPT: <selected script>\\\\nREASON: <reasoning why the script fits the user prompt>\\\\n\"', 'input': \"Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` split the dataframe into several dataframes based on if the value of adata.obs is true or false for the following columns: 'REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B'. Using each adata generated above, compute the cosine similarities of the observations in each of the 10 new AnnData objects relative to the observations in the original adata where 'recipe' == 'target'. Then, please generate a separate visualization for each distribution of cosine similarities. Once you make plots for all 10 Transcription Factors, please train a model to predict the distance based on the inclusion or exclusion of each TF and visualize the model. Do not use any loops or conditional statements. Please write the code to study each transcription factor separately.\", 'output': 'content=\"SCRIPT: scanpy_brad\\\\nREASON: The user\\'s query involves loading an AnnData object, manipulating it by splitting into several dataframes based on specific conditions, computing cosine similarities, and generating visualizations for each distribution of cosine similarities. The `scanpy_brad` script is the best match because it allows for the manipulation, visualization, and exploration of single-cell datasets using libraries like scanpy, seaborn, and matplotlib.pyplot. These capabilities are essential for performing the requested operations on the AnnData object and creating the necessary visualizations. The `models_brad` script, on the other hand, is focused on building predictive models and does not provide the necessary tools for data manipulation and visualization as described in the user\\'s query.\" response_metadata={\\'token_usage\\': {\\'completion_tokens\\': 147, \\'prompt_tokens\\': 474, \\'total_tokens\\': 621}, \\'model_name\\': \\'gpt-4o-2024-05-13\\', \\'system_fingerprint\\': \\'fp_c832e4513b\\', \\'finish_reason\\': \\'stop\\', \\'logprobs\\': None} id=\\'run-86163a8c-af4a-4c22-b82b-e753c21b1955-0\\' usage_metadata={\\'input_tokens\\': 474, \\'output_tokens\\': 147, \\'total_tokens\\': 621}', 'parsedOutput': {'scriptName': '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py', 'scriptType': 'python', 'scriptPath': '/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts'}, 'purpose': 'Select which code to run'}, {'llm': \"client=<openai.resources.chat.completions.Completions object at 0x147ee873af90> async_client=<openai.resources.chat.completions.AsyncCompletions object at 0x147ef87436d0> model_name='gpt-4o' temperature=0.0 openai_api_key=SecretStr('**********') openai_proxy=''\", 'memory': 'None', 'prompt': 'input_variables=[\\'history\\', \\'input\\'] template=\\'Current conversation:\\\\n{history}\\\\n\\\\n**PYTHON SCRIPT**\\\\nYou must run this python script:\\\\n/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\\\n\\\\n**PYTHON SCRIPT DOCUMENTATION**:\\\\nThis is the doc string of this python script:\\\\n\"\"\"\\\\nThis script executes a series of commands on an AnnData object or single cell datasets loaded from a .h5ad file. You can write your own code to manipulate, visualize, or otherwise study these data. The results will be saved back to disk. You can use the scanpy, seaborn, matplotlib.pyplot, and other standard libraries to process .h5ad files, create visualizations, and explore a dataset with this file.\\\\n\\\\nArguments (four arguments):\\\\n    1. output directory: chatstatus[\\\\\\'output-directory\\\\\\']\\\\n    2. output file: <name of output file>\\\\n    3. input file: <file created in previous step>\\\\n    4. scanpy commands: a list of scanpy commands to be executed on the AnnData object (provided as a single string with commands separated by a delimiter, e.g., \\\\\\';\\\\\\')\\\\n\\\\nBased on the arguments, the input file will be loaded, then your commands will be executed, and finally, the output or resulting ann data object will be saved to the corrrectou output file and directory. Your code is not responsible for loading the .h5ad object, that will happen automatically, and when loaded, the object will be called adata. Your scanpy commands can operate directly on the adata object that will be loaded for you.\\\\n\\\\nAdditionally, the following imports are already provided for you and can be used in your code:\\\\n```\\\\nimport scanpy as sc\\\\nimport seaborn as sns\\\\nimport matplotlib.pyplot as plt\\\\nimport os\\\\nimport numpy as np\\\\nimport pandas as pd\\\\nfrom scipy.spatial.distance import cdist\\\\n```\\\\n\\\\n**Usage**\\\\nBRAD Line:\\\\n```\\\\nsubprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\\\\\'output-directory\\\\\\'], <output file>, <input file>, \"<scanpy commands>\"], capture_output=True, text=True)\\\\n```\\\\n\\\\n**Examples**\\\\nUse the below examples to help generate your code.\\\\n\\\\n*Example 1*\\\\nUser Prompt: Run scanpy preprocessing and UMAP visualization on XXX.h5ad and save the UMAP plot\\\\nResponse Code:\\\\n```\\\\nresponse = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\\\\\'output-directory\\\\\\'], \"XXX-modified.h5ad\", \"<path/to/data>/XXX.h5ad\", \"sc.pp.neighbors(adata); sc.tl.umap(adata); sc.pl.umap(adata, save=\\\\\\'umap.png\\\\\\')\"], capture_output=True, text=True)\\\\n```\\\\nExplination: the adata object will be loaded in memory already. The command \"sc.pp.neighbors(adata)\" will preprocess the data, then the command \"sc.tl.umap(adata)\" will perform UMAP, and finally the command \"sc.pl.umap(adata, \\\\\\'umap.png\\\\\\')\" will save the UMAP to a well named file.\\\\n\\\\n*Example 2*\\\\nUser Prompt: Perform PCA and clustering on the dataset YYY.h5ad and save the PCA plot\\\\nResponse Code:\\\\n```\\\\nresponse = subprocess.run([sys.executable, \"<path/to/script/>/scanpy_brad.py\", chatstatus[\\\\\\'output-directory\\\\\\'], \"YYY-modified.h5ad\", \"<path/to/data>/YYY.h5ad\", \"sc.pp.pca(adata); sc.tl.leiden(adata); sc.pl.pca(adata, save=\\\\\\'pca.png\\\\\\')\"], capture_output=True, text=True)\\\\n```\\\\nExplination: the adata object will be loaded in memory already. The command \"sc.pp.pca(adata)\" will preprocess the data, then the command \"sc.tl.leiden(adata)\" will perform the leiden algorithm, and finally the command \"sc.pl.pca(adata, save=\\\\\\'pca.png\\\\\\')\" will save the PCA to a well named file.\\\\n\\\\n*Example 3*\\\\nUser Prompt: compute the distance between all cells that are not source or target to all source and target cells.\\\\nReponse Code:\\\\n```\\\\nadata_X = adata.X\\\\nmask = ~adata.obs[\\\\\\'recipe\\\\\\'].isin([\\\\\\'source\\\\\\', \\\\\\'target\\\\\\'])\\\\nrecipe_embeddings = adata[mask]\\\\nsource_embeddings = adata[adata.obs[\\\\\\'recipe\\\\\\'] == \\\\\\'source\\\\\\']\\\\ntarget_embeddings = adata[adata.obs[\\\\\\'recipe\\\\\\'] == \\\\\\'target\\\\\\']\\\\nrecipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X)\\\\nrecipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X)\\\\nadata_dist = sc.AnnData(X=np.concatenate((recipe_distances_source, recipe_distances_target), axis=1), var=pd.concat((source_embeddings.obs, target_embeddings.obs)), obs=recipe_embeddings.obs)\\\\nadata_dist.write(os.path.join(chatstatus[\\\\\\'output-directory\\\\\\'], <file name>.h5ad))\\\\n```\\\\n\\\\nEach plot should have a title. Also if a legend is used, please place the legend on the right hand side.\\\\n\\\\n**OUTPUT FILE NAME INSTRUCTIONS**\\\\n1. Output path should be chatstatus[\\\\\\'output-directory\\\\\\']\\\\n2. Output file name should be `<descriptive name>.h5ad`\\\\n\"\"\"\\\\n\\\\n\\\\n**CALL PYTHON SCRIPTS FROM PYTHON**:\\\\nUse the `subprocess` module to call any Python script from another Python script. Here are some examples to call a few common Python scripts:\\\\n\\\\nTo call a Python script `example_script.py` which has no arguments:\\\\n\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\']], capture_output=True, text=True)\\\\n```\\\\n\\\\nTo call a Python script `example_script.py` with one argument:\\\\n\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\'], \\\\\\'arg1\\\\\\'], capture_output=True, text=True)\\\\n```\\\\n\\\\nTo call a Python script `example_script.py` with two arguments:\\\\n```\\\\nExecute: subprocess.run([sys.executable, \\\\\\'<full path to script/> example_script.py\\\\\\', chatstatus[\\\\\\'output-directory\\\\\\'], \\\\\\'arg1\\\\\\', \\\\\\'arg2\\\\\\'], capture_output=True, text=True)\\\\n```\\\\n\\\\nNote that chatstatus[\\\\\\'output-directory\\\\\\'] is ALWAYS passed as the first argument.\\\\n\\\\nThe following files were previously created by BRAD and could be used as input to a function if necessary:\\\\nlog.json\\\\n\\\\nQuery:{input}\\\\n\\\\n**PYTHON SCRIPT**\\\\nPlease use the correct script name and path. You must run this python script:\\\\n/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\\\\n\\\\n**INSTRUCTIONS**\\\\n1. Given the user query and the documentation, identify each of the arguments found in the user\\\\\\'s query that should be passed to the Python script.\\\\n2. Using the `subprocess` module, provide the one line of code to execute the desired Python script with the given arguments. Assume the necessary modules (`subprocess` and `sys`) are already imported.\\\\n3. The last line of your response should say \"Execute: <Python code to execute>\"\\\\n4. Do not include any extra words or characters. Format the response/output as:\\\\n    Arguments: \\\\n    Python Code Explanation: <2 sentences maximum>\\\\n    Execute: <your code here>\\\\n\\\\n**IMPORTANT**\\\\nThe code to execute from your response must be formatted as:\\\\n    Execute: subprocess.call([sys.executable, \\\\\\'<path to python script>\\\\\\', \\\\\\'<argument 1>\\\\\\', \\\\\\'<argument 2>\\\\\\', ..., \\\\\\'<argument n>\\\\\\'], capture_output=True, text=True))\\\\nThis output should be exactly one line and no longer. Stop the response after this line.\\\\n\\'', 'input': \"Load the anndata object from the file located at `/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad` split the dataframe into several dataframes based on if the value of adata.obs is true or false for the following columns: 'REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B'. Using each adata generated above, compute the cosine similarities of the observations in each of the 10 new AnnData objects relative to the observations in the original adata where 'recipe' == 'target'. Then, please generate a separate visualization for each distribution of cosine similarities. Once you make plots for all 10 Transcription Factors, please train a model to predict the distance based on the inclusion or exclusion of each TF and visualize the model. Do not use any loops or conditional statements. Please write the code to study each transcription factor separately.\", 'output': 'Arguments:\\n1. output directory: chatstatus[\\'output-directory\\']\\n2. output file: \"HSC-reprogrammed-DE-modified.h5ad\"\\n3. input file: \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\"\\n4. scanpy commands: \"REL_adata = adata[adata.obs[\\'REL\\']]; STAT5A_adata = adata[adata.obs[\\'STAT5A\\']]; FOS_adata = adata[adata.obs[\\'FOS\\']]; ETV6_adata = adata[adata.obs[\\'ETV6\\']]; MEF2C_adata = adata[adata.obs[\\'MEF2C\\']]; RUNX3_adata = adata[adata.obs[\\'RUNX3\\']]; FOSB_adata = adata[adata.obs[\\'FOSB\\']]; GATA2_adata = adata[adata.obs[\\'GATA2\\']]; IKZF1_adata = adata[adata.obs[\\'IKZF1\\']]; GFI1B_adata = adata[adata.obs[\\'GFI1B\\']]; target_embeddings = adata[adata.obs[\\'recipe\\'] == \\'target\\']; REL_cosine_sim = 1 - cdist(REL_adata.X, target_embeddings.X, metric=\\'cosine\\'); STAT5A_cosine_sim = 1 - cdist(STAT5A_adata.X, target_embeddings.X, metric=\\'cosine\\'); FOS_cosine_sim = 1 - cdist(FOS_adata.X, target_embeddings.X, metric=\\'cosine\\'); ETV6_cosine_sim = 1 - cdist(ETV6_adata.X, target_embeddings.X, metric=\\'cosine\\'); MEF2C_cosine_sim = 1 - cdist(MEF2C_adata.X, target_embeddings.X, metric=\\'cosine\\'); RUNX3_cosine_sim = 1 - cdist(RUNX3_adata.X, target_embeddings.X, metric=\\'cosine\\'); FOSB_cosine_sim = 1 - cdist(FOSB_adata.X, target_embeddings.X, metric=\\'cosine\\'); GATA2_cosine_sim = 1 - cdist(GATA2_adata.X, target_embeddings.X, metric=\\'cosine\\'); IKZF1_cosine_sim = 1 - cdist(IKZF1_adata.X, target_embeddings.X, metric=\\'cosine\\'); GFI1B_cosine_sim = 1 - cdist(GFI1B_adata.X, target_embeddings.X, metric=\\'cosine\\'); sns.histplot(REL_cosine_sim.flatten()).set(title=\\'REL Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'REL_cosine_similarity.png\\')); sns.histplot(STAT5A_cosine_sim.flatten()).set(title=\\'STAT5A Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'STAT5A_cosine_similarity.png\\')); sns.histplot(FOS_cosine_sim.flatten()).set(title=\\'FOS Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'FOS_cosine_similarity.png\\')); sns.histplot(ETV6_cosine_sim.flatten()).set(title=\\'ETV6 Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'ETV6_cosine_similarity.png\\')); sns.histplot(MEF2C_cosine_sim.flatten()).set(title=\\'MEF2C Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'MEF2C_cosine_similarity.png\\')); sns.histplot(RUNX3_cosine_sim.flatten()).set(title=\\'RUNX3 Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'RUNX3_cosine_similarity.png\\')); sns.histplot(FOSB_cosine_sim.flatten()).set(title=\\'FOSB Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'FOSB_cosine_similarity.png\\')); sns.histplot(GATA2_cosine_sim.flatten()).set(title=\\'GATA2 Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'GATA2_cosine_similarity.png\\')); sns.histplot(IKZF1_cosine_sim.flatten()).set(title=\\'IKZF1 Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'IKZF1_cosine_similarity.png\\')); sns.histplot(GFI1B_cosine_sim.flatten()).set(title=\\'GFI1B Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'GFI1B_cosine_similarity.png\\'))\"\\n\\nPython Code Explanation: The code loads the AnnData object, splits it based on the specified columns, computes cosine similarities, and generates visualizations for each transcription factor.\\n\\nExecute: subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"HSC-reprogrammed-DE-modified.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"REL_adata = adata[adata.obs[\\'REL\\']]; STAT5A_adata = adata[adata.obs[\\'STAT5A\\']]; FOS_adata = adata[adata.obs[\\'FOS\\']]; ETV6_adata = adata[adata.obs[\\'ETV6\\']]; MEF2C_adata = adata[adata.obs[\\'MEF2C\\']]; RUNX3_adata = adata[adata.obs[\\'RUNX3\\']]; FOSB_adata = adata[adata.obs[\\'FOSB\\']]; GATA2_adata = adata[adata.obs[\\'GATA2\\']]; IKZF1_adata = adata[adata.obs[\\'IKZF1\\']]; GFI1B_adata = adata[adata.obs[\\'GFI1B\\']]; target_embeddings = adata[adata.obs[\\'recipe\\'] == \\'target\\']; REL_cosine_sim = 1 - cdist(REL_adata.X, target_embeddings.X, metric=\\'cosine\\'); STAT5A_cosine_sim = 1 - cdist(STAT5A_adata.X, target_embeddings.X, metric=\\'cosine\\'); FOS_cosine_sim = 1 - cdist(FOS_adata.X, target_embeddings.X, metric=\\'cosine\\'); ETV6_cosine_sim = 1 - cdist(ETV6_adata.X, target_embeddings.X, metric=\\'cosine\\'); MEF2C_cosine_sim = 1 - cdist(MEF2C_adata.X, target_embeddings.X, metric=\\'cosine\\'); RUNX3_cosine_sim = 1 - cdist(RUNX3_adata.X, target_embeddings.X, metric=\\'cosine\\'); FOSB_cosine_sim = 1 - cdist(FOSB_adata.X, target_embeddings.X, metric=\\'cosine\\'); GATA2_cosine_sim = 1 - cdist(GATA2_adata.X, target_embeddings.X, metric=\\'cosine\\'); IKZF1_cosine_sim = 1 - cdist(IKZF1_adata.X, target_embeddings.X, metric=\\'cosine\\'); GFI1B_cosine_sim = 1 - cdist(GFI1B_adata.X, target_embeddings.X, metric=\\'cosine\\'); sns.histplot(REL_cosine_sim.flatten()).set(title=\\'REL Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'REL_cosine_similarity.png\\')); sns.histplot(STAT5A_cosine_sim.flatten()).set(title=\\'STAT5A Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'STAT5A_cosine_similarity.png\\')); sns.histplot(FOS_cosine_sim.flatten()).set(title=\\'FOS Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'FOS_cosine_similarity.png\\')); sns.histplot(ETV6_cosine_sim.flatten()).set(title=\\'ETV6 Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'ETV6_cosine_similarity.png\\')); sns.histplot(MEF2C_cosine_sim.flatten()).set(title=\\'MEF2C Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'MEF2C_cosine_similarity.png\\')); sns.histplot(RUNX3_cosine_sim.flatten()).set(title=\\'RUNX3 Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'RUNX3_cosine_similarity.png\\')); sns.histplot(FOSB_cosine_sim.flatten()).set(title=\\'FOSB Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'FOSB_cosine_similarity.png\\')); sns.histplot(GATA2_cosine_sim.flatten()).set(title=\\'GATA2 Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'GATA2_cosine_similarity.png\\')); sns.histplot(IKZF1_cosine_sim.flatten()).set(title=\\'IKZF1 Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'IKZF1_cosine_similarity.png\\')); sns.histplot(GFI1B_cosine_sim.flatten()).set(title=\\'GFI1B Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'GFI1B_cosine_similarity.png\\'))\"], capture_output=True, text=True)', 'parsedOutput': {'code': 'subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"HSC-reprogrammed-DE-modified.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"REL_adata = adata[adata.obs[\\'REL\\']]; STAT5A_adata = adata[adata.obs[\\'STAT5A\\']]; FOS_adata = adata[adata.obs[\\'FOS\\']]; ETV6_adata = adata[adata.obs[\\'ETV6\\']]; MEF2C_adata = adata[adata.obs[\\'MEF2C\\']]; RUNX3_adata = adata[adata.obs[\\'RUNX3\\']]; FOSB_adata = adata[adata.obs[\\'FOSB\\']]; GATA2_adata = adata[adata.obs[\\'GATA2\\']]; IKZF1_adata = adata[adata.obs[\\'IKZF1\\']]; GFI1B_adata = adata[adata.obs[\\'GFI1B\\']]; target_embeddings = adata[adata.obs[\\'recipe\\'] == \\'target\\']; REL_cosine_sim = 1 - cdist(REL_adata.X, target_embeddings.X, metric=\\'cosine\\'); STAT5A_cosine_sim = 1 - cdist(STAT5A_adata.X, target_embeddings.X, metric=\\'cosine\\'); FOS_cosine_sim = 1 - cdist(FOS_adata.X, target_embeddings.X, metric=\\'cosine\\'); ETV6_cosine_sim = 1 - cdist(ETV6_adata.X, target_embeddings.X, metric=\\'cosine\\'); MEF2C_cosine_sim = 1 - cdist(MEF2C_adata.X, target_embeddings.X, metric=\\'cosine\\'); RUNX3_cosine_sim = 1 - cdist(RUNX3_adata.X, target_embeddings.X, metric=\\'cosine\\'); FOSB_cosine_sim = 1 - cdist(FOSB_adata.X, target_embeddings.X, metric=\\'cosine\\'); GATA2_cosine_sim = 1 - cdist(GATA2_adata.X, target_embeddings.X, metric=\\'cosine\\'); IKZF1_cosine_sim = 1 - cdist(IKZF1_adata.X, target_embeddings.X, metric=\\'cosine\\'); GFI1B_cosine_sim = 1 - cdist(GFI1B_adata.X, target_embeddings.X, metric=\\'cosine\\'); sns.histplot(REL_cosine_sim.flatten()).set(title=\\'REL Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'REL_cosine_similarity.png\\')); sns.histplot(STAT5A_cosine_sim.flatten()).set(title=\\'STAT5A Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'STAT5A_cosine_similarity.png\\')); sns.histplot(FOS_cosine_sim.flatten()).set(title=\\'FOS Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'FOS_cosine_similarity.png\\')); sns.histplot(ETV6_cosine_sim.flatten()).set(title=\\'ETV6 Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'ETV6_cosine_similarity.png\\')); sns.histplot(MEF2C_cosine_sim.flatten()).set(title=\\'MEF2C Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'MEF2C_cosine_similarity.png\\')); sns.histplot(RUNX3_cosine_sim.flatten()).set(title=\\'RUNX3 Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'RUNX3_cosine_similarity.png\\')); sns.histplot(FOSB_cosine_sim.flatten()).set(title=\\'FOSB Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'FOSB_cosine_similarity.png\\')); sns.histplot(GATA2_cosine_sim.flatten()).set(title=\\'GATA2 Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'GATA2_cosine_similarity.png\\')); sns.histplot(IKZF1_cosine_sim.flatten()).set(title=\\'IKZF1 Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'IKZF1_cosine_similarity.png\\')); sns.histplot(GFI1B_cosine_sim.flatten()).set(title=\\'GFI1B Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'GFI1B_cosine_similarity.png\\'))\"], capture_output=True, text=True)'}, 'purpose': 'Format function call'}, {'func': 'pythonCaller.execute_python_code', 'code': 'response = subprocess.run([sys.executable, \"/home/jpic/RAG-DEV/tutorials/geneformer/analysis-scripts/scanpy_brad.py\", chatstatus[\\'output-directory\\'], \"HSC-reprogrammed-DE-modified.h5ad\", \"/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad\", \"REL_adata = adata[adata.obs[\\'REL\\']]; STAT5A_adata = adata[adata.obs[\\'STAT5A\\']]; FOS_adata = adata[adata.obs[\\'FOS\\']]; ETV6_adata = adata[adata.obs[\\'ETV6\\']]; MEF2C_adata = adata[adata.obs[\\'MEF2C\\']]; RUNX3_adata = adata[adata.obs[\\'RUNX3\\']]; FOSB_adata = adata[adata.obs[\\'FOSB\\']]; GATA2_adata = adata[adata.obs[\\'GATA2\\']]; IKZF1_adata = adata[adata.obs[\\'IKZF1\\']]; GFI1B_adata = adata[adata.obs[\\'GFI1B\\']]; target_embeddings = adata[adata.obs[\\'recipe\\'] == \\'target\\']; REL_cosine_sim = 1 - cdist(REL_adata.X, target_embeddings.X, metric=\\'cosine\\'); STAT5A_cosine_sim = 1 - cdist(STAT5A_adata.X, target_embeddings.X, metric=\\'cosine\\'); FOS_cosine_sim = 1 - cdist(FOS_adata.X, target_embeddings.X, metric=\\'cosine\\'); ETV6_cosine_sim = 1 - cdist(ETV6_adata.X, target_embeddings.X, metric=\\'cosine\\'); MEF2C_cosine_sim = 1 - cdist(MEF2C_adata.X, target_embeddings.X, metric=\\'cosine\\'); RUNX3_cosine_sim = 1 - cdist(RUNX3_adata.X, target_embeddings.X, metric=\\'cosine\\'); FOSB_cosine_sim = 1 - cdist(FOSB_adata.X, target_embeddings.X, metric=\\'cosine\\'); GATA2_cosine_sim = 1 - cdist(GATA2_adata.X, target_embeddings.X, metric=\\'cosine\\'); IKZF1_cosine_sim = 1 - cdist(IKZF1_adata.X, target_embeddings.X, metric=\\'cosine\\'); GFI1B_cosine_sim = 1 - cdist(GFI1B_adata.X, target_embeddings.X, metric=\\'cosine\\'); sns.histplot(REL_cosine_sim.flatten()).set(title=\\'REL Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'REL_cosine_similarity.png\\')); sns.histplot(STAT5A_cosine_sim.flatten()).set(title=\\'STAT5A Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'STAT5A_cosine_similarity.png\\')); sns.histplot(FOS_cosine_sim.flatten()).set(title=\\'FOS Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'FOS_cosine_similarity.png\\')); sns.histplot(ETV6_cosine_sim.flatten()).set(title=\\'ETV6 Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'ETV6_cosine_similarity.png\\')); sns.histplot(MEF2C_cosine_sim.flatten()).set(title=\\'MEF2C Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'MEF2C_cosine_similarity.png\\')); sns.histplot(RUNX3_cosine_sim.flatten()).set(title=\\'RUNX3 Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'RUNX3_cosine_similarity.png\\')); sns.histplot(FOSB_cosine_sim.flatten()).set(title=\\'FOSB Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'FOSB_cosine_similarity.png\\')); sns.histplot(GATA2_cosine_sim.flatten()).set(title=\\'GATA2 Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'GATA2_cosine_similarity.png\\')); sns.histplot(IKZF1_cosine_sim.flatten()).set(title=\\'IKZF1 Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'IKZF1_cosine_similarity.png\\')); sns.histplot(GFI1B_cosine_sim.flatten()).set(title=\\'GFI1B Cosine Similarity\\'); plt.savefig(os.path.join(chatstatus[\\'output-directory\\'], \\'GFI1B_cosine_similarity.png\\'))\"], capture_output=True, text=True)', 'purpose': 'execute python code'}, {'func': 'utils.makeNamesConsistent', 'files': [{'old-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-40-37/RUNX3_cosine_similarity.png', 'new-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-40-37/S1-RUNX3_cosine_similarity.png'}, {'old-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-40-37/GFI1B_cosine_similarity.png', 'new-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-40-37/S1-GFI1B_cosine_similarity.png'}, {'old-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-40-37/MEF2C_cosine_similarity.png', 'new-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-40-37/S1-MEF2C_cosine_similarity.png'}, {'old-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-40-37/ETV6_cosine_similarity.png', 'new-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-40-37/S1-ETV6_cosine_similarity.png'}, {'old-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-40-37/FOSB_cosine_similarity.png', 'new-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-40-37/S1-FOSB_cosine_similarity.png'}, {'old-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-40-37/REL_cosine_similarity.png', 'new-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-40-37/S1-REL_cosine_similarity.png'}, {'old-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-40-37/FOS_cosine_similarity.png', 'new-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-40-37/S1-FOS_cosine_similarity.png'}, {'old-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-40-37/GATA2_cosine_similarity.png', 'new-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-40-37/S1-GATA2_cosine_similarity.png'}, {'old-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-40-37/HSC-reprogrammed-DE-modified.h5ad', 'new-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-40-37/S1-HSC-reprogrammed-DE-modified.h5ad'}, {'old-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-40-37/IKZF1_cosine_similarity.png', 'new-name': '/home/jpic/RAG-DEV/tutorials/geneformer/logs/2024-08-03_01-40-37/S1-IKZF1_cosine_similarity.png'}]}]}\n",
      "2024-08-03 01:41:40,656 - INFO - current llm calls=3\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n"
     ]
    }
   ],
   "source": [
    "llm = llms.load_openai(model_name='gpt-4o')\n",
    "# llm = llms.load_nvidia() #model_name='meta/llama-3.1-405b-instruct')\n",
    "BOT3 = brad.chatbot(llm=llm, interactive=False, config='/home/jpic/RAG-DEV/tutorials/geneformer/config-sc.json')\n",
    "BOT3.chat() # invoke('/force PLANNER please load and run the BOT3 pipeline')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "ddacb282-8098-44f7-a813-be42956b8b87",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'cdist' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[41], line 12\u001b[0m\n\u001b[1;32m     10\u001b[0m GFI1B_adata \u001b[38;5;241m=\u001b[39m adata[adata\u001b[38;5;241m.\u001b[39mobs[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mGFI1B\u001b[39m\u001b[38;5;124m'\u001b[39m]]\n\u001b[1;32m     11\u001b[0m target_embeddings \u001b[38;5;241m=\u001b[39m adata[adata\u001b[38;5;241m.\u001b[39mobs[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mrecipe\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtarget\u001b[39m\u001b[38;5;124m'\u001b[39m]\n\u001b[0;32m---> 12\u001b[0m REL_distances \u001b[38;5;241m=\u001b[39m \u001b[43mcdist\u001b[49m(REL_adata\u001b[38;5;241m.\u001b[39mX, target_embeddings\u001b[38;5;241m.\u001b[39mX)\n\u001b[1;32m     13\u001b[0m STAT5A_distances \u001b[38;5;241m=\u001b[39m cdist(STAT5A_adata\u001b[38;5;241m.\u001b[39mX, target_embeddings\u001b[38;5;241m.\u001b[39mX)\n\u001b[1;32m     14\u001b[0m FOS_distances \u001b[38;5;241m=\u001b[39m cdist(FOS_adata\u001b[38;5;241m.\u001b[39mX, target_embeddings\u001b[38;5;241m.\u001b[39mX)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'cdist' is not defined"
     ]
    }
   ],
   "source": [
    "REL_adata = adata[adata.obs['REL']]\n",
    "STAT5A_adata = adata[adata.obs['STAT5A']]\n",
    "FOS_adata = adata[adata.obs['FOS']]\n",
    "ETV6_adata = adata[adata.obs['ETV6']]\n",
    "MEF2C_adata = adata[adata.obs['MEF2C']]\n",
    "RUNX3_adata = adata[adata.obs['RUNX3']]\n",
    "FOSB_adata = adata[adata.obs['FOSB']]\n",
    "GATA2_adata = adata[adata.obs['GATA2']]\n",
    "IKZF1_adata = adata[adata.obs['IKZF1']]\n",
    "GFI1B_adata = adata[adata.obs['GFI1B']]\n",
    "target_embeddings = adata[adata.obs['recipe'] == 'target']\n",
    "REL_distances = cdist(REL_adata.X, target_embeddings.X)\n",
    "STAT5A_distances = cdist(STAT5A_adata.X, target_embeddings.X)\n",
    "FOS_distances = cdist(FOS_adata.X, target_embeddings.X)\n",
    "ETV6_distances = cdist(ETV6_adata.X, target_embeddings.X)\n",
    "MEF2C_distances = cdist(MEF2C_adata.X, target_embeddings.X)\n",
    "RUNX3_distances = cdist(RUNX3_adata.X, target_embeddings.X)\n",
    "FOSB_distances = cdist(FOSB_adata.X, target_embeddings.X)\n",
    "GATA2_distances = cdist(GATA2_adata.X, target_embeddings.X)\n",
    "IKZF1_distances = cdist(IKZF1_adata.X, target_embeddings.X)\n",
    "GFI1B_distances = cdist(GFI1B_adata.X, target_embeddings.X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2f32afe-2bc9-4319-b100-63413de74b66",
   "metadata": {},
   "outputs": [],
   "source": [
    "REL_adata = adata[adata.obs['REL']]\n",
    "STAT5A_adata = adata[adata.obs['STAT5A']]\n",
    "FOS_adata = adata[adata.obs['FOS']]\n",
    "ETV6_adata = adata[adata.obs['ETV6']]\n",
    "MEF2C_adata = adata[adata.obs['MEF2C']]\n",
    "RUNX3_adata = adata[adata.obs['RUNX3']]\n",
    "FOSB_adata = adata[adata.obs['FOSB']]\n",
    "GATA2_adata = adata[adata.obs['GATA2']]\n",
    "IKZF1_adata = adata[adata.obs['IKZF1']]\n",
    "GFI1B_adata = adata[adata.obs['GFI1B']]\n",
    "target_embeddings = adata[adata.obs['recipe'] == 'target']\n",
    "REL_distances = cdist(REL_adata.X, target_embeddings.X)\n",
    "STAT5A_distances = cdist(STAT5A_adata.X, target_embeddings.X)\n",
    "FOS_distances = cdist(FOS_adata.X, target_embeddings.X)\n",
    "ETV6_distances = cdist(ETV6_adata.X, target_embeddings.X)\n",
    "MEF2C_distances = cdist(MEF2C_adata.X, target_embeddings.X)\n",
    "RUNX3_distances = cdist(RUNX3_adata.X, target_embeddings.X)\n",
    "FOSB_distances = cdist(FOSB_adata.X, target_embeddings.X)\n",
    "GATA2_distances = cdist(GATA2_adata.X, target_embeddings.X)\n",
    "IKZF1_distances = cdist(IKZF1_adata.X, target_embeddings.X)\n",
    "GFI1B_distances = cdist(GFI1B_adata.X, target_embeddings.X)\n",
    "sns.histplot(REL_distances); \n",
    "plt.savefig('REL_distances.png');\n",
    "sns.histplot(STAT5A_distances);\n",
    "plt.savefig('STAT5A_distances.png');\n",
    "sns.histplot(FOS_distances);\n",
    "plt.savefig('FOS_distances.png');\n",
    "sns.histplot(ETV6_distances);\n",
    "plt.savefig('ETV6_distances.png');\n",
    "sns.histplot(MEF2C_distances);\n",
    "plt.savefig('MEF2C_distances.png');\n",
    "sns.histplot(RUNX3_distances);\n",
    "plt.savefig('RUNX3_distances.png');\n",
    "sns.histplot(FOSB_distances);\n",
    "plt.savefig('FOSB_distances.png');\n",
    "sns.histplot(GATA2_distances);\n",
    "plt.savefig('GATA2_distances.png');\n",
    "sns.histplot(IKZF1_distances);\n",
    "plt.savefig('IKZF1_distances.png'); sns.histplot(GFI1B_distances); plt.savefig('GFI1B_distances.png'); from sklearn.linear_model import LinearRegression; X = adata.obs[['REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B']]; y = cdist(adata.X, target_embeddings.X).mean(axis=1); model = LinearRegression().fit(X, y); plt.figure(); plt.plot(model.coef_); plt.savefig('model_coefficients.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b0339879-e3b2-4758-a03f-ac0c26ae8b92",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'np' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[8], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01msklearn\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtree\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m DecisionTreeRegressor\n\u001b[0;32m----> 2\u001b[0m X \u001b[38;5;241m=\u001b[39m \u001b[43mnp\u001b[49m\u001b[38;5;241m.\u001b[39mconcatenate([adata\u001b[38;5;241m.\u001b[39mX, adata\u001b[38;5;241m.\u001b[39mobs[[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mREL\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mSTAT5A\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mFOS\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mETV6\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mMEF2C\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mRUNX3\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mFOSB\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mGATA2\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mIKZF1\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mGFI1B\u001b[39m\u001b[38;5;124m'\u001b[39m]]], axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m)\n\u001b[1;32m      3\u001b[0m y \u001b[38;5;241m=\u001b[39m cdist(adata[adata\u001b[38;5;241m.\u001b[39mobs[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mrecipe\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtarget\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39mX, adata\u001b[38;5;241m.\u001b[39mX)\u001b[38;5;241m.\u001b[39mmin(axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0\u001b[39m)\n\u001b[1;32m      4\u001b[0m X_train, X_test, y_train, y_test \u001b[38;5;241m=\u001b[39m train_test_split(X, y, test_size\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0.2\u001b[39m, random_state\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m42\u001b[39m)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'np' is not defined"
     ]
    }
   ],
   "source": [
    "from sklearn.tree import DecisionTreeRegressor\n",
    "X = np.concatenate([adata.X, adata.obs[['REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B']]], axis=1)\n",
    "y = cdist(adata[adata.obs['recipe'] == 'target'].X, adata.X).min(axis=0)\n",
    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)\n",
    "model = DecisionTreeRegressor()\n",
    "model.fit(X_train, y_train)\n",
    "y_pred = model.predict(X_test)\n",
    "results = {'model': model, 'r2_score': r2_score(y_test, y_pred), 'mean_squared_error': mean_squared_error(y_test, y_pred), 'mean_absolute_error': mean_absolute_error(y_test, y_pred)}\n",
    "with open(os.path.join(output_directory, 'decision_tree_metrics.txt'), 'w') as f: f.write(str(results))\n",
    "    \n",
    "plt.figure()\n",
    "plt.scatter(y_test, y_pred)\n",
    "plt.xlabel('True Values')\n",
    "plt.ylabel('Predictions')\n",
    "# plt.savefig(os.path.join(output_directory, 'decision_tree_scatter.png'))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c759e3f9-3d72-41ff-8f5e-8feb4f6cf97a",
   "metadata": {},
   "source": [
    "# Testing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "9f0812a1-b03e-4079-8b6f-b9945ecd33ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import argparse\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy.spatial.distance import cdist\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.ensemble import RandomForestRegressor\n",
    "from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4ddfba56-4c93-46f7-888b-2933f62bdc8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "from sklearn.cluster import KMeans\n",
    "import anndata as ad\n",
    "from scipy.spatial.distance import cdist\n",
    "\n",
    "adata = sc.read_h5ad('/nfs/turbo/umms-indikar/shared/projects/geneformer/fib15k/jpic/HSC-reprogrammed-DE.h5ad')\n",
    "# adata = sc.read_h5ad('distances.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "cba373a1-2b8b-4b8a-9526-2ddd8c7cd050",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0, 0.5, 'Predictions')"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjMAAAGwCAYAAABcnuQpAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAAArhUlEQVR4nO3deXRU9f3/8dckgSyQDKZklRBTDF9AKPuqrEIElYKxP8ClhQoKKmhK/aKglrRSIrZSWwG1WIh+LTXaAlpLqXEhgEhZImUtBIkSJfmGRTJsSUhyf3/wzdSQEJLJTGY+yfNxzj1n5t47n/sOwz3zOvd+7udjsyzLEgAAgKH8vF0AAABAQxBmAACA0QgzAADAaIQZAABgNMIMAAAwGmEGAAAYjTADAACMFuDtAjytoqJCx44dU2hoqGw2m7fLAQAAdWBZls6cOaPY2Fj5+dV+7aXJh5ljx44pLi7O22UAAAAX5OXlqV27drXu0+TDTGhoqKRL/xhhYWFergYAANSFw+FQXFyc83e8Nk0+zFTeWgoLCyPMAABgmLp0EaEDMAAAMBphBgAAGI0wAwAAjEaYAQAARiPMAAAAoxFmAACA0QgzAADAaIQZAABgNMIMAAAwWpMfARgAALhPeYWlbbmnVFB0QafOlSq8daCiw4LULyFc/n7emdCZMAMAAGpVXmFp65GTen1Lrj7693FdrLCq7RMVGqifj7tBo7vGNHp9hBkAAHBF6/fm64nVe3T6/MVa9/vfMyWa8Ua2Xr63V6MHGvrMAACAGq3bna8Zb2RfNch826Nv7lJ5DVduPIkwAwAAqlm3+5geWpVd78+VlFVo86HjHqjoyggzAACgiktB5jOXP7980xE3VnN1hBkAAOD09va8BgUZSSoqrvttKXegAzAAAJAk9V2QqeNnSxvcTvd2djdUU3eEGQAAmpnKsWIKzxQrMvTSGDFDns3U8bPuuaLy5G03uKWduiLMAADQjKzfm6+f/3W/8ouKPdL+yM6RCm7p75G2r4QwAwBAM7F+b74efCNbnnpw+nvtwvTq5L4eav3KCDMAADQD5RWWfv7X/R4LMr+b0F3f79XOQ63XjjADAEAzsC33lEduLdkkHVwwRi0DvPeANI9mAwDQDBSe8UwfmaV39/JqkJG4MgMAQJNWXmFp48FCpby5y+1t3z/4Ot36vcafWPJyhBkAAJqo9XvzNXPVZyrzwFxJ9w9O0JO3dXF7u64gzAAA0AQ1dEqCKwlv1VILxnX1iSsylQgzAAA0Me9mf61H3trl9nZ/MjJRM0ckyt/P5va2G4IwAwBAE1A5qu8z7+3T/vwzbm07pKW/Fk/ortFdfedqzLcRZgAAMNz6vflKfXefChwlbm3X30+aNTxRs272vasx30aYAQDAUOUVll788JBe+PCwW9v1k7RySl/d1DHCp0NMJcIMAAAGWr83X0+s3qPT590zOWSlFn5SzsLb3NqmpxFmAAAwzPq9+ZrxRrbb241rE6RNT9zs9nY9jTADAIBByisszXn7X25v9/kfdNedfbwzt1JDEWYAADBEeYWl3gvel6Ok3K3tLpnUQ7f3uNatbTYmwgwAAD6uvMLSko8O67cfHpK7B/OdPiTB6CAjEWYAAPBp6/fm64m/7NHpC+7t6BsYYNNvJvT0qZF8XUWYAQDAx1QOgJe5v0ArPvnC7e33u+4a/emBgUY8dl0XhBkAALysMrwUFF3Q5pzjWr/vf3Wu1L39Yiq19Lc1qSAjEWYAAPCq9Xvz9fO/7ld+UXGjHO93d/VsUkFGIswAAOA16/fm68E3suXmPr01ig4LVOr3b/DZ+ZUagjADAIAXlFdY+vlf93s0yNgkLZ7QXdH2YPVLCG9yV2QqEWYAAPCCbbmnPHprKbiFnw48M8Zj7fsSP28XAABAc1NaVqGM7Uc91n5oSz/t/floj7Xva7gyAwBAI0pbt1/LN+W6ffC7b/vVhB5N9pZSTQgzAAA0krR1+/XKxlyPtR9jD9L8sV2aZCff2hBmAABoBKVlFfq9h4LMj2+8Tkldopt0J9/aEGYAAGigykHvCs8UKzI0qEqoqNy2aN1ejzy5tOzunrr1e7EeaNkchBkAABqgpkHvwlu10B09rlVYcAv9adtRFThKPHLs6UMSmn2QkQgzAADUW+XVlg/2F+gPNcyddOrcxRrXu5NN0rv/ytec0Z2b5a2lbyPMAABQD409/cCVWJLyi4q1LfeUBnb4jldr8TavjjOTlpamvn37KjQ0VJGRkRo/frwOHjxYZR/LspSamqrY2FgFBwdr2LBh2rdvn5cqBgA0Z5XTD3g7yHxb4RnfqcVbvBpmsrKy9PDDD2vr1q3KzMxUWVmZkpKSdO7cOec+zz33nBYvXqwlS5Zo+/btio6O1qhRo3TmzBkvVg4AaG4aY/oBV0SGBnm7BK+zWZblM9/L8ePHFRkZqaysLA0ZMkSWZSk2NlYpKSl6/PHHJUklJSWKiorSokWLNH369GptlJSUqKTkPx2tHA6H4uLiVFRUpLCwsEb7WwAATcunn5/UXcu3ersMJ5ukaHuQNj8+okn2mXE4HLLb7XX6/fap6QyKiookSeHh4ZKk3NxcFRQUKCkpyblPYGCghg4dqi1bttTYRlpamux2u3OJi4vzfOEAgCbPm7dzLo8qle/nj+3SJINMfflMmLEsS7Nnz9ZNN92krl27SpIKCgokSVFRUVX2jYqKcm673Ny5c1VUVORc8vLyPFs4AKBZaOzbOTH2IL18by+9fG8vRdurHjvaHqSX7u3V7Eb6vRKfeZpp5syZ2r17tzZv3lxtm81WNXVallVtXaXAwEAFBgZ6pEYAQPPVLyFcbYJb6PSFix47hr9Neu4H3RXbJrjKwHujukRfcVA++EiYmTVrlt59911t3LhR7dq1c66Pjo6WdOkKTUzMf9JnYWFhtas1AAC42+Uj+04ZdJ1e+DDHY8dbek/NV1v8/WzN/vHr2ng1zFiWpVmzZmnNmjXasGGDEhISqmxPSEhQdHS0MjMz1bNnT0lSaWmpsrKytGjRIm+UDABoJmoaT8Ye7JmfzeY6QaS7eDXMPPzww1q1apXeeecdhYaGOvvB2O12BQcHy2azKSUlRQsXLlRiYqISExO1cOFChYSE6O677/Zm6QCAJqxyPJnLH/ctulDm1uN0aNtKC+7oxm2jBvJqmHnppZckScOGDauyfuXKlZoyZYokac6cObpw4YIeeughffPNN+rfv7/ef/99hYaGNnK1AIDmoDHHk5k54npuH7mB128zXY3NZlNqaqpSU1M9XxAAoNnblnuq0Ub4jbYHN8pxmjqfeTQbAABf0FjjyUSHBapfQnijHKup84mnmQAA8LbKJ5f+vqfmcczcLfX7N9BPxk0IMwCAZm/93nylvrtfBQ7PX5VpE9JCzyZ348klNyLMAACatfV78zXjjWyPH2d8j1j9v95xGtDhO1yRcTPCDACg2SqvsPTTt/7l0WNUTgj5/IQehBgPoQMwAKBZKq+w9NOMz3SutNxjx2BCyMbBlRkAQJN3+bQE35wrVcpbu1RaVuHR40Yzsm+jIMwAAIxweSCp66i5NU1L4EmtA/01sU+cRnaJZmTfRkKYAQD4vJoCSV3mM7rStASeEBTgpweHddDMEYkEmEZGmAEA+LQrBZKComI9+Ea2Xrq35pmmG3NagrCgAO14apRaBtAV1Rv4VwcA+KzaAknlup//db/KK6rv0ZjTEjyb/D2CjBfxLw8A8FlXCySWpPyiYm3LPVVtW+b+xhnJd/qQBN36PTr4ehO3mQAAPquu8yRdvt97u77Wik++8EBF/xHeqoUWjOuqW78X69Hj4OoIMwAAnxUZGlTv/X75t/1avinXUyXp4WEddFNiBE8q+RDCDADAZ/VLCFeMPUgFRcU19pupHF23X0K4yisszVqVrXV7PXd7KcYepNlJ/0WI8TH0mQEA+Cx/P5vmj+0i6T+j6Vb69ui6mfsL1G3+eo8FGdv/LYzk65sIMwAAnza6a4xeureXou1VbzlF24P00r29JEkz3sjW+YueG8238liM5OubuM0EAPB5o7vGaFSX6GojAF8oLVe31H947Lhtglto6T29NOC7zHTtywgzAAAj+PvZNLDDd5zvp6Zv04f/Pu6RY1XGlmfv7KYbr2/rkWPAfQgzAADjDH3uI3156oLH2meCSLMQZgAARrlv5Ta3Bpl2bYKUlvw9+fnZdOJsSb0msYRvIMwAAIzx3q5j+uige28t/WTUf2lwxwi3tonGxdNMAAAjlFdYeuzP/3J7u7Ftgt3eJhoXYQYAYIStn59UcZl7H7+O+b8B92A2wgwAwAhbjpxwe5sMgtc0EGYAAEY49o17n16678breFqpiSDMAACM4O6+LaO6RLu1PXgPYQYAYIRBHdwzeJ1N9JVpaggzAAAjDOjwHbUJaXHV/VoH+mvqTfHOySG/7duTU9JXpukgzAAAjODvZ9Ozyd1q3ecnIxP1r/m36Onbu9Y6OSV9ZZoWm2VZlreL8CSHwyG73a6ioiKFhYV5uxwAQAOt35uv1Hf3qcBR4lwXHRao1O/fUC2klFdY1San5IqMGerz+02YAQAYh5DS9NXn95vpDAAAxrl8Bm00b/SZAQAARiPMAAAAoxFmAACA0QgzAADAaIQZAABgNMIMAAAwGmEGAAAYjTADAACMRpgBAABGI8wAAACjEWYAAIDRCDMAAMBohBkAAGA0wgwAADAaYQYAABiNMAMAAIxGmAEAAEYjzAAAAKMRZgAAgNEIMwAAwGiEGQAAYDTCDAAAMBphBgAAGI0wAwAAjEaYAQAARiPMAAAAoxFmAACA0QgzAADAaIQZAABgNMIMAAAwmlfDzMaNGzV27FjFxsbKZrNp7dq1VbZPmTJFNputyjJgwADvFAsAAHySV8PMuXPn1L17dy1ZsuSK+4wePVr5+fnOZd26dY1YIQAA8HUB3jz4mDFjNGbMmFr3CQwMVHR0dJ3bLCkpUUlJifO9w+FwuT4AAOD7fL7PzIYNGxQZGamOHTvq/vvvV2FhYa37p6WlyW63O5e4uLhGqhQAAHiDzbIsy9tFSJLNZtOaNWs0fvx457qMjAy1bt1a8fHxys3N1dNPP62ysjLt3LlTgYGBNbZT05WZuLg4FRUVKSwszNN/BgAAcAOHwyG73V6n32+v3ma6mokTJzpfd+3aVX369FF8fLz+9re/KTk5ucbPBAYGXjHoAACApsfnbzN9W0xMjOLj45WTk+PtUgAAgI8wKsycPHlSeXl5iomJ8XYpAADAR3j1NtPZs2d1+PBh5/vc3Fzt2rVL4eHhCg8PV2pqqu68807FxMToiy++0Lx589S2bVvdcccdXqwaAAD4Eq+GmR07dmj48OHO97Nnz5YkTZ48WS+99JL27Nmj119/XadPn1ZMTIyGDx+ujIwMhYaGeqtkAADgY3zmaSZPqU9vaAAA4Bvq8/ttVJ8ZAACAyxFmAACA0QgzAADAaIQZAABgNMIMAAAwGmEGAAAYjTADAACMRpgBAABGI8wAAACjEWYAAIDRCDMAAMBohBkAAGA0wgwAADCaS2EmOztbe/bscb5/5513NH78eM2bN0+lpaVuKw4AAOBqXAoz06dP16FDhyRJR44c0aRJkxQSEqK3335bc+bMcWuBAAAAtXEpzBw6dEg9evSQJL399tsaMmSIVq1apfT0dP3lL39xZ30AAAC1cinMWJaliooKSdIHH3ygW2+9VZIUFxenEydOuK86AACAq3ApzPTp00cLFizQ//zP/ygrK0u33XabJCk3N1dRUVFuLRAAAKA2LoWZF154QdnZ2Zo5c6aefPJJXX/99ZKkP//5zxo0aJBbCwQAAKiNzbIsy12NFRcXy9/fXy1atHBXkw3mcDhkt9tVVFSksLAwb5cDAADqoD6/3wENOVBpaakKCwud/WcqtW/fviHNAgAA1JlLYebQoUOaOnWqtmzZUmW9ZVmy2WwqLy93S3EAAABX41KY+fGPf6yAgAC99957iomJkc1mc3ddAAAAdeJSmNm1a5d27typTp06ubseAACAenHpaaYuXbowngwAAPAJLoWZRYsWac6cOdqwYYNOnjwph8NRZQEAAGgsLj2a7ed3KQNd3lfGFzsA82g2AADm8fij2R9//LFLhQEAALibS2Fm6NCh7q4DAADAJS4Pmnf69Gn94Q9/0IEDB2Sz2dSlSxfdd999stvt7qwPAACgVi51AN6xY4c6dOig3/zmNzp16pROnDihxYsXq0OHDsrOznZ3jQAAAFfkUgfgwYMH6/rrr9fy5csVEHDp4k5ZWZmmTZumI0eOaOPGjW4v1FV0AAYAwDz1+f12KcwEBwfrs88+qzZo3v79+9WnTx+dP3++vk16DGEGAADz1Of326XbTGFhYTp69Gi19Xl5eQoNDXWlSQAAAJe4FGYmTpyoqVOnKiMjQ3l5efrqq6/05ptvatq0abrrrrvcXSMAAMAVufQ0069//WvZbDb96Ec/UllZmSSpRYsWevDBB/Xss8+6tUAAAIDauNRnptL58+f1+eefy7IsXX/99QoJCXFnbW5BnxkAAMzj8RGAK4WEhKhbt24NaQIAAKBB6hxmkpOTlZ6errCwMCUnJ9e67+rVqxtcGAAAQF3UOczY7XbnxJJhYWHVJpkEAADwhgb1mTEBfWYAADCPx8eZGTFihE6fPl3jgUeMGOFKkwAAAC5xKcxs2LBBpaWl1dYXFxdr06ZNDS4KAACgrur1NNPu3budr/fv36+CggLn+/Lycq1fv17XXnut+6oDAAC4inqFmR49eshms8lms9V4Oyk4OFgvvvii24oDAAC4mnqFmdzcXFmWpe9+97vatm2bIiIinNtatmypyMhI+fv7u71IAACAK6lXmImPj5ckVVRUeKQYAACA+nKpA3BaWppWrFhRbf2KFSu0aNGiBhcFAABQVy6FmVdeeUWdOnWqtv6GG27Qyy+/3OCiAAAA6sqlMFNQUKCYmJhq6yMiIpSfn9/gogAAAOrKpTATFxenTz75pNr6Tz75RLGxsQ0uCgAAoK5cmjV72rRpSklJ0cWLF52PaH/44YeaM2eOfvrTn7q1QAAAgNq4FGbmzJmjU6dO6aGHHnKOBBwUFKTHH39cc+fOdWuBAAAAtWnQRJNnz57VgQMHFBwcrMTERAUGBrqzNrdgokkAAMxTn99vl67MVGrdurX69u3bkCYAAAAapM5hJjk5Wenp6QoLC1NycnKt+65evbrBhQEAANRFncOM3W6XzWZzvgYAAPAFDeozYwL6zAAAYJ76/H67NM4MAACAr6jzbaaePXs6bzNdTXZ2tssFAQAA1Eedw8z48eOdr4uLi7Vs2TJ16dJFAwcOlCRt3bpV+/bt00MPPeT2IgEAAK6kzmFm/vz5ztfTpk3TI488omeeeabaPnl5ee6rDgAA4Cpc6gBst9u1Y8cOJSYmVlmfk5OjPn36qKioyG0FNhQdgAEAMI/HOwAHBwdr8+bN1dZv3rxZQUFBdW5n48aNGjt2rGJjY2Wz2bR27doq2y3LUmpqqmJjYxUcHKxhw4Zp3759rpQMAACaKJdGAE5JSdGDDz6onTt3asCAAZIu9ZlZsWKFfvazn9W5nXPnzql79+768Y9/rDvvvLPa9ueee06LFy9Wenq6OnbsqAULFmjUqFE6ePCgQkNDXSkdAAA0MS6PM/PWW2/pt7/9rQ4cOCBJ6ty5sx599FFNmDDBtUJsNq1Zs8bZ0diyLMXGxiolJUWPP/64JKmkpERRUVFatGiRpk+fXqd2uc0EAIB5GmVupgkTJrgcXOoiNzdXBQUFSkpKcq4LDAzU0KFDtWXLliuGmZKSEpWUlDjfOxwOj9UIAAC8z+VB806fPq1XX31V8+bN06lTpyRdGl/m66+/dkthBQUFkqSoqKgq66OiopzbapKWlia73e5c4uLi3FIPAADwTS6Fmd27d6tjx45atGiRfvWrX+n06dOSpDVr1mju3LnurK/aQH2WZdU6eN/cuXNVVFTkXHhUHACAps2lMDN79mxNmTJFOTk5VZ5eGjNmjDZu3OiWwqKjoyWp2lWYwsLCaldrvi0wMFBhYWFVFgAA0HS5FGa2b99eY5+Va6+9ttZbQPWRkJCg6OhoZWZmOteVlpYqKytLgwYNcssxAACA+VzqABwUFFRjx9qDBw8qIiKizu2cPXtWhw8fdr7Pzc3Vrl27FB4ervbt2yslJUULFy5UYmKiEhMTtXDhQoWEhOjuu+92pWwAANAEuRRmxo0bp1/84hd66623JF3q13L06FE98cQTNY4XcyU7duzQ8OHDne9nz54tSZo8ebLS09M1Z84cXbhwQQ899JC++eYb9e/fX++//z5jzAAAACeXxplxOBy69dZbtW/fPp05c0axsbEqKCjQwIEDtW7dOrVq1coTtbqEcWYAADCPx8eZCQsL0+bNm/XRRx8pOztbFRUV6tWrl0aOHOlSwQAAAK6qd5gpKytTUFCQdu3apREjRmjEiBGeqAsAAKBO6v00U0BAgOLj41VeXu6JegAAAOrFpUezn3rqKc2dO9c58i8AAIC3uNRn5ne/+50OHz6s2NhYxcfHV+vwm52d7ZbiAAAArsalMDN+/HjZbDa5OOE2AACA29QrzJw/f17//d//rbVr1+rixYu6+eab9eKLL6pt27aeqg8AAKBW9eozM3/+fKWnp+u2227TXXfdpQ8++EAPPvigp2oDAAC4qnpdmVm9erX+8Ic/aNKkSZKke+65RzfeeKPKy8vl7+/vkQIBAABqU68rM3l5eRo8eLDzfb9+/RQQEKBjx465vTAAAIC6qFeYKS8vV8uWLausCwgIUFlZmVuLAgAAqKt63WayLEtTpkxRYGCgc11xcbFmzJhR5fHs1atXu69CAACAWtQrzEyePLnaunvvvddtxQAAANRXvcLMypUrPVUHAACAS1yazgAAAMBXEGYAAIDRCDMAAMBohBkAAGA0wgwAADAaYQYAABiNMAMAAIxGmAEAAEYjzAAAAKMRZgAAgNEIMwAAwGiEGQAAYDTCDAAAMBphBgAAGI0wAwAAjEaYAQAARiPMAAAAoxFmAACA0QgzAADAaIQZAABgNMIMAAAwGmEGAAAYjTADAACMRpgBAABGI8wAAACjEWYAAIDRCDMAAMBohBkAAGA0wgwAADAaYQYAABiNMAMAAIxGmAEAAEYjzAAAAKMRZgAAgNEIMwAAwGiEGQAAYDTCDAAAMBphBgAAGI0wAwAAjEaYAQAARiPMAAAAoxFmAACA0QgzAADAaIQZAABgNMIMAAAwGmEGAAAYjTADAACMRpgBAABGI8wAAACj+XSYSU1Nlc1mq7JER0d7uywAAOBDArxdwNXccMMN+uCDD5zv/f39vVgNAADwNT4fZgICArgaAwAArsinbzNJUk5OjmJjY5WQkKBJkybpyJEjte5fUlIih8NRZQEAAE2XT4eZ/v376/XXX9c//vEPLV++XAUFBRo0aJBOnjx5xc+kpaXJbrc7l7i4uEasGAAANDabZVmWt4uoq3PnzqlDhw6aM2eOZs+eXeM+JSUlKikpcb53OByKi4tTUVGRwsLCGqtUAADQAA6HQ3a7vU6/3z7fZ+bbWrVqpW7duiknJ+eK+wQGBiowMLARqwIAAN7k07eZLldSUqIDBw4oJibG26UAAAAf4dNh5rHHHlNWVpZyc3P1z3/+Uz/4wQ/kcDg0efJkb5cGAAB8hE/fZvrqq69011136cSJE4qIiNCAAQO0detWxcfHe7s0AADgI3w6zLz55pveLgEAAPg4n77NBAAAcDWEGQAAYDTCDAAAMBphBgAAGI0wAwAAjEaYAQAARiPMAAAAoxFmAACA0QgzAADAaIQZAABgNMIMAAAwGmEGAAAYjTADAACMRpgBAABGI8wAAACjEWYAAIDRCDMAAMBohBkAAGA0wgwAADAaYQYAABiNMAMAAIxGmAEAAEYjzAAAAKMRZgAAgNEIMwAAwGiEGQAAYDTCDAAAMBphBgAAGI0wAwAAjEaYAQAARiPMAAAAoxFmAACA0QgzAADAaIQZAABgNMIMAAAwGmEGAAAYjTADAACMRpgBAABGI8wAAACjEWYAAIDRCDMAAMBohBkAAGA0wgwAADAaYQYAABiNMAMAAIxGmAEAAEYjzAAAAKMRZgAAgNEIMwAAwGiEGQAAYDTCDAAAMBphBgAAGI0wAwAAjEaYAQAARiPMAAAAoxFmAACA0QgzAADAaIQZAABgNMIMAAAwGmEGAAAYjTADAACMRpgBAABGC/B2AaY6dbZUk36/RYVnShUZ2lJvPjBI4a1berssAACaHSOuzCxbtkwJCQkKCgpS7969tWnTJq/W03dBpnotyNShwnM6feGiDhWeU68Fmeq7INOrdQEA0Bz5fJjJyMhQSkqKnnzySX322WcaPHiwxowZo6NHj3qlnr4LMnX8bGmN246fLSXQAADQyHw+zCxevFhTp07VtGnT1LlzZ73wwguKi4vTSy+91Oi1nDpbesUgU+n42VKduso+AADAfXw6zJSWlmrnzp1KSkqqsj4pKUlbtmyp8TMlJSVyOBxVFneZ9Puaj+nqfgAAoOF8OsycOHFC5eXlioqKqrI+KipKBQUFNX4mLS1NdrvducTFxbmtnsIzdbviUtf9AABAw/l0mKlks9mqvLcsq9q6SnPnzlVRUZFzycvLc1sdkaF1e1qprvsBAICG8+kw07ZtW/n7+1e7ClNYWFjtak2lwMBAhYWFVVnc5c0HBrl1PwAA0HA+HWZatmyp3r17KzOz6hNCmZmZGjSo8QNDeOuWirjKWDIRrVsy3gwAAI3Ip8OMJM2ePVuvvvqqVqxYoQMHDugnP/mJjh49qhkzZnilnu1PjbpioIlo3VLbnxrVyBUBANC8+fwIwBMnTtTJkyf1i1/8Qvn5+eratavWrVun+Ph4r9W0/alRjAAMAICPsFmWZXm7CE9yOByy2+0qKipya/8ZAADgOfX5/fb520wAAAC1IcwAAACjEWYAAIDRCDMAAMBohBkAAGA0wgwAADAaYQYAABiNMAMAAIxGmAEAAEbz+ekMGqpygGOHw+HlSgAAQF1V/m7XZaKCJh9mzpw5I0mKi4vzciUAAKC+zpw5I7vdXus+TX5upoqKCh07dkyhoaGy2WxubdvhcCguLk55eXnM++RlfBe+g+/Ct/B9+A6+i/qxLEtnzpxRbGys/Pxq7xXT5K/M+Pn5qV27dh49RlhYGP8xfQTfhe/gu/AtfB++g++i7q52RaYSHYABAIDRCDMAAMBohJkGCAwM1Pz58xUYGOjtUpo9vgvfwXfhW/g+fAffhec0+Q7AAACgaePKDAAAMBphBgAAGI0wAwAAjEaYAQAARiPMuGjZsmVKSEhQUFCQevfurU2bNnm7pGYpNTVVNputyhIdHe3tspqFjRs3auzYsYqNjZXNZtPatWurbLcsS6mpqYqNjVVwcLCGDRumffv2eafYJu5q38WUKVOqnScDBgzwTrFNXFpamvr27avQ0FBFRkZq/PjxOnjwYJV9ODfcjzDjgoyMDKWkpOjJJ5/UZ599psGDB2vMmDE6evSot0trlm644Qbl5+c7lz179ni7pGbh3Llz6t69u5YsWVLj9ueee06LFy/WkiVLtH37dkVHR2vUqFHO+dLgPlf7LiRp9OjRVc6TdevWNWKFzUdWVpYefvhhbd26VZmZmSorK1NSUpLOnTvn3IdzwwMs1Fu/fv2sGTNmVFnXqVMn64knnvBSRc3X/Pnzre7du3u7jGZPkrVmzRrn+4qKCis6Otp69tlnneuKi4stu91uvfzyy16osPm4/LuwLMuaPHmyNW7cOK/U09wVFhZakqysrCzLsjg3PIUrM/VUWlqqnTt3Kikpqcr6pKQkbdmyxUtVNW85OTmKjY1VQkKCJk2apCNHjni7pGYvNzdXBQUFVc6TwMBADR06lPPESzZs2KDIyEh17NhR999/vwoLC71dUrNQVFQkSQoPD5fEueEphJl6OnHihMrLyxUVFVVlfVRUlAoKCrxUVfPVv39/vf766/rHP/6h5cuXq6CgQIMGDdLJkye9XVqzVnkucJ74hjFjxuiPf/yjPvroIz3//PPavn27RowYoZKSEm+X1qRZlqXZs2frpptuUteuXSVxbnhKk58121NsNluV95ZlVVsHzxszZozzdbdu3TRw4EB16NBBr732mmbPnu3FyiBxnviKiRMnOl937dpVffr0UXx8vP72t78pOTnZi5U1bTNnztTu3bu1efPmats4N9yLKzP11LZtW/n7+1dL0IWFhdWSNhpfq1at1K1bN+Xk5Hi7lGat8okyzhPfFBMTo/j4eM4TD5o1a5beffddffzxx2rXrp1zPeeGZxBm6qlly5bq3bu3MjMzq6zPzMzUoEGDvFQVKpWUlOjAgQOKiYnxdinNWkJCgqKjo6ucJ6WlpcrKyuI88QEnT55UXl4e54kHWJalmTNnavXq1froo4+UkJBQZTvnhmdwm8kFs2fP1g9/+EP16dNHAwcO1O9//3sdPXpUM2bM8HZpzc5jjz2msWPHqn379iosLNSCBQvkcDg0efJkb5fW5J09e1aHDx92vs/NzdWuXbsUHh6u9u3bKyUlRQsXLlRiYqISExO1cOFChYSE6O677/Zi1U1Tbd9FeHi4UlNTdeeddyomJkZffPGF5s2bp7Zt2+qOO+7wYtVN08MPP6xVq1bpnXfeUWhoqPMKjN1uV3BwsGw2G+eGJ3j1WSqDLV261IqPj7datmxp9erVy/nYHRrXxIkTrZiYGKtFixZWbGyslZycbO3bt8/bZTULH3/8sSWp2jJ58mTLsi49gjp//nwrOjraCgwMtIYMGWLt2bPHu0U3UbV9F+fPn7eSkpKsiIgIq0WLFlb79u2tyZMnW0ePHvV22U1STd+DJGvlypXOfTg33M9mWZbV+BEKAADAPegzAwAAjEaYAQAARiPMAAAAoxFmAACA0QgzAADAaIQZAABgNMIMAAAwGmEGAAAYjTADoFkZNmyYUlJSvF0GADcizAColc1mq3WZMmVKo9QxduxYjRw5ssZtn376qWw2m7KzsxulFgC+hYkmAdQqPz/f+TojI0M/+9nPdPDgQee64ODgKvtfvHhRLVq0cHsdU6dOVXJysr788kvFx8dX2bZixQr16NFDvXr1cvtxAfg+rswAqFV0dLRzsdvtstlszvfFxcVq06aN3nrrLQ0bNkxBQUF64403lJqaqh49elRp54UXXtB1111XZd3KlSvVuXNnBQUFqVOnTlq2bNkV67j99tsVGRmp9PT0KuvPnz+vjIwMTZ06VSdPntRdd92ldu3aKSQkRN26ddOf/vSnWv8+m82mtWvXVlnXpk2bKsf5+uuvNXHiRF1zzTX6zne+o3HjxumLL75wbt+wYYP69eunVq1aqU2bNrrxxhv15Zdf1npcAO5DmAHQYI8//rgeeeQRHThwQLfcckudPrN8+XI9+eST+uUvf6kDBw5o4cKFevrpp/Xaa6/VuH9AQIB+9KMfKT09Xd+eH/ftt99WaWmp7rnnHhUXF6t379567733tHfvXj3wwAP64Q9/qH/+858u/23nz5/X8OHD1bp1a23cuFGbN29W69atNXr0aJWWlqqsrEzjx4/X0KFDtXv3bn366ad64IEHZLPZXD4mgPrhNhOABktJSVFycnK9PvPMM8/o+eefd34uISFB+/fv1yuvvKLJkyfX+Jn77rtPv/rVr7RhwwYNHz5c0qVbTMnJybrmmmt0zTXX6LHHHnPuP2vWLK1fv15vv/22+vfv79Lf9uabb8rPz0+vvvqqM6CsXLlSbdq00YYNG9SnTx8VFRXp9ttvV4cOHSRJnTt3dulYAFxDmAHQYH369KnX/sePH1deXp6mTp2q+++/37m+rKxMdrv9ip/r1KmTBg0apBUrVmj48OH6/PPPtWnTJr3//vuSpPLycj377LPKyMjQ119/rZKSEpWUlKhVq1au/WGSdu7cqcOHDys0NLTK+uLiYn3++edKSkrSlClTdMstt2jUqFEaOXKkJkyYoJiYGJePCaB+CDMAGuzysODn51flVpB0qWNwpYqKCkmXbjVdfsXE39+/1mNNnTpVM2fO1NKlS7Vy5UrFx8fr5ptvliQ9//zz+s1vfqMXXnhB3bp1U6tWrZSSkqLS0tIrtmez2a5aa+/evfXHP/6x2mcjIiIkXbpS88gjj2j9+vXKyMjQU089pczMTA0YMKDWvwWAexBmALhdRESECgoKZFmW89bMrl27nNujoqJ07bXX6siRI7rnnnvq1faECRP06KOPatWqVXrttdd0//33O4+xadMmjRs3Tvfee6+kS0EkJyen1ts+ERERVZ7YysnJ0fnz553ve/XqpYyMDEVGRiosLOyK7fTs2VM9e/bU3LlzNXDgQK1atYowAzQSOgADcLthw4bp+PHjeu655/T5559r6dKl+vvf/15ln9TUVKWlpem3v/2tDh06pD179mjlypVavHhxrW23bt1aEydO1Lx583Ts2LEq49xcf/31yszM1JYtW3TgwAFNnz5dBQUFtbY3YsQILVmyRNnZ2dqxY4dmzJhR5dHye+65R23bttW4ceO0adMm5ebmKisrS48++qi++uor5ebmau7cufr000/15Zdf6v3339ehQ4foNwM0IsIMALfr3Lmzli1bpqVLl6p79+7atm1blY65kjRt2jS9+uqrSk9PV7du3TR06FClp6crISHhqu1PnTpV33zzjUaOHKn27ds71z/99NPq1auXbrnlFg0bNkzR0dEaP358rW09//zziouL05AhQ3T33XfrscceU0hIiHN7SEiINm7cqPbt2ys5OVmdO3fWfffdpwsXLigsLEwhISH697//rTvvvFMdO3bUAw88oJkzZ2r69On1+0cD4DKbdfnNYgAAAINwZQYAABiNMAMAAIxGmAEAAEYjzAAAAKMRZgAAgNEIMwAAwGiEGQAAYDTCDAAAMBphBgAAGI0wAwAAjEaYAQAARvv/68XaCU8kyvAAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from sklearn.tree import DecisionTreeRegressor\n",
    "X = np.concatenate([adata.X, adata.obs[['REL', 'STAT5A', 'FOS', 'ETV6', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B']]], axis=1)\n",
    "y = cdist(adata[adata.obs['recipe'] == 'target'].X, adata.X).min(axis=0)\n",
    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)\n",
    "model = DecisionTreeRegressor()\n",
    "model.fit(X_train, y_train)\n",
    "y_pred = model.predict(X_test)\n",
    "results = {'model': model,\n",
    "           'r2_score': r2_score(y_test, y_pred), 'mean_squared_error': mean_squared_error(y_test, y_pred), 'mean_absolute_error': mean_absolute_error(y_test, y_pred)}\n",
    "# with open(os.path.join(output_directory, 'decision_tree_metrics.txt'), 'w') as f: f.write(str(results))\n",
    "    \n",
    "plt.figure()\n",
    "plt.scatter(y_test, y_pred)\n",
    "plt.xlabel('True Values')\n",
    "plt.ylabel('Predictions')\n",
    "# plt.savefig(os.path.join(output_directory, 'decision_tree_scatter.png'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88b03790-d18c-4be2-a298-2a61058091db",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-01 22:21:06,811 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,814 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,814 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,814 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,815 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,815 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,815 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,816 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,816 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,817 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,817 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,817 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,818 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,818 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,818 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,819 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,819 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,820 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,820 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,820 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,820 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,821 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,821 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,821 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,822 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,822 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,822 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,823 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,823 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,824 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,824 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,825 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,825 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,825 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,826 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,826 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,827 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,827 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,827 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,827 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,828 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,828 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,829 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,829 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,830 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,830 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,830 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,831 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,831 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,831 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,831 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,832 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,832 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,833 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,833 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,833 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,834 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,834 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,835 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,835 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,835 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,835 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,836 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,836 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,836 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,837 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,837 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,837 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,838 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,838 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,839 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,839 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,839 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,840 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,840 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,840 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,841 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,841 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,842 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,842 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,842 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,843 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,843 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,843 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,844 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,844 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,845 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,845 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,845 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,846 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,846 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,846 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,846 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,847 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,847 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,848 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,848 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,848 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,849 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,849 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,850 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,850 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,850 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,851 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,851 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,852 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,852 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,852 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,853 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,853 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,853 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,854 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,854 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,854 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,854 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,855 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,855 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,856 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,856 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,856 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,857 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,857 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,858 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,858 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,858 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,859 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,859 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,859 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,860 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,860 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,860 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,861 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,861 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,862 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,862 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,862 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,863 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,863 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,864 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,864 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,864 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,865 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,865 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,865 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,866 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,866 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,866 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,867 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,867 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,867 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,868 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,868 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,869 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,869 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,869 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,870 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,870 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,870 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,871 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,871 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,871 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,871 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,872 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,872 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,873 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,873 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,873 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,874 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,874 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,875 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,875 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,875 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,876 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,876 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,876 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,877 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,877 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,878 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,878 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,878 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,879 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,879 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,879 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,880 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,880 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,880 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,881 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,881 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,881 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,882 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,882 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,883 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,883 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,883 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,883 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,884 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,884 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,885 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,885 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,885 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,886 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,886 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,886 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,887 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,887 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,888 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,888 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,888 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,889 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,889 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,889 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,890 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,890 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,891 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,891 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,891 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,891 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,892 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,892 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,892 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,893 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,893 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,893 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,894 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,894 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,894 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,894 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,895 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,896 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,896 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,896 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,897 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,897 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,897 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,898 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,898 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,898 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,899 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,899 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,899 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,900 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,900 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,901 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,901 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,901 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,901 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,902 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,902 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,903 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,903 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,903 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,904 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,904 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,904 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,905 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,905 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,906 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,906 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,906 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,907 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,907 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,907 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,908 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,908 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,908 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,909 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,909 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,909 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,910 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,910 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,911 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,911 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,911 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,911 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,912 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,912 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,913 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,913 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,913 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,914 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,914 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,914 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,915 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,915 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,915 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,916 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,916 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,916 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,917 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,917 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,917 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,918 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,918 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,919 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,919 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,919 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,920 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,920 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,920 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,920 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,921 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,922 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,922 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,922 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,922 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,923 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,923 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,923 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,924 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,924 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,924 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,925 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,925 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,926 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,926 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,927 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,927 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,927 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,928 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,928 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,928 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,929 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,929 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,930 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,930 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,930 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,931 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,931 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,931 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,932 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,932 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,932 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,933 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,933 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,933 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,934 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,934 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,934 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,935 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,935 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,936 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,936 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,936 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,937 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,937 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,937 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,938 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,938 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,938 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,939 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,939 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,939 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,940 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,940 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,941 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,941 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,941 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,942 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,942 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,942 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,943 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,943 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,943 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,944 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,944 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,944 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,945 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,945 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,946 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,946 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,946 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,946 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,947 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,947 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,948 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,948 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,948 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,949 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,949 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,949 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,950 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,950 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,951 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,951 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,951 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,952 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,952 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,952 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,953 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,953 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,953 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,954 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,954 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,955 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,955 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,955 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,955 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,956 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,956 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,957 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,957 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,957 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,958 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,958 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,958 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,959 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,959 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,959 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,960 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,960 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,960 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,961 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,961 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,962 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,962 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,962 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,963 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,963 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,963 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,964 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,964 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,964 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,965 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,965 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,966 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,966 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,966 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,967 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,967 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,967 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,968 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,968 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,968 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,969 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,969 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,969 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,970 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,970 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,970 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,971 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,971 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,971 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,972 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,972 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,972 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,972 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,973 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,974 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,974 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,974 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,975 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,975 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,975 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,976 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,976 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,977 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,977 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,977 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,978 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,978 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,978 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,978 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,979 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,979 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,980 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,980 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,980 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,981 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,981 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,981 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,982 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,982 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,983 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,983 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,983 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,984 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,984 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,984 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,985 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,985 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,985 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,986 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,986 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,986 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,987 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,987 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,988 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,988 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,988 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,989 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,989 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,989 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,990 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,990 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,990 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,991 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,991 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,991 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,992 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,992 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,992 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,993 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,993 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,993 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,994 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,994 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,995 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,995 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,995 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,996 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,996 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,996 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,997 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,997 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,997 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,998 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,998 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,998 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,999 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:06,999 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,000 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,000 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,000 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,001 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,001 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,001 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,002 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,002 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,002 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,003 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,003 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,003 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,004 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,004 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,005 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,005 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,005 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,006 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,006 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,006 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,007 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,007 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,007 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,008 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,008 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,009 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,009 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,009 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,010 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,010 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,010 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,011 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,011 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,011 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,012 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,012 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,012 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,013 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,013 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,013 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,014 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,014 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,015 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,015 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,015 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,015 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,016 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,016 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,017 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,017 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,017 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,018 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,018 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,018 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,019 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,019 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,019 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,020 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,020 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,021 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,021 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,021 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,022 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,022 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,023 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,023 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,023 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,023 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,024 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,024 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,024 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,025 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,025 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,026 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,026 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,026 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,027 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,027 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,027 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,028 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,028 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,028 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,028 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,029 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,029 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,029 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,029 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,030 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,030 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,030 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,031 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,031 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,031 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,031 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,032 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,033 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,034 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,034 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,034 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,035 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,035 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,035 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,036 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,036 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,037 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,037 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,037 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,038 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,038 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,038 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,039 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,039 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,039 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,040 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,040 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,041 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,041 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,041 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,041 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,042 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,042 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,042 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,043 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,043 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,044 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,044 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,044 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,045 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,045 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,045 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,046 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,046 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,047 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,047 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,047 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,048 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,048 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,048 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,049 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,049 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,049 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,050 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,050 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,051 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,051 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,051 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,051 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,052 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,052 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,052 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,052 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,053 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,053 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,053 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,054 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,055 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,055 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,055 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,056 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,056 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,056 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,057 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,057 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,057 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,058 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,058 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,059 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,059 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,059 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,060 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,060 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,061 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,061 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,061 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,061 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,062 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,062 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,063 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,063 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,063 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,064 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,064 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,064 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,065 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,065 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,065 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,066 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,066 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,067 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,067 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,067 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,067 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,068 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,068 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,069 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,069 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,069 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,069 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,070 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,070 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,071 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,071 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,071 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,072 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,072 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,073 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,073 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,073 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,074 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,074 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,074 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,075 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,075 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,076 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,076 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,076 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,077 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,077 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,077 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,077 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,078 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,078 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,078 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,079 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,079 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,080 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,080 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,080 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,081 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,081 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,082 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,082 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,082 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,083 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,083 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,083 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,084 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,084 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,085 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,085 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,085 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,085 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,086 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,086 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,087 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,087 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,087 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,088 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,088 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,088 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,089 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,089 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,090 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,090 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,090 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,091 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,091 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,092 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,092 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,092 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,092 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,093 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,093 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,093 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,093 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,094 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,094 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,095 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,095 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,096 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,096 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,096 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,097 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,097 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,098 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,098 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,098 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,099 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,099 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,099 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,100 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,100 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,100 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,101 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,101 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,101 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,102 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,102 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,103 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,103 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,103 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,104 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,104 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,104 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,104 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,105 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,105 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,106 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,106 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,106 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,107 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,107 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,107 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,108 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,108 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,109 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,109 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,109 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,109 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,110 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,111 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,111 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,111 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,112 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,112 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,112 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,113 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,113 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,113 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,114 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,114 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,115 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,115 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,115 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,116 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,116 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,116 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,117 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,117 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,118 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,118 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,118 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,119 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,119 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,119 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,120 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,120 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,120 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,121 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,121 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,121 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,121 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,122 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,122 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,122 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,123 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,123 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,123 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,124 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,125 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,125 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,125 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,125 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,126 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,126 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,127 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,127 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,127 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,128 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,128 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,129 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,129 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,129 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,130 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,130 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,130 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,131 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,131 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,131 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,132 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,132 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,132 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,132 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,133 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,134 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,134 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,134 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,135 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,135 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,135 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,136 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,136 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,137 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,137 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,137 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,138 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,138 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,138 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,138 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,139 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,139 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,139 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,140 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,140 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,141 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,141 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,141 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,142 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,142 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,143 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,143 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,143 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,144 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,144 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,144 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,144 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,145 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,145 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,145 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,145 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,146 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,147 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,147 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,148 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,148 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,148 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,148 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,149 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,149 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,149 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,149 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,150 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,150 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,150 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,151 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,152 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,152 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,152 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,153 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,153 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,154 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,154 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,154 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,154 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,155 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,155 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,156 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,156 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,156 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,157 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,157 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,158 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,158 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,158 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,159 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,159 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,159 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,160 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,160 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,160 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,161 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,161 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,162 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,162 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,162 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,162 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,163 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,163 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,164 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,164 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,164 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,165 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,165 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,165 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,166 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,166 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,167 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,167 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,167 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,168 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,168 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,168 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,169 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,169 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,170 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,170 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,170 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,171 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,171 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,171 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,172 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,172 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,172 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,173 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,173 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,173 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,174 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,174 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,175 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,175 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,175 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,176 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,176 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,176 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,177 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,177 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,177 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,178 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,178 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,179 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,179 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,179 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,179 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,180 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,180 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,181 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,181 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,181 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,182 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,182 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,182 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,183 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,183 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,184 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,184 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,184 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,184 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,185 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,185 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,186 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,186 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,187 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,187 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,187 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,188 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,188 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,188 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,189 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,189 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,189 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,190 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,190 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,191 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,191 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,191 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,192 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,192 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,192 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,193 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,193 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,194 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,194 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,194 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,195 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,195 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,195 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,196 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,196 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,196 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,197 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,197 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,198 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,198 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,198 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,199 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,199 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,199 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,200 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,200 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,201 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,201 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,201 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,202 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,202 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,202 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,203 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,203 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,203 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,204 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,204 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,204 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,205 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,205 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,206 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,206 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,206 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,206 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,207 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,207 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,207 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,207 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,208 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,209 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,209 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,209 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,210 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,210 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,210 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,211 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,211 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,211 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,212 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,212 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,213 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,213 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,213 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,214 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,214 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,214 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,215 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,215 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,215 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,216 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,216 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,216 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,217 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,217 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,218 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,218 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,218 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,218 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,219 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,219 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,220 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,220 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,220 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,220 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,221 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,221 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,222 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,222 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,222 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,223 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,223 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,224 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,224 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,224 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,225 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,225 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,226 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,226 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,226 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,227 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,227 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,227 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,228 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,228 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,228 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,228 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,229 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,229 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,230 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,230 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,230 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,231 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,231 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,232 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,232 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,232 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,233 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,233 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,233 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,234 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,234 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,235 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,235 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,235 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,235 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,236 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,236 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,236 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,237 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,237 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,238 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,238 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,238 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,239 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,239 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,240 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,240 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,240 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,240 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,241 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,241 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,242 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,242 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,242 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,242 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,243 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,243 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,244 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,244 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,245 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,245 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,245 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,246 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,246 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,246 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,246 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,247 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,247 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,247 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,248 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,248 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,249 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,249 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,250 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,250 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,250 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,251 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,251 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,251 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,252 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,252 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,252 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,253 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,253 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,253 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,254 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,254 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,255 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,255 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,255 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,255 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,256 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,256 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,256 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,257 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,257 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,258 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,258 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,258 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,259 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,259 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,259 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,260 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,260 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,261 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,261 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,261 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,262 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,262 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,262 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,263 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,263 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,263 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,264 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,264 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,265 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,265 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,265 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,266 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,266 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,266 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,266 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,267 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,268 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,268 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,268 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,269 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,269 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,269 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,269 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,270 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,270 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,271 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,271 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,271 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,272 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,272 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,272 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,273 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,273 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,273 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,274 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,274 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,274 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,275 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,275 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,276 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,276 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,277 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,277 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,277 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,278 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,278 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,278 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,279 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,279 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,279 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,280 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,280 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,281 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,281 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,281 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,282 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,282 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,282 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,282 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,283 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,283 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,284 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,284 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,285 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,285 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,285 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,285 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,286 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,286 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,287 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,287 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,287 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,288 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,288 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,289 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,289 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,289 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,289 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,290 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,290 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,291 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,291 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,291 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,292 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,292 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,292 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,293 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,293 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,293 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,294 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,294 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,295 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,295 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,295 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,296 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,296 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,296 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,297 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,297 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,298 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,298 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,298 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,299 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,299 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,299 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,300 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,300 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,300 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,301 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,301 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,301 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,302 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,302 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,302 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,303 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,303 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,304 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,304 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,304 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,304 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,305 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,305 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,306 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,306 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,306 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,307 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,307 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,308 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,308 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,308 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,309 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,309 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,309 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,309 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,310 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,310 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,311 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,311 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,311 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,312 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,312 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,313 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,313 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,313 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,314 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,314 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,314 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,315 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,315 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,315 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,316 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,316 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,316 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,317 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,317 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,317 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,318 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,318 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,318 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,319 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,319 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,319 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,320 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,320 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,321 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,321 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,321 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,322 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,322 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,322 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,323 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,323 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,324 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,324 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,324 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,325 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,325 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,325 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,326 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,326 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,326 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,327 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,327 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,328 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,328 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,328 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,329 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,329 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,329 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,329 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,330 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,330 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,330 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,331 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,331 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,332 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,332 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,332 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,333 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,333 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,334 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,334 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,334 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,334 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,335 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,335 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,336 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,336 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,337 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,337 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,337 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,338 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,338 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,338 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,339 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,339 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,339 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,340 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,340 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,341 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,341 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,341 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,342 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,342 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,342 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,343 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,343 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,343 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,344 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,344 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,344 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,345 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,345 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,346 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,346 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,346 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,347 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,347 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,347 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,348 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,348 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,348 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,349 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,349 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,350 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,350 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,350 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,350 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,351 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,351 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,352 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,352 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,352 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,353 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,353 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,353 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,354 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,354 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,355 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,355 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,355 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,356 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,356 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,356 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,357 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,357 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,358 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,358 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,358 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,359 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,359 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,359 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,360 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,360 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,360 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,360 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,361 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,361 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,361 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,362 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,363 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,363 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,363 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,364 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,364 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,364 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,365 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,365 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,365 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,366 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,366 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,366 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,367 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,367 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,367 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,368 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,368 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,369 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,369 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,369 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,370 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,370 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,370 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,371 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,371 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,371 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,371 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,372 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,372 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,372 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,373 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,373 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,374 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,374 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,375 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,375 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,375 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,376 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,376 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,376 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,377 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,377 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,378 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,378 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,378 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,378 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,379 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,379 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,379 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,380 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,380 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,381 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,381 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,382 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,382 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,382 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,383 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,383 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,383 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,384 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,384 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,385 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,385 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,385 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,386 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,386 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,386 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,387 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,387 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,387 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,388 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,388 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,389 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,389 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,389 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,389 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,390 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,390 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,391 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,391 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,391 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,392 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,392 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,392 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,393 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,393 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,393 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,394 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,394 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,395 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,395 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,395 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,395 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,396 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,396 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,396 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,397 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,397 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,398 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,398 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,398 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,399 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,399 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,399 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,400 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,400 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,401 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,401 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,401 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,401 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,402 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,402 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,403 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,403 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,404 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,404 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,404 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,404 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,405 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,405 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,405 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,405 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,406 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,406 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,407 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,407 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,408 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,408 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,408 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,408 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,409 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,409 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,410 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,410 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,410 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,411 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,411 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,412 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,412 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,412 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,412 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,413 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,413 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,414 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,414 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,414 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,415 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,415 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,416 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,416 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,416 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,417 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,417 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,417 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,418 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,418 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,418 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,419 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,419 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,420 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,420 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,420 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,421 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,421 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,421 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,422 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,422 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,423 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,423 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,423 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,423 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,424 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,424 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,424 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,425 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,425 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,426 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,426 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,426 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,426 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,427 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,427 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,427 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,428 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,428 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,428 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,428 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,429 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,430 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,430 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,431 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,431 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,431 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,431 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,432 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,432 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,433 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,433 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,433 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,434 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,434 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,435 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,435 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,435 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,436 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,436 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,436 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,437 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,437 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,437 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,438 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,438 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,439 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,439 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,439 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,440 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,440 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,440 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,441 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,441 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,441 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,442 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,442 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,442 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,443 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,443 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,443 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,443 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,444 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,445 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,445 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,445 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,446 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,446 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,447 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,447 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,447 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,447 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,448 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,448 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,448 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,449 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,449 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,449 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,450 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,450 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,450 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,451 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,451 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,452 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,452 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,452 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,453 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,453 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,453 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,454 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,454 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,455 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,455 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,455 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,456 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,456 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,456 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,457 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,457 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,458 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,458 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,458 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,459 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,459 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,459 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,460 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,460 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,461 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,461 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,461 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,462 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,462 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,462 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,463 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,463 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,463 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,464 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,464 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,464 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,465 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,465 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,465 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,466 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,466 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,467 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,467 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,467 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,468 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,468 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,469 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,469 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,469 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,470 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,470 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,470 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,470 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,471 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,471 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,471 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,471 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,472 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,473 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,473 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,473 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,474 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,474 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,474 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,475 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,475 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,476 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,476 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,476 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,477 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,477 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,477 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,478 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,478 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,478 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,479 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,479 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,479 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,480 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,480 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,480 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,481 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,481 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,482 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,482 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,482 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,483 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,483 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,483 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,484 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,484 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,485 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,485 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,485 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,485 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,486 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,486 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,486 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,487 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,487 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,488 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,488 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,488 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,488 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,489 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,489 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,489 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,490 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,491 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,491 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,491 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,492 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,492 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,492 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,493 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,493 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,493 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,494 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,494 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,494 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,494 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,495 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,496 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,496 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,496 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,496 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,497 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,497 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,498 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,498 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,498 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,499 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,499 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,500 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,500 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,500 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,501 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,501 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,501 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,502 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,502 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,502 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,503 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,503 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,503 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,504 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,504 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,504 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,505 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,505 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,506 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,506 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,506 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,507 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,507 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,507 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,508 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,508 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,508 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,509 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,509 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,510 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,510 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,510 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,511 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,511 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,511 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,512 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,512 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,512 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,513 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,513 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,514 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,514 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,514 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,515 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,515 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,515 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,515 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,516 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,516 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,517 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,517 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,517 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,518 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,518 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,519 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,519 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,519 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,519 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,520 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,520 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,521 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,521 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,521 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,522 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,522 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,523 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,523 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,523 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,523 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,524 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,524 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,524 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,525 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,525 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,526 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,526 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,526 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,526 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,527 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,527 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,528 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,528 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,528 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,529 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,529 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,529 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,530 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,530 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,530 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,531 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,531 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,532 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,532 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,532 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,533 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,533 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,533 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,534 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,534 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,535 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,535 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,535 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,536 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,536 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,536 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,537 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,537 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,537 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,538 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,538 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,539 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,539 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,539 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,539 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,540 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,540 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,540 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,541 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,541 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,541 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,541 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,542 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,543 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,543 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,544 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,544 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,544 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,545 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,545 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,545 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,546 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,546 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,546 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,547 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,547 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,547 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,548 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,548 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,549 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,549 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,549 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,550 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,550 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,550 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,551 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,551 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,551 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,551 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,552 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,552 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,552 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,553 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,553 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,553 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,553 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,554 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,555 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,555 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,556 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,556 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,556 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,557 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,557 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,557 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,558 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,558 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,558 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,558 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,559 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,560 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,560 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,560 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,561 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,561 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,561 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,562 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,562 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,562 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,563 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,563 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,564 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,564 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,564 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,565 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,565 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,565 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,565 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,566 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,566 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,567 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,567 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,567 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,568 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,568 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,569 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,569 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,569 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,570 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,570 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,570 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,571 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,571 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,571 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,572 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,572 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,572 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,573 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,573 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,574 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,574 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,574 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,574 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,575 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,575 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,575 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,576 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,576 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,577 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,577 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,577 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,578 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,578 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,578 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,578 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,579 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,580 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,580 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,580 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,581 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,581 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,581 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,582 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,582 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,582 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,583 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,583 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,584 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,584 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,584 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,585 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,585 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,585 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,586 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,586 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,586 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,587 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,587 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,588 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,588 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,588 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,589 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,589 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,589 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,590 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,590 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,590 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,591 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,591 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,591 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,591 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,592 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,592 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,592 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,592 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,593 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,594 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,594 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,595 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,595 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,595 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,596 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,596 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,596 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,597 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,597 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,598 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,598 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,598 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,599 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,599 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,599 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,600 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,600 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,600 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,601 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,601 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,601 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,602 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,602 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,602 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,602 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,603 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,603 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,603 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,604 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,605 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,605 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,605 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,606 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,606 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,607 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,607 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,607 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,608 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,608 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,608 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,609 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,609 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,609 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,610 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,610 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,610 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,611 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,611 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,611 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,612 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,612 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,613 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,613 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,613 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,614 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,614 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,614 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,615 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,615 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n",
      "2024-08-01 22:21:07,616 - INFO - Fontsize 0.03 < 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt\n"
     ]
    }
   ],
   "source": [
    "from sklearn.tree import plot_tree\n",
    "plt.figure(figsize=(20,10))\n",
    "plot_tree(model, filled=True, rounded=True)\n",
    "# plt.savefig(os.path.join(output_directory, 'decision_tree_visualization.png'))\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "5dc934d2-80fc-4d32-92ee-ce3c2af07e93",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([22.17468851, 18.70258825, 21.12446492, ..., 20.37118612,\n",
       "       18.56246956, 21.2590161 ])"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f34940ee-9b8d-455c-a8ce-1ad9309e5a07",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_X = adata.X\n",
    "mask = ~adata.obs['recipe'].isin(['source', 'target'])\n",
    "recipe_embeddings = adata[mask]\n",
    "source_embeddings = adata[adata.obs['recipe'] == 'source']\n",
    "target_embeddings = adata[adata.obs['recipe'] == 'target']\n",
    "recipe_distances_source = cdist(recipe_embeddings.X, source_embeddings.X)\n",
    "recipe_distances_target = cdist(recipe_embeddings.X, target_embeddings.X)\n",
    "adata_dist = sc.AnnData(X=np.concatenate((recipe_distances_source, recipe_distances_target), axis=1), var=pd.concat((source_embeddings.obs, target_embeddings.obs)), obs=recipe_embeddings.obs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "09f60911-f7be-4b95-9c77-8967eb56db6a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(6325, 50)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.concatenate((recipe_distances_source, recipe_distances_target), axis=1).shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "83dbe60c-4621-468b-bd2e-cc2b9ec29ddb",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_dist = sc.AnnData(X=np.concatenate((recipe_distances_source, recipe_distances_target), axis=1), var=pd.concat((source_embeddings.obs, target_embeddings.obs)), obs=recipe_embeddings.obs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "f943b8cf-df64-4a50-afe5-69e091fa839c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs Ã— n_vars = 6325 Ã— 50\n",
       "    obs: 'recipe', 'filename', 'REL', 'STAT5A', 'FOS', 'ETV6', 'source', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B', 'target'\n",
       "    var: 'recipe', 'filename', 'REL', 'STAT5A', 'FOS', 'ETV6', 'source', 'MEF2C', 'RUNX3', 'FOSB', 'GATA2', 'IKZF1', 'GFI1B', 'target'"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_dist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c68bc644-76d7-4de5-ab19-b5dd7e0306f1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(50, 14)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.concat((source_embeddings.obs, target_embeddings.obs)).shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "0102f4c8-c3ff-44ed-9d72-61c4b9c68dd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "source_embeddings = adata[adata.obs['recipe'] == 'source'].X\n",
    "target_embeddings = adata[adata.obs['recipe'] == 'target'].X\n",
    "mask = ~adata.obs['recipe'].isin(['source', 'target'])\n",
    "recipe_embeddings = adata[mask]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "c4f8d6a9-9a33-4472-b8e7-3e142bcacb39",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_895424/1752194696.py\u001b[0m in \u001b[0;36m?\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0msource_distances\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcdist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msource_embeddings\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrecipe_embeddings\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/.local/lib/python3.11/site-packages/scipy/spatial/distance.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(XA, XB, metric, out, **kwargs)\u001b[0m\n\u001b[1;32m   2972\u001b[0m     \u001b[0;31m# between all pairs of vectors in XA and XB using the distance metric 'abc'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2973\u001b[0m     \u001b[0;31m# but with a more succinct, verifiable, but less efficient implementation.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2974\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2975\u001b[0m     \u001b[0mXA\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0masarray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mXA\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2976\u001b[0;31m     \u001b[0mXB\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0masarray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mXB\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2977\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2978\u001b[0m     \u001b[0ms\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mXA\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2979\u001b[0m     \u001b[0msB\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mXB\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.11/site-packages/anndata/_core/anndata.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self, index)\u001b[0m\n\u001b[1;32m   1083\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__getitem__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mIndex\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0mAnnData\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1084\u001b[0m         \u001b[0;34m\"\"\"Returns a sliced view of the object.\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1085\u001b[0m         \u001b[0moidx\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvidx\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_normalize_indices\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1086\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mAnnData\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moidx\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0moidx\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvidx\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mvidx\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0masview\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/.local/lib/python3.11/site-packages/anndata/_core/anndata.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self, X, obs, var, uns, obsm, varm, layers, raw, dtype, shape, filename, filemode, asview, obsp, varp, oidx, vidx)\u001b[0m\n\u001b[1;32m    265\u001b[0m     ):\n\u001b[1;32m    266\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0masview\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    267\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mAnnData\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    268\u001b[0m                 \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"`X` has to be an AnnData object.\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 269\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_init_as_view\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moidx\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvidx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    270\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    271\u001b[0m             self._init_as_actual(\n\u001b[1;32m    272\u001b[0m                 \u001b[0mX\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mX\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.11/site-packages/anndata/_core/anndata.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self, adata_ref, oidx, vidx)\u001b[0m\n\u001b[1;32m    321\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_obsp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0madata_ref\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobsp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_view\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moidx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    322\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_varp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0madata_ref\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvarp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_view\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvidx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    323\u001b[0m         \u001b[0;31m# fix categories\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    324\u001b[0m         \u001b[0muns\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0madata_ref\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_uns\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 325\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_remove_unused_categories\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0madata_ref\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mobs_sub\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0muns\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    326\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_remove_unused_categories\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0madata_ref\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvar\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvar_sub\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0muns\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    327\u001b[0m         \u001b[0;31m# set attributes\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    328\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_obs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mDataFrameView\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobs_sub\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mview_args\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"obs\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.11/site-packages/anndata/_core/anndata.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self, df_full, df_sub, uns)\u001b[0m\n\u001b[1;32m   1092\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf_full\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mCategoricalDtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1093\u001b[0m                 \u001b[0;32mcontinue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1094\u001b[0m             \u001b[0mall_categories\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf_full\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcat\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcategories\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1095\u001b[0m             \u001b[0;32mwith\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moption_context\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"mode.chained_assignment\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1096\u001b[0;31m                 \u001b[0mdf_sub\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf_sub\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcat\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mremove_unused_categories\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1097\u001b[0m             \u001b[0;31m# also correct the colors...\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1098\u001b[0m             \u001b[0mcolor_key\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34mf\"{k}_colors\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1099\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mcolor_key\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0muns\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/sw/pkgs/arc/python3.11-anaconda/2024.02-1/lib/python3.11/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self, key, value)\u001b[0m\n\u001b[1;32m   4087\u001b[0m             \u001b[0;31m# Column to set is duplicated\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4088\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_setitem_array\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4089\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4090\u001b[0m             \u001b[0;31m# set column\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 4091\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_set_item\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/sw/pkgs/arc/python3.11-anaconda/2024.02-1/lib/python3.11/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self, key, value)\u001b[0m\n\u001b[1;32m   4296\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4297\u001b[0m         \u001b[0mSeries\u001b[0m\u001b[0;34m/\u001b[0m\u001b[0mTimeSeries\u001b[0m \u001b[0mwill\u001b[0m \u001b[0mbe\u001b[0m \u001b[0mconformed\u001b[0m \u001b[0mto\u001b[0m \u001b[0mthe\u001b[0m \u001b[0mDataFrames\u001b[0m \u001b[0mindex\u001b[0m \u001b[0mto\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4298\u001b[0m         \u001b[0mensure\u001b[0m \u001b[0mhomogeneity\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4299\u001b[0m         \"\"\"\n\u001b[0;32m-> 4300\u001b[0;31m         \u001b[0mvalue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrefs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_sanitize_column\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   4301\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4302\u001b[0m         if (\n\u001b[1;32m   4303\u001b[0m             \u001b[0mkey\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/sw/pkgs/arc/python3.11-anaconda/2024.02-1/lib/python3.11/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self, value)\u001b[0m\n\u001b[1;32m   5032\u001b[0m         \u001b[0;32massert\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mDataFrame\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5033\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mis_dict_like\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5034\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mSeries\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5035\u001b[0m                 \u001b[0mvalue\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mSeries\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5036\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0m_reindex_for_setitem\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5037\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5038\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mis_list_like\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5039\u001b[0m             \u001b[0mcom\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrequire_length_match\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/sw/pkgs/arc/python3.11-anaconda/2024.02-1/lib/python3.11/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(value, index)\u001b[0m\n\u001b[1;32m  12296\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m  12297\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mequals\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m  12298\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0musing_copy_on_write\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mSeries\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m  12299\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_values\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_references\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m> 12300\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_values\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m  12301\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m  12302\u001b[0m     \u001b[0;31m# GH#4107\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m  12303\u001b[0m     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "source_distances = cdist(source_embeddings, recipe_embeddings)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "5f5f53f6-b4b6-4a0b-8c0d-4da4913c699b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "KeyboardInterrupt\n",
      "\n"
     ]
    }
   ],
   "source": [
    "source_distances = cdist(source_embeddings, recipe_embeddings)\n",
    "target_distances = cdist(target_embeddings, recipe_embeddings)\n",
    "pd.DataFrame(source_distances).to_csv(os.path.join(chatstatus['output-directory'], 'source_distances.csv'))\n",
    "pd.DataFrame(target_distances).to_csv(os.path.join(chatstatus['output-directory'], 'target_distances.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d5f4b1e-60ea-4b35-a8a8-55e8acdd3632",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame(source_distances)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
