{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "fc8ba19c-2b23-45ad-8e07-0b2834b9116b",
   "metadata": {},
   "source": [
    "# Benchmarking LLMs for Scraping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "cf7a1fd7-ce63-47b3-9076-be8dfd0c45ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "from BRAD import llms\n",
    "from BRAD import brad"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "899ce708-fcf7-4e43-8c62-d3064b754659",
   "metadata": {},
   "source": [
    "## Nvidia Hosted Models"
   ]
  },
  {
   "cell_type": "raw",
   "id": "873da4b3-cf0d-4dc4-99fe-463ddcd5a2e7",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "02d0c17f-0aac-48d4-ad1c-4225ea2a3524",
   "metadata": {},
   "outputs": [],
   "source": [
    "# nvidia llms\n",
    "nvidia_llms = [\n",
    "    'meta/llama-3.1-8b-instruct',\n",
    "    'meta/llama-3.1-405b-instruct',\n",
    "    'nvidia/nemotron-4-340b-instruct',\n",
    "    'google/gemma-2-9b-it'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "379f4421-d5df-40ca-b7fc-91a228b00977",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "meta/llama-3.1-8b-instruct\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter your NVIDIA API key:  ········\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32m2024-07-23 16:39:35 INFO semantic_router.utils.logger local\u001b[0m\n",
      "2024-07-23 16:39:47,907 - INFO - SCRAPE\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome to RAG! The chat log from this conversation will be saved to /nfs/turbo/umms-indikar/shared/projects/BRAD/paper-experiments/webscrape-by-llm/pore-c/2024-07-23_16-39-34/log.json. How can I help?\n",
      "==================================================\n",
      "BRAD >> 0: \n",
      "\n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation:\n",
      "\n",
      "\n",
      "Query:what are the most relevant literature on arxiv regarding Pore-C?\n",
      "\n",
      "From the query, decide if ARXIV, PUBMED, or BIORXIV should be searched, and propose at least 10 search terms for this query and database. Separate each term with a comma, and provide no extra information/explination for either the database or search terms. Do not include general purpose search terms such as \"recent advances\", \"novel\", \"current research\", \"latest trends\", \"theoretical foundations\", or other general query terms. Also, do not use very general fields as serch terms such as: \"dynamical systems\", \"machine learning\", or \"single cell\". It is correct to use these terms with modifications to be more specific, so that \"biological dynamical systems\", \"machine learning for genomics\", or \"single cell clustering\" would be good terms to search.\n",
      "\n",
      "The following search terms have been previously uses and identical terms should be avoided in this search: []\n",
      "\n",
      "**Format your output as follows. Do not include any addition information:**\n",
      "Database: <ARXIV, PUBMED, or BIORXIV>\n",
      "Search Terms: <improved search terms>\n",
      "\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-07-23 16:39:52,253 - INFO - {'database': 'ARXIV', 'search_terms': ['Pore-C', 'Pore CNN', 'Adaptive Pore CNN', 'Hierarchical Porous Channel CNN', 'Hierarchical Porous Channel Attention CNN', 'Deep Pore-CNN', 'Pore Architectures', 'Graph-pore CNN', 'Topology-pore CNN', 'Mixture of Pores CNN', 'Pore-composite network', 'Porous Residual U-Net', 'Porous Attention U-Net']}\n",
      "2024-07-23 16:39:52,253 - INFO - \n",
      "\n",
      "\n",
      "route\n",
      "\n",
      "\n",
      "\n",
      "2024-07-23 16:39:52,254 - INFO - SCRAPE\n",
      "2024-07-23 16:39:52,254 - INFO - \n",
      "\n",
      "PLANNER:\n",
      "\n",
      "{'module': 'SCRAPE', 'steps': [{'llm': 'meta/llama-3.1-8b-instruct', 'memory': 'chat_memory=InMemoryChatMessageHistory(messages=[HumanMessage(content=\\'what are the most relevant literature on arxiv regarding Pore-C?\\'), AIMessage(content=\"ARXIV\\\\nPore-C, Pore CNN, Adaptive Pore CNN, Hierarchical Porous Channel CNN, Hierarchical Porous Channel Attention CNN, Deep Pore-CNN, Pore Architectures, Graph-pore CNN, Topology-pore CNN, Mixture of Pores CNN, Pore-composite network, Porous Residual U-Net, Porous Attention U-Net\\\\n\\\\nPUBMED\\\\nPore-C, Pore-C networks, CTC branching and anastomosing channels, Porous biofactory, Adaptable hierarchical pore-channel networks, Pore-composite networks, Progressive porous networks, Hierarchical pore-channel networks, Pore-granule based analysis, Pore multi-compartment networks, Porous topological models, Pore-forming toxins experimental applications\\\\n\\\\nBIORXIV\\\\nPore-channel networks, Hierarchical pore-channel architecture, Hierarchical pore-based analysis, Topology of pore-cellular models, Adaptable pore-channel networks, Progressive pore-based networks, Porous channel networks, Bio Cell and Pore CN, Pore channels and occluding junctions, Pore-based sub-compartments, Porous models for Imaging analysis of the cells\\' poles and Culcote cells\")]) ai_prefix=\\'BRAD\\'', 'prompt': 'input_variables=[\\'history\\', \\'input\\'] template=\\'Current conversation:\\\\n{history}\\\\n\\\\nQuery:{input}\\\\n\\\\nFrom the query, decide if ARXIV, PUBMED, or BIORXIV should be searched, and propose at least 10 search terms for this query and database. Separate each term with a comma, and provide no extra information/explination for either the database or search terms. Do not include general purpose search terms such as \"recent advances\", \"novel\", \"current research\", \"latest trends\", \"theoretical foundations\", or other general query terms. Also, do not use very general fields as serch terms such as: \"dynamical systems\", \"machine learning\", or \"single cell\". It is correct to use these terms with modifications to be more specific, so that \"biological dynamical systems\", \"machine learning for genomics\", or \"single cell clustering\" would be good terms to search.\\\\n\\\\nThe following search terms have been previously uses and identical terms should be avoided in this search: []\\\\n\\\\n**Format your output as follows. Do not include any addition information:**\\\\nDatabase: <ARXIV, PUBMED, or BIORXIV>\\\\nSearch Terms: <improved search terms>\\\\n\\'', 'input': 'what are the most relevant literature on arxiv regarding Pore-C?', 'output': \"ARXIV\\nPore-C, Pore CNN, Adaptive Pore CNN, Hierarchical Porous Channel CNN, Hierarchical Porous Channel Attention CNN, Deep Pore-CNN, Pore Architectures, Graph-pore CNN, Topology-pore CNN, Mixture of Pores CNN, Pore-composite network, Porous Residual U-Net, Porous Attention U-Net\\n\\nPUBMED\\nPore-C, Pore-C networks, CTC branching and anastomosing channels, Porous biofactory, Adaptable hierarchical pore-channel networks, Pore-composite networks, Progressive porous networks, Hierarchical pore-channel networks, Pore-granule based analysis, Pore multi-compartment networks, Porous topological models, Pore-forming toxins experimental applications\\n\\nBIORXIV\\nPore-channel networks, Hierarchical pore-channel architecture, Hierarchical pore-based analysis, Topology of pore-cellular models, Adaptable pore-channel networks, Progressive pore-based networks, Porous channel networks, Bio Cell and Pore CN, Pore channels and occluding junctions, Pore-based sub-compartments, Porous models for Imaging analysis of the cells' poles and Culcote cells\", 'parsedOutput': {'database': 'ARXIV', 'search_terms': ['Pore-C', 'Pore CNN', 'Adaptive Pore CNN', 'Hierarchical Porous Channel CNN', 'Hierarchical Porous Channel Attention CNN', 'Deep Pore-CNN', 'Pore Architectures', 'Graph-pore CNN', 'Topology-pore CNN', 'Mixture of Pores CNN', 'Pore-composite network', 'Porous Residual U-Net', 'Porous Attention U-Net']}, 'purpose': 'identify how to web scrape'}]}\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "meta/llama-3.1-405b-instruct\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32m2024-07-23 16:39:52 INFO semantic_router.utils.logger local\u001b[0m\n",
      "2024-07-23 16:39:56,276 - INFO - SCRAPE\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome to RAG! The chat log from this conversation will be saved to /nfs/turbo/umms-indikar/shared/projects/BRAD/paper-experiments/webscrape-by-llm/pore-c/2024-07-23_16-39-52/log.json. How can I help?\n",
      "==================================================\n",
      "BRAD >> 0: \n",
      "\n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation:\n",
      "\n",
      "\n",
      "Query:what are the most relevant literature on arxiv regarding Pore-C?\n",
      "\n",
      "From the query, decide if ARXIV, PUBMED, or BIORXIV should be searched, and propose at least 10 search terms for this query and database. Separate each term with a comma, and provide no extra information/explination for either the database or search terms. Do not include general purpose search terms such as \"recent advances\", \"novel\", \"current research\", \"latest trends\", \"theoretical foundations\", or other general query terms. Also, do not use very general fields as serch terms such as: \"dynamical systems\", \"machine learning\", or \"single cell\". It is correct to use these terms with modifications to be more specific, so that \"biological dynamical systems\", \"machine learning for genomics\", or \"single cell clustering\" would be good terms to search.\n",
      "\n",
      "The following search terms have been previously uses and identical terms should be avoided in this search: []\n",
      "\n",
      "**Format your output as follows. Do not include any addition information:**\n",
      "Database: <ARXIV, PUBMED, or BIORXIV>\n",
      "Search Terms: <improved search terms>\n",
      "\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-07-23 16:40:01,320 - INFO - {'database': 'ARXIV', 'search_terms': ['pore-C', 'Single-cell RNA sequencing', 'Spatial transcriptomics', '3D genome organization', 'genome structure', 'Chromatin architecture', 'Whole-genome modeling', 'Genome simulation', 'Chromatin Conformation Capture', 'Micro-C', 'genome-scale modeling', '3D nucleus modeling', 'genome organization simulations.']}\n",
      "2024-07-23 16:40:01,321 - INFO - \n",
      "\n",
      "\n",
      "route\n",
      "\n",
      "\n",
      "\n",
      "2024-07-23 16:40:01,321 - INFO - SCRAPE\n",
      "2024-07-23 16:40:01,321 - INFO - \n",
      "\n",
      "PLANNER:\n",
      "\n",
      "{'module': 'SCRAPE', 'steps': [{'llm': 'meta/llama-3.1-405b-instruct', 'memory': \"chat_memory=InMemoryChatMessageHistory(messages=[HumanMessage(content='what are the most relevant literature on arxiv regarding Pore-C?'), AIMessage(content='Database: ARXIV\\\\nSearch Terms: pore-C, Single-cell RNA sequencing, Spatial transcriptomics, 3D genome organization, genome structure, Chromatin architecture, Whole-genome modeling, Genome simulation, Chromatin Conformation Capture, Micro-C, genome-scale modeling, 3D nucleus modeling, genome organization simulations.')]) ai_prefix='BRAD'\", 'prompt': 'input_variables=[\\'history\\', \\'input\\'] template=\\'Current conversation:\\\\n{history}\\\\n\\\\nQuery:{input}\\\\n\\\\nFrom the query, decide if ARXIV, PUBMED, or BIORXIV should be searched, and propose at least 10 search terms for this query and database. Separate each term with a comma, and provide no extra information/explination for either the database or search terms. Do not include general purpose search terms such as \"recent advances\", \"novel\", \"current research\", \"latest trends\", \"theoretical foundations\", or other general query terms. Also, do not use very general fields as serch terms such as: \"dynamical systems\", \"machine learning\", or \"single cell\". It is correct to use these terms with modifications to be more specific, so that \"biological dynamical systems\", \"machine learning for genomics\", or \"single cell clustering\" would be good terms to search.\\\\n\\\\nThe following search terms have been previously uses and identical terms should be avoided in this search: []\\\\n\\\\n**Format your output as follows. Do not include any addition information:**\\\\nDatabase: <ARXIV, PUBMED, or BIORXIV>\\\\nSearch Terms: <improved search terms>\\\\n\\'', 'input': 'what are the most relevant literature on arxiv regarding Pore-C?', 'output': 'Database: ARXIV\\nSearch Terms: pore-C, Single-cell RNA sequencing, Spatial transcriptomics, 3D genome organization, genome structure, Chromatin architecture, Whole-genome modeling, Genome simulation, Chromatin Conformation Capture, Micro-C, genome-scale modeling, 3D nucleus modeling, genome organization simulations.', 'parsedOutput': {'database': 'ARXIV', 'search_terms': ['pore-C', 'Single-cell RNA sequencing', 'Spatial transcriptomics', '3D genome organization', 'genome structure', 'Chromatin architecture', 'Whole-genome modeling', 'Genome simulation', 'Chromatin Conformation Capture', 'Micro-C', 'genome-scale modeling', '3D nucleus modeling', 'genome organization simulations.']}, 'purpose': 'identify how to web scrape'}]}\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "nvidia/nemotron-4-340b-instruct\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32m2024-07-23 16:40:01 INFO semantic_router.utils.logger local\u001b[0m\n",
      "2024-07-23 16:40:05,779 - INFO - SCRAPE\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome to RAG! The chat log from this conversation will be saved to /nfs/turbo/umms-indikar/shared/projects/BRAD/paper-experiments/webscrape-by-llm/pore-c/2024-07-23_16-40-01/log.json. How can I help?\n",
      "==================================================\n",
      "BRAD >> 0: \n",
      "\n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation:\n",
      "\n",
      "\n",
      "Query:what are the most relevant literature on arxiv regarding Pore-C?\n",
      "\n",
      "From the query, decide if ARXIV, PUBMED, or BIORXIV should be searched, and propose at least 10 search terms for this query and database. Separate each term with a comma, and provide no extra information/explination for either the database or search terms. Do not include general purpose search terms such as \"recent advances\", \"novel\", \"current research\", \"latest trends\", \"theoretical foundations\", or other general query terms. Also, do not use very general fields as serch terms such as: \"dynamical systems\", \"machine learning\", or \"single cell\". It is correct to use these terms with modifications to be more specific, so that \"biological dynamical systems\", \"machine learning for genomics\", or \"single cell clustering\" would be good terms to search.\n",
      "\n",
      "The following search terms have been previously uses and identical terms should be avoided in this search: []\n",
      "\n",
      "**Format your output as follows. Do not include any addition information:**\n",
      "Database: <ARXIV, PUBMED, or BIORXIV>\n",
      "Search Terms: <improved search terms>\n",
      "\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-07-23 16:40:12,211 - INFO - {'database': 'ARXIV', 'search_terms': ['\"Pore-C\"', '\"genome architecture\"', '\"chromosome conformation\"', '\"nanopore sequencing\"', '\"3D genome structure\"', '\"long-range genomic interactions\"', '\"single-molecule sequencing\"', '\"spatial organization of chromatin\"', '\"genome folding\"', '\"Pore-C analysis\"']}\n",
      "2024-07-23 16:40:12,211 - INFO - \n",
      "\n",
      "\n",
      "route\n",
      "\n",
      "\n",
      "\n",
      "2024-07-23 16:40:12,212 - INFO - SCRAPE\n",
      "2024-07-23 16:40:12,212 - INFO - \n",
      "\n",
      "PLANNER:\n",
      "\n",
      "{'module': 'SCRAPE', 'steps': [{'llm': 'nvidia/nemotron-4-340b-instruct', 'memory': 'chat_memory=InMemoryChatMessageHistory(messages=[HumanMessage(content=\\'what are the most relevant literature on arxiv regarding Pore-C?\\'), AIMessage(content=\\'Database: ARXIV\\\\nSearch Terms: \"Pore-C\" , \"genome architecture\" , \"chromosome conformation\" , \"nanopore sequencing\" , \"3D genome structure\" , \"long-range genomic interactions\" , \"single-molecule sequencing\" , \"spatial organization of chromatin\" , \"genome folding\" , \"Pore-C analysis\"\\\\n\\')]) ai_prefix=\\'BRAD\\'', 'prompt': 'input_variables=[\\'history\\', \\'input\\'] template=\\'Current conversation:\\\\n{history}\\\\n\\\\nQuery:{input}\\\\n\\\\nFrom the query, decide if ARXIV, PUBMED, or BIORXIV should be searched, and propose at least 10 search terms for this query and database. Separate each term with a comma, and provide no extra information/explination for either the database or search terms. Do not include general purpose search terms such as \"recent advances\", \"novel\", \"current research\", \"latest trends\", \"theoretical foundations\", or other general query terms. Also, do not use very general fields as serch terms such as: \"dynamical systems\", \"machine learning\", or \"single cell\". It is correct to use these terms with modifications to be more specific, so that \"biological dynamical systems\", \"machine learning for genomics\", or \"single cell clustering\" would be good terms to search.\\\\n\\\\nThe following search terms have been previously uses and identical terms should be avoided in this search: []\\\\n\\\\n**Format your output as follows. Do not include any addition information:**\\\\nDatabase: <ARXIV, PUBMED, or BIORXIV>\\\\nSearch Terms: <improved search terms>\\\\n\\'', 'input': 'what are the most relevant literature on arxiv regarding Pore-C?', 'output': 'Database: ARXIV\\nSearch Terms: \"Pore-C\" , \"genome architecture\" , \"chromosome conformation\" , \"nanopore sequencing\" , \"3D genome structure\" , \"long-range genomic interactions\" , \"single-molecule sequencing\" , \"spatial organization of chromatin\" , \"genome folding\" , \"Pore-C analysis\"\\n', 'parsedOutput': {'database': 'ARXIV', 'search_terms': ['\"Pore-C\"', '\"genome architecture\"', '\"chromosome conformation\"', '\"nanopore sequencing\"', '\"3D genome structure\"', '\"long-range genomic interactions\"', '\"single-molecule sequencing\"', '\"spatial organization of chromatin\"', '\"genome folding\"', '\"Pore-C analysis\"']}, 'purpose': 'identify how to web scrape'}]}\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "google/gemma-2-9b-it\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32m2024-07-23 16:40:12 INFO semantic_router.utils.logger local\u001b[0m\n",
      "2024-07-23 16:40:19,346 - INFO - SCRAPE\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome to RAG! The chat log from this conversation will be saved to /nfs/turbo/umms-indikar/shared/projects/BRAD/paper-experiments/webscrape-by-llm/pore-c/2024-07-23_16-40-12/log.json. How can I help?\n",
      "==================================================\n",
      "BRAD >> 0: \n",
      "\n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation:\n",
      "\n",
      "\n",
      "Query:what are the most relevant literature on arxiv regarding Pore-C?\n",
      "\n",
      "From the query, decide if ARXIV, PUBMED, or BIORXIV should be searched, and propose at least 10 search terms for this query and database. Separate each term with a comma, and provide no extra information/explination for either the database or search terms. Do not include general purpose search terms such as \"recent advances\", \"novel\", \"current research\", \"latest trends\", \"theoretical foundations\", or other general query terms. Also, do not use very general fields as serch terms such as: \"dynamical systems\", \"machine learning\", or \"single cell\". It is correct to use these terms with modifications to be more specific, so that \"biological dynamical systems\", \"machine learning for genomics\", or \"single cell clustering\" would be good terms to search.\n",
      "\n",
      "The following search terms have been previously uses and identical terms should be avoided in this search: []\n",
      "\n",
      "**Format your output as follows. Do not include any addition information:**\n",
      "Database: <ARXIV, PUBMED, or BIORXIV>\n",
      "Search Terms: <improved search terms>\n",
      "\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-07-23 16:40:20,178 - INFO - {'database': 'ARXIV', 'search_terms': ['Pore-C', 'long-read sequencing', 'nanopore sequencing', 'single-molecule sequencing', 'DNA sequencing', 'RNA sequencing', 'genomics', 'transcriptomics', 'bioinformatics', 'computational biology']}\n",
      "2024-07-23 16:40:20,179 - INFO - \n",
      "\n",
      "\n",
      "route\n",
      "\n",
      "\n",
      "\n",
      "2024-07-23 16:40:20,179 - INFO - SCRAPE\n",
      "2024-07-23 16:40:20,179 - INFO - \n",
      "\n",
      "PLANNER:\n",
      "\n",
      "{'module': 'SCRAPE', 'steps': [{'llm': 'google/gemma-2-9b-it', 'memory': \"chat_memory=InMemoryChatMessageHistory(messages=[HumanMessage(content='what are the most relevant literature on arxiv regarding Pore-C?'), AIMessage(content='Database: ARXIV\\\\nSearch Terms: Pore-C, long-read sequencing, nanopore sequencing, single-molecule sequencing, DNA sequencing, RNA sequencing, genomics, transcriptomics, bioinformatics, computational biology')]) ai_prefix='BRAD'\", 'prompt': 'input_variables=[\\'history\\', \\'input\\'] template=\\'Current conversation:\\\\n{history}\\\\n\\\\nQuery:{input}\\\\n\\\\nFrom the query, decide if ARXIV, PUBMED, or BIORXIV should be searched, and propose at least 10 search terms for this query and database. Separate each term with a comma, and provide no extra information/explination for either the database or search terms. Do not include general purpose search terms such as \"recent advances\", \"novel\", \"current research\", \"latest trends\", \"theoretical foundations\", or other general query terms. Also, do not use very general fields as serch terms such as: \"dynamical systems\", \"machine learning\", or \"single cell\". It is correct to use these terms with modifications to be more specific, so that \"biological dynamical systems\", \"machine learning for genomics\", or \"single cell clustering\" would be good terms to search.\\\\n\\\\nThe following search terms have been previously uses and identical terms should be avoided in this search: []\\\\n\\\\n**Format your output as follows. Do not include any addition information:**\\\\nDatabase: <ARXIV, PUBMED, or BIORXIV>\\\\nSearch Terms: <improved search terms>\\\\n\\'', 'input': 'what are the most relevant literature on arxiv regarding Pore-C?', 'output': 'Database: ARXIV\\nSearch Terms: Pore-C, long-read sequencing, nanopore sequencing, single-molecule sequencing, DNA sequencing, RNA sequencing, genomics, transcriptomics, bioinformatics, computational biology', 'parsedOutput': {'database': 'ARXIV', 'search_terms': ['Pore-C', 'long-read sequencing', 'nanopore sequencing', 'single-molecule sequencing', 'DNA sequencing', 'RNA sequencing', 'genomics', 'transcriptomics', 'bioinformatics', 'computational biology']}, 'purpose': 'identify how to web scrape'}]}\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "for i in range(len(nvidia_llms)):\n",
    "    print(nvidia_llms[i])\n",
    "    llm = llms.load_nvidia(nvidia_llms[i])\n",
    "    bot = brad.chatbot(llm=llm, interactive=False)\n",
    "    bot.invoke('/force SCRAPE what are the most relevant literature on arxiv regarding Pore-C?')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e757dc47-e7c6-4d7b-80c7-1274cfd86d46",
   "metadata": {},
   "source": [
    "## OpenAI Hosted Models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0f4439a1-35ea-4c57-ae48-78667e452597",
   "metadata": {},
   "outputs": [],
   "source": [
    "# nvidia llms\n",
    "openAI_llms = [\n",
    "    'gpt-4-turbo',\n",
    "    'gpt-4o',\n",
    "    'gpt-3.5-turbo'\n",
    "]"
   ]
  },
  {
   "cell_type": "raw",
   "id": "4271c3d5-70ff-4bf9-be71-c2bd34e81fa9",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b4ac89bc-a801-4bc2-93f9-7e31f38f99fe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gpt-4-turbo\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter your Open AI API key:  ········\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32m2024-07-23 16:41:15 INFO semantic_router.utils.logger local\u001b[0m\n",
      "2024-07-23 16:41:18,874 - INFO - SCRAPE\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome to RAG! The chat log from this conversation will be saved to /nfs/turbo/umms-indikar/shared/projects/BRAD/paper-experiments/webscrape-by-llm/pore-c/2024-07-23_16-41-14/log.json. How can I help?\n",
      "==================================================\n",
      "BRAD >> 0: \n",
      "\n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation:\n",
      "\n",
      "\n",
      "Query:what are the most relevant literature on arxiv regarding Pore-c?\n",
      "\n",
      "From the query, decide if ARXIV, PUBMED, or BIORXIV should be searched, and propose at least 10 search terms for this query and database. Separate each term with a comma, and provide no extra information/explination for either the database or search terms. Do not include general purpose search terms such as \"recent advances\", \"novel\", \"current research\", \"latest trends\", \"theoretical foundations\", or other general query terms. Also, do not use very general fields as serch terms such as: \"dynamical systems\", \"machine learning\", or \"single cell\". It is correct to use these terms with modifications to be more specific, so that \"biological dynamical systems\", \"machine learning for genomics\", or \"single cell clustering\" would be good terms to search.\n",
      "\n",
      "The following search terms have been previously uses and identical terms should be avoided in this search: []\n",
      "\n",
      "**Format your output as follows. Do not include any addition information:**\n",
      "Database: <ARXIV, PUBMED, or BIORXIV>\n",
      "Search Terms: <improved search terms>\n",
      "\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-07-23 16:41:22,099 - INFO - HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2024-07-23 16:41:22,111 - INFO - {'database': 'PUBMED', 'search_terms': ['Pore-c technology', 'Pore-c chromatin conformation', 'Pore-c genome architecture', 'Pore-c sequencing', 'Pore-c data analysis', 'Pore-c 3D genome', 'Pore-c chromosomal interactions', 'Pore-c spatial genome', 'Pore-c DNA sequencing', 'Pore-c molecular biology']}\n",
      "2024-07-23 16:41:22,112 - INFO - \n",
      "\n",
      "\n",
      "route\n",
      "\n",
      "\n",
      "\n",
      "2024-07-23 16:41:22,112 - INFO - SCRAPE\n",
      "2024-07-23 16:41:22,113 - INFO - \n",
      "\n",
      "PLANNER:\n",
      "\n",
      "{'module': 'SCRAPE', 'steps': [{'llm': 'gpt-4-turbo', 'memory': \"chat_memory=InMemoryChatMessageHistory(messages=[HumanMessage(content='what are the most relevant literature on arxiv regarding Pore-c?'), AIMessage(content='Database: PUBMED\\\\nSearch Terms: Pore-c technology, Pore-c chromatin conformation, Pore-c genome architecture, Pore-c sequencing, Pore-c data analysis, Pore-c 3D genome, Pore-c chromosomal interactions, Pore-c spatial genome, Pore-c DNA sequencing, Pore-c molecular biology')]) ai_prefix='BRAD'\", 'prompt': 'input_variables=[\\'history\\', \\'input\\'] template=\\'Current conversation:\\\\n{history}\\\\n\\\\nQuery:{input}\\\\n\\\\nFrom the query, decide if ARXIV, PUBMED, or BIORXIV should be searched, and propose at least 10 search terms for this query and database. Separate each term with a comma, and provide no extra information/explination for either the database or search terms. Do not include general purpose search terms such as \"recent advances\", \"novel\", \"current research\", \"latest trends\", \"theoretical foundations\", or other general query terms. Also, do not use very general fields as serch terms such as: \"dynamical systems\", \"machine learning\", or \"single cell\". It is correct to use these terms with modifications to be more specific, so that \"biological dynamical systems\", \"machine learning for genomics\", or \"single cell clustering\" would be good terms to search.\\\\n\\\\nThe following search terms have been previously uses and identical terms should be avoided in this search: []\\\\n\\\\n**Format your output as follows. Do not include any addition information:**\\\\nDatabase: <ARXIV, PUBMED, or BIORXIV>\\\\nSearch Terms: <improved search terms>\\\\n\\'', 'input': 'what are the most relevant literature on arxiv regarding Pore-c?', 'output': 'Database: PUBMED\\nSearch Terms: Pore-c technology, Pore-c chromatin conformation, Pore-c genome architecture, Pore-c sequencing, Pore-c data analysis, Pore-c 3D genome, Pore-c chromosomal interactions, Pore-c spatial genome, Pore-c DNA sequencing, Pore-c molecular biology', 'parsedOutput': {'database': 'PUBMED', 'search_terms': ['Pore-c technology', 'Pore-c chromatin conformation', 'Pore-c genome architecture', 'Pore-c sequencing', 'Pore-c data analysis', 'Pore-c 3D genome', 'Pore-c chromosomal interactions', 'Pore-c spatial genome', 'Pore-c DNA sequencing', 'Pore-c molecular biology']}, 'purpose': 'identify how to web scrape'}]}\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "gpt-4o\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32m2024-07-23 16:41:22 INFO semantic_router.utils.logger local\u001b[0m\n",
      "2024-07-23 16:41:27,092 - INFO - SCRAPE\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome to RAG! The chat log from this conversation will be saved to /nfs/turbo/umms-indikar/shared/projects/BRAD/paper-experiments/webscrape-by-llm/pore-c/2024-07-23_16-41-22/log.json. How can I help?\n",
      "==================================================\n",
      "BRAD >> 0: \n",
      "\n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation:\n",
      "\n",
      "\n",
      "Query:what are the most relevant literature on arxiv regarding Pore-c?\n",
      "\n",
      "From the query, decide if ARXIV, PUBMED, or BIORXIV should be searched, and propose at least 10 search terms for this query and database. Separate each term with a comma, and provide no extra information/explination for either the database or search terms. Do not include general purpose search terms such as \"recent advances\", \"novel\", \"current research\", \"latest trends\", \"theoretical foundations\", or other general query terms. Also, do not use very general fields as serch terms such as: \"dynamical systems\", \"machine learning\", or \"single cell\". It is correct to use these terms with modifications to be more specific, so that \"biological dynamical systems\", \"machine learning for genomics\", or \"single cell clustering\" would be good terms to search.\n",
      "\n",
      "The following search terms have been previously uses and identical terms should be avoided in this search: []\n",
      "\n",
      "**Format your output as follows. Do not include any addition information:**\n",
      "Database: <ARXIV, PUBMED, or BIORXIV>\n",
      "Search Terms: <improved search terms>\n",
      "\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-07-23 16:41:28,865 - INFO - HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2024-07-23 16:41:28,867 - INFO - {'database': 'ARXIV', 'search_terms': ['Pore-C sequencing', 'Pore-C technology', 'Pore-C applications', 'Pore-C data analysis', 'Pore-C bioinformatics', 'Pore-C genomics', 'Pore-C epigenetics', 'Pore-C structural variation', 'Pore-C chromatin interactions', 'Pore-C multi-omics']}\n",
      "2024-07-23 16:41:28,868 - INFO - \n",
      "\n",
      "\n",
      "route\n",
      "\n",
      "\n",
      "\n",
      "2024-07-23 16:41:28,868 - INFO - SCRAPE\n",
      "2024-07-23 16:41:28,868 - INFO - \n",
      "\n",
      "PLANNER:\n",
      "\n",
      "{'module': 'SCRAPE', 'steps': [{'llm': 'gpt-4o', 'memory': \"chat_memory=InMemoryChatMessageHistory(messages=[HumanMessage(content='what are the most relevant literature on arxiv regarding Pore-c?'), AIMessage(content='Database: ARXIV\\\\nSearch Terms: Pore-C sequencing, Pore-C technology, Pore-C applications, Pore-C data analysis, Pore-C bioinformatics, Pore-C genomics, Pore-C epigenetics, Pore-C structural variation, Pore-C chromatin interactions, Pore-C multi-omics')]) ai_prefix='BRAD'\", 'prompt': 'input_variables=[\\'history\\', \\'input\\'] template=\\'Current conversation:\\\\n{history}\\\\n\\\\nQuery:{input}\\\\n\\\\nFrom the query, decide if ARXIV, PUBMED, or BIORXIV should be searched, and propose at least 10 search terms for this query and database. Separate each term with a comma, and provide no extra information/explination for either the database or search terms. Do not include general purpose search terms such as \"recent advances\", \"novel\", \"current research\", \"latest trends\", \"theoretical foundations\", or other general query terms. Also, do not use very general fields as serch terms such as: \"dynamical systems\", \"machine learning\", or \"single cell\". It is correct to use these terms with modifications to be more specific, so that \"biological dynamical systems\", \"machine learning for genomics\", or \"single cell clustering\" would be good terms to search.\\\\n\\\\nThe following search terms have been previously uses and identical terms should be avoided in this search: []\\\\n\\\\n**Format your output as follows. Do not include any addition information:**\\\\nDatabase: <ARXIV, PUBMED, or BIORXIV>\\\\nSearch Terms: <improved search terms>\\\\n\\'', 'input': 'what are the most relevant literature on arxiv regarding Pore-c?', 'output': 'Database: ARXIV\\nSearch Terms: Pore-C sequencing, Pore-C technology, Pore-C applications, Pore-C data analysis, Pore-C bioinformatics, Pore-C genomics, Pore-C epigenetics, Pore-C structural variation, Pore-C chromatin interactions, Pore-C multi-omics', 'parsedOutput': {'database': 'ARXIV', 'search_terms': ['Pore-C sequencing', 'Pore-C technology', 'Pore-C applications', 'Pore-C data analysis', 'Pore-C bioinformatics', 'Pore-C genomics', 'Pore-C epigenetics', 'Pore-C structural variation', 'Pore-C chromatin interactions', 'Pore-C multi-omics']}, 'purpose': 'identify how to web scrape'}]}\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "gpt-3.5-turbo\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32m2024-07-23 16:41:29 INFO semantic_router.utils.logger local\u001b[0m\n",
      "2024-07-23 16:41:36,669 - INFO - SCRAPE\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome to RAG! The chat log from this conversation will be saved to /nfs/turbo/umms-indikar/shared/projects/BRAD/paper-experiments/webscrape-by-llm/pore-c/2024-07-23_16-41-29/log.json. How can I help?\n",
      "==================================================\n",
      "BRAD >> 0: \n",
      "\n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation:\n",
      "\n",
      "\n",
      "Query:what are the most relevant literature on arxiv regarding Pore-c?\n",
      "\n",
      "From the query, decide if ARXIV, PUBMED, or BIORXIV should be searched, and propose at least 10 search terms for this query and database. Separate each term with a comma, and provide no extra information/explination for either the database or search terms. Do not include general purpose search terms such as \"recent advances\", \"novel\", \"current research\", \"latest trends\", \"theoretical foundations\", or other general query terms. Also, do not use very general fields as serch terms such as: \"dynamical systems\", \"machine learning\", or \"single cell\". It is correct to use these terms with modifications to be more specific, so that \"biological dynamical systems\", \"machine learning for genomics\", or \"single cell clustering\" would be good terms to search.\n",
      "\n",
      "The following search terms have been previously uses and identical terms should be avoided in this search: []\n",
      "\n",
      "**Format your output as follows. Do not include any addition information:**\n",
      "Database: <ARXIV, PUBMED, or BIORXIV>\n",
      "Search Terms: <improved search terms>\n",
      "\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-07-23 16:41:38,234 - INFO - HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2024-07-23 16:41:38,235 - INFO - {'database': 'ARXIV', 'search_terms': ['Pore-c analysis', 'Nanopore sequencing', 'Pore-c applications', 'Pore-c technology', 'Pore-c methodology', 'Pore-c performance', 'Pore-c review', 'Pore-c challenges', 'Pore-c advancements', 'Pore-c developments']}\n",
      "2024-07-23 16:41:38,236 - INFO - \n",
      "\n",
      "\n",
      "route\n",
      "\n",
      "\n",
      "\n",
      "2024-07-23 16:41:38,236 - INFO - SCRAPE\n",
      "2024-07-23 16:41:38,237 - INFO - \n",
      "\n",
      "PLANNER:\n",
      "\n",
      "{'module': 'SCRAPE', 'steps': [{'llm': 'gpt-3.5-turbo', 'memory': \"chat_memory=InMemoryChatMessageHistory(messages=[HumanMessage(content='what are the most relevant literature on arxiv regarding Pore-c?'), AIMessage(content='Database: ARXIV\\\\nSearch Terms: Pore-c analysis, Nanopore sequencing, Pore-c applications, Pore-c technology, Pore-c methodology, Pore-c performance, Pore-c review, Pore-c challenges, Pore-c advancements, Pore-c developments')]) ai_prefix='BRAD'\", 'prompt': 'input_variables=[\\'history\\', \\'input\\'] template=\\'Current conversation:\\\\n{history}\\\\n\\\\nQuery:{input}\\\\n\\\\nFrom the query, decide if ARXIV, PUBMED, or BIORXIV should be searched, and propose at least 10 search terms for this query and database. Separate each term with a comma, and provide no extra information/explination for either the database or search terms. Do not include general purpose search terms such as \"recent advances\", \"novel\", \"current research\", \"latest trends\", \"theoretical foundations\", or other general query terms. Also, do not use very general fields as serch terms such as: \"dynamical systems\", \"machine learning\", or \"single cell\". It is correct to use these terms with modifications to be more specific, so that \"biological dynamical systems\", \"machine learning for genomics\", or \"single cell clustering\" would be good terms to search.\\\\n\\\\nThe following search terms have been previously uses and identical terms should be avoided in this search: []\\\\n\\\\n**Format your output as follows. Do not include any addition information:**\\\\nDatabase: <ARXIV, PUBMED, or BIORXIV>\\\\nSearch Terms: <improved search terms>\\\\n\\'', 'input': 'what are the most relevant literature on arxiv regarding Pore-c?', 'output': 'Database: ARXIV\\nSearch Terms: Pore-c analysis, Nanopore sequencing, Pore-c applications, Pore-c technology, Pore-c methodology, Pore-c performance, Pore-c review, Pore-c challenges, Pore-c advancements, Pore-c developments', 'parsedOutput': {'database': 'ARXIV', 'search_terms': ['Pore-c analysis', 'Nanopore sequencing', 'Pore-c applications', 'Pore-c technology', 'Pore-c methodology', 'Pore-c performance', 'Pore-c review', 'Pore-c challenges', 'Pore-c advancements', 'Pore-c developments']}, 'purpose': 'identify how to web scrape'}]}\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "for i in range(len(openAI_llms)):\n",
    "    print(openAI_llms[i])\n",
    "    llm = llms.load_openai(openAI_llms[i])\n",
    "    bot = brad.chatbot(llm=llm, interactive=False)\n",
    "    bot.invoke('/force SCRAPE what are the most relevant literature on arxiv regarding Pore-c?')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a76c613b-8058-4ec7-b934-338ede7cf006",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2bad20e4-b068-45fa-955e-a14452dce3d6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9d120175-5eaa-4e36-a7bf-d8b956427d3a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter your Open AI API key:  ········\n"
     ]
    }
   ],
   "source": [
    "llm = llms.load_openai(openAI_llms[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "eacb7416-74a2-4cb7-9c50-b152638097e6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'gpt-4-turbo'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "llm.model_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9a09c56b-0b5d-4ede-8892-a69f5c9c974e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter your Open AI API key:  ········\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'gpt-4o'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "llm = llms.load_openai(openAI_llms[1])\n",
    "llm.model_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3664ea6e-1810-4f61-98f4-b7ea53d8093f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gpt-4-turbo\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter your Open AI API key:  ········\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32m2024-07-23 15:29:39 INFO semantic_router.utils.logger local\u001b[0m\n",
      "2024-07-23 15:29:45,309 - INFO - SCRAPE\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome to RAG! The chat log from this conversation will be saved to /nfs/turbo/umms-indikar/shared/projects/BRAD/paper-experiments/webscrape-by-llm/2024-07-23_15-29-37/log.json. How can I help?\n",
      "==================================================\n",
      "BRAD >> 0: \n",
      "\n",
      "\n",
      "\u001b[1m> Entering new ConversationChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mCurrent conversation:\n",
      "\n",
      "\n",
      "Query:what are the most relevant literature on arxiv regarding cellular reprogramming?\n",
      "\n",
      "From the query, decide if ARXIV, PUBMED, or BIORXIV should be searched, and propose at least 10 search terms for this query and database. Separate each term with a comma, and provide no extra information/explination for either the database or search terms. Do not include general purpose search terms such as \"recent advances\", \"novel\", \"current research\", \"latest trends\", \"theoretical foundations\", or other general query terms. Also, do not use very general fields as serch terms such as: \"dynamical systems\", \"machine learning\", or \"single cell\". It is correct to use these terms with modifications to be more specific, so that \"biological dynamical systems\", \"machine learning for genomics\", or \"single cell clustering\" would be good terms to search.\n",
      "\n",
      "The following search terms have been previously uses and identical terms should be avoided in this search: []\n",
      "\n",
      "**Format your output as follows. Do not include any addition information:**\n",
      "Database: <ARXIV, PUBMED, or BIORXIV>\n",
      "Search Terms: <improved search terms>\n",
      "\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-07-23 15:29:48,825 - INFO - HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2024-07-23 15:29:48,838 - INFO - {'database': 'ARXIV', 'search_terms': ['cellular reprogramming techniques', 'induced pluripotent stem cells', 'transcription factors in reprogramming', 'epigenetic reprogramming', 'direct cell reprogramming', 'reprogramming efficiency', 'cellular reprogramming and aging', 'reprogramming and disease modeling', 'gene editing in cellular reprogramming', 'cellular reprogramming and regeneration']}\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n"
     ]
    },
    {
     "ename": "UnboundLocalError",
     "evalue": "cannot access local variable 'e' where it is not associated with a value",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "File \u001b[0;32m~/RAG-DEV/BRAD/scraper.py:82\u001b[0m, in \u001b[0;36mwebScraping\u001b[0;34m(chatstatus)\u001b[0m\n\u001b[1;32m     81\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m---> 82\u001b[0m     llmType \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mstr\u001b[39m(\u001b[43mllm\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmodel\u001b[49m)\n\u001b[1;32m     83\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m e:\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'ChatOpenAI' object has no attribute 'model'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mUnboundLocalError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[6], line 5\u001b[0m\n\u001b[1;32m      3\u001b[0m llm \u001b[38;5;241m=\u001b[39m llms\u001b[38;5;241m.\u001b[39mload_openai(openAI_llms[i])\n\u001b[1;32m      4\u001b[0m bot \u001b[38;5;241m=\u001b[39m brad\u001b[38;5;241m.\u001b[39mchatbot(llm\u001b[38;5;241m=\u001b[39mllm, interactive\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n\u001b[0;32m----> 5\u001b[0m \u001b[43mbot\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43minvoke\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m/force SCRAPE what are the most relevant literature on arxiv regarding cellular reprogramming?\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/RAG-DEV/BRAD/brad.py:270\u001b[0m, in \u001b[0;36mchatbot.invoke\u001b[0;34m(self, query)\u001b[0m\n\u001b[1;32m    264\u001b[0m output_files \u001b[38;5;241m=\u001b[39m utils\u001b[38;5;241m.\u001b[39moutputFiles(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mchatstatus)\n\u001b[1;32m    266\u001b[0m \u001b[38;5;66;03m# if not self.chatstatus['interactive']:\u001b[39;00m\n\u001b[1;32m    267\u001b[0m \u001b[38;5;66;03m#     self.chatstatus['queue'] = [[]] # don't let it us an empty queue if it is not interactive\u001b[39;00m\n\u001b[1;32m    268\u001b[0m \n\u001b[1;32m    269\u001b[0m \u001b[38;5;66;03m# Query module\u001b[39;00m\n\u001b[0;32m--> 270\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mchatstatus \u001b[38;5;241m=\u001b[39m \u001b[43mmodule\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mchatstatus\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    272\u001b[0m \u001b[38;5;66;03m# Remove the item that was executed. We need must do it after running it for the current file naming system.\u001b[39;00m\n\u001b[1;32m    273\u001b[0m log\u001b[38;5;241m.\u001b[39mdebugLog(\u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124mroute\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m'\u001b[39m, chatstatus\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mchatstatus)\n",
      "File \u001b[0;32m~/RAG-DEV/BRAD/scraper.py:83\u001b[0m, in \u001b[0;36mwebScraping\u001b[0;34m(chatstatus)\u001b[0m\n\u001b[1;32m     81\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m     82\u001b[0m     llmType \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mstr\u001b[39m(llm\u001b[38;5;241m.\u001b[39mmodel)\n\u001b[0;32m---> 83\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[43me\u001b[49m:\n\u001b[1;32m     84\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m     85\u001b[0m         llmType \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mstr\u001b[39m(llm\u001b[38;5;241m.\u001b[39mmodel_name)\n",
      "\u001b[0;31mUnboundLocalError\u001b[0m: cannot access local variable 'e' where it is not associated with a value"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "042a4042-d53a-4a1f-aa67-a4a066a61047",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SyncPage[Model](data=[Model(id='dall-e-3', created=1698785189, object='model', owned_by='system'), Model(id='gpt-4o-2024-05-13', created=1715368132, object='model', owned_by='system'), Model(id='gpt-4-turbo-2024-04-09', created=1712601677, object='model', owned_by='system'), Model(id='gpt-4-1106-preview', created=1698957206, object='model', owned_by='system'), Model(id='dall-e-2', created=1698798177, object='model', owned_by='system'), Model(id='gpt-4o', created=1715367049, object='model', owned_by='system'), Model(id='gpt-4-turbo', created=1712361441, object='model', owned_by='system'), Model(id='tts-1-hd-1106', created=1699053533, object='model', owned_by='system'), Model(id='tts-1-hd', created=1699046015, object='model', owned_by='system'), Model(id='gpt-4-0125-preview', created=1706037612, object='model', owned_by='system'), Model(id='babbage-002', created=1692634615, object='model', owned_by='system'), Model(id='gpt-4-turbo-preview', created=1706037777, object='model', owned_by='system'), Model(id='text-embedding-3-small', created=1705948997, object='model', owned_by='system'), Model(id='text-embedding-3-large', created=1705953180, object='model', owned_by='system'), Model(id='gpt-3.5-turbo', created=1677610602, object='model', owned_by='openai'), Model(id='gpt-3.5-turbo-instruct', created=1692901427, object='model', owned_by='system'), Model(id='gpt-3.5-turbo-instruct-0914', created=1694122472, object='model', owned_by='system'), Model(id='whisper-1', created=1677532384, object='model', owned_by='openai-internal'), Model(id='text-embedding-ada-002', created=1671217299, object='model', owned_by='openai-internal'), Model(id='gpt-3.5-turbo-16k', created=1683758102, object='model', owned_by='openai-internal'), Model(id='davinci-002', created=1692634301, object='model', owned_by='system'), Model(id='tts-1-1106', created=1699053241, object='model', owned_by='system'), Model(id='gpt-3.5-turbo-0125', created=1706048358, object='model', owned_by='system'), Model(id='tts-1', created=1681940951, object='model', owned_by='openai-internal'), Model(id='gpt-3.5-turbo-1106', created=1698959748, object='model', owned_by='system'), Model(id='gpt-4o-mini-2024-07-18', created=1721172717, object='model', owned_by='system'), Model(id='gpt-4o-mini', created=1721172741, object='model', owned_by='system'), Model(id='gpt-4-0613', created=1686588896, object='model', owned_by='openai'), Model(id='gpt-4', created=1687882411, object='model', owned_by='openai')], object='list')\n"
     ]
    }
   ],
   "source": [
    "from openai import OpenAI\n",
    "import os\n",
    "\n",
    "client = OpenAI(\n",
    "    api_key = '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'\n",
    ")\n",
    "    \n",
    "models = client.models.list()\n",
    "\n",
    "print(models)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0edfaf93-a700-4811-9db2-ded484de69d2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dall-e-3\n",
      "gpt-4o-2024-05-13\n",
      "gpt-4-turbo-2024-04-09\n",
      "gpt-4-1106-preview\n",
      "dall-e-2\n",
      "gpt-4o\n",
      "gpt-4-turbo\n",
      "tts-1-hd-1106\n",
      "tts-1-hd\n",
      "gpt-4-0125-preview\n",
      "babbage-002\n",
      "gpt-4-turbo-preview\n",
      "text-embedding-3-small\n",
      "text-embedding-3-large\n",
      "gpt-3.5-turbo\n",
      "gpt-3.5-turbo-instruct\n",
      "gpt-3.5-turbo-instruct-0914\n",
      "whisper-1\n",
      "text-embedding-ada-002\n",
      "gpt-3.5-turbo-16k\n",
      "davinci-002\n",
      "tts-1-1106\n",
      "gpt-3.5-turbo-0125\n",
      "tts-1\n",
      "gpt-3.5-turbo-1106\n",
      "gpt-4o-mini-2024-07-18\n",
      "gpt-4o-mini\n",
      "gpt-4-0613\n",
      "gpt-4\n"
     ]
    }
   ],
   "source": [
    "for model in models:\n",
    "    print(model.id)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
