'''Get gene former embeddings of your data
Get gene former embeddings of your data

Arguments (three arguments):
    1. output path: chatstatus['output-directory']
    2. output file: <name of output file>
    3. input file: <name of the input file>

Usage:
    Command Line:
    ```
    python <path/to/script/>geneformerEmbedding.py <output path> <output file> <input file>
    ```
                                                       |              |           |
                                                   Argument 1     Argument 2   Argument 3
    BRAD Line:
    ```
    subprocess.call([sys.executable, '<path/to/script/>/geneformerEmbedding.py', chatstatus['output-directory'], <output file>, <input file>])
    ```

Output file name should be `S<Step number>-EF-Embeddings.csv`

This script uses Geneformer to embed gene expression data.

We could add more documentation for this and make it a more complicated method too.
'''
import os
import sys
from geneformer import EmbExtractor

def main():
    model = "/nfs/turbo/umms-indikar/shared/projects/geneformer/fine_tuned_models/geneformer-6L-30M_CellClassifier_cardiomyopathies_220224"

    outputPath = sys.argv[1]
    outputFile = sys.argv[2]
    inputFile  = sys.argv[3]
    # outputFile = os.path.join(outputPath, outputFile)

    data_path = "/nfs/turbo/umms-indikar/shared/projects/geneformer/example_input_files/cell_classification/disease_classification/human_dcm_hcm_nf.dataset/"
    
    # initiate EmbExtractor
    embex = EmbExtractor(model_type="CellClassifier",
                         num_classes=3,
                         max_ncells=100,
                         emb_layer=-1,
                         emb_label=["cell_type", "length", "disease"],
                         labels_to_plot=["cell_type", "length", "disease"],
                         forward_batch_size=10,
                         nproc=16)
    
    # extracts embedding from input data
    # input data is tokenized rank value encodings generated by Geneformer tokenizer (see tokenizing_scRNAseq_data.ipynb)
    # example dataset: https://huggingface.co/datasets/ctheodoris/Genecorpus-30M/tree/main/example_input_files/cell_classification/disease_classification/human_dcm_hcm_nf.dataset
    embs = embex.extract_embs(model,
                              data_path,
                              outputPath,
                              outputFile)

if __name__ == "__main__":
    main()
