{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "90c59371-1957-447c-922e-bd8e416ed1fe",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Retrieval Augmented Generation (RAG) with BRAD\n",
    "\n",
    "Given a collection of documents (pdfs), the RAG first builds a database by splitting up the text into equal size ('chunk_size'). These chunks can optionally be set to have some overlap ('chunk_overlap'). Both of these values within BRAD are set to 700 and 200 respectively but can be manually changed for different applications. Then, we vectorize these chunks using an embedding model from HuggingFace (Note - this step may take a while). Then, given a query, it embeds the query in the same embedding space and finds the top k chunks (preset to 4) with the closest cosine similarity to the query in the embedding space and uses these chunks as a basis for the response."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f185bba2-a3de-4013-9e58-9d0856829d16",
   "metadata": {},
   "source": [
    "# Literature Databases"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "996b2a1b-48cb-428a-a4ea-e5c47a617b8c",
   "metadata": {},
   "source": [
    "## Building a Database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d8a78fec-c55f-4063-9768-cf101482e39a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import os\n",
    "from langchain.document_loaders import DirectoryLoader\n",
    "from langchain.document_loaders import UnstructuredPDFLoader\n",
    "from langchain.vectorstores import Chroma\n",
    "from langchain.embeddings import HuggingFaceEmbeddings\n",
    "import chromadb\n",
    "from langchain_community.document_loaders import PyPDFLoader\n",
    "from langchain_text_splitters import RecursiveCharacterTextSplitter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8c0a7a24-4c07-4657-a7b3-fe7e35252c13",
   "metadata": {},
   "outputs": [],
   "source": [
    "from BRAD import rag"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "223ee6b9-5c46-46fe-a21f-90abd7cc65f8",
   "metadata": {},
   "source": [
    "## Fixed Size Chunking"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "1c3a171a-a7c7-41aa-b649-8abc7e6314be",
   "metadata": {},
   "outputs": [],
   "source": [
    "docsPath='papers/'\n",
    "dbName='database'\n",
    "dbPath='databases/'\n",
    "HuggingFaceEmbeddingsModel = 'BAAI/bge-base-en-v1.5'\n",
    "chunk_size=[700]\n",
    "chunk_overlap=[200]\n",
    "v=False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5da62144-d4bf-4c17-b2cc-3111507c0755",
   "metadata": {},
   "outputs": [],
   "source": [
    "local = os.getcwd()  ## Get local dir\n",
    "os.chdir(local)      ## shift the work dir to local dir\n",
    "\n",
    "print('\\nWork Directory: {}'.format(local)) if v else None\n",
    "\n",
    "#%% Phase 1 - Load DB\n",
    "embeddings_model = HuggingFaceEmbeddings(model_name=HuggingFaceEmbeddingsModel)\n",
    "\n",
    "print('\\nDocuments loading from:', docsPath) if v else None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f70b31b-a1ec-4f7a-b296-9e5dccfc5b5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "text_loader_kwargs={'autodetect_encoding': True}\n",
    "loader = DirectoryLoader(docsPath,\n",
    "                         glob=\"**/*.pdf\",\n",
    "                         loader_cls=PyPDFLoader, \n",
    "                         show_progress=True,\n",
    "                         use_multithreading=True)\n",
    "docs_data = loader.load()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "331f6eff-8e19-4a73-822f-6115e0312828",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘databases/DB_cosine_cSize_700_cOver_200/*’: No such file or directory\n"
     ]
    }
   ],
   "source": [
    "print('\\nDocuments loaded...') if v else None\n",
    "\n",
    "for i in range(len(chunk_size)):\n",
    "    for j in range(len(chunk_overlap)):\n",
    "        text_splitter = RecursiveCharacterTextSplitter(chunk_size = chunk_size[i],\n",
    "                                                        chunk_overlap = chunk_overlap[j],\n",
    "                                                        separators=[\" \", \",\", \"\\n\", \". \"])\n",
    "        data_splits = text_splitter.split_documents(docs_data)\n",
    "        \n",
    "        print('Documents split into chunks...') if v else None\n",
    "        print('Initializing Chroma Database...') if v else None\n",
    "\n",
    "        dbName = \"DB_cosine_cSize_%d_cOver_%d\" %(chunk_size[i], chunk_overlap[j])\n",
    "\n",
    "        p2_2 = subprocess.run('mkdir  %s/*'%(dbPath+dbName), shell=True)\n",
    "        _client_settings = chromadb.PersistentClient(path=(dbPath+dbName))\n",
    "\n",
    "        vectordb = Chroma.from_documents(documents           = data_splits,\n",
    "                                         embedding           = embeddings_model,\n",
    "                                         client              = _client_settings,\n",
    "                                         collection_name     = dbName,\n",
    "                                         collection_metadata = {\"hnsw:space\": \"cosine\"})\n",
    "\n",
    "        print('Completed Chroma Database: ', dbName) if v else None\n",
    "        del vectordb, text_splitter, data_splits"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1450bc00-151e-4e0c-a72d-a337d24baaf0",
   "metadata": {},
   "source": [
    "### Contextual Chunking"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9e244cd2-a553-483f-b09b-8072b7a05179",
   "metadata": {},
   "outputs": [],
   "source": [
    "from semantic_router.splitters import RollingWindowSplitter\n",
    "from semantic_router.utils.logger import logger"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "dbdfb6d8-b982-4fc3-bf91-407ccb110be3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "HuggingFaceEmbeddingsModel = 'BAAI/bge-base-en-v1.5'\n",
    "encoder = HuggingFaceEmbeddings(model_name=HuggingFaceEmbeddingsModel)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "d7e59272-88fc-42d2-8987-ff0a5cfe9058",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from getpass import getpass\n",
    "from semantic_router.encoders import OpenAIEncoder\n",
    "\n",
    "os.environ[\"OPENAI_API_KEY\"] = os.getenv(\"OPENAI_API_KEY\") or getpass(\n",
    "    \"OpenAI API key: \"\n",
    ")\n",
    "# encoder = OpenAIEncoder(openai_base_url='https://integrate.api.nvidia.com/v1')\n",
    "encoder = OpenAIEncoder(name=\"text-embedding-3-small\", openai_base_url='https://integrate.api.nvidia.com/v1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "4bad077a-7a3a-494d-98ab-54e65fa6559f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3740256/3643352218.py:1: UserWarning: Splitters are being deprecated. They have moved to their own package. Please migrate to the `semantic-chunkers` package. More information can be found at:\n",
      "https://github.com/aurelio-labs/semantic-chunkers\n",
      "  splitter = RollingWindowSplitter(\n"
     ]
    }
   ],
   "source": [
    "splitter = RollingWindowSplitter(\n",
    "    encoder=encoder,\n",
    "    dynamic_threshold=True,\n",
    "    min_split_tokens=100,\n",
    "    max_split_tokens=1000,\n",
    "    window_size=2,\n",
    "    plot_splits=True,  # set this to true to visualize chunking\n",
    "    enable_statistics=True  # to print chunking stats\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a3d059bd-6baf-4903-9e7d-687107586313",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Dataset({\n",
       "    features: ['id', 'title', 'summary', 'source', 'authors', 'categories', 'comment', 'journal_ref', 'primary_category', 'published', 'updated', 'content', 'references'],\n",
       "    num_rows: 2673\n",
       "})"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from datasets import load_dataset\n",
    "\n",
    "dataset = load_dataset(\"jamescalam/ai-arxiv2\", split=\"train\")\n",
    "dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "e32d0a8f-2a32-4112-a603-25b2e72d24b8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[33m2024-06-17 09:42:04 WARNING semantic_router.utils.logger Retrying in 2 seconds...\u001b[0m\n",
      "\u001b[33m2024-06-17 09:42:08 WARNING semantic_router.utils.logger Retrying in 4 seconds...\u001b[0m\n",
      "\u001b[33m2024-06-17 09:42:16 WARNING semantic_router.utils.logger Retrying in 8 seconds...\u001b[0m\n",
      "\u001b[33m2024-06-17 09:42:32 WARNING semantic_router.utils.logger Retrying in 16 seconds...\u001b[0m\n",
      "\u001b[33m2024-06-17 09:43:05 WARNING semantic_router.utils.logger Retrying in 32 seconds...\u001b[0m\n",
      "\u001b[33m2024-06-17 09:44:09 WARNING semantic_router.utils.logger Retrying in 64 seconds...\u001b[0m\n",
      "\u001b[31m2024-06-17 09:44:09 ERROR semantic_router.utils.logger Error encoding documents ['4 2 0 2', 'n a J 8 ] G L . s c [', '1 v 8 8 0 4 0 . 1 0 4 2 : v i X r a', '# Mixtral of Experts', 'Albert Q.', 'Jiang, Alexandre Sablayrolles, Antoine Roux, Arthur Mensch, Blanche Savary, Chris Bamford, Devendra Singh Chaplot, Diego de las Casas, Emma Bou Hanna, Florian Bressand, Gianna Lengyel, Guillaume Bour, Guillaume Lample, LÃ©lio Renard Lavaud, Lucile Saulnier, Marie-Anne Lachaux, Pierre Stock, Sandeep Subramanian, Sophia Yang, Szymon Antoniak, Teven Le Scao, ThÃ©ophile Gervet, Thibaut Lavril, Thomas Wang, TimothÃ©e Lacroix, William El Sayed', 'Abstract', 'We introduce Mixtral 8x7B, a Sparse Mixture of Experts (SMoE) language model.', 'Mixtral has the same architecture as Mistral 7B, with the difference that each layer is composed of 8 feedforward blocks (i.e. experts).', 'For every token, at each layer, a router network selects two experts to process the current state and combine their outputs.', 'Even though each token only sees two experts, the selected experts can be different at each timestep.', 'As a result, each token has access to 47B parameters, but only uses 13B active parameters during inference.', 'Mixtral was trained with a context size of 32k tokens and it outperforms or matches Llama 2 70B and GPT-3.5 across all evaluated benchmarks.', 'In particular, Mixtral vastly outperforms Llama 2 70B on mathematics, code generation, and multilingual benchmarks.', 'We also provide a model fine- tuned to follow instructions, Mixtral 8x7B â', 'Instruct, that surpasses GPT-3.5 Turbo, Claude-2.1, Gemini Pro, and Llama 2 70B â', 'chat model on human bench- marks.', 'Both the base and instruct models are released under the Apache 2.0 license.', 'Code: https://github.com/mistralai/mistral-src Webpage: https://mistral.ai/news/mixtral-of-experts/', '# Introduction', 'In this paper, we present Mixtral 8x7B, a sparse mixture of experts model (SMoE) with open weights, licensed under Apache 2.0.', 'Mixtral outperforms Llama 2 70B and GPT-3.5 on most benchmarks.', 'As it only uses a subset of its parameters for every token, Mixtral allows faster inference speed at low batch-sizes, and higher throughput at large batch-sizes.', 'Mixtral is a sparse mixture-of-experts network.', 'It is a decoder-only model where the feedforward block picks from a set of 8 distinct groups of parameters.', 'At every layer, for every token, a router network chooses two of these groups (the â', 'expertsâ', ') to process the token and combine their output additively.', 'This technique increases the number of parameters of a model while controlling cost and latency, as the model only uses a fraction of the total set of parameters per token.', 'Mixtral is pretrained with multilingual data using a context size of 32k tokens.', 'It either matches or exceeds the performance of Llama 2 70B and GPT-3.5, over several benchmarks.', 'In particular,', 'Mixture of Experts Layer i gating inputs af outputs router expert', 'Figure 1:', 'Mixture of Experts Layer.', 'Each input vector is assigned to 2 of the 8 experts by a router.', 'The layerâ', 's output is the weighted sum of the outputs of the two selected experts.', 'In Mixtral, an expert is a standard feedforward block as in a vanilla transformer architecture.', 'Mixtral demonstrates superior capabilities in mathematics, code generation, and tasks that require multilingual understanding, significantly outperforming Llama 2 70B in these domains.', 'Experiments show that Mixtral is able to successfully retrieve information from its context window of 32k tokens, regardless of the sequence length and the location of the information in the sequence.', 'We also present Mixtral 8x7B â', 'Instruct, a chat model fine-tuned to follow instructions using supervised fine-tuning and Direct Preference Optimization [25].', 'Its performance notably surpasses that of GPT-3.5 Turbo, Claude-2.1, Gemini Pro, and Llama 2 70B â', 'chat model on human evaluation benchmarks.', 'Mixtral â', 'Instruct also demonstrates reduced biases, and a more balanced sentiment profile in benchmarks such as BBQ, and BOLD.', 'We release both Mixtral 8x7B and Mixtral 8x7B â', 'Instruct under the Apache 2.0 license1, free for academic and commercial usage, ensuring broad accessibility and potential for diverse applications.', 'To enable the community to run Mixtral with a fully open-source stack, we submitted changes to the vLLM project, which integrates Megablocks CUDA kernels for efficient inference.', 'Skypilot also allows the deployment of vLLM endpoints on any instance in the cloud.', '# 2 Architectural details', 'Mixtral is based on a transformer architecture [31] and uses the same modifications as described in [18], with the notable exceptions that Mix- tral supports a fully dense context length of 32k tokens, and the feed- forward blocks are replaced by Mixture-of-Expert layers (Section 2.1).', 'The model architecture parameters are summarized in Table 1.', 'Parameter Value', 'dim n_layers head_dim hidden_dim n_heads n_kv_heads context_len vocab_size num_experts top_k_experts', '# 2.1 Sparse Mixture of Experts', 'We present a brief overview of the Mixture of Experts layer (Figure 1).', 'For a more in-depth overview, see [12].', 'The output of the MoE module for a given input x is determined by the weighted sum of the outputs of the expert networks, where the weights are given by the gating networkâ', 's output. i.e. given n expert networks {E0, Ei, ..., Enâ', '1}, the output of the expert layer is given by:', 'Table 1:', 'Model architecture.', '# j nâ', 'G(x)i Â· Ei(x). i=0', 'Here, G(x)i denotes the n-dimensional output of the gating network for the i-th expert, and Ei(x) is the output of the i-th expert network.', 'If the gating vector is sparse, we can avoid computing the outputs of experts whose gates are zero.', 'There are multiple alternative ways of implementing G(x) [6, 15, 35], but a simple and performant one is implemented by taking the softmax over the Top-K logits of a linear layer [28].', 'We use', 'G(x) := Softmax(TopK(x Â· Wg)),', 'where (TopK(â', '))i := â', 'i if â', 'i is among the top-K coordinates of logits â', 'â', 'Rn and (TopK(â', '))i := â', 'â', 'otherwise.', 'The value of K â', 'the number of experts used per token â', 'is a hyper-parameter that modu- lates the amount of compute used to process each token.', 'If one increases n while keeping K fixed, one', '# 1https://mistral.ai/news/mixtral-of-experts/', '2', '4096 32 128 14336 32 8 32768 32000 8 2', 'can increase the modelâ', 's parameter count while keeping its computational cost effectively constant.', 'This motivates a distinction between the modelâ', 's total parameter count (commonly referenced as the sparse parameter count), which grows with n, and the number of parameters used for processing an individual token (called the active parameter count), which grows with K up to n.', 'MoE layers can be run efficiently on single GPUs with high performance specialized kernels.', 'For example, Megablocks [13] casts the feed-forward network (FFN) operations of the MoE layer as large sparse matrix multiplications, significantly enhancing the execution speed and naturally handling cases where different experts get a variable number of tokens assigned to them.', 'Moreover, the MoE layer can be distributed to multiple GPUs through standard Model Parallelism techniques, and through a particular kind of partitioning strategy called Expert Parallelism (EP) [28].', 'During the MoE layerâ', 's execution, tokens meant to be processed by a specific expert are routed to the corresponding GPU for processing, and the expertâ', 's output is returned to the original token location.', 'Note that EP introduces challenges in load balancing, as it is essential to distribute the workload evenly across the GPUs to prevent overloading individual GPUs or hitting computational bottlenecks.', 'In a Transformer model, the MoE layer is applied independently per token and replaces the feed-forward (FFN) sub-block of the transformer block.', 'For Mixtral we use the same SwiGLU architecture as the expert function Ei(x) and set K = 2.', 'This means each token is routed to two SwiGLU sub-blocks with different sets of weights.', 'Taking this all together, the output y for an input token x is computed as:', 'n-1 y= Ss Softmax(Top2(a - W,)); - SwiGLU;(a). i=0', 'This formulation is similar to the GShard architecture [21], with the exceptions that we replace all FFN sub-blocks by MoE layers while GShard replaces every other block, and that GShard uses a more elaborate gating strategy for the second expert assigned to each token.', '# 3 Results', 'We compare Mixtral to Llama, and re-run all benchmarks with our own evaluation pipeline for fair comparison.', 'We measure performance on a wide variety of tasks categorized as follow:', 'â', '¢ Commonsense Reasoning (0-shot):', 'Hellaswag [32], Winogrande [26], PIQA [3], SIQA [27], OpenbookQA [22], ARC-Easy, ARC-Challenge [8], CommonsenseQA [30]', 'World Knowledge (5-shot):', 'NaturalQuestions [20], TriviaQA [19] â', '¢ Reading Comprehension (0-shot):', 'BoolQ [7], QuAC [5] â', '¢ Math:', 'GSM8K [9] (8-shot) with maj@8 and MATH [17] (4-shot) with maj@4 â', '¢ Code:', 'Humaneval [4] (0-shot) and MBPP [1] (3-shot) â', '¢ Popular aggregated results:', 'MMLU [16] (5-shot), BBH [29] (3-shot), and AGI Eval [34]', '(3-5-shot, English multiple-choice questions only)', '80 SE Mistral 78 = LLaMA27B = Sl LLaMA134B, jam Mistral 78 = LlaMA27B Ss LLAMA 1348, cee Mixtral 8x78 Sm LLaMA213BÂ° mmm LLaMA2 70B je Mixtral 8x78 mm LlaMA2138 lm LLaMA2 708 70 50 60 50 20 40 10 BH Code MMU Knowledge Reasoning â', 'Comprehension AGI Eval Math â', 'Accuracy (%)', 'Figure 2:', 'Performance of Mixtral and different Llama models on a wide range of benchmarks.', 'All models were re-evaluated on all metrics with our evaluation pipeline for accurate comparison.', 'Mixtral outperforms or matches Llama 2 70B on all benchmarks.', 'In particular, it is vastly superior in mathematics and code generation.', '3', 'Active Params MMLU HellaS WinoG PIQA Arc-e Arc-c NQ TriQA HumanE MBPP Math GSM8K 7B 44.4% 77.1% 69.5% 77.9% 68.7% 43.2% 17.5% 56.6% 11.6% 26.1% 3.9% 16.0% 13B 55.6% 80.7% 72.9% 80.8% 75.2% 48.8% 16.7% 64.0% 18.9% 35.4% 6.0% 34.3% 33B 56.8% 83.7% 76.2% 82.2% 79.6% 54.4% 24.1% 68.5% 25.0% 40.9% 8.4% 44.1% 70B 69.9% 85.4% 80.4% 82.6% 79.9% 56.5% 25.4% 73.0% 29.3% 49.8% 13.8% 69.6% 7B 62.5% 81.0% 74.2% 82.2% 80.5% 54.9% 23.2% 62.5% 26.2% 50.2% 12.7% 50.0% 13B 70.6% 84.4% 77.2% 83.6% 83.1% 59.7% 30.6% 71.5% 40.2% 60.7% 28.4% 74.4%', 'Table 2:', 'Comparison of Mixtral with Llama.', 'Mixtral outperforms or matches Llama 2 70B performance on almost all popular benchmarks while using 5x fewer active parameters during inference.', '70 Mixtral 8x7B. â', 'Mixtral 8x7B Mixtral 8x7B 355 =o = Es & E60!', 'Mistral 78 % 2681 Mistral 78 3 3 s0 5 = A % 66 50 g 4 45 64 78 138 348708 78 138 348708 78 138 348 70B S66 Mixtral 8x7B 50 Mixtral 8x7B 5 = 564 340 g al Mistral 78 ee Mistral 78 3 5 Â§ 30 5 eo â', '= Mistral Â° 20 â', 'e LlaMA2 78 (138 348 70B 7B (138 348 708 7B Â«13B 34B 708 Active Params Active Params Active Params', 'Figure 3:', 'Results on MMLU, commonsense reasoning, world knowledge and reading comprehension, math and code for Mistral (7B/8x7B) vs Llama 2 (7B/13B/70B).', 'Mixtral largely outperforms Llama 2 70B on all benchmarks, except on reading comprehension benchmarks while using 5x lower active parameters.', 'It is also vastly superior to Llama 2 70B on code and math.', 'Detailed results for Mixtral, Mistral 7B and Llama 2 7B/13B/70B and Llama 1 34B2 are reported in Table 2.', 'Figure 2 compares the performance of Mixtral with the Llama models in different categories.', 'Mixtral surpasses Llama 2 70B across most metrics.', 'In particular, Mixtral displays a superior performance in code and mathematics benchmarks.', 'Size and Efficiency.', 'We compare our performance to the Llama 2 family, aiming to understand Mixtral modelsâ', 'efficiency in the cost-performance spectrum (see Figure 3).', 'As a sparse Mixture- of-Experts model, Mixtral only uses 13B active parameters for each token.', 'With 5x lower active parameters, Mixtral is able to outperform Llama 2 70B across most categories.', 'Note that this analysis focuses on the active parameter count (see Section 2.1), which is directly proportional to the inference compute cost, but does not consider the memory costs and hardware utilization.', 'The memory costs for serving Mixtral are proportional to its sparse parameter count, 47B, which is still smaller than Llama 2 70B.', 'As for device utilization, we note that the SMoEs layer introduces additional overhead due to the routing mechanism and due to the increased memory loads when running more than one expert per device.', 'They are more suitable for batched workloads where one can reach a good degree of arithmetic intensity.', 'Comparison with Llama 2 70B and GPT-3.5.', 'In Table 3, we report the performance of Mixtral 8x7B compared to Llama 2 70B and GPT-3.5.', 'We observe that Mixtral performs similarly or above the two other models.', 'On MMLU, Mixtral obtains a better performance, despite its significantly smaller capacity (47B tokens compared to 70B).', 'For MT Bench, we report the performance of the latest GPT-3.5-Turbo model available, gpt-3.5-turbo-1106.', '2Since Llama 2 34B was not open-sourced, we report results for Llama 1 34B.', '4', 'LLaMA 2 70B GPT-3.5 MMLU (MCQ in 57 subjects) 69.9% 70.0% 70.6% HellaSwag (10-shot) 87.1% 85.5% 86.7% ARC Challenge (25-shot) 85.1% 85.2% 85.8% WinoGrande (5-shot) 83.2% 81.6% 81.2% MBPP (pass@1) 49.8% 52.2% 60.7% GSM-8K (5-shot) 53.6% 57.1% 58.4% MT Bench (for Instruct Models) 6.86 8.32 8.30', '# Mixtral 8x7B', 'Table 3:', 'Comparison of Mixtral with Llama 2 70B and GPT-3.5.', 'Mixtral outperforms or matches Llama 2 70B and GPT-3.5 performance on most metrics.', 'Evaluation Differences.', 'On some benchmarks, there are some differences between our evaluation protocol and the one reported in the Llama 2 paper: 1) on MBPP, we use the hand-verified subset 2) on TriviaQA, we do not provide Wikipedia contexts.', '# 3.1 Multilingual benchmarks', 'Compared to Mistral 7B, we significantly upsample the proportion of multilingual data during pretraining.', 'The extra capacity allows Mixtral to perform well on multilingual benchmarks while maintaining a high accuracy in English.', 'In particular, Mixtral significantly outperforms Llama 2 70B in French, German, Spanish, and Italian, as shown in Table 4.', 'Active Params French Arc-c HellaS MMLU German Arc-c HellaS MMLU Spanish Arc-c HellaS MMLU Italian Arc-c HellaS MMLU 33B 70B 13B 42.9% 65.4% 49.0% 39.3% 68.1% 49.9% 49.9% 72.5% 64.3% 49.4% 70.9% 65.1% 58.2% 77.4% 70.9% 54.3% 73.0% 71.5% 55.4% 77.6% 72.5% 52.8% 75.1% 70.9% 41.1% 63.3% 48.7% 47.3% 68.7% 64.2% 45.7% 69.8% 52.3% 50.5% 74.5% 66.0%', 'Table 4:', 'Comparison of Mixtral with Llama on Multilingual Benchmarks.', 'On ARC Challenge, Hellaswag, and MMLU, Mixtral outperforms Llama 2 70B on 4 languages:', 'French, German, Spanish, and Italian.', '# 3.2 Long range performance', 'To assess the capabilities of Mixtral to tackle long context, we evaluate it on the passkey retrieval task introduced in [23], a synthetic task designed to measure the ability of the model to retrieve a passkey inserted randomly in a long prompt.', 'Results in Figure 4 (Left) show that Mixtral achieves a 100% retrieval accuracy regardless of the context length or the position of passkey in the sequence.', 'Figure 4 (Right) shows that the perplexity of Mixtral on a subset of the proof-pile dataset [2] decreases monotonically as the size of the context increases.', 'Passkey Performance ry 0.8 0.6 04 0.2 0.0 OK 4K 8K 12K 16K 20K 24K 28K Seq Len Passkey Loc', '3.8 â', 'Mixtral_8x7B 3.5 32 > $3.0 i] 228 fos a 2.0 0 5k 10k 15k 20k 25k 30k Context length', 'Passkey Performance ry 3.8 â', 'Mixtral_8x7B 3.5 0.8 32 > 0.6 $3.0 i] 228 04 fos 0.2 a 2.0 0.0 OK 4K 8K 12K 16K 20K 24K 28K 0 5k 10k 15k 20k 25k 30k Seq Len Context length', 'Figure 4:', 'Long range performance of Mixtral. (Left) Mixtral has 100% retrieval accuracy of the Passkey task regardless of the location of the passkey and length of the input sequence. (Right) The perplexity of Mixtral on the proof-pile dataset decreases monotonically as the context length increases.', '5', '# 3.3 Bias Benchmarks', 'To identify possible flaws to be corrected by fine-tuning / preference modeling, we measure the base model performance on Bias Benchmark for QA (BBQ) [24] and Bias in Open-Ended Language Generation Dataset (BOLD) [10].', 'BBQ is a dataset of hand-written question sets that target attested social biases against nine differ- ent socially-relevant categories: age, dis- ability status, gender identity, nationality, physical appearance, race/ethnicity, religion, socio-economic status, sexual orientation.', 'BOLD is a large-scale dataset that consists of 23,679 English text generation prompts for bias benchmarking across five domains.', 'Llama 2 70B Mixtral 8x7B BBQ accuracy 51.5% 56.0% BOLD sentiment score (avg Â± std) gender profession religious_ideology political_ideology race 0.293 Â± 0.073 0.218 Â± 0.073 0.188 Â± 0.133 0.149 Â± 0.140 0.232 Â± 0.049 0.323 Â±0.045 0.243 Â± 0.087 0.144 Â± 0.089 0.186 Â± 0.146 0.232 Â± 0.052', 'Figure 5:', 'Bias Benchmarks.', 'Compared Llama 2 70B, Mixtral presents less bias (higher accuracy on BBQ, lower std on BOLD) and displays more positive sentiment (higher avg on BOLD).', 'We benchmark Llama 2 and Mixtral on BBQ and BOLD with our evaluation framework and report the results in Table 5.', 'Compared to Llama 2, Mixtral presents less bias on the BBQ benchmark (56.0% vs 51.5%).', 'For each group in BOLD, a higher average sentiment score means more positive sentiments and a lower standard deviation indicates less bias within the group.', 'Overall, Mixtral displays more positive sentiments than Llama 2, with similar variances within each group.', '# Instruction Fine-tuning', 'We train Mixtral â', 'Instruct using supervised fine-tuning (SFT) on an instruction dataset followed by Direct Preference Optimization (DPO) [25] on a paired feedback dataset.', 'Mixtral â', 'Instruct reaches a score of 8.30 on MT-Bench [33] (see Table 2), making it the best open-weights model as of December 2023.', 'Independent human evaluation conducted by LMSys is reported in Figure 63 and shows that Mixtral â', 'Instruct outperforms GPT-3.5-Turbo, Gemini Pro, Claude-2.1, and Llama 2 70B chat.', 'vs Arena Elo rating 1 MT-bench (score) License 1243 9.32 Proprietary 1192 8.96 Proprietary 1158 9.18 Proprietary Glaude-4 1149 7.9 Proprietary Claude-2.0 1131 8.06 Proprietary 1121 eS) Apache 2.0 Glaude-2.4 1117 8.18 Proprietary GPT-3..5-Turbo-9613 1117 8.39 Proprietary Gemini..Pro 1141 Proprietary Glas ta 1110 7.85 Proprietary Tulu-2-0P0-708 1110 7.89 AI2 ImpACT Low-risk Yi-34B-Chat 1110 Yi License GPT-3.5:Turbo-0314 1105 7.94 Proprietary Llama-2-79b-chat 1077 6.86 Llama 2 Community', 'Figure 6:', 'LMSys Leaderboard. (Screenshot from Dec 22, 2023) Mixtral 8x7B Instruct v0.1 achieves an Arena Elo rating of 1121 outperforming Claude-2.1 (1117), all versions of GPT-3.5-Turbo (1117 best), Gemini Pro (1111), and Llama-2-70b-chat (1077).', 'Mixtral is currently the best open-weights model by a large margin.', '3https://huggingface.co/spaces/lmsys/chatbot-arena-leaderboard', '6', '# 5 Routing analysis', 'In this section, we perform a small analysis on the expert selection by the router.', 'In particular, we are interested to see if during training some experts specialized to some specific domains (e.g. mathematics, biology, philosophy, etc.).', 'To investigate this, we measure the distribution of selected experts on different subsets of The Pile validation dataset [14].', 'Results are presented in Figure 7, for layers 0, 15, and 31 (layers 0 and 31 respectively being the first and the last layers of the model).', 'Surprisingly, we do not observe obvious patterns in the assignment of experts based on the topic.', 'For instance, at all layers, the distribution of expert assignment is very similar for ArXiv papers (written in Latex), for biology (PubMed Abstracts), and for Philosophy (PhilPapers) documents.', 'Only for DM Mathematics we note a marginally different distribution of experts.', 'This divergence is likely a consequence of the datasetâ', 's synthetic nature and its limited coverage of the natural language spectrum, and is particularly noticeable at the first and last layers, where the hidden states are very correlated to the input and output embeddings respectively.', 'This suggests that the router does exhibit some structured syntactic behavior.', 'Figure 8 shows examples of text from different domains (Python code, mathematics, and English), where each token is highlighted with a background color corresponding to its selected expert.', 'The figure shows that words such as â', 'selfâ', 'in Python and â', 'Questionâ', 'in English often get routed through the same expert even though they involve multiple tokens.', 'Similarly, in code, the indentation tokens are always assigned to the same experts, particularly at the first and last layers where the hidden states are more correlated to the input and output of the model.', 'We also note from Figure 8 that consecutive tokens are often assigned the same experts.', 'In fact, we observe some degree of positional locality in The Pile datasets.', 'Table 5 shows the proportion of con- secutive tokens that get the same expert assignments per domain and layer.', 'The proportion of repeated', '0.20 0.15 0.10 0.05 layer: 15 0.20 0.15 0.10 0.05 layer: 31 Selection proportion 0.20 0.15 0.10 0.05 Expert ID | | ArXiv | Github | | PhilPapers | StackExchange | | DM Mathematics | | Gutenberg | | PubMed Abstracts | | Wikipedia (en)', 'Figure 7:', 'Proportion of tokens assigned to each expert on different domains from The Pile dataset for layers 0, 15, and 31.', 'The gray dashed vertical line marks 1/8, i.e. the proportion expected with uniform sampling.', 'Here, we consider experts that are either selected as a first or second choice by the router.', 'A breakdown of the proportion of assignments done in each case cane be seen in Figure 9 in the Appendix.', '7', 'Layer 0 First choice Layer 15 Layer 31 Layer 0 First or second choice Layer 15 Layer 31 ArXiv DM Mathematics Github Gutenberg PhilPapers PubMed Abstracts StackExchange Wikipedia (en) 14.0% 14.1% 14.9% 13.9% 13.6% 14.2% 13.6% 14.4% 27.9% 28.4% 28.1% 26.1% 25.3% 24.6% 27.2% 23.6% 22.7% 19.7% 19.7% 26.3% 22.1% 22.0% 23.6% 25.3% 46.5% 44.9% 49.9% 49.5% 46.9% 48.6% 48.2% 49.8% 62.3% 67.0% 66.9% 63.1% 61.9% 61.6% 64.6% 62.1% 52.9% 44.5% 49.2% 52.2% 51.3% 51.8% 53.6% 51.8%', 'Table 5:', 'Percentage of expert assignment repetitions.', 'We evaluate the proportion of times the same expert is assigned to a token i and its following token i+1.', 'We report whether the first chosen expert is the same, or whether the same expert is observed as first or second choice in consecutive tokens.', 'For reference, the expected proportion of repetitions in the case of random assignments is 1 5 7 â', '46% for â', 'First and second choiceâ', '.', 'Repetitions at the first layer are close to random, but are significantly higher at layers 15 and 31.', 'The high number of repetitions shows that expert choice exhibits high temporal locality at these layers.', 'consecutive assignments is significantly higher than random for higher layers.', 'This has implications in how one might optimize the model for fast training and inference.', 'For example, cases with high locality are more likely to cause over-subscription of certain experts when doing Expert Parallelism.', 'Conversely, this locality can be leveraged for caching, as is done in [11].', 'A more complete view of these same expert frequency is provided for all layers and across datasets in Figure 10 in the Appendix.', '# 6 Conclusion', 'In this paper, we introduced Mixtral 8x7B, the first mixture-of-experts network to reach a state-of-the- art performance among open-source models.', 'Mixtral 8x7B Instruct outperforms Claude-2.1, Gem- ini Pro, and GPT-3.5 Turbo on human evaluation benchmarks.', 'Because it only uses two experts at each time step, Mixtral only uses 13B active parameters per token while outperforming the previous best model using 70B parameters per token (Llama 2 70B).', 'We are making our trained and fine-tuned mod- els publicly available under the Apache 2.0 license.', 'By sharing our models, we aim to facilitate the de- velopment of new techniques and applications that can benefit a wide range of industries and domains.', 'Layer 0 Layer 15 Layer 31 class MoeLayer(nn.', 'Module) : â', \"init__(self, experts//List [nn.Modutel,) | Super (V7 init assert len(experts) > 0 self. experts = nn.ModuleList((experts) self. gate = gate self.args = moe_args def forward(self, inputs: torch.Tensor): inputs _squashed = inputs. view(-1,_ inputs.| gate_logits = self.gatel inputs_squashed) weights, selected_experts = torch. topk( gate_logits, Self-args.nun_experts_Ã© weights! = nri.|funct ional softinax'( weights, din=1, dtype=torch. float, ).type_as|(inputs) results| = torch. zeros_ ike! linputs_squashe for i, expert in enunerate(self. experts): batch_idx,! nth_expert = torch. wnere( results [batch_idx] += weights [batch_i input s_squashed [batch_idx] ) return resutts:.view las{(inputs) class NoeLayer (nn.\", 'Module) = def _ init__(self, experts!', \"List'{nri.Modulelly Super (Tz init_t assert len (experts) > 9) self.experts = nn.\", 'ModuleList((experits)) def forward(self, inputs: torch.', 'Tensor)?! inputs_squashed = inputs.View(-1) inputs) gate_logits = self.gatel inputs_squashed) weights, selected_experts = torch. topk( getellogits, self.argssnun_experts pe weightsâ', '= nn. functionallsoftmax(Â® Weights, dtypextorch. floaty ) type_as (inputs) results| = torch. zerdsillikel(input siiequashe| for i, expert in enumerate (self. experts): batch idx, nth_expert = torch.where(s results [batch_idx] += weights [batch_iÂ¢ inputs|_squashed[batch idx], y return resultsiiview jas (inputs) class| MoeLayer(nn.', 'Module): def init__(self, expertsâ', 'List|fifi.Modulel) Super(Ve_init_O) assert len(experts) > 0 self, experts = nn.ModuleListl(@xperits)) self. gate = gate Self.args = moe_args def forward(self, inputs: torch.', 'Tensor): inputs_squashed = inputs.view(=1, inputs) gate_logits = self.gate( inputs_squashed) weights, selected_experts = torch. topk( gate_logits, self.argssfum_experts_pe weights) nni.unct iorial.isoftinax( YP Yiitype_as (inputs) results = torch. zerosillikel(inputslisquashe| for i, expert in enunerate(self.experts): batch_idx, nth_expert = torch.where(s results [batch_idx] += weights [batch_iÂ¢ inputs_squashed [batch_idx] ) return) results\\\\iviewilas|(inputs)) Tuestiond] Solve â', 'AINr 27K SLIT! and SORT, lanswers 4 Question?â', 'Calculate Baiasoazusaaly 4111270 iAnswer: -841469015.544 (Question!', 'Letâ', 'x(gy = 94g # Hl Let! q(clJ= Zee #] IAnswer:', 'S4ea - 30 â', 'Question#!', 'Solve Azer Â¥ 27HE = Ate and 1505 lanswer:) 4 Calculate ~eaieseiaz. saa Â¥ 417127. ~841469015.544 â', 'Answer: (Questor â', 'Answer: etâ', 'x(q) = 9*g Â¥ Wl Let! ql)! = 2eele Sara â', '30 question Solve -42Â¥e1E B7eC= â', 'Ad67 and 130%] answers \\\\questionÂ®| calculate savesona2.saq + auaz7.', 'Answer: -847469015.544 â', 'OÂ¥o)H A Let q(el = (questiond!', 'Let! x(a) = awed | Answers 54a ~ â', 'A model airplane flies stower when flying into tt jwind and faster with wind at its back. when Launcl Iright angles to the wind,â', 'cross wind,| its groun Icompared with! flying in still air is (A) the same (B) greater (C) less (0)! either! grea lor less dependingâ', 'on wind speed i nodelaitp ane) URE slover when flying into eH lind and faster with wind at its back.', 'When) launch Tight angles to the wind, a cross wind,. its) grounc Compared with â', 'lying in stitt air is (A) the same (18) greater) (C) less (D)! either grea lor less depending on wind speed H model airplane flies slower! when flying inte th wind and faster with wind at its backâ', '.', 'When Launcl [right angles to the wind, a cross wind, its grounc Icompared with flying in still air is (A) the sane (B) greater (C) less (0)! either gree jor less depending on wind speed', 'Figure 8:', 'Text samples where each token is colored with the first expert choice.', 'The selection of experts appears to be more aligned with the syntax rather than the domain, especially at the initial and final layers.', '8', '# Acknowledgements', 'We thank the CoreWeave and Scaleway teams for technical support as we trained our models.', 'We are grateful to NVIDIA for supporting us in integrating TensorRT-LLM and Triton and working alongside us to make a sparse mixture of experts compatible with TensorRT-LLM.', '# References', '[1] Jacob Austin, Augustus Odena, Maxwell Nye, Maarten Bosma, Henryk Michalewski, David Dohan, Ellen Jiang, Carrie Cai, Michael Terry, Quoc Le, et al.', 'Program synthesis with large language models. arXiv preprint arXiv:2108.07732, 2021.', '[2] Zhangir Azerbayev, Hailey Schoelkopf, Keiran Paster, Marco Dos Santos, Stephen McAleer, Albert Q Jiang, Jia Deng, Stella Biderman, and Sean Welleck.', 'Llemma:', 'An open language model for mathematics. arXiv preprint arXiv:2310.10631, 2023.', '[3] Yonatan Bisk, Rowan Zellers, Jianfeng Gao, Yejin Choi, et al.', 'Piqa:', 'Reasoning about phys- ical commonsense in natural language.', 'In Proceedings of the AAAI conference on artificial intelligence, pages 7432â', '7439, 2020.', '[4] Mark Chen, Jerry Tworek, Heewoo Jun, Qiming Yuan, Henrique Ponde de Oliveira Pinto, Jared Kaplan, Harri Edwards, Yuri Burda, Nicholas Joseph, Greg Brockman, et al.', 'Evaluating large language models trained on code. arXiv preprint arXiv:2107.03374, 2021.', '[5] Eunsol Choi, He He, Mohit Iyyer, Mark Yatskar, Wen-tau Yih, Yejin Choi, Percy Liang, and Luke Zettlemoyer.', 'Quac:', 'Question answering in context. arXiv preprint arXiv:1808.07036, 2018.', '[6] Aidan Clark, Diego De Las Casas, Aurelia Guy, Arthur Mensch, Michela Paganini, Jordan Hoffmann, Bogdan Damoc, Blake Hechtman, Trevor Cai, Sebastian Borgeaud, et al.', 'Unified scaling laws for routed language models.', 'In International Conference on Machine Learning, pages 4057â', '4086.', 'PMLR, 2022.', '[7] Christopher Clark, Kenton Lee, Ming-Wei Chang, Tom Kwiatkowski, Michael Collins, and Kristina Toutanova.', 'Boolq:', 'Exploring the surprising difficulty of natural yes/no questions. arXiv preprint arXiv:1905.10044, 2019.', '[8] Peter Clark, Isaac Cowhey, Oren Etzioni, Tushar Khot, Ashish Sabharwal, Carissa Schoenick, and Oyvind Tafjord.', 'Think you have solved question answering? try arc, the ai2 reasoning challenge. arXiv preprint arXiv:1803.05457, 2018.', '[9] Karl Cobbe, Vineet Kosaraju, Mohammad Bavarian, Mark Chen, Heewoo Jun, Lukasz Kaiser, Matthias Plappert, Jerry Tworek, Jacob Hilton, Reiichiro Nakano, et al.', 'Training verifiers to solve math word problems. arXiv preprint arXiv:2110.14168, 2021.', '[10] Jwala Dhamala, Tony Sun, Varun Kumar, Satyapriya Krishna, Yada Pruksachatkun, Kai-Wei Chang, and Rahul Gupta.', 'Bold:', 'Dataset and metrics for measuring biases in open-ended language generation.', 'In Proceedings of the 2021 ACM conference on fairness, accountability, and transparency, pages 862â', '872, 2021.', '[11] Artyom Eliseev and Denis Mazur.', 'Fast inference of mixture-of-experts language models with offloading. arXiv preprint arXiv:2312.17238, 2023.', '[12] William Fedus, Jeff Dean, and Barret Zoph.', 'A review of sparse expert models in deep learning. arXiv preprint arXiv:2209.01667, 2022.', '[13] Trevor Gale, Deepak Narayanan, Cliff Young, and Matei Zaharia.', 'Megablocks:', 'Efficient sparse training with mixture-of-experts. arXiv preprint arXiv:2211.15841, 2022.', '[14] Leo Gao, Stella Biderman, Sid Black, Laurence Golding, Travis Hoppe, Charles Foster, Jason Phang, Horace He, Anish Thite, Noa Nabeshima, et al.', 'The pile:', 'An 800gb dataset of diverse text for language modeling. arXiv preprint arXiv:2101.00027, 2020.', '[15] Hussein Hazimeh, Zhe Zhao, Aakanksha Chowdhery, Maheswaran Sathiamoorthy, Yihua Chen, Rahul Mazumder, Lichan Hong, and Ed Chi.', 'Dselect-k:', 'Differentiable selection in the mixture of experts with applications to multi-task learning.', 'Advances in Neural Information Processing Systems, 34:29335â', '29347, 2021.', '9', '[16] Dan Hendrycks, Collin Burns, Steven Basart, Andy Zou, Mantas Mazeika, Dawn Song, and Jacob Steinhardt.', 'Measuring massive multitask language understanding. arXiv preprint arXiv:2009.03300, 2020.', '[17] Dan Hendrycks, Collin Burns, Saurav Kadavath, Akul Arora, Steven Basart, Eric Tang, Dawn Song, and Jacob Steinhardt.', 'Measuring mathematical problem solving with the math dataset. arXiv preprint arXiv:2103.03874, 2021.', '[18] Albert Q Jiang, Alexandre Sablayrolles, Arthur Mensch, Chris Bamford, Devendra Singh Chaplot, Diego de las Casas, Florian Bressand, Gianna Lengyel, Guillaume Lample, Lucile Saulnier, et al.', 'Mistral 7b. arXiv preprint arXiv:2310.06825, 2023.', '[19] Mandar Joshi, Eunsol Choi, Daniel S Weld, and Luke Zettlemoyer.', 'Triviaqa:', 'A large scale distantly supervised challenge dataset for reading comprehension. arXiv preprint arXiv:1705.03551, 2017.', '[20] Tom Kwiatkowski, Jennimaria Palomaki, Olivia Redfield, Michael Collins, Ankur Parikh, Chris Alberti, Danielle Epstein, Illia Polosukhin, Jacob Devlin, Kenton Lee, et al.', 'Natural questions: a benchmark for question answering research.', 'Transactions of the Association for Computational Linguistics, pages 453â', '466, 2019.', '[21] Dmitry Lepikhin, HyoukJoong Lee, Yuanzhong Xu, Dehao Chen, Orhan Firat, Yanping Huang, Maxim Krikun, Noam Shazeer, and Zhifeng Chen.', 'Gshard:', 'Scaling giant models with condi- tional computation and automatic sharding. arXiv preprint arXiv:2006.16668, 2020.', '[22] Todor Mihaylov, Peter Clark, Tushar Khot, and Ashish Sabharwal.', 'Can a suit of armor conduct electricity? a new dataset for open book question answering. arXiv preprint arXiv:1809.02789, 2018.', '[23] Amirkeivan Mohtashami and Martin Jaggi.', 'Landmark attention:', 'Random-access infinite context length for transformers. arXiv preprint arXiv:2305.16300, 2023.', '[24] Alicia Parrish, Angelica Chen, Nikita Nangia, Vishakh Padmakumar, Jason Phang, Jana Thomp- son, Phu Mon Htut, and Samuel R Bowman.', 'Bbq:', 'A hand-built bias benchmark for question answering. arXiv preprint arXiv:2110.08193, 2021.', '[25] Rafael Rafailov, Archit Sharma, Eric Mitchell, Stefano Ermon, Christopher D Manning, and Chelsea Finn.', 'Direct preference optimization:', 'Your language model is secretly a reward model. arXiv preprint arXiv:2305.18290, 2023.', '[26] Keisuke Sakaguchi, Ronan Le Bras, Chandra Bhagavatula, and Yejin Choi.', 'Winogrande:', 'An adversarial winograd schema challenge at scale.', 'Communications of the ACM, pages 99â', '106, 2021.', '[27] Maarten Sap, Hannah Rashkin, Derek Chen, Ronan LeBras, and Yejin Choi.', 'Socialiqa:', 'Com- monsense reasoning about social interactions. arXiv preprint arXiv:1904.09728, 2019.', '[28] Noam Shazeer, Azalia Mirhoseini, Krzysztof Maziarz, Andy Davis, Quoc Le, Geoffrey Hinton, and Jeff Dean.', 'Outrageously large neural networks:', 'The sparsely-gated mixture-of-experts layer. arXiv preprint arXiv:1701.06538, 2017.', '[29] Mirac Suzgun, Nathan Scales, Nathanael SchÃ¤rli, Sebastian Gehrmann, Yi Tay, Hyung Won Chung, Aakanksha Chowdhery, Quoc V Le, Ed H Chi, Denny Zhou, , and Jason Wei.', 'Challenging big-bench tasks and whether chain-of-thought can solve them. arXiv preprint arXiv:2210.09261, 2022.', '[30] Alon Talmor, Jonathan Herzig, Nicholas Lourie, and Jonathan Berant.', 'Commonsenseqa:', 'A ques- tion answering challenge targeting commonsense knowledge. arXiv preprint arXiv:1811.00937, 2018.', '[31] Ashish Vaswani, Noam Shazeer, Niki Parmar, Jakob Uszkoreit, Llion Jones, Aidan N Gomez, Å', 'ukasz Kaiser, and Illia Polosukhin.', 'Attention is all you need.', 'Advances in neural information processing systems, 30, 2017.', '[32] Rowan Zellers, Ari Holtzman, Yonatan Bisk, Ali Farhadi, and Yejin Choi.', 'Hellaswag:', 'Can a machine really finish your sentence? arXiv preprint arXiv:1905.07830, 2019.', '[33] Lianmin Zheng, Wei-Lin Chiang, Ying Sheng, Siyuan Zhuang, Zhanghao Wu, Yonghao Zhuang, Zi Lin, Zhuohan Li, Dacheng Li, Eric Xing, et al.', 'Judging llm-as-a-judge with mt-bench and chatbot arena. arXiv preprint arXiv:2306.05685, 2023.', '10', '[34] Wanjun Zhong, Ruixiang Cui, Yiduo Guo, Yaobo Liang, Shuai Lu, Yanlin Wang, Amin Saied, Weizhu Chen, and Nan Duan.', 'Agieval:', 'A human-centric benchmark for evaluating foundation models. arXiv preprint arXiv:2304.06364, 2023.', '[35] Yanqi Zhou, Tao Lei, Hanxiao Liu, Nan Du, Yanping Huang, Vincent Zhao, Andrew M Dai, Quoc V Le, James Laudon, et al.', 'Mixture-of-experts with expert choice routing.', 'Advances in Neural Information Processing Systems, 35:7103â', '7114, 2022.', '11', '# Either choice', '0', 'Layer -- 0.3 0.2 0 Layer 0 -- First choice 0.3 Layer 0 -- Second choice 0.3 < 2 t Layer 15 -- First choice fe} Q 0.3 Â° a 0.2 el (el er rere! ie it len | ie} o 0 v Layer 15 -- Second choice 8 03 0.2 0 Layer 31 -- Either choice', '# Expert ID', 'ArXiv Github PhilPapers.', 'StackExchange |_| | |_| | | DM Mathematics | Gutenberg || PubMed Abstracts | Wikipedia (en)', 'Figure 9:', 'Proportion of tokens assigned to each expert on different subsets from The Pile dataset, separated by whether the expert was selected as first or second choice, or either.', 'The â', 'Either choiceâ', 'case is equivalent to Figure 7.', 'The gray dashed vertical line marks 1', '12', 'First choice 9 w is) Â° N a Â° N is) Â° An wu 0.7 0.6 Proportion of repeated assignments 0.5 Layer source â', 'e ArXiv â', 'eâ', 'DM Mathematics â', 'e Github â', 'eâ', 'Gutenberg â', 'eâ', 'PhilPapers â', 'eâ', 'PubMed â', 'e- StackExchange â', 'e-â', 'Wikipedia (en)', '# Abstracts', 'Figure 10:', 'Repeated consecutive assignments per MoE layer.', 'Repeated assignments occur a lot more often than they would with uniform assignments (materialized by the dashed lines).', 'Patterns are similar across datasets with less repetitions for DM Mathematics.', '13']: No embeddings returned. Error: 404 page not found\u001b[0m\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "No embeddings returned. Error: 404 page not found",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[20], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m logger\u001b[38;5;241m.\u001b[39msetLevel(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mWARNING\u001b[39m\u001b[38;5;124m\"\u001b[39m)  \u001b[38;5;66;03m# reduce logs from splitter\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m splits \u001b[38;5;241m=\u001b[39m \u001b[43msplitter\u001b[49m\u001b[43m(\u001b[49m\u001b[43m[\u001b[49m\u001b[43mdataset\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcontent\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/semantic_router/splitters/rolling_window.py:93\u001b[0m, in \u001b[0;36mRollingWindowSplitter.__call__\u001b[0;34m(self, docs)\u001b[0m\n\u001b[1;32m     87\u001b[0m         logger\u001b[38;5;241m.\u001b[39minfo(\n\u001b[1;32m     88\u001b[0m             \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mSingle document exceeds the maximum token limit \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     89\u001b[0m             \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mof \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmax_split_tokens\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m. \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     90\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mSplitting to sentences before semantically splitting.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     91\u001b[0m         )\n\u001b[1;32m     92\u001b[0m     docs \u001b[38;5;241m=\u001b[39m split_to_sentences(docs[\u001b[38;5;241m0\u001b[39m])\n\u001b[0;32m---> 93\u001b[0m encoded_docs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_encode_documents\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdocs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     94\u001b[0m similarities \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_calculate_similarity_scores(encoded_docs)\n\u001b[1;32m     95\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdynamic_threshold:\n",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/semantic_router/splitters/rolling_window.py:125\u001b[0m, in \u001b[0;36mRollingWindowSplitter._encode_documents\u001b[0;34m(self, docs)\u001b[0m\n\u001b[1;32m    123\u001b[0m batch_docs \u001b[38;5;241m=\u001b[39m docs[i : i \u001b[38;5;241m+\u001b[39m max_docs_per_batch]\n\u001b[1;32m    124\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 125\u001b[0m     batch_embeddings \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mencoder\u001b[49m\u001b[43m(\u001b[49m\u001b[43mbatch_docs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    126\u001b[0m     embeddings\u001b[38;5;241m.\u001b[39mextend(batch_embeddings)\n\u001b[1;32m    127\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/semantic_router/encoders/openai.py:112\u001b[0m, in \u001b[0;36mOpenAIEncoder.__call__\u001b[0;34m(self, docs, truncate)\u001b[0m\n\u001b[1;32m    106\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m (\n\u001b[1;32m    107\u001b[0m     \u001b[38;5;129;01mnot\u001b[39;00m embeds\n\u001b[1;32m    108\u001b[0m     \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(embeds, CreateEmbeddingResponse)\n\u001b[1;32m    109\u001b[0m     \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m embeds\u001b[38;5;241m.\u001b[39mdata\n\u001b[1;32m    110\u001b[0m ):\n\u001b[1;32m    111\u001b[0m     logger\u001b[38;5;241m.\u001b[39minfo(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mReturned embeddings: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00membeds\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m--> 112\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mNo embeddings returned. Error: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00merror_message\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    114\u001b[0m embeddings \u001b[38;5;241m=\u001b[39m [embeds_obj\u001b[38;5;241m.\u001b[39membedding \u001b[38;5;28;01mfor\u001b[39;00m embeds_obj \u001b[38;5;129;01min\u001b[39;00m embeds\u001b[38;5;241m.\u001b[39mdata]\n\u001b[1;32m    115\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m embeddings\n",
      "\u001b[0;31mValueError\u001b[0m: No embeddings returned. Error: 404 page not found"
     ]
    }
   ],
   "source": [
    "logger.setLevel(\"WARNING\")  # reduce logs from splitter\n",
    "splits = splitter([dataset[\"content\"][0]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "22df4d6b-61ad-42e7-ae14-4e2c4d478fc4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on class RollingWindowSplitter in module semantic_router.splitters.rolling_window:\n",
      "\n",
      "class RollingWindowSplitter(semantic_router.splitters.base.BaseSplitter)\n",
      " |  RollingWindowSplitter(encoder: semantic_router.encoders.base.BaseEncoder, name='rolling_window_splitter', threshold_adjustment=0.01, dynamic_threshold: bool = True, window_size=5, min_split_tokens=100, max_split_tokens=300, split_tokens_tolerance=10, plot_splits=False, enable_statistics=False) -> None\n",
      " |  \n",
      " |  Method resolution order:\n",
      " |      RollingWindowSplitter\n",
      " |      semantic_router.splitters.base.BaseSplitter\n",
      " |      pydantic.v1.main.BaseModel\n",
      " |      pydantic.v1.utils.Representation\n",
      " |      builtins.object\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __call__(self, docs: List[str]) -> List[semantic_router.schema.DocumentSplit]\n",
      " |      Split documents into smaller chunks based on semantic similarity.\n",
      " |      \n",
      " |      :param docs: list of text documents to be split, if only wanted to\n",
      " |          split a single document, pass it as a list with a single element.\n",
      " |      \n",
      " |      :return: list of DocumentSplit objects containing the split documents.\n",
      " |  \n",
      " |  __init__(self, encoder: semantic_router.encoders.base.BaseEncoder, name='rolling_window_splitter', threshold_adjustment=0.01, dynamic_threshold: bool = True, window_size=5, min_split_tokens=100, max_split_tokens=300, split_tokens_tolerance=10, plot_splits=False, enable_statistics=False)\n",
      " |      Create a new model by parsing and validating input data from keyword arguments.\n",
      " |      \n",
      " |      Raises ValidationError if the input data cannot be parsed to form a valid model.\n",
      " |  \n",
      " |  plot_sentence_similarity_scores(self, docs: List[str], threshold: float, window_size: int)\n",
      " |  \n",
      " |  plot_similarity_scores(self, similarities: List[float], split_indices: List[int], splits: List[semantic_router.schema.DocumentSplit])\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Static methods defined here:\n",
      " |  \n",
      " |  __json_encoder__ = pydantic_encoder(obj: Any) -> Any\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes defined here:\n",
      " |  \n",
      " |  __abstractmethods__ = frozenset()\n",
      " |  \n",
      " |  __annotations__ = {}\n",
      " |  \n",
      " |  __class_vars__ = set()\n",
      " |  \n",
      " |  __config__ = <class 'pydantic.v1.config.Config'>\n",
      " |  \n",
      " |  __custom_root_type__ = False\n",
      " |  \n",
      " |  __exclude_fields__ = None\n",
      " |  \n",
      " |  __fields__ = {'encoder': ModelField(name='encoder', type=BaseEncoder, ...\n",
      " |  \n",
      " |  __hash__ = None\n",
      " |  \n",
      " |  __include_fields__ = None\n",
      " |  \n",
      " |  __post_root_validators__ = []\n",
      " |  \n",
      " |  __pre_root_validators__ = []\n",
      " |  \n",
      " |  __private_attributes__ = {}\n",
      " |  \n",
      " |  __schema_cache__ = {}\n",
      " |  \n",
      " |  __signature__ = <Signature (encoder: semantic_router.encoders.ba...t_s...\n",
      " |  \n",
      " |  __validators__ = {}\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from semantic_router.splitters.base.BaseSplitter:\n",
      " |  \n",
      " |  print(self, document_splits: List[semantic_router.schema.DocumentSplit]) -> None\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes inherited from semantic_router.splitters.base.BaseSplitter:\n",
      " |  \n",
      " |  Config = <class 'semantic_router.splitters.base.BaseSplitter.Config'>\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from pydantic.v1.main.BaseModel:\n",
      " |  \n",
      " |  __eq__(self, other: Any) -> bool\n",
      " |      Return self==value.\n",
      " |  \n",
      " |  __getstate__(self) -> 'DictAny'\n",
      " |      Helper for pickle.\n",
      " |  \n",
      " |  __iter__(self) -> 'TupleGenerator'\n",
      " |      so `dict(model)` works\n",
      " |  \n",
      " |  __repr_args__(self) -> 'ReprArgs'\n",
      " |      Returns the attributes to show in __str__, __repr__, and __pretty__ this is generally overridden.\n",
      " |      \n",
      " |      Can either return:\n",
      " |      * name - value pairs, e.g.: `[('foo_name', 'foo'), ('bar_name', ['b', 'a', 'r'])]`\n",
      " |      * or, just values, e.g.: `[(None, 'foo'), (None, ['b', 'a', 'r'])]`\n",
      " |  \n",
      " |  __setattr__(self, name, value)\n",
      " |      Implement setattr(self, name, value).\n",
      " |  \n",
      " |  __setstate__(self, state: 'DictAny') -> None\n",
      " |  \n",
      " |  copy(self: 'Model', *, include: Union[ForwardRef('AbstractSetIntStr'), ForwardRef('MappingIntStrAny'), NoneType] = None, exclude: Union[ForwardRef('AbstractSetIntStr'), ForwardRef('MappingIntStrAny'), NoneType] = None, update: Optional[ForwardRef('DictStrAny')] = None, deep: bool = False) -> 'Model'\n",
      " |      Duplicate a model, optionally choose which fields to include, exclude and change.\n",
      " |      \n",
      " |      :param include: fields to include in new model\n",
      " |      :param exclude: fields to exclude from new model, as with values this takes precedence over include\n",
      " |      :param update: values to change/add in the new model. Note: the data is not validated before creating\n",
      " |          the new model: you should trust this data\n",
      " |      :param deep: set to `True` to make a deep copy of the model\n",
      " |      :return: new model instance\n",
      " |  \n",
      " |  dict(self, *, include: Union[ForwardRef('AbstractSetIntStr'), ForwardRef('MappingIntStrAny'), NoneType] = None, exclude: Union[ForwardRef('AbstractSetIntStr'), ForwardRef('MappingIntStrAny'), NoneType] = None, by_alias: bool = False, skip_defaults: Optional[bool] = None, exclude_unset: bool = False, exclude_defaults: bool = False, exclude_none: bool = False) -> 'DictStrAny'\n",
      " |      Generate a dictionary representation of the model, optionally specifying which fields to include or exclude.\n",
      " |  \n",
      " |  json(self, *, include: Union[ForwardRef('AbstractSetIntStr'), ForwardRef('MappingIntStrAny'), NoneType] = None, exclude: Union[ForwardRef('AbstractSetIntStr'), ForwardRef('MappingIntStrAny'), NoneType] = None, by_alias: bool = False, skip_defaults: Optional[bool] = None, exclude_unset: bool = False, exclude_defaults: bool = False, exclude_none: bool = False, encoder: Optional[Callable[[Any], Any]] = None, models_as_dict: bool = True, **dumps_kwargs: Any) -> str\n",
      " |      Generate a JSON representation of the model, `include` and `exclude` arguments as per `dict()`.\n",
      " |      \n",
      " |      `encoder` is an optional function to supply as `default` to json.dumps(), other arguments as per `json.dumps()`.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Class methods inherited from pydantic.v1.main.BaseModel:\n",
      " |  \n",
      " |  __get_validators__() -> 'CallableGenerator' from pydantic.v1.main.ModelMetaclass\n",
      " |  \n",
      " |  __try_update_forward_refs__(**localns: Any) -> None from pydantic.v1.main.ModelMetaclass\n",
      " |      Same as update_forward_refs but will not raise exception\n",
      " |      when forward references are not defined.\n",
      " |  \n",
      " |  construct(_fields_set: Optional[ForwardRef('SetStr')] = None, **values: Any) -> 'Model' from pydantic.v1.main.ModelMetaclass\n",
      " |      Creates a new model setting __dict__ and __fields_set__ from trusted or pre-validated data.\n",
      " |      Default values are respected, but no other validation is performed.\n",
      " |      Behaves as if `Config.extra = 'allow'` was set since it adds all passed values\n",
      " |  \n",
      " |  from_orm(obj: Any) -> 'Model' from pydantic.v1.main.ModelMetaclass\n",
      " |  \n",
      " |  parse_file(path: Union[str, pathlib.Path], *, content_type: str = None, encoding: str = 'utf8', proto: pydantic.v1.parse.Protocol = None, allow_pickle: bool = False) -> 'Model' from pydantic.v1.main.ModelMetaclass\n",
      " |  \n",
      " |  parse_obj(obj: Any) -> 'Model' from pydantic.v1.main.ModelMetaclass\n",
      " |  \n",
      " |  parse_raw(b: Union[str, bytes], *, content_type: str = None, encoding: str = 'utf8', proto: pydantic.v1.parse.Protocol = None, allow_pickle: bool = False) -> 'Model' from pydantic.v1.main.ModelMetaclass\n",
      " |  \n",
      " |  schema(by_alias: bool = True, ref_template: str = '#/definitions/{model}') -> 'DictStrAny' from pydantic.v1.main.ModelMetaclass\n",
      " |  \n",
      " |  schema_json(*, by_alias: bool = True, ref_template: str = '#/definitions/{model}', **dumps_kwargs: Any) -> str from pydantic.v1.main.ModelMetaclass\n",
      " |  \n",
      " |  update_forward_refs(**localns: Any) -> None from pydantic.v1.main.ModelMetaclass\n",
      " |      Try to update ForwardRefs on fields based on this Model, globalns and localns.\n",
      " |  \n",
      " |  validate(value: Any) -> 'Model' from pydantic.v1.main.ModelMetaclass\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors inherited from pydantic.v1.main.BaseModel:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables\n",
      " |  \n",
      " |  __fields_set__\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from pydantic.v1.utils.Representation:\n",
      " |  \n",
      " |  __pretty__(self, fmt: Callable[[Any], Any], **kwargs: Any) -> Generator[Any, NoneType, NoneType]\n",
      " |      Used by devtools (https://python-devtools.helpmanual.io/) to provide a human readable representations of objects\n",
      " |  \n",
      " |  __repr__(self) -> str\n",
      " |      Return repr(self).\n",
      " |  \n",
      " |  __repr_name__(self) -> str\n",
      " |      Name of the instance's class, used in __repr__.\n",
      " |  \n",
      " |  __repr_str__(self, join_str: str) -> str\n",
      " |  \n",
      " |  __rich_repr__(self) -> 'RichReprResult'\n",
      " |      Get fields for Rich library\n",
      " |  \n",
      " |  __str__(self) -> str\n",
      " |      Return str(self).\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(RollingWindowSplitter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f225f88a-193c-450d-93f2-b0b4a98e24f9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c200ec04-77e8-41fd-bc96-c2783d99c3fc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a08f320c-d93f-4835-ae78-3d1fb78cbfd2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "80a0b5a0-4345-4dbe-aaf8-ad94ef6373da",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'UnstructuredPDFLoader' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 6\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mlangchain_chroma\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m Chroma\n\u001b[1;32m      5\u001b[0m \u001b[38;5;66;03m# Load the document, split it into chunks, embed each chunk and load it into the vector store.\u001b[39;00m\n\u001b[0;32m----> 6\u001b[0m raw_documents \u001b[38;5;241m=\u001b[39m \u001b[43mUnstructuredPDFLoader\u001b[49m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mpapers/poreC.pdf\u001b[39m\u001b[38;5;124m'\u001b[39m)\u001b[38;5;241m.\u001b[39mload()\n\u001b[1;32m      7\u001b[0m text_splitter \u001b[38;5;241m=\u001b[39m CharacterTextSplitter(chunk_size\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1000\u001b[39m, chunk_overlap\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0\u001b[39m)\n\u001b[1;32m      8\u001b[0m documents \u001b[38;5;241m=\u001b[39m text_splitter\u001b[38;5;241m.\u001b[39msplit_documents(raw_documents)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'UnstructuredPDFLoader' is not defined"
     ]
    }
   ],
   "source": [
    "from langchain_community.document_loaders import TextLoader\n",
    "from langchain_text_splitters import CharacterTextSplitter\n",
    "from langchain_chroma import Chroma\n",
    "\n",
    "# Load the document, split it into chunks, embed each chunk and load it into the vector store.\n",
    "raw_documents = UnstructuredPDFLoader('papers/Pore-C.pdf').load()\n",
    "text_splitter = CharacterTextSplitter(chunk_size=1000, chunk_overlap=0)\n",
    "documents = text_splitter.split_documents(raw_documents)\n",
    "db = Chroma.from_documents(documents, embeddings_model)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2874d5d-9e1d-458d-bf48-42f17bc1eb23",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8448927c-1b90-4d81-957a-bd656f66651f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "57ab6987-2a02-451c-85eb-0adb175e080d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import os\n",
    "from langchain.document_loaders import DirectoryLoader\n",
    "from langchain.document_loaders import UnstructuredPDFLoader\n",
    "from langchain.vectorstores import Chroma\n",
    "from langchain.embeddings import HuggingFaceEmbeddings\n",
    "import chromadb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "961d24d4-3126-4a3d-b1a1-7410cb31598b",
   "metadata": {},
   "outputs": [],
   "source": [
    "docsPath='papers/'\n",
    "dbName='tutorialDatabase'\n",
    "dbPath='databases/'\n",
    "HuggingFaceEmbeddingsModel = 'BAAI/bge-base-en-v1.5'\n",
    "chunk_size=[700]\n",
    "chunck_overlap=[200]\n",
    "v=True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "36b90dda-3415-4830-8749-69bd5394e6a2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Work Directory: /home/jpic/RAG-DEV/tutorials/RAG-with-BRAD\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Documents loading from: papers/\n"
     ]
    }
   ],
   "source": [
    "dbPath   += dbName\n",
    "\n",
    "local = os.getcwd()  ## Get local dir\n",
    "os.chdir(local)      ## shift the work dir to local dir\n",
    "\n",
    "print('\\nWork Directory: {}'.format(local)) if v else None\n",
    "\n",
    "#%% Phase 1 - Load DB\n",
    "embeddings_model = HuggingFaceEmbeddings(model_name=HuggingFaceEmbeddingsModel)\n",
    "\n",
    "print('\\nDocuments loading from:', docsPath) if v else None\n",
    "\n",
    "text_loader_kwargs={'autodetect_encoding': True}\n",
    "loader = DirectoryLoader(docsPath,\n",
    "                         glob=\"**/*.pdf\",\n",
    "                         loader_cls=UnstructuredPDFLoader, \n",
    "                         #loader_kwargs=text_loader_kwargs,\n",
    "                         show_progress=True,\n",
    "                         )\n",
    "# docs_data = loader.load()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40f6c90b-f991-4f93-ae43-cccf6d3efef9",
   "metadata": {},
   "outputs": [],
   "source": [
    "docs_data = loader.load()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d08b476c-4132-4701-bfc2-a0cf42deb30d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on package pdfminer:\n",
      "\n",
      "NAME\n",
      "    pdfminer\n",
      "\n",
      "PACKAGE CONTENTS\n",
      "    _saslprep\n",
      "    arcfour\n",
      "    ascii85\n",
      "    ccitt\n",
      "    cmapdb\n",
      "    converter\n",
      "    data_structures\n",
      "    encodingdb\n",
      "    fontmetrics\n",
      "    glyphlist\n",
      "    high_level\n",
      "    image\n",
      "    jbig2\n",
      "    latin_enc\n",
      "    layout\n",
      "    lzw\n",
      "    pdfcolor\n",
      "    pdfdevice\n",
      "    pdfdocument\n",
      "    pdffont\n",
      "    pdfinterp\n",
      "    pdfpage\n",
      "    pdfparser\n",
      "    pdftypes\n",
      "    psparser\n",
      "    runlength\n",
      "    settings\n",
      "    utils\n",
      "\n",
      "FILE\n",
      "    (built-in)\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import pdfminer\n",
    "help(pdfminer)\n",
    "from pdfminer import psparser"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e838605-5e5a-40bc-a02b-e5b0f053e930",
   "metadata": {},
   "source": [
    "### WIKI Retrieval"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "e014ac0c-ffa1-4282-9351-f1eff48063eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain_community.retrievers import WikipediaRetriever\n",
    "retriever = WikipediaRetriever(top_k_results=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "bd59252c-b938-4631-ad6b-b36ca564d63f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on WikipediaRetriever in module langchain_community.retrievers.wikipedia object:\n",
      "\n",
      "class WikipediaRetriever(langchain_core.retrievers.BaseRetriever, langchain_community.utilities.wikipedia.WikipediaAPIWrapper)\n",
      " |  WikipediaRetriever(*, wiki_client: Any = None, top_k_results: int = 3, lang: str = 'en', load_all_available_meta: bool = False, doc_content_chars_max: int = 4000, name: Optional[str] = None, tags: Optional[List[str]] = None, metadata: Optional[Dict[str, Any]] = None) -> None\n",
      " |  \n",
      " |  `Wikipedia API` retriever.\n",
      " |  \n",
      " |  It wraps load() to get_relevant_documents().\n",
      " |  It uses all WikipediaAPIWrapper arguments without any change.\n",
      " |  \n",
      " |  Method resolution order:\n",
      " |      WikipediaRetriever\n",
      " |      langchain_core.retrievers.BaseRetriever\n",
      " |      langchain_core.runnables.base.RunnableSerializable\n",
      " |      langchain_core.load.serializable.Serializable\n",
      " |      langchain_community.utilities.wikipedia.WikipediaAPIWrapper\n",
      " |      pydantic.v1.main.BaseModel\n",
      " |      pydantic.v1.utils.Representation\n",
      " |      langchain_core.runnables.base.Runnable\n",
      " |      typing.Generic\n",
      " |      abc.ABC\n",
      " |      builtins.object\n",
      " |  \n",
      " |  Static methods defined here:\n",
      " |  \n",
      " |  __json_encoder__ = pydantic_encoder(obj: Any) -> Any\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes defined here:\n",
      " |  \n",
      " |  __abstractmethods__ = frozenset()\n",
      " |  \n",
      " |  __annotations__ = {}\n",
      " |  \n",
      " |  __class_vars__ = set()\n",
      " |  \n",
      " |  __config__ = <class 'pydantic.v1.config.Config'>\n",
      " |  \n",
      " |  __custom_root_type__ = False\n",
      " |  \n",
      " |  __exclude_fields__ = None\n",
      " |  \n",
      " |  __fields__ = {'doc_content_chars_max': ModelField(name='doc_content_ch...\n",
      " |  \n",
      " |  __hash__ = None\n",
      " |  \n",
      " |  __include_fields__ = None\n",
      " |  \n",
      " |  __parameters__ = ()\n",
      " |  \n",
      " |  __post_root_validators__ = [(False, <function WikipediaAPIWrapper.vali...\n",
      " |  \n",
      " |  __pre_root_validators__ = []\n",
      " |  \n",
      " |  __private_attributes__ = {}\n",
      " |  \n",
      " |  __schema_cache__ = {}\n",
      " |  \n",
      " |  __signature__ = <Signature (*, wiki_client: Any = None, top_k_re...tad...\n",
      " |  \n",
      " |  __validators__ = {}\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from langchain_core.retrievers.BaseRetriever:\n",
      " |  \n",
      " |  async aget_relevant_documents(self, query: 'str', *, callbacks: 'Callbacks' = None, tags: 'Optional[List[str]]' = None, metadata: 'Optional[Dict[str, Any]]' = None, run_name: 'Optional[str]' = None, **kwargs: 'Any') -> 'List[Document]'\n",
      " |      [*Deprecated*] Asynchronously get documents relevant to a query.\n",
      " |      \n",
      " |      Users should favor using `.ainvoke` or `.abatch` rather than\n",
      " |      `aget_relevant_documents directly`.\n",
      " |      \n",
      " |      Args:\n",
      " |          query: string to find relevant documents for\n",
      " |          callbacks: Callback manager or list of callbacks\n",
      " |          tags: Optional list of tags associated with the retriever. Defaults to None\n",
      " |              These tags will be associated with each call to this retriever,\n",
      " |              and passed as arguments to the handlers defined in `callbacks`.\n",
      " |          metadata: Optional metadata associated with the retriever. Defaults to None\n",
      " |              This metadata will be associated with each call to this retriever,\n",
      " |              and passed as arguments to the handlers defined in `callbacks`.\n",
      " |          run_name: Optional name for the run.\n",
      " |      \n",
      " |      Returns:\n",
      " |          List of relevant documents\n",
      " |      \n",
      " |      Notes\n",
      " |      -----\n",
      " |      .. deprecated:: langchain-core==0.1.46\n",
      " |         Use ainvoke instead.\n",
      " |  \n",
      " |  async ainvoke(self, input: 'str', config: 'Optional[RunnableConfig]' = None, **kwargs: 'Any') -> 'List[Document]'\n",
      " |      Asynchronously invoke the retriever to get relevant documents.\n",
      " |      \n",
      " |      Main entry point for asynchronous retriever invocations.\n",
      " |      \n",
      " |      Args:\n",
      " |          input: The query string\n",
      " |          config: Configuration for the retriever\n",
      " |          **kwargs: Additional arguments to pass to the retriever\n",
      " |      \n",
      " |      Returns:\n",
      " |          List of relevant documents\n",
      " |      \n",
      " |      Examples:\n",
      " |      \n",
      " |      .. code-block:: python\n",
      " |      \n",
      " |          await retriever.ainvoke(\"query\")\n",
      " |  \n",
      " |  get_relevant_documents(self, query: 'str', *, callbacks: 'Callbacks' = None, tags: 'Optional[List[str]]' = None, metadata: 'Optional[Dict[str, Any]]' = None, run_name: 'Optional[str]' = None, **kwargs: 'Any') -> 'List[Document]'\n",
      " |      [*Deprecated*] Retrieve documents relevant to a query.\n",
      " |      \n",
      " |      Users should favor using `.invoke` or `.batch` rather than\n",
      " |      `get_relevant_documents directly`.\n",
      " |      \n",
      " |      Args:\n",
      " |          query: string to find relevant documents for\n",
      " |          callbacks: Callback manager or list of callbacks\n",
      " |          tags: Optional list of tags associated with the retriever. Defaults to None\n",
      " |              These tags will be associated with each call to this retriever,\n",
      " |              and passed as arguments to the handlers defined in `callbacks`.\n",
      " |          metadata: Optional metadata associated with the retriever. Defaults to None\n",
      " |              This metadata will be associated with each call to this retriever,\n",
      " |              and passed as arguments to the handlers defined in `callbacks`.\n",
      " |          run_name: Optional name for the run.\n",
      " |      \n",
      " |      Returns:\n",
      " |          List of relevant documents\n",
      " |      \n",
      " |      Notes\n",
      " |      -----\n",
      " |      .. deprecated:: langchain-core==0.1.46\n",
      " |         Use invoke instead.\n",
      " |  \n",
      " |  invoke(self, input: 'str', config: 'Optional[RunnableConfig]' = None, **kwargs: 'Any') -> 'List[Document]'\n",
      " |      Invoke the retriever to get relevant documents.\n",
      " |      \n",
      " |      Main entry point for synchronous retriever invocations.\n",
      " |      \n",
      " |      Args:\n",
      " |          input: The query string\n",
      " |          config: Configuration for the retriever\n",
      " |          **kwargs: Additional arguments to pass to the retriever\n",
      " |      \n",
      " |      Returns:\n",
      " |          List of relevant documents\n",
      " |      \n",
      " |      Examples:\n",
      " |      \n",
      " |      .. code-block:: python\n",
      " |      \n",
      " |          retriever.invoke(\"query\")\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Class methods inherited from langchain_core.retrievers.BaseRetriever:\n",
      " |  \n",
      " |  __init_subclass__(**kwargs: 'Any') -> 'None' from pydantic.v1.main.ModelMetaclass\n",
      " |      This method is called when a class is subclassed.\n",
      " |      \n",
      " |      The default implementation does nothing. It may be\n",
      " |      overridden to extend subclasses.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes inherited from langchain_core.retrievers.BaseRetriever:\n",
      " |  \n",
      " |  Config = <class 'langchain_core.retrievers.BaseRetriever.Config'>\n",
      " |      Configuration for this pydantic object.\n",
      " |  \n",
      " |  \n",
      " |  __orig_bases__ = (langchain_core.runnables.base.RunnableSerializab...i...\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from langchain_core.runnables.base.RunnableSerializable:\n",
      " |  \n",
      " |  configurable_alternatives(self, which: 'ConfigurableField', *, default_key: 'str' = 'default', prefix_keys: 'bool' = False, **kwargs: 'Union[Runnable[Input, Output], Callable[[], Runnable[Input, Output]]]') -> 'RunnableSerializable[Input, Output]'\n",
      " |      Configure alternatives for runnables that can be set at runtime.\n",
      " |      \n",
      " |      .. code-block:: python\n",
      " |      \n",
      " |          from langchain_anthropic import ChatAnthropic\n",
      " |          from langchain_core.runnables.utils import ConfigurableField\n",
      " |          from langchain_openai import ChatOpenAI\n",
      " |      \n",
      " |          model = ChatAnthropic(\n",
      " |              model_name=\"claude-3-sonnet-20240229\"\n",
      " |          ).configurable_alternatives(\n",
      " |              ConfigurableField(id=\"llm\"),\n",
      " |              default_key=\"anthropic\",\n",
      " |              openai=ChatOpenAI()\n",
      " |          )\n",
      " |      \n",
      " |          # uses the default model ChatAnthropic\n",
      " |          print(model.invoke(\"which organization created you?\").content)\n",
      " |      \n",
      " |          # uses ChatOpenaAI\n",
      " |          print(\n",
      " |              model.with_config(\n",
      " |                  configurable={\"llm\": \"openai\"}\n",
      " |              ).invoke(\"which organization created you?\").content\n",
      " |          )\n",
      " |  \n",
      " |  configurable_fields(self, **kwargs: 'AnyConfigurableField') -> 'RunnableSerializable[Input, Output]'\n",
      " |      Configure particular runnable fields at runtime.\n",
      " |      \n",
      " |      .. code-block:: python\n",
      " |      \n",
      " |          from langchain_core.runnables import ConfigurableField\n",
      " |          from langchain_openai import ChatOpenAI\n",
      " |      \n",
      " |          model = ChatOpenAI(max_tokens=20).configurable_fields(\n",
      " |              max_tokens=ConfigurableField(\n",
      " |                  id=\"output_token_number\",\n",
      " |                  name=\"Max tokens in the output\",\n",
      " |                  description=\"The maximum number of tokens in the output\",\n",
      " |              )\n",
      " |          )\n",
      " |      \n",
      " |          # max_tokens = 20\n",
      " |          print(\n",
      " |              \"max_tokens_20: \",\n",
      " |              model.invoke(\"tell me something about chess\").content\n",
      " |          )\n",
      " |      \n",
      " |          # max_tokens = 200\n",
      " |          print(\"max_tokens_200: \", model.with_config(\n",
      " |              configurable={\"output_token_number\": 200}\n",
      " |              ).invoke(\"tell me something about chess\").content\n",
      " |          )\n",
      " |  \n",
      " |  to_json(self) -> 'Union[SerializedConstructor, SerializedNotImplemented]'\n",
      " |      Serialize the runnable to JSON.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors inherited from langchain_core.runnables.base.RunnableSerializable:\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from langchain_core.load.serializable.Serializable:\n",
      " |  \n",
      " |  __repr_args__(self) -> Any\n",
      " |      Returns the attributes to show in __str__, __repr__, and __pretty__ this is generally overridden.\n",
      " |      \n",
      " |      Can either return:\n",
      " |      * name - value pairs, e.g.: `[('foo_name', 'foo'), ('bar_name', ['b', 'a', 'r'])]`\n",
      " |      * or, just values, e.g.: `[(None, 'foo'), (None, ['b', 'a', 'r'])]`\n",
      " |  \n",
      " |  to_json_not_implemented(self) -> langchain_core.load.serializable.SerializedNotImplemented\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Class methods inherited from langchain_core.load.serializable.Serializable:\n",
      " |  \n",
      " |  get_lc_namespace() -> List[str] from pydantic.v1.main.ModelMetaclass\n",
      " |      Get the namespace of the langchain object.\n",
      " |      \n",
      " |      For example, if the class is `langchain.llms.openai.OpenAI`, then the\n",
      " |      namespace is [\"langchain\", \"llms\", \"openai\"]\n",
      " |  \n",
      " |  is_lc_serializable() -> bool from pydantic.v1.main.ModelMetaclass\n",
      " |      Is this class serializable?\n",
      " |  \n",
      " |  lc_id() -> List[str] from pydantic.v1.main.ModelMetaclass\n",
      " |      A unique identifier for this class for serialization purposes.\n",
      " |      \n",
      " |      The unique identifier is a list of strings that describes the path\n",
      " |      to the object.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Readonly properties inherited from langchain_core.load.serializable.Serializable:\n",
      " |  \n",
      " |  lc_attributes\n",
      " |      List of attribute names that should be included in the serialized kwargs.\n",
      " |      \n",
      " |      These attributes must be accepted by the constructor.\n",
      " |  \n",
      " |  lc_secrets\n",
      " |      A map of constructor argument names to secret ids.\n",
      " |      \n",
      " |      For example,\n",
      " |          {\"openai_api_key\": \"OPENAI_API_KEY\"}\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from langchain_community.utilities.wikipedia.WikipediaAPIWrapper:\n",
      " |  \n",
      " |  lazy_load(self, query: str) -> Iterator[langchain_core.documents.base.Document]\n",
      " |      Run Wikipedia search and get the article text plus the meta information.\n",
      " |      See\n",
      " |      \n",
      " |      Returns: a list of documents.\n",
      " |  \n",
      " |  load(self, query: str) -> List[langchain_core.documents.base.Document]\n",
      " |      Run Wikipedia search and get the article text plus the meta information.\n",
      " |      See\n",
      " |      \n",
      " |      Returns: a list of documents.\n",
      " |  \n",
      " |  run(self, query: str) -> str\n",
      " |      Run Wikipedia search and get page summaries.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Class methods inherited from langchain_community.utilities.wikipedia.WikipediaAPIWrapper:\n",
      " |  \n",
      " |  validate_environment(values: Dict) -> Dict from pydantic.v1.main.ModelMetaclass\n",
      " |      Validate that the python package exists in environment.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from pydantic.v1.main.BaseModel:\n",
      " |  \n",
      " |  __eq__(self, other: Any) -> bool\n",
      " |      Return self==value.\n",
      " |  \n",
      " |  __getstate__(self) -> 'DictAny'\n",
      " |      Helper for pickle.\n",
      " |  \n",
      " |  __init__(__pydantic_self__, **data: Any) -> None\n",
      " |      Create a new model by parsing and validating input data from keyword arguments.\n",
      " |      \n",
      " |      Raises ValidationError if the input data cannot be parsed to form a valid model.\n",
      " |  \n",
      " |  __iter__(self) -> 'TupleGenerator'\n",
      " |      so `dict(model)` works\n",
      " |  \n",
      " |  __setattr__(self, name, value)\n",
      " |      Implement setattr(self, name, value).\n",
      " |  \n",
      " |  __setstate__(self, state: 'DictAny') -> None\n",
      " |  \n",
      " |  copy(self: 'Model', *, include: Union[ForwardRef('AbstractSetIntStr'), ForwardRef('MappingIntStrAny'), NoneType] = None, exclude: Union[ForwardRef('AbstractSetIntStr'), ForwardRef('MappingIntStrAny'), NoneType] = None, update: Optional[ForwardRef('DictStrAny')] = None, deep: bool = False) -> 'Model'\n",
      " |      Duplicate a model, optionally choose which fields to include, exclude and change.\n",
      " |      \n",
      " |      :param include: fields to include in new model\n",
      " |      :param exclude: fields to exclude from new model, as with values this takes precedence over include\n",
      " |      :param update: values to change/add in the new model. Note: the data is not validated before creating\n",
      " |          the new model: you should trust this data\n",
      " |      :param deep: set to `True` to make a deep copy of the model\n",
      " |      :return: new model instance\n",
      " |  \n",
      " |  dict(self, *, include: Union[ForwardRef('AbstractSetIntStr'), ForwardRef('MappingIntStrAny'), NoneType] = None, exclude: Union[ForwardRef('AbstractSetIntStr'), ForwardRef('MappingIntStrAny'), NoneType] = None, by_alias: bool = False, skip_defaults: Optional[bool] = None, exclude_unset: bool = False, exclude_defaults: bool = False, exclude_none: bool = False) -> 'DictStrAny'\n",
      " |      Generate a dictionary representation of the model, optionally specifying which fields to include or exclude.\n",
      " |  \n",
      " |  json(self, *, include: Union[ForwardRef('AbstractSetIntStr'), ForwardRef('MappingIntStrAny'), NoneType] = None, exclude: Union[ForwardRef('AbstractSetIntStr'), ForwardRef('MappingIntStrAny'), NoneType] = None, by_alias: bool = False, skip_defaults: Optional[bool] = None, exclude_unset: bool = False, exclude_defaults: bool = False, exclude_none: bool = False, encoder: Optional[Callable[[Any], Any]] = None, models_as_dict: bool = True, **dumps_kwargs: Any) -> str\n",
      " |      Generate a JSON representation of the model, `include` and `exclude` arguments as per `dict()`.\n",
      " |      \n",
      " |      `encoder` is an optional function to supply as `default` to json.dumps(), other arguments as per `json.dumps()`.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Class methods inherited from pydantic.v1.main.BaseModel:\n",
      " |  \n",
      " |  __get_validators__() -> 'CallableGenerator' from pydantic.v1.main.ModelMetaclass\n",
      " |  \n",
      " |  __try_update_forward_refs__(**localns: Any) -> None from pydantic.v1.main.ModelMetaclass\n",
      " |      Same as update_forward_refs but will not raise exception\n",
      " |      when forward references are not defined.\n",
      " |  \n",
      " |  construct(_fields_set: Optional[ForwardRef('SetStr')] = None, **values: Any) -> 'Model' from pydantic.v1.main.ModelMetaclass\n",
      " |      Creates a new model setting __dict__ and __fields_set__ from trusted or pre-validated data.\n",
      " |      Default values are respected, but no other validation is performed.\n",
      " |      Behaves as if `Config.extra = 'allow'` was set since it adds all passed values\n",
      " |  \n",
      " |  from_orm(obj: Any) -> 'Model' from pydantic.v1.main.ModelMetaclass\n",
      " |  \n",
      " |  parse_file(path: Union[str, pathlib.Path], *, content_type: str = None, encoding: str = 'utf8', proto: pydantic.v1.parse.Protocol = None, allow_pickle: bool = False) -> 'Model' from pydantic.v1.main.ModelMetaclass\n",
      " |  \n",
      " |  parse_obj(obj: Any) -> 'Model' from pydantic.v1.main.ModelMetaclass\n",
      " |  \n",
      " |  parse_raw(b: Union[str, bytes], *, content_type: str = None, encoding: str = 'utf8', proto: pydantic.v1.parse.Protocol = None, allow_pickle: bool = False) -> 'Model' from pydantic.v1.main.ModelMetaclass\n",
      " |  \n",
      " |  schema(by_alias: bool = True, ref_template: str = '#/definitions/{model}') -> 'DictStrAny' from pydantic.v1.main.ModelMetaclass\n",
      " |  \n",
      " |  schema_json(*, by_alias: bool = True, ref_template: str = '#/definitions/{model}', **dumps_kwargs: Any) -> str from pydantic.v1.main.ModelMetaclass\n",
      " |  \n",
      " |  update_forward_refs(**localns: Any) -> None from pydantic.v1.main.ModelMetaclass\n",
      " |      Try to update ForwardRefs on fields based on this Model, globalns and localns.\n",
      " |  \n",
      " |  validate(value: Any) -> 'Model' from pydantic.v1.main.ModelMetaclass\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors inherited from pydantic.v1.main.BaseModel:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables\n",
      " |  \n",
      " |  __fields_set__\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from pydantic.v1.utils.Representation:\n",
      " |  \n",
      " |  __pretty__(self, fmt: Callable[[Any], Any], **kwargs: Any) -> Generator[Any, NoneType, NoneType]\n",
      " |      Used by devtools (https://python-devtools.helpmanual.io/) to provide a human readable representations of objects\n",
      " |  \n",
      " |  __repr__(self) -> str\n",
      " |      Return repr(self).\n",
      " |  \n",
      " |  __repr_name__(self) -> str\n",
      " |      Name of the instance's class, used in __repr__.\n",
      " |  \n",
      " |  __repr_str__(self, join_str: str) -> str\n",
      " |  \n",
      " |  __rich_repr__(self) -> 'RichReprResult'\n",
      " |      Get fields for Rich library\n",
      " |  \n",
      " |  __str__(self) -> str\n",
      " |      Return str(self).\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from langchain_core.runnables.base.Runnable:\n",
      " |  \n",
      " |  __or__(self, other: 'Union[Runnable[Any, Other], Callable[[Any], Other], Callable[[Iterator[Any]], Iterator[Other]], Mapping[str, Union[Runnable[Any, Other], Callable[[Any], Other], Any]]]') -> 'RunnableSerializable[Input, Other]'\n",
      " |      Compose this runnable with another object to create a RunnableSequence.\n",
      " |  \n",
      " |  __ror__(self, other: 'Union[Runnable[Other, Any], Callable[[Other], Any], Callable[[Iterator[Other]], Iterator[Any]], Mapping[str, Union[Runnable[Other, Any], Callable[[Other], Any], Any]]]') -> 'RunnableSerializable[Other, Output]'\n",
      " |      Compose this runnable with another object to create a RunnableSequence.\n",
      " |  \n",
      " |  async abatch(self, inputs: 'List[Input]', config: 'Optional[Union[RunnableConfig, List[RunnableConfig]]]' = None, *, return_exceptions: 'bool' = False, **kwargs: 'Optional[Any]') -> 'List[Output]'\n",
      " |      Default implementation runs ainvoke in parallel using asyncio.gather.\n",
      " |      \n",
      " |      The default implementation of batch works well for IO bound runnables.\n",
      " |      \n",
      " |      Subclasses should override this method if they can batch more efficiently;\n",
      " |      e.g., if the underlying runnable uses an API which supports a batch mode.\n",
      " |  \n",
      " |  async abatch_as_completed(self, inputs: 'List[Input]', config: 'Optional[Union[RunnableConfig, List[RunnableConfig]]]' = None, *, return_exceptions: 'bool' = False, **kwargs: 'Optional[Any]') -> 'AsyncIterator[Tuple[int, Union[Output, Exception]]]'\n",
      " |      Run ainvoke in parallel on a list of inputs,\n",
      " |      yielding results as they complete.\n",
      " |  \n",
      " |  assign(self, **kwargs: 'Union[Runnable[Dict[str, Any], Any], Callable[[Dict[str, Any]], Any], Mapping[str, Union[Runnable[Dict[str, Any], Any], Callable[[Dict[str, Any]], Any]]]]') -> 'RunnableSerializable[Any, Any]'\n",
      " |      Assigns new fields to the dict output of this runnable.\n",
      " |      Returns a new runnable.\n",
      " |      \n",
      " |      .. code-block:: python\n",
      " |      \n",
      " |          from langchain_community.llms.fake import FakeStreamingListLLM\n",
      " |          from langchain_core.output_parsers import StrOutputParser\n",
      " |          from langchain_core.prompts import SystemMessagePromptTemplate\n",
      " |          from langchain_core.runnables import Runnable\n",
      " |          from operator import itemgetter\n",
      " |      \n",
      " |          prompt = (\n",
      " |              SystemMessagePromptTemplate.from_template(\"You are a nice assistant.\")\n",
      " |              + \"{question}\"\n",
      " |          )\n",
      " |          llm = FakeStreamingListLLM(responses=[\"foo-lish\"])\n",
      " |      \n",
      " |          chain: Runnable = prompt | llm | {\"str\": StrOutputParser()}\n",
      " |      \n",
      " |          chain_with_assign = chain.assign(hello=itemgetter(\"str\") | llm)\n",
      " |      \n",
      " |          print(chain_with_assign.input_schema.schema())\n",
      " |          # {'title': 'PromptInput', 'type': 'object', 'properties':\n",
      " |          {'question': {'title': 'Question', 'type': 'string'}}}\n",
      " |          print(chain_with_assign.output_schema.schema()) #\n",
      " |          {'title': 'RunnableSequenceOutput', 'type': 'object', 'properties':\n",
      " |          {'str': {'title': 'Str',\n",
      " |          'type': 'string'}, 'hello': {'title': 'Hello', 'type': 'string'}}}\n",
      " |  \n",
      " |  async astream(self, input: 'Input', config: 'Optional[RunnableConfig]' = None, **kwargs: 'Optional[Any]') -> 'AsyncIterator[Output]'\n",
      " |      Default implementation of astream, which calls ainvoke.\n",
      " |      Subclasses should override this method if they support streaming output.\n",
      " |  \n",
      " |  astream_events(self, input: 'Any', config: 'Optional[RunnableConfig]' = None, *, version: \"Literal['v1']\", include_names: 'Optional[Sequence[str]]' = None, include_types: 'Optional[Sequence[str]]' = None, include_tags: 'Optional[Sequence[str]]' = None, exclude_names: 'Optional[Sequence[str]]' = None, exclude_types: 'Optional[Sequence[str]]' = None, exclude_tags: 'Optional[Sequence[str]]' = None, **kwargs: 'Any') -> 'AsyncIterator[StreamEvent]'\n",
      " |      [*Beta*] Generate a stream of events.\n",
      " |      \n",
      " |      Use to create an iterator over StreamEvents that provide real-time information\n",
      " |      about the progress of the runnable, including StreamEvents from intermediate\n",
      " |      results.\n",
      " |      \n",
      " |      A StreamEvent is a dictionary with the following schema:\n",
      " |      \n",
      " |      - ``event``: **str** - Event names are of the\n",
      " |          format: on_[runnable_type]_(start|stream|end).\n",
      " |      - ``name``: **str** - The name of the runnable that generated the event.\n",
      " |      - ``run_id``: **str** - randomly generated ID associated with the given execution of\n",
      " |          the runnable that emitted the event.\n",
      " |          A child runnable that gets invoked as part of the execution of a\n",
      " |          parent runnable is assigned its own unique ID.\n",
      " |      - ``tags``: **Optional[List[str]]** - The tags of the runnable that generated\n",
      " |          the event.\n",
      " |      - ``metadata``: **Optional[Dict[str, Any]]** - The metadata of the runnable\n",
      " |          that generated the event.\n",
      " |      - ``data``: **Dict[str, Any]**\n",
      " |      \n",
      " |      \n",
      " |      Below is a table that illustrates some evens that might be emitted by various\n",
      " |      chains. Metadata fields have been omitted from the table for brevity.\n",
      " |      Chain definitions have been included after the table.\n",
      " |      \n",
      " |      +----------------------+------------------+---------------------------------+-----------------------------------------------+-------------------------------------------------+\n",
      " |      | event                | name             | chunk                           | input                                         | output                                          |\n",
      " |      +======================+==================+=================================+===============================================+=================================================+\n",
      " |      | on_chat_model_start  | [model name]     |                                 | {\"messages\": [[SystemMessage, HumanMessage]]} |                                                 |\n",
      " |      +----------------------+------------------+---------------------------------+-----------------------------------------------+-------------------------------------------------+\n",
      " |      | on_chat_model_stream | [model name]     | AIMessageChunk(content=\"hello\") |                                               |                                                 |\n",
      " |      +----------------------+------------------+---------------------------------+-----------------------------------------------+-------------------------------------------------+\n",
      " |      | on_chat_model_end    | [model name]     |                                 | {\"messages\": [[SystemMessage, HumanMessage]]} | {\"generations\": [...], \"llm_output\": None, ...} |\n",
      " |      +----------------------+------------------+---------------------------------+-----------------------------------------------+-------------------------------------------------+\n",
      " |      | on_llm_start         | [model name]     |                                 | {'input': 'hello'}                            |                                                 |\n",
      " |      +----------------------+------------------+---------------------------------+-----------------------------------------------+-------------------------------------------------+\n",
      " |      | on_llm_stream        | [model name]     | 'Hello'                         |                                               |                                                 |\n",
      " |      +----------------------+------------------+---------------------------------+-----------------------------------------------+-------------------------------------------------+\n",
      " |      | on_llm_end           | [model name]     |                                 | 'Hello human!'                                |                                                 |\n",
      " |      +----------------------+------------------+---------------------------------+-----------------------------------------------+-------------------------------------------------+\n",
      " |      | on_chain_start       | format_docs      |                                 |                                               |                                                 |\n",
      " |      +----------------------+------------------+---------------------------------+-----------------------------------------------+-------------------------------------------------+\n",
      " |      | on_chain_stream      | format_docs      | \"hello world!, goodbye world!\"  |                                               |                                                 |\n",
      " |      +----------------------+------------------+---------------------------------+-----------------------------------------------+-------------------------------------------------+\n",
      " |      | on_chain_end         | format_docs      |                                 | [Document(...)]                               | \"hello world!, goodbye world!\"                  |\n",
      " |      +----------------------+------------------+---------------------------------+-----------------------------------------------+-------------------------------------------------+\n",
      " |      | on_tool_start        | some_tool        |                                 | {\"x\": 1, \"y\": \"2\"}                            |                                                 |\n",
      " |      +----------------------+------------------+---------------------------------+-----------------------------------------------+-------------------------------------------------+\n",
      " |      | on_tool_stream       | some_tool        | {\"x\": 1, \"y\": \"2\"}              |                                               |                                                 |\n",
      " |      +----------------------+------------------+---------------------------------+-----------------------------------------------+-------------------------------------------------+\n",
      " |      | on_tool_end          | some_tool        |                                 |                                               | {\"x\": 1, \"y\": \"2\"}                              |\n",
      " |      +----------------------+------------------+---------------------------------+-----------------------------------------------+-------------------------------------------------+\n",
      " |      | on_retriever_start   | [retriever name] |                                 | {\"query\": \"hello\"}                            |                                                 |\n",
      " |      +----------------------+------------------+---------------------------------+-----------------------------------------------+-------------------------------------------------+\n",
      " |      | on_retriever_chunk   | [retriever name] | {documents: [...]}              |                                               |                                                 |\n",
      " |      +----------------------+------------------+---------------------------------+-----------------------------------------------+-------------------------------------------------+\n",
      " |      | on_retriever_end     | [retriever name] |                                 | {\"query\": \"hello\"}                            | {documents: [...]}                              |\n",
      " |      +----------------------+------------------+---------------------------------+-----------------------------------------------+-------------------------------------------------+\n",
      " |      | on_prompt_start      | [template_name]  |                                 | {\"question\": \"hello\"}                         |                                                 |\n",
      " |      +----------------------+------------------+---------------------------------+-----------------------------------------------+-------------------------------------------------+\n",
      " |      | on_prompt_end        | [template_name]  |                                 | {\"question\": \"hello\"}                         | ChatPromptValue(messages: [SystemMessage, ...]) |\n",
      " |      +----------------------+------------------+---------------------------------+-----------------------------------------------+-------------------------------------------------+\n",
      " |      \n",
      " |      Here are declarations associated with the events shown above:\n",
      " |      \n",
      " |      `format_docs`:\n",
      " |      \n",
      " |      .. code-block:: python\n",
      " |      \n",
      " |          def format_docs(docs: List[Document]) -> str:\n",
      " |              '''Format the docs.'''\n",
      " |              return \", \".join([doc.page_content for doc in docs])\n",
      " |      \n",
      " |          format_docs = RunnableLambda(format_docs)\n",
      " |      \n",
      " |      `some_tool`:\n",
      " |      \n",
      " |      .. code-block:: python\n",
      " |      \n",
      " |          @tool\n",
      " |          def some_tool(x: int, y: str) -> dict:\n",
      " |              '''Some_tool.'''\n",
      " |              return {\"x\": x, \"y\": y}\n",
      " |      \n",
      " |      `prompt`:\n",
      " |      \n",
      " |      .. code-block:: python\n",
      " |      \n",
      " |          template = ChatPromptTemplate.from_messages(\n",
      " |              [(\"system\", \"You are Cat Agent 007\"), (\"human\", \"{question}\")]\n",
      " |          ).with_config({\"run_name\": \"my_template\", \"tags\": [\"my_template\"]})\n",
      " |      \n",
      " |      \n",
      " |      Example:\n",
      " |      \n",
      " |      .. code-block:: python\n",
      " |      \n",
      " |          from langchain_core.runnables import RunnableLambda\n",
      " |      \n",
      " |          async def reverse(s: str) -> str:\n",
      " |              return s[::-1]\n",
      " |      \n",
      " |          chain = RunnableLambda(func=reverse)\n",
      " |      \n",
      " |          events = [\n",
      " |              event async for event in chain.astream_events(\"hello\", version=\"v1\")\n",
      " |          ]\n",
      " |      \n",
      " |          # will produce the following events (run_id has been omitted for brevity):\n",
      " |          [\n",
      " |              {\n",
      " |                  \"data\": {\"input\": \"hello\"},\n",
      " |                  \"event\": \"on_chain_start\",\n",
      " |                  \"metadata\": {},\n",
      " |                  \"name\": \"reverse\",\n",
      " |                  \"tags\": [],\n",
      " |              },\n",
      " |              {\n",
      " |                  \"data\": {\"chunk\": \"olleh\"},\n",
      " |                  \"event\": \"on_chain_stream\",\n",
      " |                  \"metadata\": {},\n",
      " |                  \"name\": \"reverse\",\n",
      " |                  \"tags\": [],\n",
      " |              },\n",
      " |              {\n",
      " |                  \"data\": {\"output\": \"olleh\"},\n",
      " |                  \"event\": \"on_chain_end\",\n",
      " |                  \"metadata\": {},\n",
      " |                  \"name\": \"reverse\",\n",
      " |                  \"tags\": [],\n",
      " |              },\n",
      " |          ]\n",
      " |      \n",
      " |      Args:\n",
      " |          input: The input to the runnable.\n",
      " |          config: The config to use for the runnable.\n",
      " |          version: The version of the schema to use.\n",
      " |                   Currently only version 1 is available.\n",
      " |                   No default will be assigned until the API is stabilized.\n",
      " |          include_names: Only include events from runnables with matching names.\n",
      " |          include_types: Only include events from runnables with matching types.\n",
      " |          include_tags: Only include events from runnables with matching tags.\n",
      " |          exclude_names: Exclude events from runnables with matching names.\n",
      " |          exclude_types: Exclude events from runnables with matching types.\n",
      " |          exclude_tags: Exclude events from runnables with matching tags.\n",
      " |          kwargs: Additional keyword arguments to pass to the runnable.\n",
      " |              These will be passed to astream_log as this implementation\n",
      " |              of astream_events is built on top of astream_log.\n",
      " |      \n",
      " |      Returns:\n",
      " |          An async stream of StreamEvents.\n",
      " |      \n",
      " |      Notes\n",
      " |      -----\n",
      " |      .. beta::\n",
      " |         This API is in beta and may change in the future.\n",
      " |  \n",
      " |  async astream_log(self, input: 'Any', config: 'Optional[RunnableConfig]' = None, *, diff: 'bool' = True, with_streamed_output_list: 'bool' = True, include_names: 'Optional[Sequence[str]]' = None, include_types: 'Optional[Sequence[str]]' = None, include_tags: 'Optional[Sequence[str]]' = None, exclude_names: 'Optional[Sequence[str]]' = None, exclude_types: 'Optional[Sequence[str]]' = None, exclude_tags: 'Optional[Sequence[str]]' = None, **kwargs: 'Any') -> 'Union[AsyncIterator[RunLogPatch], AsyncIterator[RunLog]]'\n",
      " |      Stream all output from a runnable, as reported to the callback system.\n",
      " |      This includes all inner runs of LLMs, Retrievers, Tools, etc.\n",
      " |      \n",
      " |      Output is streamed as Log objects, which include a list of\n",
      " |      jsonpatch ops that describe how the state of the run has changed in each\n",
      " |      step, and the final state of the run.\n",
      " |      \n",
      " |      The jsonpatch ops can be applied in order to construct state.\n",
      " |      \n",
      " |      Args:\n",
      " |          input: The input to the runnable.\n",
      " |          config: The config to use for the runnable.\n",
      " |          diff: Whether to yield diffs between each step, or the current state.\n",
      " |          with_streamed_output_list: Whether to yield the streamed_output list.\n",
      " |          include_names: Only include logs with these names.\n",
      " |          include_types: Only include logs with these types.\n",
      " |          include_tags: Only include logs with these tags.\n",
      " |          exclude_names: Exclude logs with these names.\n",
      " |          exclude_types: Exclude logs with these types.\n",
      " |          exclude_tags: Exclude logs with these tags.\n",
      " |  \n",
      " |  async atransform(self, input: 'AsyncIterator[Input]', config: 'Optional[RunnableConfig]' = None, **kwargs: 'Optional[Any]') -> 'AsyncIterator[Output]'\n",
      " |      Default implementation of atransform, which buffers input and calls astream.\n",
      " |      Subclasses should override this method if they can start producing output while\n",
      " |      input is still being generated.\n",
      " |  \n",
      " |  batch(self, inputs: 'List[Input]', config: 'Optional[Union[RunnableConfig, List[RunnableConfig]]]' = None, *, return_exceptions: 'bool' = False, **kwargs: 'Optional[Any]') -> 'List[Output]'\n",
      " |      Default implementation runs invoke in parallel using a thread pool executor.\n",
      " |      \n",
      " |      The default implementation of batch works well for IO bound runnables.\n",
      " |      \n",
      " |      Subclasses should override this method if they can batch more efficiently;\n",
      " |      e.g., if the underlying runnable uses an API which supports a batch mode.\n",
      " |  \n",
      " |  batch_as_completed(self, inputs: 'List[Input]', config: 'Optional[Union[RunnableConfig, List[RunnableConfig]]]' = None, *, return_exceptions: 'bool' = False, **kwargs: 'Optional[Any]') -> 'Iterator[Tuple[int, Union[Output, Exception]]]'\n",
      " |      Run invoke in parallel on a list of inputs,\n",
      " |      yielding results as they complete.\n",
      " |  \n",
      " |  bind(self, **kwargs: 'Any') -> 'Runnable[Input, Output]'\n",
      " |      Bind arguments to a Runnable, returning a new Runnable.\n",
      " |      \n",
      " |      Useful when a runnable in a chain requires an argument that is not\n",
      " |      in the output of the previous runnable or included in the user input.\n",
      " |      \n",
      " |      Example:\n",
      " |      \n",
      " |      .. code-block:: python\n",
      " |      \n",
      " |          from langchain_community.chat_models import ChatOllama\n",
      " |          from langchain_core.output_parsers import StrOutputParser\n",
      " |      \n",
      " |          llm = ChatOllama(model='llama2')\n",
      " |      \n",
      " |          # Without bind.\n",
      " |          chain = (\n",
      " |              llm\n",
      " |              | StrOutputParser()\n",
      " |          )\n",
      " |      \n",
      " |          chain.invoke(\"Repeat quoted words exactly: 'One two three four five.'\")\n",
      " |          # Output is 'One two three four five.'\n",
      " |      \n",
      " |          # With bind.\n",
      " |          chain = (\n",
      " |              llm.bind(stop=[\"three\"])\n",
      " |              | StrOutputParser()\n",
      " |          )\n",
      " |      \n",
      " |          chain.invoke(\"Repeat quoted words exactly: 'One two three four five.'\")\n",
      " |          # Output is 'One two'\n",
      " |  \n",
      " |  config_schema(self, *, include: 'Optional[Sequence[str]]' = None) -> 'Type[BaseModel]'\n",
      " |      The type of config this runnable accepts specified as a pydantic model.\n",
      " |      \n",
      " |      To mark a field as configurable, see the `configurable_fields`\n",
      " |      and `configurable_alternatives` methods.\n",
      " |      \n",
      " |      Args:\n",
      " |          include: A list of fields to include in the config schema.\n",
      " |      \n",
      " |      Returns:\n",
      " |          A pydantic model that can be used to validate config.\n",
      " |  \n",
      " |  get_graph(self, config: 'Optional[RunnableConfig]' = None) -> 'Graph'\n",
      " |      Return a graph representation of this runnable.\n",
      " |  \n",
      " |  get_input_schema(self, config: 'Optional[RunnableConfig]' = None) -> 'Type[BaseModel]'\n",
      " |      Get a pydantic model that can be used to validate input to the runnable.\n",
      " |      \n",
      " |      Runnables that leverage the configurable_fields and configurable_alternatives\n",
      " |      methods will have a dynamic input schema that depends on which\n",
      " |      configuration the runnable is invoked with.\n",
      " |      \n",
      " |      This method allows to get an input schema for a specific configuration.\n",
      " |      \n",
      " |      Args:\n",
      " |          config: A config to use when generating the schema.\n",
      " |      \n",
      " |      Returns:\n",
      " |          A pydantic model that can be used to validate input.\n",
      " |  \n",
      " |  get_name(self, suffix: 'Optional[str]' = None, *, name: 'Optional[str]' = None) -> 'str'\n",
      " |      Get the name of the runnable.\n",
      " |  \n",
      " |  get_output_schema(self, config: 'Optional[RunnableConfig]' = None) -> 'Type[BaseModel]'\n",
      " |      Get a pydantic model that can be used to validate output to the runnable.\n",
      " |      \n",
      " |      Runnables that leverage the configurable_fields and configurable_alternatives\n",
      " |      methods will have a dynamic output schema that depends on which\n",
      " |      configuration the runnable is invoked with.\n",
      " |      \n",
      " |      This method allows to get an output schema for a specific configuration.\n",
      " |      \n",
      " |      Args:\n",
      " |          config: A config to use when generating the schema.\n",
      " |      \n",
      " |      Returns:\n",
      " |          A pydantic model that can be used to validate output.\n",
      " |  \n",
      " |  get_prompts(self, config: 'Optional[RunnableConfig]' = None) -> 'List[BasePromptTemplate]'\n",
      " |  \n",
      " |  map(self) -> 'Runnable[List[Input], List[Output]]'\n",
      " |      Return a new Runnable that maps a list of inputs to a list of outputs,\n",
      " |      by calling invoke() with each input.\n",
      " |      \n",
      " |      Example:\n",
      " |      \n",
      " |          .. code-block:: python\n",
      " |      \n",
      " |                  from langchain_core.runnables import RunnableLambda\n",
      " |      \n",
      " |                  def _lambda(x: int) -> int:\n",
      " |                      return x + 1\n",
      " |      \n",
      " |                  runnable = RunnableLambda(_lambda)\n",
      " |                  print(runnable.map().invoke([1, 2, 3])) # [2, 3, 4]\n",
      " |  \n",
      " |  pick(self, keys: 'Union[str, List[str]]') -> 'RunnableSerializable[Any, Any]'\n",
      " |      Pick keys from the dict output of this runnable.\n",
      " |      \n",
      " |      Pick single key:\n",
      " |          .. code-block:: python\n",
      " |      \n",
      " |              import json\n",
      " |      \n",
      " |              from langchain_core.runnables import RunnableLambda, RunnableMap\n",
      " |      \n",
      " |              as_str = RunnableLambda(str)\n",
      " |              as_json = RunnableLambda(json.loads)\n",
      " |              chain = RunnableMap(str=as_str, json=as_json)\n",
      " |      \n",
      " |              chain.invoke(\"[1, 2, 3]\")\n",
      " |              # -> {\"str\": \"[1, 2, 3]\", \"json\": [1, 2, 3]}\n",
      " |      \n",
      " |              json_only_chain = chain.pick(\"json\")\n",
      " |              json_only_chain.invoke(\"[1, 2, 3]\")\n",
      " |              # -> [1, 2, 3]\n",
      " |      \n",
      " |      Pick list of keys:\n",
      " |          .. code-block:: python\n",
      " |      \n",
      " |              from typing import Any\n",
      " |      \n",
      " |              import json\n",
      " |      \n",
      " |              from langchain_core.runnables import RunnableLambda, RunnableMap\n",
      " |      \n",
      " |              as_str = RunnableLambda(str)\n",
      " |              as_json = RunnableLambda(json.loads)\n",
      " |              def as_bytes(x: Any) -> bytes:\n",
      " |                  return bytes(x, \"utf-8\")\n",
      " |      \n",
      " |              chain = RunnableMap(\n",
      " |                  str=as_str,\n",
      " |                  json=as_json,\n",
      " |                  bytes=RunnableLambda(as_bytes)\n",
      " |              )\n",
      " |      \n",
      " |              chain.invoke(\"[1, 2, 3]\")\n",
      " |              # -> {\"str\": \"[1, 2, 3]\", \"json\": [1, 2, 3], \"bytes\": b\"[1, 2, 3]\"}\n",
      " |      \n",
      " |              json_and_bytes_chain = chain.pick([\"json\", \"bytes\"])\n",
      " |              json_and_bytes_chain.invoke(\"[1, 2, 3]\")\n",
      " |              # -> {\"json\": [1, 2, 3], \"bytes\": b\"[1, 2, 3]\"}\n",
      " |  \n",
      " |  pipe(self, *others: 'Union[Runnable[Any, Other], Callable[[Any], Other]]', name: 'Optional[str]' = None) -> 'RunnableSerializable[Input, Other]'\n",
      " |      Compose this Runnable with Runnable-like objects to make a RunnableSequence.\n",
      " |      \n",
      " |      Equivalent to `RunnableSequence(self, *others)` or `self | others[0] | ...`\n",
      " |      \n",
      " |      Example:\n",
      " |          .. code-block:: python\n",
      " |      \n",
      " |              from langchain_core.runnables import RunnableLambda\n",
      " |      \n",
      " |              def add_one(x: int) -> int:\n",
      " |                  return x + 1\n",
      " |      \n",
      " |              def mul_two(x: int) -> int:\n",
      " |                  return x * 2\n",
      " |      \n",
      " |              runnable_1 = RunnableLambda(add_one)\n",
      " |              runnable_2 = RunnableLambda(mul_two)\n",
      " |              sequence = runnable_1.pipe(runnable_2)\n",
      " |              # Or equivalently:\n",
      " |              # sequence = runnable_1 | runnable_2\n",
      " |              # sequence = RunnableSequence(first=runnable_1, last=runnable_2)\n",
      " |              sequence.invoke(1)\n",
      " |              await sequence.ainvoke(1)\n",
      " |              # -> 4\n",
      " |      \n",
      " |              sequence.batch([1, 2, 3])\n",
      " |              await sequence.abatch([1, 2, 3])\n",
      " |              # -> [4, 6, 8]\n",
      " |  \n",
      " |  stream(self, input: 'Input', config: 'Optional[RunnableConfig]' = None, **kwargs: 'Optional[Any]') -> 'Iterator[Output]'\n",
      " |      Default implementation of stream, which calls invoke.\n",
      " |      Subclasses should override this method if they support streaming output.\n",
      " |  \n",
      " |  transform(self, input: 'Iterator[Input]', config: 'Optional[RunnableConfig]' = None, **kwargs: 'Optional[Any]') -> 'Iterator[Output]'\n",
      " |      Default implementation of transform, which buffers input and then calls stream.\n",
      " |      Subclasses should override this method if they can start producing output while\n",
      " |      input is still being generated.\n",
      " |  \n",
      " |  with_config(self, config: 'Optional[RunnableConfig]' = None, **kwargs: 'Any') -> 'Runnable[Input, Output]'\n",
      " |      Bind config to a Runnable, returning a new Runnable.\n",
      " |  \n",
      " |  with_fallbacks(self, fallbacks: 'Sequence[Runnable[Input, Output]]', *, exceptions_to_handle: 'Tuple[Type[BaseException], ...]' = (<class 'Exception'>,), exception_key: 'Optional[str]' = None) -> 'RunnableWithFallbacksT[Input, Output]'\n",
      " |      Add fallbacks to a runnable, returning a new Runnable.\n",
      " |      \n",
      " |      Example:\n",
      " |      \n",
      " |          .. code-block:: python\n",
      " |      \n",
      " |              from typing import Iterator\n",
      " |      \n",
      " |              from langchain_core.runnables import RunnableGenerator\n",
      " |      \n",
      " |      \n",
      " |              def _generate_immediate_error(input: Iterator) -> Iterator[str]:\n",
      " |                  raise ValueError()\n",
      " |                  yield \"\"\n",
      " |      \n",
      " |      \n",
      " |              def _generate(input: Iterator) -> Iterator[str]:\n",
      " |                  yield from \"foo bar\"\n",
      " |      \n",
      " |      \n",
      " |              runnable = RunnableGenerator(_generate_immediate_error).with_fallbacks(\n",
      " |                  [RunnableGenerator(_generate)]\n",
      " |                  )\n",
      " |              print(''.join(runnable.stream({}))) #foo bar\n",
      " |      \n",
      " |      Args:\n",
      " |          fallbacks: A sequence of runnables to try if the original runnable fails.\n",
      " |          exceptions_to_handle: A tuple of exception types to handle.\n",
      " |          exception_key: If string is specified then handled exceptions will be passed\n",
      " |              to fallbacks as part of the input under the specified key. If None,\n",
      " |              exceptions will not be passed to fallbacks. If used, the base runnable\n",
      " |              and its fallbacks must accept a dictionary as input.\n",
      " |      \n",
      " |      Returns:\n",
      " |          A new Runnable that will try the original runnable, and then each\n",
      " |          fallback in order, upon failures.\n",
      " |  \n",
      " |  with_listeners(self, *, on_start: 'Optional[Listener]' = None, on_end: 'Optional[Listener]' = None, on_error: 'Optional[Listener]' = None) -> 'Runnable[Input, Output]'\n",
      " |      Bind lifecycle listeners to a Runnable, returning a new Runnable.\n",
      " |      \n",
      " |      on_start: Called before the runnable starts running, with the Run object.\n",
      " |      on_end: Called after the runnable finishes running, with the Run object.\n",
      " |      on_error: Called if the runnable throws an error, with the Run object.\n",
      " |      \n",
      " |      The Run object contains information about the run, including its id,\n",
      " |      type, input, output, error, start_time, end_time, and any tags or metadata\n",
      " |      added to the run.\n",
      " |      \n",
      " |      Example:\n",
      " |      \n",
      " |      .. code-block:: python\n",
      " |          from langchain_core.runnables import RunnableLambda\n",
      " |          import time\n",
      " |      \n",
      " |          def test_runnable(time_to_sleep : int):\n",
      " |              time.sleep(time_to_sleep)\n",
      " |      \n",
      " |          def fn_start(run_obj : Runnable):\n",
      " |              print(\"start_time:\", run_obj.start_time)\n",
      " |      \n",
      " |          def fn_end(run_obj : Runnable):\n",
      " |              print(\"end_time:\", run_obj.end_time)\n",
      " |      \n",
      " |          RunnableLambda(test_runnable).with_listeners(\n",
      " |              on_start=fn_start,\n",
      " |              on_end=fn_end\n",
      " |              ).invoke(2)\n",
      " |  \n",
      " |  with_retry(self, *, retry_if_exception_type: 'Tuple[Type[BaseException], ...]' = (<class 'Exception'>,), wait_exponential_jitter: 'bool' = True, stop_after_attempt: 'int' = 3) -> 'Runnable[Input, Output]'\n",
      " |      Create a new Runnable that retries the original runnable on exceptions.\n",
      " |      \n",
      " |      Example:\n",
      " |      \n",
      " |      .. code-block:: python\n",
      " |          from langchain_core.runnables import RunnableLambda\n",
      " |      \n",
      " |          count = 0\n",
      " |      \n",
      " |      \n",
      " |          def _lambda(x: int) -> None:\n",
      " |              global count\n",
      " |              count = count + 1\n",
      " |              if x == 1:\n",
      " |                  raise ValueError(\"x is 1\")\n",
      " |              else:\n",
      " |                   pass\n",
      " |      \n",
      " |      \n",
      " |          runnable = RunnableLambda(_lambda)\n",
      " |          try:\n",
      " |              runnable.with_retry(\n",
      " |                  stop_after_attempt=2,\n",
      " |                  retry_if_exception_type=(ValueError,),\n",
      " |              ).invoke(1)\n",
      " |          except ValueError:\n",
      " |              pass\n",
      " |      \n",
      " |          assert (count == 2)\n",
      " |      \n",
      " |      \n",
      " |      Args:\n",
      " |          retry_if_exception_type: A tuple of exception types to retry on\n",
      " |          wait_exponential_jitter: Whether to add jitter to the wait time\n",
      " |                                   between retries\n",
      " |          stop_after_attempt: The maximum number of attempts to make before giving up\n",
      " |      \n",
      " |      Returns:\n",
      " |          A new Runnable that retries the original runnable on exceptions.\n",
      " |  \n",
      " |  with_types(self, *, input_type: 'Optional[Type[Input]]' = None, output_type: 'Optional[Type[Output]]' = None) -> 'Runnable[Input, Output]'\n",
      " |      Bind input and output types to a Runnable, returning a new Runnable.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Readonly properties inherited from langchain_core.runnables.base.Runnable:\n",
      " |  \n",
      " |  InputType\n",
      " |      The type of input this runnable accepts specified as a type annotation.\n",
      " |  \n",
      " |  OutputType\n",
      " |      The type of output this runnable produces specified as a type annotation.\n",
      " |  \n",
      " |  config_specs\n",
      " |      List configurable fields for this runnable.\n",
      " |  \n",
      " |  input_schema\n",
      " |      The type of input this runnable accepts specified as a pydantic model.\n",
      " |  \n",
      " |  output_schema\n",
      " |      The type of output this runnable produces specified as a pydantic model.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes inherited from langchain_core.runnables.base.Runnable:\n",
      " |  \n",
      " |  name = None\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Class methods inherited from typing.Generic:\n",
      " |  \n",
      " |  __class_getitem__(params) from pydantic.v1.main.ModelMetaclass\n",
      " |      Parameterizes a generic class.\n",
      " |      \n",
      " |      At least, parameterizing a generic class is the *main* thing this method\n",
      " |      does. For example, for some generic class `Foo`, this is called when we\n",
      " |      do `Foo[int]` - there, with `cls=Foo` and `params=int`.\n",
      " |      \n",
      " |      However, note that this method is also called when defining generic\n",
      " |      classes in the first place with `class Foo(Generic[T]): ...`.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(retriever)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "8c7c4aa2-26d1-4498-83fd-766e635ae153",
   "metadata": {},
   "outputs": [],
   "source": [
    "docs = retriever.invoke('kronecker product')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "cd97957f-ebde-4c76-8362-a3f61ed705e6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(docs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58b922b8-c7cb-4b3f-b192-71264c737b22",
   "metadata": {},
   "source": [
    "### Building ONLINE DB TUTORIAL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "80332f8b-55a7-4125-9fc6-67b409d7c4d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import bs4\n",
    "from langchain import hub\n",
    "from langchain_chroma import Chroma\n",
    "from langchain_community.document_loaders import WebBaseLoader\n",
    "from langchain_core.output_parsers import StrOutputParser\n",
    "from langchain_core.runnables import RunnablePassthrough\n",
    "from langchain_text_splitters import RecursiveCharacterTextSplitter\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "9006b9c7-9a69-491e-8607-c5d5d7989d6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load, chunk and index the contents of the blog.\n",
    "loader = WebBaseLoader(\n",
    "    web_paths=(\"https://lilianweng.github.io/posts/2023-06-23-agent/\",),\n",
    "    bs_kwargs=dict(\n",
    "        parse_only=bs4.SoupStrainer(\n",
    "            class_=(\"post-content\", \"post-title\", \"post-header\")\n",
    "        )\n",
    "    ),\n",
    ")\n",
    "docs = loader.load()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "6747d35f-f892-4fff-a8b2-dedc1b6c8fe1",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'OpenAIEmbeddings' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[21], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m text_splitter \u001b[38;5;241m=\u001b[39m RecursiveCharacterTextSplitter(chunk_size\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1000\u001b[39m, chunk_overlap\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m200\u001b[39m)\n\u001b[1;32m      2\u001b[0m splits \u001b[38;5;241m=\u001b[39m text_splitter\u001b[38;5;241m.\u001b[39msplit_documents(docs)\n\u001b[0;32m----> 3\u001b[0m vectorstore \u001b[38;5;241m=\u001b[39m Chroma\u001b[38;5;241m.\u001b[39mfrom_documents(documents\u001b[38;5;241m=\u001b[39msplits, embedding\u001b[38;5;241m=\u001b[39m\u001b[43mOpenAIEmbeddings\u001b[49m())\n\u001b[1;32m      5\u001b[0m \u001b[38;5;66;03m# Retrieve and generate using the relevant snippets of the blog.\u001b[39;00m\n\u001b[1;32m      6\u001b[0m retriever \u001b[38;5;241m=\u001b[39m vectorstore\u001b[38;5;241m.\u001b[39mas_retriever()\n",
      "\u001b[0;31mNameError\u001b[0m: name 'OpenAIEmbeddings' is not defined"
     ]
    }
   ],
   "source": [
    "text_splitter = RecursiveCharacterTextSplitter(chunk_size=1000, chunk_overlap=200)\n",
    "splits = text_splitter.split_documents(docs)\n",
    "vectorstore = Chroma.from_documents(documents=splits, embedding=OpenAIEmbeddings())\n",
    "\n",
    "# Retrieve and generate using the relevant snippets of the blog.\n",
    "retriever = vectorstore.as_retriever()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "5a009d59-81de-4a59-8584-ac0c9c659338",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[Document(page_content='\\n\\n      LLM Powered Autonomous Agents\\n    \\nDate: June 23, 2023  |  Estimated Reading Time: 31 min  |  Author: Lilian Weng\\n\\n\\nBuilding agents with LLM (large language model) as its core controller is a cool concept. Several proof-of-concepts demos, such as AutoGPT, GPT-Engineer and BabyAGI, serve as inspiring examples. The potentiality of LLM extends beyond generating well-written copies, stories, essays and programs; it can be framed as a powerful general problem solver.\\nAgent System Overview#\\nIn a LLM-powered autonomous agent system, LLM functions as the agent’s brain, complemented by several key components:\\n\\nPlanning\\n\\nSubgoal and decomposition: The agent breaks down large tasks into smaller, manageable subgoals, enabling efficient handling of complex tasks.\\nReflection and refinement: The agent can do self-criticism and self-reflection over past actions, learn from mistakes and refine them for future steps, thereby improving the quality of final results.\\n\\n\\nMemory\\n\\nShort-term memory: I would consider all the in-context learning (See Prompt Engineering) as utilizing short-term memory of the model to learn.\\nLong-term memory: This provides the agent with the capability to retain and recall (infinite) information over extended periods, often by leveraging an external vector store and fast retrieval.\\n\\n\\nTool use\\n\\nThe agent learns to call external APIs for extra information that is missing from the model weights (often hard to change after pre-training), including current information, code execution capability, access to proprietary information sources and more.\\n\\n\\n\\n\\nFig. 1. Overview of a LLM-powered autonomous agent system.\\nComponent One: Planning#\\nA complicated task usually involves many steps. An agent needs to know what they are and plan ahead.\\nTask Decomposition#\\nChain of thought (CoT; Wei et al. 2022) has become a standard prompting technique for enhancing model performance on complex tasks. The model is instructed to “think step by step” to utilize more test-time computation to decompose hard tasks into smaller and simpler steps. CoT transforms big tasks into multiple manageable tasks and shed lights into an interpretation of the model’s thinking process.\\nTree of Thoughts (Yao et al. 2023) extends CoT by exploring multiple reasoning possibilities at each step. It first decomposes the problem into multiple thought steps and generates multiple thoughts per step, creating a tree structure. The search process can be BFS (breadth-first search) or DFS (depth-first search) with each state evaluated by a classifier (via a prompt) or majority vote.\\nTask decomposition can be done (1) by LLM with simple prompting like \"Steps for XYZ.\\\\n1.\", \"What are the subgoals for achieving XYZ?\", (2) by using task-specific instructions; e.g. \"Write a story outline.\" for writing a novel, or (3) with human inputs.\\nAnother quite distinct approach, LLM+P (Liu et al. 2023), involves relying on an external classical planner to do long-horizon planning. This approach utilizes the Planning Domain Definition Language (PDDL) as an intermediate interface to describe the planning problem. In this process, LLM (1) translates the problem into “Problem PDDL”, then (2) requests a classical planner to generate a PDDL plan based on an existing “Domain PDDL”, and finally (3) translates the PDDL plan back into natural language. Essentially, the planning step is outsourced to an external tool, assuming the availability of domain-specific PDDL and a suitable planner which is common in certain robotic setups but not in many other domains.\\nSelf-Reflection#\\nSelf-reflection is a vital aspect that allows autonomous agents to improve iteratively by refining past action decisions and correcting previous mistakes. It plays a crucial role in real-world tasks where trial and error are inevitable.\\nReAct (Yao et al. 2023) integrates reasoning and acting within LLM by extending the action space to be a combination of task-specific discrete actions and the language space. The former enables LLM to interact with the environment (e.g. use Wikipedia search API), while the latter prompting LLM to generate reasoning traces in natural language.\\nThe ReAct prompt template incorporates explicit steps for LLM to think, roughly formatted as:\\nThought: ...\\nAction: ...\\nObservation: ...\\n... (Repeated many times)\\n\\nFig. 2.  Examples of reasoning trajectories for knowledge-intensive tasks (e.g. HotpotQA, FEVER) and decision-making tasks (e.g. AlfWorld Env, WebShop). (Image source: Yao et al. 2023).\\nIn both experiments on knowledge-intensive tasks and decision-making tasks, ReAct works better than the Act-only baseline where Thought: … step is removed.\\nReflexion (Shinn & Labash 2023) is a framework to equips agents with dynamic memory and self-reflection capabilities to improve reasoning skills. Reflexion has a standard RL setup, in which the reward model provides a simple binary reward and the action space follows the setup in ReAct where the task-specific action space is augmented with language to enable complex reasoning steps. After each action $a_t$, the agent computes a heuristic $h_t$ and optionally may decide to reset the environment to start a new trial depending on the self-reflection results.\\n\\nFig. 3. Illustration of the Reflexion framework. (Image source: Shinn & Labash, 2023)\\nThe heuristic function determines when the trajectory is inefficient or contains hallucination and should be stopped. Inefficient planning refers to trajectories that take too long without success. Hallucination is defined as encountering a sequence of consecutive identical actions that lead to the same observation in the environment.\\nSelf-reflection is created by showing two-shot examples to LLM and each example is a pair of (failed trajectory, ideal reflection for guiding future changes in the plan). Then reflections are added into the agent’s working memory, up to three, to be used as context for querying LLM.\\n\\nFig. 4. Experiments on AlfWorld Env and HotpotQA. Hallucination is a more common failure than inefficient planning in AlfWorld. (Image source: Shinn & Labash, 2023)\\nChain of Hindsight (CoH; Liu et al. 2023) encourages the model to improve on its own outputs by explicitly presenting it with a sequence of past outputs, each annotated with feedback. Human feedback data is a collection of $D_h = \\\\{(x, y_i , r_i , z_i)\\\\}_{i=1}^n$, where $x$ is the prompt, each $y_i$ is a model completion, $r_i$ is the human rating of $y_i$, and $z_i$ is the corresponding human-provided hindsight feedback. Assume the feedback tuples are ranked by reward, $r_n \\\\geq r_{n-1} \\\\geq \\\\dots \\\\geq r_1$ The process is supervised fine-tuning where the data is a sequence in the form of $\\\\tau_h = (x, z_i, y_i, z_j, y_j, \\\\dots, z_n, y_n)$, where $\\\\leq i \\\\leq j \\\\leq n$. The model is finetuned to only predict $y_n$ where conditioned on the sequence prefix, such that the model can self-reflect to produce better output based on the feedback sequence. The model can optionally receive multiple rounds of instructions with human annotators at test time.\\nTo avoid overfitting, CoH adds a regularization term to maximize the log-likelihood of the pre-training dataset. To avoid shortcutting and copying (because there are many common words in feedback sequences), they randomly mask 0% - 5% of past tokens during training.\\nThe training dataset in their experiments is a combination of WebGPT comparisons, summarization from human feedback and human preference dataset.\\n\\nFig. 5. After fine-tuning with CoH, the model can follow instructions to produce outputs with incremental improvement in a sequence. (Image source: Liu et al. 2023)\\nThe idea of CoH is to present a history of sequentially improved outputs  in context and train the model to take on the trend to produce better outputs. Algorithm Distillation (AD; Laskin et al. 2023) applies the same idea to cross-episode trajectories in reinforcement learning tasks, where an algorithm is encapsulated in a long history-conditioned policy. Considering that an agent interacts with the environment many times and in each episode the agent gets a little better, AD concatenates this learning history and feeds that into the model. Hence we should expect the next predicted action to lead to better performance than previous trials. The goal is to learn the process of RL instead of training a task-specific policy itself.\\n\\nFig. 6. Illustration of how Algorithm Distillation (AD) works. (Image source: Laskin et al. 2023).\\nThe paper hypothesizes that any algorithm that generates a set of learning histories can be distilled into a neural network by performing behavioral cloning over actions. The history data is generated by a set of source policies, each trained for a specific task. At the training stage, during each RL run, a random task is sampled and a subsequence of multi-episode history is used for training, such that the learned policy is task-agnostic.\\nIn reality, the model has limited context window length, so episodes should be short enough to construct multi-episode history. Multi-episodic contexts of 2-4 episodes are necessary to learn a near-optimal in-context RL algorithm. The emergence of in-context RL requires long enough context.\\nIn comparison with three baselines, including ED (expert distillation, behavior cloning with expert trajectories instead of learning history), source policy (used for generating trajectories for distillation by UCB), RL^2 (Duan et al. 2017; used as upper bound since it needs online RL), AD demonstrates in-context RL with performance getting close to RL^2 despite only using offline RL and learns much faster than other baselines. When conditioned on partial training history of the source policy, AD also improves much faster than ED baseline.\\n\\nFig. 7. Comparison of AD, ED, source policy and RL^2 on environments that require memory and exploration. Only binary reward is assigned. The source policies are trained with A3C for \"dark\" environments and DQN for watermaze.(Image source: Laskin et al. 2023)\\nComponent Two: Memory#\\n(Big thank you to ChatGPT for helping me draft this section. I’ve learned a lot about the human brain and data structure for fast MIPS in my conversations with ChatGPT.)\\nTypes of Memory#\\nMemory can be defined as the processes used to acquire, store, retain, and later retrieve information. There are several types of memory in human brains.\\n\\n\\nSensory Memory: This is the earliest stage of memory, providing the ability to retain impressions of sensory information (visual, auditory, etc) after the original stimuli have ended. Sensory memory typically only lasts for up to a few seconds. Subcategories include iconic memory (visual), echoic memory (auditory), and haptic memory (touch).\\n\\n\\nShort-Term Memory (STM) or Working Memory: It stores information that we are currently aware of and needed to carry out complex cognitive tasks such as learning and reasoning. Short-term memory is believed to have the capacity of about 7 items (Miller 1956) and lasts for 20-30 seconds.\\n\\n\\nLong-Term Memory (LTM): Long-term memory can store information for a remarkably long time, ranging from a few days to decades, with an essentially unlimited storage capacity. There are two subtypes of LTM:\\n\\nExplicit / declarative memory: This is memory of facts and events, and refers to those memories that can be consciously recalled, including episodic memory (events and experiences) and semantic memory (facts and concepts).\\nImplicit / procedural memory: This type of memory is unconscious and involves skills and routines that are performed automatically, like riding a bike or typing on a keyboard.\\n\\n\\n\\n\\nFig. 8. Categorization of human memory.\\nWe can roughly consider the following mappings:\\n\\nSensory memory as learning embedding representations for raw inputs, including text, image or other modalities;\\nShort-term memory as in-context learning. It is short and finite, as it is restricted by the finite context window length of Transformer.\\nLong-term memory as the external vector store that the agent can attend to at query time, accessible via fast retrieval.\\n\\nMaximum Inner Product Search (MIPS)#\\nThe external memory can alleviate the restriction of finite attention span.  A standard practice is to save the embedding representation of information into a vector store database that can support fast maximum inner-product search (MIPS). To optimize the retrieval speed, the common choice is the approximate nearest neighbors (ANN)\\u200b algorithm to return approximately top k nearest neighbors to trade off a little accuracy lost for a huge speedup.\\nA couple common choices of ANN algorithms for fast MIPS:\\n\\nLSH (Locality-Sensitive Hashing): It introduces a hashing function such that similar input items are mapped to the same buckets with high probability, where the number of buckets is much smaller than the number of inputs.\\nANNOY (Approximate Nearest Neighbors Oh Yeah): The core data structure are random projection trees, a set of binary trees where each non-leaf node represents a hyperplane splitting the input space into half and each leaf stores one data point. Trees are built independently and at random, so to some extent, it mimics a hashing function. ANNOY search happens in all the trees to iteratively search through the half that is closest to the query and then aggregates the results. The idea is quite related to KD tree but a lot more scalable.\\nHNSW (Hierarchical Navigable Small World): It is inspired by the idea of small world networks where most nodes can be reached by any other nodes within a small number of steps; e.g. “six degrees of separation” feature of social networks. HNSW builds hierarchical layers of these small-world graphs, where the bottom layers contain the actual data points. The layers in the middle create shortcuts to speed up search. When performing a search, HNSW starts from a random node in the top layer and navigates towards the target. When it can’t get any closer, it moves down to the next layer, until it reaches the bottom layer. Each move in the upper layers can potentially cover a large distance in the data space, and each move in the lower layers refines the search quality.\\nFAISS (Facebook AI Similarity Search): It operates on the assumption that in high dimensional space, distances between nodes follow a Gaussian distribution and thus there should exist clustering of data points. FAISS applies vector quantization by partitioning the vector space into clusters and then refining the quantization within clusters. Search first looks for cluster candidates with coarse quantization and then further looks into each cluster with finer quantization.\\nScaNN (Scalable Nearest Neighbors): The main innovation in ScaNN is anisotropic vector quantization. It quantizes a data point $x_i$ to $\\\\tilde{x}_i$ such that the inner product $\\\\langle q, x_i \\\\rangle$ is as similar to the original distance of $\\\\angle q, \\\\tilde{x}_i$ as possible, instead of picking the closet quantization centroid points.\\n\\n\\nFig. 9. Comparison of MIPS algorithms, measured in recall@10. (Image source: Google Blog, 2020)\\nCheck more MIPS algorithms and performance comparison in ann-benchmarks.com.\\nComponent Three: Tool Use#\\nTool use is a remarkable and distinguishing characteristic of human beings. We create, modify and utilize external objects to do things that go beyond our physical and cognitive limits. Equipping LLMs with external tools can significantly extend the model capabilities.\\n\\nFig. 10. A picture of a sea otter using rock to crack open a seashell, while floating in the water. While some other animals can use tools, the complexity is not comparable with humans. (Image source: Animals using tools)\\nMRKL (Karpas et al. 2022), short for “Modular Reasoning, Knowledge and Language”, is a neuro-symbolic architecture for autonomous agents. A MRKL system is proposed to contain a collection of “expert” modules and the general-purpose LLM works as a router to route inquiries to the best suitable expert module. These modules can be neural (e.g. deep learning models) or symbolic (e.g. math calculator, currency converter, weather API).\\nThey did an experiment on fine-tuning LLM to call a calculator, using arithmetic as a test case. Their experiments showed that it was harder to solve verbal math problems than explicitly stated math problems because LLMs (7B Jurassic1-large model) failed to extract the right arguments for the basic arithmetic reliably. The results highlight when the external symbolic tools can work reliably, knowing when to and how to use the tools are crucial, determined by the LLM capability.\\nBoth TALM (Tool Augmented Language Models; Parisi et al. 2022) and Toolformer (Schick et al. 2023) fine-tune a LM to learn to use external tool APIs. The dataset is expanded based on whether a newly added API call annotation can improve the quality of model outputs. See more details in the “External APIs” section of Prompt Engineering.\\nChatGPT Plugins and OpenAI API  function calling are good examples of LLMs augmented with tool use capability working in practice. The collection of tool APIs can be provided by other developers (as in Plugins) or self-defined (as in function calls).\\nHuggingGPT (Shen et al. 2023) is a framework to use ChatGPT as the task planner to select models available in HuggingFace platform according to the model descriptions and summarize the response based on the execution results.\\n\\nFig. 11. Illustration of how HuggingGPT works. (Image source: Shen et al. 2023)\\nThe system comprises of 4 stages:\\n(1) Task planning: LLM works as the brain and parses the user requests into multiple tasks. There are four attributes associated with each task: task type, ID, dependencies, and arguments. They use few-shot examples to guide LLM to do task parsing and planning.\\nInstruction:\\n\\nThe AI assistant can parse user input to several tasks: [{\"task\": task, \"id\", task_id, \"dep\": dependency_task_ids, \"args\": {\"text\": text, \"image\": URL, \"audio\": URL, \"video\": URL}}]. The \"dep\" field denotes the id of the previous task which generates a new resource that the current task relies on. A special tag \"-task_id\" refers to the generated text image, audio and video in the dependency task with id as task_id. The task MUST be selected from the following options: {{ Available Task List }}. There is a logical relationship between tasks, please note their order. If the user input can\\'t be parsed, you need to reply empty JSON. Here are several cases for your reference: {{ Demonstrations }}. The chat history is recorded as {{ Chat History }}. From this chat history, you can find the path of the user-mentioned resources for your task planning.\\n\\n(2) Model selection: LLM distributes the tasks to expert models, where the request is framed as a multiple-choice question. LLM is presented with a list of models to choose from. Due to the limited context length, task type based filtration is needed.\\nInstruction:\\n\\nGiven the user request and the call command, the AI assistant helps the user to select a suitable model from a list of models to process the user request. The AI assistant merely outputs the model id of the most appropriate model. The output must be in a strict JSON format: \"id\": \"id\", \"reason\": \"your detail reason for the choice\". We have a list of models for you to choose from {{ Candidate Models }}. Please select one model from the list.\\n\\n(3) Task execution: Expert models execute on the specific tasks and log results.\\nInstruction:\\n\\nWith the input and the inference results, the AI assistant needs to describe the process and results. The previous stages can be formed as - User Input: {{ User Input }}, Task Planning: {{ Tasks }}, Model Selection: {{ Model Assignment }}, Task Execution: {{ Predictions }}. You must first answer the user\\'s request in a straightforward manner. Then describe the task process and show your analysis and model inference results to the user in the first person. If inference results contain a file path, must tell the user the complete file path.\\n\\n(4) Response generation: LLM receives the execution results and provides summarized results to users.\\nTo put HuggingGPT into real world usage, a couple challenges need to solve: (1) Efficiency improvement is needed as both LLM inference rounds and interactions with other models slow down the process; (2) It relies on a long context window to communicate over complicated task content; (3) Stability improvement of LLM outputs and external model services.\\nAPI-Bank (Li et al. 2023) is a benchmark for evaluating the performance of tool-augmented LLMs. It contains 53 commonly used API tools, a complete tool-augmented LLM workflow, and 264 annotated dialogues that involve 568 API calls. The selection of APIs is quite diverse, including search engines, calculator, calendar queries, smart home control, schedule management, health data management, account authentication workflow and more. Because there are a large number of APIs, LLM first has access to API search engine to find the right API to call and then uses the corresponding documentation to make a call.\\n\\nFig. 12. Pseudo code of how LLM makes an API call in API-Bank. (Image source: Li et al. 2023)\\nIn the API-Bank workflow, LLMs need to make a couple of decisions and at each step we can evaluate how accurate that decision is. Decisions include:\\n\\nWhether an API call is needed.\\nIdentify the right API to call: if not good enough, LLMs need to iteratively modify the API inputs (e.g. deciding search keywords for Search Engine API).\\nResponse based on the API results: the model can choose to refine and call again if results are not satisfied.\\n\\nThis benchmark evaluates the agent’s tool use capabilities at three levels:\\n\\nLevel-1 evaluates the ability to call the API. Given an API’s description, the model needs to determine whether to call a given API, call it correctly, and respond properly to API returns.\\nLevel-2 examines the ability to retrieve the API. The model needs to search for possible APIs that may solve the user’s requirement and learn how to use them by reading documentation.\\nLevel-3 assesses the ability to plan API beyond retrieve and call. Given unclear user requests (e.g. schedule group meetings, book flight/hotel/restaurant for a trip), the model may have to conduct multiple API calls to solve it.\\n\\nCase Studies#\\nScientific Discovery Agent#\\nChemCrow (Bran et al. 2023) is a domain-specific example in which LLM is augmented with 13 expert-designed tools to accomplish tasks across organic synthesis, drug discovery, and materials design. The workflow, implemented in LangChain, reflects what was previously described in the ReAct and MRKLs and combines CoT reasoning with tools relevant to the tasks:\\n\\nThe LLM is provided with a list of tool names, descriptions of their utility, and details about the expected input/output.\\nIt is then instructed to answer a user-given prompt using the tools provided when necessary. The instruction suggests the model to follow the ReAct format - Thought, Action, Action Input, Observation.\\n\\nOne interesting observation is that while the LLM-based evaluation concluded that GPT-4 and ChemCrow perform nearly equivalently, human evaluations with experts oriented towards the completion and chemical correctness of the solutions showed that ChemCrow outperforms GPT-4 by a large margin. This indicates a potential problem with using LLM to evaluate its own performance on domains that requires deep expertise. The lack of expertise may cause LLMs not knowing its flaws and thus cannot well judge the correctness of task results.\\nBoiko et al. (2023) also looked into LLM-empowered agents for scientific discovery, to handle autonomous design, planning, and performance of complex scientific experiments. This agent can use tools to browse the Internet, read documentation, execute code, call robotics experimentation APIs and leverage other LLMs.\\nFor example, when requested to \"develop a novel anticancer drug\", the model came up with the following reasoning steps:\\n\\ninquired about current trends in anticancer drug discovery;\\nselected a target;\\nrequested a scaffold targeting these compounds;\\nOnce the compound was identified, the model attempted its synthesis.\\n\\nThey also discussed the risks, especially with illicit drugs and bioweapons. They developed a test set containing a list of known chemical weapon agents and asked the agent to synthesize them. 4 out of 11 requests (36%) were accepted to obtain a synthesis solution and the agent attempted to consult documentation to execute the procedure. 7 out of 11 were rejected and among these 7 rejected cases, 5 happened after a Web search while 2 were rejected based on prompt only.\\nGenerative Agents Simulation#\\nGenerative Agents (Park, et al. 2023) is super fun experiment where 25 virtual characters, each controlled by a LLM-powered agent, are living and interacting in a sandbox environment, inspired by The Sims. Generative agents create believable simulacra of human behavior for interactive applications.\\nThe design of generative agents combines LLM with memory, planning and reflection mechanisms to enable agents to behave conditioned on past experience, as well as to interact with other agents.\\n\\nMemory stream: is a long-term memory module (external database) that records a comprehensive list of agents’ experience in natural language.\\n\\nEach element is an observation, an event directly provided by the agent.\\n- Inter-agent communication can trigger new natural language statements.\\n\\n\\nRetrieval model: surfaces the context to inform the agent’s behavior, according to relevance, recency and importance.\\n\\nRecency: recent events have higher scores\\nImportance: distinguish mundane from core memories. Ask LM directly.\\nRelevance: based on how related it is to the current situation / query.\\n\\n\\nReflection mechanism: synthesizes memories into higher level inferences over time and guides the agent’s future behavior. They are higher-level summaries of past events (<- note that this is a bit different from self-reflection above)\\n\\nPrompt LM with 100 most recent observations and to generate 3 most salient high-level questions given a set of observations/statements. Then ask LM to answer those questions.\\n\\n\\nPlanning & Reacting: translate the reflections and the environment information into actions\\n\\nPlanning is essentially in order to optimize believability at the moment vs in time.\\nPrompt template: {Intro of an agent X}. Here is X\\'s plan today in broad strokes: 1)\\nRelationships between agents and observations of one agent by another are all taken into consideration for planning and reacting.\\nEnvironment information is present in a tree structure.\\n\\n\\n\\n\\nFig. 13. The generative agent architecture. (Image source: Park et al. 2023)\\nThis fun simulation results in emergent social behavior, such as information diffusion, relationship memory (e.g. two agents continuing the conversation topic) and coordination of social events (e.g. host a party and invite many others).\\nProof-of-Concept Examples#\\nAutoGPT has drawn a lot of attention into the possibility of setting up autonomous agents with LLM as the main controller. It has quite a lot of reliability issues given the natural language interface, but nevertheless a cool proof-of-concept demo. A lot of code in AutoGPT is about format parsing.\\nHere is the system message used by AutoGPT, where {{...}} are user inputs:\\nYou are {{ai-name}}, {{user-provided AI bot description}}.\\nYour decisions must always be made independently without seeking user assistance. Play to your strengths as an LLM and pursue simple strategies with no legal complications.\\n\\nGOALS:\\n\\n1. {{user-provided goal 1}}\\n2. {{user-provided goal 2}}\\n3. ...\\n4. ...\\n5. ...\\n\\nConstraints:\\n1. ~4000 word limit for short term memory. Your short term memory is short, so immediately save important information to files.\\n2. If you are unsure how you previously did something or want to recall past events, thinking about similar events will help you remember.\\n3. No user assistance\\n4. Exclusively use the commands listed in double quotes e.g. \"command name\"\\n5. Use subprocesses for commands that will not terminate within a few minutes\\n\\nCommands:\\n1. Google Search: \"google\", args: \"input\": \"<search>\"\\n2. Browse Website: \"browse_website\", args: \"url\": \"<url>\", \"question\": \"<what_you_want_to_find_on_website>\"\\n3. Start GPT Agent: \"start_agent\", args: \"name\": \"<name>\", \"task\": \"<short_task_desc>\", \"prompt\": \"<prompt>\"\\n4. Message GPT Agent: \"message_agent\", args: \"key\": \"<key>\", \"message\": \"<message>\"\\n5. List GPT Agents: \"list_agents\", args:\\n6. Delete GPT Agent: \"delete_agent\", args: \"key\": \"<key>\"\\n7. Clone Repository: \"clone_repository\", args: \"repository_url\": \"<url>\", \"clone_path\": \"<directory>\"\\n8. Write to file: \"write_to_file\", args: \"file\": \"<file>\", \"text\": \"<text>\"\\n9. Read file: \"read_file\", args: \"file\": \"<file>\"\\n10. Append to file: \"append_to_file\", args: \"file\": \"<file>\", \"text\": \"<text>\"\\n11. Delete file: \"delete_file\", args: \"file\": \"<file>\"\\n12. Search Files: \"search_files\", args: \"directory\": \"<directory>\"\\n13. Analyze Code: \"analyze_code\", args: \"code\": \"<full_code_string>\"\\n14. Get Improved Code: \"improve_code\", args: \"suggestions\": \"<list_of_suggestions>\", \"code\": \"<full_code_string>\"\\n15. Write Tests: \"write_tests\", args: \"code\": \"<full_code_string>\", \"focus\": \"<list_of_focus_areas>\"\\n16. Execute Python File: \"execute_python_file\", args: \"file\": \"<file>\"\\n17. Generate Image: \"generate_image\", args: \"prompt\": \"<prompt>\"\\n18. Send Tweet: \"send_tweet\", args: \"text\": \"<text>\"\\n19. Do Nothing: \"do_nothing\", args:\\n20. Task Complete (Shutdown): \"task_complete\", args: \"reason\": \"<reason>\"\\n\\nResources:\\n1. Internet access for searches and information gathering.\\n2. Long Term memory management.\\n3. GPT-3.5 powered Agents for delegation of simple tasks.\\n4. File output.\\n\\nPerformance Evaluation:\\n1. Continuously review and analyze your actions to ensure you are performing to the best of your abilities.\\n2. Constructively self-criticize your big-picture behavior constantly.\\n3. Reflect on past decisions and strategies to refine your approach.\\n4. Every command has a cost, so be smart and efficient. Aim to complete tasks in the least number of steps.\\n\\nYou should only respond in JSON format as described below\\nResponse Format:\\n{\\n    \"thoughts\": {\\n        \"text\": \"thought\",\\n        \"reasoning\": \"reasoning\",\\n        \"plan\": \"- short bulleted\\\\n- list that conveys\\\\n- long-term plan\",\\n        \"criticism\": \"constructive self-criticism\",\\n        \"speak\": \"thoughts summary to say to user\"\\n    },\\n    \"command\": {\\n        \"name\": \"command name\",\\n        \"args\": {\\n            \"arg name\": \"value\"\\n        }\\n    }\\n}\\nEnsure the response can be parsed by Python json.loads\\nGPT-Engineer is another project to create a whole repository of code given a task specified in natural language. The GPT-Engineer is instructed to think over a list of smaller components to build and ask for user input to clarify questions as needed.\\nHere are a sample conversation for task clarification sent to OpenAI ChatCompletion endpoint used by GPT-Engineer. The user inputs are wrapped in {{user input text}}.\\n[\\n  {\\n    \"role\": \"system\",\\n    \"content\": \"You will read instructions and not carry them out, only seek to clarify them.\\\\nSpecifically you will first summarise a list of super short bullets of areas that need clarification.\\\\nThen you will pick one clarifying question, and wait for an answer from the user.\\\\n\"\\n  },\\n  {\\n    \"role\": \"user\",\\n    \"content\": \"We are writing {{a Super Mario game in python. MVC components split in separate files. Keyboard control.}}\\\\n\"\\n  },\\n  {\\n    \"role\": \"assistant\",\\n    \"content\": \"Summary of areas that need clarification:\\\\n1. Specifics of the Super Mario game (e.g. level design, characters, gameplay mechanics)\\\\n2. Details about the MVC components (e.g. which components are in each file)\\\\n3. Keyboard control implementation (e.g. which keys to use, how to handle input)\\\\n\\\\nClarifying question:\\\\nCan you provide more details about the Super Mario game, such as level design, characters, and gameplay mechanics?\"\\n  },\\n  {\\n    \"role\": \"user\",\\n    \"content\": \"{{There are 10 levels in total. The main character is a plumber named Mario, who can walk and jump. It is a classical platform game just like Super Mario. The main character moves from left to right, trying to get to the destination, where there are many obstacles and attacks from enemies in the process.}}\\\\n\\\\nIs anything else unclear? If yes, only answer in the form:\\\\n{remaining unclear areas} remaining questions.\\\\n{Next question}\\\\nIf everything is sufficiently clear, only answer \\\\\"Nothing more to clarify.\\\\\".\"\\n  },\\n  {\\n    \"role\": \"assistant\",\\n    \"content\": \"Remaining unclear areas: 2 remaining questions.\\\\nCan you provide more information about how the MVC components are split into separate files?\"\\n  },\\n  {\\n    \"role\": \"user\",\\n    \"content\": \"{{Make your own assumptions and state them explicitly before starting}}\"\\n  }\\n]\\nThen after these clarification, the agent moved into the code writing mode with a different system message.\\nSystem message:\\n\\nYou will get instructions for code to write.\\nYou will write a very long answer. Make sure that every detail of the architecture is, in the end, implemented as code.\\nMake sure that every detail of the architecture is, in the end, implemented as code.\\nThink step by step and reason yourself to the right decisions to make sure we get it right.\\nYou will first lay out the names of the core classes, functions, methods that will be necessary, as well as a quick comment on their purpose.\\nThen you will output the content of each file including ALL code.\\nEach file must strictly follow a markdown code block format, where the following tokens must be replaced such that\\nFILENAME is the lowercase file name including the file extension,\\nLANG is the markup code block language for the code’s language, and CODE is the code:\\nFILENAME\\nCODE\\nYou will start with the “entrypoint” file, then go to the ones that are imported by that file, and so on.\\nPlease note that the code should be fully functional. No placeholders.\\nFollow a language and framework appropriate best practice file naming convention.\\nMake sure that files contain all imports, types etc. Make sure that code in different files are compatible with each other.\\nEnsure to implement all code, if you are unsure, write a plausible implementation.\\nInclude module dependency or package manager dependency definition file.\\nBefore you finish, double check that all parts of the architecture is present in the files.\\nUseful to know:\\nYou almost always put different classes in different files.\\nFor Python, you always create an appropriate requirements.txt file.\\nFor NodeJS, you always create an appropriate package.json file.\\nYou always add a comment briefly describing the purpose of the function definition.\\nYou try to add comments explaining very complex bits of logic.\\nYou always follow the best practices for the requested languages in terms of describing the code written as a defined\\npackage/project.\\nPython toolbelt preferences:\\n\\npytest\\ndataclasses\\n\\n\\nConversatin samples:\\n[\\n  {\\n    \"role\": \"system\",\\n    \"content\": \"You will get instructions for code to write.\\\\nYou will write a very long answer. Make sure that every detail of the architecture is, in the end, implemented as code.\\\\nMake sure that every detail of the architecture is, in the end, implemented as code.\\\\n\\\\nThink step by step and reason yourself to the right decisions to make sure we get it right.\\\\nYou will first lay out the names of the core classes, functions, methods that will be necessary, as well as a quick comment on their purpose.\\\\n\\\\nThen you will output the content of each file including ALL code.\\\\nEach file must strictly follow a markdown code block format, where the following tokens must be replaced such that\\\\nFILENAME is the lowercase file name including the file extension,\\\\nLANG is the markup code block language for the code\\'s language, and CODE is the code:\\\\n\\\\nFILENAME\\\\n```LANG\\\\nCODE\\\\n```\\\\n\\\\nYou will start with the \\\\\"entrypoint\\\\\" file, then go to the ones that are imported by that file, and so on.\\\\nPlease note that the code should be fully functional. No placeholders.\\\\n\\\\nFollow a language and framework appropriate best practice file naming convention.\\\\nMake sure that files contain all imports, types etc. Make sure that code in different files are compatible with each other.\\\\nEnsure to implement all code, if you are unsure, write a plausible implementation.\\\\nInclude module dependency or package manager dependency definition file.\\\\nBefore you finish, double check that all parts of the architecture is present in the files.\\\\n\\\\nUseful to know:\\\\nYou almost always put different classes in different files.\\\\nFor Python, you always create an appropriate requirements.txt file.\\\\nFor NodeJS, you always create an appropriate package.json file.\\\\nYou always add a comment briefly describing the purpose of the function definition.\\\\nYou try to add comments explaining very complex bits of logic.\\\\nYou always follow the best practices for the requested languages in terms of describing the code written as a defined\\\\npackage/project.\\\\n\\\\n\\\\nPython toolbelt preferences:\\\\n- pytest\\\\n- dataclasses\\\\n\"\\n  },\\n #  … same conversation as earlier, ended with \"Make your own assumptions and state them explicitly before starting\".\\n  {\\n    \"role\": \"assistant\",\\n    \"content\": \"Assumptions:\\\\n1. Model: The model will contain the game\\'s data, such as level information, character states, and enemy positions.\\\\n2. View: The view will handle the game\\'s visuals, including rendering the game objects, backgrounds, and updating the display.\\\\n3. Controller: The controller will manage user input, such as keyboard controls, and update the model accordingly.\\\\n\\\\nWith these assumptions in mind, is there any specific aspect of the keyboard control implementation you would like me to clarify?\"\\n  },\\n  {\\n    \"role\": \"user\",\\n    \"content\": \"Please now remember the steps:\\\\n\\\\nThink step by step and reason yourself to the right decisions to make sure we get it right.\\\\nFirst lay out the names of the core classes, functions, methods that will be necessary, As well as a quick comment on their purpose.\\\\n\\\\nThen you will output the content of each file including ALL code.\\\\nEach file must strictly follow a markdown code block format, where the following tokens must be replaced such that\\\\nFILENAME is the lowercase file name including the file extension,\\\\nLANG is the markup code block language for the code\\'s language, and CODE is the code:\\\\n\\\\nFILENAME\\\\n```LANG\\\\nCODE\\\\n```\\\\n\\\\nPlease note that the code should be fully functional. No placeholders.\\\\n\\\\nYou will start with the \\\\\"entrypoint\\\\\" file, then go to the ones that are imported by that file, and so on.\\\\nFollow a language and framework appropriate best practice file naming convention.\\\\nMake sure that files contain all imports, types etc. The code should be fully functional. Make sure that code in different files are compatible with each other.\\\\nBefore you finish, double check that all parts of the architecture is present in the files.\\\\n\"\\n  }\\n]\\nChallenges#\\nAfter going through key ideas and demos of building LLM-centered agents, I start to see a couple common limitations:\\n\\n\\nFinite context length: The restricted context capacity limits the inclusion of historical information, detailed instructions, API call context, and responses. The design of the system has to work with this limited communication bandwidth, while mechanisms like self-reflection to learn from past mistakes would benefit a lot from long or infinite context windows. Although vector stores and retrieval can provide access to a larger knowledge pool, their representation power is not as powerful as full attention.\\n\\n\\nChallenges in long-term planning and task decomposition: Planning over a lengthy history and effectively exploring the solution space remain challenging. LLMs struggle to adjust plans when faced with unexpected errors, making them less robust compared to humans who learn from trial and error.\\n\\n\\nReliability of natural language interface: Current agent system relies on natural language as an interface between LLMs and external components such as memory and tools. However, the reliability of model outputs is questionable, as LLMs may make formatting errors and occasionally exhibit rebellious behavior (e.g. refuse to follow an instruction). Consequently, much of the agent demo code focuses on parsing model output.\\n\\n\\nCitation#\\nCited as:\\n\\nWeng, Lilian. (Jun 2023). “LLM-powered Autonomous Agents”. Lil’Log. https://lilianweng.github.io/posts/2023-06-23-agent/.\\n\\nOr\\n@article{weng2023agent,\\n  title   = \"LLM-powered Autonomous Agents\",\\n  author  = \"Weng, Lilian\",\\n  journal = \"lilianweng.github.io\",\\n  year    = \"2023\",\\n  month   = \"Jun\",\\n  url     = \"https://lilianweng.github.io/posts/2023-06-23-agent/\"\\n}\\nReferences#\\n[1] Wei et al. “Chain of thought prompting elicits reasoning in large language models.” NeurIPS 2022\\n[2] Yao et al. “Tree of Thoughts: Dliberate Problem Solving with Large Language Models.” arXiv preprint arXiv:2305.10601 (2023).\\n[3] Liu et al. “Chain of Hindsight Aligns Language Models with Feedback\\n“ arXiv preprint arXiv:2302.02676 (2023).\\n[4] Liu et al. “LLM+P: Empowering Large Language Models with Optimal Planning Proficiency” arXiv preprint arXiv:2304.11477 (2023).\\n[5] Yao et al. “ReAct: Synergizing reasoning and acting in language models.” ICLR 2023.\\n[6] Google Blog. “Announcing ScaNN: Efficient Vector Similarity Search” July 28, 2020.\\n[7] https://chat.openai.com/share/46ff149e-a4c7-4dd7-a800-fc4a642ea389\\n[8] Shinn & Labash. “Reflexion: an autonomous agent with dynamic memory and self-reflection” arXiv preprint arXiv:2303.11366 (2023).\\n[9] Laskin et al. “In-context Reinforcement Learning with Algorithm Distillation” ICLR 2023.\\n[10] Karpas et al. “MRKL Systems A modular, neuro-symbolic architecture that combines large language models, external knowledge sources and discrete reasoning.” arXiv preprint arXiv:2205.00445 (2022).\\n[11] Nakano et al. “Webgpt: Browser-assisted question-answering with human feedback.” arXiv preprint arXiv:2112.09332 (2021).\\n[12] Parisi et al. “TALM: Tool Augmented Language Models”\\n[13] Schick et al. “Toolformer: Language Models Can Teach Themselves to Use Tools.” arXiv preprint arXiv:2302.04761 (2023).\\n[14] Weaviate Blog. Why is Vector Search so fast? Sep 13, 2022.\\n[15] Li et al. “API-Bank: A Benchmark for Tool-Augmented LLMs” arXiv preprint arXiv:2304.08244 (2023).\\n[16] Shen et al. “HuggingGPT: Solving AI Tasks with ChatGPT and its Friends in HuggingFace” arXiv preprint arXiv:2303.17580 (2023).\\n[17] Bran et al. “ChemCrow: Augmenting large-language models with chemistry tools.” arXiv preprint arXiv:2304.05376 (2023).\\n[18] Boiko et al. “Emergent autonomous scientific research capabilities of large language models.” arXiv preprint arXiv:2304.05332 (2023).\\n[19] Joon Sung Park, et al. “Generative Agents: Interactive Simulacra of Human Behavior.” arXiv preprint arXiv:2304.03442 (2023).\\n[20] AutoGPT. https://github.com/Significant-Gravitas/Auto-GPT\\n[21] GPT-Engineer. https://github.com/AntonOsika/gpt-engineer\\n', metadata={'source': 'https://lilianweng.github.io/posts/2023-06-23-agent/'})]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "docs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea926c46-1953-460d-bb26-78b413a92fd7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "273a30f5-e56b-4468-a888-2a6954443e24",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4c3b3c7-7db0-4644-be16-c2ec8ecea457",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e58d78d7-a6b4-4a3c-a879-5a05707992a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import poppler #-utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a1e87f9f-327a-4cc0-b27f-40727d52000a",
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'pdfinfo'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[3], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mpdfinfo\u001b[39;00m\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'pdfinfo'"
     ]
    }
   ],
   "source": [
    "import pdfinfo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "754bd6f9-3d42-470d-b196-04dca88b1ce6",
   "metadata": {},
   "outputs": [
    {
     "ename": "PDFInfoNotInstalledError",
     "evalue": "Unable to get page count. Is poppler installed and in PATH?",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/pdf2image/pdf2image.py:581\u001b[0m, in \u001b[0;36mpdfinfo_from_path\u001b[0;34m(pdf_path, userpw, ownerpw, poppler_path, rawdates, timeout, first_page, last_page)\u001b[0m\n\u001b[1;32m    580\u001b[0m     env[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mLD_LIBRARY_PATH\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m poppler_path \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m:\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m+\u001b[39m env\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mLD_LIBRARY_PATH\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m--> 581\u001b[0m proc \u001b[38;5;241m=\u001b[39m Popen(command, env\u001b[38;5;241m=\u001b[39menv, stdout\u001b[38;5;241m=\u001b[39mPIPE, stderr\u001b[38;5;241m=\u001b[39mPIPE)\n\u001b[1;32m    583\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n",
      "File \u001b[0;32m/sw/pkgs/arc/python3.11-anaconda/2024.02-1/lib/python3.11/subprocess.py:1026\u001b[0m, in \u001b[0;36mPopen.__init__\u001b[0;34m(self, args, bufsize, executable, stdin, stdout, stderr, preexec_fn, close_fds, shell, cwd, env, universal_newlines, startupinfo, creationflags, restore_signals, start_new_session, pass_fds, user, group, extra_groups, encoding, errors, text, umask, pipesize, process_group)\u001b[0m\n\u001b[1;32m   1023\u001b[0m             \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstderr \u001b[38;5;241m=\u001b[39m io\u001b[38;5;241m.\u001b[39mTextIOWrapper(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstderr,\n\u001b[1;32m   1024\u001b[0m                     encoding\u001b[38;5;241m=\u001b[39mencoding, errors\u001b[38;5;241m=\u001b[39merrors)\n\u001b[0;32m-> 1026\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_execute_child(args, executable, preexec_fn, close_fds,\n\u001b[1;32m   1027\u001b[0m                         pass_fds, cwd, env,\n\u001b[1;32m   1028\u001b[0m                         startupinfo, creationflags, shell,\n\u001b[1;32m   1029\u001b[0m                         p2cread, p2cwrite,\n\u001b[1;32m   1030\u001b[0m                         c2pread, c2pwrite,\n\u001b[1;32m   1031\u001b[0m                         errread, errwrite,\n\u001b[1;32m   1032\u001b[0m                         restore_signals,\n\u001b[1;32m   1033\u001b[0m                         gid, gids, uid, umask,\n\u001b[1;32m   1034\u001b[0m                         start_new_session, process_group)\n\u001b[1;32m   1035\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m:\n\u001b[1;32m   1036\u001b[0m     \u001b[38;5;66;03m# Cleanup if the child failed starting.\u001b[39;00m\n",
      "File \u001b[0;32m/sw/pkgs/arc/python3.11-anaconda/2024.02-1/lib/python3.11/subprocess.py:1950\u001b[0m, in \u001b[0;36mPopen._execute_child\u001b[0;34m(self, args, executable, preexec_fn, close_fds, pass_fds, cwd, env, startupinfo, creationflags, shell, p2cread, p2cwrite, c2pread, c2pwrite, errread, errwrite, restore_signals, gid, gids, uid, umask, start_new_session, process_group)\u001b[0m\n\u001b[1;32m   1949\u001b[0m         err_msg \u001b[38;5;241m=\u001b[39m os\u001b[38;5;241m.\u001b[39mstrerror(errno_num)\n\u001b[0;32m-> 1950\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m child_exception_type(errno_num, err_msg, err_filename)\n\u001b[1;32m   1951\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m child_exception_type(err_msg)\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'pdfinfo'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mPDFInfoNotInstalledError\u001b[0m                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mpdf2image\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;241m*\u001b[39m\n\u001b[0;32m----> 2\u001b[0m images \u001b[38;5;241m=\u001b[39m convert_from_path(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mpapers/Pore-C.pdf\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/pdf2image/pdf2image.py:127\u001b[0m, in \u001b[0;36mconvert_from_path\u001b[0;34m(pdf_path, dpi, output_folder, first_page, last_page, fmt, jpegopt, thread_count, userpw, ownerpw, use_cropbox, strict, transparent, single_file, output_file, poppler_path, grayscale, size, paths_only, use_pdftocairo, timeout, hide_annotations)\u001b[0m\n\u001b[1;32m    124\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(poppler_path, PurePath):\n\u001b[1;32m    125\u001b[0m     poppler_path \u001b[38;5;241m=\u001b[39m poppler_path\u001b[38;5;241m.\u001b[39mas_posix()\n\u001b[0;32m--> 127\u001b[0m page_count \u001b[38;5;241m=\u001b[39m pdfinfo_from_path(\n\u001b[1;32m    128\u001b[0m     pdf_path, userpw, ownerpw, poppler_path\u001b[38;5;241m=\u001b[39mpoppler_path\n\u001b[1;32m    129\u001b[0m )[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mPages\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m    131\u001b[0m \u001b[38;5;66;03m# We start by getting the output format, the buffer processing function and if we need pdftocairo\u001b[39;00m\n\u001b[1;32m    132\u001b[0m parsed_fmt, final_extension, parse_buffer_func, use_pdfcairo_format \u001b[38;5;241m=\u001b[39m _parse_format(\n\u001b[1;32m    133\u001b[0m     fmt, grayscale\n\u001b[1;32m    134\u001b[0m )\n",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/pdf2image/pdf2image.py:607\u001b[0m, in \u001b[0;36mpdfinfo_from_path\u001b[0;34m(pdf_path, userpw, ownerpw, poppler_path, rawdates, timeout, first_page, last_page)\u001b[0m\n\u001b[1;32m    604\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m d\n\u001b[1;32m    606\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mOSError\u001b[39;00m:\n\u001b[0;32m--> 607\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m PDFInfoNotInstalledError(\n\u001b[1;32m    608\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mUnable to get page count. Is poppler installed and in PATH?\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    609\u001b[0m     )\n\u001b[1;32m    610\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m:\n\u001b[1;32m    611\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m PDFPageCountError(\n\u001b[1;32m    612\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mUnable to get page count.\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;132;01m{\u001b[39;00merr\u001b[38;5;241m.\u001b[39mdecode(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mutf8\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;250m \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mignore\u001b[39m\u001b[38;5;124m'\u001b[39m)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    613\u001b[0m     )\n",
      "\u001b[0;31mPDFInfoNotInstalledError\u001b[0m: Unable to get page count. Is poppler installed and in PATH?"
     ]
    }
   ],
   "source": [
    "from pdf2image import *\n",
    "images = convert_from_path('papers/Pore-C.pdf')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "446b6d42-d50e-47b4-ad55-3f6e88cff5b1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/bin/bash: apt-get: command not found\n"
     ]
    }
   ],
   "source": [
    "!apt-get update"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0ae318e6-ef5e-4a7f-b669-ac99775f16d1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Work Directory: /home/jpic/RAG-DEV/tutorials/RAG-with-BRAD\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Documents loading from: papers/\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/1 [00:00<?, ?it/s]Error loading file papers/Pore-C.pdf\n",
      "100%|██████████| 1/1 [00:00<00:00,  1.11it/s]"
     ]
    },
    {
     "ename": "PDFInfoNotInstalledError",
     "evalue": "Unable to get page count. Is poppler installed and in PATH?",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/pdf2image/pdf2image.py:581\u001b[0m, in \u001b[0;36mpdfinfo_from_path\u001b[0;34m(pdf_path, userpw, ownerpw, poppler_path, rawdates, timeout, first_page, last_page)\u001b[0m\n\u001b[1;32m    580\u001b[0m     env[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mLD_LIBRARY_PATH\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m poppler_path \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m:\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m+\u001b[39m env\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mLD_LIBRARY_PATH\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m--> 581\u001b[0m proc \u001b[38;5;241m=\u001b[39m Popen(command, env\u001b[38;5;241m=\u001b[39menv, stdout\u001b[38;5;241m=\u001b[39mPIPE, stderr\u001b[38;5;241m=\u001b[39mPIPE)\n\u001b[1;32m    583\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n",
      "File \u001b[0;32m/sw/pkgs/arc/python3.11-anaconda/2024.02-1/lib/python3.11/subprocess.py:1026\u001b[0m, in \u001b[0;36mPopen.__init__\u001b[0;34m(self, args, bufsize, executable, stdin, stdout, stderr, preexec_fn, close_fds, shell, cwd, env, universal_newlines, startupinfo, creationflags, restore_signals, start_new_session, pass_fds, user, group, extra_groups, encoding, errors, text, umask, pipesize, process_group)\u001b[0m\n\u001b[1;32m   1023\u001b[0m             \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstderr \u001b[38;5;241m=\u001b[39m io\u001b[38;5;241m.\u001b[39mTextIOWrapper(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstderr,\n\u001b[1;32m   1024\u001b[0m                     encoding\u001b[38;5;241m=\u001b[39mencoding, errors\u001b[38;5;241m=\u001b[39merrors)\n\u001b[0;32m-> 1026\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_execute_child(args, executable, preexec_fn, close_fds,\n\u001b[1;32m   1027\u001b[0m                         pass_fds, cwd, env,\n\u001b[1;32m   1028\u001b[0m                         startupinfo, creationflags, shell,\n\u001b[1;32m   1029\u001b[0m                         p2cread, p2cwrite,\n\u001b[1;32m   1030\u001b[0m                         c2pread, c2pwrite,\n\u001b[1;32m   1031\u001b[0m                         errread, errwrite,\n\u001b[1;32m   1032\u001b[0m                         restore_signals,\n\u001b[1;32m   1033\u001b[0m                         gid, gids, uid, umask,\n\u001b[1;32m   1034\u001b[0m                         start_new_session, process_group)\n\u001b[1;32m   1035\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m:\n\u001b[1;32m   1036\u001b[0m     \u001b[38;5;66;03m# Cleanup if the child failed starting.\u001b[39;00m\n",
      "File \u001b[0;32m/sw/pkgs/arc/python3.11-anaconda/2024.02-1/lib/python3.11/subprocess.py:1950\u001b[0m, in \u001b[0;36mPopen._execute_child\u001b[0;34m(self, args, executable, preexec_fn, close_fds, pass_fds, cwd, env, startupinfo, creationflags, shell, p2cread, p2cwrite, c2pread, c2pwrite, errread, errwrite, restore_signals, gid, gids, uid, umask, start_new_session, process_group)\u001b[0m\n\u001b[1;32m   1949\u001b[0m         err_msg \u001b[38;5;241m=\u001b[39m os\u001b[38;5;241m.\u001b[39mstrerror(errno_num)\n\u001b[0;32m-> 1950\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m child_exception_type(errno_num, err_msg, err_filename)\n\u001b[1;32m   1951\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m child_exception_type(err_msg)\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'pdfinfo'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mPDFInfoNotInstalledError\u001b[0m                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[3], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m rag\u001b[38;5;241m.\u001b[39mcreate_database(docsPath\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mpapers/\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[1;32m      2\u001b[0m                     dbName\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtutorialDatabase\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[1;32m      3\u001b[0m                     dbPath\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mdatabases/\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[1;32m      4\u001b[0m                     HuggingFaceEmbeddingsModel \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mBAAI/bge-base-en-v1.5\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[1;32m      5\u001b[0m                     chunk_size\u001b[38;5;241m=\u001b[39m[\u001b[38;5;241m700\u001b[39m],\n\u001b[1;32m      6\u001b[0m                     chunck_overlap\u001b[38;5;241m=\u001b[39m[\u001b[38;5;241m200\u001b[39m],\n\u001b[1;32m      7\u001b[0m                     v\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n",
      "File \u001b[0;32m~/RAG-DEV/BRAD/rag.py:309\u001b[0m, in \u001b[0;36mcreate_database\u001b[0;34m(docsPath, dbName, dbPath, HuggingFaceEmbeddingsModel, chunk_size, chunck_overlap, v)\u001b[0m\n\u001b[1;32m    302\u001b[0m text_loader_kwargs\u001b[38;5;241m=\u001b[39m{\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mautodetect_encoding\u001b[39m\u001b[38;5;124m'\u001b[39m: \u001b[38;5;28;01mTrue\u001b[39;00m}\n\u001b[1;32m    303\u001b[0m loader \u001b[38;5;241m=\u001b[39m DirectoryLoader(docsPath,\n\u001b[1;32m    304\u001b[0m                          glob\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m**/*.pdf\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m    305\u001b[0m                          loader_cls\u001b[38;5;241m=\u001b[39mUnstructuredPDFLoader, \n\u001b[1;32m    306\u001b[0m                          loader_kwargs\u001b[38;5;241m=\u001b[39mtext_loader_kwargs,\n\u001b[1;32m    307\u001b[0m                          show_progress\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m,\n\u001b[1;32m    308\u001b[0m                          use_multithreading\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[0;32m--> 309\u001b[0m docs_data \u001b[38;5;241m=\u001b[39m loader\u001b[38;5;241m.\u001b[39mload()\n\u001b[1;32m    311\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124mDocuments loaded...\u001b[39m\u001b[38;5;124m'\u001b[39m) \u001b[38;5;28;01mif\u001b[39;00m v \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m    313\u001b[0m chunk_size \u001b[38;5;241m=\u001b[39m [\u001b[38;5;241m700\u001b[39m] \u001b[38;5;66;03m#Chunk size \u001b[39;00m\n",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/langchain_community/document_loaders/directory.py:117\u001b[0m, in \u001b[0;36mDirectoryLoader.load\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    115\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mload\u001b[39m(\u001b[38;5;28mself\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m List[Document]:\n\u001b[1;32m    116\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Load documents.\"\"\"\u001b[39;00m\n\u001b[0;32m--> 117\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mlist\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mlazy_load())\n",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/langchain_community/document_loaders/directory.py:178\u001b[0m, in \u001b[0;36mDirectoryLoader.lazy_load\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    169\u001b[0m             futures\u001b[38;5;241m.\u001b[39mappend(\n\u001b[1;32m    170\u001b[0m                 executor\u001b[38;5;241m.\u001b[39msubmit(\n\u001b[1;32m    171\u001b[0m                     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_lazy_load_file_to_non_generator(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_lazy_load_file),\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    175\u001b[0m                 )\n\u001b[1;32m    176\u001b[0m             )\n\u001b[1;32m    177\u001b[0m         \u001b[38;5;28;01mfor\u001b[39;00m future \u001b[38;5;129;01min\u001b[39;00m concurrent\u001b[38;5;241m.\u001b[39mfutures\u001b[38;5;241m.\u001b[39mas_completed(futures):\n\u001b[0;32m--> 178\u001b[0m             \u001b[38;5;28;01mfor\u001b[39;00m item \u001b[38;5;129;01min\u001b[39;00m future\u001b[38;5;241m.\u001b[39mresult():\n\u001b[1;32m    179\u001b[0m                 \u001b[38;5;28;01myield\u001b[39;00m item\n\u001b[1;32m    180\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "File \u001b[0;32m/sw/pkgs/arc/python3.11-anaconda/2024.02-1/lib/python3.11/concurrent/futures/_base.py:449\u001b[0m, in \u001b[0;36mFuture.result\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    447\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m CancelledError()\n\u001b[1;32m    448\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_state \u001b[38;5;241m==\u001b[39m FINISHED:\n\u001b[0;32m--> 449\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m__get_result()\n\u001b[1;32m    451\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_condition\u001b[38;5;241m.\u001b[39mwait(timeout)\n\u001b[1;32m    453\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_state \u001b[38;5;129;01min\u001b[39;00m [CANCELLED, CANCELLED_AND_NOTIFIED]:\n",
      "File \u001b[0;32m/sw/pkgs/arc/python3.11-anaconda/2024.02-1/lib/python3.11/concurrent/futures/_base.py:401\u001b[0m, in \u001b[0;36mFuture.__get_result\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    399\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_exception:\n\u001b[1;32m    400\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 401\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_exception\n\u001b[1;32m    402\u001b[0m     \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[1;32m    403\u001b[0m         \u001b[38;5;66;03m# Break a reference cycle with the exception in self._exception\u001b[39;00m\n\u001b[1;32m    404\u001b[0m         \u001b[38;5;28mself\u001b[39m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "File \u001b[0;32m/sw/pkgs/arc/python3.11-anaconda/2024.02-1/lib/python3.11/concurrent/futures/thread.py:58\u001b[0m, in \u001b[0;36m_WorkItem.run\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m     55\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m\n\u001b[1;32m     57\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m---> 58\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfn(\u001b[38;5;241m*\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mkwargs)\n\u001b[1;32m     59\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mBaseException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m exc:\n\u001b[1;32m     60\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfuture\u001b[38;5;241m.\u001b[39mset_exception(exc)\n",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/langchain_community/document_loaders/directory.py:189\u001b[0m, in \u001b[0;36mDirectoryLoader._lazy_load_file_to_non_generator.<locals>.non_generator\u001b[0;34m(item, path, pbar)\u001b[0m\n\u001b[1;32m    188\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mnon_generator\u001b[39m(item: Path, path: Path, pbar: Optional[Any]) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m List:\n\u001b[0;32m--> 189\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m [x \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m func(item, path, pbar)]\n",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/langchain_community/document_loaders/directory.py:189\u001b[0m, in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m    188\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mnon_generator\u001b[39m(item: Path, path: Path, pbar: Optional[Any]) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m List:\n\u001b[0;32m--> 189\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m [x \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m func(item, path, pbar)]\n",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/langchain_community/document_loaders/directory.py:220\u001b[0m, in \u001b[0;36mDirectoryLoader._lazy_load_file\u001b[0;34m(self, item, path, pbar)\u001b[0m\n\u001b[1;32m    218\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    219\u001b[0m         logger\u001b[38;5;241m.\u001b[39merror(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mError loading file \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mstr\u001b[39m(item)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m--> 220\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m e\n\u001b[1;32m    221\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[1;32m    222\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m pbar:\n",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/langchain_community/document_loaders/directory.py:210\u001b[0m, in \u001b[0;36mDirectoryLoader._lazy_load_file\u001b[0;34m(self, item, path, pbar)\u001b[0m\n\u001b[1;32m    208\u001b[0m loader \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mloader_cls(\u001b[38;5;28mstr\u001b[39m(item), \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mloader_kwargs)\n\u001b[1;32m    209\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 210\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m subdoc \u001b[38;5;129;01min\u001b[39;00m loader\u001b[38;5;241m.\u001b[39mlazy_load():\n\u001b[1;32m    211\u001b[0m         \u001b[38;5;28;01myield\u001b[39;00m subdoc\n\u001b[1;32m    212\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mNotImplementedError\u001b[39;00m:\n",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/langchain_community/document_loaders/unstructured.py:88\u001b[0m, in \u001b[0;36mUnstructuredBaseLoader.lazy_load\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m     86\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mlazy_load\u001b[39m(\u001b[38;5;28mself\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m Iterator[Document]:\n\u001b[1;32m     87\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Load file.\"\"\"\u001b[39;00m\n\u001b[0;32m---> 88\u001b[0m     elements \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_get_elements()\n\u001b[1;32m     89\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_post_process_elements(elements)\n\u001b[1;32m     90\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmode \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124melements\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/langchain_community/document_loaders/pdf.py:73\u001b[0m, in \u001b[0;36mUnstructuredPDFLoader._get_elements\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m     70\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_get_elements\u001b[39m(\u001b[38;5;28mself\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m List:\n\u001b[1;32m     71\u001b[0m     \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01munstructured\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mpartition\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mpdf\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m partition_pdf\n\u001b[0;32m---> 73\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m partition_pdf(filename\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfile_path, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39munstructured_kwargs)\n",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/unstructured/documents/elements.py:591\u001b[0m, in \u001b[0;36mprocess_metadata.<locals>.decorator.<locals>.wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    589\u001b[0m \u001b[38;5;129m@functools\u001b[39m\u001b[38;5;241m.\u001b[39mwraps(func)\n\u001b[1;32m    590\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mwrapper\u001b[39m(\u001b[38;5;241m*\u001b[39margs: _P\u001b[38;5;241m.\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs: _P\u001b[38;5;241m.\u001b[39mkwargs) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;28mlist\u001b[39m[Element]:\n\u001b[0;32m--> 591\u001b[0m     elements \u001b[38;5;241m=\u001b[39m func(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[1;32m    592\u001b[0m     call_args \u001b[38;5;241m=\u001b[39m get_call_args_applying_defaults(func, \u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[1;32m    594\u001b[0m     regex_metadata: \u001b[38;5;28mdict\u001b[39m[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstr\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstr\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m call_args\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mregex_metadata\u001b[39m\u001b[38;5;124m\"\u001b[39m, {})\n",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/unstructured/file_utils/filetype.py:618\u001b[0m, in \u001b[0;36madd_filetype.<locals>.decorator.<locals>.wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    616\u001b[0m \u001b[38;5;129m@functools\u001b[39m\u001b[38;5;241m.\u001b[39mwraps(func)\n\u001b[1;32m    617\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mwrapper\u001b[39m(\u001b[38;5;241m*\u001b[39margs: _P\u001b[38;5;241m.\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs: _P\u001b[38;5;241m.\u001b[39mkwargs) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m List[Element]:\n\u001b[0;32m--> 618\u001b[0m     elements \u001b[38;5;241m=\u001b[39m func(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[1;32m    619\u001b[0m     params \u001b[38;5;241m=\u001b[39m get_call_args_applying_defaults(func, \u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[1;32m    620\u001b[0m     include_metadata \u001b[38;5;241m=\u001b[39m params\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124minclude_metadata\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mTrue\u001b[39;00m)\n",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/unstructured/file_utils/filetype.py:582\u001b[0m, in \u001b[0;36madd_metadata.<locals>.wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    580\u001b[0m \u001b[38;5;129m@functools\u001b[39m\u001b[38;5;241m.\u001b[39mwraps(func)\n\u001b[1;32m    581\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mwrapper\u001b[39m(\u001b[38;5;241m*\u001b[39margs: _P\u001b[38;5;241m.\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs: _P\u001b[38;5;241m.\u001b[39mkwargs) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m List[Element]:\n\u001b[0;32m--> 582\u001b[0m     elements \u001b[38;5;241m=\u001b[39m func(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[1;32m    583\u001b[0m     call_args \u001b[38;5;241m=\u001b[39m get_call_args_applying_defaults(func, \u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[1;32m    584\u001b[0m     include_metadata \u001b[38;5;241m=\u001b[39m call_args\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124minclude_metadata\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mTrue\u001b[39;00m)\n",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/unstructured/chunking/dispatch.py:74\u001b[0m, in \u001b[0;36madd_chunking_strategy.<locals>.wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m     71\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"The decorated function is replaced with this one.\"\"\"\u001b[39;00m\n\u001b[1;32m     73\u001b[0m \u001b[38;5;66;03m# -- call the partitioning function to get the elements --\u001b[39;00m\n\u001b[0;32m---> 74\u001b[0m elements \u001b[38;5;241m=\u001b[39m func(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[1;32m     76\u001b[0m \u001b[38;5;66;03m# -- look for a chunking-strategy argument --\u001b[39;00m\n\u001b[1;32m     77\u001b[0m call_args \u001b[38;5;241m=\u001b[39m get_call_args_applying_defaults(func, \u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/unstructured/partition/pdf.py:192\u001b[0m, in \u001b[0;36mpartition_pdf\u001b[0;34m(filename, file, include_page_breaks, strategy, infer_table_structure, ocr_languages, languages, include_metadata, metadata_filename, metadata_last_modified, chunking_strategy, hi_res_model_name, extract_images_in_pdf, extract_image_block_types, extract_image_block_output_dir, extract_image_block_to_payload, date_from_file_object, starting_page_number, extract_forms, form_extraction_skip_tables, **kwargs)\u001b[0m\n\u001b[1;32m    188\u001b[0m exactly_one(filename\u001b[38;5;241m=\u001b[39mfilename, file\u001b[38;5;241m=\u001b[39mfile)\n\u001b[1;32m    190\u001b[0m languages \u001b[38;5;241m=\u001b[39m check_language_args(languages \u001b[38;5;129;01mor\u001b[39;00m [], ocr_languages) \u001b[38;5;129;01mor\u001b[39;00m [\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124meng\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[0;32m--> 192\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m partition_pdf_or_image(\n\u001b[1;32m    193\u001b[0m     filename\u001b[38;5;241m=\u001b[39mfilename,\n\u001b[1;32m    194\u001b[0m     file\u001b[38;5;241m=\u001b[39mfile,\n\u001b[1;32m    195\u001b[0m     include_page_breaks\u001b[38;5;241m=\u001b[39minclude_page_breaks,\n\u001b[1;32m    196\u001b[0m     strategy\u001b[38;5;241m=\u001b[39mstrategy,\n\u001b[1;32m    197\u001b[0m     infer_table_structure\u001b[38;5;241m=\u001b[39minfer_table_structure,\n\u001b[1;32m    198\u001b[0m     languages\u001b[38;5;241m=\u001b[39mlanguages,\n\u001b[1;32m    199\u001b[0m     metadata_last_modified\u001b[38;5;241m=\u001b[39mmetadata_last_modified,\n\u001b[1;32m    200\u001b[0m     hi_res_model_name\u001b[38;5;241m=\u001b[39mhi_res_model_name,\n\u001b[1;32m    201\u001b[0m     extract_images_in_pdf\u001b[38;5;241m=\u001b[39mextract_images_in_pdf,\n\u001b[1;32m    202\u001b[0m     extract_image_block_types\u001b[38;5;241m=\u001b[39mextract_image_block_types,\n\u001b[1;32m    203\u001b[0m     extract_image_block_output_dir\u001b[38;5;241m=\u001b[39mextract_image_block_output_dir,\n\u001b[1;32m    204\u001b[0m     extract_image_block_to_payload\u001b[38;5;241m=\u001b[39mextract_image_block_to_payload,\n\u001b[1;32m    205\u001b[0m     date_from_file_object\u001b[38;5;241m=\u001b[39mdate_from_file_object,\n\u001b[1;32m    206\u001b[0m     starting_page_number\u001b[38;5;241m=\u001b[39mstarting_page_number,\n\u001b[1;32m    207\u001b[0m     extract_forms\u001b[38;5;241m=\u001b[39mextract_forms,\n\u001b[1;32m    208\u001b[0m     form_extraction_skip_tables\u001b[38;5;241m=\u001b[39mform_extraction_skip_tables,\n\u001b[1;32m    209\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs,\n\u001b[1;32m    210\u001b[0m )\n",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/unstructured/partition/pdf.py:321\u001b[0m, in \u001b[0;36mpartition_pdf_or_image\u001b[0;34m(filename, file, is_image, include_page_breaks, strategy, infer_table_structure, ocr_languages, languages, metadata_last_modified, hi_res_model_name, extract_images_in_pdf, extract_image_block_types, extract_image_block_output_dir, extract_image_block_to_payload, date_from_file_object, starting_page_number, extract_forms, form_extraction_skip_tables, **kwargs)\u001b[0m\n\u001b[1;32m    318\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m strategy \u001b[38;5;241m==\u001b[39m PartitionStrategy\u001b[38;5;241m.\u001b[39mOCR_ONLY:\n\u001b[1;32m    319\u001b[0m     \u001b[38;5;66;03m# NOTE(robinson): Catches file conversion warnings when running with PDFs\u001b[39;00m\n\u001b[1;32m    320\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m warnings\u001b[38;5;241m.\u001b[39mcatch_warnings():\n\u001b[0;32m--> 321\u001b[0m         elements \u001b[38;5;241m=\u001b[39m _partition_pdf_or_image_with_ocr(\n\u001b[1;32m    322\u001b[0m             filename\u001b[38;5;241m=\u001b[39mfilename,\n\u001b[1;32m    323\u001b[0m             file\u001b[38;5;241m=\u001b[39mfile,\n\u001b[1;32m    324\u001b[0m             include_page_breaks\u001b[38;5;241m=\u001b[39minclude_page_breaks,\n\u001b[1;32m    325\u001b[0m             languages\u001b[38;5;241m=\u001b[39mlanguages,\n\u001b[1;32m    326\u001b[0m             is_image\u001b[38;5;241m=\u001b[39mis_image,\n\u001b[1;32m    327\u001b[0m             metadata_last_modified\u001b[38;5;241m=\u001b[39mmetadata_last_modified \u001b[38;5;129;01mor\u001b[39;00m last_modification_date,\n\u001b[1;32m    328\u001b[0m             starting_page_number\u001b[38;5;241m=\u001b[39mstarting_page_number,\n\u001b[1;32m    329\u001b[0m             \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs,\n\u001b[1;32m    330\u001b[0m         )\n\u001b[1;32m    331\u001b[0m         out_elements \u001b[38;5;241m=\u001b[39m _process_uncategorized_text_elements(elements)\n\u001b[1;32m    333\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m out_elements\n",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/unstructured/partition/pdf.py:773\u001b[0m, in \u001b[0;36m_partition_pdf_or_image_with_ocr\u001b[0;34m(filename, file, include_page_breaks, languages, is_image, metadata_last_modified, starting_page_number, **kwargs)\u001b[0m\n\u001b[1;32m    771\u001b[0m         elements\u001b[38;5;241m.\u001b[39mextend(page_elements)\n\u001b[1;32m    772\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 773\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m page_number, image \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(\n\u001b[1;32m    774\u001b[0m         convert_pdf_to_images(filename, file), start\u001b[38;5;241m=\u001b[39mstarting_page_number\n\u001b[1;32m    775\u001b[0m     ):\n\u001b[1;32m    776\u001b[0m         page_elements \u001b[38;5;241m=\u001b[39m _partition_pdf_or_image_with_ocr_from_image(\n\u001b[1;32m    777\u001b[0m             image\u001b[38;5;241m=\u001b[39mimage,\n\u001b[1;32m    778\u001b[0m             languages\u001b[38;5;241m=\u001b[39mlanguages,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    782\u001b[0m             \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs,\n\u001b[1;32m    783\u001b[0m         )\n\u001b[1;32m    784\u001b[0m         elements\u001b[38;5;241m.\u001b[39mextend(page_elements)\n",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/unstructured/partition/pdf_image/pdf_image_utils.py:389\u001b[0m, in \u001b[0;36mconvert_pdf_to_images\u001b[0;34m(filename, file, chunk_size)\u001b[0m\n\u001b[1;32m    387\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    388\u001b[0m     f_bytes \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m--> 389\u001b[0m     info \u001b[38;5;241m=\u001b[39m pdf2image\u001b[38;5;241m.\u001b[39mpdfinfo_from_path(filename)\n\u001b[1;32m    391\u001b[0m total_pages \u001b[38;5;241m=\u001b[39m info[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mPages\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m    392\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m start_page \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(\u001b[38;5;241m1\u001b[39m, total_pages \u001b[38;5;241m+\u001b[39m \u001b[38;5;241m1\u001b[39m, chunk_size):\n",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/pdf2image/pdf2image.py:607\u001b[0m, in \u001b[0;36mpdfinfo_from_path\u001b[0;34m(pdf_path, userpw, ownerpw, poppler_path, rawdates, timeout, first_page, last_page)\u001b[0m\n\u001b[1;32m    604\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m d\n\u001b[1;32m    606\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mOSError\u001b[39;00m:\n\u001b[0;32m--> 607\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m PDFInfoNotInstalledError(\n\u001b[1;32m    608\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mUnable to get page count. Is poppler installed and in PATH?\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    609\u001b[0m     )\n\u001b[1;32m    610\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m:\n\u001b[1;32m    611\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m PDFPageCountError(\n\u001b[1;32m    612\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mUnable to get page count.\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;132;01m{\u001b[39;00merr\u001b[38;5;241m.\u001b[39mdecode(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mutf8\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;250m \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mignore\u001b[39m\u001b[38;5;124m'\u001b[39m)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    613\u001b[0m     )\n",
      "\u001b[0;31mPDFInfoNotInstalledError\u001b[0m: Unable to get page count. Is poppler installed and in PATH?"
     ]
    }
   ],
   "source": [
    "rag.create_database(docsPath='papers/',\n",
    "                    dbName='tutorialDatabase',\n",
    "                    dbPath='databases/',\n",
    "                    HuggingFaceEmbeddingsModel = 'BAAI/bge-base-en-v1.5',\n",
    "                    chunk_size=[700],\n",
    "                    chunck_overlap=[200],\n",
    "                    v=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1615be1a-a5bd-431e-a7fe-31658a025979",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "240b49b1-d1bf-4422-b492-5d53decea369",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "375f881e-d553-432a-8a53-386d70a583a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "from BRAD import rag"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ed985f98-0e6f-41ec-a2a0-8f93ca499d86",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import os\n",
    "from langchain.document_loaders import DirectoryLoader\n",
    "from langchain.document_loaders import UnstructuredPDFLoader\n",
    "from langchain.vectorstores import Chroma\n",
    "from langchain.embeddings import HuggingFaceEmbeddings\n",
    "import chromadb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9cba2d72-08bd-4706-803b-6a8a61955364",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_database(self, docsFile=None, docsPath='/nfs/turbo/umms-indikar/shared/projects/RAG/papers/', dbName=None, dbPath='/nfs/turbo/umms-indikar/shared/projects/RAG/databases/', HuggingFaceEmbeddingsModel = 'BAAI/bge-base-en-v1.5', chunk_size=[700], chunck_overlap=[200], v=False):\n",
    "    # Handle arguments\n",
    "    docsPath += docsFile\n",
    "    dbPath   += dbName\n",
    "    \n",
    "    local = os.getcwd()  ## Get local dir\n",
    "    os.chdir(local)      ## shift the work dir to local dir\n",
    "    \n",
    "    print('\\nWork Directory: {}'.format(local)) if v else None\n",
    "\n",
    "    #%% Phase 1 - Load DB\n",
    "    embeddings_model = HuggingFaceEmbeddings(model_name=HuggingFaceEmbeddingsModel)\n",
    "    \n",
    "    print('\\nDocuments loading from:', docsPath) if v else None\n",
    "\n",
    "    text_loader_kwargs={'autodetect_encoding': True}\n",
    "    loader = DirectoryLoader(docsPath,\n",
    "                             glob=\"**/*.pdf\",\n",
    "                             loader_cls=UnstructuredPDFLoader, \n",
    "                             loader_kwargs=text_loader_kwargs,\n",
    "                             show_progress=True,\n",
    "                             use_multithreading=True)\n",
    "    docs_data = loader.load()\n",
    "\n",
    "    print('\\nDocuments loaded...') if v else None\n",
    "    \n",
    "    chunk_size = [700] #Chunk size \n",
    "    chunk_overlap = [200] #Chunk overlap\n",
    "\n",
    "    for i in range(len(chunk_size)):\n",
    "        for j in range(len(chunk_overlap)):\n",
    "            text_splitter = RecursiveCharacterTextSplitter(chunk_size = chunk_size[i],\n",
    "                                                            chunk_overlap = chunk_overlap[j],\n",
    "                                                            separators=[\" \", \",\", \"\\n\", \". \"])\n",
    "            data_splits = text_splitter.split_documents(docs_data)\n",
    "            \n",
    "            print('Documents split into chunks...') if v else None\n",
    "            print('Initializing Chroma Database...') if v else None\n",
    "\n",
    "            dbName = \"DB_cosine_cSize_%d_cOver_%d\" %(chunk_size[i], chunk_overlap[j])\n",
    "\n",
    "            p2_2 = subprocess.run('mkdir  %s/*'%(dbPath+dbName), shell=True)\n",
    "            _client_settings = chromadb.PersistentClient(path=(dbPath+dbName))\n",
    "\n",
    "            vectordb = Chroma.from_documents(documents           = data_splits,\n",
    "                                             embedding           = embeddings_model,\n",
    "                                             client              = _client_settings,\n",
    "                                             collection_name     = dbName,\n",
    "                                             collection_metadata = {\"hnsw:space\": \"cosine\"})\n",
    "\n",
    "            print('Completed Chroma Database: ', dbName) if v else None\n",
    "            del vectordb, text_splitter, data_splits"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a53ec77-96e9-4356-8c8a-f16ea21f2606",
   "metadata": {},
   "source": [
    "# Connecting Literature Databases to BRAD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "24efb3c9-c28e-41c5-9c97-b1099eed9bf5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter your NVIDIA API key:  ········\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/langchain_nvidia_ai_endpoints/_statics.py:313: UserWarning: Model mistral_7b is deprecated. Using mistralai/mistral-7b-instruct-v0.2 instead.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "from BRAD import llms\n",
    "llm = llms.load_nvidia()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e54223fe-6cb3-4af3-858e-9547be80ee9b",
   "metadata": {},
   "source": [
    "## Specifiying the Database"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "984c8386-b70d-4a19-919e-f06ae079742a",
   "metadata": {},
   "source": [
    "When running `brad.chat()`, there is an option to use a previously saved database. **ADD how to s Type Y to supplement your query with the database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "d4d1f2e0-b7c5-4ca8-812f-d664f2df10bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import os\n",
    "from langchain.document_loaders import DirectoryLoader\n",
    "from langchain.document_loaders import UnstructuredPDFLoader\n",
    "from langchain.vectorstores import Chroma\n",
    "from langchain.embeddings import HuggingFaceEmbeddings\n",
    "import chromadb\n",
    "from langchain_community.document_loaders import PyPDFLoader\n",
    "from langchain_text_splitters import RecursiveCharacterTextSplitter\n",
    "from langchain import PromptTemplate, LLMChain\n",
    "from langchain.memory import ConversationBufferMemory\n",
    "from langchain.chains import ConversationChain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "26bcd5a8-370d-4365-9dc8-f4802a73dc30",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "# Load the database\n",
    "persist_directory = '/nfs/turbo/umms-indikar/shared/projects/RAG/databases/DigitalLibrary-10-June-2024/'\n",
    "embeddings_model = HuggingFaceEmbeddings(model_name='BAAI/bge-base-en-v1.5')\n",
    "db_name = \"DigitalLibrary\"\n",
    "_client_settings = chromadb.PersistentClient(path=(persist_directory + db_name))\n",
    "vectordb = Chroma(persist_directory=persist_directory, embedding_function=embeddings_model, client=_client_settings, collection_name=db_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e3794e4-2168-406d-8fa6-bf0bf2dab499",
   "metadata": {},
   "source": [
    "## Using the RAG system"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93ad94aa-b832-4905-a607-5aae8fefeb5d",
   "metadata": {},
   "source": [
    "## Viewing the Documents from BRAD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83d942b5-2f8b-4a34-a87d-021255d96ad7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "b5a97f0d-9499-4bd8-8d6d-107457efaeaa",
   "metadata": {},
   "source": [
    "# MultiQuery RAG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2468f1e2-6308-4e45-b509-1ef48480ea67",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome to RAG! The chat log from this conversation will be saved to /home/jpic/BRAD/2024-06-16-23:53:20.661368.json. How can I help?\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sun 16 Jun 2024 11:53:20 PM EDT INFO local\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  /force RAG what cellular processes is the MYOD gene involved in?\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sun 16 Jun 2024 11:53:42 PM EDT INFO RAG\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==================================================\n",
      "RAG >> 1: "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/langchain_core/_api/deprecation.py:119: LangChainDeprecationWarning: The method `BaseRetriever.get_relevant_documents` was deprecated in langchain-core 0.1.46 and will be removed in 0.3.0. Use invoke instead.\n",
      "  warn_deprecated(\n",
      "Sun 16 Jun 2024 11:53:43 PM EDT INFO Generated queries: ['1. Which cellular functions does the MYOD gene influence or regulate?', '2. Can you identify the specific cellular pathways where the MYOD gene plays a role?', '3. What are the major cellular activities associated with the expression of the MYOD gene?']\n",
      "/home/jpic/.local/lib/python3.11/site-packages/langchain_core/_api/deprecation.py:119: LangChainDeprecationWarning: The method `Chain.__call__` was deprecated in langchain 0.1.0 and will be removed in 0.3.0. Use invoke instead.\n",
      "  warn_deprecated(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "expressed, how does Myod regulate skeletal muscle cell\n",
      "differentiation? In one sense, the answer seems fairly simple:Myod is a transcription factor with binding sites in theregulatory regions of many genes that are expressed in skeletalmuscle. Myod forms heterodimers with the nearly ubiquitousE-protein sub-family of bHLH proteins through the interactionof the HLH domains (see Fig. 1) (Lassar et al., 1991; Murre etal., 1989). The basic regions act as sequence-speciﬁc DNA-binding domains that recognize a binding site with the simplecore consensus sequence of CANNTG, termed an E-box, andshow additional preferences for internal and ﬂankingsequences (Blackwell and Weintraub, 1990). Myod has a singleamino-terminal acidic-activation domain, as determined by its\n",
      "fusion to the heterologous DNA-binding domain of the Gal4\n",
      "protein (Weintraub et al., 1991b), whereas E-proteins have amore complex mix of activation and repression domains(Markus et al., 2002) (see Fig. 1). Therefore, the simple\n",
      "\n",
      "There is a positive correlation between regulated MyoD\n",
      "binding (i.e., sites that are preferentially bound in myoblasts or\n",
      "differentiated myotubes) and gene expression. This suggests\n",
      "that sites regulating gene transcription in myotubes require addi-tional factors to modulate MyoD binding. Indeed, the sites asso-\n",
      "ciated with myotube-expressed genes are enriched for motifs of\n",
      "factors that are activated by MyoD and function with MyoD ina positive feed-forward circuit, as demonstrated previouslywith Mef2 ( Penn et al., 2004 ) and Myog ( Cao et al., 2006 ). In addi-\n",
      "tion, the Pbx/Meis complex cooperates with MyoD in activatinga subset of genes ( Berkes et al., 2004; Maves et al., 2007 ).\n",
      "Another ﬁnding was that genes decreasing expression with\n",
      "differentiation were associated with decreased MyoD binding\n",
      "at sites enriched for RP58 and AP1 motifs. A recent study (pub-lished after our analysis was complete) identiﬁed RP58 as a gene\n",
      "activated by MyoD during muscle differentiation that\n",
      "\n",
      "Cell, Vol. 75, 1241-1244, December 31, 1993, Copyright @ 1993 by Cell Press \n",
      "The MyoD Family and Myogenesis: \n",
      "Redundancy, Networks, \n",
      "and Thresholds \n",
      "Harold Weintraub \n",
      "Fred Hutchinson Cancer Research Center \n",
      "Howard Hughes Medical Institute Laboratory \n",
      "1124 Columbia Street \n",
      "Seattle, Washington 98104 \n",
      "The members of the MyoD family of myogenic transcrip- \n",
      "tional regulators (MyoD, Myf-5, MRF4, myogenin) have \n",
      "the remarkable property that, when expressed from a con- \n",
      "stitutive promoter, they can convert a large number of dif- \n",
      "ferent cell types into muscle (reviewed by Buckingham, \n",
      "1992; Emerson, 1990; Olson, 1990; Weintraub et al., \n",
      "1991). In contrast, mice with homozygous gene-targeted \n",
      "mutations in either MyoD or Myf-5 produce fairly normal \n",
      "amounts of muscle (Rudnicki et al., 1992; Braun et al., \n",
      "1992). This apparent paradox has now been resolved by \n",
      "two recent studies. In one, a simple genetic cross has \n",
      "shown that in the double homozygous mutants, in which \n",
      "both MyoD and Myfd are\n",
      "\n",
      "of mRNA is expressed in cells or embryos to inhibit the \n",
      "expression of specific mRNAs, allows the generation of lack-of-function phenotypes and permits the assessment \n",
      "of the role of specific genes in development in the absence \n",
      "of classical developmental genetics. \n",
      "More recently, in a remarkable series of experiments, \n",
      "Hal identified a single gene, myoD, which can elicit the \n",
      "entire prog'ram of muscle differentiation when introduced \n",
      "into nonmuscle cell types. He demonstrated that muscle \n",
      "cell differentiation could be elicited in fibroblasts following \n",
      "the transfer of genomic DNA from muscle cells. This led \n",
      "Hal to the isolation of the myoD gene, which encodes a \n",
      "transcriptional activator that promotes myogenesis by di- \n",
      "rectly binding to a control region shared by muscle-specific \n",
      "genes. Moreover, MyoD activates its own transcription, \n",
      "creating an autoregulatory loop that stabilizes the commit- \n",
      "ment to myogenesis. Structure-function analysis allowed \n",
      "the definition of a complex\n",
      "\n",
      "2687 Review\n",
      "Genetic experiments have shown that either Myod or Myf5\n",
      "can specify the skeletal muscle lineage in mice: mouse muscle\n",
      "develops relatively normally without either Myod or Myf5 ; but\n",
      "disrupting both genes results in the absence of skeletal musclecells, indicating that these genes are necessary to establish a\n",
      "viable muscle lineage (Rudnicki et al., 1993). Myog appearsto function ‘downstream’ of Myf5 and Myod to activate musclegene expression, and appears to be crucial for thedifferentiation of muscle cells in vivo. The disruption of Myog\n",
      "in mice prevents muscle differentiation in vivo, despite thecontinued expression of Myod , whereas cells cultured from the\n",
      "muscle primordium of Myog null mice form skeletal muscle as\n",
      "efficiently as cells derived from wild-type mice (Hasty et al.,1993; Nabeshima et al., 1993). This indicates that the functionsof Myod in terminal differentiation are suppressed by in vivosignaling events. By contrast, Myog is ‘blind’ to these\n",
      "suppressive signals\n",
      "\n",
      "as described thus far is designed to commit any single \n",
      "cell to an all-or-none decision, once levels of myogenic \n",
      "regulatory gene activity approach the threshold. In terms \n",
      "of the intricacies of their control networks, the MyoD (and \n",
      "achaete-scute) gene families share many parallels to the \n",
      "cyclin-C/n gene family, which is designed to control the \n",
      "START decision for the cell cycle. \n",
      "While it is not clear what potential myoblasts become \n",
      "(or, indeed, whether they even survive) in the MyoD- and \n",
      "My&!&deficient mice, it is notable that additional fat is ob- \n",
      "served in place of muscle. It has been known for some \n",
      "time that 1 OTB cell lines can give rise to both muscle and \n",
      "fat. Also interesting is the fact that while MyoD can cause \n",
      "growth arrest in Gl , the various mutants did not seem to \n",
      "develop tumors. However, the number of mice may not \n",
      "be large enough and their lives long enough to evaluate \n",
      "the consequences fully. Diversion of myogenic cells to \n",
      "an alternative lineage might also\n",
      "\n",
      "Harter,2003). In contrast to the differentiating muscle cell, whereMyod is associated with HATs, in the myoblast (the replicatingmuscle precursor cell), Myod is associated with histonedeacetylases (HDACs) and might actively suppress geneexpression (Fulco et al., 2003; Mal and Harter, 2003; Mal etal., 2001).\n",
      "These ﬁndings strongly suggest that Myod acts to negatively\n",
      "regulate the transcription of some genes in the myoblast andthat muscle differentiation is initiated when Myod switchesfrom its association with repressive factors to activatingfactors. The differentiation of Myod- expressing cell lines, such\n",
      "as the C2C12 myoblast cell line (Silberstein et al., 1986), canbe induced in culture by removing serum or it can be preventedby adding growth factors, such as Fgf or Tgf β(De Angelis et\n",
      "al., 1998; Li et al., 1992), indicating that mitogen stimulationin vivo might sustain a myoblast state. In addition, Notchsignaling represses both Myod transcription and Myod protein\n",
      "activity (Kopan et\n",
      "\n",
      "a combination of promoter-speciﬁc regulation of Myod binding and activity.\n",
      "Because Myod initiates the myogenic differentiation\n",
      "program and that program temporally regulates the activity of\n",
      "Myod, it follows that Myod programs the regulation of its ownactivity. It does this, at least in part, through a feed-forward\n",
      "Development\n",
      "\n",
      "is active (Penn et al., 2004). The feed-forward mechanism regulates the acti vity of Myod\n",
      "at a subset of promoters and imposes a temporal order on Myod-mediated gene expression. This diagram uses the graphical languag e BioD\n",
      "(Cook et al., 2001).\n",
      "Development\n",
      "\n",
      "transcription and Myod protein\n",
      "activity (Kopan et al., 1994; Nofziger et al., 1999), andprobably contributes to regulating differentiation in vivo. It isinteresting that while we have identiﬁed several mechanismsthat might delay myoblast differentiation, such as mitogens andNotch signaling, we do not yet have a good understanding ofthe events that occur in vivo to overcome these inhibitorysignals and to induce differentiation at a speciﬁc time andplace.\n",
      "A feed-forward circuit as a quantal step\n",
      "How does a single transcription factor execute an entireprogram of cell differentiation? How does a single event result\n",
      "in a predictable and complex response? Microarray expressionstudies of cultured C2C12 cells have shown that expressionlevels of many RNAs change during skeletal muscledifferentiation (Delgado et al., 2003; Tomczak et al., 2004). To\n",
      "determine how many of these changes are caused by theexpression and activity of Myod, we assessed gene expressionchanges in ﬁbroblasts with an\n",
      "Human: what cellular processes is the MYOD gene involved in?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      " The MYOD gene is involved in muscle differentiation. It is a transcription factor that regulates skeletal muscle cell differentiation by binding to specific DNA sequences, known as E-boxes, and forming heterodimers with E-proteins. Myod activation leads to the transcription of genes required for muscle cell differentiation and the suppression of genes that are not needed in differentiated muscle cells. The switch from Myod's association with repressive factors to activating factors initiates the myogenic differentiation program, which regulates the activity of Myod itself through a feed-forward mechanism. This feed-forward mechanism imposes a temporal order on Myod-mediated gene expression and enables Myod to execute an entire program of cell differentiation.\n",
      "Source: The circuitry of a master switch.pdf\n",
      "Source: Genome-wide MyoD Binding.pdf\n",
      "Source: The MyoD Family and Myogenesis Redundancy, Networks, and Thresholds.pdf\n",
      "Source: Harold Weintraub.pdf\n",
      "Source: The circuitry of a master switch.pdf\n",
      "Source: The MyoD Family and Myogenesis Redundancy, Networks, and Thresholds.pdf\n",
      "Source: The circuitry of a master switch.pdf\n",
      "Source: The circuitry of a master switch.pdf\n",
      "Source: The circuitry of a master switch.pdf\n",
      "Source: The circuitry of a master switch.pdf\n",
      "==================================================\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Input >>  q\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Thanks for chatting today! I hope to talk soon, and don't forget that a record of this conversation is available at: /home/jpic/BRAD/2024-06-16-23:53:20.661368.json\n"
     ]
    }
   ],
   "source": [
    "from BRAD import brad\n",
    "brad.chat(ragvectordb=vectordb)"
   ]
  },
  {
   "cell_type": "raw",
   "id": "a2815a04-1bef-4454-9368-98689924818d",
   "metadata": {},
   "source": [
    "/force RAG what cellular processes is the MYOD gene involved in?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33ce017a-2299-4c52-91ae-650b90ed17e8",
   "metadata": {},
   "source": [
    "# Contextual Compression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "e3f74302-faa3-48da-96f1-05962c23f1fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "chatstatus = {\n",
    "    'config' : {\n",
    "        'debug':True\n",
    "    }\n",
    "}\n",
    "query = 'What cellular processes is MYOD involved in?'\n",
    "documentSearch = vectordb.similarity_search_with_relevance_scores(query=query, k=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "4f27956c-e022-4443-a6b0-7281312a8f81",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.7678906917572021"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "documentSearch[0][1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "c0b2c00b-442b-4ae2-9084-bcd0c49073d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def summarizeDocumentTemplate():\n",
    "    template = \"\"\"**INSTRUCTIONS**\n",
    "You are an assistant responsible for compressing the important information in a document.\n",
    "You will be given a users query and a piece of text. Summarize the text with the following aims:\n",
    "1. remove information that is not complete ideas or unrelated to the topic of the user\n",
    "2. improve the clarity of the writing and information\n",
    "If there is no relevant information, say \"None\"\n",
    "\n",
    "**USER QUERY**\n",
    "{user_query}\n",
    "\n",
    "**TEXT**\n",
    "{text}\n",
    "\n",
    "**OUTPUT**\n",
    "<put summary output here>\n",
    "\"\"\"\n",
    "    return template"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "fe4ea774-9df2-401e-940d-79631cb92e1f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "============\n",
      "a combination of promoter-speciﬁc regulation of Myod binding and activity.\n",
      "Because Myod initiates the myogenic differentiation\n",
      "program and that program temporally regulates the activity of\n",
      "Myod, it follows that Myod programs the regulation of its ownactivity. It does this, at least in part, through a feed-forward\n",
      "Development\n",
      "Summary: MYOD is involved in initiating the myogenic differentiation program, which temporally regulates its own activity through a feed-forward mechanism.\n",
      "============\n",
      "transcription and Myod protein\n",
      "activity (Kopan et al., 1994; Nofziger et al., 1999), andprobably contributes to regulating differentiation in vivo. It isinteresting that while we have identiﬁed several mechanismsthat might delay myoblast differentiation, such as mitogens andNotch signaling, we do not yet have a good understanding ofthe events that occur in vivo to overcome these inhibitorysignals and to induce differentiation at a speciﬁc time andplace.\n",
      "A feed-forward circuit as a quantal step\n",
      "How does a single transcription factor execute an entireprogram of cell differentiation? How does a single event result\n",
      "in a predictable and complex response? Microarray expressionstudies of cultured C2C12 cells have shown that expressionlevels of many RNAs change during skeletal muscledifferentiation (Delgado et al., 2003; Tomczak et al., 2004). To\n",
      "determine how many of these changes are caused by theexpression and activity of Myod, we assessed gene expressionchanges in ﬁbroblasts with an\n",
      "Summary: MYOD is a transcription factor involved in muscle differentiation as indicated by its role in gene transcription and regulation of myoblast differentiation both in vitro and in vivo (Kopan et al., 1994; Nofziger et al., 1999). Despite known inhibitors of differentiation like mitogens and Notch signaling, the precise mechanisms allowing differentiation to occur at specific times and places remain unclear. As for the execution of an entire program of cell differentation by a single transcription factor, research has shown that expression levels of many RNAs change during skeletal muscle differentiation in cultured C2C12 cells. To investigate the role of MYOD in these changes, gene expression was assessed in fibroblasts.\n",
      "============\n",
      "compartments. Thus, myogenin-independent myoblast \n",
      "identity seems to be definable by the generation of approx- \n",
      "imately the right number of cells in the right positions of \n",
      "the body under the direction of MyoD-Myf-5 function. \n",
      "Regulation of Myogenic Genes In Vitro and In Vivo \n",
      "Tissue culture experiments also support the notion that Minireview \n",
      "myogenin acts downstream of MyoD and Myf-5 since most \n",
      "myoblasts in culture express MyoD or Myf-5, but turn on \n",
      "myogenin only when induced to differentiate. Cell culture \n",
      "experiments have also contributed several additional lay- \n",
      "ers to the regulatory circuitry, showing that MyoD can acti- \n",
      "vate its own transcription as well as that of myogenin; \n",
      "myogenin can activate itself and as well as MyoD; and \n",
      "Myfd can activate itself, MyoD, and myogenin. Activation \n",
      "of myogenin seems to be mediated, in part, via a down- \n",
      "stream muscle transcription factor, MEF2 (Edmondson et \n",
      "al., 1992), so that in this case what is downstream and \n",
      "what is upstream may\n",
      "Summary: MyoD is a key transcription factor involved in muscle development. It regulates the expression of myogenin and other muscle-specific genes during both in vitro (tissue culture) and in vivo (living organism) processes. MyoD can also activate its own transcription and that of myogenin, while myogenin can activate MyoD and itself. The activation of myogenin is mediated in part by a muscle transcription factor called MEF2.\n",
      "============\n",
      "Cell, Vol. 75, 1241-1244, December 31, 1993, Copyright @ 1993 by Cell Press \n",
      "The MyoD Family and Myogenesis: \n",
      "Redundancy, Networks, \n",
      "and Thresholds \n",
      "Harold Weintraub \n",
      "Fred Hutchinson Cancer Research Center \n",
      "Howard Hughes Medical Institute Laboratory \n",
      "1124 Columbia Street \n",
      "Seattle, Washington 98104 \n",
      "The members of the MyoD family of myogenic transcrip- \n",
      "tional regulators (MyoD, Myf-5, MRF4, myogenin) have \n",
      "the remarkable property that, when expressed from a con- \n",
      "stitutive promoter, they can convert a large number of dif- \n",
      "ferent cell types into muscle (reviewed by Buckingham, \n",
      "1992; Emerson, 1990; Olson, 1990; Weintraub et al., \n",
      "1991). In contrast, mice with homozygous gene-targeted \n",
      "mutations in either MyoD or Myf-5 produce fairly normal \n",
      "amounts of muscle (Rudnicki et al., 1992; Braun et al., \n",
      "1992). This apparent paradox has now been resolved by \n",
      "two recent studies. In one, a simple genetic cross has \n",
      "shown that in the double homozygous mutants, in which \n",
      "both MyoD and Myfd are\n",
      "Summary: MYOD is a member of the MyoD family of myogenic transcriptional regulators, including Myf-5, MRF4, and myogenin. When expressed from a constitutive promoter, MYOD can convert various cell types into muscle. However, in mice with homogenous gene-targeted mutations for both MYOD and Myf-5, normal muscle amounts are produced, which was previously considered a paradox. This paradox has been resolved by recent studies that show in double homogenous mutants, both MyoD and Myf-5 are required for full muscle conversion and development.\n",
      "============\n",
      "expressed, how does Myod regulate skeletal muscle cell\n",
      "differentiation? In one sense, the answer seems fairly simple:Myod is a transcription factor with binding sites in theregulatory regions of many genes that are expressed in skeletalmuscle. Myod forms heterodimers with the nearly ubiquitousE-protein sub-family of bHLH proteins through the interactionof the HLH domains (see Fig. 1) (Lassar et al., 1991; Murre etal., 1989). The basic regions act as sequence-speciﬁc DNA-binding domains that recognize a binding site with the simplecore consensus sequence of CANNTG, termed an E-box, andshow additional preferences for internal and ﬂankingsequences (Blackwell and Weintraub, 1990). Myod has a singleamino-terminal acidic-activation domain, as determined by its\n",
      "fusion to the heterologous DNA-binding domain of the Gal4\n",
      "protein (Weintraub et al., 1991b), whereas E-proteins have amore complex mix of activation and repression domains(Markus et al., 2002) (see Fig. 1). Therefore, the simple\n",
      "Summary: MYOD is a transcription factor that regulates skeletal muscle cell differentiation by binding to the regulatory regions of genes expressed in muscle cells. It forms heterodimers with E-proteins through the interaction of their HLH domains, and recognizes DNA binding sites with the E-box consensus sequence (CANNTG). MYOD has an acidic-activation domain, while E-proteins have a more complex mix of activation and repression domains. (summary: MYOD is a muscle-specific transcription factor that regulates muscle gene expression by binding to the E-box sequence in partnership with E-proteins, using its acidic-activation domain to regulate gene expression.)\n",
      "============\n",
      "Harter,2003). In contrast to the differentiating muscle cell, whereMyod is associated with HATs, in the myoblast (the replicatingmuscle precursor cell), Myod is associated with histonedeacetylases (HDACs) and might actively suppress geneexpression (Fulco et al., 2003; Mal and Harter, 2003; Mal etal., 2001).\n",
      "These ﬁndings strongly suggest that Myod acts to negatively\n",
      "regulate the transcription of some genes in the myoblast andthat muscle differentiation is initiated when Myod switchesfrom its association with repressive factors to activatingfactors. The differentiation of Myod- expressing cell lines, such\n",
      "as the C2C12 myoblast cell line (Silberstein et al., 1986), canbe induced in culture by removing serum or it can be preventedby adding growth factors, such as Fgf or Tgf β(De Angelis et\n",
      "al., 1998; Li et al., 1992), indicating that mitogen stimulationin vivo might sustain a myoblast state. In addition, Notchsignaling represses both Myod transcription and Myod protein\n",
      "activity (Kopan et\n",
      "Summary: Myod is involved in regulating gene expression during muscle cell differentiation. In the replicating muscle precursor cell (myoblast), Myod associates with histone deacetylases (HDACs) and suppresses gene expression. However, during muscle differentiation, Myod switches to associating with histone acetyltransferases (HATs) and activates gene expression. This switch is important for initiating muscle differentiation, and it can be induced in cell lines such as C2C12 myoblasts by removing serum or adding growth factors like Fgf or Tgf β. Notch signaling also represses Myod's transcription and activity.\n",
      "============\n",
      "identiﬁcation of Myod1 (Lassar et al.,1986), a basic-helix-loop-helix (bHLH) transcription factorthat can induce skeletal muscle differentiation in cells frommany different lineages.The identiﬁcation of Myod as a master regulatory gene of\n",
      "skeletal muscle differentiation provided the opportunity to\n",
      "assess the roles of tissue-speciﬁc transcription factors and\n",
      "chromatin-associated proteins in regulating cell differentiation.\n",
      "It also provided the opportunity to address another importantquestion: how does a single transcription factor execute anentire program of cell differentiation? One reason to beinterested in the answer to this question is that it is a speciﬁc\n",
      "example of the broader question: how does a single event result\n",
      "in a predictable, complex response? It is likely thatunderstanding the orchestration of skeletal muscle geneexpression by Myod will provide insights into the regulationof other complex biological events.\n",
      "This review describes generally how Myod regulates the\n",
      "program\n",
      "Summary: MYOD, identified by Lassar et al. (1986), is a bHLH transcription factor that triggers skeletal muscle differentiation. Its discovery led to investigating the roles of muscle-specific transcription factors and chromatin-associated proteins in cell differentiation. Furthermore, it raised the question of how a single transcription factor manages an entire program of cell differentiation. This review focuses on how MYOD orchestrates the muscle gene expression process.\n",
      "============\n",
      "of mRNA is expressed in cells or embryos to inhibit the \n",
      "expression of specific mRNAs, allows the generation of lack-of-function phenotypes and permits the assessment \n",
      "of the role of specific genes in development in the absence \n",
      "of classical developmental genetics. \n",
      "More recently, in a remarkable series of experiments, \n",
      "Hal identified a single gene, myoD, which can elicit the \n",
      "entire prog'ram of muscle differentiation when introduced \n",
      "into nonmuscle cell types. He demonstrated that muscle \n",
      "cell differentiation could be elicited in fibroblasts following \n",
      "the transfer of genomic DNA from muscle cells. This led \n",
      "Hal to the isolation of the myoD gene, which encodes a \n",
      "transcriptional activator that promotes myogenesis by di- \n",
      "rectly binding to a control region shared by muscle-specific \n",
      "genes. Moreover, MyoD activates its own transcription, \n",
      "creating an autoregulatory loop that stabilizes the commit- \n",
      "ment to myogenesis. Structure-function analysis allowed \n",
      "the definition of a complex\n",
      "Summary: MYOD is a gene that encodes a transcriptional activator involved in muscle differentiation. When introduced into non-muscle cell types, it promotes myogenesis by binding to a control region shared by muscle-specific genes and activating its own transcription, creating an autoregulatory loop that strengthens the commitment to muscle cell development.\n",
      "============\n",
      "are\n",
      "directly correlated, allows us to consider control mechanisms ofthe chromosomal network that derive the network geometryfrom an initial configuration toward a specific cell type. This isparticularly exemplified by transcription factors that have broad\n",
      "influence on cell fate, such as MyoD. In addition to its role inmyoblasts as a transcription factor regulating expression of\n",
      "skeletal muscle genes, MyoD can convert fibroblasts to skeletalmuscle cells by activating the skeletal muscle differentiationprogram. Recent studies indicate that MyoD binds and induces\n",
      "histone modifications at tens of thousands of sites in the myoblast\n",
      "prior to transcription of most skeletal muscle genes, suggesting apotentially more global role in cell specification (10). Another“master regulator ”is GATA-1, a zinc finger transcription factor\n",
      "essential to maintenance of the erythroid and megakaryocytelineages (11). GATA-1 may have a global impact on nuclearorganization by catalyzing interactions within and\n",
      "Summary: MyoD is a transcription factor that plays a significant role in cell specification by converting fibroblasts into skeletal muscle cells. It does this by activating the skeletal muscle differentiation program and inducing histone modifications at tens of thousands of sites in the myoblast before the transcription of most skeletal muscle genes. This suggests a more global role in cell specification.\n",
      "============\n",
      "that read upstream tem- \n",
      "poral and positional cues [see below]), for defining the \n",
      "myoblast state and positioning the proper number of cells \n",
      "in the muscle-forming regions of the body, and for receiv- \n",
      "ing inhibitory signals from the environment; that MyoD \n",
      "may be primarily responsible for stabilizing the determined \n",
      "state via autoactivation; and that myogenin, once acti- \n",
      "vated by MyoD or Myfd, is used for bypassing negative \n",
      "control in vivo and for actually activating most (but not all) \n",
      "downstream muscle structural genes. MRM, which has many features in common with myogenin (e.g., late onset \n",
      "of expression), may also have a partially overlapping func- \n",
      "tion with myogenin, which could contribute to the residual \n",
      "myogenesis present in myogenin-deficient mice. Under \n",
      "certain conditions (e.g., overexpression or growth in vitro \n",
      "versus in vivo), these distinctions can become blurred. \n",
      "A major problem in postulating totally redundant func- \n",
      "tions for MyoD and Myf-5 is that over\n",
      "Summary: MYOD is involved in defining muscle cell state and positioning muscles in the body through receiving cues and giving inhibitory signals. It plays a role in stabilizing the muscle cell state through autoactivation. Myogenin, which is activated by MYOD or Myf-d, drives the activation of most muscle structural genes. MRM, which shares features with myogenin, may have a partial overlap in function with myogenin. The distinctions between the functions of MYOD and Myf-5 can blur under certain conditions. However, it is important to note that MYOD and Myf-5 do not have completely redundant functions.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[(Document(page_content='MYOD is involved in initiating the myogenic differentiation program, which temporally regulates its own activity through a feed-forward mechanism.', metadata={'page': 4, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/The circuitry of a master switch.pdf'}),\n",
       "  0.7678906917572021),\n",
       " (Document(page_content='MYOD is a transcription factor involved in muscle differentiation as indicated by its role in gene transcription and regulation of myoblast differentiation both in vitro and in vivo (Kopan et al., 1994; Nofziger et al., 1999). Despite known inhibitors of differentiation like mitogens and Notch signaling, the precise mechanisms allowing differentiation to occur at specific times and places remain unclear. As for the execution of an entire program of cell differentation by a single transcription factor, research has shown that expression levels of many RNAs change during skeletal muscle differentiation in cultured C2C12 cells. To investigate the role of MYOD in these changes, gene expression was assessed in fibroblasts.', metadata={'page': 4, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/The circuitry of a master switch.pdf'}),\n",
       "  0.7660456895828247),\n",
       " (Document(page_content='MyoD is a key transcription factor involved in muscle development. It regulates the expression of myogenin and other muscle-specific genes during both in vitro (tissue culture) and in vivo (living organism) processes. MyoD can also activate its own transcription and that of myogenin, while myogenin can activate MyoD and itself. The activation of myogenin is mediated in part by a muscle transcription factor called MEF2.', metadata={'page': 0, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/The MyoD Family and Myogenesis Redundancy, Networks, and Thresholds.pdf'}),\n",
       "  0.7584931254386902),\n",
       " (Document(page_content='MYOD is a member of the MyoD family of myogenic transcriptional regulators, including Myf-5, MRF4, and myogenin. When expressed from a constitutive promoter, MYOD can convert various cell types into muscle. However, in mice with homogenous gene-targeted mutations for both MYOD and Myf-5, normal muscle amounts are produced, which was previously considered a paradox. This paradox has been resolved by recent studies that show in double homogenous mutants, both MyoD and Myf-5 are required for full muscle conversion and development.', metadata={'page': 0, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/The MyoD Family and Myogenesis Redundancy, Networks, and Thresholds.pdf'}),\n",
       "  0.7571537494659424),\n",
       " (Document(page_content='MYOD is a transcription factor that regulates skeletal muscle cell differentiation by binding to the regulatory regions of genes expressed in muscle cells. It forms heterodimers with E-proteins through the interaction of their HLH domains, and recognizes DNA binding sites with the E-box consensus sequence (CANNTG). MYOD has an acidic-activation domain, while E-proteins have a more complex mix of activation and repression domains. (summary: MYOD is a muscle-specific transcription factor that regulates muscle gene expression by binding to the E-box sequence in partnership with E-proteins, using its acidic-activation domain to regulate gene expression.)', metadata={'page': 3, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/The circuitry of a master switch.pdf'}),\n",
       "  0.7549338340759277),\n",
       " (Document(page_content=\"Myod is involved in regulating gene expression during muscle cell differentiation. In the replicating muscle precursor cell (myoblast), Myod associates with histone deacetylases (HDACs) and suppresses gene expression. However, during muscle differentiation, Myod switches to associating with histone acetyltransferases (HATs) and activates gene expression. This switch is important for initiating muscle differentiation, and it can be induced in cell lines such as C2C12 myoblasts by removing serum or adding growth factors like Fgf or Tgf β. Notch signaling also represses Myod's transcription and activity.\", metadata={'page': 4, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/The circuitry of a master switch.pdf'}),\n",
       "  0.7548009157180786),\n",
       " (Document(page_content='MYOD, identified by Lassar et al. (1986), is a bHLH transcription factor that triggers skeletal muscle differentiation. Its discovery led to investigating the roles of muscle-specific transcription factors and chromatin-associated proteins in cell differentiation. Furthermore, it raised the question of how a single transcription factor manages an entire program of cell differentiation. This review focuses on how MYOD orchestrates the muscle gene expression process.', metadata={'page': 0, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/The circuitry of a master switch.pdf'}),\n",
       "  0.7541691064834595),\n",
       " (Document(page_content='MYOD is a gene that encodes a transcriptional activator involved in muscle differentiation. When introduced into non-muscle cell types, it promotes myogenesis by binding to a control region shared by muscle-specific genes and activating its own transcription, creating an autoregulatory loop that strengthens the commitment to muscle cell development.', metadata={'page': 0, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/Harold Weintraub.pdf'}),\n",
       "  0.7537667751312256),\n",
       " (Document(page_content='MyoD is a transcription factor that plays a significant role in cell specification by converting fibroblasts into skeletal muscle cells. It does this by activating the skeletal muscle differentiation program and inducing histone modifications at tens of thousands of sites in the myoblast before the transcription of most skeletal muscle genes. This suggests a more global role in cell specification.', metadata={'page': 2, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/Dynamics and control of state-dependent networks for probing genomic organization.pdf'}),\n",
       "  0.752388060092926),\n",
       " (Document(page_content='MYOD is involved in defining muscle cell state and positioning muscles in the body through receiving cues and giving inhibitory signals. It plays a role in stabilizing the muscle cell state through autoactivation. Myogenin, which is activated by MYOD or Myf-d, drives the activation of most muscle structural genes. MRM, which shares features with myogenin, may have a partial overlap in function with myogenin. The distinctions between the functions of MYOD and Myf-5 can blur under certain conditions. However, it is important to note that MYOD and Myf-5 do not have completely redundant functions.', metadata={'page': 2, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/The MyoD Family and Myogenesis Redundancy, Networks, and Thresholds.pdf'}),\n",
       "  0.7507756948471069)]"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def contextualCompression(documentSearch, chatstatus):\n",
    "    \"\"\"\n",
    "    Summarizes the content of documents based on a user query, updating the \n",
    "    document search results with these summaries.\n",
    "\n",
    "    Args:\n",
    "        documentSearch (list): A list of documents where each document is a tuple, \n",
    "                               and the first element of the tuple has an attribute \n",
    "                               `page_content` containing the text content of the document.\n",
    "        chatstatus (dict): BRAD chatstatus used to track debuging\n",
    "\n",
    "    Returns:\n",
    "        list: The modified `documentSearch` list with updated `page_content` for each \n",
    "              document, replaced by their summaries.\n",
    "\n",
    "    Example:\n",
    "        documentSearch = [(Document(page_content=\"...\"),), ...]\n",
    "        chatstatus = {'config': {'debug': True}}\n",
    "        updatedDocs = contextualCompression(documentSearch, chatstatus)\n",
    "    \"\"\"\n",
    "    template = summarizeDocumentTemplate()\n",
    "    PROMPT = PromptTemplate(input_variables=[\"user_query\"], template=template)\n",
    "    reducedDocs = []\n",
    "    for i, doc in enumerate(documentSearch):\n",
    "        pageContent = doc[0].page_content\n",
    "        prompt = PROMPT.format(text=pageContent, user_query=query)\n",
    "        res = llm.invoke(input=prompt)\n",
    "        summary = res.content.strip()\n",
    "        if chatstatus['config']['debug']:\n",
    "            print('============')\n",
    "            print(pageContent)\n",
    "            print('Summary: ' + summary)\n",
    "        doc[0].page_content = summary\n",
    "        documentSearch[i] = doc\n",
    "    return documentSearch\n",
    "\n",
    "contextualCompression(documentSearch, chatstatus)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "663ae67b-0223-4d86-ae59-de8603049793",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[((Document(page_content='MYOD is involved in initiating the myogenic differentiation program, which temporally regulates its own activity through a feed-forward mechanism.', metadata={'page': 4, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/The circuitry of a master switch.pdf'}),\n",
       "   0.7678906917572021),\n",
       "  0.7678906917572021),\n",
       " ((Document(page_content='MYOD is a transcription factor involved in muscle differentiation as indicated by its role in gene transcription and regulation of myoblast differentiation both in vitro and in vivo (Kopan et al., 1994; Nofziger et al., 1999). Despite known inhibitors of differentiation like mitogens and Notch signaling, the precise mechanisms allowing differentiation to occur at specific times and places remain unclear. As for the execution of an entire program of cell differentation by a single transcription factor, research has shown that expression levels of many RNAs change during skeletal muscle differentiation in cultured C2C12 cells. To investigate the role of MYOD in these changes, gene expression was assessed in fibroblasts.', metadata={'page': 4, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/The circuitry of a master switch.pdf'}),\n",
       "   0.7660456895828247),\n",
       "  0.7660456895828247),\n",
       " ((Document(page_content='MyoD is a key transcription factor involved in muscle development. It regulates the expression of myogenin and other muscle-specific genes during both in vitro (tissue culture) and in vivo (living organism) processes. MyoD can also activate its own transcription and that of myogenin, while myogenin can activate MyoD and itself. The activation of myogenin is mediated in part by a muscle transcription factor called MEF2.', metadata={'page': 0, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/The MyoD Family and Myogenesis Redundancy, Networks, and Thresholds.pdf'}),\n",
       "   0.7584931254386902),\n",
       "  0.7584931254386902),\n",
       " ((Document(page_content='MYOD is a member of the MyoD family of myogenic transcriptional regulators, including Myf-5, MRF4, and myogenin. When expressed from a constitutive promoter, MYOD can convert various cell types into muscle. However, in mice with homogenous gene-targeted mutations for both MYOD and Myf-5, normal muscle amounts are produced, which was previously considered a paradox. This paradox has been resolved by recent studies that show in double homogenous mutants, both MyoD and Myf-5 are required for full muscle conversion and development.', metadata={'page': 0, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/The MyoD Family and Myogenesis Redundancy, Networks, and Thresholds.pdf'}),\n",
       "   0.7571537494659424),\n",
       "  0.7571537494659424),\n",
       " ((Document(page_content='MYOD is a transcription factor that regulates skeletal muscle cell differentiation by binding to the regulatory regions of genes expressed in muscle cells. It forms heterodimers with E-proteins through the interaction of their HLH domains, and recognizes DNA binding sites with the E-box consensus sequence (CANNTG). MYOD has an acidic-activation domain, while E-proteins have a more complex mix of activation and repression domains. (summary: MYOD is a muscle-specific transcription factor that regulates muscle gene expression by binding to the E-box sequence in partnership with E-proteins, using its acidic-activation domain to regulate gene expression.)', metadata={'page': 3, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/The circuitry of a master switch.pdf'}),\n",
       "   0.7549338340759277),\n",
       "  0.7549338340759277),\n",
       " ((Document(page_content=\"Myod is involved in regulating gene expression during muscle cell differentiation. In the replicating muscle precursor cell (myoblast), Myod associates with histone deacetylases (HDACs) and suppresses gene expression. However, during muscle differentiation, Myod switches to associating with histone acetyltransferases (HATs) and activates gene expression. This switch is important for initiating muscle differentiation, and it can be induced in cell lines such as C2C12 myoblasts by removing serum or adding growth factors like Fgf or Tgf β. Notch signaling also represses Myod's transcription and activity.\", metadata={'page': 4, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/The circuitry of a master switch.pdf'}),\n",
       "   0.7548009157180786),\n",
       "  0.7548009157180786),\n",
       " ((Document(page_content='MYOD, identified by Lassar et al. (1986), is a bHLH transcription factor that triggers skeletal muscle differentiation. Its discovery led to investigating the roles of muscle-specific transcription factors and chromatin-associated proteins in cell differentiation. Furthermore, it raised the question of how a single transcription factor manages an entire program of cell differentiation. This review focuses on how MYOD orchestrates the muscle gene expression process.', metadata={'page': 0, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/The circuitry of a master switch.pdf'}),\n",
       "   0.7541691064834595),\n",
       "  0.7541691064834595),\n",
       " ((Document(page_content='MYOD is a gene that encodes a transcriptional activator involved in muscle differentiation. When introduced into non-muscle cell types, it promotes myogenesis by binding to a control region shared by muscle-specific genes and activating its own transcription, creating an autoregulatory loop that strengthens the commitment to muscle cell development.', metadata={'page': 0, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/Harold Weintraub.pdf'}),\n",
       "   0.7537667751312256),\n",
       "  0.7537667751312256),\n",
       " ((Document(page_content='MyoD is a transcription factor that plays a significant role in cell specification by converting fibroblasts into skeletal muscle cells. It does this by activating the skeletal muscle differentiation program and inducing histone modifications at tens of thousands of sites in the myoblast before the transcription of most skeletal muscle genes. This suggests a more global role in cell specification.', metadata={'page': 2, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/Dynamics and control of state-dependent networks for probing genomic organization.pdf'}),\n",
       "   0.752388060092926),\n",
       "  0.752388060092926),\n",
       " ((Document(page_content='MYOD is involved in defining muscle cell state and positioning muscles in the body through receiving cues and giving inhibitory signals. It plays a role in stabilizing the muscle cell state through autoactivation. Myogenin, which is activated by MYOD or Myf-d, drives the activation of most muscle structural genes. MRM, which shares features with myogenin, may have a partial overlap in function with myogenin. The distinctions between the functions of MYOD and Myf-5 can blur under certain conditions. However, it is important to note that MYOD and Myf-5 do not have completely redundant functions.', metadata={'page': 2, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/The MyoD Family and Myogenesis Redundancy, Networks, and Thresholds.pdf'}),\n",
       "   0.7507756948471069),\n",
       "  0.7507756948471069)]"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "documentSearch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "b768a47f-cdc6-42e6-a477-94caba610948",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(Document(page_content='a combination of promoter-speciﬁc regulation of Myod binding and activity.\\nBecause Myod initiates the myogenic differentiation\\nprogram and that program temporally regulates the activity of\\nMyod, it follows that Myod programs the regulation of its ownactivity. It does this, at least in part, through a feed-forward\\nDevelopment', metadata={'page': 4, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/The circuitry of a master switch.pdf'}),\n",
       "  0.7678906917572021),\n",
       " (Document(page_content='transcription and Myod protein\\nactivity (Kopan et al., 1994; Nofziger et al., 1999), andprobably contributes to regulating differentiation in vivo. It isinteresting that while we have identiﬁed several mechanismsthat might delay myoblast differentiation, such as mitogens andNotch signaling, we do not yet have a good understanding ofthe events that occur in vivo to overcome these inhibitorysignals and to induce differentiation at a speciﬁc time andplace.\\nA feed-forward circuit as a quantal step\\nHow does a single transcription factor execute an entireprogram of cell differentiation? How does a single event result\\nin a predictable and complex response? Microarray expressionstudies of cultured C2C12 cells have shown that expressionlevels of many RNAs change during skeletal muscledifferentiation (Delgado et al., 2003; Tomczak et al., 2004). To\\ndetermine how many of these changes are caused by theexpression and activity of Myod, we assessed gene expressionchanges in ﬁbroblasts with an', metadata={'page': 4, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/The circuitry of a master switch.pdf'}),\n",
       "  0.7660456895828247),\n",
       " (Document(page_content='compartments. Thus, myogenin-independent myoblast \\nidentity seems to be definable by the generation of approx- \\nimately the right number of cells in the right positions of \\nthe body under the direction of MyoD-Myf-5 function. \\nRegulation of Myogenic Genes In Vitro and In Vivo \\nTissue culture experiments also support the notion that Minireview \\nmyogenin acts downstream of MyoD and Myf-5 since most \\nmyoblasts in culture express MyoD or Myf-5, but turn on \\nmyogenin only when induced to differentiate. Cell culture \\nexperiments have also contributed several additional lay- \\ners to the regulatory circuitry, showing that MyoD can acti- \\nvate its own transcription as well as that of myogenin; \\nmyogenin can activate itself and as well as MyoD; and \\nMyfd can activate itself, MyoD, and myogenin. Activation \\nof myogenin seems to be mediated, in part, via a down- \\nstream muscle transcription factor, MEF2 (Edmondson et \\nal., 1992), so that in this case what is downstream and \\nwhat is upstream may', metadata={'page': 0, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/The MyoD Family and Myogenesis Redundancy, Networks, and Thresholds.pdf'}),\n",
       "  0.7584931254386902),\n",
       " (Document(page_content='Cell, Vol. 75, 1241-1244, December 31, 1993, Copyright @ 1993 by Cell Press \\nThe MyoD Family and Myogenesis: \\nRedundancy, Networks, \\nand Thresholds \\nHarold Weintraub \\nFred Hutchinson Cancer Research Center \\nHoward Hughes Medical Institute Laboratory \\n1124 Columbia Street \\nSeattle, Washington 98104 \\nThe members of the MyoD family of myogenic transcrip- \\ntional regulators (MyoD, Myf-5, MRF4, myogenin) have \\nthe remarkable property that, when expressed from a con- \\nstitutive promoter, they can convert a large number of dif- \\nferent cell types into muscle (reviewed by Buckingham, \\n1992; Emerson, 1990; Olson, 1990; Weintraub et al., \\n1991). In contrast, mice with homozygous gene-targeted \\nmutations in either MyoD or Myf-5 produce fairly normal \\namounts of muscle (Rudnicki et al., 1992; Braun et al., \\n1992). This apparent paradox has now been resolved by \\ntwo recent studies. In one, a simple genetic cross has \\nshown that in the double homozygous mutants, in which \\nboth MyoD and Myfd are', metadata={'page': 0, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/The MyoD Family and Myogenesis Redundancy, Networks, and Thresholds.pdf'}),\n",
       "  0.7571537494659424),\n",
       " (Document(page_content='expressed, how does Myod regulate skeletal muscle cell\\ndifferentiation? In one sense, the answer seems fairly simple:Myod is a transcription factor with binding sites in theregulatory regions of many genes that are expressed in skeletalmuscle. Myod forms heterodimers with the nearly ubiquitousE-protein sub-family of bHLH proteins through the interactionof the HLH domains (see Fig. 1) (Lassar et al., 1991; Murre etal., 1989). The basic regions act as sequence-speciﬁc DNA-binding domains that recognize a binding site with the simplecore consensus sequence of CANNTG, termed an E-box, andshow additional preferences for internal and ﬂankingsequences (Blackwell and Weintraub, 1990). Myod has a singleamino-terminal acidic-activation domain, as determined by its\\nfusion to the heterologous DNA-binding domain of the Gal4\\nprotein (Weintraub et al., 1991b), whereas E-proteins have amore complex mix of activation and repression domains(Markus et al., 2002) (see Fig. 1). Therefore, the simple', metadata={'page': 3, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/The circuitry of a master switch.pdf'}),\n",
       "  0.7549338340759277),\n",
       " (Document(page_content='Harter,2003). In contrast to the differentiating muscle cell, whereMyod is associated with HATs, in the myoblast (the replicatingmuscle precursor cell), Myod is associated with histonedeacetylases (HDACs) and might actively suppress geneexpression (Fulco et al., 2003; Mal and Harter, 2003; Mal etal., 2001).\\nThese ﬁndings strongly suggest that Myod acts to negatively\\nregulate the transcription of some genes in the myoblast andthat muscle differentiation is initiated when Myod switchesfrom its association with repressive factors to activatingfactors. The differentiation of Myod- expressing cell lines, such\\nas the C2C12 myoblast cell line (Silberstein et al., 1986), canbe induced in culture by removing serum or it can be preventedby adding growth factors, such as Fgf or Tgf β(De Angelis et\\nal., 1998; Li et al., 1992), indicating that mitogen stimulationin vivo might sustain a myoblast state. In addition, Notchsignaling represses both Myod transcription and Myod protein\\nactivity (Kopan et', metadata={'page': 4, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/The circuitry of a master switch.pdf'}),\n",
       "  0.7548009157180786),\n",
       " (Document(page_content='identiﬁcation of Myod1 (Lassar et al.,1986), a basic-helix-loop-helix (bHLH) transcription factorthat can induce skeletal muscle differentiation in cells frommany different lineages.The identiﬁcation of Myod as a master regulatory gene of\\nskeletal muscle differentiation provided the opportunity to\\nassess the roles of tissue-speciﬁc transcription factors and\\nchromatin-associated proteins in regulating cell differentiation.\\nIt also provided the opportunity to address another importantquestion: how does a single transcription factor execute anentire program of cell differentiation? One reason to beinterested in the answer to this question is that it is a speciﬁc\\nexample of the broader question: how does a single event result\\nin a predictable, complex response? It is likely thatunderstanding the orchestration of skeletal muscle geneexpression by Myod will provide insights into the regulationof other complex biological events.\\nThis review describes generally how Myod regulates the\\nprogram', metadata={'page': 0, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/The circuitry of a master switch.pdf'}),\n",
       "  0.7541691064834595),\n",
       " (Document(page_content=\"of mRNA is expressed in cells or embryos to inhibit the \\nexpression of specific mRNAs, allows the generation of lack-of-function phenotypes and permits the assessment \\nof the role of specific genes in development in the absence \\nof classical developmental genetics. \\nMore recently, in a remarkable series of experiments, \\nHal identified a single gene, myoD, which can elicit the \\nentire prog'ram of muscle differentiation when introduced \\ninto nonmuscle cell types. He demonstrated that muscle \\ncell differentiation could be elicited in fibroblasts following \\nthe transfer of genomic DNA from muscle cells. This led \\nHal to the isolation of the myoD gene, which encodes a \\ntranscriptional activator that promotes myogenesis by di- \\nrectly binding to a control region shared by muscle-specific \\ngenes. Moreover, MyoD activates its own transcription, \\ncreating an autoregulatory loop that stabilizes the commit- \\nment to myogenesis. Structure-function analysis allowed \\nthe definition of a complex\", metadata={'page': 0, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/Harold Weintraub.pdf'}),\n",
       "  0.7537667751312256),\n",
       " (Document(page_content='are\\ndirectly correlated, allows us to consider control mechanisms ofthe chromosomal network that derive the network geometryfrom an initial configuration toward a specific cell type. This isparticularly exemplified by transcription factors that have broad\\ninfluence on cell fate, such as MyoD. In addition to its role inmyoblasts as a transcription factor regulating expression of\\nskeletal muscle genes, MyoD can convert fibroblasts to skeletalmuscle cells by activating the skeletal muscle differentiationprogram. Recent studies indicate that MyoD binds and induces\\nhistone modifications at tens of thousands of sites in the myoblast\\nprior to transcription of most skeletal muscle genes, suggesting apotentially more global role in cell specification (10). Another“master regulator ”is GATA-1, a zinc finger transcription factor\\nessential to maintenance of the erythroid and megakaryocytelineages (11). GATA-1 may have a global impact on nuclearorganization by catalyzing interactions within and', metadata={'page': 2, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/Dynamics and control of state-dependent networks for probing genomic organization.pdf'}),\n",
       "  0.752388060092926),\n",
       " (Document(page_content='that read upstream tem- \\nporal and positional cues [see below]), for defining the \\nmyoblast state and positioning the proper number of cells \\nin the muscle-forming regions of the body, and for receiv- \\ning inhibitory signals from the environment; that MyoD \\nmay be primarily responsible for stabilizing the determined \\nstate via autoactivation; and that myogenin, once acti- \\nvated by MyoD or Myfd, is used for bypassing negative \\ncontrol in vivo and for actually activating most (but not all) \\ndownstream muscle structural genes. MRM, which has many features in common with myogenin (e.g., late onset \\nof expression), may also have a partially overlapping func- \\ntion with myogenin, which could contribute to the residual \\nmyogenesis present in myogenin-deficient mice. Under \\ncertain conditions (e.g., overexpression or growth in vitro \\nversus in vivo), these distinctions can become blurred. \\nA major problem in postulating totally redundant func- \\ntions for MyoD and Myf-5 is that over', metadata={'page': 2, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/The MyoD Family and Myogenesis Redundancy, Networks, and Thresholds.pdf'}),\n",
       "  0.7507756948471069)]"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "documentSearch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "c54c8c25-33b4-4933-8f78-b5c7dbef780c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Myod is involved in initiating the myogenic differentiation program, which regulates its own activity.',\n",
       " 'MYOD is involved in the process of cell differentiation in muscular tissues, specifically during skeletal muscle differentiation. It does so by regulating gene expression, as evidenced by changes in expression levels of many RNAs observed in microarray studies.',\n",
       " 'MYOD is involved in the regulation of myogenin expression during muscle cell differentiation.',\n",
       " 'MYOD is involved in the process of muscle conversion in cells, as expressed from a constitutive promoter, it can transform different cell types into muscle. However, homOzygous gene-targeted mutants of MYOD or Myf-5 produce normal amounts of muscle in mice. The recent studies resolved this paradox by showing that both MyoD and Myf-5 are required in the double homOzygous mutants for proper muscle development.',\n",
       " 'Myod is a transcription factor involved in skeletal muscle cell differentiation. It forms heterodimers with E-proteins through the interaction of their HLH domains and binds to E-boxes to regulate the expression of related genes.',\n",
       " 'In the myoblast cell, MYOD is associated with histone deacetylases (HDACs) and might suppress gene expression through this association (Fulco et al., 2003; Mal and Harter, 2003; Mal et al., 2001). This suggests that MYOD acts as a negative regulator of gene transcription in the myoblast, and muscle differentiation is initiated when MYOD switches from its association with repressive factors to activating factors (Silberstein et al., 1986; De Angelis et al., 1998; Li et al., 1992). Therefore, MYOD is involved in regulating the transcription of genes during myoblast development and differentiation.',\n",
       " 'MYOD is a master regulatory gene involved in skeletal muscle differentiation. It regulates this process by orchestrating the expression of skeletal muscle genes. (Summary)',\n",
       " 'MyoD is a transcriptional activator involved in muscle differentiation. It directly binds to the control region of muscle-specific genes, promoting myogenesis by activating its own transcription, creating an autoregulatory loop.',\n",
       " 'MyoD is involved in the process of converting fibroblasts to skeletal muscle cells, known as skeletal muscle differentiation. It does this by activating the skeletal muscle differentiation program and binding to and inducing histone modifications at tens of thousands of sites in the myoblast before the transcription of most skeletal muscle genes.',\n",
       " 'MYOD is involved in defining the myoblast state and positioning cells in muscle-forming regions of the body by activating the downstream muscle structural genes, primarily myogenin.']"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reducedText"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b43fc7d5-4ca2-41c0-b787-5ec3d33bb012",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "0d2c88f0-d276-4baf-b11e-691b80a80dcb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MyoD is involved in defining the myoblast state, positioning cells in muscle-forming regions, and receiving inhibitory signals from the environment. It primarily stabilizes the determined state via autoactivation. Myogenin, which is activated by MyoD, is used for actual activation of most muscle structural genes. MRM, which shares features with myogenin, may have a partially overlapping function with myogenin. The distinctions between their functions can blur under certain conditions.'"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res.content.strip()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "8e276351-1f80-4fb9-9d1b-4dfc52416423",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ChatMessage(content=\" I'm just a computer program, so I don't have the ability to feel emotions like a human does. I'm here to help answer any questions you have to the best of my ability. Is there a specific topic you'd like to know more about?\", response_metadata={'role': 'assistant', 'content': \" I'm just a computer program, so I don't have the ability to feel emotions like a human does. I'm here to help answer any questions you have to the best of my ability. Is there a specific topic you'd like to know more about?\", 'token_usage': {'prompt_tokens': 14, 'total_tokens': 70, 'completion_tokens': 56}, 'model_name': 'mistralai/mistral-7b-instruct-v0.2'}, id='run-41ae1e56-16c0-42e8-a37e-e87909df9eb5-0', role='assistant')"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prompt = PROMPT.format(text='YY', user_query='XX')\n",
    "llm.invoke(input=prompt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "289ec2a3-52e7-40b8-ace5-d38f0a536ed1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "**INSTRUCTIONS**\n",
      "You are an assistant responsible for compressing the important information in a document.\n",
      "You will be given a users query and a piece of text. Summarize the text to contain only the information\n",
      "relevant to answering the users question. If no information is in the text is related, return None in the\n",
      "summary section.\n",
      "\n",
      "**USER QUERY**\n",
      "XX\n",
      "\n",
      "**TEXT**\n",
      "YY\n",
      "\n",
      "**OUTPUT**\n",
      "Summary: <output here>\n",
      "\n"
     ]
    }
   ],
   "source": [
    "prompt = PROMPT.format(text='YY', user_query='XX')\n",
    "print(prompt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e749d290-490f-4168-b653-6a563de46df5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "8bd3f090-ef37-43ef-a689-0a7adf0ee19c",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'tuple' object has no attribute 'page_content'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[32], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mdoc\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpage_content\u001b[49m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'tuple' object has no attribute 'page_content'"
     ]
    }
   ],
   "source": [
    "doc.page_content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "31785321-0bff-41d2-ae7d-e6a7c9cb55f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ChatNVIDIA(model='mistralai/mistral-7b-instruct-v0.2')"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "llm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6626ec50-d68a-4edd-b8b7-49f0b78a3a45",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68e1d696-0e5c-47bc-867b-f88b96774a70",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "83fc58e0-916c-4a8a-a291-b8a1532fd352",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'pretty_print_docs' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[8], line 8\u001b[0m\n\u001b[1;32m      5\u001b[0m compression_retriever \u001b[38;5;241m=\u001b[39m ContextualCompressionRetriever(base_compressor\u001b[38;5;241m=\u001b[39mcompressor, base_retriever\u001b[38;5;241m=\u001b[39mretriever)\n\u001b[1;32m      7\u001b[0m compressed_docs \u001b[38;5;241m=\u001b[39m compression_retriever\u001b[38;5;241m.\u001b[39mget_relevant_documents(query\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mWhat cellular processes is MYOD involved in?\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m----> 8\u001b[0m \u001b[43mpretty_print_docs\u001b[49m(compressed_docs)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'pretty_print_docs' is not defined"
     ]
    }
   ],
   "source": [
    "from langchain.retrievers import ContextualCompressionRetriever\n",
    "from langchain.retrievers.document_compressors import LLMChainExtractor\n",
    "retriever = vectordb.as_retriever()\n",
    "compressor = LLMChainExtractor.from_llm(llm)\n",
    "compression_retriever = ContextualCompressionRetriever(base_compressor=compressor, base_retriever=retriever)\n",
    "\n",
    "compressed_docs = compression_retriever.get_relevant_documents(query='What cellular processes is MYOD involved in?')\n",
    "pretty_print_docs(compressed_docs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "23255eff-ef3a-46ed-82a3-310638e46c49",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Document 1:\n",
      "Myod initiates the myogenic differentiation program and is involved in regulating its own activity through a feed-forward mechanism. (Context: a combination of promoter-specific regulation of Myod binding and activity. Because Myod initiates the myogenic differentiation program and that program temporally regulates the activity of Myod, it follows that Myod programs the regulation of its own activity. It does this, at least in part, through a feed-forward mechanism.)\n",
      "----------------------------------------------------------------------------------------------------\n",
      "Document 2:\n",
      "transcription and Myod protein activity (Kopan et al., 1994; Nofziger et al., 1999)\n",
      "Myod expression and activity leads to changes in gene expression during skeletal muscle differentiation (Delgado et al., 2003; Tomczak et al., 2004)\n",
      "----------------------------------------------------------------------------------------------------\n",
      "Document 3:\n",
      "MyoD is involved in the regulation of myogenin as well as its own transcription. Myogenin can also activate MyoD and itself. MEF2 is a downstream muscle transcription factor involved in the activation of myogenin.\n",
      "----------------------------------------------------------------------------------------------------\n",
      "Document 4:\n",
      "The members of the MyoD family of myogenic transcriptional regulators, including MyoD, have the property of converting different cell types into muscle when expressed from a constitutive promoter (Weintraub, 1993). MyoD and Myf-5 are important in myogenesis, but mice with homogenous gene-targeted mutations in either MyoD or Myf-5 produce fairly normal amounts of muscle (Rudnicki et al., 1992; Braun et al., 1992). The paradox of the large role of MyoD in converting cells into muscle but normal muscle production in knockout mice was recently resolved by showing that in the double homoygous mutants, both MyoD and Myfd are involved in the process.\n",
      "\n",
      "Relevant context: \"The members of the MyoD family of myogenic transcriptional regulators... have the property of converting different cell types into muscle... In contrast, mice with homoezygous gene-targeted mutations in either MyoD or Myf-5 produce fairly normal amounts of muscle... This paradox has now been resolved... in the double homoezygous mutants, both MyoD and Myfd are involved in the process.\"\n"
     ]
    }
   ],
   "source": [
    "pretty_print_docs(compressed_docs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "90b81210-bd25-4584-b3db-3d111f4c148c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[Document(page_content='Myod initiates the myogenic differentiation program and is involved in regulating its own activity through a feed-forward mechanism. (Context: a combination of promoter-specific regulation of Myod binding and activity. Because Myod initiates the myogenic differentiation program and that program temporally regulates the activity of Myod, it follows that Myod programs the regulation of its own activity. It does this, at least in part, through a feed-forward mechanism.)', metadata={'page': 4, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/The circuitry of a master switch.pdf'}),\n",
       " Document(page_content='transcription and Myod protein activity (Kopan et al., 1994; Nofziger et al., 1999)\\nMyod expression and activity leads to changes in gene expression during skeletal muscle differentiation (Delgado et al., 2003; Tomczak et al., 2004)', metadata={'page': 4, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/The circuitry of a master switch.pdf'}),\n",
       " Document(page_content='MyoD is involved in the regulation of myogenin as well as its own transcription. Myogenin can also activate MyoD and itself. MEF2 is a downstream muscle transcription factor involved in the activation of myogenin.', metadata={'page': 0, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/The MyoD Family and Myogenesis Redundancy, Networks, and Thresholds.pdf'}),\n",
       " Document(page_content='The members of the MyoD family of myogenic transcriptional regulators, including MyoD, have the property of converting different cell types into muscle when expressed from a constitutive promoter (Weintraub, 1993). MyoD and Myf-5 are important in myogenesis, but mice with homogenous gene-targeted mutations in either MyoD or Myf-5 produce fairly normal amounts of muscle (Rudnicki et al., 1992; Braun et al., 1992). The paradox of the large role of MyoD in converting cells into muscle but normal muscle production in knockout mice was recently resolved by showing that in the double homoygous mutants, both MyoD and Myfd are involved in the process.\\n\\nRelevant context: \"The members of the MyoD family of myogenic transcriptional regulators... have the property of converting different cell types into muscle... In contrast, mice with homoezygous gene-targeted mutations in either MyoD or Myf-5 produce fairly normal amounts of muscle... This paradox has now been resolved... in the double homoezygous mutants, both MyoD and Myfd are involved in the process.\"', metadata={'page': 0, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/The MyoD Family and Myogenesis Redundancy, Networks, and Thresholds.pdf'})]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compressed_docs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8857087f-7ae5-45c3-86ca-efd5be95abca",
   "metadata": {},
   "outputs": [],
   "source": [
    "def pretty_print_docs(docs):\n",
    "    print(f\"\\n{'-' * 100}\\n\".join([f\"Document {i+1}:\\n\" + d.page_content for i, d in enumerate(docs)]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8f8eb561-a653-43f0-b03c-eb0dc8f52831",
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "from langchain.retrievers.multi_query import MultiQueryRetriever\n",
    "\n",
    "logging.basicConfig()\n",
    "logging.getLogger(\"langchain.retrievers.multi_query\").setLevel(logging.INFO)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "35a9f53e-2ae4-4744-90e0-06f4ee76268a",
   "metadata": {},
   "outputs": [],
   "source": [
    "retriever = MultiQueryRetriever.from_llm(\n",
    "    retriever=vectordb.as_retriever(), llm=llm\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "081b72ef-5c84-498c-a1c6-438412380964",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'MultiQueryRetriever' object has no attribute 'get_relevant_documents_with_scores'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[20], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m docs \u001b[38;5;241m=\u001b[39m \u001b[43mretriever\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_relevant_documents_with_scores\u001b[49m(query\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mWhat genes are related to MYOD?\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'MultiQueryRetriever' object has no attribute 'get_relevant_documents_with_scores'"
     ]
    }
   ],
   "source": [
    "docs = retriever.get_relevant_documents_with_scores(query='What genes are related to MYOD?')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "d033246d-8769-432a-8cda-ce606c4113d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[Document(page_content='There is a positive correlation between regulated MyoD\\nbinding (i.e., sites that are preferentially bound in myoblasts or\\ndifferentiated myotubes) and gene expression. This suggests\\nthat sites regulating gene transcription in myotubes require addi-tional factors to modulate MyoD binding. Indeed, the sites asso-\\nciated with myotube-expressed genes are enriched for motifs of\\nfactors that are activated by MyoD and function with MyoD ina positive feed-forward circuit, as demonstrated previouslywith Mef2 ( Penn et al., 2004 ) and Myog ( Cao et al., 2006 ). In addi-\\ntion, the Pbx/Meis complex cooperates with MyoD in activatinga subset of genes ( Berkes et al., 2004; Maves et al., 2007 ).\\nAnother ﬁnding was that genes decreasing expression with\\ndifferentiation were associated with decreased MyoD binding\\nat sites enriched for RP58 and AP1 motifs. A recent study (pub-lished after our analysis was complete) identiﬁed RP58 as a gene\\nactivated by MyoD during muscle differentiation that', metadata={'page': 9, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/Genome-wide MyoD Binding.pdf'}),\n",
       " Document(page_content=\"of mRNA is expressed in cells or embryos to inhibit the \\nexpression of specific mRNAs, allows the generation of lack-of-function phenotypes and permits the assessment \\nof the role of specific genes in development in the absence \\nof classical developmental genetics. \\nMore recently, in a remarkable series of experiments, \\nHal identified a single gene, myoD, which can elicit the \\nentire prog'ram of muscle differentiation when introduced \\ninto nonmuscle cell types. He demonstrated that muscle \\ncell differentiation could be elicited in fibroblasts following \\nthe transfer of genomic DNA from muscle cells. This led \\nHal to the isolation of the myoD gene, which encodes a \\ntranscriptional activator that promotes myogenesis by di- \\nrectly binding to a control region shared by muscle-specific \\ngenes. Moreover, MyoD activates its own transcription, \\ncreating an autoregulatory loop that stabilizes the commit- \\nment to myogenesis. Structure-function analysis allowed \\nthe definition of a complex\", metadata={'page': 0, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/Harold Weintraub.pdf'}),\n",
       " Document(page_content='expressed, how does Myod regulate skeletal muscle cell\\ndifferentiation? In one sense, the answer seems fairly simple:Myod is a transcription factor with binding sites in theregulatory regions of many genes that are expressed in skeletalmuscle. Myod forms heterodimers with the nearly ubiquitousE-protein sub-family of bHLH proteins through the interactionof the HLH domains (see Fig. 1) (Lassar et al., 1991; Murre etal., 1989). The basic regions act as sequence-speciﬁc DNA-binding domains that recognize a binding site with the simplecore consensus sequence of CANNTG, termed an E-box, andshow additional preferences for internal and ﬂankingsequences (Blackwell and Weintraub, 1990). Myod has a singleamino-terminal acidic-activation domain, as determined by its\\nfusion to the heterologous DNA-binding domain of the Gal4\\nprotein (Weintraub et al., 1991b), whereas E-proteins have amore complex mix of activation and repression domains(Markus et al., 2002) (see Fig. 1). Therefore, the simple', metadata={'page': 3, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/The circuitry of a master switch.pdf'}),\n",
       " Document(page_content='In addition, the binding and activity of Myod isregulated by other factors to achieve a temporal patterning ofgene expression through a feed-forward mechanism. It is likelythat other master regulatory factors, such as Neurod inneurogenesis and E-proteins in lymphocyte differentiation,regulate and are regulated through similar mechanisms: asimple DNA-binding site that permits a large sampling of thetranscriptional potential of the genome, and super-imposedpromoter-speciﬁc regulation to achieve a coherent pattern ofgene expression.\\nIt is interesting to contrast the role of Myod in muscle cell\\ndifferentiation with that of Pax6 in eye development, both ofwhich have been termed master regulatory genes (Pichaud and\\nDesplan, 2002; Weintraub et al., 1991a). Myod regulates thedifferentiation of a single cell type, whereas Pax6 is essentialfor the development of a complex organ comprising multipledifferent specialized cell types. Within the context establishedby Pax6 and by the other homeobox', metadata={'page': 8, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/The circuitry of a master switch.pdf'}),\n",
       " Document(page_content='or in MyoD-MEFs, respectively (data not shown), reﬂecting\\na larger number of peaks in primary myotubes. Examination ofseveral MyoD-regulated genes also showed that the MyoD-\\nbinding regions in the three different types of muscle cells\\nwere very similar, although some binding regions varied amongthe muscle cell types ( Figures 4 C–4F). Therefore, the MyoD-\\nbinding regions are extremely similar in C2C12 cells, primarymuscle cells, and MEFs transduced with MyoD.\\nFunctional Assessment of Genome-wide MyoD Binding\\nTo determine whether all identiﬁed MyoD-binding sites representclassic enhancer elements, we selected 25 isolated peaks\\ncovering conserved regions that increased, decreased, or did\\nnot change peak height with differentiation, and cloned regions(ranging from 204 to 567 nt under the center of the peak thatmaximized the inclusion of evolutionarily conserved sequences)\\ninto a reporter construct that contains the SV40 promoter driving\\nluciferase (pGL3 promoter), and enhancer activity', metadata={'page': 7, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/Genome-wide MyoD Binding.pdf'}),\n",
       " Document(page_content='func- \\ntions for MyoD and Myf-5 is that over evolutionary time, \\nan inhibitory mutation in either MyoDor Myf-5 should theo- \\nretically be tolerated (because of backup from the other \\ngene); hence, one would expect the accumulation of muta- \\ntions and the disappearance of one or the other of MyoD \\nor My-f-5 Presumably, this does not occur because each \\ngene also has a subset of functions that are unique. Direct \\nevidence for unique functions comes from the fact that \\nMyf-dnegative mice fail to develop normal rib cages (pre- \\nsumably owing to some Myfd-dependent secondary influ- \\nence of muscle on cartilage morphology), and MyoD- \\nnegative mice show a viability lower than expected. \\nMyoD-negative mice also overproduce Myfd, suggesting \\nthat MyoD might also negatively regulate Myf-5 (see Figure \\n3). Unique functions for MyoD and Myf-5 might not only \\noccur at the level of their respective target genes, but the \\ntwo genes might be associated with enhancers that “read” \\ndifferent upstream', metadata={'page': 2, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/The MyoD Family and Myogenesis Redundancy, Networks, and Thresholds.pdf'}),\n",
       " Document(page_content=',mastergenes(likeMyoD)wouldbeactivatedto\\nencodecelltype.Confirmationofthisviewwillrequirea\\ndetailedanalysisoftheregulationoftheendogenousMyoD\\ngeneincombinationwithananalysisofthetrans-acting\\nelementsresponsiblefortheinitialactivationofMyoDduring\\ndevelopment.\\nWethankourcolleaguesforadvice,HelenDevittfortypingthe\\nmanuscript,PeiFengChengandGuyRosmanfortechnicalhelp,and\\ntheNationalInstitutesofHealth(H.W.andA.D.M.),theLucille\\nMarkeyFoundation(A.B.L.),andtheFondsdelaRechercheen\\nSantdduQuebec(M.A.A.)forsupport.\\n1.Lassar,A.B.,Paterson,B.M.&Weintraub,H.(1986)Cell47,\\n649-656.\\n2.Davis,R.L.,Weintraub,H.&Lassar,A.B.(1987)Cell51,\\n987-1000.\\n3.Tapscott,S.J.,Davis,R.L.,Thayer,M.J.,Cheng,P.-F.,\\nWeintraub,H.&Lassar,A.B.(1988)Science242,405-411.\\n4.Sasoon,D.,Wright,W.,Lin,V.,Lassar,A.,Weintraub,H.&\\nBuckingham,M.(1989)Nature(London),inpress.\\n5.Murre,C.,McCaw,P.S.&Baltimore,D.(1989)Cell56,\\n777-783.\\n6.Lassar,A.B.,Buskin,J.,Lockshon,D.,Davis,R.,Hauschka,\\nS.&Weintraub,H.(1989)Cell,inpress.\\n7.Thayer,M.J.', metadata={'page': 4, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/Activation of muscle-specific genes in pigment, nerve, fat, liver, and fibroblast cell lines by forced expression of MyoD.pdf'}),\n",
       " Document(page_content='together with othertranscription factors in a feed-forward circuit and often bindsto an E-box site prior to transcriptional activation, which is deter-\\nmined by the recruitment of cofactors at the speciﬁc promoter\\n(Bergstrom et al., 2002; Penn et al., 2004; Rampalli et al.,\\n2007 ). Together, these studies indicate that MyoD-binding\\nafﬁnity is sensitive to core and ﬂanking E-box sequences, andthat MyoD transcriptional activity can be regulated indepen-dently of binding afﬁnity. Therefore, it is possible that MyoDmight bind at sites in the genome without regulating regional\\ngene transcription; however, it is not known how frequently this\\nmight occur, nor whether it might have any other biologicalactivity.\\nIn this study, we have used chromatin immunoprecipitation\\nand high-throughput sequencing (ChIP-seq) to identify genome-wide binding of MyoD in several skeletal muscle cell types.\\nAs anticipated, MyoD binds to many of the genes transcription-\\nally regulated during skeletal muscle', metadata={'page': 0, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/Genome-wide MyoD Binding.pdf'}),\n",
       " Document(page_content='both Myod and Myog are\\nexpressed during skeletal muscle differentiation. In addition to itsearly and transient expression in the somite, Mrf4 is also expressed in\\nthe terminally differentiated muscle cells, and it is likely that Myodand Myog regulate this late expression of Mrf4 . A transgene that\\ndrives Mrf4 expression from the Myog promoter can partly\\ncompensate for the loss of Myog (Zhu and Miller, 1997),\\ndemonstrating a partly redundant role of Mrf4 and Myog in terminal\\ndifferentiation.\\nDevelopment', metadata={'page': 2, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/The circuitry of a master switch.pdf'}),\n",
       " Document(page_content='gene expressionchanges in ﬁbroblasts with an inducible Myod protein andobserved a similarly large number of expression changes with~5% of the genes tiled on the array (Bergstrom et al., 2002).Many of the RNAs that showed increased expression inresponse to Myod were muscle-speciﬁc genes, such as skeletalmuscle myosins and actins, but many were genes expressed innumerous different lineages, such as the Mef2 transcriptionfactors. Surprisingly, some RNAs induced by Myod coded forproteins that inhibit Myod activity, such as the Id proteins\\n(Benezra et al., 1990). This could partly be due to a non-autonomous inhibitory-surround mechanism. For example, theNotch ligand Delta is an early target of Myod and itsexpression is followed by the expression of the Notch regulatedgene Hes1 (Bergstrom et al., 2002), indicating that Myod -\\nexpressing cells might inhibit muscle differentiation in theirneighbors through the Notch signaling pathway.\\nNot all genes are simultaneously expressed in response', metadata={'page': 4, 'source': '/nfs/turbo/umms-indikar/shared/projects/RAG/papers/DigitalLibrary-9-June-2024/The circuitry of a master switch.pdf'})]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "docs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9c98aa9f-aa34-4275-ae62-573cc36cf99b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on method from_llm in module langchain.retrievers.multi_query:\n",
      "\n",
      "from_llm(retriever: langchain_core.retrievers.BaseRetriever, llm: langchain_core.language_models.base.BaseLanguageModel, prompt: langchain_core.prompts.prompt.PromptTemplate = PromptTemplate(input_variables=['question'], template='You are an AI language model assistant. Your task is \\n    to generate 3 different versions of the given user \\n    question to retrieve relevant documents from a vector  database. \\n    By generating multiple perspectives on the user question, \\n    your goal is to help the user overcome some of the limitations \\n    of distance-based similarity search. Provide these alternative \\n    questions separated by newlines. Original question: {question}'), parser_key: Optional[str] = None, include_original: bool = False) -> 'MultiQueryRetriever' method of pydantic.v1.main.ModelMetaclass instance\n",
      "    Initialize from llm using default template.\n",
      "    \n",
      "    Args:\n",
      "        retriever: retriever to query documents from\n",
      "        llm: llm for query generation using DEFAULT_QUERY_PROMPT\n",
      "        include_original: Whether to include the original query in the list of\n",
      "            generated queries.\n",
      "    \n",
      "    Returns:\n",
      "        MultiQueryRetriever\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(MultiQueryRetriever.from_llm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "193dfcd3-e071-44c6-b6b3-73cfe19c2a0c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "178659d7-1911-4076-af6e-10a600861eea",
   "metadata": {},
   "source": [
    "# Building a L"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
